// @ali/4ever-mo/dist-node/__tests__/helper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.wrapTextToValue = wrapTextToValue;
exports.wrapTextToASL = wrapTextToASL;
exports.wrapStringToText = wrapStringToText;
exports.regenerateJsonML = regenerateJsonML;
exports.regenerateValue = regenerateValue;
exports.createValueWithBlocks = createValueWithBlocks;
exports.createValueWithContent = createValueWithContent;

var _everCangjie = require("@ali/4ever-cangjie");

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _serializer = _interopRequireDefault(require("./serializer"));

function wrapTextToValue(text) {
  return _everCangjie.Value.create({
    document: _everCangjie.Document.create({
      nodes: [_everPluginParagraph.Paragraph.create({
        nodes: [text]
      })]
    })
  });
}

function wrapTextToASL(text) {
  return ['root', {}, ['p', {}, text]];
}

function wrapStringToText(string) {
  return ['span', {
    'data-type': 'text'
  }, ['span', {
    'data-type': 'leaf'
  }, string]];
}

function regenerateJsonML(jsonml) {
  const value = _serializer.default.jsonMLToValue(jsonml);

  return _serializer.default.valueToJsonML(value);
}

function regenerateValue(value) {
  const jsonml = _serializer.default.valueToJsonML(value);

  return _serializer.default.jsonMLToValue(jsonml);
}

function createValueWithBlocks(blocks, selection) {
  return _everCangjie.Value.create({
    document: _everCangjie.Document.create({
      key: 'doc',
      nodes: blocks
    }),
    selection
  });
}

function createValueWithContent(content, selection) {
  return createValueWithBlocks([_everCangjie.Block.create({
    key: 'p',
    type: 'paragraph',
    nodes: content
  })], selection);
}
