// @ali/4ever-bi/dist-node/plugins/indent/components/next/createIncreaseIndentButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createIncreaseIndentButton;

var React = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everComponent = require("@ali/4ever-component");

var _hotKeyTips = require("../../utils/hotKeyTips");

var _actions = require("../../actions");

const _createElement = /*#__PURE__*/React.createElement;
;
;
;

var _ref = /*#__PURE__*/_createElement(_everComponent.IconIncreaseIndent, null);

function createIncreaseIndentButton(props) {
  const {
    controller,
    locale,
    icon,
    title,
    tooltip,
    badgeDot,
    disabled
  } = props;

  const handleClick = () => controller.run('onAction', (0, _actions.increaseIndent)());

  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const isDisabled = disabled || isCollapsedAtEdge || controller.query('isSelectionInHr') || controller.query('isSelectionInListSymbol');
  const buttonConfig = {
    icon: icon || _ref,
    onClick: handleClick,
    title: title || locale?.rightButtonTitle,
    tooltip: {
      title: title || locale?.rightButtonTitle,
      shortCut: _hotKeyTips.INCREASE_HOT_KEY_TIP,
      ...tooltip
    },
    badgeDot,
    disabled: isDisabled,
    role: _everBamboo.PluginRoles.indentIncrease,
    testid: "bi-toolbar-incindent"
  };
  return buttonConfig;
}
