// @ali/4ever-bi/dist-node/plugins/fonts/utils/getFontsTip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getFontsTip;
// unnecessary and useless, just for consistent api
const WEUI_FONTS_MAP = {
  Helvetica: 'Helvetica',
  default: 'dropdownDefault',
  Arial: 'dropdownArial',
  ComicSansMS: 'dropdownComicSans',
  CourierNew: 'dropdownCourierNew',
  Georgia: 'dropdownGeorgia',
  Impact: 'dropdownImpact',
  TimesNewRoman: 'dropdownTimesNewRoman',
  TrebuchetMS: 'dropdownTrebuchet',
  Verdana: 'dropdownVerdana',
  FangSong: 'dropdownFangSong',
  STFangsong: 'dropdownSTFangsong',
  STSong: 'dropdownSTSong',
  STKaiti: 'dropdownKaiti',
  SimSun: 'dropdownSimSun',
  MicrosoftYaHei: 'dropdownMSYaHei',
  KaiTi: 'dropdownKaiTi',
  KaitiSC: 'dropdownKaitiSC',
  SimHei: 'dropdownSimHei',
  HeitiSC: 'dropdownHeitiSC',
  FZHeiB01S: 'dropdownFZHei',
  FZKaiZ03S: 'dropdownFZKai',
  FZFangSongZ02S: 'dropdownFZFangSong'
};

function getFontsTip(fontsValue, locale) {
  if (!locale) return fontsValue;
  const fonts = fontsValue.replace(/ |-/g, '');
  const key = WEUI_FONTS_MAP[fonts];
  return locale[key] || fontsValue;
}
