// @ali/we-design-next/dist-node/form/color-picker/ColorPickerGradientRow.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PresetGradient = void 0;
var _react = _interopRequireDefault(require("react"));
var _common = require("../../common");
var _styled = require("./styled");
var _isWhite = require("./isWhite");
const _createElement = /*#__PURE__*/_react.default.createElement;
const gradients = ['#FFFFFF', '#28B4BE__#3935DD', '#2449FE__#CA4BA7', '#FF7100__#F300AD', '#B16A4D__#D5A638', '#1890FF__#4FF3E0', '#F3AA00__#E43E29', '#EB47FF__#F95101', '#10CCF7__#AB44FB', '#F55B2D__#510101'];
const PresetGradient = /*#__PURE__*/_react.default.memo(props => {
  const {
    value,
    onSelect
  } = props;
  return /*#__PURE__*/_createElement(_styled.Row, null, gradients.map((g, i) => {
    const style = {};
    if (g.indexOf('__') === -1) {
      style.background = g;
      if ((0, _isWhite.isWhite)(g)) {
        style.border = `2px solid ${_common.COLORS.OTHERS.BORDER}`;
      }
    } else {
      const [c1, c2] = g.split('__');
      style.backgroundImage = `linear-gradient(90deg, ${c1}, ${c2})`;
      style.backgroundOrigin = 'border-box';
    }
    return /*#__PURE__*/_createElement(_styled.ColorSquareWrap, {
      onClick: () => onSelect(g),
      style: style
    }, g === value ? /*#__PURE__*/_createElement("svg", {
      width: "14",
      height: "10",
      viewBox: "0 0 16 12",
      fill: "none"
    }, /*#__PURE__*/_createElement("path", {
      d: "M13.965 2.15533L5.63499 10.485L1.46999 6.32034",
      stroke: i === 0 ? '#000000' : '#ffffff',
      strokeOpacity: "0.6",
      strokeWidth: "2.38",
      strokeLinecap: "round",
      strokeLinejoin: "round"
    })) : null);
  }));
});
exports.PresetGradient = PresetGradient;
