// @ali/we-design-next/dist-node/form/color-picker/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColorTextA = exports.ColorSquareWrap = exports.ColorPickerWrapper = exports.AutoColorWrapper = exports.ArrowRight = exports.AdvanceGroup = void 0;
Object.defineProperty(exports, "Divider", {
  enumerable: true,
  get: function () {
    return _common.Divider;
  }
});
exports.SketchPickerWrap = exports.Row = exports.RecentGroup = exports.NullColorSquareWrap = exports.NoColorWrapper = exports.HeaderGroup = exports.Group = void 0;
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _common = require("../../common");
const _createElement = /*#__PURE__*/_react.default.createElement;
const ColorPickerWrapper = /*#__PURE__*/_styledComponents.default.div(["user-select:none;margin:4px 8px;"]);
exports.ColorPickerWrapper = ColorPickerWrapper;
const HeaderGroup = /*#__PURE__*/_styledComponents.default.div(["position:relative;display:flex;justify-content:space-between;align-items:center;padding:4px;"]);
exports.HeaderGroup = HeaderGroup;
const AutoColorWrapper = /*#__PURE__*/_styledComponents.default.div(["padding:4px;display:flex;align-items:center;overflow:hidden;border-radius:4px;"]);
exports.AutoColorWrapper = AutoColorWrapper;
const NoColorWrapper = /*#__PURE__*/(0, _styledComponents.default)(AutoColorWrapper)(["&:hover{background:", ";}transition:background 0.1s ease-in-out;"], _common.COLORS.OTHERS.DAO_HOVER_BG_COLOR);
exports.NoColorWrapper = NoColorWrapper;
const Group = /*#__PURE__*/_styledComponents.default.div(["position:relative;padding:4px 8px 8px 8px;"]);
exports.Group = Group;
const RecentGroup = /*#__PURE__*/_styledComponents.default.div(["position:relative;padding:4px 8px;"]);
exports.RecentGroup = RecentGroup;
const Row = /*#__PURE__*/_styledComponents.default.div(["position:relative;display:flex;width:100%;min-height:6px;margin-bottom:3px;&:last-child{margin-bottom:0;}"]);
exports.Row = Row;
const AdvanceGroup = _common.RowItem;
exports.AdvanceGroup = AdvanceGroup;
const ColorSquareWrap = /*#__PURE__*/_styledComponents.default.span(["position:relative;display:inline-block;width:20px;height:20px;box-sizing:border-box;margin-right:3px;&:last-child{margin-right:0;}border:2px solid ", ";border-radius:4px 4px;flex:0 0 auto;background-color:", ";&:hover{padding:2px 2px;border:2px solid ", ";& svg{top:2px;left:2px;}}& svg{position:absolute;top:2px;left:2px;width:12px;height:12px;}& ._wdn-color-picker-text-a{display:none;}&:hover ._wdn-color-picker-text-a{display:flex;}"], ({
  alwaysShowBorder
}) => alwaysShowBorder ? _common.COLORS.OTHERS.BORDER : _common.COLORS.TRANSPARENT, _common.COLORS.WHITE, _common.COLORS.OTHERS.PRIMARY_HOVER);
exports.ColorSquareWrap = ColorSquareWrap;
const NullColorSquareWrap = /*#__PURE__*/(0, _styledComponents.default)(ColorSquareWrap)(["&:after{content:'';display:block;position:absolute;top:8px;left:-2px;width:22px;height:0;border-bottom:1px solid ", ";transform:rotate(-45deg);}&:hover{border:1px solid ", ";}border:1px solid ", ";overflow:hidden;margin:0;"], _common.COLORS.COMMON.DANGER1_COLOR, _common.COLORS.OTHERS.BORDER, _common.COLORS.OTHERS.BORDER);
exports.NullColorSquareWrap = NullColorSquareWrap;
const ColorTextA = /*#__PURE__*/_styledComponents.default.span(["position:absolute;left:1px;right:0;bottom:0;top:1px;display:flex;justify-content:center;align-items:center;font-size:12px;"]);
exports.ColorTextA = ColorTextA;
const SketchPickerWrap = /*#__PURE__*/_styledComponents.default.div(["input{box-sizing:content-box;}"]);
exports.SketchPickerWrap = SketchPickerWrap;
const ArrowRight = /*#__PURE__*/_react.default.memo(() => /*#__PURE__*/_createElement("svg", {
  width: "12",
  height: "12",
  viewBox: "0 0 12 12",
  fill: "none"
}, /*#__PURE__*/_createElement("path", {
  fillRule: "evenodd",
  clipRule: "evenodd",
  d: "M7.54293 6.25004L3.39648 2.10359L4.10359 1.39648L8.95714 6.25004L4.10359 11.1036L3.39648 10.3965L7.54293 6.25004Z",
  fill: _common.COLORS.LIGHT.LEVEL1_BASE_COLOR,
  fillOpacity: "0.6"
})));
exports.ArrowRight = ArrowRight;
