// @ali/4ever-plugin-refblock/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.default;
  }
});
Object.defineProperty(exports, "RefBlock", {
  enumerable: true,
  get: function () {
    return _models.default;
  }
});
Object.defineProperty(exports, "RefBlockStatus", {
  enumerable: true,
  get: function () {
    return _models.RefBlockStatus;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.default;
  }
});
Object.defineProperty(exports, "RefBlockRole", {
  enumerable: true,
  get: function () {
    return _roles.RefBlockRole;
  }
});
Object.defineProperty(exports, "biFactory", {
  enumerable: true,
  get: function () {
    return _bi.default;
  }
});
Object.defineProperty(exports, "zhiFactory", {
  enumerable: true,
  get: function () {
    return _zhi.default;
  }
});
exports.utils = exports.styles = void 0;

require("./utils/types");

require("./bi/types");

require("./zhi/types");

var _index = _interopRequireDefault(require("./mo/index"));

var _models = _interopRequireWildcard(require("./mo/models"));

var _index2 = _interopRequireDefault(require("./mo/serializer/index"));

var _roles = require("./utils/roles");

var _styles = _interopRequireWildcard(require("./utils/styles"));

exports.styles = _styles;

var _utils = _interopRequireWildcard(require("./utils"));

exports.utils = _utils;

var _bi = _interopRequireDefault(require("./bi"));

var _zhi = _interopRequireDefault(require("./zhi"));
