// @ali/we-design-next/dist-node/form/switch/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SwitchInner = exports.SwitchButton = void 0;
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _colors = require("../../common/consts/colors");
const widthSmall = 26;
const heightSmall = 16;
const widthNormal = 44;
const heightNormal = 22;
const SwitchButton = /*#__PURE__*/_styledComponents.default.button(["border:0;display:inline-block;position:relative;outline:none;transition:all .36s cubic-bezier(.78,.14,.15,.86);", " border-radius:", "px;background-color:rgba(17,31,44,0.12);overflow:hidden;", " ", ""], ({
  size
}) => {
  if (size === 'small') {
    return `
        width: ${widthSmall}px;
        height: ${heightSmall}px;
      `;
  }
  return `
      width: ${widthNormal}px;
      height: ${heightNormal}px;
    `;
}, heightNormal, ({
  checked
}) => {
  return checked ? `
      background-color: ${_colors.COLORS.COMMON.BLUE1_COLOR};
    ` : '';
}, ({
  disabled,
  checked
}) => {
  return disabled ? `
      cursor: not-allowed;
      background-color: ${checked ? _colors.COLORS.COMMON.BLUE1_HOVER_COLOR : 'rgba(0, 0, 0, 0.04)'};
    ` : '';
});
exports.SwitchButton = SwitchButton;
const SwitchInner = /*#__PURE__*/_styledComponents.default.span(["display:block;left:2px;top:2px;transition:all .36s cubic-bezier(.78,.14,.15,.86);", " border-radius:50%;position:absolute;background-color:", ";", ""], ({
  checked,
  size
}) => {
  const height = size === 'small' ? heightSmall : heightNormal;
  return checked ? `
      left: calc(100% - ${height - 2}px);
    ` : '';
}, _colors.COLORS.WHITE, ({
  size,
  theme
}) => {
  if (size === 'small') {
    return `
        width: ${heightSmall - 4}px;
        height: ${heightSmall - 4}px;
      `;
  }
  return `
      width: ${heightNormal - 4}px;
      height: ${heightNormal - 4}px;
    `;
});
exports.SwitchInner = SwitchInner;
