// @ali/we-design-next/dist-web/form/radio/Radio.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { useDefaultProp } from "../../common/react/useDefaultProp";
import { RadioWrap, RadioInput, RadioIcon, RadioText } from "./styled";
export var Radio = /*#__PURE__*/React.memo( /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _props$defaultChecked = props.defaultChecked,
    defaultChecked = _props$defaultChecked === void 0 ? false : _props$defaultChecked,
    _checked = props.checked,
    onChange = props.onChange,
    _props$disabled = props.disabled,
    disabled = _props$disabled === void 0 ? false : _props$disabled,
    children = props.children;
  var _useDefaultProp = useDefaultProp(defaultChecked, _checked, onChange, function (e) {
      return e.target.checked;
    }),
    checked = _useDefaultProp[0],
    setChecked = _useDefaultProp[1];
  return /*#__PURE__*/_createElement(RadioWrap, {
    disabled: !!disabled
  }, /*#__PURE__*/_createElement(RadioInput, {
    ref: ref,
    type: "radio",
    onChange: setChecked,
    disabled: disabled,
    checked: checked
  }), /*#__PURE__*/_createElement(RadioIcon, {
    disabled: !!disabled,
    checked: !!checked
  }), children && /*#__PURE__*/_createElement(RadioText, null, children));
}));
