// @ali/4ever-factory/dist-node/groupFactory/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.groupFactory = groupFactory;

var _everPluginGroupFactory = require("@ali/4ever-plugin-group-factory");

var _everPluginHeadingGroup = require("@ali/4ever-plugin-heading-group");

var _createOnAction = _interopRequireDefault(require("./handlers/createOnAction"));

var factoryCommands = _interopRequireWildcard(require("./commands"));

function groupFactory(createPlugin) {
  return function createGroupPlugin(bizConfig) {
    const myPlugin = createPlugin(bizConfig);
    const {
      onAction,
      onKeyDown,
      queries,
      commands,
      groupNodes,
      renderNode,
      renderMark,
      groupType
    } = myPlugin;
    const config = bizConfig[groupType] || {};
    return {
      shouldNodeComponentUpdate: (0, _everPluginGroupFactory.createShouldComponentUpdate)(),
      renderNode: renderNode || (0, _everPluginGroupFactory.createRenderNode)(groupType),
      // @ts-ignore
      renderMark: renderMark || (0, _everPluginHeadingGroup.createRenderMark)(config),
      groupNodes,
      onAction: (0, _createOnAction.default)(onAction),
      onKeyDown: onKeyDown || (0, _everPluginGroupFactory.createOnKeyDown)(),
      onCangjieSelect: (0, _everPluginGroupFactory.createOnCangjieSelect)(config),
      commands: { ...factoryCommands,
        ...commands
      },
      queries
    };
  };
}

groupFactory.actions = _everPluginGroupFactory.actions;
groupFactory.commands = factoryCommands;
