// @ali/4ever-plugin-calendar-card/dist-web/utils/type.js
// 1-未响应，2-已接受，10-已拒绝, 20-待定；目前仅显示未响应跟已拒绝人员
export var ReceiverTypeEnum;

(function (ReceiverTypeEnum) {
  ReceiverTypeEnum[ReceiverTypeEnum["noresponse"] = 1] = "noresponse";
  ReceiverTypeEnum[ReceiverTypeEnum["refused"] = 10] = "refused";
})(ReceiverTypeEnum || (ReceiverTypeEnum = {}));

export var AttachmentTypeEnum;

(function (AttachmentTypeEnum) {
  AttachmentTypeEnum[AttachmentTypeEnum["text"] = 1] = "text";
  AttachmentTypeEnum[AttachmentTypeEnum["audio"] = 2] = "audio";
  AttachmentTypeEnum[AttachmentTypeEnum["image"] = 3] = "image";
  AttachmentTypeEnum[AttachmentTypeEnum["file"] = 4] = "file";
  AttachmentTypeEnum[AttachmentTypeEnum["link"] = 5] = "link";
  AttachmentTypeEnum[AttachmentTypeEnum["redenvelop"] = 6] = "redenvelop";
  AttachmentTypeEnum[AttachmentTypeEnum["dingfile"] = 7] = "dingfile";
})(AttachmentTypeEnum || (AttachmentTypeEnum = {}));

export var CalendarType; // 接受者列表

(function (CalendarType) {
  CalendarType["SHANHUI"] = "shanhui";
})(CalendarType || (CalendarType = {}));
