// @ali/4ever-plugin-mention/dist-web/zhi/mention.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React, { useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { marksToStyle as convertMarks } from '@ali/4ever-plugin-marks';
import { BtnWrapper } from "../components/placeholder";
import MentionComponent from "../components/mention";
import { MentionNameWrapper } from "../components/styled";
export default function Mention(props) {
  var node = props.node,
      onMentionClick = props.onMentionClick,
      onMentionHover = props.onMentionHover,
      getCurrentUser = props.getCurrentUser,
      locale = props.locale,
      rest = _objectWithoutPropertiesLoose(props, ["node", "onMentionClick", "onMentionHover", "getCurrentUser", "locale"]);

  var _node$data = node.data,
      name = _node$data.name,
      login = _node$data.login,
      id = _node$data.id,
      isPlaceholder = _node$data.isPlaceholder;
  var text = node.getFirstText();
  var leaf = text && text.leaves[0];
  var marks = leaf ? leaf.marks : [];
  var style = convertMarks(marks);
  var isSelf = getCurrentUser ? getCurrentUser() === login : false;
  var handleMentionClick = useCallback(function (event) {
    event.preventDefault();
    event.stopPropagation();

    if (onMentionClick) {
      // @ts-ignore
      onMentionClick(node.data);
    }
  }, [node.data, onMentionClick]);
  var handleMentionHover = useCallback(function (event) {
    if (onMentionHover) {
      // @ts-ignore
      onMentionHover(node.data);
    }
  }, [node.data, onMentionHover]);

  if (isPlaceholder) {
    return /*#__PURE__*/_createElement(BtnWrapper, {
      isReader: true
    }, "@ ", locale.mentionPlaceholder);
  }

  return /*#__PURE__*/_createElement(MentionComponent, _extends({}, rest, {
    // @ts-ignore
    id: id,
    "data-type": node.type,
    style: style,
    onClick: handleMentionClick,
    onMouseEnter: handleMentionHover,
    isSelf: isSelf,
    node: node
  }), "@", /*#__PURE__*/_createElement(MentionNameWrapper, null, name));
}
