// @ali/4ever-plugin-calendar-card/dist-node/utils/components/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  ContainerWrapper: true,
  BodyWrapper: true,
  CommentTextWrapper: true,
  SubjectPlaceholder: true,
  TimePlaceholder: true,
  TimeTextParagraph: true,
  PlaceTextParagraph: true,
  CommentTextParagraph: true,
  RefreshTipsTextWrapper: true
};
Object.defineProperty(exports, "RefreshTipsTextWrapper", {
  enumerable: true,
  get: function () {
    return _styled.RefreshTipsTextWrapper;
  }
});
exports.CommentTextParagraph = exports.PlaceTextParagraph = exports.TimeTextParagraph = exports.TimePlaceholder = exports.SubjectPlaceholder = exports.CommentTextWrapper = exports.BodyWrapper = exports.ContainerWrapper = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _weDesign = require("@ali/we-design");

var _paragraphFactory = require("./paragraphFactory");

var _textParagraphFactory = _interopRequireDefault(require("./textParagraphFactory"));

var _constants = require("../constants");

var _loadingWarpper = require("./loadingWarpper");

Object.keys(_loadingWarpper).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _loadingWarpper[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _loadingWarpper[key];
    }
  });
});

var _extraArea = require("./extraArea");

Object.keys(_extraArea).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _extraArea[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _extraArea[key];
    }
  });
});

var _receiversParagraph = require("./receiversParagraph");

Object.keys(_receiversParagraph).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _receiversParagraph[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _receiversParagraph[key];
    }
  });
});

var _subjectParagraph = require("./subjectParagraph");

Object.keys(_subjectParagraph).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _subjectParagraph[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _subjectParagraph[key];
    }
  });
});

var _styled = require("./extraArea/styled");

const _createElement = /*#__PURE__*/_react.default.createElement;

const ContainerWrapper = /*#__PURE__*/_styledComponents.default.div(["cursor:default;color:#111F2C;font-size:14px;user-select:text;background-color:", ";border-radius:8px;.weicon{color:rgba(23,26,29,0.6);&.weicon-CalendarBetaNormal > svg{& > path{fill:rgba(23,26,29,0.6)!important;}}&.weicon-PositionNormal > svg{& > path{fill:rgba(23,26,29,0.6)!important;}}}&.placeholder{box-shadow:0px 0px 1px 0px rgba(0,0,0,0.12) inset;}&.placeholder > div{color:rgba(17,31,44,0.4);}&.card-calendar-selected{overflow:hidden;background-color:", ";box-shadow:0px 0px 0px 1px #006AFE,0px 0px 0px 3px rgba(0,106,254,0.12);}&:hover{overflow:hidden;}", "{padding-left:0;}"], props => props.backgroundColor || '#F6F6F7', props => props.backgroundColor || '#F5F6F8', _paragraphFactory.IconInnerWrapper);

exports.ContainerWrapper = ContainerWrapper;

const BodyWrapper = /*#__PURE__*/_styledComponents.default.div(["padding:", "px;position:relative;& > div{margin-top:", "px;}& > div:first-child{margin-top:0;}"], _constants.CALENDAR_CARD_PADDING_LEN, _constants.CALENDAR_CARD_PARAGRAPH_MARGIN_TOP);

exports.BodyWrapper = BodyWrapper;

const CommentTextWrapper = /*#__PURE__*/_styledComponents.default.span(["overflow:hidden;text-overflow:ellipsis;display:-webkit-box;-webkit-line-clamp:6;-webkit-box-orient:vertical;"]);

exports.CommentTextWrapper = CommentTextWrapper;

const SubjectPlaceholder = /*#__PURE__*/_styledComponents.default.span(["color:rgba(17,31,44,0.4);"]);

exports.SubjectPlaceholder = SubjectPlaceholder;

const TimePlaceholder = /*#__PURE__*/_styledComponents.default.span(["color:rgba(17,31,44,0.4);cursor:pointer;"]);

exports.TimePlaceholder = TimePlaceholder;
const TimeTextParagraph = (0, _textParagraphFactory.default)( /*#__PURE__*/_createElement(_weDesign.ClockNormalNormal, {
  style: _constants.DEFAULT_ICON_STYLE
}));
exports.TimeTextParagraph = TimeTextParagraph;
const PlaceTextParagraph = (0, _textParagraphFactory.default)( /*#__PURE__*/_createElement(_weDesign.PositionNormalNormal, {
  style: _constants.DEFAULT_ICON_STYLE
}));
exports.PlaceTextParagraph = PlaceTextParagraph;
const CommentTextParagraph = (0, _textParagraphFactory.default)( /*#__PURE__*/_createElement(_weDesign.DingdriveDingdocBetaNormalNormal, {
  style: _constants.DEFAULT_ICON_STYLE
}));
exports.CommentTextParagraph = CommentTextParagraph;
