// @ali/4ever-bi/dist-web/plugins/vertAlign/components/next/subButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { PluginRoles } from '@ali/4ever-bamboo';
import { SubscriptButton } from '@ali/4ever-component';
import { toggleSub } from "../../actions";
import { SHORTCUT } from "../../utils";
;
;
export default function SubButton(props) {
  var controller = props.controller,
      locale = props.locale,
      tooltipProp = props.tooltip,
      rest = _objectWithoutPropertiesLoose(props, ["controller", "locale", "tooltip"]);

  var isActive = controller.query('isSub');
  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInMention') || controller.query('isSelectionInLink') || controller.query('isSelectionInListSymbol') || controller.query('isSelectionInLinkCard');
  var handleClick = React.useCallback(function () {
    return controller.run('onAction', toggleSub());
  }, [controller]);
  var tooltip = React.useMemo(function () {
    return _extends({
      title: (locale == null ? void 0 : locale.subButtonTitle) || '',
      shortCut: SHORTCUT.sub,
      description: "Markdown: ~" + ((locale == null ? void 0 : locale.text) || 'Text') + "~ " + ((locale == null ? void 0 : locale.space) || 'Space')
    }, tooltipProp);
  }, [tooltipProp]);
  return /*#__PURE__*/_createElement(SubscriptButton, _extends({
    testid: "bi-toolbar-sub",
    active: isActive,
    disabled: !!isDisabled,
    onClick: handleClick,
    title: locale == null ? void 0 : locale.subButtonTitle,
    tooltip: tooltip,
    role: PluginRoles.vertAlignSub
  }, rest));
}
