// @ali/4ever-zhi/dist-node/plugins/onlineVideo/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ErrorDesc = exports.CardContentWrapper = exports.CardIconWrapper = exports.ContentHeader = exports.CardContentInnerWrapper = exports.EmbedFileCardWrapper = exports.YoukuPlayer = exports.YoukuPlayerPlaceholder = exports.YoukuPlayerWrapper = exports.VideoWrapper = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

const VideoWrapper = /*#__PURE__*/_styledComponents.default.div(["width:100%;height:", ";border:1px solid #ECEDEF;border-radius:6px;video{width:100%;height:100%;background:#000;outline:none;border-radius:6px;}"], props => props.isMobile === true ? '246px' : '436px');

exports.VideoWrapper = VideoWrapper;

const YoukuPlayerWrapper = /*#__PURE__*/_styledComponents.default.div(["position:relative;width:100%;height:", ";"], props => props.isMobile === true ? '246px' : '436px');

exports.YoukuPlayerWrapper = YoukuPlayerWrapper;

const YoukuPlayerPlaceholder = /*#__PURE__*/_styledComponents.default.div(["width:100%;height:45px;line-height:45px;background:#F7F8F9;border-radius:10px;background-color:#F7F8F9;text-align:center;span{font-size:12px;color:rgba(17,31,44,0.4);}"]);

exports.YoukuPlayerPlaceholder = YoukuPlayerPlaceholder;

const YoukuPlayer = /*#__PURE__*/_styledComponents.default.iframe(["width:100%;height:100%;border:none;border-radius:6px;"]);

exports.YoukuPlayer = YoukuPlayer;

const EmbedFileCardWrapper = /*#__PURE__*/_styledComponents.default.div(["width:100%;height:54px;line-height:24px;border:1px solid #D9DBDD;padding:8px 16px;display:flex;border-radius:6px;background-color:#fff;box-sizing:border-box;"]);

exports.EmbedFileCardWrapper = EmbedFileCardWrapper;

const CardContentInnerWrapper = /*#__PURE__*/_styledComponents.default.div(["display:block;margin-top:6px;"]);

exports.CardContentInnerWrapper = CardContentInnerWrapper;

const ContentHeader = /*#__PURE__*/_styledComponents.default.div(["font-size:14px;color:#111F2C;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;line-height:22px;"]);

exports.ContentHeader = ContentHeader;

const CardIconWrapper = /*#__PURE__*/_styledComponents.default.div(["margin-right:16px;flex:0;margin-top:6px;"]);

exports.CardIconWrapper = CardIconWrapper;

const CardContentWrapper = /*#__PURE__*/_styledComponents.default.div(["flex:1;min-width:0;"]);

exports.CardContentWrapper = CardContentWrapper;

const ErrorDesc = /*#__PURE__*/_styledComponents.default.div(["font-size:12px;color:#D0021B;"]);

exports.ErrorDesc = ErrorDesc;
