// @ali/4ever-plugin-paragraph/dist-web/index.js
import { default as moFactory } from "./mo/index";
import { default as zhiFactory } from "./zhi/index";
import { default as biFactory } from "./bi/index";
var mobileZhiFactory = zhiFactory;
var mobileBiFactory = biFactory; // facotry

export { moFactory, biFactory, zhiFactory, mobileBiFactory, mobileZhiFactory }; // Element Model

export { default as Paragraph } from "./mo/models";
export { ELEMENT_TYPE } from "./mo/meta"; // serializer

export { default as createSerializationPlugin } from "./mo/serializer/index";
export { createServerSerializationPlugin } from "./mo/serializer/index"; // actions

export { setParagraphSpacing, setParagraphInd, PARAGRAPH_SET_SPACING, PARAGRAPH_SET_IND } from "./bi/actions"; // types

// other export for api compatibility
export { default as isPureParagraphValue } from "./mo/serializer/utils/isPureParagraphValue";
export { default as ReadParagraph } from "./components/ReadParagraph";
export { default as EditParagraph } from "./components/EditParagraph";
export { Component as P } from "./components/styled";
import * as OldBambooPlugin from "./utils/bamboo";
export {
/** @deprecated */
OldBambooPlugin };
export { ParagraphRole } from "./utils/role";
