// @ali/4ever-bi/dist-node/plugins/headingGroupNew/handlers/createRenderGroup.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _everMo = require("@ali/4ever-mo");

var _everBamboo = require("@ali/4ever-bamboo");

const _createElement = /*#__PURE__*/React.createElement;
const {
  GroupNew
} = _everBamboo.GroupFactory;
const {
  Heading1,
  Heading2,
  Heading3,
  Heading4,
  Heading5,
  Heading6
} = _everMo.Heading;
const headingList = [Heading1, Heading2, Heading3, Heading4, Heading5, Heading6];

function createRenderGroup(config) {
  return function renderGroup(props, _, next) {
    const {
      node,
      attributes,
      children,
      visible = true,
      controller,
      group,
      isCollapsed
    } = props;

    if (group.type === 'heading1-group' || group.type === 'heading2-group' || group.type === 'heading3-group') {
      return /*#__PURE__*/_createElement(GroupNew, {
        node: node,
        attributes: attributes,
        controller: controller,
        isSelected: false,
        visible: visible,
        isCollapsed: isCollapsed
      }, children());
    }

    return next();
  };
}

var _default = createRenderGroup;
exports.default = _default;
