// @ali/4ever-plugin-color/dist-node/bi/components/colorButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ColorButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _actions = require("../actions");

var _constants = require("../constants");

var _everPluginPref = require("@ali/4ever-plugin-pref");

const _createElement = /*#__PURE__*/React.createElement;

function ColorButton(props) {
  const {
    controller,
    locale,
    ...rest
  } = props;
  const color = controller.query('getColor');
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const handleChange = React.useCallback(colorValue => controller.run('onAction', (0, _actions.setColor)(colorValue)), [controller]);
  const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInLinkCard');
  const getColorPanelMode = React.useCallback(() => controller.query('getColorPanelMode')?.mode, [controller]);
  const handleSelectValueChange = React.useCallback(mode => {
    controller.run('onAction', (0, _everPluginPref.setColorPanelMode)({
      mode
    }));
  }, [controller]);
  return /*#__PURE__*/_createElement(_weToolbar.ColorPicker, (0, _extends2.default)({}, rest, {
    testid: "bi-toolbar-color",
    btnType: _weToolbar.EColorPickerIcon.font,
    disabled: !!isDisabled,
    color: color,
    onColorPicker: handleChange,
    defaultColor: _constants.DEFAULT_COLOR,
    automaticColor: "#404040",
    shortcutText: locale.buttonTitle,
    showToggleSelect: true,
    showGradient: true,
    getColorPanelMode: getColorPanelMode,
    onSelectValueChange: handleSelectValueChange
  }));
}
