// @ali/4ever-mo/dist-node/utils/nodeTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isListValue = isListValue;
exports.isMentionValue = isMentionValue;
exports.isParagraphValue = isParagraphValue;
exports.isTableValue = isTableValue;
exports.isTableRowValue = isTableRowValue;
exports.isTableCellValue = isTableCellValue;
Object.defineProperty(exports, "isPureParagraphValue", {
  enumerable: true,
  get: function () {
    return _everPluginParagraph.isPureParagraphValue;
  }
});

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _everPluginMention = require("@ali/4ever-plugin-mention");

var _everPluginTable = require("@ali/4ever-plugin-table");

var _everPluginList = require("@ali/4ever-plugin-list");

function isListValue(node) {
  // list 数据类型为 paragraph，区别在于 data 中存在 list 字段
  return _everPluginList.List.isList(node);
}

function isMentionValue(node) {
  return _everPluginMention.Mention.isMention(node);
}

function isParagraphValue(node) {
  return _everPluginParagraph.Paragraph.isParagraph(node);
}

function isTableValue(node) {
  return _everPluginTable.Table.isTable(node);
}

function isTableRowValue(node) {
  return _everPluginTable.TableRow.isTableRow(node);
}

function isTableCellValue(node) {
  return _everPluginTable.TableCell.isTableCell(node);
}
