// @ali/4ever-plugin-template-button/dist-node/bi/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.biFactory = void 0;

var _queries = require("./queries");

var _commands = require("./commands");

var _actions = require("./actions");

var _onKeyDown = require("./handlers/onKeyDown");

var _mo = require("../mo");

var _onAction = require("./handlers/onAction");

var _createMenubar = require("./handlers/createMenubar");

var _createToolbar = require("./handlers/createToolbar");

var _onCangjieSelect = require("./handlers/onCangjieSelect");

var _createRenderNode = require("./handlers/createRenderNode");

var _onCangjieCopyOrCut = require("./handlers/onCangjieCopyOrCut");

var _createNormalizeNode = require("./handlers/createNormalizeNode");

var _createRenderToolbar = require("./handlers/createRenderToolbar");

var _createRenderHoverToolbar = require("./handlers/createRenderHoverToolbar");

const biFactory = configs => {
  const config = configs['templateButton'] || {};
  return { ...(0, _mo.moFactory)({}),
    queries: _queries.queries,
    actions: _actions.actions,
    isActions: _actions.isActions,
    models: {
      klass: 'block',
      type: _mo.TemplateButton.TYPE
    },
    onKeyDown: _onKeyDown.onKeyDown,
    menubar: (0, _createMenubar.createMenubar)(),
    onCangjieSelect: _onCangjieSelect.onCangjieSelect,
    onCangjieCut: _onCangjieCopyOrCut.onCangjieCopyOrCut,
    onCangjieCopy: _onCangjieCopyOrCut.onCangjieCopyOrCut,
    normalizeNode: (0, _createNormalizeNode.createNormalizeNode)(),
    toolbar: (0, _createToolbar.createToolbar)(config),
    commands: (0, _commands.createCommands)(config),
    onAction: (0, _onAction.createOnAction)(config),
    renderNode: (0, _createRenderNode.createRenderNode)(config),
    renderToolbar: (0, _createRenderToolbar.createRenderToolbar)(config),
    renderHoverToolbar: (0, _createRenderHoverToolbar.createRenderHoverToolbar)(config)
  };
};

exports.biFactory = biFactory;
