// build-plugin-ice-ssr/lib/vite/ssrPlugin.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var ssrDevHandler_1 = require("./ssrDevHandler");
var externalPackages = [
    // server 入口依赖
    'cheerio',
    '@loadable/server',
    'chalk',
    'parseurl',
    // @ice/runtime 依赖
    'axios',
    'path-to-regexp',
    // @ice/store 依赖
    'react-redux',
    'lodash.isfunction',
];
var vitePluginSSRServer = function (ssrEntry) {
    return {
        name: 'vite-plugin-ssr',
        enforce: 'pre',
        config: function () {
            return {
                ssr: {
                    // 主要由 server 入口和 noExternal 引入的 cjs 依赖，基于 cjs 规范，可以直接通过 ssrImport，如果不设置将通过 esm 方式执行导致 require / module 是 undefined 报错
                    external: externalPackages,
                    // 设置 noExternal 逻辑，以下包均只存在 esm 产物，但未设置 type: "module"，设置 noExternal 将不会通过 ssrImport 直接执行
                    // TODO：后续需要以下包以相应规范产出 cjs 模块
                    noExternal: ['create-app-shared', 'react-app-renderer', '@ice/runtime', '@ice/store', 'miniapp-history'],
                },
            };
        },
        configResolved: function (resolvedConfig) {
            resolvedConfig.resolve.alias.push({
                // universal-env 写法非标准
                // TODO：universal-env 打包出标准的 ESM 产物
                find: /universal-env$/,
                replacement: '@uni/env',
            });
        },
        configureServer: function (server) { return __awaiter(void 0, void 0, void 0, function () {
            var handler;
            return __generator(this, function (_a) {
                handler = (0, ssrDevHandler_1.default)(server, { ssrEntry: ssrEntry });
                return [2 /*return*/, function () { return server.middlewares.use(handler); }];
            });
        }); },
    };
};
exports.default = vitePluginSSRServer;
