// @ali/we-ui/dist-web/components/modal/styled.js
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/explicit-function-return-type */
import styled from 'styled-components';
import { Button } from "../button";
import { withTheme } from "../theme";
export var ConfirmButton = withTheme(styled(Button)(["margin-left:", ";"], function (_ref) {
  var theme = _ref.theme;
  return theme.ui.margin;
}));
export var Root = /*#__PURE__*/styled.div(["box-sizing:border-box;*{box-sizing:border-box;}"]);
export var Title = /*#__PURE__*/styled.div([""]);
export var CloseButton = withTheme(styled.div(["position:absolute;right:14px;top:10px;padding:10px;line-height:1;font-size:16px;cursor:pointer;color:", ";transition:all .24s ease;&:hover{color:", ";}"], function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.colorText3;
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.colorText4;
}));
export var Header = withTheme(styled.div(["font-weight:500;font-size:", ";margin-bottom:", ";"], function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.fontSizeLarge;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.ui.marginHug;
}));
export var Body = withTheme(styled.div(["font-size:", ";"], function (_ref6) {
  var theme = _ref6.theme;
  return theme.ui.fontSize;
}));
export var Footer = withTheme(styled.div(["margin-top:", ";"], function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.marginHug;
}));
export var FooterWrap = /*#__PURE__*/styled.div(["display:flex;justify-content:flex-end;"]);
export var Content = withTheme(styled.div(["position:relative;padding:", ";box-shadow:", ";background-color:", ";border-radius:", ";width:", ";"], function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.paddingLarge;
}, function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.boxShadow;
}, function (_ref10) {
  var theme = _ref10.theme;
  return theme.ui.colorGray1;
}, function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.radiusBig;
}, function (_ref12) {
  var width = _ref12.width;
  return typeof width === 'number' ? width + "px" : width;
}));
export var Mask = withTheme(styled.div(["display:", ";position:fixed;top:0;right:0;bottom:0;left:0;z-index:", ";height:100%;background-color:", ";"], function (_ref13) {
  var visible = _ref13.visible;
  return visible ? 'block' : 'none';
}, function (_ref14) {
  var theme = _ref14.theme;
  return theme.ui.zIndexModalMask;
}, function (_ref15) {
  var theme = _ref15.theme;
  return theme.ui.modalMaskBackground;
}));
export var Wrap = withTheme(styled.div(["display:", ";position:fixed;top:0;right:0;bottom:0;left:0;z-index:", ";overflow:auto;outline:0;justify-content:center;align-items:center;"], function (_ref16) {
  var visible = _ref16.visible;
  return visible ? 'flex' : 'none';
}, function (_ref17) {
  var theme = _ref17.theme;
  return theme.ui.zIndexModal;
}));
export var TitleTextWrap = withTheme(styled.span(["margin-left:", ";"], function (_ref18) {
  var theme = _ref18.theme;
  return theme.ui.margin;
}));
export var TitleIconWrap = withTheme(styled.span(["font-size:22px;color:", ";"], function (_ref19) {
  var theme = _ref19.theme,
      colorKey = _ref19.colorKey;
  return colorKey ? theme.ui[colorKey] : theme.ui.colorText3;
}));
export var TitleIconTextWrap = /*#__PURE__*/styled.div(["display:flex;align-items:center;"]);
