// @ali/4ever-plugin-subject/dist-node/utils/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isSubjectHeading = isSubjectHeading;
exports.isSubject = isSubject;
exports.createSubject = createSubject;
exports.isEmptyNode = isEmptyNode;
exports.isNormalParagraph = isNormalParagraph;
exports.uniqueId = uniqueId;

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _everCangjie = require("@ali/4ever-cangjie");

/**
 * 判断节点是否是议题标题
 * @param node
 * @returns
 */
function isSubjectHeading(node) {
  return _everCangjie.Block.isBlock(node) && node.type === 'heading-1' && !!node.data.subjectPr;
}
/**
 * 判断节点是否是议题组
 */


function isSubject(node) {
  return _everCangjie.Block.isBlock(node) && node.type === 'subject-group';
}

function createSubject(props) {
  return _everCangjie.Block.create({ ...props,
    type: 'subject-group'
  });
}
/**
 * 判断节点是否是空节点
 * @param node
 * @returns
 */


function isEmptyNode(node) {
  return node.nodes.length === 0 || node.nodes.length === 1 && _everPluginParagraph.Paragraph.isParagraph(node) && _everCangjie.Text.isEmptyText(node.nodes[0]);
}

function isNormalParagraph(node) {
  return _everPluginParagraph.Paragraph.isParagraph(node) && !node.data.list;
}
/**
 * 简单生成 uniqueId
 * @see https://stackoverflow.com/questions/105034/how-to-create-a-guid-uuid#answer-44078785
 * @returns
 */


function uniqueId() {
  return Date.now().toString(36) + Math.random().toString(36).substring(2);
}
