// @ali/we-design/dist-web/components/theme/params/colors.js
import _extends from "@babel/runtime/helpers/extends";

/**
 * ยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยท>
 * ๐ @module : ๐ ่ฒๆฟ้ข่ฒๅ้
 * ยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยท>
 */
var commonColors = {
  // ็ฐ่ฒ
  colorGray1: '#ffffff',
  colorGray2: 'rgba(17, 31, 44, 0.06)',
  colorGray3: '#f5f5f5',
  colorGray4: '#e8e8e8',
  colorGray5: '#d9d9d9',
  colorGray6: '#BFBFBF',
  // fade(#000, 25%);
  colorGray7: '#8C8C8C',
  // fade(#000, 45%);
  colorGray8: '#595959',
  // fade(#000, 65%);
  colorGray9: '#262626',
  // fade(#000, 85%);
  colorGray10: '#000000',
  // ๅๅฒ็บฟ้ข่ฒ
  colorLine2: 'rgba(17, 31, 44, 0.12)',
  colorLine1: 'rgba(17, 31, 44, 0.08)',
  // ๆๅญ้ข่ฒ
  colorText5: 'rgba(17, 31, 44, 1)',
  colorText4: 'rgba(17, 31, 44, 0.56)',
  colorText3: 'rgba(17, 31, 44, 0.4)',
  colorText2: 'rgba(17, 31, 44, 0.24)',
  colorText1: 'rgba(17, 31, 44, 0.12)',
  colorBlue1: '#0089FF',
  // ้็จ่่ฒ
  colorBlue2: '#94CCFB',
  // ่่ฒๆต่ฒ
  colorBlue3: '#DEEFFD',
  // ่่ฒๆตx2่ฒ
  colorRed1: '#FB602D',
  // ้็จ็บข่ฒ
  colorRed2: '#F9BCA7',
  // ็บข่ฒๆต่ฒ
  colorRed3: '#FDEAE4',
  // ็บข่ฒๆตx2่ฒ
  colorOrange1: '#FF9F00',
  // ้็จๆฉ่ฒ
  colorOrange2: '#FAD595',
  // ๆฉ่ฒๆต่ฒ
  colorOrange3: '#FDF2DE',
  // ๆฉ่ฒๆตx2่ฒ
  colorGreen1: '#00B853',
  // ้็จ็ปฟ่ฒ
  colorGreen2: '#94DFB7',
  // ็ปฟ่ฒๆต่ฒ
  colorGreen3: '#DEF5E9' // ็ปฟ่ฒๆตx2่ฒ

};
/* ้็จ key ็ปไปถ่กจๅ๏ผ7.24๏ผ */

var desktopCommon = {
  commonBaseColor1: 'rgba(17, 31, 44, 1)',
  commonBaseColor2: 'rgba(17, 31, 44, 0.56)',
  commonBaseColor3: 'rgba(17, 31, 44, 0.4)',
  commonBaseColor4: 'rgba(17, 31, 44, 0.24)',
  commonLinkColor: 'rgba(80, 116, 154, 1)',
  commonLinkPressColor: 'rgba(80, 116, 154, 0.4)',
  commonLinkHardColor: 'rgba(17, 31, 44, 0.12)',
  commonLinkHardColor20: 'rgba(17, 31, 44, 0.2)',
  commonLineLightColor: 'rgba(17, 31, 44, 0.08)',
  commonFGColor: '#FFFFFF',
  commonFGz1Color: '#FFFFFF',
  commonBgColor: '#F6F6F7',
  commonBgLightColor: '#F7F8Fa',
  commonBgHoverColor: 'rgba(17, 31, 44, 0.08)',
  commonBgPressColor: 'rgba(17, 31, 44, 0.12)',
  commonActionIconColor: '#171A1D' // commonActionIconColor: 'red',

};
var bizColors = {
  // ไธป้ข่ฒ
  primaryColor: commonColors.colorBlue1,
  primaryColor1: '#f0faff',
  primaryColor2: '#d6f0ff',
  primaryColor3: 'rgba(0,106,254,1)',
  primaryHoverColor: '#38A2FF',
  primaryHoverColor2: 'rgba(46,133,255,1)',
  primaryDisabledColor: commonColors.colorBlue2,
  primaryPressColor: '#027CE6',
  primaryPressColor2: 'rgba(0,79,198,1)',
  successColor: '#52c41a',
  infoColor: '#1890ff',
  errorColor: '#f5222d',
  errorColor2: '#FDEAE4',
  warningColor: '#faad14',
  warningColor2: '#FDF2DE',
  // ๆฐดๅฐ้ข่ฒ
  stampColor: 'rgba(17,31,44,0.6)',
  // ่ๆฏ้ข่ฒ
  backgroundColor: '#F2F4F5',
  backgroundColorLight: '#f7f8fa',
  backgroundColorPress: 'rgba(17,31,44,0.12)',
  backgroundColorHover: 'rgba(17,31,44,0.5)',
  hoverbg: 'rgb(237, 236, 238)',
  // ็ฆๆญข
  disabledColor: desktopCommon.commonBaseColor4,
  backgroundDisabledColor: commonColors.colorGray2,
  // ้พๆฅ้ข่ฒ
  linkColor: 'rgba(80,116,154,1)',
  linkHoverColor: '#1890FF',
  linkPressColor: 'rgba(80,116,154,0.12)',
  // ่พนๆก้ข่ฒ
  borderColor: 'rgba(0,0,0,0.15)',
  borderColor2: commonColors.colorGray5,
  // #d9d9d9
  placeholderColor: 'rgba(25, 31, 37, 0.4)',
  disabledBackgroundColor: 'rgba(0, 0, 0, 0.04)',
  disabledBorderColor: 'rgba(0, 0, 0, 0.15)',
  colorFill4: '#ECEDEF',
  colorFill3: '#EDF0F2',
  colorFill2: '#F2F4F5',
  colorFill1: '#F6F8F9',
  normalColor: '#d9d9d9',
  toolbarButton4: 'rgba(25,31,37,0.04)',
  toolbarButtonHover: 'rgba(25,31,37,0.08)',
  toolbarButtonClick: 'rgba(25,31,37,0.16)',
  toolbarButtonSelect: 'rgba(25,31,37,0.12)',
  toolbarButtonSelectHover: 'rgba(25,31,37,0.16)',
  toolbarButtonIcon: 'rgba(58,60,63,1)',
  toolbarButtonIcon56: 'rgba(17,31,44,.56)',
  toolbarButtonIconDisable: '#A9AEB3',
  toolbarButtonSelectDisable: 'rgba(25,31,37,0.08)',
  toolbarButtonTransparent: 'rgba(0,0,0,0)',
  //template
  templateLevel2BaseColor: 'rgba(23, 26, 29, 0.6)',
  templateLevel1BaseColor: 'rgba(23, 26, 29, 1)'
};
export default _extends({}, commonColors, bizColors, desktopCommon);
