// @ali/4ever-bi/dist-web/plugins/sticker/actions.js
export var STICKER_INSERT = Symbol('STICKER_INSERT');
export var STICKER_SET = Symbol('STICKER_SET');
export var STICKER_PANEL_SET_VISIBLE = Symbol('STICKER_PANEL_SET_VISIBLE');
export var STICKER_MOVE_CARET = Symbol('STICKER_MOVE_CARET');
export var STICKER_SET_RANGE = Symbol('STICKER_SET_RANGE');
export var INSERT_MARKDOWN_EMOJI = Symbol('INSERT_MARKDOWN_EMOJI');
export function insertSticker(payload) {
  return {
    type: STICKER_INSERT,
    payload: payload
  };
}
export function setSticker(payload) {
  return {
    type: STICKER_SET,
    payload: payload
  };
}
export function setStickerPanelVisible(payload) {
  return {
    type: STICKER_PANEL_SET_VISIBLE,
    payload: payload
  };
}
export function moveCaret(payload) {
  return {
    type: STICKER_MOVE_CARET,
    payload: payload
  };
}
export function setRange(payload) {
  return {
    type: STICKER_SET_RANGE,
    payload: payload
  };
}
export function insertMarkdownEmoji(payload) {
  return {
    type: INSERT_MARKDOWN_EMOJI,
    payload: payload
  };
}
