// @ali/we-ui/dist-web/components/time-picker/styled.js
/* eslint-disable @typescript-eslint/explicit-function-return-type */
import styled from 'styled-components';
import { withTheme } from "../theme";
import { Ellipsis } from "../../utils/styled.mixin";
export var PanelWrap = withTheme(styled.div(["display:flex;width:100%;box-sizing:border-box;font-size:", ";color:", ";*{box-sizing:border-box;}"], function (_ref) {
  var theme = _ref.theme;
  return theme.ui.fontSize;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.textColor;
}));
export var PanelColumnWrap = withTheme(styled.div(["flex:1 auto;border-right:1px solid ", ";&:last-child{border-right:0;}"], function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.borderColor;
}));
export var PanelColumnTitle = withTheme(styled.div(["text-align:center;height:32px;line-height:32px;"]));
export var PanelColumnBody = withTheme(styled.div([""]));
export var NumberSliderWrap = withTheme(styled.div(["height:224px;padding-bottom:192px;overflow-y:hidden;&:hover{overflow-y:auto;}"]));
export var NumberSliderNumber = withTheme(styled.div(["height:32px;line-height:32px;text-align:center;user-select:none;&:hover{background-color:", ";cursor:pointer;}&.scrolling{background-color:transparent;cursor:default;}&.actived{background-color:", ";cursor:default;}"], function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.backgroundColor;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.ui.backgroundColor;
}));
export var Trigger = withTheme(styled.div(["box-sizing:border-box;display:inline-block;width:100%;min-width:150px;height:", ";line-height:", ";border:1px solid ", ";transition:all .24s ease;padding:0 ", ";font-size:0 ", ";user-select:none;&:hover{border-color:", ";cursor:pointer;}.calendar-icon{margin-left:", ";}&.placeholder{color:", ";}&.disabled{color:", ";background-color:", ";border-color:", ";cursor:not-allowed;}*{box-sizing:border-box;}"], function (_ref6) {
  var theme = _ref6.theme;
  return theme.ui.elementHeight;
}, function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.elementHeight;
}, function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.borderColor;
}, function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.padding;
}, function (_ref10) {
  var theme = _ref10.theme;
  return theme.ui.fontSize;
}, function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.primaryColor;
}, function (_ref12) {
  var theme = _ref12.theme;
  return theme.ui.marginSmall;
}, function (_ref13) {
  var theme = _ref13.theme;
  return theme.ui.colorText2;
}, function (_ref14) {
  var theme = _ref14.theme;
  return theme.ui.disabledColor;
}, function (_ref15) {
  var theme = _ref15.theme;
  return theme.ui.disabledBackgroundColor;
}, function (_ref16) {
  var theme = _ref16.theme;
  return theme.ui.borderColor;
}));
export var TriggerWrap = withTheme(styled.div(["display:flex;align-items:center;.text{flex:1 auto;width:0;", "}"], Ellipsis));
export var OverlayWrap = withTheme(styled.div(["width:240px;"]));
export var OverlayInputs = withTheme(styled.div(["padding:", ";"], function (_ref17) {
  var theme = _ref17.theme;
  return theme.ui.padding;
}));
export var TimePanelWrap = withTheme(styled.div(["padding:", ";"], function (_ref18) {
  var theme = _ref18.theme;
  return theme.ui.padding;
}));
