// @ali/4ever-plugin-list/dist-node/bi/mobile/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _queries = require("../queries");

var _actions = require("../actions");

var _onAction = _interopRequireDefault(require("./onAction"));

var _onKeyDown = _interopRequireDefault(require("../handlers/onKeyDown"));

var _listWrapper = _interopRequireDefault(require("../components/listWrapper"));

var _onCangjieInput = _interopRequireDefault(require("../handlers/onCangjieInput"));

var _placeholder = _interopRequireDefault(require("../placeholder"));

var _onCangjiePaste = _interopRequireDefault(require("../handlers/onCangjiePaste"));

var _createRenderParagraph = require("../queries/createRenderParagraph");

var _biCommands = require("../commands/biCommands");

var _mo = _interopRequireDefault(require("../../mo"));

var _utils = require("../../utils");

const _createElement = /*#__PURE__*/React.createElement;

const mobileBiFactory = config => {
  const listConfig = config.list || {};
  return { ...(0, _mo.default)({}),

    renderEditable(props, controller, next) {
      return (
        /*#__PURE__*/
        // @ts-ignore
        _createElement(_listWrapper.default, (0, _extends2.default)({}, props, {
          controller: controller
        }), /*#__PURE__*/_createElement(_utils.ListSymbolWarmupWrapper, {
          controller: controller
        }, next()))
      );
    },

    toolbar: (0, _toolbar.default)(listConfig),
    commands: _biCommands.biCommands,
    queries: { ..._queries.queries,
      renderParagraph: (0, _createRenderParagraph.createRenderParagraph)(config)
    },
    actions: _actions.actions,
    onAction: _onAction.default,
    injectToNode: (0, _utils.createInjectToNode)(listConfig),
    onCangjieInput: _onCangjieInput.default,
    onCangjiePaste: (0, _onCangjiePaste.default)(),
    onKeyDown: _onKeyDown.default,
    getNodeStyle: _utils.getNodeStyle,
    models: {
      klass: 'block',
      type: 'paragraph',
      data: {
        ind: [undefined, Object],
        spacing: [undefined, Object],
        jc: [undefined, String],
        blockquote: [undefined, Boolean],
        list: Object,
        calloutPr: [undefined, Object],
        subjectPr: [undefined, Object],
        shd: [undefined, Object]
      }
    },
    placeholder: (0, _placeholder.default)(listConfig)
  };
};

var _default = mobileBiFactory;
exports.default = _default;
