// @ali/we-design-next/dist-node/office/title-cover/style.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.titleCoverStyle = exports.TitleCoverWrapper = exports.SettingButton = exports.RepositionTips = exports.ButtonsWrapper = void 0;
var _styledComponents = _interopRequireDefault(require("styled-components"));
const titleCoverStyle = {
  boxShadow: '0px -2px 8px 0px rgb(0 0 0 / 6%)'
};
exports.titleCoverStyle = titleCoverStyle;
const TitleCoverWrapper = /*#__PURE__*/_styledComponents.default.div(["position:relative;height:234px;box-sizing:border-box;outline:none;background-color:rgb(250,251,252);"]);
exports.TitleCoverWrapper = TitleCoverWrapper;
const RepositionTips = /*#__PURE__*/_styledComponents.default.div(["position:absolute;top:50%;left:50%;transform:translate(-50%,-50%);background:rgba(0,0,0,0.4);border-radius:3px;color:white;font-size:12px;padding:0 20px;pointer-events:none;text-align:center;opacity:", ";transition:opacity ease-in 0.1s;"], p => p.visible ? 1 : 0);
exports.RepositionTips = RepositionTips;
const SettingButton = /*#__PURE__*/_styledComponents.default.div(["user-select:none;transition:background 20ms ease-in 0s;display:inline-flex;align-items:center;padding:4px 6px;font-weight:500;color:rgba(55,53,47,0.6);fill:rgba(55,53,47,0.6);cursor:pointer;background-color:white;font-size:12px;&:hover{background:rgb(239,239,238);}&:first-child{border-top-left-radius:3px;border-bottom-left-radius:3px;}&:last-child{border-top-right-radius:3px;border-bottom-right-radius:3px;}"]);
exports.SettingButton = SettingButton;
const ButtonsWrapper = /*#__PURE__*/_styledComponents.default.div(["display:inline-flex;box-shadow:rgb(15 15 15 / 10%) 0px 0px 0px 1px,rgb(15 15 15 / 10%) 0px 2px 4px;white-space:nowrap;overflow:hidden;"]);
exports.ButtonsWrapper = ButtonsWrapper;
