// @ali/we-design-next/dist-node/office/title-cover/TitleCover.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TitleCover = void 0;
var _react = _interopRequireWildcard(require("react"));
var _locale = require("../../locale");
var _mergeRestProps = require("../../common/react/mergeRestProps");
var _common = require("../../common");
var _cover = require("./components/cover");
var _style = require("./style");
var _utils = require("./utils");
var _constants = require("./constants");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const _createElement = /*#__PURE__*/_react.default.createElement;
const TitleCover = /*#__PURE__*/_react.default.memo( /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    cover,
    renderError,
    ...rest
  } = props;
  const {
    src,
    thumbSrc,
    transform = {}
  } = cover;
  const getCover = (0, _common.useEventCallback)(() => cover);
  const {
    offsetY: dataPosition
  } = transform;
  const [position, setPosition] = (0, _react.useState)(_constants.DEFAULT_POSITION);
  const imageRef = (0, _react.useRef)(null);
  const [isReposition, setIsReposition] = (0, _react.useState)(false);
  const onChangeFnsRef = (0, _react.useRef)([]);
  (0, _react.useImperativeHandle)(ref, () => ({
    onChange: fn => {
      onChangeFnsRef.current.push(fn);
      return () => {
        const i = onChangeFnsRef.current.indexOf(fn);
        if (i !== -1) {
          onChangeFnsRef.current.splice(i, 1);
        }
      };
    },
    setRenderPosition: renderP => {
      setPosition(renderP);
    },
    startReposition: () => {
      setIsReposition(true);
    },
    endReposition: () => {
      setIsReposition(false);
    },
    getImageRef: () => imageRef,
    getCover
  }));
  (0, _react.useEffect)(() => {
    setPosition((0, _utils.getSafePosition)(dataPosition));
  }, [dataPosition]);
  (0, _react.useEffect)(() => {
    for (const fn of onChangeFnsRef.current) {
      fn({
        isReposition,
        position
      });
    }
  }, [isReposition, position]);
  const safePosition = (0, _utils.getSafePosition)(position);
  const t = (0, _locale.useTranslate)();
  const style = {
    cursor: isReposition ? "move" : "auto"
  };
  return /*#__PURE__*/_createElement(_style.TitleCoverWrapper, (0, _mergeRestProps.mergeRestProps)(rest, 'wdn-title-cover', style), /*#__PURE__*/_createElement(_cover.Cover, {
    ref: imageRef,
    src: src,
    thumbSrc: thumbSrc,
    position: safePosition,
    renderError: renderError
  }), /*#__PURE__*/_createElement(_style.RepositionTips, {
    visible: isReposition
  }, t("wdn_title_cover_drag_reposition")));
}));
exports.TitleCover = TitleCover;
