// @ali/4ever-factory/dist-node/tagFactory/components/Container/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _everPluginTag = require("@ali/4ever-plugin-tag");

var _lodash = require("lodash");

const _createElement = /*#__PURE__*/_react.default.createElement;

const Container = props => {
  const {
    children,
    verticalAlign,
    isSelected,
    restrictSize
  } = props;
  const node = props.node;
  const {
    height,
    tagType
  } = node.data;
  const {
    width
  } = node.data;

  const refWrapper = _react.default.useRef(null);

  let cssHeight = 'auto';
  let cssWidth = 'auto';

  if (restrictSize) {
    cssHeight = (0, _lodash.isNumber)(height) && height > 0 ? `${height}px` : height;
    cssWidth = (0, _lodash.isNumber)(width) && width > 0 ? `${width}px` : width;
  }

  const tagStyle = {
    height: cssHeight,
    width: cssWidth,
    verticalAlign,
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    maxWidth: '100%'
  };

  const className = _react.default.useMemo(() => (0, _classnames.default)({
    selected: isSelected
  }), [isSelected]);

  return /*#__PURE__*/_createElement(_everPluginTag.TagWrapper, {
    ref: refWrapper,
    "data-testid": tagType,
    "data-type": _everPluginTag.Tag.TYPE,
    style: tagStyle,
    className: className
  }, children);
};

var _default = Container;
exports.default = _default;
