// @ali/4ever-plugin-sticker/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "StickerManager", {
  enumerable: true,
  get: function () {
    return _sticker.default;
  }
});
Object.defineProperty(exports, "StickerItem", {
  enumerable: true,
  get: function () {
    return _stickerItem.default;
  }
});
Object.defineProperty(exports, "DEFAULT_SZ", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_SZ;
  }
});
Object.defineProperty(exports, "DEFAULT_STICKERS", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_STICKERS;
  }
});
Object.defineProperty(exports, "MOVE_SET", {
  enumerable: true,
  get: function () {
    return _constants.MOVE_SET;
  }
});
Object.defineProperty(exports, "STICKER_NODE_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.STICKER_NODE_TYPE;
  }
});
Object.defineProperty(exports, "STICKER_PANEL_COLUMMS", {
  enumerable: true,
  get: function () {
    return _constants.STICKER_PANEL_COLUMMS;
  }
});
Object.defineProperty(exports, "STICKER_PANEL_HEIGHT", {
  enumerable: true,
  get: function () {
    return _constants.STICKER_PANEL_HEIGHT;
  }
});
Object.defineProperty(exports, "STICKER_PANEL_STYLE", {
  enumerable: true,
  get: function () {
    return _constants.STICKER_PANEL_STYLE;
  }
});
Object.defineProperty(exports, "STICKER_PLACEHOLDER_DATA", {
  enumerable: true,
  get: function () {
    return _constants.STICKER_PLACEHOLDER_DATA;
  }
});
Object.defineProperty(exports, "STICKER_SZ_IN_PANEL", {
  enumerable: true,
  get: function () {
    return _constants.STICKER_SZ_IN_PANEL;
  }
});
Object.defineProperty(exports, "DATA_STICKER", {
  enumerable: true,
  get: function () {
    return _constants.DATA_STICKER;
  }
});
Object.defineProperty(exports, "DEFAULT_LANG", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_LANG;
  }
});
Object.defineProperty(exports, "StickerRole", {
  enumerable: true,
  get: function () {
    return _role.StickerRole;
  }
});
Object.defineProperty(exports, "StickerType", {
  enumerable: true,
  get: function () {
    return _meta.ELEMENT_TYPE;
  }
});
Object.defineProperty(exports, "getEmojiData", {
  enumerable: true,
  get: function () {
    return _getEmojiData.getEmojiData;
  }
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.default;
  }
});
Object.defineProperty(exports, "Sticker", {
  enumerable: true,
  get: function () {
    return _models.default;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.createServerSerializationPlugin;
  }
});
Object.defineProperty(exports, "StickerPanel", {
  enumerable: true,
  get: function () {
    return _stickerPanel.default;
  }
});
Object.defineProperty(exports, "StickerPannelWrapper", {
  enumerable: true,
  get: function () {
    return _styled.StickerPannelWrapper;
  }
});
Object.defineProperty(exports, "StickerSplitLine", {
  enumerable: true,
  get: function () {
    return _styled.StickerSplitLine;
  }
});
Object.defineProperty(exports, "StickerPanelItem", {
  enumerable: true,
  get: function () {
    return _styled.StickerPanelItem;
  }
});
Object.defineProperty(exports, "InlinePlaceholderWrapper", {
  enumerable: true,
  get: function () {
    return _styled.InlinePlaceholderWrapper;
  }
});
Object.defineProperty(exports, "InlinePlaceholderContainer", {
  enumerable: true,
  get: function () {
    return _styled.InlinePlaceholderContainer;
  }
});
Object.defineProperty(exports, "PanelWrapper", {
  enumerable: true,
  get: function () {
    return _styled.PanelWrapper;
  }
});
exports.OldBambooPlugin = void 0;

var _sticker = _interopRequireDefault(require("./utils/sticker"));

var _stickerItem = _interopRequireDefault(require("./utils/stickerItem"));

var _constants = require("./utils/constants");

var _role = require("./utils/role");

var _meta = require("./utils/meta");

var _getEmojiData = require("./utils/getEmojiData");

var _index = _interopRequireDefault(require("./mo/index"));

var _models = _interopRequireDefault(require("./mo/models"));

var _index2 = _interopRequireWildcard(require("./mo/serializer/index"));

var _stickerPanel = _interopRequireDefault(require("./bi/components/stickerPanel"));

var _styled = require("./bi/components/styled");

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;
