// @ali/we-design-next/dist-node/display/popover/align.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.align = void 0;
var _domAlign = _interopRequireDefault(require("dom-align"));
const align = (source, target, placement, [m, s] = [0, 0], resize) => {
  let align = ['tc', 'bc']; // 对应defaultProps的bottom
  let offset = [0, 0];
  switch (placement) {
    case 'topLeft':
      align = ['bl', 'tl'];
      offset = [s, -m];
      break;
    case 'top':
      align = ['bc', 'tc'];
      offset = [s, -m];
      break;
    case 'topRight':
      align = ['br', 'tr'];
      offset = [-s, -m];
      break;
    case 'leftTop':
      align = ['tr', 'tl'];
      offset = [-m, s];
      break;
    case 'rightTop':
      align = ['tl', 'tr'];
      offset = [m, s];
      break;
    case 'left':
      align = ['cr', 'cl'];
      offset = [-m, s];
      break;
    case 'right':
      align = ['cl', 'cr'];
      offset = [m, s];
      break;
    case 'leftBottom':
      align = ['br', 'bl'];
      offset = [-m, -s];
      break;
    case 'rightBottom':
      align = ['bl', 'br'];
      offset = [m, -s];
      break;
    case 'bottomLeft':
      align = ['tl', 'bl'];
      offset = [s, m];
      break;
    case 'bottom':
      align = ['tc', 'bc'];
      offset = [s, m];
      break;
    case 'bottomRight':
      align = ['tr', 'br'];
      offset = [-s, m];
      break;
    default:
      break;
  }
  (0, _domAlign.default)(source, target, {
    points: align,
    offset: offset,
    overflow: {
      adjustX: true,
      adjustY: true,
      resizeWidth: resize,
      resizeHeight: resize
    }
  });
};
exports.align = align;
