// @ali/4ever-bi/dist-node/plugins/image/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.selectImage = selectImage;
exports.removeImage = removeImage;
exports.updateImage = updateImage;
exports.setImageInjection = setImageInjection;
exports.toggleImageSidebar = toggleImageSidebar;
exports.DEPRECATED_ACTIONS = exports.actions = exports.TOGGLE_IMAGE_SIDEBAR = exports.INSERT_IMAGE_PLACEHOLDER = exports.IMAGE_SET_INJECTION = exports.IMAGE_UPDATE = exports.IMAGE_REMOVE = exports.IMAGE_SELECT = void 0;
const IMAGE_SELECT = Symbol('IMAGE_SELECT');
exports.IMAGE_SELECT = IMAGE_SELECT;
const IMAGE_REMOVE = Symbol('IMAGE_REMOVE');
exports.IMAGE_REMOVE = IMAGE_REMOVE;
const IMAGE_UPDATE = Symbol('IMAGE_UPDATE');
exports.IMAGE_UPDATE = IMAGE_UPDATE;
const IMAGE_SET_INJECTION = Symbol('IMAGE_SET_INJECTION');
/**
 * 插入上传图片的 placeholder
 */

exports.IMAGE_SET_INJECTION = IMAGE_SET_INJECTION;
const INSERT_IMAGE_PLACEHOLDER = Symbol('INSERT_IMAGE_PLACEHOLDER');
/**
 * 唤起图片格式侧边栏
 */

exports.INSERT_IMAGE_PLACEHOLDER = INSERT_IMAGE_PLACEHOLDER;
const TOGGLE_IMAGE_SIDEBAR = Symbol('TOGGLE_IMAGE_SIDEBAR');
exports.TOGGLE_IMAGE_SIDEBAR = TOGGLE_IMAGE_SIDEBAR;

function selectImage(placeholderKey) {
  return {
    type: IMAGE_SELECT,
    payload: {
      placeholderKey
    }
  };
}

function removeImage(image) {
  return {
    type: IMAGE_REMOVE,
    payload: {
      node: image
    }
  };
}

function updateImage(image, data, container) {
  return {
    type: IMAGE_UPDATE,
    payload: {
      node: image,
      data,
      container
    }
  };
}

function setImageInjection(key, data) {
  return {
    type: IMAGE_SET_INJECTION,
    payload: {
      key,
      data
    }
  };
}

function toggleImageSidebar(payload) {
  return {
    type: TOGGLE_IMAGE_SIDEBAR,
    payload
  };
}

const actions = {
  selectImage,
  removeImage,
  updateImage,
  setImageInjection,
  toggleImageSidebar
};
/** @deprecated */

exports.actions = actions;
const DEPRECATED_ACTIONS = {
  IMAGE_SELECT,
  IMAGE_REMOVE,
  IMAGE_UPDATE,
  IMAGE_SET_INJECTION,
  INSERT_IMAGE_PLACEHOLDER,
  TOGGLE_IMAGE_SIDEBAR,
  selectImage,
  removeImage,
  updateImage,
  setImageInjection,
  toggleImageSidebar
};
exports.DEPRECATED_ACTIONS = DEPRECATED_ACTIONS;
