// @ali/4ever-bi/dist-node/plugins/image/createRenderNode.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderNode;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _image = _interopRequireDefault(require("./components/image"));

const _createElement = /*#__PURE__*/React.createElement;

function createRenderNode(config, state) {
  return {
    image: (props, controller) => {
      const {
        children,
        parent,
        node,
        ...rest
      } = props;
      const {
        id
      } = node.data;
      const {
        key
      } = node;
      const pluginState = state[id] || state[key] || {};
      return (
        /*#__PURE__*/
        // @ts-ignore
        _createElement(_image.default, (0, _extends2.default)({}, rest, {
          controller: controller,
          pluginState: pluginState,
          imageConfig: config || {},
          node: node,
          parent: parent
        }), children())
      );
    }
  };
}
