// @ali/4ever-bi/dist-node/plugins/image/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "actions", {
  enumerable: true,
  get: function () {
    return _actions.DEPRECATED_ACTIONS;
  }
});
exports.default = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _sidebar = _interopRequireDefault(require("./sidebar"));

var _createRenderNode = _interopRequireDefault(require("./createRenderNode"));

var _createRenderEditable = _interopRequireDefault(require("./createRenderEditable"));

var _actions = require("./actions");

var _insertImage = _interopRequireDefault(require("./commands/insertImage"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onQuery = _interopRequireDefault(require("./handlers/onQuery"));

var _onDrop = _interopRequireDefault(require("./handlers/onDrop"));

var _onCangjiePaste = _interopRequireDefault(require("./handlers/onCangjiePaste"));

var _quickInsertConfig = _interopRequireDefault(require("./quickInsertConfig"));

var _insertImageFile = _interopRequireDefault(require("./commands/insertImageFile"));

var _onCommand = _interopRequireDefault(require("./handlers/onCommand"));

var _createContextMenu = _interopRequireDefault(require("./handlers/createContextMenu"));

var _createRenderHoverToolbar = _interopRequireDefault(require("./createRenderHoverToolbar"));

var _onCangjieInput = _interopRequireDefault(require("./handlers/onCangjieInput"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var utils = _interopRequireWildcard(require("./utils"));

var _queries = require("./queries");

var _createRenderToolbar = _interopRequireDefault(require("./createRenderToolbar"));

var _everPluginImage = require("@ali/4ever-plugin-image");

const state = {};

function createImagePlugin(config) {
  const imageConfig = config.image || {};
  return { ...(0, _everPluginImage.moFactory)({}),
    sidebar: (0, _sidebar.default)(config),
    toolbar: (0, _toolbar.default)(imageConfig),
    menubar: (0, _menubar.default)(imageConfig),
    contextMenu: (0, _createContextMenu.default)(imageConfig, config.sideBarTrigger),
    renderNode: (0, _createRenderNode.default)(imageConfig, state),
    renderEditable: (0, _createRenderEditable.default)(imageConfig),
    renderToolbar: (0, _createRenderToolbar.default)(imageConfig.locale || {}),
    renderHoverToolbar: (0, _createRenderHoverToolbar.default)(imageConfig),
    commands: {
      insertImage: _insertImage.default,
      // @ts-ignore
      insertImageFile: (0, _insertImageFile.default)(imageConfig, state)
    },
    queries: _queries.queries,
    actions: _actions.actions,
    onCommand: (0, _onCommand.default)(config, state),
    onAction: (0, _onAction.default)(config),
    onQuery: (0, _onQuery.default)(imageConfig),
    onCangjiePaste: (0, _onCangjiePaste.default)(config, state),
    onCangjieInput: (0, _onCangjieInput.default)(imageConfig, state),
    onDrop: (0, _onDrop.default)(imageConfig, state),
    onKeyDown: _onKeyDown.default,
    quickInsert: _quickInsertConfig.default,
    models: {
      klass: 'inline',
      type: 'image'
    }
  };
}

createImagePlugin.actions = _actions.DEPRECATED_ACTIONS;
createImagePlugin.utils = utils;
var _default = createImagePlugin;
exports.default = _default;
