// @ali/4ever-plugin-marks/dist-web/utils/createMarkFactory/createRenderToolbar.js
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { ItemSlot, SlotType } from '@ali/4ever-component';
import createMarkButton from "./components/next/createMarkButton";
export default function factory(payload) {
  var type = payload.type,
      hotKey = payload.hotKey,
      delimiter = payload.delimiter,
      defaultIcon = payload.toolbarIcon,
      localeKeys = payload.localeKeys;
  var Component = createMarkButton({
    type: type,
    hotKey: hotKey,
    defaultIcon: defaultIcon,
    localeKeys: localeKeys,
    delimiter: delimiter
  });
  return function createRenderToolbar(locale) {
    return function (props, controller) {
      var buttonName = props.buttonName,
          extraProps = props.extraProps;

      if (buttonName === type) {
        return /*#__PURE__*/_createElement(ItemSlot, {
          display: SlotType.iconButton
        }, /*#__PURE__*/_createElement(Component, _extends({
          controller: controller,
          locale: locale
        }, extraProps)));
      }

      return null;
    };
  };
}
