// @ali/4ever-plugin-marks/dist-web/utils/createMarkFactory/createContextMenu.js
import { capitalize } from 'lodash-es';
import { toggleMark } from "./actions";
export default function createContextMenu(type, hotKey) {
  return function contextMenuPlugin(pluginConfig) {
    var locale = pluginConfig.locale || {};
    return function contextMenu(controller, next) {
      var menus = [];
      var _controller$value = controller.value,
          selection = _controller$value.selection,
          document = _controller$value.document;
      var canToggle = false;

      try {
        // getFragmentAtRange throws unknow error
        // https://aone.alibaba-inc.com/v2/project/995011/bug/43531316
        var fragement = document.getFragmentAtRange(selection);
        canToggle = Boolean(fragement.text);
      } catch (e) {// ignore
      }

      if (canToggle && selection.isExpanded && controller.query("canToggle" + capitalize(type))) {
        menus.push({
          name: locale.buttonTitle,
          action: toggleMark(type),
          key: "mark-toggle-" + type,
          options: {
            keepMenu: true,
            hotKey: hotKey,
            group: 2,
            checked: controller.query("is" + capitalize(type))
          }
        });
      }

      var others = next() || [];
      return [].concat(menus, others);
    };
  };
}
