// @ali/we-ui/dist-web/components/table-wrap-select/TableWrapSelect.js
import React, { useMemo, useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import useControlledValue from "../../hooks/useControlledValue";
import { useTranslate } from "../../locale/locales";
import { OverflowNormal, ClipNormal, WrapNormal } from "../icons";
import { Select, SelectOption } from "../select";
import { OptionWrap, TriggerWrap } from "./styled";

var TableWrapSelect = function TableWrapSelect(props) {
  var locale = props.locale,
      style = props.style,
      onChange = props.onChange,
      title = props.title,
      titles = props.titles,
      zIndex = props.zIndex,
      className = props.className,
      showTitle = props.showTitle,
      disabled = props.disabled;

  var _useControlledValue = useControlledValue(props.value, props.defaultValue),
      value = _useControlledValue[0],
      setValue = _useControlledValue[1];

  var t = useTranslate(locale); // 垂直对齐的选项

  var options = useMemo(function () {
    return [{
      title: t('we_ui_wrap_overflow'),
      value: 'overflow',
      Icon: OverflowNormal
    }, {
      title: t('we_ui_wrap_auto-wrap'),
      value: 'auto-wrap',
      Icon: WrapNormal
    }, {
      title: t('we_ui_wrap_line-clip'),
      value: 'line-clip',
      Icon: ClipNormal
    }];
  }, [t]); // 该组件的trigger区域只显示Icon

  var renderTrigger = useCallback(function (align) {
    var index = options.findIndex(function (option) {
      return option.value === align;
    });
    var currentOption = options[index];
    return /*#__PURE__*/_createElement(TriggerWrap, {
      className: "we-ui-table-wrap-select-trigger",
      style: {
        minWidth: 16
      }
    }, currentOption && /*#__PURE__*/_createElement(React.Fragment, null, /*#__PURE__*/_createElement(currentOption.Icon, null), showTitle && /*#__PURE__*/_createElement("span", {
      className: "title"
    }, currentOption.title)));
  }, [options, showTitle]);
  var handleChange = useCallback(function (e) {
    if (onChange) {
      onChange(e);
    }

    setValue(e);
  }, [setValue, onChange]);
  return /*#__PURE__*/_createElement(Select, {
    value: value,
    onChange: handleChange,
    style: style,
    renderTrigger: renderTrigger,
    zIndex: zIndex,
    title: title,
    className: className,
    disabled: disabled,
    testid: "TableWrapSelect"
  }, options.map(function (option) {
    return /*#__PURE__*/_createElement(SelectOption, {
      value: option.value,
      key: option.value,
      title: titles[option.value]
    }, /*#__PURE__*/_createElement(OptionWrap, null, /*#__PURE__*/_createElement(option.Icon, {
      className: "icon"
    }), /*#__PURE__*/_createElement("span", null, option.title)));
  }));
};

TableWrapSelect.defaultProps = {
  defaultValue: 'overflow',
  titles: {}
};
export default TableWrapSelect;
