// @ali/4ever-bi/dist-node/plugins/refer/commands/insertRefer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

var _v = _interopRequireDefault(require("uuid/v4"));

var _default = (controller, files = []) => {
  controller.command(_everCangjie.Commands.focus);

  if (controller.query('isCollapsedInBlockVoid')) {
    controller.command(_everCangjie.Commands.insertBlock, 'paragraph');
  }

  files.forEach(file => {
    controller.command(_everCangjie.Commands.insertInline, {
      type: 'refer',
      data: {
        id: (0, _v.default)(),
        name: file.fileName,
        type: 'dingpan',
        fileType: file.fileType,
        corpId: file.corpId,
        spaceId: file.spaceId,
        fileId: file.fileId,
        fileSize: file.fileSize,
        viewType: file.viewType || 'link',
        height: file.height
      }
    }).command(_everCangjie.Commands.moveToStartOfNextText).command(_everCangjie.Commands.insertText, ' ');
  });
  return controller.flush();
};

exports.default = _default;
