// @ali/4ever-bi/dist-node/plugins/link/utils/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getCurrentLink", {
  enumerable: true,
  get: function () {
    return _getCurrentLink.default;
  }
});
Object.defineProperty(exports, "hasLink", {
  enumerable: true,
  get: function () {
    return _hasLink.default;
  }
});
Object.defineProperty(exports, "isLink", {
  enumerable: true,
  get: function () {
    return _isLink.default;
  }
});
Object.defineProperty(exports, "isSelectionInLink", {
  enumerable: true,
  get: function () {
    return _isSelectionInLink.default;
  }
});
Object.defineProperty(exports, "isSelectionInLinkCard", {
  enumerable: true,
  get: function () {
    return _isSelectionInLinkCard.default;
  }
});
Object.defineProperty(exports, "getLinkDefaultSz", {
  enumerable: true,
  get: function () {
    return _getLinkDefaultSz.default;
  }
});
Object.defineProperty(exports, "isLinkCardNode", {
  enumerable: true,
  get: function () {
    return _isLinkCard.isLinkCardNode;
  }
});
Object.defineProperty(exports, "isLinkNode", {
  enumerable: true,
  get: function () {
    return _isLinkCard.isLinkNode;
  }
});
Object.defineProperty(exports, "getCardInfo", {
  enumerable: true,
  get: function () {
    return _getCardInfo.default;
  }
});
Object.defineProperty(exports, "getLinkCardState", {
  enumerable: true,
  get: function () {
    return _getLinkCardState.getLinkCardState;
  }
});

var _getCurrentLink = _interopRequireDefault(require("./getCurrentLink"));

var _hasLink = _interopRequireDefault(require("./hasLink"));

var _isLink = _interopRequireDefault(require("./isLink"));

var _isSelectionInLink = _interopRequireDefault(require("./isSelectionInLink"));

var _isSelectionInLinkCard = _interopRequireDefault(require("./isSelectionInLinkCard"));

var _getLinkDefaultSz = _interopRequireDefault(require("./getLinkDefaultSz"));

var _isLinkCard = require("./isLinkCard");

var _getCardInfo = _interopRequireDefault(require("./getCardInfo"));

var _getLinkCardState = require("./getLinkCardState");
