// @ali/4ever-component/dist-web/Placeholder/PlaceholderInlineToolbar.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Button, DeleteBetaNormal } from '@ali/we-design';
import { InlineToolbar } from "../Toolbar/inlineToolbar";

var _ref2 = /*#__PURE__*/_createElement(DeleteBetaNormal, null);

var PlaceholderInlineToolbar = function PlaceholderInlineToolbar(props) {
  var onDeletePlaceholder = function onDeletePlaceholder(e) {
    e.preventDefault();
    e.stopPropagation();
    props.handleRemovePlaceholder();
  };

  var _ref = /*#__PURE__*/_createElement(Button, {
    testid: "ph-delete",
    onClick: onDeletePlaceholder,
    bordered: false,
    icon: _ref2
  });

  var DelPlaceholderButton = function DelPlaceholderButton() {
    return _ref;
  };

  var customToolButtons = {
    'placeholder.del': DelPlaceholderButton
  };
  var layout = [{
    toolButtons: [{
      name: 'placeholder.del'
    }]
  }];
  return /*#__PURE__*/_createElement(InlineToolbar, {
    visible: true,
    triggerNode: props.triggerNode || undefined,
    customToolButtons: customToolButtons,
    inlineLayout: layout
  });
};

export default /*#__PURE__*/React.memo(PlaceholderInlineToolbar);
