// @ali/we-design-next/dist-web/form/slider/styled.js
import styled, { css } from 'styled-components';
export var SliderContainer = /*#__PURE__*/styled.div(["user-select:none;position:relative;min-height:30px;", ""], function (_ref) {
  var isMove = _ref.isMove;
  return isMove ? 'cursor: pointer;' : '';
});
export var StyledSliderLine = /*#__PURE__*/styled.div(["position:relative;top:9px;width:100%;height:4px;border-radius:10px;"]);
export var LineContent = /*#__PURE__*/styled.div(["position:absolute;height:100%;border-radius:9px;"]);
export var StyledSliderMarks = /*#__PURE__*/styled.div(["position:absolute;top:7px;"]);
export var Mark = /*#__PURE__*/styled.div(["width:1px;height:8px;"]);
export var MarkText = /*#__PURE__*/styled.div(["font-size:13px;position:relative;left:-50%;"]);
var tipStyle = /*#__PURE__*/css(["position:absolute;bottom:20px;left:7px;padding:6px 4px;font-size:12px;white-space:nowrap;border-radius:4px;transform:translate(-50%,0);"]);
export var StyledSliderPoint = /*#__PURE__*/styled.div(["position:absolute;top:11px;width:12px;height:12px;border-radius:100px;transform:translate(-50%,-50%);", ""], function (_ref2) {
  var isMove = _ref2.isMove;
  return isMove ? css(["width:14px;height:14px;"]) : css(["&:hover{width:14px;height:14px;box-shadow:none;[role=\"hide\"]{", ";display:block;}}"], tipStyle);
});
export var PointTip = /*#__PURE__*/styled.div(["&[role=\"move\"]{", ";}&[role=\"hide\"]{display:none;}"], tipStyle);
