// @ali/4ever-plugin-calendar-card/dist-web/utils/components/receiversParagraph/styled.js
import styled from 'styled-components';
import { CALENDAR_CARD_MAX_RECEIVER_WIDTH, RECEIVER_MARGIN, AVATAR_WRAPPER_SZ, NAME_WRAPPER_PADDING, RECEIVER_WRAPPER_MARGIN } from "../../constants";
export var ReceiverWrapper = /*#__PURE__*/styled.span(["height:24px;line-height:24px;width:auto;cursor:", ";background-color:", ";border-radius:4px;display:inline-flex;margin-right:", "px;margin-bottom:", "px;max-width:", "px;&:hover{background-color:", ";color:", ";svg{color:", ";}}"], function (props) {
  return props.isReadonly ? 'auto' : 'pointer';
}, function (props) {
  return props.backgroundColor || 'rgba(17, 31, 44, 0.04)';
}, RECEIVER_WRAPPER_MARGIN, RECEIVER_MARGIN, CALENDAR_CARD_MAX_RECEIVER_WIDTH, function (props) {
  return props.backgroundColor || (props.isReadonly ? 'rgba(17,31,44,0.04)' : 'rgba(17,31,44,0.08)');
}, function (props) {
  return props.textHoverColor || '';
}, function (props) {
  return props.textHoverColor || '';
});
export var AvatarWrapper = /*#__PURE__*/styled.span(["height:", "px;line-height:", "px;width:", "px;display:flex;align-items:center;padding-left:2px;"], AVATAR_WRAPPER_SZ, AVATAR_WRAPPER_SZ, AVATAR_WRAPPER_SZ);
export var NameWrapper = /*#__PURE__*/styled.span(["overflow:hidden;display:inline-flex;flex-grow:1;padding-left:", "px;height:24px;justify-content:center;flex-direction:column;"], NAME_WRAPPER_PADDING);
export var AuxiliaryLoadingWrapper = /*#__PURE__*/styled.img(["left:-999999999px;position:absolute;visibility:hidden;"]);
export var EllipsisWrapper = /*#__PURE__*/styled.span(["white-space:nowrap;overflow:hidden;text-overflow:ellipsis;"]);
export var TypeName = /*#__PURE__*/styled.span(["color:#9fa7b0;"]);
export var TipWrapper = /*#__PURE__*/styled.span(["min-width:6px;line-height:24px;height:24px;display:inline-flex;flex:none;margin-right:2px;"]);
export var AddReceiverIconWrapper = /*#__PURE__*/styled.span(["width:20px;height:20px;align-items:center;display:inline-flex;.weicon{width:20px;height:20px;color:red;& > svg{width:20px;height:20px;}}"]);
export var SquareAvatarWrapper = /*#__PURE__*/styled.span(["border-radius:4px;background-image:url(", ");background-size:cover;background-position:center center;"], function (props) {
  return props.url || '';
});
export var TipContentWrapper = /*#__PURE__*/styled.span(["width:auto;font-size:12px;text-overflow:ellipsis;white-space:nowrap;"]);
export var SquareTextAvatarWrapper = /*#__PURE__*/styled.span(["width:20px;height:20px;border-radius:4px;overflow:hidden;& > span{width:100%;height:100%;color:#ffffff;display:flex;justify-content:center;align-items:center;font-size:12px;background-color:#0089ff;}"]); // 兼容iOS端显示

export var TipTextWrapper = /*#__PURE__*/styled.span(["transform:scale(0.9);display:inline-block;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;color:rgb(132,134,138);"]);
export var ContainerWapper = /*#__PURE__*/styled.div(["display:flex;flex-wrap:wrap;"]);
export var TotalBodyWapper = /*#__PURE__*/styled.span(["margin:", ";display:flex;"], function (props) {
  return props.margin;
});
export var TotalIconWapper = /*#__PURE__*/styled.span(["display:inline-flex;align-items:center;"]);
