// @ali/4ever-plugin-table/dist-web/bi/mobile/index.js
import _extends from "@babel/runtime/helpers/extends";
import toolbar from "./toolbar";
import createRenderNode from "./createRenderNode";
import createOnAction from "./handlers/createOnAction";
import onCangjieInput from "../handlers/onCangjieInput";
import onKeyDown from "./handlers/onKeyDown";
import onCangjiePaste from "../handlers/onCangjiePaste";
import queries from "./queries";
import insertTable from "../commands/insertTable";
import createOnCangjieCopy from "../handlers/createOnCangjieCopy";
import createOnCangjieCut from "../handlers/createOnCangjieCut";
import { moFactory } from "../../mo";
import getNodeStyle from "../../common/handlers/getNodeStyle";
var commands = {
  insertTable: insertTable
};
export var mobileBiFactory = function mobileBiFactory(configs) {
  var tableConfig = configs.table || {};
  var _configs$clipboard = configs.clipboard,
      clipboard = _configs$clipboard === void 0 ? {} : _configs$clipboard;

  var mobileConfig = _extends({}, configs, {
    draggable: {
      disabled: true
    },
    table: tableConfig
  });

  return _extends({}, moFactory({}), {
    toolbar: toolbar(tableConfig),
    // @ts-ignore
    renderNode: createRenderNode(mobileConfig),
    onAction: createOnAction(tableConfig),
    onCangjieInput: onCangjieInput,
    onKeyDown: onKeyDown,
    onCangjieCopy: createOnCangjieCopy(true),
    onCangjieCut: createOnCangjieCut(true),
    // TODO: 类型生效后这里就有 typing 问题 了
    onCangjiePaste: onCangjiePaste(clipboard, tableConfig),
    getNodeStyle: getNodeStyle,
    queries: queries,
    commands: commands,
    models: [{
      klass: 'block',
      type: 'table'
    }, {
      klass: 'block',
      type: 'table-cell'
    }, {
      klass: 'block',
      type: 'table-row'
    }]
  });
};
