// @ali/4ever-plugin-table/dist-web/bi/mobile/queries.js
import canInsertTable from "../queries/canInsertTable";
import { isSelectionInTable } from "../queries/isSelectionInTable";
import { isSelectionInTableCell } from "../queries/isSelectionInTableCell";
import calcMinTableColWidth from "../../common/queries/calcMinTableColWidth";
import getSelectionContainers from "../../common/queries/getSelectionContainers";
import isNodeInTable from "../../common/queries/isNodeInTable";
import { isTableNode } from "../../common/queries/isTableNode";
import tableSelection from "../queries/tableSelection";
import getSelectionFromTableSelection from "../queries/getSelectionFromTableSelection";
import viewTableSelections from "../queries/viewTableSelections";
import isHideSelection from "../queries/isHideSelection";
import isTableFromSameOrigin from "../queries/isTableFromSameOrigin";
import isSelectWholeTable from "../queries/isSelectWholeTable";
var queries = {
  calcMinTableColWidth: calcMinTableColWidth,
  canInsertTable: canInsertTable,
  isSelectionInTable: isSelectionInTable,
  isSelectionInTableCell: isSelectionInTableCell,
  getSelectionContainers: getSelectionContainers,
  isNodeInTable: isNodeInTable,
  isTableNode: isTableNode,
  tableSelection: tableSelection,
  viewTableSelections: viewTableSelections,
  getSelectionFromTableSelection: getSelectionFromTableSelection,
  isHideSelection: isHideSelection,
  isTableFromSameOrigin: isTableFromSameOrigin,
  isSelectWholeTable: isSelectWholeTable
};
export default queries;
