// @ali/4ever-plugin-color/dist-node/bi/components/next/colorButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ColorButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _actions = require("../../actions");

var _constants = require("../../constants");

var _everPluginPref = require("@ali/4ever-plugin-pref");

var _everComponent = require("@ali/4ever-component");

var _weToolbar = require("@ali/we-toolbar");

const _createElement = /*#__PURE__*/React.createElement;
;

function ColorButton(props) {
  const {
    controller,
    locale,
    ...rest
  } = props;
  const color = controller.query('getColor');
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const handleChange = React.useCallback(colorValue => controller.run('onAction', (0, _actions.setColor)(colorValue)), [controller]);
  const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInLinkCard');
  const currentPreset = controller.query('getColorPanelMode')?.mode;
  const handleSelectValueChange = React.useCallback(mode => {
    controller.run('onAction', (0, _everPluginPref.setColorPanelMode)({
      mode: mode === 'art' ? _weToolbar.EColorPanelMode.art : _weToolbar.EColorPanelMode.normal
    }));
  }, [controller]);
  return /*#__PURE__*/_createElement(_everComponent.FontColorPicker, (0, _extends2.default)({
    testid: "bi-toolbar-color",
    disabled: !!isDisabled,
    color: color,
    onColorChange: handleChange,
    defaultColor: _constants.DEFAULT_COLOR,
    autoColor: _constants.DEFAULT_COLOR,
    title: locale.buttonTitle,
    tooltip: locale.buttonTitle,
    currentPreset: currentPreset,
    onCurrentPresetChange: handleSelectValueChange
  }, rest));
}
