// @ali/4ever-plugin-clipboard/dist-node/bi/components/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PanelWrapper = exports.ModeItemContent = exports.ModeItemIcon = exports.ModeItem = exports.ModeMenu = exports.PanelWrapper4Custom = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

const PanelWrapper = /*#__PURE__*/_styledComponents.default.div(["position:absolute;z-index:999;border-radius:4px;background-color:#ffffff;box-shadow:0px 0px 1px rgba(0,0,0,0.24),0px 8px 24px rgba(0,0,0,0.16);font-size:14px;"]);

exports.PanelWrapper = PanelWrapper;

const PanelWrapper4Custom = /*#__PURE__*/_styledComponents.default.div(["color:white;position:absolute;z-index:99;"]);

exports.PanelWrapper4Custom = PanelWrapper4Custom;

const ModeMenu = /*#__PURE__*/_styledComponents.default.div(["min-width:169px;border-radius:8px;overflow:hidden;padding:8px 0;&:hover{cursor:pointer;}"]);

exports.ModeMenu = ModeMenu;

const ModeItem = /*#__PURE__*/_styledComponents.default.div(["width:100%;height:32px;padding:0 8px;line-height:32px;&:hover{background-color:rgba(17,31,44,0.08);}"]);

exports.ModeItem = ModeItem;

const ModeItemIcon = /*#__PURE__*/_styledComponents.default.span(["display:inline-block;width:20px;padding-right:8px;"]);

exports.ModeItemIcon = ModeItemIcon;

const ModeItemContent = /*#__PURE__*/_styledComponents.default.span(["display:inline-block;"]);

exports.ModeItemContent = ModeItemContent;
