// @ali/4ever-plugin-table/dist-node/zhi/handlers/createRenderNode.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createRenderNode = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _lodash = require("lodash");

var _models = _interopRequireDefault(require("../../mo/models"));

var _tableRow = _interopRequireDefault(require("../../mo/models/tableRow"));

var _tableCell = _interopRequireDefault(require("../../mo/models/tableCell"));

var _table = _interopRequireDefault(require("../components/table"));

var _tableRow2 = _interopRequireDefault(require("../../components/tableRow"));

var _tableCell2 = _interopRequireDefault(require("../../components/tableCell"));

var _tableWrapper = _interopRequireDefault(require("../components/tableWrapper"));

const _createElement = /*#__PURE__*/React.createElement;

const createRenderNode = (config, isMobile) => {
  return {
    [_models.default.TYPE]: props => {
      const {
        node
      } = props;

      if (!_models.default.isValidTable(node)) {
        return /*#__PURE__*/_createElement(React.Fragment, null);
      }

      return /*#__PURE__*/_createElement(TableWithWrapper, {
        config: config,
        props: props,
        isMobile: isMobile
      });
    },
    [_tableRow.default.TYPE]: props => {
      return /*#__PURE__*/_createElement(_tableRow2.default, props);
    },
    [_tableCell.default.TYPE]: props => {
      return /*#__PURE__*/_createElement(_tableCell2.default, props);
    }
  };
};

exports.createRenderNode = createRenderNode;

function TableWithWrapper(p) {
  const {
    config,
    props,
    isMobile = false
  } = p;
  const {
    attributes,
    visible
  } = props;
  const otherProps = (0, _lodash.omit)(props, 'attributes');
  return /*#__PURE__*/_createElement(_tableWrapper.default, (0, _extends2.default)({}, attributes, {
    visible: visible,
    "data-type": "table",
    isMobile: isMobile
  }), /*#__PURE__*/_createElement(_table.default, (0, _extends2.default)({}, otherProps, {
    isMobile: isMobile,
    config: config
  })));
}
