// @ali/4ever-plugin-marks/dist-node/bi/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.biFactory = void 0;

var _index = require("../mo/index");

var _renderMark = _interopRequireDefault(require("../common/renderMark"));

var _react = _interopRequireDefault(require("react"));

var _commands = require("./commands");

const biFactory = () => {
  return { ...(0, _index.moFactory)({}),
    renderMark: _renderMark.default,
    models: [],
    commands: _commands.commands,

    renderCaret(props, controller, next) {
      const child = next();

      if (!child) {
        return null;
      }

      const {
        value
      } = controller;
      const {
        document,
        selection
      } = value;
      const marks = !selection || selection.isExpanded ? [] : selection.marks || document.getMarksAtRange(selection);
      const style = marks.reduce((s, mark) => {
        if (mark.type === 'italic') {
          return { ...s,
            transformOrigin: 'right 80%',
            transform: 'rotate(13deg)'
          };
        }

        return s;
      }, {}); // TODO: 各家处理光标的策略不同，后续可以根据临时选区的sz推算光标的高度

      return /*#__PURE__*/_react.default.cloneElement(child, {
        style
      });
    }

  };
};

exports.biFactory = biFactory;
