// @ali/4ever-plugin-list/dist-web/mo/serializer/htmlToJsonML/utils/index.js
import _extends from "@babel/runtime/helpers/extends";
import { getTagName, getAttributes, getChildren, paragraphTags } from '@ali/4ever-utils';
export { isMsWordList, transfromIndLeft, removeSymbolText } from "./msWordUtils";
export var listTags = {
  ol: 'ol',
  ul: 'ul',
  li: 'li'
}; // Refactor: 与 Bi 共用

export function getRandomId() {
  if (process.env.NODE_ENV === 'test') return 'test';
  return Math.random().toString(36).slice(2);
}
export function shouldFlattenChildren(li) {
  var children = getChildren(li);
  return children.some(function (child) {
    var childTagName = getTagName(child);
    return !!paragraphTags[childTagName] || childTagName === 'table';
  });
}
export function getFlattenedChildren(li) {
  var children = getChildren(li);
  return children.map(function (child, index) {
    if (index === 0) {
      if (getTagName(child) === 'p') {
        return ['p', _extends({}, getAttributes(li), getAttributes(child))].concat(getChildren(child));
      }

      return ['p', _extends({}, getAttributes(li)), child];
    }

    return child;
  });
}
