// @ali/4ever-plugin-calendar-card/dist-web/utils/components/extraArea/RefreshTips.js
import React, { useRef, useEffect } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { ReportFillNormal } from '@ali/we-design';
import preventDefaultOnClick from "../../uitls/preventDefaultOnClick";
import { RefreshTipsTextWrapper, RefreshTipsWrapper } from "./styled";

var _ref = /*#__PURE__*/_createElement(ReportFillNormal, null);

var RefreshTips = function RefreshTips(props) {
  var offsetX = props.offsetX,
      refreshTips = props.refreshTips,
      tipsNode = props.tipsNode,
      onChangeOffsetX = props.onChangeOffsetX;
  var tipsRef = useRef(null);
  useEffect(function () {
    if (!tipsRef.current) return;
    var offsetWidth = tipsRef.current.offsetWidth;
    onChangeOffsetX(48 - offsetWidth / 2 - 12 * 2 + 2);
  }, [onChangeOffsetX]);
  return /*#__PURE__*/_createElement(RefreshTipsWrapper, {
    ref: tipsRef,
    offsetX: Math.abs(offsetX),
    onClick: preventDefaultOnClick,
    onMouseDown: preventDefaultOnClick
  }, tipsNode || /*#__PURE__*/_createElement(React.Fragment, null, _ref, /*#__PURE__*/_createElement(RefreshTipsTextWrapper, null, refreshTips)));
};

export default RefreshTips;
