// @ali/4ever-bi/dist-node/plugins/common/commonPlugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommonPlugin = void 0;

var _everUtils = require("@ali/4ever-utils");

// reexport for bi api, 插件改为从 4ever-utils 引入
const CommonPlugin = {
  utils: {
    toggleRangesMark: (controller, mark) => controller.command('toggleRangesMark', mark),
    isAtEndOfParagraph: _everUtils.isAtEndOfParagraph,
    isAtStartOfParagraph: _everUtils.isAtStartOfParagraph,
    //isSection,
    isSelectionAtSameBlock: _everUtils.isSelectionAtSameBlock,
    //isSingleSection,
    getPointAndBlockOnSelection: _everUtils.getPointAndBlockOnSelection,
    //unwrapSingleSection,
    changeNullSelectionMark: _everUtils.changeNullSelectionMark,
    generateKey: _everUtils.generateKey,
    getClosestParagraph: _everUtils.getClosestParagraph,
    getRandomId: _everUtils.getRandomId,
    hasMark: _everUtils.hasMark,
    isCollapsedInEmptyText: _everUtils.isCollapsedInEmptyText,
    isInsertInline: _everUtils.isInsertInline,
    re: _everUtils.re,
    getMatches: _everUtils.getMatches,
    getOffset: _everUtils.getOffset,
    getMarkTextMatchInfo: _everUtils.getMarkTextMatchInfo
  }
};
exports.CommonPlugin = CommonPlugin;
