// @ali/we-ui/dist-web/components/table-set/TableSet.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Button } from "../button";
import { TablePanel } from "./TablePanel";
import { Dropdown } from "../dropdown";
export var TableSet = /*#__PURE__*/function (_React$PureComponent) {
  _inheritsLoose(TableSet, _React$PureComponent);

  function TableSet() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$PureComponent.call.apply(_React$PureComponent, [this].concat(args)) || this;
    _this.state = {
      active: false
    };

    _this.getTablePanel = function () {
      var _this$props = _this.props,
          row = _this$props.row,
          col = _this$props.col;
      return /*#__PURE__*/_createElement(TablePanel, {
        row: row,
        col: col,
        onPick: _this.handlePick
      });
    };

    _this.handlePick = function (row, col) {
      var onChange = _this.props.onChange;

      if (onChange) {
        onChange(row, col);
      }

      _this.setState({
        active: false
      });
    };

    _this.handleChangeActive = function (active) {
      _this.setState({
        active: active
      });
    };

    return _this;
  }

  var _proto = TableSet.prototype;

  _proto.render = function render() {
    var _this$props2 = this.props,
        btnIcon = _this$props2.btnIcon,
        style = _this$props2.style,
        _this$props2$bordered = _this$props2.bordered,
        bordered = _this$props2$bordered === void 0 ? false : _this$props2$bordered,
        zIndex = _this$props2.zIndex,
        className = _this$props2.className,
        disabled = _this$props2.disabled,
        title = _this$props2.title,
        size = _this$props2.size;
    var active = this.state.active;
    return /*#__PURE__*/_createElement(Dropdown, {
      visible: active,
      overlay: this.getTablePanel(),
      onVisibleChange: this.handleChangeActive,
      zIndex: zIndex,
      title: title
    }, /*#__PURE__*/_createElement(Button, {
      active: active,
      style: style,
      icon: btnIcon,
      bordered: bordered,
      className: className,
      disabled: disabled,
      size: size,
      "data-testid": "button-TableSet"
    }));
  };

  return TableSet;
}(React.PureComponent);
