// @ali/4ever-factory/dist-web/embedFactory/mobileEmbedPluginFactory.js
import _extends from "@babel/runtime/helpers/extends";
import { upperFirst } from 'lodash-es';
import insertEmbed from "./commands/insertEmbed";
import insertFile from "./commands/insertFile";
import * as actions from "./actions";
import onAction from "./handlers/onAction";
export { actions };
export default function mobileEmbedPluginFactory(params, moFactory) {
  var createRenderNode = params.createRenderNode,
      commands = params.commands,
      configName = params.configName,
      _params$actions = params.actions,
      customActions = _params$actions === void 0 ? {} : _params$actions;
  return function (config) {
    var _extends2;

    var pluginConfig = config[configName] || {};
    var uploadFile = pluginConfig.uploadFile,
        _pluginConfig$enableF = pluginConfig.enableFilePreview,
        enableFilePreview = _pluginConfig$enableF === void 0 ? false : _pluginConfig$enableF; // 目前存在 insertVideo, insertAudio, insertEmbed 三个 command

    var insertCommandName = "insert" + upperFirst(configName);
    return _extends({}, moFactory({}), {
      key: configName,
      renderNode: createRenderNode(config),
      commands: _extends((_extends2 = {}, _extends2[insertCommandName] = insertEmbed(uploadFile, enableFilePreview, configName), _extends2.insertFile = insertFile, _extends2), commands),
      onAction: onAction,
      models: {
        klass: 'block',
        type: 'embed'
      }
    });
  };
}
