// @ali/4ever-plugin-quickinsert/dist-node/bi/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.insertQuickInsert = insertQuickInsert;
exports.showQuickInsert = showQuickInsert;
exports.setQuickInsert = setQuickInsert;
exports.resetQuickInsert = resetQuickInsert;
exports.removeQuickInsert = removeQuickInsert;
exports.QUICK_INSERT_REMOVE = exports.QUICK_INSERT_RESET = exports.QUICK_INSERT_SET = exports.QUICK_INSERT_SHOW = exports.QUICK_INSERT_INSERT = void 0;
const QUICK_INSERT_INSERT = Symbol('insert-quick-insert');
exports.QUICK_INSERT_INSERT = QUICK_INSERT_INSERT;
const QUICK_INSERT_SHOW = Symbol('show-quick-insert');
exports.QUICK_INSERT_SHOW = QUICK_INSERT_SHOW;
const QUICK_INSERT_SET = Symbol('set-quick-insert');
exports.QUICK_INSERT_SET = QUICK_INSERT_SET;
const QUICK_INSERT_RESET = Symbol('reset-quick-insert');
exports.QUICK_INSERT_RESET = QUICK_INSERT_RESET;
const QUICK_INSERT_REMOVE = Symbol('remove-quick-insert');
exports.QUICK_INSERT_REMOVE = QUICK_INSERT_REMOVE;

function insertQuickInsert(plugin) {
  return {
    type: QUICK_INSERT_INSERT,
    payload: {
      plugin
    }
  };
}

function showQuickInsert() {
  return {
    type: QUICK_INSERT_SHOW,
    payload: {}
  };
}

function setQuickInsert(plugin) {
  return {
    type: QUICK_INSERT_SET,
    payload: {
      plugin
    }
  };
}

function resetQuickInsert(plugin) {
  return {
    type: QUICK_INSERT_RESET,
    payload: {
      plugin
    }
  };
}

function removeQuickInsert() {
  return {
    type: QUICK_INSERT_REMOVE,
    payload: {}
  };
}
