// @ali/we-ui/dist-web/components/list-set/ListSet.js
import React, { useMemo, useState, useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { SplitButton } from "../button";
import { BulletedlistNormal, NumberedlistNormal } from "../icons";
import styled from 'styled-components';
import { withTheme } from "../theme";
import { classnames } from "../../utils";
import { useControlledValue } from "../../hooks";
export var DEFAULT_BULLET_LIST_ARRAY = ['SCIR_ECIR_SREC', 'SFDI_SRIG_SREC', 'EREC_EREC_EREC', 'SARR_SDIO_SCIR', 'SSTA_ECIR_SREC', 'SRIG_ECIR_SREC'];
export var DEFAULT_NUMBER_LIST_ARRAY = ['DEC_LEN_LROM_P', 'DEC_LEN_LROM_H', 'DEC_DEC_DEC_P', 'UEN_LEN_LROM_P', 'UROM_UEN_DEC_P', 'DDEC_LEN_LROM_P', 'ZH_DEC_LEN_D', 'ZH_DEC_LEN_A'];
export var LIST_STYLE_MAP = {
  SCIR_ECIR_SREC: 'https://img.alicdn.com/tfs/TB1mWdYbAcx_u4jSZFlXXXnUFXa-78-78.svg',
  SFDI_SRIG_SREC: 'https://img.alicdn.com/tfs/TB1hpepOpP7gK0jSZFjXXc5aXXa-78-78.svg',
  EREC_EREC_EREC: 'https://img.alicdn.com/tfs/TB1cJR4czMZ7e4jSZFOXXX7epXa-78-78.svg',
  SARR_SDIO_SCIR: 'https://img.alicdn.com/tfs/TB1aaWcOEY1gK0jSZFMXXaWcVXa-78-78.svg',
  SSTA_ECIR_SREC: 'https://img.alicdn.com/tfs/TB1VoX8OxD1gK0jSZFsXXbldVXa-78-78.svg',
  SRIG_ECIR_SREC: 'https://img.alicdn.com/tfs/TB1cqp4OAL0gK0jSZFtXXXQCXXa-78-78.svg',
  DEC_LEN_LROM_P: 'https://img.alicdn.com/tfs/TB19lt9OEz1gK0jSZLeXXb9kVXa-78-78.svg',
  DEC_LEN_LROM_H: 'https://img.alicdn.com/tfs/TB18HydOAL0gK0jSZFAXXcA9pXa-78-78.svg',
  DEC_DEC_DEC_P: 'https://img.alicdn.com/tfs/TB1gHycOqL7gK0jSZFBXXXZZpXa-78-78.svg',
  UEN_LEN_LROM_P: 'https://img.alicdn.com/tfs/TB1VaOnOBr0gK0jSZFnXXbRRXXa-78-78.svg',
  UROM_UEN_DEC_P: 'https://img.alicdn.com/tfs/TB1xcqhOAY2gK0jSZFgXXc5OFXa-78-78.svg',
  DDEC_LEN_LROM_P: 'https://img.alicdn.com/tfs/TB1yjive8Bh1e4jSZFhXXcC9VXa-78-78.svg',
  ZH_DEC_LEN_D: 'https://img.alicdn.com/tfs/TB1bFQZ0aL7gK0jSZFBXXXZZpXa-82-82.svg',
  ZH_DEC_LEN_A: 'https://img.alicdn.com/tfs/TB11.iDj4vbeK8jSZPfXXariXXa-82-82.svg'
};
var OverlayContent = /*#__PURE__*/styled.div(["width:272px;display:flex;flex-wrap:wrap;justify-content:flex-start;padding:4px;"]);
var ListStyleItemWrap = /*#__PURE__*/styled.div([""]);
var ListStyleItem = withTheme(styled.img(["width:80px;height:80px;margin:4px;border:1px solid rgba(17,31,44,0.12);border-radius:2px;cursor:pointer;&:hover{border:1px solid ", ";}&.actived{border:2px solid ", ";}"], function (_ref) {
  var theme = _ref.theme;
  return theme.ui.primaryColor;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.primaryColor;
}));

var _ref3 = /*#__PURE__*/_createElement(BulletedlistNormal, null);

var _ref4 = /*#__PURE__*/_createElement(NumberedlistNormal, null);

export var ListSet = function ListSet(props) {
  var _props$mode = props.mode,
      mode = _props$mode === void 0 ? 'ul' : _props$mode,
      defaultListStyle = props.defaultListStyle,
      listStyle = props.listStyle,
      onChange = props.onChange,
      onClick = props.onClick,
      className = props.className,
      style = props.style,
      disabled = props.disabled,
      title = props.title,
      active = props.active,
      size = props.size;
  var listStyleArray = useMemo(function () {
    return mode === 'ul' ? DEFAULT_BULLET_LIST_ARRAY : DEFAULT_NUMBER_LIST_ARRAY;
  }, [mode]);

  var _useControlledValue = useControlledValue(listStyle, defaultListStyle || listStyleArray[0]),
      value = _useControlledValue[0],
      setValue = _useControlledValue[1];

  var _useState = useState(false),
      overlayVisible = _useState[0],
      setOverlayVisible = _useState[1];

  var handleChange = useCallback(function (e) {
    if (onChange) {
      onChange(e);
    }

    setValue(e);
    setOverlayVisible(!overlayVisible);
  }, [onChange, setValue, overlayVisible]);
  var handleVisibleChange = useCallback(function (visible) {
    setOverlayVisible(visible);
  }, []);
  var handleClickLeft = useCallback(function () {
    if (onClick) {
      onClick(value);
    }
  }, [onClick, value]);

  var getOverlay = function getOverlay() {
    return /*#__PURE__*/_createElement(OverlayContent, null, listStyleArray.map(function (item) {
      return /*#__PURE__*/_createElement(ListStyleItemWrap, {
        key: item,
        onClick: function onClick() {
          return handleChange(item);
        },
        role: "button"
      }, /*#__PURE__*/_createElement(ListStyleItem, {
        src: LIST_STYLE_MAP[item],
        alt: "list-style-item",
        className: classnames({
          actived: item === value
        }),
        "data-testid": "option-" + item
      }));
    }));
  };

  var renderTriggerIcon = useMemo(function () {
    return mode === 'ul' ? _ref3 : _ref4;
  }, [mode]);
  return /*#__PURE__*/_createElement(SplitButton, {
    overlay: getOverlay(),
    onClickLeft: handleClickLeft,
    className: className,
    disabled: disabled,
    style: style,
    active: overlayVisible,
    onActiveChange: handleVisibleChange,
    testid: "ListSet",
    title: title,
    highlight: active,
    size: size
  }, renderTriggerIcon);
};
