// @ali/4ever-bi/dist-web/plugins/link/commands/switchLinkDisplayType.js
import _extends from "@babel/runtime/helpers/extends";
import { Commands } from '@ali/4ever-cangjie';
import removeinjectedCardInfo from "./removeinjectedCardInfo";
export default function switchLinkDisplayType(controller, node, displayType) {
  var _cardInfo$displayType;

  var _node$data = node.data,
      href = _node$data.href,
      _node$data$cardInfo = _node$data.cardInfo,
      cardInfo = _node$data$cardInfo === void 0 ? {} : _node$data$cardInfo;
  var originalType = (_cardInfo$displayType = cardInfo == null ? void 0 : cardInfo.displayType) != null ? _cardInfo$displayType : 'link'; // 更新需要控制粒度，避免生成额外 op

  if (originalType !== displayType) {
    // 如果更新了卡片信息
    controller.command(Commands.setNodeByKey, node.key, {
      type: 'link',
      data: {
        href: href,
        cardInfo: _extends({}, cardInfo, {
          displayType: displayType
        })
      }
    }).command(Commands.moveToEndOfNode, node);

    if (displayType === 'link') {
      // 样式切换为链接时，清理节点的 injection 内容
      controller.command(removeinjectedCardInfo, node);
    }
  }

  return controller.command(Commands.moveToStartOfNextText).command(Commands.focus);
}
