// @ali/4ever-plugin-find-and-replace/dist-node/common/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toggleFindAndReplace = toggleFindAndReplace;
exports.replaceText = replaceText;
exports.updateFindText = updateFindText;
exports.updateFindFragment = updateFindFragment;
exports.updateSearchResult = updateSearchResult;
exports.UPDATE_SEARCH_RESULT = exports.UPDATE_FIND_FRAGMENT = exports.UPDATE_FIND_TEXT = exports.REPLACE_TEXT_AT_DECORATION = exports.FIND_AND_REPLACE_TOGGLE = void 0;
const FIND_AND_REPLACE_TOGGLE = Symbol('FIND_AND_REPLACE_TOGGLE');
exports.FIND_AND_REPLACE_TOGGLE = FIND_AND_REPLACE_TOGGLE;
const REPLACE_TEXT_AT_DECORATION = Symbol('REPLACE_TEXT_AT_DECORATION');
exports.REPLACE_TEXT_AT_DECORATION = REPLACE_TEXT_AT_DECORATION;
const UPDATE_FIND_TEXT = Symbol('UPDATE_FIND_TEXT');
exports.UPDATE_FIND_TEXT = UPDATE_FIND_TEXT;
const UPDATE_FIND_FRAGMENT = Symbol('UPDATE_FIND_FRAGMENT');
exports.UPDATE_FIND_FRAGMENT = UPDATE_FIND_FRAGMENT;
const UPDATE_SEARCH_RESULT = Symbol('UPDATE_SEARCH_RESULT');
exports.UPDATE_SEARCH_RESULT = UPDATE_SEARCH_RESULT;

function toggleFindAndReplace(visible) {
  return {
    type: FIND_AND_REPLACE_TOGGLE,
    payload: {
      visible
    }
  };
}

function replaceText(focus, text) {
  return {
    type: REPLACE_TEXT_AT_DECORATION,
    payload: {
      focus,
      text
    }
  };
}

function updateFindText(findText) {
  return {
    type: UPDATE_FIND_TEXT,
    payload: {
      findText
    }
  };
}

function updateFindFragment(findFragment) {
  return {
    type: UPDATE_FIND_FRAGMENT,
    payload: {
      findFragment
    }
  };
}

function updateSearchResult(focus) {
  return {
    type: UPDATE_SEARCH_RESULT,
    payload: {
      focus
    }
  };
}
