// @ali/4ever-plugin-find-and-replace/dist-node/common/createContextMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = crateContextMenu;

var _actions = require("./actions");

function crateContextMenu(config, ctlManager) {
  const {
    locale
  } = config || {};
  return function (controller, next) {
    const menus = [];
    const {
      selection
    } = controller.value;
    const tableSelection = controller.query('tableSelection');

    const handleClick = () => {
      const {
        mainController
      } = ctlManager;

      if (!mainController) {
        return;
      }

      mainController.run('onAction', (0, _actions.toggleFindAndReplace)());
      let fragment = '';
      ctlManager.run(ctl => {
        const text = ctl.value.fragment?.text;

        if (text) {
          fragment = text;
        }
      });

      if (fragment) {
        mainController.run('onAction', (0, _actions.updateFindFragment)(fragment));
      }
    };

    if (selection && selection.isExpanded && !tableSelection) {
      menus.push({
        name: locale?.buttonTitle,
        onClick: handleClick,
        key: 'find',
        options: {
          hotKey: 'mod+f',
          group: 6
        }
      });
    }

    const others = next() || [];
    return [...menus, ...others];
  };
}
