// @ali/4ever-plugin-find-and-replace/dist-node/common/createRenderToolbar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderToolbar;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everComponent = require("@ali/4ever-component");

var _FindAndReplaceButton = require("../components/next/FindAndReplaceButton");

const _createElement = /*#__PURE__*/React.createElement;
const buttons = {
  findAndReplace: _FindAndReplaceButton.FindAndReplaceButton
};

function createRenderToolbar(locale, ctlManager) {
  return (props, controller) => {
    const {
      buttonName,
      extraProps
    } = props;
    const Component = buttons[buttonName];

    if (Component) {
      return /*#__PURE__*/_createElement(_everComponent.ItemSlot, {
        display: _everComponent.SlotType.iconButton
      }, /*#__PURE__*/_createElement(Component, (0, _extends2.default)({
        controller: controller,
        locale: locale,
        ctlManager: ctlManager
      }, extraProps)));
    }

    return null;
  };
}
