// @ali/4ever-plugin-color/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.moFactory;
  }
});
Object.defineProperty(exports, "biFactory", {
  enumerable: true,
  get: function () {
    return _index2.biFactory;
  }
});
Object.defineProperty(exports, "mobileBiFactory", {
  enumerable: true,
  get: function () {
    return _mobile.mobileBiFactory;
  }
});
Object.defineProperty(exports, "Color", {
  enumerable: true,
  get: function () {
    return _models.default;
  }
});
Object.defineProperty(exports, "ELEMENT_TYPE", {
  enumerable: true,
  get: function () {
    return _meta.ELEMENT_TYPE;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index3.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index3.createServerSerializationPlugin;
  }
});
Object.defineProperty(exports, "oldQueries", {
  enumerable: true,
  get: function () {
    return _queries.queries;
  }
});
Object.defineProperty(exports, "oldCommands", {
  enumerable: true,
  get: function () {
    return _commands.commands;
  }
});
Object.defineProperty(exports, "oldActions", {
  enumerable: true,
  get: function () {
    return _actions.actions;
  }
});
Object.defineProperty(exports, "COLOR_SET", {
  enumerable: true,
  get: function () {
    return _actions.COLOR_SET;
  }
});
Object.defineProperty(exports, "ColorRole", {
  enumerable: true,
  get: function () {
    return _role.ColorRole;
  }
});
exports.OldBambooPlugin = void 0;

var _index = require("./mo/index");

var _index2 = require("./bi/index");

var _mobile = require("./bi/mobile");

require("./utils/types");

var _models = _interopRequireDefault(require("./mo/models"));

var _meta = require("./utils/meta");

var _index3 = _interopRequireWildcard(require("./mo/serializer/index"));

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;

var _queries = require("./bi/queries");

var _commands = require("./bi/commands");

var _actions = require("./bi/actions");

var _role = require("./utils/role");
