// @ali/4ever-plugin-template-button/dist-web/components/container.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import CloseView from "./closeView";
import OpenView from "./openView";

var Container = function Container(props) {
  var _config$locale;

  var node = props.node,
      controller = props.controller,
      child = props.child,
      config = props.config,
      isSelected = props.isSelected,
      _props$editable = props.editable,
      editable = _props$editable === void 0 ? true : _props$editable;
  var metadata = node.data.metadata;
  var placeholder = (config == null ? void 0 : (_config$locale = config.locale) == null ? void 0 : _config$locale.namePlaceholder) || 'Add a new item';

  if (metadata.isOpen) {
    return /*#__PURE__*/_createElement(OpenView, _extends({}, metadata, {
      placeholder: placeholder,
      controller: controller,
      nodeKey: node.key,
      child: child,
      config: config,
      editable: editable,
      isSelected: isSelected
    }));
  } else {
    return /*#__PURE__*/_createElement(CloseView, _extends({}, metadata, {
      placeholder: placeholder,
      controller: controller,
      node: node,
      isSelected: isSelected,
      editable: editable,
      config: config
    }));
  }
};

export { Container };
