// @ali/4ever-bi/dist-node/plugins/sz/mobile/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _setSize = _interopRequireDefault(require("../commands/setSize"));

var _actions = require("../actions");

var _onAction = _interopRequireDefault(require("../handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("../handlers/onKeyDown"));

var _queries = require("../queries");

const commands = {
  setSize: _setSize.default
};

function createSzPlugin(config) {
  const szConfig = config.sz || {};
  return {
    toolbar: (0, _toolbar.default)(szConfig),
    commands,
    queries: _queries.queries,
    actions: _actions.actions,
    isActions: _actions.isActions,
    onAction: _onAction.default,
    onKeyDown: _onKeyDown.default,
    models: {
      klass: 'mark',
      type: 'sz'
    }
  };
}

createSzPlugin.actions = _actions.DEPRECATED_ACTIONS;
createSzPlugin.commands = commands;
createSzPlugin.queries = _queries.queries;
var _default = createSzPlugin;
exports.default = _default;
