// @ali/we-design-next/dist-node/basic/list/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListWrap = exports.ListItemWrap = exports.ListItemSuffix = exports.ListItemPrefix = exports.ListItemGroupTitle = exports.ListItemContent = exports.ListDivider = void 0;
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _common = require("../../common");
var _text = require("../text");
const ListWrap = /*#__PURE__*/_styledComponents.default.div(["", " outline:none;min-width:150px;overflow-x:hidden;overflow-y:overlay;"], _common.CSS_SCROLLBAR);
exports.ListWrap = ListWrap;
const ListDivider = /*#__PURE__*/_styledComponents.default.div(["height:1px;margin:4px;background-color:", ";"], _common.COLORS.OTHERS.DIVIDER);
exports.ListDivider = ListDivider;
const ListItemWrap = /*#__PURE__*/_styledComponents.default.div(["", " ", " display:flex;align-items:center;user-select:none;min-height:34px;padding:0 8px;margin:4px;border-radius:4px;", " color:", ";"], _text.TEXT_CSS.TEXT, _common.CSS_CURSOR_DISABLE, ({
  isActive
}) => isActive ? `background: ${_common.COLORS.OTHERS.DAO_HOVER_BG_COLOR};` : null, ({
  disabled
}) => disabled ? _common.COLORS.LIGHT.LEVEL4_BASE_COLOR : 'unset');
exports.ListItemWrap = ListItemWrap;
const ListItemPrefix = /*#__PURE__*/_styledComponents.default.div(["flex-grow:0;margin-right:4px;display:flex;align-items:center;min-width:20px;"]);
exports.ListItemPrefix = ListItemPrefix;
const ListItemSuffix = /*#__PURE__*/_styledComponents.default.div(["flex-grow:0;margin-left:16px;display:flex;align-items:center;min-width:20px;"]);
exports.ListItemSuffix = ListItemSuffix;
const ListItemContent = /*#__PURE__*/_styledComponents.default.div(["flex-grow:1;word-break:keep-all;"]);
exports.ListItemContent = ListItemContent;
const ListItemGroupTitle = /*#__PURE__*/_styledComponents.default.div(["", " color:", ";line-height:36px;height:36px;padding-left:12px;user-select:none;"], _text.TEXT_CSS.TEXT_SUB, _common.COLORS.LIGHT.LEVEL2_BASE_COLOR);
exports.ListItemGroupTitle = ListItemGroupTitle;
