// @ali/4ever-plugin-calendar-card/dist-node/utils/components/subjectParagraph/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SubjectParagraph = void 0;

var _react = _interopRequireDefault(require("react"));

var _lodash = require("lodash");

var _weDesign = require("@ali/we-design");

var _paragraphFactory = _interopRequireDefault(require("../paragraphFactory"));

var _constants = require("../../constants");

var _preventDefaultOnClick = _interopRequireDefault(require("../../uitls/preventDefaultOnClick"));

var _styled = require("./styled");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement(_weDesign.CalendarBetaNormalNormal, {
  style: _constants.DEFAULT_ICON_STYLE
});

var _ref2 = /*#__PURE__*/_createElement(_weDesign.ArrowRightNormal, null);

const SubjectParagraph = props => {
  const {
    innerWidth,
    subject,
    jumpToCalendar,
    isReadonly,
    isPlaceholder
  } = props;
  const Paragraph = (0, _paragraphFactory.default)(_ref);
  const realWidth = (0, _lodash.isNumber)(innerWidth) ? innerWidth - _constants.CALENDAR_CARD_ICON_WIDTH - _constants.CALENDAR_CARD_EXTRA_ICON_WIDTH * 2 : innerWidth;
  return /*#__PURE__*/_createElement(Paragraph, null, /*#__PURE__*/_createElement(_styled.ContainerWrapper, {
    width: realWidth,
    isReadonly: isPlaceholder || isReadonly
  }, /*#__PURE__*/_createElement(_styled.TextWrapper, {
    onMouseDown: jumpToCalendar,
    onClick: _preventDefaultOnClick.default,
    "data-testid": "calendar-subject"
  }, subject), /*#__PURE__*/_createElement(_styled.IconWrapper, null, isReadonly || isPlaceholder ? null : _ref2)));
};

exports.SubjectParagraph = SubjectParagraph;
