// @ali/4ever-zhi/dist-node/__tests__/reader.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Reader;
exports.MobileReader = MobileReader;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everCangjie = require("@ali/4ever-cangjie");

var _everMo = require("@ali/4ever-mo");

var _ = require("..");

const _createElement = /*#__PURE__*/React.createElement;
const mo = (0, _everMo.Serializer)({
  plugins: (0, _everMo.createMoPlugins)({})
});

var _ref = /*#__PURE__*/_createElement("div", null, "zhi\u4E2D\u7684\u65E5\u7A0B\u5361\u7247-demo");

var _ref2 = /*#__PURE__*/_createElement("div", null, "zhi\u4E2D\u7684URL\u5361\u7247-demo");

var _ref3 = /*#__PURE__*/_createElement("div", null, "tag-demo\u6807\u7B7E");

var _ref4 = /*#__PURE__*/_createElement("div", null, "zhi\u4E2D\u7684\u9879\u76EE\u770B\u677F\u5361\u7247");

function Reader({
  previewImage,
  transformImageURL,
  openLink,
  onMentionClick,
  onMentionHover,
  voteCardConfig,
  content,
  enableCard,
  config = {},
  ...rest
}) {
  const demoCardPlugin = (0, _.cardFactory)(() => {
    return {
      cardType: 'demo',
      renderNode: () => {
        return _ref;
      }
    };
  });
  const demoURLCardPlugin = (0, _.cardFactory)(() => {
    return {
      cardType: 'url',
      renderNode: () => {
        return _ref2;
      }
    };
  });
  const demoTagPlugin = (0, _.tagFactory)(() => {
    return {
      tagType: 'demo',
      verticalAlign: 'middle',
      renderNode: () => {
        return _ref3;
      }
    };
  });
  const demoProjectCardPlugin = (0, _.cardFactory)(() => {
    return {
      cardType: 'projectCard',
      renderNode: () => {
        return _ref4;
      }
    };
  });

  const ErrorComponent = () => {
    throw new Error('error ErrorComponent');
  };

  var _ref5 = /*#__PURE__*/_createElement(ErrorComponent, null);

  const errorCardWithChildren = (0, _.cardFactory)(() => {
    return {
      cardType: 'errorChildren',
      renderNode: () => {
        return _ref5;
      }
    };
  });
  const errorCard = (0, _.cardFactory)(() => {
    return {
      cardType: 'error',
      renderNode: () => {
        throw new Error('error errorCard');
      }
    };
  });
  const errorTag = (0, _.tagFactory)(() => {
    return {
      tagType: 'error',
      renderNode: () => {
        throw new Error('error errorTag');
      }
    };
  });

  var _ref6 = /*#__PURE__*/_createElement(ErrorComponent, null);

  const errorTagWithChildren = (0, _.tagFactory)(() => {
    return {
      tagType: 'errorChildren',
      renderNode: () => {
        return _ref6;
      }
    };
  });
  const createZhiPlugins = (0, _.createCustomizedZhiPlugins)({ ..._.plugins,
    demoCardPlugin,
    demoProjectCardPlugin,
    demoTagPlugin,
    demoURLCardPlugin,
    GroupTagPlugin: _.GroupTagPlugin,
    ListGroupPlugin: _.ListGroupPlugin,
    errorCard,
    errorCardWithChildren,
    errorTag,
    errorTagWithChildren,
    CalendarCardPlugin: _.CalendarCardPlugin,
    voteCard: _.VoteCardPlugin,
    ImTagPlugin: _.ImTagPlugin,
    FormulaPlugin: _.FormulaPlugin
  });
  const plugins = createZhiPlugins({
    image: {
      previewImage,
      transformImageURL
    },
    link: {
      openLink
    },
    mention: {
      onMentionClick,
      onMentionHover
    },
    voteCard: voteCardConfig,
    table: { ...(config.table || {})
    }
  }); // eslint-disable-next-line

  const slateValue = _everCangjie.Value.isValue(content) ? content : mo.jsonMLToValue(content, {
    normalize: false
  });
  return /*#__PURE__*/_createElement(_.Reader, (0, _extends2.default)({
    content: slateValue,
    plugins: plugins
  }, rest));
}

function MobileReader({
  openLink,
  enableCard,
  content,
  ...rest
}) {
  const createZhiPlugins = (0, _.createMobileCustomizedZhiPlugins)({ ..._.mobilePlugins
  });
  const plugins = createZhiPlugins({
    link: {
      openLink
    }
  }); // eslint-disable-next-line

  const slateValue = _everCangjie.Value.isValue(content) ? content : mo.jsonMLToValue(content, {
    normalize: false
  });
  return /*#__PURE__*/_createElement(_.Reader, (0, _extends2.default)({
    content: slateValue,
    plugins: plugins
  }, rest));
}
