// @ali/dingtalk-kit/lib/apis/weex/platform/getEnv.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var shared_1 = require("../../../constants/platform/shared");
var env;
function getDingTalkVersion() {
    var _ddVersion = weex.config.env.appVersion;
    var version = shared_1.DEFAULT_VERSION;
    if (typeof _ddVersion === 'string') {
        version = _ddVersion || shared_1.DEFAULT_VERSION;
    }
    else {
        version = _ddVersion && _ddVersion[1] || shared_1.DEFAULT_VERSION;
    }
    return version;
}
function getEnv() {
    if (!env) {
        var hasWeex = typeof weex === 'object';
        var appName = hasWeex ? (weex.config.env.appName || '') : '';
        var isDingTalk = appName.toLowerCase().indexOf('dingtalk') > -1;
        var isMac = false;
        var isWin = false;
        var platform = hasWeex ? (weex.config.env.platform || '').toLowerCase() : '';
        var isAndroid = platform === 'android';
        var isIOS = platform === 'ios';
        var dingTalkVersion = hasWeex ? getDingTalkVersion() : shared_1.DEFAULT_VERSION;
        env = {
            isDingTalk: isDingTalk,
            isMac: isMac,
            isWin: isWin,
            dingTalkVersion: dingTalkVersion,
            isAndroid: isAndroid,
            isIOS: isIOS
        };
    }
    return env;
}
exports.getEnv = getEnv;
