// @ali/4ever-component/dist-web/MobileToolbar/buttons/styled.js
/* eslint-disable @typescript-eslint/consistent-type-definitions */

/* eslint-disable no-confusing-arrow */
import styled from 'styled-components';
import { BOTTOM_BAR_SIZE, ICON_BUTOTN_WIDTH } from "../constant";
export var NewLine = /*#__PURE__*/styled.div(["height:0;width:100%;"]);
export var Divider = /*#__PURE__*/styled.div(["width:12px;"]);
export var AlignIcon = /*#__PURE__*/styled.div(["flex:1;height:0;min-width:", "px;"], ICON_BUTOTN_WIDTH);
export var GroupTitle = /*#__PURE__*/styled.div(["width:100%;margin-bottom:8px;line-height:18px;font-size:12px;color:rgba(23,26,29,0.4);"]);
// 底部一级大按钮
var BoxBig = /*#__PURE__*/styled.div(["width:32px;height:32px;border-radius:4px;display:flex;align-items:center;position:relative;justify-content:center;opacity:", ";& > span{font-size:24px;}& path{transition:fill 0.2s;}&::after{content:'';position:absolute;left:4px;bottom:-3px;width:24px;height:3px;background:transparent;transition:background 0.2s;border-radius:1.5px;}", ""], function (props) {
  return props.isDisabled ? '0.24' : '1';
}, function (props) {
  return props.isActive && !props.isFocused ? "\n  & path {\n    fill: #0089FF !important;\n  }\n  &::after {\n    content: '';\n    position: absolute;\n    left: 4px;\n    bottom: -3px;\n    width: 24px;\n    height: 3px;\n    background: #0089FF;\n    transition: background 0.2s;\n    border-radius: 1.5px;\n  }\n  " : '';
}); // Panel 的上下图标结构。

var BoxIcon = /*#__PURE__*/styled.div(["display:flex;flex-direction:column;justify-content:center;align-items:center;width:", "px;opacity:", ";& > div:first-of-type{width:44px;height:44px;display:flex;justify-content:center;background:rgba(126,134,142,0.12);align-items:center;border-radius:6px;& > span{font-size:24px;}}& > div:last-of-type{line-height:20px;margin-top:4px;color:#171a1d;font-size:14px;}"], ICON_BUTOTN_WIDTH, function (props) {
  return props.isDisabled ? '0.5' : '1';
});
var BoxLine = /*#__PURE__*/styled.div(["height:100%;display:flex;flex-direction:row;justify-content:flex-start;align-items:center;font-size:17px;color:#171a1d;position:relative;&::after{content:'';position:absolute;right:0;bottom:0;width:calc(100% - 40px);height:1px;background:rgba(126,134,142,0.16);}& >:first-child{width:24px;height:24px;font-size:0;display:flex;justify-content:center;align-items:center;margin-right:16px;& > span{font-size:24px;}}"]); // Panel 的按钮。加粗、斜体之类

var BoxHoriz = /*#__PURE__*/styled.div(["width:100%;height:100%;transition:background 0.2s;border-radius:", ";display:flex;flex-direction:row;justify-content:center;align-items:center;background:", ";opacity:", ";& > span{font-size:20px;}"], function (props) {
  if (props.isStart && props.isEnd) {
    return '4px';
  }

  if (props.isStart) {
    return '4px 0 0 4px';
  }

  if (props.isEnd) {
    return '0 4px 4px 0';
  }

  return '0';
}, function (props) {
  return props.isActive ? 'rgba(0, 137, 255, 0.12)' : 'rgba(126, 134, 142, 0.12)';
}, function (props) {
  return props.isDisabled ? '0.5' : '1';
});
var OuterBig = /*#__PURE__*/styled.div(["cursor:pointer;height:", "px;width:", "px;min-width:", "px;display:flex;align-items:center;justify-content:center;", ""], BOTTOM_BAR_SIZE, BOTTOM_BAR_SIZE, BOTTOM_BAR_SIZE, function (props) {
  return props.isTouched || props.isFocused ? "\n  & " + BoxBig + " {\n   background: rgba(126, 134, 142, 0.16);\n  }\n  " : '';
});
var OuterIcon = /*#__PURE__*/styled.div(["flex:", ";cursor:pointer;margin-right:1px;margin-bottom:12px;display:flex;align-items:center;justify-content:center;", ""], function (props) {
  return props.flexWidth;
}, function (props) {
  return props.isTouched ? "\n  & " + BoxHoriz + " {\n   background: rgba(126, 134, 142, 0.24);\n  }\n  & " + BoxIcon + " > div:first-of-type {\n    background: rgba(126, 134, 142, 0.24);\n  }\n  " : '';
});
var OuterLine = /*#__PURE__*/styled.div(["cursor:pointer;width:100%;height:56px;padding-left:16px;transition:background 0.2s;", ""], function (props) {
  return props.isTouched ? "\n    background: rgba(126, 134, 142, 0.12);\n" : '';
});
var OuterHoriz = /*#__PURE__*/styled(OuterIcon)(["height:44px;"]);
export var Outers = {
  big: OuterBig,
  horiz: OuterHoriz,
  line: OuterLine,
  icon: OuterIcon
};
export var Boxes = {
  big: BoxBig,
  horiz: BoxHoriz,
  line: BoxLine,
  icon: BoxIcon
};
export var ImageIcon = /*#__PURE__*/styled.img(["width:24px;height:24px;"]);
