// @ali/4ever-bi/dist-node/plugins/sticker/components/stickerMenu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inlineMenu = exports.menu = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../actions");

var _constants = require("../constants");

const _createElement = /*#__PURE__*/_react.default.createElement;
;

function WithStickerMenu(Component) {
  return function StickerMenu(props) {
    const {
      menuItem,
      controller
    } = props;
    const {
      value
    } = controller;
    const {
      title,
      ...rest
    } = menuItem;
    const {
      isCollapsedAtEdge
    } = controller.value.selection;
    const isDisabled = controller.query('isDisableSticker') || isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInLink') || controller.query('isSelectionInMention') || controller.query('isSelectionInLinkCard');
    const handleClick = (0, _react.useCallback)(() => {
      controller.run('onAction', (0, _actions.insertSticker)(_constants.STICKER_PLACEHOLDER_DATA));
      controller.run('onAction', (0, _actions.setStickerPanelVisible)({
        visible: true
      }));
    }, [controller]);
    return /*#__PURE__*/_createElement(Component, (0, _extends2.default)({}, rest, {
      disabled: isDisabled,
      onClick: handleClick,
      role: _everBamboo.PluginRoles.sticker
    }), title || '');
  };
}

const menu = WithStickerMenu(_weDesign.MenuItem);
exports.menu = menu;
const inlineMenu = WithStickerMenu(_weDesign.InlineMenuItem);
exports.inlineMenu = inlineMenu;
