// @ali/4ever-bi/dist-node/plugins/sticker/components/stickerButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = StickerButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _weDesign = require("@ali/we-design");

var _everComponent = require("@ali/4ever-component");

var _actions = require("../actions");

var _constants = require("../constants");

const _createElement = /*#__PURE__*/_react.default.createElement;

function StickerButton(props) {
  const {
    controller,
    locale,
    toolbarMode,
    ...rest
  } = props;
  const {
    value
  } = controller;
  const isDisabled = controller.query('isDisableSticker') || controller.query('isSelectionInImage') || controller.query('isSelectionInLinkCard');
  const handleClick = (0, _react.useCallback)(() => {
    controller.run('onAction', (0, _actions.insertSticker)(_constants.STICKER_PLACEHOLDER_DATA));
    controller.run('onAction', (0, _actions.setStickerPanelVisible)({
      visible: true
    }));
  }, []);
  const viewSize = toolbarMode === _everComponent.ToolbarMode.single ? ' 20' : ' 24'; // @ts-ignore 组件没有 viewBox 属性

  const icon = /*#__PURE__*/_createElement(_weDesign.ToolbarEmojiNormal, {
    viewBox: `0 0${viewSize.repeat(2)}`
  });

  return /*#__PURE__*/_createElement(_weToolbar.EmojiMenuButton, (0, _extends2.default)({}, rest, {
    testid: "bi-toolbar-sticker" // @ts-ignore 组件没有 icon 属性
    ,
    icon: icon,
    tooltip: locale.buttonTitle,
    onClick: handleClick,
    disabled: isDisabled,
    autoClose: false,
    showArrow: false
  }));
}
