// @ali/4ever-bi/dist-node/plugins/sticker/components/stickerWrapper.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _utils = require("../utils");

const _createElement = /*#__PURE__*/_react.default.createElement;
const {
  DEFAULT_SZ,
  StickerItem,
  getStickerStyle
} = _everBamboo.StickerPlugin;

const EmptyFunc = () => null;

function StickerWrapper(props) {
  const {
    controller,
    node,
    sticker,
    StickerPlaceholder = EmptyFunc
  } = props;
  const {
    code: stickerCode
  } = node.data;
  const {
    sz = DEFAULT_SZ,
    highlight
  } = (0, _react.useMemo)(() => getStickerStyle(node), [node]);
  const onClick = (0, _react.useCallback)(e => {
    e.preventDefault();
    e.stopPropagation();
    const {
      offsetX
    } = e.nativeEvent;
    const {
      offsetLeft,
      offsetWidth
    } = e.target;
    const isRightSide = offsetX > 0 && offsetX - offsetLeft > offsetWidth / 2;
    (0, _utils.moveCaretToText)({
      controller,
      offset: isRightSide ? 1 : -1
    });
  }, [controller.value.selection]);
  return /*#__PURE__*/_createElement(StickerItem, {
    key: node.key,
    code: stickerCode,
    backgroundColor: highlight,
    placeholder: /*#__PURE__*/_createElement(StickerPlaceholder, props),
    sz: _everBamboo.unitsConverter.toPX(`${sz * 1.5}pt`),
    sticker: sticker,
    onClick: onClick
  });
}

var _default = /*#__PURE__*/_react.default.memo(StickerWrapper);

exports.default = _default;
