// @ali/4ever-bi/dist-node/plugins/sticker/components/styled.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "StickerPannelWrapper", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.StickerPannelWrapper;
  }
});
Object.defineProperty(exports, "StickerSplitLine", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.StickerSplitLine;
  }
});
Object.defineProperty(exports, "StickerPanelItem", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.StickerPanelItem;
  }
});
Object.defineProperty(exports, "InlinePlaceholderWrapper", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.InlinePlaceholderWrapper;
  }
});
Object.defineProperty(exports, "InlinePlaceholderContainer", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.InlinePlaceholderContainer;
  }
});
Object.defineProperty(exports, "PanelWrapper", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.PanelWrapper;
  }
});

var _everPluginSticker = require("@ali/4ever-plugin-sticker");
