// @ali/4ever-plugin-sticker/dist-web/utils/sticker.js
import { flatten } from 'lodash-es';
import { DEFAULT_LANG, SINGLE, SPRITE } from "./constants";

var Sticker = /*#__PURE__*/function () {
  // 表情映射表
  // 表情语言
  // 表情列表
  function Sticker(_ref) {
    var customSticker = _ref.customSticker,
        _ref$lang = _ref.lang,
        lang = _ref$lang === void 0 ? DEFAULT_LANG : _ref$lang;
    this.stickerMap = void 0;
    this.stickerLang = void 0;
    this.stickerList = void 0;
    this.stickerLang = lang;
    this.stickerList = this.getStickerList(customSticker);
    this.stickerMap = new Map(this.stickerList.map(function (s) {
      return [s.code, s];
    }));
  }

  var _proto = Sticker.prototype;

  _proto.getStickerList = function getStickerList(customSticker) {
    var _this = this;

    return flatten(customSticker.map(function (cusStk) {
      return _this.parseStickers(cusStk);
    }));
  };

  _proto.parseStickers = function parseStickers(customSticker) {
    if ((customSticker == null ? void 0 : customSticker.type) === SPRITE) {
      return customSticker.stickers.map(function (stk) {
        stk.src = customSticker.src;
        stk.spriteWidth = customSticker.width;
        stk.spriteHeight = customSticker.height;
        return stk;
      });
    } // TODO 目前仅支持雪碧图，单个表情或动图待支持 @灯少


    if ((customSticker == null ? void 0 : customSticker.type) === SINGLE) {
      return customSticker.stickers.map(function (stk) {
        stk.x = 0;
        stk.y = 0;
        stk.spriteWidth = stk.width;
        stk.spriteHeight = stk.height;
        return stk;
      });
    }

    return [];
  };

  _proto.getInfoByCode = function getInfoByCode(code) {
    return this.stickerMap.get(code);
  };

  _proto.getStickerSource = function getStickerSource(code) {
    var stickerInfo = this.getInfoByCode(code);
    return stickerInfo == null ? void 0 : stickerInfo.src;
  };

  _proto.getStickerLocale = function getStickerLocale(code, lang) {
    var stickerInfo = this.getInfoByCode(code);
    return (stickerInfo == null ? void 0 : stickerInfo.locale[lang || this.stickerLang]) || '';
  };

  return Sticker;
}();

export default Sticker;
