// @ali/4ever-plugin-list/dist-web/utils/transformNumber/index.js
import arabic2Chinese, { CHINESE_NUMBER_LIST, CHINESE_UNIT_LIST } from "./arabic2Chinese";
import arabic2Letter, { LETTER_LIST } from "./arabic2Letter";
import arabic2Roma, { ROMA_LIST } from "./arabic2Roma";
import arabic2Darabic from "./arabic2Darabic";
import chinese2Arabic from "./chinese2Arabic";
import roman2Arabic from "./roma2Arabic";
import letter2Arabic from "./letter2Arabic";

var transformNumber = function transformNumber(num, type) {
  var result = '';

  switch (type) {
    case 'decimalZero':
      result = arabic2Darabic(num);
      break;

    case 'chineseCountingThousand':
    case 'chineseCounting':
    case 'japaneseCounting':
      result = arabic2Chinese(num);
      break;

    case 'upperLetter':
    case 'lowerLetter':
      result = arabic2Letter(num, type.slice(0, 5));
      break;

    case 'upperRoman':
    case 'lowerRoman':
      result = arabic2Roma(num, type.slice(0, 5));
      break;

    default:
      result = String(num);
      break;
  }

  return result;
};

var transformString = function transformString(str, type) {
  var result = 0;

  switch (type) {
    case 'chinese':
      result = chinese2Arabic(str);
      break;

    case 'roman':
      result = roman2Arabic(str);
      break;

    case 'letter':
      result = letter2Arabic(str);
      break;

    default:
      return parseInt(str, 10);
  }

  return result;
};

export { arabic2Chinese, arabic2Darabic, arabic2Letter, arabic2Roma, transformNumber, transformString, CHINESE_NUMBER_LIST, CHINESE_UNIT_LIST, LETTER_LIST, ROMA_LIST };
