// @ali/4ever-plugin-color-blocks/dist-web/mo/serializer/jsonMLToHTML.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { getTagName, getAttributes, getChildren } from '@ali/4ever-utils';
import { ELEMENT_TYPE } from "../../utils/meta";

function getStyle(metadata) {
  var style = {
    padding: '10px 14px'
  };

  var _ref = metadata || {},
      _ref$bgcolor = _ref.bgcolor,
      bgcolor = _ref$bgcolor === void 0 ? '#FEF8E8' : _ref$bgcolor,
      border = _ref.border;

  if (bgcolor) {
    style.backgroundColor = bgcolor;
  }

  if (border) {
    style.border = "1px solid " + border;
    style.borderRadius = '8px';
  }

  return style;
}

export default {
  name: 'container',
  match: function match(node) {
    var tag = getTagName(node);
    var attrs = getAttributes(node);
    return tag === 'container' && (attrs == null ? void 0 : attrs.subType) === ELEMENT_TYPE;
  },
  convert: function convert(node, path, _convert) {
    var attrs = getAttributes(node) || {};
    var children = getChildren(node).map(function (child, index) {
      return _convert(child, [].concat(path, [index]));
    });
    var style = getStyle(attrs == null ? void 0 : attrs.metadata);
    return /*#__PURE__*/_createElement("div", {
      "data-type": "container",
      style: style
    }, children);
  }
};
