// @ali/4ever-bi/dist-node/plugins/quickPreview/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _onCangjieFocus = _interopRequireDefault(require("./hooks/onCangjieFocus"));

var _createRenderContent = _interopRequireDefault(require("./hooks/createRenderContent"));

var _onAction = _interopRequireDefault(require("./hooks/onAction"));

var _actions = _interopRequireWildcard(require("./actions"));

var _isQuickPreview = _interopRequireDefault(require("./queries/isQuickPreview"));

var _shouldRenderH5Controls = _interopRequireDefault(require("./queries/shouldRenderH5Controls"));

var _selectionColor = _interopRequireDefault(require("./queries/selectionColor"));

function QuickPreviewPlugin(configs) {
  return {
    queries: {
      isQuickPreview: _isQuickPreview.default,
      selectionColor: _selectionColor.default,
      shouldRenderH5Controls: _shouldRenderH5Controls.default
    },
    onAction: (0, _onAction.default)(configs?.quickPreview),
    onCangjieFocus: _onCangjieFocus.default,
    renderContent: (0, _createRenderContent.default)(configs?.quickPreview),
    actions: {
      enterPreview: _actions.enterPreview,
      exitPreview: _actions.exitPreview
    }
  };
}

QuickPreviewPlugin.Actions = _actions.default;
var _default = QuickPreviewPlugin;
exports.default = _default;
