// @ali/4ever-plugin-mention/dist-web/bi/quickInsert/createQuickInsertConfig.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { MentionBetaNormal } from '@ali/we-design';
import { insertQuickInsert } from '@ali/4ever-plugin-quickinsert';
import MentionSuggestion from "../components/suggestionForQuickInsert";

var _ref = /*#__PURE__*/_createElement(MentionBetaNormal, null);

export default function createQuickInsertConfig( // @ts-ignore
mentionConfig) {
  var quickInsert = {
    onMenuClick: function onMenuClick(controller) {
      return controller.run('onAction', insertQuickInsert('quickInsertMention'));
    },
    renderIcon: function renderIcon() {
      return _ref;
    },
    renderPanel: function renderPanel(props) {
      var text = props.text,
          controller = props.controller,
          rest = _objectWithoutPropertiesLoose(props, ["text", "controller"]);

      return /*#__PURE__*/_createElement(MentionSuggestion, _extends({}, rest, mentionConfig, {
        controller: controller,
        text: text.slice(1)
      }));
    }
  };
  return quickInsert;
}
