// @ali/we-ui/dist-web/components/dropdown/Dropdown.js
import React, { useMemo } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Trigger } from "../trigger";
export var Dropdown = function Dropdown(props) {
  var children = props.children,
      trigger = props.trigger,
      overlay = props.overlay,
      getContainer = props.getContainer,
      onVisibleChange = props.onVisibleChange,
      placement = props.placement,
      overlayStyle = props.overlayStyle,
      visible = props.visible,
      zIndex = props.zIndex,
      overlayClassName = props.overlayClassName,
      offsetX = props.offsetX,
      offsetY = props.offsetY,
      pageX = props.pageX,
      pageY = props.pageY,
      triggerNode = props.triggerNode,
      title = props.title,
      testid = props.testid; // trigger's action

  var action = useMemo(function () {
    return trigger === 'contextMenu' ? [trigger] : [trigger, 'focus'];
  }, [trigger]);
  return /*#__PURE__*/_createElement(Trigger, {
    action: action,
    overlay: overlay,
    getContainer: getContainer,
    onVisibleChange: onVisibleChange,
    placement: placement,
    visible: visible,
    overlayStyle: overlayStyle,
    zIndex: zIndex,
    overlayClassName: overlayClassName,
    offsetX: offsetX,
    offsetY: offsetY,
    pageX: pageX,
    pageY: pageY,
    triggerNode: triggerNode,
    title: title,
    testid: testid
  }, children);
};
Dropdown.defaultProps = {
  trigger: 'click'
};
