// @ali/we-design-next/dist-web/basic/button/Button.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
var _excluded = ["children", "size", "type", "icon", "disabled"];
import React, { memo, forwardRef } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { mergeRestProps } from "../../common";
import { ButtonWrap } from "./styled";
export var Button = /*#__PURE__*/memo( /*#__PURE__*/forwardRef(function (props, ref) {
  var children = props.children,
    size = props.size,
    _props$type = props.type,
    type = _props$type === void 0 ? 'normal' : _props$type,
    icon = props.icon,
    disabled = props.disabled,
    rest = _objectWithoutPropertiesLoose(props, _excluded);
  return /*#__PURE__*/_createElement(ButtonWrap, _extends({
    ref: ref,
    size: size,
    disabled: disabled,
    btnType: type,
    content: icon && children ? 'icon-text' : children ? 'only-text' : 'only-icon'
  }, mergeRestProps(rest, 'wdn-button')), icon, children);
}));
Button.displayName = 'Button';
