// @ali/4ever-zhi/dist-node/plugins/voteCard/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.votingActive = votingActive;
exports.voteSubmit = voteSubmit;
exports.voteEditing = voteEditing;
exports.cancelEditing = cancelEditing;
exports.voteSave = voteSave;
exports.VOTE_SAVE = exports.VOTE_CANCEL_EDITING = exports.VOTE_EDITING = exports.VOTE_SUBMIT = exports.VOTING_ACTIVE = void 0;
const VOTING_ACTIVE = Symbol('VOTE_ACTIVE');
exports.VOTING_ACTIVE = VOTING_ACTIVE;
const VOTE_SUBMIT = Symbol('VOTE_SUBMIT');
exports.VOTE_SUBMIT = VOTE_SUBMIT;
const VOTE_EDITING = Symbol('VOTE_EDITING');
exports.VOTE_EDITING = VOTE_EDITING;
const VOTE_CANCEL_EDITING = Symbol('VOTE_CANCEL_EDITING');
exports.VOTE_CANCEL_EDITING = VOTE_CANCEL_EDITING;
const VOTE_SAVE = Symbol('VOTE_SAVE');
exports.VOTE_SAVE = VOTE_SAVE;

function votingActive(node) {
  return {
    type: VOTING_ACTIVE,
    payload: {
      node
    }
  };
}

function voteSubmit(payload) {
  return {
    type: VOTE_SUBMIT,
    payload
  };
}

function voteEditing(node) {
  return {
    type: VOTE_EDITING,
    payload: {
      node
    }
  };
}

function cancelEditing(node) {
  return {
    type: VOTE_CANCEL_EDITING,
    payload: {
      node
    }
  };
}

function voteSave(form) {
  return {
    type: VOTE_SAVE,
    payload: {
      form
    }
  };
}
