// @ali/4ever-component/dist-node/HoverToolbar/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HoverToolbarHooks = exports.HoverToolbarContexts = void 0;

var _PopoverVisibleSetContext = require("./PopoverVisibleSetContext");

var _HoverToolbarVisibleContext = require("./HoverToolbarVisibleContext");

var _GetContainerContext = require("./GetContainerContext");

var _PopoverVisibleContext = require("./PopoverVisibleContext");

const HoverToolbarContexts = {
  VisiblePopoverSetContext: _PopoverVisibleSetContext.VisiblePopoverSetContext,
  HoverToolbarVisibleContext: _HoverToolbarVisibleContext.HoverToolbarVisibleContext,
  GetContainerContext: _GetContainerContext.GetContainerContext
};
exports.HoverToolbarContexts = HoverToolbarContexts;
const HoverToolbarHooks = {
  useIsPopoverVisible: _PopoverVisibleSetContext.useIsPopoverVisible,
  useIsHoverToolbarVisible: _HoverToolbarVisibleContext.useIsHoverToolbarVisible,
  useGetContainer: _GetContainerContext.useGetContainer,
  usePopoverVisible: _PopoverVisibleContext.usePopoverVisible
};
exports.HoverToolbarHooks = HoverToolbarHooks;
