// @ali/4ever-component/dist-node/MobileToolbar/constant.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.itemToKey = exports.SUBLINE_MARGIN_BOTTOM = exports.SUBLINE_MIN_LEFT = exports.THREED_TRAMSFORM = exports.ALIGN_PLACEHOLDER = exports.DIVIDER_VERT = exports.DIVIDER = exports.NEW_LINE = exports.DIVIDER_NONE = exports.ICON_BUTOTN_WIDTH = exports.PANEL_HEADER_SIZE = exports.Z_INDEX = exports.BOTOTM_PANEL_HEIGHT = exports.BOTTOM_BAR_SIZE = void 0;
const BOTTOM_BAR_SIZE = 44;
exports.BOTTOM_BAR_SIZE = BOTTOM_BAR_SIZE;
const BOTOTM_PANEL_HEIGHT = 400;
exports.BOTOTM_PANEL_HEIGHT = BOTOTM_PANEL_HEIGHT;
const Z_INDEX = 1900;
exports.Z_INDEX = Z_INDEX;
const PANEL_HEADER_SIZE = 56; // icon 类的 button 宽度

exports.PANEL_HEADER_SIZE = PANEL_HEADER_SIZE;
const ICON_BUTOTN_WIDTH = 68; // 无渲染内容的分割，用户区别两个连续的按钮，做 round border 的效果

exports.ICON_BUTOTN_WIDTH = ICON_BUTOTN_WIDTH;
const DIVIDER_NONE = 'mobileToolbar_tinySeperator'; // 新行

exports.DIVIDER_NONE = DIVIDER_NONE;
const NEW_LINE = 'mobileToolbar_newLine'; // 小分割（约 12px）

exports.NEW_LINE = NEW_LINE;
const DIVIDER = 'mobileToolbar_divider'; // 小分割（竖向）

exports.DIVIDER = DIVIDER;
const DIVIDER_VERT = 'mobileToolbar_divider_vert'; // 空白的 icon，用于最后一行的左对齐

exports.DIVIDER_VERT = DIVIDER_VERT;
const ALIGN_PLACEHOLDER = 'mobileToolbar_emptyIcon';
exports.ALIGN_PLACEHOLDER = ALIGN_PLACEHOLDER;
const THREED_TRAMSFORM = 'translate3d(0, 0, 0)'; // 浮出二级菜单的最小左侧距离

exports.THREED_TRAMSFORM = THREED_TRAMSFORM;
const SUBLINE_MIN_LEFT = 24; // 浮出二级菜单和底部的 margin

exports.SUBLINE_MIN_LEFT = SUBLINE_MIN_LEFT;
const SUBLINE_MARGIN_BOTTOM = 8; // name -> itemKey。移除 [. 空格 -]，转小写

exports.SUBLINE_MARGIN_BOTTOM = SUBLINE_MARGIN_BOTTOM;

const itemToKey = item => {
  const name = typeof item === 'object' ? item.name : item;
  return String(name).replace(/ |\.|-/g, '_').toLowerCase();
};

exports.itemToKey = itemToKey;
var _default = {
  DIVIDER,
  DIVIDER_NONE,
  DIVIDER_VERT,
  NEW_LINE
};
exports.default = _default;
