// @ali/4ever-plugin-basic/dist-node/bi/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _mergeData = _interopRequireDefault(require("./commands/mergeData"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onMouseUp = _interopRequireDefault(require("./handlers/onMouseUp"));

var _queries = require("./queries");

var _actions = require("./actions");

var _meta = require("../utils/meta");

const factory = () => {
  const plugin = {
    key: _meta.key,
    commands: {
      mergeData: _mergeData.default
    },
    actions: _actions.actions,
    onAction: _onAction.default,
    isActions: _actions.isActions,
    onMouseUp: _onMouseUp.default,
    onKeyDown: _onKeyDown.default,
    queries: _queries.queries,
    models: []
  };
  return plugin;
};

var _default = factory;
exports.default = _default;
