// @ali/4ever-bi/dist-web/plugins/link/components/card/LinkCardInlineToolbar.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Button } from '@ali/we-design';
import { RefreshIcon } from "./RefreshIcon";
import { InlineToolbar } from "../../../../index";
var defaultImageLayout = [{
  toolButtons: [{
    name: 'linkCard.refresh'
  }]
}];

var getToolButton = function getToolButton(name, icon, handler, testid) {
  return function () {
    return /*#__PURE__*/_createElement(Button, {
      testid: testid,
      onClick: handler,
      title: name,
      bordered: false,
      icon: icon
    });
  };
};

var _ref = /*#__PURE__*/_createElement(RefreshIcon, null);

var LinkCardToolbarToolbar = function LinkCardToolbarToolbar(props) {
  var triggerNode = props.triggerNode,
      locale = props.locale,
      onRefresh = props.onRefresh;
  var getCustomToolButtons = React.useMemo(function () {
    return {
      'linkCard.refresh': getToolButton(locale.refresh, _ref, onRefresh, 'bi-toolbar-linkCard-refres')
    };
  }, [onRefresh, locale]);
  return /*#__PURE__*/_createElement(InlineToolbar, {
    visible: true,
    triggerNode: triggerNode,
    inlineLayout: defaultImageLayout,
    customToolButtons: getCustomToolButtons
  });
};

export default LinkCardToolbarToolbar;
