// @ali/we-ui/dist-web/components/toolbar/normal/styled.js
import styled, { css } from 'styled-components';
import { withTheme } from "../../theme";
import { Ellipsis } from "../../../utils/styled.mixin";
var BoxCommon = /*#__PURE__*/css(["", ""], function (_ref) {
  var disabled = _ref.disabled;
  return disabled && "\n  position: relative;\n  opacity: .5;\n  &::before{\n    content: '';\n    position: absolute;\n    width: 100%;\n    top: 0;\n    bottom: 0;\n    z-index: 1;\n    cursor: not-allowed;\n  }\n  &:hover{\n    .tool-button{\n      background-color:initial;\n    }\n  }\n  ";
});
export var ToolsShow = /*#__PURE__*/styled.div(["flex:0 1 auto;white-space:nowrap;display:flex;"]);
var dividerStyle = /*#__PURE__*/css(["content:'';display:block;position:absolute;top:50%;transform:translateY(-50%);background-color:", ";"], function (_ref2) {
  var ui = _ref2.theme.ui;
  return ui.commonLineLightColor;
});
export var ToolsMore = withTheme(styled.div(["flex:0 0 auto;padding-right:8px;padding-left:9px;display:flex;align-items:center;position:relative;::before{", " left:4px;width:1px;height:", ";}"], dividerStyle, function (_ref3) {
  var isSingleLine = _ref3.isSingleLine;
  return isSingleLine ? '24px' : '42px';
}));
export var Content = withTheme(styled.div(["flex:auto;display:flex;&.left{justify-content:flex-start;}&.center{justify-content:center;}&.right{justify-content:flex-end;}"]));
export var ToolbarWrapper = withTheme(styled.div(["display:flex;position:relative;flex:auto;"]));
export var ToolbarFrame = withTheme(styled.div(["display:flex;flex-direction:row;align-items:center;padding:0 20px;.toggle-button{fontSize:12px;color:", ";}"], function (_ref4) {
  var ui = _ref4.theme.ui;
  return ui.commonBaseColor3;
}));
export var BoxVerticalWrap = /*#__PURE__*/styled.div(["display:flex;flex-direction:column;align-items:center;box-sizing:border-box;width:42px;height:58px;padding:0px 4px;cursor:pointer;", " & >:first-child{width:32px;height:32px;margin-top:2px;.weicon{font-size:20px;}.select-arrow{font-size:12px;transform:scale(.67)}}"], BoxCommon);
export var BoxHorizontalWrap = /*#__PURE__*/styled.div(["display:flex;flex-direction:row;align-items:center;width:fit-content;cursor:pointer;", ""], BoxCommon);
export var BoxSingleWrap = /*#__PURE__*/styled.div(["display:flex;align-items:center;cursor:pointer;margin:0 2px;", ""], BoxCommon);
export var ToolBoxText = withTheme(styled.div(["width:50px;text-align:center;font-size:12px;transform:scale(.9);margin-top:1px;color:", ";user-select:none;", ";"], function (p) {
  return p.theme.ui.commonActionIconColor;
}, Ellipsis));
export var ToolSubGroupWrap = withTheme(styled.div(["display:flex;flex-direction:row;align-items:center;position:relative;height:", ";&+&{padding-left:", ";margin-top:", "};&+&::before{", " left:4px;width:", ";height:24px;}"], function (p) {
  return p.isSingleLine ? '36px' : undefined;
}, function (p) {
  return p.isSingleLine ? '9px' : undefined;
}, function (p) {
  return p.isSingleLine ? undefined : '1px';
}, dividerStyle, function (p) {
  return p.isSingleLine ? '1px' : '0px';
}));
export var ToolGroupWrap = withTheme(styled.div(["display:flex;visibility:", ";width:", ";flex-direction:column;justify-content:center;flex-wrap:", ";height:", "px;position:relative;&+&{padding-left:13px;}&+&::before{", " width:1px;left:6px;height:42px;}"], function (p) {
  return p.isComplexLayout && !p.width ? 'hidden' : 'visible';
}, function (p) {
  return p.width ? p.width + "px" : undefined;
}, function (p) {
  return p.isComplexLayout ? 'wrap' : 'nowrap';
}, function (p) {
  return p.height;
}, dividerStyle));
export var ToolbarDivider = withTheme(styled.div(["display:block;position:relative;height:24px;width:16px;::before{", " width:1px;height:16px;}"], dividerStyle));
