// @ali/we-ui/dist-web/components/toolbar/normal/Toolbar.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Tooltip } from "../../tooltip";
import { Button } from "../../button";
import { DoubletoolbarNormal } from "../../icons";
import { ToolbarFrame } from "./styled";
export var EToolbarMode;

(function (EToolbarMode) {
  EToolbarMode["singleLine"] = "singleLine";
  EToolbarMode["doubleLine"] = "doubleLine";
})(EToolbarMode || (EToolbarMode = {}));

export var Toolbar = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(Toolbar, _React$Component);

  function Toolbar(_props) {
    var _this;

    _this = _React$Component.call(this, _props) || this;

    _this.handleModeChange = function () {
      // setState是异步执行
      _this.setState(function (preState, props) {
        var innerMode = preState.innerMode;
        var newMode = innerMode === EToolbarMode.singleLine ? EToolbarMode.doubleLine : EToolbarMode.singleLine;
        var onModeChange = props.onModeChange; // 受控模式

        if (onModeChange) {
          onModeChange(newMode);
        }

        return {
          innerMode: newMode
        };
      });
    };

    _this.state = {
      innerMode: _props.mode || _props.defaultMode || EToolbarMode.singleLine
    };
    return _this;
  }

  var _proto = Toolbar.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        mode = _this$props.mode,
        singleLineToolbar = _this$props.singleLineToolbar,
        doubleLineToolbar = _this$props.doubleLineToolbar;
    var innerMode = this.state.innerMode;
    var activeMode = mode || innerMode;
    var isSingleLine = activeMode === EToolbarMode.singleLine;
    var iconRotate = isSingleLine ? 0 : 180;
    return /*#__PURE__*/_createElement(ToolbarFrame, null, isSingleLine ? singleLineToolbar : doubleLineToolbar, /*#__PURE__*/_createElement(Tooltip, {
      title: ""
    }, /*#__PURE__*/_createElement(Button, {
      bordered: false,
      shape: "rect",
      onClick: this.handleModeChange,
      icon: /*#__PURE__*/_createElement(DoubletoolbarNormal, {
        className: "toggle-button",
        rotate: iconRotate
      })
    })));
  };

  return Toolbar;
}(React.Component);
