// @ali/4ever-bi/dist-web/plugins/image/mobile/index.js
import _extends from "@babel/runtime/helpers/extends";
import toolbar from "./toolbar";
import renderNode from "./renderNode";
import insertImage from "./insertImage";
import { DEPRECATED_ACTIONS, actions } from "../actions";
import insertImageFile from "../commands/insertImageFile";
import createOnCommand from "../handlers/onCommand";
import createRenderEditable from "../createRenderEditable";
import { queries } from "../queries";
import createOnCangjieInput from "../handlers/onCangjieInput";
import { moFactory } from '@ali/4ever-plugin-image';
var state = {};
var commands = {};

function createImagePlugin(config) {
  var imageConfig = config.image || {};
  commands = {
    insertImage: insertImage(state),
    insertImageFile: insertImageFile(imageConfig, state)
  };
  return _extends({}, moFactory({}), {
    toolbar: toolbar(imageConfig),
    renderNode: renderNode(imageConfig, state),
    renderEditable: createRenderEditable(imageConfig),
    onCommand: createOnCommand(config, state),
    onCangjieInput: createOnCangjieInput(imageConfig, state),
    commands: commands,
    queries: queries,
    actions: actions,
    models: {
      klass: 'inline',
      type: 'image'
    }
  });
}

createImagePlugin.actions = DEPRECATED_ACTIONS;
createImagePlugin.commands = commands;
export default createImagePlugin;
