// @ali/4ever-plugin-paragraph/dist-web/components/SpecialIndent.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Select, SelectOption as Option } from '@ali/we-design';
import { IndentType } from '@ali/4ever-utils';
export default function SpecialIndent(props) {
  var locale = props.locale,
      handleIndentTypeChange = props.handleIndentTypeChange,
      _props$type = props.type,
      type = _props$type === void 0 ? IndentType.none : _props$type;
  var indentList = [{
    title: locale.noneIndent,
    key: IndentType.none
  }, {
    title: locale.firstLineIndent,
    key: IndentType.firstLine
  }, {
    title: locale.hangIndent,
    key: IndentType.hang
  }];

  var handleChange = function handleChange(indent) {
    handleIndentTypeChange(indent);
  };

  return /*#__PURE__*/_createElement(Select, {
    className: "bi-sidebar-indent",
    value: type,
    onChange: handleChange,
    defaultValue: IndentType.none
  }, indentList.map(function (indent) {
    return /*#__PURE__*/_createElement(Option, {
      value: indent.key,
      key: indent.key
    }, /*#__PURE__*/_createElement("span", null, indent.title));
  }));
}
