// @ali/4ever-zhi/dist-web/plugins/link/linkCard/createRenderNode.js
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { URLLinkCard } from '@ali/we-biz-link-card';
import { LinkCardPlugin } from '@ali/4ever-bi';
import { LinkPlugin, useScrollableContent } from '@ali/4ever-bamboo';
var normalizeLink = LinkPlugin.normalizeLink,
    openLink = LinkPlugin.openLink;
var copyBlockLinkCard = LinkCardPlugin.actions.copyBlockLinkCard;
var Wrapper = /*#__PURE__*/styled.div(["cursor:pointer;"]);

function LinkCardNode(props) {
  var nodeProps = props.nodeProps,
      controller = props.controller,
      linkConfig = props.linkConfig,
      isMobile = props.isMobile;
  var locale = linkConfig.locale,
      _linkConfig$protocolA = linkConfig.protocolAllowList,
      protocolAllowList = _linkConfig$protocolA === void 0 ? ['http:', 'https:', 'dingtalk:', 'taobao:'] : _linkConfig$protocolA,
      openLinkOfConfig = linkConfig.openLink;
  var node = nodeProps.node,
      isSelected = nodeProps.isSelected;
  var selection = controller.value.selection;
  var _ref = node.data.metadata,
      href = _ref.href,
      cardInfo = _ref.cardInfo,
      text = _ref.text;
  var scrollContent = useScrollableContent() || document.body;
  var getContainer = React.useCallback(function () {
    return scrollContent;
  }, [scrollContent]);
  var urlInfo = React.useMemo(function () {
    var isTitleEqualsHref = text.trim() === href.trim();
    var actualTitle = isTitleEqualsHref ? cardInfo.title : text;
    return {
      title: actualTitle || '',
      desc: cardInfo.desc || '',
      imgURL: cardInfo.imgURL || ''
    };
  }, [cardInfo, text, href]);
  var linkCardLocale = React.useMemo(function () {
    var tooltipLocale = {
      editTooltip: locale.edit,
      copyTooltip: locale.copy,
      unlinkTooltip: locale.unlink,
      deleteTooltip: locale["delete"],
      switchStyleTooltip: locale.styleText
    };
    return _extends({}, locale, tooltipLocale);
  }, [locale]);
  var handleVisit = React.useCallback(function () {
    if (openLinkOfConfig) {
      var normalizedLink = normalizeLink(href, protocolAllowList);
      openLinkOfConfig(normalizedLink);
    } else {
      openLink(href, protocolAllowList);
    }
  }, [href, protocolAllowList, openLinkOfConfig]);
  var handleCopy = React.useCallback(function (e) {
    e.preventDefault();
    e.stopPropagation(); // 卡片形式下，使用仓颉复制

    controller.run('onAction', copyBlockLinkCard(node));
  }, [href, node, controller]);
  return /*#__PURE__*/_createElement(Wrapper, null, /*#__PURE__*/_createElement(URLLinkCard, {
    getContainer: getContainer,
    isSelected: isSelected && selection.isCollapsed,
    cardStyle: 'indoc',
    state: 'normal',
    url: href,
    readonly: true,
    locale: linkCardLocale,
    urlInfo: urlInfo,
    isMobile: isMobile // @ts-ignore
    ,
    onCopy: controller.canCopy ? handleCopy : undefined,
    onOpenURL: handleVisit,
    editBtnTestID: "link-toolbar-edit",
    copyBtnTestID: "link-toolbar-copy",
    cutBtnTestID: "link-toolbar-cut",
    unlinkBtnTestID: "link-toolbar-unlink",
    deleteBtnTestID: "link-toolbar-remove",
    switchStyleBtnTestID: "link-toolbar-switch",
    linkStyleBtnTestID: "link-toolbar-switch-link",
    cardStyleBtnTestID: "link-toolbar-switch-card",
    cardTestID: "link-card"
  }));
}

export default function createRenderNode(linkConfig, isMobile) {
  return function (nodeProps, controller) {
    return /*#__PURE__*/_createElement(LinkCardNode, {
      linkConfig: linkConfig,
      isMobile: isMobile,
      nodeProps: nodeProps,
      controller: controller
    });
  };
}
