// @ali/4ever-zhi/dist-web/plugins/link/linkCard/mobile.js
import _extends from "@babel/runtime/helpers/extends";
import { zhiCardFactory as cardFactory } from '@ali/4ever-factory';
import locale from "../../../locales/zh_CN";
import createRenderNode from "./createRenderNode";
import { LinkCardPlugin } from '@ali/4ever-bi';
import { LinkPlugin } from '@ali/4ever-bamboo';
var copyLink = LinkCardPlugin.utils.copyLink;
var LINK_CARD_COPY_LINK = LinkCardPlugin.actions.LINK_CARD_COPY_LINK;
var createZhiCalendarCardPlugin = cardFactory(function (config) {
  var _config$clipboard;

  var linkConfig = _extends({
    locale: _extends({}, locale.link.locale)
  }, config.link);

  var serializer = (_config$clipboard = config.clipboard) == null ? void 0 : _config$clipboard.serializer;
  return {
    cardType: 'link',
    isMobile: true,
    renderNode: createRenderNode(linkConfig, true),
    queries: {
      getMinTableColWidth: LinkPlugin.getMinTableColWidthForBlockLinkCard
    },
    onAction: function onAction(action, controller, next) {
      var type = action.type,
          payload = action.payload;

      if (type === LINK_CARD_COPY_LINK) {
        var _ref = payload,
            node = _ref.node;
        return copyLink(controller, node, serializer || controller.serializer);
      }

      return next();
    }
  };
});
export default createZhiCalendarCardPlugin;
