// @ali/we-design-next/dist-web/display/tooltip/Tooltip.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Popover } from "../popover/Popover";
import { TooltipWrapper, TooltipTitle, TooltipDescription } from "./styled";
export var tooltipOverlayRestStyle = {
  border: 'unset',
  boxShadow: 'unset',
  backgroundColor: 'transparent'
};
export var TooltipContent = /*#__PURE__*/React.memo(function (_ref) {
  var title = _ref.title,
    description = _ref.description;
  return /*#__PURE__*/_createElement(TooltipWrapper, null, title && /*#__PURE__*/_createElement(TooltipTitle, null, title), description && /*#__PURE__*/_createElement(TooltipDescription, null, description));
});
export var Tooltip = /*#__PURE__*/React.memo(function (props) {
  var children = props.children,
    title = props.title,
    description = props.description,
    placement = props.placement,
    getContainer = props.getContainer;
  return /*#__PURE__*/_createElement(Popover, {
    content: /*#__PURE__*/_createElement(TooltipContent, {
      title: title,
      description: description
    }),
    style: tooltipOverlayRestStyle,
    placement: placement,
    placementOffset: [4, 0],
    trigger: "hover",
    getContainer: getContainer,
    animation: true
  }, children);
});
