// @ali/we-design-next/dist-web/form/multiselect/styled.js
var _path;
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled, { css } from 'styled-components';
import { COLORS, CSS_BORDER_BOX, CSS_DISABLE_SELECT, CSS_SCROLLBAR } from "../../common";
var CSS_BORDERED_ENABLED = /*#__PURE__*/css(["&:hover{border-color:", ";}&:active{border-color:", ";}&:focus-within{border-color:", ";}"], COLORS.OTHERS.PRIMARY_HOVER, COLORS.COMMON.BLUE1_ACTIVE_COLOR, COLORS.DING_BLUR);
var CSS_BORDERED_DISABLED = /*#__PURE__*/css(["border-color:", ";"], COLORS.COMMON.OVERLAY_COLOR);
var CSS_BORDERED_ERRORED = /*#__PURE__*/css(["border-color:", ";&:focus-within{border-color:", ";}"], COLORS.COMMON.DANGER1_COLOR, COLORS.DING_BLUR);
var CSS_VISIBLE = /*#__PURE__*/css(["border-color:", ";box-shadow:0px 0px 0px 2px rgba(0,106,254,0.12);"], COLORS.DING_BLUR);
var COLOR_DISABLED = 'rgba(17, 31, 44, 0.24)';
export var MultiSelectInputWrap = /*#__PURE__*/styled.div(["", ";position:relative;display:flex;align-items:center;width:100%;height:", ";padding-left:4px;border-radius:4px;border:1px solid ", ";", ";cursor:", ";", ";outline:none;color:", ";background-color:", ";", " &:focus-within{", "}.input{height:", ";max-width:100%;min-width:40px;padding-left:1px;background-color:", ";> input{padding-left:0;}}"], CSS_BORDER_BOX, function (_ref) {
  var size = _ref.size;
  return size === 'normal' ? '32px' : '28px';
}, COLORS.COMMON.OVERLAY_COLOR, function (_ref2) {
  var disabled = _ref2.disabled,
    error = _ref2.error;
  return disabled ? CSS_BORDERED_DISABLED : error ? CSS_BORDERED_ERRORED : CSS_BORDERED_ENABLED;
}, function (_ref3) {
  var disabled = _ref3.disabled;
  return disabled ? 'not-allowed' : 'text';
}, CSS_DISABLE_SELECT, function (_ref4) {
  var disabled = _ref4.disabled;
  return disabled ? COLOR_DISABLED : COLORS.LIGHT.LEVEL1_BASE_COLOR;
}, function (_ref5) {
  var disabled = _ref5.disabled;
  return disabled ? '#F7F6F6' : 'unset';
}, function (_ref6) {
  var visible = _ref6.visible;
  return visible ? CSS_VISIBLE : null;
}, function (_ref7) {
  var disabled = _ref7.disabled;
  return !disabled ? 'box-shadow: 0px 0px 0px 2px rgba(0, 106, 254, 0.12)' : null;
}, function (_ref8) {
  var size = _ref8.size;
  return size === 'normal' ? '30px' : '26px';
}, function (_ref9) {
  var disabled = _ref9.disabled;
  return disabled ? '#F7F6F6' : 'unset';
});
export var OverflowWrap = /*#__PURE__*/styled.div(["display:flex;flex:auto;max-width:100%;"]);
export var SuffixWrap = /*#__PURE__*/styled.span(["display:flex;margin-left:4px;margin-right:4px;line-height:0;cursor:default;"]);
export var OverlayWrap = /*#__PURE__*/styled.div(["max-height:400px;overflow-y:overlay;", ";"], CSS_SCROLLBAR);
export var ErrorTip = /*#__PURE__*/styled.div(["position:absolute;top:", ";left:0;margin-top:4px;font-size:12px;line-height:18px;color:", ";"], function (_ref10) {
  var size = _ref10.size;
  return size === 'normal' ? '32px' : '28px';
}, COLORS.COMMON.DANGER1_COLOR);
export var DisableMask = /*#__PURE__*/styled.div(["position:absolute;top:0px;left:0px;width:100%;height:100%;cursor:no-drop;"]);
export var SingleTagWrap = /*#__PURE__*/styled.span(["", ";display:flex;align-items:center;height:", ";margin:3px 3px 3px 0;padding:4px;font-size:13px;background:#EBECED;border-radius:4px;", ";cursor:default;.wdn-icon{color:", ";.hover{color:unset;}}"], CSS_BORDER_BOX, function (_ref11) {
  var size = _ref11.size;
  return size === 'normal' ? '26px' : '22px';
}, CSS_DISABLE_SELECT, COLORS.LIGHT.LEVEL3_BASE_COLOR);
export var TagContentWrap = /*#__PURE__*/styled.span(["margin-right:3px;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;"]);
export var ItemWrap = /*#__PURE__*/styled.div(["flex:none;align-self:center;max-width:100%;"]);
export var ArrowDown = /*#__PURE__*/React.memo(function (props) {
  return /*#__PURE__*/_createElement("svg", {
    width: "12",
    height: "12",
    viewBox: "0 0 12 12",
    stroke: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/_createElement("path", {
    d: "M6.00004 7.79293L1.85359 3.64648L1.14648 4.35359L6.00004 9.20714L10.8536 4.35359L10.1465 3.64648L6.00004 7.79293Z",
    fill: props.disabled ? COLOR_DISABLED : COLORS.LIGHT.LEVEL1_BASE_COLOR
  }));
});
export var ArrowUp = /*#__PURE__*/React.memo(function (props) {
  return /*#__PURE__*/_createElement("svg", {
    width: "12",
    height: "12",
    viewBox: "0 0 12 12",
    stroke: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/_createElement("path", {
    d: "M5.99996 4.20707L10.1464 8.35352L10.8535 7.64641L5.99996 2.79286L1.14641 7.64641L1.85352 8.35351L5.99996 4.20707Z",
    fill: props.disabled ? COLOR_DISABLED : COLORS.LIGHT.LEVEL1_BASE_COLOR
  }));
});
export var ClearIcon = /*#__PURE__*/React.memo(function () {
  return /*#__PURE__*/_createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    width: "12",
    height: "12",
    viewBox: "0 0 12 12",
    fill: "none"
  }, /*#__PURE__*/_createElement("path", {
    d: "M12 6C12 9.31371 9.31371 12 6 12C2.68629 12 0 9.31371 0 6C0 2.68629 2.68629 0 6 0C9.31371 0 12 2.68629 12 6Z",
    fill: COLORS.LIGHT.LEVEL1_BASE_COLOR,
    "fill-opacity": "0.24"
  }), _path || (_path = /*#__PURE__*/_createElement("path", {
    d: "M5.29294 5.99996L3.14651 3.85353L3.85361 3.14642L6.00005 5.29286L8.14646 3.14644L8.85357 3.85355L6.70715 5.99996L8.85359 8.1464L8.14648 8.85351L6.00005 6.70707L3.85359 8.85353L3.14648 8.14642L5.29294 5.99996Z",
    fill: "${COLORS.LIGHT.LEVEL1_BASE_COLOR}",
    "fill-opacity": "0.6"
  })));
});
