// @ali/4ever-plugin-list/dist-node/mo/serializer/htmlToJsonML/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRandomId = getRandomId;
exports.shouldFlattenChildren = shouldFlattenChildren;
exports.getFlattenedChildren = getFlattenedChildren;
Object.defineProperty(exports, "isMsWordList", {
  enumerable: true,
  get: function () {
    return _msWordUtils.isMsWordList;
  }
});
Object.defineProperty(exports, "transfromIndLeft", {
  enumerable: true,
  get: function () {
    return _msWordUtils.transfromIndLeft;
  }
});
Object.defineProperty(exports, "removeSymbolText", {
  enumerable: true,
  get: function () {
    return _msWordUtils.removeSymbolText;
  }
});
exports.listTags = void 0;

var _everUtils = require("@ali/4ever-utils");

var _msWordUtils = require("./msWordUtils");

const listTags = {
  ol: 'ol',
  ul: 'ul',
  li: 'li'
}; // Refactor: 与 Bi 共用

exports.listTags = listTags;

function getRandomId() {
  if (process.env.NODE_ENV === 'test') return 'test';
  return Math.random().toString(36).slice(2);
}

function shouldFlattenChildren(li) {
  const children = (0, _everUtils.getChildren)(li);
  return children.some(child => {
    const childTagName = (0, _everUtils.getTagName)(child);
    return !!_everUtils.paragraphTags[childTagName] || childTagName === 'table';
  });
}

function getFlattenedChildren(li) {
  const children = (0, _everUtils.getChildren)(li);
  return children.map((child, index) => {
    if (index === 0) {
      if ((0, _everUtils.getTagName)(child) === 'p') {
        return ['p', { ...(0, _everUtils.getAttributes)(li),
          ...(0, _everUtils.getAttributes)(child)
        }, ...(0, _everUtils.getChildren)(child)];
      }

      return ['p', { ...(0, _everUtils.getAttributes)(li)
      }, child];
    }

    return child;
  });
}
