// @ali/4ever-plugin-clipboard/dist-web/bi/utils/isParagraph.js
// 这几个方法原先在 bi/src/plugins/common/utils/isParagraph.ts 中，依赖 heading、paragraph 插件。但没有其他地方用到，先直接迁移到 clipboard 中
import { Block, Document } from '@ali/4ever-cangjie';
import { Paragraph } from '@ali/4ever-plugin-paragraph';
import { Heading } from '@ali/4ever-plugin-heading'; // section可以看作paragraph、heading的超集

export function isSection(block) {
  return Heading.isHeading(block) || Paragraph.isParagraph(block);
}
/**
 * 检测 fragment 是否仅包含一个 普通段落或标题
 * section可以看作paragraph、heading的超集
 * @param fragment
 */

export function isSingleSection(fragment) {
  var firstBlock = fragment.nodes[0];
  return fragment.nodes.length === 1 && isSection(firstBlock);
}
/**
 * 从 fragment 中把 single paragraph 解构出来
 * @param fragment
 */

export function unwrapSingleSection(fragment) {
  var firstBlock = fragment.nodes[0];

  if (isSection(firstBlock)) {
    var node = Block.create({
      // 类型不变（例如复制标题的一部分，粘贴还是标题）
      type: firstBlock.type,
      data: {},
      nodes: firstBlock.nodes
    });
    return Document.create({
      nodes: [node]
    });
  }

  return fragment;
}
