// @ali/4ever-bi/dist-web/plugins/hoverToolbar/components/MoreMenuButton.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { TBIconButton, Popover, List } from '@ali/we-design-next';
import { IconMore, HoverToolbarHooks } from '@ali/4ever-component';
import { PluginRoles } from '@ali/4ever-bamboo';

var _ref2 = /*#__PURE__*/_createElement(IconMore, null);

export var MoreMenuButton = function MoreMenuButton(props) {
  var items = props.items,
      onOverlayVisibleChange = props.onOverlayVisibleChange,
      onSelect = props.onSelect,
      locale = props.locale;
  var getContainer = HoverToolbarHooks.useGetContainer();

  var _HoverToolbarHooks$us = HoverToolbarHooks.usePopoverVisible('more', onOverlayVisibleChange),
      isOverlayVisible = _HoverToolbarHooks$us[0],
      setOverlayVisible = _HoverToolbarHooks$us[1];

  var handleSelect = React.useCallback(function (key) {
    onSelect(key);
    setOverlayVisible(false);
  }, [onSelect, setOverlayVisible]);
  var toggleVisible = React.useCallback(function () {
    setOverlayVisible(!isOverlayVisible);
  }, [setOverlayVisible, isOverlayVisible]);
  var renderItem = React.useCallback(function (_ref) {
    var key = _ref.key;
    if (!items) return null;
    var item = items == null ? void 0 : items.find(function (t) {
      return typeof t !== 'string' && t.key === key;
    });
    if (!item) return null;
    return /*#__PURE__*/_createElement(List.Item, item);
  }, [items]);
  return /*#__PURE__*/_createElement(Popover, {
    placement: "bottomLeft",
    content: /*#__PURE__*/_createElement(List, {
      "data-testid": "bi-toolbar-hover-more-options",
      items: items,
      onSelect: handleSelect,
      "data-role": "hoverToolbarMoreList",
      renderItem: renderItem
    }),
    onVisibleChange: setOverlayVisible,
    visible: isOverlayVisible,
    getContainer: getContainer,
    "data-testid": "overlay-bi-toolbar-hover-more"
  }, /*#__PURE__*/_createElement(TBIconButton, {
    active: isOverlayVisible,
    icon: _ref2,
    onClick: toggleVisible,
    title: (locale == null ? void 0 : locale.more) || 'More',
    tooltip: isOverlayVisible ? '' : (locale == null ? void 0 : locale.more) || 'More',
    testid: "bi-toolbar-hover-more",
    role: PluginRoles.hoverToolbarMore
  }));
};
