// @ali/we-design/dist-web/components/form/Form.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import Form, { useForm } from 'rc-field-form';
var defaultValueForContext = {
  layout: 'horizontal'
};
export var FormContext = /*#__PURE__*/React.createContext(defaultValueForContext);

var Main = function Main(_ref, ref) {
  var form = _ref.form,
      initialValues = _ref.initialValues,
      onFinish = _ref.onFinish,
      onFinishFailed = _ref.onFinishFailed,
      onValuesChange = _ref.onValuesChange,
      children = _ref.children,
      labelSpan = _ref.labelSpan,
      rest = _objectWithoutPropertiesLoose(_ref, ["form", "initialValues", "onFinish", "onFinishFailed", "onValuesChange", "children", "labelSpan"]);

  var _defaultValueForConte = _extends({}, defaultValueForContext, rest),
      layout = _defaultValueForConte.layout;

  return /*#__PURE__*/_createElement(FormContext.Provider, {
    value: {
      layout: layout,
      labelSpan: labelSpan
    }
  }, /*#__PURE__*/_createElement(Form, {
    onValuesChange: onValuesChange,
    onFinishFailed: onFinishFailed,
    onFinish: onFinish,
    initialValues: initialValues,
    form: form,
    ref: ref
  }, children));
};

export default Main;
export { useForm };
