// @ali/4ever-component/dist-web/MobileToolbar/layout/index.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import fontLayout from "./font";
import insertLayout from "./insert";
import alignLayout from "./align";

function getButton(focusIcon, icon) {
  return function (_ref) {
    var Wrapper = _ref.Wrapper,
        panelShowing = _ref.panelShowing,
        rest = _objectWithoutPropertiesLoose(_ref, ["Wrapper", "panelShowing"]);

    return /*#__PURE__*/_createElement(Wrapper, _extends({
      type: "big"
    }, rest, {
      isActive: panelShowing,
      icon: panelShowing ? focusIcon : icon
    }));
  };
}

export var mobileLayout = [{
  key: 'group1',
  items: [{
    name: 'insert',
    Component: getButton('plus', 'plus'),
    panel: {
      title: 'Insert',
      content: insertLayout
    }
  }, {
    name: 'font',
    Component: getButton('font', 'font'),
    panel: {
      title: 'Fonts',
      content: fontLayout
    }
  }, {
    name: 'align',
    Component: getButton('alignLeft', 'alignLeft'),
    panel: {
      title: 'Align',
      content: alignLayout
    }
  }]
}, {
  key: 'group2',
  items: ['image.mobileBig', 'list.mobileTask', 'mention.mobile', 'history.undoMobile', 'history.redoMobile']
}, {
  key: 'fixed',
  fixedOnRight: true,
  items: ['keyboard']
}];
