// @ali/4ever-bi/dist-node/plugins/clearStyle/components/clearStyleButtonMobile.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MobileClearButton = MobileClearButton;
exports.MobileClearButtonBig = MobileClearButtonBig;

var React = _interopRequireWildcard(require("react"));

var _actions = require("../actions");

const _createElement = /*#__PURE__*/React.createElement;

function MobileClearButton(props) {
  const {
    controller,
    Wrapper,
    item,
    locale
  } = props;
  const handleClick = React.useCallback(() => {
    controller.run('onAction', (0, _actions.clearStyles)());
  }, [controller]);
  return /*#__PURE__*/_createElement(Wrapper, {
    type: 'horiz',
    item: item,
    width: 2.1,
    onClick: handleClick
  }, /*#__PURE__*/_createElement("div", null, locale.buttonTitle));
}

function MobileClearButtonBig(props) {
  const {
    controller,
    Wrapper,
    item
  } = props;
  const handleClick = React.useCallback(() => {
    controller.run('onAction', (0, _actions.clearStyles)());
  }, [controller]);
  return /*#__PURE__*/_createElement(Wrapper, {
    type: 'big',
    icon: "clearStyle",
    item: item,
    onClick: handleClick
  });
}
