// @ali/4ever-bi/dist-node/plugins/clearStyle/components/clearStyleButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ClearStyleButton;
exports.CLEAR_STYLE_HOT_KEY = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everCangjie = require("@ali/4ever-cangjie");

var _weToolbar = require("@ali/we-toolbar");

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../actions");

const _createElement = /*#__PURE__*/React.createElement;
const MOD = _everCangjie.environment.IS_MAC ? '⌘' : 'Ctrl';
const CLEAR_STYLE_HOT_KEY = `${MOD}+\\`;
exports.CLEAR_STYLE_HOT_KEY = CLEAR_STYLE_HOT_KEY;

function ClearStyleButton(props) {
  const {
    controller,
    locale,
    ...rest
  } = props;
  const handleClick = React.useCallback(() => {
    controller.run('onAction', (0, _actions.clearStyles)());
  }, [controller]);
  return /*#__PURE__*/_createElement(_weToolbar.EraserButton, (0, _extends2.default)({}, rest, {
    onClick: handleClick,
    shortcutText: CLEAR_STYLE_HOT_KEY,
    testid: 'bi-toolbar-clearstyle',
    role: _everBamboo.PluginRoles.clear
  }));
}
