// @ali/4ever-plugin-table/dist-web/bi/constants.js
export { MIN_COL_WIDTH, MIN_ROW_HEIGHT, QUICK_INSERT_BUTTON_WIDTH_HEIGHT, TOOLBAR_ITEM_SIZE_MOBILE } from "../utils/constants";
export var DOC_WIDTH = 598; // 行列插入指示线宽度 or 高度

export var INSERT_INDICATOR_LINE_SIZE = 2; // 降低以处理过于敏感容易误触的问题

export var DETECT_DELTA = 2; //  hover 出现插入按钮在行工具栏左侧区域的宽度
// 列工具栏同理

export var HOVER_SHOW_INSERT_DELTA = 15; // 当行高/列宽小于 32px 时，
// 行上下侧 8px 或列左右侧 8px 区域 hover 才出现插入按钮

export var MIN_HOVER_SHOW_INSERT_DELTA = 8;
export var DISTANCE_BETWEEN_INSERT_BUTTON_AND_TOOLBAR = -2;
export var DISTANCE_BETWEEN_DELETE_BUTTON_AND_TOOLBAR = 2;
export var TOOLBAR_ITEM_SIZE = 8;
export var THEME = {
  toolbarBorderColor: '#D6D6D6',
  toolbarBgColor: {
    "default": '#F5F6F7',
    hover: '#e2e4e6b3',
    active: '#87B5F6'
  },
  toolbarItemBorderColor: 'rgba(126, 134, 142, 0.08)',
  blue0: 'rgba(0, 106, 254, 0.12)',
  // blue1: 'rgba(126,134,142,0.08)',
  blue1: 'rgba(0, 122, 255, 0.5)',
  blue2: 'rgba(81, 152, 255, 0.48)',
  blue3: 'rgb(0, 106, 254)',
  dragBlue0: '#0089FF',
  dragBlue1: 'rgba(0, 106, 254, 0.24)',
  red0: 'rgba(255, 0, 0, 0.15)',
  red1: 'rgba(255, 0, 0, 0.67)',
  selectionBgColor: 'rgba(0, 122, 255, 0.1)',
  deleteBgColor: 'rgba(255, 0, 0, 0.15)',
  cellBorderColor: '#D9DBDD',
  rowColIndicatorColor: '#3296FA'
};
export var DRAG_TRIGGER_TIME = 200; // 表头 styleId

export var TABLE_HEADER_STYLE_ID = 'tableHeader';
export var DEFAULT_TABLE_ALIGNMENT = 'middle';
