// @ali/4ever-plugin-table/dist-web/bi/index.js
import _extends from "@babel/runtime/helpers/extends";
import toolbar from "./toolbar";
import menubar from "./menubar";
import createOnAction from "./handlers/createOnAction";
import onCangjieInput from "./handlers/onCangjieInput";
import createOnKeyDown from "./handlers/createOnKeyDown";
import onCangjiePaste from "./handlers/onCangjiePaste";
import quickInsert from "./quickInsertConfig";
import queries from "./queries";
import createOnCangjieCopy from "./handlers/createOnCangjieCopy";
import createOnCangjieCut from "./handlers/createOnCangjieCut";
import onMouseDown from "./handlers/onMouseDown";
import createContextMenu from "./createContextMenu";
import createRenderNode from "./createRenderNode";
import createRenderToolbar from "./createRenderToolbar";
import shouldNodeComponentUpdate from "../common/handlers/shouldNodeComponentUpdate";
import selectionTrigger from "../utils/utils/selectionTrigger";
import getNodeStyle from "../common/handlers/getNodeStyle";
import { moFactory } from "../mo";
import { actions, isActions } from "./actions";
import * as commands from "./commands";
var contextMenu = createContextMenu();
export { mobileBiFactory } from "./mobile";
export var biFactory = function biFactory(configs) {
  var tableConfig = configs.table || {};
  var _configs$clipboard = configs.clipboard,
      clipboard = _configs$clipboard === void 0 ? {} : _configs$clipboard;
  return _extends({}, moFactory({}), {
    toolbar: toolbar(tableConfig),
    renderToolbar: createRenderToolbar(tableConfig),
    menubar: menubar(),
    renderNode: createRenderNode(configs),
    shouldNodeComponentUpdate: shouldNodeComponentUpdate,
    commands: commands,
    actions: actions,
    isActions: isActions,
    onAction: createOnAction(tableConfig),
    onCangjieInput: onCangjieInput,
    onKeyDown: createOnKeyDown(tableConfig),
    onCangjieCopy: createOnCangjieCopy(),
    onCangjieSelect: function onCangjieSelect(event, _, next) {
      selectionTrigger.set(event.detail.trigger);
      return next();
    },
    onCangjieCut: createOnCangjieCut(),
    onCangjiePaste: onCangjiePaste(clipboard, tableConfig),
    onMouseDown: onMouseDown,
    quickInsert: quickInsert,
    queries: queries,
    getNodeStyle: getNodeStyle,
    contextMenu: contextMenu(tableConfig),
    models: [{
      klass: 'block',
      type: 'table'
    }, {
      klass: 'block',
      type: 'table-cell'
    }, {
      klass: 'block',
      type: 'table-row'
    }]
  });
};
