// @ali/4ever-plugin-table/dist-web/bi/toolbar.js
import TableButton from "./components/toolbarButtons/tableButton";
import TablePickerButton from "./components/toolbarButtons/tablePickerButton";
import TableMergeButton from "./components/toolbarButtons/tableMergeButton";
import TableSimpleMergeButton from "./components/toolbarButtons/tableSimpleMergeButton";
import TableSimpleSplitButton from "./components/toolbarButtons/tableSimpleSplitButton";
import TableFillButton from "./components/toolbarButtons/tableFillButton";
import TableBorderSet from "./components/toolbarButtons/tableBorderSet";
import TableVAlignButton from "./components/toolbarButtons/tableVAlignButton";
import TableFitWidthButton from "./components/toolbarButtons/tableFitWidthButton";
import TableDeleteButton from "./components/toolbarButtons/tableDeleteButton";
import TableVAlignTopButton from "./components/toolbarButtons/tableVAlignTopButton";
import TableVAlignMiddleButton from "./components/toolbarButtons/tableVAlignMiddleButton";
import TableVAlignBottomButton from "./components/toolbarButtons/tableVAlignBottomButton";
import isSelectionInTable from "./utils/isSelectionInTable";
import TableFillButtonFocus from "./components/toolbarButtons/tableFillButtonFocus";
import TableThemeButton from "./components/toolbarButtons/tableThemeButton";
export default function toolbar(tableConfig) {
  var tableLocaleConfig = tableConfig.locale || {};
  return {
    buttons: {
      "default": TableButton,
      picker: TablePickerButton,
      merge: TableMergeButton,
      mergeSimple: TableSimpleMergeButton,
      splitSimple: TableSimpleSplitButton,
      fill: TableFillButton,
      fillFocus: TableFillButtonFocus,
      vAlign: TableVAlignButton,
      vAlignTop: TableVAlignTopButton,
      vAlignMiddle: TableVAlignMiddleButton,
      vAlignBottom: TableVAlignBottomButton,
      border: TableBorderSet,
      fitWidth: TableFitWidthButton,
      "delete": TableDeleteButton,
      theme: TableThemeButton
    },
    locale: tableLocaleConfig,
    getTarget: function getTarget(controller) {
      var hasHoverToolbar = controller.query('hasHoverToolbar', 'table');

      if (hasHoverToolbar) {
        var tableSelection = controller.query('tableSelection');
        return !!tableSelection ? 'tableSelection' : '';
      }

      if (!isSelectionInTable(controller.value)) {
        return '';
      }

      return 'table';
    }
  };
}
