// @ali/4ever-component/dist-node/ToolbarNext/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultDoubleLayout = exports.defaultSingleLayout = exports.defaultDisableTargets = void 0;
const defaultDisableTargets = ['hr', 'code'];
exports.defaultDisableTargets = defaultDisableTargets;
const defaultSingleLayout = [{
  items: [{
    name: 'history.undo'
  }, {
    name: 'history.redo'
  }, {
    name: 'formatPainter'
  }, {
    name: 'clearStyle'
  }]
}, {
  items: [{
    name: 'insertMore'
  }]
}, {
  items: [{
    name: 'heading'
  }, {
    name: 'fonts'
  }, {
    name: 'sz'
  }]
}, {
  items: [{
    name: 'bold'
  }, {
    name: 'italic'
  }, {
    name: 'underline'
  }, {
    name: 'strike'
  }, {
    name: 'inlineCode'
  }]
}, {
  items: [{
    name: 'color'
  }, {
    name: 'highlight'
  }]
}, {
  items: [{
    name: 'list.ulist'
  }, {
    name: 'list.olist'
  }, {
    name: 'list.tlist'
  }, {
    name: 'indent.decrease'
  }, {
    name: 'indent.increase'
  }, {
    name: 'alignment'
  }, {
    name: 'lineheight'
  }]
}, {
  items: [{
    name: 'sticker'
  }, {
    name: 'image'
  }, {
    name: 'table.picker',
    disableTargets: ['table']
  }, {
    name: 'embed'
  }, {
    name: 'link'
  }]
}, {
  items: [{
    name: 'findAndReplace'
  }]
}, {
  items: [{
    name: 'inlineComment'
  }, {
    name: 'voiceTranslation'
  }, {
    name: 'flashminute'
  }, {
    name: 'translate'
  }, {
    name: 'mentions'
  }, {
    name: 'dingTask'
  }, {
    name: 'revision'
  }]
}];
exports.defaultSingleLayout = defaultSingleLayout;
const defaultDoubleLayout = [{
  blocks: [{
    lines: [{
      items: [{
        name: 'history.undo'
      }]
    }, {
      items: [{
        name: 'history.redo'
      }]
    }]
  }]
}, {
  blocks: [{
    items: [{
      name: 'formatPainter'
    }, {
      name: 'clearStyle'
    }]
  }]
}, {
  blocks: [{
    items: [{
      name: 'insertMore'
    }]
  }]
}, {
  blocks: [{
    lines: [{
      items: [{
        name: 'fonts.wide'
      }, {
        name: 'sz'
      }, {
        name: 'sz.bigger'
      }, {
        name: 'sz.smaller'
      }]
    }, {
      items: [{
        name: 'bold'
      }, {
        name: 'italic'
      }, {
        name: 'underline'
      }, {
        name: 'strike'
      }, {
        name: 'vertAlign.supMark'
      }, {
        name: 'vertAlign.subMark'
      }, {
        name: 'inlineCode'
      }, {
        name: 'color'
      }, {
        name: 'highlight'
      }]
    }]
  }]
}, {
  blocks: [{
    lines: [{
      items: [{
        name: 'table.border'
      }, {
        name: 'table.vAlign'
      }]
    }, {
      items: [{
        name: 'table.fill'
      }, {
        name: 'table.merge'
      }]
    }]
  }]
}, {
  blocks: [{
    lines: [{
      items: [{
        name: 'list.ulist'
      }, {
        name: 'list.olist'
      }, {
        name: 'list.tlist'
      }, {
        name: 'indent.decrease'
      }, {
        name: 'indent.increase'
      }]
    }, {
      items: [{
        name: 'alignment.left'
      }, {
        name: 'alignment.center'
      }, {
        name: 'alignment.right'
      }, {
        name: 'alignment.justify'
      }, {
        name: 'lineheight'
      }]
    }]
  }, {
    items: [{
      name: 'heading.group'
    }]
  }, {
    items: [{
      name: 'sticker'
    }, {
      name: 'image'
    }, {
      name: 'table.picker',
      disableTargets: ['table']
    }, {
      name: 'embed'
    }, {
      name: 'link'
    }]
  }]
}, {
  blocks: [{
    items: [{
      name: 'findAndReplace'
    }]
  }]
}, {
  blocks: [{
    items: [{
      name: 'inlineComment'
    }, {
      name: 'voiceTranslation'
    }, {
      name: 'flashminute'
    }, {
      name: 'translate'
    }, {
      name: 'mentions'
    }, {
      name: 'dingTask'
    }, {
      name: 'revision'
    }]
  }]
}];
exports.defaultDoubleLayout = defaultDoubleLayout;
