// @ali/4ever-component/dist-node/ToolbarNext/ItemKeys.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ItemKey = void 0;
let ItemKey;
exports.ItemKey = ItemKey;

(function (ItemKey) {
  ItemKey["findAndReplace"] = "findAndReplace";
  ItemKey["alignment"] = "alignment";
  ItemKey["alignment.left"] = "alignment.left";
  ItemKey["alignment.center"] = "alignment.center";
  ItemKey["alignment.right"] = "alignment.right";
  ItemKey["alignment.justify"] = "alignment.justify";
  ItemKey["bold"] = "bold";
  ItemKey["italic"] = "italic";
  ItemKey["strike"] = "strike";
  ItemKey["comment"] = "comment";
  ItemKey["formatPainter"] = "formatPainter";
  ItemKey["history.undo"] = "history.undo";
  ItemKey["history.redo"] = "history.redo";
  ItemKey["image"] = "image";
  ItemKey["indent.decrease"] = "indent.decrease";
  ItemKey["indent.increase"] = "indent.increase";
  ItemKey["inlineCode"] = "inlineCode";
  ItemKey["clearStyle"] = "clearStyle";
  ItemKey["link"] = "link";
  ItemKey["list.tlist"] = "list.tlist";
  ItemKey["list.olist"] = "list.olist";
  ItemKey["list.ulist"] = "list.ulist";
  ItemKey["mentions"] = "mentions";
  ItemKey["sticker"] = "sticker";
  ItemKey["sz"] = "sz";
  ItemKey["sz.bigger"] = "sz.bigger";
  ItemKey["sz.smaller"] = "sz.smaller";
  ItemKey["translate"] = "translate";
  ItemKey["underline"] = "underline";
  ItemKey["heading"] = "heading";
  ItemKey["heading.group"] = "heading.group";
  ItemKey["color"] = "color";
  ItemKey["highlight"] = "highlight";
  ItemKey["lineheight"] = "lineheight";
  ItemKey["table.picker"] = "table.picker";
  ItemKey["vertAlign.subMark"] = "vertAlign.subMark";
  ItemKey["vertAlign.supMark"] = "vertAlign.supMark";
  ItemKey["vertAlign"] = "vertAlign";
  ItemKey["fonts"] = "fonts";
  ItemKey["fonts.wide"] = "fonts.wide";
  ItemKey["table.border"] = "table.border";
  ItemKey["table.vAlign"] = "table.vAlign";
  ItemKey["table.fill"] = "table.fill";
  ItemKey["table.merge"] = "table.merge";
  ItemKey["table.theme"] = "table.theme";
})(ItemKey || (exports.ItemKey = ItemKey = {}));
