// @ali/we-design-next/dist-node/form/date-picker/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SwitchOptionWrapper = exports.SpaceBar = exports.SPACE_HEIGHT = exports.NowWrapper = exports.NOW_HEIGHT = exports.DatePickerWrapper = void 0;
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _common = require("../../common");
const NOW_HEIGHT = 32;
exports.NOW_HEIGHT = NOW_HEIGHT;
const SPACE_HEIGHT = 20;
exports.SPACE_HEIGHT = SPACE_HEIGHT;
const DatePickerWrapper = /*#__PURE__*/_styledComponents.default.div(["width:fit-content;height:fit-content;background:", ";"], _common.COLORS.WHITE);
exports.DatePickerWrapper = DatePickerWrapper;
const NowWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;justify-content:center;height:", "px;"], NOW_HEIGHT);
exports.NowWrapper = NowWrapper;
const SpaceBar = /*#__PURE__*/_styledComponents.default.div(["width:100%;height:", "px;"], SPACE_HEIGHT);
exports.SpaceBar = SpaceBar;
const SwitchOptionWrapper = /*#__PURE__*/_styledComponents.default.div(["float:left;height:32px;display:flex;align-items:center;font-weight:400;font-size:14px;line-height:20px;color:", ";"], _common.COLORS.LIGHT.LEVEL1_BASE_COLOR);
exports.SwitchOptionWrapper = SwitchOptionWrapper;
