// @ali/4ever-bi/dist-web/plugins/voteCard/index.js
import _extends from "@babel/runtime/helpers/extends";
import { cardFactory } from '@ali/4ever-factory';
import locale from "../../locales/zh_CN";
import { VoteCardPlugin } from '@ali/4ever-bamboo';
import createOnCangjieCopy from "./handlers/createOnCangjieCopy";
import createRenderNode from "./createRenderNode";
import createMenubar from "./handlers/createMenu";
import { copyNodes } from '@ali/4ever-plugin-clipboard';
var createOnAction = VoteCardPlugin.createOnAction,
    createQueries = VoteCardPlugin.createQueries;

var createPlugin = function createPlugin(bizConfig) {
  var voteCardConfig = _extends({}, locale.voteCard, bizConfig.voteCard); // @ts-ignore


  var renderNode = createRenderNode(voteCardConfig); // @ts-ignore

  var onAction = createOnAction(voteCardConfig, copyNodes);
  return {
    cardType: 'vote',
    onAction: onAction,
    renderNode: renderNode,
    onCangjieCopy: createOnCangjieCopy(),
    menubar: {
      menu: createMenubar()
    },
    queries: createQueries()
  };
};

var createBiVoteCardPlugin = cardFactory(createPlugin);
export default createBiVoteCardPlugin;
