// @ali/4ever-plugin-table/dist-web/bi/components/cornerToolbar/tableWidthFitContainerButton.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { TableRole as PluginRoles } from "../../../utils/role";
import { Tooltip } from '@ali/we-design';
import { AdaptNormal } from '@ali/we-icon';
var Button = /*#__PURE__*/styled.span(["position:relative;display:flex;align-items:center;justify-content:center;color:rgba(17,31,44,0.72);margin:0;padding:0;width:28px;height:28px;font-size:16px;cursor:pointer;line-height:16px;border-radius:4px;&:hover{background-color:rgba(17,31,44,0.08);}"]);

var TableWidthFitContainerButton = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(TableWidthFitContainerButton, _React$Component);

  function TableWidthFitContainerButton() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;

    _this.handleMouseDown = function (e) {
      e.preventDefault(); // Prevent blur.

      e.stopPropagation();
      var onMenuEvent = _this.props.onMenuEvent;
      onMenuEvent('table-width-fit-container');
    };

    return _this;
  }

  var _proto = TableWidthFitContainerButton.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        style = _this$props.style,
        locale = _this$props.locale;
    return /*#__PURE__*/_createElement(Button, {
      style: style,
      onMouseDown: this.handleMouseDown,
      onTouchStart: this.handleMouseDown,
      "data-role": PluginRoles.tableFitWidth,
      "data-testid": "table-width-fit-button"
    }, /*#__PURE__*/_createElement(Tooltip, {
      placement: "left",
      title: locale == null ? void 0 : locale.widthFitContainer
    }, /*#__PURE__*/_createElement("span", null, /*#__PURE__*/_createElement(AdaptNormal, {
      style: {
        fontSize: '20px',
        color: '#171A1D'
      }
    }))));
  };

  return TableWidthFitContainerButton;
}(React.Component);

export { TableWidthFitContainerButton as default };
