// @ali/4ever-plugin-inline-code/dist-node/mo/serializer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createSerializer;
exports.createServerSerializationPlugin = createServerSerializationPlugin;

var _jsonMLToValue = _interopRequireDefault(require("./jsonMLToValue"));

var _valueToJsonML = _interopRequireDefault(require("./valueToJsonML"));

var _markdownToHTML = _interopRequireDefault(require("./markdownToHTML"));

var _htmlToMarkdown = _interopRequireDefault(require("./htmlToMarkdown"));

function createSerializer() {
  // inlineCode灰度过程中：
  // 关闭所有新InlineCode入口:编辑、粘贴html、粘贴markdown
  // 但无论新旧，均需具有新InlineCode模型的解析能力:故保留jsonMLToValue&valueToJsonML
  // 关闭新inlineCode的粘贴html入口:去除htmlToJsonML
  // 保留markdownToHTML：新旧InlineCode的markdown解析格式是一致的，复用即可
  return {
    jsonMLToValue: _jsonMLToValue.default,
    valueToJsonML: _valueToJsonML.default,
    //TODO inline-code全量后需要将该入口打开
    // htmlToJsonML,
    markdownToHTML: _markdownToHTML.default,
    htmlToMarkdown: _htmlToMarkdown.default
  };
}

function createServerSerializationPlugin() {
  return {
    valueToJsonML: _valueToJsonML.default,
    jsonMLToValue: _jsonMLToValue.default
  };
}
