// @ali/4ever-plugin-tag/dist-web/mo/serializer/jsonMLToHTML/index.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { getTagName, getAttributes } from '@ali/4ever-utils';
export default function createJsonMLToHTMLFactory(plugins) {
  return function createJsonMLToHTML(configs) {
    return {
      name: 'tag',
      match: function match(node) {
        var tag = getTagName(node);
        var attrs = getAttributes(node);
        return tag === 'tag' && !!attrs && attrs.tagType;
      },
      convert: function convert(node) {
        // @ts-ignore
        var attrs = getAttributes(node) || {};
        var _ref = attrs,
            metadata = _ref.metadata,
            tagType = _ref.tagType;

        var _ref2 = metadata || {
          imTitle: ''
        },
            imTitle = _ref2.imTitle;

        var targetConvert = plugins[tagType];
        if (targetConvert) return targetConvert(attrs, configs);
        return /*#__PURE__*/_createElement(React.Fragment, null, /*#__PURE__*/_createElement("span", {
          "data-type": "tag"
        }, imTitle));
      }
    };
  };
}
