// @ali/4ever-zhi/dist-node/plugins/audio/createRenderNode.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderNode;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _audio = _interopRequireDefault(require("./components/audio"));

const _createElement = /*#__PURE__*/React.createElement;
const {
  EmbedFailedCard
} = _everBamboo.EmbedPlugin;

function createRenderNode(config) {
  const audioConfig = config.audio || {};
  const {
    locale: audioLocaleConfig = {},
    downloadFile,
    transformAudioURL,
    onMediaLoadSuccess,
    onMediaLoadError,
    allowDownload
  } = audioConfig;
  return {
    embed: (props, controller, next) => {
      const {
        node
      } = props;

      if (node.data.type === 'audio' && node.data.src) {
        return /*#__PURE__*/_createElement(_audio.default, (0, _extends2.default)({}, props, {
          controller: controller,
          onDownload: downloadFile,
          locale: audioLocaleConfig,
          transformAudioURL: transformAudioURL,
          onMediaLoadSuccess: onMediaLoadSuccess,
          onMediaLoadError: onMediaLoadError,
          allowDownload: allowDownload
        }));
      }

      if (node.data.type === 'audio' && node.data.errorType) {
        return /*#__PURE__*/_createElement(EmbedFailedCard, {
          node: node,
          locale: audioLocaleConfig
        });
      }

      return next();
    }
  };
}
