// @ali/4ever-component/dist-node/ToolbarNext/basicComponents/buttons/fontSizeMaxSelect/FontSizeButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _style = require("./style");

const _createElement = /*#__PURE__*/_react.default.createElement;

const FontSizeButton = ({
  option,
  value,
  minFontSize,
  maxFontSize,
  defaultFontFamily
}) => {
  return /*#__PURE__*/_createElement(_style.GalleryButton, {
    key: option.type,
    value: option.type,
    testid: option.type,
    active: option.type === value,
    className: (0, _classnames.default)({
      active: option.type === value
    })
  }, /*#__PURE__*/_createElement(_style.TitleWrap, {
    type: option.type,
    style: {
      fontSize: `${Math.min(Math.max(option.default * 0.75, minFontSize), maxFontSize)}pt`,
      fontWeight: option.fontWeight,
      lineHeight: '29px',
      height: '29px',
      textOverflow: 'clip',
      maxWidth: '65px',
      overflow: 'hidden'
    }
  }, option.title), /*#__PURE__*/_createElement(_style.DesWrap, null, /*#__PURE__*/_createElement(_style.TextWrap, null, option.title === 'AaBbCcDdEe' ? option.des : `${defaultFontFamily} ${option.type === 'blockquote' ? '-' : option.default}`)));
};

var _default = FontSizeButton;
exports.default = _default;
