// @ali/4ever-component/dist-node/ToolbarNext/basicComponents/buttons/fontSizeMaxSelect/style.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TabBarWrapper = exports.TabBarTab = exports.TabsWrap = exports.NextButton = exports.PrevButton = exports.Nav = exports.NavWrap = exports.NavContainer = exports.TabPaneWrap = exports.TableStyleSet = exports.IconWrapper = exports.ContainerMenuDown = exports.TextWrap = exports.DesWrap = exports.TitleWrap = exports.TableButton = exports.GalleryButton = exports.GalleryWrap = exports.DisableWrap = void 0;

var _styledComponents = _interopRequireWildcard(require("styled-components"));

const DisableWrap = /*#__PURE__*/_styledComponents.default.div(["display:flex;", ""], ({
  disabled
}) => disabled && (0, _styledComponents.css)(["opacity:0.35;position:relative;&:before{content:'';position:absolute;top:0;left:0;right:0;bottom:0;z-index:1;cursor:not-allowed;}"]));

exports.DisableWrap = DisableWrap;

const GalleryWrap = /*#__PURE__*/_styledComponents.default.div(["display:flex;flex-direction:column;align-items:center;border:1px solid rgba(17,31,44,0.12);box-sizing:border-box;border-radius:4px;user-select:none;"]);

exports.GalleryWrap = GalleryWrap;

const GalleryButton = /*#__PURE__*/_styledComponents.default.div(["flex-direction:column;display:flex;height:44px;width:70px;cursor:default;align-items:center;justify-content:space-between;&.active{border-radius:4px;background-color:rgba(25,31,37,0.12);}", ""], ({
  active
}) => (0, _styledComponents.css)([":hover{border-radius:4px;background-color:", ";}:active{border-radius:4px;background-color:rgba(25,31,37,0.16);}"], active ? 'rgba(25,31,37,0.16)' : 'rgba(25,31,37,0.08)'));

exports.GalleryButton = GalleryButton;

const TableButton = /*#__PURE__*/_styledComponents.default.div(["flex-direction:column;cursor:default;display:flex;height:44px;margin-top:10px;width:70px;align-items:center;justify-content:space-between;margin-right:5px;&.active{border-radius:4px;background-color:rgba(25,31,37,0.12);}", ""], ({
  active
}) => (0, _styledComponents.css)([":hover{border-radius:4px;background-color:", ";}:active{border-radius:4px;background-color:rgba(25,31,37,0.16);}"], active ? 'rgba(25,31,37,0.16)' : 'rgba(25,31,37,0.08)'));

exports.TableButton = TableButton;

const TitleWrap = /*#__PURE__*/_styledComponents.default.span(["text-overflow:ellipsis;text-align:center;color:#171A1D;opacity:", ";white-space:nowrap;display:flex;align-items:center;", ";"], ({
  type
}) => type === 'blockquote' ? 0.4 : 1, ({
  type
}) => type === 'blockquote' && (0, _styledComponents.css)(["&:before{content:'';margin-right:2px;border-left:1px solid #171A1D;height:45%;display:inline-block;line-height:29px;vertical-align:middle;}"]));

exports.TitleWrap = TitleWrap;

const DesWrap = /*#__PURE__*/_styledComponents.default.span(["text-overflow:ellipsis;width:68px;color:rgba(17,31,44,0.56);height:14px;line-height:14px;text-align:center;font-size:12px;white-space:nowrap;margin-bottom:1px;background:#f6f6f7;border-radius:3px;"]);

exports.DesWrap = DesWrap;

const TextWrap = /*#__PURE__*/_styledComponents.default.div(["font-size:10px;transform:scale(0.9);"]);

exports.TextWrap = TextWrap;

const ContainerMenuDown = /*#__PURE__*/_styledComponents.default.div(["position:relative;display:flex;justify-content:center;"]);

exports.ContainerMenuDown = ContainerMenuDown;

const IconWrapper = /*#__PURE__*/_styledComponents.default.div(["width:34px;height:15px;cursor:default;display:flex;position:absolute;left:", ";justify-content:center;align-items:center;box-sizing:content-box;", ";&:hover{border:1px solid transparent;border-bottom-right-radius:4px;border-bottom-left-radius:4px;border-top:0;background-color:#e8e8eb;color:rgba(58,60,63,1);}&:active{border:1px solid transparent;border-bottom-right-radius:4px;border-bottom-left-radius:4px;border-top:0;background-color:#d5d6d9;}"], p => p.left, ({
  active
}) => active && (0, _styledComponents.css)(["background-color:rgba(255,255,255,1);border:1px solid rgba(17,31,44,0.12);border-top:0;border-bottom-right-radius:4px;border-bottom-left-radius:4px;"]));

exports.IconWrapper = IconWrapper;

const TableStyleSet = /*#__PURE__*/_styledComponents.default.div(["padding:0 5px 12px 12px;width:300px;display:flex;flex-wrap:wrap;box-sizing:content-box;background-color:white;border:1px solid rgba(17,31,44,0.08);box-shadow:1px 3px 8px 0 rgba(0,0,0,0.1);border-radius:6px;user-select:none;"]);

exports.TableStyleSet = TableStyleSet;

const TabPaneWrap = /*#__PURE__*/_styledComponents.default.div(["flex-shrink:0;width:100%;outline:none;"]);

exports.TabPaneWrap = TabPaneWrap;

const NavContainer = /*#__PURE__*/_styledComponents.default.div(["position:relative;display:flex;font-size:14px;line-height:1.5715;white-space:nowrap;transition:padding 0.3s cubic-bezier(0.645,0.045,0.355,1);flex:1;min-width:0;font-size:14px;padding:3px 17px;"]);

exports.NavContainer = NavContainer;

const NavWrap = /*#__PURE__*/_styledComponents.default.div(["flex:1 1 auto;align-self:stretch;overflow:hidden;white-space:nowrap;transform:translate(0);"]);

exports.NavWrap = NavWrap;

const Nav = /*#__PURE__*/_styledComponents.default.div(["position:relative;"]);

exports.Nav = Nav;
const PrevNextButton = /*#__PURE__*/(0, _styledComponents.css)(["display:flex;justify-content:center;align-items:center;position:absolute;z-index:500;width:0;height:100%;color:rgba(17,31,44,0.4);cursor:pointer;opacity:0;transition:width 0.3s cubic-bezier(0.645,0.045,0.355,1) 0s,opacity 0.3s cubic-bezier(0.645,0.045,0.355,1) 0s,color 0.3s cubic-bezier(0.645,0.045,0.355,1) 0s;user-select:none;pointer-events:none;&:hover{color:rgba(17,31,44,0.72);}&.disabled{color:rgba(17,31,44,0.24);cursor:not-allowed;}"]);

const PrevButton = /*#__PURE__*/_styledComponents.default.div(["left:0;", " &.show{opacity:1;pointer-events:auto;width:16px;height:36px;margin:4px 0 4px 0;cursor:default;&:hover{border-top-right-radius:4px;border-bottom-right-radius:4px;background-color:rgba(17,31,44,0.08);}&:active{border-top-right-radius:4px;border-bottom-right-radius:4px;background-color:rgba(17,31,44,0.12);}}"], PrevNextButton);

exports.PrevButton = PrevButton;

const NextButton = /*#__PURE__*/_styledComponents.default.div(["right:0;", " &.show{opacity:1;pointer-events:auto;width:16px;height:36px;margin:4px 0 4px 0;cursor:default;&:hover{border-top-left-radius:4px;border-bottom-left-radius:4px;background-color:rgba(17,31,44,0.08);}&:active{border-top-left-radius:4px;border-bottom-left-radius:4px;background-color:rgba(17,31,44,0.12);}}"], PrevNextButton);

exports.NextButton = NextButton;

const TabsWrap = /*#__PURE__*/_styledComponents.default.div(["transition:transform 0.3s ease-in-out,width 0.2s ease-in-out,left 0.3s ease-in-out;overflow:hidden;"]);

exports.TabsWrap = TabsWrap;

const TabBarTab = /*#__PURE__*/_styledComponents.default.div(["outline:none;position:relative;display:inline-block;height:100%;text-decoration:none;cursor:pointer;transition:color 0.3s cubic-bezier(0.645,0.045,0.355,1);color:rgba(17,31,44,1);user-select:none;margin:0;cursor:default;padding-right:3px;padding-left:3px;&.actived{font-weight:500;}&.disabled{color:rgba(17,31,44,0.24);cursor:not-allowed;}"]);

exports.TabBarTab = TabBarTab;

const TabBarWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;"]);

exports.TabBarWrapper = TabBarWrapper;
