// @ali/4ever-component/dist-node/ToolbarNext/basicComponents/buttons/fontSizeMaxSelect/TabPane.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TabPane = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _style = require("./style");

var _utils = require("./utils");

const _createElement = /*#__PURE__*/_react.default.createElement;

const TabPane = props => {
  const {
    className,
    actived,
    forceRender,
    key,
    id,
    children,
    ...restProps
  } = props; // 此缓存是为了让最近显示过的tab项不销毁

  const isActived = (0, _react.useRef)(!!actived);
  isActived.current = isActived.current || !!actived;
  const shouldRender = isActived.current || forceRender;
  const tabKeyExists = key && String(key).length > 0;
  const uuid = tabKeyExists && (id ? `${key}-${id}` : `${key}`);
  return /*#__PURE__*/_createElement(_style.TabPaneWrap, (0, _extends2.default)({
    role: "tabpanel",
    "aria-hidden": actived ? 'false' : 'true',
    tabIndex: actived ? 0 : -1,
    className: (0, _classnames.default)(className, {
      actived
    }),
    id: uuid && `tabpane-${uuid}`,
    "aria-labelledby": uuid && `tab-${uuid}`
  }, (0, _utils.getDataAttr)(restProps)), shouldRender ? children : null);
};

exports.TabPane = TabPane;
TabPane.defaultProps = {
  disabled: false
};
