// @ali/4ever-plugin-image/dist-web/utils/utils/getImageType.js
export var ImageTypeEnum; // (List_of_file_signatures)[https://en.wikipedia.org/wiki/List_of_file_signatures]

(function (ImageTypeEnum) {
  ImageTypeEnum["GIF"] = "gif";
  ImageTypeEnum["JPEG"] = "jpeg";
  ImageTypeEnum["PNG"] = "png";
  ImageTypeEnum["WEBP"] = "webp";
  ImageTypeEnum["TIFF"] = "tiff";
  ImageTypeEnum["TIF"] = "tif";
  ImageTypeEnum["SVG"] = "svg";
  ImageTypeEnum["UNCERTAIN"] = "uncertain_type";
})(ImageTypeEnum || (ImageTypeEnum = {}));

function getTypeFromMagicNumber(signature) {
  switch (signature) {
    case '47494638':
      return ImageTypeEnum.GIF;

    case 'FFD8FFDB':
    case 'FFD8FFE0':
    case 'FFD8FFE1':
    case 'FFD8FFEE':
      return ImageTypeEnum.JPEG;

    case '89504E47':
      return ImageTypeEnum.PNG;

    case '52494646':
      return ImageTypeEnum.WEBP;

    case '4D4D002A':
      return ImageTypeEnum.TIFF;
    // // TIFF, big-endian type

    case '49492A00':
      return ImageTypeEnum.TIF;
    // TIFF, little-endian type

    default:
      return ImageTypeEnum.UNCERTAIN;
  }
}

export var getImageTypeFromBuffer = function getImageTypeFromBuffer(buffer) {
  var uint = new Uint8Array(buffer);
  var bytes = [];
  uint.forEach(function (_byte) {
    // byte = 0 ~ 15 时，其十六进制数前需补0
    var _char = ("0" + _byte.toString(16)).slice(-2);

    bytes.push(_char);
  });
  var magicNumber = bytes.join('').toUpperCase();
  var typeFromMagicNumber = getTypeFromMagicNumber(magicNumber);
  return typeFromMagicNumber;
};
export var getImageTypeFromBlob = function getImageTypeFromBlob(blob) {
  return new Promise(function (resolve) {
    var subBlob = blob.slice(0, 4);
    var reader = new FileReader();

    reader.onload = function (event) {
      var fileStart = event.target.result;
      var type = getImageTypeFromBuffer(fileStart);
      resolve(type);
    };

    reader.onerror = function () {
      resolve(ImageTypeEnum.UNCERTAIN);
    };

    reader.readAsArrayBuffer(subBlob);
  });
};
