// @ali/4ever-plugin-subject/dist-node/utils/components/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InnerCardWrapperForMention = exports.InnerCardWrapper = exports.InnerTextWrapper = exports.SubjectPlaceholderWrapper = exports.DingWrapper = exports.InlinePlaceholder = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _weDesign = require("@ali/we-design");

const InlinePlaceholder = /*#__PURE__*/_styledComponents.default.span(["color:rgba(23,26,29,0.4);pointer-events:none;"]);

exports.InlinePlaceholder = InlinePlaceholder;

const DingWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;align-items:center;justify-content:center;color:", ";width:24px;height:24px;padding:2px;z-index:9;cursor:pointer;border-radius:4px;&:hover{background:rgba(126,134,142,0.12);}"], _weDesign.theme.commonBaseColor3);

exports.DingWrapper = DingWrapper;

const SubjectPlaceholderWrapper = /*#__PURE__*/_styledComponents.default.div(["position:absolute;top:0;display:flex;flex-direction:row;height:100%;pointer-events:none;align-items:center;color:#171A1D;"]);

exports.SubjectPlaceholderWrapper = SubjectPlaceholderWrapper;

const InnerTextWrapper = /*#__PURE__*/_styledComponents.default.span(["display:inline-flex;color:rgba(23,26,29,0.4);left:2px;align-items:center;top:0;bottom:0;z-index:1;white-space:nowrap;pointer-events:none;"]); // 引入位置五花八门，需要 important 以规范在 placeholder 中的样式


exports.InnerTextWrapper = InnerTextWrapper;

const InnerCardWrapper = /*#__PURE__*/_styledComponents.default.span(["pointer-events:auto;div{color:rgb(110,110,110);background:#F5F6F7;font-size:14px;}span{color:rgb(110,110,110);}:hover div{background:rgb(240,240,241);}div:hover{background:rgb(240,240,241);}.weicon svg{font-size:18px;}.inline-placeholder-icon-box .weicon svg{fill:rgb(110,110,110);}.inline-placeholder-icon-box{font-size:15px;height:15px;}"]);

exports.InnerCardWrapper = InnerCardWrapper;

const InnerCardWrapperForMention = /*#__PURE__*/_styledComponents.default.span(["border-radius:4px;margin:0 3px;padding:0 2px;pointer-events:auto;color:rgb(110,110,110);background:#F5F6F7;:hover{background:rgb(240,240,241);}"]);

exports.InnerCardWrapperForMention = InnerCardWrapperForMention;
