// @ali/4ever-bi/dist-node/plugins/hr/commands/insertHr.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

var _everUtils = require("@ali/4ever-utils");

var _updateHr = _interopRequireDefault(require("./updateHr"));

var _constants = require("../constants");

var _utils = require("../utils");

var _default = (controller, payload = {}) => {
  const {
    value
  } = controller;
  const {
    focusBlock
  } = value;

  if (focusBlock?.type === 'hr') {
    return (0, _updateHr.default)(controller, focusBlock, payload);
  }

  const {
    color,
    type,
    sz
  } = payload;
  const hrColor = color ?? _constants.DEFAULT_HR_COLOR;
  controller.command(_everCangjie.Commands.insertBlock, {
    type: 'hr',
    data: {
      color: hrColor,
      type,
      width: '100%',
      // 默认占满一行
      sz: sz || 2,
      // size 默认为 2
      id: (0, _everUtils.generateKey)()
    }
  }).command(_everCangjie.Commands.insertBlock, 'paragraph'); // 插入前是空行则删除

  if ((0, _utils.shouldDeleteParagraph)(focusBlock)) {
    controller.command(_everCangjie.Commands.removeNodeByKey, focusBlock.key);
  }

  return controller;
};

exports.default = _default;
