// @ali/4ever-plugin-table/dist-web/bi/actions/index.js
import { setTableMinColsWidthData, adjustColsWidth, isTableSetTableMinColsWidthDataAction, isTableAdjustColsWidthAction, TABLE_ADJUST_COLS_WIDTH, TABLE_SET_MIN_COLS_WIDTH_DATA } from "../../utils/actions";
import { clickTableToolbar, isTableClickTableToolbarAction, TABLE_CLICK_TOOLBAR } from "./clickTableToolbar";
import { deleteTable, isTableDeleteTableAction, TABLE_DELETE_TABLE } from "./deleteTable";
import { deleteTableCols, isTableDeleteColsAction, TABLE_DELETE_COLS } from "./deleteTableCols";
import { deleteTableRows, isTableDeleteRowsAction, TABLE_DELETE_ROWS } from "./deleteTableRows";
import { emptyTableSelection, isTableEmptySelectionAction, TABLE_EMPTY_SELECTION } from "./emptyTableSelection";
import { insertTableRowBySelections, isTableInsertRowBySelectionsAction, TABLE_INSERT_ROW_BY_SELECTIONS } from "./inserTableRowsBySelections";
import { insertTable, isTableInsertTableAction, TABLE_INSERT_TABLE } from "./insertTable";
import { insertTableCol, isTableInsertColAction, TABLE_INSERT_COL } from "./insertTableCol";
import { insertTableRow, isTableInsertRowAction, TABLE_INSERT_ROW } from "./insertTableRow";
import { mergeTableCells, isTableMergeCellsAction, TABLE_MERGE_CELLS } from "./mergeTableCells";
import { moveTableCols, isTableMoveColsAction, TABLE_MOVE_COLS } from "./moveTableCols";
import { moveTableRows, isTableMoveRowsAction, TABLE_MOVE_ROWS } from "./moveTableRows";
import { removeTableColHeader, isTableRemoveTableColHeaderAction, TABLE_REMOVE_COL_HEADER } from "./removeTableColHeader";
import { removeTableRowHeader, isTableRemoveTableRowHeaderAction, TABLE_REMOVE_ROW_HEADER } from "./removeTableRowHeader";
import { resizeTableColWidth, isTableResizeTableColWidthAction, TABLE_RESIZE_COL_WIDTH } from "./resizeTableColWidth";
import { resizeTableHeight, isTableResizeTableHeightAction, TABLE_RESIZE_HEIGHT } from "./resizeTableHeight";
import { resizeTableWidth, isTableResizeTableWidthAction, TABLE_RESIZE_WIDTH } from "./resizeTableWidth";
import { resizeTableWidthFitContainer, isTableResizeTableWidthFitContainerAction, TABLE_RESIZE_WIDTH_FIT_CONTAINER } from "./resizeTableWidthFitContainer";
import { scrollTable, isTableScrollTableAction, TABLE_SCROLL } from "./scrollTable";
import { selectTable, isTableSelectTableAction, TABLE_SELECT_TABLE } from "./selectTable";
import { selectTableCols, isTableSelectTableColsAction, TABLE_SELECT_COLS } from "./selectTableCols";
import { selectTableRows, isTableSelectTableRowsAction, TABLE_SELECT_ROWS } from "./selectTableRows";
import { selectTableSelection, isTableSelectTableSelectionAction, TABLE_SELECT_SELECTIONS } from "./selectTableSelection";
import { setTableAutofitWidthDisable, isTableSetTableAutofitWidthDisableAction, TABLE_DISABLE_AUTOFIT_WIDTH } from "./setTableAutofitWidthDisable";
import { setTableAutofitWidthEnable, isTablesetTableAutofitWidthEnableAction, TABLE_ENABLE_AUTOFIT_WIDTH } from "./setTableAutofitWidthEnable";
import { setTableBorder, isTableSetTableBorderAction, TABLE_SET_BORDER } from "./setTableBorder";
import { setTableCellFill, isTableSetTableCellFillAction, TABLE_SET_CELL_FILL } from "./setTableCellFill";
import { setTableCellVAlign, isTableSetTableCellVAlignAction, TABLE_SET_CELL_VALIGN } from "./setTableCellVAlign";
import { setTableColHeader, isTableSetTableColHeaderAction, TABLE_SET_COL_HEADER } from "./setTableColHeader";
import { setTableRowHeader, isTableSetTableRowHeaderAction, TABLE_SET_ROW_HEADER } from "./setTableRowHeader";
import { setTableRowHeight, isTableSetTableRowHeightAction, TABLE_SET_ROW_HEIGHT } from "./setTableRowHeight";
import { splitTableCell, isTableSplitTableCellAction, TABLE_SPLIT_CELL } from "./splitTableCell";
import { tableBlur, isTableBlurAction, TABLE_BLUR } from "./tableBlur";
export { // common
setTableMinColsWidthData, adjustColsWidth // bi
, clickTableToolbar, deleteTable, deleteTableCols, deleteTableRows, emptyTableSelection, insertTableRowBySelections, insertTable, insertTableCol, insertTableRow, mergeTableCells, moveTableCols, moveTableRows, removeTableColHeader, removeTableRowHeader, resizeTableColWidth, resizeTableHeight, resizeTableWidth, resizeTableWidthFitContainer, scrollTable, selectTable, selectTableCols, selectTableRows, selectTableSelection, setTableAutofitWidthDisable, setTableAutofitWidthEnable, setTableBorder, setTableCellFill, setTableCellVAlign, setTableColHeader, setTableRowHeader, setTableRowHeight, splitTableCell, tableBlur };
export var actions = {
  // common
  setTableMinColsWidthData: setTableMinColsWidthData,
  adjustColsWidth: adjustColsWidth,
  // bi
  clickTableToolbar: clickTableToolbar,
  deleteTable: deleteTable,
  deleteTableCols: deleteTableCols,
  deleteTableRows: deleteTableRows,
  emptyTableSelection: emptyTableSelection,
  insertTableRowBySelections: insertTableRowBySelections,
  insertTable: insertTable,
  insertTableCol: insertTableCol,
  insertTableRow: insertTableRow,
  mergeTableCells: mergeTableCells,
  moveTableCols: moveTableCols,
  moveTableRows: moveTableRows,
  removeTableColHeader: removeTableColHeader,
  removeTableRowHeader: removeTableRowHeader,
  resizeTableColWidth: resizeTableColWidth,
  resizeTableHeight: resizeTableHeight,
  resizeTableWidth: resizeTableWidth,
  resizeTableWidthFitContainer: resizeTableWidthFitContainer,
  scrollTable: scrollTable,
  selectTable: selectTable,
  selectTableCols: selectTableCols,
  selectTableRows: selectTableRows,
  selectTableSelection: selectTableSelection,
  setTableAutofitWidthDisable: setTableAutofitWidthDisable,
  setTableAutofitWidthEnable: setTableAutofitWidthEnable,
  setTableBorder: setTableBorder,
  setTableCellFill: setTableCellFill,
  setTableCellVAlign: setTableCellVAlign,
  setTableColHeader: setTableColHeader,
  setTableRowHeader: setTableRowHeader,
  setTableRowHeight: setTableRowHeight,
  splitTableCell: splitTableCell,
  tableBlur: tableBlur
};
export var isActions = {
  // common
  isTableSetTableMinColsWidthDataAction: isTableSetTableMinColsWidthDataAction,
  isTableAdjustColsWidthAction: isTableAdjustColsWidthAction,
  // bi
  isTableClickTableToolbarAction: isTableClickTableToolbarAction,
  isTableDeleteTableAction: isTableDeleteTableAction,
  isTableDeleteColsAction: isTableDeleteColsAction,
  isTableDeleteRowsAction: isTableDeleteRowsAction,
  isTableEmptySelectionAction: isTableEmptySelectionAction,
  isTableInsertRowBySelectionsAction: isTableInsertRowBySelectionsAction,
  isTableInsertTableAction: isTableInsertTableAction,
  isTableInsertColAction: isTableInsertColAction,
  isTableInsertRowAction: isTableInsertRowAction,
  isTableMergeCellsAction: isTableMergeCellsAction,
  isTableMoveColsAction: isTableMoveColsAction,
  isTableMoveRowsAction: isTableMoveRowsAction,
  isTableRemoveTableColHeaderAction: isTableRemoveTableColHeaderAction,
  isTableRemoveTableRowHeaderAction: isTableRemoveTableRowHeaderAction,
  isTableResizeTableColWidthAction: isTableResizeTableColWidthAction,
  isTableResizeTableHeightAction: isTableResizeTableHeightAction,
  isTableResizeTableWidthAction: isTableResizeTableWidthAction,
  isTableResizeTableWidthFitContainerAction: isTableResizeTableWidthFitContainerAction,
  isTableScrollTableAction: isTableScrollTableAction,
  isTableSelectTableAction: isTableSelectTableAction,
  isTableSelectTableColsAction: isTableSelectTableColsAction,
  isTableSelectTableRowsAction: isTableSelectTableRowsAction,
  isTableSelectTableSelectionAction: isTableSelectTableSelectionAction,
  isTableSetTableAutofitWidthDisableAction: isTableSetTableAutofitWidthDisableAction,
  isTablesetTableAutofitWidthEnableAction: isTablesetTableAutofitWidthEnableAction,
  isTableSetTableBorderAction: isTableSetTableBorderAction,
  isTableSetTableCellFillAction: isTableSetTableCellFillAction,
  isTableSetTableCellVAlignAction: isTableSetTableCellVAlignAction,
  isTableSetTableColHeaderAction: isTableSetTableColHeaderAction,
  isTableSetTableRowHeaderAction: isTableSetTableRowHeaderAction,
  isTableSetTableRowHeightAction: isTableSetTableRowHeightAction,
  isTableSplitTableCellAction: isTableSplitTableCellAction,
  isTableBlurAction: isTableBlurAction
};
export var actionTypes = {
  // common
  TABLE_ADJUST_COLS_WIDTH: TABLE_ADJUST_COLS_WIDTH,
  TABLE_SET_MIN_COLS_WIDTH_DATA: TABLE_SET_MIN_COLS_WIDTH_DATA,
  // bi
  TABLE_BLUR: TABLE_BLUR,
  TABLE_CLICK_TOOLBAR: TABLE_CLICK_TOOLBAR,
  TABLE_DELETE_COLS: TABLE_DELETE_COLS,
  TABLE_DELETE_ROWS: TABLE_DELETE_ROWS,
  TABLE_DELETE_TABLE: TABLE_DELETE_TABLE,
  TABLE_DISABLE_AUTOFIT_WIDTH: TABLE_DISABLE_AUTOFIT_WIDTH,
  TABLE_EMPTY_SELECTION: TABLE_EMPTY_SELECTION,
  TABLE_ENABLE_AUTOFIT_WIDTH: TABLE_ENABLE_AUTOFIT_WIDTH,
  TABLE_INSERT_COL: TABLE_INSERT_COL,
  TABLE_INSERT_ROW: TABLE_INSERT_ROW,
  TABLE_INSERT_ROW_BY_SELECTIONS: TABLE_INSERT_ROW_BY_SELECTIONS,
  TABLE_INSERT_TABLE: TABLE_INSERT_TABLE,
  TABLE_MERGE_CELLS: TABLE_MERGE_CELLS,
  TABLE_MOVE_COLS: TABLE_MOVE_COLS,
  TABLE_MOVE_ROWS: TABLE_MOVE_ROWS,
  TABLE_REMOVE_COL_HEADER: TABLE_REMOVE_COL_HEADER,
  TABLE_REMOVE_ROW_HEADER: TABLE_REMOVE_ROW_HEADER,
  TABLE_RESIZE_COL_WIDTH: TABLE_RESIZE_COL_WIDTH,
  TABLE_RESIZE_HEIGHT: TABLE_RESIZE_HEIGHT,
  TABLE_RESIZE_WIDTH: TABLE_RESIZE_WIDTH,
  TABLE_RESIZE_WIDTH_FIT_CONTAINER: TABLE_RESIZE_WIDTH_FIT_CONTAINER,
  TABLE_SCROLL: TABLE_SCROLL,
  TABLE_SELECT_COLS: TABLE_SELECT_COLS,
  TABLE_SELECT_ROWS: TABLE_SELECT_ROWS,
  TABLE_SELECT_SELECTIONS: TABLE_SELECT_SELECTIONS,
  TABLE_SELECT_TABLE: TABLE_SELECT_TABLE,
  TABLE_SET_BORDER: TABLE_SET_BORDER,
  TABLE_SET_CELL_FILL: TABLE_SET_CELL_FILL,
  TABLE_SET_CELL_VALIGN: TABLE_SET_CELL_VALIGN,
  TABLE_SET_COL_HEADER: TABLE_SET_COL_HEADER,
  TABLE_SET_ROW_HEADER: TABLE_SET_ROW_HEADER,
  TABLE_SET_ROW_HEIGHT: TABLE_SET_ROW_HEIGHT,
  TABLE_SPLIT_CELL: TABLE_SPLIT_CELL
};
