// @ali/4ever-bi/dist-web/plugins/formula/utils/importFont.js
/* eslint-disable max-len */

/* eslint-disable camelcase */
import KaTeX_AMS_Regular_woff2 from "../fonts/KaTeX_AMS-Regular.woff2";
import KaTeX_AMS_Regular_woff from "../fonts/KaTeX_AMS-Regular.woff";
import KaTeX_AMS_Regular_ttf from "../fonts/KaTeX_AMS-Regular.ttf";
import KaTeX_Caligraphic_Bold_woff2 from "../fonts/KaTeX_Caligraphic-Bold.woff2";
import KaTeX_Caligraphic_Bold_woff from "../fonts/KaTeX_Caligraphic-Bold.woff";
import KaTeX_Caligraphic_Bold_ttf from "../fonts/KaTeX_Caligraphic-Bold.ttf";
import KaTeX_Caligraphic_Regular_woff2 from "../fonts/KaTeX_Caligraphic-Regular.woff2";
import KaTeX_Caligraphic_Regular_woff from "../fonts/KaTeX_Caligraphic-Regular.woff";
import KaTeX_Caligraphic_Regular_ttf from "../fonts/KaTeX_Caligraphic-Regular.ttf";
import KaTeX_Fraktur_Bold_woff2 from "../fonts/KaTeX_Fraktur-Bold.woff2";
import KaTeX_Fraktur_Bold_woff from "../fonts/KaTeX_Fraktur-Bold.woff";
import KaTeX_Fraktur_Bold_ttf from "../fonts/KaTeX_Fraktur-Bold.ttf";
import KaTeX_Fraktur_Regular_woff2 from "../fonts/KaTeX_Fraktur-Regular.woff2";
import KaTeX_Fraktur_Regular_woff from "../fonts/KaTeX_Fraktur-Regular.woff";
import KaTeX_Fraktur_Regular_ttf from "../fonts/KaTeX_Fraktur-Regular.ttf";
import KaTeX_Main_Bold_woff2 from "../fonts/KaTeX_Main-Bold.woff2";
import KaTeX_Main_Bold_woff from "../fonts/KaTeX_Main-Bold.woff";
import KaTeX_Main_Bold_ttf from "../fonts/KaTeX_Main-Bold.ttf";
import KaTeX_Main_BoldItalic_woff2 from "../fonts/KaTeX_Main-BoldItalic.woff2";
import KaTeX_Main_BoldItalic_woff from "../fonts/KaTeX_Main-BoldItalic.woff";
import KaTeX_Main_BoldItalic_ttf from "../fonts/KaTeX_Main-BoldItalic.ttf";
import KaTeX_Main_Italic_woff2 from "../fonts/KaTeX_Main-Italic.woff2";
import KaTeX_Main_Italic_woff from "../fonts/KaTeX_Main-Italic.woff";
import KaTeX_Main_Italic_ttf from "../fonts/KaTeX_Main-Italic.ttf";
import KaTeX_Main_Regular_woff2 from "../fonts/KaTeX_Main-Regular.woff2";
import KaTeX_Main_Regular_woff from "../fonts/KaTeX_Main-Regular.woff";
import KaTeX_Main_Regular_ttf from "../fonts/KaTeX_Main-Regular.ttf";
import KaTeX_Math_BoldItalic_woff2 from "../fonts/KaTeX_Math-BoldItalic.woff2";
import KaTeX_Math_BoldItalic_woff from "../fonts/KaTeX_Math-BoldItalic.woff";
import KaTeX_Math_BoldItalic_ttf from "../fonts/KaTeX_Math-BoldItalic.ttf";
import KaTeX_Math_Italic_woff2 from "../fonts/KaTeX_Math-Italic.woff2";
import KaTeX_Math_Italic_woff from "../fonts/KaTeX_Math-Italic.woff";
import KaTeX_Math_Italic_ttf from "../fonts/KaTeX_Math-Italic.ttf";
import KaTeX_SansSerif_Bold_woff2 from "../fonts/KaTeX_SansSerif-Bold.woff2";
import KaTeX_SansSerif_Bold_woff from "../fonts/KaTeX_SansSerif-Bold.woff";
import KaTeX_SansSerif_Bold_ttf from "../fonts/KaTeX_SansSerif-Bold.ttf";
import KaTeX_SansSerif_Italic_woff2 from "../fonts/KaTeX_SansSerif-Italic.woff2";
import KaTeX_SansSerif_Italic_woff from "../fonts/KaTeX_SansSerif-Italic.woff";
import KaTeX_SansSerif_Italic_ttf from "../fonts/KaTeX_SansSerif-Italic.ttf";
import KaTeX_SansSerif_Regular_woff2 from "../fonts/KaTeX_SansSerif-Regular.woff2";
import KaTeX_SansSerif_Regular_woff from "../fonts/KaTeX_SansSerif-Regular.woff";
import KaTeX_SansSerif_Regular_ttf from "../fonts/KaTeX_SansSerif-Regular.ttf";
import KaTeX_Script_Regular_woff2 from "../fonts/KaTeX_Script-Regular.woff2";
import KaTeX_Script_Regular_woff from "../fonts/KaTeX_Script-Regular.woff";
import KaTeX_Script_Regular_ttf from "../fonts/KaTeX_Script-Regular.ttf";
import KaTeX_Size1_Regular_woff2 from "../fonts/KaTeX_Size1-Regular.woff2";
import KaTeX_Size1_Regular_woff from "../fonts/KaTeX_Size1-Regular.woff";
import KaTeX_Size1_Regular_ttf from "../fonts/KaTeX_Size1-Regular.ttf";
import KaTeX_Size2_Regular_woff2 from "../fonts/KaTeX_Size2-Regular.woff2";
import KaTeX_Size2_Regular_woff from "../fonts/KaTeX_Size2-Regular.woff";
import KaTeX_Size2_Regular_ttf from "../fonts/KaTeX_Size2-Regular.ttf";
import KaTeX_Size3_Regular_woff2 from "../fonts/KaTeX_Size3-Regular.woff2";
import KaTeX_Size3_Regular_woff from "../fonts/KaTeX_Size3-Regular.woff";
import KaTeX_Size3_Regular_ttf from "../fonts/KaTeX_Size3-Regular.ttf";
import KaTeX_Size4_Regular_woff2 from "../fonts/KaTeX_Size4-Regular.woff2";
import KaTeX_Size4_Regular_woff from "../fonts/KaTeX_Size4-Regular.woff";
import KaTeX_Size4_Regular_ttf from "../fonts/KaTeX_Size4-Regular.ttf";
import KaTeX_Typewriter_Regular_woff2 from "../fonts/KaTeX_Typewriter-Regular.woff2";
import KaTeX_Typewriter_Regular_woff from "../fonts/KaTeX_Typewriter-Regular.woff";
import KaTeX_Typewriter_Regular_ttf from "../fonts/KaTeX_Typewriter-Regular.ttf";
export var FONTS = [{
  family: 'KaTeX_AMS',
  source: "url(" + KaTeX_AMS_Regular_woff2 + ") format('woff2'), url(" + KaTeX_AMS_Regular_woff + ") format('woff'), url(" + KaTeX_AMS_Regular_ttf + ") format('truetype')",
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Caligraphic',
  source: "url(" + KaTeX_Caligraphic_Bold_woff2 + ") format('woff2'), url(" + KaTeX_Caligraphic_Bold_woff + ") format('woff'), url(" + KaTeX_Caligraphic_Bold_ttf + ") format('truetype')",
  descriptors: {
    weight: 'bold',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Caligraphic',
  source: "url(" + KaTeX_Caligraphic_Regular_woff2 + ") format('woff2'), url(" + KaTeX_Caligraphic_Regular_woff + ") format('woff'), url(" + KaTeX_Caligraphic_Regular_ttf + ") format('truetype') ",
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Fraktur',
  source: "url(" + KaTeX_Fraktur_Bold_woff2 + ") format('woff2'), url(" + KaTeX_Fraktur_Bold_woff + ") format('woff'), url(" + KaTeX_Fraktur_Bold_ttf + ") format('truetype')",
  descriptors: {
    weight: 'bold',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Fraktur',
  source: "url(" + KaTeX_Fraktur_Regular_woff2 + ") format('woff2'), url(" + KaTeX_Fraktur_Regular_woff + ") format('woff'), url(" + KaTeX_Fraktur_Regular_ttf + ") format('truetype')",
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Main',
  source: "url(" + KaTeX_Main_Bold_woff2 + ") format('woff2'), url(" + KaTeX_Main_Bold_woff + ") format('woff'), url(" + KaTeX_Main_Bold_ttf + ") format('truetype')",
  descriptors: {
    weight: 'bold',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Main',
  source: "url(" + KaTeX_Main_BoldItalic_woff2 + ") format('woff2'), url(" + KaTeX_Main_BoldItalic_woff + ") format('woff'), url(" + KaTeX_Main_BoldItalic_ttf + ") format('truetype')",
  descriptors: {
    weight: 'bold',
    style: 'italic'
  }
}, {
  family: 'KaTeX_Main',
  source: "url(" + KaTeX_Main_Italic_woff2 + ") format('woff2'), url(" + KaTeX_Main_Italic_woff + ") format('woff'), url(" + KaTeX_Main_Italic_ttf + ") format('truetype')",
  descriptors: {
    weight: 'normal',
    style: 'italic'
  }
}, {
  family: 'KaTeX_Main',
  source: "url(" + KaTeX_Main_Regular_woff2 + ") format('woff2'), url(" + KaTeX_Main_Regular_woff + ") format('woff'), url(" + KaTeX_Main_Regular_ttf + ") format('truetype')",
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Math',
  source: "url(" + KaTeX_Math_BoldItalic_woff2 + ") format('woff2'), url(" + KaTeX_Math_BoldItalic_woff + ") format('woff'), url(" + KaTeX_Math_BoldItalic_ttf + ") format('truetype')",
  descriptors: {
    weight: 'bold',
    style: 'italic'
  }
}, {
  family: 'KaTeX_Math',
  source: "url(" + KaTeX_Math_Italic_woff2 + ") format('woff2'), url(" + KaTeX_Math_Italic_woff + ") format('woff'), url(" + KaTeX_Math_Italic_ttf + ") format('truetype')",
  descriptors: {
    weight: 'normal',
    style: 'italic'
  }
}, {
  family: 'KaTeX_SansSerif',
  source: "url(" + KaTeX_SansSerif_Bold_woff2 + ") format('woff2'), url(" + KaTeX_SansSerif_Bold_woff + ") format('woff'), url(" + KaTeX_SansSerif_Bold_ttf + ") format('truetype')",
  descriptors: {
    weight: 'bold',
    style: 'normal'
  }
}, {
  family: 'KaTeX_SansSerif',
  source: "url(" + KaTeX_SansSerif_Italic_woff2 + ") format('woff2'), url(" + KaTeX_SansSerif_Italic_woff + ") format('woff'), url(" + KaTeX_SansSerif_Italic_ttf + ") format('truetype')",
  descriptors: {
    weight: 'normal',
    style: 'italic'
  }
}, {
  family: 'KaTeX_SansSerif',
  source: "url(" + KaTeX_SansSerif_Regular_woff2 + ") format('woff2'), url(" + KaTeX_SansSerif_Regular_woff + ") format('woff'), url(" + KaTeX_SansSerif_Regular_ttf + ") format('truetype')",
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Script',
  source: "url(" + KaTeX_Script_Regular_woff2 + ") format('woff2'), url(" + KaTeX_Script_Regular_woff + ") format('woff'), url(" + KaTeX_Script_Regular_ttf + ") format('truetype')",
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Size1',
  source: "url(" + KaTeX_Size1_Regular_woff2 + ") format('woff2'), url(" + KaTeX_Size1_Regular_woff + ") format('woff'), url(" + KaTeX_Size1_Regular_ttf + ") format('truetype')",
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Size2',
  source: "url(" + KaTeX_Size2_Regular_woff2 + ") format('woff2'), url(" + KaTeX_Size2_Regular_woff + ") format('woff'), url(" + KaTeX_Size2_Regular_ttf + ") format('truetype')",
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Size3',
  source: "url(" + KaTeX_Size3_Regular_woff2 + ") format('woff2'), url(" + KaTeX_Size3_Regular_woff + ") format('woff'), url(" + KaTeX_Size3_Regular_ttf + ") format('truetype')",
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Size4',
  source: "url(" + KaTeX_Size4_Regular_woff2 + ") format('woff2'), url(" + KaTeX_Size4_Regular_woff + ") format('woff'), url(" + KaTeX_Size4_Regular_ttf + ") format('truetype')",
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Typewriter',
  source: "url(" + KaTeX_Typewriter_Regular_woff2 + ") format('woff2'), url(" + KaTeX_Typewriter_Regular_woff + ") format('woff'), url(" + KaTeX_Typewriter_Regular_ttf + ") format('truetype')",
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}];

function importFont() {
  var styleElement = document.createElement('style');
  document.head.appendChild(styleElement);
  var styleSheet = styleElement.sheet;
  return function insertRule(style) {
    if (!styleSheet) {
      return;
    }

    if (typeof style === 'string') {
      styleSheet.insertRule(style, 0);
      return;
    }

    if (typeof style === 'object') {
      var family = style.family,
          source = style.source,
          descriptors = style.descriptors;
      var rule = "\n        @font-face {\n          font-family: " + family + ";\n          font-style: " + descriptors.style + ";\n          font-weight: " + descriptors.weight + ";\n          src: " + source + ";\n        }\n      ";
      styleSheet.insertRule(rule, 0);
    }
  };
}

export var dynamicalFontImport = importFont();
