// @ali/we-design-next/dist-web/form/checkbox/Checkbox.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { CheckboxIcon } from "./CheckboxIcon";
import { CheckboxWrap, CheckboxInput, CheckboxText } from "./styled";
import { useDefaultProp } from "../../common/react/useDefaultProp";
export var Checkbox = /*#__PURE__*/React.memo( /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _props$defaultChecked = props.defaultChecked,
    defaultChecked = _props$defaultChecked === void 0 ? false : _props$defaultChecked,
    _checked = props.checked,
    onChange = props.onChange,
    children = props.children,
    _props$indeterminate = props.indeterminate,
    indeterminate = _props$indeterminate === void 0 ? false : _props$indeterminate,
    _props$disabled = props.disabled,
    disabled = _props$disabled === void 0 ? false : _props$disabled;
  var _useDefaultProp = useDefaultProp(defaultChecked, _checked, onChange, function (e) {
      return e.target.checked;
    }),
    checked = _useDefaultProp[0],
    setChecked = _useDefaultProp[1];
  return /*#__PURE__*/_createElement(CheckboxWrap, {
    disabled: disabled,
    checked: checked
  }, /*#__PURE__*/_createElement(CheckboxInput, {
    ref: ref,
    type: "checkbox",
    onChange: setChecked,
    disabled: disabled,
    checked: checked
  }), /*#__PURE__*/_createElement(CheckboxIcon, {
    disabled: disabled,
    checked: checked,
    indeterminate: indeterminate
  }), children && /*#__PURE__*/_createElement(CheckboxText, null, children));
}));
