// @ali/4ever-plugin-comment/dist-web/utils/actions/index.js
import { createActionHandlers } from '@ali/4ever-cangjie';
export var ACTION_COMMENT_CLICK = Symbol('comment-click');
export var ACTION_COMMENT_ADD = Symbol('comment-add');
export var ACTION_COMMENT_SHOW = Symbol('comment-show');
export var ACTION_COMMENT_CANCEL = Symbol('comment-cancel');
export var ACTION_COMMENT_SET_ID = Symbol('comment-setid');
export var ACTION_COMMENT_SET_COMMENTS = Symbol('comment-comments');
export var ACTION_COMMENT_SET_SIDE_BAR = Symbol('comment-set-sidebar');
export var ACTION_COMMENT_SELECTED_NODES = Symbol('comment-selected-nodes');
/**
 * 显示评论
 */

export var showComment = function showComment(contentId, options) {
  return {
    type: ACTION_COMMENT_SHOW,
    payload: {
      contentId: contentId,
      options: options
    }
  };
};
/**
 * 隐藏评论
 */

export var hideComment = function hideComment() {
  return {
    type: ACTION_COMMENT_SHOW,
    payload: undefined
  };
};
/**
 * 添加评论
 */

export var addComment = function addComment() {
  return {
    type: ACTION_COMMENT_ADD
  };
};
/**
 * 取消正在添加的评论
 */

export var cancelComment = function cancelComment() {
  return {
    type: ACTION_COMMENT_CANCEL
  };
};
/**
 * 设置显示的 comment （showComment 会触发滚动，这里没有副作用）
 */

export var setActiveId = function setActiveId(contentId, isAutoFocus) {
  if (isAutoFocus === void 0) {
    isAutoFocus = true;
  }

  return {
    type: ACTION_COMMENT_SET_ID,
    payload: {
      contentId: contentId,
      isAutoFocus: isAutoFocus
    }
  };
};
/**
 * 设置 comments 数据
 */

export var setComments = function setComments(comments) {
  return {
    type: ACTION_COMMENT_SET_COMMENTS,
    payload: {
      comments: comments
    }
  };
};
/**
 * 设置右侧栏是否可见
 */

export var setSidebarVisible = function setSidebarVisible(visible) {
  return {
    type: ACTION_COMMENT_SET_SIDE_BAR,
    payload: {
      visible: visible
    }
  };
};

var _createActionHandlers = createActionHandlers(ACTION_COMMENT_SELECTED_NODES),
    commentSelectedNodes = _createActionHandlers[0],
    isCommentSelectedNodes = _createActionHandlers[1];

export { commentSelectedNodes, isCommentSelectedNodes };
