// @ali/we-design/dist-web/components/avatar/AvatarPopover.js
var _SplitLine;

import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { Dropdown } from "../dropdown";
import { Dot, dropdownOverlayStyle } from "./styled";
;
var PopoverWrapper = /*#__PURE__*/styled.div([""]);
var AvatarOuterWrapper = /*#__PURE__*/styled.div(["display:flex;align-items:center;"]);
var ProfileInfoWrapper = /*#__PURE__*/styled.div(["display:flex;justify-content:space-between;padding:8px 12px;"]);
var ProfileInfoInnerLeft = /*#__PURE__*/styled.div(["display:flex;align-items:center;justify-content:flex-start;margin-right:8px;"]);
var ProfileInfoInnerCenter = /*#__PURE__*/styled.div(["display:flex;min-width:105px;align-items:center;justify-content:space-between;"]);
var ProfileInfoInnerRight = /*#__PURE__*/styled.div(["display:flex;align-items:center;justify-content:flex-end;"]);
var NickText = /*#__PURE__*/styled.span(["flex:1;width:50%;max-width:96px;font-weight:400;font-size:13px;color:#171a1d;text-overflow:ellipsis;white-space:nowrap;overflow:hidden;margin-right:8px;"]);
var StatusText = /*#__PURE__*/styled.span(["font-weight:400;font-size:12px;margin-right:4px;color:rgba(23,26,29,0.4);"]);
var SplitLine = /*#__PURE__*/styled.div(["height:1px;margin:0 4px;background:rgba(126,134,142,0.16);"]);

var ProfileInfo = function ProfileInfo(props) {
  var name = props.name,
      color = props.color,
      status = props.status,
      children = props.children;
  return /*#__PURE__*/_createElement(ProfileInfoWrapper, null, /*#__PURE__*/_createElement(ProfileInfoInnerLeft, null, children), /*#__PURE__*/_createElement(ProfileInfoInnerCenter, null, /*#__PURE__*/_createElement(NickText, null, name), status && /*#__PURE__*/_createElement(StatusText, null, status)), /*#__PURE__*/_createElement(ProfileInfoInnerRight, null, color && /*#__PURE__*/_createElement(Dot, {
    color: color
  })));
};

var AvatarPopover = function AvatarPopover(props) {
  var overlay = props.overlay,
      color = props.color,
      status = props.status,
      name = props.name,
      children = props.children;
  var avatar = children && React.Children.toArray(children).map(function (child) {
    return /*#__PURE__*/React.cloneElement(child, {
      shape: 'rect',
      popoverOverlay: false,
      active: false,
      size: 'normal',
      onClick: undefined,
      color: undefined
    });
  })[0];
  return /*#__PURE__*/_createElement(Dropdown, {
    zIndex: 1001,
    overlayStyle: dropdownOverlayStyle,
    overlay: /*#__PURE__*/_createElement(PopoverWrapper, null, /*#__PURE__*/_createElement(ProfileInfo, {
      color: color,
      status: status,
      name: name
    }, avatar), overlay && (_SplitLine || (_SplitLine = /*#__PURE__*/_createElement(SplitLine, null))), overlay),
    trigger: "hover",
    placement: "bottomLeft",
    delayShow: 200
  }, /*#__PURE__*/_createElement(AvatarOuterWrapper, null, children));
};

export default AvatarPopover;
