// @ali/we-design/dist-web/components/avatar/Avatar.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { AvatarGroup } from "./AvatarGroup";
import AvatarPopover from "./AvatarPopover";
import AvatarContent from "./AvatarContent";
export var Avatar = Object.assign( /*#__PURE__*/React.memo(function (props) {
  var _ref = props,
      status = _ref.status,
      name = _ref.name,
      color = _ref.color,
      popoverOverlay = _ref.popoverOverlay,
      rest = _objectWithoutPropertiesLoose(_ref, ["status", "name", "color", "popoverOverlay"]);

  var avatarContent = /*#__PURE__*/_createElement(AvatarContent, _extends({}, rest, {
    color: color,
    name: name
  }));

  return !popoverOverlay ? avatarContent : /*#__PURE__*/_createElement(AvatarPopover, {
    name: name,
    status: status,
    color: color,
    overlay: /*#__PURE__*/React.isValidElement(popoverOverlay) ? popoverOverlay : null
  }, avatarContent);
}), {
  Group: AvatarGroup
});
Avatar.defaultProps = {
  size: 'normal'
};
Avatar.displayName = 'Avatar';
