// @ali/we-design/dist-web/components/avatar/styled.js
import styled, { css, keyframes } from 'styled-components';
import { resetComponent, optionWrapStyle } from "../../utils/styled.mixin";
import { withTheme } from "../theme";
var directionCss = /*#__PURE__*/css(["", ""], function (_ref) {
  var direction = _ref.direction;

  if (direction === 'rtl') {
    return 'flex-direction: row-reverse;';
  }

  return '';
});
var avatarWidth = /*#__PURE__*/css(["width:", ";"], function (_ref2) {
  var theme = _ref2.theme,
      size = _ref2.size;
  var sizeString = '';
  if (size === 'small') sizeString = 'Small';
  if (size === 'large') sizeString = 'Large';
  return theme.ui["elementHeight" + sizeString];
});
var avatarHeight = /*#__PURE__*/css(["height:", ";"], function (_ref3) {
  var theme = _ref3.theme,
      size = _ref3.size;
  var sizeString = '';
  if (size === 'small') sizeString = 'Small';
  if (size === 'large') sizeString = 'Large';
  return theme.ui["elementHeight" + sizeString];
});
var breath = /*#__PURE__*/keyframes(["0%{box-shadow:0 0px 0px 1px rgba(255,184,0,0.4),0 0px 0px 1px rgba(255,184,0,0.4) inset;}100%{box-shadow:0 0px 0px 4px rgba(255,168,0,0.4),0 0px 0px 4px rgba(255,168,0,0.4) inset;}"]);
export var AvatarWrap = withTheme(styled.div(["", " display:inline-block;position:relative;", ";", ";cursor:", ";&.is-active::before{position:absolute;top:-2px;left:-2px;width:100%;height:100%;box-sizing:content-box;content:'';border:2px solid #ffa800;box-shadow:0px 0px 0px 4px rgba(255,184,0,0.24);}&.is-active.circle::before{border-radius:50%;}&.is-active.rect::before{border-radius:5px;}&.is-active.enable-active-animation::before{animation:", " 1500ms ease-in-out infinite alternate;}&.is-active::after{width:10px;height:3px;background-color:rgb(255,168,0);position:absolute;top:-10px;left:50%;transform:translate(-50%,0);content:'';}"], resetComponent, avatarWidth, avatarHeight, function (p) {
  return p.clickable ? 'pointer' : 'auto';
}, breath));
var divider = /*#__PURE__*/css(["content:'';display:block;position:absolute;background-color:", ";"], function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.colorGray1;
});
var leftDivider = /*#__PURE__*/css(["", " left:0;top:0;bottom:0;width:1px;"], divider);
var rightDivider = /*#__PURE__*/css(["", " right:0;top:0;bottom:0;width:1px;"], divider);
var topDivider = /*#__PURE__*/css(["", " right:0;left:0;top:0;height:1px;"], divider);
export var AvatarInner = withTheme(styled.div(["box-sizing:border-box;position:relative;width:100%;height:100%;overflow:hidden;border:", ";&:hover,&:active{border:", ";}&.circle{border-radius:50%;}&.rect{border-radius:5px;}& > .avatar_2_1,& > .avatar_3_1{transform:translate(-50%,0);}& > .avatar_2_2{transform:translate(50%,0);&:after{", "}}& > .avatar_3_2{transform:translate(50%,-50%);&:after{", "}}& > .avatar_3_3{transform:translate(50%,50%);&:after{", "}&:before{", "}}& > .avatar_4_1{transform:translate(-50%,-50%);&:after{", "}}& > .avatar_4_2{transform:translate(-50%,50%);&:after{", "}&:before{", "}}& > .avatar_4_3{transform:translate(50%,-50%);}& > .avatar_4_4{transform:translate(50%,50%);&:after{", "}}"], function (p) {
  return typeof p.color === 'string' ? "1px solid " + p.theme.ui.colorGray1 : 'none';
}, function (p) {
  return typeof p.color === 'string' ? "1px solid " + (p.color || p.theme.ui.colorGray1) : 'none';
}, leftDivider, leftDivider, leftDivider, topDivider, rightDivider, rightDivider, topDivider, topDivider));
export var AvatarCorner = withTheme(styled.div(["position:absolute;bottom:0;right:0;width:8px;height:8px;border-radius:50%;overflow:hidden;background-color:", ";border:1px solid ", ";box-sizing:border-box;"], function (_ref5) {
  var theme = _ref5.theme;
  return theme.ui.colorGray1;
}, function (_ref6) {
  var theme = _ref6.theme;
  return theme.ui.colorGray1;
}));
export var AvatarColor = withTheme(styled.div(["width:100%;height:100%;background-color:", ";"], function (_ref7) {
  var backgroundColor = _ref7.backgroundColor;
  return backgroundColor;
}));
export var NameWrap = withTheme(styled.div(["width:100%;height:100%;background-color:", ";color:", ";display:flex;justify-content:center;align-items:center;font-size:14px;"], function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.primaryColor;
}, function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.colorGray1;
}));
export var TextSpan = /*#__PURE__*/styled.span(["margin:0 3px 0 7px;"]);
export var AvatarGroupWrap = withTheme(styled.div(["display:inline-block;.avatar-item:not(:last-child){", "}"], function (_ref10) {
  var direction = _ref10.direction;

  if (direction === 'rtl') {
    return ' margin-left: -11px';
  }

  return ' margin-right: -11px';
}));
export var GroupAvatarList = withTheme(styled.ul(["", ";", ";width:100%;", ";display:flex;align-items:center;position:relative;"], resetComponent, directionCss, avatarHeight));
export var GroupAvatarItem = withTheme(styled.li(["", ";", ";position:relative;flex:1 auto;display:flex;align-items:center;&.more .more-item-span{color:rgba(23,26,29,0.6);border:1px solid #fff;background:linear-gradient(180deg,#e9eef4 0%,#dce3ed 100%);&:hover,&.more-box-active{border:1px solid rgba(126,134,142,0.12);}}"], resetComponent, avatarHeight));
export var GroupMoreItem = withTheme(styled.div(["box-sizing:border-box;", ";", ";background-color:#e2e4e5;border-radius:50%;border:1px solid ", ";display:flex;justify-content:center;align-items:center;cursor:pointer;i{display:block;background-color:", ";width:2px;height:2px;margin-right:2px;border-radius:50%;&:last-child{margin-right:0;}}"], avatarHeight, avatarWidth, function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.colorGray1;
}, function (_ref12) {
  var theme = _ref12.theme;
  return theme.ui.textColor;
}));
export var GroupMoreItemSpan = withTheme(styled.span(["font-size:12px;transform:scale(0.75);user-select:none;"]));
export var CountWrap = withTheme(styled.div(["font-weight:400;font-size:13px;line-height:18px;color:rgba(17,31,44,0.56);padding:0 10px 8px 12px;"]));
export var OverlayAvatarItem = withTheme(styled.div(["", " justify-content:space-between;padding:3px 8px;margin:0px 2px 0 4px;border-radius:4px;.status-icon{font-size:14px;}& + &{margin-bottom:2px;}"], optionWrapStyle));
export var OverlayAvatarName = withTheme(styled.div(["width:96px;margin-left:8px;padding-right:16px;text-align:left;font-size:13px;color:#171a1d;font-weight:400;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;"]));
export var OverlayWrap = withTheme(styled.div(["padding:11px 2px 8px 0px;min-width:100px;"]));
export var ListWrap = withTheme(styled.div(["width:100%;max-height:", ";overflow-y:auto;&::-webkit-scrollbar{width:6px;background:transparent;}&::-webkit-scrollbar-thumb{border-radius:6px;background:#c7c8c9;}"], function (p) {
  return (p.overlayMaxHeight || 247) + "px";
}));
export var OverlayAvatarWrap = withTheme(styled.div(["display:flex;align-items:center;.avatar-in-list > .avatar-border-wrapper{border:none;&:hover{border:none;}}"]));
export var OverlayAvatarStatus = withTheme(styled.span(["margin-left:auto;color:", ";"], function (_ref13) {
  var theme = _ref13.theme;
  return theme.ui.commonBaseColor3;
}));
export var Dot = /*#__PURE__*/styled.span(["display:inline-block;width:", "px;height:", "px;border-radius:50%;background:", ";"], function (p) {
  return p.size || 8;
}, function (p) {
  return p.size || 8;
}, function (p) {
  return p.color;
});
export var dropdownOverlayStyle = {
  overflow: 'hidden'
};
