// @ali/4ever-plugin-table/dist-node/mo/models/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

var _tableCell = _interopRequireDefault(require("./tableCell"));

var _tableRow = _interopRequireDefault(require("./tableRow"));

var _meta = require("../../utils/meta");

class Table extends _everCangjie.Block {
  static create(properties) {
    if (typeof properties === 'string') {
      throw new Error('`Table.create` does not accept string');
    }

    if (Table.isTable(properties)) {
      return properties;
    }

    return super.create({ ...properties,
      type: Table.TYPE
    });
  }

  static isTable(maybeTable) {
    return _everCangjie.Block.isBlock(maybeTable) && maybeTable.type === Table.TYPE;
  }

  static isAutofitWidth(table) {
    return table.data?.tblW?.type === 'pct';
  }

  static isRowHeader(table) {
    const {
      nodes
    } = table;
    return Array.isArray(nodes) && nodes.length > 0 && !!nodes[0]?.data?.isTblHeader;
  }

  static isColumnHeader(table) {
    return !!table.data?.isColumnHeader;
  }

  static isValidTable(maybeTable) {
    const {
      colsWidth
    } = maybeTable.data || {};
    return this.isTable(maybeTable) && Array.isArray(colsWidth) && !!colsWidth.length && !!maybeTable.nodes?.length && !(maybeTable.nodes.length === 1 && _everCangjie.Text.isText(maybeTable.nodes[0]));
  }

}

exports.default = Table;
Table.TYPE = _meta.ELEMENT_TYPE_TABLE;
Table.TableRow = _tableRow.default;
Table.TableCell = _tableCell.default;
