// @ali/4ever-bi/dist-node/plugins/imTag/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getMeasureWidth", {
  enumerable: true,
  get: function () {
    return _ImTag.getMeasureWidth;
  }
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _everFactory = require("@ali/4ever-factory");

var _ImTag = require("./components/ImTag");

const _createElement = /*#__PURE__*/React.createElement;
const createBiImTagPlugin = (0, _everFactory.tagFactory)(bizConfig => {
  return {
    tagType: 'im',
    renderNode: (props, controller) => {
      return /*#__PURE__*/_createElement(_ImTag.ImTag, {
        props: props,
        controller: controller,
        imTagConfig: bizConfig.imTag
      });
    },
    models: {
      klass: 'inline',
      type: 'tag'
    }
  };
}, () => ({
  key: 'imTag'
}));
var _default = createBiImTagPlugin;
exports.default = _default;
