// @ali/4ever-bi/dist-node/plugins/quickInsert/components/errorBoundary.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

const _createElement = /*#__PURE__*/React.createElement;

const Wrapper = /*#__PURE__*/_styledComponents.default.div(["overflow:auto;height:100%;"]);

var _ref = /*#__PURE__*/_createElement("h2", null, "Something went wrong.");

var _ref2 = /*#__PURE__*/_createElement("br", null);

class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      error: null,
      errorInfo: null
    };
  }

  componentDidCatch(error, errorInfo) {
    this.setState({
      error,
      errorInfo
    });
  }

  render() {
    const {
      errorInfo,
      error
    } = this.state;
    const {
      children
    } = this.props;

    if (errorInfo) {
      return /*#__PURE__*/_createElement(Wrapper, null, _ref, /*#__PURE__*/_createElement("div", {
        style: {
          whiteSpace: 'pre-wrap'
        }
      }, error && error.toString(), _ref2, errorInfo.componentStack));
    }

    return children;
  }

}

var _default = ErrorBoundary;
exports.default = _default;
