// @ali/4ever-bi/dist-node/plugins/quickInsert/components/menu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _weDesign = require("@ali/we-design");

/* eslint-disable react/no-array-index-key */
const _createElement = /*#__PURE__*/React.createElement;

const MenuWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;background:", ";height:32px;padding:0px 6px;width:100%;overflow:visible;"], () => _weDesign.theme.colorFill1);

const MenuGroup = /*#__PURE__*/_styledComponents.default.div(["display:flex;height:100%;"]);

const MenuGroupWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;align-items:center;:not(:last-child){margin-right:6px;}"]);

const MenuItemWrapper = /*#__PURE__*/_styledComponents.default.div(["font-size:17px;display:flex;position:relative;align-items:center;justify-content:center;padding:0px 4px;border-bottom:2px solid transparent;height:100%;border-bottom-color:", ";color:", ";cursor:", ";user-select:none;:not(:last-child){margin-right:6px;}"], props => props.actived ? _weDesign.theme.colorBlue1 : 'transparent', props => props.actived ? _weDesign.theme.colorBlue1 : _weDesign.theme.colorText3, props => props.disabled ? 'not-allowed' : 'pointer');

const Divider = /*#__PURE__*/_styledComponents.default.div(["background-color:", ";width:1px;height:17px;margin-left:6px;"], () => _weDesign.theme.colorLine2);

const MenuItem = /*#__PURE__*/React.memo(props => {
  const {
    item,
    onClick,
    disabled,
    actived
  } = props;
  const handleClick = React.useCallback(() => {
    if (!disabled && !actived) {
      onClick(item.key);
    }
  }, [item, onClick, disabled, actived]);
  return /*#__PURE__*/_createElement(MenuItemWrapper, {
    "data-testid": "group-item",
    onClick: handleClick,
    disabled: disabled,
    actived: actived
  }, /*#__PURE__*/_createElement(_weDesign.Tooltip, {
    placement: "bottom",
    title: item.title
  }, item.renderIcon({
    actived,
    disabled
  })));
});

var _ref = /*#__PURE__*/_createElement(Divider, null);

const Menu = /*#__PURE__*/React.memo(props => {
  const {
    groups,
    value,
    onChange,
    editorValue
  } = props;
  return /*#__PURE__*/_createElement(MenuWrapper, null, groups.map(({
    key,
    items
  }, index) => /*#__PURE__*/_createElement(MenuGroupWrapper, {
    key: key
  }, /*#__PURE__*/_createElement(MenuGroup, {
    "data-testid": "group"
  }, items.map(item => /*#__PURE__*/_createElement(MenuItem, {
    key: item.key,
    item: item,
    actived: value === item.key,
    disabled: item.disabled(editorValue),
    onClick: onChange
  }))), index !== groups.length - 1 && _ref)));
});
var _default = Menu;
exports.default = _default;
