// @ali/4ever-zhi/dist-web/plugins/imTag/index.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { ImTagPlugin } from '@ali/4ever-bamboo';
import { zhiTagFactory as tagFactory } from '@ali/4ever-factory';
var ImTagContainer = ImTagPlugin.ImTagContainer,
    IMTAG_MEASURE_STRING = ImTagPlugin.IMTAG_MEASURE_STRING;
var createZhiImTagPlugin = tagFactory(function (config) {
  var imTagConfig = config.imTag;

  var _ref = imTagConfig || {},
      _ref$locale = _ref.locale,
      locale = _ref$locale === void 0 ? {
    imFrom: IMTAG_MEASURE_STRING
  } : _ref$locale,
      jumpToConversition = _ref.jumpToConversition;

  return {
    tagType: 'im',
    renderNode: function renderNode(props) {
      var node = props.node;
      var _ref2 = node.data,
          height = _ref2.height,
          width = _ref2.width,
          metadata = _ref2.metadata;

      var _ref3 = metadata || {},
          _ref3$imTitle = _ref3.imTitle,
          imTitle = _ref3$imTitle === void 0 ? '' : _ref3$imTitle,
          cid = _ref3.cid;

      var onClick = function onClick() {
        if (jumpToConversition && cid) jumpToConversition(cid);
      };

      return /*#__PURE__*/_createElement(ImTagContainer, {
        width: width,
        height: height,
        text: locale.imFrom + " " + imTitle,
        onClick: onClick
      });
    }
  };
});
export default createZhiImTagPlugin;
