// @ali/4ever-bi/dist-node/plugins/columns/mobile/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everBamboo = require("@ali/4ever-bamboo");

var _setColumnsItemData = _interopRequireDefault(require("../commands/setColumnsItemData"));

const {
  hooks,
  createQueries
} = _everBamboo.ColumnsPlugin;

function createColumnsPlugin(allConfigs) {
  const {
    createRenderNode,
    createOnAction,
    createOnCangjieSelect,
    createOnKeyDown,
    createPlaceholder,
    createOnCut,
    beforeCopySelectedNodes
  } = hooks;
  const configs = allConfigs.columns;
  return {
    commands: {
      setColumnsItemData: _setColumnsItemData.default
    },
    queries: createQueries(configs),
    renderNode: createRenderNode(configs),
    onAction: createOnAction(configs),
    onCangjieSelect: createOnCangjieSelect(configs),
    onKeyDown: createOnKeyDown(),
    onCut: createOnCut(),
    // @ts-ignore
    placeholder: createPlaceholder(configs),
    beforeCopySelectedNodes
  };
}

var _default = createColumnsPlugin;
exports.default = _default;
