// @ali/we-toolbar/dist-web/biz/splitButton/indentButton/IndentButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";

var _ToolbarIncreaseInden;

import React, { useState, useCallback, useMemo } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { ToolbarIncreaseIndentNormal, ToolbarReduceIndentNormal, Menu, MenuItem } from '@ali/we-design';
import { useTranslate } from '@ali/we-util';
import SplitButton from "../../../common/splitButton";
export var IndentButton = function IndentButton(props) {
  var disabled = props.disabled,
      onSelect = props.onSelect,
      onClick = props.onClick,
      title = props.title,
      titles = props.titles,
      locale = props.locale,
      size = props.size,
      activeLeft = props.active,
      icon = props.icon,
      restProps = _objectWithoutPropertiesLoose(props, ["disabled", "onSelect", "onClick", "title", "titles", "locale", "size", "active", "icon"]);

  var _useState = useState(false),
      active = _useState[0],
      setActive = _useState[1];

  var t = useTranslate(locale);
  var options = [{
    title: t('we_toolbar_increase_indent'),
    Icon: ToolbarIncreaseIndentNormal,
    value: 'increase'
  }, {
    title: t('we_toolbar_reduce_indent'),
    Icon: ToolbarReduceIndentNormal,
    value: 'decrease'
  }];
  var handleClickOption = useCallback(function (e) {
    if (onSelect) onSelect(e, options.find(function (o) {
      return o.value === e;
    }));
  }, [onSelect, options]);
  var handleClickLeft = useCallback(function () {
    if (onClick) onClick();
  }, [onClick]);
  var getOverlay = useMemo(function () {
    return /*#__PURE__*/_createElement(Menu, {
      onClick: function onClick(value) {
        if (value === 'increase' || value === 'decrease') handleClickOption(value);
      }
    }, options.map(function (item) {
      return /*#__PURE__*/_createElement(MenuItem, {
        menuKey: item.value,
        icon: /*#__PURE__*/_createElement(item.Icon, null),
        testid: "option-" + item.value
      }, item.title);
    }));
  }, [options, handleClickOption, titles]);
  return /*#__PURE__*/_createElement(SplitButton, _extends({
    overlay: getOverlay,
    onActiveChange: setActive,
    activeLeft: activeLeft,
    disabled: disabled,
    testid: "IndentButton",
    title: title,
    size: size,
    icon: icon || _ToolbarIncreaseInden || (_ToolbarIncreaseInden = /*#__PURE__*/_createElement(ToolbarIncreaseIndentNormal, null))
  }, restProps, {
    onClickLeft: handleClickLeft,
    active: active
  }));
};
IndentButton.defaultProps = {
  titles: {}
};
