// @ali/we-design-next/dist-web/form/date-picker/panels/styled.js
var _rect, _rect2;
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from "styled-components";
import { PANEL_WIDTH, BODY_HEIGHT, HEADER_HEIGHT } from "./DatePanel/styled";
import { COLORS, CSS_BORDER_BOX } from "../../../common";
export var HeaderWrapper = /*#__PURE__*/styled.div(["", ";height:", "px;display:flex;justify-content:space-between;align-items:center;padding:16px 20px 9px 24px;font-weight:500;font-size:16px;line-height:24px;color:", ";user-select:none;"], CSS_BORDER_BOX, HEADER_HEIGHT, COLORS.LIGHT.LEVEL1_BASE_COLOR);
export var LeftSide = /*#__PURE__*/styled.div(["width:70%;font-weight:500;font-size:16px;line-height:24px;color:", ";"], COLORS.LIGHT.LEVEL1_BASE_COLOR);
export var RightSide = /*#__PURE__*/styled.div(["display:flex;"]);
export var TableWrapper = /*#__PURE__*/styled.div(["", ";height:", "px;padding:0 14px 0px;"], CSS_BORDER_BOX, BODY_HEIGHT);
export var CalendarTable = /*#__PURE__*/styled.table(["width:100%;height:100%;border-collapse:collapse;border-spacing:0;table-layout:fixed;border:none;user-select:none;th{text-align:center;}"]);
export var Td = /*#__PURE__*/styled.td([""]);
export var TBody = /*#__PURE__*/styled.tbody(["min-width:", "px;padding:0 14px 20px;"], PANEL_WIDTH);
export var Cell = /*#__PURE__*/styled.div(["box-sizing:border-box;text-align:center;min-width:24px;height:24px;border-radius:12px;margin:3px auto;font-size:14px;line-height:22px;font-weight:400;border:1px solid ", ";color:", ";&:hover{background-color:#F2F4F5;border-color:#F2F4F5;cursor:default;}&.year{width:50px;}&.month{width:40px;}&.date{width:24px;}&.notCurrentPeriod{color:rgba(17,31,44,0.4);}&.isToday{position:relative;color:", ";background:", ";&:hover{border-color:", ";}}&.selected{color:", ";background-color:", ";border-color:", ";&:before{display:none;}}&.disabled{color:rgba(17,31,44,0.24);background-color:rgba(0,0,0,0.04) !important;cursor:not-allowed;}"], COLORS.TRANSPARENT, COLORS.LIGHT.LEVEL1_BASE_COLOR, COLORS.DING_BLUR, COLORS.COMMON.BLUR3_COLOR, COLORS.DING_BLUR, COLORS.WHITE, COLORS.DING_BLUR, COLORS.DING_BLUR);
export var FooterWrap = /*#__PURE__*/styled.div(["position:relative;overflow:hidden;padding:12px 16px 12px 20px;border-top:1px solid ", ";"], COLORS.OTHERS.LINE_LIGHT_COLOR);
export var ButtonWrap = /*#__PURE__*/styled.div(["float:right;"]);
export var ArrowLeft = /*#__PURE__*/React.memo(function () {
  return /*#__PURE__*/_createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    width: "16",
    height: "16",
    viewBox: "0 0 16 16",
    fill: "none"
  }, _rect || (_rect = /*#__PURE__*/_createElement("rect", {
    opacity: "0.01",
    width: "16",
    height: "16",
    transform: "matrix(1 -8.74228e-08 -8.74228e-08 -1 0 16)",
    fill: "#D8D8D8"
  })), /*#__PURE__*/_createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M11.3932 14.1815C11.1224 14.4684 10.679 14.4944 10.3772 14.2506L10.333 14.212L4.58966 8.7901C4.1546 8.37938 4.13486 7.69374 4.54558 7.25867L4.58932 7.2149L10.3328 1.78819C10.6339 1.50371 11.1085 1.51717 11.393 1.81825C11.6639 2.10499 11.6646 2.5492 11.404 2.83659L11.363 2.87849L5.9402 8.00234L11.3627 13.1213C11.6496 13.3921 11.6756 13.8355 11.4318 14.1374L11.3932 14.1815Z",
    fill: COLORS.LIGHT.LEVEL1_BASE_COLOR
  }));
});
export var ArrowRight = /*#__PURE__*/React.memo(function () {
  return /*#__PURE__*/_createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    width: "16",
    height: "16",
    viewBox: "0 0 16 16",
    fill: "none"
  }, _rect2 || (_rect2 = /*#__PURE__*/_createElement("rect", {
    opacity: "0.01",
    width: "16",
    height: "16",
    transform: "matrix(-1 0 0 1 16 0)",
    fill: "#D8D8D8"
  })), /*#__PURE__*/_createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M4.60676 1.81848C4.87757 1.53162 5.32101 1.50557 5.62281 1.74937L5.66698 1.78796L11.4103 7.2099C11.8454 7.62062 11.8651 8.30627 11.4544 8.74133L11.4107 8.7851L5.66722 14.2118C5.36614 14.4963 4.89146 14.4828 4.60698 14.1817C4.33606 13.895 4.33536 13.4508 4.59599 13.1634L4.63705 13.1215L10.0598 7.99766L4.63728 2.8787C4.35042 2.6079 4.32437 2.16446 4.56817 1.86265L4.60676 1.81848Z",
    fill: COLORS.LIGHT.LEVEL1_BASE_COLOR
  }));
});
