// @ali/we-biz-url-card/dist-web/index.js
import _taggedTemplateLiteralLoose from '@babel/runtime/helpers/taggedTemplateLiteralLoose';
import React, { useState, useRef, useEffect } from 'react';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import styled from 'styled-components';
import { Button, LoadingNormal, Dropdown, FooterHelpNormal, ExclamationCircleBetaNormal } from '@ali/we-design';
import { IconButton } from '@ali/we-toolbar';

function _templateObject4() {
  var data = _taggedTemplateLiteralLoose(["\n  opacity: 0.3;\n  flex: 1;\n  border: 1px solid rgba(126, 134, 142, 0.24);\n  box-sizing: border-box;\n  border-radius: 4px;\n  padding: 7px 6px;\n  min-width: 0;\n  font-size: 14px;\n  font-family: PingFang SC;\n  height: 32px;\n\n  &:focus::placeholder {\n    color: rgba(23, 26, 29, 0.24);\n  }\n\n  &:focus {\n    outline: none;\n    opacity: 1;\n    border-color: #2d84fa;\n    color: 171a1d;\n  }\n\n  &[data-error] {\n    opacity: 1;\n    border-color: #fb602d;\n  }\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteralLoose(["\n  display: flex;\n  align-items: center;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose([""]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n  height: ", "px;\n  box-sizing: border-box;\n  white-space: normal;\n  cursor: auto;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}
var Container = styled.div(_templateObject(), function (p) {
  return p.height;
});
var InputWrapper = styled.div(_templateObject2());
var InputContent = styled.div(_templateObject3());
var InputInput = styled.input(_templateObject4()); // 占位符的输入组件

var MainInput = function MainInput(_ref, ref) {
  var style = _ref.style,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      placeholder = _ref.placeholder,
      value = _ref.value,
      showError = _ref.showError,
      btnText = _ref.btnText,
      onChange = _ref.onChange,
      onConfirm = _ref.onConfirm,
      onFocus = _ref.onFocus,
      onBlur = _ref.onBlur,
      isLoading = _ref.isLoading;

  var _useState = useState(false),
      showButton = _useState[0],
      setShowButton = _useState[1];

  var handleInput = function handleInput(evt) {
    var value = evt.target.value; // 出现按钮

    if (value.length > 0) {
      setShowButton(true);
    } else {
      setShowButton(false);
    }

    onChange(value);
  }; // Press enter to submit


  var handleKeyDown = function handleKeyDown(evt) {
    if (evt.key === 'Enter') {
      onConfirm();
    }
  };

  return /*#__PURE__*/React.createElement(InputWrapper, {
    style: style
  }, /*#__PURE__*/React.createElement(InputContent, null, /*#__PURE__*/React.createElement(InputInput, {
    ref: ref,
    disabled: disabled ? true : undefined,
    value: value,
    type: "string",
    placeholder: placeholder,
    "data-error": showError ? true : null,
    onChange: handleInput,
    onKeyDown: handleKeyDown,
    onFocus: onFocus,
    onBlur: onBlur
  }), showButton && /*#__PURE__*/React.createElement(Button, {
    onClick: onConfirm,
    size: "large",
    type: "primary",
    disabled: showError,
    style: {
      marginLeft: '12px',
      overflow: 'hidden'
    }
  }, isLoading ? /*#__PURE__*/React.createElement(LoadingNormal, {
    style: {
      marginLeft: '2px'
    }
  }) : btnText)));
};
var Input = /*#__PURE__*/React.forwardRef(MainInput);

var getIframeURL = function getIframeURL(html) {
  try {
    var div = document.createElement('div');
    div.innerHTML = html;
    return div.children[0].src;
  } catch (err) {
    return '';
  }
}; // ‒ https://ditu.amap.com/place/B0FFFKOO8E 直接从地址栏复制
var youkuURLRegexp = /^https?:\/\/v\.youku\.com\/v_show\/id_(.+?)\.html(\?.*)?$/;
var youkuPlayerURLRegexp = /^https?:\/\/player\.youku\.com\/embed\/.+$/; // 将输入的内容转换为优酷 iframe 的 URL 格式，如果不能转换，返回空
// 输入格式：
// - https://player.youku.com/embed/XMTY1MjYwODYzNg==
// - https://v.youku.com/v_show/id_XNDk0MTAzNTc5Ng==.html?spm=a2hbt.13141534.1_3.d_1_2&s=cafe37eba40744ba9ca9
// - https://v.youku.com/v_show/id_XNDk0MTAzNTc5Ng==.html
// - <iframe height=498 width=510 src='https://player.youku.com/embed/XNDk0MTAzNTc5Ng==' frameborder=0 'allowfullscreen'></iframe>
// 输出格式： https://player.youku.com/embed/XNDgzNDM2NTM1Ng==

var transformYoukuInput = function transformYoukuInput(input) {
  // iframe
  if (input.startsWith('<iframe ')) {
    var url = getIframeURL(input);
    return youkuPlayerURLRegexp.test(url) ? url : '';
  }

  if (youkuPlayerURLRegexp.test(input)) {
    return input;
  }

  var match = input.match(youkuURLRegexp);
  if (match == null) return '';
  return "https://player.youku.com/embed/" + match[1];
};

function _templateObject9() {
  var data = _taggedTemplateLiteralLoose(["\n  width: 260px;\n  background: #ffffff;\n  box-shadow: 0px 0px 1px rgba(0, 0, 0, 0.24), 0px 8px 24px rgba(0, 0, 0, 0.16);\n  padding: 12px;\n  border-radius: 4px;\n\n  > img {\n    max-width: 100%;\n    display: block;\n  }\n"]);

  _templateObject9 = function _templateObject9() {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = _taggedTemplateLiteralLoose(["\n  display: flex;\n  align-items: center;\n  min-width: 0;\n  flex: 1;\n  color: rgba(23, 26, 29, 1);\n  // \u6587\u5B57\u4F1A\u8BBE\u7F6E line-height\uFF0C\u8FD9\u91CC\u91CD\u7F6E\uFF0C\u5426\u5219\u5BF9\u9F50\u4F1A\u6709\u95EE\u9898\n  line-height: normal;\n"]);

  _templateObject8 = function _templateObject8() {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = _taggedTemplateLiteralLoose(["\n  height: 100%;\n  background-image: url('", "');\n  background-position: 50% 50%;\n  font-size: 14px;\n  display: flex;\n  flex-direction: column;\n\n  &.tran-enter {\n    background-size: ", ";\n  }\n\n  &.tran-enter-active {\n    background-size: ", ";\n    transition: background-size 2000ms;\n  }\n\n  &.tran-enter-done {\n    background-size: cover;\n  }\n"]);

  _templateObject7 = function _templateObject7() {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = _taggedTemplateLiteralLoose(["\n  background: linear-gradient(\n    180deg,\n    rgba(17, 31, 44, 0) 0%,\n    rgba(17, 31, 44, 0.12) 24.48%,\n    rgba(17, 31, 44, 0.12) 79.69%,\n    rgba(17, 31, 44, 0) 100%\n  );\n  position: absolute;\n  left: ", "px;\n  top: 1px;\n  width: 1px;\n  height: 100%;\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = _taggedTemplateLiteralLoose(["\n  border-radius: 8px;\n  background: white;\n  box-shadow: 0px 5px 15px rgba(17, 31, 44, 0.1);\n  padding: ", ";\n  position: relative;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4$1() {
  var data = _taggedTemplateLiteralLoose(["\n  padding: ", ";\n\n  &.tran-enter {\n    transform: scale(0.8);\n  }\n\n  &.tran-enter-active {\n    transform: scale(1);\n    transition: transform 200ms;\n  }\n\n  &.tran-enter-done {\n    transform: scale(1);\n  }\n"]);

  _templateObject4$1 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3$1() {
  var data = _taggedTemplateLiteralLoose(["\n  width: 24px;\n  height: 24px;\n  border-radius: 50%;\n  text-align: center;\n  font-size: 13px;\n  line-height: 24px;\n  margin-right: 10px;\n  color: white;\n  flex-shrink: 0;\n  background: ", ";\n"]);

  _templateObject3$1 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2$1() {
  var data = _taggedTemplateLiteralLoose(["\n  display: flex;\n  align-items: center;\n  position: relative; // \u76D6\u8FC7\u7AD6\u7EBF\n"]);

  _templateObject2$1 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$1() {
  var data = _taggedTemplateLiteralLoose(["\n  flex: 1;\n  display: flex;\n  align-items: center;\n  justify-content: space-around;\n\n  img {\n    width: ", "px;\n  }\n"]);

  _templateObject$1 = function _templateObject() {
    return data;
  };

  return data;
}
var BG_URL = 'https://img.alicdn.com/imgextra/i2/O1CN01Ww89xd1uCVRWYin5T_!!6000000006001-2-tps-1324-800.png';
var config = {
  bg: BG_URL,
  gif: 'https://img.alicdn.com/imgextra/i3/O1CN015Bwpd21hh6tcGbTED_!!6000000004308-1-tps-400-400.gif',
  url: 'https://www.youku.com/',
  transform: transformYoukuInput,
  gifWidth: {
    pc: 200,
    mobile: 80
  }
};
var GifContainer = styled.div(_templateObject$1(), function (p) {
  return p.gifWidth;
});
var WithNumberPrefixWrapper = styled.div(_templateObject2$1());
var WithNumberPrefixContent = styled.div(_templateObject3$1(), function (p) {
  return p.active ? '#007aff' : '#E5E5EA';
});

var WithNumberPrefix = function WithNumberPrefix(_ref) {
  var num = _ref.num,
      children = _ref.children,
      style = _ref.style,
      _ref$active = _ref.active,
      active = _ref$active === void 0 ? false : _ref$active;
  return /*#__PURE__*/React.createElement(WithNumberPrefixWrapper, {
    style: style
  }, /*#__PURE__*/React.createElement(WithNumberPrefixContent, {
    active: active
  }, num), children);
};

var ContentWrapper = styled.div(_templateObject4$1(), function (p) {
  return !p.isMobile ? "20px 20px 0px 20px" : '9px 6px 0 6px';
});
var Content = styled.div(_templateObject5(), function (p) {
  return !p.isMobile ? '18px 16px 16px 16px' : '10px 12px';
}); // 竖线

var VerticalLine = styled.div(_templateObject6(), function (p) {
  return !p.isMobile ? 28 : 24;
});
var Wrapper = styled.div(_templateObject7(), function (p) {
  return p.bgURL;
}, function (p) {
  return !p.isMobile ? 'auto 200%' : '200%';
}, function (p) {
  return !p.isMobile ? 'auto 100%' : '100%';
});
var FirstLineWrapper = styled.div(_templateObject8());
var PopupOverlayWrapper = styled.div(_templateObject9());

var _ref3 = /*#__PURE__*/React.createElement("img", {
  src: "https://img.alicdn.com/imgextra/i1/O1CN01NHjF2q1wOzTMdYO73_!!6000000006299-2-tps-484-224.png"
});

var PopupOverlay = function PopupOverlay(_ref2) {
  var title = _ref2.title,
      desc = _ref2.desc,
      textAtTop = _ref2.textAtTop;
  var textNode = /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("div", {
    style: {
      fontSize: '16px',
      lineHeight: '160%',
      color: '#171A1D',
      marginTop: textAtTop ? 0 : '12px',
      marginBottom: '3px'
    }
  }, title), /*#__PURE__*/React.createElement("div", {
    style: {
      fontSize: '14px',
      lineHeight: '20px',
      color: 'rgba(23, 26, 29, 0.6)',
      marginBottom: textAtTop ? '12px' : '0px'
    }
  }, desc));
  return /*#__PURE__*/React.createElement(PopupOverlayWrapper, null, textAtTop && textNode, _ref3, !textAtTop && textNode);
};

var _ref5 = /*#__PURE__*/React.createElement("div", null, "\u4F18\u9177\u89C6\u9891\u6682\u4E0D\u652F\u6301\u624B\u673A\u7AEF\u63D2\u5165");

var _ref6 = /*#__PURE__*/React.createElement(FooterHelpNormal, null);

var YoukuPlaceholder = function YoukuPlaceholder(_ref4) {
  var readonly = _ref4.readonly,
      onURLChange = _ref4.onURLChange,
      openURLInBrowser = _ref4.openURLInBrowser,
      onGoToSite = _ref4.onGoToSite,
      _ref4$isMobile = _ref4.isMobile,
      isMobile = _ref4$isMobile === void 0 ? false : _ref4$isMobile,
      locale = _ref4.locale;

  var _useState = useState(1),
      step = _useState[0],
      setStep = _useState[1];

  var _useState2 = useState(''),
      inputStr = _useState2[0],
      setInputStr = _useState2[1];

  var _useState3 = useState(false),
      showError = _useState3[0],
      setShowError = _useState3[1];

  var inputRef = useRef(null); // 用于保持 GIF 图地址唯一，否则第二次插入时不会播放

  var timeStr = useRef(Date.now().toString());

  var _useState4 = useState(false),
      inProp = _useState4[0],
      setInProp = _useState4[1];

  useEffect(function () {
    setInProp(true);
  }, []); // 一旦点击以后，就 activate step 2

  var toURL = function toURL() {
    openURLInBrowser == null ? void 0 : openURLInBrowser(config.url);
    onGoToSite == null ? void 0 : onGoToSite();

    if (!readonly) {
      var _inputRef$current;

      setStep(2);
      (_inputRef$current = inputRef.current) == null ? void 0 : _inputRef$current.focus();
    }
  };

  var onInputChange = function onInputChange(newValue) {
    setStep(2);
    setInputStr(newValue);

    if (showError) {
      setShowError(false);
    }
  };

  var onConfirm = function onConfirm() {
    var result = config.transform(inputStr); // 输入的值无效

    if (result === '') {
      var _inputRef$current2;

      setShowError(true);
      inputRef && ((_inputRef$current2 = inputRef.current) == null ? void 0 : _inputRef$current2.select());
    } else {
      onURLChange == null ? void 0 : onURLChange(result);
    }
  };

  var handleFocus = function handleFocus() {
    setStep(2);

    if (showError && inputRef && inputRef.current) {
      inputRef.current.select();
    }
  };

  var handleBlur = function handleBlur() {
    setShowError(false);
  };

  return /*#__PURE__*/React.createElement(CSSTransition, {
    timeout: 2000,
    "in": inProp,
    classNames: "tran"
  }, /*#__PURE__*/React.createElement(Wrapper, {
    isMobile: isMobile,
    bgURL: config.bg
  }, /*#__PURE__*/React.createElement(CSSTransition, {
    "in": inProp,
    timeout: 200,
    classNames: "tran"
  }, /*#__PURE__*/React.createElement(ContentWrapper, {
    isMobile: isMobile
  }, /*#__PURE__*/React.createElement(Content, {
    isMobile: isMobile
  }, /*#__PURE__*/React.createElement(VerticalLine, {
    isMobile: isMobile
  }), /*#__PURE__*/React.createElement(WithNumberPrefix, {
    num: "1",
    active: step === 1,
    style: {
      marginBottom: (isMobile ? 8 : 16) + 'px'
    }
  }, /*#__PURE__*/React.createElement(FirstLineWrapper, null, isMobile ? _ref5 : /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("span", null, locale.open), "\u2009", /*#__PURE__*/React.createElement(Button, {
    type: "text",
    onClick: toURL
  }, locale.name), "\u2009", /*#__PURE__*/React.createElement("span", null, locale.actionText), /*#__PURE__*/React.createElement("div", {
    style: {
      marginLeft: 'auto'
    }
  }, /*#__PURE__*/React.createElement(Dropdown, {
    overlayStyle: {
      border: 'none',
      boxShadow: 'none'
    },
    overlay:
    /*#__PURE__*/

    /**
     * TODO(xunyi): 这里包装了一层，否则会有 warning
     * figure out why
     */
    React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(PopupOverlay, {
      title: locale.copyTitle,
      desc: locale.copyDesc,
      textAtTop: false
    })),
    trigger: "click",
    placement: "bottomRight",
    offsetY: 4
  }, /*#__PURE__*/React.createElement(IconButton, {
    icon: _ref6,
    showTitle: true,
    title: locale.linkText
  })))))), /*#__PURE__*/React.createElement(WithNumberPrefix, {
    num: "2",
    active: step === 2
  }, /*#__PURE__*/React.createElement(Dropdown, {
    visible: showError,
    offsetY: 4,
    overlayStyle: {
      border: 'none',
      boxShadow: 'none'
    },
    overlay:
    /*#__PURE__*/

    /**
     * TODO(xunyi): 这里包装了一层，否则会有 warning
     * figure out why
     */
    React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(PopupOverlay, {
      textAtTop: true,
      title: locale.linkError,
      desc: locale.linkErrorDesc
    }))
  }, /*#__PURE__*/React.createElement(Input, {
    ref: inputRef,
    value: inputStr,
    disabled: isMobile,
    onChange: onInputChange,
    onConfirm: onConfirm,
    onFocus: handleFocus,
    onBlur: handleBlur,
    showError: showError,
    btnText: locale.confirm,
    isLoading: false,
    style: {
      minWidth: 0,
      flex: '1'
    },
    placeholder: isMobile ? '' : locale.placeholder
  })))))), /*#__PURE__*/React.createElement(GifContainer, {
    gifWidth: isMobile ? config.gifWidth.mobile : config.gifWidth.pc
  }, /*#__PURE__*/React.createElement("img", {
    src: config.gif + ("?_id=" + timeStr.current)
  }))));
};

var _ref2 = /*#__PURE__*/React.createElement("svg", {
  width: "25",
  height: "25",
  viewBox: "0 0 25 25",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/React.createElement("path", {
  d: "M17.353 8.62731C16.2273 7.1644 14.459 6.22168 12.4704 6.22168C9.07003 6.22168 6.31348 8.97823 6.31348 12.3786C6.31348 15.779 9.07003 18.5355 12.4704 18.5355C14.8695 18.5355 16.9481 17.1633 17.9643 15.161",
  stroke: "#111F2C",
  strokeOpacity: "0.85"
}), /*#__PURE__*/React.createElement("path", {
  d: "M17.308 5.34229V8.42075H14.2295",
  stroke: "#111F2C",
  strokeOpacity: "0.85"
}));

var RefreshIcon = function RefreshIcon() {
  return _ref2;
};

var _ref3$1 = /*#__PURE__*/React.createElement("svg", {
  width: "25",
  height: "25",
  viewBox: "0 0 25 25",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/React.createElement("circle", {
  cx: "12.6166",
  cy: "12.3788",
  r: "6.53648",
  stroke: "#111F2C",
  strokeOpacity: "0.85"
}), /*#__PURE__*/React.createElement("path", {
  fillRule: "evenodd",
  clipRule: "evenodd",
  d: "M12.7362 10.8675C12.301 10.8582 11.8811 11.0274 11.5739 11.3358C11.2656 11.643 11.0965 12.0629 11.1056 12.4981L9.53809 15.6325L12.6725 14.0653C13.1078 14.0746 13.5279 13.9054 13.8351 13.597C14.1434 13.2898 14.3125 12.8698 14.3034 12.4347L15.8706 9.30029L12.7362 10.8675ZM12.0846 12.4664C12.0846 12.1208 12.359 11.8376 12.7044 11.8267C13.0498 11.8376 13.3242 12.1208 13.3242 12.4664C13.3242 12.812 13.0498 13.0951 12.7044 13.1061C12.359 13.0951 12.0846 12.812 12.0846 12.4664Z",
  fill: "#111F2C",
  fillOpacity: "0.85"
}));

var BrowserIcon = function BrowserIcon() {
  return _ref3$1;
};

function _templateObject6$1() {
  var data = _taggedTemplateLiteralLoose(["\n  height: 100%;\n  display: flex;\n  align-items: center;\n  background: white;\n"]);

  _templateObject6$1 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5$1() {
  var data = _taggedTemplateLiteralLoose(["\n  border: 1px solid rgba(17, 31, 44, 0.08);\n  box-shadow: 0px 10px 20px rgba(0, 0, 0, 0.12);\n  border-radius: 4px;\n  display: flex;\n  align-items: center;\n  background: white;\n"]);

  _templateObject5$1 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4$2() {
  var data = _taggedTemplateLiteralLoose(["\n  padding-bottom: 10px;\n  position: absolute;\n  left: 0;\n  top: -35px;\n  z-index: 9999;\n"]);

  _templateObject4$2 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3$2() {
  var data = _taggedTemplateLiteralLoose(["\n  padding: 2px;\n  box-sizing: border-box;\n  display: flex;\n  align-items: center;\n  cursor: pointer;\n\n  > div {\n    width: 24px;\n    height: 24px;\n    border-radius: 4px;\n    &:hover {\n      background-color: rgba(25, 31, 37, 0.08);\n    }\n    > svg {\n      width: 100%;\n      height: 100%;\n    }\n  }\n"]);

  _templateObject3$2 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2$2() {
  var data = _taggedTemplateLiteralLoose(["\n  position: absolute;\n  left: 0;\n  top: 0;\n  width: 100%;\n  height: 100%;\n\n  &.tran-enter {\n    opacity: 0;\n  }\n\n  &.tran-enter-active {\n    opacity: 1;\n    transition: opacity 200ms ease-in-out;\n  }\n\n  &.tran-exit {\n    opacity: 1;\n  }\n\n  &.tran-exit-active {\n    opacity: 0;\n    transition: opacity 200ms ease-in-out;\n  }\n"]);

  _templateObject2$2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$2() {
  var data = _taggedTemplateLiteralLoose(["\n  position: relative;\n  height: 100%;\n  border-radius: 8px;\n  overflow: hidden;\n"]);

  _templateObject$2 = function _templateObject() {
    return data;
  };

  return data;
}
var ContentWrapper$1 = styled.div(_templateObject$2());
var Content$1 = styled.div(_templateObject2$2());
var ToolbarItemWrapper = styled.div(_templateObject3$2());

var ToolbarItem = function ToolbarItem(_ref) {
  var children = _ref.children,
      onClick = _ref.onClick;
  return /*#__PURE__*/React.createElement(ToolbarItemWrapper, {
    onClick: onClick
  }, /*#__PURE__*/React.createElement("div", null, children));
}; // 底部留出一段距离，这样鼠标划出以后，toolbar 不会消失


var ToolbarWrapper = styled.div(_templateObject4$2());
var ToolbarContent = styled.div(_templateObject5$1());

var _ref3$2 = /*#__PURE__*/React.createElement(RefreshIcon, null);

var _ref4 = /*#__PURE__*/React.createElement(BrowserIcon, null);

var Toolbar = function Toolbar(_ref2) {
  var onRefresh = _ref2.onRefresh,
      onOpenURL = _ref2.onOpenURL;
  return /*#__PURE__*/React.createElement(ToolbarWrapper, null, /*#__PURE__*/React.createElement(ToolbarContent, null, /*#__PURE__*/React.createElement(ToolbarItem, {
    onClick: onRefresh
  }, _ref3$2), /*#__PURE__*/React.createElement(ToolbarItem, {
    onClick: onOpenURL
  }, _ref4)));
};
var getHeight = function getHeight(_ref5) {
  var isMobile = _ref5.isMobile;
  return isMobile ? 200 : 400;
};
var YoukuContentWrapper = styled.div(_templateObject6$1());
var YoukuContent = /*#__PURE__*/React.forwardRef(function (_ref6, ref) {
  var url = _ref6.url;
  var iframeRef = useRef(null);
  React.useImperativeHandle(ref, function () {
    return {
      refresh: function refresh() {
        if (iframeRef.current) {
          iframeRef.current.src += '';
        }
      }
    };
  });
  return /*#__PURE__*/React.createElement(YoukuContentWrapper, null, /*#__PURE__*/React.createElement("iframe", {
    onLoad: function onLoad(e) {
      return e.target.blur();
    },
    sandbox: "allow-forms allow-scripts allow-same-origin",
    allowFullScreen: true,
    ref: iframeRef,
    style: {
      border: 'none',
      width: '100%',
      height: '100%'
    },
    src: url
  }));
});
var YoukuCard = function YoukuCard(_ref7) {
  var _ref7$isMobile = _ref7.isMobile,
      isMobile = _ref7$isMobile === void 0 ? false : _ref7$isMobile,
      _ref7$readonly = _ref7.readonly,
      readonly = _ref7$readonly === void 0 ? false : _ref7$readonly,
      url = _ref7.url,
      locale = _ref7.locale,
      onURLChange = _ref7.onURLChange,
      onGoToSite = _ref7.onGoToSite,
      openURLInBrowser = _ref7.openURLInBrowser;
  var height = getHeight({
    isMobile: isMobile
  });

  var _useState = useState(false),
      showToolbar = _useState[0],
      setShowToolbar = _useState[1];

  var showPlaceholder = url === '';
  var contentRef = useRef(null);

  var toolbarVisible = function () {
    if (readonly) return false;
    if (showPlaceholder) return false;
    if (isMobile) return false;
    return showToolbar;
  }();

  var handleRefresh = function handleRefresh() {
    var _contentRef$current;

    (_contentRef$current = contentRef.current) == null ? void 0 : _contentRef$current.refresh();
  };

  var handleOpenURL = function handleOpenURL() {
    openURLInBrowser == null ? void 0 : openURLInBrowser(url);
  };

  return /*#__PURE__*/React.createElement(Container, {
    onMouseEnter: function onMouseEnter() {
      return setShowToolbar(true);
    },
    onMouseLeave: function onMouseLeave() {
      return setShowToolbar(false);
    },
    height: height
  }, /*#__PURE__*/React.createElement(ContentWrapper$1, null, /*#__PURE__*/React.createElement(TransitionGroup, null, /*#__PURE__*/React.createElement(CSSTransition, {
    key: showPlaceholder.toString(),
    timeout: 500,
    classNames: "tran"
  }, /*#__PURE__*/React.createElement(Content$1, null, showPlaceholder ? /*#__PURE__*/React.createElement(YoukuPlaceholder, {
    isMobile: isMobile,
    onURLChange: onURLChange,
    onGoToSite: onGoToSite,
    openURLInBrowser: openURLInBrowser,
    readonly: readonly,
    locale: locale
  }) : /*#__PURE__*/React.createElement(YoukuContent, {
    url: url,
    ref: contentRef
  }))))), toolbarVisible && /*#__PURE__*/React.createElement(Toolbar, {
    onRefresh: handleRefresh,
    onOpenURL: handleOpenURL
  }));
};

var Exported = YoukuCard;
Exported.getHeight = getHeight;
Exported.bgURL = BG_URL;

var _ref = /*#__PURE__*/React.createElement("span", null, "\u6682\u4E0D\u652F\u6301\u5BFC\u51FA");

var NoSupportPlaceholder = /*#__PURE__*/React.memo(function () {
  return /*#__PURE__*/React.createElement("div", {
    style: {
      height: '100%',
      minHeight: ' 200px',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      background: 'white',
      border: '1px solid rgba(17, 31, 44, 0.08)',
      borderRadius: 8
    } //TODO 服务端ssl 不支持国际化，暂时不做

  }, /*#__PURE__*/React.createElement("div", {
    style: {
      display: 'flex',
      justifyContent: 'center'
    }
  }, /*#__PURE__*/React.createElement(ExclamationCircleBetaNormal, {
    style: {
      fontSize: 23,
      marginRight: 4
    }
  }), _ref));
});

export { NoSupportPlaceholder, Exported as YoukuCard };
