// @ali/4ever-bi/dist-web/plugins/callout/components/colorPanel.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { ColorPanel as ColorPanelComp } from '@ali/we-design';
import { CalloutPlugin } from '@ali/4ever-bamboo';
import { setMetaData } from "../actions";
var DEFAULTS = CalloutPlugin.DEFAULTS;
var ColorPanelWrapper = /*#__PURE__*/styled.div(["background:white;padding:8px;border-radius:4px;"]);

var blockEvent = function blockEvent(event) {
  event.stopPropagation();
  event.preventDefault();
};

export function ColorPanel(_ref) {
  var color = _ref.color,
      controller = _ref.controller,
      getNode = _ref.getNode,
      bgcolor = _ref.bgcolor,
      border = _ref.border,
      locale = _ref.locale,
      showSticker = _ref.showSticker,
      onValueChange = _ref.onValueChange;

  var handleChange = function handleChange(key) {
    return function (value) {
      var data = {}; // @ts-ignore

      data[key] = value;
      var node = getNode();
      controller.run('onCangjieBlur');
      controller.run('onAction', setMetaData({
        node: node,
        data: data
      }));
      onValueChange && onValueChange(key, value);
    };
  };

  var checkboxLayout = [{
    title: (locale == null ? void 0 : locale.showEmoji) || 'Show Sticker',
    defaultValue: true,
    value: Boolean(showSticker),
    onValueChange: function onValueChange(value) {
      handleChange('showstk')(value);
    }
  }];
  return /*#__PURE__*/_createElement(ColorPanelWrapper, {
    onMouseDown: blockEvent,
    onClick: blockEvent
  }, /*#__PURE__*/_createElement(ColorPanelComp, {
    color: color,
    defaultColor: DEFAULTS.color,
    onColorClick: handleChange('color'),
    borderColor: border,
    defaultBorderColor: DEFAULTS.border,
    onBorderColorClick: handleChange('border'),
    backgroundColor: bgcolor,
    defaultBackgroundColor: DEFAULTS.bgcolor,
    onBackgroundColorClick: handleChange('bgcolor'),
    checkboxLayout: checkboxLayout,
    enableResetButton: true
  }));
}
