// @ali/4ever-bi/dist-web/plugins/callout/components/styled.js
import styled from 'styled-components';
export var PaddingHolder = /*#__PURE__*/styled.div(["display:flex;visibility:", ";opacity:", ";flex-direction:row;padding:0;position:absolute;z-index:200;top:3px;right:3px;"], function (props) {
  return props.showInnerToolbar ? 'visible' : 'hidden';
}, function (props) {
  return props.visibile ? '1' : '0';
});
export var CalloutInnerToolbarOuterWrapper = /*#__PURE__*/styled.div(["padding:", ";background-color:", ";border-radius:4px;"], function (_ref) {
  var bgcolor = _ref.bgcolor;
  return bgcolor === 'white' ? '0' : '1px';
}, function (_ref2) {
  var bgcolor = _ref2.bgcolor;
  return bgcolor;
});
export var CalloutInnerToolbarWrapper = /*#__PURE__*/styled.div(["cursor:pointer;border-radius:4px;transition:background-color 250ms;background-color:rgba(126,134,142,0.15);&:hover{background-color:rgba(126,134,142,0.12);}& div .icon-button{cursor:pointer;}"]);
export var CalloutSettingPanelWrapper = /*#__PURE__*/styled.div(["display:flex;flex-direction:column;width:125px;height:100px;border-radius:8px;background:#ffffff;box-shadow:0px 0px 1px rgba(0,0,0,0.24),0px 8px 24px rgba(0,0,0,0.16);"]);
export var CalloutColorSettingPanelWrapper = /*#__PURE__*/styled.div(["display:inline-block;flex-direction:column;justify-content:center;align-items:center;flex-wrap:wrap;width:125px;height:71px;padding-top:7px;"]);
export var CalloutColorSettingPanelLineWrapper = /*#__PURE__*/styled.div(["display:inline-block;width:125px;padding-left:7px;"]);
export var CalloutClearSettingPanelWrapper = /*#__PURE__*/styled.div(["display:flex;flex-direction:row;justify-content:center;align-items:center;width:121px;height:28px;color:#0089ff;border-top:solid 1px rgba(126,134,142,0.16);font-size:12px;margin:0 2px;cursor:pointer;"]);
export var CalloutColorSettingItem = /*#__PURE__*/styled.div(["box-sizing:content-box;display:inline-block;text-align:center;line-height:22px;width:22px;height:22px;border-radius:4px;margin:2px 2px;color:rgba(23,26,29,0.4);font-size:17px;background-color:", ";border:", ";cursor:pointer;"], function (props) {
  return props.color;
}, function (props) {
  return props.color === '#FFFFFF' ? '1px solid #EEEEEE' : '1px solid rgba(255,255,255,0)';
});
export var Placeholder = /*#__PURE__*/styled.div(["position:absolute;color:rgba(23,26,29,0.4);padding-left:24px;top:0;left:2px;pointer-events:none;"]);
export var EmptyContainer = /*#__PURE__*/styled.div([""]);
