// @ali/we-design-next/dist-web/display/popover/align.js
import domAlign from 'dom-align';
// import domAlign from '../../common/domAlign';

export var align = function align(source, target, placement, _temp, resize) {
  var _ref = _temp === void 0 ? [0, 0] : _temp,
    m = _ref[0],
    s = _ref[1];
  var align = ['tc', 'bc']; // 对应defaultProps的bottom
  var offset = [0, 0];
  switch (placement) {
    case 'topLeft':
      align = ['bl', 'tl'];
      offset = [s, -m];
      break;
    case 'top':
      align = ['bc', 'tc'];
      offset = [s, -m];
      break;
    case 'topRight':
      align = ['br', 'tr'];
      offset = [-s, -m];
      break;
    case 'leftTop':
      align = ['tr', 'tl'];
      offset = [-m, s];
      break;
    case 'rightTop':
      align = ['tl', 'tr'];
      offset = [m, s];
      break;
    case 'left':
      align = ['cr', 'cl'];
      offset = [-m, s];
      break;
    case 'right':
      align = ['cl', 'cr'];
      offset = [m, s];
      break;
    case 'leftBottom':
      align = ['br', 'bl'];
      offset = [-m, -s];
      break;
    case 'rightBottom':
      align = ['bl', 'br'];
      offset = [m, -s];
      break;
    case 'bottomLeft':
      align = ['tl', 'bl'];
      offset = [s, m];
      break;
    case 'bottom':
      align = ['tc', 'bc'];
      offset = [s, m];
      break;
    case 'bottomRight':
      align = ['tr', 'br'];
      offset = [-s, m];
      break;
    default:
      break;
  }
  domAlign(source, target, {
    points: align,
    offset: offset,
    overflow: {
      adjustX: true,
      adjustY: true,
      resizeWidth: resize,
      resizeHeight: resize
    }
  });
};
