// @ali/we-design/dist-web/components/layout/Layout.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { SidePanel } from "./SidePanel";
var Container = /*#__PURE__*/styled.div(["display:flex;flex-direction:column;height:100vh;"]);
var Main = /*#__PURE__*/styled.div(["flex:1;display:flex;flex-direction:row;height:0;"]);
var Center = /*#__PURE__*/styled.div(["flex:1;width:0;"]);
export var Layout = function Layout(_ref) {
  var leftPanel = _ref.leftPanel,
      allowLeftPanelResize = _ref.allowLeftPanelResize,
      allowRightPanelResize = _ref.allowRightPanelResize,
      _ref$maxLeftPanelWidt = _ref.maxLeftPanelWidth,
      maxLeftPanelWidth = _ref$maxLeftPanelWidt === void 0 ? Infinity : _ref$maxLeftPanelWidt,
      _ref$minLeftPanelWidt = _ref.minLeftPanelWidth,
      minLeftPanelWidth = _ref$minLeftPanelWidt === void 0 ? 0 : _ref$minLeftPanelWidt,
      leftPanelWidth = _ref.leftPanelWidth,
      rightPanelWidth = _ref.rightPanelWidth,
      _ref$maxRightPanelWid = _ref.maxRightPanelWidth,
      maxRightPanelWidth = _ref$maxRightPanelWid === void 0 ? Infinity : _ref$maxRightPanelWid,
      _ref$minRightPanelWid = _ref.minRightPanelWidth,
      minRightPanelWidth = _ref$minRightPanelWid === void 0 ? 0 : _ref$minRightPanelWid,
      rightPanel = _ref.rightPanel,
      topPanel = _ref.topPanel,
      bottomPanel = _ref.bottomPanel,
      children = _ref.children,
      allowLeftPanelClose = _ref.allowLeftPanelClose,
      allowRightPanelClose = _ref.allowRightPanelClose,
      onLeftPanelClose = _ref.onLeftPanelClose,
      onRightPanelClose = _ref.onRightPanelClose,
      onSidePanelWidthChange = _ref.onSidePanelWidthChange,
      style = _ref.style;
  return /*#__PURE__*/_createElement(Container, {
    style: style
  }, topPanel ? /*#__PURE__*/_createElement("div", null, topPanel) : null, /*#__PURE__*/_createElement(Main, null, leftPanel ? /*#__PURE__*/_createElement(SidePanel, {
    closable: allowLeftPanelClose,
    onClose: onLeftPanelClose,
    width: leftPanelWidth,
    isLeft: true,
    onPanelWidthChange: onSidePanelWidthChange,
    maxWidth: maxLeftPanelWidth,
    minWidth: minLeftPanelWidth,
    allowResize: allowLeftPanelResize
  }, leftPanel) : null, /*#__PURE__*/_createElement(Center, null, children), rightPanel ? /*#__PURE__*/_createElement(SidePanel, {
    closable: allowRightPanelClose,
    onClose: onRightPanelClose,
    width: rightPanelWidth,
    maxWidth: maxRightPanelWidth,
    minWidth: minRightPanelWidth,
    onPanelWidthChange: onSidePanelWidthChange,
    allowResize: allowRightPanelResize
  }, rightPanel) : null), bottomPanel ? /*#__PURE__*/_createElement("div", null, bottomPanel) : null);
};
