// @ali/4ever-bi/dist-web/plugins/columns/mobile/index.js
import { ColumnsPlugin } from '@ali/4ever-bamboo';
import setColumnsItemData from "../commands/setColumnsItemData";
var hooks = ColumnsPlugin.hooks,
    createQueries = ColumnsPlugin.createQueries;

function createColumnsPlugin(allConfigs) {
  var createRenderNode = hooks.createRenderNode,
      createOnAction = hooks.createOnAction,
      createOnCangjieSelect = hooks.createOnCangjieSelect,
      createOnKeyDown = hooks.createOnKeyDown,
      createPlaceholder = hooks.createPlaceholder,
      createOnCut = hooks.createOnCut,
      beforeCopySelectedNodes = hooks.beforeCopySelectedNodes;
  var configs = allConfigs.columns;
  return {
    commands: {
      setColumnsItemData: setColumnsItemData
    },
    queries: createQueries(configs),
    renderNode: createRenderNode(configs),
    onAction: createOnAction(configs),
    onCangjieSelect: createOnCangjieSelect(configs),
    onKeyDown: createOnKeyDown(),
    onCut: createOnCut(),
    // @ts-ignore
    placeholder: createPlaceholder(configs),
    beforeCopySelectedNodes: beforeCopySelectedNodes
  };
}

export default createColumnsPlugin;
