// @ali/we-ui/dist-web/components/find-replace/styled.js
import styled, { css } from 'styled-components';
import { withTheme } from "../theme";
import { dropdownBoxStyle, Placeholder } from "../../utils/styled.mixin";
export var ModalWrapper = withTheme(styled.div(["", " padding:0px 16px 16px 16px;box-sizing:border-box;width:350px;color:#414C56;font-size:14px;box-shadow:", ";z-index:2;user-select:none;@media print{display:none;}"], dropdownBoxStyle, function (_ref) {
  var focused = _ref.focused;
  return focused ? '0 2px 10px rgba(0, 0, 0, 0.12)' : '';
}));
export var ModalInput = withTheme(styled.input(["box-sizing:border-box;position:relative;display:inline-block;outline:0;list-style:none;width:100%;height:32px;padding:4px 11px;background-image:none;border:1px solid #d9d9d9;border-radius:4px;flex-grow:1;&:hover{border:solid 1px #3296fa;}", ""], function (_ref2) {
  var theme = _ref2.theme;
  return Placeholder(theme.ui.colorText2);
}));
export var ModalText = /*#__PURE__*/styled.div(["user-select:none;display:inline-block;text-align:center;"]);
export var ModalCloseButton = withTheme(styled.span(["position:absolute;top:16px;right:16px;cursor:pointer;color:", ";&:hover{color:#3296fa;}"], function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.commonBaseColor3;
}));
export var PanelWrapper = withTheme(styled.div(["display:flex;flex-direction:column;"]));
var operationPanelStyle = /*#__PURE__*/css(["display:flex;flex-direction:row;justify-content:space-between;align-items:center;"]);
export var OperationPanel = /*#__PURE__*/styled.div(["", " margin-top:12px;padding:8px 0px;"], operationPanelStyle);
export var FlexItemOperationPanel = /*#__PURE__*/styled.div(["", " margin-left:8px;"], operationPanelStyle);
export var FocusPanel = /*#__PURE__*/styled.div(["display:flex;flex-direction:row;align-items:center;justify-content:space-between;flex:none;width:116px;margin-left:auto;"]);
export var OperationButtonWrapper = /*#__PURE__*/styled.div(["display:flex;flex-direction:row;"]);
export var Description = withTheme(styled.div(["color:", ";font-size:14px;text-align:left;padding:8px 0px;"], function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.colorText4;
}));
var rowWrapperStyle = /*#__PURE__*/css(["margin-top:8px;"]);
export var RowWrapper = /*#__PURE__*/styled.div(["", ""], rowWrapperStyle);
export var FlexRowWrapper = /*#__PURE__*/styled.div(["", " display:flex;"], rowWrapperStyle);
