// @ali/4ever-bi/dist-node/plugins/embed/mobile/embed.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everFactory = require("@ali/4ever-factory");

const _createElement = /*#__PURE__*/_react.default.createElement;
const {
  MobileEmbedCard
} = _everBamboo.EmbedPlugin;

class Embed extends _react.PureComponent {
  constructor(...args) {
    super(...args);
    this.ref = /*#__PURE__*/_react.default.createRef();
    this.previewRef = /*#__PURE__*/_react.default.createRef();

    this.getNodeData = () => {
      const {
        node
      } = this.props;
      const {
        data
      } = node || {};
      return data;
    };
  }

  renderCard() {
    const {
      name,
      src,
      type,
      size
    } = this.getNodeData();
    const {
      onPreview
    } = this.props;
    return /*#__PURE__*/_createElement(MobileEmbedCard, {
      name: name,
      onPreview: onPreview,
      src: src,
      type: type,
      size: size
    });
  }

  renderInlineToolbar() {
    const {
      controller,
      node,
      locale
    } = this.props;
    return /*#__PURE__*/_createElement(_everFactory.EmbedInlineToolbar, {
      controller: controller,
      node: node,
      locale: locale,
      triggerNode: this.ref.current
    });
  }

  render() {
    const {
      attributes,
      isSelected
    } = this.props;
    return /*#__PURE__*/_createElement(_react.default.Fragment, null, isSelected ? this.renderInlineToolbar() : null, /*#__PURE__*/_createElement("div", (0, _extends2.default)({}, attributes, {
      ref: this.ref
    }), this.renderCard()));
  }

}

const Container = props => {
  const scrollableContainer = (0, _everBamboo.useScrollableContainer)();
  return /*#__PURE__*/_createElement(Embed, (0, _extends2.default)({}, props, {
    scrollableContainer: scrollableContainer
  }));
};

var _default = Container;
exports.default = _default;
