// @ali/4ever-bi/dist-node/plugins/laserPointer/createRenderNode.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderNode;

var React = _interopRequireWildcard(require("react"));

var _highlight = _interopRequireDefault(require("./components/highlight"));

var _utils = require("./utils");

const _createElement = /*#__PURE__*/React.createElement;

function createRenderNode(config) {
  const renderNode = (props, controller, next) => {
    const {
      node,
      attributes
    } = props;

    if (!config?.isLaserPointerEnabled || !config.isLaserPointerEnabled()) {
      return next();
    }

    const {
      getRole,
      renderHighlight,
      getHighlightType
    } = config;
    const type = getHighlightType(node, controller);

    if (type) {
      const highlighted = (0, _utils.isHighlighted)(node.key, controller.value.injections);
      return /*#__PURE__*/_createElement(_highlight.default, {
        type: type,
        controller: controller,
        node: node,
        domRef: attributes.ref,
        getRole: getRole,
        highlighted: highlighted,
        renderHighlight: renderHighlight
      }, next());
    }

    return next();
  };

  return renderNode;
}
