// @ali/4ever-plugin-tag/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.default;
  }
});
Object.defineProperty(exports, "Tag", {
  enumerable: true,
  get: function () {
    return _models.default;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.createServerSerializationPlugin;
  }
});
Object.defineProperty(exports, "createSerializationWithFormulaPlugin", {
  enumerable: true,
  get: function () {
    return _with.default;
  }
});
Object.defineProperty(exports, "useResetWidthByParentWidth", {
  enumerable: true,
  get: function () {
    return _useParentWidth.useResetWidthByParentWidth;
  }
});
Object.defineProperty(exports, "TagWrapper", {
  enumerable: true,
  get: function () {
    return _styled.TagWrapper;
  }
});
exports.OldBambooPlugin = void 0;

var _index = _interopRequireDefault(require("./mo/index"));

var _models = _interopRequireDefault(require("./mo/models"));

var _index2 = _interopRequireWildcard(require("./mo/serializer/index"));

var _with = _interopRequireDefault(require("./mo/serializer/with.formula"));

var _useParentWidth = require("./utils/useParentWidth");

var _styled = require("./components/styled");

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;
