// @ali/4ever-plugin-code/dist-node/mo/serializer/htmlToJsonML/parseFromWeb.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

var _constants = require("./constants");

var _utils = require("./utils");

// 识别如下两种结构作为代码块
// <pre><code>xxxxx</code></pre>
// <pre>xxxx</pre>
function parseFromWeb() {
  let index = 0;
  return {
    name: 'code',

    onOpenTag(state, name) {
      if (name !== 'pre') return false;
      state.push(['code', {
        syntax: _constants.PLAIN_TEXT,
        theme: _constants.THEME_DEFAULT
      }]);
      return true;
    },

    onCloseTag(state, name) {
      if (name !== 'pre') return false;
      state.pop();
      const codeBlockDom = state.htmlDom.getElementsByTagName('pre')[index];
      state.append(['code', {
        syntax: (0, _utils.getSyntax)(codeBlockDom),
        theme: _constants.THEME_DEFAULT
      }, (0, _everCangjie.createTextWithString)((0, _utils.getTextContent)(codeBlockDom) || '')]);
      index++;
      return true;
    }

  };
}

var _default = parseFromWeb;
exports.default = _default;
