// @ali/4ever-plugin-code/dist-node/mo/serializer/htmlToJsonML/parseFromYuQue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

var _constants = require("./constants");

function parseFromYuQue() {
  let index = 0;
  return {
    name: 'code',

    onOpenTag(state, name, attr) {
      if (attr[_constants.DATA_LAKE_CARD] !== _constants.CODEBLOCK) return false;
      state.push(['code', {
        syntax: attr[_constants.DATA_LANGUAGE] || _constants.PLAIN_TEXT,
        theme: _constants.THEME_DEFAULT
      }]);
      return true;
    },

    onCloseTag(state, name, attr) {
      if (attr[_constants.DATA_LAKE_CARD] !== _constants.CODEBLOCK) return false;
      state.pop();
      const codeBlockDom = state.htmlDom.querySelectorAll(`[${_constants.DATA_LAKE_CARD}=${_constants.CODEBLOCK}]`)[index];
      state.append(['code', {
        syntax: attr[_constants.DATA_LANGUAGE] || _constants.PLAIN_TEXT,
        theme: _constants.THEME_DEFAULT
      }, (0, _everCangjie.createTextWithString)(codeBlockDom?.textContent || '')]);
      index++;
      return true;
    }

  };
}

var _default = parseFromYuQue;
exports.default = _default;
