// @ali/4ever-plugin-code/dist-web/bi/actions.js
export var CODE_CHANGE = Symbol('change-code');
export var CODE_FOCUS_TO_CODE_EDITOR = Symbol('focus-to-code-editor');
export var CODE_FOCUS_TO_TEXT_EDITOR = Symbol('focus-to-text-editor');
export var CODE_REMOVE = Symbol('remove-code-block');
export var CODE_INSERT = Symbol('insert-code-block');
export var CODE_CHANGE_DATA = Symbol('change-code-data');
export var CODE_BLUR = Symbol('blur-code');
export var CODE_TRANSFORM = Symbol('code-transform');
export function changeCode(payload) {
  return {
    type: CODE_CHANGE,
    payload: payload
  };
}
export function changeCodeData(node, data) {
  return {
    type: CODE_CHANGE_DATA,
    payload: {
      data: data,
      node: node
    }
  };
}
export function focusToCodeEditor(node) {
  return {
    type: CODE_FOCUS_TO_CODE_EDITOR,
    payload: {
      node: node
    }
  };
}
export function focusToTextEditor(node, direction) {
  return {
    type: CODE_FOCUS_TO_TEXT_EDITOR,
    payload: {
      direction: direction,
      node: node
    }
  };
}
export function removeCode(node) {
  return {
    type: CODE_REMOVE,
    payload: {
      node: node
    }
  };
}
export function insertCode(payload) {
  if (payload === void 0) {
    payload = {};
  }

  return {
    type: CODE_INSERT,
    payload: payload
  };
}
export function transformCode() {
  return {
    type: CODE_TRANSFORM,
    payload: {}
  };
}
export function cancelCode() {
  return {
    type: CODE_TRANSFORM,
    payload: {}
  };
}
export function blurCode() {
  return {
    type: CODE_BLUR,
    payload: {}
  };
}
