// @ali/4ever-plugin-code/dist-web/bi/createToolbar.js
import ToolbarButton from "./components/ToolbarButton";
import LanguageSelect from "./components/ToolbarLanguageSelect";
import ThemeSelect from "./components/ToolbarThemeSelect";
import ToolbarLineWrap from "./components/ToolbarLineWrap";
import ToolbarMore from "./components/ToolbarMore";
export default function createToolbar(config) {
  return {
    buttons: {
      "default": ToolbarButton,
      button: ToolbarButton,
      syntax: LanguageSelect,
      theme: ThemeSelect,
      wrap: ToolbarLineWrap,
      more: ToolbarMore
    },
    locale: config.locale,
    getTarget: function getTarget(controller) {
      var _controller$value = controller.value,
          selection = _controller$value.selection,
          document = _controller$value.document;

      if (selection && selection.isCollapsed) {
        var block = document.getClosestBlock(selection.focus.key);

        if ((block == null ? void 0 : block.type) === 'code') {
          return 'code';
        }
      }

      return undefined;
    }
  };
}
