// @ali/4ever-mo/dist-web/plugins/index.js
import { createSerializationPlugin as marks } from '@ali/4ever-plugin-marks';
import { createSerializationPlugin as bold } from '@ali/4ever-plugin-bold';
import { createSerializationPlugin as blockquote } from '@ali/4ever-plugin-blockquote';
import { createSerializationPlugin as br } from '@ali/4ever-plugin-br';
import { createSerializationPlugin as code } from '@ali/4ever-plugin-code';
import { createSerializationPlugin as heading } from '@ali/4ever-plugin-heading';
import { createSerializationPlugin as hr } from '@ali/4ever-plugin-hr';
import { createSerializationPlugin as image } from '@ali/4ever-plugin-image';
import { createSerializationPlugin as italic } from '@ali/4ever-plugin-italic';
import { createSerializationPlugin as link } from '@ali/4ever-plugin-link';
import { createSerializationPlugin as list } from '@ali/4ever-plugin-list';
import { createSerializationPlugin as mention } from '@ali/4ever-plugin-mention';
import { createSerializationPlugin as onlineVideo } from '@ali/4ever-plugin-online-video';
import { createSerializationPlugin as textBackground } from '@ali/4ever-plugin-text-background';
import { createSerializationPlugin as formula } from '@ali/4ever-plugin-formula';
import { createSerializationPlugin as paragraph } from '@ali/4ever-plugin-paragraph';
import { createSerializationPlugin as placeholder } from '@ali/4ever-plugin-placeholder';
import { createSerializationPlugin as inlinePlaceholder } from '@ali/4ever-plugin-inline-placeholder';
import { createSerializationPlugin as refer } from '@ali/4ever-plugin-refer';
import { createSerializationPlugin as sticker } from '@ali/4ever-plugin-sticker';
import { createSerializationPlugin as strike } from '@ali/4ever-plugin-strike';
import { createSerializationPlugin as vertAlign } from '@ali/4ever-plugin-vert-align';
import { createSerializationPlugin as underline } from '@ali/4ever-plugin-underline';
import { createSerializationPlugin as table } from '@ali/4ever-plugin-table';
import { createSerializationPlugin as card } from '@ali/4ever-plugin-card';
import { createSerializationPlugin as tag, createSerializationWithFormulaPlugin as tagWithFormula } from '@ali/4ever-plugin-tag';
import { createSerializationPlugin as embed } from '@ali/4ever-plugin-embed';
import { createSerializationPlugin as container } from '@ali/4ever-plugin-container';
import { createSerializationPlugin as colorBlock } from '@ali/4ever-plugin-color-blocks';
import { createSerializationPlugin as templateButton } from '@ali/4ever-plugin-template-button';
import { createSerializationPlugin as refblock } from '@ali/4ever-plugin-refblock';
import { createSerializationPlugin as inlineCode } from '@ali/4ever-plugin-inline-code';
export { marks, bold, blockquote, code, br, tag, tagWithFormula, card, embed, heading, hr, image, italic, textBackground, formula // container 的子类型要放在 container 前面。container 有兜底逻辑
, templateButton // child of container
, colorBlock // child of container
, container, link, list, mention, onlineVideo, refer, table, strike, inlineCode, vertAlign, underline, sticker, placeholder, inlinePlaceholder, refblock, paragraph };
export default {
  marks: marks,
  bold: bold,
  formula: formula,
  blockquote: blockquote,
  code: code,
  br: br,
  tag: tag,
  card: card,
  embed: embed,
  heading: heading,
  hr: hr,
  image: image,
  italic: italic,
  colorBlock: colorBlock,
  textBackground: textBackground,
  container: container,
  link: link,
  list: list,
  mention: mention,
  onlineVideo: onlineVideo,
  refer: refer,
  table: table,
  strike: strike,
  inlineCode: inlineCode,
  vertAlign: vertAlign,
  underline: underline,
  sticker: sticker,
  placeholder: placeholder,
  inlinePlaceholder: inlinePlaceholder,
  refblock: refblock,
  // paragraph 放在次后
  paragraph: paragraph
};
