// @ali/4ever-plugin-vote-card/dist-web/utils/actions/index.js
import _extends from "@babel/runtime/helpers/extends";
export var VOTING_ACTIVE = Symbol('VOTE_ACTIVE');
export var VOTE_EDITING = Symbol('VOTE_EDITING');
export var VOTE_CANCEL_EDITING = Symbol('VOTE_CANCEL_EDITING');
export var VOTE_CANCEL_VOTING = Symbol('VOTE_CANCEL_VOTING');
export var VOTE_SAVE = Symbol('VOTE_SAVE');
export var VOTE_DELETE = Symbol('VOTE_DELETE');
export var VOTE_COPY = Symbol('VOTE_COPY');
export var VOTE_CUT = Symbol('VOTE_CUT');
export var VOTE_SUBMIT = Symbol('VOTE_SUBMIT');
export var VOTE_MODIFY = Symbol('VOTE_MODIFY');
export var VOTE_SAVE_CP = Symbol('VOTE_SAVE_CP');
export function votingActive(node) {
  return {
    type: VOTING_ACTIVE,
    payload: {
      node: node
    }
  };
}
export function voteEditing(node) {
  return {
    type: VOTE_EDITING,
    payload: {
      node: node
    }
  };
}
export function cancelEditing(node, data) {
  return {
    type: VOTE_CANCEL_EDITING,
    payload: _extends({
      node: node
    }, data)
  };
}
export function voteSave(_ref) {
  var node = _ref.node,
      data = _ref.data,
      setData = _ref.setData;
  return {
    type: VOTE_SAVE,
    payload: _extends({
      node: node,
      setData: setData
    }, data)
  };
}
export function voteDelete(nodeKey) {
  return {
    type: VOTE_DELETE,
    payload: {
      key: nodeKey
    }
  };
}
export function voteCopy(nodeKey) {
  return {
    type: VOTE_COPY,
    payload: {
      key: nodeKey
    }
  };
}
export function voteCut(nodeKey) {
  return {
    type: VOTE_CUT,
    payload: {
      key: nodeKey
    }
  };
}
export function voteSubmit(payload) {
  return {
    type: VOTE_SUBMIT,
    payload: payload
  };
}
export function voteModify(nodeKey) {
  return {
    type: VOTE_MODIFY,
    payload: {
      key: nodeKey
    }
  };
}
export function voteSaveCP(nodeKey, payload) {
  return {
    type: VOTE_SAVE_CP,
    payload: {
      key: nodeKey,
      payload: payload
    }
  };
}
export function cancelVoting(node) {
  return {
    type: VOTE_CANCEL_VOTING,
    payload: {
      node: node
    }
  };
}
