// @ali/4ever-plugin-table/dist-web/bi/components/tableSelection/index.js
import React, { useEffect, useState } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import equal from 'fast-deep-equal';
import { useSelectingHots, useSelectionData } from '@ali/4ever-cangjie';
import { THEME } from "../../constants";
import SelectionRect from "./selectionRect";
import { usePixelColsWidth } from "../../../utils/hooks/usePixelColsWidth";

var TableSelection = function TableSelection(props) {
  var node = props.node,
      controller = props.controller,
      propSelection = props.selection,
      isHighlight = props.isHighlight,
      tableRef = props.tableRef,
      _props$scale = props.scale,
      scale = _props$scale === void 0 ? 1 : _props$scale,
      getTableSelection = props.getTableSelection,
      forceUpdate = props.forceUpdate;

  var _usePixelColsWidth = usePixelColsWidth(),
      colsWidth = _usePixelColsWidth[0];

  var _useState = useState(propSelection),
      selection = _useState[0],
      setSelection = _useState[1];

  useEffect(function () {
    setSelection(propSelection);
  }, [propSelection]);
  useSelectionData(controller.selectionData$, null, function () {
    var tempSelection = getTableSelection();

    if (!equal(selection, tempSelection)) {
      setSelection(tempSelection);
    }
  }, []);
  var onHots = React.useCallback(function (hots) {
    // PERF: 非当前表格热区不处理
    if (hots.some(function (hot) {
      return node.key === hot.node.key || node.hasNode(hot.node.key);
    })) {
      var tableSelection = getTableSelection();
      setSelection(tableSelection);
    }
  }, [node, getTableSelection]);
  useSelectingHots(controller, onHots);
  return /*#__PURE__*/_createElement(SelectionRect, {
    testid: "table-selection-wrapper",
    table: node,
    tableRef: tableRef,
    tableSelection: selection,
    scale: scale,
    colsWidth: colsWidth,
    style: {
      backgroundColor: isHighlight ? THEME.red0 : THEME.blue0,
      borderColor: isHighlight ? THEME.red1 : THEME.blue3
    },
    forceUpdate: forceUpdate
  });
};

export default TableSelection;
