// @ali/4ever-component/dist-web/Toolbar/interface.js
/**
 * 编辑器提供的默认工具栏interface
 * 数据结构参照 we-ui toolbar：
 * https://pds.alibaba-inc.com/alidocs/we-ui/?path=/docs/toolbar-%E7%A4%BA%E4%BE%8B--resize
 * 单行配置(嵌套数组)：
 * SingleToolbarProps: [{
 *    toolButtons: [{
 *      name: 'aaa'
 *    }, {
 *      name: 'bbb'
 *    }]
 *  }, {
 *    toolButtons: [{
 *      name: 'ccc'
 *    }]
 *  }]
 * 显示效果：|- 工具栏 -|
 * |- aaa bbb | ccc -|
 * **********************************************
 * 双行配置(2层嵌套数组)：
 * DoubleToolbarProps: [{
 *    toolRows: [{
 *      toolButtons: [{
 *        name: 'aaa'
 *      }, {
 *        name: 'bbb'
 *      }]
 *    }, {
 *      toolButtons: [{
 *        name: 'ccc'
 *      }, {
 *        name: 'ddd'
 *      }]
 *    }]
 *  }, {
 *    toolRows: [{
 *      toolButtons: [{
 *        name: 'eee',
 *        boxLayout: ToolBoxLayoutMode.vertical
 *      }]
 *    }]
 *  }]
 * 显示效果：|- 工具栏 -|
 * |- aaa bbb |     -|
 * |-         | eee -|
 * |- ccc ddd |     -|
 */
import { EToolbarMode, EToolBoxLayout } from '@ali/we-ui';
/**
 * 工具栏按钮布局枚举
 */

export var ToolBoxLayoutMode;
/**
 * 单个工具栏组件配置
 */

(function (ToolBoxLayoutMode) {
  ToolBoxLayoutMode[ToolBoxLayoutMode["vertical"] = EToolBoxLayout.Vertical] = "vertical";
  ToolBoxLayoutMode[ToolBoxLayoutMode["single"] = EToolBoxLayout.Single] = "single";
  ToolBoxLayoutMode[ToolBoxLayoutMode["horizontal"] = EToolBoxLayout.Horizontal] = "horizontal";
})(ToolBoxLayoutMode || (ToolBoxLayoutMode = {}));

/**
 * 工具栏显示模式
 */
export var ToolbarMode;

(function (ToolbarMode) {
  ToolbarMode[ToolbarMode["double"] = EToolbarMode.doubleLine] = "double";
  ToolbarMode[ToolbarMode["single"] = EToolbarMode.singleLine] = "single";
  ToolbarMode[ToolbarMode["inline"] = 2] = "inline";
})(ToolbarMode || (ToolbarMode = {}));
