// @ali/4ever-plugin-code/dist-web/mo/serializer/htmlToJsonML/parseFromYuQue.js
import { createTextWithString } from '@ali/4ever-cangjie';
import { CODEBLOCK, DATA_LAKE_CARD, DATA_LANGUAGE, PLAIN_TEXT, THEME_DEFAULT } from "./constants";

function parseFromYuQue() {
  var index = 0;
  return {
    name: 'code',
    onOpenTag: function onOpenTag(state, name, attr) {
      if (attr[DATA_LAKE_CARD] !== CODEBLOCK) return false;
      state.push(['code', {
        syntax: attr[DATA_LANGUAGE] || PLAIN_TEXT,
        theme: THEME_DEFAULT
      }]);
      return true;
    },
    onCloseTag: function onCloseTag(state, name, attr) {
      if (attr[DATA_LAKE_CARD] !== CODEBLOCK) return false;
      state.pop();
      var codeBlockDom = state.htmlDom.querySelectorAll("[" + DATA_LAKE_CARD + "=" + CODEBLOCK + "]")[index];
      state.append(['code', {
        syntax: attr[DATA_LANGUAGE] || PLAIN_TEXT,
        theme: THEME_DEFAULT
      }, createTextWithString((codeBlockDom == null ? void 0 : codeBlockDom.textContent) || '')]);
      index++;
      return true;
    }
  };
}

export default parseFromYuQue;
