// @ali/we-design-next/dist-node/display/modal/methods.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.warn = exports.success = exports.info = exports.confirm = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactDom = _interopRequireDefault(require("react-dom"));
var _button = require("../../basic/button");
var _Modal = require("./Modal");
var _ModalFooter = require("./ModalFooter");
var _text = require("../../basic/text");
var _div;
const _createElement = /*#__PURE__*/_react.default.createElement;
const iconInfo = /*#__PURE__*/_createElement("svg", {
  width: "44",
  height: "44",
  viewBox: "0 0 44 44",
  fill: "none",
  stroke: "none",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/_createElement("circle", {
  cx: "22",
  cy: "22",
  r: "20",
  fill: "#0089FF"
}), /*#__PURE__*/_createElement("path", {
  d: "M20 22C20 20.8954 20.8954 20 22 20C23.1046 20 24 20.8954 24 22V32C24 33.1046 23.1046 34 22 34C20.8954 34 20 33.1046 20 32V22Z",
  fill: "white"
}), /*#__PURE__*/_createElement("circle", {
  cx: "22",
  cy: "14",
  r: "3",
  fill: "white"
}));
const iconWarning = /*#__PURE__*/_createElement("svg", {
  width: "44",
  height: "44",
  viewBox: "0 0 44 44",
  fill: "none",
  stroke: "none",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/_createElement("circle", {
  cx: "22",
  cy: "22",
  r: "20",
  fill: "#FF9200"
}), /*#__PURE__*/_createElement("path", {
  d: "M20 13C20 11.8954 20.8954 11 22 11C23.1046 11 24 11.8954 24 13V23C24 24.1046 23.1046 25 22 25C20.8954 25 20 24.1046 20 23V13Z",
  fill: "white"
}), /*#__PURE__*/_createElement("circle", {
  cx: "22",
  cy: "31",
  r: "3",
  fill: "white"
}));
const iconSuccess = /*#__PURE__*/_createElement("svg", {
  width: "44",
  height: "44",
  viewBox: "0 0 44 44",
  fill: "none",
  stroke: "none",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/_createElement("circle", {
  cx: "22",
  cy: "22",
  r: "20",
  fill: "#00B042"
}), /*#__PURE__*/_createElement("path", {
  d: "M31 17L20 28L14 22",
  stroke: "white",
  "stroke-width": "4",
  "stroke-linecap": "round",
  "stroke-linejoin": "round"
}));
const ModalMethod = /*#__PURE__*/_react.default.memo(props => {
  const {
    type,
    message,
    title,
    options = [],
    confirmText,
    cancelText,
    onConfirm,
    onCancel,
    mask
  } = props;
  return /*#__PURE__*/_createElement(_Modal.Modal, {
    style: {
      width: 360
    },
    visible: true,
    header: null,
    content: /*#__PURE__*/_createElement("div", {
      style: {
        display: 'flex'
      }
    }, /*#__PURE__*/_createElement("div", {
      style: {
        width: '40px',
        marginTop: 10
      }
    }, type === 'info' || type === 'confirm' ? iconInfo : type === 'success' ? iconSuccess : type === 'warning' ? iconWarning : null), /*#__PURE__*/_createElement("div", {
      style: {
        paddingLeft: '16px'
      }
    }, title ? /*#__PURE__*/_createElement(_text.H1, null, title) : null, /*#__PURE__*/_createElement(_text.Text, null, message), options.length > 0 ? /*#__PURE__*/_createElement("div", null, options.map(opt => _div || (_div = /*#__PURE__*/_createElement("div", null)))) : null)),
    footer: /*#__PURE__*/_createElement(_ModalFooter.ModalFooter, {
      cancel: type === 'confirm' || type === 'warning' ? /*#__PURE__*/_createElement(_button.Button, {
        children: cancelText || undefined,
        onClick: () => {
          onCancel && onCancel();
          destory();
        }
      }) : null,
      confirm: /*#__PURE__*/_createElement(_button.Button, {
        children: confirmText || undefined,
        onClick: () => {
          onConfirm && onConfirm();
          destory();
        }
      })
    }),
    mask: mask
  });
});
const method = options => {
  _reactDom.default.render( /*#__PURE__*/_createElement(ModalMethod, options), ensureContainer());
};
const destory = () => {
  _reactDom.default.render( /*#__PURE__*/_createElement(_react.default.Fragment, null), ensureContainer());
};
const ensureContainer = () => {
  const id = 'wdn-modal-method-root';
  const divEle = document.querySelector(`div#${id}`);
  if (!divEle) {
    const ele = document.createElement('div');
    ele.id = id;
    ele.style.position = 'absolute';
    ele.style.width = '100%';
    ele.style.top = '0px';
    ele.style.left = '0px';
    document.body.appendChild(ele);
  }
  return document.querySelector(`div#${id}`);
};
const info = (message, options = {}) => method({
  type: 'info',
  message,
  ...options
});
exports.info = info;
const warn = (message, options = {}) => method({
  type: 'warning',
  message,
  ...options
});
exports.warn = warn;
const success = (message, options = {}) => method({
  type: 'success',
  message,
  ...options
});
exports.success = success;
const confirm = (message, options = {}) => method({
  type: 'confirm',
  message,
  ...options
});
exports.confirm = confirm;
