// @ali/we-design-next/dist-node/display/modal/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModalMask = exports.ModalHeaderTitle = exports.ModalHeaderClose = exports.ModalHeader = exports.ModalFooterAction = exports.ModalFooter = exports.ModalContent = exports.ModalContainer = exports.MoadlRoot = void 0;
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _styled = require("../../basic/text/styled");
var _common = require("../../common");
const MoadlRoot = /*#__PURE__*/_styledComponents.default.div(["box-sizing:border-box;"]);
exports.MoadlRoot = MoadlRoot;
const ModalContainer = /*#__PURE__*/_styledComponents.default.div(["", " background-color:", ";box-sizing:border-box;position:fixed;left:0;top:0;max-width:calc(100vw - 32px);min-width:300px;margin:0 auto;border:0;border-radius:4px;padding:16px;transform:translate(-9999px,-9999px);z-index:600;"], _common.CSS_BOX_SHADOWS.L, _common.COLORS.LIGHT.DAVINCI_BG_NEW);
exports.ModalContainer = ModalContainer;
const ModalMask = /*#__PURE__*/_styledComponents.default.div(["display:block;position:fixed;top:0;right:0;bottom:0;left:0;z-index:500;height:100%;background-color:", ";"], ({
  isTransparent
}) => isTransparent ? _common.COLORS.TRANSPARENT : 'rgba(0, 0, 0, 0.4)');
exports.ModalMask = ModalMask;
const ModalHeader = /*#__PURE__*/_styledComponents.default.div(["", " display:flex;align-items:center;font-weight:500;user-select:none;line-height:16px;"], _styled.CSS_H1);
exports.ModalHeader = ModalHeader;
const ModalHeaderTitle = /*#__PURE__*/_styledComponents.default.div(["flex-grow:1;padding-bottom:16px;", ""], ({
  move
}) => move ? `cursor: move;` : null);
exports.ModalHeaderTitle = ModalHeaderTitle;
const ModalHeaderClose = /*#__PURE__*/_styledComponents.default.div(["margin-bottom:16px;& > .wdn-icon{border-radius:4px;transition:background 0.2s ease-out;}& > .wdn-icon:hover{background:rgba(228,230,233,1);}"]);
exports.ModalHeaderClose = ModalHeaderClose;
const ModalContent = /*#__PURE__*/_styledComponents.default.div(["min-height:100px;", ""], _styled.CSS_TEXT);
exports.ModalContent = ModalContent;
const ModalFooter = /*#__PURE__*/_styledComponents.default.div(["padding-top:16px;display:flex;justify-content:flex-end;align-items:center;& > .wdn-button{margin-left:12px;vertical-align:middle;}"]);
exports.ModalFooter = ModalFooter;
const ModalFooterAction = /*#__PURE__*/_styledComponents.default.div(["flex-grow:1;"]);
exports.ModalFooterAction = ModalFooterAction;
