// @ali/we-design/dist-web/components/message/styled.js
/* eslint-disable @typescript-eslint/explicit-function-return-type */
import styled from 'styled-components';
import { withTheme } from "../theme";
import { Ellipsis } from "../../utils/styled.mixin";
export var ContainerRoot = withTheme(styled.div(["position:fixed;left:0px;top:calc(16px + env(safe-area-inset-top));width:100%;box-sizing:border-box;z-index:", ";font-size:", ";color:", ";*{box-sizing:border-box;}"], function (_ref) {
  var theme = _ref.theme;
  return theme.ui.zIndexMessage;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.fontSize;
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.colorText4;
}));
export var Container = withTheme(styled.div(["position:absolute;left:50%;transform:translate(-50%,0);text-align:center;"]));
export var MessageBox = withTheme(styled.div(["margin-bottom:", ";.animated{transition:all .24s cubic-bezier(.645,.045,.355,1);}&.message-enter{.animated{opacity:0;transform:translate(0,-100%);}}&.message-enter-active{.animated{opacity:1;transform:translate(0,0);}}&.message-leave{.animated{opacity:1;transform:translate(0,0);}}&.message-leave-active{.animated{opacity:0;transform:translate(0,-100%);}}"], function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.margin;
}));
export var MessageWrap = withTheme(styled.div(["display:flex;justify-content:center;align-items:center;height:36px;padding:8px 10px 8px 8px;background-color:rgba(69,72,74,0.9);box-shadow:", ";border:1px solid ", ";border-radius:4px;font-size:13px;width:fit-content;max-width:445px;.icon{margin-right:", ";font-size:20px;&.success-icon{color:", ";}&.info-icon,&.loading-icon{color:", ";}&.error-icon{color:", ";}&.warning-icon{color:", ";}}"], function (_ref5) {
  var theme = _ref5.theme;
  return theme.ui.boxShadow;
}, function (_ref6) {
  var theme = _ref6.theme;
  return theme.ui.borderColor;
}, function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.marginSmall;
}, function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.successColor;
}, function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.infoColor;
}, function (_ref10) {
  var theme = _ref10.theme;
  return theme.ui.errorColor;
}, function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.warningColor;
}));
export var MessageContent = withTheme(styled.span(["line-height:1.5;color:rgba(238,242,248,1);", ""], Ellipsis));
export var MessageFlex = withTheme(styled.span(["display:flex;justify-content:center;align-items:center;max-width:100%;"]));
