// @ali/we-design-next/dist-node/common/domAlign/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  alignElement: true,
  alignPoint: true,
  getRectOfSource: true,
  getRectOfTarget: true,
  getDestinationRect: true,
  alignRect: true
};
Object.defineProperty(exports, "alignElement", {
  enumerable: true,
  get: function () {
    return _alignElement.default;
  }
});
Object.defineProperty(exports, "alignPoint", {
  enumerable: true,
  get: function () {
    return _alignPoint.default;
  }
});
Object.defineProperty(exports, "alignRect", {
  enumerable: true,
  get: function () {
    return _alignRect.alignRect;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "getDestinationRect", {
  enumerable: true,
  get: function () {
    return _alignRect.getDestinationRect;
  }
});
Object.defineProperty(exports, "getRectOfSource", {
  enumerable: true,
  get: function () {
    return _alignRect.getRectOfSource;
  }
});
Object.defineProperty(exports, "getRectOfTarget", {
  enumerable: true,
  get: function () {
    return _alignRect.getRectOfTarget;
  }
});
var _alignElement = _interopRequireDefault(require("./align/alignElement"));
var _alignPoint = _interopRequireDefault(require("./align/alignPoint"));
var _alignRect = require("./align/alignRect");
var _interface = require("./interface");
Object.keys(_interface).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interface[key];
    }
  });
});
var _default = _alignElement.default;
exports.default = _default;
