// @ali/4ever-bi/dist-node/plugins/formula/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _wrapper = _interopRequireDefault(require("./components/wrapper"));

var _view = _interopRequireDefault(require("./components/view"));

var _placeholder = _interopRequireDefault(require("./components/placeholder"));

var _everFactory = require("@ali/4ever-factory");

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _types = require("./types");

require("./fonts.css");

var _everPluginFormula = require("@ali/4ever-plugin-formula");

const _createElement = /*#__PURE__*/_react.default.createElement;

function createRenderNode(config) {
  return function renderNode(props) {
    return /*#__PURE__*/_createElement(_wrapper.default, (0, _extends2.default)({}, props, {
      config: config
    }));
  };
}

const createFormulaPlugin = (0, _everFactory.tagFactory)(config => {
  const renderNode = createRenderNode(config.formula);
  return {
    tagType: _types.TagType.Formula,
    renderNode,
    onAction: _onAction.default
  };
}, _everPluginFormula.moFactory);
createFormulaPlugin.components = {
  FormulaView: _view.default,
  FormulaPlaceholder: _placeholder.default
};
var _default = createFormulaPlugin;
exports.default = _default;
