// @ali/4ever-zhi/dist-web/plugins/comment/mobile.js
import { CommentPlugin } from '@ali/4ever-bamboo';
var hooks = CommentPlugin.hooks,
    types = CommentPlugin.types,
    actions = CommentPlugin.actions,
    commands = CommentPlugin.commands,
    queries = CommentPlugin.queries;
var lazyRenderStrategy = queries.lazyRenderStrategy;
var commentSelectedNodes = actions.commentSelectedNodes;

function createCommentPlugin(allConfigs) {
  var createContextMenu = hooks.createContextMenu,
      createOnKeyDown = hooks.createOnKeyDown,
      createOnMouseMove = hooks.createOnMouseMove,
      createOnAction = hooks.createOnAction,
      createOnChange = hooks.createOnChange,
      createRenderMark = hooks.createRenderMark,
      createRenderNode = hooks.createRenderNode,
      createRenderContent = hooks.createRenderContent,
      createRenderEditable = hooks.createRenderEditable,
      createOnCangjieSelect = hooks.createOnCangjieSelect,
      createOnCangjieCopyOrCut = hooks.createOnCangjieCopyOrCut,
      createGetNodeSelectToolbarItems = hooks.createGetNodeSelectToolbarItems;
  var configs = allConfigs.comment;
  return {
    queries: {
      lazyRenderStrategy: lazyRenderStrategy
    },
    contextMenu: createContextMenu(configs),
    onKeyDown: createOnKeyDown(configs),
    onMouseMove: createOnMouseMove(configs),
    onAction: createOnAction(configs),
    onChange: createOnChange(configs),
    renderMark: createRenderMark(configs),
    renderNode: createRenderNode(configs),
    renderContent: createRenderContent(configs),
    renderEditable: createRenderEditable(configs),
    onCangjieSelect: createOnCangjieSelect(configs),
    onCangjieCopy: createOnCangjieCopyOrCut(),
    onCangjieCut: createOnCangjieCopyOrCut(),
    getNodeSelectToolbarItems: createGetNodeSelectToolbarItems((configs == null ? void 0 : configs.locale) || {}),
    commands: commands,
    actions: {
      commentSelectedNodes: commentSelectedNodes
    }
  };
}

createCommentPlugin.types = types;
createCommentPlugin.actions = actions;
createCommentPlugin.commands = commands;
createCommentPlugin.queries = queries;
export default createCommentPlugin;
