// @ali/4ever-plugin-draggable/dist-node/bi/components/dragHander.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _dargDropdown = _interopRequireDefault(require("./dargDropdown"));

/* eslint-disable react-hooks/exhaustive-deps */
const _createElement = /*#__PURE__*/_react.default.createElement;

var _default = config => {
  return ({
    controller,
    currentBlock,
    scale = 1
  }) => {
    const {
      composing
    } = controller.value;
    const isNewLine = controller.query('isInsertBtnDisplay', {
      node: currentBlock,
      composing
    });
    const isHideDragButton = controller.query('isHideDragButton', {
      node: currentBlock
    });
    const nodeText = currentBlock?.text;
    const showBtn = (nodeText || !isNewLine) && !isHideDragButton;

    if (showBtn) {
      return /*#__PURE__*/_createElement(_dargDropdown.default, {
        controller: controller,
        currentBlock: currentBlock,
        draggableConfig: config.draggable,
        disableDrag: !config.draggable?.enabled,
        scale: scale
      });
    }

    return null;
  };
};

exports.default = _default;
