// @ali/4ever-plugin-draggable/dist-node/bi/components/useTransformMenus.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useTransformMenus;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _weDesign = require("@ali/we-design");

var _everCangjie = require("@ali/4ever-cangjie");

var _everPluginCode = require("@ali/4ever-plugin-code");

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _everPluginHeading = require("@ali/4ever-plugin-heading");

var _everPluginList = require("@ali/4ever-plugin-list");

var _everPluginCallout = require("@ali/4ever-plugin-callout");

var _everPluginBlockquote = require("@ali/4ever-plugin-blockquote");

var _everUtils = require("@ali/4ever-utils");

var _weIcon = require("@ali/we-icon");

var _everPluginSubject = require("@ali/4ever-plugin-subject");

var _everComponent = require("@ali/4ever-component");

var _isConvertSelected = _interopRequireDefault(require("../utils/isConvertSelected"));

const _createElement = /*#__PURE__*/_react.default.createElement;
const commonHeadStyle = {
  padding: '6px 8px',
  height: 'auto',
  fontWeight: 500
};

const Selected = /*#__PURE__*/_styledComponents.default.span(["margin-left:20px;svg{fill:#0089FF;}"]);

var _ref = /*#__PURE__*/_createElement(Selected, null, /*#__PURE__*/_createElement(_weIcon.SelectedNormalNormal, null));

var _ref2 = /*#__PURE__*/_createElement(Selected, null, /*#__PURE__*/_createElement(_weIcon.SelectedNormalNormal, null));

var _ref3 = /*#__PURE__*/_createElement(_weIcon.ChangetoNormal, null);

var _ref4 = /*#__PURE__*/_createElement(_weIcon.DocToolbarTextNormal, null);

var _ref5 = /*#__PURE__*/_createElement(_weIcon.DocToolbarH1Normal, null);

var _ref6 = /*#__PURE__*/_createElement(_weIcon.DocToolbarH2Normal, null);

var _ref7 = /*#__PURE__*/_createElement(_weIcon.DocToolbarH3Normal, null);

var _ref8 = /*#__PURE__*/_createElement(_weIcon.DocToolbarH4Normal, null);

var _ref9 = /*#__PURE__*/_createElement(_weIcon.ToolbarOrderListNormal, null);

var _ref10 = /*#__PURE__*/_createElement(_weIcon.ToolbarUnorderListNormal, null);

var _ref11 = /*#__PURE__*/_createElement(_weIcon.ToolbarCheckBoxNormal, null);

var _ref12 = /*#__PURE__*/_createElement(_weDesign.QuoteBetaNormal, null);

var _ref13 = /*#__PURE__*/_createElement(_everComponent.IconSquareBrackets, null);

var _ref14 = /*#__PURE__*/_createElement(_weIcon.DocHighBlockNormal, null);

function useTransformMenus(locale, currentBlock, useCustomBlock) {
  const controller = _everCangjie.Controller.useController();

  const disableSubjectHeading = (0, _react.useMemo)(() => {
    return (0, _everPluginSubject.isSubjectHeading)(currentBlock);
  }, [currentBlock]);
  const disableMenu = disableSubjectHeading || controller.query('isListSymbolSelected');

  const isCode = _everPluginCode.Code.isCode(currentBlock);

  const extras = (0, _react.useMemo)(() => {
    const extrasMap = {};
    ['paragraph', 'h1', 'h2', 'h3', 'h4', 'olist', 'ulist', 'tlist', 'blockquote', 'code', 'callout'].forEach(key => {
      if (!useCustomBlock && (0, _isConvertSelected.default)(key, controller)) {
        // 代表没有指定 currentBlock，由当前选区决定（选区工具栏场景）
        extrasMap[key] = _ref;
      } else if (currentBlock && (0, _isConvertSelected.default)(key, controller, currentBlock)) {
        extrasMap[key] = _ref2;
      }
    });
    return extrasMap;
  }, [currentBlock, controller.value, useCustomBlock]);
  const menus = (0, _react.useMemo)(() => {
    return {
      type: _everComponent.TYPE_SUB_MENU,
      icon: _ref3,
      title: locale?.transform,
      maxHeight: 550,
      overlayStyle: {
        minWidth: 173,
        overflow: 'auto'
      },
      menus: [{
        type: _everComponent.TYPE_GROUP,
        title: locale?.block,
        menus: [{
          type: _everComponent.TYPE_MENU_ITEM,
          menuKey: 'heading.paragraph',
          title: locale?.paragraph,
          icon: _ref4,
          action: 'format',
          disabled: disableMenu,
          role: _everPluginParagraph.ParagraphRole.paragraph,
          extra: extras['paragraph']
        }, {
          type: _everComponent.TYPE_MENU_ITEM,
          menuKey: 'heading.heading1',
          title: locale?.h1,
          icon: _ref5,
          action: 'format',
          disabled: disableMenu,
          role: _everPluginHeading.HeadingRole.heading1,
          extra: extras['h1'],
          style: {
            fontSize: _everUtils.theme['heading-1']['font-size'],
            lineHeight: _everUtils.theme['heading-1']['line-height'],
            ...commonHeadStyle
          }
        }, {
          type: _everComponent.TYPE_MENU_ITEM,
          title: locale?.h2,
          menuKey: 'heading.heading2',
          icon: _ref6,
          action: 'format',
          disabled: disableMenu,
          role: _everPluginHeading.HeadingRole.heading2,
          extra: extras['h2'],
          style: {
            fontSize: _everUtils.theme['heading-2']['font-size'],
            lineHeight: _everUtils.theme['heading-2']['line-height'],
            ...commonHeadStyle
          }
        }, {
          type: _everComponent.TYPE_MENU_ITEM,
          menuKey: 'heading.heading3',
          title: locale?.h3,
          icon: _ref7,
          action: 'format',
          disabled: disableMenu,
          role: _everPluginHeading.HeadingRole.heading3,
          extra: extras['h3'],
          style: {
            fontSize: _everUtils.theme['heading-3']['font-size'],
            lineHeight: _everUtils.theme['heading-3']['line-height'],
            ...commonHeadStyle
          }
        }, {
          type: _everComponent.TYPE_MENU_ITEM,
          menuKey: 'heading.heading4',
          title: locale?.h4,
          icon: _ref8,
          action: 'format',
          disabled: disableMenu,
          role: _everPluginHeading.HeadingRole.heading4,
          extra: extras['h4'],
          style: {
            fontSize: _everUtils.theme['heading-4']['font-size'],
            lineHeight: _everUtils.theme['heading-4']['line-height'],
            ...commonHeadStyle
          }
        }]
      }, {
        type: _everComponent.TYPE_GROUP,
        title: locale?.list,
        menus: [{
          type: _everComponent.TYPE_MENU_ITEM,
          menuKey: 'list.olist',
          title: locale?.olist,
          icon: _ref9,
          action: 'format',
          disabled: disableSubjectHeading,
          role: _everPluginList.ListRole.listOrdered,
          extra: extras['olist']
        }, {
          type: _everComponent.TYPE_MENU_ITEM,
          menuKey: 'list.ulist',
          title: locale?.ulist,
          icon: _ref10,
          action: 'format',
          disabled: disableSubjectHeading,
          role: _everPluginList.ListRole.listUnordered,
          extra: extras['ulist']
        }, {
          type: _everComponent.TYPE_MENU_ITEM,
          menuKey: 'list.tlist',
          title: locale?.tlist,
          icon: _ref11,
          action: 'format',
          disabled: disableSubjectHeading,
          role: _everPluginList.ListRole.listTask,
          extra: extras['tlist']
        }]
      }, {
        type: _everComponent.TYPE_GROUP,
        title: locale?.block,
        menus: [{
          type: _everComponent.TYPE_MENU_ITEM,
          title: locale?.blockquote,
          menuKey: 'blockquote',
          icon: _ref12,
          action: 'format',
          disabled: disableMenu || isCode,
          role: _everPluginBlockquote.BlockquoteRole.blockquote,
          extra: extras['blockquote']
        }, {
          type: _everComponent.TYPE_MENU_ITEM,
          title: locale?.code,
          menuKey: 'code',
          icon: _ref13,
          action: 'format',
          disabled: disableMenu,
          role: _everPluginCode.CodeRole.code,
          // 为 undefined 会导致显示快捷键
          extra: extras['code'] || ''
        }, {
          type: _everComponent.TYPE_MENU_ITEM,
          title: locale?.callout,
          menuKey: 'callout',
          icon: _ref14,
          action: 'format',
          disabled: disableMenu || isCode,
          role: _everPluginCallout.CalloutRole.callout,
          extra: extras['callout']
        }]
      }]
    };
  }, [extras, locale, disableMenu, isCode, disableSubjectHeading]);
  return {
    extras,
    menus
  };
}
