// @ali/4ever-plugin-code/dist-node/utils/constants/languages.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createCodeLanguages = exports.DEFAULT_LANGUAGE = void 0;
const DEFAULT_LANGUAGE = 'plaintext';
exports.DEFAULT_LANGUAGE = DEFAULT_LANGUAGE;

const createCodeLanguages = () => [{
  title: 'Plain Text',
  key: DEFAULT_LANGUAGE,
  testid: 'option-default'
}, {
  title: 'C',
  key: 'text/x-csrc',
  mode: 'clike'
}, {
  title: 'C++',
  key: 'text/x-c++src',
  alias: ['cpp'],
  mode: 'clike'
}, {
  title: 'C#',
  key: 'text/x-csharp',
  alias: ['csharp', 'cs'],
  mode: 'clike'
}, {
  title: 'Clojure',
  key: 'text/x-clojure',
  mode: 'clojure'
}, {
  title: 'CMake',
  key: 'text/x-cmake',
  mode: 'cmake'
}, {
  title: 'CSS',
  key: 'text/css',
  mode: 'css'
}, {
  title: 'Dart',
  key: 'application/dart',
  mode: 'dart'
}, {
  title: 'Dockerfile',
  key: 'text/x-dockerfile',
  mode: 'dockerfile'
}, {
  title: 'Diff',
  key: 'text/x-diff',
  mode: 'diff'
}, {
  title: 'Erlang',
  key: 'text/x-erlang',
  mode: 'erlang'
}, {
  title: 'Fortran',
  key: 'text/x-fortran',
  mode: 'fortran'
}, {
  title: 'F#',
  key: 'text/x-fsharp',
  mode: 'mllike'
}, {
  title: 'Go',
  key: 'text/x-go',
  mode: 'go'
}, {
  title: 'Groovy',
  key: 'text/x-groovy',
  mode: 'groovy'
}, {
  title: 'HTML',
  key: 'htmlmixed',
  mode: 'htmlmixed',
  testid: 'option-htmlmixed'
}, {
  title: 'HTTP',
  key: 'message/http',
  mode: 'http'
}, {
  title: 'Java',
  key: 'text/x-java',
  mode: 'clike'
}, {
  title: 'JavaScript(JSON)',
  key: 'text/javascript',
  alias: ['js', 'javascript', 'json', 'ecmascript', 'node'],
  mode: 'javascript'
}, {
  title: 'JSX',
  key: 'text/jsx',
  mode: 'jsx'
}, {
  title: 'Kotlin',
  key: 'text/x-kotlin',
  mode: 'clike'
}, {
  title: 'LESS',
  key: 'text/x-less',
  mode: 'css'
}, {
  title: 'Lisp',
  key: 'text/x-common-lisp',
  mode: 'commonlisp'
}, {
  title: 'Lua',
  key: 'text/x-lua',
  mode: 'lua'
}, {
  title: 'Markdown',
  key: 'text/x-markdown',
  mode: 'markdown'
}, {
  title: 'MATLAB(Octave)',
  key: 'text/x-octave',
  mode: 'octave'
}, {
  title: 'Nginx',
  key: 'text/x-nginx-conf',
  mode: 'nginx'
}, {
  title: 'Object C',
  key: 'text/x-objectivec',
  alias: ['oc', 'objective-c', 'objc'],
  mode: 'clike'
}, {
  title: 'OCaml',
  key: 'text/x-ocaml',
  mode: 'mllike'
}, {
  title: 'Perl',
  key: 'text/x-perl',
  mode: 'perl'
}, {
  title: 'PHP',
  key: 'text/x-php',
  mode: 'php'
}, {
  title: 'Python',
  key: 'text/x-python',
  alias: ['py'],
  mode: 'python'
}, {
  title: 'ProtoBuf',
  key: 'text/x-protobuf',
  mode: 'protobuf'
}, {
  title: 'R',
  key: 'text/x-rsrc',
  mode: 'r'
}, {
  title: 'Ruby',
  key: 'text/x-ruby',
  mode: 'ruby'
}, {
  title: 'Rust',
  key: 'text/x-rustsrc',
  mode: 'rust'
}, {
  title: 'Sass',
  key: 'text/x-sass',
  mode: 'sass'
}, {
  title: 'Scala',
  key: 'text/x-scala',
  mode: 'clike'
}, {
  title: 'Scss',
  key: 'text/x-scss',
  mode: 'css'
}, {
  title: 'Shell',
  key: 'text/x-sh',
  mode: 'shell'
}, {
  title: 'SQL',
  key: 'text/x-sql',
  mode: 'sql'
}, {
  title: 'Swift',
  key: 'text/x-swift',
  mode: 'swift'
}, {
  title: 'TSX',
  key: 'text/typescript-jsx',
  mode: 'jsx'
}, {
  title: 'TypeScript',
  key: 'text/typescript',
  alias: ['ts'],
  mode: 'javascript'
}, {
  title: 'VB.NET',
  key: 'text/x-vb',
  alis: ['vb'],
  mode: 'vb'
}, {
  title: 'VBScript',
  key: 'text/vbscript',
  mode: 'vbscript'
}, {
  title: 'VHDL',
  key: 'text/x-vhdl',
  mode: 'vhdl'
}, {
  title: 'Vue.js',
  key: 'text/x-vue',
  mode: 'vue'
}, {
  title: 'XML',
  key: 'application/xml',
  mode: 'xml'
}, {
  title: 'YAML',
  key: 'text/x-yaml',
  mode: 'yaml'
}, {
  title: 'APL',
  key: 'text/apl',
  mode: 'apl',
  show: false
}, {
  title: 'PGP',
  key: 'application/pgp',
  mode: 'asciiarmor',
  show: false
}, {
  title: 'ASN.1',
  key: 'text/x-ttcn-asn',
  mode: 'asn.1',
  show: false
}, {
  title: 'Asterisk',
  key: 'text/x-asterisk',
  mode: 'asterisk',
  show: false
}, {
  title: 'Brainfuck',
  key: 'text/x-brainfuck',
  mode: 'brainfuck',
  show: false
}, {
  title: 'Cobol',
  key: 'text/x-cobol',
  mode: 'cobol',
  show: false
}, {
  title: 'ClojureScript',
  key: 'text/x-clojurescript',
  mode: 'clojure',
  show: false
}, {
  title: 'Closure Stylesheets (GSS)',
  key: 'text/x-gss',
  mode: 'css',
  show: false
}, {
  title: 'CoffeeScript',
  key: 'application/vnd.coffeescript',
  alias: ['coffee', 'coffee-script'],
  mode: 'coffeescript',
  show: false
}, {
  title: 'Cypher',
  key: 'application/x-cypher-query',
  mode: 'cypher',
  show: false
}, {
  title: 'Cython',
  key: 'text/x-cython',
  mode: 'python',
  show: false
}, {
  title: 'Crystal',
  key: 'text/x-crystal',
  mode: 'crystal',
  show: false
}, {
  title: 'CQL',
  key: 'text/x-cassandra',
  mode: 'sql',
  show: false
}, {
  title: 'D',
  key: 'text/x-d',
  mode: 'd',
  show: false
}, {
  title: 'Django',
  key: 'text/x-django',
  mode: 'django',
  show: false
}, {
  title: 'DTD',
  key: 'application/xml-dtd',
  mode: 'dtd',
  show: false
}, {
  title: 'Dylan',
  key: 'text/x-dylan',
  mode: 'dylan',
  show: false
}, {
  title: 'EBNF',
  key: 'text/x-ebnf',
  mode: 'ebnf',
  show: false
}, {
  title: 'ECL',
  key: 'text/x-ecl',
  mode: 'ecl',
  show: false
}, {
  title: 'edn',
  key: 'application/edn',
  mode: 'clojure',
  show: false
}, {
  title: 'Eiffel',
  key: 'text/x-eiffel',
  mode: 'eiffel',
  show: false
}, {
  title: 'Elm',
  key: 'text/x-elm',
  mode: 'elm',
  show: false
}, {
  title: 'Embedded Javascript',
  key: 'application/x-ejs',
  mode: 'htmlembedded',
  show: false
}, {
  title: 'Embedded Ruby',
  key: 'application/x-erb',
  mode: 'htmlembedded',
  show: false
}, {
  title: 'Esper',
  key: 'text/x-esper',
  mode: 'sql',
  show: false
}, {
  title: 'Factor',
  key: 'text/x-factor',
  mode: 'factor',
  show: false
}, {
  title: 'FCL',
  key: 'text/x-fcl',
  mode: 'fcl',
  show: false
}, {
  title: 'Forth',
  key: 'text/x-forth',
  mode: 'forth',
  show: false
}, {
  title: 'Gas',
  key: 'text/x-gas',
  mode: 'gas',
  show: false
}, {
  title: 'Gherkin',
  key: 'text/x-feature',
  mode: 'gherkin',
  show: false
}, {
  title: 'GitHub Flavored Markdown',
  key: 'text/x-gfm',
  mode: 'gfm',
  show: false
}, {
  title: 'HAML',
  key: 'text/x-haml',
  mode: 'haml',
  show: false
}, {
  title: 'Haskell',
  key: 'text/x-haskell',
  mode: 'haskell',
  show: false
}, {
  title: 'Haskell (Literate)',
  key: 'text/x-literate-haskell',
  mode: 'haskell-literate',
  show: false
}, {
  title: 'Haxe',
  key: 'text/x-haxe',
  mode: 'haxe',
  show: false
}, {
  title: 'HXML',
  key: 'text/x-hxml',
  mode: 'haxe',
  show: false
}, {
  title: 'ASP.NET',
  key: 'application/x-aspx',
  alias: ['asp', 'aspx'],
  mode: 'htmlembedded',
  show: false
}, {
  title: 'IDL',
  key: 'text/x-idl',
  mode: 'idl',
  show: false
}, {
  title: 'Pug',
  key: 'text/x-pug',
  alias: ['jade'],
  mode: 'pug',
  show: false
}, {
  title: 'Java Server Pages',
  key: 'application/x-jsp',
  alias: ['jsp'],
  mode: 'htmlembedded',
  show: false
}, {
  title: 'JSON-LD',
  key: 'application/ld+json',
  alias: ['jsonld'],
  mode: 'javascript',
  show: false
}, {
  title: 'Jinja2',
  key: 'text/jinja2',
  mode: 'jinja2',
  show: false
}, {
  title: 'Julia',
  key: 'text/x-julia',
  mode: 'julia',
  show: false
}, {
  title: 'LiveScript',
  key: 'text/x-livescript',
  alias: ['ls'],
  mode: 'livescript',
  show: false
}, {
  title: 'mIRC',
  key: 'text/mirc',
  mode: 'mirc',
  show: false
}, {
  title: 'MariaDB SQL',
  key: 'text/x-mariadb',
  mode: 'sql',
  show: false
}, {
  title: 'Mathematica',
  key: 'text/x-mathematica',
  mode: 'mathematica',
  show: false
}, {
  title: 'Modelica',
  key: 'text/x-modelica',
  mode: 'modelica',
  show: false
}, {
  title: 'MUMPS',
  key: 'text/x-mumps',
  mode: 'mumps',
  show: false
}, {
  title: 'MS SQL',
  key: 'text/x-mssql',
  mode: 'sql',
  show: false
}, {
  title: 'mbox',
  key: 'application/mbox',
  mode: 'mbox',
  show: false
}, {
  title: 'MySQL',
  key: 'text/x-mysql',
  mode: 'sql',
  show: false
}, {
  title: 'NSIS',
  key: 'text/x-nsis',
  mode: 'nsis',
  show: false
}, {
  title: 'NTriples',
  key: 'application/n-triples',
  mode: 'ntriples',
  show: false
}, {
  title: 'Objective-C++',
  key: 'text/x-objectivec++',
  alias: ['objective-c++', 'objc++'],
  mode: 'clike',
  show: false
}, {
  title: 'Oz',
  key: 'text/x-oz',
  mode: 'oz',
  show: false
}, {
  title: 'Pascal',
  key: 'text/x-pascal',
  mode: 'pascal',
  show: false
}, {
  title: 'PEG.js',
  key: 'null',
  mode: 'pegjs',
  show: false
}, {
  title: 'Pig',
  key: 'text/x-pig',
  mode: 'pig',
  show: false
}, {
  title: 'PLSQL',
  key: 'text/x-plsql',
  mode: 'sql',
  show: false
}, {
  title: 'PostgreSQL',
  key: 'text/x-pgsql',
  mode: 'sql',
  show: false
}, {
  title: 'PowerShell',
  key: 'application/x-powershell',
  mode: 'powershell',
  show: false
}, {
  title: 'Properties files',
  key: 'text/x-properties',
  alias: ['ini', 'properties'],
  mode: 'properties',
  show: false
}, {
  title: 'Puppet',
  key: 'text/x-puppet',
  mode: 'puppet',
  show: false
}, {
  title: 'Q',
  key: 'text/x-q',
  mode: 'q',
  show: false
}, {
  title: 'reStructuredText',
  key: 'text/x-rst',
  alias: ['rst'],
  mode: 'rst',
  show: false
}, {
  title: 'RPM Changes',
  key: 'text/x-rpm-changes',
  mode: 'rpm',
  show: false
}, {
  title: 'RPM Spec',
  key: 'text/x-rpm-spec',
  mode: 'rpm',
  show: false
}, {
  title: 'SAS',
  key: 'text/x-sas',
  mode: 'sas',
  show: false
}, {
  title: 'Scheme',
  key: 'text/x-scheme',
  mode: 'scheme',
  show: false
}, {
  title: 'Sieve',
  key: 'application/sieve',
  mode: 'sieve',
  show: false
}, {
  title: 'Slim',
  key: 'application/x-slim',
  mode: 'slim',
  show: false
}, {
  title: 'Smalltalk',
  key: 'text/x-stsrc',
  mode: 'smalltalk',
  show: false
}, {
  title: 'Smarty',
  key: 'text/x-smarty',
  mode: 'smarty',
  show: false
}, {
  title: 'Solr',
  key: 'text/x-solr',
  mode: 'solr',
  show: false
}, {
  title: 'SML',
  key: 'text/x-sml',
  mode: 'mllike',
  show: false
}, {
  title: 'Soy',
  key: 'text/x-soy',
  alias: ['closure template'],
  mode: 'soy',
  show: false
}, {
  title: 'SPARQL',
  key: 'application/sparql-query',
  alias: ['sparul'],
  mode: 'sparql',
  show: false
}, {
  title: 'Spreadsheet',
  key: 'text/x-spreadsheet',
  alias: ['excel', 'formula'],
  mode: 'spreadsheet',
  show: false
}, {
  title: 'SQLite',
  key: 'text/x-sqlite',
  mode: 'sql',
  show: false
}, {
  title: 'Squirrel',
  key: 'text/x-squirrel',
  mode: 'clike',
  show: false
}, {
  title: 'Stylus',
  key: 'text/x-styl',
  mode: 'stylus',
  show: false
}, {
  title: 'sTeX',
  key: 'text/x-stex',
  mode: 'stex',
  show: false
}, {
  title: 'LaTeX',
  key: 'text/x-latex',
  alias: ['tex'],
  mode: 'stex',
  show: false
}, {
  title: 'SystemVerilog',
  key: 'text/x-systemverilog',
  mode: 'verilog',
  show: false
}, {
  title: 'Tcl',
  key: 'text/x-tcl',
  mode: 'tcl',
  show: false
}, {
  title: 'Textile',
  key: 'text/x-textile',
  mode: 'textile',
  show: false
}, {
  title: 'TiddlyWiki',
  key: 'text/x-tiddlywiki',
  mode: 'tiddlywiki',
  show: false
}, {
  title: 'Tiki wiki',
  key: 'text/tiki',
  mode: 'tiki',
  show: false
}, {
  title: 'TOML',
  key: 'text/x-toml',
  mode: 'toml',
  show: false
}, {
  title: 'Tornado',
  key: 'text/x-tornado',
  mode: 'tornado',
  show: false
}, {
  title: 'troff',
  key: 'text/troff',
  mode: 'troff',
  show: false
}, {
  title: 'TTCN',
  key: 'text/x-ttcn',
  mode: 'ttcn',
  show: false
}, {
  title: 'TTCN_CFG',
  key: 'text/x-ttcn-cfg',
  mode: 'ttcn-cfg',
  show: false
}, {
  title: 'Turtle',
  key: 'text/turtle',
  mode: 'turtle',
  show: false
}, {
  title: 'Twig',
  key: 'text/x-twig',
  mode: 'twig',
  show: false
}, {
  title: 'Web IDL',
  key: 'text/x-webidl',
  mode: 'webidl',
  show: false
}, {
  title: 'Velocity',
  key: 'text/velocity',
  mode: 'velocity',
  show: false
}, {
  title: 'Verilog',
  key: 'text/x-verilog',
  mode: 'verilog',
  show: false
}, {
  title: 'XQuery',
  key: 'application/xquery',
  mode: 'xquery',
  show: false
}, {
  title: 'Yacas',
  key: 'text/x-yacas',
  mode: 'yacas',
  show: false
}, {
  title: 'Z80',
  key: 'text/x-z80',
  mode: 'z80',
  show: false
}, {
  title: 'mscgen',
  key: 'text/x-mscgen',
  mode: 'mscgen',
  show: false
}, {
  title: 'xu',
  key: 'text/x-xu',
  mode: 'mscgen',
  show: false
}, {
  title: 'msgenny',
  key: 'text/x-msgenny',
  mode: 'mscgen',
  show: false
}];

exports.createCodeLanguages = createCodeLanguages;
