// @ali/4ever-plugin-code/dist-node/utils/constants/themes.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createCodeThemes = exports.createThemeMenuItems = exports.CodeWrapper = exports.DEFAULT_THEME = void 0;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _everUtils = require("@ali/4ever-utils");

const _createElement = /*#__PURE__*/React.createElement;
const DEFAULT_THEME = 'default';
exports.DEFAULT_THEME = DEFAULT_THEME;

const Title = /*#__PURE__*/_styledComponents.default.span(["display:inline-flex;align-items:center;vertical-align:middle;"]);

const Icon = /*#__PURE__*/_styledComponents.default.span(["background:", ";width:16px;height:16px;border-radius:3px;display:inline-block;margin-right:10px;border:1px solid #ecedef;"], props => props.color);

const CodeWrapper = /*#__PURE__*/_styledComponents.default.div(["margin-left:", "px;margin-right:", "px;"], _everUtils.codeTheme['margin-left'], _everUtils.codeTheme['margin-right']);

exports.CodeWrapper = CodeWrapper;

var _ref = /*#__PURE__*/_createElement(Icon, {
  color: "#f6f8f9"
});

var _ref2 = /*#__PURE__*/_createElement(Icon, {
  color: "#111f2c"
});

const createThemeMenuItems = (locale, key) => [{
  key: DEFAULT_THEME,
  prefix: _ref,
  title: locale.lightTheme || 'Light',
  selected: key === DEFAULT_THEME,
  testid: 'option-default'
}, {
  key: 'dracula',
  prefix: _ref2,
  title: locale.darkTheme || 'Dark',
  selected: key === 'dracula',
  testid: 'option-dracula'
}];

exports.createThemeMenuItems = createThemeMenuItems;

var _ref3 = /*#__PURE__*/_createElement(Icon, {
  color: "#f6f8f9"
});

var _ref4 = /*#__PURE__*/_createElement(Icon, {
  color: "#111f2c"
});

const createCodeThemes = locale => [{
  title: /*#__PURE__*/_createElement(Title, null, _ref3, " ", locale.lightTheme || 'Light'),
  key: DEFAULT_THEME
}, {
  title: /*#__PURE__*/_createElement(Title, null, _ref4, " ", locale.darkTheme || 'Dark'),
  key: 'dracula'
}];

exports.createCodeThemes = createCodeThemes;
