// @ali/dingtalk-jsapi/platform/pc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ddSdk_1 = require("../lib/ddSdk");
var env_1 = require("../lib/env");
var h5Pc_1 = require("../lib/bridge/h5Pc");
var eapp_1 = require("../lib/bridge/eapp");
var sdk_1 = require("../lib/sdk");
var h5PcEvent_1 = require("../lib/bridge/h5PcEvent");
ddSdk_1.ddSdk.setPlatform({
    platform: env_1.ENV_ENUM.pc,
    bridgeInit: function () {
        var env = env_1.getENV();
        switch (env.appType) {
            case sdk_1.APP_TYPE.MINI_APP:
                return Promise.resolve(eapp_1.default);
            default:
                return h5Pc_1.h5PcBridgeInit().then(function () {
                    return h5Pc_1.default;
                });
        }
    },
    authMethod: 'config',
    authParamsDeal: function (params) {
        var finalParams = Object.assign({}, params);
        finalParams.url = window.location.href.split('#')[0];
        return finalParams;
    },
    event: {
        on: function (type, handler) {
            if (env_1.getENV().appType === sdk_1.APP_TYPE.WEB) {
                return h5PcEvent_1.on(type, handler);
            }
        },
        off: function (type, handler) {
            if (env_1.getENV().appType === sdk_1.APP_TYPE.WEB) {
                return h5PcEvent_1.off(type, handler);
            }
        },
    },
});
