// @ali/4ever-zhi/dist-node/plugins/image/image.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _styled = require("./styled");

const _createElement = /*#__PURE__*/_react.default.createElement;

const Image = props => {
  const imageRef = _react.default.useRef(null);

  const {
    controller,
    node,
    imageConfig
  } = props;
  const {
    src
  } = node.data;
  const {
    previewImage
  } = imageConfig;

  const [isLoaded, setIsLoaded] = _react.default.useState(false);

  const handleZoomImage = (0, _react.useCallback)(event => {
    if (_everBamboo.ImagePlugin.isParentClickable(controller, node)) {
      return;
    }

    event.preventDefault();
    event.stopPropagation();

    if (previewImage) {
      previewImage(src, node);
    }
  }, [previewImage, src, node, controller]);

  const handlers = _react.default.useMemo(() => {
    return {
      onClick: handleZoomImage
    };
  }, [handleZoomImage]);

  const cursorStyle = !!imageConfig.previewImage && isLoaded ? 'zoom-in' : 'auto';
  return /*#__PURE__*/_createElement(_styled.ImageWrapper, {
    cursorStyle: cursorStyle
  }, /*#__PURE__*/_createElement(_everBamboo.ImagePlugin.Img, (0, _extends2.default)({
    ref: imageRef,
    node: node,
    controller: controller,
    imageConfig: imageConfig
  }, handlers, {
    onLoad: () => setIsLoaded(true)
  })));
};

var _default = Image;
exports.default = _default;
