// @ali/4ever-plugin-image/dist-node/mo/serializer/htmlToJsonML.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everUtils = require("@ali/4ever-utils");

var _lodash = require("lodash");

var _everCangjie = require("@ali/4ever-cangjie");

var _default = {
  name: 'img',

  onOpenTag(state, name, attrs) {
    // 防止误解析来自 IM 的表情
    if (name !== 'img' || (0, _everUtils.isEmojiClsName)(attrs?.class)) return false;
    const {
      style: attrStyle,
      styleObj: style,
      width: attrWidth,
      height: attrHeight,
      'data-width': dataWidth,
      'data-height': dataHeight,
      'data-4ever-bamboo-src': bambooSrc,
      ...restAttrs
    } = attrs;
    const {
      width: styleWidth,
      height: styleHeight
    } = style; // 宽度取值优先级：dataWidth(存在 boambooSrc 时) > styleWidth > attrWidth
    // 高度同理

    const isBamboo = !!bambooSrc;
    const width = isBamboo && dataWidth || styleWidth || attrWidth;
    const height = isBamboo && dataHeight || styleHeight || attrHeight;
    const rawAttrs = { ...(0, _lodash.pick)(restAttrs, ['width', 'height', 'src', 'style', 'title', 'alt']),
      width: parseInt(width, 10),
      height: parseInt(height, 10),
      src: bambooSrc || attrs.src || ''
    };

    if (style.opacity) {
      rawAttrs.opacity = +style.opacity;
    }

    const img = ['img', rawAttrs, (0, _everCangjie.createEmptyText)()];
    state.push(img);
    return true;
  },

  onCloseTag(state, name, attrs) {
    if (name !== 'img' || (0, _everUtils.isEmojiClsName)(attrs?.class)) return false;
    const img = state.pop();
    const parent = state.peek();
    const parentTagName = (0, _everUtils.getTagName)(parent);

    if (!_everUtils.paragraphTags[parentTagName] && parentTagName !== 'a') {
      state.append(['p', (0, _everCangjie.createEmptyText)(), img, (0, _everCangjie.createEmptyText)()]);
    } else {
      state.append(img);
    }

    return true;
  }

};
exports.default = _default;
