// @ali/4ever-plugin-image/dist-node/mo/serializer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createSerializer;
exports.createServerSerializationPlugin = createServerSerializationPlugin;

var _htmlToJsonML = _interopRequireDefault(require("./htmlToJsonML"));

var _jsonMLToValue = _interopRequireDefault(require("./jsonMLToValue"));

var _valueToJsonML = _interopRequireDefault(require("./valueToJsonML"));

var _jsonMLToHTML = _interopRequireDefault(require("./jsonMLToHTML"));

var _markdownToHTML = _interopRequireDefault(require("./markdownToHTML"));

var _valueToText = _interopRequireDefault(require("./valueToText"));

var _htmlToMarkdown = _interopRequireDefault(require("./htmlToMarkdown"));

function createSerializer(configs) {
  return {
    htmlToJsonML: _htmlToJsonML.default,
    htmlToMarkdown: _htmlToMarkdown.default,
    jsonMLToValue: _jsonMLToValue.default,
    valueToJsonML: _valueToJsonML.default,
    jsonMLToHTML: (0, _jsonMLToHTML.default)(configs.image),
    markdownToHTML: _markdownToHTML.default,
    valueToText: (0, _valueToText.default)(configs.image?.locale)
  };
}

function createServerSerializationPlugin() {
  return {
    valueToJsonML: _valueToJsonML.default,
    jsonMLToValue: _jsonMLToValue.default
  };
}
