// @ali/4ever-plugin-image/dist-node/mo/serializer/jsonMLToHTML.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everUtils = require("@ali/4ever-utils");

const _createElement = /*#__PURE__*/React.createElement;

const Image = props => {
  const {
    src,
    style,
    path,
    config,
    ...rest
  } = props;
  return /*#__PURE__*/_createElement("img", (0, _extends2.default)({
    src: (0, _everUtils.sanitizeUrl)(src, config?.protocolAllowList),
    style: style,
    key: path.toString()
  }, rest));
};

var _default = config => ({
  name: 'img',

  match(node) {
    return (0, _everUtils.getTagName)(node) === 'img';
  },

  convert(node, path) {
    // @ts-ignore
    const {
      src,
      width,
      height,
      opacity,
      outline
    } = (0, _everUtils.getAttributes)(node);
    let border;

    if (outline && outline.width) {
      border = `${outline.width}px solid ${outline.color ? outline.color : '#000000'}`;
    }

    if (!src) return null;
    return /*#__PURE__*/_createElement(Image, {
      src: src,
      style: {
        width,
        height,
        opacity,
        border
      },
      path: path,
      config: config
    });
  }

});

exports.default = _default;
