// @ali/4ever-bi/dist-web/plugins/image/actions.js
export var IMAGE_SELECT = Symbol('IMAGE_SELECT');
export var IMAGE_REMOVE = Symbol('IMAGE_REMOVE');
export var IMAGE_UPDATE = Symbol('IMAGE_UPDATE');
export var IMAGE_SET_INJECTION = Symbol('IMAGE_SET_INJECTION');
/**
 * 插入上传图片的 placeholder
 */

export var INSERT_IMAGE_PLACEHOLDER = Symbol('INSERT_IMAGE_PLACEHOLDER');
/**
 * 唤起图片格式侧边栏
 */

export var TOGGLE_IMAGE_SIDEBAR = Symbol('TOGGLE_IMAGE_SIDEBAR');
export function selectImage(placeholderKey) {
  return {
    type: IMAGE_SELECT,
    payload: {
      placeholderKey: placeholderKey
    }
  };
}
export function removeImage(image) {
  return {
    type: IMAGE_REMOVE,
    payload: {
      node: image
    }
  };
}
export function updateImage(image, data, container) {
  return {
    type: IMAGE_UPDATE,
    payload: {
      node: image,
      data: data,
      container: container
    }
  };
}
export function setImageInjection(key, data) {
  return {
    type: IMAGE_SET_INJECTION,
    payload: {
      key: key,
      data: data
    }
  };
}
export function toggleImageSidebar(payload) {
  return {
    type: TOGGLE_IMAGE_SIDEBAR,
    payload: payload
  };
}
export var actions = {
  selectImage: selectImage,
  removeImage: removeImage,
  updateImage: updateImage,
  setImageInjection: setImageInjection,
  toggleImageSidebar: toggleImageSidebar
};
/** @deprecated */

export var DEPRECATED_ACTIONS = {
  IMAGE_SELECT: IMAGE_SELECT,
  IMAGE_REMOVE: IMAGE_REMOVE,
  IMAGE_UPDATE: IMAGE_UPDATE,
  IMAGE_SET_INJECTION: IMAGE_SET_INJECTION,
  INSERT_IMAGE_PLACEHOLDER: INSERT_IMAGE_PLACEHOLDER,
  TOGGLE_IMAGE_SIDEBAR: TOGGLE_IMAGE_SIDEBAR,
  selectImage: selectImage,
  removeImage: removeImage,
  updateImage: updateImage,
  setImageInjection: setImageInjection,
  toggleImageSidebar: toggleImageSidebar
};
