// @ali/4ever-bi/dist-web/plugins/image/config.js
import zhLocale from "../../locales/zh_CN";
/**
 * 获取图片工具栏的布局
 * @returns
 */

export function getToolbarLayout() {
  return [{
    toolButtons: [{
      name: 'image.download'
    }, {
      name: 'image.reset'
    }, {
      name: 'image.border'
    }]
  }, {
    toolButtons: [{
      name: 'image.crop'
    }, {
      name: 'image.rotateReverse'
    }, {
      name: 'image.rotate'
    }]
  }];
}
export function getSidebarLayout(config) {
  var zhImageLocale = zhLocale.image.locale;
  var _config$locale = config.locale,
      locale = _config$locale === void 0 ? zhImageLocale : _config$locale;
  return [{
    renderTitle: function renderTitle() {
      return locale.sizeAndStyle;
    },
    items: ['image.sizeAndStyle'],
    index: 1,
    key: 'imageSizeAndStyle'
  }, {
    renderTitle: function renderTitle() {
      return locale.recommendStyle;
    },
    items: ['image.recommendStyle'],
    index: 5,
    key: 'imageRecommendStyle'
  }, {
    renderTitle: function renderTitle() {
      return locale.alignment;
    },
    items: ['alignment'],
    index: 3,
    key: 'alignmentDefault'
  } // 添加链接的交互未优化，暂时不添加
  // {
  //   renderTitle: () => locale.addLink,
  //   items: [
  //     'link.setInlineHref',
  //   ],
  //   index: 7,
  // },
  ];
}
export var getRecommendStyleItems = function getRecommendStyleItems() {
  return [{
    data: {
      outline: {
        width: 1,
        type: 'solid',
        color: 'rgba(126, 134, 142, 0.3)'
      },
      radius: 0,
      shadow: ''
    },
    demo: {
      outline: {
        width: 1,
        type: 'solid',
        color: 'rgba(23, 26, 29, 0.4)'
      },
      radius: 0,
      shadow: ''
    }
  }, {
    data: {
      outline: {
        width: 1,
        type: 'solid',
        color: 'rgba(126, 134, 142, 0.3)'
      },
      radius: 8,
      shadow: ''
    },
    demo: {
      outline: {
        width: 1,
        type: 'solid',
        color: 'rgba(23, 26, 29, 0.4)'
      },
      radius: 4,
      shadow: ''
    }
  }, {
    data: {
      outline: {
        width: 4,
        type: 'solid',
        color: '#171A1D'
      },
      radius: 0,
      shadow: ''
    },
    demo: {
      outline: {
        width: 2,
        type: 'solid',
        color: '#171A1D'
      },
      radius: 0,
      shadow: ''
    }
  }, {
    data: {
      outline: {},
      radius: 0,
      shadow: 'rgba(0, 0, 0, 0.24) 0px 0px 1px, rgba(0, 0, 0, 0.12) 0px 4px 16px'
    },
    demo: {
      outline: {},
      radius: 0,
      shadow: '0px 0px 1px rgba(0, 0, 0, 0.16), 0px 2px 4px rgba(23, 26 , 29, 0.24)'
    }
  }, {
    data: {
      outline: {},
      radius: 8,
      shadow: 'rgba(0, 0, 0, 0.24) 0px 0px 1px, rgba(0, 0, 0, 0.12) 0px 4px 16px'
    },
    demo: {
      outline: {},
      radius: 4,
      shadow: '0px 0px 1px rgba(0, 0, 0, 0.16), 0px 2px 4px rgba(23, 26 , 29, 0.24)'
    }
  }, {
    data: {
      outline: {
        width: 4,
        type: 'solid',
        color: '#FFFFFF'
      },
      radius: 0,
      shadow: '0px 0px 1px rgba(0, 0, 0, 0.24), 0px 2px 10px rgba(0, 0, 0, 0.24)'
    },
    demo: {
      outline: {
        width: 2,
        type: 'solid',
        color: '#FFFFFF'
      },
      radius: 0,
      shadow: '0px 0px 1px rgba(0, 0, 0, 0.16), 0px 1px 2px rgba(23, 26, 29, 0.4)'
    }
  }];
};
