// @ali/4ever-bi/dist-web/plugins/image/createRenderNode.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import Image from "./components/image";
export default function createRenderNode(config, state) {
  return {
    image: function image(props, controller) {
      var children = props.children,
          parent = props.parent,
          node = props.node,
          rest = _objectWithoutPropertiesLoose(props, ["children", "parent", "node"]);

      var id = node.data.id;
      var key = node.key;
      var pluginState = state[id] || state[key] || {};
      return (
        /*#__PURE__*/
        // @ts-ignore
        _createElement(Image, _extends({}, rest, {
          controller: controller,
          pluginState: pluginState,
          imageConfig: config || {},
          node: node,
          parent: parent
        }), children())
      );
    }
  };
}
