// @ali/4ever-zhi/dist-node/plugins/formula/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _everBi = require("@ali/4ever-bi");

var _everFactory = require("@ali/4ever-factory");

var _everPluginFormula = require("@ali/4ever-plugin-formula");

const _createElement = /*#__PURE__*/React.createElement;
const {
  FormulaPlaceholder,
  FormulaView
} = _everBi.FormulaPlugin.components;
const createZhiFormulaTagPlugin = (0, _everFactory.zhiTagFactory)(config => {
  const formulaConfig = config.formula || {};
  const {
    locale = {}
  } = formulaConfig;

  var _ref = /*#__PURE__*/_createElement(FormulaPlaceholder, {
    locale: locale,
    notAllowed: true,
    readOnly: true
  });

  return {
    tagType: 'formula',
    renderNode: props => {
      const node = props.node;
      const {
        formula
      } = node.data.metadata;

      if (formula) {
        return /*#__PURE__*/_createElement(FormulaView, {
          formula: formula,
          isEditing: false,
          readOnly: true
        });
      }

      return _ref;
    }
  };
}, _everPluginFormula.moFactory);
var _default = createZhiFormulaTagPlugin;
exports.default = _default;
