// @ali/we-design/dist-web/components/color-panel/ColorPreview/styled.js
import styled, { css } from 'styled-components';
import { DEFAULT_FONT_COLOR, DEFAULT_BG_COLOR } from "../constants";
import isGradientColor, { GRADIENT_FLAG } from "../utils/isGradientColor";
export var ColorPreviewWrapper = /*#__PURE__*/styled.div(["position:absolute;top:0;width:fit-content;padding:8px;background:#f5f6f7;box-shadow:rgb(0 0 0 / 24%) 0px 0px 1px,rgb(10 30 65 / 16%) 0px 8px 24px;border-radius:4px;font-size:12px;white-space:nowrap;"]);
export var ColorPreviewContent = /*#__PURE__*/styled.div(["background:#ffffff;padding:4px 2px;"]);
export var Text = /*#__PURE__*/styled.span(["padding-left:2px;padding-right:2px;color:", ";line-height:20px;"], DEFAULT_FONT_COLOR); // 渐变色和背景色同时设置在一个 span 上时有冲突，所以这里拆分为「背景层」和「颜色层」

export var BackgroundLayout = /*#__PURE__*/styled(Text)(["", ";"], function (_ref) {
  var backgroundColor = _ref.backgroundColor;
  return css(["background-color:", ";"], backgroundColor || DEFAULT_BG_COLOR);
});
export var ColorLayout = /*#__PURE__*/styled.span(["", ""], function (_ref2) {
  var color = _ref2.color;

  if (color && isGradientColor(color)) {
    var _color$split = color.split(GRADIENT_FLAG),
        firstColor = _color$split[0],
        secondColor = _color$split[1];

    return css(["-webkit-text-fill-color:transparent;-webkit-background-clip:text;background-image:linear-gradient(90deg,", ",", ");"], firstColor, secondColor);
  } else {
    return css(["color:", ";"], color || DEFAULT_FONT_COLOR);
  }
});
export var Tip = /*#__PURE__*/styled.span(["display:inline-block;color:#707274;margin-top:8px;"]);
