// @ali/4ever-plugin-code/dist-node/components/SimpleCode.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _lodash = require("lodash");

var _everUtils = require("@ali/4ever-utils");

var _dimension = require("../utils/constants/dimension");

const _createElement = /*#__PURE__*/React.createElement;

const Code = /*#__PURE__*/_styledComponents.default.pre(["background-color:", ";border-color:", ";padding:10px;border-radius:5px;height:", ";line-height:", ";overflow:auto;-webkit-overflow-scrolling:touch;", " ", " code{color:", ";}"], ({
  theme
}) => theme === 'default' ? 'rgba(246,248,249,0.8)' : '#282a36', ({
  theme
}) => theme === 'default' ? 'rgba(17, 31, 44, 0.08)' : 'rgba(17, 31, 44, 1)', ({
  height
}) => height ? `${height}px` : 'inherit', _everUtils.codeTheme['line-height'], props => props.lineWrapping && (0, _styledComponents.css)(["word-wrap:break-word;white-space:pre-wrap;word-break:normal;"]), props => !props.lineWrapping && (0, _styledComponents.css)(["word-wrap:normal;"]), ({
  theme
}) => theme === 'default' ? 'rgba(17, 31, 44, 0.56)' : '#FFFFFF');

const SimpleCode = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    code,
    height,
    theme,
    onTouchStart = _lodash.noop,
    onMouseDown = _lodash.noop,
    onClick = _lodash.noop,
    node,
    lineWrapping,
    print
  } = props;
  return /*#__PURE__*/_createElement(Code, {
    style: {
      marginLeft: _everUtils.codeTheme['margin-left'],
      marginRight: _everUtils.codeTheme['margin-right'],
      marginTop: _everUtils.codeTheme['margin-top'],
      marginBottom: _everUtils.codeTheme['margin-bottom'],
      maxHeight: !print ? _dimension.MAX_HEIGHT : 'none',
      ..._everUtils.SpacingUtil.toStyle(node)
    },
    ref: ref,
    height: height,
    theme: theme,
    lineWrapping: lineWrapping,
    onTouchStart: onTouchStart,
    onMouseDown: onMouseDown,
    onClick: onClick,
    "data-hover-box": "borderBox",
    "data-plugin-scrollable": true
  }, /*#__PURE__*/_createElement("code", null, code));
});
var _default = SimpleCode;
exports.default = _default;
