// @ali/4ever-bi/dist-web/plugins/heading/index.js
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import toolbar from "./toolbar";
import menubar from "./menubar";
import HeadingWrapper from "./components/headingWrapper";
import onAction from "./handlers/onAction";
import onKeyDown from "./handlers/onKeyDown";
import onCangjieInput from "./handlers/onCangjieInput";
import createOnCangjiePaste from "./handlers/createOnCangjiePaste";
import * as actions from "./actions";
import placeholder from "./placeholder";
import getHeadingModels from "./utils/getHeadingModels";
import { createRenderNode } from "./createRenderNode";
import { queries } from "./queries";
import { commands } from "./commands";
import { moFactory } from '@ali/4ever-plugin-heading';
import createRenderToolbar from "./createRenderToolbar";

function createHeadingPlugin(config) {
  var headingConfig = config.heading || {};
  return _extends({}, moFactory({}), {
    // @ts-ignore
    renderEditable: function renderEditable(props, controller, next) {
      if (headingConfig.onHeadingsChange) {
        return /*#__PURE__*/_createElement(HeadingWrapper, _extends({}, props, {
          controller: controller,
          onHeadingsChange: headingConfig.onHeadingsChange
        }), next());
      }

      return next();
    },
    renderNode: createRenderNode(),
    toolbar: toolbar(headingConfig),
    renderToolbar: createRenderToolbar(headingConfig.locale || {}),
    menubar: menubar(),
    actions: actions,
    commands: commands,
    queries: queries,
    onAction: onAction,
    onCangjiePaste: createOnCangjiePaste(config),
    onKeyDown: onKeyDown,
    onCangjieInput: onCangjieInput,
    models: getHeadingModels(),
    placeholder: placeholder(headingConfig)
  });
}

createHeadingPlugin.actions = actions;
createHeadingPlugin.commands = commands;
createHeadingPlugin.queries = queries;
export default createHeadingPlugin;
