// @ali/4ever-plugin-image/dist-node/utils/utils/getImageType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getImageTypeFromBlob = exports.getImageTypeFromBuffer = exports.ImageTypeEnum = void 0;
let ImageTypeEnum; // (List_of_file_signatures)[https://en.wikipedia.org/wiki/List_of_file_signatures]

exports.ImageTypeEnum = ImageTypeEnum;

(function (ImageTypeEnum) {
  ImageTypeEnum["GIF"] = "gif";
  ImageTypeEnum["JPEG"] = "jpeg";
  ImageTypeEnum["PNG"] = "png";
  ImageTypeEnum["WEBP"] = "webp";
  ImageTypeEnum["TIFF"] = "tiff";
  ImageTypeEnum["TIF"] = "tif";
  ImageTypeEnum["SVG"] = "svg";
  ImageTypeEnum["UNCERTAIN"] = "uncertain_type";
})(ImageTypeEnum || (exports.ImageTypeEnum = ImageTypeEnum = {}));

function getTypeFromMagicNumber(signature) {
  switch (signature) {
    case '47494638':
      return ImageTypeEnum.GIF;

    case 'FFD8FFDB':
    case 'FFD8FFE0':
    case 'FFD8FFE1':
    case 'FFD8FFEE':
      return ImageTypeEnum.JPEG;

    case '89504E47':
      return ImageTypeEnum.PNG;

    case '52494646':
      return ImageTypeEnum.WEBP;

    case '4D4D002A':
      return ImageTypeEnum.TIFF;
    // // TIFF, big-endian type

    case '49492A00':
      return ImageTypeEnum.TIF;
    // TIFF, little-endian type

    default:
      return ImageTypeEnum.UNCERTAIN;
  }
}

const getImageTypeFromBuffer = buffer => {
  const uint = new Uint8Array(buffer);
  const bytes = [];
  uint.forEach(byte => {
    // byte = 0 ~ 15 时，其十六进制数前需补0
    const char = `0${byte.toString(16)}`.slice(-2);
    bytes.push(char);
  });
  const magicNumber = bytes.join('').toUpperCase();
  const typeFromMagicNumber = getTypeFromMagicNumber(magicNumber);
  return typeFromMagicNumber;
};

exports.getImageTypeFromBuffer = getImageTypeFromBuffer;

const getImageTypeFromBlob = blob => {
  return new Promise(resolve => {
    const subBlob = blob.slice(0, 4);
    const reader = new FileReader();

    reader.onload = event => {
      const fileStart = event.target.result;
      const type = getImageTypeFromBuffer(fileStart);
      resolve(type);
    };

    reader.onerror = () => {
      resolve(ImageTypeEnum.UNCERTAIN);
    };

    reader.readAsArrayBuffer(subBlob);
  });
};

exports.getImageTypeFromBlob = getImageTypeFromBlob;
