// @ali/4ever-plugin-highlight/dist-web/bi/components/highlightButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { ColorPicker, EColorPickerIcon } from '@ali/we-toolbar';
;
import { setHighlight } from "../actions";
import { DEFAULT_COLOR } from "../constants";
import { setColorPanelMode } from '@ali/4ever-plugin-pref';
import { HighlightRole } from "../../utils/role";
export default function HighlightButton(props) {
  var controller = props.controller,
      locale = props.locale,
      rest = _objectWithoutPropertiesLoose(props, ["controller", "locale"]);

  var currentHighlight = controller.query('getHighlight');
  var highlight = typeof currentHighlight === 'string' ? currentHighlight : DEFAULT_COLOR;
  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInLinkCard');
  var handleChange = React.useCallback(function (color) {
    controller.run('onAction', setHighlight(color));
  }, [controller]);
  var getColorPanelMode = React.useCallback(function () {
    var _controller$query;

    return (_controller$query = controller.query('getColorPanelMode')) == null ? void 0 : _controller$query.mode;
  }, [controller]);
  var handleSelectValueChange = React.useCallback(function (mode) {
    controller.run('onAction', setColorPanelMode({
      mode: mode
    }));
  }, [controller]);
  return /*#__PURE__*/_createElement(ColorPicker, _extends({}, rest, {
    testid: "bi-toolbar-highlight",
    disabled: !!isDisabled,
    color: highlight,
    onColorPicker: handleChange,
    defaultColor: '',
    automaticColor: "",
    clearColor: "",
    btnType: EColorPickerIcon.backgroundfont,
    shortcutText: locale.buttonTitle,
    role: HighlightRole.highlight,
    showToggleSelect: true,
    getColorPanelMode: getColorPanelMode,
    onSelectValueChange: handleSelectValueChange
  }));
}
