// @ali/4ever-zhi/dist-web/createCustomizedZhiPlugins.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";

/**
 * @packageDocumentation
 * @module Reader
 */
import { createSchema } from '@ali/4ever-cangjie';
import { sortPlugins } from '@ali/4ever-utils';
import { zhiFactory as clipboard } from '@ali/4ever-plugin-clipboard';
import { zhiFactory as paragraph } from '@ali/4ever-plugin-paragraph';
import { zhiFactory as container } from '@ali/4ever-plugin-container';
import { createUnknownCardZhiPlugin as createUnknownCardPlugin, createUnknownTagZhiPlugin as createUnknownTagPlugin } from '@ali/4ever-factory';
/**
 * 创建自定义插件。传入自定义的插件 Map，返回一个插件创建函数（createZhiPlugins）。
 * 传入内容格式：{ PluginName: PluginFactory }
 */

function createCustomizedZhiPlugins(plugins) {
  var paragraphPlugin = plugins.paragraph || paragraph;

  var _plugins$unknownCard = plugins.unknownCard,
      unknownCard = _plugins$unknownCard === void 0 ? createUnknownCardPlugin : _plugins$unknownCard,
      _plugins$unknownTag = plugins.unknownTag,
      unknownTag = _plugins$unknownTag === void 0 ? createUnknownTagPlugin : _plugins$unknownTag,
      rest = _objectWithoutPropertiesLoose(plugins, ["unknownCard", "unknownTag"]);

  var allPlugins = _extends({}, rest, {
    container: container,
    // unknown 卡片/标签必须放在最后做兜底显示
    unknownCard: unknownCard,
    unknownTag: unknownTag,
    clipboard: clipboard,
    paragraph: paragraphPlugin
  });

  return function createZhiPlugins(configs) {
    if (configs === void 0) {
      configs = {};
    }

    var result = [];
    sortPlugins(allPlugins).forEach(function (_ref) {
      var factory = _ref.factory;
      var plugin = factory(configs);
      result.push(plugin);
    });
    var schema = createSchema(result);
    result.push({
      schema: schema
    });
    return result;
  };
}

export default createCustomizedZhiPlugins;
