// @ali/4ever-zhi/dist-web/index.js
export { default as createZhiPlugins } from "./createZhiPlugins";
export { default as createMobileZhiPlugins } from "./createMobileZhiPlugins";
export { default as createCustomizedZhiPlugins, default as createMobileCustomizedZhiPlugins } from "./createCustomizedZhiPlugins";
export { default as plugins } from "./plugins";
export { default as mobilePlugins } from "./plugins/mobile";
export { default as Reader } from "./reader";
export { default as locale_en_US } from "./locales/en_US";
export { default as locale_ja_JP } from "./locales/ja_JP";
export { default as locale_zh_CN } from "./locales/zh_CN";
export { default as locale_zh_HK } from "./locales/zh_HK";
export { default as locale_zh_TW } from "./locales/zh_TW"; // Plugins

export { zhiFactory as MarksPlugin, mobileZhiFactory as MobileMarksPlugin } from '@ali/4ever-plugin-marks';
export { default as AudioPlugin } from "./plugins/audio";
export { default as MobileAudioPlugin } from "./plugins/audio/mobile";
export { default as BlockquotePlugin, default as MobileBlockquotePlugin } from "./plugins/blockquote";
export { zhiFactory as ClipboardPlugin, mobileZhiFactory as MobileClipboardPlugin } from '@ali/4ever-plugin-clipboard';
export { default as FormulaPlugin } from "./plugins/formula";
export { mobileZhiFactory as MobileCodePlugin, zhiFactory as CodePlugin } from '@ali/4ever-plugin-code';
export { zhiFactory as InlineCodePlugin } from '@ali/4ever-plugin-inline-code';
export { default as ColumnsPlugin } from "./plugins/columns";
export { default as MobileColumnsPlugin } from "./plugins/columns/mobile";
export { default as CommentPlugin } from "./plugins/comment";
export { default as MobileCommentPlugin } from "./plugins/comment/mobile";
export { zhiFactory as ContextMenuPlugin } from '@ali/4ever-plugin-pc-contex-menu';
export { default as EmbedPlugin } from "./plugins/embed";
export { default as MobileEmbedPlugin } from "./plugins/embed/mobile";
export { default as GroupTagPlugin, default as MobileGroupTagPlugin } from "./plugins/groupTag";
export { default as HeadingPlugin, default as MobileHeadingPlugin } from "./plugins/heading";
export { default as HrPlugin, default as MobileHrPlugin } from "./plugins/hr";
export { default as ImagePlugin } from "./plugins/image";
export { default as MobileImagePlugin } from "./plugins/image/mobile";
export { default as LinkPlugin } from "./plugins/link";
export { default as LinkCardPlugin } from "./plugins/link/linkCard/index";
export { default as MobileLinkPlugin } from "./plugins/link/mobile";
export { default as MobileLinkCardPlugin } from "./plugins/link/linkCard/mobile";
export { zhiFactory as ListPlugin, mobileZhiFactory as MobileListPlugin } from '@ali/4ever-plugin-list';
export { default as VideoPlugin } from "./plugins/video";
export { default as MobileVideoPlugin } from "./plugins/video/mobile";
export { zhiFactory as MentionPlugin, mobileZhiFactory as MobileMentionPlugin } from '@ali/4ever-plugin-mention';
export { default as OnlineVideoPlugin, default as MobileOnlineVideoPlugin } from "./plugins/onlineVideo";
export { zhiFactory as ParagraphPlugin, mobileZhiFactory as MobileParagraphPlugin } from '@ali/4ever-plugin-paragraph';
export { default as PlaceholderPlugin, default as MobilePlaceholderPlugin } from "./plugins/placeholder";
export { default as ReferPlugin, default as MobileReferPlugin } from "./plugins/refer";
export { zhiFactory as TablePlugin, mobileZhiFactory as MobileTablePlugin } from '@ali/4ever-plugin-table';
export { default as StickerPlugin, default as MobileStickerPlugin } from "./plugins/sticker";
export { default as ImTagPlugin, default as MobileImTagPlugin } from "./plugins/imTag";
export { default as CalendarCardPlugin } from "./plugins/calendarCard";
export { default as VoteCardPlugin } from "./plugins/voteCard";
export { createFindAndReplacePlugin as FindAndReplacePlugin } from '@ali/4ever-plugin-find-and-replace';
export { default as ProjectCardPlugin } from "./plugins/projectCard";
export { default as MobileCalendarCardPlugin } from "./plugins/calendarCard/mobile";
export { default as URLCardPlugin } from "./plugins/urlCard";
export { default as CalloutPlugin } from "./plugins/callout";
export { default as HeadingGroupPlugin } from "./plugins/headingGroup";
export { default as HeadingGroupNewPlugin } from "./plugins/headingGroupNew";
export { default as ListGroupPlugin } from "./plugins/listGroup";
export { default as SubjectPlugin } from "./plugins/subject";
export { zhiCardFactory as cardFactory } from '@ali/4ever-factory';
export { zhiGroupFactory as groupFactory } from '@ali/4ever-factory';
export { zhiTagFactory as tagFactory } from '@ali/4ever-factory';
export * from "./types";
