// @ali/we-design/dist-web/components/dropdown/Dropdown.js
import React, { useMemo } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Trigger } from "../trigger";
export var Dropdown = function Dropdown(props) {
  var children = props.children,
      trigger = props.trigger,
      overlay = props.overlay,
      getContainer = props.getContainer,
      onVisibleChange = props.onVisibleChange,
      placement = props.placement,
      overlayStyle = props.overlayStyle,
      visible = props.visible,
      zIndex = props.zIndex,
      overlayClassName = props.overlayClassName,
      offsetX = props.offsetX,
      offsetY = props.offsetY,
      pageX = props.pageX,
      pageY = props.pageY,
      triggerNode = props.triggerNode,
      title = props.title,
      testid = props.testid,
      delayDestroy = props.delayDestroy,
      delayShow = props.delayShow,
      delayHide = props.delayHide,
      disableDomAlign = props.disableDomAlign,
      overlayAutoAlign = props.overlayAutoAlign,
      overlayAutoAlignDelay = props.overlayAutoAlignDelay,
      _props$role = props.role,
      role = _props$role === void 0 ? '' : _props$role; // trigger's action

  var action = useMemo(function () {
    return trigger === 'contextMenu' ? [trigger] : [trigger];
  }, [trigger]);
  return /*#__PURE__*/_createElement(Trigger, {
    action: action,
    overlay: overlay,
    getContainer: getContainer,
    onVisibleChange: onVisibleChange,
    placement: placement,
    visible: visible,
    overlayStyle: overlayStyle,
    zIndex: zIndex,
    overlayClassName: overlayClassName,
    offsetX: offsetX,
    offsetY: offsetY,
    pageX: pageX,
    pageY: pageY,
    triggerNode: triggerNode,
    title: title,
    delayDestroy: delayDestroy,
    delayShow: delayShow,
    delayHide: delayHide,
    testid: testid,
    disableDomAlign: disableDomAlign,
    overlayAutoAlign: overlayAutoAlign,
    overlayAutoAlignDelay: overlayAutoAlignDelay,
    role: role
  }, children);
};
Dropdown.defaultProps = {
  trigger: 'click'
};
