// @ali/4ever-plugin-table/dist-web/bi/components/inlineToolbarButtons/inlineToolbarDeleteButton.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

/* eslint-disable react/no-find-dom-node */
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { ToolbarTableDeleteNormal, Tooltip } from '@ali/we-design';
import { ButtonWrapper, IconWrapper } from "../styled";
export var DeleteWrapper = /*#__PURE__*/styled(ButtonWrapper)(["&:hover{color:rgba(255,0,0,0.67);box-sizing:border-box;}"]);

var InlineToolbarDeleteButton = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(InlineToolbarDeleteButton, _React$Component);

  function InlineToolbarDeleteButton() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;

    _this.handleMouseDown = function (e) {
      e.preventDefault();
      e.stopPropagation();
      var onDelete = _this.props.onDelete;
      onDelete();
    };

    _this.handleMouseEnter = function () {
      var onHighlightSelection = _this.props.onHighlightSelection;
      onHighlightSelection();
    };

    _this.handleMouseLeave = function () {
      var onCancelHighlightSelection = _this.props.onCancelHighlightSelection;
      onCancelHighlightSelection();
    };

    return _this;
  }

  var _proto = InlineToolbarDeleteButton.prototype;

  _proto.componentWillUnmount = function componentWillUnmount() {
    var onCancelHighlightSelection = this.props.onCancelHighlightSelection;
    onCancelHighlightSelection();
  };

  _proto.render = function render() {
    var _this$props = this.props,
        mode = _this$props.mode,
        locale = _this$props.locale,
        scale = _this$props.scale;
    var placement = mode === 'row' ? 'left' : 'top';
    var tip = mode === 'row' ? locale.toolbarDeleteRow : locale.toolbarDeleteCol;
    return /*#__PURE__*/_createElement(DeleteWrapper, {
      scale: scale,
      onMouseDown: this.handleMouseDown,
      onMouseEnter: this.handleMouseEnter,
      onMouseLeave: this.handleMouseLeave,
      "data-testid": "table-delete-" + mode + "-button",
      size: 16
    }, /*#__PURE__*/_createElement(Tooltip, {
      placement: placement,
      title: tip
    }, /*#__PURE__*/_createElement(IconWrapper, null, /*#__PURE__*/_createElement(ToolbarTableDeleteNormal, {
      style: {
        fontSize: 16 / scale,
        width: 16 / scale + "px",
        height: 16 / scale + "px"
      }
    }))));
  };

  return InlineToolbarDeleteButton;
}(React.Component);

export { InlineToolbarDeleteButton as default };
