// @ali/4ever-plugin-table/dist-web/bi/commands/insertTable.js
import { Commands, TextPoint, Selection } from '@ali/4ever-cangjie';
import createTable from "../utils/createTable";
import createTableData from "../utils/createTableData";
import { biActions as basicActions } from '@ali/4ever-plugin-basic';
import { setTableRowHeader } from "../actions";
export default function insertTable(controller, payload) {
  var rowSize = payload.rowSize,
      colSize = payload.colSize,
      containerWidth = payload.containerWidth,
      defaultAutofit = payload.defaultAutofit,
      defaultHeader = payload.defaultHeader,
      setTableHeaderStyle = payload.setTableHeaderStyle;
  var tableData = createTableData(controller, {
    colSize: colSize,
    containerWidth: containerWidth,
    defaultAutofit: defaultAutofit
  });
  var table = createTable({
    rowSize: rowSize,
    colSize: colSize,
    data: tableData
  });
  var point = TextPoint.create({
    key: table.getFirstNode().key,
    offset: 0
  });
  var newSelection = Selection.create({
    anchor: point,
    focus: point
  });

  if (defaultHeader) {
    // NOTE: 需要先执行 setTableHeaderStyle
    setTableHeaderStyle == null ? void 0 : setTableHeaderStyle();
  }

  controller.run('onAction', basicActions.createInsertBlockWithoutExtraBlankAction({
    node: table
  }));
  controller.command(Commands.select, newSelection); // 自动定位到第一个单元格

  if (defaultHeader) {
    var newTable = controller.value.document.getNode(table.key);
    controller.run('onAction', setTableRowHeader(newTable));
  }

  return controller;
}
