// @ali/4ever-plugin-list/dist-web/bi/components/next/orderedListButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React, { useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { environment } from '@ali/4ever-cangjie';
import { ListSet } from '@ali/4ever-component';
import isOrderedList from "../../queries/isOrderedList";
import { getListStyleType, OLLIST_HOT_KEY_TIP } from "../../utils";
import { toggleList } from "../../actions";
import { ListRole } from "../../../utils/role";
import { DEFAULT_NUMBER_LIST_ARRAY as numberLists } from "../../../utils/listStyleMapping";
var isMobile = environment.IS_MOBILE;
export default function OrderedListButton(props) {
  var controller = props.controller,
      locale = props.locale,
      rest = _objectWithoutPropertiesLoose(props, ["controller", "locale"]);

  var listStyleType = getListStyleType(controller);
  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var isDisabled = isCollapsedAtEdge || controller.query('isSelectionInHr') || // fix: https://aone.alibaba-inc.com/v2/project/1090534/bug/35667506
  controller.query('isSelectionInSubjectHeading');
  var isActive = isOrderedList(controller);
  var handleClick = useCallback(function (listType) {
    controller.run('onAction', toggleList('olist', isMobile || numberLists[listType] === undefined ? numberLists[0] : listType));
  }, [controller]);
  var tooltip = React.useMemo(function () {
    return {
      title: locale.orderedList,
      shortCut: OLLIST_HOT_KEY_TIP,
      description: 'Markdown: 1. ' + (locale == null ? void 0 : locale.space) || 'Space'
    };
  }, []);
  return /*#__PURE__*/_createElement(ListSet, _extends({
    mode: "ol",
    active: isActive,
    testid: "bi-toolbar-ollist",
    disabled: !!isDisabled,
    defaultListStyle: listStyleType,
    onChange: handleClick,
    onClick: handleClick,
    role: ListRole.listOrdered,
    title: locale.orderedList,
    tooltip: tooltip
  }, rest));
}
