// @ali/4ever-plugin-list/dist-web/bi/components/next/unorderedListButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React, { useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { environment } from '@ali/4ever-cangjie';
import { ListSet } from '@ali/4ever-component';
import isUnorderedList from "../../queries/isUnorderedList";
import { toggleList } from "../../actions";
import { getListStyleType, ULLIST_HOT_KEY_TIP } from "../../utils";
import { ListRole } from "../../../utils/role";
import { DEFAULT_BULLET_LIST_ARRAY as bulletLists } from "../../../utils/listStyleMapping";
var isMobile = environment.IS_MOBILE;
export default function UnorderedListButton(props) {
  var controller = props.controller,
      locale = props.locale,
      _props$enableSticker = props.enableSticker,
      enableSticker = _props$enableSticker === void 0 ? false : _props$enableSticker,
      rest = _objectWithoutPropertiesLoose(props, ["controller", "locale", "enableSticker"]);

  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var isDisabled = isCollapsedAtEdge || controller.query('isSelectionInHr') || // fix: https://aone.alibaba-inc.com/v2/project/1090534/bug/35667506
  controller.query('isSelectionInSubjectHeading');
  var listStyleType = getListStyleType(controller);
  var isActive = isUnorderedList(controller);
  var handleClick = useCallback(function (listType) {
    return controller.run('onAction', toggleList('ulist', isMobile || bulletLists[listType] === undefined ? bulletLists[0] : listType));
  }, [controller]);
  var tooltip = React.useMemo(function () {
    return {
      title: locale.unorderedList,
      shortCut: ULLIST_HOT_KEY_TIP,
      description: 'Markdown: * ' + ((locale == null ? void 0 : locale.space) || 'Space')
    };
  }, []);
  return /*#__PURE__*/_createElement(ListSet, _extends({
    mode: "ul",
    active: isActive,
    testid: "bi-toolbar-ullist",
    disabled: !!isDisabled,
    defaultListStyle: listStyleType,
    onChange: handleClick,
    onClick: handleClick,
    enableSticker: enableSticker,
    role: ListRole.listUnordered,
    title: locale.unorderedList,
    tooltip: tooltip
  }, rest));
}
