// @ali/we-design/dist-web/components/button/SplitButton.js
import React, { useCallback, useMemo } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import useControlledValue from "../../hooks/useControlledValue";
import { Dropdown } from "../../components/dropdown";
import { ArrowDownLineNormal } from "../../components/icons";
import { SplitButtonWrap, SplitButtonLeft, SplitButtonRight, SplitButtonInner } from "./styled";
import { classnames } from "../../utils";
import { DropdownAnimation } from '@ali/we-util';
var SplitButton = /*#__PURE__*/React.memo(function (props) {
  var disabled = props.disabled,
      overlay = props.overlay,
      overlayStyle = props.overlayStyle,
      className = props.className,
      onActiveChange = props.onActiveChange,
      onVisibleChange = props.onVisibleChange,
      children = props.children,
      style = props.style,
      size = props.size,
      bordered = props.bordered,
      onClickLeft = props.onClickLeft,
      zIndex = props.zIndex,
      testid = props.testid,
      title = props.title,
      highlight = props.highlight,
      getContainer = props.getContainer,
      _props$role = props.role,
      role = _props$role === void 0 ? '' : _props$role,
      _props$roleRight = props.roleRight,
      roleRight = _props$roleRight === void 0 ? '' : _props$roleRight;

  var _useControlledValue = useControlledValue(props.active, props.defaultActive),
      active = _useControlledValue[0],
      setActive = _useControlledValue[1];
  /**
   * 下拉面板是否可见修改函数
   *
   * @param {boolean} visible
   * @return {void}
   */


  var handleVisibleChange = useCallback(function (visible) {
    onVisibleChange && onVisibleChange(visible);

    if (!disabled) {
      if (onActiveChange) {
        onActiveChange(visible);
      }

      setActive(visible);
    }
  }, [onActiveChange, setActive, disabled]);
  /**
   * 点击左部区域
   *
   * @param {React.MouseEvent<HTMLElement>}
   * @return {void}
   */

  var handleClickLeft = useCallback(function (e) {
    if (!disabled) {
      if (!active) {
        e.stopPropagation();
      }

      if (onClickLeft) {
        onClickLeft(e);
      }
    }
  }, [disabled, active, onClickLeft]);
  var overlayMemo = useMemo(function () {
    return /*#__PURE__*/React.cloneElement(overlay, {
      'data-testid': testid && "overlay-" + testid
    });
  }, [overlay, testid]);
  var visible = disabled ? false : active;
  return /*#__PURE__*/_createElement(Dropdown, {
    title: title,
    visible: visible,
    overlay: /*#__PURE__*/_createElement(DropdownAnimation, {
      visible: !!visible,
      "data-testid": "splite-overlay-" + testid
    }, overlayMemo),
    overlayStyle: overlayStyle,
    onVisibleChange: handleVisibleChange,
    zIndex: zIndex,
    getContainer: getContainer
  }, /*#__PURE__*/_createElement(SplitButtonWrap, {
    active: active,
    disabled: disabled,
    className: classnames(className, {
      active: active || highlight
    }),
    size: size,
    style: style,
    bordered: bordered
  }, /*#__PURE__*/_createElement(SplitButtonInner, null, /*#__PURE__*/_createElement(SplitButtonLeft, {
    disabled: disabled,
    onClick: handleClickLeft,
    "data-testid": testid && "splitbutton-left-" + testid,
    className: "splitbutton-left",
    "data-role": role
  }, children), /*#__PURE__*/_createElement(SplitButtonRight, {
    active: active,
    disabled: disabled,
    className: "splitbutton-right",
    "data-role": roleRight
  }, /*#__PURE__*/_createElement(ArrowDownLineNormal, {
    className: classnames('arrow-icon', {
      opened: active || highlight
    }),
    "data-testid": testid && "splitbutton-arrow-" + testid
  })))));
});
SplitButton.defaultProps = {
  defaultActive: false,
  size: 'normal',
  bordered: false
};
export { SplitButton };
