// @ali/we-design/dist-web/components/button/styled.js
/* eslint-disable no-nested-ternary */

/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/explicit-function-return-type */
// /* eslint-disable @typescript-eslint/explicit-function-return-type */
import styled, { css } from 'styled-components';
import { withTheme } from "../../components/theme";
import { resetComponent, Ellipsis } from "../../utils/styled.mixin";
var commonButton = /*#__PURE__*/css(["", " display:inline-block;width:fit-content;border:1px solid ", ";outline:0;height:", ";", ";min-width:", ";padding:", ";cursor:", ";transition:all 0.2s ease-out;user-select:none;"], resetComponent, function (_ref) {
  var theme = _ref.theme;
  return theme.ui.commonLineLightColor;
}, function (_ref2) {
  var theme = _ref2.theme,
      size = _ref2.size;
  var sizeString = '';
  if (size === 'small') sizeString = 'Small';
  if (size === 'large') sizeString = 'Large';
  return theme.ui["elementHeight" + sizeString];
}, function (_ref3) {
  var theme = _ref3.theme,
      onlyIcon = _ref3.onlyIcon,
      size = _ref3.size;
  var isRect = onlyIcon;
  var width = theme.ui["elementHeight" + (size === 'small' ? 'Small' : size === 'large' ? 'Large' : '')];
  return isRect ? "\n      width: " + width + ";\n    " : '';
}, function (_ref4) {
  var theme = _ref4.theme,
      size = _ref4.size,
      onlyIcon = _ref4.onlyIcon;
  var sizeString = '';
  if (onlyIcon) return 'auto';
  if (size === 'small') sizeString = 'Small';
  if (size === 'large') sizeString = 'Large';
  return theme.ui["elementHeight" + sizeString];
}, function (_ref5) {
  var theme = _ref5.theme,
      onlyIcon = _ref5.onlyIcon;
  if (onlyIcon) return '0';
  return "0 " + theme.ui.paddingLarge;
}, function (_ref6) {
  var disabled = _ref6.disabled;
  return disabled ? 'not-allowed' : 'pointer';
});
var roundButton = /*#__PURE__*/css(["border-radius:", ";"], function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.elementHeightSmall;
});
var rectButton = /*#__PURE__*/css(["border-radius:", ";"], function (_ref8) {
  var size = _ref8.size;
  if (size === 'large') return '5px';
  return '4px';
});
var circleButton = /*#__PURE__*/css(["border-radius:50%;"]);
var primaryButton = /*#__PURE__*/css(["background-color:rgb(0,137,255);color:", ";", ";"], function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.colorGray1;
}, function (_ref10) {
  var theme = _ref10.theme,
      disabled = _ref10.disabled;

  if (!disabled) {
    return "\n        &:hover{\n          // border-color: " + theme.ui.primaryHoverColor2 + ";\n          // background-color: " + theme.ui.primaryHoverColor2 + ";\n          border-color: rgb(57,163,255);\n          background-color: rgb(57,163,255);\n        }\n\n        &:active, &.active {\n          // border-color: " + theme.ui.primaryPressColor2 + ";\n          // background-color: " + theme.ui.primaryPressColor2 + ";\n          border-color: rgb(0,101,213);\n          background-color: rgb(0,101,213);\n        }\n      ";
  }

  return "\n      // border-color: " + theme.ui.primaryDisabledColor + ";\n      // background-color: " + theme.ui.primaryDisabledColor + ";\n      border-color: transparent;\n      background-color: rgba(0,137,255,0.48);\n      color: rgba(255, 255, 255, 0.4);\n    ";
});
var normalButton = /*#__PURE__*/css(["border-color:", ";", ";", ";"], function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.borderColor;
}, function (_ref12) {
  var onlyIcon = _ref12.onlyIcon;
  return !onlyIcon && 'background-color: white';
}, function (_ref13) {
  var theme = _ref13.theme,
      disabled = _ref13.disabled,
      bordered = _ref13.bordered,
      onlyIcon = _ref13.onlyIcon;

  if (!disabled) {
    return "\n        &:hover {\n          // color: " + theme.ui.buttonNormalHoverTextColor + ";\n          color: rgb(57,163,255);\n          " + (bordered && "\n              // border-color: " + theme.ui.buttonNormalHoverBorderColor + ";\n              border-color:rgba(0, 137, 255, 0.48);\n            ") + ";\n        }\n\n        &:active {\n          // background-color: " + theme.ui.buttonNormalClickBgColor2 + ";\n          // color: " + theme.ui.buttonNormalClickTextColor + ";\n          color: rgb(0,101,213);\n          background-color: rgba(0, 137, 255, 0.12);\n          " + (bordered && "\n              // border-color: " + theme.ui.buttonNormalClickBorderColor + ";\n              border-color: rgba(0, 137, 255, 0.48);\n            ") + ";\n        }\n\n        &.active {\n          // background-color: " + theme.ui.buttonNormalClickBgColor2 + ";\n          // color: " + theme.ui.buttonNormalClickTextColor + ";\n          color: rgb(0,101,213);\n          background-color: rgba(0, 137, 255, 0.12);\n          " + (bordered && "\n              // border-color: " + theme.ui.buttonNormalClickBorderColor + ";\n              border-color: rgba(0, 137, 255, 0.48);\n            ") + ";\n        }\n      ";
  }

  return "\n      border-color: " + theme.ui.buttonNormalDisabledBorderColor + ";\n      // background-color: transparent;\n      // color: " + theme.ui.disabledColor + ";\n      background-color: " + (onlyIcon ? 'transparent' : 'white') + ";\n      color: rgba(23,26,29,0.24);\n    ";
});
var dangerButton = /*#__PURE__*/css(["background-color:", ";border-color:", ";color:", ";", ";"], function (_ref14) {
  var theme = _ref14.theme;
  return theme.ui.buttonDangerBgColor;
}, function (_ref15) {
  var theme = _ref15.theme;
  return theme.ui.buttonDangerBgColor;
}, function (_ref16) {
  var theme = _ref16.theme;
  return theme.ui.colorGray1;
}, function (_ref17) {
  var theme = _ref17.theme,
      disabled = _ref17.disabled;

  if (!disabled) {
    return "\n        &:hover {\n          background-color: " + theme.ui.buttonDangerHoverBgColor + ";\n        }\n\n        &:active, &.active {\n          background-color: " + theme.ui.buttonDangerActiveBgColor + ";\n        }\n      ";
  }

  return "\n      border-color: " + theme.ui.buttonDangerDisabledBgColor + ";\n      background-color: " + theme.ui.buttonDangerDisabledBgColor + ";\n    ";
});
var ghostButton = /*#__PURE__*/css(["background-color:transparent;border-color:", ";color:", ";", ";"], function (_ref18) {
  var theme = _ref18.theme;
  return theme.ui.colorGray1;
}, function (_ref19) {
  var theme = _ref19.theme;
  return theme.ui.colorGray1;
}, function (_ref20) {
  var theme = _ref20.theme,
      disabled = _ref20.disabled;

  if (!disabled) {
    return "\n        &:hover {\n          background-color: " + theme.ui.buttonGhostHoverBgColor + ";\n          border-color: " + theme.ui.buttonGhostHoverBorderColor + ";\n        }\n\n        &:active, &.active {\n          background-color: " + theme.ui.buttonGhostActiveBgColor + ";\n          border-color: " + theme.ui.buttonGhostActiveBorderColor + ";\n        }\n      ";
  }

  return "\n      border-color: " + theme.ui.buttonGhostDisabledBorderColor + ";\n      color: " + theme.ui.buttonGhostDisabledBorderColor + ";\n    ";
});
var textButton = /*#__PURE__*/css(["background-color:transparent;border-color:transparent;color:rgb(0,137,255);padding:0;", ";"], function (_ref21) {
  var theme = _ref21.theme,
      disabled = _ref21.disabled;

  if (!disabled) {
    return "\n        &:hover {\n          // color: " + theme.ui.buttonPrimaryHoverBgColor + ";\n          color: rgb(57,163,255);\n        }\n\n        &:active, &.active {\n          // color: " + theme.ui.buttonPrimaryActiveBgColor + ";\n          color: rgb(0,101,213);\n        }\n      ";
  }

  return "\n      // color: " + theme.ui.buttonPrimaryDisabledBgColor + ";\n      color: rgba(23,26,29,0.4);\n    ";
});
export var ButtonInner = /*#__PURE__*/styled.div(["opacity:", ";box-sizing:border-box;height:100%;display:flex;align-items:center;justify-content:center;", ";"], function (_ref22) {
  var visible = _ref22.visible;
  return visible ? '1' : '0';
}, Ellipsis);
export var ButtonChildrenWrap = /*#__PURE__*/styled.div(["max-width:100%;", ""], Ellipsis);
export var ButtonWrap = withTheme(styled.div(["position:relative;", " ", " ", " ", " ", " ", " ", " ", " ", " ", " ", " ", ""], resetComponent, commonButton, function (_ref23) {
  var disabled = _ref23.disabled,
      theme = _ref23.theme,
      onlyIcon = _ref23.onlyIcon;
  return disabled ? '' : onlyIcon && ".weicon {color: " + theme.ui.commonActionIconColor + ";}";
}, function (_ref24) {
  var shape = _ref24.shape;
  return shape === 'round' && roundButton;
}, function (_ref25) {
  var shape = _ref25.shape;
  return shape === 'rect' && rectButton;
}, function (_ref26) {
  var shape = _ref26.shape;
  return shape === 'circle' && circleButton;
}, function (_ref27) {
  var type = _ref27.type;
  return type === 'primary' && primaryButton;
}, function (_ref28) {
  var type = _ref28.type;
  return type === 'normal' && normalButton;
}, function (_ref29) {
  var type = _ref29.type;
  return type === 'danger' && dangerButton;
}, function (_ref30) {
  var type = _ref30.type;
  return type === 'ghost' && ghostButton;
}, function (_ref31) {
  var type = _ref31.type;
  return type === 'text' && textButton;
}, function (_ref32) {
  var bordered = _ref32.bordered;
  return !bordered && "\n    border-color: transparent !important;\n\n  ";
}));
export var SplitButtonWrap = withTheme(styled.div(["", " ", " ", " border:", ";border-radius:", "};padding:0;overflow:hidden;", ";&.leftSelected .splitbutton-left{background-color:", ";}&.leftSelected .splitbutton-right{background-color:", ";}"], resetComponent, commonButton, function (_ref33) {
  var disabled = _ref33.disabled,
      theme = _ref33.theme;
  return disabled ? '' : ".weicon {color: " + theme.ui.commonActionIconColor + ";}";
}, function (_ref34) {
  var bordered = _ref34.bordered,
      theme = _ref34.theme;
  return bordered ? "1px solid " + theme.ui.commonLineLightColor : 'none';
}, function (_ref35) {
  var theme = _ref35.theme;
  return theme.ui.radius;
}, function (_ref36) {
  var theme = _ref36.theme,
      disabled = _ref36.disabled;

  if (!disabled) {
    return "\n        &:hover, &.active {\n          background-color: " + theme.ui.splitButtonBgColor + ";\n        }\n      ";
  }

  return "\n      cursor: not-allowed;\n      background-color: transparent;\n      color: " + theme.ui.disabledColor + ";\n\n      // \u63D0\u5347\u6B64\u89C4\u5219\u7684\u4F18\u5148\u7EA7\n      .arrow-icon.arrow-icon {\n        color: " + theme.ui.disabledColor + ";\n      }\n    ";
}, function (_ref37) {
  var theme = _ref37.theme;
  return theme.ui.splitButtonLeftActiveBgColor;
}, function (_ref38) {
  var theme = _ref38.theme;
  return theme.ui.splitButtonBgColor;
}));
export var SplitButtonInner = /*#__PURE__*/styled.div(["box-sizing:border-box;height:100%;display:flex;align-items:stretch;justify-content:center;", ""], Ellipsis);
export var SplitButtonLeft = withTheme(styled.div(["display:flex;align-items:center;justify-content:center;padding:0 4px 0 4px;margin:2px 0 2px 0;transition:background-color 0.1s ease-in-out;", ""], function (_ref39) {
  var disabled = _ref39.disabled,
      theme = _ref39.theme;

  if (!disabled) {
    return "\n          &:hover {\n            padding-right: 3px;\n            border-right: 1px solid " + theme.ui.commonLineLightColor + ";\n          }\n          &:active {\n            background-color: " + theme.ui.splitButtonLeftActiveBgColor + ";\n          }\n        ";
  }

  return 'opacity:0.5';
}));
export var SplitButtonRight = withTheme(styled.div(["display:flex;align-items:center;justify-content:center;width:10px;padding:0 4px 0 1px;box-sizing:content-box;", " .arrow-icon{transform:", ";font-size:12px;color:", ";}"], function (_ref40) {
  var disabled = _ref40.disabled,
      theme = _ref40.theme,
      active = _ref40.active;

  if (!disabled) {
    if (active) {
      return "background-color: " + theme.ui.splitButtonActiveBgColor + ";";
    }

    return "\n          &:hover {\n            background-color: " + theme.ui.splitButtonHoverBgColor + ";\n          }\n        ";
  }

  return '';
}, function (_ref41) {
  var active = _ref41.active;
  return active ? 'rotate(180deg)' : 'rotate(0)';
}, function (_ref42) {
  var theme = _ref42.theme;
  return theme.ui.commonBaseColor3;
}));
export var LoadingWrapper = /*#__PURE__*/styled.div(["opacity:", ";display:flex;align-items:center;justify-content:center;position:absolute;top:0;right:0;bottom:0;left:0;"], function (_ref43) {
  var visible = _ref43.visible;
  return visible ? '1' : '0';
});
export var IWaveWrap = withTheme(styled.div(["display:inline-block;height:", ";@keyframes waveEffect-normal{100%{box-shadow:0 0 0 3px #1890ff;}}@keyframes waveEffect-primary{100%{box-shadow:0 0 0 3px #1890ff;}}@keyframes waveEffect-danger{100%{box-shadow:0 0 0 3px red;}}@keyframes waveEffect-ghost{100%{box-shadow:0 0 0 3px #1890ff;}}@keyframes fadeEffect{100%{opacity:0;}}.we-click-wave-animating-normal{position:relative;}.we-click-wave-animating-primary{position:relative;}.we-click-wave-animating-danger{position:relative;}.we-click-wave-animating-ghost{position:relative;}.we-click-wave-animating-normal::after{position:absolute;top:0;right:0;bottom:0;left:0;display:block;border-radius:inherit;box-shadow:0 0 0 0 #1890ff;opacity:0.2;animation:fadeEffect 2s cubic-bezier(0.08,0.82,0.17,1),waveEffect-normal 0.4s cubic-bezier(0.08,0.82,0.17,1);animation-fill-mode:forwards;pointer-events:none;content:'';}.we-click-wave-animating-primary::after{position:absolute;top:0;right:0;bottom:0;left:0;display:block;border-radius:inherit;box-shadow:0 0 0 0 #1890ff;opacity:0.2;animation:fadeEffect 2s cubic-bezier(0.08,0.82,0.17,1),waveEffect-primary 0.4s cubic-bezier(0.08,0.82,0.17,1);animation-fill-mode:forwards;pointer-events:none;content:'';}.we-click-wave-animating-danger::after{position:absolute;top:0;right:0;bottom:0;left:0;display:block;border-radius:inherit;box-shadow:0 0 0 0 red;opacity:0.2;animation:fadeEffect 2s cubic-bezier(0.08,0.82,0.17,1),waveEffect-danger 0.4s cubic-bezier(0.08,0.82,0.17,1);animation-fill-mode:forwards;pointer-events:none;content:'';}.we-click-wave-animating-ghost::after{position:absolute;top:0;right:0;bottom:0;left:0;display:block;border-radius:inherit;box-shadow:0 0 0 0 #1890ff;opacity:0.2;animation:fadeEffect 2s cubic-bezier(0.08,0.82,0.17,1),waveEffect-ghost 0.4s cubic-bezier(0.08,0.82,0.17,1);animation-fill-mode:forwards;pointer-events:none;content:'';}"], function (_ref44) {
  var theme = _ref44.theme,
      size = _ref44.size;
  var sizeString = '';
  if (size === 'small') sizeString = 'Small';
  if (size === 'large') sizeString = 'Large';
  return theme.ui["elementHeight" + sizeString];
}));
