// @ali/4ever-plugin-container/dist-node/mo/models/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Container = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

var _meta = require("../../utils/meta");

class Container extends _everCangjie.Block {
  static createContainer(properties) {
    if (typeof properties === 'string') {
      return super.create({
        type: Container.TYPE,
        data: {
          subType: properties
        }
      });
    }

    return super.create({ ...properties,
      type: Container.TYPE
    });
  }

  static isContainer(maybeContainer, subType) {
    if (!_everCangjie.Block.isBlock(maybeContainer)) {
      return false;
    }

    const matchSubType = subType ? maybeContainer.data.subType === subType : true;
    return maybeContainer.type === Container.TYPE && matchSubType;
  }
  /**
   * @deprecated
   * 临时冗余，待 container 迁移为 cangjie 内置插件后移除
   */


  static isVoid(controller, node) {
    const isTemplateButton = Container.isContainer(node, 'templateButton');
    const {
      injections
    } = controller.value;
    const isTemplateButtonOpen = injections.find(inj => inj.data?.metadata?.isOpen !== undefined && inj.key === node.key)?.data?.metadata?.isOpen ?? false;
    return isTemplateButton && !isTemplateButtonOpen;
  }

}

exports.Container = Container;
Container.TYPE = _meta.ELEMENT_TYPE;
