// @ali/4ever-bamboo/dist-node/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  List: true,
  styledComponents: true,
  H1: true,
  H2: true,
  H3: true,
  H4: true,
  H5: true,
  H6: true,
  getHeadings: true,
  createHeadingId: true,
  P: true,
  TD: true,
  useContainerWidthResize: true,
  isPureTableNode: true,
  isTableNode: true,
  BlockquoteWrapper: true,
  Placeholder: true,
  PlaceholderWrapper: true,
  InlinePlaceholder: true,
  ErrorBoundary: true,
  Select: true,
  InputNumberItem: true,
  CustomInputNumberItem: true,
  useScrollableContainer: true,
  useScrollableContent: true,
  convertMarks: true,
  throttleByIdleCallback: true,
  copyToClipboard: true,
  UnitConvert: true,
  isFirstPart: true,
  addDefaultUnit: true,
  transformThemeObjToCssString: true,
  isNaN: true,
  ptToPx: true,
  unitsConverter: true,
  sample: true,
  Sample: true,
  SimpleMapCache: true,
  fontRenderFallback: true,
  ShdUtil: true,
  UnderlineUtil: true,
  calculatedPluginSize: true,
  SpacingUtil: true,
  IndUtil: true,
  SzUtil: true,
  CommonFormatUtil: true,
  GradientUtil: true,
  isEmptyParagraph: true,
  zIndex: true,
  isSingleTableCell: true,
  unwrapSingleTableCell: true,
  fontSize: true,
  lineHeight: true,
  blockMarginStyleInContainer: true,
  theme: true
};
exports.styledComponents = styledComponents;
Object.defineProperty(exports, "H1", {
  enumerable: true,
  get: function () {
    return _everPluginHeading.H1;
  }
});
Object.defineProperty(exports, "H2", {
  enumerable: true,
  get: function () {
    return _everPluginHeading.H2;
  }
});
Object.defineProperty(exports, "H3", {
  enumerable: true,
  get: function () {
    return _everPluginHeading.H3;
  }
});
Object.defineProperty(exports, "H4", {
  enumerable: true,
  get: function () {
    return _everPluginHeading.H4;
  }
});
Object.defineProperty(exports, "H5", {
  enumerable: true,
  get: function () {
    return _everPluginHeading.H5;
  }
});
Object.defineProperty(exports, "H6", {
  enumerable: true,
  get: function () {
    return _everPluginHeading.H6;
  }
});
Object.defineProperty(exports, "getHeadings", {
  enumerable: true,
  get: function () {
    return _everPluginHeading.getHeadings;
  }
});
Object.defineProperty(exports, "createHeadingId", {
  enumerable: true,
  get: function () {
    return _everPluginHeading.createHeadingId;
  }
});
Object.defineProperty(exports, "P", {
  enumerable: true,
  get: function () {
    return _everPluginParagraph.P;
  }
});
Object.defineProperty(exports, "TD", {
  enumerable: true,
  get: function () {
    return _everPluginTable.TD;
  }
});
Object.defineProperty(exports, "useContainerWidthResize", {
  enumerable: true,
  get: function () {
    return _everPluginTable.useContainerWidthResize;
  }
});
Object.defineProperty(exports, "isPureTableNode", {
  enumerable: true,
  get: function () {
    return _everPluginTable.isPureTableNode;
  }
});
Object.defineProperty(exports, "isTableNode", {
  enumerable: true,
  get: function () {
    return _everPluginTable.isTableNode;
  }
});
Object.defineProperty(exports, "BlockquoteWrapper", {
  enumerable: true,
  get: function () {
    return _everPluginBlockquote.BlockquoteWrapper;
  }
});
Object.defineProperty(exports, "Placeholder", {
  enumerable: true,
  get: function () {
    return _everComponent.BambooPlaceholder;
  }
});
Object.defineProperty(exports, "PlaceholderWrapper", {
  enumerable: true,
  get: function () {
    return _everComponent.PlaceholderWrapper;
  }
});
Object.defineProperty(exports, "InlinePlaceholder", {
  enumerable: true,
  get: function () {
    return _everComponent.InlinePlaceholder;
  }
});
Object.defineProperty(exports, "ErrorBoundary", {
  enumerable: true,
  get: function () {
    return _everComponent.ErrorBoundary;
  }
});
Object.defineProperty(exports, "Select", {
  enumerable: true,
  get: function () {
    return _everComponent.Select;
  }
});
Object.defineProperty(exports, "InputNumberItem", {
  enumerable: true,
  get: function () {
    return _everComponent.InputNumberItem;
  }
});
Object.defineProperty(exports, "CustomInputNumberItem", {
  enumerable: true,
  get: function () {
    return _everComponent.CustomInputNumberItem;
  }
});
Object.defineProperty(exports, "useScrollableContainer", {
  enumerable: true,
  get: function () {
    return _everCangjie.useScrollableContainer;
  }
});
Object.defineProperty(exports, "useScrollableContent", {
  enumerable: true,
  get: function () {
    return _everCangjie.useScrollableContent;
  }
});
Object.defineProperty(exports, "convertMarks", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.marksToStyle;
  }
});
Object.defineProperty(exports, "throttleByIdleCallback", {
  enumerable: true,
  get: function () {
    return _everUtils.throttleByIdleCallback;
  }
});
Object.defineProperty(exports, "copyToClipboard", {
  enumerable: true,
  get: function () {
    return _everUtils.copyToClipboard;
  }
});
Object.defineProperty(exports, "UnitConvert", {
  enumerable: true,
  get: function () {
    return _everUtils.UnitConvert;
  }
});
Object.defineProperty(exports, "isFirstPart", {
  enumerable: true,
  get: function () {
    return _everUtils.isFirstPart;
  }
});
Object.defineProperty(exports, "addDefaultUnit", {
  enumerable: true,
  get: function () {
    return _everUtils.addDefaultUnit;
  }
});
Object.defineProperty(exports, "transformThemeObjToCssString", {
  enumerable: true,
  get: function () {
    return _everUtils.transformThemeObjToCssString;
  }
});
Object.defineProperty(exports, "isNaN", {
  enumerable: true,
  get: function () {
    return _everUtils.isNaN;
  }
});
Object.defineProperty(exports, "ptToPx", {
  enumerable: true,
  get: function () {
    return _everUtils.ptToPx;
  }
});
Object.defineProperty(exports, "unitsConverter", {
  enumerable: true,
  get: function () {
    return _everUtils.unitsConverter;
  }
});
Object.defineProperty(exports, "sample", {
  enumerable: true,
  get: function () {
    return _everUtils.sample;
  }
});
Object.defineProperty(exports, "Sample", {
  enumerable: true,
  get: function () {
    return _everUtils.Sample;
  }
});
Object.defineProperty(exports, "SimpleMapCache", {
  enumerable: true,
  get: function () {
    return _everUtils.SimpleMapCache;
  }
});
Object.defineProperty(exports, "fontRenderFallback", {
  enumerable: true,
  get: function () {
    return _everUtils.fontRenderFallback;
  }
});
Object.defineProperty(exports, "ShdUtil", {
  enumerable: true,
  get: function () {
    return _everUtils.ShdUtil;
  }
});
Object.defineProperty(exports, "UnderlineUtil", {
  enumerable: true,
  get: function () {
    return _everUtils.UnderlineUtil;
  }
});
Object.defineProperty(exports, "calculatedPluginSize", {
  enumerable: true,
  get: function () {
    return _everUtils.calculatedPluginSize;
  }
});
Object.defineProperty(exports, "SpacingUtil", {
  enumerable: true,
  get: function () {
    return _everUtils.SpacingUtil;
  }
});
Object.defineProperty(exports, "IndUtil", {
  enumerable: true,
  get: function () {
    return _everUtils.IndUtil;
  }
});
Object.defineProperty(exports, "SzUtil", {
  enumerable: true,
  get: function () {
    return _everUtils.SzUtil;
  }
});
Object.defineProperty(exports, "CommonFormatUtil", {
  enumerable: true,
  get: function () {
    return _everUtils.CommonFormatUtil;
  }
});
Object.defineProperty(exports, "GradientUtil", {
  enumerable: true,
  get: function () {
    return _everUtils.GradientUtil;
  }
});
Object.defineProperty(exports, "isEmptyParagraph", {
  enumerable: true,
  get: function () {
    return _everUtils.isEmptyParagraph;
  }
});
Object.defineProperty(exports, "zIndex", {
  enumerable: true,
  get: function () {
    return _everUtils.zIndex;
  }
});
Object.defineProperty(exports, "isSingleTableCell", {
  enumerable: true,
  get: function () {
    return _everUtils.isSingleTableCell;
  }
});
Object.defineProperty(exports, "unwrapSingleTableCell", {
  enumerable: true,
  get: function () {
    return _everUtils.unwrapSingleTableCell;
  }
});
Object.defineProperty(exports, "fontSize", {
  enumerable: true,
  get: function () {
    return _everUtils.fontSize;
  }
});
Object.defineProperty(exports, "lineHeight", {
  enumerable: true,
  get: function () {
    return _everUtils.lineHeight;
  }
});
Object.defineProperty(exports, "blockMarginStyleInContainer", {
  enumerable: true,
  get: function () {
    return _everUtils.blockMarginStyleInContainer;
  }
});
Object.defineProperty(exports, "theme", {
  enumerable: true,
  get: function () {
    return _everUtils.theme;
  }
});
exports.List = void 0;

var _everPluginList = require("@ali/4ever-plugin-list");

var _everPluginHeading = require("@ali/4ever-plugin-heading");

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _everPluginTable = require("@ali/4ever-plugin-table");

var _everPluginBlockquote = require("@ali/4ever-plugin-blockquote");

var _everComponent = require("@ali/4ever-component");

var _everCangjie = require("@ali/4ever-cangjie");

var _plugins = require("./plugins");

Object.keys(_plugins).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _plugins[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _plugins[key];
    }
  });
});

var _everPluginMarks = require("@ali/4ever-plugin-marks");

var _everUtils = require("@ali/4ever-utils");

// reexport
const List = {
  Ul: _everPluginList.Ul,
  Ol: _everPluginList.Ol,
  Li: _everPluginList.Li
};
exports.List = List;
const typeToComponent = {
  'table-cell': _everPluginTable.TD,
  'heading-1': _everPluginHeading.H1,
  'heading-2': _everPluginHeading.H2,
  'heading-3': _everPluginHeading.H3,
  'heading-4': _everPluginHeading.H4,
  'heading-5': _everPluginHeading.H5,
  'heading-6': _everPluginHeading.H6,
  paragraph: _everPluginParagraph.P
};

function styledComponents(type) {
  return typeToComponent[type];
}
