// @ali/4ever-plugin-indent/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "IndentRole", {
  enumerable: true,
  get: function () {
    return _role.IndentRole;
  }
});
Object.defineProperty(exports, "INDENT_SIZE", {
  enumerable: true,
  get: function () {
    return _constants.INDENT_SIZE;
  }
});
Object.defineProperty(exports, "INDENT_SIZE_CHARS", {
  enumerable: true,
  get: function () {
    return _constants.INDENT_SIZE_CHARS;
  }
});
Object.defineProperty(exports, "MIN_INDENT_SIZE", {
  enumerable: true,
  get: function () {
    return _constants.MIN_INDENT_SIZE;
  }
});
Object.defineProperty(exports, "MAX_INDENT", {
  enumerable: true,
  get: function () {
    return _constants.MAX_INDENT;
  }
});
Object.defineProperty(exports, "MAX_INDENT_SIZE", {
  enumerable: true,
  get: function () {
    return _constants.MAX_INDENT_SIZE;
  }
});
Object.defineProperty(exports, "MAX_INDENT_SIZE_CHARS", {
  enumerable: true,
  get: function () {
    return _constants.MAX_INDENT_SIZE_CHARS;
  }
});
Object.defineProperty(exports, "getHighestSelectedBlocks", {
  enumerable: true,
  get: function () {
    return _getHighestSelectedBlocks.getHighestSelectedBlocks;
  }
});
exports.OldBambooPlugin = void 0;

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;

var _role = require("./utils/role");

var _constants = require("./utils/constants");

var _getHighestSelectedBlocks = require("./utils/getHighestSelectedBlocks");
