// @ali/4ever-plugin-template-button/dist-node/mo/models/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TemplateButton = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

var _everPluginContainer = require("@ali/4ever-plugin-container");

var _meta = require("../../utils/meta");

class TemplateButton extends _everPluginContainer.Container {
  static isTemplateButton(node) {
    return _everPluginContainer.Container.isContainer(node, TemplateButton.SUB_TYPE);
  }

  static isOpen(controller, key) {
    const {
      injections
    } = controller.value;
    return injections.find(inj => inj.data?.metadata?.isOpen !== undefined && inj.key === key)?.data?.metadata?.isOpen ?? false;
  }

  static isVoid(controller, node) {
    return TemplateButton.isTemplateButton(node) && !TemplateButton.isOpen(controller, node.key);
  }

  static createTemplateButton(metadata = {}, listId = Math.random().toString(36).slice(2)) {
    return _everPluginContainer.Container.createContainer({
      data: {
        subType: TemplateButton.SUB_TYPE,
        metadata: {
          isOnce: false,
          direction: 'bottom',
          title: '',
          ...metadata
        }
      },
      nodes: [_everCangjie.Block.create({
        type: 'paragraph',
        data: {
          list: {
            level: 0,
            isChecked: false,
            isOrdered: false,
            isTaskList: true,
            listId
          }
        },
        nodes: [_everCangjie.Text.create()]
      })]
    });
  }

}

exports.TemplateButton = TemplateButton;
TemplateButton.SUB_TYPE = _meta.ELEMENT_TYPE;
