// @ali/4ever-plugin-clipboard/dist-node/bi/handlers/createOnEditable.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _ClipboardPad = _interopRequireDefault(require("../../utils/ClipboardPad"));

var _lazyPastePanel = _interopRequireDefault(require("../components/lazyPastePanel"));

const _createElement = /*#__PURE__*/_react.default.createElement;

const createOnEditable = clipboard => {
  let mounted = true;
  return (props, controller, next) => {
    const {
      customPanel
    } = controller.value.data;

    const clipboardPad = _ClipboardPad.default.getCachedClipboardPad(clipboard?.key);

    if (mounted) {
      clipboardPad?.initPluginModels(controller);
      mounted = false;
    }

    return /*#__PURE__*/_createElement(_react.default.Fragment, null, next(), /*#__PURE__*/_createElement(_lazyPastePanel.default, {
      config: clipboard,
      customPanel: customPanel,
      controller: controller,
      clipboardPad: clipboardPad,
      locale: clipboard.locale,
      beforeInsertMarkdown: clipboard.beforeInsertMarkdown,
      afterInsertMarkdown: clipboard.afterInsertMarkdown,
      getContainer: clipboard.getContainer
    }));
  };
};

var _default = createOnEditable;
exports.default = _default;
