// @ali/4ever-bi/dist-web/plugins/formatPainter/components/next/FormatPainterButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { BrushButton, useDefaultSz } from '@ali/4ever-component';
import { PluginRoles } from '@ali/4ever-bamboo';
;
;
;
import { copyFormat, clearFormatClipboard } from "../../actions";
export default function FormatPainterButton(props) {
  var controller = props.controller,
      locale = props.locale,
      tooltipProp = props.tooltip,
      rest = _objectWithoutPropertiesLoose(props, ["controller", "locale", "tooltip"]);

  var value = controller.value;
  var defaultSz = useDefaultSz();
  var formatClipboard = value.data.formatClipboard;
  var isActive = formatClipboard && formatClipboard.trigger === 'toolbar';
  var isCollapsedAtEdge = value.selection.isCollapsedAtEdge;
  var isDisabled = isCollapsedAtEdge || value.data.edgeSelection || controller.query('isSelectionInHr') || controller.query('isSelectionInImage') || controller.query('isSelectionInMention') || controller.query('isSelectionInListSymbol') || controller.query('isSelectionInLinkCard');
  var handleClick = React.useCallback(function () {
    if (!formatClipboard) {
      controller.run('onAction', copyFormat('single', defaultSz));
      return;
    }

    controller.run('onAction', clearFormatClipboard());
  }, [controller, defaultSz, formatClipboard]);
  var handleDoubleClick = React.useCallback(function () {
    controller.run('onAction', copyFormat('multiple', defaultSz));
  }, [controller, defaultSz]);
  var tooltip = React.useMemo(function () {
    return _extends({
      title: locale == null ? void 0 : locale.buttonTitle,
      description: (locale == null ? void 0 : locale.doubleTitle) || ''
    }, tooltipProp);
  }, [tooltipProp]);
  return /*#__PURE__*/_createElement(BrushButton, _extends({
    testid: "bi-toolbar-painter",
    active: isActive,
    disabled: isDisabled,
    onClick: handleClick,
    onDoubleClick: handleDoubleClick,
    role: PluginRoles.formatPainter,
    title: locale == null ? void 0 : locale.buttonTitle,
    tooltip: tooltip
  }, rest));
}
