// @ali/4ever-plugin-list/dist-node/mo/serializer/htmlToJsonML/parseFromMsWord.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everUtils = require("@ali/4ever-utils");

var _everCangjie = require("@ali/4ever-cangjie");

var _msWordParseSymbol = _interopRequireDefault(require("./utils/msWordParseSymbol"));

var _utils = require("./utils");

const MSO_LIST = 'msoList';

function parseFromMsWord() {
  let listData;
  const msoListMap = {};
  const temp = {
    // 是否在列表中
    wordList: false,
    // 是否是开始位置
    fromStart: true,
    // 上一个新增text内容
    preTxt: '',
    preStyle: '',
    // 上一个列表level
    lastLevel: 0,
    isOrdered: null
  };
  return {
    name: 'list',
    onOpenTag: (state, name, attr) => {
      let inlineStyle;

      if (name === 'ol' || name === 'ul') {
        temp.isOrdered = name === 'ol';
      }

      if ((0, _utils.isMsWordList)(temp.preStyle) && attr.style?.includes('mso')) {
        inlineStyle = (0, _everCangjie.toStyleObject)(attr.style);

        if (inlineStyle.fontSize) {
          listData.list.symbolStyle = { ...listData.list.symbolStyle,
            sz: parseInt(inlineStyle.fontSize)
          };
        }

        if (inlineStyle.color) {
          listData.list.symbolStyle = { ...listData.list.symbolStyle,
            color: inlineStyle.color
          };
        }

        temp.preStyle = '';
      }

      if (!(0, _utils.isMsWordList)(attr?.style)) {
        return false;
      }

      temp.preStyle = attr.style;
      inlineStyle = inlineStyle || (0, _everCangjie.toStyleObject)(attr.style);

      if (!inlineStyle[MSO_LIST]) {
        return false;
      }

      state.wrapList('ordered-list');
      temp.wordList = true;
      const mso = inlineStyle[MSO_LIST].split(' ');
      const level = Math.max(parseInt(mso[1].slice(5)) - 1, 0);
      const listId = mso[2];
      const type = name.includes('h') ? name : 'p';
      let indLeft = (0, _utils.transfromIndLeft)(inlineStyle.marginLeft || '', level);

      if (inlineStyle.textIndent) {
        const textIndent = _everUtils.unitsConverter.toPX(inlineStyle.textIndent);

        if (textIndent > 0) {
          indLeft += textIndent;
        }
      }

      state.currentListId = listId;
      state.currentListLevel = level;

      if (!msoListMap[listId]) {
        msoListMap[listId] = {};
      }

      if (msoListMap[listId][level] === undefined || level > temp.lastLevel) {
        msoListMap[listId][level] = 1;
      } else {
        msoListMap[listId][level]++;
      }

      listData = {
        list: {
          listId,
          isOrdered: state.isParentOrderedList(),
          level
        },
        ind: {
          left: indLeft
        },
        jc: inlineStyle.textAlign || 'left'
      };

      if (name.includes('h')) {
        listData.list.symbolStyle = {
          bold: true
        };
      }

      const li = [type, listData];
      state.push(li);
      return true;
    },
    onText: (state, txt) => {
      if (!temp.wordList) return false;

      if (temp.fromStart && txt.trim() !== '') {
        // 通过解析text内容，获取列表类型，更新列表信息
        temp.fromStart = false;
        const listId = state.currentListId;
        const level = state.currentListLevel;
        if (!listId || level === undefined || level === null) return false;
        const [format, text, type] = (0, _msWordParseSymbol.default)(txt, msoListMap[listId][level], level, temp.isOrdered);

        if (format !== 'bullet') {
          state.unwrapList();
          state.wrapList('ordered-list');
        }

        listData.list.isOrdered = format !== 'bullet';
        listData.list.listStyleType = type;
        listData.list.listStyle = { ...listData.list.listStyle,
          format,
          text
        };
      }

      if (temp.preTxt.trim() !== '' && !(!listData.list.isOrdered && temp.preTxt.length !== 1)) {
        // 删除项目符号相关text，仅保留列表内容
        (0, _utils.removeSymbolText)(state, temp.preTxt);
      }

      temp.preTxt += txt;
      return false;
    },
    onCloseTag: (state, name, attr) => {
      if (name === 'ol' || name === 'ul') {
        temp.isOrdered = null;
      }

      if (!(0, _utils.isMsWordList)(attr?.style)) {
        return false;
      }

      temp.wordList = false;
      temp.fromStart = true;
      state.unwrapList();
      const li = state.pop();

      if ((0, _utils.shouldFlattenChildren)(li)) {
        const flattenedChildren = (0, _utils.getFlattenedChildren)(li);
        state.append(...flattenedChildren);
      } else if ((0, _everUtils.getChildren)(li).length) {
        state.append((0, _everUtils.normalizeJsonMLChildren)(li));
      }

      temp.preTxt = '';
      return true;
    }
  };
}

var _default = parseFromMsWord;
exports.default = _default;
