// @ali/4ever-plugin-embed/dist-web/utils/components/audio/audioPreview.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import EmbedPreviewHeader from "../common/embedPreviewHeader";
import EmbedDownloadButton from "../common/embedDownloadButton";
import AudioPlayer from "./audioPlayer";
import { FileTypeMap } from "../../fileTypeHelper";
import { AudioWrapper } from "./styled";
export default function AudioPreview(props) {
  var onMediaLoadSuccess = props.onMediaLoadSuccess,
      onMediaLoadError = props.onMediaLoadError,
      transformAudioURL = props.transformAudioURL,
      locale = props.locale,
      allowDownload = props.allowDownload,
      name = props.name,
      size = props.size,
      src = props.src,
      onDownloadFile = props.onDownloadFile,
      isMobile = props.isMobile;
  return /*#__PURE__*/_createElement(AudioWrapper, {
    "data-type": "embed-audio"
  }, /*#__PURE__*/_createElement(EmbedPreviewHeader, {
    type: FileTypeMap.Audio,
    name: name,
    size: size,
    renderExtra: function renderExtra() {
      return /*#__PURE__*/_createElement(EmbedDownloadButton, {
        title: locale == null ? void 0 : locale.title,
        onDownloadFile: onDownloadFile,
        allowDownload: allowDownload
      });
    },
    isMobile: isMobile
  }), /*#__PURE__*/_createElement(AudioPlayer, {
    transformAudioURL: transformAudioURL,
    onMediaLoadSuccess: onMediaLoadSuccess,
    onMediaLoadError: onMediaLoadError,
    src: src,
    isMobile: isMobile
  }));
}
