// @ali/4ever-zhi/dist-node/plugins/headingGroup/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everBamboo = require("@ali/4ever-bamboo");

var _everFactory = require("@ali/4ever-factory");

var _createGroupNodes = _interopRequireDefault(require("./handlers/createGroupNodes"));

var _createRenderNode = _interopRequireDefault(require("./handlers/createRenderNode"));

const {
  createRenderMark
} = _everBamboo.GroupFactory;
const HeadingGroupPlugin = (0, _everFactory.zhiGroupFactory)(config => {
  const {
    headingGroup = {}
  } = config;
  const {
    level = 3
  } = headingGroup;
  return {
    groupType: 'headingGroup',
    groupNodes: (0, _createGroupNodes.default)(headingGroup),
    renderNode: (0, _createRenderNode.default)(headingGroup),
    renderMark: createRenderMark({
      autoUnfold: headingGroup.autoUnfold,
      inject: true
    }),
    queries: {
      getHeadingGroupLevel: () => level,
      isHeadingGroupAvailable: () => true,
      ..._everBamboo.HeadingGroupPlugin.queries
    }
  };
}); // @ts-ignore

HeadingGroupPlugin.actions = _everBamboo.HeadingGroupPlugin.actions; // @ts-ignore

HeadingGroupPlugin.queries = _everBamboo.HeadingGroupPlugin.queries;
var _default = HeadingGroupPlugin;
exports.default = _default;
