// @ali/4ever-bi/dist-node/plugins/heading/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _headingWrapper = _interopRequireDefault(require("./components/headingWrapper"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _onCangjieInput = _interopRequireDefault(require("./handlers/onCangjieInput"));

var _createOnCangjiePaste = _interopRequireDefault(require("./handlers/createOnCangjiePaste"));

var actions = _interopRequireWildcard(require("./actions"));

var _placeholder = _interopRequireDefault(require("./placeholder"));

var _getHeadingModels = _interopRequireDefault(require("./utils/getHeadingModels"));

var _createRenderNode = require("./createRenderNode");

var _queries = require("./queries");

var _commands = require("./commands");

var _everPluginHeading = require("@ali/4ever-plugin-heading");

var _createRenderToolbar = _interopRequireDefault(require("./createRenderToolbar"));

const _createElement = /*#__PURE__*/React.createElement;

function createHeadingPlugin(config) {
  const headingConfig = config.heading || {};
  return { ...(0, _everPluginHeading.moFactory)({}),

    // @ts-ignore
    renderEditable(props, controller, next) {
      if (headingConfig.onHeadingsChange) {
        return /*#__PURE__*/_createElement(_headingWrapper.default, (0, _extends2.default)({}, props, {
          controller: controller,
          onHeadingsChange: headingConfig.onHeadingsChange
        }), next());
      }

      return next();
    },

    renderNode: (0, _createRenderNode.createRenderNode)(),
    toolbar: (0, _toolbar.default)(headingConfig),
    renderToolbar: (0, _createRenderToolbar.default)(headingConfig.locale || {}),
    menubar: (0, _menubar.default)(),
    actions,
    commands: _commands.commands,
    queries: _queries.queries,
    onAction: _onAction.default,
    onCangjiePaste: (0, _createOnCangjiePaste.default)(config),
    onKeyDown: _onKeyDown.default,
    onCangjieInput: _onCangjieInput.default,
    models: (0, _getHeadingModels.default)(),
    placeholder: (0, _placeholder.default)(headingConfig)
  };
}

createHeadingPlugin.actions = actions;
createHeadingPlugin.commands = _commands.commands;
createHeadingPlugin.queries = _queries.queries;
var _default = createHeadingPlugin;
exports.default = _default;
