// @ali/we-ui/dist-web/components/color-picker/sketch/common/Checkboard.js
import _extends from "@babel/runtime/helpers/extends";
import React, { isValidElement } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import reactCSS from 'reactcss';
import * as checkboardHelper from "../helpers/checkboard";
export var Checkboard = function Checkboard(_ref) {
  var white = _ref.white,
      grey = _ref.grey,
      size = _ref.size,
      renderers = _ref.renderers,
      borderRadius = _ref.borderRadius,
      boxShadow = _ref.boxShadow,
      children = _ref.children;
  var styles = reactCSS({
    "default": {
      grid: {
        borderRadius: borderRadius,
        boxShadow: boxShadow,
        absolute: '0px 0px 0px 0px',
        background: "url(" + checkboardHelper.get(white, grey, size, renderers.canvas) + ") center left"
      }
    }
  });
  return /*#__PURE__*/isValidElement(children) ? /*#__PURE__*/React.cloneElement(children, _extends({}, children.props, {
    style: _extends({}, children.props.style, styles.grid)
  })) : /*#__PURE__*/_createElement("div", {
    style: styles.grid
  });
};
Checkboard.defaultProps = {
  size: 8,
  white: 'transparent',
  grey: 'rgba(0,0,0,.08)',
  renderers: {}
};
export default Checkboard;
