// @ali/4ever-component/dist-node/MobileToolbar/panels/layoutContent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _buttons = _interopRequireDefault(require("../buttons"));

var _constant = require("../constant");

var _styled = require("./styled");

/* eslint-disable @typescript-eslint/consistent-type-definitions */
const _createElement = /*#__PURE__*/React.createElement;

function Group({
  items,
  controller,
  plugins,
  borderless
}) {
  return /*#__PURE__*/_createElement(_styled.GroupWrapper, {
    borderless: borderless
  }, items.map((item, index) => {
    const isStart = index === 0 || items[index - 1] === _constant.DIVIDER || items[index - 1] === _constant.DIVIDER_NONE || items[index - 1] === _constant.NEW_LINE;
    const isEnd = index === items.length - 1 || items[index + 1] === _constant.DIVIDER || items[index + 1] === _constant.DIVIDER_NONE || items[index + 1] === _constant.NEW_LINE;
    return /*#__PURE__*/_createElement(_buttons.default, {
      key: typeof item === 'string' ? `${item}_${index}` : item.name,
      item: item,
      isStart: isStart,
      isEnd: isEnd,
      controller: controller,
      plugins: plugins
    });
  }));
}

var _ref = /*#__PURE__*/_createElement(_styled.SectionDivider, null);

function LayoutPanel({
  controller,
  plugins,
  layout
}) {
  return /*#__PURE__*/_createElement(_styled.ContentWrapper, {
    "data-toolbar-panel-layout": true
  }, layout.map((group, index) => {
    return /*#__PURE__*/_createElement(React.Fragment, {
      key: group.key
    }, index !== 0 && _ref, /*#__PURE__*/_createElement(Group, {
      items: group.items,
      borderless: group.borderless,
      controller: controller,
      plugins: plugins
    }));
  }));
}

var _default = LayoutPanel;
exports.default = _default;
