// @ali/4ever-component/dist-node/MobileToolbar/panels/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IconWrapper = exports.HeaderText = exports.HeaderWrapper = exports.GroupWrapper = exports.SectionDivider = exports.ContentWrapper = exports.CloseButtonWrapper = exports.PanelContentWrapper = exports.BottomPanelWrapper = exports.HEADER_BG = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constant = require("../constant");

/* eslint-disable no-confusing-arrow */
const PANEL_TRANSITION = 'opacity 250ms, transform 250ms';
const HEADER_BG = '#f8f8f8';
exports.HEADER_BG = HEADER_BG;

const BottomPanelWrapper = /*#__PURE__*/_styledComponents.default.div(["position:fixed;z-index:", ";font-size:14px;transition:", ";left:0;bottom:0;width:100%;height:", ";background:white;border-radius:16px 16px 0 0;box-shadow:rgb(0 0 0 / 12%) 0px -16px 16px -16px,rgb(0 0 0 / 32%) 0px -1px 1px -1px;padding-bottom:0;padding-bottom:constant(safe-area-inset-bottom);padding-bottom:env(safe-area-inset-bottom);-webkit-tap-highlight-color:rgba(0,0,0,0);@media only screen and (min-width:580px){width:460px;left:calc(50vw - 230px);border-radius:16px 16px 0 0;overflow:hidden;box-shadow:0 0 16px 0 rgba(0,0,0,0.08);}"], _constant.Z_INDEX, PANEL_TRANSITION, props => props.height || `${_constant.BOTOTM_PANEL_HEIGHT}px`);

exports.BottomPanelWrapper = BottomPanelWrapper;

const PanelContentWrapper = /*#__PURE__*/_styledComponents.default.div(["width:100%;height:100%;overflow-y:scroll;-webkit-overflow-scrolling:touch;-ms-overflow-style:none;scrollbar-width:none;&::-webkit-scrollbar{display:none;}"]);

exports.PanelContentWrapper = PanelContentWrapper;

const CloseButtonWrapper = /*#__PURE__*/_styledComponents.default.div(["width:", "px;height:", "px;position:absolute;top:0;left:0;display:flex;align-items:center;border-radius:16px;justify-content:center;", ""], _constant.PANEL_HEADER_SIZE, _constant.PANEL_HEADER_SIZE, props => props.isTouched ? '& > img { opacity: 0.8 }' : '');

exports.CloseButtonWrapper = CloseButtonWrapper;

const ContentWrapper = /*#__PURE__*/_styledComponents.default.div(["width:100%;margin-top:", "px;flex:1;min-height:250px;background:white;overflow-y:scroll;"], _constant.BOTTOM_BAR_SIZE);

exports.ContentWrapper = ContentWrapper;

const SectionDivider = /*#__PURE__*/_styledComponents.default.div(["height:8px;background:#f4f4f7;"]);

exports.SectionDivider = SectionDivider;

const GroupWrapper = /*#__PURE__*/_styledComponents.default.div(["", " display:flex;flex-direction:row;align-items:center;justify-content:space-between;flex-wrap:wrap;"], props => props.borderless ? '' : `
  margin: 0 16px;
  padding-top: 16px;
`);

exports.GroupWrapper = GroupWrapper;

const HeaderWrapper = /*#__PURE__*/_styledComponents.default.div(["height:", "px;display:flex;transition:background 250ms;border-bottom:0.5px solid rgba(126,134,142,0.16);flex-direction:row;align-items:center;justify-content:center;"], _constant.PANEL_HEADER_SIZE);

exports.HeaderWrapper = HeaderWrapper;

const HeaderText = /*#__PURE__*/_styledComponents.default.div(["margin-left:", ";font-size:15px;color:#171A1D;font-weight:600;"], props => props.hasIcon ? '6px' : '0');

exports.HeaderText = HeaderText;

const IconWrapper = /*#__PURE__*/_styledComponents.default.div(["margin-left:16px;line-height:0;"]);

exports.IconWrapper = IconWrapper;
