// @ali/4ever-plugin-sticker/dist-web/index.js
import { default as StickerManager } from "./utils/sticker";
import StickerItem from "./utils/stickerItem";
import { DEFAULT_SZ, DEFAULT_STICKERS, MOVE_SET, STICKER_NODE_TYPE, STICKER_PANEL_COLUMMS, STICKER_PANEL_HEIGHT, STICKER_PANEL_STYLE, STICKER_PLACEHOLDER_DATA, STICKER_SZ_IN_PANEL, DATA_STICKER, DEFAULT_LANG } from "./utils/constants";
import { StickerRole } from "./utils/role";
export { ELEMENT_TYPE as StickerType } from "./utils/meta";
export { getEmojiData } from "./utils/getEmojiData"; // facotry

export { default as moFactory } from "./mo/index"; // Element Model

export { default as Sticker } from "./mo/models"; // serializer

export { default as createSerializationPlugin } from "./mo/serializer/index";
export { createServerSerializationPlugin } from "./mo/serializer/index"; // other export for api compatiable

export { default as StickerPanel } from "./bi/components/stickerPanel";
export { StickerPannelWrapper, StickerSplitLine, StickerPanelItem, InlinePlaceholderWrapper, InlinePlaceholderContainer, PanelWrapper } from "./bi/components/styled";
export { StickerManager, DEFAULT_SZ, DEFAULT_STICKERS, MOVE_SET, STICKER_NODE_TYPE, STICKER_PANEL_COLUMMS, STICKER_PANEL_HEIGHT, STICKER_PANEL_STYLE, STICKER_PLACEHOLDER_DATA, STICKER_SZ_IN_PANEL, DATA_STICKER, DEFAULT_LANG, StickerItem, StickerRole };
import * as OldBambooPlugin from "./utils/bamboo";
export {
/** @deprecated */
OldBambooPlugin };
