// @ali/4ever-plugin-refblock/dist-node/zhi/createRenderNode.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderNode;

var React = _interopRequireWildcard(require("react"));

var _lodash = require("lodash");

var _models = _interopRequireDefault(require("../mo/models"));

var _RefBlock = _interopRequireDefault(require("../components/RefBlock"));

const _createElement = /*#__PURE__*/React.createElement;

const detach = () => Promise.reject("forbidden detach in readonly mode");

function createRenderNode(config) {
  const {
    locale,
    toolbarAnimationDuration,
    enableToolbar,
    getReferenceList,
    getDocKey,
    renderRefBlockInOtherStates,
    renderLoading,
    onRefBlockMounted,
    getDocURL
  } = config;
  return {
    [_models.default.TYPE]: (props, controller) => {
      const {
        children
      } = props;
      return /*#__PURE__*/_createElement(_RefBlock.default, {
        isFocused: false,
        locale: locale,
        enableToolbar: enableToolbar,
        readOnly: true,
        renderProps: props,
        toolbarAnimationDuration: toolbarAnimationDuration,
        getReferenceList: getReferenceList,
        getDocKey: getDocKey,
        detach: detach,
        onRefBlockMounted: onRefBlockMounted,
        renderRefBlockInOtherStates: renderRefBlockInOtherStates,
        copyRefBlock: _lodash.noop,
        renderLoading: renderLoading,
        getURL: getDocURL
      }, children());
    }
  };
}
