// @ali/4ever-plugin-marks/dist-node/utils/createMarkFactory/components/createMarkMenu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createMarkMenu;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _lodash = require("lodash");

var _weDesign = require("@ali/we-design");

var _canToggleMark = _interopRequireDefault(require("../queries/canToggleMark"));

var _actions = require("../actions");

const _createElement = /*#__PURE__*/React.createElement;

function createMarkMenu(type) {
  return function MarkButton(props) {
    const {
      menuItem,
      controller
    } = props;
    const {
      title,
      ...rest
    } = menuItem;
    const isDisabled = !(0, _canToggleMark.default)(controller);
    const isActive = controller.query(`is${(0, _lodash.capitalize)(type)}`);

    const handleClick = () => controller.run('onAction', (0, _actions.toggleMark)(type));

    return /*#__PURE__*/_createElement(_weDesign.MenuItem, (0, _extends2.default)({
      onClick: handleClick,
      selected: isActive,
      disabled: isDisabled,
      role: type
    }, rest), title);
  };
}
