// @ali/4ever-utils/dist-node/utils/Cache.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

/**
 * 一般 K/V Cache 类型
 */

/**
 * 简单 K/V cache 缓存
 * @param capacity 容量 - (真实的缓存容量会大 1 位)
 */
class SimpleCache {
  /**
   * 缓存池容量
   */

  /**
   * 当前缓存池的大小
   */

  /**
   * 备份缓存池
   */

  /**
   * 当前缓存池
   */
  constructor(capacity = Infinity) {
    this.capacity = void 0;
    this.size = 0;
    this.backup = Object.create(null);
    this.cache = Object.create(null);
    this.capacity = capacity;
  }

  update(key, value) {
    this.size += 1;
    this.cache[key] = value;

    if (this.size >= this.capacity) {
      this.size = 0;
      this.backup = this.cache;
      this.cache = Object.create(null);
    }
  }

  get(key) {
    let value = this.cache[key];

    if (value !== undefined) {
      return value;
    }

    value = this.backup[key];

    if (value !== undefined) {
      this.update(key, value);
      return value;
    }

    return null;
  }

  set(key, value) {
    if (this.cache[key] !== undefined) {
      this.cache[key] = value;
    } else {
      this.update(key, value);
    }
  }

  has(key) {
    return key in this.cache || key in this.backup;
  }

  remove(key) {
    let value = this.cache[key];

    if (value !== undefined) {
      this.size -= 1;
      delete this.cache[key];
    } else {
      value = this.backup[key];
    }

    delete this.backup[key];
    return value || null;
  }

  clear() {
    this.size = 0;
    this.cache = Object.create(null);
    this.backup = Object.create(null);
  }

}

exports.default = SimpleCache;
