// @ali/4ever-utils/dist-node/utils/mime-types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.APPLICATION_VSCODE = exports.APPLICATION_CANGJIE_DRAG_FRAGMENT = exports.APPLICATION_GOOGLE_SHEET_CLIP = exports.APPLICATION_GOOGLE_DOCS_CLIP = exports.APPLICATION_DING_SHEET = exports.APPLICATION_SLATE_FRAGMENT = exports.APPLICATION_CANGJIE_FRAGMENT = exports.TEXT_LINK = exports.TEXT_RTF = exports.TEXT_HTML = exports.TEXT_PLAIN = void 0;
const TEXT_PLAIN = 'text/plain';
exports.TEXT_PLAIN = TEXT_PLAIN;
const TEXT_HTML = 'text/html';
exports.TEXT_HTML = TEXT_HTML;
const TEXT_RTF = 'text/rtf';
exports.TEXT_RTF = TEXT_RTF;
const TEXT_LINK = 'text/uri-list';
/**
 * 仓颉编辑器
 */

exports.TEXT_LINK = TEXT_LINK;
const APPLICATION_CANGJIE_FRAGMENT = 'application/x-cangjie-fragment';
/**
 * Slate 的 fragment
 */

exports.APPLICATION_CANGJIE_FRAGMENT = APPLICATION_CANGJIE_FRAGMENT;
const APPLICATION_SLATE_FRAGMENT = 'application/x-slate-fragment';
/**
 * 钉钉表格
 */

exports.APPLICATION_SLATE_FRAGMENT = APPLICATION_SLATE_FRAGMENT;
const APPLICATION_DING_SHEET = 'application/ding-sheets-clip+json';
/**
 * Google Docs
 */

exports.APPLICATION_DING_SHEET = APPLICATION_DING_SHEET;
const APPLICATION_GOOGLE_DOCS_CLIP = 'application/x-vnd.google-docs-document-slice-clip+wrapped';
/**
 * Google Sheet: 这个 MIME type 只针对于 G-Sheet 中的 range 拷贝，如果是整个 Sheet 拷贝则没有
 */

exports.APPLICATION_GOOGLE_DOCS_CLIP = APPLICATION_GOOGLE_DOCS_CLIP;
const APPLICATION_GOOGLE_SHEET_CLIP = 'application/x-vnd.google-docs-embedded-grid_range_clip+wrapped';
/**
 * 仓颉编辑器拖拽数据
 */

exports.APPLICATION_GOOGLE_SHEET_CLIP = APPLICATION_GOOGLE_SHEET_CLIP;
const APPLICATION_CANGJIE_DRAG_FRAGMENT = 'application/x-cangjie-drag-fragment';
/**
 * VS Code
 */

exports.APPLICATION_CANGJIE_DRAG_FRAGMENT = APPLICATION_CANGJIE_DRAG_FRAGMENT;
const APPLICATION_VSCODE = 'vscode-editor-data';
exports.APPLICATION_VSCODE = APPLICATION_VSCODE;
