// @ali/4ever-plugin-image/dist-node/utils/bamboo.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  base64ToArrayBuffer: true,
  calcComplexImage: true,
  isParentClickable: true,
  getBorder: true,
  Image: true,
  Img: true,
  Loading: true,
  InlinePlaceholder: true,
  MobileImage: true,
  updateInlineImage: true,
  Uploading: true,
  ImageManager: true
};
Object.defineProperty(exports, "base64ToArrayBuffer", {
  enumerable: true,
  get: function () {
    return _base64ToArrayBuffer.default;
  }
});
Object.defineProperty(exports, "calcComplexImage", {
  enumerable: true,
  get: function () {
    return _calcComplexImage.default;
  }
});
Object.defineProperty(exports, "isParentClickable", {
  enumerable: true,
  get: function () {
    return _isParentClickable.default;
  }
});
Object.defineProperty(exports, "getBorder", {
  enumerable: true,
  get: function () {
    return _getBorder.default;
  }
});
Object.defineProperty(exports, "Image", {
  enumerable: true,
  get: function () {
    return _image.default;
  }
});
Object.defineProperty(exports, "Img", {
  enumerable: true,
  get: function () {
    return _img.default;
  }
});
Object.defineProperty(exports, "Loading", {
  enumerable: true,
  get: function () {
    return _loading.default;
  }
});
Object.defineProperty(exports, "InlinePlaceholder", {
  enumerable: true,
  get: function () {
    return _inlinePlaceholder.default;
  }
});
Object.defineProperty(exports, "MobileImage", {
  enumerable: true,
  get: function () {
    return _mobile.default;
  }
});
Object.defineProperty(exports, "updateInlineImage", {
  enumerable: true,
  get: function () {
    return _updateInlineImage.default;
  }
});
Object.defineProperty(exports, "Uploading", {
  enumerable: true,
  get: function () {
    return _uploading.default;
  }
});
Object.defineProperty(exports, "ImageManager", {
  enumerable: true,
  get: function () {
    return _imageManager.default;
  }
});

var _base64ToArrayBuffer = _interopRequireDefault(require("./utils/base64ToArrayBuffer"));

var _calcComplexImage = _interopRequireDefault(require("./utils/calcComplexImage"));

var _isParentClickable = _interopRequireDefault(require("./utils/isParentClickable"));

var _getBorder = _interopRequireWildcard(require("./utils/getBorder"));

Object.keys(_getBorder).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getBorder[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getBorder[key];
    }
  });
});

var _image = _interopRequireDefault(require("./components/image"));

var _img = _interopRequireDefault(require("./img"));

var _loading = _interopRequireDefault(require("./components/loading"));

var _inlinePlaceholder = _interopRequireDefault(require("./components/inlinePlaceholder"));

var _mobile = _interopRequireDefault(require("./mobile"));

var _updateInlineImage = _interopRequireDefault(require("./commands/updateInlineImage"));

var _uploading = _interopRequireDefault(require("./components/uploading"));

var _imageManager = _interopRequireDefault(require("./imageManager"));

var _constants = require("./constants");

Object.keys(_constants).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _constants[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _constants[key];
    }
  });
});

var _aslService = require("./utils/aslService");

Object.keys(_aslService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _aslService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _aslService[key];
    }
  });
});

var _urlService = require("./utils/urlService");

Object.keys(_urlService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _urlService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _urlService[key];
    }
  });
});

var _isNodeInTable = require("./utils/isNodeInTable");

Object.keys(_isNodeInTable).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _isNodeInTable[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _isNodeInTable[key];
    }
  });
});

var _isNodeInTask = require("./utils/isNodeInTask");

Object.keys(_isNodeInTask).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _isNodeInTask[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _isNodeInTask[key];
    }
  });
});

var _getImageMaxSize = require("./utils/getImageMaxSize");

Object.keys(_getImageMaxSize).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getImageMaxSize[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getImageMaxSize[key];
    }
  });
});

var _getImageMaxWidth = require("./utils/getImageMaxWidth");

Object.keys(_getImageMaxWidth).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getImageMaxWidth[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getImageMaxWidth[key];
    }
  });
});

var _getZoomImageData = require("./utils/getZoomImageData");

Object.keys(_getZoomImageData).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getZoomImageData[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getZoomImageData[key];
    }
  });
});

var _imageOrientation = require("./utils/imageOrientation");

Object.keys(_imageOrientation).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _imageOrientation[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _imageOrientation[key];
    }
  });
});

var _getImageType = require("./utils/getImageType");

Object.keys(_getImageType).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getImageType[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getImageType[key];
    }
  });
});

var _styled = require("./components/styled");

Object.keys(_styled).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _styled[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _styled[key];
    }
  });
});

var _types = require("./components/types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});

var _types2 = require("./types");

Object.keys(_types2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types2[key];
    }
  });
});

var _role = require("./role");

Object.keys(_role).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _role[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _role[key];
    }
  });
});

var _hooks = require("./hooks");

Object.keys(_hooks).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _hooks[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _hooks[key];
    }
  });
});
