// @ali/4ever-plugin-image/dist-node/utils/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HIDE_RELOAD_TIP_MAX_WIDTH = exports.DEFAULT_STYLE = exports.IMAGE_HOVER_BORDER = exports.IMAGE_EDIT_BORDER = exports.IMAGE_DATA_WRAPPER_CLASS_NAME = exports.IMAGE_BOX_SHADOW = exports.HOVER_INNER_LINE_WIDTH = exports.HOVER_LINE_WIDTH = exports.DEFAULT_RADIUS = exports.DEFAULT_HEIGHT = exports.DEFAULT_WIDTH = exports.PAGE_WIDTH = void 0;
const PAGE_WIDTH = 648;
exports.PAGE_WIDTH = PAGE_WIDTH;
const DEFAULT_WIDTH = 300;
exports.DEFAULT_WIDTH = DEFAULT_WIDTH;
const DEFAULT_HEIGHT = 300;
exports.DEFAULT_HEIGHT = DEFAULT_HEIGHT;
const DEFAULT_RADIUS = 0; // 图片高亮边框的宽度

exports.DEFAULT_RADIUS = DEFAULT_RADIUS;
const HOVER_LINE_WIDTH = 2; // 图片选中时的高亮边框

exports.HOVER_LINE_WIDTH = HOVER_LINE_WIDTH;
const HOVER_INNER_LINE_WIDTH = 1;
exports.HOVER_INNER_LINE_WIDTH = HOVER_INNER_LINE_WIDTH;
const IMAGE_BOX_SHADOW = 'rgb(0 0 0 / 20%) 1px 1px 10px 0px'; // 标记用于图片落库宽高的元素

exports.IMAGE_BOX_SHADOW = IMAGE_BOX_SHADOW;
const IMAGE_DATA_WRAPPER_CLASS_NAME = 'image-data-wrapper'; // 选中状态框

exports.IMAGE_DATA_WRAPPER_CLASS_NAME = IMAGE_DATA_WRAPPER_CLASS_NAME;
const IMAGE_EDIT_BORDER = 'editor-image-edit-border'; // hover 时的边框，在部分浏览器中支持圆角

exports.IMAGE_EDIT_BORDER = IMAGE_EDIT_BORDER;
const IMAGE_HOVER_BORDER = 'editor-image-hover';
exports.IMAGE_HOVER_BORDER = IMAGE_HOVER_BORDER;
const DEFAULT_STYLE = {
  outline: {},
  radius: DEFAULT_RADIUS,
  shadow: ''
};
exports.DEFAULT_STYLE = DEFAULT_STYLE;
const HIDE_RELOAD_TIP_MAX_WIDTH = 120;
exports.HIDE_RELOAD_TIP_MAX_WIDTH = HIDE_RELOAD_TIP_MAX_WIDTH;
