// @ali/4ever-plugin-image/dist-node/utils/img.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _image = _interopRequireDefault(require("./components/image"));

const _createElement = /*#__PURE__*/React.createElement;

var _default = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    node,
    controller,
    imageConfig,
    isTransferring,
    uploadPlaceholder,
    uploadProgress,
    onLoad,
    ...rest
  } = props;
  return /*#__PURE__*/_createElement(_image.default, (0, _extends2.default)({}, rest, {
    ref: ref,
    node: node,
    controller: controller,
    imageConfig: imageConfig,
    isTransferring: isTransferring,
    uploadPlaceholder: uploadPlaceholder,
    uploadProgress: uploadProgress,
    onLoad: onLoad
  }));
});

exports.default = _default;
