// @ali/we-toolbar/dist-web/common/cardLayout/placeholder-old/styled.js
import styled, { css, withTheme } from 'styled-components';
var light = {
  level3_base_color: 'rgba(23, 26, 29, 0.4)',
  level1_base_color: '#171A1D',
  level2_base_color: 'rgba(23, 26, 29, 0.6)',
  hover_bg_color: 'rgba(126, 134, 142, 0.12)',
  border_normal: 'rgba(0, 0, 0, 0.16)',
  line_light_color: 'rgba(126, 134, 142, 0.16)',
  bg_doc_color: 'rgba(245,246,248,1)'
};
export var PlaceholderWrapper = /*#__PURE__*/withTheme(styled.div(["width:100%;height:", ";display:flex;border:1px dashed ", ";color:", ";box-sizing:border-box;border-radius:8px;cursor:default;align-items:center;", ""], function (_ref) {
  var height = _ref.height;
  return height + "px";
}, light.border_normal, light.level3_base_color, function (_ref2) {
  var clickable = _ref2.clickable,
      isSelected = _ref2.isSelected,
      visible = _ref2.visible;
  return clickable ? isSelected ? css(["border-color:transparent;;"]) : visible ? css(["background-color:", ";border-color:transparent;.placeholder-title{color:", ";}.placeholder-icon{color:", ";}"], light.hover_bg_color, light.level1_base_color, light.level1_base_color) : css(["&:hover{border-color:", ";}&:hover .placeholder-title{color:", ";}&:hover .placeholder-icon{color:", ";}&:active{background-color:", ";border-color:transparent;}&:active .placeholder-title{color:", ";}&:active .placeholder-icon{color:", ";}background-color:", ";"], light.level3_base_color, light.level1_base_color, light.level1_base_color, light.line_light_color, light.level1_base_color, light.level1_base_color, isSelected ? light.hover_bg_color : 'transparent') : css([""]);
}));
export var IconContainer = /*#__PURE__*/withTheme(styled.div(["display:flex;width:40px;height:40px;background:", ";border-radius:4px;flex-shrink:0;align-items:center;justify-content:center;"], light.hover_bg_color));
export var ClickAreaContainer = /*#__PURE__*/withTheme(styled.div(["width:100%;margin:9px 20px 9px 11px;overflow:hidden;display:flex;box-sizing:border-box;flex-direction:row;"]));
export var RightContainer = /*#__PURE__*/withTheme(styled.div(["width:100%;padding-left:12px;display:flex;overflow:hidden;flex-direction:column;justify-content:", ";"], function (_ref3) {
  var isSingleLine = _ref3.isSingleLine;
  return isSingleLine ? 'center' : 'flex-start';
}));
export var TitleWrapper = /*#__PURE__*/withTheme(styled.div(["display:flex;align-items:center;"]));
export var Title = /*#__PURE__*/withTheme(styled.div(["overflow:hidden;white-space:nowrap;width:", ";text-overflow:ellipsis;font-weight:", ";", ";"], function (_ref4) {
  var isSingleLine = _ref4.isSingleLine,
      toolbarWidth = _ref4.toolbarWidth,
      toolbarVisible = _ref4.toolbarVisible;
  return !isSingleLine && toolbarWidth && toolbarVisible ? "calc(100% - " + toolbarWidth + "px)" : isSingleLine + " " + toolbarWidth;
}, function (_ref5) {
  var isSingleLine = _ref5.isSingleLine;
  return isSingleLine ? 'normal' : 500;
}, function (_ref6) {
  var type = _ref6.type;
  return type === 'export' ? css(["height:20px;font-size:14px;line-height:20px;"]) : css(["height:22px;font-size:16px;line-height:22px;"]);
}));
export var DesCription = /*#__PURE__*/withTheme(styled.div(["overflow:hidden;white-space:nowrap;text-overflow:ellipsis;font-weight:normal;font-size:12px;height:16px;line-height:16px;margin-top:2px;"]));
