// @ali/4ever-bi/dist-node/plugins/sz/components/SzMobileButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MobileSizeInc = MobileSizeInc;
exports.MobileSizeDec = MobileSizeDec;
exports.MobileSizeNow = MobileSizeNow;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _everComponent = require("@ali/4ever-component");

var _actions = require("../actions");

var _everPluginSz = require("@ali/4ever-plugin-sz");

const _createElement = /*#__PURE__*/React.createElement;

const SzWrapper = /*#__PURE__*/_styledComponents.default.div(["flex:0.8;margin-right:1px;margin-bottom:12px;display:flex;flex-direction:column;align-items:center;"]);

const Icon = /*#__PURE__*/_styledComponents.default.img(["width:20px;height:20px;"]);

const Line1 = /*#__PURE__*/_styledComponents.default.div(["font-size:12px;color:#171a1d;line-height:16px;"]);

const Line2 = /*#__PURE__*/_styledComponents.default.div(["font-size:17px;font-weight:500;color:#171a1d;line-height:20px;"]);

function MobileSizeInc(props) {
  const {
    controller,
    item,
    Wrapper
  } = props;
  const defaultSz = (0, _everComponent.useDefaultSz)();
  const sz = controller.query('getSize', defaultSz);
  const handleClick = React.useCallback(() => {
    const newSize = _everPluginSz.szList.find(s => s > sz) || _everPluginSz.szList[_everPluginSz.szList.length - 1];

    controller.run('onAction', (0, _actions.setSz)(newSize));
  }, [controller, sz]);
  return /*#__PURE__*/_createElement(Wrapper, {
    type: "horiz",
    item: item,
    onClick: handleClick,
    icon: "sizePlus"
  });
}

function MobileSizeDec(props) {
  const {
    controller,
    item,
    Wrapper
  } = props;
  const defaultSz = (0, _everComponent.useDefaultSz)();
  const sz = controller.query('getSize', defaultSz);
  const handleClick = React.useCallback(() => {
    const newSize = [..._everPluginSz.szList].reverse().find(s => s < sz) || _everPluginSz.szList[0];

    controller.run('onAction', (0, _actions.setSz)(newSize));
  }, [controller, sz]);
  return /*#__PURE__*/_createElement(Wrapper, {
    type: "horiz",
    item: item,
    onClick: handleClick,
    icon: "sizeMinus"
  });
}

function MobileSizeNow(props) {
  const {
    controller,
    locale
  } = props;
  const defaultSz = (0, _everComponent.useDefaultSz)();
  const sz = controller.query('getSize', defaultSz);
  return /*#__PURE__*/_createElement(SzWrapper, null, /*#__PURE__*/_createElement(Line1, null, locale?.titleMobile), /*#__PURE__*/_createElement(Line2, null, sz));
}
