// @ali/4ever-bi/dist-node/plugins/columns/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everBamboo = require("@ali/4ever-bamboo");

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _setColumnsItemData = _interopRequireDefault(require("./commands/setColumnsItemData"));

var _leftToolbarHandlers = require("./leftToolbarHandlers");

var _createRenderHoverToolbar = _interopRequireDefault(require("./createRenderHoverToolbar"));

const {
  hooks,
  createQueries
} = _everBamboo.ColumnsPlugin;

function createColumnsPlugin(allConfigs) {
  const {
    createOnAction,
    createOnCangjieSelect,
    createOnKeyDown,
    createPlaceholder,
    createOnCut,
    beforeCopySelectedNodes
  } = hooks;
  const configs = allConfigs.columns;
  return {
    // @ts-ignore
    toolbar: (0, _toolbar.default)(),
    commands: {
      setColumnsItemData: _setColumnsItemData.default
    },
    // @ts-ignore
    leftToolbarHandlers: _leftToolbarHandlers.leftToolbarHandlers,
    menubar: (0, _menubar.default)(configs),
    queries: createQueries(configs),
    renderNode: (0, _everBamboo.ColumnsRenderNodeEdit)(configs, _leftToolbarHandlers.handleDragChange),
    onAction: createOnAction(configs),
    onCangjieSelect: createOnCangjieSelect(configs),
    onKeyDown: createOnKeyDown(),
    onCut: createOnCut(),
    placeholder: createPlaceholder(configs),
    renderHoverToolbar: (0, _createRenderHoverToolbar.default)(configs),
    beforeCopySelectedNodes
  };
}

var _default = createColumnsPlugin;
exports.default = _default;
