// @ali/we-design/dist-web/components/menu/MenuItem.js
import _extends from "@babel/runtime/helpers/extends";

var _SelectedNormalNormal, _SelectedNormalNormal2;

import React, { useCallback, useContext, useEffect, useMemo } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { classnames } from "../../utils";
import { MenuItemWrap, MenuItemInner, MenuItemIconWrap, MenuItemContentWrap, MenuItemExtraWrap } from "./styled";
import { useController, useMenuKeyContext, useHotkeyController, useHotKeyScrollRef, useMousePosition } from "./hooks";
import { TextIndentContext, MenuKeyContext, PathContext, SubMenuContext } from "./contexts";
import { MENU_VISIBLE_TOGGLE_DELAY } from "./consts";
import { Tooltip } from "../tooltip";
import { SelectedNormalNormal } from '@ali/we-icon'; // eslint-disable-next-line max-lines-per-function

export var MenuItem = /*#__PURE__*/React.memo(function (props) {
  var children = props.children,
      disabled = props.disabled,
      disabledTip = props.disabledTip,
      tooltipZIndex = props.tooltipZIndex,
      style = props.style,
      selected = props.selected,
      onClick = props.onClick,
      hoverable = props.hoverable,
      height = props.height,
      testid = props.testid,
      icon = props.icon,
      extra = props.extra,
      description = props.description,
      showSelect = props.showSelect,
      _props$role = props.role,
      role = _props$role === void 0 ? '' : _props$role; // 接口

  var controller = useController('MenuItem'); // 快捷键触达的位置key

  var hotkeyController = useHotkeyController('MenuItem');
  var subMenuCtr = useContext(SubMenuContext);
  var hotkeyPosition = hotkeyController.hotkeyPosition,
      isHotkeyType = hotkeyController.isHotkeyType,
      registerValidNode = hotkeyController.registerValidNode,
      unRegisterValidNode = hotkeyController.unRegisterValidNode;

  var _useMousePosition = useMousePosition(),
      setMousePosition = _useMousePosition.setMousePosition,
      mousePosition = _useMousePosition.mousePosition; // 被 Group 包裹需要缩进


  var textIndent = useContext(TextIndentContext); // menuKey处理

  var _useMenuKeyContext = useMenuKeyContext(props.menuKey),
      menuKeyCtxValue = _useMenuKeyContext[0],
      menuKey = _useMenuKeyContext[1]; // 父组件的key路径


  var parentPath = useContext(PathContext); // 支持快捷键时 菜单过长元素自动定位

  var elRef = useHotKeyScrollRef(menuKey, isHotkeyType, hotkeyPosition); // 是否选中，props.selected优先级最高

  var isSelected = typeof selected === 'boolean' ? selected : controller.isMenuItemSelected(menuKey); // 自动关闭，当前组件的props优先级更高

  var autoClose = typeof props.autoClose === 'boolean' ? props.autoClose : controller.autoClose;
  /**
   * 点击 Item
   *
   * @param {React.MouseEvent<HTMLElement>} event
   * @return {void}
   */

  var handleClick = useCallback(function (event) {
    // 自动关闭处理
    if (autoClose) {
      controller.closeAllMenu();
    }

    if (disabled) {
      return;
    } // 点击处理


    if (onClick) {
      onClick(menuKey, event);
    } // 选中处理


    controller.selectMenuItem(menuKey);

    if (controller.clickMenuItem) {
      controller.clickMenuItem(menuKey, event);
    }
  }, [disabled, autoClose, controller, onClick, menuKey]);
  /**
   * 如果有快捷键时新增的鼠标进入事件
   */

  var hotkeyMouseEnterHander = useCallback(function (event) {
    if (!disabled && mousePosition !== event.clientY) {
      hotkeyController.setHotkeyType(false);
      setMousePosition(event.clientY);
      hotkeyController.setHotkeyPosition(hotkeyController.getActionNode(menuKey));
    }
  }, [disabled, menuKey, mousePosition]);
  /**
   * 鼠标移入 Item
   * 应该收起其他菜单
   *
   * @return {void}
   */

  var handleMouseEnter = useCallback(function (event) {
    if (controller.shortcutKey) {
      hotkeyMouseEnterHander(event);
    }

    if (controller.showMenuTimer) {
      window.clearTimeout(controller.showMenuTimer);
    }

    controller.showMenuTimer = window.setTimeout(function () {
      controller.showSubMenu(parentPath);
    }, MENU_VISIBLE_TOGGLE_DELAY);
  }, [parentPath, controller, hotkeyMouseEnterHander]); // 整合height样式

  var styleCombined = useMemo(function () {
    var combined = _extends({}, style || {});

    if (typeof height === 'number' || height === 'auto') {
      combined.height = height;
    }

    if (controller.minWidth) {
      combined.minWidth = controller.minWidth;
    }

    return combined;
  }, [height, style, controller.minWidth]);
  /**
   * 快捷键：注册有效节点
   */

  useEffect(function () {
    if (!disabled && controller.shortcutKey && elRef.current) {
      registerValidNode({
        menuKey: menuKey,
        path: "" + (parentPath ? parentPath + "::" : '') + menuKey,
        onClick: handleClick,
        parentPath: parentPath
      });
      return function () {
        unRegisterValidNode({
          menuKey: menuKey,
          path: "" + (parentPath ? parentPath + "::" : '') + menuKey
        });
      };
    }
  }, [disabled, controller.shortcutKey]);
  /**
   * showSelect 为 false 或者为 true 都是指定并且有意义的，需要用 undefined 来做判断是否有效
   */

  var showSelectFinal = showSelect !== undefined ? showSelect : subMenuCtr.showSelect !== undefined ? subMenuCtr.showSelect : controller.showSelect;
  return /*#__PURE__*/_createElement(Tooltip, {
    disabled: !(disabled && disabledTip),
    title: disabledTip,
    placement: 'top',
    zIndex: tooltipZIndex
  }, /*#__PURE__*/_createElement(MenuItemWrap, {
    path: "" + (parentPath ? parentPath + "::" : '') + menuKey,
    menuId: controller.menuId,
    className: classnames('vertical', {
      selected: isSelected,
      disabled: disabled,
      doublueLine: !!description,
      hotkeyHover: hotkeyPosition && menuKey === hotkeyPosition.menuKey && !disabled && hoverable && controller.shortcutKey
    }),
    ref: elRef,
    hoverable: hoverable,
    isHotkeyType: isHotkeyType,
    textIndent: textIndent,
    onClick: handleClick,
    style: styleCombined,
    onMouseEnter: handleMouseEnter,
    "data-testid": testid ? "menu-item-" + testid : undefined,
    "data-role": role,
    showSelect: showSelectFinal,
    showSelectOnRight: controller.showSelectOnRight
  }, showSelectFinal && !controller.showSelectOnRight && (_SelectedNormalNormal || (_SelectedNormalNormal = /*#__PURE__*/_createElement(SelectedNormalNormal, {
    className: "icon selected-icon"
  }))), /*#__PURE__*/_createElement(MenuKeyContext.Provider, {
    value: menuKeyCtxValue
  }, /*#__PURE__*/_createElement(MenuItemInner, {
    className: classnames({
      selected: isSelected
    })
  }, /*#__PURE__*/_createElement("div", {
    style: {
      display: 'flex',
      flex: 1,
      width: '100%',
      alignItems: 'center'
    }
  }, icon ? /*#__PURE__*/_createElement(MenuItemIconWrap, {
    disabled: disabled,
    doublueLine: !!description
  }, icon) : null, /*#__PURE__*/_createElement(MenuItemContentWrap, null, /*#__PURE__*/_createElement("div", null, children), !!description && /*#__PURE__*/_createElement("div", {
    className: "description"
  }, description))), /*#__PURE__*/_createElement(MenuItemExtraWrap, {
    isShow: !!extra
  }, extra || ''), !extra && controller.showSelectOnRight && (_SelectedNormalNormal2 || (_SelectedNormalNormal2 = /*#__PURE__*/_createElement(SelectedNormalNormal, {
    className: "icon selected-icon"
  })))))));
});
MenuItem.defaultProps = {
  disabled: false,
  hoverable: true
};
