// @ali/we-design/dist-web/components/menu/styled.js
import styled, { css, keyframes } from 'styled-components';
import { Dropdown } from "../dropdown";
import { withTheme } from "../theme";
var hoverActive = /*#__PURE__*/css(["&:hover{background-color:", ";cursor:default;}"], function (_ref) {
  var theme = _ref.theme;
  return theme.ui.backgroundColor;
});
export var TriggerTitleWrap = withTheme(styled.div(["color:", ";"], function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.commonBaseColor1;
}));
export var MenuItemExtraWrap = /*#__PURE__*/styled.div(["font-size:12px;color:rgba(17,31,44,0.4);margin-left:", ";line-height:1;"], function (_ref3) {
  var isShow = _ref3.isShow;
  return isShow ? '20px' : '0px';
});
var handleItemCommon = /*#__PURE__*/css(["position:relative;min-width:150px;height:34px;font-size:13px;line-height:20px;padding:0 ", ";margin:0 ", ";color:#111f2c;border-radius:4px;", " .icon{position:absolute;top:50%;right:23px;transform:translate(120%,-50%);&.selected-icon{position:relative;transform:none;top:auto;right:auto;}}&.disabled{color:rgba(23,26,29,0.24);", "{color:rgba(23,26,29,0.24);}.description,.weicon,", "{color:rgba(23,26,29,0.24);}&:hover{background-color:transparent;cursor:not-allowed;}}"], function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.paddingMedium;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.ui.marginTiny;
}, function (_ref6) {
  var isHotkeyType = _ref6.isHotkeyType,
      hoverable = _ref6.hoverable,
      theme = _ref6.theme;
  return hoverable !== false && "\n    &:hover {\n      background-color: " + (isHotkeyType ? 'inherit' : theme.ui.commonBgHoverColor) + ";\n      cursor: default;\n    }\n  ";
}, TriggerTitleWrap, MenuItemExtraWrap);
export var MenuRoot = withTheme(styled.div(["box-sizing:border-box;font-size:13px;color:", ";&.vertical{min-width:150px;padding:4px 0;}&.horizontal{display:flex;}"], function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.commonActionIconColor;
}));
export var Divider = withTheme(styled.div(["height:1px;margin:", ";background-color:", ";"], function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.marginTiny;
}, function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.commonLineLightColor;
}));
export var SubMenuWrap = withTheme(styled.div.attrs(function (_ref10) {
  var path = _ref10.path,
      menuId = _ref10.menuId;
  return {
    'data-path': path,
    'data-menuid': menuId
  };
})(["user-select:none;"]));
export var SubMenuInnerWrap = withTheme(styled.div(["box-sizing:border-box;", " &.horizontal{height:24px;vertical-align:middle;display:inline-block;padding:", ";margin:0 2px;border-radius:4px;}&.hotkeyHover{background-color:", " !important;}&.vertical{", "}&.opened{background-color:", ";}.arrow-icon{color:rgba(17,31,44,0.56);&.disabled{color:rgba(23,26,29,0.24);}}"], hoverActive, function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.paddingTiny;
}, function (_ref12) {
  var theme = _ref12.theme;
  return theme.ui.commonBgHoverColor;
}, handleItemCommon, function (_ref13) {
  var theme = _ref13.theme;
  return theme.ui.colorLine1;
}));
export var TriggerWrap = withTheme(styled.div(["display:flex;align-items:center;height:100%;&.hasArrow{padding-right:24px;}", ""], function (_ref14) {
  var showSelect = _ref14.showSelect,
      showSelectOnRight = _ref14.showSelectOnRight;
  return showSelect && !showSelectOnRight && css(["padding-left:28px;"]);
}));
export var TriggerIconWrap = withTheme(styled.div(["display:flex;align-items:center;height:100%;margin-right:8px;color:", ";span:first-child{font-size:20px;}"], function (_ref15) {
  var theme = _ref15.theme;
  return theme.ui.commonActionIconColor;
}));
export var MenuItemInner = withTheme(styled.div(["display:flex;align-items:center;width:100%;justify-content:space-between;"]));
export var MenuItemIconWrap = withTheme(styled.div(["display:flex;align-items:center;height:100%;margin-right:8px;color:", ";", " ", " span:first-child{font-size:20px;}"], function (_ref16) {
  var theme = _ref16.theme;
  return theme.ui.commonActionIconColor;
}, function (_ref17) {
  var disabled = _ref17.disabled;
  return disabled && css(["color:rgba(23,26,29,0.24);"]);
}, function (_ref18) {
  var doublueLine = _ref18.doublueLine;
  return doublueLine && css([".weicon{font-size:32px;}"]);
}));
export var MenuItemContentWrap = withTheme(styled.div([".description{font-size:12px;color:rgba(17,31,44,0.4);line-height:17px;}"]));
export var MenuItemWrap = withTheme(styled.div.attrs(function (_ref19) {
  var path = _ref19.path,
      menuId = _ref19.menuId;
  return {
    'data-path': path,
    'data-menuid': menuId
  };
})(["user-select:none;.selected-icon{", " color:#171A1D;font-size:20px;margin-right:", ";}&.hotkeyHover{background-color:", " !important;}&.vertical{", ";display:flex;align-items:center;&.selected{.selected-icon{font-size:20px;", "}}&.doublueLine{height:53px;}}&.horizontal{", " display:inline-block;padding:", ";margin:0 ", ";border-radius:", ";}"], function (_ref20) {
  var showSelect = _ref20.showSelect;
  return showSelect ? css(["visibility:hidden;"]) : " display: none;";
}, function (_ref21) {
  var showSelectOnRight = _ref21.showSelectOnRight;
  return showSelectOnRight ? '0px' : '8px';
}, function (_ref22) {
  var theme = _ref22.theme;
  return theme.ui.commonBgHoverColor;
}, handleItemCommon, function (_ref23) {
  var showSelect = _ref23.showSelect;
  return showSelect ? css(["visibility:visible;"]) : css(["display:inline-block;"]);
}, hoverActive, function (_ref24) {
  var theme = _ref24.theme;
  return theme.ui.paddingTiny;
}, function (_ref25) {
  var theme = _ref25.theme;
  return theme.ui.marginTiny;
}, function (_ref26) {
  var theme = _ref26.theme;
  return theme.ui.radiusMedium;
}));
export var MenuItemGroupWrap = /*#__PURE__*/styled.div(["user-select:none;"]);
export var MenuItemGroupTitle = withTheme(styled.div(["font-size:12px;color:", ";line-height:36px;height:36px;padding-left:", ";"], function (_ref27) {
  var theme = _ref27.theme;
  return theme.ui.colorText4;
}, function (_ref28) {
  var theme = _ref28.theme;
  return theme.ui.padding;
}));
export var MenuItemGroupList = /*#__PURE__*/styled.div([""]);
export var SubMenuDropdown = /*#__PURE__*/styled(Dropdown)(["width:100%;"]);
var scaleUpTr = /*#__PURE__*/keyframes(["0%{transform:scale(0.8);opacity:0;transform-origin:0% 0%;}100%{transform:scale(1);opacity:1;transform-origin:0% 0%;}"]);
var hide = /*#__PURE__*/keyframes(["100%{opacity:1;}0%{opacity:0;}"]);
export var SubMenuOverlay = withTheme(styled.div(["width:100%;box-shadow:0px 0px 1px rgba(0,0,0,0.24),0px 8px 24px rgba(10,30,65,0.16);border-radius:4px;padding:4px 0;background-color:#fff;&.scaleUp{animation:", " 100ms cubic-bezier(0.39,0.575,0.565,1) forwards;}&.hide{transition:", " 2s ease-out;}"], scaleUpTr, hide));
export var SubMenuTransitionWrap = /*#__PURE__*/styled.div(["&.menu-transition-leave{animation:hide 0.1s ease-out both;}"]);
export var InlineMenuItemIconWrapper = withTheme(styled.div(["width:28px;height:28px;padding:4px;cursor:pointer;color:", ";display:flex;justify-content:center;align-items:center;overflow:hidden;.weicon{font-size:20px;}&:hover{background-color:rgba(126,134,142,0.12);}&.hotkeyHover{background-color:", " !important;}&.disabled{color:rgba(23,26,29,0.24);.weicon{color:rgba(23,26,29,0.24);}&:hover{background-color:transparent;cursor:not-allowed;}}"], function (_ref29) {
  var theme = _ref29.theme;
  return theme.ui.commonActionIconColor;
}, function (_ref30) {
  var theme = _ref30.theme;
  return theme.ui.commonBgHoverColor;
}));
export var InlineMenuItemContainer = /*#__PURE__*/styled.div(["display:flex;max-width:250px;min-width:170px;justify-content:flex-start;flex-wrap:wrap;padding:0 0 0 9px;& > *{margin-right:1px;}& > div{border-radius:4px;}"]);
