// @ali/4ever-bi/dist-node/plugins/placeholder/mobile/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _createRenderNode = _interopRequireDefault(require("./createRenderNode"));

var actions = _interopRequireWildcard(require("../actions"));

var _onAction = _interopRequireDefault(require("../handlers/onAction"));

var _everPluginPlaceholder = require("@ali/4ever-plugin-placeholder");

function createPlaceholderPlugin(config) {
  const placeholderConfig = config.placeholder || {};
  return { ...(0, _everPluginPlaceholder.moFactory)({}),
    renderNode: (0, _createRenderNode.default)(placeholderConfig),
    onAction: _onAction.default,
    models: [{
      klass: 'block',
      type: 'ph'
    }, {
      klass: 'inline',
      type: 'iph'
    }]
  };
}

createPlaceholderPlugin.actions = actions;
var _default = createPlaceholderPlugin;
exports.default = _default;
