// @ali/4ever-bi/dist-web/plugins/quickInsert/components/errorBoundary.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
var Wrapper = /*#__PURE__*/styled.div(["overflow:auto;height:100%;"]);

var _ref = /*#__PURE__*/_createElement("h2", null, "Something went wrong.");

var _ref2 = /*#__PURE__*/_createElement("br", null);

var ErrorBoundary = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(ErrorBoundary, _React$Component);

  function ErrorBoundary(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;
    _this.state = {
      error: null,
      errorInfo: null
    };
    return _this;
  }

  var _proto = ErrorBoundary.prototype;

  _proto.componentDidCatch = function componentDidCatch(error, errorInfo) {
    this.setState({
      error: error,
      errorInfo: errorInfo
    });
  };

  _proto.render = function render() {
    var _this$state = this.state,
        errorInfo = _this$state.errorInfo,
        error = _this$state.error;
    var children = this.props.children;

    if (errorInfo) {
      return /*#__PURE__*/_createElement(Wrapper, null, _ref, /*#__PURE__*/_createElement("div", {
        style: {
          whiteSpace: 'pre-wrap'
        }
      }, error && error.toString(), _ref2, errorInfo.componentStack));
    }

    return children;
  };

  return ErrorBoundary;
}(React.Component);

export default ErrorBoundary;
