// @ali/4ever-plugin-template-button/dist-node/components/container.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Container = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _closeView = _interopRequireDefault(require("./closeView"));

var _openView = _interopRequireDefault(require("./openView"));

const _createElement = /*#__PURE__*/_react.default.createElement;

const Container = props => {
  const {
    node,
    controller,
    child,
    config,
    isSelected,
    editable = true
  } = props;
  const {
    data: {
      metadata
    }
  } = node;
  const placeholder = config?.locale?.namePlaceholder || 'Add a new item';

  if (metadata.isOpen) {
    return /*#__PURE__*/_createElement(_openView.default, (0, _extends2.default)({}, metadata, {
      placeholder: placeholder,
      controller: controller,
      nodeKey: node.key,
      child: child,
      config: config,
      editable: editable,
      isSelected: isSelected
    }));
  } else {
    return /*#__PURE__*/_createElement(_closeView.default, (0, _extends2.default)({}, metadata, {
      placeholder: placeholder,
      controller: controller,
      node: node,
      isSelected: isSelected,
      editable: editable,
      config: config
    }));
  }
};

exports.Container = Container;
