// @ali/4ever-plugin-template-button/dist-node/components/styled.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ViewWrapper = exports.OpenViewFooter = exports.TemplateInputArea = exports.TemplateInput = exports.OpenViewContentGroup = exports.OpenViewContent = exports.OpenViewHeader = exports.OpenViewContainer = exports.Title = exports.Wrapper = exports.Right = exports.Left = void 0;

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _everComponent = require("@ali/4ever-component");

const Left = /*#__PURE__*/_styledComponents.default.div(["display:flex;align-items:center;border:1px solid rgba(126,134,142,0.24);background-color:#fff;box-sizing:border-box;border-radius:4px;line-height:24px;max-width:100%;overflow:hidden;", " svg{height:18px;width:24px;color:rgba(23,26,29,0.6);margin-top:-1px;transform:rotate(", ");>path{fill:rgba(23,26,29,0.6)!important;}}"], ({
  isOpen
}) => isOpen && (0, _styledComponents.css)(["border-color:transparent;"]), ({
  direction
}) => direction === 'top' ? '90deg' : '-90deg');

exports.Left = Left;

const Right = /*#__PURE__*/_styledComponents.default.div(["display:none;position:absolute;right:5px;align-items:center;height:24px;line-height:24px;padding:0 2px;.inline-toolbar-container>div{background:transparent;}"]);

exports.Right = Right;
const WrapperSelectCss = /*#__PURE__*/(0, _styledComponents.css)(["background:rgba(126,134,142,0.12);border-color:rgba(126,134,142,0.24);border-radius:4px;"]);

const Wrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;position:relative;padding:5px;cursor:default;color:rgba(23,26,29,0.6);box-sizing:border-box;border:1px solid transparent;&:hover{padding-right:60px;}&:hover ", "{display:flex;}", ""], Right, p => !p.enableBorderHighlight ? (0, _styledComponents.css)(["", " ", " &:hover ", "{background:#fff;}"], p.isSelected && !p.isOpen ? WrapperSelectCss : '', !p.isOpen ? (0, _styledComponents.css)(["&:hover{", "}"], WrapperSelectCss) : '', Left) : (0, _styledComponents.css)(["", " ", ""], p.isHover ? _everComponent.hoverCSS.noBorderBoxHoverCSS : '', p.isSelected ? _everComponent.hoverCSS.noBorderBoxSelectCSS : ''));

exports.Wrapper = Wrapper;

const Title = /*#__PURE__*/_styledComponents.default.div(["text-overflow:ellipsis;overflow:hidden;white-space:nowrap;line-height:24px;font-size:11pt;"]);

exports.Title = Title;

const OpenViewContainer = /*#__PURE__*/_styledComponents.default.div(["border:1px solid transparent;box-shadow:0px 0px 1px rgba(0,0,0,0.16),0px 1px 4px rgba(23,26,29,0.16);border-radius:8px;overflow:hidden;margin-top:2px;"]);

exports.OpenViewContainer = OpenViewContainer;

const OpenViewHeader = /*#__PURE__*/_styledComponents.default.div(["overflow:hidden;padding:12px 22px;>span{color:#000;font-size:16px;font-weight:bold;}"]);

exports.OpenViewHeader = OpenViewHeader;

const OpenViewContent = /*#__PURE__*/_styledComponents.default.div(["overflow:hidden;padding:12px 22px;"]);

exports.OpenViewContent = OpenViewContent;

const OpenViewContentGroup = /*#__PURE__*/_styledComponents.default.div(["margin:4px 0 16px;>h4{margin-bottom:4px;font-size:14px;color:#000;}>div{}"]);

exports.OpenViewContentGroup = OpenViewContentGroup;

const TemplateInput = /*#__PURE__*/_styledComponents.default.input(["width:100%;height:40px;border:1px solid rgba(126,134,142,0.24);border-radius:4px;padding:5px 12px;box-sizing:border-box;color:#2c2c2c;font-size:14px;color:#2c2c2c;caret-color:#000;&:focus{outline:none;}"]);

exports.TemplateInput = TemplateInput;

const TemplateInputArea = /*#__PURE__*/_styledComponents.default.div(["border:1px solid rgba(126,134,142,0.24);box-sizing:border-box;border-radius:4px;padding:5px 12px;"]);

exports.TemplateInputArea = TemplateInputArea;

const OpenViewFooter = /*#__PURE__*/_styledComponents.default.div(["overflow:hidden;padding:0 0 14px 22px;"]);

exports.OpenViewFooter = OpenViewFooter;

const ViewWrapper = /*#__PURE__*/_styledComponents.default.div(["position:relative;border-radius:6px;", " ", ""], p => p.isHover ? _everComponent.hoverCSS.borderBoxHoverCSS : '', p => p.isActive ? _everComponent.hoverCSS.borderBoxActiveCSS : '');

exports.ViewWrapper = ViewWrapper;
