// @ali/we-design/dist-web/components/modal/styled.js
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/explicit-function-return-type */
import styled from 'styled-components';
import { Button } from "../../components/button";
import { withTheme } from "../theme";
export var CancelButton = withTheme(styled(Button)(["border-radius:", ";height:32px;"], function (_ref) {
  var theme = _ref.theme;
  return theme.ui.radius;
}));
export var ConfirmButton = withTheme(styled(CancelButton)(["margin-left:", ";"], function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.margin;
}));
export var Root = /*#__PURE__*/styled.div(["box-sizing:border-box;*{box-sizing:border-box;}"]);
export var Title = /*#__PURE__*/styled.div([""]);
export var CloseButton = withTheme(styled.div(["position:absolute;right:16px;top:16px;padding:2px;line-height:1;font-size:20px;cursor:pointer;transition:all 0.1s ease;display:flex;justify-content:center;align-items:center;&:hover{background:rgba(126,134,142,0.12);border-radius:", ";}& span{font-size:20px;}"], function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.radius;
}));
export var Header = withTheme(styled.div(["font-weight:500;font-size:", ";margin-bottom:", ";line-height:24px;"], function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.fontSizeLarge;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.ui.marginLarge;
}));
export var Body = withTheme(styled.div(["font-size:", ";"], function (_ref6) {
  var theme = _ref6.theme;
  return theme.ui.fontSize;
}));
export var Footer = withTheme(styled.div(["margin-top:", ";"], function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.marginLarge;
}));
export var FooterWrap = /*#__PURE__*/styled.div(["display:flex;justify-content:flex-end;align-items:center;"]);
export var Content = withTheme(styled.div(["min-width:312px;user-select:none;position:relative;pointer-events:auto;padding:", ";box-shadow:0px 12px 32px 0px rgba(0,0,0,0.24),0px 0px 1px 0px rgba(0,0,0,0.24);background-color:#f5f6f8;border-radius:", ";width:", ";"], function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.paddingLarge;
}, function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.radiusMedium;
}, function (_ref10) {
  var width = _ref10.width;
  return typeof width === 'number' ? width + "px" : width;
}));
export var Mask = withTheme(styled.div(["display:", ";position:fixed;top:0;right:0;bottom:0;left:0;z-index:", ";height:100%;background-color:", ";"], function (_ref11) {
  var visible = _ref11.visible;
  return visible ? 'block' : 'none';
}, function (_ref12) {
  var theme = _ref12.theme;
  return theme.ui.zIndexModalMask;
}, function (_ref13) {
  var theme = _ref13.theme;
  return theme.ui.modalMaskBackground;
}));
export var Wrap = withTheme(styled.div(["display:", ";position:fixed;top:0;right:0;bottom:0;left:0;z-index:", ";overflow:auto;outline:0;justify-content:center;align-items:center;pointer-events:none;"], function (_ref14) {
  var visible = _ref14.visible;
  return visible ? 'flex' : 'none';
}, function (_ref15) {
  var theme = _ref15.theme;
  return theme.ui.zIndexModal;
}));
export var TitleTextWrap = withTheme(styled.span(["margin-left:", ";"], function (_ref16) {
  var theme = _ref16.theme;
  return theme.ui.margin;
}));
export var TitleIconWrap = withTheme(styled.span(["font-size:22px;color:", ";"], function (_ref17) {
  var theme = _ref17.theme,
      colorKey = _ref17.colorKey;
  return colorKey ? theme.ui[colorKey] : theme.ui.colorText3;
}));
export var TitleIconTextWrap = /*#__PURE__*/styled.div(["display:flex;align-items:center;"]);
