// @ali/4ever-plugin-sticker/dist-node/utils/bamboo.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getStickerStyle", {
  enumerable: true,
  get: function () {
    return _getStickerStyle.default;
  }
});
Object.defineProperty(exports, "getEmojiData", {
  enumerable: true,
  get: function () {
    return _getEmojiData.getEmojiData;
  }
});
Object.defineProperty(exports, "Sticker", {
  enumerable: true,
  get: function () {
    return _sticker.default;
  }
});
Object.defineProperty(exports, "StickerItem", {
  enumerable: true,
  get: function () {
    return _stickerItem.default;
  }
});
Object.defineProperty(exports, "DEFAULT_SZ", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_SZ;
  }
});
Object.defineProperty(exports, "DEFAULT_STICKERS", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_STICKERS;
  }
});
Object.defineProperty(exports, "StickerRole", {
  enumerable: true,
  get: function () {
    return _role.StickerRole;
  }
});
Object.defineProperty(exports, "StickerType", {
  enumerable: true,
  get: function () {
    return _meta.ELEMENT_TYPE;
  }
});

var _getStickerStyle = _interopRequireDefault(require("./getStickerStyle"));

var _getEmojiData = require("./getEmojiData");

var _sticker = _interopRequireDefault(require("./sticker"));

var _stickerItem = _interopRequireDefault(require("./stickerItem"));

var _constants = require("./constants");

var _role = require("./role");

var _meta = require("./meta");
