// @ali/4ever-plugin-sticker/dist-node/utils/constants.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_STICKERS = exports.DEFAULT_SZ = exports.DEFAULT_LANG = exports.SPRITE = exports.SINGLE = exports.DATA_STICKER = exports.STICKER_SZ_IN_PANEL = exports.STICKER_PLACEHOLDER_DATA = exports.STICKER_PANEL_STYLE = exports.STICKER_PANEL_HEIGHT = exports.STICKER_PANEL_COLUMMS = exports.STICKER_NODE_TYPE = exports.MOVE_SET = void 0;

var _models = _interopRequireDefault(require("../mo/models"));

var _everUtils = require("@ali/4ever-utils");

/* eslint-disable max-lines */
// 自定义表情类型, sprite -> 雪碧图， single -> 单图
const SINGLE = 'single';
exports.SINGLE = SINGLE;
const SPRITE = 'sprite'; // 默认语言

exports.SPRITE = SPRITE;
const DEFAULT_LANG = 'zh_CN'; // 默认大小

exports.DEFAULT_LANG = DEFAULT_LANG;
const DEFAULT_SZ = 11; // 默认表情配置

exports.DEFAULT_SZ = DEFAULT_SZ;
const DEFAULT_STICKERS = {
  type: 'sprite',
  src: 'https://img.alicdn.com/imgextra/i1/O1CN01sJOJLH1yb1Cy75YJe_!!6000000006596-2-tps-672-2592.png',
  width: 980,
  height: 3780,
  stickers: [{
    stickerId: 'sticker_001',
    code: '优先级: 1',
    x: 0,
    y: 0,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Priority: 1',
      ja_JP: '優先度: 1',
      zh_CN: '优先级: 1',
      zh_HK: '優先級：1',
      zh_TW: '優先級: 1'
    }
  }, {
    stickerId: 'sticker_002',
    code: '优先级: 2',
    x: -140,
    y: 0,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Priority: 2',
      ja_JP: '優先度: 2',
      zh_CN: '优先级: 2',
      zh_HK: '優先級：2',
      zh_TW: '優先級: 2'
    }
  }, {
    stickerId: 'sticker_003',
    code: '优先级: 3',
    x: -280,
    y: 0,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Priority: 3',
      ja_JP: '優先度: 3',
      zh_CN: '优先级: 3',
      zh_HK: '優先級：3',
      zh_TW: '優先級: 3'
    }
  }, {
    stickerId: 'sticker_004',
    code: '优先级: 4',
    x: -420,
    y: 0,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Priority: 4',
      ja_JP: '優先度: 4',
      zh_CN: '优先级: 4',
      zh_HK: '優先級：4',
      zh_TW: '優先級: 4'
    }
  }, {
    stickerId: 'sticker_005',
    code: '优先级: 5',
    x: -560,
    y: 0,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Priority: 5',
      ja_JP: '優先度: 5',
      zh_CN: '优先级: 5',
      zh_HK: '優先級：5',
      zh_TW: '優先級: 5'
    }
  }, {
    stickerId: 'sticker_006',
    code: '优先级: 6',
    x: -700,
    y: 0,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Priority: 6',
      ja_JP: '優先度: 6',
      zh_CN: '优先级: 6',
      zh_HK: '優先級：6',
      zh_TW: '優先級: 6'
    }
  }, {
    stickerId: 'sticker_007',
    code: '优先级: 7',
    x: -840,
    y: 0,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Priority: 7',
      ja_JP: '優先度: 7',
      zh_CN: '优先级: 7',
      zh_HK: '優先級：7',
      zh_TW: '優先級: 7'
    }
  }, {
    stickerId: 'sticker_008',
    code: '进度: 未开始',
    x: 0,
    y: -140,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Progress: Not started',
      ja_JP: '進捗状況: ていない',
      zh_CN: '进度: 未开始',
      zh_HK: '進度: 未开始',
      zh_TW: '進度: 未開始'
    }
  }, {
    stickerId: 'sticker_009',
    code: '进度: 20%',
    x: -140,
    y: -140,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Progress: 20%',
      ja_JP: '進捗状況: 20%',
      zh_CN: '进度: 20%',
      zh_HK: '進度: 20%',
      zh_TW: '進度: 20%'
    }
  }, {
    stickerId: 'sticker_010',
    code: '进度: 40%',
    x: -280,
    y: -140,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Progress: 40%',
      ja_JP: '進捗状況: 40%',
      zh_CN: '进度: 40%',
      zh_HK: '進度: 40%',
      zh_TW: '進度: 40%'
    }
  }, {
    stickerId: 'sticker_011',
    code: '进度: 50%',
    x: -420,
    y: -140,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Progress: 50%',
      ja_JP: '進捗状況: 50%',
      zh_CN: '进度: 50%',
      zh_HK: '進度: 50%',
      zh_TW: '進度: 50%'
    }
  }, {
    stickerId: 'sticker_012',
    code: '进度: 70%',
    x: -560,
    y: -140,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Progress: 70%',
      ja_JP: '進捗状況: 70%',
      zh_CN: '进度: 70%',
      zh_HK: '進度: 70%',
      zh_TW: '進度: 70%'
    }
  }, {
    stickerId: 'sticker_013',
    code: '进度: 90%',
    x: -700,
    y: -140,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Progress: 90%',
      ja_JP: '進捗状況: 90%',
      zh_CN: '进度: 90%',
      zh_HK: '進度: 90%',
      zh_TW: '進度: 90%'
    }
  }, {
    stickerId: 'sticker_014',
    code: '进度: 已完成',
    x: -840,
    y: -140,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Progress: Completed',
      ja_JP: '進捗状況: 完了',
      zh_CN: '进度: 已完成',
      zh_HK: '進度: 已完成',
      zh_TW: '進度: 已完成'
    }
  }, {
    stickerId: 'sticker_split_line_01'
  }, {
    stickerId: 'sticker_015',
    code: '微笑',
    x: 0,
    y: -280,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Smile',
      ja_JP: 'スマイル',
      zh_CN: '微笑',
      zh_HK: '微笑',
      zh_TW: '微笑'
    }
  }, {
    stickerId: 'sticker_016',
    code: '憨笑',
    x: -140,
    y: -280,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Wow',
      ja_JP: 'ヨダレ',
      zh_CN: '憨笑',
      zh_HK: '憨笑',
      zh_TW: '憨笑'
    }
  }, {
    stickerId: 'sticker_017',
    code: '大笑',
    x: -280,
    y: -280,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Laugh',
      ja_JP: '笑い',
      zh_CN: '大笑',
      zh_HK: '大笑',
      zh_TW: '大笑'
    }
  }, {
    stickerId: 'sticker_018',
    code: '加油',
    x: -420,
    y: -280,
    width: 140,
    height: 140,
    locale: {
      en_US: 'YouCanDoIt',
      ja_JP: '頑張って',
      zh_CN: '加油',
      zh_HK: '加油',
      zh_TW: '加油'
    }
  }, {
    stickerId: 'sticker_019',
    code: '色',
    x: -560,
    y: -280,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Yum',
      ja_JP: 'ハート目',
      zh_CN: '色',
      zh_HK: '色',
      zh_TW: '色'
    }
  }, {
    stickerId: 'sticker_020',
    code: '偷笑',
    x: -700,
    y: -280,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Chuckle',
      ja_JP: 'クスクス',
      zh_CN: '偷笑',
      zh_HK: '偷笑',
      zh_TW: '偷笑'
    }
  }, {
    stickerId: 'sticker_021',
    code: '跳舞',
    x: -840,
    y: -280,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Yay',
      ja_JP: '嬉しい',
      zh_CN: '跳舞',
      zh_HK: '跳舞',
      zh_TW: '跳舞'
    }
  }, {
    stickerId: 'sticker_022',
    code: '捂脸哭',
    x: 0,
    y: -420,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Facepalm',
      ja_JP: '目を覆う',
      zh_CN: '捂脸哭',
      zh_HK: '捂臉哭',
      zh_TW: '捂臉哭'
    }
  }, {
    stickerId: 'sticker_023',
    code: '笑哭',
    x: -140,
    y: -420,
    width: 140,
    height: 140,
    locale: {
      en_US: 'LaughAndCry',
      ja_JP: '泣き笑い',
      zh_CN: '笑哭',
      zh_HK: '笑哭',
      zh_TW: '笑哭'
    }
  }, {
    stickerId: 'sticker_024',
    code: '流泪',
    x: -280,
    y: -420,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Sob',
      ja_JP: '泣き',
      zh_CN: '流泪',
      zh_HK: '流淚',
      zh_TW: '流淚'
    }
  }, {
    stickerId: 'sticker_025',
    code: '疑问',
    x: -420,
    y: -420,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Question',
      ja_JP: '疑問',
      zh_CN: '疑问',
      zh_HK: '疑問',
      zh_TW: '疑問'
    }
  }, {
    stickerId: 'sticker_026',
    code: '傻笑',
    x: -560,
    y: -420,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Oops',
      ja_JP: '汗',
      zh_CN: '傻笑',
      zh_HK: '傻笑',
      zh_TW: '傻笑'
    }
  }, {
    stickerId: 'sticker_027',
    code: '流鼻血',
    x: -700,
    y: -420,
    width: 140,
    height: 140,
    locale: {
      en_US: 'BloodyNose',
      ja_JP: '鼻血',
      zh_CN: '流鼻血',
      zh_HK: '流鼻血',
      zh_TW: '流鼻血'
    }
  }, {
    stickerId: 'sticker_028',
    code: '狗子',
    x: -840,
    y: -420,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Puppy',
      ja_JP: '犬',
      zh_CN: '狗子',
      zh_HK: '狗子',
      zh_TW: '狗子'
    }
  }, {
    stickerId: 'sticker_029',
    code: '赞',
    x: 0,
    y: -560,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Like',
      ja_JP: 'いいね！',
      zh_CN: '赞',
      zh_HK: '贊',
      zh_TW: '贊'
    }
  }, {
    stickerId: 'sticker_030',
    code: 'OK',
    x: -140,
    y: -560,
    width: 140,
    height: 140,
    locale: {
      en_US: 'OK',
      ja_JP: 'OK',
      zh_CN: 'OK',
      zh_HK: 'OK',
      zh_TW: 'OK'
    }
  }, {
    stickerId: 'sticker_031',
    code: '抱拳',
    x: -280,
    y: -560,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Salute',
      ja_JP: 'よろしく',
      zh_CN: '抱拳',
      zh_HK: '抱拳',
      zh_TW: '抱拳'
    }
  }, {
    stickerId: 'sticker_032',
    code: '向上',
    x: -420,
    y: -560,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Up',
      ja_JP: '上',
      zh_CN: '向上',
      zh_HK: '向上',
      zh_TW: '向上'
    }
  }, {
    stickerId: 'sticker_033',
    code: '向下',
    x: -560,
    y: -560,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Down',
      ja_JP: '下',
      zh_CN: '向下',
      zh_HK: '向下',
      zh_TW: '向下'
    }
  }, {
    stickerId: 'sticker_034',
    code: '向左',
    x: -700,
    y: -560,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Left',
      ja_JP: '左',
      zh_CN: '向左',
      zh_HK: '向左',
      zh_TW: '向左'
    }
  }, {
    stickerId: 'sticker_035',
    code: '向右',
    x: -840,
    y: -560,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Right',
      ja_JP: '右',
      zh_CN: '向右',
      zh_HK: '向右',
      zh_TW: '向右'
    }
  }, {
    stickerId: 'sticker_036',
    code: '资料',
    x: 0,
    y: -700,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Information',
      ja_JP: '情報',
      zh_CN: '资料',
      zh_HK: '資料',
      zh_TW: '資料'
    }
  }, {
    stickerId: 'sticker_037',
    code: '本子',
    x: -140,
    y: -700,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Book',
      ja_JP: '本',
      zh_CN: '本子',
      zh_HK: '本子',
      zh_TW: '本子'
    }
  }, {
    stickerId: 'sticker_038',
    code: '笔记本',
    x: -280,
    y: -700,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Notebook',
      ja_JP: 'ノートブック',
      zh_CN: '笔记本',
      zh_HK: '筆記本',
      zh_TW: '筆記本'
    }
  }, {
    stickerId: 'sticker_039',
    code: '折线图',
    x: -420,
    y: -700,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Line chart',
      ja_JP: '折れ線グラフ',
      zh_CN: '折线图',
      zh_HK: '折線圖',
      zh_TW: '折線圖'
    }
  }, {
    stickerId: 'sticker_040',
    code: '柱状图',
    x: -560,
    y: -700,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Histogram',
      ja_JP: 'ヒストグラム',
      zh_CN: '柱状图',
      zh_HK: '柱狀圖',
      zh_TW: '柱狀圖'
    }
  }, {
    stickerId: 'sticker_041',
    code: '羽毛笔',
    x: -700,
    y: -700,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Quill',
      ja_JP: 'クイル',
      zh_CN: '羽毛笔',
      zh_HK: '羽毛筆',
      zh_TW: '羽毛筆'
    }
  }, {
    stickerId: 'sticker_042',
    code: '钢笔',
    x: -840,
    y: -700,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Pen',
      ja_JP: 'ペン',
      zh_CN: '钢笔',
      zh_HK: '鋼筆',
      zh_TW: '鋼筆'
    }
  }, {
    stickerId: 'sticker_043',
    code: '警告',
    x: 0,
    y: -840,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Warning',
      ja_JP: '警告',
      zh_CN: '警告',
      zh_HK: '警告',
      zh_TW: '警告'
    }
  }, {
    stickerId: 'sticker_044',
    code: '问号',
    x: -140,
    y: -840,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Question mark',
      ja_JP: '疑問符',
      zh_CN: '问号',
      zh_HK: '問號',
      zh_TW: '問號'
    }
  }, {
    stickerId: 'sticker_045',
    code: '禁止',
    x: -280,
    y: -840,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Prohibition',
      ja_JP: '禁止',
      zh_CN: '禁止',
      zh_HK: '禁止',
      zh_TW: '禁止'
    }
  }, {
    stickerId: 'sticker_046',
    code: '禁行',
    x: -420,
    y: -840,
    width: 140,
    height: 140,
    locale: {
      en_US: 'No entry',
      ja_JP: '通路なし',
      zh_CN: '禁行',
      zh_HK: '禁行',
      zh_TW: '禁行'
    }
  }, {
    stickerId: 'sticker_047',
    code: '锁',
    x: -560,
    y: -840,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Lock',
      ja_JP: 'ロック',
      zh_CN: '锁',
      zh_HK: '鎖',
      zh_TW: '鎖'
    }
  }, {
    stickerId: 'sticker_048',
    code: '气泡',
    x: -700,
    y: -840,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Bubble',
      ja_JP: 'バブル',
      zh_CN: '气泡',
      zh_HK: '氣泡',
      zh_TW: '氣泡'
    }
  }, {
    stickerId: 'sticker_049',
    code: '沙漏',
    x: -840,
    y: -840,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Hourglass',
      ja_JP: '砂時計',
      zh_CN: '沙漏',
      zh_HK: '沙漏',
      zh_TW: '沙漏'
    }
  }, {
    stickerId: 'sticker_050',
    code: '公文包',
    x: 0,
    y: -980,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Briefcase',
      ja_JP: 'ブリーフケース',
      zh_CN: '公文包',
      zh_HK: '公文包',
      zh_TW: '公文包'
    }
  }, {
    stickerId: 'sticker_051',
    code: '火箭',
    x: -140,
    y: -980,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Rocket',
      ja_JP: 'ロケット',
      zh_CN: '火箭',
      zh_HK: '火箭',
      zh_TW: '火箭'
    }
  }, {
    stickerId: 'sticker_052',
    code: '火',
    x: -280,
    y: -980,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Fire',
      ja_JP: '火',
      zh_CN: '火',
      zh_HK: '火',
      zh_TW: '火'
    }
  }, {
    stickerId: 'sticker_053',
    code: '奖牌',
    x: -420,
    y: -980,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Medal',
      ja_JP: 'メダル',
      zh_CN: '奖牌',
      zh_HK: '獎牌',
      zh_TW: '獎牌'
    }
  }, {
    stickerId: 'sticker_054',
    code: '灯泡',
    x: -560,
    y: -980,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Bulb',
      ja_JP: '電球',
      zh_CN: '灯泡',
      zh_HK: '燈泡',
      zh_TW: '燈泡'
    }
  }, {
    stickerId: 'sticker_055',
    code: '钉子',
    x: -700,
    y: -980,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Nail',
      ja_JP: 'ネイル',
      zh_CN: '钉子',
      zh_HK: '釘子',
      zh_TW: '釘子'
    }
  }, {
    stickerId: 'sticker_056',
    code: '旗子',
    x: -840,
    y: -980,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Flag',
      ja_JP: 'フラグ',
      zh_CN: '旗子',
      zh_HK: '旗子',
      zh_TW: '旗子'
    }
  }, {
    stickerId: 'sticker_057',
    code: '茶',
    x: 0,
    y: -1120,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Tea',
      ja_JP: 'お茶',
      zh_CN: '茶',
      zh_HK: '茶',
      zh_TW: '茶'
    }
  }, {
    stickerId: 'sticker_058',
    code: '休假',
    x: -140,
    y: -1120,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Vacation',
      ja_JP: '休暇',
      zh_CN: '休假',
      zh_HK: '度假',
      zh_TW: '度假'
    }
  }, {
    stickerId: 'sticker_059',
    code: '气球',
    x: -280,
    y: -1120,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Balloon',
      ja_JP: 'バルーン',
      zh_CN: '气球',
      zh_HK: '氣球',
      zh_TW: '氣球'
    }
  }, {
    stickerId: 'sticker_060',
    code: '锦鲤',
    x: -420,
    y: -1120,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Koi',
      ja_JP: '鯉',
      zh_CN: '锦鲤',
      zh_HK: '錦鯉',
      zh_TW: '錦鯉'
    }
  }, {
    stickerId: 'sticker_061',
    code: '咖啡',
    x: -560,
    y: -1120,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Coffee',
      ja_JP: 'コーヒー',
      zh_CN: '咖啡',
      zh_HK: '咖啡',
      zh_TW: '咖啡'
    }
  }, {
    stickerId: 'sticker_062',
    code: '奶茶',
    x: -700,
    y: -1120,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Milk tea',
      ja_JP: 'ミルクティー',
      zh_CN: '奶茶',
      zh_HK: '奶茶',
      zh_TW: '奶茶'
    }
  }, {
    stickerId: 'sticker_063',
    code: '调色板',
    x: -840,
    y: -1120,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Palette',
      ja_JP: 'パレット',
      zh_CN: '调色板',
      zh_HK: '調色板',
      zh_TW: '調色板'
    }
  }, {
    stickerId: 'sticker_064',
    code: '感谢',
    x: 0,
    y: -1260,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Thanks',
      ja_JP: 'ありがとう',
      zh_CN: '感谢',
      zh_HK: '感謝',
      zh_TW: '感謝'
    }
  }, {
    stickerId: 'sticker_065',
    code: '打招呼',
    x: -140,
    y: -1260,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Hi',
      ja_JP: 'やあ',
      zh_CN: '打招呼',
      zh_HK: '打招呼',
      zh_TW: '打招呼'
    }
  }, {
    stickerId: 'sticker_066',
    code: '666',
    x: -280,
    y: -1260,
    width: 140,
    height: 140,
    locale: {
      en_US: '666',
      ja_JP: '666',
      zh_CN: '666',
      zh_HK: '666',
      zh_TW: '666'
    }
  }, {
    stickerId: 'sticker_067',
    code: '握手',
    x: -420,
    y: -1260,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Shake',
      ja_JP: '握手',
      zh_CN: '握手',
      zh_HK: '握手',
      zh_TW: '握手'
    }
  }, {
    stickerId: 'sticker_068',
    code: '胜利',
    x: -560,
    y: -1260,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Victory',
      ja_JP: '勝利',
      zh_CN: '胜利',
      zh_HK: '勝利',
      zh_TW: '勝利'
    }
  }, {
    stickerId: 'sticker_069',
    code: '一点点',
    x: -700,
    y: -1260,
    width: 140,
    height: 140,
    locale: {
      en_US: 'ALittle',
      ja_JP: 'すこし',
      zh_CN: '一点点',
      zh_HK: '一點點',
      zh_TW: '一點點'
    }
  }, {
    stickerId: 'sticker_070',
    code: '鼓掌',
    x: -840,
    y: -1260,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Clap',
      ja_JP: '拍手',
      zh_CN: '鼓掌',
      zh_HK: '鼓掌',
      zh_TW: '鼓掌'
    }
  }, {
    stickerId: 'sticker_071',
    code: '送花花',
    x: 0,
    y: -1400,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Flower',
      ja_JP: 'お花',
      zh_CN: '送花花',
      zh_HK: '送花花',
      zh_TW: '送花花'
    }
  }, {
    stickerId: 'sticker_072',
    code: '比心',
    x: -140,
    y: -1400,
    width: 140,
    height: 140,
    locale: {
      en_US: 'FingerHeart',
      ja_JP: 'ハート',
      zh_CN: '比心',
      zh_HK: '比心',
      zh_TW: '比心'
    }
  }, {
    stickerId: 'sticker_073',
    code: '加一',
    x: -280,
    y: -1400,
    width: 140,
    height: 140,
    locale: {
      en_US: 'PlusOne',
      ja_JP: 'プラスワン',
      zh_CN: '加一',
      zh_HK: '加一',
      zh_TW: '加一'
    }
  }, {
    stickerId: 'sticker_074',
    code: '100分',
    x: -420,
    y: -1400,
    width: 140,
    height: 140,
    locale: {
      en_US: '100',
      ja_JP: '100分',
      zh_CN: '100分',
      zh_HK: '100分',
      zh_TW: '100分'
    }
  }, {
    stickerId: 'sticker_075',
    code: 'KPI',
    x: -560,
    y: -1400,
    width: 140,
    height: 140,
    locale: {
      en_US: 'KPI',
      ja_JP: 'KPI',
      zh_CN: 'KPI',
      zh_HK: 'KPI',
      zh_TW: 'KPI'
    }
  }, {
    stickerId: 'sticker_076',
    code: '对勾',
    x: -700,
    y: -1400,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Check',
      ja_JP: 'チェック',
      zh_CN: '对勾',
      zh_HK: '對勾',
      zh_TW: '對勾'
    }
  }, {
    stickerId: 'sticker_077',
    code: '爱心',
    x: -840,
    y: -1400,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Love',
      ja_JP: 'ハート',
      zh_CN: '爱心',
      zh_HK: '愛心',
      zh_TW: '愛心'
    }
  }, {
    stickerId: 'sticker_078',
    code: '可爱',
    x: 0,
    y: -1540,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Lovely',
      ja_JP: 'かわいい',
      zh_CN: '可爱',
      zh_HK: '可愛',
      zh_TW: '可愛'
    }
  }, {
    stickerId: 'sticker_079',
    code: '发呆',
    x: -140,
    y: -1540,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Dazed',
      ja_JP: 'ぼんやり',
      zh_CN: '发呆',
      zh_HK: '發呆',
      zh_TW: '發呆'
    }
  }, {
    stickerId: 'sticker_080',
    code: '老板',
    x: -280,
    y: -1540,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Boss',
      ja_JP: 'ボス',
      zh_CN: '老板',
      zh_HK: '老闆',
      zh_TW: '老闆'
    }
  }, {
    stickerId: 'sticker_081',
    code: '害羞',
    x: -420,
    y: -1540,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Shy',
      ja_JP: 'シャイ',
      zh_CN: '害羞',
      zh_HK: '害羞',
      zh_TW: '害羞'
    }
  }, {
    stickerId: 'sticker_082',
    code: '闭嘴',
    x: -560,
    y: -1540,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Silence',
      ja_JP: '沈黙',
      zh_CN: '闭嘴',
      zh_HK: '閉嘴',
      zh_TW: '閉嘴'
    }
  }, {
    stickerId: 'sticker_083',
    code: '睡',
    x: -700,
    y: -1540,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Sleepy',
      ja_JP: '眠い',
      zh_CN: '睡',
      zh_HK: '睡',
      zh_TW: '睡'
    }
  }, {
    stickerId: 'sticker_084',
    code: '大哭',
    x: -840,
    y: -1540,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Cry',
      ja_JP: '泣き',
      zh_CN: '大哭',
      zh_HK: '大哭',
      zh_TW: '大哭'
    }
  }, {
    stickerId: 'sticker_085',
    code: '尴尬',
    x: 0,
    y: -1680,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Awkward',
      ja_JP: '困る',
      zh_CN: '尴尬',
      zh_HK: '尴尬',
      zh_TW: '尴尬'
    }
  }, {
    stickerId: 'sticker_086',
    code: '调皮',
    x: -140,
    y: -1680,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Tongueout',
      ja_JP: 'てへ',
      zh_CN: '调皮',
      zh_HK: '調皮',
      zh_TW: '調皮'
    }
  }, {
    stickerId: 'sticker_087',
    code: '惊讶',
    x: -420,
    y: -1680,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Scowl',
      ja_JP: 'びっくり',
      zh_CN: '惊讶',
      zh_HK: '驚訝',
      zh_TW: '驚訝'
    }
  }, {
    stickerId: 'sticker_088',
    code: '流汗',
    x: -560,
    y: -1680,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Sweat',
      ja_JP: '汗',
      zh_CN: '流汗',
      zh_HK: '流汗',
      zh_TW: '流汗'
    }
  }, {
    stickerId: 'sticker_089',
    code: '广播',
    x: -700,
    y: -1680,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Shout',
      ja_JP: '叫ぶ',
      zh_CN: '广播',
      zh_HK: '廣播',
      zh_TW: '廣播'
    }
  }, {
    stickerId: 'sticker_090',
    code: '自信',
    x: -840,
    y: -1680,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Self-confident',
      ja_JP: '自信',
      zh_CN: '自信',
      zh_HK: '自信',
      zh_TW: '自信'
    }
  }, {
    stickerId: 'sticker_091',
    code: '你强',
    x: 0,
    y: -1820,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Awesome',
      ja_JP: 'すごい',
      zh_CN: '你强',
      zh_HK: '你强',
      zh_TW: '你强'
    }
  }, {
    stickerId: 'sticker_092',
    code: '怒吼',
    x: -140,
    y: -1820,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Pumped',
      ja_JP: '怒号',
      zh_CN: '怒吼',
      zh_HK: '怒吼',
      zh_TW: '怒吼'
    }
  }, {
    stickerId: 'sticker_093',
    code: '惊愕',
    x: -280,
    y: -1820,
    width: 140,
    height: 140,
    locale: {
      en_US: 'What?!',
      ja_JP: '驚き',
      zh_CN: '惊愕',
      zh_HK: '惊愕',
      zh_TW: '惊愕'
    }
  }, {
    stickerId: 'sticker_094',
    code: '快哭了',
    x: -420,
    y: -1820,
    width: 140,
    height: 140,
    locale: {
      en_US: 'TearingUp',
      ja_JP: '泣きそう',
      zh_CN: '快哭了',
      zh_HK: '快哭了',
      zh_TW: '快哭了'
    }
  }, {
    stickerId: 'sticker_095',
    code: '无聊',
    x: -560,
    y: -1820,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Bored',
      ja_JP: '退屈',
      zh_CN: '无聊',
      zh_HK: '無聊',
      zh_TW: '無聊'
    }
  }, {
    stickerId: 'sticker_096',
    code: '吐',
    x: -700,
    y: -1820,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Puke',
      ja_JP: '吐きそう',
      zh_CN: '吐',
      zh_HK: '吐',
      zh_TW: '吐'
    }
  }, {
    stickerId: 'sticker_097',
    code: '算账',
    x: -840,
    y: -1820,
    width: 140,
    height: 140,
    locale: {
      en_US: 'SoMuch',
      ja_JP: '計算中',
      zh_CN: '算账',
      zh_HK: '算账',
      zh_TW: '算账'
    }
  }, {
    stickerId: 'sticker_098',
    code: '晕',
    x: 0,
    y: -1960,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Dizzy',
      ja_JP: 'めまい',
      zh_CN: '晕',
      zh_HK: '暈',
      zh_TW: '暈'
    }
  }, {
    stickerId: 'sticker_099',
    code: '摸摸',
    x: -140,
    y: -1960,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Comfort',
      ja_JP: 'よしよし',
      zh_CN: '摸摸',
      zh_HK: '摸摸',
      zh_TW: '摸摸'
    }
  }, {
    stickerId: 'sticker_100',
    code: '飞吻',
    x: -280,
    y: -1960,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Blowkiss',
      ja_JP: '投げキッス',
      zh_CN: '飞吻',
      zh_HK: '飛吻',
      zh_TW: '飛吻'
    }
  }, {
    stickerId: 'sticker_101',
    code: '鄙视',
    x: -420,
    y: -1960,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Dislike',
      ja_JP: 'ガッカリ',
      zh_CN: '鄙视',
      zh_HK: '鄙視',
      zh_TW: '鄙視'
    }
  }, {
    stickerId: 'sticker_102',
    code: '嘘',
    x: -560,
    y: -1960,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Shhh',
      ja_JP: '静かに',
      zh_CN: '嘘',
      zh_HK: '嘘',
      zh_TW: '嘘'
    }
  }, {
    stickerId: 'sticker_103',
    code: '思考',
    x: -700,
    y: -1960,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Hmm…',
      ja_JP: 'うーん...',
      zh_CN: '思考',
      zh_HK: '思考',
      zh_TW: '思考'
    }
  }, {
    stickerId: 'sticker_104',
    code: '亲亲',
    x: -840,
    y: -1960,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Kiss',
      ja_JP: 'キッス',
      zh_CN: '亲亲',
      zh_HK: '親親',
      zh_TW: '親親'
    }
  }, {
    stickerId: 'sticker_105',
    code: '感冒',
    x: 0,
    y: -2100,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Pollution',
      ja_JP: '風邪',
      zh_CN: '感冒',
      zh_HK: '感冒',
      zh_TW: '感冒'
    }
  }, {
    stickerId: 'sticker_106',
    code: '对不起',
    x: -140,
    y: -2100,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Sorry',
      ja_JP: 'ごめん',
      zh_CN: '对不起',
      zh_HK: '對不起',
      zh_TW: '對不起'
    }
  }, {
    stickerId: 'sticker_107',
    code: '再见',
    x: -280,
    y: -2100,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Wave',
      ja_JP: 'さようなら',
      zh_CN: '再见',
      zh_HK: '再見',
      zh_TW: '再見'
    }
  }, {
    stickerId: 'sticker_108',
    code: '投降',
    x: -420,
    y: -2100,
    width: 140,
    height: 140,
    locale: {
      en_US: 'GiveUp',
      ja_JP: 'あきらめ',
      zh_CN: '投降',
      zh_HK: '投降',
      zh_TW: '投降'
    }
  }, {
    stickerId: 'sticker_109',
    code: '哼',
    x: -560,
    y: -2100,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Grumpy',
      ja_JP: 'ふん！',
      zh_CN: '哼',
      zh_HK: '哼',
      zh_TW: '哼'
    }
  }, {
    stickerId: 'sticker_110',
    code: '欠扁',
    x: -700,
    y: -2100,
    width: 140,
    height: 140,
    locale: {
      en_US: 'FaceSlap',
      ja_JP: 'ビンタ',
      zh_CN: '欠扁',
      zh_HK: '欠扁',
      zh_TW: '欠扁'
    }
  }, {
    stickerId: 'sticker_111',
    code: '坏笑',
    x: -840,
    y: -2100,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Trick',
      ja_JP: 'ニヤリ',
      zh_CN: '坏笑',
      zh_HK: '坏笑',
      zh_TW: '坏笑'
    }
  }, {
    stickerId: 'sticker_112',
    code: '拜托',
    x: 0,
    y: -2240,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Please',
      ja_JP: 'お願い',
      zh_CN: '拜托',
      zh_HK: '拜託',
      zh_TW: '拜託'
    }
  }, {
    stickerId: 'sticker_113',
    code: '可怜',
    x: -140,
    y: -2240,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Aww…',
      ja_JP: 'ウルウル...',
      zh_CN: '可怜',
      zh_HK: '可憐',
      zh_TW: '可憐'
    }
  }, {
    stickerId: 'sticker_114',
    code: '舒服',
    x: -280,
    y: -2240,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Relax',
      ja_JP: 'リラックス',
      zh_CN: '舒服',
      zh_HK: '舒服',
      zh_TW: '舒服'
    }
  }, {
    stickerId: 'sticker_115',
    code: '爱意',
    x: -420,
    y: -2240,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Romantic',
      ja_JP: 'ロマンチック',
      zh_CN: '爱意',
      zh_HK: '愛意',
      zh_TW: '愛意'
    }
  }, {
    stickerId: 'sticker_116',
    code: '财迷',
    x: -560,
    y: -2240,
    width: 140,
    height: 140,
    locale: {
      en_US: 'MoneyMoney',
      ja_JP: 'マネー',
      zh_CN: '财迷',
      zh_HK: '財迷',
      zh_TW: '財迷'
    }
  }, {
    stickerId: 'sticker_117',
    code: '迷惑',
    x: -700,
    y: -2240,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Puzzled',
      ja_JP: '困惑',
      zh_CN: '迷惑',
      zh_HK: '迷惑',
      zh_TW: '迷惑'
    }
  }, {
    stickerId: 'sticker_118',
    code: 'PK',
    x: -840,
    y: -2240,
    width: 140,
    height: 140,
    locale: {
      en_US: 'PK',
      ja_JP: 'PK',
      zh_CN: 'PK',
      zh_HK: 'PK',
      zh_TW: 'PK'
    }
  }, {
    stickerId: 'sticker_119',
    code: '委屈',
    x: 0,
    y: -2380,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Worried',
      ja_JP: 'モジモジ',
      zh_CN: '委屈',
      zh_HK: '委屈',
      zh_TW: '委屈'
    }
  }, {
    stickerId: 'sticker_120',
    code: '灵感',
    x: -140,
    y: -2380,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Idea',
      ja_JP: 'ひらめき',
      zh_CN: '灵感',
      zh_HK: '靈感',
      zh_TW: '靈感'
    }
  }, {
    stickerId: 'sticker_121',
    code: '天使',
    x: -280,
    y: -2380,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Angel',
      ja_JP: '天使',
      zh_CN: '天使',
      zh_HK: '天使',
      zh_TW: '天使'
    }
  }, {
    stickerId: 'sticker_122',
    code: '鬼脸',
    x: -420,
    y: -2380,
    width: 140,
    height: 140,
    locale: {
      en_US: 'SillyFace',
      ja_JP: '変顔',
      zh_CN: '鬼脸',
      zh_HK: '鬼臉',
      zh_TW: '鬼臉'
    }
  }, {
    stickerId: 'sticker_123',
    code: '凄凉',
    x: -560,
    y: -2380,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Phew',
      ja_JP: '寒い',
      zh_CN: '凄凉',
      zh_HK: '凄凉',
      zh_TW: '凄凉'
    }
  }, {
    stickerId: 'sticker_124',
    code: '郁闷',
    x: -700,
    y: -2380,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Tired',
      ja_JP: 'ゆううつ',
      zh_CN: '郁闷',
      zh_HK: '鬱悶',
      zh_TW: '鬱悶'
    }
  }, {
    stickerId: 'sticker_125',
    code: '吃瓜',
    x: -840,
    y: -2380,
    width: 140,
    height: 140,
    locale: {
      en_US: 'EatingMelon',
      ja_JP: 'スイカ',
      zh_CN: '吃瓜',
      zh_HK: '吃瓜',
      zh_TW: '吃瓜'
    }
  }, {
    stickerId: 'sticker_126',
    code: '嘿嘿',
    x: 0,
    y: -2520,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Smirk',
      ja_JP: 'へへ',
      zh_CN: '嘿嘿',
      zh_HK: '嘿嘿',
      zh_TW: '嘿嘿'
    }
  }, {
    stickerId: 'sticker_127',
    code: '抠鼻',
    x: -140,
    y: -2520,
    width: 140,
    height: 140,
    locale: {
      en_US: 'NosePick',
      ja_JP: '鼻をほじる',
      zh_CN: '抠鼻',
      zh_HK: '摳鼻',
      zh_TW: '摳鼻'
    }
  }, {
    stickerId: 'sticker_128',
    code: '呲牙',
    x: -280,
    y: -2520,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Grin',
      ja_JP: '笑',
      zh_CN: '呲牙',
      zh_HK: '呲牙',
      zh_TW: '呲牙'
    }
  }, {
    stickerId: 'sticker_129',
    code: '彩虹',
    x: -420,
    y: -2520,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Rainbow',
      ja_JP: 'レインボー',
      zh_CN: '彩虹',
      zh_HK: '彩虹',
      zh_TW: '彩虹'
    }
  }, {
    stickerId: 'sticker_130',
    code: '耶',
    x: -560,
    y: -2520,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Yeah',
      ja_JP: 'イエーイ',
      zh_CN: '耶',
      zh_HK: '耶',
      zh_TW: '耶'
    }
  }, {
    stickerId: 'sticker_131',
    code: '捂眼睛',
    x: -700,
    y: -2520,
    width: 140,
    height: 140,
    locale: {
      en_US: 'CannotLook',
      ja_JP: '目隠し',
      zh_CN: '捂眼睛',
      zh_HK: '捂眼睛',
      zh_TW: '捂眼睛'
    }
  }, {
    stickerId: 'sticker_132',
    code: '推眼镜',
    x: -840,
    y: -2520,
    width: 140,
    height: 140,
    locale: {
      en_US: 'PushGlasses',
      ja_JP: 'メガネ',
      zh_CN: '推眼镜',
      zh_HK: '推眼镜',
      zh_TW: '推眼镜'
    }
  }, {
    stickerId: 'sticker_133',
    code: '暗中观察',
    x: 0,
    y: -2660,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Peep',
      ja_JP: 'こっそり',
      zh_CN: '暗中观察',
      zh_HK: '暗中觀察',
      zh_TW: '暗中觀察'
    }
  }, {
    stickerId: 'sticker_134',
    code: '开心',
    x: -140,
    y: -2660,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Happy',
      ja_JP: 'ハッピー',
      zh_CN: '开心',
      zh_HK: '開心',
      zh_TW: '開心'
    }
  }, {
    stickerId: 'sticker_135',
    code: '惊喜',
    x: -280,
    y: -2660,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Surprised',
      ja_JP: 'ビックリ',
      zh_CN: '惊喜',
      zh_HK: '驚喜',
      zh_TW: '驚喜'
    }
  }, {
    stickerId: 'sticker_136',
    code: '回头',
    x: -420,
    y: -2660,
    width: 140,
    height: 140,
    locale: {
      en_US: 'LookBack',
      ja_JP: '振り返り',
      zh_CN: '回头',
      zh_HK: '回頭',
      zh_TW: '回頭'
    }
  }, {
    stickerId: 'sticker_137',
    code: '发怒',
    x: -560,
    y: -2660,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Steamed',
      ja_JP: '怒り',
      zh_CN: '发怒',
      zh_HK: '發怒',
      zh_TW: '發怒'
    }
  }, {
    stickerId: 'sticker_138',
    code: '忍者',
    x: -700,
    y: -2660,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Sneaky',
      ja_JP: '忍者',
      zh_CN: '忍者',
      zh_HK: '忍者',
      zh_TW: '忍者'
    }
  }, {
    stickerId: 'sticker_139',
    code: '衰',
    x: -840,
    y: -2660,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Grr',
      ja_JP: '困惑',
      zh_CN: '衰',
      zh_HK: '衰',
      zh_TW: '衰'
    }
  }, {
    stickerId: 'sticker_140',
    code: '脑暴',
    x: 0,
    y: -2800,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Brainstorming',
      ja_JP: 'ブレインストーム',
      zh_CN: '脑暴',
      zh_HK: '腦暴',
      zh_TW: '腦暴'
    }
  }, {
    stickerId: 'sticker_141',
    code: '冷笑',
    x: -140,
    y: -2800,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Distressed',
      ja_JP: '寒い',
      zh_CN: '冷笑',
      zh_HK: '冷笑',
      zh_TW: '冷笑'
    }
  }, {
    stickerId: 'sticker_142',
    code: '黑眼圈',
    x: -280,
    y: -2800,
    width: 140,
    height: 140,
    locale: {
      en_US: 'DarkCircle',
      ja_JP: '目のクマ',
      zh_CN: '黑眼圈',
      zh_HK: '黑眼圈',
      zh_TW: '黑眼圈'
    }
  }, {
    stickerId: 'sticker_143',
    code: '恭喜',
    x: -420,
    y: -2800,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Congrats',
      ja_JP: '祝う',
      zh_CN: '恭喜',
      zh_HK: '恭喜',
      zh_TW: '恭喜'
    }
  }, {
    stickerId: 'sticker_144',
    code: '费解',
    x: -560,
    y: -2800,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Confuse',
      ja_JP: '難しい',
      zh_CN: '费解',
      zh_HK: '不解',
      zh_TW: '不解'
    }
  }, {
    stickerId: 'sticker_145',
    code: '收到',
    x: -700,
    y: -2800,
    width: 140,
    height: 140,
    locale: {
      en_US: 'RogerThat',
      ja_JP: '了解',
      zh_CN: '收到',
      zh_HK: '收到',
      zh_TW: '收到'
    }
  }, {
    stickerId: 'sticker_146',
    code: '炸弹',
    x: -840,
    y: -2800,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Uh-Oh',
      ja_JP: '爆弾',
      zh_CN: '炸弹',
      zh_HK: '炸彈',
      zh_TW: '炸彈'
    }
  }, {
    stickerId: 'sticker_147',
    code: '白眼',
    x: 0,
    y: -2940,
    width: 140,
    height: 140,
    locale: {
      en_US: 'RollEyes',
      ja_JP: '白目',
      zh_CN: '白眼',
      zh_HK: '白眼',
      zh_TW: '白眼'
    }
  }, {
    stickerId: 'sticker_148',
    code: '一团乱麻',
    x: -140,
    y: -2940,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Overwhelmed',
      ja_JP: '滅茶苦茶',
      zh_CN: '一团乱麻',
      zh_HK: '一團亂麻',
      zh_TW: '一團亂麻'
    }
  }, {
    stickerId: 'sticker_149',
    code: '无奈',
    x: -280,
    y: -2940,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Disappointed',
      ja_JP: '失望',
      zh_CN: '无奈',
      zh_HK: '無奈',
      zh_TW: '無奈'
    }
  }, {
    stickerId: 'sticker_150',
    code: '敲打',
    x: -420,
    y: -2940,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Hammer',
      ja_JP: 'ハンマー',
      zh_CN: '敲打',
      zh_HK: '敲打',
      zh_TW: '敲打'
    }
  }, {
    stickerId: 'sticker_151',
    code: '专注',
    x: -560,
    y: -2940,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Concentrate',
      ja_JP: '集中',
      zh_CN: '专注',
      zh_HK: '專注',
      zh_TW: '專注'
    }
  }, {
    stickerId: 'sticker_152',
    code: '忙疯了',
    x: -700,
    y: -2940,
    width: 140,
    height: 140,
    locale: {
      en_US: 'CrazyBusy',
      ja_JP: '超忙しい',
      zh_CN: '忙疯了',
      zh_HK: '忙瘋了',
      zh_TW: '忙瘋了'
    }
  }, {
    stickerId: 'sticker_153',
    code: '鞠躬',
    x: -840,
    y: -2940,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Bow',
      ja_JP: 'お願い',
      zh_CN: '鞠躬',
      zh_HK: '鞠躬',
      zh_TW: '鞠躬'
    }
  }, {
    stickerId: 'sticker_154',
    code: '摊手',
    x: 0,
    y: -3080,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Smugshrug',
      ja_JP: '仕方ない',
      zh_CN: '摊手',
      zh_HK: '攤手',
      zh_TW: '攤手'
    }
  }, {
    stickerId: 'sticker_155',
    code: '抱抱',
    x: -140,
    y: -3080,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Hug',
      ja_JP: 'ハグ',
      zh_CN: '抱抱',
      zh_HK: '抱抱',
      zh_TW: '抱抱'
    }
  }, {
    stickerId: 'sticker_156',
    code: '举手',
    x: -280,
    y: -3080,
    width: 140,
    height: 140,
    locale: {
      en_US: 'RaiseHand',
      ja_JP: '手を挙げる',
      zh_CN: '举手',
      zh_HK: '舉手',
      zh_TW: '舉手'
    }
  }, {
    stickerId: 'sticker_157',
    code: '跪了',
    x: -420,
    y: -3080,
    width: 140,
    height: 140,
    locale: {
      en_US: 'YouWin',
      ja_JP: '正座',
      zh_CN: '跪了',
      zh_HK: '跪了',
      zh_TW: '跪了'
    }
  }, {
    stickerId: 'sticker_158',
    code: '猫咪',
    x: -560,
    y: -3080,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Kitty',
      ja_JP: '猫',
      zh_CN: '猫咪',
      zh_HK: '貓',
      zh_TW: '貓'
    }
  }, {
    stickerId: 'sticker_159',
    code: '二哈',
    x: -700,
    y: -3080,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Doggy',
      ja_JP: 'ハスキー犬',
      zh_CN: '二哈',
      zh_HK: '二哈',
      zh_TW: '二哈'
    }
  }, {
    stickerId: 'sticker_160',
    code: '三多',
    x: 0,
    y: -3220,
    width: 140,
    height: 140,
    locale: {
      en_US: 'SanDuo',
      ja_JP: 'SanDuo',
      zh_CN: '三多',
      zh_HK: '三多',
      zh_TW: '三多'
    }
  }, {
    stickerId: 'sticker_161',
    code: '承让',
    x: -140,
    y: -3220,
    width: 140,
    height: 140,
    locale: {
      en_US: 'LetMeWin',
      ja_JP: 'お願い',
      zh_CN: '承让',
      zh_HK: '承讓',
      zh_TW: '承讓'
    }
  }, {
    stickerId: 'sticker_162',
    code: '撒花',
    x: -280,
    y: -3220,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Celebration',
      ja_JP: '花吹雪',
      zh_CN: '撒花',
      zh_HK: '撒花',
      zh_TW: '撒花'
    }
  }, {
    stickerId: 'sticker_163',
    code: '邮件',
    x: -420,
    y: -3220,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Mail',
      ja_JP: 'メール',
      zh_CN: '邮件',
      zh_HK: '郵件',
      zh_TW: '郵件'
    }
  }, {
    stickerId: 'sticker_164',
    code: '文档',
    x: -560,
    y: -3220,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Document',
      ja_JP: '文書',
      zh_CN: '文档',
      zh_HK: '文檔',
      zh_TW: '文檔'
    }
  }, {
    stickerId: 'sticker_165',
    code: '演示',
    x: -700,
    y: -3220,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Presentation',
      ja_JP: 'プレゼン',
      zh_CN: '演示',
      zh_HK: '演示',
      zh_TW: '演示'
    }
  }, {
    stickerId: 'sticker_166',
    code: '表格',
    x: -840,
    y: -3220,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Sheet',
      ja_JP: '表計算',
      zh_CN: '表格',
      zh_HK: '表格',
      zh_TW: '表格'
    }
  }, {
    stickerId: 'sticker_167',
    code: '生日快乐',
    x: 0,
    y: -3360,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Birthday',
      ja_JP: '誕生日',
      zh_CN: '生日快乐',
      zh_HK: '生日快樂',
      zh_TW: '生日快樂'
    }
  }, {
    stickerId: 'sticker_168',
    code: '心碎',
    x: -140,
    y: -3360,
    width: 140,
    height: 140,
    locale: {
      en_US: 'BrokenHeart',
      ja_JP: 'ブロークンハート',
      zh_CN: '心碎',
      zh_HK: '心碎',
      zh_TW: '心碎'
    }
  }, {
    stickerId: 'sticker_169',
    code: '红包',
    x: -280,
    y: -3360,
    width: 140,
    height: 140,
    locale: {
      en_US: 'RedPacket',
      ja_JP: 'レッドポケット',
      zh_CN: '红包',
      zh_HK: '紅包',
      zh_TW: '紅包'
    }
  }, {
    stickerId: 'sticker_170',
    code: '嘴唇',
    x: -420,
    y: -3360,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Lips',
      ja_JP: 'リップ',
      zh_CN: '嘴唇',
      zh_HK: '嘴唇',
      zh_TW: '嘴唇'
    }
  }, {
    stickerId: 'sticker_171',
    code: '鲜花',
    x: -560,
    y: -3360,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Rose',
      ja_JP: 'ローズ',
      zh_CN: '鲜花',
      zh_HK: '鮮花',
      zh_TW: '鮮花'
    }
  }, {
    stickerId: 'sticker_172',
    code: '残花',
    x: -700,
    y: -3360,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Wilted',
      ja_JP: '散る',
      zh_CN: '残花',
      zh_HK: '殘花',
      zh_TW: '殘花'
    }
  }, {
    stickerId: 'sticker_173',
    code: '干杯',
    x: -840,
    y: -3360,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Cheers',
      ja_JP: '乾杯',
      zh_CN: '干杯',
      zh_HK: '乾杯',
      zh_TW: '乾杯'
    }
  }, {
    stickerId: 'sticker_174',
    code: '出差',
    x: 0,
    y: -3500,
    width: 140,
    height: 140,
    locale: {
      en_US: 'BusinessTrip',
      ja_JP: '出張',
      zh_CN: '出差',
      zh_HK: '出差',
      zh_TW: '出差'
    }
  }, {
    stickerId: 'sticker_175',
    code: '时间',
    x: -140,
    y: -3500,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Time',
      ja_JP: '時間',
      zh_CN: '时间',
      zh_HK: '時間',
      zh_TW: '時間'
    }
  }, {
    stickerId: 'sticker_176',
    code: '福',
    x: -280,
    y: -3500,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Luck',
      ja_JP: '福',
      zh_CN: '福',
      zh_HK: '福',
      zh_TW: '福'
    }
  }, {
    stickerId: 'sticker_177',
    code: '月饼',
    x: -420,
    y: -3500,
    width: 140,
    height: 140,
    locale: {
      en_US: 'MoonCake',
      ja_JP: '月餅',
      zh_CN: '月饼',
      zh_HK: '月餅',
      zh_TW: '月餅'
    }
  }, {
    stickerId: 'sticker_178',
    code: '礼物',
    x: -560,
    y: -3500,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Present',
      ja_JP: 'プレゼント',
      zh_CN: '礼物',
      zh_HK: '禮物',
      zh_TW: '禮物'
    }
  }, {
    stickerId: 'sticker_179',
    code: '幼苗',
    x: -700,
    y: -3500,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Seedling',
      ja_JP: '苗',
      zh_CN: '幼苗',
      zh_HK: '幼苗',
      zh_TW: '幼苗'
    }
  }, {
    stickerId: 'sticker_180',
    code: '烟花',
    x: -840,
    y: -3500,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Fireworks',
      ja_JP: '花火',
      zh_CN: '烟花',
      zh_HK: '煙花',
      zh_TW: '煙花'
    }
  }, {
    stickerId: 'sticker_181',
    code: '灯笼',
    x: 0,
    y: -3640,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Lantern',
      ja_JP: 'ランタン',
      zh_CN: '灯笼',
      zh_HK: '燈籠',
      zh_TW: '燈籠'
    }
  }, {
    stickerId: 'sticker_182',
    code: '爆竹',
    x: -140,
    y: -3640,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Firecracker',
      ja_JP: '爆竹',
      zh_CN: '爆竹',
      zh_HK: '爆竹',
      zh_TW: '爆竹'
    }
  }, {
    stickerId: 'sticker_183',
    code: '鸡腿',
    x: -280,
    y: -3640,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Chicken',
      ja_JP: '鶏の脚',
      zh_CN: '鸡腿',
      zh_HK: '雞腿',
      zh_TW: '雞腿'
    }
  }, {
    stickerId: 'sticker_184',
    code: '高铁',
    x: -420,
    y: -3640,
    width: 140,
    height: 140,
    locale: {
      en_US: 'High-speed rail',
      ja_JP: '高速レール',
      zh_CN: '高铁',
      zh_HK: '高鐵',
      zh_TW: '高鐵'
    }
  }, {
    stickerId: 'sticker_185',
    code: '三连',
    x: -560,
    y: -3640,
    width: 140,
    height: 140,
    locale: {
      en_US: 'ThreeCompanies',
      ja_JP: '3社',
      zh_CN: '三连',
      zh_HK: '三連',
      zh_TW: '三連'
    }
  }, {
    stickerId: 'sticker_186',
    code: 'OKR',
    x: -700,
    y: -3640,
    width: 140,
    height: 140,
    locale: {
      en_US: 'OKR',
      ja_JP: 'OKR',
      zh_CN: 'OKR',
      zh_HK: 'OKR',
      zh_TW: 'OKR'
    }
  }, {
    stickerId: 'sticker_187',
    code: 'Done',
    x: -840,
    y: -3640,
    width: 140,
    height: 140,
    locale: {
      en_US: 'Done',
      ja_JP: 'Done',
      zh_CN: 'Done',
      zh_HK: 'Done',
      zh_TW: 'Done'
    }
  }]
};
exports.DEFAULT_STICKERS = DEFAULT_STICKERS;
const {
  MOVE_UPWARD,
  MOVE_DOWNWARD
} = _everUtils.SELECTION_TRIGGER;
const MOVE_SET = {
  [MOVE_UPWARD]: {
    queryRange: 'getUpsideRange',
    getTargetMethod: 'getPreviousNode',
    moveMethod: 'moveToEndOfNode'
  },
  [MOVE_DOWNWARD]: {
    queryRange: 'getDownsideRange',
    getTargetMethod: 'getNextNode',
    moveMethod: 'moveToStartOfNode'
  }
}; // 节点类型

exports.MOVE_SET = MOVE_SET;
const STICKER_NODE_TYPE = _models.default.TYPE; // 表情选择面板样式

exports.STICKER_NODE_TYPE = STICKER_NODE_TYPE;
const STICKER_PANEL_STYLE = {
  border: '1px solid rgba(126, 134, 142, 0.16)',
  boxShadow: '0px 8px 24px rgba(0, 0, 0, 0.16)',
  borderRadius: '8px',
  overflow: 'hidden'
};
exports.STICKER_PANEL_STYLE = STICKER_PANEL_STYLE;
const STICKER_SZ_IN_PANEL = 24; // 单行显示的表情个数

exports.STICKER_SZ_IN_PANEL = STICKER_SZ_IN_PANEL;
const STICKER_PANEL_COLUMMS = 7;
exports.STICKER_PANEL_COLUMMS = STICKER_PANEL_COLUMMS;
const DATA_STICKER = 'data-sticker'; // 表情面板高度

exports.DATA_STICKER = DATA_STICKER;
const STICKER_PANEL_HEIGHT = 256;
exports.STICKER_PANEL_HEIGHT = STICKER_PANEL_HEIGHT;
const STICKER_PLACEHOLDER_DATA = {
  code: '',
  width: DEFAULT_SZ,
  height: DEFAULT_SZ
};
exports.STICKER_PLACEHOLDER_DATA = STICKER_PLACEHOLDER_DATA;
