// @ali/4ever-plugin-marks/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  mobileBiFactory: true,
  mobileZhiFactory: true,
  moFactory: true,
  zhiFactory: true,
  biFactory: true,
  createSerializationPlugin: true,
  createServerSerializationPlugin: true,
  createMarkFactory: true,
  createMarkFactoryMobile: true,
  MARK_TOGGLE: true,
  MarksComponent: true,
  setStyleWithMark: true,
  marksToStyle: true,
  MarksOptionContext: true,
  markToAttr: true,
  marksToAttrs: true,
  attrsToMarks: true
};
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.moFactory;
  }
});
Object.defineProperty(exports, "zhiFactory", {
  enumerable: true,
  get: function () {
    return _index2.zhiFactory;
  }
});
Object.defineProperty(exports, "biFactory", {
  enumerable: true,
  get: function () {
    return _index3.biFactory;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _serializer.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _serializer.createServerSerializationPlugin;
  }
});
Object.defineProperty(exports, "createMarkFactory", {
  enumerable: true,
  get: function () {
    return _createMarkFactory.createMarkFactory;
  }
});
Object.defineProperty(exports, "createMarkFactoryMobile", {
  enumerable: true,
  get: function () {
    return _mobile.createMarkFactoryMobile;
  }
});
Object.defineProperty(exports, "MARK_TOGGLE", {
  enumerable: true,
  get: function () {
    return _actions.MARK_TOGGLE;
  }
});
Object.defineProperty(exports, "MarksComponent", {
  enumerable: true,
  get: function () {
    return _marks.MarksComponent;
  }
});
Object.defineProperty(exports, "setStyleWithMark", {
  enumerable: true,
  get: function () {
    return _marks.setStyleWithMark;
  }
});
Object.defineProperty(exports, "marksToStyle", {
  enumerable: true,
  get: function () {
    return _marks.marksToStyle;
  }
});
Object.defineProperty(exports, "MarksOptionContext", {
  enumerable: true,
  get: function () {
    return _marks.MarksOptionContext;
  }
});
Object.defineProperty(exports, "markToAttr", {
  enumerable: true,
  get: function () {
    return _convert.markToAttr;
  }
});
Object.defineProperty(exports, "marksToAttrs", {
  enumerable: true,
  get: function () {
    return _convert.marksToAttrs;
  }
});
Object.defineProperty(exports, "attrsToMarks", {
  enumerable: true,
  get: function () {
    return _convert.attrsToMarks;
  }
});
exports.mobileZhiFactory = exports.mobileBiFactory = void 0;

var _index = require("./mo/index");

var _index2 = require("./zhi/index");

var _index3 = require("./bi/index");

var _index4 = require("./mo/models/index");

Object.keys(_index4).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _index4[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _index4[key];
    }
  });
});

var _serializer = _interopRequireWildcard(require("./mo/serializer"));

var _createMarkFactory = require("./utils/createMarkFactory");

var _mobile = require("./utils/createMarkFactory/mobile");

var _actions = require("./utils/createMarkFactory/actions");

var _marks = require("./components/marks");

var _convert = require("./utils/convert");

const mobileBiFactory = _index3.biFactory;
exports.mobileBiFactory = mobileBiFactory;
const mobileZhiFactory = _index2.zhiFactory; // facotry

exports.mobileZhiFactory = mobileZhiFactory;
