// @ali/4ever-mo/dist-node/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  pluginUtils: true,
  BasicPlugin: true,
  Marks: true,
  Serializer: true,
  moSerializer: true,
  createMoPlugins: true,
  createCustomizedMoPlugins: true,
  plugins: true,
  ParagraphPlugin: true,
  BrPlugin: true,
  CodePlugin: true,
  EmbedPlugin: true,
  HrPlugin: true,
  ImagePlugin: true,
  LinkPlugin: true,
  ListPlugin: true,
  MentionPlugin: true,
  OnlineVideoPlugin: true,
  PlaceholderPlugin: true,
  InlineCodePlugin: true,
  ReferPlugin: true,
  TablePlugin: true,
  TagPlugin: true,
  RefBlockPlugin: true,
  BoldMark: true,
  CapsMark: true,
  ColorMark: true,
  FontsMark: true,
  HighlightMark: true,
  InlineCodeMark: true,
  ItalicMark: true,
  SmallCapsMark: true,
  StrikeMark: true,
  SzMark: true,
  UnderlineMark: true,
  VertAlignMark: true,
  ShdMark: true,
  Common: true,
  Interfaces: true,
  State: true,
  DATA_CLIPBORAD_CANGJIE: true,
  SLASH_PLACEHOLDER: true,
  HtmlParser: true,
  Card: true,
  Tag: true,
  Embed: true,
  Code: true,
  Heading: true,
  Hr: true,
  Image: true,
  Link: true,
  Mention: true,
  InlineCode: true,
  Paragraph: true,
  Table: true,
  TableRow: true,
  TableCell: true,
  OnlineVideo: true,
  Placeholder: true,
  InlinePlaceholder: true,
  Refer: true,
  Sticker: true,
  Container: true,
  TemplateButton: true,
  ColorBlocks: true,
  schema: true,
  strictSchema: true,
  inlineTypes: true,
  blockTypes: true,
  createStrictSchema: true,
  ServerSerializer: true,
  createServerMoPlugins: true,
  serverPlugins: true,
  moServerSerializer: true,
  VERSION: true,
  KEY_OF_VERSION: true,
  DEFAULT_VERSION: true,
  migrate: true,
  getVersionFromCP: true,
  getVersionFromDocument: true,
  assertHasMigrations: true,
  normalizeMigrateError: true,
  MissingMigrationError: true,
  migrations: true,
  createMigration: true,
  corrections: true
};
Object.defineProperty(exports, "Serializer", {
  enumerable: true,
  get: function () {
    return _serializer.default;
  }
});
Object.defineProperty(exports, "moSerializer", {
  enumerable: true,
  get: function () {
    return _serializer.moSerializer;
  }
});
Object.defineProperty(exports, "createMoPlugins", {
  enumerable: true,
  get: function () {
    return _createMoPlugins.default;
  }
});
Object.defineProperty(exports, "createCustomizedMoPlugins", {
  enumerable: true,
  get: function () {
    return _createCustomizedMoPlugins.default;
  }
});
Object.defineProperty(exports, "plugins", {
  enumerable: true,
  get: function () {
    return _plugins.default;
  }
});
Object.defineProperty(exports, "ParagraphPlugin", {
  enumerable: true,
  get: function () {
    return _plugins.paragraph;
  }
});
Object.defineProperty(exports, "BrPlugin", {
  enumerable: true,
  get: function () {
    return _plugins.br;
  }
});
Object.defineProperty(exports, "CodePlugin", {
  enumerable: true,
  get: function () {
    return _plugins.code;
  }
});
Object.defineProperty(exports, "EmbedPlugin", {
  enumerable: true,
  get: function () {
    return _plugins.embed;
  }
});
Object.defineProperty(exports, "HrPlugin", {
  enumerable: true,
  get: function () {
    return _plugins.hr;
  }
});
Object.defineProperty(exports, "ImagePlugin", {
  enumerable: true,
  get: function () {
    return _plugins.image;
  }
});
Object.defineProperty(exports, "LinkPlugin", {
  enumerable: true,
  get: function () {
    return _plugins.link;
  }
});
Object.defineProperty(exports, "ListPlugin", {
  enumerable: true,
  get: function () {
    return _plugins.list;
  }
});
Object.defineProperty(exports, "MentionPlugin", {
  enumerable: true,
  get: function () {
    return _plugins.mention;
  }
});
Object.defineProperty(exports, "OnlineVideoPlugin", {
  enumerable: true,
  get: function () {
    return _plugins.onlineVideo;
  }
});
Object.defineProperty(exports, "PlaceholderPlugin", {
  enumerable: true,
  get: function () {
    return _plugins.placeholder;
  }
});
Object.defineProperty(exports, "InlineCodePlugin", {
  enumerable: true,
  get: function () {
    return _plugins.inlineCode;
  }
});
Object.defineProperty(exports, "ReferPlugin", {
  enumerable: true,
  get: function () {
    return _plugins.refer;
  }
});
Object.defineProperty(exports, "TablePlugin", {
  enumerable: true,
  get: function () {
    return _plugins.table;
  }
});
Object.defineProperty(exports, "TagPlugin", {
  enumerable: true,
  get: function () {
    return _plugins.tagWithFormula;
  }
});
Object.defineProperty(exports, "RefBlockPlugin", {
  enumerable: true,
  get: function () {
    return _plugins.refblock;
  }
});
Object.defineProperty(exports, "BoldMark", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.Bold;
  }
});
Object.defineProperty(exports, "CapsMark", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.Caps;
  }
});
Object.defineProperty(exports, "ColorMark", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.Color;
  }
});
Object.defineProperty(exports, "FontsMark", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.Fonts;
  }
});
Object.defineProperty(exports, "HighlightMark", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.Highlight;
  }
});
Object.defineProperty(exports, "InlineCodeMark", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.InlineCode;
  }
});
Object.defineProperty(exports, "ItalicMark", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.Italic;
  }
});
Object.defineProperty(exports, "SmallCapsMark", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.SmallCaps;
  }
});
Object.defineProperty(exports, "StrikeMark", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.Strike;
  }
});
Object.defineProperty(exports, "SzMark", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.Sz;
  }
});
Object.defineProperty(exports, "UnderlineMark", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.Underline;
  }
});
Object.defineProperty(exports, "VertAlignMark", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.VertAlign;
  }
});
Object.defineProperty(exports, "ShdMark", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.Shd;
  }
});
Object.defineProperty(exports, "Common", {
  enumerable: true,
  get: function () {
    return _everCangjie.MoCommon;
  }
});
Object.defineProperty(exports, "Interfaces", {
  enumerable: true,
  get: function () {
    return _everCangjie.MoInterfaces;
  }
});
Object.defineProperty(exports, "State", {
  enumerable: true,
  get: function () {
    return _everCangjie.MoState;
  }
});
Object.defineProperty(exports, "DATA_CLIPBORAD_CANGJIE", {
  enumerable: true,
  get: function () {
    return _everCangjie.DATA_CLIPBORAD_CANGJIE;
  }
});
Object.defineProperty(exports, "SLASH_PLACEHOLDER", {
  enumerable: true,
  get: function () {
    return _everCangjie.SLASH_PLACEHOLDER;
  }
});
Object.defineProperty(exports, "HtmlParser", {
  enumerable: true,
  get: function () {
    return _everCangjie.HtmlParser;
  }
});
Object.defineProperty(exports, "Card", {
  enumerable: true,
  get: function () {
    return _everPluginCard.Card;
  }
});
Object.defineProperty(exports, "Tag", {
  enumerable: true,
  get: function () {
    return _everPluginTag.Tag;
  }
});
Object.defineProperty(exports, "Embed", {
  enumerable: true,
  get: function () {
    return _everPluginEmbed.Embed;
  }
});
Object.defineProperty(exports, "Code", {
  enumerable: true,
  get: function () {
    return _everPluginCode.Code;
  }
});
Object.defineProperty(exports, "Heading", {
  enumerable: true,
  get: function () {
    return _everPluginHeading.Heading;
  }
});
Object.defineProperty(exports, "Hr", {
  enumerable: true,
  get: function () {
    return _everPluginHr.Hr;
  }
});
Object.defineProperty(exports, "Image", {
  enumerable: true,
  get: function () {
    return _everPluginImage.Image;
  }
});
Object.defineProperty(exports, "Link", {
  enumerable: true,
  get: function () {
    return _everPluginLink.Link;
  }
});
Object.defineProperty(exports, "Mention", {
  enumerable: true,
  get: function () {
    return _everPluginMention.Mention;
  }
});
Object.defineProperty(exports, "InlineCode", {
  enumerable: true,
  get: function () {
    return _everPluginInlineCode.InlineCode;
  }
});
Object.defineProperty(exports, "Paragraph", {
  enumerable: true,
  get: function () {
    return _everPluginParagraph.Paragraph;
  }
});
Object.defineProperty(exports, "Table", {
  enumerable: true,
  get: function () {
    return _everPluginTable.Table;
  }
});
Object.defineProperty(exports, "TableRow", {
  enumerable: true,
  get: function () {
    return _everPluginTable.TableRow;
  }
});
Object.defineProperty(exports, "TableCell", {
  enumerable: true,
  get: function () {
    return _everPluginTable.TableCell;
  }
});
Object.defineProperty(exports, "OnlineVideo", {
  enumerable: true,
  get: function () {
    return _everPluginOnlineVideo.OnlineVideo;
  }
});
Object.defineProperty(exports, "Placeholder", {
  enumerable: true,
  get: function () {
    return _everPluginPlaceholder.Placeholder;
  }
});
Object.defineProperty(exports, "InlinePlaceholder", {
  enumerable: true,
  get: function () {
    return _everPluginInlinePlaceholder.InlinePlaceholder;
  }
});
Object.defineProperty(exports, "Refer", {
  enumerable: true,
  get: function () {
    return _everPluginRefer.Refer;
  }
});
Object.defineProperty(exports, "Sticker", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.Sticker;
  }
});
Object.defineProperty(exports, "Container", {
  enumerable: true,
  get: function () {
    return _everPluginContainer.Container;
  }
});
Object.defineProperty(exports, "TemplateButton", {
  enumerable: true,
  get: function () {
    return _everPluginTemplateButton.TemplateButton;
  }
});
Object.defineProperty(exports, "ColorBlocks", {
  enumerable: true,
  get: function () {
    return _everPluginColorBlocks.ColorBlocks;
  }
});
Object.defineProperty(exports, "schema", {
  enumerable: true,
  get: function () {
    return _schema.default;
  }
});
Object.defineProperty(exports, "strictSchema", {
  enumerable: true,
  get: function () {
    return _strictSchema.default;
  }
});
Object.defineProperty(exports, "inlineTypes", {
  enumerable: true,
  get: function () {
    return _strictSchema.inlineTypes;
  }
});
Object.defineProperty(exports, "blockTypes", {
  enumerable: true,
  get: function () {
    return _strictSchema.blockTypes;
  }
});
Object.defineProperty(exports, "createStrictSchema", {
  enumerable: true,
  get: function () {
    return _strictSchema.createStrictSchemaOld;
  }
});
Object.defineProperty(exports, "ServerSerializer", {
  enumerable: true,
  get: function () {
    return _serializerValueJsonML.default;
  }
});
Object.defineProperty(exports, "createServerMoPlugins", {
  enumerable: true,
  get: function () {
    return _serializerValueJsonML.createServerMoPlugins;
  }
});
Object.defineProperty(exports, "serverPlugins", {
  enumerable: true,
  get: function () {
    return _serializerValueJsonML.serverPlugins;
  }
});
Object.defineProperty(exports, "moServerSerializer", {
  enumerable: true,
  get: function () {
    return _serializerValueJsonML.moServerSerializer;
  }
});
Object.defineProperty(exports, "VERSION", {
  enumerable: true,
  get: function () {
    return _version.VERSION;
  }
});
Object.defineProperty(exports, "KEY_OF_VERSION", {
  enumerable: true,
  get: function () {
    return _version.KEY_OF_VERSION;
  }
});
Object.defineProperty(exports, "DEFAULT_VERSION", {
  enumerable: true,
  get: function () {
    return _version.DEFAULT_VERSION;
  }
});
Object.defineProperty(exports, "migrate", {
  enumerable: true,
  get: function () {
    return _migrate.default;
  }
});
Object.defineProperty(exports, "getVersionFromCP", {
  enumerable: true,
  get: function () {
    return _migrate.getVersionFromCP;
  }
});
Object.defineProperty(exports, "getVersionFromDocument", {
  enumerable: true,
  get: function () {
    return _migrate.getVersionFromDocument;
  }
});
Object.defineProperty(exports, "assertHasMigrations", {
  enumerable: true,
  get: function () {
    return _migrate.assertHasMigrations;
  }
});
Object.defineProperty(exports, "normalizeMigrateError", {
  enumerable: true,
  get: function () {
    return _migrate.normalizeError;
  }
});
Object.defineProperty(exports, "MissingMigrationError", {
  enumerable: true,
  get: function () {
    return _migrate.MissingMigrationError;
  }
});
Object.defineProperty(exports, "migrations", {
  enumerable: true,
  get: function () {
    return _migrations.default;
  }
});
Object.defineProperty(exports, "createMigration", {
  enumerable: true,
  get: function () {
    return _createMigration.default;
  }
});
Object.defineProperty(exports, "corrections", {
  enumerable: true,
  get: function () {
    return _corrections.default;
  }
});
exports.default = exports.Marks = exports.BasicPlugin = exports.pluginUtils = void 0;

var _serializer = _interopRequireWildcard(require("./serializer"));

var _createMoPlugins = _interopRequireDefault(require("./createMoPlugins"));

var _createCustomizedMoPlugins = _interopRequireDefault(require("./createCustomizedMoPlugins"));

var _plugins = _interopRequireWildcard(require("./plugins"));

var _utils = require("./plugins/utils");

var _everPluginMarks = require("@ali/4ever-plugin-marks");

var _everCangjie = require("@ali/4ever-cangjie");

var _type = require("./type");

Object.keys(_type).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _type[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _type[key];
    }
  });
});

var _everPluginCard = require("@ali/4ever-plugin-card");

var _everPluginTag = require("@ali/4ever-plugin-tag");

var _everPluginEmbed = require("@ali/4ever-plugin-embed");

var _everPluginCode = require("@ali/4ever-plugin-code");

var _everPluginHeading = require("@ali/4ever-plugin-heading");

var _everPluginHr = require("@ali/4ever-plugin-hr");

var _everPluginImage = require("@ali/4ever-plugin-image");

var _everPluginLink = require("@ali/4ever-plugin-link");

var _everPluginMention = require("@ali/4ever-plugin-mention");

var _everPluginInlineCode = require("@ali/4ever-plugin-inline-code");

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _everPluginTable = require("@ali/4ever-plugin-table");

var _everPluginOnlineVideo = require("@ali/4ever-plugin-online-video");

var _everPluginPlaceholder = require("@ali/4ever-plugin-placeholder");

var _everPluginInlinePlaceholder = require("@ali/4ever-plugin-inline-placeholder");

var _everPluginRefer = require("@ali/4ever-plugin-refer");

var _everPluginSticker = require("@ali/4ever-plugin-sticker");

var _everPluginContainer = require("@ali/4ever-plugin-container");

var _everPluginTemplateButton = require("@ali/4ever-plugin-template-button");

var _everPluginColorBlocks = require("@ali/4ever-plugin-color-blocks");

var _schema = _interopRequireDefault(require("./schema"));

var _strictSchema = _interopRequireWildcard(require("./strictSchema"));

var _serializerValueJsonML = _interopRequireWildcard(require("./serializerValueJsonML"));

var _version = require("./version");

var _migrate = _interopRequireWildcard(require("./migrate"));

var _migrations = _interopRequireDefault(require("./migrations"));

var _createMigration = _interopRequireDefault(require("./migrations/createMigration"));

var _corrections = _interopRequireDefault(require("./corrections"));

const pluginUtils = {
  createEmptyParagraph: _utils.createEmptyParagraph,
  createEmptyText: _utils.createEmptyText,
  attrsToMarks: _utils.attrsToMarks,
  markToAttr: _utils.markToAttr,
  marksToAttrs: _utils.marksToAttrs
};
exports.pluginUtils = pluginUtils;

// 防止 breaking change 产生大规模影响
const BasicPlugin = () => ({});

exports.BasicPlugin = BasicPlugin;
// Marks Models
const Marks = {
  Bold: _everPluginMarks.Bold,
  Caps: _everPluginMarks.Caps,
  Color: _everPluginMarks.Color,
  Fonts: _everPluginMarks.Fonts,
  Highlight: _everPluginMarks.Highlight,
  InlineCode: _everPluginMarks.InlineCode,
  Italic: _everPluginMarks.Italic,
  SmallCaps: _everPluginMarks.SmallCaps,
  Strike: _everPluginMarks.Strike,
  Dstrike: _everPluginMarks.Dstrike,
  Sz: _everPluginMarks.Sz,
  Underline: _everPluginMarks.Underline,
  VertAlign: _everPluginMarks.VertAlign,
  Shd: _everPluginMarks.Shd,
  Spacing: _everPluginMarks.Spacing,
  Kern: _everPluginMarks.Kern,
  Em: _everPluginMarks.Em
};
exports.Marks = Marks;
var _default = {
  Serializer: _serializer.default,
  createMoPlugins: _createMoPlugins.default,
  createCustomizedMoPlugins: _createCustomizedMoPlugins.default
};
exports.default = _default;
