// @ali/4ever-mo/dist-node/strictSchema.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createStrictSchemaOld = exports.default = exports.inlineTypes = exports.blockTypes = void 0;

var _schema = _interopRequireWildcard(require("./schema"));

var _everCangjie = require("@ali/4ever-cangjie");

const blockTypes = Object.keys(_schema.default.blocks);
exports.blockTypes = blockTypes;
const inlineTypes = Object.keys(_schema.default.inlines);
/** @deprecated use controller.strictSchema instead  */

exports.inlineTypes = inlineTypes;
const strictSchema = (0, _everCangjie.createStrictSchema)(_schema.moPlugins);
var _default = strictSchema;
/**
 * @deprecated use controller.strictSchema instead
 *
 * 这个方法 pi 里面还在用，先兼容
 */

exports.default = _default;

const createStrictSchemaOld = opts => {
  const {
    blockTypes = [],
    inlineTypes = [],
    inTableBlockTypes = []
  } = opts;
  const s = { ...strictSchema,
    document: { ...strictSchema.document,
      nodes: [{
        match: blockTypes.map(t => ({
          type: t
        })),
        min: 1
      }]
    },
    blocks: { ...strictSchema.blocks,
      'table-cell': { ...strictSchema.blocks['table-cell'],
        nodes: [{
          match: inTableBlockTypes.map(t => ({
            type: t
          })),
          min: 1
        }]
      }
    }
  };
  return s;
};

exports.createStrictSchemaOld = createStrictSchemaOld;
