// @ali/4ever-bi/dist-node/plugins/quickInsert/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "QUICK_INSERT_INSERT", {
  enumerable: true,
  get: function () {
    return _everPluginQuickinsert.QUICK_INSERT_INSERT;
  }
});
Object.defineProperty(exports, "QUICK_INSERT_SHOW", {
  enumerable: true,
  get: function () {
    return _everPluginQuickinsert.QUICK_INSERT_SHOW;
  }
});
Object.defineProperty(exports, "QUICK_INSERT_SET", {
  enumerable: true,
  get: function () {
    return _everPluginQuickinsert.QUICK_INSERT_SET;
  }
});
Object.defineProperty(exports, "QUICK_INSERT_RESET", {
  enumerable: true,
  get: function () {
    return _everPluginQuickinsert.QUICK_INSERT_RESET;
  }
});
Object.defineProperty(exports, "QUICK_INSERT_REMOVE", {
  enumerable: true,
  get: function () {
    return _everPluginQuickinsert.QUICK_INSERT_REMOVE;
  }
});
Object.defineProperty(exports, "insertQuickInsert", {
  enumerable: true,
  get: function () {
    return _everPluginQuickinsert.insertQuickInsert;
  }
});
Object.defineProperty(exports, "showQuickInsert", {
  enumerable: true,
  get: function () {
    return _everPluginQuickinsert.showQuickInsert;
  }
});
Object.defineProperty(exports, "setQuickInsert", {
  enumerable: true,
  get: function () {
    return _everPluginQuickinsert.setQuickInsert;
  }
});
Object.defineProperty(exports, "resetQuickInsert", {
  enumerable: true,
  get: function () {
    return _everPluginQuickinsert.resetQuickInsert;
  }
});
Object.defineProperty(exports, "removeQuickInsert", {
  enumerable: true,
  get: function () {
    return _everPluginQuickinsert.removeQuickInsert;
  }
});

var _everPluginQuickinsert = require("@ali/4ever-plugin-quickinsert");
