// @ali/4ever-bi/dist-node/plugins/quickInsert/createRenderEditable.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _quickInsertTrigger = _interopRequireDefault(require("./components/quickInsertTrigger"));

var _everUtils = require("@ali/4ever-utils");

const _createElement = /*#__PURE__*/React.createElement;

function createRenderEditable(groups, insertMap, config) {
  // @ts-ignore
  return function renderEditable(_, controller, next) {
    const isNecessary = controller.query('isSelectionInQuickInsert');

    if (!isNecessary) {
      return next();
    }

    const decoration = controller.query('getQuickInsertDecoration');
    const value = decoration?.mark.data.plugin; // 这里应该根据注册的插件传入type

    const text = (0, _everUtils.getQuickInsertKeyword)(controller, {
      type: 'mention'
    });
    return /*#__PURE__*/_createElement(React.Fragment, null, next(), /*#__PURE__*/_createElement(_quickInsertTrigger.default, {
      groups: groups,
      value: value,
      text: text,
      controller: controller,
      insertMap: insertMap,
      config: config
    }));
  };
}

var _default = createRenderEditable;
exports.default = _default;
