// @ali/4ever-plugin-paragraph/dist-web/utils/index.js
import _extends from "@babel/runtime/helpers/extends";
import { DEFAULT_INDENT } from '@ali/4ever-utils';
import { ELEMENT_TYPE } from "../mo/meta"; // TOOD: 与 heading 解耦

var NODE_CAN_SET_SPACING = {
  paragraph: true,
  'heading-1': true,
  'heading-2': true,
  'heading-3': true,
  'heading-4': true,
  'heading-5': true,
  'heading-6': true,
  hr: true,
  code: true,
  embed: true,
  card: true
};

function canSetParagraphAttr(type) {
  return !!NODE_CAN_SET_SPACING[type];
}

var generalParagraphType = {};
[ELEMENT_TYPE, 'heading-1', 'heading-2', 'heading-3', 'heading-4', 'heading-5', 'heading-6'].forEach(function (type) {
  generalParagraphType[type] = true;
});
var DEFAULT_SPACING = {
  line: 1,
  lineRule: 'auto',
  before: 0,
  beforeLines: 0,
  beforeAutospacing: false,
  after: 0,
  afterLines: 0,
  afterAutospacing: false
};
export var DEFAULT_INDENT_FULL = _extends({}, DEFAULT_INDENT, {
  right: 0,
  rightChars: 0,
  start: 0,
  end: 0
});
export { canSetParagraphAttr, generalParagraphType, DEFAULT_SPACING };
