// @ali/dingtalk-kit/lib/sharedApi/platform/getEnv.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ua_1 = require("../../util/ua");
var shared_1 = require("../../constants/platform/shared");
var getDingTalkVersion_1 = require("../../util/getDingTalkVersion");
var env;
function getEnv() {
    if (!env) {
        var ua = ua_1.default();
        var isDingTalk = shared_1.dingTalkReg.test(ua);
        var isIPad = shared_1.iPadReg.test(ua);
        var isMac = ua.indexOf(shared_1.macKeywords) !== -1 && !isIPad;
        var isWin = ua.indexOf(shared_1.winKeywords) !== -1;
        var isAndroid = ua.indexOf(shared_1.androidKeywords) !== -1;
        var isIOS = shared_1.iOSReg.test(ua);
        var dingTalkVersion = getDingTalkVersion_1.default(ua, {
            isMac: isMac,
            isWin: isWin
        });
        env = {
            isDingTalk: isDingTalk,
            isIPad: isIPad,
            isMac: isMac,
            isWin: isWin,
            dingTalkVersion: dingTalkVersion,
            isAndroid: isAndroid,
            isIOS: isIOS
        };
    }
    return env;
}
exports.getEnv = getEnv;
