// @ali/4ever-bi/dist-node/plugins/alignment/components/next/createAlignmentButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createAlignmentButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _lodash = require("lodash");

var _everComponent = require("@ali/4ever-component");

var _actions = require("../../actions");

var _getTipsByType = require("../../utils/getTipsByType");

const _createElement = /*#__PURE__*/React.createElement;
;

function createAlignmentButton(type, titleKey, tooltipKey) {
  const AlignButton = props => {
    const {
      controller,
      disabled,
      locale,
      title,
      tooltip: tooltipProp,
      ...rest
    } = props;
    const isActive = controller.query('getAlignment') === type;
    const handleClick = React.useCallback(() => controller.run('onAction', (0, _actions.setAlignment)(type)), [controller]);
    const {
      isCollapsedAtEdge
    } = controller.value.selection;
    const isDisabled = disabled || isCollapsedAtEdge || controller.query('isSelectionInHr') || controller.query('isSelectionInTaskList') || controller.query('isSelectionInListSymbol');
    let Button;

    switch (type) {
      case 'left':
        Button = _everComponent.AlignLeftButton;
        break;

      case 'right':
        Button = _everComponent.AlignRightButton;
        break;

      case 'center':
        Button = _everComponent.AlignCenterButton;
        break;

      default:
        Button = _everComponent.AlignJustifyButton;
        break;
    }

    const tooltip = React.useMemo(() => ({
      title: title || titleKey && locale && locale[titleKey],
      shortCut: _getTipsByType.shortcut[type],
      description: tooltipKey && locale && locale[tooltipKey],
      ...tooltipProp
    }), [type, title]);
    return /*#__PURE__*/_createElement(Button, (0, _extends2.default)({
      testid: `bi-toolbar-align_${type}`,
      active: isActive,
      onClick: handleClick,
      disabled: isDisabled,
      role: `align${(0, _lodash.capitalize)(type)}`,
      title: title || titleKey && locale && locale[titleKey],
      tooltip: tooltip
    }, rest));
  };

  return AlignButton;
}
