// @ali/4ever-plugin-color/dist-web/bi/components/colorAndHighlightButton/styled.js
import styled from 'styled-components';
import { DEFAULT_COLOR } from "../../constants";
export var ColorAndHighlightPanelWrapper = /*#__PURE__*/styled.div(["width:183px;"]);
export var SectionTitle = /*#__PURE__*/styled.div(["font-size:12px;line-height:17px;padding-bottom:3px;color:rgba(23,26,29,0.6);"]);
export var PanelSection = /*#__PURE__*/styled.div(["padding:10px 8px 6px 8px;border-bottom:1px solid rgba(17,31,44,0.08);"]);
export var ColorBox = /*#__PURE__*/styled.div(["display:flex;flex-direction:row;flex-wrap:wrap;justify-content:space-between;"]);
export var ColorBoxItem = /*#__PURE__*/styled.div(["position:relative;width:24px;height:24px;margin:2px auto;display:flex;justify-content:center;align-items:center;border-radius:5px;cursor:pointer;", " ", ";&:hover{border:1px solid ", ";}"], function (p) {
  return p.bgColor && "background-color: " + p.bgColor + ";";
}, function (p) {
  return p.isSelected && 'border: 1px solid #0089ff;';
}, function (p) {
  return p.isSelected ? '#0089FF' : 'rgba(126, 134, 142, 0.12)';
});
export var TextColorItem = /*#__PURE__*/styled.div(["width:24px;height:24px;box-sizing:border-box;border-radius:4px;color:", ";display:flex;align-items:center;justify-content:center;", ";"], function (p) {
  return p.color || DEFAULT_COLOR;
}, function (p) {
  return !p.isSelected && 'border: 1px solid rgba(126, 134, 142, 0.16);';
});
export var ClearItem = /*#__PURE__*/styled.div(["position:relative;width:22px;height:22px;border-radius:4px;border:1px solid rgba(126,134,142,0.16);&::after{content:'';position:absolute;width:23px;height:2px;left:-1px;top:9.4px;background:#e41100;transform:rotate(45deg);}"]);
export var ClearBox = /*#__PURE__*/styled.div(["height:28px;line-height:28px;margin:2px;color:#0089ff;font-size:13px;text-align:center;cursor:pointer;border-radius:4px;&:hover{background:rgba(126,134,142,0.12);}"]);
