// @ali/we-ui/dist-web/components/radio/Radio.js
import React, { useCallback, useContext } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { RadioWrap, RadioContent, RadioInner, RadioText } from "./styled";
import useControlledValue from "../../hooks/useControlledValue";
import { classnames } from "../../utils";
import { RadioContext } from "./RadioContext";
var Radio = /*#__PURE__*/React.memo( /*#__PURE__*/React.forwardRef(function (props, ref) {
  var children = props.children,
      onChange = props.onChange,
      value = props.value; // eslint-disable-next-line prefer-const

  var _useControlledValue = useControlledValue(props.checked, false),
      checked = _useControlledValue[0],
      setChecked = _useControlledValue[1];

  var provider = useContext(RadioContext); // 如果有Group，直接根据 Group的选中项判断

  if (provider && provider.checkedValue) {
    checked = provider.checkedValue === value;
  } // disabled状态，如果有Group包住，且Group的disabled为true，则disabled为true


  var disabled = provider && provider.disabled ? true : props.disabled; // Group 修改参数函数

  var onToggleValue = provider && provider.onToggleValue;
  /**
   * 修改选中状态
   *
   * @param {React.ChangeEvent<HTMLInputElement>} e
   * @return {void}
   */

  var handleChange = useCallback(function (e) {
    if (onToggleValue && value) {
      onToggleValue({
        checked: e.target.checked,
        value: value
      });
    } else {
      if (onChange) {
        onChange(e);
      }

      setChecked(e.target.checked);
    }
  }, [onChange, onToggleValue, value, setChecked]);
  return /*#__PURE__*/_createElement(RadioWrap, {
    checked: checked,
    disabled: disabled
  }, /*#__PURE__*/_createElement(RadioContent, {
    disabled: disabled,
    checked: checked
  }, /*#__PURE__*/_createElement("input", {
    ref: ref,
    type: "radio",
    onChange: handleChange,
    disabled: disabled,
    checked: checked
  }), /*#__PURE__*/_createElement(RadioInner, {
    className: classnames({
      disabled: disabled
    })
  })), children && /*#__PURE__*/_createElement(RadioText, null, children));
}));
Radio.displayName = 'Radio';
export { Radio };
