// @ali/4ever-bi/dist-node/plugins/sticker/mobile/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everBamboo = require("@ali/4ever-bamboo");

var _renderNode = _interopRequireDefault(require("../renderNode"));

var _onCangjieSelect = _interopRequireDefault(require("../handlers/onCangjieSelect"));

var _onAction = _interopRequireDefault(require("../handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("../handlers/onKeyDown"));

var _constants = require("../constants");

var _everPluginSticker = require("@ali/4ever-plugin-sticker");

const {
  DEFAULT_STICKERS,
  Sticker
} = _everBamboo.StickerPlugin;

function createStickerPlugin(config) {
  const stickerConfig = {
    lang: _constants.DEFAULT_LANG,
    customSticker: [DEFAULT_STICKERS],
    ...config.sticker
  };
  const sticker = new Sticker(stickerConfig);
  const params = {
    stickerConfig,
    sticker
  };
  return { ...(0, _everPluginSticker.moFactory)({}),
    // @ts-ignore
    renderNode: (0, _renderNode.default)(params),
    onAction: _onAction.default,
    onKeyDown: _onKeyDown.default,
    onCangjieSelect: _onCangjieSelect.default,
    models: {
      klass: 'inline',
      type: _constants.STICKER_NODE_TYPE
    }
  };
}

createStickerPlugin.defaultSticker = DEFAULT_STICKERS;
var _default = createStickerPlugin;
exports.default = _default;
