// @ali/dingtalk-jsapi/lib/apiHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.genBizStoreParamsDealFn = exports.genBoolResultDealFn = exports.forceChangeParamsDealFn = exports.genDefaultParamsDealFn = exports.addDefaultCorpIdParamsDeal = exports.addWatchParamsDeal = void 0;
var addWatchParamsDeal = function (params) {
    var finalParams = Object.assign({}, params);
    finalParams.watch = true;
    return finalParams;
};
exports.addWatchParamsDeal = addWatchParamsDeal;
var addDefaultCorpIdParamsDeal = function (params) {
    var finalParams = Object.assign({}, params);
    finalParams.corpId = 'corpId';
    return finalParams;
};
exports.addDefaultCorpIdParamsDeal = addDefaultCorpIdParamsDeal;
var genDefaultParamsDealFn = function (defaultParams) {
    var defaultParamsObj = Object.assign({}, defaultParams);
    return function (params) {
        return Object.assign({}, defaultParamsObj, params);
    };
};
exports.genDefaultParamsDealFn = genDefaultParamsDealFn;
var forceChangeParamsDealFn = function (forceParams) {
    var forceParamsObj = Object.assign({}, forceParams);
    return function (params) {
        return Object.assign(params, forceParamsObj);
    };
};
exports.forceChangeParamsDealFn = forceChangeParamsDealFn;
var genBoolResultDealFn = function (boolKeyList) {
    return function (params) {
        var finalParams = Object.assign({}, params);
        boolKeyList.forEach(function (key) {
            if (finalParams[key] !== undefined) {
                finalParams[key] = !!finalParams[key];
            }
        });
        return finalParams;
    };
};
exports.genBoolResultDealFn = genBoolResultDealFn;
var genBizStoreParamsDealFn = function (params) {
    var finalParams = Object.assign({}, params);
    if (typeof finalParams.params !== 'string') {
        finalParams.params = JSON.stringify(finalParams);
        return finalParams;
    }
    else {
        return finalParams;
    }
};
exports.genBizStoreParamsDealFn = genBizStoreParamsDealFn;
