// @ali/4ever-bi/dist-node/plugins/hoverToolbar/components/MoreMenuButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MoreMenuButton = void 0;

var _react = _interopRequireDefault(require("react"));

var _weDesignNext = require("@ali/we-design-next");

var _everComponent = require("@ali/4ever-component");

var _everBamboo = require("@ali/4ever-bamboo");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement(_everComponent.IconMore, null);

const MoreMenuButton = props => {
  const {
    items,
    onOverlayVisibleChange,
    onSelect,
    locale
  } = props;

  const getContainer = _everComponent.HoverToolbarHooks.useGetContainer();

  const [isOverlayVisible, setOverlayVisible] = _everComponent.HoverToolbarHooks.usePopoverVisible('more', onOverlayVisibleChange);

  const handleSelect = _react.default.useCallback(key => {
    onSelect(key);
    setOverlayVisible(false);
  }, [onSelect, setOverlayVisible]);

  const toggleVisible = _react.default.useCallback(() => {
    setOverlayVisible(!isOverlayVisible);
  }, [setOverlayVisible, isOverlayVisible]);

  const renderItem = _react.default.useCallback(({
    key
  }) => {
    if (!items) return null;
    const item = items?.find(t => typeof t !== 'string' && t.key === key);
    if (!item) return null;
    return /*#__PURE__*/_createElement(_weDesignNext.List.Item, item);
  }, [items]);

  return /*#__PURE__*/_createElement(_weDesignNext.Popover, {
    placement: "bottomLeft",
    content: /*#__PURE__*/_createElement(_weDesignNext.List, {
      "data-testid": "bi-toolbar-hover-more-options",
      items: items,
      onSelect: handleSelect,
      "data-role": "hoverToolbarMoreList",
      renderItem: renderItem
    }),
    onVisibleChange: setOverlayVisible,
    visible: isOverlayVisible,
    getContainer: getContainer,
    "data-testid": "overlay-bi-toolbar-hover-more"
  }, /*#__PURE__*/_createElement(_weDesignNext.TBIconButton, {
    active: isOverlayVisible,
    icon: _ref,
    onClick: toggleVisible,
    title: locale?.more || 'More',
    tooltip: isOverlayVisible ? '' : locale?.more || 'More',
    testid: "bi-toolbar-hover-more",
    role: _everBamboo.PluginRoles.hoverToolbarMore
  }));
};

exports.MoreMenuButton = MoreMenuButton;
