// @ali/4ever-plugin-vote-card/dist-web/utils/components/headerToolbar.js
import React, { useCallback, useMemo, useRef } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { InlineToolbar, IconButton } from '@ali/we-toolbar';
import { EditNormal } from '@ali/we-icon';
import { voteCopy, voteDelete, voteEditing, voteCut } from "../actions";
import { HeaderWrapper } from "./styled";

var HeaderToolbar = function HeaderToolbar(props) {
  var controller = props.controller,
      node = props.node;
  var containerRef = useRef(null);
  var getContainer = useCallback(function () {
    return containerRef.current || window.document.body;
  }, []);
  var handleEdit = useCallback(function () {
    controller.run('onAction', voteEditing(node));
  }, []);
  var handleDelete = useCallback(function () {
    controller.run('onAction', voteDelete(node.key));
  }, [controller]);
  var handleCopy = useCallback(function () {
    controller.run('onAction', voteCopy(node.key));
  }, [controller]);
  var handleCut = useCallback(function () {
    controller.run('onAction', voteCut(node.key));
  }, [controller]);
  var moreMenuConfig = useMemo(function () {
    return {
      options: [[{
        name: 'delete',
        onClick: handleDelete,
        testId: 'templateButton-delete'
      }, {
        name: 'copy',
        onClick: handleCopy,
        testId: 'templateButton-copy'
      }, {
        name: 'cut',
        onClick: handleCut,
        testId: 'templateButton-cut'
      }]]
    };
  }, [handleCopy, handleCut, handleDelete]);
  return /*#__PURE__*/_createElement(HeaderWrapper, {
    className: "header-toolbar",
    "data-testid": "card-vote-header-toolbar",
    ref: containerRef
  }, /*#__PURE__*/_createElement(InlineToolbar, {
    getContainer: getContainer,
    moreMenuConfig: moreMenuConfig,
    width: 400
  }, /*#__PURE__*/_createElement(IconButton, {
    onClick: handleEdit,
    icon: /*#__PURE__*/_createElement(EditNormal, {
      style: {
        cursor: 'pointer'
      },
      "data-testid": "card-vote-header-toolbar-edit"
    })
  })));
};

export default HeaderToolbar;
