// @ali/we-design/dist-web/components/find-replace/FindAndReplace.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _ArrowLeftNormal, _span, _ArrowRightNormal;

import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { ArrowLeftNormal, ArrowRightNormal, ClosedNormal } from "../../components/icons";
import { isEscHotKey, isEnterHotKey, isShiftEnterHotKey, isComposing } from "../../utils";
import { Tabs, TabPane } from "../../components/tabs";
import { TranslateComp } from '@ali/we-util';
import { ModalWrapper, ModalText, ModalCloseButton, ModalInput, PanelWrapper, OperationPanel, FocusPanel, OperationButtonWrapper, Description, RowWrapper, FlexRowWrapper // FlexItemOperationPanel,
} from "./styled";
import { Button } from "../../components/button";
import { TranslateProvider } from '@ali/we-util';
export var FindAndReplace = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(FindAndReplace, _React$Component);

  function FindAndReplace(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;
    _this.inputRef = /*#__PURE__*/React.createRef();

    _this.handleReplaceCurrent = function () {
      var onReplaceCurrent = _this.props.onReplaceCurrent;
      var _this$state$replaceTe = _this.state.replaceText,
          replaceText = _this$state$replaceTe === void 0 ? '' : _this$state$replaceTe;

      if (onReplaceCurrent) {
        onReplaceCurrent(replaceText);
      }
    };

    _this.handleReplaceAll = function () {
      var onReplaceAll = _this.props.onReplaceAll;
      var _this$state$replaceTe2 = _this.state.replaceText,
          replaceText = _this$state$replaceTe2 === void 0 ? '' : _this$state$replaceTe2;

      if (onReplaceAll) {
        onReplaceAll(replaceText);
      }
    };

    _this.handleFindInputChange = function (event) {
      var _event$target$value = event.target.value,
          text = _event$target$value === void 0 ? '' : _event$target$value;
      var onFindInputChange = _this.props.onFindInputChange;

      if (onFindInputChange) {
        onFindInputChange(text);
      }
    };

    _this.handleClose = function () {
      var _this$props = _this.props,
          onClose = _this$props.onClose,
          onFindInputChange = _this$props.onFindInputChange;
      onClose && onClose();
      onFindInputChange('');

      _this.setState({
        replaceText: ''
      });
    };

    _this.handleFindInputKeyDown = function (event) {
      var _this$props2 = _this.props,
          onFocusIndexChange = _this$props2.onFocusIndexChange,
          focusIndex = _this$props2.focusIndex;

      if (isShiftEnterHotKey(event)) {
        onFocusIndexChange(focusIndex - 1);
      } else if (!isComposing(event) && isEnterHotKey(event) && onFocusIndexChange) {
        onFocusIndexChange(focusIndex + 1);
      } else if (isEscHotKey(event)) {
        _this.handleClose();
      }
    };

    _this.handleFocusPrevMatchedItem = function () {
      var _this$props3 = _this.props,
          onFocusIndexChange = _this$props3.onFocusIndexChange,
          focusIndex = _this$props3.focusIndex;

      if (onFocusIndexChange) {
        onFocusIndexChange(focusIndex - 1);
      }
    };

    _this.handleFocusNextMatchedItem = function () {
      var _this$props4 = _this.props,
          onFocusIndexChange = _this$props4.onFocusIndexChange,
          focusIndex = _this$props4.focusIndex;

      if (onFocusIndexChange) {
        onFocusIndexChange(focusIndex + 1);
      }
    };

    _this.handleReplaceInputChange = function (event) {
      var _event$target$value2 = event.target.value,
          text = _event$target$value2 === void 0 ? '' : _event$target$value2;

      _this.setState({
        replaceText: text
      });
    };

    _this.handleReplaceInputKeyDown = function (event) {
      var _this$props5 = _this.props,
          onReplaceCurrent = _this$props5.onReplaceCurrent,
          disableReplace = _this$props5.disableReplace;
      var replaceText = _this.state.replaceText;

      if (!isComposing(event) && isEnterHotKey(event) && !disableReplace && onReplaceCurrent) {
        onReplaceCurrent(replaceText);
      } else if (isEscHotKey(event)) {
        _this.handleClose();
      }
    };

    _this.focusInput = function () {
      if (_this.inputRef.current) {
        _this.inputRef.current.focus();
      }
    };

    _this.selectInput = function () {
      if (_this.inputRef.current) {
        _this.inputRef.current.select();
      }
    };

    _this.renderFocusIndex = function () {
      var _this$props6 = _this.props,
          matchedCount = _this$props6.matchedCount,
          focusIndex = _this$props6.focusIndex,
          testid = _this$props6.testid;
      return /*#__PURE__*/_createElement(FocusPanel, null, /*#__PURE__*/_createElement(Button, {
        shape: "circle",
        disabled: !matchedCount,
        onClick: _this.handleFocusPrevMatchedItem,
        icon: _ArrowLeftNormal || (_ArrowLeftNormal = /*#__PURE__*/_createElement(ArrowLeftNormal, null)),
        tabIndex: "-1",
        testid: testid ? "ArrowLeftNormal-" + testid : undefined
      }), /*#__PURE__*/_createElement(ModalText, {
        style: {
          wordBreak: 'keep-all',
          whiteSpace: 'nowrap'
        }
      }, matchedCount > 0 ? /*#__PURE__*/_createElement("span", null, focusIndex + "/" + matchedCount) : // : <TranslateComp locale={locale} textKey={'we_ui_find_and_replace_result'} />
      _span || (_span = /*#__PURE__*/_createElement("span", null, "0/0"))), /*#__PURE__*/_createElement(Button, {
        shape: "circle",
        disabled: !matchedCount,
        onClick: _this.handleFocusNextMatchedItem,
        icon: _ArrowRightNormal || (_ArrowRightNormal = /*#__PURE__*/_createElement(ArrowRightNormal, null)),
        tabIndex: "-1",
        testid: testid ? "ArrowRightNormal-" + testid : undefined
      }));
    };

    _this.renderFindPanel = function () {
      var findText = _this.props.findText;
      return /*#__PURE__*/_createElement(PanelWrapper, null, /*#__PURE__*/_createElement(FlexRowWrapper, null, /*#__PURE__*/_createElement(TranslateProvider, null, function (t) {
        return /*#__PURE__*/_createElement(ModalInput, {
          autoFocus: true,
          "data-testid": "J_findValueInput",
          ref: _this.inputRef,
          value: findText,
          onChange: _this.handleFindInputChange,
          onKeyDown: _this.handleFindInputKeyDown,
          placeholder: t('we_ui_find_and_replace_find_placeholder')
        });
      })), /*#__PURE__*/_createElement(OperationPanel, null, _this.renderFocusIndex()));
    };

    _this.renderReplacePanel = function () {
      var _this$props7 = _this.props,
          findText = _this$props7.findText,
          locale = _this$props7.locale,
          matchedCount = _this$props7.matchedCount,
          _this$props7$disableR = _this$props7.disableReplace,
          disableReplace = _this$props7$disableR === void 0 ? false : _this$props7$disableR,
          _this$props7$disableR2 = _this$props7.disableReplaceAll,
          disableReplaceAll = _this$props7$disableR2 === void 0 ? false : _this$props7$disableR2;
      var replaceText = _this.state.replaceText;
      return /*#__PURE__*/_createElement(PanelWrapper, null, /*#__PURE__*/_createElement(Description, null, /*#__PURE__*/_createElement(TranslateComp, {
        locale: locale,
        textKey: 'we_ui_find_and_replace_find_description'
      })), /*#__PURE__*/_createElement(TranslateProvider, null, function (t) {
        return /*#__PURE__*/_createElement(ModalInput, {
          autoFocus: true,
          "data-testid": "J_findValueInput",
          ref: _this.inputRef,
          value: findText,
          onChange: _this.handleFindInputChange,
          onKeyDown: _this.handleFindInputKeyDown,
          placeholder: t('we_ui_find_and_replace_find_placeholder')
        });
      }), /*#__PURE__*/_createElement(RowWrapper, null, /*#__PURE__*/_createElement(Description, null, /*#__PURE__*/_createElement(TranslateComp, {
        locale: locale,
        textKey: 'we_ui_find_and_replace_replace_description'
      })), /*#__PURE__*/_createElement(TranslateProvider, null, function (t) {
        return /*#__PURE__*/_createElement(ModalInput, {
          "data-testid": "J_replaceValueInput",
          value: replaceText,
          onChange: _this.handleReplaceInputChange,
          onKeyDown: _this.handleReplaceInputKeyDown,
          placeholder: t('we_ui_find_and_replace_replace_placeholder')
        });
      })), /*#__PURE__*/_createElement(OperationPanel, null, /*#__PURE__*/_createElement(OperationButtonWrapper, null, /*#__PURE__*/_createElement(Button, {
        shape: "round",
        disabled: matchedCount === 0 || disableReplace,
        testid: "J_replaceOne",
        onClick: _this.handleReplaceCurrent
      }, /*#__PURE__*/_createElement(TranslateComp, {
        locale: locale,
        textKey: 'we_ui_find_and_replace_replace'
      })), /*#__PURE__*/_createElement(Button, {
        shape: "round",
        disabled: matchedCount === 0 || disableReplaceAll,
        style: {
          marginLeft: 10
        },
        testid: "J_replaceAll",
        onClick: _this.handleReplaceAll
      }, /*#__PURE__*/_createElement(TranslateComp, {
        locale: locale,
        textKey: 'we_ui_find_and_replace_replace_all'
      }))), _this.renderFocusIndex()));
    };

    _this.state = {
      replaceText: props.defaultReplaceText || ''
    };
    return _this;
  }

  var _proto = FindAndReplace.prototype;

  _proto.render = function render() {
    var _this$props8 = this.props,
        activeTabKey = _this$props8.activeTabKey,
        _this$props8$focused = _this$props8.focused,
        focused = _this$props8$focused === void 0 ? false : _this$props8$focused,
        isFindOnly = _this$props8.isFindOnly,
        locale = _this$props8.locale,
        defaultTabKey = _this$props8.defaultTabKey,
        testid = _this$props8.testid,
        domRef = _this$props8.domRef,
        onTabClick = _this$props8.onTabClick;
    return /*#__PURE__*/_createElement(ModalWrapper, {
      focused: focused,
      ref: domRef
    }, /*#__PURE__*/_createElement(Tabs, {
      defaultActiveKey: defaultTabKey,
      activeKey: activeTabKey,
      onTabClick: onTabClick
    }, /*#__PURE__*/_createElement(TabPane, {
      tab: /*#__PURE__*/_createElement(TranslateComp, {
        locale: locale,
        textKey: 'we_ui_find_and_replace_find'
      }),
      key: "find"
    }, this.renderFindPanel()), isFindOnly ? null : /*#__PURE__*/_createElement(TabPane, {
      tab: /*#__PURE__*/_createElement(TranslateComp, {
        locale: locale,
        textKey: 'we_ui_find_and_replace_replace'
      }),
      key: "replace"
    }, this.renderReplacePanel())), /*#__PURE__*/_createElement(ModalCloseButton, {
      onMouseDown: this.handleClose
    }, /*#__PURE__*/_createElement(ClosedNormal, {
      testid: testid
    })));
  };

  return FindAndReplace;
}(React.Component);
