// @ali/4ever-plugin-code/dist-web/utils/constants/themes.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { codeTheme } from '@ali/4ever-utils';
export var DEFAULT_THEME = 'default';
var Title = /*#__PURE__*/styled.span(["display:inline-flex;align-items:center;vertical-align:middle;"]);
var Icon = /*#__PURE__*/styled.span(["background:", ";width:16px;height:16px;border-radius:3px;display:inline-block;margin-right:10px;border:1px solid #ecedef;"], function (props) {
  return props.color;
});
export var CodeWrapper = /*#__PURE__*/styled.div(["margin-left:", "px;margin-right:", "px;"], codeTheme['margin-left'], codeTheme['margin-right']);

var _ref = /*#__PURE__*/_createElement(Icon, {
  color: "#f6f8f9"
});

var _ref2 = /*#__PURE__*/_createElement(Icon, {
  color: "#111f2c"
});

export var createThemeMenuItems = function createThemeMenuItems(locale, key) {
  return [{
    key: DEFAULT_THEME,
    prefix: _ref,
    title: locale.lightTheme || 'Light',
    selected: key === DEFAULT_THEME,
    testid: 'option-default'
  }, {
    key: 'dracula',
    prefix: _ref2,
    title: locale.darkTheme || 'Dark',
    selected: key === 'dracula',
    testid: 'option-dracula'
  }];
};

var _ref3 = /*#__PURE__*/_createElement(Icon, {
  color: "#f6f8f9"
});

var _ref4 = /*#__PURE__*/_createElement(Icon, {
  color: "#111f2c"
});

export var createCodeThemes = function createCodeThemes(locale) {
  return [{
    title: /*#__PURE__*/_createElement(Title, null, _ref3, " ", locale.lightTheme || 'Light'),
    key: DEFAULT_THEME
  }, {
    title: /*#__PURE__*/_createElement(Title, null, _ref4, " ", locale.darkTheme || 'Dark'),
    key: 'dracula'
  }];
};
