// @ali/we-ui/dist-web/components/theme/params/font.js
/**
 * ยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยท>
 * ๐ @module : ๐ ๅญไฝ็ธๅณๅ้
 * ยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยทยท>
 */
import color from "./colors";
var font = {
  fontFamily: "~\"'Chinese Quote', -appleSystem, BlinkMacSystemFont, 'Segoe UI', Roboto, 'PingFang SC', 'Hiragino Sans GB', 'Microsoft YaHei', 'Helvetica Neue', Helvetica, Arial, sansSerif\"",
  // eslint-disable-line
  codeFamily: 'Consolas, Menlo, Courier, monospace',
  // ๐ https://ant.design/docs/spec/font-cn#ๅญ้ถไธ่ก้ซ
  fontSize: '14px',
  fontSizeSmall: '12px',
  fontSizeMedium: '14px',
  fontSizeLarge: '16px',
  fontSizeHug: '18px',
  fontSizeH1: '30px',
  fontSizeH2: '24px',
  fontSizeH3: '20px',
  fontSizeH4: '18px',
  fontSizeMenu: '13px',
  fontSizeIcon: '15px',
  // ๐ https://ant.design/docs/spec/font-cn#ๅญไฝ้ข่ฒ
  titleColor: color.colorText5,
  // #262626 85%
  textColor: color.colorText5,
  // #595959 65%
  textColor2: color.colorGray7,
  // #8C8C8C 45%
  textColor3: color.colorGray6,
  // #BFBFBF 25%
  textColorDisable: color.colorText1 // #BFBFBF 25%

};
export default font;
