// @ali/4ever-plugin-comment/dist-node/utils/components/simplePopup/textInput.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TextInput;

var _react = _interopRequireDefault(require("react"));

var _everCangjie = require("@ali/4ever-cangjie");

const _createElement = /*#__PURE__*/_react.default.createElement;
const buttonDefaultStyle = {
  background: '#2364E8',
  border: 'none',
  color: 'white',
  borderRadius: '4px',
  padding: '4px 8px'
};
const inputDiv = {
  marginTop: '8px',
  textAlign: 'right'
};

function TextInput(props) {
  const {
    contentId,
    sendComment,
    onAdd
  } = props;

  const controller = _everCangjie.Controller.useController();

  const textRef = _react.default.useRef(null);

  const [isSending, setIsSending] = _react.default.useState(false);

  const buttonStyle = { ...buttonDefaultStyle,
    opacity: isSending ? 0.5 : 1
  };

  const send = () => {
    const content = textRef.current?.value;

    if (content) {
      if (!sendComment) {
        throw Error('`Comment` plugin should specify sendComment callback');
      }

      setIsSending(true);
      sendComment({
        content,
        contentId
      }).then(() => {
        onAdd();

        if (textRef.current) {
          textRef.current.value = '';
        }

        setIsSending(false);
      }).catch(error => {
        setIsSending(false);
      });
    }
  };

  const handleKeyDown = event => {
    if (event.keyCode === 13 && event.metaKey) {
      event.preventDefault();
      send();
    }
  };

  return /*#__PURE__*/_createElement("div", null, /*#__PURE__*/_createElement("textarea", {
    "data-testid": "comment-input",
    ref: textRef,
    style: {
      width: '100%'
    },
    autoFocus: true,
    onKeyDown: handleKeyDown,
    placeholder: "\u6309 enter \u53D1\u9001\uFF0Cshift+enter \u6362\u884C"
  }), /*#__PURE__*/_createElement("div", {
    style: inputDiv
  }, /*#__PURE__*/_createElement("button", {
    "data-testid": "comment-send",
    onClick: send,
    style: buttonStyle
  }, isSending ? '发送中' : '评论')));
}
