// @ali/we-design-next/dist-web/form/select/styled.js
import styled, { css } from 'styled-components';
import { COLORS, CSS_BORDER_BOX, CSS_CURSOR_DISABLE, CSS_DISABLE_SELECT } from "../../common";
import { TEXT_CSS } from "../../basic/text";
var CSS_BORDERED_ENABLED = /*#__PURE__*/css(["&:hover{border-color:", ";}&:active{border-color:", ";}&:focus-within{border-color:", ";}"], COLORS.OTHERS.PRIMARY_HOVER, COLORS.COMMON.BLUE1_ACTIVE_COLOR, COLORS.DING_BLUR);
var CSS_BORDERED_DISABLED = /*#__PURE__*/css(["border-color:", ";"], COLORS.COMMON.OVERLAY_COLOR);
var CSS_BORDERED_ERRORED = /*#__PURE__*/css(["border-color:", ";"], COLORS.COMMON.DANGER1_COLOR);
var CSS_VISIBLE = /*#__PURE__*/css(["border-color:", ";box-shadow:0px 0px 0px 2px rgba(0,106,254,0.12);"], COLORS.DING_BLUR);
export var visibleStyle = {
  borderColor: COLORS.DING_BLUR,
  boxShadow: '0px 0px 0px 2px rgba(0, 106, 254, 0.12)'
};
var HOVER_TRANSITION_BG = /*#__PURE__*/css(["&:hover{background:", ";}&:active{background:", ";}transition:background-color 0.1s ease-in-out;"], COLORS.OTHERS.DAO_HOVER_BG_COLOR, COLORS.OTHERS.BG_ACTIVE);
var CSS_VISIBLE_BG = /*#__PURE__*/css(["background:", ";transition:background-color 0.1s ease-in-out;"], COLORS.OTHERS.BG_FOCUS);
export var InputWrap = /*#__PURE__*/styled.div(["position:relative;border-radius:4px;height:", ";"], function (_ref) {
  var size = _ref.size;
  return size === 'normal' ? '32px' : '28px';
});
export var SelectNormal = /*#__PURE__*/styled.div(["", " ", " ", ";display:flex;align-items:center;width:100%;height:", ";padding:4px 4px 4px 8px;border-radius:4px;", ";outline:none;color:", ";background-color:", ";border:", ";", ";", ";&:focus-within{", "}"], CSS_BORDER_BOX, TEXT_CSS.TEXT, CSS_CURSOR_DISABLE, function (_ref2) {
  var size = _ref2.size;
  return size === 'normal' ? '32px' : '28px';
}, CSS_DISABLE_SELECT, function (_ref3) {
  var disabled = _ref3.disabled;
  return disabled ? COLORS.LIGHT.LEVEL4_BASE_COLOR : COLORS.LIGHT.LEVEL1_BASE_COLOR;
}, function (_ref4) {
  var disabled = _ref4.disabled,
    bordered = _ref4.bordered,
    visible = _ref4.visible;
  return bordered ? disabled ? '#F7F6F6' : 'unset' : visible ? COLORS.OTHERS.BG_FOCUS : 'unset';
}, function (_ref5) {
  var bordered = _ref5.bordered;
  return bordered ? "1px solid " + COLORS.COMMON.OVERLAY_COLOR : '1px solid transparent';
}, function (_ref6) {
  var disabled = _ref6.disabled,
    bordered = _ref6.bordered,
    hasError = _ref6.hasError,
    visible = _ref6.visible;
  return !bordered ? disabled ? null : visible ? CSS_VISIBLE_BG : HOVER_TRANSITION_BG : disabled ? CSS_BORDERED_DISABLED : hasError ? CSS_BORDERED_ERRORED : CSS_BORDERED_ENABLED;
}, function (_ref7) {
  var visible = _ref7.visible,
    bordered = _ref7.bordered;
  return visible && bordered ? CSS_VISIBLE : null;
}, function (_ref8) {
  var disabled = _ref8.disabled,
    bordered = _ref8.bordered;
  return !disabled && bordered ? 'box-shadow: 0px 0px 0px 2px rgba(0, 106, 254, 0.12)' : null;
});
export var SelectNormalValue = /*#__PURE__*/styled.div(["flex-grow:1;min-width:10px;"]);
export var SelectNormalValuePlaceholder = /*#__PURE__*/styled.span(["margin:0 4px;color:", ";"], COLORS.LIGHT.LEVEL4_BASE_COLOR);
export var SuffixWrap = /*#__PURE__*/styled.span(["display:inline-block;margin-left:4px;margin-right:4px;line-height:0;cursor:default;"]);
export var errorStyle = {
  borderColor: COLORS.COMMON.DANGER1_COLOR
};
export var ErrorTip = /*#__PURE__*/styled.div(["position:absolute;top:", ";left:0;margin-top:4px;font-size:12px;line-height:18px;color:", ";"], function (_ref9) {
  var size = _ref9.size;
  return size === 'normal' ? '32px' : '28px';
}, COLORS.COMMON.DANGER1_COLOR);
