// @ali/4ever-plugin-refblock/dist-node/bi/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _index = _interopRequireDefault(require("../mo/index"));

var _lodash = require("lodash");

var queries = _interopRequireWildcard(require("./queries"));

var commands = _interopRequireWildcard(require("./commands"));

var _actions = require("./actions");

var _createOnPaste = _interopRequireDefault(require("./handlers/createOnPaste"));

var _createOnAction = _interopRequireDefault(require("./handlers/createOnAction"));

var _onCangjieCopyOrCut = _interopRequireDefault(require("./handlers/onCangjieCopyOrCut"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _createRenderNode = _interopRequireDefault(require("./handlers/createRenderNode"));

var commonQueries = _interopRequireWildcard(require("../common/queries"));

const biFactory = configs => {
  const refblockConfig = configs.refblock;
  /* istanbul ignore next */

  if (!refblockConfig) {
    return (0, _index.default)({});
  }

  const {
    locale,
    disableNestedInRefBlock = _lodash.stubFalse,
    disableDropTo = _lodash.stubFalse
  } = refblockConfig;
  const disableNodeInRefBlock = queries.createDisableNodeInRefBlock(disableNestedInRefBlock);
  return { ...(0, _index.default)({}),
    commands,
    queries: { ...commonQueries,
      disableDrag: queries.disableDrag,
      disableDrop: queries.createDisableDrop(locale.cannotMoveTo, locale.cannotDropTo, disableNodeInRefBlock, disableDropTo),
      disableNodeInRefBlock,
      isContainer: queries.isContainer,
      shouldSelectAll: queries.shouldSelectAll
    },
    actions: _actions.actions,
    isActions: _actions.isActions,
    onCangjiePaste: (0, _createOnPaste.default)(configs),
    onCangjieCopy: _onCangjieCopyOrCut.default,
    onCangjieCut: _onCangjieCopyOrCut.default,
    onKeyDown: _onKeyDown.default,
    onAction: (0, _createOnAction.default)(refblockConfig, configs.clipboard?.serializer),
    renderNode: (0, _createRenderNode.default)(refblockConfig)
  };
};

var _default = biFactory;
exports.default = _default;
