// @ali/4ever-plugin-calendar-card/dist-web/utils/components/subjectParagraph/index.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { isNumber } from 'lodash-es';
import { ArrowRightNormal, CalendarBetaNormalNormal } from '@ali/we-design';
import paragraphFactory from "../paragraphFactory";
import { CALENDAR_CARD_ICON_WIDTH, CALENDAR_CARD_EXTRA_ICON_WIDTH, DEFAULT_ICON_STYLE } from "../../constants";
import preventDefaultOnClick from "../../uitls/preventDefaultOnClick";
import { ContainerWrapper, IconWrapper, TextWrapper } from "./styled";

var _ref = /*#__PURE__*/_createElement(CalendarBetaNormalNormal, {
  style: DEFAULT_ICON_STYLE
});

var _ref2 = /*#__PURE__*/_createElement(ArrowRightNormal, null);

export var SubjectParagraph = function SubjectParagraph(props) {
  var innerWidth = props.innerWidth,
      subject = props.subject,
      jumpToCalendar = props.jumpToCalendar,
      isReadonly = props.isReadonly,
      isPlaceholder = props.isPlaceholder;
  var Paragraph = paragraphFactory(_ref);
  var realWidth = isNumber(innerWidth) ? innerWidth - CALENDAR_CARD_ICON_WIDTH - CALENDAR_CARD_EXTRA_ICON_WIDTH * 2 : innerWidth;
  return /*#__PURE__*/_createElement(Paragraph, null, /*#__PURE__*/_createElement(ContainerWrapper, {
    width: realWidth,
    isReadonly: isPlaceholder || isReadonly
  }, /*#__PURE__*/_createElement(TextWrapper, {
    onMouseDown: jumpToCalendar,
    onClick: preventDefaultOnClick,
    "data-testid": "calendar-subject"
  }, subject), /*#__PURE__*/_createElement(IconWrapper, null, isReadonly || isPlaceholder ? null : _ref2)));
};
