// @ali/we-design/dist-web/components/tabs/InkTabBarNodeTop.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

/* eslint-disable @typescript-eslint/explicit-function-return-type */

/* eslint-disable react/static-property-placement */
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { classnames } from "../../utils";
import { setTransform, isTransform3dSupported, getLeft, getActiveIndex, noop } from "./utils";
import { InkTabBar } from "./styled";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function _componentDidUpdate(component, init) {
  var _component$props = component.props,
      styles = _component$props.styles,
      panes = _component$props.panes,
      activeKey = _component$props.activeKey,
      type = _component$props.type;
  var rootNode = component.props.getRef('root');
  var wrapNode = component.props.getRef('nav') || rootNode;
  var inkBarNode = component.props.getRef('inkBar');
  var activeTab = component.props.getRef('activeTab');
  var inkBarNodeStyle = inkBarNode.style;

  if (init) {
    // prevent mount animation
    inkBarNodeStyle.display = 'none';
  }

  if (activeTab) {
    var tabNode = activeTab;
    var transformSupported = isTransform3dSupported(inkBarNodeStyle); // Reset current style

    setTransform(inkBarNodeStyle, '');
    inkBarNodeStyle.width = '';
    inkBarNodeStyle.height = '';
    inkBarNodeStyle.left = '';
    inkBarNodeStyle.top = '';
    inkBarNodeStyle.bottom = '';
    inkBarNodeStyle.right = '';
    var left = getLeft(tabNode, wrapNode);
    var width = tabNode.offsetWidth; // If tabNode'width equal to wrapNode'width when tabBarPosition is top or bottom
    // It means no css working, then ink bar should not have width until css is loaded
    // Fix https://github.com/ant-design/ant-design/issues/7564

    if (width === rootNode.offsetWidth) {
      width = 0;
    } else if (styles != null && styles.inkBar && (styles == null ? void 0 : styles.inkBar.width) !== undefined) {
      // 传入 styles ，则使用传入的宽度，并重算 left
      width = parseInt(styles == null ? void 0 : styles.inkBar.width, 10);

      if (width) {
        left += (tabNode.offsetWidth - width) / 2;
      }
    }

    if (type === 'd-design') {
      left += 14;
      width -= 28;
    } // 可关闭时左右 padding 不等，为使线条在文字内容中间对齐，需左移 (24 - 8) / 2 = 8px


    if (type === 'editable-line') {
      left -= 8;
    } // 可关闭时，偏移到tab title 的中间，去除关闭按钮的影响, 需左移动 （2 + 16) / 2 = 9px


    if (type === 'editable-biz') {
      left -= 9;
    } // use 3d gpu to optimize render


    if (transformSupported) {
      setTransform(inkBarNodeStyle, "translate3d(" + left + "px,0,0)");
    } else {
      inkBarNodeStyle.left = left + "px";
    }

    inkBarNodeStyle.width = width + "px";
  }

  var activeIndex = getActiveIndex(panes, activeKey);
  inkBarNodeStyle.display = activeIndex !== -1 ? 'block' : 'none';
}

var InkTabBarNodeTop = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(InkTabBarNodeTop, _React$Component);

  function InkTabBarNodeTop() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.timeoutID = -1;
    return _this;
  }

  var _proto = InkTabBarNodeTop.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var _this2 = this;

    // ref https://github.com/ant-design/ant-design/issues/8678
    // ref https://github.com/react-component/tabs/issues/135
    // InkTabBarNode need parent/root ref for calculating position
    // since parent componentDidMount triggered after child componentDidMount
    // we're doing a quick fix here to use setTimeout to calculate position
    // after parent/root component mounted
    this.timeoutID = window.setTimeout(function () {
      _componentDidUpdate(_this2, true);
    }, 0);
  };

  _proto.componentDidUpdate = function componentDidUpdate() {
    _componentDidUpdate(this);
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    clearTimeout(this.timeoutID);
  };

  _proto.render = function render() {
    var _this$props = this.props,
        saveRef = _this$props.saveRef,
        animated = _this$props.animated,
        type = _this$props.type;
    return /*#__PURE__*/_createElement(InkTabBar, {
      className: classnames(animated && 'animated', type),
      ref: saveRef('inkBar')
    });
  };

  return InkTabBarNodeTop;
}(React.Component);

InkTabBarNodeTop.defaultProps = {
  styles: {},
  saveRef: noop
};
export { InkTabBarNodeTop as default };
