// @ali/we-design/dist-web/components/tabs/styled.js
/* eslint-disable no-nested-ternary */

/* eslint-disable @typescript-eslint/explicit-function-return-type */
import styled, { css } from 'styled-components';
import { withTheme } from "../theme";
import { resetComponent } from "../../utils/styled.mixin";
var HorizontalTransition = /*#__PURE__*/css(["transition:transform ", ",width ", ",left ", ";"], function (_ref) {
  var theme = _ref.theme;
  return theme.ui.transEaseInOutSlow;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.transEaseInOut;
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.transEaseInOutSlow;
});
var VerticalTransition = /*#__PURE__*/css(["transition:transform ", ",height ", ",top ", ",width ", ",left ", ";"], function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.transEaseInOutSlow;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.ui.transEaseInOut;
}, function (_ref6) {
  var theme = _ref6.theme;
  return theme.ui.transEaseInOutSlow;
}, function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.transEaseInOut;
}, function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.transEaseInOutSlow;
});
export var TabsWrap = withTheme(styled.div(["", " &.animated{overflow:hidden;}"], HorizontalTransition));
export var TabBarRoot = withTheme(styled.div(["margin:0 0;outline:none;transition:padding ", ";position:relative;display:flex;flex-flow:row nowrap;align-items:center;&.capsule{padding:", " 0;}&.custom{padding:3px 0;}&.card,&.editable-card{border-bottom:1px solid ", ";padding:0;margin:0 0 ", ";}"], function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.transEaseInOutSlow;
}, function (_ref10) {
  var theme = _ref10.theme;
  return theme.ui.paddingMedium;
}, function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.borderColor;
}, function (_ref12) {
  var theme = _ref12.theme;
  return theme.ui.margin;
}));
export var NavContainer = withTheme(styled.div(["position:relative;display:flex;&.left{flex-direction:column;height:100%;width:auto;}font-size:14px;line-height:1.5715;white-space:nowrap;transition:padding ", ";flex:1;min-width:0;font-size:", ";&.scrolling:not(.biz,.editable-biz .d-design,){padding-right:22px;padding-left:22px;}&.custom{padding-right:17px;padding-left:17px;}&.small{font-size:", ";}&.large{font-size:", ";}&.editable-card,&.card{margin-bottom:0;}&.editable-biz,&.d-design,&.biz{margin:0 ", ";}"], function (_ref13) {
  var theme = _ref13.theme;
  return theme.ui.tabsTransChange;
}, function (_ref14) {
  var theme = _ref14.theme;
  return theme.ui.fontSizeMedium;
}, function (_ref15) {
  var theme = _ref15.theme;
  return theme.ui.fontSizeSmall;
}, function (_ref16) {
  var theme = _ref16.theme;
  return theme.ui.fontSizeLarge;
}, function (_ref17) {
  var theme = _ref17.theme;
  return theme.ui.margin;
}));
export var NavWrap = withTheme(styled.div(["flex:1 1 auto;align-self:stretch;overflow:hidden;white-space:nowrap;transform:translate(0);&.editable-card,&.card{margin-bottom:0;}"]));
export var Nav = withTheme(styled.div(["position:relative;&.animated{transition:transform ", ";}&.line,&.editable-line{margin:8px 0;}&.biz:not(.left),&.d-design:not(.left),&.editable-biz:not(.left){margin-bottom:5px;margin-top:5px;}"], function (_ref18) {
  var theme = _ref18.theme;
  return theme.ui.tabsTransChange;
}));
export var TabBarWrapper = /*#__PURE__*/styled.div(["display:flex;&.left{display:flex;flex-direction:column;}"]);
export var NavBarDivider = withTheme(styled.div(["position:absolute;left:0;right:0;bottom:4px;border-bottom:1px solid rgba(0,0,0,0.12);&.biz,&.d-design,&.editable-biz{bottom:0px;border-bottom:1px solid rgba(126,134,142,0.16);left:", ";right:", ";}&.capsule,&.custom,&.card,&.d-design,&.editable-card{display:none;}"], function (_ref19) {
  var theme = _ref19.theme;
  return theme.ui.margin;
}, function (_ref20) {
  var theme = _ref20.theme;
  return theme.ui.margin;
}));
export var TabBarTab = withTheme(styled.div(["outline:none;position:relative;display:inline-block;height:100%;margin:0 ", " 0 0;text-decoration:none;cursor:pointer;transition:color ", ";color:", ";user-select:none;&.small{padding:", " 0;}&.large{padding:", " 0;}&.actived{font-weight:500;}&.disabled{color:", ";cursor:not-allowed;}&.line,&.editable-line{line-height:1.43;padding:0;color:", ";&.actived{color:", ";}&.single-tab{width:calc(100% - 6px);margin-right:6px;margin-left:0;}.line-tab-close-button{font-size:", ";position:absolute;top:50%;right:4px;transform:translate(0,-50%);opacity:0;transition:opacity ", ";color:", ";&:hover{color:", ";}}&:hover,&.actived{.line-tab-close-button{opacity:1;}}}&.capsule{font-size:", ";line-height:1.34;padding:", " 0;&.actived{color:", ";}}&.custom{margin:0;cursor:default;padding-right:3px;padding-left:3px;}&.editable-card,&.card{position:relative;margin:0;margin-right:-1px;line-height:22px;border:1px solid ", ";border-bottom:0;padding:", " ", ";background:", ";&.actived{color:", ";background-color:", ";border-bottom-color:", ";}.card-tab-close-button{position:absolute;right:5px;top:50%;transform:translate(0,-50%);opacity:0;transition:opacity ", ";color:", ";&:hover{color:", ";}}&:hover,&.actived{.card-tab-close-button{opacity:1;}}}&.left{width:70px;height:32px;padding-left:5px;}&.editable-biz,&.d-design,&.biz{margin:0 2px;}&.d-design{margin:0;padding:0;.line-tab-close-button{position:absolute;top:50%;right:-7px;transform:translate(0,-50%);opacity:0;transition:opacity ", ";color:", ";}&:hover{.line-tab-close-button{opacity:1;color:", ";}}}"], function (_ref21) {
  var theme = _ref21.theme;
  return theme.ui.marginSmall;
}, function (_ref22) {
  var theme = _ref22.theme;
  return theme.ui.tabsTransChange;
}, function (_ref23) {
  var theme = _ref23.theme;
  return theme.ui.commonBaseColor1;
}, function (_ref24) {
  var theme = _ref24.theme;
  return theme.ui.paddingSmall;
}, function (_ref25) {
  var theme = _ref25.theme;
  return theme.ui.paddingLarge;
}, function (_ref26) {
  var theme = _ref26.theme;
  return theme.ui.disabledColor;
}, function (_ref27) {
  var theme = _ref27.theme;
  return theme.ui.commonBaseColor2;
}, function (_ref28) {
  var theme = _ref28.theme;
  return theme.ui.commonBaseColor1;
}, function (_ref29) {
  var theme = _ref29.theme;
  return theme.ui.fontSizeLarge;
}, function (_ref30) {
  var theme = _ref30.theme;
  return theme.ui.tabsTransChange;
}, function (_ref31) {
  var theme = _ref31.theme;
  return theme.ui.commonBaseColor3;
}, function (_ref32) {
  var theme = _ref32.theme;
  return theme.ui.commonBaseColor1;
}, function (_ref33) {
  var theme = _ref33.theme;
  return theme.ui.fontSizeSmall;
}, function (_ref34) {
  var theme = _ref34.theme;
  return theme.ui.paddingTiny;
}, function (_ref35) {
  var theme = _ref35.theme;
  return theme.ui.primaryColor;
}, function (_ref36) {
  var theme = _ref36.theme;
  return theme.ui.borderColor;
}, function (_ref37) {
  var theme = _ref37.theme;
  return theme.ui.paddingMedium;
}, function (_ref38) {
  var theme = _ref38.theme;
  return theme.ui.paddingHug;
}, function (_ref39) {
  var theme = _ref39.theme;
  return theme.ui.backgroundColor;
}, function (_ref40) {
  var theme = _ref40.theme;
  return theme.ui.primaryColor;
}, function (_ref41) {
  var theme = _ref41.theme;
  return theme.ui.tabsCardHeadActivedBackground;
}, function (_ref42) {
  var theme = _ref42.theme;
  return theme.ui.tabsCardHeadActivedBackground;
}, function (_ref43) {
  var theme = _ref43.theme;
  return theme.ui.tabsTransChange;
}, function (_ref44) {
  var theme = _ref44.theme;
  return theme.ui.commonBaseColor3;
}, function (_ref45) {
  var theme = _ref45.theme;
  return theme.ui.commonBaseColor1;
}, function (_ref46) {
  var theme = _ref46.theme;
  return theme.ui.tabsTransChange;
}, function (_ref47) {
  var theme = _ref47.theme;
  return theme.ui.commonBaseColor1;
}, function (_ref48) {
  var theme = _ref48.theme;
  return theme.ui.commonBaseColor1;
}));
export var TabBarTabItemWrap = withTheme(styled.div(["&.editable-line{padding:6px 24px 6px 8px;}&.line{padding:6px 16px;}&.left{padding-left:7px;}&.line,&.editable-line{border-radius:", ";&:hover{background-color:", ";}&:active{background-color:", ";}&.single-tab{background-color:transparent;}}&.capsule{padding:", " ", ";border-radius:", ";background-color:", ";&.actived{background-color:rgba(0,122,255,0.16);}}&.custom{padding:0;margin:0;}&.biz,&.editable-biz{display:flex;align-items:center;padding:7px 4px;border-radius:4px;color:rgba(23,26,29,0.6);&.single{cursor:auto;}&:not(.single):hover{background:rgba(126,134,142,0.12);.biz-tab-close-button{opacity:1;}}&:not(.disabled) > .biz-tab-close-button:hover{color:#171a1d;}&.actived{color:#171a1d;}.biz-tab-close-button{color:rgba(23,26,29,0.6);margin-left:2px;opacity:0;}}&.d-design{display:flex;align-items:center;padding:8px 14px;border-radius:4px;color:rgba(23,26,29,0.6);>span{align-items:center;display:flex;}&.single{cursor:auto;}&:not(.single):hover{color:#171A1D;.biz-tab-close-button{opacity:1;}}&:not(.disabled) > .line-tab-close-button:hover{color:#171a1d;}&.actived{color:#171a1d;}.line-tab-close-button{color:rgba(23,26,29,0.6);margin-left:2px;opacity:0;}}"], function (_ref49) {
  var theme = _ref49.theme;
  return theme.ui.radiusLarge;
}, function (_ref50) {
  var theme = _ref50.theme;
  return theme.ui.commonBgHoverColor;
}, function (_ref51) {
  var theme = _ref51.theme;
  return theme.ui.commonBgPressColor;
}, function (_ref52) {
  var theme = _ref52.theme;
  return theme.ui.paddingTiny;
}, function (_ref53) {
  var theme = _ref53.theme;
  return theme.ui.padding;
}, function (_ref54) {
  var theme = _ref54.theme;
  return theme.ui.radiusBig;
}, function (_ref55) {
  var theme = _ref55.theme;
  return theme.ui.commonBgHoverColor;
}));
export var InkTabBar = withTheme(styled.div(["position:absolute;bottom:-4px;left:0;z-index:1;width:0;height:5px;border-radius:", ";background-color:", ";&.animated{", "}&.biz,&.d-design,&.editable-biz{width:28px;height:4px;border-radius:4px;background:#171a1d;}&.d-design{width:28px;height:3px;border-radius:2px;background:#171a1d;}"], function (_ref56) {
  var theme = _ref56.theme;
  return theme.ui.marginMini;
}, function (_ref57) {
  var theme = _ref57.theme;
  return theme.ui.primaryColor;
}, HorizontalTransition));
export var InkTabBarVertical = withTheme(styled.div(["position:absolute;top:6.5px;z-index:1;height:0;width:4px;border-radius:", ";background-color:", ";&.animated{", "}&.biz,&.d-design,&.editable-biz{border-radius:2px;background:#171a1d;}"], function (_ref58) {
  var theme = _ref58.theme;
  return theme.ui.marginMini;
}, function (_ref59) {
  var theme = _ref59.theme;
  return theme.ui.primaryColor;
}, VerticalTransition));
var PrevNextButton = /*#__PURE__*/css(["display:flex;justify-content:center;align-items:center;position:absolute;z-index:", ";width:0;height:100%;color:rgba(0,0,0,0.45);cursor:pointer;opacity:0;transition:width ", ",opacity ", ",color ", ";user-select:none;pointer-events:none;color:", ";&.show{opacity:1;width:", ";height:100%;pointer-events:auto;}&:hover{color:", ";}&.disabled{color:", ";cursor:not-allowed;}"], function (_ref60) {
  var theme = _ref60.theme;
  return theme.ui.zIndexNormal;
}, function (_ref61) {
  var theme = _ref61.theme;
  return theme.ui.tabsTransChange;
}, function (_ref62) {
  var theme = _ref62.theme;
  return theme.ui.tabsTransChange;
}, function (_ref63) {
  var theme = _ref63.theme;
  return theme.ui.tabsTransChange;
}, function (_ref64) {
  var theme = _ref64.theme;
  return theme.ui.commonBaseColor3;
}, function (_ref65) {
  var theme = _ref65.theme;
  return theme.ui.tabsScrollingSize;
}, function (_ref66) {
  var theme = _ref66.theme;
  return theme.ui.commonActionIconColor;
}, function (_ref67) {
  var theme = _ref67.theme;
  return theme.ui.disabledColor;
});
export var PrevButton = withTheme(styled.div(["left:0;", ""], PrevNextButton));
export var NextButton = withTheme(styled.div(["right:0;", ""], PrevNextButton));
export var TabPaneWrap = withTheme(styled.div(["flex-shrink:0;width:100%;outline:none;"]));
export var TabContentWrap = withTheme(styled.div(["&.animated{display:flex;flex-direction:row;transition:transform ", ";}.inactived-pane{height:0;opacity:0;overflow:hidden;}"], function (_ref68) {
  var theme = _ref68.theme;
  return theme.ui.tabsTransChange;
}));
export var MoreButtonWrapper = withTheme(styled.div(["position:relative;display:flex;box-sizing:border-box;justify-content:center;align-items:center;z-index:", ";cursor:pointer;user-select:none;flex:0 0 36px;&.top{margin-bottom:5px;& > .more-button-inner-icon{transform:translate(0,2px);}&.editable-biz,&.d-design,&.biz{margin-bottom:1px;& > .more-button-inner-icon{transform:translate(0,1px);}}}&::after{content:'';position:absolute;pointer-events:none;transition:opacity 0.3s;}&.top::after{left:-30px;top:0;bottom:0;width:30px;box-shadow:inset -10px 0 8px -8px rgb(0 0 0 / 8%);}&.left::after{top:-30px;left:0;right:0;height:30px;box-shadow:inset 0 -10px 8px -8px rgb(0 0 0 / 8%);}"], function (_ref69) {
  var theme = _ref69.theme;
  return theme.ui.zIndexNormal;
}));
export var CloseAllDivider = /*#__PURE__*/styled.div(["position:absolute;height:23px;width:0px;border-left:1px solid rgba(126,134,142,0.16);top:50%;left:0px;transform:translate(0,-50%);"]);
export var CloseAllWrapper = /*#__PURE__*/styled.div(["position:relative;display:flex;align-items:center;justify-content:center;flex:0 0 36px;"]);
export var TabOptionButton = /*#__PURE__*/styled.div(["display:flex;align-items:center;justify-content:center;width:24px;height:24px;border-radius:4px;&:hover{background:rgba(126,134,142,0.12);cursor:pointer;}"]);
export var DropdownWrap = withTheme(styled.ul(["", " position:relative;overflow-y:auto;max-height:70vh;text-align:center;"], resetComponent));
export var DropdownItemWrap = withTheme(styled.li(["min-width:60px;font-size:", ";padding:", " ", ";&:hover{background-color:", ";cursor:pointer;}&.actived{color:", ";}&.disabled{color:", ";&:hover{background-color:transparent;cursor:not-allowed;}}"], function (_ref70) {
  var theme = _ref70.theme;
  return theme.ui.fontSizeMedium;
}, function (_ref71) {
  var theme = _ref71.theme;
  return theme.ui.paddingSmall;
}, function (_ref72) {
  var theme = _ref72.theme;
  return theme.ui.paddingMassive;
}, function (_ref73) {
  var theme = _ref73.theme;
  return theme.ui.backgroundColor;
}, function (_ref74) {
  var theme = _ref74.theme;
  return theme.ui.primaryColor;
}, function (_ref75) {
  var theme = _ref75.theme;
  return theme.ui.disabledColor;
}));
