// @ali/we-design/dist-web/components/tabs/TabBarRootNode.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";

/* eslint-disable react/static-property-placement */
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { getDataAttr, noop } from "./utils";
import { TabBarRoot } from "./styled";
import { classnames } from "../../utils";

var TabBarRootNode = function TabBarRootNode(props) {
  var onKeyDown = props.onKeyDown,
      style = props.tabBarStyle,
      children = props.children,
      saveRef = props.saveRef,
      type = props.type,
      testid = props.testid,
      restProps = _objectWithoutPropertiesLoose(props, ["onKeyDown", "tabBarStyle", "children", "saveRef", "type", "testid"]);

  return /*#__PURE__*/_createElement(TabBarRoot, _extends({
    className: classnames(type),
    role: "tablist",
    tabIndex: -1,
    ref: saveRef('root'),
    onKeyDown: onKeyDown,
    style: style,
    "data-testid": testid ? "TabBarRootNode_" + testid : undefined
  }, getDataAttr(restProps)), children);
};

TabBarRootNode.defaultProps = {
  onKeyDown: noop
};
export default TabBarRootNode;
