// @ali/4ever-plugin-color/dist-node/bi/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setColor = setColor;
exports.setColorAtRange = setColorAtRange;
exports.isSetColorAction = isSetColorAction;
exports.isSetColorAtRangeAction = isSetColorAtRangeAction;
exports.isActions = exports.actions = exports.COLOR_SET_AT_RANGE = exports.COLOR_SET = void 0;
const COLOR_SET = Symbol('COLOR_SET');
exports.COLOR_SET = COLOR_SET;
const COLOR_SET_AT_RANGE = Symbol('COLOR_SET_AT_RANGE');
exports.COLOR_SET_AT_RANGE = COLOR_SET_AT_RANGE;

function setColor(color) {
  return {
    type: COLOR_SET,
    payload: {
      color
    }
  };
}

function setColorAtRange(range, color) {
  return {
    type: COLOR_SET_AT_RANGE,
    payload: {
      range,
      color
    }
  };
}

function isSetColorAction(action) {
  return action.type === COLOR_SET;
}

function isSetColorAtRangeAction(action) {
  return action.type === COLOR_SET_AT_RANGE;
}

const actions = {
  setColor,
  setColorAtRange
};
exports.actions = actions;
const isActions = {
  isSetColorAction,
  isSetColorAtRangeAction
};
exports.isActions = isActions;
