// @ali/4ever-plugin-color/dist-node/bi/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SIMPLE_HIGHLIGHTS = exports.INITIAL_CAN_HIDE = exports.INITIAL_FILL = exports.DEFAUL_HIGHLIGHT = exports.INITIAL_HIGHLIGHT = exports.SIMPLE_COLORS = exports.DEFAULT_COLOR = void 0;
const DEFAULT_COLOR = '#404040';
exports.DEFAULT_COLOR = DEFAULT_COLOR;
const SIMPLE_COLORS = ['#404040', '#CB494B', '#FF9B00', '#FABD14', '#00B042', '#FFFFFF', '#317ED0', '#0089FF', '#B963D3', '#EC4589', '#AC6F42', '#888888']; // 极简模式颜色工具栏初始背景色

exports.SIMPLE_COLORS = SIMPLE_COLORS;
const INITIAL_HIGHLIGHT = '#FBF3DC'; // 默认背景色为透明

exports.INITIAL_HIGHLIGHT = INITIAL_HIGHLIGHT;
const DEFAUL_HIGHLIGHT = '';
exports.DEFAUL_HIGHLIGHT = DEFAUL_HIGHLIGHT;
const INITIAL_FILL = '#FDBE3D'; // 默认可隐藏

exports.INITIAL_FILL = INITIAL_FILL;
const INITIAL_CAN_HIDE = true;
exports.INITIAL_CAN_HIDE = INITIAL_CAN_HIDE;
const SIMPLE_HIGHLIGHTS = ['#FDE0E2', '#FFD8C1', '#F9E4CF', '#FFF5AC', '#FBF3DC', '#E6F7E7', '#DAF6F8', '#DFECFA', '#ECDEF5', '#EDE9DC', '#F0F0F0', '#171A1D', '#FF5219', '#FF9200', '#00B042', '#0089FF', '#B963D3'];
exports.SIMPLE_HIGHLIGHTS = SIMPLE_HIGHLIGHTS;
