// @ali/4ever-plugin-color/dist-node/bi/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.biFactory = void 0;

var _index = require("../mo/index");

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _commands = require("./commands");

var _queries = require("./queries");

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _actions = require("./actions");

var _createRenderToolbar = _interopRequireDefault(require("./createRenderToolbar"));

const biFactory = configs => {
  const colorConfig = configs.color || {};
  return { ...(0, _index.moFactory)({}),
    actions: _actions.actions,
    isActions: _actions.isActions,
    toolbar: (0, _toolbar.default)(colorConfig),
    renderToolbar: (0, _createRenderToolbar.default)(colorConfig.locale),
    commands: _commands.commands,
    queries: _queries.queries,
    onAction: _onAction.default
  };
};

exports.biFactory = biFactory;
