// @ali/4ever-bi/dist-node/plugins/link/linkCard/mobile/createRenderNode.js
"use strict";

var _interopRequireWildcard3 = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderNode;

var _interopRequireWildcard2 = _interopRequireDefault(require("@babel/runtime/helpers/interopRequireWildcard"));

var React = _interopRequireWildcard3(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _everComponent = require("@ali/4ever-component");

var _actions = require("../actions");

var _getBlockLinkCardState = require("../utils/getBlockLinkCardState");

var _everBamboo = require("@ali/4ever-bamboo");

var _zh_CN = _interopRequireDefault(require("../../../../locales/zh_CN"));

const _createElement = /*#__PURE__*/React.createElement;
const {
  normalizeLink,
  openLink
} = _everBamboo.LinkPlugin;

const Wrapper = /*#__PURE__*/_styledComponents.default.div(["cursor:pointer;border-radius:9px;"]);

function LinkCardNode(props) {
  const {
    nodeProps,
    controller,
    linkConfig,
    isMobile,
    URLLinkCard
  } = props;
  const {
    locale = _zh_CN.default.link.locale,
    protocolAllowList = ['http:', 'https:', 'dingtalk:', 'taobao:'],
    onCopyLink,
    getLinkInfo,
    openLink: openLinkOfConfig
  } = linkConfig;
  const {
    node,
    isSelected
  } = nodeProps;
  const {
    selection
  } = controller.value;
  const {
    text,
    href,
    cardInfo
  } = node.data.metadata;
  const scrollContent = (0, _everBamboo.useScrollableContent)() || document.body;
  const getContainer = React.useCallback(() => {
    return scrollContent;
  }, [scrollContent]);
  const urlInfo = React.useMemo(() => {
    const isTitleEqualsHref = text.trim() === href.trim();
    const actualTitle = isTitleEqualsHref ? cardInfo.title : text;
    return {
      title: actualTitle || '',
      desc: cardInfo.desc || '',
      imgURL: cardInfo.imgURL || ''
    };
  }, [cardInfo, text, href]);
  const linkCardLocale = React.useMemo(() => {
    const tooltipLocale = {
      editTooltip: locale.edit,
      copyTooltip: locale.copy,
      unlinkTooltip: locale.unlink,
      deleteTooltip: locale.delete,
      switchStyleTooltip: locale.styleText
    };
    return { ...locale,
      ...tooltipLocale
    };
  }, [locale]);
  const state = React.useMemo(() => {
    return (0, _getBlockLinkCardState.getBlockLinkCardState)(node);
  }, [node]);
  const handleVisit = React.useCallback(() => {
    if (openLinkOfConfig) {
      const normalizedLink = normalizeLink(href, protocolAllowList);
      openLinkOfConfig(normalizedLink);
    } else {
      openLink(href, protocolAllowList);
    }
  }, [href, protocolAllowList, openLinkOfConfig]);
  const handleCopy = React.useCallback(() => {
    // 卡片形式下，使用仓颉复制
    controller.run('onAction', (0, _actions.copyBlockLinkCard)(node));

    if (onCopyLink && typeof onCopyLink === 'function') {
      onCopyLink(href);
    }
  }, [href, onCopyLink, node, controller]);
  const handleCut = React.useCallback(() => {
    controller.run('onAction', (0, _actions.cutBlockLinkCard)(node));
  }, [controller, node]);
  const handleRemove = React.useCallback(() => {
    controller.run('onAction', (0, _actions.deleteBlockLinkCard)(node));
  }, [controller, node]);
  const handleUnlink = React.useCallback(() => {
    controller.run('onAction', (0, _actions.unlinkBlockCardData)(node));
  }, [controller, node]);
  const handleSwitchStyle = React.useCallback(e => {
    if (e === 'popup') {
      controller.run('onAction', (0, _actions.convertBlockLinkCardToTextLink)(node));
    }
  }, [controller, node]);
  const handleRefresh = React.useCallback(() => {
    controller.run('onAction', (0, _actions.updateBlockLinkCardData)(node, text, href));
  }, [controller, getLinkInfo, text, href, node]);
  return /*#__PURE__*/_createElement(Wrapper, null, /*#__PURE__*/_createElement(URLLinkCard, {
    getContainer: getContainer,
    isSelected: isSelected && selection.isCollapsed,
    cardStyle: 'indoc',
    state: state,
    url: href,
    readonly: false,
    locale: linkCardLocale,
    urlInfo: urlInfo,
    isMobile: isMobile // @ts-ignore
    ,
    onRefresh: handleRefresh,
    onCopy: handleCopy,
    onOpenURL: handleVisit,
    onSwitchStyle: handleSwitchStyle,
    onDelete: handleRemove,
    onCut: handleCut,
    onEdit: () => {},
    onUnlink: handleUnlink,
    editBtnTestID: "link-toolbar-edit",
    copyBtnTestID: "link-toolbar-copy",
    cutBtnTestID: "link-toolbar-cut",
    unlinkBtnTestID: "link-toolbar-unlink",
    deleteBtnTestID: "link-toolbar-remove",
    switchStyleBtnTestID: "link-toolbar-switch",
    linkStyleBtnTestID: "link-toolbar-switch-link",
    cardStyleBtnTestID: "link-toolbar-switch-card",
    cardTestID: "link-card"
  }));
}

function createRenderNode(linkConfig, isMobile) {
  const LazyLoad = linkConfig.lazyload || _everComponent.SimpleLazyLoad;
  const URLLinkCard = LazyLoad(() => Promise.resolve().then(() => (0, _interopRequireWildcard2.default)(require('@ali/we-biz-link-card'))).then(({
    URLLinkCard
  }) => URLLinkCard));
  return (nodeProps, controller) => {
    return /*#__PURE__*/_createElement(LinkCardNode, {
      linkConfig: linkConfig,
      isMobile: isMobile,
      nodeProps: nodeProps,
      URLLinkCard: URLLinkCard,
      controller: controller
    });
  };
}
