// @ali/4ever-bi/dist-node/plugins/link/linkCard/mobile/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _zh_CN = _interopRequireDefault(require("../../../../locales/zh_CN"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everFactory = require("@ali/4ever-factory");

var _createRenderNode = _interopRequireDefault(require("./createRenderNode"));

var _onAction = _interopRequireDefault(require("../onAction"));

const createBiLinkCardPlugin = (0, _everFactory.cardFactory)(bizConfig => {
  const linkCardConfig = {
    locale: { ..._zh_CN.default.link.locale
    },
    ...bizConfig.link
  };
  const serializer = bizConfig.clipboard?.serializer;
  const renderNode = (0, _createRenderNode.default)(linkCardConfig, true);
  return {
    cardType: 'link',
    onAction: (0, _onAction.default)(linkCardConfig, serializer),
    isMobile: true,
    renderNode,
    queries: {
      getMinTableColWidth: _everBamboo.LinkPlugin.getMinTableColWidthForBlockLinkCard
    }
  };
});
var _default = createBiLinkCardPlugin;
exports.default = _default;
