// @ali/4ever-plugin-group-factory/dist-node/utils/handlers/createRenderNode.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _everCangjie = require("@ali/4ever-cangjie");

var _group = _interopRequireDefault(require("../components/group"));

const _createElement = /*#__PURE__*/React.createElement;

function createRenderNode(groupType) {
  return function renderNode(props, controller, next) {
    const {
      node,
      attributes,
      isSelected,
      children,
      visible
    } = props;

    if (_everCangjie.Block.isBlock(node) && groupType === node.type) {
      return /*#__PURE__*/_createElement(_group.default, {
        node: node,
        attributes: attributes,
        controller: controller,
        isSelected: isSelected,
        visible: visible
      }, children());
    }

    return next();
  };
}

var _default = createRenderNode;
exports.default = _default;
