// @builder/mpa-config/lib/generate/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ReactGenerator_1 = require("./ReactGenerator");
var RaxGenerator_1 = require("./RaxGenerator");
function generatePageFiles(api, options) {
    var framework = options.framework, pageEntry = options.pageEntry, isAppEntry = options.isAppEntry;
    var generator;
    if (framework === 'react') {
        generator = new ReactGenerator_1.default(api, options);
    }
    else if (framework === 'rax') {
        generator = new RaxGenerator_1.default(api, options);
    }
    var userConfig = api.context.userConfig;
    // 在分析运行时依赖的场景下，不能直接执行 generator
    // 需要在分析完成并执行 disableRuntimePlugins 逻辑后，再执行生成
    var generateTasks = [];
    generateTasks.push(function () {
        generator.generateRunAppFile(userConfig);
    });
    // Do not modify the page entry when entryPath ends with app.ts
    if (isAppEntry) {
        return {
            generator: generator,
            generateTasks: generateTasks,
            entryPath: pageEntry,
            runAppPath: generator.runAppPath,
            routesFilePath: generator.routesFilePath,
        };
    }
    generateTasks.push(function () {
        generator.generateEntryFile();
    });
    return {
        generator: generator,
        generateTasks: generateTasks,
        entryPath: generator.entryPath,
        runAppPath: generator.runAppPath,
        routesFilePath: generator.routesFilePath,
    };
}
exports.default = generatePageFiles;
