// @ali/we-design/dist-web/components/lines/common.js
import _taggedTemplateLiteralLoose from "@babel/runtime/helpers/taggedTemplateLiteralLoose";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10, _templateObject11, _templateObject12, _templateObject13, _templateObject14;

export var lineWidthByBuiltIn = [1, 2, 4, 6];
export var lineWidthByCustom = [4, 6]; // 线段属性

export var commonLineMap = {
  single: {
    content: '',
    repeat: 0,
    sizeList: lineWidthByBuiltIn,
    defaultSize: 2,
    unitHeight: 6
  },
  dotted: {
    content: String.raw(_templateObject || (_templateObject = _taggedTemplateLiteralLoose(["e7f5"], ["\\e7f5"]))),
    repeat: 800,
    sizeList: lineWidthByBuiltIn,
    defaultSize: 2,
    unitHeight: 16 / 3
  },
  dashed: {
    content: String.raw(_templateObject2 || (_templateObject2 = _taggedTemplateLiteralLoose(["e7f8"], ["\\e7f8"]))),
    repeat: 400,
    sizeList: lineWidthByBuiltIn,
    defaultSize: 2,
    unitHeight: 16 / 2
  },
  dotDash: {
    content: String.raw(_templateObject3 || (_templateObject3 = _taggedTemplateLiteralLoose(["e7f4"], ["\\e7f4"]))),
    repeat: 400,
    sizeList: lineWidthByBuiltIn,
    defaultSize: 2,
    unitHeight: 16
  },
  dotDotDash: {
    content: String.raw(_templateObject4 || (_templateObject4 = _taggedTemplateLiteralLoose(["e7f6"], ["\\e7f6"]))),
    repeat: 400,
    sizeList: lineWidthByBuiltIn,
    defaultSize: 2,
    unitHeight: 16
  },
  "double": {
    content: String.raw(_templateObject5 || (_templateObject5 = _taggedTemplateLiteralLoose(["e7ed"], ["\\e7ed"]))),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 4
  },
  thickThinSmallGap: {
    content: String.raw(_templateObject6 || (_templateObject6 = _taggedTemplateLiteralLoose(["e7ea"], ["\\e7ea"]))),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 16 / 6
  },
  thinThickThinMediumGap: {
    content: String.raw(_templateObject7 || (_templateObject7 = _taggedTemplateLiteralLoose(["e7ec"], ["\\e7ec"]))),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 16 / 6
  },
  dashStroked: {
    content: String.raw(_templateObject8 || (_templateObject8 = _taggedTemplateLiteralLoose(["e7f3"], ["\\e7f3"]))),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 2,
    unitHeight: 16 / 7
  },
  wave: {
    content: String.raw(_templateObject9 || (_templateObject9 = _taggedTemplateLiteralLoose(["e7f0"], ["\\e7f0"]))),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 10,
    scale: 0.2
  },
  doubleWave: {
    content: String.raw(_templateObject10 || (_templateObject10 = _taggedTemplateLiteralLoose(["e7ef"], ["\\e7ef"]))),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 6,
    scale: 0.2
  },
  dashDotStroked: {
    content: String.raw(_templateObject11 || (_templateObject11 = _taggedTemplateLiteralLoose(["e7e7"], ["\\e7e7"]))),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 16 / 5
  },
  widthDoubleWave: {
    content: String.raw(_templateObject12 || (_templateObject12 = _taggedTemplateLiteralLoose(["e7f1"], ["\\e7f1"]))),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 10,
    scale: 0.2
  },
  widthWave: {
    content: String.raw(_templateObject13 || (_templateObject13 = _taggedTemplateLiteralLoose(["e7f2"], ["\\e7f2"]))),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 10,
    scale: 0.2
  },
  roundDot: {
    content: String.raw(_templateObject14 || (_templateObject14 = _taggedTemplateLiteralLoose(["e7f9"], ["\\e7f9"]))),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 16 / 3
  }
};
