// @ali/4ever-plugin-draggable/dist-web/bi/utils/domUtils.js
import { findDOMNodeByKey } from '@ali/4ever-utils';
import getClosestDom from "./getClosestDom";
export var DRAG_HIGHLIGHT_CLASSNAME = 'draggable-highlight';
export var DISABLE_DRAG_HIGHLIGHT_CLASSNAME = 'disable-draggable-highlight';
export var DRAG_THUMBNAIL_ID = 'drag-custom-thumbnail';
export var DRAG_HIGHLIGHT_SHADOW = 'shadow-highlight';
export var DRAG_HOLDER_WRAPPER = 'drag-holder-wrapper';
export var DRAG_HIGHLIGHT_TABLE_CLASSNAME = 'draggable-highlight-table';
export function removeDragThumbnail() {
  var thumbnail = document.querySelector("#" + DRAG_THUMBNAIL_ID);

  if (thumbnail && document.body.contains(thumbnail)) {
    document.body.removeChild(thumbnail);
  }
}
export function removeDragHolderWrapper() {
  var thumbnail = document.querySelector("." + DRAG_HOLDER_WRAPPER);

  if (thumbnail && document.body.contains(thumbnail)) {
    document.body.removeChild(thumbnail);
  }
} // 移除整个document的所有高亮
// TODO: 多实例？？

export function removeDragHightlight() {
  var draggingElements = Array.from(document.querySelectorAll("." + DRAG_HIGHLIGHT_CLASSNAME));
  draggingElements.forEach(function (ele) {
    var _ele$classList;

    if ((_ele$classList = ele.classList) != null && _ele$classList.remove) {
      ele.classList.remove(DRAG_HIGHLIGHT_CLASSNAME, DRAG_HIGHLIGHT_SHADOW);
    }
  });
}
export function addDragHighlight(itemDom, disableHighlight) {
  var _itemDom$classList;

  if (itemDom != null && (_itemDom$classList = itemDom.classList) != null && _itemDom$classList.add) {
    itemDom.classList.add(DRAG_HIGHLIGHT_CLASSNAME);
    disableHighlight && itemDom.classList.add(DISABLE_DRAG_HIGHLIGHT_CLASSNAME);

    if (itemDom != null && itemDom.querySelector('[data-type="card"]')) {
      itemDom.classList.add(DRAG_HIGHLIGHT_SHADOW);
    }
  }
} // 表格需要定制化，不添加背景色

export function addDragTableHighlight(itemDom) {
  var _itemDom$classList2;

  if (itemDom != null && (_itemDom$classList2 = itemDom.classList) != null && _itemDom$classList2.add) {
    itemDom.classList.add(DRAG_HIGHLIGHT_TABLE_CLASSNAME);

    if (itemDom != null && itemDom.querySelector('[data-type="card"]')) {
      itemDom.classList.add(DRAG_HIGHLIGHT_SHADOW);
    }
  }
} // 移除表格的高亮

export function removeDragTableHightlightSingle(itemDom) {
  var _itemDom$classList3;

  if (itemDom != null && (_itemDom$classList3 = itemDom.classList) != null && _itemDom$classList3.remove) {
    itemDom.classList.remove(DRAG_HIGHLIGHT_TABLE_CLASSNAME, DRAG_HIGHLIGHT_SHADOW);
  }
} // 移除指定节点的高亮

export function removeDragHightlightSingle(itemDom) {
  var _itemDom$classList4;

  if (itemDom != null && (_itemDom$classList4 = itemDom.classList) != null && _itemDom$classList4.remove) {
    itemDom.classList.remove(DRAG_HIGHLIGHT_CLASSNAME, DRAG_HIGHLIGHT_SHADOW, DISABLE_DRAG_HIGHLIGHT_CLASSNAME);
  }
}
/**
 * @param blocks
 * @deprecated
 */

export function addDargHighlight(blocks, disableHighlight) {
  for (var i = 0; i < blocks.length; i++) {
    var _itemDom$classList5;

    var itemDom = getClosestDom(blocks[i]);

    if (itemDom != null && (_itemDom$classList5 = itemDom.classList) != null && _itemDom$classList5.add) {
      itemDom.classList.add(DRAG_HIGHLIGHT_CLASSNAME);
      disableHighlight && itemDom.classList.add(DISABLE_DRAG_HIGHLIGHT_CLASSNAME);

      if (itemDom != null && itemDom.querySelector('[data-type="card"]')) {
        itemDom.classList.add(DRAG_HIGHLIGHT_SHADOW);
      }
    }
  }
}
export { findDOMNodeByKey };
