// @ali/4ever-plugin-draggable/dist-web/bi/utils/getDropHolderDataFromViewKeys.js
import getClosestDom from "./getClosestDom";
import { DROP_POSITION } from "../types";
export default function getDropHolderDataFromViewKeys(controller, event, viewKeys) {
  var document = controller.value.document;
  var startViewKey = viewKeys[0];
  var endViewKey = viewKeys[viewKeys.length - 1];
  if (startViewKey === endViewKey) return undefined;
  var startViewNode = document.getNode(startViewKey);
  var endViewNode = document.getNode(endViewKey);
  var startNode = getClosestDom(startViewNode);
  var endNode = getClosestDom(endViewNode);
  if (!startNode || !endNode) return undefined; // 处理跨页的情形

  var startNodeRect = startNode.getBoundingClientRect();
  var endNodeRect = endNode.getBoundingClientRect();
  var clientY = event.clientY;

  if (clientY > startNodeRect.top && clientY < startNodeRect.top + startNodeRect.height) {
    // 在前
    return {
      rect: {
        top: startNodeRect.top,
        left: startNodeRect.left,
        width: startNodeRect.width,
        height: startNodeRect.height
      },
      pagePosition: DROP_POSITION.before
    };
  } else {
    // 在后
    return {
      rect: {
        top: endNodeRect.top,
        left: endNodeRect.left,
        width: endNodeRect.width,
        height: endNodeRect.height
      },
      pagePosition: DROP_POSITION.after
    };
  }
}
