// @ali/we-design-next/dist-web/form/date-picker/panels/DateTimePanel/DateTimePanel.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
var _excluded = ["value", "showTime", "curPanel", "onSelect", "showNow", "timeFormat"];
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { DatePanel } from "../DatePanel";
import { TimePanel } from "../TimePanel";
import { DateTimePanelWrapper } from "./styled";
export var DateTimePanel = /*#__PURE__*/React.memo(function (props) {
  var value = props.value,
    showTime = props.showTime,
    curPanel = props.curPanel,
    onSelect = props.onSelect,
    showNow = props.showNow,
    timeFormat = props.timeFormat,
    rest = _objectWithoutPropertiesLoose(props, _excluded);
  return /*#__PURE__*/_createElement(DateTimePanelWrapper, null, /*#__PURE__*/_createElement(DatePanel, _extends({
    curPanel: curPanel,
    value: value,
    onSelect: onSelect,
    showTime: showTime,
    showNow: showNow
  }, rest)), showTime && curPanel === 'date' ? /*#__PURE__*/_createElement(TimePanel, {
    value: value,
    onSelect: onSelect,
    showNow: showNow,
    timeFormat: timeFormat
  }) : null);
});
