// @ali/4ever-plugin-pc-contex-menu/dist-node/utils/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Inflat = exports.HotKey = exports.Name = exports.Icon = exports.MenuItem = exports.MenuGroup = exports.MenuWrapper = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _everCangjie = require("@ali/4ever-cangjie");

var _util = require("./util");

const MENU_SIZE = 174;

const MenuWrapper = /*#__PURE__*/_styledComponents.default.div(["position:absolute;z-index:", ";opacity:0;background-color:#fff;box-shadow:0px 0px 1px rgba(0,0,0,0.24),0px 8px 24px rgba(0,0,0,0.16);border-radius:4px;will-change:top,right,bottom,left;"], _everCangjie.constants.Z_INDEX_CONTEXTMENU);

exports.MenuWrapper = MenuWrapper;

const MenuGroup = /*#__PURE__*/_styledComponents.default.div([""]);

exports.MenuGroup = MenuGroup;

const MenuItem = /*#__PURE__*/_styledComponents.default.span(["display:", ";flex-direction:row;align-items:center;width:", ";border-radius:8px;"], ({
  inline
}) => inline ? 'inline-flex' : 'flex', ({
  inline
}) => inline ? 'auto' : (0, _util.toPx)(MENU_SIZE));

exports.MenuItem = MenuItem;

const Icon = /*#__PURE__*/_styledComponents.default.span(["margin-right:4p;line-height:0;"]);

exports.Icon = Icon;

const Name = /*#__PURE__*/_styledComponents.default.span([""]);

exports.Name = Name;

const HotKey = /*#__PURE__*/_styledComponents.default.span(["color:", ";"], ({
  disabled
}) => disabled ? 'unset' : 'rgba(23, 26, 29, 0.4)');

exports.HotKey = HotKey;

const Inflat = /*#__PURE__*/_styledComponents.default.span(["flex:1;"]);

exports.Inflat = Inflat;
