// @ali/4ever-bi/dist-web/plugins/onlineVideo/components/OnlineVideoInlineToolbar.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Button, DeleteBetaNormal } from '@ali/we-design';
import { removeOnlineVideo } from "../actions";
import { InlineToolbar } from "../../../index";

var _ref = /*#__PURE__*/_createElement(DeleteBetaNormal, null);

/* eslint-disable */
var OnlineVideoInlineToolbar = function OnlineVideoInlineToolbar(props) {
  var _props$locale = props.locale,
      locale = _props$locale === void 0 ? {} : _props$locale,
      triggerNode = props.triggerNode;

  var onDeleteOnlineVideo = function onDeleteOnlineVideo(e) {
    e.preventDefault();
    e.stopPropagation();
    var controller = props.controller,
        node = props.node;
    controller.run('onAction', removeOnlineVideo(node));
  };

  var DelOnlineVideoButton = function DelOnlineVideoButton() {
    return /*#__PURE__*/_createElement(Button, {
      className: "onlinevideo-toolbar-delete",
      onClick: onDeleteOnlineVideo,
      title: locale["delete"],
      bordered: false,
      icon: _ref
    });
  };

  var customToolButtons = {
    'onlineVideo.del': DelOnlineVideoButton
  };
  var onlineVideoLayout = [{
    toolButtons: [{
      name: 'onlineVideo.del'
    }]
  }];
  return /*#__PURE__*/_createElement(InlineToolbar, {
    visible: true,
    triggerNode: triggerNode || undefined,
    customToolButtons: customToolButtons,
    inlineLayout: onlineVideoLayout
  });
};

export default /*#__PURE__*/React.memo(OnlineVideoInlineToolbar);
