// @ali/4ever-bi/dist-node/plugins/clearStyle/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _actions = require("./actions");

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _createRenderToolbar = _interopRequireDefault(require("./createRenderToolbar"));

var _commands = require("./commands");

function createClearStylePlugin(config) {
  const clearStyleConfig = config.clearStyle || {};
  return {
    toolbar: (0, _toolbar.default)(clearStyleConfig),
    menubar: (0, _menubar.default)(clearStyleConfig),
    renderToolbar: (0, _createRenderToolbar.default)(clearStyleConfig.locale),
    commands: (0, _commands.createCommands)(clearStyleConfig),
    actions: _actions.actions,
    onAction: _onAction.default,
    onKeyDown: _onKeyDown.default,
    models: []
  };
}

createClearStylePlugin.actions = _actions.DEPRECATED_ACTIONS;
var _default = createClearStylePlugin;
exports.default = _default;
