// @ali/4ever-plugin-template-button/dist-node/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.moFactory;
  }
});
Object.defineProperty(exports, "zhiFactory", {
  enumerable: true,
  get: function () {
    return _index2.zhiFactory;
  }
});
Object.defineProperty(exports, "biFactory", {
  enumerable: true,
  get: function () {
    return _index3.biFactory;
  }
});
Object.defineProperty(exports, "mobileZhiFactory", {
  enumerable: true,
  get: function () {
    return _mobile.mobileZhiFactory;
  }
});
Object.defineProperty(exports, "mobileBiFactory", {
  enumerable: true,
  get: function () {
    return _mobile2.mobileBiFactory;
  }
});
Object.defineProperty(exports, "TemplateButton", {
  enumerable: true,
  get: function () {
    return _models.TemplateButton;
  }
});
Object.defineProperty(exports, "ELEMENT_TYPE", {
  enumerable: true,
  get: function () {
    return _meta.ELEMENT_TYPE;
  }
});
Object.defineProperty(exports, "TemplateButtonRole", {
  enumerable: true,
  get: function () {
    return _role.TemplateButtonRole;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index4.default;
  }
});
exports.insertTemplateButton = void 0;

var _index = require("./mo/index");

var _index2 = require("./zhi/index");

var _index3 = require("./bi/index");

var _mobile = require("./zhi/mobile");

var _mobile2 = require("./bi/mobile");

var _actions = require("./bi/actions");

require("./utils/types");

var _models = require("./mo/models");

var _meta = require("./utils/meta");

var _role = require("./utils/role");

var _index4 = _interopRequireDefault(require("./mo/serializer/index"));

// make types combination work
// facotry
// Element Model
// serializer
// @deprecated
const insertTemplateButton = _actions.actions.insertTemplateButton;
exports.insertTemplateButton = insertTemplateButton;
