// @ali/4ever-bi/dist-node/plugins/video/createRenderNode.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderNode;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _video = _interopRequireDefault(require("./components/video"));

const _createElement = /*#__PURE__*/React.createElement;

function createRenderNode(config) {
  const videoConfig = config.video || {};
  const {
    locale: videoLocaleConfig = {},
    downloadFile,
    transformVideoURL,
    onMediaLoadSuccess,
    onMediaLoadError,
    setTempPosterUrl,
    renderVideoMask,
    allowDownload,
    allowFullscreen
  } = videoConfig;
  return {
    embed: (props, controller, next) => {
      const {
        node
      } = props;

      if (node.data.type === 'video' && node.data.src) {
        return /*#__PURE__*/_createElement(_video.default, (0, _extends2.default)({}, props, {
          controller: controller,
          onDownload: downloadFile,
          locale: videoLocaleConfig,
          transformVideoURL: transformVideoURL,
          onMediaLoadSuccess: onMediaLoadSuccess,
          onMediaLoadError: onMediaLoadError,
          setTempPosterUrl: setTempPosterUrl,
          renderVideoMask: renderVideoMask,
          allowDownload: allowDownload,
          allowFullscreen: allowFullscreen
        }));
      }

      return next();
    }
  };
}
