// @ali/4ever-plugin-clipboard/dist-node/bi/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.biFactory = void 0;

var _createOnCangjiePaste = _interopRequireDefault(require("./handlers/createOnCangjiePaste"));

var _createOnCangjieCopyOrCut = _interopRequireDefault(require("../common/handlers/createOnCangjieCopyOrCut"));

var _onCangjieInput = _interopRequireDefault(require("./handlers/onCangjieInput"));

var _ClipboardPad = _interopRequireDefault(require("../utils/ClipboardPad"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _createOnEditable = _interopRequireDefault(require("./handlers/createOnEditable"));

var _copyNodes = require("../common/queries/copyNodes");

var _getClipboardData = _interopRequireDefault(require("../common/queries/getClipboardData"));

var _meta = require("../utils/meta");

var _commands = require("./commands");

var _actions = require("./actions");

const biFactory = configs => {
  const clipboard = configs.clipboard || {};
  const onCangjieCopyOrCut = (0, _createOnCangjieCopyOrCut.default)(clipboard);
  return {
    onConstruct: (controller, next) => {
      _ClipboardPad.default.getClipboardPad({
        serializer: controller.serializer,
        schema: controller.strictSchema,
        ...clipboard
      }, clipboard.key);

      return next();
    },
    key: _meta.key,
    queries: {
      copyNodes: _copyNodes.copyNodes,
      getClipboardData: (0, _getClipboardData.default)(clipboard)
    },
    actions: _actions.actions,
    commands: _commands.commands,
    onAction: (0, _onAction.default)(clipboard),
    onCangjieInput: _onCangjieInput.default,
    onKeyDown: _onKeyDown.default,
    onCangjieCopy: onCangjieCopyOrCut,
    onCangjieCut: onCangjieCopyOrCut,
    onCangjiePaste: (0, _createOnCangjiePaste.default)(clipboard),
    renderEditable: (0, _createOnEditable.default)(clipboard)
  };
};

exports.biFactory = biFactory;
