// @ali/4ever-plugin-columns/dist-node/utils/hooks/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createRenderNode", {
  enumerable: true,
  get: function () {
    return _renderNode.default;
  }
});
Object.defineProperty(exports, "createPlaceholder", {
  enumerable: true,
  get: function () {
    return _placeholder.default;
  }
});
Object.defineProperty(exports, "createOnAction", {
  enumerable: true,
  get: function () {
    return _onAction.default;
  }
});
Object.defineProperty(exports, "createOnCangjieSelect", {
  enumerable: true,
  get: function () {
    return _onCangjieSelect.default;
  }
});
Object.defineProperty(exports, "createOnKeyDown", {
  enumerable: true,
  get: function () {
    return _onKeyDown.default;
  }
});
Object.defineProperty(exports, "createOnCut", {
  enumerable: true,
  get: function () {
    return _onCut.default;
  }
});
Object.defineProperty(exports, "beforeCopySelectedNodes", {
  enumerable: true,
  get: function () {
    return _beforeCopySelectedNodes.default;
  }
});

var _renderNode = _interopRequireDefault(require("./renderNode"));

var _placeholder = _interopRequireDefault(require("./placeholder"));

var _onAction = _interopRequireDefault(require("./onAction"));

var _onCangjieSelect = _interopRequireDefault(require("./onCangjieSelect"));

var _onKeyDown = _interopRequireDefault(require("./onKeyDown"));

var _onCut = _interopRequireDefault(require("./onCut"));

var _beforeCopySelectedNodes = _interopRequireDefault(require("./beforeCopySelectedNodes"));
