// @ali/4ever-bi/dist-node/plugins/hr/components/getHrMenuList.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HrInlineMenu = exports.HrMenu = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../actions");

var _constants = require("../constants");

const _createElement = /*#__PURE__*/React.createElement;
const {
  hrLines
} = _everBamboo.HrPlugin;

const HrMenu = props => {
  const {
    menuItem,
    controller
  } = props;

  const handleClick = key => {
    const lineType = key.split('_')[1];
    const line = hrLines.find(item => {
      return item.lineType === lineType;
    });

    if (line) {
      controller.run('onAction', (0, _actions.insertHr)({
        type: line.lineType,
        color: _constants.DEFAULT_HR_COLOR,
        sz: line.size
      }));
    }
  };

  const lines = hrLines.map(line => {
    return /*#__PURE__*/_createElement(_weDesign.MenuItem, {
      testid: `hr_${line.lineType}`,
      menuKey: `hr_${line.lineType}`,
      onClick: handleClick,
      key: line.lineType
    }, /*#__PURE__*/_createElement(_weDesign.CommonLine, {
      lineType: line.lineType,
      length: 134,
      size: line.size
    }));
  });
  return /*#__PURE__*/_createElement(_weDesign.SubMenu, (0, _extends2.default)({}, menuItem, {
    overlayClassName: "menu-hr-overlay",
    title: menuItem.title || '',
    maxHeight: "600"
  }), lines);
};

exports.HrMenu = HrMenu;

const HrInlineMenu = props => {
  const {
    menuItem,
    controller
  } = props;
  const { ...rest
  } = menuItem;

  const handleClick = () => {
    const line = hrLines[0];
    controller.run('onAction', (0, _actions.insertHr)({
      type: line.lineType,
      color: _constants.DEFAULT_HR_COLOR,
      sz: line.size
    }));
  };

  return /*#__PURE__*/_createElement(_weDesign.InlineMenuItem, (0, _extends2.default)({}, rest, {
    onClick: handleClick
  }));
};

exports.HrInlineMenu = HrInlineMenu;
