// @ali/4ever-bi/dist-node/plugins/hr/components/hr.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everCangjie = require("@ali/4ever-cangjie");

var _weDesign = require("@ali/we-design");

var _updateHr = _interopRequireDefault(require("../commands/updateHr"));

var _hrControl = _interopRequireDefault(require("./hrControl"));

var _index = require("../../../index");

const _createElement = /*#__PURE__*/React.createElement;
const hrLayout = [{
  toolButtons: [{
    name: 'hr.color'
  }, {
    name: 'hr.style'
  }, {
    name: 'hr.weight'
  }]
}];

const Hr = props => {
  const {
    isSelected,
    payload,
    attributes
  } = props;

  const controller = _everCangjie.Controller.useController();

  const disableInlineToolbar = controller.query('hasHoverToolbar', 'hr');
  return /*#__PURE__*/_createElement(_index.InlineToolbar, {
    visible: !disableInlineToolbar && !_everCangjie.environment.IS_MOBILE && isSelected,
    inlineLayout: hrLayout
  }, /*#__PURE__*/_createElement(_hrControl.default, (0, _extends2.default)({
    attributes: attributes,
    onHrChange: _updateHr.default,
    isSelected: isSelected
  }, props), /*#__PURE__*/_createElement(_weDesign.CommonLine, payload)));
};

var _default = Hr;
exports.default = _default;
