// @ali/4ever-bi/dist-node/plugins/hr/components/quickInsertPanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _weDesign = require("@ali/we-design");

var _actions = require("../actions");

const _createElement = /*#__PURE__*/React.createElement;

const LineWrapper = /*#__PURE__*/_styledComponents.default.div(["width:100%;padding:10px;&:hover{background-color:", ";}"], () => _weDesign.theme.colorFill4);

const Line = /*#__PURE__*/React.memo(props => {
  const {
    lineStyle,
    onClick
  } = props;
  const defaultSize = _weDesign.commonLineMap[lineStyle].defaultSize;
  const handleClick = React.useCallback((style, width) => {
    onClick(style, width);
  }, [onClick]);
  return /*#__PURE__*/_createElement(LineWrapper, {
    onClick: () => handleClick(lineStyle, defaultSize)
  }, /*#__PURE__*/_createElement(_weDesign.CommonLine, {
    color: "black",
    size: defaultSize,
    lineType: lineStyle
  }));
});

const QuickInsertPanel = props => {
  const {
    controller,
    close
  } = props;
  const handleInsert = React.useCallback((lineStyle, width) => {
    close();
    controller.run('onAction', (0, _actions.insertHr)({
      type: lineStyle,
      color: 'black',
      sz: width
    }));
  }, [close, controller]);
  return /*#__PURE__*/_createElement("div", null, Object.keys(_weDesign.commonLineMap).map(lineStyle => /*#__PURE__*/_createElement(Line, {
    key: lineStyle,
    onClick: handleInsert,
    lineStyle: lineStyle
  })));
};

var _default = QuickInsertPanel;
exports.default = _default;
