// @ali/4ever-bi/dist-web/plugins/formula/index.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import Wrapper from "./components/wrapper";
import FormulaView from "./components/view";
import FormulaPlaceholder from "./components/placeholder";
import { tagFactory } from '@ali/4ever-factory';
import onAction from "./handlers/onAction";
import { TagType } from "./types";
import "./fonts.css";
import { moFactory } from '@ali/4ever-plugin-formula';

function createRenderNode(config) {
  return function renderNode(props) {
    return /*#__PURE__*/_createElement(Wrapper, _extends({}, props, {
      config: config
    }));
  };
}

var createFormulaPlugin = tagFactory(function (config) {
  var renderNode = createRenderNode(config.formula);
  return {
    tagType: TagType.Formula,
    renderNode: renderNode,
    onAction: onAction
  };
}, moFactory);
createFormulaPlugin.components = {
  FormulaView: FormulaView,
  FormulaPlaceholder: FormulaPlaceholder
};
export default createFormulaPlugin;
