// @ali/4ever-bi/dist-node/plugins/indent/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DEFAULT_INDENT", {
  enumerable: true,
  get: function () {
    return _everUtils.DEFAULT_INDENT;
  }
});
Object.defineProperty(exports, "IndentTrigger", {
  enumerable: true,
  get: function () {
    return _everUtils.IndentTrigger;
  }
});
Object.defineProperty(exports, "IndentType", {
  enumerable: true,
  get: function () {
    return _everUtils.IndentType;
  }
});
Object.defineProperty(exports, "INDENT_SIZE", {
  enumerable: true,
  get: function () {
    return _everPluginIndent.INDENT_SIZE;
  }
});
Object.defineProperty(exports, "INDENT_SIZE_CHARS", {
  enumerable: true,
  get: function () {
    return _everPluginIndent.INDENT_SIZE_CHARS;
  }
});
Object.defineProperty(exports, "MIN_INDENT_SIZE", {
  enumerable: true,
  get: function () {
    return _everPluginIndent.MIN_INDENT_SIZE;
  }
});
Object.defineProperty(exports, "MAX_INDENT", {
  enumerable: true,
  get: function () {
    return _everPluginIndent.MAX_INDENT;
  }
});
Object.defineProperty(exports, "MAX_INDENT_SIZE", {
  enumerable: true,
  get: function () {
    return _everPluginIndent.MAX_INDENT_SIZE;
  }
});
Object.defineProperty(exports, "MAX_INDENT_SIZE_CHARS", {
  enumerable: true,
  get: function () {
    return _everPluginIndent.MAX_INDENT_SIZE_CHARS;
  }
});

var _everUtils = require("@ali/4ever-utils");

var _everPluginIndent = require("@ali/4ever-plugin-indent");
