// @ali/4ever-bi/dist-node/plugins/subject/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toggleFinished = toggleFinished;
exports.insertSubject = insertSubject;
exports.insertBlock = insertBlock;
exports.highlightSubject = highlightSubject;
exports.HIGHLIGHT_SUBJECT = exports.INSERT_BLOCK = exports.INSERT_SUBJECT = exports.TOGGLE_FINISHED = void 0;

/**
 * 设置议题完成状态
 */
const TOGGLE_FINISHED = Symbol('TOGGLE_FINISHED');
exports.TOGGLE_FINISHED = TOGGLE_FINISHED;

function toggleFinished(payload) {
  return {
    type: TOGGLE_FINISHED,
    payload
  };
}
/**
 * 插入议题
 */


const INSERT_SUBJECT = Symbol('INSERT_SUBJECT');
exports.INSERT_SUBJECT = INSERT_SUBJECT;

function insertSubject(payload) {
  return {
    type: INSERT_SUBJECT,
    payload
  };
}
/**
 * 插入空段落
 */


const INSERT_BLOCK = Symbol('INSERT_BLOCK');
exports.INSERT_BLOCK = INSERT_BLOCK;

function insertBlock(payload) {
  return {
    type: INSERT_BLOCK,
    payload
  };
}
/**
 * 高亮议题区块
 */


const HIGHLIGHT_SUBJECT = Symbol('HIGHLIGHT_SUBJECT');
exports.HIGHLIGHT_SUBJECT = HIGHLIGHT_SUBJECT;

function highlightSubject(payload) {
  return {
    type: HIGHLIGHT_SUBJECT,
    payload
  };
}
