// @ali/4ever-bi/dist-node/plugins/subject/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "isSubjectHeading", {
  enumerable: true,
  get: function () {
    return _everPluginSubject.isSubjectHeading;
  }
});
Object.defineProperty(exports, "isSubject", {
  enumerable: true,
  get: function () {
    return _everPluginSubject.isSubject;
  }
});
Object.defineProperty(exports, "createSubject", {
  enumerable: true,
  get: function () {
    return _everPluginSubject.createSubject;
  }
});
Object.defineProperty(exports, "isEmptyNode", {
  enumerable: true,
  get: function () {
    return _everPluginSubject.isEmptyNode;
  }
});
Object.defineProperty(exports, "isNormalParagraph", {
  enumerable: true,
  get: function () {
    return _everPluginSubject.isNormalParagraph;
  }
});
Object.defineProperty(exports, "uniqueId", {
  enumerable: true,
  get: function () {
    return _everPluginSubject.uniqueId;
  }
});

var _everPluginSubject = require("@ali/4ever-plugin-subject");
