// @ali/4ever-plugin-mention/dist-node/zhi/handlers/createRenderNode.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createRenderNode = createRenderNode;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _mention = _interopRequireDefault(require("../mention"));

const _createElement = /*#__PURE__*/React.createElement;

function createRenderNode(mentionConfig) {
  const {
    onMentionClick,
    onMentionHover,
    getCurrentUser,
    locale = {}
  } = mentionConfig;

  const renderNode = (props, _) => {
    const {
      node
    } = props;
    const {
      attributes
    } = props;
    return /*#__PURE__*/_createElement(_mention.default, (0, _extends2.default)({}, attributes, {
      node: node,
      onMentionClick: onMentionClick,
      onMentionHover: onMentionHover,
      getCurrentUser: getCurrentUser,
      locale: locale
    }));
  };

  return {
    mention: renderNode
  };
}
