// @ali/4ever-plugin-inline-code/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "biFactory", {
  enumerable: true,
  get: function () {
    return _bi.default;
  }
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _mo.default;
  }
});
Object.defineProperty(exports, "zhiFactory", {
  enumerable: true,
  get: function () {
    return _zhi.default;
  }
});
Object.defineProperty(exports, "InlineCode", {
  enumerable: true,
  get: function () {
    return _models.default;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _serializer.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _serializer.createServerSerializationPlugin;
  }
});
Object.defineProperty(exports, "INLINECODE_TOGGLE", {
  enumerable: true,
  get: function () {
    return _actions.INLINECODE_TOGGLE;
  }
});
Object.defineProperty(exports, "toggleInlineCode", {
  enumerable: true,
  get: function () {
    return _actions.toggleInlineCode;
  }
});
Object.defineProperty(exports, "InlineCodeRole", {
  enumerable: true,
  get: function () {
    return _role.InlineCodeRole;
  }
});
exports.mobileBiFactory = exports.mobileZhiFactory = void 0;

var _bi = _interopRequireDefault(require("./bi"));

var _mo = _interopRequireDefault(require("./mo"));

var _zhi = _interopRequireDefault(require("./zhi"));

var _models = _interopRequireDefault(require("./mo/models"));

var _serializer = _interopRequireWildcard(require("./mo/serializer"));

var _actions = require("./bi/actions");

var _role = require("./role");

const mobileBiFactory = _bi.default;
exports.mobileBiFactory = mobileBiFactory;
const mobileZhiFactory = _zhi.default; // factory

exports.mobileZhiFactory = mobileZhiFactory;
