// @ali/4ever-bi/dist-node/plugins/blockquote/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BLOCKQUOTE_TOGGLE", {
  enumerable: true,
  get: function () {
    return _everUtils.BLOCKQUOTE_TOGGLE;
  }
});
Object.defineProperty(exports, "BLOCKQUOTE_ENABLE", {
  enumerable: true,
  get: function () {
    return _everUtils.BLOCKQUOTE_ENABLE;
  }
});
Object.defineProperty(exports, "BLOCKQUOTE_TRANSFORM", {
  enumerable: true,
  get: function () {
    return _everUtils.BLOCKQUOTE_TRANSFORM;
  }
});
Object.defineProperty(exports, "BLOCKQUOTE_DISABLE", {
  enumerable: true,
  get: function () {
    return _everUtils.BLOCKQUOTE_DISABLE;
  }
});
Object.defineProperty(exports, "toggleBlockquote", {
  enumerable: true,
  get: function () {
    return _everUtils.toggleBlockquote;
  }
});
Object.defineProperty(exports, "transformBlockquote", {
  enumerable: true,
  get: function () {
    return _everUtils.transformBlockquote;
  }
});
Object.defineProperty(exports, "enableBlockquote", {
  enumerable: true,
  get: function () {
    return _everUtils.enableBlockquote;
  }
});
Object.defineProperty(exports, "disableBlockquote", {
  enumerable: true,
  get: function () {
    return _everUtils.disableBlockquote;
  }
});

var _everUtils = require("@ali/4ever-utils");
