// @ali/4ever-bi/dist-node/plugins/blockquote/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _renderNode = _interopRequireDefault(require("./renderNode"));

var _toggleQuote = _interopRequireDefault(require("./commands/toggleQuote"));

var _quickInsertConfig = _interopRequireDefault(require("./quickInsertConfig"));

var actions = _interopRequireWildcard(require("./actions"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _onCangjieInput = _interopRequireDefault(require("./handlers/onCangjieInput"));

var _queries = require("./queries");

var _everPluginBlockquote = require("@ali/4ever-plugin-blockquote");

const commands = {
  toggleQuote: _toggleQuote.default
};

function createBlockquotePlugin(config) {
  const blockquoteConfig = config.blockquote || {};
  return { ...(0, _everPluginBlockquote.moFactory)({}),
    toolbar: (0, _toolbar.default)(blockquoteConfig),
    menubar: (0, _menubar.default)(),
    renderNode: _renderNode.default,
    commands,
    quickInsert: _quickInsertConfig.default,
    onAction: _onAction.default,
    onKeyDown: _onKeyDown.default,
    queries: _queries.queries,
    onCangjieInput: _onCangjieInput.default,
    models: {
      data: {
        shd: [undefined, Object]
      }
    }
  };
}

createBlockquotePlugin.actions = actions;
createBlockquotePlugin.commands = commands;
var _default = createBlockquotePlugin;
exports.default = _default;
