// @ali/4ever-bi/dist-node/plugins/callout/components/next/stickerSwitch.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StickerSwitchButton = void 0;

var _react = _interopRequireDefault(require("react"));

var _everComponent = require("@ali/4ever-component");

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../../actions");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement(_everComponent.IconEmoji, null);

const StickerSwitchButton = props => {
  const {
    controller,
    locale,
    getNode
  } = props;
  const node = getNode();

  const [showStk, setShowStk] = _react.default.useState(node?.data?.metadata?.showstk || false);

  const handleClick = _react.default.useCallback(() => {
    setShowStk(v => {
      const data = {}; // @ts-ignore

      data.showstk = !v;
      const node = getNode();
      controller.run('onCangjieBlur');
      node && controller.run('onAction', (0, _actions.setMetaData)({
        node,
        data,
        needBlur: true
      }));
      return !v;
    });
  }, []);

  if (!node) return null;
  return /*#__PURE__*/_createElement(_everComponent.IconButton, {
    onClick: handleClick,
    active: showStk,
    icon: _ref,
    tooltip: locale?.showEmoji || 'Show sticker',
    role: _everBamboo.PluginRoles.calloutSticker
  });
};

exports.StickerSwitchButton = StickerSwitchButton;
