// @ali/4ever-plugin-mention/dist-web/bi/commands/setMention.js
import { Commands, Inline } from '@ali/4ever-cangjie';
import removeMention from "./removeMention";
import { nearsetAt } from "../utils";
export default (function (controller, suggestion, at) {
  if (!suggestion) return controller;
  var value = controller.value;
  var selection = value.selection,
      document = value.document,
      injections = value.injections;
  var start = selection.getStart(document);
  var offset = nearsetAt(value.endText.text, start.offset);
  var mentionLength = start.offset - offset;
  var inline = Inline.create({
    type: 'mention',
    data: {
      id: suggestion.id,
      login: suggestion.login,
      name: suggestion.name,
      avatarUrl: suggestion.avatarUrl
    }
  }); // 如果是 placeholder 类型的 mention 做 setData

  if (at) {
    var excludeInjections = injections.filter(function (i) {
      return i.key !== at;
    });
    return controller.setInjections([].concat(excludeInjections)).command(Commands.setNodeByKey, at, {
      data: inline.data
    }).command(Commands.focus).command(Commands.moveToStartOfNextText);
  }

  controller.command(Commands.deleteBackward, mentionLength).command(Commands.focus).command(Commands.insertInline, inline).command(Commands.moveToStartOfNextText);
  return removeMention(controller);
});
