// @ali/4ever-bi/dist-node/plugins/blockquote/mobile/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _renderNode = _interopRequireDefault(require("../renderNode"));

var _toggleQuote = _interopRequireDefault(require("../commands/toggleQuote"));

var actions = _interopRequireWildcard(require("../actions"));

var _onAction = _interopRequireDefault(require("../handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("../handlers/onKeyDown"));

var _onCangjieInput = _interopRequireDefault(require("../handlers/onCangjieInput"));

var _queries = require("../queries");

var _everPluginBlockquote = require("@ali/4ever-plugin-blockquote");

const commands = {
  toggleQuote: _toggleQuote.default
};

function createBlockquotePlugin(config) {
  const blockquoteConfig = config.blockquote || {};
  return { ...(0, _everPluginBlockquote.moFactory)({}),
    renderNode: _renderNode.default,
    commands,
    onAction: _onAction.default,
    onKeyDown: _onKeyDown.default,
    onCangjieInput: _onCangjieInput.default,
    queries: _queries.queries,
    models: {
      data: {
        shd: [undefined, Object]
      }
    }
  };
}

createBlockquotePlugin.actions = actions;
createBlockquotePlugin.commands = commands;
var _default = createBlockquotePlugin;
exports.default = _default;
