// @ali/we-design-next/dist-web/basic/split-button/SplitButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
var _excluded = ["secondary", "children", "icon", "secondaryVisible", "onSecondaryVisibleChange"];
import React, { memo, forwardRef } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { ArrowUpDown12, useDefaultProp, proxyRestProps } from "../../common";
import { Popover } from "../../display/popover";
import { Button, Main, Secondary } from "./styled";
export var SplitButton = /*#__PURE__*/memo( /*#__PURE__*/forwardRef(function (props, ref) {
  var _props$secondary = props.secondary,
    secondary = _props$secondary === void 0 ? /*#__PURE__*/_createElement(React.Fragment, null) : _props$secondary,
    children = props.children,
    icon = props.icon,
    _secondaryVisible = props.secondaryVisible,
    onSecondaryVisibleChange = props.onSecondaryVisibleChange,
    buttonProps = _objectWithoutPropertiesLoose(props, _excluded);
  var _useDefaultProp = useDefaultProp(false, _secondaryVisible, onSecondaryVisibleChange),
    secondaryVisible = _useDefaultProp[0],
    setSecondaryVisible = _useDefaultProp[1];
  var _buttonProps$type = buttonProps.type,
    type = _buttonProps$type === void 0 ? 'normal' : _buttonProps$type,
    _buttonProps$size = buttonProps.size,
    size = _buttonProps$size === void 0 ? 'normal' : _buttonProps$size;
  return /*#__PURE__*/_createElement(Popover, {
    content: secondary,
    visible: secondaryVisible,
    onVisibleChange: setSecondaryVisible,
    placement: "bottomRight",
    placementOffset: [4, 0]
  }, /*#__PURE__*/_createElement(Button, buttonProps, /*#__PURE__*/_createElement(Main, _extends({
    size: size,
    btnType: type
  }, proxyRestProps(buttonProps, 'main')), icon, children), /*#__PURE__*/_createElement(Secondary, _extends({
    size: size,
    btnType: type,
    onClick: function onClick(e) {
      setSecondaryVisible(!secondaryVisible);
      e.stopPropagation();
    }
  }, proxyRestProps(buttonProps, 'secondary')), /*#__PURE__*/_createElement(ArrowUpDown12, {
    dir: secondaryVisible ? 'up' : 'down'
  }))));
}));
SplitButton.displayName = 'SplitButton';
