// @ali/4ever-bi/dist-web/plugins/callout/actions.js
export var INSERT_CALLOUT = Symbol('INSERT_CALLOUT');
export var DELETE_CALLOUT = Symbol('DELETE_CALLOUT');
export var SET_CALLOUT_STICKER_CODE = Symbol('SET_CALLOUT_STICKER_CODE');
export var SET_CALLOUT_BACKGROUND = Symbol('SET_CALLOUT_BACKGROUND');
export var SET_CALLOUT_SELECTED = Symbol('SET_CALLOUT_SELECTED');
export var SET_META_DATA = Symbol('SET_META_DATA');
export var CALLOUT_DATA_CHANGED = Symbol('CALLOUT_DATA_CHANGED');
export var CLEAR_CALLOUT_SELECTED = Symbol('CLEAR_CALLOUT_SELECTED');
export var TRANSFORM_CALLOUT = Symbol('TRANSFORM_CALLOUT');
export var CONVERT_KEYS_TO_CONTAINER = Symbol('CONVERT_KEYS_TO_CONTAINER');
export function insertCallout(_payload) {
  return {
    type: INSERT_CALLOUT,
    payload: _payload || {
      text: ''
    }
  };
}
export function transformCallout() {
  return {
    type: TRANSFORM_CALLOUT
  };
}
export function deleteCallout(payload) {
  return {
    type: DELETE_CALLOUT,
    payload: payload
  };
}
export function setCalloutStickerCode(payload) {
  return {
    type: SET_CALLOUT_STICKER_CODE,
    payload: payload
  };
}
export function setCalloutBackground(payload) {
  return {
    type: SET_CALLOUT_BACKGROUND,
    payload: payload
  };
}
export function setCalloutSelected(payload) {
  return {
    type: SET_CALLOUT_SELECTED,
    payload: payload
  };
}
export function setMetaData(payload) {
  return {
    type: SET_META_DATA,
    payload: payload
  };
}
export function clearCalloutSelected() {
  return {
    type: CLEAR_CALLOUT_SELECTED
  };
}
export function convertKeysToContainer(payload) {
  return {
    type: CONVERT_KEYS_TO_CONTAINER,
    payload: payload
  };
}
export function broadcastDataChange(obj) {
  var sc = obj.sticker,
      bg = obj.bgcolor,
      border = obj.border,
      showstk = obj.showstk;
  var payload = {
    sc: sc,
    bg: bg,
    border: border,
    showstk: showstk
  };
  return {
    type: CALLOUT_DATA_CHANGED,
    payload: payload
  };
}
