// @ali/we-toolbar/dist-web/common/cardLayout/placeholder/placeholderH5.js
import React, { useMemo, useRef } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { CardPlaceholderContainer } from "./styled";
import { CardContainer } from "../container";
import { uuid } from '@ali/we-design';
var delay = 200;
export var PlaceholderH5 = function PlaceholderH5(props) {
  var _props$height = props.height,
      height = _props$height === void 0 ? 65 : _props$height,
      title = props.title,
      icon = props.icon,
      desc = props.desc,
      _props$type = props.type,
      type = _props$type === void 0 ? 'export' : _props$type,
      _props$isSelected = props.isSelected,
      isSelected = _props$isSelected === void 0 ? false : _props$isSelected,
      _props$clickable = props.clickable,
      clickable = _props$clickable === void 0 ? true : _props$clickable,
      thresholdWidth = props.thresholdWidth,
      onMouseDown = props.onMouseDown;
  var id = useMemo(function () {
    return uuid();
  }, []);
  var cardPlaceholderContainerRef = useRef(null);
  return type === 'normal' ? /*#__PURE__*/_createElement(CardPlaceholderContainer, {
    delay: delay,
    height: height,
    isSelected: isSelected,
    clickable: clickable,
    onMouseDown: onMouseDown,
    ref: cardPlaceholderContainerRef,
    "data-id": "card-placeholder-container-" + id,
    type: type
  }, /*#__PURE__*/_createElement(CardContainer, {
    className: "card-container",
    thresholdWidth: thresholdWidth
  }, icon && /*#__PURE__*/_createElement("div", {
    className: "placeholder-icon-box"
  }, icon), /*#__PURE__*/_createElement("div", {
    className: "placeholder-right"
  }, /*#__PURE__*/_createElement("div", {
    className: "placeholder-top"
  }, /*#__PURE__*/_createElement("div", {
    className: "placeholder-title",
    style: !desc ? {
      lineHeight: '38px'
    } : {}
  }, title)), desc && /*#__PURE__*/_createElement("div", {
    className: "placeholder-desc"
  }, /*#__PURE__*/_createElement("div", {
    className: "desc-text"
  }, desc))))) : /*#__PURE__*/_createElement(CardPlaceholderContainer, {
    delay: delay,
    height: height,
    clickable: clickable,
    onMouseDown: onMouseDown,
    "data-id": "card-placeholder-container-" + id,
    type: type
  }, /*#__PURE__*/_createElement(CardContainer, {
    className: "card-container"
  }, icon && /*#__PURE__*/_createElement("div", {
    className: "placeholder-icon-box"
  }, icon), /*#__PURE__*/_createElement("div", {
    className: "placeholder-right"
  }, /*#__PURE__*/_createElement("div", {
    className: "placeholder-top"
  }, /*#__PURE__*/_createElement("div", {
    className: "placeholder-title"
  }, title)))));
};
