// @ali/we-toolbar/dist-web/common/cardLayout/placeholder/styled.js
import styled, { css, keyframes } from 'styled-components';
var light = {
  level3_base_color: 'rgba(23, 26, 29, 0.4)',
  level1_base_color: 'rgba(23,26,29,0.6)',
  level2_base_color: 'rgba(23, 26, 29, 0.8)',
  hover_bg_color: 'rgba(126, 134, 142, 0.12)',
  border_normal: 'rgba(0, 0, 0, 0.16)',
  line_light_color: 'rgba(126, 134, 142, 0.16)',
  bg_doc_color: 'rgba(245,246,248,1)',
  bg_doc_light_color: 'rgba(245,246,248,1)',
  blue1_color: 'rgba(0,137,255,1)',
  bg_select: 'rgba(126,134,142,0.10)'
};
var changeBgColor = /*#__PURE__*/keyframes(["0%{background-color:", ";}60%{background-color:", ";}to{background-color:", ";}}"], light.line_light_color, light.line_light_color, light.bg_select);
var selectBorder = /*#__PURE__*/css(["border:1px solid ", ";box-shadow:0 0 0 3px rgba(0,106,254,0.12);"], light.blue1_color);
var normalBorder = /*#__PURE__*/css(["border:1px dashed ", ";"], light.border_normal);
export var CardPlaceholderContainer = /*#__PURE__*/styled.div(["display:inline-block;box-sizing:border-box;width:100%;height:", ";border-radius:8px;color:", ";transition:", "ms;transition-timing-function:step-end;", " cursor:pointer;", " &:hover{.card-container{.placeholder-icon-box{font-size:16px;svg{color:", ";}}.placeholder-title{color:", ";}}", "}", " ", " ", " &.select{animation:", " 250ms cubic-bezier(0.39,0.575,0.565,1) forwards;border-color:", ";.card-container{.placeholder-title{color:", ";}.placeholder-icon-box{font-size:16px;svg{color:", ";}}}", "}&:active{background-color:", ";border-color:", ";.card-container{.placeholder-icon-box{font-size:16px;svg{color:", ";transition:0;}}.placeholder-title{color:", ";transition:0;}}", "}*{box-sizing:border-box;}.card-container{padding:0 0 12px 12px;width:100%;height:100%;display:flex;.placeholder-icon-box{font-size:16px;flex:none;width:40px;height:40px;", " margin-top:12px;margin-right:12px;border-radius:4px;img{object-fit:contain;width:100%;height:100%;}svg{transition:", "ms;transition-timing-function:step-end;font-size:1.2em;color:", ";}}.placeholder-right{flex:1 1 auto;height:100%;overflow:hidden;display:flex;flex-direction:column;.placeholder-top{width:100%;display:flex;flex:1 1 auto;.placeholder-title{flex:1 1 auto;padding-right:12px;margin-top:12px;align-self:center;transition:", "ms;transition-timing-function:step-end;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;font-style:normal;", " line-height:22px;}.placeholder-toolbar{margin:4px 4px 0 0;flex:none;}}.placeholder-desc{flex:none;width:100%;padding-right:12px;overflow:hidden;display:inline-flex;font-style:normal;font-weight:normal;font-size:12px;line-height:16px;", " .desc-text{flex:0 1 auto;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;}.refresh{flex:none;margin-left:5px;color:", ";&:hover{color:rgba(0,101,213,1);}cursor:pointer;}}}}"], function (_ref) {
  var height = _ref.height;
  return typeof height === 'number' ? height + 'px' : height;
}, light.level3_base_color, function (_ref2) {
  var delay = _ref2.delay;
  return delay;
}, function (_ref3) {
  var type = _ref3.type;
  return type === 'failure' ? css(["background:#f6f9fe;"]) : css(["background:", ";"], light.bg_doc_color);
}, function (_ref4) {
  var isSelected = _ref4.isSelected,
      type = _ref4.type;
  return isSelected ? selectBorder : type === 'normal' && normalBorder;
}, light.level2_base_color, light.level2_base_color, function (_ref5) {
  var isSelected = _ref5.isSelected,
      type = _ref5.type;
  return isSelected ? selectBorder : type === 'normal' && normalBorder;
}, function (_ref6) {
  var type = _ref6.type,
      isSelected = _ref6.isSelected;
  return !isSelected && type === 'normal' && css(["&:hover{border:1px dashed ", ";}"], light.blue1_color);
}, function (_ref7) {
  var isSelected = _ref7.isSelected;
  return !isSelected && css(["&.active{background:", ";border:1px dashed ", ";}"], light.line_light_color, light.border_normal);
}, function (_ref8) {
  var clickable = _ref8.clickable;
  return !clickable && css(["pointer-events:none;"]);
}, changeBgColor, light.line_light_color, light.level2_base_color, light.level2_base_color, function (_ref9) {
  var isSelected = _ref9.isSelected;
  return isSelected && selectBorder;
}, light.line_light_color, light.line_light_color, light.level1_base_color, light.level1_base_color, function (_ref10) {
  var isSelected = _ref10.isSelected;
  return isSelected && selectBorder;
}, function (_ref11) {
  var type = _ref11.type;
  return type !== 'failure' && css(["padding:11px 10px;background:", ";"], light.hover_bg_color);
}, function (_ref12) {
  var delay = _ref12.delay;
  return delay;
}, light.level3_base_color, function (_ref13) {
  var delay = _ref13.delay;
  return delay;
}, function (_ref14) {
  var type = _ref14.type;
  return type === 'normal' ? css(["font-size:16px;"]) : type === 'failure' ? css(["font-weight:500;font-size:12px;color:#111f2c;"]) : css(["font-size:14px;line-height:38px;color:", ";"], light.level1_base_color);
}, function (_ref15) {
  var type = _ref15.type;
  return type === 'failure' ? css(["color:#111f2c;"]) : css(["color:rgba(17,31,44,0.4);"]);
}, light.blue1_color);
