// @ali/we-design-next/dist-node/display/menu/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SuffixPlaceholder = exports.MenuWrap = exports.MenuItemWrap = exports.MenuItemTitleDescription = exports.MenuItemSuffix = exports.MenuItemPrefix = exports.MenuItemGroupTitle = exports.MenuItemGroup = exports.MenuItemDivider = exports.MenuItemDescription = exports.MenuItemContent = void 0;
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _text = require("../../basic/text");
var _common = require("../../common");
const MenuWrap = /*#__PURE__*/_styledComponents.default.div(["", " outline:none;min-width:150px;overflow-x:hidden;overflow-y:overlay;"], _common.CSS_SCROLLBAR);
exports.MenuWrap = MenuWrap;
const MenuItemWrap = /*#__PURE__*/_styledComponents.default.div(["", " user-select:none;height:", ";line-height:20px;padding:0 8px;margin:4px;display:flex;align-items:center;border-radius:4px;", " cursor:", ";color:", ";"], _text.TEXT_CSS.TEXT, ({
  hasDescription
}) => hasDescription ? '53px' : '34px', ({
  isActive
}) => isActive ? `background: ${_common.COLORS.OTHERS.DAO_HOVER_BG_COLOR};` : null, ({
  disabled
}) => disabled ? 'not-allowed' : 'default', ({
  disabled
}) => disabled ? _common.COLORS.LIGHT.LEVEL4_BASE_COLOR : 'unset');
exports.MenuItemWrap = MenuItemWrap;
const MenuItemContent = /*#__PURE__*/_styledComponents.default.div(["flex-grow:1;word-break:keep-all;"]);
exports.MenuItemContent = MenuItemContent;
const MenuItemTitleDescription = /*#__PURE__*/_styledComponents.default.div(["margin-top:8px;"]);
exports.MenuItemTitleDescription = MenuItemTitleDescription;
const MenuItemDescription = /*#__PURE__*/_styledComponents.default.div(["", " margin-bottom:8px;color:", ";"], _text.TEXT_CSS.TEXT_SUB, ({
  disabled
}) => disabled ? _common.COLORS.LIGHT.LEVEL4_BASE_COLOR : _common.COLORS.LIGHT.LEVEL2_BASE_COLOR);
exports.MenuItemDescription = MenuItemDescription;
const MenuItemPrefix = /*#__PURE__*/_styledComponents.default.div(["flex-grow:0;margin-right:4px;position:relative;top:0.2em;"]);
exports.MenuItemPrefix = MenuItemPrefix;
const MenuItemSuffix = /*#__PURE__*/_styledComponents.default.div(["flex-grow:0;margin-left:16px;position:relative;top:0.2em;"]);
exports.MenuItemSuffix = MenuItemSuffix;
const MenuItemDivider = /*#__PURE__*/_styledComponents.default.div(["height:1px;margin:4px;background-color:", ";"], _common.COLORS.OTHERS.DIVIDER);
exports.MenuItemDivider = MenuItemDivider;
const MenuItemGroup = /*#__PURE__*/_styledComponents.default.div(["&::after{display:block;margin:0 4px;height:1px;background-color:", ";content:'';}&:last-child{&::after{display:none;}}"], _common.COLORS.OTHERS.DIVIDER);
exports.MenuItemGroup = MenuItemGroup;
const MenuItemGroupTitle = /*#__PURE__*/_styledComponents.default.div(["", " color:", ";line-height:36px;height:36px;padding-left:12px;user-select:none;"], _text.TEXT_CSS.TEXT_SUB, _common.COLORS.LIGHT.LEVEL2_BASE_COLOR);
exports.MenuItemGroupTitle = MenuItemGroupTitle;
const SuffixPlaceholder = /*#__PURE__*/_styledComponents.default.div(["margin-left:16px;width:16px;height:16px;"]);
exports.SuffixPlaceholder = SuffixPlaceholder;
