// @ali/4ever-plugin-list/dist-node/bi/commands/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  list2Paragraph: true,
  toggleList: true,
  toggleWrapInList: true,
  splitListItem: true,
  increaseItemDepth: true,
  increaseItemByNode: true,
  removeListNode: true,
  toggleCheckStatus: true,
  removeListSymbol: true,
  transform: true
};
Object.defineProperty(exports, "list2Paragraph", {
  enumerable: true,
  get: function () {
    return _list2Paragraph.default;
  }
});
Object.defineProperty(exports, "toggleList", {
  enumerable: true,
  get: function () {
    return _toggleList.default;
  }
});
Object.defineProperty(exports, "toggleWrapInList", {
  enumerable: true,
  get: function () {
    return _toggleWrapInList.default;
  }
});
Object.defineProperty(exports, "splitListItem", {
  enumerable: true,
  get: function () {
    return _splitListItem.default;
  }
});
Object.defineProperty(exports, "increaseItemDepth", {
  enumerable: true,
  get: function () {
    return _increaseItemDepth.default;
  }
});
Object.defineProperty(exports, "increaseItemByNode", {
  enumerable: true,
  get: function () {
    return _increaseItemDepth.increaseItemByNode;
  }
});
Object.defineProperty(exports, "removeListNode", {
  enumerable: true,
  get: function () {
    return _removeListNode.default;
  }
});
Object.defineProperty(exports, "toggleCheckStatus", {
  enumerable: true,
  get: function () {
    return _toggleCheckStatus.default;
  }
});
Object.defineProperty(exports, "removeListSymbol", {
  enumerable: true,
  get: function () {
    return _removeListSymbol.default;
  }
});
Object.defineProperty(exports, "transform", {
  enumerable: true,
  get: function () {
    return _transform.default;
  }
});

var _list2Paragraph = _interopRequireDefault(require("./list2Paragraph"));

var _toggleList = _interopRequireDefault(require("./toggleList"));

var _toggleWrapInList = _interopRequireDefault(require("./toggleWrapInList"));

var _splitListItem = _interopRequireDefault(require("./splitListItem"));

var _increaseItemDepth = _interopRequireWildcard(require("./increaseItemDepth"));

var _removeListNode = _interopRequireDefault(require("./removeListNode"));

var _toggleCheckStatus = _interopRequireDefault(require("./toggleCheckStatus"));

var _removeListSymbol = _interopRequireDefault(require("./removeListSymbol"));

var _transform = _interopRequireDefault(require("./transform"));

var _numbering = require("./numbering");

Object.keys(_numbering).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _numbering[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _numbering[key];
    }
  });
});
