// @ali/4ever-plugin-comment/dist-node/utils/bamboo.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  hooks: true,
  types: true,
  actions: true,
  commands: true,
  queries: true,
  models: true
};
exports.models = exports.queries = exports.commands = exports.actions = exports.types = exports.hooks = void 0;

var _hooks = _interopRequireWildcard(require("./hooks"));

exports.hooks = _hooks;

var _types = _interopRequireWildcard(require("./types"));

exports.types = _types;

var _actions = _interopRequireWildcard(require("./actions"));

exports.actions = _actions;

var _commands = _interopRequireWildcard(require("./commands"));

exports.commands = _commands;

var _queries = _interopRequireWildcard(require("./queries"));

exports.queries = _queries;

var _models = _interopRequireWildcard(require("./models"));

exports.models = _models;

var _role = require("./role");

Object.keys(_role).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _role[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _role[key];
    }
  });
});
