// @ali/4ever-plugin-paragraph/dist-web/components/sidebar/common.js
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Select } from '@ali/we-design';
import { isNaN } from '@ali/4ever-utils';

var createNumberFormater = function createNumberFormater(fix) {
  var f = fix || 100;
  return function (number) {
    if (number === '') return '';
    if (!number || typeof number === 'number') return "" + (number || 0);
    if (/\.$/.test(number)) return number;
    var res = Math.round(parseFloat(number) * f) / f;
    if (isNaN(res)) return '';
    return "" + res;
  };
};

var unitSelectStyle = {
  width: 50,
  height: 28,
  fontSize: 13
};
export var formatNumber10 = createNumberFormater(10);
export var formatNumber100 = createNumberFormater(100);
export function parseInputNumberPara(arg) {
  // 点击 上下 button 为 number
  if (typeof arg === 'number') return arg; // 按键盘上下键为 string

  if (typeof arg === 'string') return parseFloat(arg); // 按 enter 为 event

  if (arg != null && arg.target) return parseFloat(arg.target.value);
  return NaN;
}
export var commonInputNumberProps = {
  min: 0,
  placeholder: '',
  step: 1
};
export function checkInputValue(value, min) {
  if (isNaN(value)) return false;
  if (value < (min || 0)) return false;
  return true;
}
export var inputNumberStyle = _extends({}, unitSelectStyle, {
  width: 60,
  margin: '0 5px'
});
export var commonSelectStyle = {
  height: 28,
  width: 90,
  fontSize: 13,
  padding: '0 4px 0 0'
};
export function UnitSelect(props) {
  var value = props.value,
      units = props.units,
      onChange = props.onChange;
  var options = units.map(function (unit) {
    return /*#__PURE__*/_createElement(Select.Option, {
      value: unit.key
    }, unit.value);
  });
  return /*#__PURE__*/_createElement(Select, {
    value: value,
    style: unitSelectStyle,
    onChange: onChange
  }, options);
}
