// @ali/4ever-plugin-heading/dist-node/mo/models/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.headingTypes = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _everUtils = require("@ali/4ever-utils");

var _heading = _interopRequireDefault(require("./heading1"));

var _heading2 = _interopRequireDefault(require("./heading2"));

var _heading3 = _interopRequireDefault(require("./heading3"));

var _heading4 = _interopRequireDefault(require("./heading4"));

var _heading5 = _interopRequireDefault(require("./heading5"));

var _heading6 = _interopRequireDefault(require("./heading6"));

const headingTypes = (0, _everUtils.createDict)({
  [_heading.default.TYPE]: true,
  [_heading2.default.TYPE]: true,
  [_heading3.default.TYPE]: true,
  [_heading4.default.TYPE]: true,
  [_heading5.default.TYPE]: true,
  [_heading6.default.TYPE]: true
});
exports.headingTypes = headingTypes;
const Heading = {
  TYPES: Object.keys(headingTypes),
  Heading1: _heading.default,
  Heading2: _heading2.default,
  Heading3: _heading3.default,
  Heading4: _heading4.default,
  Heading5: _heading5.default,
  Heading6: _heading6.default,

  isHeading(maybeHeading) {
    return _everCangjie.Block.isBlock(maybeHeading) && headingTypes[maybeHeading.type];
  },

  isEmpty(block) {
    return headingTypes[block.type] && (block.data.jc || 'left') === 'left' && !block.data.ind?.left && block.text === '';
  },

  isCallout(maybeCallout) {
    return (_everPluginParagraph.Paragraph.isParagraph(maybeCallout) || Heading.isHeading(maybeCallout)) && !!maybeCallout.data.calloutPr;
  }

};
var _default = Heading;
exports.default = _default;
