// @ali/we-design-next/dist-node/display/drawer/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TitleBox = exports.Sticker = exports.HeaderBox = exports.FooterBox = exports.DrawerBox = exports.Background = void 0;
var _styledComponents = _interopRequireDefault(require("styled-components"));
const Sticker = /*#__PURE__*/_styledComponents.default.div(["position:fixed;left:0;bottom:0;height:0;width:100%;z-index:999;"]);
exports.Sticker = Sticker;
const DrawerBox = /*#__PURE__*/_styledComponents.default.div(["position:fixed;bottom:0;width:100%;display:flex;flex-direction:column;background-color:#fff;padding-bottom:env(safe-area-inset-bottom);border-top-left-radius:16px;border-top-right-radius:16px;box-shadow:0px 0px 1px 0px rgba(0,0,0,0.12),0px 8px 24px 0px rgba(53,64,90,0.12);transform:translateY(100%);"]);
exports.DrawerBox = DrawerBox;
const Background = /*#__PURE__*/_styledComponents.default.div(["position:fixed;left:0;bottom:0;background-color:", ";width:100%;height:100%;display:", ";"], p => p.mask ? 'rgba(0, 0, 0, 0.4)' : 'transparent', p => p.isVisible ? 'block' : 'none');
exports.Background = Background;
const HeaderBox = /*#__PURE__*/_styledComponents.default.div(["position:relative;width:100%;padding-top:16px;z-index:1;&.wdn-show-touch-bar::after{content:' ';position:absolute;top:6px;left:50%;transform:translateX(-50%);height:4px;width:48px;border-radius:2px;overflow:hidden;background:#D8D8D8;}"]);
exports.HeaderBox = HeaderBox;
const TitleBox = /*#__PURE__*/_styledComponents.default.div(["flex-shrink:0;position:relative;"]);
exports.TitleBox = TitleBox;
const FooterBox = /*#__PURE__*/_styledComponents.default.div(["flex-shrink:0;"]);
exports.FooterBox = FooterBox;
