// @ali/4ever-plugin-table/dist-web/utils/utils/horizontalScroll.js
import _createClass from "@babel/runtime/helpers/createClass";

/**
 * 创建一个scroller的dom
 */
import { throttle } from 'lodash-es';
import fastdom from 'fastdom';
import { TOOLBAR_ITEM_SIZE } from "../constants"; // 滚动条高度

var THUMB_HEIGH = 6;
var TEST_ID = 'table-horizontal-scrollbar';
/**
 * 滚动条显示模式：
 *  - always：强制常驻
 *  - default：默认只在需要的时候展示
 */

var ScrollMode;

(function (ScrollMode) {
  ScrollMode["always"] = "always";
  ScrollMode["default"] = "default";
})(ScrollMode || (ScrollMode = {}));

var Scroller = /*#__PURE__*/function () {
  /**
   * 表格滚动容器
   */

  /**
   * 滚动条宽度是否占满，占满则不显示滚动条
   */

  /**
   * 滚动条显示模式：默认只在需要的时候展示
   */

  /**
   * 需要添加的滚动条dom
   */

  /**
   * 需要添加的滚动条杆dom
   */

  /**
   * 需要添加的滚动条滑块dom
   */

  /**
   * 滚动时检查是否需要隐藏或者展示滚动条
   */

  /**
   * 销毁用于同步点击滚动条移动表格的事件监听器
   */

  /**
   * 监听目标表格dom变化的observer
   */

  /**
   * 滚动条挂载容器
   */

  /**
   * @param targetTableWrapperEl 表格滚动容器
   * @param getBottom 获取滚动条到页面底部的距离
   * @param mountDom 滚动条挂载容器
   * @param mode 滚动条显示模式：默认只在需要的时候展示
   */

  /** 是否需要重新设置bottom */

  /** 当前bottom */
  function Scroller(targetTableWrapperEl, getBottom, mountDom, mode) {
    var _this = this;

    if (mode === void 0) {
      mode = ScrollMode["default"];
    }

    this.targetTableWrapperEl = void 0;
    this.fullwidth = void 0;
    this.mode = void 0;
    this.scrollBar = void 0;
    this.bar = void 0;
    this.thumb = void 0;
    this.checkIfScrollVisible = void 0;
    this.handleTableWrapperScroll = void 0;
    this.syncDestoryHandler = void 0;
    this.tableElObserver = void 0;
    this.mountDom = void 0;
    this.getBottom = void 0;
    this.needResetBottom = void 0;
    this.currentBottom = void 0;

    if (!targetTableWrapperEl) {
      throw new Error('need have table element');
    }

    this.targetTableWrapperEl = targetTableWrapperEl;
    this.fullwidth = false;
    this.mode = mode;
    this.mountDom = mountDom || document.body;
    this.getBottom = getBottom;
    this.needResetBottom = !!getBottom; // 如果传了getBottom则说明需要对bottom进行重置操作

    /**
     * 添加滚动条dom节点
     */

    var footerHeight = getBottom ? getBottom() : 0;
    var scroller = document.createElement('div');
    scroller.setAttribute('data-testid', TEST_ID);
    scroller.style.height = THUMB_HEIGH + "px";
    scroller.style.position = 'fixed';
    scroller.style.bottom = footerHeight + "px";
    this.currentBottom = footerHeight; // z-index 要高于内容区域 以及 右侧分屏面板的 z-index

    scroller.style.zIndex = '200';
    scroller.style.opacity = '0';
    this.scrollBar = scroller;
    this.resetScroller();
    var bar = document.createElement('div');
    bar.style.height = '100%';
    this.bar = bar;
    scroller.appendChild(bar);
    var thumb = document.createElement('div');
    thumb.style.height = '100%';
    thumb.style.backgroundColor = 'rgb(205, 205, 210)';
    thumb.style.borderRadius = '5px';
    bar.appendChild(thumb);
    this.thumb = thumb;

    try {
      this.mountDom.appendChild(this.scrollBar);
    } catch (e) {
      document.body.appendChild(this.scrollBar);
    }
    /**
     * 初始化配置
     */


    var instance = this;
    this.checkIfScrollVisible = throttle(function () {
      fastdom.measure(function () {
        var viewHeight = window.innerHeight || document.documentElement.clientHeight; // 当前可视区域超出表格最下方或者最上方时，隐藏

        var _this$targetTableWrap = _this.targetTableWrapperEl.getBoundingClientRect(),
            top = _this$targetTableWrap.top,
            bottom = _this$targetTableWrap.bottom; // 当表格上方已经处于视口边缘一定距离阈值内时，不必再显示滚动条


        if (bottom <= viewHeight || top >= viewHeight - THUMB_HEIGH * 2 - TOOLBAR_ITEM_SIZE) {
          instance.hideScroller();
        } else {
          // 需要重新设置一次当前宽度
          instance.resetBar(false);
          instance.resetScroller(); // 显示当前的bar

          instance.showScroller();
        }
      });
    }, 1000 / 60);
    document.addEventListener('scroll', this.checkIfScrollVisible, true); // 全局判断是否需要显示scroller

    this.handleTableWrapperScroll = throttle(function () {
      fastdom.measure(function () {
        instance.resetThumbPosition();
      });
    }, 1000 / 60); // 自动同步,table => scroller

    targetTableWrapperEl.addEventListener('scroll', this.handleTableWrapperScroll, true); // 自动同步 scroller => table

    this.syncDestoryHandler = this.initScrollSyncHandler(); // 监听table的dom变化，自动重新设置

    this.tableElObserver = new MutationObserver(function () {
      setTimeout(function () {
        fastdom.measure(function () {
          instance.resetBar();
          instance.resetScroller();
          instance.resetThumbPosition();
          instance.checkIfScrollVisible();
        });
      });
    });
    this.tableElObserver.observe(targetTableWrapperEl, {
      childList: true,
      subtree: true,
      // Todo: @muyu 这里不确定只观察父节点变更是否会有bug，因此子节点变化也一并观察，可能会对性能有影响。
      attributes: true,
      attributeFilter: ['style'] // 只关注目标dom节点的style变化

    }); // bar宽度自动重制

    setTimeout(function () {
      fastdom.measure(function () {
        instance.resetBar();
        instance.resetScroller();
        instance.resetThumbPosition();
        instance.checkIfScrollVisible();
        scroller.style.opacity = '1';
      });
    }, 500);
  }
  /**
   * 自动设置Bar
   * @param {boolean} changeScrollerVisible 是否开启自动设置滚动条显示与否
   */


  var _proto = Scroller.prototype;

  _proto.resetBar = function resetBar(changeScrollerVisible) {
    if (changeScrollerVisible === void 0) {
      changeScrollerVisible = true;
    }

    var targetTableWrapperEl = this.targetTableWrapperEl;
    var widthPercentage = targetTableWrapperEl.clientWidth * 100 / targetTableWrapperEl.scrollWidth;
    var thumbWidth = Math.min(widthPercentage, 100);
    this.thumb.style.width = thumbWidth + "%";
    this.fullwidth = thumbWidth >= 100;

    if (changeScrollerVisible) {
      if (this.fullwidth) {
        this.hideScroller();
      } else {
        this.checkIfScrollVisible();
      }
    }
  };

  _proto.resetThumbPosition = function resetThumbPosition() {
    // 通过transform来模拟滑块滚动
    this.thumb.style.transform = "translateX(" + this.moveX + "%)";
  } // 调整滚动条位置以及宽度
  ;

  _proto.resetScroller = function resetScroller() {
    var targetTableWrapperEl = this.targetTableWrapperEl,
        scrollBar = this.scrollBar;
    var boundingClientRect = targetTableWrapperEl.getBoundingClientRect(); // 文档Footer的高度是固定的，缩放只针对Content区域，因此只需要获取其高度并设置一次bottom即可

    if (this.needResetBottom) {
      // 但是要考虑window.innerHeight 变化的场景
      var footerHeight = this.getBottom();
      var isBottomChange = footerHeight !== this.currentBottom;

      if (isBottomChange) {
        var bottom = footerHeight || 0; // 文档由于底部Footer也是fixed，因此bottom不能设置为0，只能动态获取Footer的高度来设置bottom

        scrollBar.style.bottom = bottom + "px";
        this.currentBottom = bottom;
      }
    }

    scrollBar.style.left = boundingClientRect.left + "px";
    scrollBar.style.width = boundingClientRect.width + "px";
  } // 滑块位移计算
  ;

  /**
   * 让scroller的拖动行为和table的同步
   */
  _proto.initScrollSyncHandler = function initScrollSyncHandler() {
    var cursorDown = false;
    var tempClientX = 0;
    var rate = 1;
    var thumb = this.thumb,
        targetTableWrapperEl = this.targetTableWrapperEl,
        bar = this.bar;

    function getRate() {
      // 计算一下变换比例，拖拽走的是具体数字，但是这个实际上应该是按照比例变的
      return bar.offsetWidth / thumb.offsetWidth;
    }

    var mouseMoveDocumentHandler = throttle(function (e) {
      // 拖动的时候如果已经松开鼠标则不处理
      // cursorDown 是拖拽开启和结束的标识
      if (cursorDown === false) {
        return;
      }

      var clientX = e.clientX;
      var offset = clientX - tempClientX;
      var originTempClientX = tempClientX;
      tempClientX = clientX;
      var tempScrollleft = targetTableWrapperEl.scrollLeft;
      targetTableWrapperEl.scrollLeft += offset * rate;

      if (tempScrollleft === targetTableWrapperEl.scrollLeft) {
        tempClientX = originTempClientX;
      }
    }, 1000 / 60);

    function mouseUpDocumentHandler() {
      cursorDown = false;
      document.removeEventListener('mousemove', mouseMoveDocumentHandler);
      document.removeEventListener('mouseup', mouseUpDocumentHandler);
      document.onselectstart = null;
    }
    /**
     * 拖拽处理
     */


    function startDrag(e) {
      e.stopImmediatePropagation();
      cursorDown = true;
      document.addEventListener('mousemove', mouseMoveDocumentHandler);
      document.addEventListener('mouseup', mouseUpDocumentHandler);

      document.onselectstart = function () {
        return false;
      };
    }

    function handleThumbMouseDown(e) {
      // 右键点击以及ctrl+click事件不处理
      if (e.ctrlKey || e.button === 2) {
        return;
      }

      var clientX = e.clientX;
      tempClientX = clientX;
      rate = getRate();
      startDrag(e);
    }

    thumb.addEventListener('mousedown', handleThumbMouseDown);
    /**
     * 点击槽快速移动
     */

    function handleBarClick(e) {
      var target = e.target;

      if (target !== bar) {
        return;
      }

      rate = getRate();
      var clientX = e.clientX;
      var offset = 0;
      var thumbPosition = thumb.getBoundingClientRect();

      if (thumbPosition.left >= clientX) {
        offset = clientX - thumbPosition.left;
      } else {
        offset = clientX - thumbPosition.left - thumbPosition.width;
      }

      var targetScrollLeft = targetTableWrapperEl.scrollLeft + offset * rate;
      targetTableWrapperEl.scrollTo({
        left: targetScrollLeft,
        behavior: 'smooth'
      });
    }

    bar.addEventListener('click', handleBarClick);
    return function () {
      thumb.removeEventListener('mousedown', handleThumbMouseDown);
      bar.removeEventListener('click', handleBarClick);
      document.removeEventListener('mouseup', mouseUpDocumentHandler);
      document.removeEventListener('mousemove', mouseMoveDocumentHandler);
    };
  }
  /**
   * 显示整体
   */
  ;

  _proto.showScroller = function showScroller() {
    if (!this.fullwidth) {
      this.scrollBar.style.display = 'initial';
    }
  }
  /**
   * 隐藏整体
   */
  ;

  _proto.hideScroller = function hideScroller() {
    if (this.mode === ScrollMode.always) {
      this.scrollBar.style.display = 'initial';
    } else {
      this.scrollBar.style.display = 'none';
    }
  }
  /**
  * 添加事件监听
  */
  ;

  _proto.addEventListener = function addEventListener() {
    var _this$mountDom;

    (_this$mountDom = this.mountDom) == null ? void 0 : _this$mountDom.addEventListener('scroll', this.checkIfScrollVisible);
    this.targetTableWrapperEl.addEventListener('scroll', this.handleTableWrapperScroll);
  }
  /**
   * 移除事件监听
   */
  ;

  _proto.removeEventListener = function removeEventListener() {
    var _this$mountDom2;

    (_this$mountDom2 = this.mountDom) == null ? void 0 : _this$mountDom2.removeEventListener('scroll', this.checkIfScrollVisible);
    this.targetTableWrapperEl.removeEventListener('scroll', this.handleTableWrapperScroll);
  }
  /**
   * 销毁实例：
   * - 取消observer监听
   * - 监听器注销
   * - 滚动条dom移除
   */
  ;

  _proto.destroy = function destroy() {
    var _this$thumb$parentNod, _this$bar$parentNode, _this$scrollBar$paren;

    this.tableElObserver.disconnect();
    this.removeEventListener();
    this.syncDestoryHandler();
    (_this$thumb$parentNod = this.thumb.parentNode) == null ? void 0 : _this$thumb$parentNod.removeChild(this.thumb);
    (_this$bar$parentNode = this.bar.parentNode) == null ? void 0 : _this$bar$parentNode.removeChild(this.bar);
    (_this$scrollBar$paren = this.scrollBar.parentNode) == null ? void 0 : _this$scrollBar$paren.removeChild(this.scrollBar);
  };

  _createClass(Scroller, [{
    key: "moveX",
    get: function get() {
      var targetTableWrapperEl = this.targetTableWrapperEl;
      return targetTableWrapperEl.scrollLeft * 100 / targetTableWrapperEl.clientWidth;
    }
  }]);

  return Scroller;
}();

export { Scroller as default };
