// @ali/we-ui/dist-web/components/script-select/ScriptSelect.js
/* eslint-disable max-lines-per-function */
import React, { useCallback, useState } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import useControlledValue from "../../hooks/useControlledValue";
import { useTranslate } from "../../locale/locales";
import { SelectNormal, XDownNormal, XUpNormal } from "../icons";
import { SplitButton } from "../button";
import { Tooltip } from "../tooltip";
import { OverlayWrap, OverlayItem } from "./styled";
import { getTitleContent, getTitlePlacement } from "../select/SelectOption";

var _ref = /*#__PURE__*/_createElement(SelectNormal, {
  className: "check-icon"
});

var _ref2 = /*#__PURE__*/_createElement(XUpNormal, null);

var _ref3 = /*#__PURE__*/_createElement(XDownNormal, null);

var ScriptSelect = function ScriptSelect(props) {
  var locale = props.locale,
      style = props.style,
      onClick = props.onClick,
      className = props.className,
      title = props.title,
      titles = props.titles,
      disabled = props.disabled,
      size = props.size;

  var _useControlledValue = useControlledValue(props.value, props.defaultValue),
      value = _useControlledValue[0],
      setValue = _useControlledValue[1];

  var _useState = useState(false),
      active = _useState[0],
      setActive = _useState[1];

  var t = useTranslate(locale);
  /**
   * 点击上下标操作
   */

  var handleClickLeft = useCallback(function (event) {
    if (onClick) {
      onClick(value, event);
    }
  }, [value, onClick]);
  var isSuper = value === 'super';
  var superText = t('we_ui_script_superscript'); // 上标文案

  var subText = t('we_ui_script_subscript'); // 下标文案

  var handleChooseScript = useCallback(function (script) {
    if (onClick) {
      onClick(script);
    }

    setValue(script);
  }, [onClick, setValue]);
  var handleChooseSuper = useCallback(function () {
    if (value !== 'super') {
      handleChooseScript('super');
      setActive(false);
    }
  }, [value, handleChooseScript]);
  var handleChooseSub = useCallback(function () {
    if (value !== 'sub') {
      handleChooseScript('sub');
      setActive(false);
    }
  }, [value, handleChooseScript]);
  var chooses = [{
    title: superText,
    value: 'super',
    Icon: XUpNormal,
    onClick: handleChooseSuper
  }, {
    title: subText,
    value: 'sub',
    Icon: XDownNormal,
    onClick: handleChooseSub
  }];
  /**
   * 下拉选项
   */

  var DropContent = /*#__PURE__*/_createElement(OverlayWrap, null, chooses.map(function (item) {
    return /*#__PURE__*/_createElement(Tooltip, {
      title: getTitleContent(titles[item.value]),
      placement: getTitlePlacement(titles[item.value]),
      key: item.value
    }, /*#__PURE__*/_createElement(OverlayItem, {
      className: value === item.value ? 'selected' : '',
      onClick: item.onClick,
      "data-testid": "option-" + item.value
    }, /*#__PURE__*/_createElement(item.Icon, {
      className: "script-icon"
    }), /*#__PURE__*/_createElement("span", null, item.title), value === item.value && _ref));
  }));

  return /*#__PURE__*/_createElement(SplitButton, {
    overlay: DropContent,
    onClickLeft: handleClickLeft,
    onActiveChange: setActive,
    style: style,
    size: size,
    active: active,
    className: className,
    disabled: disabled,
    title: title || (isSuper ? superText : subText)
  }, isSuper ? _ref2 : _ref3);
};

ScriptSelect.defaultProps = {
  defaultValue: 'super',
  titles: {}
};
export default ScriptSelect;
