// @ali/4ever-plugin-find-and-replace/dist-node/components/createButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _actions = require("../common/actions");

var _everCangjie = require("@ali/4ever-cangjie");

const _createElement = /*#__PURE__*/React.createElement;
const MOD = _everCangjie.environment.IS_MAC ? '⌘' : 'Ctrl';
const SHORTCUT = `${MOD}+F`;

function createButton(ctlManager) {
  return function FindAndReplaceButton(props) {
    const {
      controller,
      locale,
      ...rest
    } = props;
    const {
      data = {}
    } = controller.value;
    const isVisible = data.findAndReplaceVisible || false;
    const isDisabled = controller.query('isDisableFindAndReplace');
    const handleClick = React.useCallback(() => {
      const {
        mainController
      } = ctlManager;

      if (!mainController) {
        return;
      }

      mainController.run('onAction', (0, _actions.toggleFindAndReplace)());
      let fragment = '';
      ctlManager.run(ctl => {
        const text = ctl.value.fragment?.text;

        if (text) {
          fragment = text;
        }
      });

      if (fragment) {
        mainController.run('onAction', (0, _actions.updateFindFragment)(fragment));
      }
    }, []);
    return /*#__PURE__*/_createElement(_weToolbar.SearchButton, (0, _extends2.default)({}, rest, {
      testid: "bi-toolbar-search",
      active: isVisible,
      onClick: handleClick,
      shortcutText: SHORTCUT,
      disabled: isDisabled,
      role: 'findandreplace'
    }));
  };
}
