// @ali/4ever-bi/dist-node/plugins/image/handlers/createContextMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everMo = require("@ali/4ever-mo");

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../actions");

function createContextMenu(config, sideBarTrigger) {
  const {
    locale = {}
  } = config;
  return function contextMenu(controller, next) {
    const {
      document,
      selection
    } = controller.value;
    const {
      isCollapsed
    } = selection;

    if (!isCollapsed && sideBarTrigger) {
      const start = selection.getStart(document);
      const end = selection.getEnd(document);
      const node1 = document.getNextText(start.key);
      const node2 = document.getPreviousText(end.key);

      const closeContextMenu = onActionController => {
        onActionController.run('onAction', _everBamboo.PcContextMenu.closeContextMenu());
      };

      if (node1 && node2 && node1 === node2) {
        const node = document.getClosestVoid(node1.key, controller);

        if (_everMo.Image.isImage(node)) {
          return [...next(), {
            name: locale?.options || 'Image options',
            key: 'image-options',
            action: (0, _actions.toggleImageSidebar)({
              node,
              toggle: true,
              callback: closeContextMenu
            }),
            options: {
              group: 2
            }
          }];
        }
      }
    }

    return next();
  };
}

var _default = createContextMenu;
exports.default = _default;
