// @ali/4ever-bi/dist-web/plugins/blockquote/index.js
import _extends from "@babel/runtime/helpers/extends";
import toolbar from "./toolbar";
import menubar from "./menubar";
import renderNode from "./renderNode";
import toggleQuote from "./commands/toggleQuote";
import quickInsert from "./quickInsertConfig";
import * as actions from "./actions";
import onAction from "./handlers/onAction";
import onKeyDown from "./handlers/onKeyDown";
import onCangjieInput from "./handlers/onCangjieInput";
import { queries } from "./queries";
import { moFactory } from '@ali/4ever-plugin-blockquote';
var commands = {
  toggleQuote: toggleQuote
};

function createBlockquotePlugin(config) {
  var blockquoteConfig = config.blockquote || {};
  return _extends({}, moFactory({}), {
    toolbar: toolbar(blockquoteConfig),
    menubar: menubar(),
    renderNode: renderNode,
    commands: commands,
    quickInsert: quickInsert,
    onAction: onAction,
    onKeyDown: onKeyDown,
    queries: queries,
    onCangjieInput: onCangjieInput,
    models: {
      data: {
        shd: [undefined, Object]
      }
    }
  });
}

createBlockquotePlugin.actions = actions;
createBlockquotePlugin.commands = commands;
export default createBlockquotePlugin;
