// @ali/4ever-plugin-color/dist-node/bi/components/colorAndHighlightButton/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ClearBox = exports.ClearItem = exports.TextColorItem = exports.ColorBoxItem = exports.ColorBox = exports.PanelSection = exports.SectionTitle = exports.ColorAndHighlightPanelWrapper = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constants = require("../../constants");

const ColorAndHighlightPanelWrapper = /*#__PURE__*/_styledComponents.default.div(["width:183px;"]);

exports.ColorAndHighlightPanelWrapper = ColorAndHighlightPanelWrapper;

const SectionTitle = /*#__PURE__*/_styledComponents.default.div(["font-size:12px;line-height:17px;padding-bottom:3px;color:rgba(23,26,29,0.6);"]);

exports.SectionTitle = SectionTitle;

const PanelSection = /*#__PURE__*/_styledComponents.default.div(["padding:10px 8px 6px 8px;border-bottom:1px solid rgba(17,31,44,0.08);"]);

exports.PanelSection = PanelSection;

const ColorBox = /*#__PURE__*/_styledComponents.default.div(["display:flex;flex-direction:row;flex-wrap:wrap;justify-content:space-between;"]);

exports.ColorBox = ColorBox;

const ColorBoxItem = /*#__PURE__*/_styledComponents.default.div(["position:relative;width:24px;height:24px;margin:2px auto;display:flex;justify-content:center;align-items:center;border-radius:5px;cursor:pointer;", " ", ";&:hover{border:1px solid ", ";}"], p => p.bgColor && `background-color: ${p.bgColor};`, p => p.isSelected && 'border: 1px solid #0089ff;', p => p.isSelected ? '#0089FF' : 'rgba(126, 134, 142, 0.12)');

exports.ColorBoxItem = ColorBoxItem;

const TextColorItem = /*#__PURE__*/_styledComponents.default.div(["width:24px;height:24px;box-sizing:border-box;border-radius:4px;color:", ";display:flex;align-items:center;justify-content:center;", ";"], p => p.color || _constants.DEFAULT_COLOR, p => !p.isSelected && 'border: 1px solid rgba(126, 134, 142, 0.16);');

exports.TextColorItem = TextColorItem;

const ClearItem = /*#__PURE__*/_styledComponents.default.div(["position:relative;width:22px;height:22px;border-radius:4px;border:1px solid rgba(126,134,142,0.16);&::after{content:'';position:absolute;width:23px;height:2px;left:-1px;top:9.4px;background:#e41100;transform:rotate(45deg);}"]);

exports.ClearItem = ClearItem;

const ClearBox = /*#__PURE__*/_styledComponents.default.div(["height:28px;line-height:28px;margin:2px;color:#0089ff;font-size:13px;text-align:center;cursor:pointer;border-radius:4px;&:hover{background:rgba(126,134,142,0.12);}"]);

exports.ClearBox = ClearBox;
