// @ali/4ever-plugin-paragraph/dist-node/zhi/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.factory = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _index = _interopRequireDefault(require("../mo/index"));

var _ReadParagraph = _interopRequireDefault(require("../components/ReadParagraph"));

var _styled = require("../components/styled");

var _queries = require("./queries");

const _createElement = /*#__PURE__*/React.createElement;

const factory = () => {
  return { ...(0, _index.default)({}),
    queries: _queries.queries,
    renderNode: {
      paragraph: props => {
        const {
          controller,
          attributes,
          children,
          node
        } = props;
        const content = controller.query('renderParagraph', props);

        if (content) {
          return content;
        }

        return /*#__PURE__*/_createElement(_ReadParagraph.default, (0, _extends2.default)({}, attributes, {
          node: node,
          Component: _styled.Component
        }), children());
      }
    }
  };
};

exports.factory = factory;
var _default = factory;
exports.default = _default;
