// @ali/4ever-plugin-paragraph/dist-node/bi/createRenderNode.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderNode;

var React = _interopRequireWildcard(require("react"));

var _EditParagraph = _interopRequireDefault(require("../components/EditParagraph"));

const _createElement = /*#__PURE__*/React.createElement;

function createRenderNode() {
  return {
    paragraph: props => {
      const {
        children,
        controller,
        node,
        attributes,
        visible
      } = props;
      /**
       * list 与 paragraph 的 type 一致，render 通过 query 解耦。
       *
       * 注：通过中间件方式解耦更合理，但不如 k/v 方式快，需要评估对性能的影响
       */

      const content = props.controller.query('renderParagraph', props);

      if (content) {
        return content;
      }

      return /*#__PURE__*/_createElement(_EditParagraph.default, {
        controller: controller,
        node: node,
        attributes: attributes,
        visible: visible
      }, children());
    }
  };
}
