// @ali/4ever-plugin-paragraph/dist-node/bi/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _interopRequireWildcard2 = _interopRequireDefault(require("@babel/runtime/helpers/interopRequireWildcard"));

var _everUtils = require("@ali/4ever-utils");

var _index = _interopRequireDefault(require("../mo/index"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _createRenderNode = _interopRequireDefault(require("./createRenderNode"));

var _queries = require("./queries");

var _actions = require("./actions");

const factory = configs => {
  const {
    paragraph = {}
  } = configs;
  const {
    locale
  } = paragraph;
  const plugin = { ...(0, _index.default)({}),
    onAction: _onAction.default,
    actions: _actions.actions,
    isActions: _actions.isActions,
    queries: _queries.queries,
    renderNode: (0, _createRenderNode.default)(),
    sidebar: async () => {
      const sidebar = await Promise.resolve().then(() => (0, _interopRequireWildcard2.default)(require("./sidebar")));
      return sidebar.default(configs);
    },
    placeholder: {
      text: locale?.placeholder,
      focusMode: true,
      when: (_, node) => {
        return (0, _everUtils.isEmptyParagraph)(node);
      }
    }
  };
  return plugin;
};

var _default = factory;
exports.default = _default;
