// @ali/4ever-bi/dist-node/plugins/link/linkCard/commands/convertBlockLinkCardToTextLink.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = convertBlockLinkCardToTextLink;

var _everCangjie = require("@ali/4ever-cangjie");

var _removeLink = _interopRequireDefault(require("./removeLink"));

function convertBlockLinkCardToTextLink(controller, node, data) {
  const {
    cardInfo
  } = node.data.metadata;
  let {
    text,
    href
  } = node.data.metadata;

  if (data) {
    text = data.text;
    href = data.href;
  } // 将 text link 转换为 block link card


  const nodeText = text.length ? text : href;

  const textLink = _everCangjie.Inline.create({
    type: 'link',
    nodes: [_everCangjie.Text.create(nodeText)],
    data: {
      href,
      cardInfo: { ...cardInfo,
        displayType: 'link'
      }
    }
  }); // 这里需要先插入，再删除，否则会报找不到 path 的异常


  controller.command(_everCangjie.Commands.moveToEndOfNode, node).command(_everCangjie.Commands.insertBlock, {
    type: _everCangjie.Block.DEFAULT_TYPE,
    nodes: [textLink]
  }).command(_removeLink.default, node);
  return controller;
}

;
