// @ali/4ever-bi/dist-node/plugins/link/linkCard/commands/unwrapLinkCard.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = unwrapLinkCard;

var _everCangjie = require("@ali/4ever-cangjie");

var _removeLink = _interopRequireDefault(require("./removeLink"));

function unwrapLinkCard(controller, node) {
  const {
    text,
    href,
    cardInfo
  } = node.data.metadata; // 将 text link 转换为 block link card

  const nodeText = text.length ? text : href;
  controller.command(_everCangjie.Commands.moveToEndOfNode, node);

  const inline = _everCangjie.Inline.create({
    type: 'link',
    nodes: [_everCangjie.Text.create(nodeText)],
    data: {
      href,
      cardInfo: { ...cardInfo,
        displayType: 'link'
      }
    }
  }); // 将 BlockLinkCard 转换为 textLink 后，调用原有的 unwrapLink 方法
  // 这里需要先插入，再删除，否则会报找不到 path 的异常


  controller.command(_everCangjie.Commands.moveToEndOfNode, node).command(_everCangjie.Commands.insertBlock, {
    type: _everCangjie.Block.DEFAULT_TYPE,
    nodes: [inline]
  }).command(_everCangjie.Commands.moveToRangeOfNode, inline).command('unwrapLink').command(_removeLink.default, node);
  return controller;
}
