// @ali/4ever-bi/dist-web/actionTypes.js
import _extends from "@babel/runtime/helpers/extends";
import { ALIGNMENT_SET } from "./plugins/alignment/actions";
import { BLOCKQUOTE_TOGGLE, BLOCKQUOTE_ENABLE, BLOCKQUOTE_DISABLE } from "./plugins/blockquote/actions";
import { CLEAR_STYLES } from "./plugins/clearStyle/actions";
import { CODE_CHANGE, CODE_FOCUS_TO_CODE_EDITOR, CODE_FOCUS_TO_TEXT_EDITOR, CODE_REMOVE, CODE_INSERT, CODE_CHANGE_DATA, CODE_BLUR } from '@ali/4ever-plugin-code';
import { COLOR_SET } from '@ali/4ever-plugin-color';
import { CLIPBOARD_SET_DECORATIONS, CLIPBOARD_HIDE_PASTEPANEL } from '@ali/4ever-plugin-clipboard';
import { MARK_TOGGLE } from '@ali/4ever-plugin-marks';
import { FONTS_SET } from "./plugins/fonts/actions";
import { FORMAT_PAINTER_COPY, FORMAT_PAINTER_CLEAR_CLIPBOARD } from "./plugins/formatPainter/actions";
import { HEADING_SET } from '@ali/4ever-plugin-heading';
import { HIGHLIGHT_SET } from '@ali/4ever-plugin-highlight';
import { HISTORY_REDO, HISTORY_UNDO } from '@ali/4ever-plugin-history';
import { HR_INSERT } from "./plugins/hr/actions";
import { IMAGE_SELECT, IMAGE_UPDATE, IMAGE_REMOVE, IMAGE_SET_INJECTION } from "./plugins/image/actions";
import { INDENT_INCREASE, INDENT_DECREASE } from "./plugins/indent/actions";
import { LINEHEIGHT_SET } from "./plugins/lineheight/actions";
import { LINK_FOCUS_EDITOR, LINK_HIDE_EDITOR, LINK_REMOVE, LINK_REMOVE_PURE, LINK_UNWRAP, LINK_UPDATE, LINK_UPDATE_PURE, LINK_WRAP, LINK_INJECT_CARDINFO, LINK_REMOVE_INJECTED_CARDINFO, LINK_COPY_LINK, LINK_SWITCH_DISPLAYTYPE, LINK_UNWRAP_PURE_LINK, LINK_SWITCH_LINK_TO_CARD, LINK_UNWRAP_CARD, LINK_EDIT_LINK_CARD, LINK_CLICK_PLACEHOLDER, LINK_HOVER_LINK_CARD } from "./plugins/link/actions";
import { FILE_SELECT, FILE_REMOVE, FILE_UPDATE, IMPORT_MEMO, CARD_SET_INJECTION, TAG_INSERT } from '@ali/4ever-factory';
import { NEWLINE_SET_DATA, MOUNT_PLUS_TOOLBAR, MOUNT_SLASH_TOOLBAR } from "./plugins/newlineGuide/actions";
import { OldActions as ListOldBiActions } from '@ali/4ever-plugin-list';
var LIST_REMOVE_LIST = ListOldBiActions.LIST_REMOVE_LIST,
    LIST_SET_CHECK_STATUS = ListOldBiActions.LIST_SET_CHECK_STATUS,
    LIST_TOGGLE_LIST = ListOldBiActions.LIST_TOGGLE_LIST,
    LIST_SELECT_SYMBOL = ListOldBiActions.LIST_SELECT_SYMBOL,
    CLEAR_RANGE_SELECTION = ListOldBiActions.CLEAR_RANGE_SELECTION,
    SELECT_LIST_SELECTIONS = ListOldBiActions.SELECT_LIST_SELECTIONS,
    CLEAR_LIST_SELECTIONS = ListOldBiActions.CLEAR_LIST_SELECTIONS;
import { OldBiActions as MentionOldBiActions } from '@ali/4ever-plugin-mention';
import { ONLINEVIDEO_INSERT, ONLINEVIDEO_REMOVE, ONLINEVIDEO_UPDATE, ONLINEVIDEO_FOCUS_EDITOR, ONLINEVIDEO_SELECT } from "./plugins/onlineVideo/actions";
import { PARAGRAPH_SET_SPACING } from '@ali/4ever-plugin-paragraph';
import { PLACEHOLDER_REMOVE } from "./plugins/placeholder/actions";
import { QUICK_INSERT_INSERT, QUICK_INSERT_REMOVE, QUICK_INSERT_RESET, QUICK_INSERT_SET, QUICK_INSERT_SHOW } from "./plugins/quickInsert/actions";
import { SZ_SET } from "./plugins/sz/actions";
import { biActionTypes as tableActions } from '@ali/4ever-plugin-table';
import { FORMULA_SET_INJECTION } from "./plugins/formula/actions";
import { oldActions } from '@ali/4ever-plugin-draggable';
import { UPDATE_HOVER_HIGHLIGHT, REMOVE_HOVER_HIGHLIGHT } from "./plugins/hoverBlock/actions";
import { TRANSLATE_INSERT, TRANSLATE_REPLACE } from "./plugins/translate/actions";
import { UNDERLINE_TOGGLE } from "./plugins/underline/actions";
import { VERT_ALIGN_TOGGLE_SUB, VERT_ALIGN_TOGGLE_SUP } from "./plugins/vertAlign/actions";
import { DINGPAN_SELECT, DINGPAN_UPDATE, DINGPAN_REMOVE } from "./plugins/refer/actions";
import { ALBUM_SELECT, VIDEO_SELECT } from "./plugins/video/actions";
import { EVENT_TRACKING } from "./plugins/eventTracking/actions";
import { MOUNT_SELECTION_TOOLBAR } from '@ali/4ever-component';
var MENTION_ACTIVE = MentionOldBiActions.MENTION_ACTIVE,
    MENTION_DEACTIVE = MentionOldBiActions.MENTION_DEACTIVE,
    MENTION_INSERT = MentionOldBiActions.MENTION_INSERT,
    MENTION_SELECT = MentionOldBiActions.MENTION_SELECT,
    MENTION_MULTIPLE_INSERT = MentionOldBiActions.MENTION_MULTIPLE_INSERT,
    MENTION_CANCEL_SELECT = MentionOldBiActions.MENTION_CANCEL_SELECT,
    MENTION_PLACEHOLDER_INSERT = MentionOldBiActions.MENTION_PLACEHOLDER_INSERT,
    MENTION_PLACEHOLDER_ACTIVE = MentionOldBiActions.MENTION_PLACEHOLDER_ACTIVE;
var CLEAR_DROPHOLDER = oldActions.CLEAR_DROPHOLDER,
    UPDATE_DRAGHANDER = oldActions.UPDATE_DRAGHANDER,
    UPDATE_DRAGHIGHLIGHT = oldActions.UPDATE_DRAGHIGHLIGHT,
    REMOVE_DRAGHIGHLIGHT_SINGLE = oldActions.REMOVE_DRAGHIGHLIGHT_SINGLE,
    UPDATE_DARGFRAGMENT = oldActions.UPDATE_DARGFRAGMENT,
    DRAG_MENU_INSERT_AFTER = oldActions.DRAG_MENU_INSERT_AFTER,
    DRAG_MENU_INSERT_BEFORE = oldActions.DRAG_MENU_INSERT_BEFORE,
    DRAG_MENU_CUT_OR_COPY = oldActions.DRAG_MENU_CUT_OR_COPY,
    DRAG_MENU_DELETE = oldActions.DRAG_MENU_DELETE,
    MOUNT_LINE_HEADER_TOOLBAR = oldActions.MOUNT_LINE_HEADER_TOOLBAR;
export default _extends({
  UPDATE_DRAGHANDER: UPDATE_DRAGHANDER,
  UPDATE_DRAGHIGHLIGHT: UPDATE_DRAGHIGHLIGHT,
  UPDATE_DARGFRAGMENT: UPDATE_DARGFRAGMENT,
  REMOVE_DRAGHIGHLIGHT_SINGLE: REMOVE_DRAGHIGHLIGHT_SINGLE,
  UPDATE_HOVER_HIGHLIGHT: UPDATE_HOVER_HIGHLIGHT,
  REMOVE_HOVER_HIGHLIGHT: REMOVE_HOVER_HIGHLIGHT,
  NEWLINE_SET_DATA: NEWLINE_SET_DATA,
  FORMULA_SET_INJECTION: FORMULA_SET_INJECTION,
  TAG_INSERT: TAG_INSERT,
  CARD_SET_INJECTION: CARD_SET_INJECTION,
  CLEAR_DROPHOLDER: CLEAR_DROPHOLDER,
  CLIPBOARD_SET_DECORATIONS: CLIPBOARD_SET_DECORATIONS,
  CLIPBOARD_HIDE_PASTEPANEL: CLIPBOARD_HIDE_PASTEPANEL,
  ALIGNMENT_SET: ALIGNMENT_SET,
  ALBUM_SELECT: ALBUM_SELECT,
  BLOCKQUOTE_TOGGLE: BLOCKQUOTE_TOGGLE,
  BLOCKQUOTE_ENABLE: BLOCKQUOTE_ENABLE,
  BLOCKQUOTE_DISABLE: BLOCKQUOTE_DISABLE,
  CLEAR_STYLES: CLEAR_STYLES,
  CODE_BLUR: CODE_BLUR,
  CODE_CHANGE: CODE_CHANGE,
  CODE_FOCUS_TO_CODE_EDITOR: CODE_FOCUS_TO_CODE_EDITOR,
  CODE_FOCUS_TO_TEXT_EDITOR: CODE_FOCUS_TO_TEXT_EDITOR,
  CODE_REMOVE: CODE_REMOVE,
  CODE_INSERT: CODE_INSERT,
  CODE_CHANGE_DATA: CODE_CHANGE_DATA,
  COLOR_SET: COLOR_SET,
  MARK_TOGGLE: MARK_TOGGLE,
  FILE_SELECT: FILE_SELECT,
  FILE_REMOVE: FILE_REMOVE,
  FILE_UPDATE: FILE_UPDATE,
  IMPORT_MEMO: IMPORT_MEMO,
  FONTS_SET: FONTS_SET,
  FORMAT_PAINTER_COPY: FORMAT_PAINTER_COPY,
  FORMAT_PAINTER_CLEAR_CLIPBOARD: FORMAT_PAINTER_CLEAR_CLIPBOARD,
  HEADING_SET: HEADING_SET,
  HIGHLIGHT_SET: HIGHLIGHT_SET,
  HISTORY_REDO: HISTORY_REDO,
  HISTORY_UNDO: HISTORY_UNDO,
  HR_INSERT: HR_INSERT,
  IMAGE_SELECT: IMAGE_SELECT,
  IMAGE_UPDATE: IMAGE_UPDATE,
  IMAGE_REMOVE: IMAGE_REMOVE,
  INDENT_INCREASE: INDENT_INCREASE,
  INDENT_DECREASE: INDENT_DECREASE,
  LINEHEIGHT_SET: LINEHEIGHT_SET,
  LINK_FOCUS_EDITOR: LINK_FOCUS_EDITOR,
  LINK_HIDE_EDITOR: LINK_HIDE_EDITOR,
  LINK_REMOVE: LINK_REMOVE,
  LINK_REMOVE_PURE: LINK_REMOVE_PURE,
  LINK_SWITCH_DISPLAYTYPE: LINK_SWITCH_DISPLAYTYPE,
  LINK_SWITCH_LINK_TO_CARD: LINK_SWITCH_LINK_TO_CARD,
  LINK_EDIT_LINK_CARD: LINK_EDIT_LINK_CARD,
  LINK_HOVER_LINK_CARD: LINK_HOVER_LINK_CARD,
  LINK_CLICK_PLACEHOLDER: LINK_CLICK_PLACEHOLDER,
  LINK_UNWRAP: LINK_UNWRAP,
  LINK_UPDATE: LINK_UPDATE,
  LINK_UPDATE_PURE: LINK_UPDATE_PURE,
  LINK_WRAP: LINK_WRAP,
  LINK_UNWRAP_CARD: LINK_UNWRAP_CARD,
  LINK_UNWRAP_PURE_LINK: LINK_UNWRAP_PURE_LINK,
  LINK_INJECT_CARDINFO: LINK_INJECT_CARDINFO,
  LINK_REMOVE_INJECTED_CARDINFO: LINK_REMOVE_INJECTED_CARDINFO,
  LINK_COPY_LINK: LINK_COPY_LINK,
  LIST_REMOVE_LIST: LIST_REMOVE_LIST,
  LIST_SET_CHECK_STATUS: LIST_SET_CHECK_STATUS,
  LIST_TOGGLE_LIST: LIST_TOGGLE_LIST,
  LIST_SELECT_SYMBOL: LIST_SELECT_SYMBOL,
  CLEAR_RANGE_SELECTION: CLEAR_RANGE_SELECTION,
  SELECT_LIST_SELECTIONS: SELECT_LIST_SELECTIONS,
  CLEAR_LIST_SELECTIONS: CLEAR_LIST_SELECTIONS,
  MENTION_ACTIVE: MENTION_ACTIVE,
  MENTION_DEACTIVE: MENTION_DEACTIVE,
  MENTION_INSERT: MENTION_INSERT,
  MENTION_SELECT: MENTION_SELECT,
  MENTION_MULTIPLE_INSERT: MENTION_MULTIPLE_INSERT,
  MENTION_CANCEL_SELECT: MENTION_CANCEL_SELECT,
  MENTION_PLACEHOLDER_INSERT: MENTION_PLACEHOLDER_INSERT,
  MENTION_PLACEHOLDER_ACTIVE: MENTION_PLACEHOLDER_ACTIVE,
  ONLINEVIDEO_INSERT: ONLINEVIDEO_INSERT,
  ONLINEVIDEO_UPDATE: ONLINEVIDEO_UPDATE,
  ONLINEVIDEO_REMOVE: ONLINEVIDEO_REMOVE,
  ONLINEVIDEO_FOCUS_EDITOR: ONLINEVIDEO_FOCUS_EDITOR,
  ONLINEVIDEO_SELECT: ONLINEVIDEO_SELECT,
  PARAGRAPH_SET_SPACING: PARAGRAPH_SET_SPACING,
  PLACEHOLDER_REMOVE: PLACEHOLDER_REMOVE,
  QUICK_INSERT_INSERT: QUICK_INSERT_INSERT,
  QUICK_INSERT_REMOVE: QUICK_INSERT_REMOVE,
  QUICK_INSERT_RESET: QUICK_INSERT_RESET,
  QUICK_INSERT_SET: QUICK_INSERT_SET,
  QUICK_INSERT_SHOW: QUICK_INSERT_SHOW,
  SZ_SET: SZ_SET,
  IMAGE_SET_INJECTION: IMAGE_SET_INJECTION
}, tableActions, {
  TRANSLATE_INSERT: TRANSLATE_INSERT,
  TRANSLATE_REPLACE: TRANSLATE_REPLACE,
  UNDERLINE_TOGGLE: UNDERLINE_TOGGLE,
  VERT_ALIGN_TOGGLE_SUB: VERT_ALIGN_TOGGLE_SUB,
  VERT_ALIGN_TOGGLE_SUP: VERT_ALIGN_TOGGLE_SUP,
  VIDEO_SELECT: VIDEO_SELECT,
  DINGPAN_SELECT: DINGPAN_SELECT,
  DINGPAN_UPDATE: DINGPAN_UPDATE,
  DINGPAN_REMOVE: DINGPAN_REMOVE,
  EVENT_TRACKING: EVENT_TRACKING,
  DRAG_MENU_INSERT_AFTER: DRAG_MENU_INSERT_AFTER,
  DRAG_MENU_INSERT_BEFORE: DRAG_MENU_INSERT_BEFORE,
  DRAG_MENU_CUT_OR_COPY: DRAG_MENU_CUT_OR_COPY,
  DRAG_MENU_DELETE: DRAG_MENU_DELETE,
  MOUNT_PLUS_TOOLBAR: MOUNT_PLUS_TOOLBAR,
  MOUNT_SLASH_TOOLBAR: MOUNT_SLASH_TOOLBAR,
  MOUNT_LINE_HEADER_TOOLBAR: MOUNT_LINE_HEADER_TOOLBAR,
  MOUNT_SELECTION_TOOLBAR: MOUNT_SELECTION_TOOLBAR
});
