// @ali/4ever-bi/dist-web/createCustomizedMobileBiPlugins.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";

/**
 * @packageDocumentation
 * @module Editor(Bi)
 */
import { createSchema } from '@ali/4ever-cangjie';
import { mobileBiFactory as exception } from '@ali/4ever-plugin-exception';
import { sortPlugins } from '@ali/4ever-utils';
import autoClose from "./plugins/autoClose";
import { createUnknownCardMobilePlugin, createUnknownTagMobilePlugin } from '@ali/4ever-factory';
import { biFactory as basic } from '@ali/4ever-plugin-basic';
import clipboard from "./plugins/clipboard/mobile";
import history from "./plugins/history/mobile";
import { mobileBiFactory as paragraph } from '@ali/4ever-plugin-paragraph';
import { biFactory as container } from '@ali/4ever-plugin-container';
export default function createCustomizedBiPlugins(plugins) {
  var paragraphPlugin = plugins.paragraph || paragraph;

  var _plugins$unknownCard = plugins.unknownCard,
      unknownCard = _plugins$unknownCard === void 0 ? createUnknownCardMobilePlugin : _plugins$unknownCard,
      _plugins$unknownTag = plugins.unknownTag,
      unknownTag = _plugins$unknownTag === void 0 ? createUnknownTagMobilePlugin : _plugins$unknownTag,
      rest = _objectWithoutPropertiesLoose(plugins, ["unknownCard", "unknownTag"]);

  var allPlugins = _extends({
    exception: exception
  }, rest, {
    container: container,
    unknownCard: unknownCard,
    unknownTag: unknownTag,
    autoClose: autoClose,
    clipboard: clipboard,
    history: history,
    paragraph: paragraphPlugin,
    basic: basic
  });

  return function createBiPlugins(configs) {
    if (configs === void 0) {
      configs = {};
    }

    var array = [];
    var hash = {}; // @ts-ignore

    sortPlugins(allPlugins).forEach(function (_ref) {
      var key = _ref.key,
          factory = _ref.factory;
      var plugin = factory(configs, hash);
      array.push(plugin);
      hash[key] = plugin;
    });
    var schema = createSchema(array);
    array.push({
      schema: schema
    });
    return {
      array: array,
      hash: hash,
      lang: configs.lang || 'zh-CN'
    };
  };
}
