// @ali/4ever-bi/dist-web/index.js
import Provider from "./provider";
import Content from "./content";
import createBiPlugins from "./createBiPlugins";
import createCustomizedBiPlugins from "./createCustomizedBiPlugins";
import actionTypes from "./actionTypes";
import plugins from "./plugins";
import locale_en_US from "./locales/en_US";
import locale_ja_JP from "./locales/ja_JP";
import locale_zh_CN from "./locales/zh_CN";
import locale_zh_HK from "./locales/zh_HK";
import locale_zh_TW from "./locales/zh_TW";
export { Resizable, MenuTitleWithNewBadge, Toolbar, InlineToolbar, ToolbarConfig, ToolBoxLayoutMode, HeaderToolbarMode, ToolbarMode, Sidebar, NewToolbar, NewToolbarConfig, SimpleToolbar // toobar next
, defaultSingleLayout, defaultDisableTargets, defaultDoubleLayout, HeadingToolbar // mobile toolbar
, MobileToolbar, mobileLayout, MobileToolbarConstant // left
, LeftToolbar // selectionBar
, Trigger as SelectionBar } from '@ali/4ever-component';
// 先删掉，待有问题再一一 export
// export * from './plugins/draggable/types';
export * from "./content";
export * from "./pluginsContext";
export * from "./createCustomizedBiPlugins";
export * from "./menu";
import * as _BiMenu from "./menu";
export { _BiMenu as BiMenu };
export { insertTag, setTagData, removeTag, tagFactory, removeCard, setCardData, setCardInjection, insertCard, cardFactory, mobileCardFactory, groupFactory } from '@ali/4ever-factory';
export { insertTemplateButton } from '@ali/4ever-plugin-template-button';
export * from "./types"; // Mobile

export { default as createMobileBiPlugins } from "./createMobileBiPlugins";
export { default as mobilePlugins } from "./plugins/mobile"; // Plugins

export { default as AlignmentPlugin } from "./plugins/alignment";
export { default as MobileAlignmentPlugin } from "./plugins/alignment/mobile";
export { default as AudioPlugin } from "./plugins/audio";
export { default as MobileAudioPlugin } from "./plugins/audio/mobile";
export { biFactory as BasicPlugin, isActions as BasicIsActions } from '@ali/4ever-plugin-basic';
export { biFactory as TemplateButton } from '@ali/4ever-plugin-template-button';
export { default as BlockquotePlugin } from "./plugins/blockquote";
export { default as MobileBlockquotePlugin } from "./plugins/blockquote/mobile";
export { default as BoldPlugin } from "./plugins/bold";
export { default as MobileBoldPlugin } from "./plugins/bold/mobile";
export { default as ClearStylePlugin } from "./plugins/clearStyle";
export { default as MobileClearStylePlugin } from "./plugins/clearStyle/mobile";
export { default as ClipBoardPlugin } from "./plugins/clipboard";
export { default as MobileClipBoardPlugin } from "./plugins/clipboard/mobile";
export { biFactory as CodePlugin, mobileBiFactory as MobileCodePlugin } from '@ali/4ever-plugin-code';
export { biFactory as InlineCodePlugin } from '@ali/4ever-plugin-inline-code';
export { default as ColumnsPlugin } from "./plugins/columns";
export { default as CommentPlugin } from "./plugins/comment";
export { default as MobileCommentPlugin } from "./plugins/comment/mobile";
export { CommonPlugin } from "./plugins/common/commonPlugin";
export { default as CalendarCardPlugin } from "./plugins/calendarCard";
export { default as MobileCalendarCardPlugin } from "./plugins/calendarCard/mobile";
export { default as VoteCardPlugin } from "./plugins/voteCard";
export { default as MobileVoteCardPlugin } from "./plugins/voteCard/mobile";
export { createFindAndReplacePlugin as FindAndReplacePlugin } from '@ali/4ever-plugin-find-and-replace';
export { default as ImTagPlugin } from "./plugins/imTag";
export { default as ColorPlugin } from "./plugins/color";
export { biFactory as ContextMenuPlugin } from '@ali/4ever-plugin-pc-contex-menu';
export { default as EmbedPlugin } from "./plugins/embed";
export { default as MobileEmbedPlugin } from "./plugins/embed/mobile";
export { default as FontsPlugin } from "./plugins/fonts";
export { default as FormatPainterPlugin } from "./plugins/formatPainter";
export { default as HeadingPlugin } from "./plugins/heading";
export { default as HighlightPlugin } from "./plugins/highlight";
export { default as HrPlugin } from "./plugins/hr";
export { default as HistoryPlugin } from "./plugins/history";
export { default as MobileHistoryPlugin } from "./plugins/history/mobile";
export { default as ImagePlugin } from "./plugins/image";
export { default as MobileImagePlugin } from "./plugins/image/mobile";
export { default as IndentPlugin } from "./plugins/indent";
export { default as ItalicPlugin } from "./plugins/italic";
export { default as LineHeightPlugin } from "./plugins/lineheight";
export { default as LinkPlugin } from "./plugins/link";
export { default as LinkCardPlugin } from "./plugins/link/linkCard";
export { default as MobileLinkPlugin } from "./plugins/link/mobile";
export { default as MobileLinkCardPlugin } from "./plugins/link/linkCard/mobile";
export { default as ListPlugin } from "./plugins/list";
export { default as MarksPlugin } from "./plugins/marks";
export { default as MentionPlugin } from "./plugins/mention";
export { default as MobileMentionPlugin } from "./plugins/mention/mobile";
export { default as OnlineVideoPlugin } from "./plugins/onlineVideo";
export { default as MobileOnlineVideoPlugin } from "./plugins/onlineVideo/mobile";
export { biFactory as ParagraphPlugin } from '@ali/4ever-plugin-paragraph';
export { default as PlaceholderPlugin } from "./plugins/placeholder";
export { default as QuickInsertPlugin } from "./plugins/quickInsert";
export { default as NewlineGuidePlugin } from "./plugins/newlineGuide";
export { default as ReferPlugin } from "./plugins/refer"; // todo rebase conflic

export { default as StrikePlugin } from "./plugins/strike";
export { default as SzPlugin } from "./plugins/sz";
export { biFactory as TablePlugin, mobileBiFactory as MobileTablePlugin, biActionTypes as TableActions } from '@ali/4ever-plugin-table';
export { default as TranslatePlugin } from "./plugins/translate";
export { default as UnderlinePlugin } from "./plugins/underline";
export { default as VertAlignPlugin } from "./plugins/vertAlign";
export { default as VideoPlugin } from "./plugins/video";
export { default as DraggablePlugin } from "./plugins/draggable";
export { default as FormulaPlugin } from "./plugins/formula";
export { default as StickerPlugin } from "./plugins/sticker";
export { default as MobileVideoPlugin } from "./plugins/video/mobile";
export { default as EventTrackingPlugin } from "./plugins/eventTracking";
export { default as SubjectPlugin } from "./plugins/subject";
export { default as HeadingGroupPlugin } from "./plugins/headingGroup";
export { default as HeadingGroupNewPlugin } from "./plugins/headingGroupNew";
export { default as ListGroupPlugin } from "./plugins/listGroup";
export { default as CalloutPlugin } from "./plugins/callout";
export { default as QuickPreviewPlugin } from "./plugins/quickPreview";
export { default as LaserPointerPlugin } from "./plugins/laserPointer";
export { default as LeftToolbarPlugin } from "./plugins/leftToolbar";
export { LaserPointerRole } from "./plugins/laserPointer/types";
import * as _LaserPointerActions from "./plugins/laserPointer/actions";
export { _LaserPointerActions as LaserPointerActions };
import * as _LaserPointerUtils from "./plugins/laserPointer/utils";
export { _LaserPointerUtils as LaserPointerUtils };
export { default as PrefPlugin } from "./plugins/pref";
export { default as createCustomizedMobileBiPlugins } from "./createCustomizedMobileBiPlugins";
export { default as HoverToolbarPlugin } from "./plugins/hoverToolbar";
export { default as HoverBlockPlugin } from "./plugins/hoverBlock";
export { Provider, Content, createBiPlugins, createCustomizedBiPlugins, actionTypes, plugins, locale_en_US, locale_ja_JP, locale_zh_CN, locale_zh_HK, locale_zh_TW };
