// @ali/4ever-component/dist-node/Menu/interface.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TYPE_QUERY_ONLY_INLINE_MENU_ITEM = exports.TYPE_INLINE_MENU_ITEM = exports.TYPE_INLINE_MENU_WRAPPER = exports.TYPE_INLINE_MENU_GROUP = exports.TYPE_MENU_ITEM = exports.TYPE_SUB_MENU = exports.TYPE_GROUP = void 0;

/**
 * bi内置的菜单栏组件
 * 菜单栏呈横向排布，MenuBar 是菜单栏入口，Menu是嵌套的菜单
 * **********************************************
 * 数据结构：
 * <MenuBar menus=[{
 *  title: '文档',
 *  children: (<Menu menus={[{
 *    type: TYPE_MENU_ITEM,
 *    title: '文档菜单项1'
 *  }]} />)
 * }, {
 *  title: '插入'
 *  children: (<Menu menus={[{
 *    type: TYPE_SUB_MENU,
 *    title: '文件',
 *    menus: [{
 *      type: TYPE_MENU_ITEM,
 *      title: '本地文件',
 *    }, {
 *      type: TYPE_MENU_ITEM,
 *      title: '在线文件',
 *    }]
 *  }]} />)
 * }] />
 * **********************************************
 * 效果：
 * |-         文档         |         插入          -|
 *            |- 文档菜单项1 - |      |- 文件 - |
 *                                     |- 本地文件 -|
 *                                     |- 在线文件 -|
 */

/**
 * 菜单组（title + 支持的菜单类型）
 */
const TYPE_GROUP = Symbol('TYPE_GROUP');
/**
 * 子菜单（可以嵌套任意支持的菜单类型）
 */

exports.TYPE_GROUP = TYPE_GROUP;
const TYPE_SUB_MENU = Symbol('TYPE_SUB_MENU');
/**
 * 菜单项（最小的菜单粒度）
 */

exports.TYPE_SUB_MENU = TYPE_SUB_MENU;
const TYPE_MENU_ITEM = Symbol('TYPE_MENU_ITEM');
/**
 * 行内菜单
 */

exports.TYPE_MENU_ITEM = TYPE_MENU_ITEM;
const TYPE_INLINE_MENU_GROUP = Symbol('TYPE_INLINE_MENU_GROUP');
/**
 * 行内菜单分组
 */

exports.TYPE_INLINE_MENU_GROUP = TYPE_INLINE_MENU_GROUP;
const TYPE_INLINE_MENU_WRAPPER = Symbol('TYPE_INLINE_MENU_WRAPPER');
/**
 * 菜单横向排列的按钮组件
 */

exports.TYPE_INLINE_MENU_WRAPPER = TYPE_INLINE_MENU_WRAPPER;
const TYPE_INLINE_MENU_ITEM = Symbol('TYPE_INLINE_MENU_ITEM');
/**
 * 仅在 query 场景显示的菜单按钮组件
 */

exports.TYPE_INLINE_MENU_ITEM = TYPE_INLINE_MENU_ITEM;
const TYPE_QUERY_ONLY_INLINE_MENU_ITEM = Symbol('TYPE_QUERY_ONLY_INLINE_MENU_ITEM');
/**
 * 支持的菜单类型
 */

exports.TYPE_QUERY_ONLY_INLINE_MENU_ITEM = TYPE_QUERY_ONLY_INLINE_MENU_ITEM;
