// @ali/4ever-component/dist-node/Menu/menubar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _style = require("./style");

const _createElement = /*#__PURE__*/_react.default.createElement;

/**
 * 菜单栏组件（横向排布）
 * @param props
 */
const MenuBar = props => {
  const {
    menus,
    style,
    ...rest
  } = props;
  const subMenus = menus.map((subMenu, index) => {
    const {
      title,
      children,
      ...subRest
    } = subMenu;
    return /*#__PURE__*/_createElement(_weDesign.SubMenu, (0, _extends2.default)({}, subRest, {
      title: title,
      key: `${subRest.testid}_${index}`
    }), children);
  });
  /**
   * 阻止默认事件，防止文档区域失焦
   */

  const handleMouseDown = (0, _react.useCallback)(ev => {
    ev.preventDefault && ev.preventDefault();
    ev.stopPropagation && ev.stopPropagation();
  }, []);
  return /*#__PURE__*/_createElement(_style.MenuBarWrapper, {
    onMouseDown: handleMouseDown,
    style: style
  }, /*#__PURE__*/_createElement(_weDesign.Menu, (0, _extends2.default)({}, rest, {
    mode: _weDesign.EMenuMode.horizontal,
    selectedKeys: [],
    destroyOverlayWhenInvisible: true
  }), subMenus));
};

var _default = MenuBar;
exports.default = _default;
