// @ali/4ever-plugin-highlight/dist-node/bi/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.biFactory = void 0;

var _meta = require("../utils/meta");

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onCangjieInput = _interopRequireDefault(require("./handlers/onCangjieInput"));

var _createRenderToolbar = _interopRequireDefault(require("./createRenderToolbar"));

var _commands = require("./commands");

var _queries = require("./queries");

var _actions = require("./actions");

const biFactory = configs => {
  const highlightConfig = configs.highlight || {};
  return {
    key: _meta.key,
    toolbar: (0, _toolbar.default)(highlightConfig),
    renderToolbar: (0, _createRenderToolbar.default)(highlightConfig.locale),
    commands: _commands.commands,
    queries: _queries.queries,
    actions: _actions.actions,
    isActions: _actions.isActions,
    onAction: _onAction.default,
    onCangjieInput: (0, _onCangjieInput.default)(),
    models: {
      klass: 'mark',
      type: 'highlight'
    }
  };
};

exports.biFactory = biFactory;
