// @ali/4ever-bi/dist-node/plugins/inlineCode/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toggleInlineCodeMark = _interopRequireDefault(require("./commands/toggleInlineCodeMark"));

var _isInlineCode = _interopRequireDefault(require("./queries/isInlineCode"));

var actions = _interopRequireWildcard(require("./actions"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _onCangjieInput = _interopRequireDefault(require("./handlers/onCangjieInput"));

const commands = {
  toggleInlineCode: _toggleInlineCodeMark.default
};
const queries = {
  isInlineCode: _isInlineCode.default
};

function createInlineCodePlugin() {
  return {
    commands,
    queries,
    actions,
    onAction: _onAction.default,
    onKeyDown: _onKeyDown.default,
    onCangjieInput: (0, _onCangjieInput.default)(),
    models: {
      klass: 'mark',
      type: 'inlineCode'
    }
  };
}

createInlineCodePlugin.actions = actions;
createInlineCodePlugin.commands = commands;
createInlineCodePlugin.queries = queries;
var _default = createInlineCodePlugin;
exports.default = _default;
