// @ali/4ever-plugin-history/dist-node/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "biFactory", {
  enumerable: true,
  get: function () {
    return _index.biFactory;
  }
});
Object.defineProperty(exports, "mobileBiFactory", {
  enumerable: true,
  get: function () {
    return _index2.mobileBiFactory;
  }
});
Object.defineProperty(exports, "HistoryRole", {
  enumerable: true,
  get: function () {
    return _role.HistoryRole;
  }
});
Object.defineProperty(exports, "DEPRECATED_ACTIONS", {
  enumerable: true,
  get: function () {
    return _actions.DEPRECATED_ACTIONS;
  }
});
Object.defineProperty(exports, "HISTORY_REDO", {
  enumerable: true,
  get: function () {
    return _actions.HISTORY_REDO;
  }
});
Object.defineProperty(exports, "HISTORY_UNDO", {
  enumerable: true,
  get: function () {
    return _actions.HISTORY_UNDO;
  }
});

var _index = require("./bi/index");

var _index2 = require("./bi/mobile/index");

var _role = require("./utils/role");

var _actions = require("./bi/actions");
