// @ali/4ever-bi/dist-node/plugins/headingGroup/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everBamboo = require("@ali/4ever-bamboo");

var _createGroupNodes = _interopRequireDefault(require("./handlers/createGroupNodes"));

var _createRenderNode = _interopRequireDefault(require("./handlers/createRenderNode"));

var queries = _interopRequireWildcard(require("./queries"));

var _everFactory = require("@ali/4ever-factory");

const {
  actions
} = _everBamboo.GroupFactory;
const DRAG_BUTTON_OFFSET = -20;
const PI_MODE = 'page_layout';
const createHeadingGroupPlugin = (0, _everFactory.groupFactory)(config => {
  const {
    headingGroup = {}
  } = config;
  const {
    level = 3
  } = headingGroup;
  const shouldOffsetTypes = [];

  for (let i = level; i > 0; i--) {
    shouldOffsetTypes.push(`heading-${i}`);
  }

  ;
  return {
    groupType: 'headingGroup',
    renderNode: (0, _createRenderNode.default)(headingGroup),
    groupNodes: (0, _createGroupNodes.default)(headingGroup),
    queries: {
      getHeadingGroupLevel: () => level,
      isHeadingGroupAvailable: () => true,
      getLeftToolbarOffset: (controller, data) => {
        const {
          node
        } = data;
        const isPi = controller.query('getLayoutMode') === PI_MODE;

        if (!isPi && shouldOffsetTypes.includes(node.type)) {
          return {
            // 这里加上4是因为左侧工具栏的offset为6，但是在标题场景下是2
            x: DRAG_BUTTON_OFFSET + 4,
            y: 0
          };
        }

        return undefined;
      },
      ...queries
    }
  };
}); // @ts-ignore

createHeadingGroupPlugin.actions = actions; // @ts-ignore

createHeadingGroupPlugin.queries = queries;
var _default = createHeadingGroupPlugin;
exports.default = _default;
