// @ali/4ever-bi/dist-node/plugins/alignment/createRenderHoverToolbar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderHoverToolbar;

var React = _interopRequireWildcard(require("react"));

var _weDesignNext = require("@ali/we-design-next");

var _alignmentSelect = require("./components/hoverToolbar/alignmentSelect");

const _createElement = /*#__PURE__*/React.createElement;

function createRenderHoverToolbar(config) {
  return (props, controller, getNode) => {
    const {
      buttonName,
      extraProps
    } = props;
    const {
      onOverlayVisibleChange
    } = extraProps;

    switch (buttonName) {
      case 'default':
        return /*#__PURE__*/_createElement(_weDesignNext.TBItemSlot, {
          display: "iconSelect"
        }, /*#__PURE__*/_createElement(_alignmentSelect.AlignmentSelect, {
          controller: controller,
          getNode: getNode,
          onOverlayVisibleChange: onOverlayVisibleChange,
          locale: config.locale || {}
        }));

      default:
        return null;
    }
  };
}
