// @ali/4ever-component/dist-web/Menu/menubar.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React, { useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { EMenuMode, Menu, SubMenu } from '@ali/we-design';
import { MenuBarWrapper } from "./style";
/**
 * 菜单栏组件（横向排布）
 * @param props
 */

var MenuBar = function MenuBar(props) {
  var menus = props.menus,
      style = props.style,
      rest = _objectWithoutPropertiesLoose(props, ["menus", "style"]);

  var subMenus = menus.map(function (subMenu, index) {
    var title = subMenu.title,
        children = subMenu.children,
        subRest = _objectWithoutPropertiesLoose(subMenu, ["title", "children"]);

    return /*#__PURE__*/_createElement(SubMenu, _extends({}, subRest, {
      title: title,
      key: subRest.testid + "_" + index
    }), children);
  });
  /**
   * 阻止默认事件，防止文档区域失焦
   */

  var handleMouseDown = useCallback(function (ev) {
    ev.preventDefault && ev.preventDefault();
    ev.stopPropagation && ev.stopPropagation();
  }, []);
  return /*#__PURE__*/_createElement(MenuBarWrapper, {
    onMouseDown: handleMouseDown,
    style: style
  }, /*#__PURE__*/_createElement(Menu, _extends({}, rest, {
    mode: EMenuMode.horizontal,
    selectedKeys: [],
    destroyOverlayWhenInvisible: true
  }), subMenus));
};

export default MenuBar;
