// @ali/4ever-bi/dist-node/plugins/alignment/mobile/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _setAlignment = _interopRequireDefault(require("../commands/setAlignment"));

var _getAlignment = _interopRequireDefault(require("../queries/getAlignment"));

var _actions = require("../actions");

var _onAction = _interopRequireDefault(require("../handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("../handlers/onKeyDown"));

const commands = {
  setAlignment: _setAlignment.default
};
const queries = {
  getAlignment: _getAlignment.default
};

function createAlignmentPlugin(config) {
  const alignmentConfig = config.alignment || {};
  return {
    toolbar: (0, _toolbar.default)(alignmentConfig),
    commands,
    queries,
    actions: _actions.actions,
    onAction: _onAction.default,
    onKeyDown: _onKeyDown.default,
    models: []
  };
}

createAlignmentPlugin.actions = _actions.DEPRECATED_ACTIONS;
createAlignmentPlugin.commands = commands;
createAlignmentPlugin.queries = queries;
var _default = createAlignmentPlugin;
exports.default = _default;
