// @ali/4ever-factory/dist-web/cardFactory/mobileCardFactory.js
import _extends from "@babel/runtime/helpers/extends";
import createRenderNode from "./createRenderNode";
import createOnKeyDown from "./handlers/createOnKeyDown";
import createOnAction from "./handlers/createOnAction";
import { removeCard, setCardData, setCardInjection, insertCard } from "./actions";
import { insertCardCommand, removeCardCommand, setCardDataCommand, setCardInjectionCommand } from "./commands";
import { Placeholder } from "./createUnknownCardMobilePlugin";
export function cardFactory(createPlugin) {
  return function createCardPlugin(bizConfig) {
    var myPlugin = createPlugin(bizConfig);
    var menubar = myPlugin.menubar,
        toolbar = myPlugin.toolbar,
        onAction = myPlugin.onAction,
        onCangjiePaste = myPlugin.onCangjiePaste,
        queries = myPlugin.queries,
        onCangjieCopy = myPlugin.onCangjieCopy;
    return {
      cardType: myPlugin.cardType,
      renderNode: createRenderNode(myPlugin, bizConfig, Placeholder),
      onKeyDown: createOnKeyDown(),
      onAction: createOnAction(onAction),
      commands: {
        insertCard: insertCardCommand,
        removeCard: removeCardCommand,
        setCardData: setCardDataCommand,
        setCardInjection: setCardInjectionCommand
      },
      queries: _extends({}, queries, {
        lazyRenderStrategy: function lazyRenderStrategy(_, node) {
          return node.type === 'card' && typeof node.data.height === 'number' ? {
            size: {
              height: node.data.height
            }
          } : undefined;
        }
      }),
      models: {
        klass: 'block',
        type: 'card'
      }
    };
  };
}
export { removeCard, setCardData, setCardInjection, insertCard };
