// @ali/4ever-bi/dist-web/plugins/blockquote/components/quickInsertPanel.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { PluginRoles } from '@ali/4ever-bamboo';
import { QuickInsertCommonPanel as CommonPanel, QuickInsertButton as Button } from '@ali/4ever-component';
import { toggleBlockquote } from "../actions";

var placeholder = /*#__PURE__*/_createElement("svg", {
  width: "64px",
  height: "64px",
  viewBox: "0 0 64 64",
  version: "1.1"
}, /*#__PURE__*/_createElement("g", {
  stroke: "none",
  strokeWidth: "1",
  fill: "none",
  fillRule: "evenodd"
}, /*#__PURE__*/_createElement("g", {
  transform: "translate(-198.000000, -195.000000)",
  fill: "#111F2C"
}, /*#__PURE__*/_createElement("g", {
  transform: "translate(194.000000, 191.000000)"
}, /*#__PURE__*/_createElement("g", {
  transform: "translate(4.000000, 4.000000)"
}, /*#__PURE__*/_createElement("rect", {
  fillOpacity: "0.12",
  x: "0",
  y: "0",
  width: "4",
  height: "64",
  rx: "2"
}), /*#__PURE__*/_createElement("rect", {
  fillOpacity: "0.08",
  transform: "translate(37.500000, 4.000000) scale(1, -1) translate(-37.500000, -4.000000) ",
  x: "11",
  y: "1",
  width: "53",
  height: "6",
  rx: "3"
}), /*#__PURE__*/_createElement("rect", {
  fillOpacity: "0.08",
  transform: "translate(37.500000, 18.000000) scale(1, -1) translate(-37.500000, -18.000000) ",
  x: "11",
  y: "15",
  width: "53",
  height: "6",
  rx: "3"
}), /*#__PURE__*/_createElement("rect", {
  fillOpacity: "0.08",
  transform: "translate(30.000000, 32.000000) scale(1, -1) translate(-30.000000, -32.000000) ",
  x: "11",
  y: "29",
  width: "38",
  height: "6",
  rx: "3"
}), /*#__PURE__*/_createElement("rect", {
  fillOpacity: "0.08",
  transform: "translate(37.500000, 47.000000) scale(1, -1) translate(-37.500000, -47.000000) ",
  x: "11",
  y: "44",
  width: "53",
  height: "6",
  rx: "3"
}), /*#__PURE__*/_createElement("rect", {
  fillOpacity: "0.08",
  transform: "translate(22.000000, 61.000000) scale(1, -1) translate(-22.000000, -61.000000) ",
  x: "11",
  y: "58",
  width: "22",
  height: "6",
  rx: "3"
}))))));

var QuickInsertPanel = function QuickInsertPanel(props) {
  var controller = props.controller,
      locale = props.locale,
      close = props.close;
  var handleInsert = React.useCallback(function () {
    close();
    controller.run('onAction', toggleBlockquote());
  }, [controller, close]);
  return /*#__PURE__*/_createElement(CommonPanel, {
    placeholder: placeholder
  }, /*#__PURE__*/_createElement(Button, {
    onClick: handleInsert,
    role: PluginRoles.blockquote
  }, locale ? locale.buttonTitle : ''));
};

export default QuickInsertPanel;
