// @ali/we-design/dist-web/components/contextmenu/useContextMenuRef.js
import React, { useState, useRef, useEffect, useMemo } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import ReactDOM from 'react-dom';
import { ContextMenu } from "./ContextMenu";
export var useContextMenuRef = function useContextMenuRef(menuNode) {
  var _ContextMenu;

  var ctxMenuRef = useRef(null);

  var _useState = useState(false),
      visible = _useState[0],
      setVisible = _useState[1];

  var _useState2 = useState({
    x: 0,
    y: 0
  }),
      position = _useState2[0],
      setPosition = _useState2[1];

  var oRoot = useMemo(function () {
    return document.createElement('div');
  }, []);
  useEffect(function () {
    /**
     * 右键菜单操作
     */
    var handleContextMenu = function handleContextMenu(e) {
      e.preventDefault();
      setPosition({
        x: e.clientX,
        y: e.clientY
      });
      setVisible(true);
    };

    var oElem = ctxMenuRef.current;

    if (oElem) {
      oElem.addEventListener('contextmenu', handleContextMenu);
    }

    return function () {
      if (oElem) {
        oElem.removeEventListener('contextmenu', handleContextMenu);
      }
    };
  }, []);
  useEffect(function () {
    ReactDOM.render(_ContextMenu || (_ContextMenu = /*#__PURE__*/_createElement(ContextMenu, {
      position: position,
      visible: visible,
      onVisibleChange: setVisible
    }, menuNode)), oRoot);
  }, [visible, position, setVisible, menuNode, oRoot]);
  return ctxMenuRef;
};
