// @ali/4ever-plugin-table/dist-web/bi/mobile/components/floatToolbar/basic.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { FloatToolbarMobile } from '@ali/we-design-next';
var Row = FloatToolbarMobile.Row,
    Divider = FloatToolbarMobile.Divider,
    Slot = FloatToolbarMobile.Slot;
export var BasicFloatToolbar = function BasicFloatToolbar(props) {
  var toolbarLayout = props.toolbarLayout,
      style = props.style;
  var shouldRender = toolbarLayout.length > 0;
  return shouldRender ? /*#__PURE__*/_createElement(FloatToolbarMobile, {
    style: _extends({
      width: 'fit-content',
      zIndex: 10000,
      position: 'fixed'
    }, style)
  }, toolbarLayout.map(function (configs, i) {
    return /*#__PURE__*/_createElement(React.Fragment, null, /*#__PURE__*/_createElement(Row, {
      key: "row_" + i
    }, configs.map(function (config) {
      return /*#__PURE__*/_createElement(Slot, {
        key: config.key,
        title: config.title,
        icon: config.icon,
        onClick: config.onClick
      });
    })), i !== toolbarLayout.length - 1 && /*#__PURE__*/_createElement(Divider, {
      key: "divider_" + i
    }));
  })) : null;
};
