// @ali/4ever-plugin-image/dist-web/mo/serializer/jsonMLToHTML.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { getTagName, getAttributes, sanitizeUrl } from '@ali/4ever-utils';

var Image = function Image(props) {
  var src = props.src,
      style = props.style,
      path = props.path,
      config = props.config,
      rest = _objectWithoutPropertiesLoose(props, ["src", "style", "path", "config"]);

  return /*#__PURE__*/_createElement("img", _extends({
    src: sanitizeUrl(src, config == null ? void 0 : config.protocolAllowList),
    style: style,
    key: path.toString()
  }, rest));
};

export default (function (config) {
  return {
    name: 'img',
    match: function match(node) {
      return getTagName(node) === 'img';
    },
    convert: function convert(node, path) {
      // @ts-ignore
      var _getAttributes = getAttributes(node),
          src = _getAttributes.src,
          width = _getAttributes.width,
          height = _getAttributes.height,
          opacity = _getAttributes.opacity,
          outline = _getAttributes.outline;

      var border;

      if (outline && outline.width) {
        border = outline.width + "px solid " + (outline.color ? outline.color : '#000000');
      }

      if (!src) return null;
      return /*#__PURE__*/_createElement(Image, {
        src: src,
        style: {
          width: width,
          height: height,
          opacity: opacity,
          border: border
        },
        path: path,
        config: config
      });
    }
  };
});
