// @ali/we-design-next/dist-node/display/gallery/Gallery.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Gallery = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _styled = require("./styled");
var _common = require("../../common");
var _RenderEmptyItem;
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const _createElement = /*#__PURE__*/_react.default.createElement;
const RenderEmptyItem = () => {
  return /*#__PURE__*/_createElement("div", {
    style: {
      width: '50px',
      height: '50px'
    }
  });
};
const Gallery = /*#__PURE__*/(0, _react.memo)(props => {
  const {
    items = [],
    columnCount = 3,
    RenderItem = () => _RenderEmptyItem || (_RenderEmptyItem = /*#__PURE__*/_createElement(RenderEmptyItem, null)),
    value,
    defaultValue,
    onChange,
    ...rest
  } = props;
  const [selected, setSelected] = (0, _common.useDefaultProp)(defaultValue, value, () => {});
  const [hoveredItem, setHoveredItem] = (0, _react.useState)();
  const itemRows = (0, _react.useMemo)(() => {
    const tmp = [];
    for (let i = 0; i < items.length; i += columnCount) {
      tmp.push(items.slice(i, i + columnCount));
    }
    return tmp;
  }, [items, columnCount]);
  const chooseItem = item => {
    setSelected(item.key);
    onChange === null || onChange === void 0 ? void 0 : onChange(item.key);
    setHoveredItem(item.key);
  };
  const hoverItem = (event, item) => {
    event.stopPropagation();
    event.preventDefault();
    setHoveredItem(item.key);
  };
  const isSelected = item => {
    return item.key === selected;
  };
  const isHovered = item => {
    return item.key === hoveredItem;
  };
  const getHoveredItem = () => {
    return items.findIndex(item => {
      return isHovered(item);
    });
  };
  const handleKeyDown = event => {
    if (!event) return;
    const curIndex = getHoveredItem();
    const keyCode = event.keyCode;
    switch (keyCode) {
      case _common.KEY_CODES.LEFT:
        if (curIndex === -1) return hoverItem(event, items[0]);
        if (curIndex > 0) {
          return hoverItem(event, items[curIndex - 1]);
        }
        break;
      case _common.KEY_CODES.UP:
        if (curIndex === -1) return hoverItem(event, items[0]);
        if (curIndex - columnCount >= 0) {
          return hoverItem(event, items[curIndex - columnCount]);
        }
        break;
      case _common.KEY_CODES.RIGHT:
        if (curIndex === -1) return hoverItem(event, items[0]);
        if (curIndex !== items.length - 1) {
          return hoverItem(event, items[curIndex + 1]);
        }
        break;
      case _common.KEY_CODES.DOWN:
        if (curIndex === -1) return hoverItem(event, items[0]);
        if (curIndex + columnCount < items.length) {
          return hoverItem(event, items[curIndex + columnCount]);
        }
        break;
      case _common.KEY_CODES.ENTER:
        if (curIndex === -1) return;
        return chooseItem === null || chooseItem === void 0 ? void 0 : chooseItem(items[curIndex]);
    }
  };
  (0, _react.useEffect)(() => {
    if (items.some(item => !item.key)) {
      console.warn('every item in items should have unique key ');
    }
  }, [items]);
  return /*#__PURE__*/_createElement(_styled.GalleryContainer, (0, _extends2.default)({
    tabIndex: -1,
    onKeyDown: handleKeyDown
  }, (0, _common.mergeRestProps)(rest, 'wdn-gallery')), itemRows.map((row, index) => /*#__PURE__*/_createElement(_styled.GalleryItemRow, null, row.map(item => /*#__PURE__*/_createElement(_styled.GalleryItemWrapper, {
    onClick: () => chooseItem(item),
    key: item.key || index
  }, /*#__PURE__*/_createElement(_styled.GalleryItem, null, RenderItem({
    data: item,
    index
  })), /*#__PURE__*/_createElement(_styled.GalleryBorder, {
    actived: isSelected(item),
    hovered: isHovered(item)
  }))))));
});
exports.Gallery = Gallery;
