// @ali/4ever-bi/dist-web/plugins/quickInsert/createRenderEditable.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import QuickInsertTrigger from "./components/quickInsertTrigger";
import { getQuickInsertKeyword } from '@ali/4ever-utils';

function createRenderEditable(groups, insertMap, config) {
  // @ts-ignore
  return function renderEditable(_, controller, next) {
    var isNecessary = controller.query('isSelectionInQuickInsert');

    if (!isNecessary) {
      return next();
    }

    var decoration = controller.query('getQuickInsertDecoration');
    var value = decoration == null ? void 0 : decoration.mark.data.plugin; // 这里应该根据注册的插件传入type

    var text = getQuickInsertKeyword(controller, {
      type: 'mention'
    });
    return /*#__PURE__*/_createElement(React.Fragment, null, next(), /*#__PURE__*/_createElement(QuickInsertTrigger, {
      groups: groups,
      value: value,
      text: text,
      controller: controller,
      insertMap: insertMap,
      config: config
    }));
  };
}

export default createRenderEditable;
