// @ali/4ever-plugin-color-blocks/dist-web/mo/models/index.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import { Block, Text } from '@ali/4ever-cangjie';
import { ELEMENT_TYPE } from "../../utils/meta";
import { Container } from '@ali/4ever-plugin-container';
export var ColorBlocks = /*#__PURE__*/function (_Container) {
  _inheritsLoose(ColorBlocks, _Container);

  function ColorBlocks() {
    return _Container.apply(this, arguments) || this;
  }

  ColorBlocks.isColorBlocks = function isColorBlocks(node) {
    return Container.isContainer(node, ColorBlocks.SUB_TYPE);
  };

  ColorBlocks.createColorBlocks = function createColorBlocks(metadata, nodes) {
    if (metadata === void 0) {
      metadata = {};
    }

    return Container.createContainer({
      data: {
        subType: ColorBlocks.SUB_TYPE,
        metadata: metadata
      },
      nodes: nodes || [Block.create({
        type: 'paragraph',
        nodes: [Text.create()]
      })]
    });
  };

  return ColorBlocks;
}(Container);
ColorBlocks.SUB_TYPE = ELEMENT_TYPE;
