// @ali/we-design-next/dist-node/basic/text/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "H1", {
  enumerable: true,
  get: function () {
    return _styled.H1;
  }
});
Object.defineProperty(exports, "H2", {
  enumerable: true,
  get: function () {
    return _styled.H2;
  }
});
exports.TEXT_CSS = void 0;
Object.defineProperty(exports, "Text", {
  enumerable: true,
  get: function () {
    return _styled.Text;
  }
});
Object.defineProperty(exports, "TextInline", {
  enumerable: true,
  get: function () {
    return _styled.TextInline;
  }
});
Object.defineProperty(exports, "TextSub", {
  enumerable: true,
  get: function () {
    return _styled.TextSub;
  }
});
Object.defineProperty(exports, "TextSubInline", {
  enumerable: true,
  get: function () {
    return _styled.TextSubInline;
  }
});
var _styled = require("./styled");
const TEXT_CSS = {
  H1: _styled.CSS_H1,
  H2: _styled.CSS_H2,
  TEXT: _styled.CSS_TEXT,
  TEXT_SUB: _styled.CSS_TEXT_SUB
};
exports.TEXT_CSS = TEXT_CSS;
