// @ali/we-design-next/dist-node/basic/text/styled.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TextSubInline = exports.TextSub = exports.TextInline = exports.Text = exports.H2 = exports.H1 = exports.CSS_TEXT_SUB = exports.CSS_TEXT = exports.CSS_H2 = exports.CSS_H1 = void 0;
var _styledComponents = _interopRequireWildcard(require("styled-components"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const CSS_H1 = /*#__PURE__*/(0, _styledComponents.css)(["font-size:16px;line-height:24px;user-select:none;"]);
exports.CSS_H1 = CSS_H1;
const CSS_H2 = /*#__PURE__*/(0, _styledComponents.css)(["font-size:14px;line-height:20px;user-select:none;"]);
exports.CSS_H2 = CSS_H2;
const CSS_TEXT = /*#__PURE__*/(0, _styledComponents.css)(["font-size:13px;line-height:20px;user-select:none;"]);
exports.CSS_TEXT = CSS_TEXT;
const CSS_TEXT_SUB = /*#__PURE__*/(0, _styledComponents.css)(["font-size:12px;line-height:18px;user-select:none;"]);
exports.CSS_TEXT_SUB = CSS_TEXT_SUB;
const H1 = /*#__PURE__*/_styledComponents.default.h1(["", ""], CSS_H1);
exports.H1 = H1;
const H2 = /*#__PURE__*/_styledComponents.default.h2(["", ""], CSS_H2);
exports.H2 = H2;
const Text = /*#__PURE__*/_styledComponents.default.div(["", ""], CSS_TEXT);
exports.Text = Text;
const TextInline = /*#__PURE__*/_styledComponents.default.span(["", ""], CSS_TEXT);
exports.TextInline = TextInline;
const TextSub = /*#__PURE__*/_styledComponents.default.div(["", ""], CSS_TEXT_SUB);
exports.TextSub = TextSub;
const TextSubInline = /*#__PURE__*/_styledComponents.default.span(["", ""], CSS_TEXT_SUB);
exports.TextSubInline = TextSubInline;
