// @ali/we-design-next/dist-web/office/guidance/MapAnchor.js
var _PosIcon, _div;
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import ReactDOM from 'react-dom';
import { PosIcon } from "./posIcon";
import { Pos } from "./style";
export var Anchor = function Anchor(props) {
  var _props$left = props.left,
    left = _props$left === void 0 ? 0 : _props$left,
    _props$top = props.top,
    top = _props$top === void 0 ? 0 : _props$top,
    _props$duration = props.duration,
    duration = _props$duration === void 0 ? 500 : _props$duration;
  var dur = (duration / 1000).toFixed(3);
  return /*#__PURE__*/_createElement(Pos, {
    style: {
      top: top - 20,
      left: left - 10,
      transition: "top " + dur + "s ease-in-out, left " + dur + "s ease-in-out"
    }
  }, _PosIcon || (_PosIcon = /*#__PURE__*/_createElement(PosIcon, null)), _div || (_div = /*#__PURE__*/_createElement("div", {
    className: "circle"
  })));
};
var current = null;
export var MapAnchor = {
  place: function place(left, top) {
    var _Anchor;
    MapAnchor.remove();
    requestAnimationFrame(function () {
      ReactDOM.render(_Anchor || (_Anchor = /*#__PURE__*/_createElement(Anchor, {
        left: left,
        top: top,
        duration: 0
      })), ensureContainer());
      current = [left, top];
    });
  },
  move: function move(left, top) {
    var _Anchor2;
    if (current === null) {
      return;
    }
    var _current = current,
      pLeft = _current[0],
      pTop = _current[1];
    var dLeft = left - pLeft;
    var dTop = top - pTop;
    var dist = Math.sqrt(dLeft * dLeft + dTop * dTop);
    var duration = dist < 200 ? 200 : dist > 1000 ? 1000 : dist;
    requestAnimationFrame(function () {
      ReactDOM.render(_Anchor2 || (_Anchor2 = /*#__PURE__*/_createElement(Anchor, {
        left: left,
        top: top,
        duration: duration
      })), ensureContainer());
      current = [left, top];
    });
  },
  remove: function remove() {
    ReactDOM.unmountComponentAtNode(ensureContainer());
    current = null;
  }
};
var ensureContainer = function ensureContainer() {
  var id = 'wdn-map-anchor-root';
  var divEle = document.querySelector("div#" + id);
  if (!divEle) {
    var ele = document.createElement('div');
    ele.id = id;
    ele.style.position = 'absolute';
    ele.style.width = '100%';
    ele.style.top = '0px';
    ele.style.left = '0px';
    document.body.appendChild(ele);
  }
  ;
  return document.querySelector("div#" + id);
};
