// @ali/4ever-bi/dist-web/plugins/sz/components/SzMobileButton.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { useDefaultSz } from '@ali/4ever-component';
import { setSz } from "../actions";
import { szList } from '@ali/4ever-plugin-sz';
var SzWrapper = /*#__PURE__*/styled.div(["flex:0.8;margin-right:1px;margin-bottom:12px;display:flex;flex-direction:column;align-items:center;"]);
var Icon = /*#__PURE__*/styled.img(["width:20px;height:20px;"]);
var Line1 = /*#__PURE__*/styled.div(["font-size:12px;color:#171a1d;line-height:16px;"]);
var Line2 = /*#__PURE__*/styled.div(["font-size:17px;font-weight:500;color:#171a1d;line-height:20px;"]);
export function MobileSizeInc(props) {
  var controller = props.controller,
      item = props.item,
      Wrapper = props.Wrapper;
  var defaultSz = useDefaultSz();
  var sz = controller.query('getSize', defaultSz);
  var handleClick = React.useCallback(function () {
    var newSize = szList.find(function (s) {
      return s > sz;
    }) || szList[szList.length - 1];
    controller.run('onAction', setSz(newSize));
  }, [controller, sz]);
  return /*#__PURE__*/_createElement(Wrapper, {
    type: "horiz",
    item: item,
    onClick: handleClick,
    icon: "sizePlus"
  });
}
export function MobileSizeDec(props) {
  var controller = props.controller,
      item = props.item,
      Wrapper = props.Wrapper;
  var defaultSz = useDefaultSz();
  var sz = controller.query('getSize', defaultSz);
  var handleClick = React.useCallback(function () {
    var newSize = [].concat(szList).reverse().find(function (s) {
      return s < sz;
    }) || szList[0];
    controller.run('onAction', setSz(newSize));
  }, [controller, sz]);
  return /*#__PURE__*/_createElement(Wrapper, {
    type: "horiz",
    item: item,
    onClick: handleClick,
    icon: "sizeMinus"
  });
}
export function MobileSizeNow(props) {
  var controller = props.controller,
      locale = props.locale;
  var defaultSz = useDefaultSz();
  var sz = controller.query('getSize', defaultSz);
  return /*#__PURE__*/_createElement(SzWrapper, null, /*#__PURE__*/_createElement(Line1, null, locale == null ? void 0 : locale.titleMobile), /*#__PURE__*/_createElement(Line2, null, sz));
}
