// @ali/4ever-plugin-code/dist-web/bi/components/ToolbarMore.js
import { InlineToolbar } from '@ali/we-toolbar';
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { ToolbarCutNormal, ToolbarCopyNormal, ToolbarDeleteNormal } from '@ali/we-icon';
import { removeCode } from "../actions";
;

var _ref = /*#__PURE__*/_createElement(ToolbarCopyNormal, null);

var _ref2 = /*#__PURE__*/_createElement(ToolbarCutNormal, null);

var _ref3 = /*#__PURE__*/_createElement(ToolbarDeleteNormal, null);

export default function ToolbarMore(props) {
  var controller = props.controller,
      _getContainer = props.getContainer,
      nodeKey = props.nodeKey,
      locale = props.locale;

  var _onClick = React.useCallback(function (type) {
    if (type === 'copy' || type === 'cut') {
      copyOrCutNode(controller, type, nodeKey);
    } else if (type === 'delete') {
      controller.run('onAction', removeCode({
        key: nodeKey
      }));
    }
  }, []);

  return /*#__PURE__*/_createElement(InlineToolbar, {
    getContainer: function getContainer() {
      return (_getContainer == null ? void 0 : _getContainer()) || document.body;
    },
    moreMenuConfig: {
      options: [[{
        title: locale.copy,
        icon: _ref,
        onClick: function onClick() {
          return _onClick('copy');
        }
      }, {
        title: locale.cut || 'Cut',
        icon: _ref2,
        onClick: function onClick() {
          return _onClick('cut');
        }
      }, {
        title: locale["delete"],
        icon: _ref3,
        onClick: function onClick() {
          return _onClick('delete');
        }
      }]]
    },
    width: 300
  });
}

function copyOrCutNode(controller, type, nodeKey) {
  var node = controller.value.document.getNode(nodeKey);

  if (node !== null) {
    var rst = controller.query('copyNodes', node);

    if (rst instanceof Promise) {
      rst["catch"](function (e) {
        return console.error('copy error', e);
      });
    }

    if (type === 'cut') {
      controller.run('onAction', removeCode({
        key: nodeKey
      }));
    }
  }
}
