// @ali/4ever-plugin-code/dist-web/bi/components/ToolbarButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { IconButton as Button } from '@ali/we-toolbar';
import { ToolbarCodeNormal } from '@ali/we-icon';
import { insertCode } from "../actions";
import { CodeRole } from "../../utils/role";
var iconStyle = {
  width: 20,
  height: 20
};

var _ref = /*#__PURE__*/_createElement(ToolbarCodeNormal, {
  style: iconStyle
});

var ToolbarButton = function ToolbarButton(props) {
  var controller = props.controller,
      locale = props.locale,
      rest = _objectWithoutPropertiesLoose(props, ["controller", "locale"]);

  var handleClick = React.useCallback(function () {
    return controller.run('onAction', insertCode());
  }, [controller]);
  return /*#__PURE__*/_createElement(Button, _extends({}, rest, {
    title: locale.buttonTitle,
    tooltip: locale.buttonTitle,
    testid: "bi-toolbar-code-block",
    icon: _ref,
    onClick: handleClick,
    role: CodeRole.code
  }));
};

export default /*#__PURE__*/React.memo(ToolbarButton);
