// @ali/we-design-next/dist-node/form/date-picker/panels/DateTimePanel/DateTimePanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DateTimePanel = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _DatePanel = require("../DatePanel");
var _TimePanel = require("../TimePanel");
var _styled = require("./styled");
const _createElement = /*#__PURE__*/_react.default.createElement;
const DateTimePanel = /*#__PURE__*/_react.default.memo(props => {
  const {
    value,
    showTime,
    curPanel,
    onSelect,
    showNow,
    timeFormat,
    ...rest
  } = props;
  return /*#__PURE__*/_createElement(_styled.DateTimePanelWrapper, null, /*#__PURE__*/_createElement(_DatePanel.DatePanel, (0, _extends2.default)({
    curPanel: curPanel,
    value: value,
    onSelect: onSelect,
    showTime: showTime,
    showNow: showNow
  }, rest)), showTime && curPanel === 'date' ? /*#__PURE__*/_createElement(_TimePanel.TimePanel, {
    value: value,
    onSelect: onSelect,
    showNow: showNow,
    timeFormat: timeFormat
  }) : null);
});
exports.DateTimePanel = DateTimePanel;
