// @ali/code-markdown/esm/editorOnly/chooseItem.js
import * as React from 'react';
import classNames from 'classnames';

var ChooseItem = function ChooseItem(props) {
  var _props$ItemList5, _props$ItemList6, _props$ItemList7;

  var _React$useState = React.useState(0),
      activeIndex = _React$useState[0],
      setActiveIndex = _React$useState[1];

  var _React$useState2 = React.useState(0),
      positonX = _React$useState2[0],
      setPositonX = _React$useState2[1];

  var _React$useState3 = React.useState(0),
      positonY = _React$useState3[0],
      setPositonY = _React$useState3[1];

  var chooseKeyEvent = function chooseKeyEvent(e) {
    var _props$ItemList;

    var chooseContainer = document.getElementById('at-mentions-dialog');

    if (props.chineseInputBool) {
      e.preventDefault();
      return;
    }

    if (e.key === 'Enter' && (_props$ItemList = props.ItemList) !== null && _props$ItemList !== void 0 && _props$ItemList.length) {
      props.selectFunc(props.ItemList[activeIndex]);
      e.preventDefault();
    }

    if (e.key === 'ArrowUp' || e.key === 'ArrowLeft') {
      if (activeIndex === 0) {
        var _props$ItemList2, _props$ItemList3;

        chooseContainer === null || chooseContainer === void 0 ? void 0 : chooseContainer.scrollTo({
          top: 40 * ((_props$ItemList2 = props.ItemList) === null || _props$ItemList2 === void 0 ? void 0 : _props$ItemList2.length),
          behavior: "smooth"
        });
        setActiveIndex(((_props$ItemList3 = props.ItemList) === null || _props$ItemList3 === void 0 ? void 0 : _props$ItemList3.length) - 1);
      } else {
        setActiveIndex(activeIndex - 1);

        if (40 * (activeIndex - 1) < (chooseContainer === null || chooseContainer === void 0 ? void 0 : chooseContainer.scrollTop)) {
          chooseContainer === null || chooseContainer === void 0 ? void 0 : chooseContainer.scrollTo({
            top: 40 * (activeIndex - 1),
            behavior: "smooth"
          });
        }
      }

      e.preventDefault();
    }

    if (e.key === 'ArrowDown' || e.key === 'ArrowRight') {
      var _props$ItemList4;

      if (activeIndex === ((_props$ItemList4 = props.ItemList) === null || _props$ItemList4 === void 0 ? void 0 : _props$ItemList4.length) - 1) {
        chooseContainer === null || chooseContainer === void 0 ? void 0 : chooseContainer.scrollTo({
          top: 0,
          behavior: "smooth"
        });
        setActiveIndex(0);
      } else {
        setActiveIndex(activeIndex + 1);

        if (activeIndex > 2) {
          chooseContainer === null || chooseContainer === void 0 ? void 0 : chooseContainer.scrollTo({
            top: 40 * (activeIndex - 2),
            behavior: "smooth"
          });
        }
      }

      e.preventDefault();
    }
  };

  React.useEffect(function () {
    window.addEventListener('keydown', chooseKeyEvent);
    return function () {
      window.removeEventListener('keydown', chooseKeyEvent);
    };
  }, [activeIndex, (_props$ItemList5 = props.ItemList) === null || _props$ItemList5 === void 0 ? void 0 : _props$ItemList5.length, props.chineseInputBool]);
  React.useEffect(function () {
    setActiveIndex(0);
    var chooseContainer = document.getElementById('at-mentions-dialog');
    chooseContainer === null || chooseContainer === void 0 ? void 0 : chooseContainer.scrollTo({
      top: 0,
      behavior: "smooth"
    });

    if (props.dialogPosition.overflowY) {
      setPositonY(props.dialogPosition.y - chooseContainer.clientHeight);
    } else {
      setPositonY(props.dialogPosition.y + 30);
    }

    if (props.dialogPosition.overflowX) {
      setPositonX(props.dialogPosition.x - (chooseContainer === null || chooseContainer === void 0 ? void 0 : chooseContainer.clientWidth));
    } else {
      setPositonX(props.dialogPosition.x + 10);
    }
  }, [(_props$ItemList6 = props.ItemList) === null || _props$ItemList6 === void 0 ? void 0 : _props$ItemList6.length, props.dialogPosition.x]);
  return /*#__PURE__*/React.createElement("div", {
    id: "at-mentions-dialog",
    className: "at-mentions__dialog",
    style: {
      top: positonY,
      left: positonX
    }
  }, (_props$ItemList7 = props.ItemList) === null || _props$ItemList7 === void 0 ? void 0 : _props$ItemList7.map(function (item, index) {
    return /*#__PURE__*/React.createElement("div", {
      className: classNames('person-item', {
        'person-item--active': index === activeIndex
      }),
      key: item.value,
      onClick: function onClick() {
        return props.selectFunc(item);
      }
    }, item.avatar && /*#__PURE__*/React.createElement("img", {
      src: item.avatar,
      className: "person-item__avatar"
    }), /*#__PURE__*/React.createElement("div", {
      className: "person-item__name"
    }, item.name, /*#__PURE__*/React.createElement("span", {
      className: "person-item__message"
    }, item.message)));
  }));
};

export default ChooseItem;
