// @ali/4ever-plugin-table/dist-node/bi/mobile/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mobileBiFactory = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _createRenderNode = _interopRequireDefault(require("./createRenderNode"));

var _createOnAction = _interopRequireDefault(require("./handlers/createOnAction"));

var _onCangjieInput = _interopRequireDefault(require("../handlers/onCangjieInput"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _onCangjiePaste = _interopRequireDefault(require("../handlers/onCangjiePaste"));

var _queries = _interopRequireDefault(require("./queries"));

var _insertTable = _interopRequireDefault(require("../commands/insertTable"));

var _createOnCangjieCopy = _interopRequireDefault(require("../handlers/createOnCangjieCopy"));

var _createOnCangjieCut = _interopRequireDefault(require("../handlers/createOnCangjieCut"));

var _mo = require("../../mo");

var _getNodeStyle = _interopRequireDefault(require("../../common/handlers/getNodeStyle"));

const commands = {
  insertTable: _insertTable.default
};

const mobileBiFactory = configs => {
  const tableConfig = configs.table || {};
  const {
    clipboard = {}
  } = configs;
  const mobileConfig = { ...configs,
    draggable: {
      disabled: true
    },
    table: tableConfig
  };
  return { ...(0, _mo.moFactory)({}),
    toolbar: (0, _toolbar.default)(tableConfig),
    // @ts-ignore
    renderNode: (0, _createRenderNode.default)(mobileConfig),
    onAction: (0, _createOnAction.default)(tableConfig),
    onCangjieInput: _onCangjieInput.default,
    onKeyDown: _onKeyDown.default,
    onCangjieCopy: (0, _createOnCangjieCopy.default)(true),
    onCangjieCut: (0, _createOnCangjieCut.default)(true),
    // TODO: 类型生效后这里就有 typing 问题 了
    onCangjiePaste: (0, _onCangjiePaste.default)(clipboard, tableConfig),
    getNodeStyle: _getNodeStyle.default,
    queries: _queries.default,
    commands,
    models: [{
      klass: 'block',
      type: 'table'
    }, {
      klass: 'block',
      type: 'table-cell'
    }, {
      klass: 'block',
      type: 'table-row'
    }]
  };
};

exports.mobileBiFactory = mobileBiFactory;
