// @ali/4ever-factory/dist-node/cardFactory/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cardFactory = cardFactory;
Object.defineProperty(exports, "removeCard", {
  enumerable: true,
  get: function () {
    return _actions.removeCard;
  }
});
Object.defineProperty(exports, "setCardData", {
  enumerable: true,
  get: function () {
    return _actions.setCardData;
  }
});
Object.defineProperty(exports, "setCardInjection", {
  enumerable: true,
  get: function () {
    return _actions.setCardInjection;
  }
});
Object.defineProperty(exports, "insertCard", {
  enumerable: true,
  get: function () {
    return _actions.insertCard;
  }
});

var _createRenderNode = _interopRequireDefault(require("./createRenderNode"));

var _createOnKeyDown = _interopRequireDefault(require("./handlers/createOnKeyDown"));

var _createOnAction = _interopRequireDefault(require("./handlers/createOnAction"));

var _actions = require("./actions");

var _commands = require("./commands");

var _createOnCangjieCopy = _interopRequireDefault(require("./handlers/createOnCangjieCopy"));

var _createOnCangjiePaste = _interopRequireDefault(require("./handlers/createOnCangjiePaste"));

var _everComponent = require("@ali/4ever-component");

var _everPluginCard = require("@ali/4ever-plugin-card");

function cardFactory(createPlugin) {
  return function createCardPlugin(bizConfig) {
    const myPlugin = createPlugin(bizConfig);
    const {
      menubar,
      toolbar,
      onAction,
      onCangjiePaste,
      queries,
      onCangjieCopy,
      renderHoverToolbar
    } = myPlugin;
    return { ...(0, _everPluginCard.moFactory)({}),
      key: myPlugin.cardType,
      cardType: myPlugin.cardType,
      menubar,
      // @ts-ignore
      toolbar,
      renderNode: (0, _createRenderNode.default)(myPlugin, bizConfig, _everComponent.Placeholder),
      onKeyDown: (0, _createOnKeyDown.default)(),
      onAction: (0, _createOnAction.default)(onAction),
      onCangjieCopy: (0, _createOnCangjieCopy.default)(onCangjieCopy),
      onCangjiePaste: (0, _createOnCangjiePaste.default)(onCangjiePaste),
      commands: {
        insertCard: _commands.insertCardCommand,
        removeCard: _commands.removeCardCommand,
        setCardData: _commands.setCardDataCommand,
        setCardInjection: _commands.setCardInjectionCommand
      },
      queries: { ...queries,

        lazyRenderStrategy(_, node) {
          return node.type === 'card' && typeof node.data.height === 'number' ? {
            size: {
              height: node.data.height
            }
          } : undefined;
        }

      },
      models: {
        klass: 'block',
        type: 'card'
      },
      renderHoverToolbar
    };
  };
}
