// @ali/4ever-factory/dist-node/cardFactory/mobileCardFactory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cardFactory = cardFactory;
Object.defineProperty(exports, "removeCard", {
  enumerable: true,
  get: function () {
    return _actions.removeCard;
  }
});
Object.defineProperty(exports, "setCardData", {
  enumerable: true,
  get: function () {
    return _actions.setCardData;
  }
});
Object.defineProperty(exports, "setCardInjection", {
  enumerable: true,
  get: function () {
    return _actions.setCardInjection;
  }
});
Object.defineProperty(exports, "insertCard", {
  enumerable: true,
  get: function () {
    return _actions.insertCard;
  }
});

var _createRenderNode = _interopRequireDefault(require("./createRenderNode"));

var _createOnKeyDown = _interopRequireDefault(require("./handlers/createOnKeyDown"));

var _createOnAction = _interopRequireDefault(require("./handlers/createOnAction"));

var _actions = require("./actions");

var _commands = require("./commands");

var _createUnknownCardMobilePlugin = require("./createUnknownCardMobilePlugin");

function cardFactory(createPlugin) {
  return function createCardPlugin(bizConfig) {
    const myPlugin = createPlugin(bizConfig);
    const {
      menubar,
      toolbar,
      onAction,
      onCangjiePaste,
      queries,
      onCangjieCopy
    } = myPlugin;
    return {
      cardType: myPlugin.cardType,
      renderNode: (0, _createRenderNode.default)(myPlugin, bizConfig, _createUnknownCardMobilePlugin.Placeholder),
      onKeyDown: (0, _createOnKeyDown.default)(),
      onAction: (0, _createOnAction.default)(onAction),
      commands: {
        insertCard: _commands.insertCardCommand,
        removeCard: _commands.removeCardCommand,
        setCardData: _commands.setCardDataCommand,
        setCardInjection: _commands.setCardInjectionCommand
      },
      queries: { ...queries,

        lazyRenderStrategy(_, node) {
          return node.type === 'card' && typeof node.data.height === 'number' ? {
            size: {
              height: node.data.height
            }
          } : undefined;
        }

      },
      models: {
        klass: 'block',
        type: 'card'
      }
    };
  };
}
