// @ali/4ever-plugin-history/dist-web/bi/components/undoButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { UndoButton as Button } from '@ali/we-toolbar';
import { environment } from '@ali/4ever-cangjie';
import { HistoryRole } from "../../utils/role";
var SHORTCUTKEY = environment.IS_MAC ? '⌘+Z' : 'Ctrl+Z';
export default function UndoButton(props) {
  var controller = props.controller,
      locale = props.locale,
      rest = _objectWithoutPropertiesLoose(props, ["controller", "locale"]);

  var isDisabled = !controller.query('canUndo');
  var handleClick = React.useCallback(function () {
    return controller.dispatch('undo');
  }, [controller]);
  return /*#__PURE__*/_createElement(Button, _extends({}, rest, {
    testid: "bi-toolbar-undo",
    disabled: isDisabled,
    onClick: handleClick,
    shortcutText: SHORTCUTKEY,
    role: HistoryRole.undo
  }));
}
