// @ali/4ever-plugin-node-select/dist-node/common/actions/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.actions = exports.isCancelSelectNode = exports.cancelSelectNode = exports.isDeleteSelectedNodes = exports.deleteSelectedNodes = exports.isCutSelectedNodes = exports.cutSelectedNodes = exports.isCopySelectedNodes = exports.copySelectedNodes = exports.DELETE_SELECTED_NODES = exports.CUT_SELECTED_NODES = exports.COPY_SELECTED_NODES = exports.CANCEL_SELECT_NODE = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

const CANCEL_SELECT_NODE = Symbol('cancel_select_node');
exports.CANCEL_SELECT_NODE = CANCEL_SELECT_NODE;
const COPY_SELECTED_NODES = Symbol('copy_SELECTED_NODES');
exports.COPY_SELECTED_NODES = COPY_SELECTED_NODES;
const CUT_SELECTED_NODES = Symbol('cut_SELECTED_NODES');
exports.CUT_SELECTED_NODES = CUT_SELECTED_NODES;
const DELETE_SELECTED_NODES = Symbol('delete_SELECTED_NODES');
exports.DELETE_SELECTED_NODES = DELETE_SELECTED_NODES;
const [copySelectedNodes, isCopySelectedNodes] = (0, _everCangjie.createActionHandlers)(COPY_SELECTED_NODES);
exports.isCopySelectedNodes = isCopySelectedNodes;
exports.copySelectedNodes = copySelectedNodes;
const [cutSelectedNodes, isCutSelectedNodes] = (0, _everCangjie.createActionHandlers)(CUT_SELECTED_NODES);
exports.isCutSelectedNodes = isCutSelectedNodes;
exports.cutSelectedNodes = cutSelectedNodes;
const [deleteSelectedNodes, isDeleteSelectedNodes] = (0, _everCangjie.createActionHandlers)(DELETE_SELECTED_NODES);
exports.isDeleteSelectedNodes = isDeleteSelectedNodes;
exports.deleteSelectedNodes = deleteSelectedNodes;
const [cancelSelectNode, isCancelSelectNode] = (0, _everCangjie.createActionHandlers)(CANCEL_SELECT_NODE);
exports.isCancelSelectNode = isCancelSelectNode;
exports.cancelSelectNode = cancelSelectNode;
const actions = {
  copySelectedNodes,
  cutSelectedNodes,
  deleteSelectedNodes,
  cancelSelectNode
};
exports.actions = actions;
