// @ali/4ever-plugin-mention/dist-node/bi/quickInsert/createQuickInsertConfig.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createQuickInsertConfig;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _everPluginQuickinsert = require("@ali/4ever-plugin-quickinsert");

var _suggestionForQuickInsert = _interopRequireDefault(require("../components/suggestionForQuickInsert"));

const _createElement = /*#__PURE__*/React.createElement;

var _ref = /*#__PURE__*/_createElement(_weDesign.MentionBetaNormal, null);

function createQuickInsertConfig( // @ts-ignore
mentionConfig) {
  const quickInsert = {
    onMenuClick: controller => controller.run('onAction', (0, _everPluginQuickinsert.insertQuickInsert)('quickInsertMention')),
    renderIcon: () => _ref,
    renderPanel: props => {
      const {
        text,
        controller,
        ...rest
      } = props;
      return /*#__PURE__*/_createElement(_suggestionForQuickInsert.default, (0, _extends2.default)({}, rest, mentionConfig, {
        controller: controller,
        text: text.slice(1)
      }));
    }
  };
  return quickInsert;
}
