// @ali/4ever-plugin-mention/dist-node/bi/quickInsert/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = creatMentionPlugin;

var _renderNode = _interopRequireDefault(require("../renderNode"));

var _onKeyDown = _interopRequireDefault(require("../handlers/onKeyDown"));

var _createOnAction = _interopRequireDefault(require("../handlers/createOnAction"));

var _onSelect = _interopRequireDefault(require("../handlers/onSelect"));

var _everPluginQuickinsert = require("@ali/4ever-plugin-quickinsert");

var _insertMention = _interopRequireDefault(require("../commands/insertMention"));

var _createQuickInsertConfig = _interopRequireDefault(require("./createQuickInsertConfig"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _isSelectionInMentionSuggestion = _interopRequireDefault(require("../queries/isSelectionInMentionSuggestion"));

var _onCangjiePaste = _interopRequireDefault(require("../handlers/onCangjiePaste"));

function creatMentionPlugin(config) {
  const mentionLocale = config.mention?.locale || {};
  const mentionConfig = {
    locale: mentionLocale,
    ...config.quickInsertMention
  };
  const mentionId = mentionConfig && mentionConfig.getMentionId ? mentionConfig.getMentionId() : '';
  const onAction = (0, _createOnAction.default)(mentionConfig);
  return {
    renderNode: (0, _renderNode.default)(mentionId, mentionConfig),
    onCangjiePaste: (0, _onCangjiePaste.default)(config),
    onAction: (action, controller, next) => {
      const {
        type,
        payload
      } = action;

      if (type === _everPluginQuickinsert.QUICK_INSERT_SET && payload?.plugin === 'quickInsertMention') {
        return controller.command(_insertMention.default);
      }

      return onAction(action, controller, next);
    },
    onKeyDown: _onKeyDown.default,
    onSelect: _onSelect.default,
    quickInsert: (0, _createQuickInsertConfig.default)(mentionConfig),
    menubar: _menubar.default,
    queries: {
      isSelectionInMentionSuggestion: _isSelectionInMentionSuggestion.default
    },
    models: []
  };
}
