// @ali/4ever-bi/dist-node/plugins/lineheight/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _getLineHeight = _interopRequireDefault(require("./queries/getLineHeight"));

var actions = _interopRequireWildcard(require("./actions"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _createRenderToolbar = _interopRequireDefault(require("./createRenderToolbar"));

const queries = {
  getLineHeight: _getLineHeight.default
};

function createLineHeightPlugin(config) {
  const lineHeightConfig = config.lineheight || {};
  const sideBarTriggerConfig = config.sideBarTrigger || {};
  return {
    toolbar: (0, _toolbar.default)(lineHeightConfig, sideBarTriggerConfig),
    renderToolbar: (0, _createRenderToolbar.default)(lineHeightConfig.locale, sideBarTriggerConfig),
    menubar: (0, _menubar.default)(),
    queries,
    onAction: _onAction.default,
    models: []
  };
}

createLineHeightPlugin.actions = actions;
createLineHeightPlugin.queries = queries;
var _default = createLineHeightPlugin;
exports.default = _default;
