// @ali/we-design-next/dist-web/basic/badge/styled.js
import { cloneElement } from 'react';
import styled from 'styled-components';
import { COLORS } from "../../common/consts/colors";
var RED_DOT_CSS = "\n  display: inline-block;\n  box-sizing: border-box;\n  width: 10px;\n  height: 10px;\n  border-radius: 5px;\n  background-color: #FF5219;\n  border: 2px solid white;\n  vertical-align: top;\n  top: 0px;\n  right: 0px;\n";
export var Dot = /*#__PURE__*/styled.span(["display:inline-block;", ""], RED_DOT_CSS);
var ClassWrapper = function ClassWrapper(_ref) {
  var children = _ref.children,
    className = _ref.className;
  return className ? /*#__PURE__*/cloneElement(children, {
    className: className + " " + (children.props.className || '')
  }) : children;
};
export var DotWrapper = /*#__PURE__*/styled(ClassWrapper)(["&::after{content:'';", "}"], RED_DOT_CSS);
var BADGE_TEXT_CSS = "\n  display: inline-block;\n  font-size: 12px;\n  line-height: 12px;\n  text-align: center;\n  height: 12px;\n  border-radius: 2px;\n  margin-left: 8px;\n  padding: 0 2px;\n  span {\n    display: inline-block;\n    transform: scale(0.83);\n  }\n";
export var New = /*#__PURE__*/styled.div(["color:rgb(255,146,0);background-color:rgba(255,146,0,0.12);", ""], BADGE_TEXT_CSS);
export var Beta = /*#__PURE__*/styled.div(["color:", ";background-color:rgba(56,155,255,0.12);", ""], COLORS.COMMON.BLUE1_HOVER_COLOR, BADGE_TEXT_CSS);
