// @ali/4ever-component/dist-node/Sidebar/inputNumber.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CustomInputNumberItem = exports.InputNumberItem = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _weDesign = require("@ali/we-design");

var _lodash = require("lodash");

const _createElement = /*#__PURE__*/_react.default.createElement;

const Wrapper = /*#__PURE__*/_styledComponents.default.div(["display:grid;grid-template-columns:1fr 28px;justify-content:space-between;"]);

const Title = /*#__PURE__*/_styledComponents.default.div(["font-size:12px;margin-bottom:4px;"]);

const InputWrapper = /*#__PURE__*/_styledComponents.default.div(["grid-column-start:1;grid-column-end:2;.image-sidebar-common-input{flex:1;}"]);

const InputUnit = /*#__PURE__*/_styledComponents.default.span(["display:inline-block;width:28px;line-height:28px;margin-left:4px;"]);

const InputNumberItem = props => {
  const {
    title,
    unit,
    onChange,
    value,
    max,
    min,
    step = 1,
    style,
    name,
    disabled,
    precision = 2,
    testid,
    role
  } = props;

  const ref = _react.default.useRef();

  const [displayValue, setDisplayValue] = _react.default.useState(value);

  const handleChange = _react.default.useCallback(change => {
    setDisplayValue(change);
  }, []);

  const handleBlur = _react.default.useCallback(() => {
    if (!(0, _lodash.isNumber)(displayValue) || isNaN(displayValue) || (0, _lodash.isNumber)(max) && displayValue > max || (0, _lodash.isNumber)(min) && displayValue < min) {
      setDisplayValue(value);
      return;
    }

    onChange(displayValue);
  }, [displayValue, onChange, value, max, min]);

  const handlePressEnter = _react.default.useCallback(() => {
    ref?.current?.blur();
  }, []);

  const handleStep = _react.default.useCallback((value, info) => {
    onChange(value);
  }, [onChange]);

  _react.default.useEffect(() => {
    setDisplayValue(value);
  }, [value]);

  return /*#__PURE__*/_createElement(Wrapper, {
    style: style
  }, /*#__PURE__*/_createElement(Title, null, title), /*#__PURE__*/_createElement(InputWrapper, {
    "data-role": role
  }, /*#__PURE__*/_createElement(_weDesign.InputNumber, {
    precision: precision,
    onChange: handleChange,
    onBlur: handleBlur,
    onPressEnter: handlePressEnter,
    onStep: handleStep,
    value: Number.isNaN(displayValue) ? undefined : displayValue,
    max: max,
    min: min,
    step: step,
    className: "image-sidebar-common-input",
    ref: ref,
    name: name,
    disabled: disabled,
    testid: testid
  })), unit && /*#__PURE__*/_createElement(InputUnit, null, unit));
};

exports.InputNumberItem = InputNumberItem;

const CustomInputNumberItem = props => {
  const {
    title,
    content,
    style
  } = props;
  return /*#__PURE__*/_createElement(Wrapper, {
    style: style
  }, /*#__PURE__*/_createElement(Title, null, title), /*#__PURE__*/_createElement(InputWrapper, null, content));
};

exports.CustomInputNumberItem = CustomInputNumberItem;
