// @ali/4ever-component/dist-node/Sidebar/select.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SelectItemSubTitle = exports.SelectItemIcon = exports.SelectItemTitle = exports.SelectItem = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _classnames = _interopRequireDefault(require("classnames"));

var _utils = require("./utils");

const _createElement = /*#__PURE__*/_react.default.createElement;

const SelectItems = /*#__PURE__*/_styledComponents.default.div(["display:flex;flex-direction:row;margin:0 4px;color:#171a1d;"]);

const SelectItem = /*#__PURE__*/_styledComponents.default.div(["display:inline-block;min-width:80px;background-color:#f9fafc;text-align:center;border:1px solid rgba(126,134,142,0.16);border-radius:4px;margin:0 4px;cursor:default;overflow:hidden;user-select:none;&.selected{box-shadow:0px 0px 0px 2px rgba(0,106,254,0.12);background-color:#ffffff;border:1px solid #0089ff;}&:hover .title{background-color:rgba(126,134,142,0.12);}&.disabled{color:rgba(17,31,44,0.24) !important;cursor:not-allowed;.title{background-color:rgba(126,134,142,0.12);}}"]);

exports.SelectItem = SelectItem;

const SelectItemTitle = /*#__PURE__*/_styledComponents.default.div(["font-size:12px;padding:0px 6px;white-space:nowrap;height:36px;display:flex;align-items:center;justify-content:left;"]);

exports.SelectItemTitle = SelectItemTitle;

const SelectItemIcon = /*#__PURE__*/_styledComponents.default.span(["font-size:0;margin-right:4px;.weicon{font-size:20px;}"]);

exports.SelectItemIcon = SelectItemIcon;

const SelectItemSubTitle = /*#__PURE__*/_styledComponents.default.div(["font-size:12px;color:rgba(17,31,44,0.56);background-color:#f6f6f7;height:16px;line-height:16px;border-radius:0px 0px 3px 3px;.subtitle{display:inline-block;transform:scale(0.9);}"]);

exports.SelectItemSubTitle = SelectItemSubTitle;

const TitleText = /*#__PURE__*/_styledComponents.default.span(["overflow:hidden;text-overflow:ellipsis;line-height:1;"]);

const StylesSelect = props => {
  const {
    items,
    onChange,
    value,
    defaultValue,
    disabled
  } = props;
  const [val, setVal] = (0, _react.useState)(value || defaultValue || '');

  const handleChange = v => {
    if (v !== val) {
      setVal(v);
      onChange(v);
    }
  };

  (0, _react.useEffect)(() => {
    setVal(value || defaultValue || '');
  }, [value, defaultValue]);

  const selectItemStyle = _react.default.useMemo(() => ({
    width: `${1 / items.length * 100}%`
  }), [items.length]);

  const content = items.map(item => {
    return /*#__PURE__*/_createElement(SelectItem, {
      className: (0, _classnames.default)({
        selected: item.key === val,
        disabled: disabled || false
      }),
      key: item.key,
      onClick: () => handleChange(item.key),
      onMouseDown: _utils.blockEvent,
      style: selectItemStyle
    }, /*#__PURE__*/_createElement(SelectItemTitle, {
      className: "title",
      style: item.titleStyle
    }, item.icon && /*#__PURE__*/_createElement(SelectItemIcon, null, item.icon), /*#__PURE__*/_createElement(TitleText, null, item.title)), item.subtitle && /*#__PURE__*/_createElement(SelectItemSubTitle, null, /*#__PURE__*/_createElement("span", {
      className: "subtitle"
    }, item.subtitle)));
  });
  return /*#__PURE__*/_createElement(SelectItems, null, content);
};

var _default = StylesSelect;
exports.default = _default;
