// @ali/4ever-component/dist-node/Sidebar/SideBar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _everCangjie = require("@ali/4ever-cangjie");

var _PluginsContext = require("../PluginsContext");

var _LazyRenderer = require("./LazyRenderer");

const _createElement = /*#__PURE__*/React.createElement;

class Sidebar extends React.Component {
  constructor(...args) {
    super(...args);

    this.renderSidebar = () => {
      const {
        type,
        plugins,
        controller
      } = this.props;
      const [pluginName, toolName = 'sidebar'] = type.split('.');
      const plugin = plugins.hash[pluginName];

      if (!plugin || !plugin.sidebar) {
        return null;
      } // lazy render


      if (typeof plugin.sidebar === 'function') {
        return /*#__PURE__*/_createElement(_LazyRenderer.LazyRenderer, {
          loader: {
            fn: () => plugin.sidebar()
          },
          render: sidebar => {
            const Component = sidebar[toolName];

            if (!Component) {
              return null;
            }

            return /*#__PURE__*/_createElement(Component, {
              controller: controller,
              locale: sidebar.locale || {},
              config: sidebar.config || {}
            });
          }
        });
      } else {
        const s = plugin.sidebar;
        const Component = s[toolName];

        if (!Component) {
          return null;
        }

        return /*#__PURE__*/_createElement(Component, {
          controller: controller,
          locale: s.locale || {},
          config: s.config || {}
        });
      }
    };
  }

  render() {
    return /*#__PURE__*/_createElement(React.Fragment, null, this.renderSidebar());
  }

}

var _default = _everCangjie.Controller.connectController((0, _PluginsContext.connectPlugins)(Sidebar));

exports.default = _default;
