// @ali/4ever-component/dist-node/Sidebar/SidebarItems.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SplitLine = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _everCangjie = require("@ali/4ever-cangjie");

var _PluginsContext = require("../PluginsContext");

var _Collapse = _interopRequireDefault(require("./Collapse"));

var _LazyRenderer = require("./LazyRenderer");

const _createElement = /*#__PURE__*/_react.default.createElement;

const SplitLine = /*#__PURE__*/_styledComponents.default.div(["padding-bottom:12px;border-top:1px solid rgba(126,134,142,0.16);"]);

exports.SplitLine = SplitLine;

var _ref = /*#__PURE__*/_createElement(SplitLine, null);

const SidebarItems = props => {
  const {
    layout = [],
    plugins,
    controller
  } = props;

  const sortedLayout = _react.default.useMemo(() => layout.sort((a, b) => (a.index || 0) - (b.index || 0)), [layout]);

  const renderItem = _react.default.useCallback(item => {
    const [pluginName, itemName = 'default'] = item.split('.');
    const plugin = plugins.hash[pluginName];

    if (!plugin || !plugin.sidebar) {
      return null;
    } // lazy render


    if (typeof plugin.sidebar === 'function') {
      return /*#__PURE__*/_createElement(_LazyRenderer.LazyRenderer, {
        loader: {
          fn: () => plugin.sidebar()
        },
        render: sidebar => {
          const Component = sidebar.items?.[itemName];

          if (!Component) {
            return null;
          }

          return /*#__PURE__*/_createElement(Component, {
            controller: controller,
            locale: sidebar.locale || {},
            config: sidebar.config || {}
          });
        }
      });
    } else {
      // TODO：待所有插件的 sidebar 迁移为异步后删除
      const s = plugin.sidebar;
      const Component = s.items?.[itemName];

      if (!Component) {
        return null;
      }

      return /*#__PURE__*/_createElement(Component, {
        controller: controller,
        locale: s.locale || {},
        config: s.config || {}
      });
    }
  }, [plugins, controller]);

  const renderGroup = _react.default.useCallback((group, isLast) => {
    const {
      renderTitle,
      items,
      defaultFold,
      useCollapse,
      key
    } = group;
    return /*#__PURE__*/_createElement(_react.default.Fragment, {
      key: key
    }, /*#__PURE__*/_createElement(_Collapse.default, {
      renderTitle: renderTitle,
      defaultFold: defaultFold,
      useCollapse: useCollapse
    }, items.map(item => renderItem(item))), !isLast && _ref);
  }, [renderItem]);

  const lastIndex = sortedLayout.length - 1;
  return /*#__PURE__*/_createElement(_react.default.Fragment, null, sortedLayout.map((group, index) => renderGroup(group, index === lastIndex)));
};

var _default = _everCangjie.Controller.connectController((0, _PluginsContext.connectPlugins)(SidebarItems));

exports.default = _default;
