// @ali/4ever-plugin-code/dist-node/bi/createRenderBlock.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderBlock;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _CodeEditor = _interopRequireDefault(require("./components/CodeEditor"));

var _themes = require("../utils/constants/themes");

const _createElement = /*#__PURE__*/React.createElement;

function createRenderBlock(config) {
  return {
    code: (props, controller) => {
      const {
        attributes,
        isSelected
      } = props;
      const node = props.node;
      const {
        cursorPos = 'end'
      } = node.data;
      return /*#__PURE__*/_createElement(_themes.CodeWrapper, (0, _extends2.default)({}, attributes, {
        "data-testid": "code-editor",
        "data-type": "code"
      }), /*#__PURE__*/_createElement(_CodeEditor.default, {
        cursorPos: cursorPos,
        onSave: config.onSave,
        isSelected: isSelected,
        controller: controller,
        node: node,
        locale: config.locale
      }));
    }
  };
}
