// @ali/4ever-plugin-code/dist-node/bi/createToolbar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createToolbar;

var _ToolbarButton = _interopRequireDefault(require("./components/ToolbarButton"));

var _ToolbarLanguageSelect = _interopRequireDefault(require("./components/ToolbarLanguageSelect"));

var _ToolbarThemeSelect = _interopRequireDefault(require("./components/ToolbarThemeSelect"));

var _ToolbarLineWrap = _interopRequireDefault(require("./components/ToolbarLineWrap"));

var _ToolbarMore = _interopRequireDefault(require("./components/ToolbarMore"));

function createToolbar(config) {
  return {
    buttons: {
      default: _ToolbarButton.default,
      button: _ToolbarButton.default,
      syntax: _ToolbarLanguageSelect.default,
      theme: _ToolbarThemeSelect.default,
      wrap: _ToolbarLineWrap.default,
      more: _ToolbarMore.default
    },
    locale: config.locale,
    getTarget: controller => {
      const {
        selection,
        document
      } = controller.value;

      if (selection && selection.isCollapsed) {
        const block = document.getClosestBlock(selection.focus.key);

        if (block?.type === 'code') {
          return 'code';
        }
      }

      return undefined;
    }
  };
}
