// @ali/4ever-plugin-code/dist-node/bi/getSnippetConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everUtils = require("@ali/4ever-utils");

var _languages = require("../utils/constants/languages");

var _actions = require("./actions");

const getSnippetConfig = (_, next) => {
  const trigger = '```[\\w\+#]*$';
  const list = (0, _languages.createCodeLanguages)();
  const config = {
    trigger,
    triggerSuffix: '^$',
    fetchItems: async () => {
      return list.map(({
        key,
        title,
        alias
      }) => ({
        key,
        title,
        alias,
        name: title
      }));
    },
    filterItems: async (text, items) => {
      const searchText = text.slice(3);

      if (!searchText) {
        return items;
      }

      return (0, _everUtils.fuzzySearch)(items, searchText, {
        key: 'title'
      });
    },
    insertSnippet: (controller, item) => {
      controller.run('onAction', (0, _actions.insertCode)({
        code: '',
        syntax: item.key
      }));
      controller.dispatch('createTriggerMarkdownAction', {
        mdType: 'code',
        info: item.key,
        from: 'snippet'
      });
    }
  };
  return {
    [trigger]: config,
    ...next()
  };
};

var _default = getSnippetConfig;
exports.default = _default;
