// @ali/4ever-plugin-code/dist-node/bi/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.biFactory = void 0;

var _createOnKeyDown = _interopRequireDefault(require("./handlers/createOnKeyDown"));

var _createOnAction = _interopRequireDefault(require("./handlers/createOnAction"));

var _createRenderBlock = _interopRequireDefault(require("./createRenderBlock"));

var _createToolbar = _interopRequireDefault(require("./createToolbar"));

var _shouldPreventEvent = _interopRequireDefault(require("./queries/shouldPreventEvent"));

var _lazyRenderStrategy = _interopRequireDefault(require("../common/queries/lazyRenderStrategy"));

var _insertCode = _interopRequireDefault(require("./commands/insertCode"));

var _removeCode = _interopRequireDefault(require("./commands/removeCode"));

var _updateCode = _interopRequireDefault(require("./commands/updateCode"));

var _quickInsertConfig = _interopRequireDefault(require("./quickInsertConfig"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _createOnCangjieSelect = _interopRequireDefault(require("./handlers/createOnCangjieSelect"));

var _index = _interopRequireDefault(require("../mo/index"));

var _createRenderHoverToolbar = _interopRequireDefault(require("./createRenderHoverToolbar"));

var _getSnippetConfig = _interopRequireDefault(require("./getSnippetConfig"));

function handleEvent(event, controller, next) {
  if ((0, _shouldPreventEvent.default)(event, controller)) {
    return controller;
  }

  return next();
}

const commands = {
  insertCode: (0, _insertCode.default)(),
  removeCode: _removeCode.default,
  updateCode: _updateCode.default
};

const biFactory = configs => {
  const codeConfig = configs.code || {};
  return { ...(0, _index.default)({}),
    commands,
    queries: {
      lazyRenderStrategy: _lazyRenderStrategy.default
    },
    onKeyDown: (0, _createOnKeyDown.default)(codeConfig),
    onAction: (0, _createOnAction.default)(codeConfig),
    onBeforeInput: handleEvent,
    onCompositionStart: handleEvent,
    onCompositionEnd: handleEvent,
    renderNode: (0, _createRenderBlock.default)(codeConfig),
    renderHoverToolbar: (0, _createRenderHoverToolbar.default)(codeConfig),
    locale: codeConfig.locale,
    toolbar: (0, _createToolbar.default)(codeConfig),
    menubar: _menubar.default,
    quickInsert: _quickInsertConfig.default,
    onCangjieSelect: (0, _createOnCangjieSelect.default)(),
    getSnippetConfig: _getSnippetConfig.default,
    models: {
      klass: 'block',
      type: 'code'
    }
  };
};

exports.biFactory = biFactory;
