// @ali/4ever-plugin-code/dist-node/bi/mobile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mobileBiFactory = void 0;

var _lazyRenderStrategy = _interopRequireDefault(require("../common/queries/lazyRenderStrategy"));

var _createOnAction = _interopRequireDefault(require("./handlers/createOnAction"));

var _createRenderBlock = _interopRequireDefault(require("./mobile/createRenderBlock"));

var _createOnKeyDown = _interopRequireDefault(require("./handlers/createOnKeyDown"));

var _index = _interopRequireDefault(require("../mo/index"));

const mobileBiFactory = configs => {
  const codeConfig = configs.code || {};
  return { ...(0, _index.default)({}),
    queries: {
      lazyRenderStrategy: _lazyRenderStrategy.default
    },
    renderNode: (0, _createRenderBlock.default)(codeConfig),
    onKeyDown: (0, _createOnKeyDown.default)(codeConfig, false),
    onAction: (0, _createOnAction.default)({}),
    locale: codeConfig.locale,
    models: {
      klass: 'block',
      type: 'code'
    }
  };
};

exports.mobileBiFactory = mobileBiFactory;
