// @ali/we-design-next/dist-node/form/input/Input.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Input = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _useDefaultProp = require("../../common/react/useDefaultProp");
var _icon = require("../../basic/icon");
var _styled = require("./styled");
var _consts = require("../../common/consts");
var _common = require("../../common/");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const _createElement = /*#__PURE__*/_react.default.createElement;
function blockEvent(e) {
  e.stopPropagation();
  e.preventDefault();
}
const Input = /*#__PURE__*/(0, _react.memo)( /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    value: _value,
    suffix,
    prefix,
    allowClear,
    disabled,
    size = 'normal',
    bordered = true,
    onEnter,
    onKeyDown,
    defaultValue = '',
    onChange,
    error,
    children,
    style,
    className,
    id,
    ['data-testid']: testid,
    ['data-role']: role,
    ...rest
  } = props;
  const [value, setValue] = (0, _useDefaultProp.useDefaultProp)(defaultValue, _value, onChange, e => e.target.value);
  const handlerClear = (0, _react.useCallback)(() => {
    setValue({
      target: {
        value: ''
      }
    });
  }, [setValue]);
  const handleKeyDown = (0, _react.useCallback)(e => {
    if (onKeyDown) {
      onKeyDown(e);
    }
    if (e.which === _consts.KEY_CODES.ENTER && onEnter) {
      onEnter(e);
    }
  }, [onKeyDown, onEnter]);
  return /*#__PURE__*/_createElement(_react.default.Fragment, null, /*#__PURE__*/_createElement(_styled.InputWrap, (0, _extends2.default)({
    size: size,
    bordered: bordered,
    disabled: !!disabled,
    error: !!error,
    tabIndex: -1 /** 目的：使得在对前后缀图标点击时，在 mouse-down 的瞬间即进入 focus 状态（而不是 mouse-up 之后）*/
  }, (0, _common.mergeRestProps)({
    className,
    style,
    id,
    ['data-testid']: testid,
    ['data-role']: role
  }, 'wdn-input')), prefix ? /*#__PURE__*/_react.default.cloneElement(prefix, {
    style: {
      cursor: 'default'
    }
  }) : null, /*#__PURE__*/_createElement(_styled.InputEle, (0, _extends2.default)({
    ref: ref,
    value: value,
    onChange: setValue,
    disabled: !!disabled,
    onKeyDown: handleKeyDown
  }, rest)), allowClear && !!value ? /*#__PURE__*/_createElement(_icon.IconCloseFill, {
    className: "icon-close",
    onClick: handlerClear,
    onMouseDown: blockEvent
  }) : null, !!suffix && !allowClear ? /*#__PURE__*/_react.default.cloneElement(suffix, {
    style: {
      cursor: 'default'
    }
  }) : null), error ? /*#__PURE__*/_createElement(_styled.InputErrorMessage, null, error) : null);
}));
exports.Input = Input;
