// @ali/4ever-component/dist-node/MobileToolbar/components/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Divider = exports.SubLineWrapper = exports.KeyboardShadow = exports.KeyboardWrapper = exports.ScrollableWrapper = exports.LineWrapper = exports.BottomSafeArea = exports.ToolbarWrapper = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constant = require("../constant");

const ToolbarWrapper = /*#__PURE__*/_styledComponents.default.div(["--sab:env(safe-area-inset-bottom);z-index:", ";position:fixed;transform:", ";left:0;right:0;bottom:0;background:white;display:flex;flex-direction:column;align-items:flex-start;-webkit-tap-highlight-color:rgba(0,0,0,0);"], _constant.Z_INDEX, _constant.THREED_TRAMSFORM);

exports.ToolbarWrapper = ToolbarWrapper;

const BottomSafeArea = /*#__PURE__*/_styledComponents.default.div(["width:100%;background:white;height:0;height:constant(safe-area-inset-bottom);height:env(safe-area-inset-bottom);"]);

exports.BottomSafeArea = BottomSafeArea;

const LineWrapper = /*#__PURE__*/_styledComponents.default.div(["position:absolute;left:0;top:0;height:", "px;padding:0 4px;width:100%;display:flex;flex-direction:row;align-items:center;transition:opacity 100ms;justify-content:space-between;box-shadow:rgb(0 0 0 / 12%) 0px -16px 16px -16px,rgb(0 0 0 / 32%) 0px -1px 1px -1px;"], _constant.BOTTOM_BAR_SIZE);

exports.LineWrapper = LineWrapper;

const ScrollableWrapper = /*#__PURE__*/_styledComponents.default.div(["height:", "px;overflow-x:scroll;display:flex;flex-direction:row;align-items:center;padding-right:8px;-ms-overflow-style:none;scrollbar-width:none;&::-webkit-scrollbar{display:none;}"], _constant.BOTTOM_BAR_SIZE);

exports.ScrollableWrapper = ScrollableWrapper;

const KeyboardWrapper = /*#__PURE__*/_styledComponents.default.div(["background:var(--common_fg_color,white);padding-left:4px;height:100%;position:relative;"]);

exports.KeyboardWrapper = KeyboardWrapper;

const KeyboardShadow = /*#__PURE__*/_styledComponents.default.div(["position:absolute;left:-8px;top:0;width:8px;height:", "px;background:linear-gradient( 270deg,var(--common_bg_color,#e6e6e8) 0%,rgba(255,255,255,0) 100% );"], _constant.BOTTOM_BAR_SIZE);

exports.KeyboardShadow = KeyboardShadow;

const SubLineWrapper = /*#__PURE__*/_styledComponents.default.div(["position:absolute;left:24px;max-width:calc(100% - ", "px);overflow-x:scroll;overflow-y:hidden;bottom:100%;margin-bottom:", "px;border-radius:8px;box-shadow:0px 0px 1px rgba(0,0,0,0.24),0px 12px 24px rgba(126,134,142,0.18);background:white;display:flex;flex-direction:row;align-items:center;&::-webkit-scrollbar{display:none;}"], _constant.SUBLINE_MIN_LEFT * 2, _constant.SUBLINE_MARGIN_BOTTOM);

exports.SubLineWrapper = SubLineWrapper;

const Divider = /*#__PURE__*/_styledComponents.default.div(["min-width:1px;width:1px;height:16px;margin:0 4px;opacity:0.6;background:rgba(126,134,142,0.24);"]);

exports.Divider = Divider;
