// @ali/4ever-component/dist-node/SelectionBar/reducer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mountSelectionToolbar = mountSelectionToolbar;
exports.default = exports.UPDATE_CAN_HIDE = exports.UPDATE_LAST_FILL = exports.UPDATE_LAST_HIGHLIGHT = exports.UPDATE_LAST_COLOR = exports.MOUNT_SELECTION_TOOLBAR = void 0;
const MOUNT_SELECTION_TOOLBAR = Symbol('MOUNT_SELECTION_TOOLBAR');
exports.MOUNT_SELECTION_TOOLBAR = MOUNT_SELECTION_TOOLBAR;
const UPDATE_LAST_COLOR = Symbol('UPDATE_LAST_COLOR');
exports.UPDATE_LAST_COLOR = UPDATE_LAST_COLOR;
const UPDATE_LAST_HIGHLIGHT = Symbol('UPDATE_LAST_HIGHLIGHT');
exports.UPDATE_LAST_HIGHLIGHT = UPDATE_LAST_HIGHLIGHT;
const UPDATE_LAST_FILL = Symbol('UPDATE_LAST_FILL');
exports.UPDATE_LAST_FILL = UPDATE_LAST_FILL;
const UPDATE_CAN_HIDE = Symbol('UPDATE_CAN_HIDE');
exports.UPDATE_CAN_HIDE = UPDATE_CAN_HIDE;

const selectionBarReducer = (state, action) => {
  if (action.type === UPDATE_LAST_COLOR) {
    return { ...state,
      lastColor: action.payload
    };
  }

  if (action.type === UPDATE_LAST_HIGHLIGHT) {
    return { ...state,
      lastHighlight: action.payload
    };
  }

  if (action.type === UPDATE_LAST_FILL) {
    return { ...state,
      lastFill: action.payload
    };
  }

  if (action.type === UPDATE_CAN_HIDE) {
    return { ...state,
      canHide: action.payload
    };
  }

  return state;
};

function mountSelectionToolbar() {
  return {
    type: MOUNT_SELECTION_TOOLBAR
  };
}

var _default = selectionBarReducer;
exports.default = _default;
