// @ali/4ever-bi/dist-web/plugins/link/mobile/createRenderNode.js
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { SimpleLazyLoad } from '@ali/4ever-component';
import Link from "./link";
import { isLinkCardNode } from "../utils";
export default function createRenderNode(linkConfig) {
  var locale = linkConfig.locale,
      protocolAllowList = linkConfig.protocolAllowList,
      openLink = linkConfig.openLink;
  var LazyLoad = linkConfig.lazyload || SimpleLazyLoad;
  var URLLinkCard = LazyLoad(function () {
    return import('@ali/we-biz-link-card').then(function (_ref) {
      var URLLinkCard = _ref.URLLinkCard;
      return URLLinkCard;
    });
  });
  return {
    link: function link(props) {
      var node = props.node;
      var linkType = 'link'; // 当 link 是卡片形式

      if (isLinkCardNode(node)) {
        var cardInfo = node.data.cardInfo;

        var _ref2 = cardInfo || {},
            displayType = _ref2.displayType;

        if (displayType === 'card') {
          linkType = 'card';
        }
      }

      return (
        /*#__PURE__*/
        // @ts-ignore
        _createElement(Link, _extends({}, props, {
          protocolAllowList: protocolAllowList || [],
          locale: locale,
          onOpenLink: openLink,
          URLLinkCard: URLLinkCard,
          displayType: linkType
        }))
      );
    }
  };
}
