// @ali/4ever-bi/dist-web/plugins/link/mobile/index.js
import _extends from "@babel/runtime/helpers/extends";
import { LinkPlugin } from '@ali/4ever-bamboo';
import createRenderNode from "./createRenderNode";
import createRenderMark from "./createRenderMark";
import createContextMenu from "./contextMenu";
import createCopyLink from "../commands/createCopyLink";
import createOpenLink from "../commands/createOpenLink";
import * as actions from "../actions";
import createOnCangjiePaste from "../handlers/createOnCangjiePaste";
import insertLink from "./insertLink";
import locale from "../../../locales/zh_CN";
import onAction from "../handlers/onAction";
import { queries } from "./queries";
import { moFactory } from '@ali/4ever-plugin-link';
var commands = {
  insertLink: insertLink
};

function createLinkPlugin(config) {
  var _config$clipboard;

  var linkConfig = _extends({
    protocolAllowList: ['http:', 'https:', 'dingtalk:', 'taobao:'],
    locale: locale.link.locale
  }, config.link || {});

  var serializer = (_config$clipboard = config.clipboard) == null ? void 0 : _config$clipboard.serializer;

  var plugin = _extends({}, moFactory({}), {
    commands: _extends({}, commands, {
      copyLink: createCopyLink(serializer),
      openLink: createOpenLink(linkConfig.protocolAllowList, linkConfig.openLink)
    }),
    renderNode: createRenderNode(linkConfig),
    contextMenu: createContextMenu(linkConfig),
    actions: actions,
    // @ts-ignore
    onAction: onAction,
    onCangjiePaste: createOnCangjiePaste(config, linkConfig),
    queries: queries,
    models: {
      klass: 'inline',
      type: 'link'
    }
  }); // 支持 purelink 识别的情况下，添加 decorateNode 和 renderMark


  if (!linkConfig.disableTextRecognize) {
    // @ts-ignore
    plugin.decorateNode = LinkPlugin.createDecorateNode(linkConfig.protocolAllowList || []);
    plugin.renderMark = createRenderMark(linkConfig);
  }

  return plugin;
}

createLinkPlugin.actions = actions;
createLinkPlugin.commands = commands;
createLinkPlugin.queries = queries;
export default createLinkPlugin;
