// @ali/4ever-bi/dist-node/plugins/hoverToolbar/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_LAYOUTS = exports.SCROLL_TIMEOUT = exports.SAFE_CONTAINER_MARGIN_RIGHT_OR_LEFT = exports.SAFE_CONTAINER_MARGIN_TOP = exports.SAFE_NODE_MARGIN_TOP = exports.HOVER_TOOLBAR_HEIGHT = void 0;
const HOVER_TOOLBAR_HEIGHT = 40; // 自适应安全距离的最大值， 工具栏下边缘与元素上边缘的最大距离

exports.HOVER_TOOLBAR_HEIGHT = HOVER_TOOLBAR_HEIGHT;
const SAFE_NODE_MARGIN_TOP = 8; // 固定安全距离，工具栏上边缘与容器上边缘的距离

exports.SAFE_NODE_MARGIN_TOP = SAFE_NODE_MARGIN_TOP;
const SAFE_CONTAINER_MARGIN_TOP = 8;
exports.SAFE_CONTAINER_MARGIN_TOP = SAFE_CONTAINER_MARGIN_TOP;
const SAFE_CONTAINER_MARGIN_RIGHT_OR_LEFT = 12;
exports.SAFE_CONTAINER_MARGIN_RIGHT_OR_LEFT = SAFE_CONTAINER_MARGIN_RIGHT_OR_LEFT;
const SCROLL_TIMEOUT = 100;
exports.SCROLL_TIMEOUT = SCROLL_TIMEOUT;
const DEFAULT_LAYOUTS = {
  image: {
    normal: [['image.download', 'image.reset', 'image.style'], ['image.crop', 'image.rotate', 'alignment'], ['comment']],
    folded: [['hoverToolbar.cut', 'hoverToolbar.copy', 'hoverToolbar.delete']]
  },
  hr: {
    normal: [['hr.color', 'hr.style', 'hr.weight']]
  },
  code: {
    normal: [['code.theme'], ['code.syntax', 'code.lineWrap'], ['comment']],
    folded: [['hoverToolbar.cut', 'hoverToolbar.copy', 'hoverToolbar.delete']]
  },
  callout: {
    normal: [['callout.style', 'callout.stickerSwitch'], ['comment']],
    folded: [['hoverToolbar.cut', 'hoverToolbar.copy', 'hoverToolbar.delete']]
  },
  columns: {
    normal: [['columns.style'], ['comment']],
    folded: [['hoverToolbar.cut', 'hoverToolbar.copy', 'hoverToolbar.delete']]
  },
  templateButton: {
    normal: [['templateButton.edit'], ['comment']],
    folded: [['hoverToolbar.copy', 'hoverToolbar.cut', 'hoverToolbar.delete']]
  }
};
exports.DEFAULT_LAYOUTS = DEFAULT_LAYOUTS;
