// @ali/4ever-bi/dist-node/plugins/hoverToolbar/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _createRenderEditable = _interopRequireDefault(require("./createRenderEditable"));

var _createOnMouseMove = _interopRequireDefault(require("./createOnMouseMove"));

var _hoverCapture = require("./utils/hoverCapture");

var _createRenderHoverToolbar = _interopRequireDefault(require("./createRenderHoverToolbar"));

var _onAction = _interopRequireDefault(require("./handles/onAction"));

var _onCangjieInput = _interopRequireDefault(require("./handles/onCangjieInput"));

var _createOnContextMenu = _interopRequireDefault(require("./createOnContextMenu"));

var _createOnQuery = _interopRequireDefault(require("./createOnQuery"));

var _constants = require("./constants");

function createHoverToolbarPlugin(allConfigs) {
  const {
    hoverToolbar: config
  } = allConfigs;
  const hoverCapture = new _hoverCapture.HoverCapture();
  const hoverToolbarConfig = { ...config,
    layouts: config.layouts || _constants.DEFAULT_LAYOUTS
  };
  return {
    onMouseMove: (0, _createOnMouseMove.default)(hoverCapture, hoverToolbarConfig),
    onContextMenu: (0, _createOnContextMenu.default)(hoverCapture, hoverToolbarConfig),
    renderEditable: (0, _createRenderEditable.default)(hoverCapture, hoverToolbarConfig),
    renderHoverToolbar: (0, _createRenderHoverToolbar.default)(hoverToolbarConfig),
    onAction: (0, _onAction.default)(),
    onCangjieInput: (0, _onCangjieInput.default)(hoverToolbarConfig, hoverCapture),
    onQuery: (0, _createOnQuery.default)(hoverToolbarConfig)
  };
}

var _default = createHoverToolbarPlugin;
exports.default = _default;
