// @ali/4ever-bi/dist-web/plugins/headingGroupNew/handlers/createRenderGroup.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Heading } from '@ali/4ever-mo';
import { GroupFactory } from '@ali/4ever-bamboo';
var GroupNew = GroupFactory.GroupNew;
var Heading1 = Heading.Heading1,
    Heading2 = Heading.Heading2,
    Heading3 = Heading.Heading3,
    Heading4 = Heading.Heading4,
    Heading5 = Heading.Heading5,
    Heading6 = Heading.Heading6;
var headingList = [Heading1, Heading2, Heading3, Heading4, Heading5, Heading6];

function createRenderGroup(config) {
  return function renderGroup(props, _, next) {
    var node = props.node,
        attributes = props.attributes,
        children = props.children,
        _props$visible = props.visible,
        visible = _props$visible === void 0 ? true : _props$visible,
        controller = props.controller,
        group = props.group,
        isCollapsed = props.isCollapsed;

    if (group.type === 'heading1-group' || group.type === 'heading2-group' || group.type === 'heading3-group') {
      return /*#__PURE__*/_createElement(GroupNew, {
        node: node,
        attributes: attributes,
        controller: controller,
        isSelected: false,
        visible: visible,
        isCollapsed: isCollapsed
      }, children());
    }

    return next();
  };
}

export default createRenderGroup;
