// @ali/4ever-bi/dist-node/plugins/newlineGuide/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "NEWLINE_REMOVE_TRIGGER_TEXT", {
  enumerable: true,
  get: function () {
    return _everUtils.NEWLINE_REMOVE_TRIGGER_TEXT;
  }
});
Object.defineProperty(exports, "NEWLINE_REMOVE_DATA", {
  enumerable: true,
  get: function () {
    return _everUtils.NEWLINE_REMOVE_DATA;
  }
});
Object.defineProperty(exports, "NEWLINE_SET_DATA", {
  enumerable: true,
  get: function () {
    return _everUtils.NEWLINE_SET_DATA;
  }
});
Object.defineProperty(exports, "MOUNT_PLUS_TOOLBAR", {
  enumerable: true,
  get: function () {
    return _everUtils.MOUNT_PLUS_TOOLBAR;
  }
});
Object.defineProperty(exports, "MOUNT_SLASH_TOOLBAR", {
  enumerable: true,
  get: function () {
    return _everUtils.MOUNT_SLASH_TOOLBAR;
  }
});
Object.defineProperty(exports, "removeNewlineTriggerText", {
  enumerable: true,
  get: function () {
    return _everUtils.removeNewlineTriggerText;
  }
});
Object.defineProperty(exports, "removeNewlineData", {
  enumerable: true,
  get: function () {
    return _everUtils.removeNewlineData;
  }
});
Object.defineProperty(exports, "setNewlineData", {
  enumerable: true,
  get: function () {
    return _everUtils.setNewlineData;
  }
});
Object.defineProperty(exports, "mountInsertMenuFromLeftToolbar", {
  enumerable: true,
  get: function () {
    return _everUtils.mountInsertMenuFromLeftToolbar;
  }
});
Object.defineProperty(exports, "mountInsertMenu", {
  enumerable: true,
  get: function () {
    return _everUtils.mountInsertMenu;
  }
});

var _everUtils = require("@ali/4ever-utils");
