// @ali/4ever-bi/dist-node/plugins/placeholder/createRenderNode.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderNode;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everComponent = require("@ali/4ever-component");

var _actions = require("./actions");

const _createElement = /*#__PURE__*/React.createElement;

function createRenderNode(config) {
  const locale = config.locale || {};
  const {
    renderText
  } = config;
  return {
    ph: (props, controller) => {
      const {
        node
      } = props;
      return /*#__PURE__*/_createElement(_everComponent.Placeholder, (0, _extends2.default)({}, props, {
        handleRemoveNode: () => controller.run('onAction', (0, _actions.removePlaceholder)(node)),
        editor: controller,
        renderText: renderText,
        locale: locale
      }));
    },
    iph: (props, controller) => {
      return /*#__PURE__*/_createElement(_everComponent.InlinePlaceholder, (0, _extends2.default)({}, props, {
        editor: controller,
        renderText: renderText
      }));
    }
  };
}
