// @ali/4ever-bi/dist-node/plugins/link/components/AnimatePortal.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _reactTransitionGroup = require("react-transition-group");

var _everComponent = require("@ali/4ever-component");

const _createElement = /*#__PURE__*/React.createElement;
const fadeInTop = /*#__PURE__*/(0, _styledComponents.keyframes)(["0%{transform:translateY(-12px);opacity:0;}100%{transform:translateY(0);opacity:1;}"]);
const fadeInTopAnimation = /*#__PURE__*/(0, _styledComponents.css)(["animation:", " 0.25s cubic-bezier(0.455,0.030,0.515,0.955) both;"], fadeInTop);
const fadeOut = /*#__PURE__*/(0, _styledComponents.keyframes)(["0%{opacity:1;}100%{opacity:0;}"]);
const fadeOutAnimation = /*#__PURE__*/(0, _styledComponents.css)(["width:fit-content;height:fit-content;animation:", " 0.15s ease-out both;"], fadeOut);

const AnimateWrapper = /*#__PURE__*/_styledComponents.default.div(["position:absolute;", " ", ""], fadeInTopAnimation, props => (props.state === 'exiting' || props.state === 'exited') && fadeOutAnimation);

const AnimatePortal = props => {
  const {
    children,
    overlay,
    visible,
    onOverlayHide,
    ...rest
  } = props;

  const animatedOverlay = /*#__PURE__*/_createElement(_reactTransitionGroup.Transition, {
    in: visible,
    timeout: {
      enter: 250,
      exit: 150
    },
    onExited: onOverlayHide,
    unmountOnExit: true
  }, state => /*#__PURE__*/_createElement(AnimateWrapper, {
    state: state
  }, /*#__PURE__*/React.cloneElement(overlay)));

  return /*#__PURE__*/_createElement(_everComponent.Portal, (0, _extends2.default)({}, rest, {
    overlay: animatedOverlay,
    visible: visible,
    destroyOverlayOnHide: false
  }), children);
};

var _default = AnimatePortal;
exports.default = _default;
