// @ali/4ever-bi/dist-node/plugins/link/components/linkMenu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inlineMenu = exports.menu = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../actions");

var _hotKeyTips = require("../utils/hotKeyTips");

const _createElement = /*#__PURE__*/React.createElement;

function withLinkMenu(Component) {
  return function Menu(props) {
    const {
      menuItem,
      controller,
      locale
    } = props;
    const {
      title,
      ...rest
    } = menuItem;

    const handleClick = () => controller.run('onAction', (0, _actions.wrapLink)());

    return /*#__PURE__*/_createElement(Component, (0, _extends2.default)({
      extra: _hotKeyTips.LINK_INSERT_HOT_KEY_TIP
    }, rest, {
      onClick: handleClick,
      role: _everBamboo.PluginRoles.link
    }), title || locale?.buttonTitle);
  };
}

const menu = withLinkMenu(_weDesign.MenuItem);
exports.menu = menu;
const inlineMenu = withLinkMenu(_weDesign.InlineMenuItem);
exports.inlineMenu = inlineMenu;
