// @ali/4ever-bi/dist-node/plugins/link/components/quickInsertPanel.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _everComponent = require("@ali/4ever-component");

var _actions = require("../actions");

const _createElement = /*#__PURE__*/React.createElement;

const placeholder = /*#__PURE__*/_createElement("svg", {
  width: "51px",
  height: "61px",
  viewBox: "0 0 51 61",
  version: "1.1"
}, /*#__PURE__*/_createElement("g", {
  stroke: "none",
  strokeWidth: "1",
  fill: "none",
  fillRule: "evenodd"
}, /*#__PURE__*/_createElement("g", {
  transform: "translate(-210.000000, -197.000000)"
}, /*#__PURE__*/_createElement("g", {
  transform: "translate(200.000000, 192.000000)"
}, /*#__PURE__*/_createElement("g", {
  strokeWidth: "1",
  fillRule: "evenodd",
  fill: "#000000",
  opacity: "0"
}, /*#__PURE__*/_createElement("g", {
  fillRule: "nonzero"
}, /*#__PURE__*/_createElement("rect", {
  x: "0",
  y: "0",
  width: "70",
  height: "70"
}))), /*#__PURE__*/_createElement("g", {
  strokeWidth: "1",
  fillRule: "evenodd",
  transform: "translate(10.000000, 5.000000)"
}, /*#__PURE__*/_createElement("g", {
  fillRule: "nonzero"
}, /*#__PURE__*/_createElement("path", {
  d: "M44.9556558,61 L6.94299161,61 C3.63160212,60.9940583 0.946272158,58.3113849 0.937052812,55.0000029 L0.888584849,6.03256355 C0.884930326,2.71885705 3.56856162,0.0299079726 6.8822685,0.0266280661 C6.88239337,0.0266279425 6.88251825,0.0266278228 6.88264347,0.0270023232 L36,0 L36,0 L50.9556558,13.9877735 L50.9556558,55 C50.9556558,58.3137085 48.2693643,61 44.9556558,61 Z",
  fill: "#F2F4F5"
}), /*#__PURE__*/_createElement("path", {
  d: "M36.0061992,7.98777529 L36,0 L36,0 L51,13.9877735 L42.0108558,13.9877735 C38.6989656,13.9831151 36.0134279,11.2996608 36.0061992,7.98777529 Z",
  fill: "#D9DBDD"
}))), /*#__PURE__*/_createElement("path", {
  d: "M38.5,18 C41.5375661,18 44,20.4624339 44,23.5 L44,30.5 C44,33.5375661 41.5375661,36 38.5,36 L37.5,36 L37.5,39 L38.5,39 C41.5375661,39 44,41.4624339 44,44.5 L44,51.5 C44,54.5375661 41.5375661,57 38.5,57 L33.5,57 C30.4624339,57 28,54.5375661 28,51.5 L28,44.5 C28,41.4624339 30.4624339,39 33.5,39 L34.5,39 L34.5,36 L33.5,36 C30.4624339,36 28,33.5375661 28,30.5 L28,23.5 C28,20.4624339 30.4624339,18 33.5,18 L38.5,18 Z M38.5,42 L37.5,42 L37.5,46.4541562 C37.5,47.2338524 36.9051119,47.8746049 36.14446,47.9472897 L36,47.9541562 C35.1715729,47.9541562 34.5,47.2825834 34.5,46.4541562 L34.5,46.4541562 L34.5,42 L33.5,42 C32.1192881,42 31,43.1192881 31,44.5 L31,51.5 C31,52.8807119 32.1192881,54 33.5,54 L38.5,54 C39.8807119,54 41,52.8807119 41,51.5 L41,44.5 C41,43.1192881 39.8807119,42 38.5,42 Z M38.5,21 L33.5,21 C32.1192881,21 31,22.1192881 31,23.5 L31,30.5 C31,31.8807119 32.1192881,33 33.5,33 L34.5,33 L34.5,29 C34.5,28.2203039 35.0948881,27.5795513 35.85554,27.5068666 L36,27.5 C36.8284271,27.5 37.5,28.1715729 37.5,29 L37.5,29 L37.5,33 L38.5,33 C39.8807119,33 41,31.8807119 41,30.5 L41,23.5 C41,22.1192881 39.8807119,21 38.5,21 Z",
  fill: "#E0E2E5",
  fillRule: "nonzero",
  transform: "translate(36.000000, 37.500000) rotate(45.000000) translate(-36.000000, -37.500000) "
})))));

const QuickInsertPanel = props => {
  const {
    controller,
    locale,
    close,
    text
  } = props;
  const handleInsert = React.useCallback(() => {
    let range;

    if (text && text.trim()) {
      const {
        selection,
        document
      } = controller.value;

      if (selection) {
        range = selection.moveStartBackward(document, text.length);
      }
    }

    close();
    controller.run('onAction', (0, _actions.wrapLink)(range));
  }, [controller, close, text]);
  return /*#__PURE__*/_createElement(_everComponent.QuickInsertCommonPanel, {
    placeholder: placeholder
  }, /*#__PURE__*/_createElement(_everComponent.QuickInsertButton, {
    onClick: handleInsert
  }, locale ? locale.buttonTitle : ''));
};

var _default = QuickInsertPanel;
exports.default = _default;
