// @ali/4ever-bi/dist-node/plugins/link/components/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToolbarIconWrapper = exports.CardToolbarWrapper = exports.ToolbarWrapper = exports.LinkHrefWrapper = exports.LinkHref = exports.EditorFooter = exports.EditorContent = exports.EditorWrapper = exports.RemoveButton = exports.SaveButton = exports.LinkText = exports.RadioWrapper = exports.ItemRadio = exports.ItemInput = exports.LinkCardWrapper = exports.PlaceholerLinkWrapper = exports.Link = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _weDesign = require("@ali/we-design");

var _everBamboo = require("@ali/4ever-bamboo");

const {
  link,
  PLACEHOLDER_LINK_PADDING_VERTICAL,
  PLACEHOLDER_LINK_PADDING_HORIZONTAL
} = _everBamboo.theme;

const Link = /*#__PURE__*/_styledComponents.default.a(["padding:0 2px;color:", ";text-decoration:underline;cursor:pointer;:hover{color:rgb(0,79,198);}"], link.color);

exports.Link = Link;

const PlaceholerLinkWrapper = /*#__PURE__*/_styledComponents.default.span(["text-indent:0;"]);

exports.PlaceholerLinkWrapper = PlaceholerLinkWrapper;

const LinkCardWrapper = /*#__PURE__*/_styledComponents.default.span(["display:inline-block;"]);

exports.LinkCardWrapper = LinkCardWrapper;

const ItemInput = /*#__PURE__*/_styledComponents.default.div(["position:relative;line-height:32px;margin:6px 0;text-align:left;flex:0 0 auto;display:flex;justify-content:space-between;label{margin-right:4px;}input{max-width:370px;color:#171A1D;line-height:20px;padding:0 6px;display:inline-block;border:1px solid rgba(126,134,142,0.16);border-radius:6px;outline:0;font-size:14px;letter-spacing:0.05em;flex:1;::placeholder{color:rgb(23,26,29,0.24);}}"]);

exports.ItemInput = ItemInput;

const ItemRadio = /*#__PURE__*/_styledComponents.default.div(["position:relative;line-height:20px;margin:6px 0;text-align:left;flex:0 0 auto;display:flex;display:flex;align-items:center;label{margin-right:4px;}"]);

exports.ItemRadio = ItemRadio;

const RadioWrapper = /*#__PURE__*/_styledComponents.default.div(["margin-left:17px;"]);

exports.RadioWrapper = RadioWrapper;

const LinkText = /*#__PURE__*/_styledComponents.default.span(["display:inline-block;min-width:200px;max-width:250px;line-height:28px;padding:0 5px;color:#414C56;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;cursor:pointer;&:hover{color:#0089FF;}"]);

exports.LinkText = LinkText;
const SaveButton = /*#__PURE__*/(0, _styledComponents.default)(_weDesign.Button)(["border-radius:6px;height:32px;"]);
exports.SaveButton = SaveButton;

const RemoveButton = /*#__PURE__*/_styledComponents.default.div(["margin-right:8px;cursor:pointer;display:flex;align-items:center;.we-icon{margin-right:2px;}"]);

exports.RemoveButton = RemoveButton;

const PortalWrapper = /*#__PURE__*/_styledComponents.default.div(["top:auto;left:-1px;right:auto;width:350px;line-height:28px;font-size:14px;font-weight:normal;text-indent:0;border-radius:2px 2px;z-index:1;padding:8px 8px 8px 12px;"]);

const EditorWrapper = /*#__PURE__*/(0, _styledComponents.default)(PortalWrapper)(["display:flex;flex-direction:column;width:400px;padding:10px 16px 0 16px;background:#FFFFFF;border-radius:4px;box-shadow:0px 0px 1px rgba(0,0,0,0.24),0px 8px 24px rgba(0,0,0,0.16);z-index:1;color:#414C56;"]);
exports.EditorWrapper = EditorWrapper;

const EditorContent = /*#__PURE__*/_styledComponents.default.div(["display:flex;flex-direction:column;"]);

exports.EditorContent = EditorContent;

const EditorFooter = /*#__PURE__*/_styledComponents.default.div(["display:flex;justify-content:flex-end;align-items:center;margin-top:10px;margin-bottom:13px;"]);

exports.EditorFooter = EditorFooter;

const LinkHref = /*#__PURE__*/_styledComponents.default.span(["display:inline-block;max-width:200px;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;"]);

exports.LinkHref = LinkHref;

const LinkHrefWrapper = /*#__PURE__*/_styledComponents.default.span(["position:relative;flex:1;margin-left:12px;color:", ";height:24px;line-height:24px;:hover{cursor:pointer;}"], _weDesign.theme.linkColor);

exports.LinkHrefWrapper = LinkHrefWrapper;
const ToolbarWrapper = /*#__PURE__*/(0, _styledComponents.default)(PortalWrapper)(["width:auto;display:flex;align-items:center;background:#FFFFFF;border:1px solid #d9d9d9;box-shadow:0 2px 10px rgba(0,0,0,0.12);z-index:1;color:#414C56;.weicon{font-size:18px;vertical-align:middle;}"]);
exports.ToolbarWrapper = ToolbarWrapper;
const CardToolbarWrapper = /*#__PURE__*/(0, _styledComponents.default)(PortalWrapper)(["padding:0;width:auto;display:flex;align-items:center;z-index:1;border-radius:12px;"]);
exports.CardToolbarWrapper = CardToolbarWrapper;

const ToolbarIconWrapper = /*#__PURE__*/_styledComponents.default.span(["display:inline-block;margin-left:10px;cursor:pointer;border-radius:50%;width:24px;height:24px;line-height:24px;text-align:center;position:relative;:hover{background-color:#ECEDEF;}"]);

exports.ToolbarIconWrapper = ToolbarIconWrapper;
