// @ali/4ever-bi/dist-web/plugins/headingGroup/index.js
import _extends from "@babel/runtime/helpers/extends";
import { GroupFactory } from '@ali/4ever-bamboo';
import createGroupNodes from "./handlers/createGroupNodes";
import createRenderNode from "./handlers/createRenderNode";
import * as queries from "./queries";
import { groupFactory } from '@ali/4ever-factory';
var actions = GroupFactory.actions;
var DRAG_BUTTON_OFFSET = -20;
var PI_MODE = 'page_layout';
var createHeadingGroupPlugin = groupFactory(function (config) {
  var _config$headingGroup = config.headingGroup,
      headingGroup = _config$headingGroup === void 0 ? {} : _config$headingGroup;
  var _headingGroup$level = headingGroup.level,
      level = _headingGroup$level === void 0 ? 3 : _headingGroup$level;
  var shouldOffsetTypes = [];

  for (var i = level; i > 0; i--) {
    shouldOffsetTypes.push("heading-" + i);
  }

  ;
  return {
    groupType: 'headingGroup',
    renderNode: createRenderNode(headingGroup),
    groupNodes: createGroupNodes(headingGroup),
    queries: _extends({
      getHeadingGroupLevel: function getHeadingGroupLevel() {
        return level;
      },
      isHeadingGroupAvailable: function isHeadingGroupAvailable() {
        return true;
      },
      getLeftToolbarOffset: function getLeftToolbarOffset(controller, data) {
        var node = data.node;
        var isPi = controller.query('getLayoutMode') === PI_MODE;

        if (!isPi && shouldOffsetTypes.includes(node.type)) {
          return {
            // 这里加上4是因为左侧工具栏的offset为6，但是在标题场景下是2
            x: DRAG_BUTTON_OFFSET + 4,
            y: 0
          };
        }

        return undefined;
      }
    }, queries)
  };
}); // @ts-ignore

createHeadingGroupPlugin.actions = actions; // @ts-ignore

createHeadingGroupPlugin.queries = queries;
export default createHeadingGroupPlugin;
