// @ali/we-design-next/dist-web/display/menu/styled.js
import styled from 'styled-components';
import { TEXT_CSS } from "../../basic/text";
import { COLORS, CSS_SCROLLBAR } from "../../common";
export var MenuWrap = /*#__PURE__*/styled.div(["", " outline:none;min-width:150px;overflow-x:hidden;overflow-y:overlay;"], CSS_SCROLLBAR);
export var MenuItemWrap = /*#__PURE__*/styled.div(["", " user-select:none;height:", ";line-height:20px;padding:0 8px;margin:4px;display:flex;align-items:center;border-radius:4px;", " cursor:", ";color:", ";"], TEXT_CSS.TEXT, function (_ref) {
  var hasDescription = _ref.hasDescription;
  return hasDescription ? '53px' : '34px';
}, function (_ref2) {
  var isActive = _ref2.isActive;
  return isActive ? "background: " + COLORS.OTHERS.DAO_HOVER_BG_COLOR + ";" : null;
}, function (_ref3) {
  var disabled = _ref3.disabled;
  return disabled ? 'not-allowed' : 'default';
}, function (_ref4) {
  var disabled = _ref4.disabled;
  return disabled ? COLORS.LIGHT.LEVEL4_BASE_COLOR : 'unset';
});
export var MenuItemContent = /*#__PURE__*/styled.div(["flex-grow:1;word-break:keep-all;"]);
export var MenuItemTitleDescription = /*#__PURE__*/styled.div(["margin-top:8px;"]);
export var MenuItemDescription = /*#__PURE__*/styled.div(["", " margin-bottom:8px;color:", ";"], TEXT_CSS.TEXT_SUB, function (_ref5) {
  var disabled = _ref5.disabled;
  return disabled ? COLORS.LIGHT.LEVEL4_BASE_COLOR : COLORS.LIGHT.LEVEL2_BASE_COLOR;
});
export var MenuItemPrefix = /*#__PURE__*/styled.div(["flex-grow:0;margin-right:4px;position:relative;top:0.2em;"]);
export var MenuItemSuffix = /*#__PURE__*/styled.div(["flex-grow:0;margin-left:16px;position:relative;top:0.2em;"]);
export var MenuItemDivider = /*#__PURE__*/styled.div(["height:1px;margin:4px;background-color:", ";"], COLORS.OTHERS.DIVIDER);
export var MenuItemGroup = /*#__PURE__*/styled.div(["&::after{display:block;margin:0 4px;height:1px;background-color:", ";content:'';}&:last-child{&::after{display:none;}}"], COLORS.OTHERS.DIVIDER);
export var MenuItemGroupTitle = /*#__PURE__*/styled.div(["", " color:", ";line-height:36px;height:36px;padding-left:12px;user-select:none;"], TEXT_CSS.TEXT_SUB, COLORS.LIGHT.LEVEL2_BASE_COLOR);
export var SuffixPlaceholder = /*#__PURE__*/styled.div(["margin-left:16px;width:16px;height:16px;"]);
