// @ali/4ever-plugin-mention/dist-web/bi/components/mentionButtonMobile.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { selectMention } from "../actions";

var isDisabled = function isDisabled(controller) {
  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  return isCollapsedAtEdge;
};

export function MobileMentionButton(props) {
  var controller = props.controller,
      Wrapper = props.Wrapper,
      item = props.item;

  var handleClick = function handleClick() {
    return controller.run('onAction', selectMention());
  };

  return /*#__PURE__*/_createElement(Wrapper, {
    type: "big",
    icon: "at",
    item: item,
    isDisabled: isDisabled(controller),
    onClick: handleClick
  });
}
export function MobileMentionInsertButton(props) {
  var controller = props.controller,
      Wrapper = props.Wrapper,
      item = props.item,
      locale = props.locale;

  var handleClick = function handleClick() {
    return controller.run('onAction', selectMention());
  };

  return /*#__PURE__*/_createElement(Wrapper, {
    type: "line",
    icon: "at",
    item: item,
    isDisabled: isDisabled(controller),
    lineText: locale == null ? void 0 : locale.buttonTitle,
    onClick: handleClick
  });
}
