// @ali/4ever-plugin-mention/dist-web/bi/components/mentionButton.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Button, MentionBetaNormal } from '@ali/we-design';
import isSelectionInMention from "../utils/isSelectionInMention";
import { activeMention } from "../actions";
import { MentionRole } from "../../utils/role";

var isDisabled = function isDisabled(controller) {
  var value = controller.value;
  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var isDisabled = isCollapsedAtEdge || isSelectionInMention(value) || controller.query('isSelectionInLink') || controller.query('isSelectionInImage') || controller.query('isSelectionInLinkCard') || controller.query('isSelectionInHr');
  return Boolean(isDisabled);
};

var _ref = /*#__PURE__*/_createElement(MentionBetaNormal, null);

export default function MentionButton(props) {
  var controller = props.controller,
      locale = props.locale;

  var handleClick = function handleClick() {
    return controller.run('onAction', activeMention());
  };

  return /*#__PURE__*/_createElement(Button, {
    title: locale.buttonTitle,
    bordered: false,
    testid: "bi-toolbar-mention",
    icon: _ref,
    disabled: isDisabled(controller),
    onClick: handleClick,
    role: MentionRole.mention
  });
}
