// @ali/4ever-plugin-mention/dist-web/bi/components/styled.js
import styled from 'styled-components';
import { theme } from '@ali/we-design';
export var Mention = /*#__PURE__*/styled.div(["position:relative;padding:0;margin:0;width:", ";max-height:", ";height:100%;overflow:auto;div:first-child{overflow-y:auto;}"], function (props) {
  return props.theme.suggestionWidth;
}, function (props) {
  return props.theme.suggestionMaxHeight;
});
export var MentionAt = /*#__PURE__*/styled.span(["position:relative;[data-cangjie-leaf]{color:", ";}"], function (props) {
  return props.active ? theme.primaryColor : 'inherit';
});
export var Toolbar = /*#__PURE__*/styled.div(["position:absolute;opacity:", ";visibility:", ";min-width:200px;line-height:28px;display:flex;flex-direction:row;font-size:14px;font-weight:normal;text-indent:0;top:auto;bottom:-46px;left:-1px;right:auto;background:", ";border-radius:8px;height:40px;border:1px solid ", ";box-shadow:", ";padding:5px 3px;z-index:1;padding:0 0;height:auto;bottom:auto;overflow:auto;"], function (props) {
  return props.active ? 1 : 0;
}, function (props) {
  return props.active ? 'visible' : 'hidden';
}, theme.colorGray1, theme.borderColor, theme.boxShadow);
export var MentionItem = /*#__PURE__*/styled.div(["padding:8px 8px 6px 8px;cursor:pointer;background:", ";"], function (props) {
  return props.actived ? theme.colorGray3 : '';
});
export var MentionItemContainer = /*#__PURE__*/styled.div(["display:flex;"]);
export var Avatar = /*#__PURE__*/styled.div(["width:36px;height:36px;display:block;padding:3px 3px;img{width:100%;height:100%;border-radius:18px;}"]);
export var ItemText = /*#__PURE__*/styled.div(["flex:1;display:block;padding-left:8px;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;line-height:36px;"]);
export var ItemName = /*#__PURE__*/styled.span(["color:", ";"], theme.colorText4);
export var SuggestionsWrapper = /*#__PURE__*/styled.div(["min-height:323px;height:100%;width:", ";max-height:", ";display:flex;flex-direction:column;& .mention-suggestion-list{::-webkit-scrollbar{background-color:transparent;width:8px;}::-webkit-scrollbar-thumb{background:rgba(194,194,194,1);border-radius:10px;}::-webkit-scrollbar-thumb:hover{background:rgba(0,0,0,0.45);}::-webkit-scrollbar-track{background:transparent;}}"], function (_ref) {
  var suggestionWidth = _ref.theme.suggestionWidth;
  return typeof suggestionWidth === 'string' ? suggestionWidth : suggestionWidth + "px";
}, function (_ref2) {
  var suggestionMaxHeight = _ref2.theme.suggestionMaxHeight;
  return typeof suggestionMaxHeight === 'string' ? suggestionMaxHeight : suggestionMaxHeight + "px";
});
export var LoadingWrapper = /*#__PURE__*/styled.div(["display:flex;justify-content:center;align-items:center;height:", "px;width:", "px;color:", ";font-size:36px;"], function (props) {
  return props.height;
}, function (props) {
  return props.width;
}, theme.colorBlue1);
export var PlaceholderInputWrapper = /*#__PURE__*/styled.div(["user-select:none;"]);
export var PlaceholderBtnWrapper = /*#__PURE__*/styled.div(["color:rgba(23,26,29,0.4);padding:3px;border-radius:6px;cursor:pointer;background:", ";&:hover{background:rgba(126,134,142,0.12);}&:active{background:rgba(126,134,142,0.16);}"], function (props) {
  return props.isSelected ? theme.colorGray4 : '';
});
export var InsertWrapper = /*#__PURE__*/styled.div(["position:absolute;display:flex;flex-direction:column;background-color:#ffffff;border:1px solid ", ";box-shadow:", ";border-radius:6px;height:375px;width:335px;"], theme.colorLine1, theme.boxShadow);
export var InputWrapper = /*#__PURE__*/styled.div(["user-select:none;max-width:400px;overflow-x:scroll;&::-webkit-scrollbar{display:none;}& .inline-placeholder{text-overflow:ellipsis;white-space:nowrap;display:inline-block;}"]);
export var InputSpanWrapper = /*#__PURE__*/styled.span(["border:none;outline:none;caret-color:#0089FF;color:#0089FF;user-select:text;"]);
export var PlaceholderTextWrapper = /*#__PURE__*/styled.span(["color:rgba(23,26,29,0.4);"]);
export var PlaceholderIcon = /*#__PURE__*/styled.span(["display:inline-block;color:#0089FF;user-select:text;border:none;outline:none;"]);
