// @ali/4ever-plugin-mention/dist-web/bi/components/suggestionItem.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { noop } from 'lodash-es';
import { MentionItem, Avatar, ItemText, ItemName, MentionItemContainer } from "./styled";

var defaultRender = function defaultRender(suggestion) {
  return /*#__PURE__*/_createElement(MentionItemContainer, null, /*#__PURE__*/_createElement(Avatar, null, /*#__PURE__*/_createElement("img", {
    src: suggestion.avatarUrl,
    alt: ""
  })), /*#__PURE__*/_createElement(ItemText, null, /*#__PURE__*/_createElement(ItemName, null, suggestion.name)));
};

var MentionSuggestionItem = function MentionSuggestionItem(props) {
  var index = props.index,
      _props$style = props.style,
      style = _props$style === void 0 ? {} : _props$style,
      data = props.data;
  var suggestions = data.suggestions,
      selectedIndex = data.selectedIndex,
      _data$renderItem = data.renderItem,
      renderItem = _data$renderItem === void 0 ? defaultRender : _data$renderItem,
      _data$onSelect = data.onSelect,
      onSelect = _data$onSelect === void 0 ? noop : _data$onSelect,
      onSelectedIndexChange = data.onSelectedIndexChange;
  var handleMouseEnter = React.useCallback(function () {
    onSelectedIndexChange(index);
  }, [index, onSelectedIndexChange]);
  var handleMouseDown = React.useCallback(function (event) {
    event.preventDefault();
    event.stopPropagation();
    onSelect(index);
  }, [index, onSelect]);
  var actived = selectedIndex === index;
  return /*#__PURE__*/_createElement(MentionItem, {
    className: "mention-suggestion-item " + (actived ? 'actived' : ''),
    "data-testid": "mention-suggestion-item",
    "data-type": "mention-suggestion-item",
    style: style,
    onMouseDown: handleMouseDown,
    onMouseEnter: handleMouseEnter,
    actived: actived
  }, renderItem(suggestions[index]));
};

export default /*#__PURE__*/React.memo(MentionSuggestionItem);
