// @ali/4ever-plugin-mention/dist-web/bi/components/suggestionForQuickInsert.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { FixedSizeList as List } from 'react-window';
import AutoSizer from 'react-virtualized-auto-sizer';
import { isNil } from 'lodash-es';
import { LoadingBetaNormal, SpinErrorCard } from '@ali/we-design';
import isKeyHotkey from 'is-hotkey';
import { deactiveMention, insertMention } from "../actions";
import MentionSuggestionItem from "./suggestionItem";
import { SuggestionsWrapper, LoadingWrapper } from "./styled";
import { useDelayedState } from '@ali/4ever-component';
var isUpHotKey = isKeyHotkey('up');
var isDownHotKey = isKeyHotkey('down');
var isLeftHotKey = isKeyHotkey('left');
var isRightHotKey = isKeyHotkey('right');

var _ref3 = /*#__PURE__*/_createElement(LoadingBetaNormal, null);

var MentionSuggestion = function MentionSuggestion(props) {
  var controller = props.controller,
      text = props.text,
      searchSuggestions = props.searchSuggestions,
      renderItem = props.renderItem,
      renderHeader = props.renderHeader,
      renderFooter = props.renderFooter,
      renderLoading = props.renderLoading,
      renderEmpty = props.renderEmpty,
      _props$offsetBottom = props.offsetBottom,
      offsetBottom = _props$offsetBottom === void 0 ? 0 : _props$offsetBottom,
      loadMoreSuggestions = props.loadMoreSuggestions,
      _props$itemHeight = props.itemHeight,
      itemHeight = _props$itemHeight === void 0 ? 50 : _props$itemHeight,
      at = props.at;

  var _React$useState = React.useState([]),
      suggestions = _React$useState[0],
      setSuggestions = _React$useState[1];

  var _React$useState2 = React.useState({}),
      errData = _React$useState2[0],
      setErrData = _React$useState2[1];

  var _React$useState3 = React.useState(0),
      selectedIndex = _React$useState3[0],
      setSelectedIndex = _React$useState3[1];

  var searching = React.useRef(false);

  var _useDelayedState = useDelayedState(false),
      delayedSearching = _useDelayedState[0],
      setDelayedSearching = _useDelayedState[1];

  var listRef = React.useRef(null);
  var lastItem = React.useRef(null);
  var editorRef = React.useRef(controller);
  var isEmptyText = text.trim().length === 0;
  var isEmptySuggestions = suggestions.length === 0;
  var handleSelect = React.useCallback(function (index) {
    var suggestion = suggestions[isNil(index) ? selectedIndex : index];
    editorRef.current.run('onAction', insertMention(suggestion, at));
  }, [suggestions, selectedIndex]);
  var handleScroll = React.useCallback(function (_ref) {
    var scrollOffset = _ref.scrollOffset;

    if (!loadMoreSuggestions || !listRef.current) {
      return;
    }
    /* 当倒数第 offsetBottom + 1 个元素进入可视区域时触发 loadMoreSuggestions */


    var startOffset = (suggestions.length - offsetBottom - 2) * itemHeight - listRef.current.props.height;

    if (startOffset > 0 && startOffset <= scrollOffset && lastItem.current !== suggestions.length) {
      lastItem.current = suggestions.length;
      loadMoreSuggestions(text, suggestions.length, setSuggestions, setErrData);
    }
  }, [loadMoreSuggestions, suggestions.length, offsetBottom, itemHeight, text]);
  var itemData = React.useMemo(function () {
    return {
      suggestions: suggestions,
      onSelect: handleSelect,
      renderItem: renderItem,
      selectedIndex: selectedIndex,
      onSelectedIndexChange: setSelectedIndex
    };
  }, [suggestions, renderItem, selectedIndex, handleSelect]);
  React.useEffect(function () {
    editorRef.current = controller;
  }, [controller]); // listen up & down

  React.useEffect(function () {
    var handleMove = function handleMove(event) {
      if (event.isComposing || searching.current) {
        return;
      }

      if (isRightHotKey(event) || isDownHotKey(event)) {
        event.preventDefault();
        setSelectedIndex(function (index) {
          return (index + 1) % suggestions.length;
        });
      }

      if (isLeftHotKey(event) || isUpHotKey(event)) {
        event.preventDefault();
        setSelectedIndex(function (index) {
          return index === 0 ? suggestions.length - 1 : index - 1;
        });
      }
    };

    document.addEventListener('keydown', handleMove);
    return function () {
      document.removeEventListener('keydown', handleMove);
    };
  }, [suggestions]);
  React.useEffect(function () {
    var handleKeyDown = function handleKeyDown(event) {
      if (!event.key) {
        return;
      }

      var key = event.key.toLowerCase();

      if (event.isComposing || searching.current) {
        return;
      }

      if (key === 'enter') {
        event.preventDefault();
        event.stopPropagation();

        if (delayedSearching) {
          return;
        }

        handleSelect();
        return editorRef.current.run('onAction', deactiveMention());
      }

      if (key === ' ' && event.keyCode === 32) {
        if (!isEmptyText && !isEmptySuggestions) {
          event.preventDefault();
          event.stopPropagation();
          handleSelect(0);
        } else {
          return editorRef.current.run('onAction', deactiveMention());
        }
      }
    };

    document.addEventListener('keydown', handleKeyDown);
    return function () {
      document.removeEventListener('keydown', handleKeyDown);
    };
  }, [handleSelect, isEmptySuggestions, isEmptyText, delayedSearching]); // reset loading when search done

  React.useEffect(function () {
    setDelayedSearching(false);
    searching.current = false;
  }, [setDelayedSearching, suggestions]); // do search when text changed

  React.useEffect(function () {
    // @ts-ignore value 没有 isComposing 属性
    if (editorRef.current.value.isComposing) {
      return;
    }

    if (!searchSuggestions) {
      // eslint-disable-next-line no-console
      console.warn('Missing `searchSuggestions`');
      return;
    }

    setDelayedSearching(true, 200);
    searching.current = true;
    searchSuggestions(text, setSuggestions, setErrData);
  }, [text, searchSuggestions, setDelayedSearching]); // scroll to selected index

  React.useEffect(function () {
    if (listRef.current) {
      listRef.current.scrollToItem(selectedIndex);
    }
  }, [selectedIndex]);
  return /*#__PURE__*/_createElement(SuggestionsWrapper, null, renderHeader && renderHeader(), /*#__PURE__*/_createElement("div", {
    style: {
      flex: 1
    }
  }, /*#__PURE__*/_createElement(AutoSizer, null, function (_ref2) {
    var height = _ref2.height,
        width = _ref2.width;
    return delayedSearching ? /*#__PURE__*/_createElement(LoadingWrapper, {
      "data-testid": "mention-suggestion-loading",
      height: height,
      width: width
    }, renderLoading ? renderLoading() : _ref3) : /*#__PURE__*/_createElement(React.Fragment, null, errData.isServiceError ? /*#__PURE__*/_createElement(SpinErrorCard, {
      style: {
        width: width + "px",
        height: height + "px",
        border: 'none'
      },
      tip: errData.serviceErrorMsg
    }) : suggestions.length === 0 && renderEmpty && renderEmpty(), suggestions.length > 0 && /*#__PURE__*/_createElement(List, {
      className: "mention-suggestion-list",
      ref: listRef,
      height: height,
      itemCount: suggestions.length,
      itemSize: itemHeight,
      itemData: itemData,
      width: width,
      onScroll: handleScroll
    }, MentionSuggestionItem));
  })), renderFooter && renderFooter());
};

export default /*#__PURE__*/React.memo(MentionSuggestion);
