// @ali/4ever-bi/dist-web/plugins/index.js
import { DocStylePlugin as docStyle } from '@ali/4ever-bamboo';
import alignment from "./alignment";
import blockquote from "./blockquote";
import bold from "./bold";
import clearStyle from "./clearStyle";
import { clipboardLatency } from '@ali/4ever-plugin-clipboard';
import { biFactory as code } from '@ali/4ever-plugin-code';
import { biFactory as br } from '@ali/4ever-plugin-br';
import { biFactory as inlinePlaceholder } from '@ali/4ever-plugin-inline-placeholder';
import { biFactory as inlineCode } from '@ali/4ever-plugin-inline-code';
import { biFactory as textBackground } from '@ali/4ever-plugin-text-background';
import { biFactory as templateButton } from '@ali/4ever-plugin-template-button';
import { biFactory as table } from '@ali/4ever-plugin-table';
import color from "./color";
import { biFactory as contextMenu } from '@ali/4ever-plugin-pc-contex-menu';
import embed from "./embed";
import fonts from "./fonts";
import formatPainter from "./formatPainter";
import heading from "./heading";
import highlight from "./highlight";
import hr from "./hr";
import image from "./image";
import indent from "./indent";
import italic from "./italic";
import lineheight from "./lineheight";
import link from "./link";
import linkCard from "./link/linkCard";
import list from "./list";
import marks from "./marks";
import mention from "./mention";
import onlineVideo from "./onlineVideo";
import placeholder from "./placeholder";
import refer from "./refer";
import sticker from "./sticker";
import strike from "./strike";
import sz from "./sz";
import translate from "./translate";
import underline from "./underline";
import vertAlign from "./vertAlign";
import video from "./video";
import draggable from "./draggable";
import audio from "./audio";
import newlineGuide from "./newlineGuide";
import { biFactory as paragraph } from '@ali/4ever-plugin-paragraph';
import letterSpacing from "./letterSpacing";
import paragraphColor from "./paragraphColor";
import clearParagraphShd from "./clearParagraphShd";
import autoSpacing from "./autoSpacing"; // import hoverToolbar from './hoverToolbar';

import hoverBlock from "./hoverBlock";
import uuid from "./uuid";
import pref from "./pref";
import { biFactory as snippets } from '@ali/4ever-plugin-snippets';
/**
 * 内置的插件集合
 */

var plugins = {
  snippets: snippets,
  br: br,
  clipboardLatency: clipboardLatency,
  // 纯数据采集，放在最前面
  clearParagraphShd: clearParagraphShd,
  // 回车时清除段落底纹，放在前面
  autoSpacing: autoSpacing,
  // 会修改用户输入，放在其它可能影响输入的插件之前
  alignment: alignment,
  bold: bold,
  blockquote: blockquote,
  clearStyle: clearStyle,
  audio: audio,
  video: video,
  embed: embed,
  color: color,
  fonts: fonts,
  formatPainter: formatPainter,
  draggable: draggable,
  // 放在需要拖拽的行内元素之前
  newlineGuide: newlineGuide,
  inlineCode: inlineCode,
  // mention 要在 heading 之前，否则 heading 处理 Enter 键会造成 mention 在标题插入重复
  mention: mention,
  heading: heading,
  highlight: highlight,
  hr: hr,
  italic: italic,
  strike: strike,
  image: image,
  lineheight: lineheight,
  link: link,
  linkCard: linkCard,
  list: list,
  code: code,
  marks: marks,
  textBackground: textBackground,
  templateButton: templateButton,
  refer: refer,
  onlineVideo: onlineVideo,
  placeholder: placeholder,
  sticker: sticker,
  sz: sz,
  // table 要放在 indent 前，indent 的 Tab 键会把 table 打挂
  table: table,
  indent: indent,
  translate: translate,
  underline: underline,
  vertAlign: vertAlign,
  contextMenu: contextMenu,
  inlinePlaceholder: inlinePlaceholder,
  paragraph: paragraph,
  letterSpacing: letterSpacing,
  paragraphColor: paragraphColor,
  // 放在其他插件之后兜底处理默认样式
  docStyle: docStyle,
  uuid: uuid,
  // hoverToolbar, @淼淼 todo 此处添加会直接导致原生的 inlineToolbar 失效，应该在应用层手动添加
  hoverBlock: hoverBlock,
  pref: pref
};
export default plugins;
