// @ali/4ever-bi/dist-web/plugins/mobile.js
import { DocStylePlugin as docStyle } from '@ali/4ever-bamboo';
import alignment from "./alignment/mobile";
import blockquote from "./blockquote/mobile";
import bold from "./bold/mobile";
import clearStyle from "./clearStyle/mobile";
import columns from "./columns/mobile";
import callout from "./callout/mobile";
import { biFactory as textBackground } from "@ali/4ever-plugin-text-background";
import { mobileBiFactory as code } from '@ali/4ever-plugin-code';
import { mobileBiFactory as inlineCode } from '@ali/4ever-plugin-inline-code';
import { mobileBiFactory as templateButton } from '@ali/4ever-plugin-template-button';
import { mobileBiFactory as table } from '@ali/4ever-plugin-table';
import color from "./color/mobile";
import heading from "./heading/mobile";
import highlight from "./highlight/mobile";
import hr from "./hr/mobile";
import image from "./image/mobile";
import indent from "./indent/mobile";
import italic from "./italic/mobile";
import link from "./link/mobile";
import linkCard from "./link/linkCard/mobile";
import list from "./list/mobile";
import marks from "./marks";
import mention from "./mention/mobile";
import onlineVideo from "./onlineVideo/mobile";
import placeholder from "./placeholder/mobile";
import strike from "./strike/mobile";
import sz from "./sz/mobile";
import sticker from "./sticker/mobile";
import underline from "./underline/mobile";
import vertAlign from "./vertAlign/mobile";
import autoSpacing from "./autoSpacing";
import { mobileBiFactory as paragraph } from '@ali/4ever-plugin-paragraph';
import voteCard from "./voteCard/mobile";
import { biFactory as br } from '@ali/4ever-plugin-br';
import { biFactory as inlinePlaceholder } from '@ali/4ever-plugin-inline-placeholder';
export default {
  br: br,
  code: code,
  autoSpacing: autoSpacing,
  // 会修改用户输入，放在其它可能影响输入的插件之前
  alignment: alignment,
  bold: bold,
  blockquote: blockquote,
  color: color,
  clearStyle: clearStyle,
  heading: heading,
  hr: hr,
  highlight: highlight,
  image: image,
  italic: italic,
  link: link,
  linkCard: linkCard,
  list: list,
  inlineCode: inlineCode,
  marks: marks,
  indent: indent,
  textBackground: textBackground,
  columns: columns,
  callout: callout,
  templateButton: templateButton,
  voteCard: voteCard,
  mention: mention,
  placeholder: placeholder,
  strike: strike,
  sz: sz,
  sticker: sticker,
  table: table,
  underline: underline,
  vertAlign: vertAlign,
  onlineVideo: onlineVideo,
  inlinePlaceholder: inlinePlaceholder,
  paragraph: paragraph,
  // 放在其他插件之后兜底处理默认样式
  docStyle: docStyle
};
