// @ali/4ever-plugin-group-factory/dist-node/utils/components/group.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everPluginHeadingGroup = require("@ali/4ever-plugin-heading-group");

var _everCangjie = require("@ali/4ever-cangjie");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _actions = require("../actions");

const _createElement = /*#__PURE__*/React.createElement;
const ContentBox = /*#__PURE__*/(0, _styledComponents.default)(_everPluginHeadingGroup.Content)(["position:relative;"]);

const TitleContent = /*#__PURE__*/_styledComponents.default.div([""]);

const GroupBox = /*#__PURE__*/_styledComponents.default.div(["position:relative;background:", ";"], props => props.backgroundColor || 'unset');

const generateRole = (role, fold) => {
  return `${role}${fold ? 'Fold' : 'UnFold'}`;
};

const Group = props => {
  const {
    node,
    attributes,
    isSelected,
    controller,
    children,
    renderExtra,
    renderBackground,
    foldBoxLeft,
    role = 'Default',
    visible
  } = props;
  const originStyle = attributes?.style;
  const style = originStyle ? { ...originStyle
  } : {};

  if (visible === false) {
    style.display = 'none';
  }

  const fold = (0, _everCangjie.useUserData)(node, 'fold') || false;
  const [foldRole, setFoleRole] = React.useState(generateRole(role, fold));
  const first = node.nodes[0];
  const {
    immediatelyUnfoldMap = {}
  } = controller.value.data;
  const immediately = !!immediatelyUnfoldMap[first.key];
  const foldRef = React.useRef(fold);
  const handleFoldChange = React.useCallback(_fold => {
    controller.run('onCangjieBlur');
    controller.run('onAction', (0, _everPluginHeadingGroup.toggleFold)({
      node: first,
      fold: _fold
    }));
    setFoleRole(generateRole(role, fold));
  }, [controller, first, fold, role]);
  React.useEffect(() => {
    foldRef.current = fold;
  }, [controller, first, fold, isSelected]); // 立即展开后需要清除标记

  React.useEffect(() => {
    if (immediately) {
      controller.run('onAction', (0, _actions.clearImmediately)({
        node: first
      }));
    }
  }, [controller, first, immediately]); // 切割议题的标题和内容块

  let titleNode;
  let contentNodes;

  if (Array.isArray(children) && children.length >= 1) {
    titleNode = children[0];
    contentNodes = children.slice(1);
  } else {
    return /*#__PURE__*/_createElement(React.Fragment, null, children);
  }

  return /*#__PURE__*/_createElement(GroupBox, (0, _extends2.default)({}, attributes, {
    "data-cangjie-group-block": true,
    "data-name": "heading-group",
    "data-type": node.type,
    style: style,
    "data-foldable": true
  }), renderBackground, /*#__PURE__*/_createElement(_everPluginHeadingGroup.Title, {
    dataName: "heading-group-title",
    fold: fold,
    foldBoxLeft: foldBoxLeft,
    onChange: handleFoldChange,
    role: foldRole
  }, /*#__PURE__*/_createElement(TitleContent, {
    "data-name": "heading-group-title-content"
  }, titleNode)), contentNodes.length > 0 && /*#__PURE__*/_createElement(ContentBox, {
    dataName: "heading-group-content",
    fold: fold
  }, renderExtra, contentNodes));
};

Group.displayName = 'Group';
var _default = Group;
exports.default = _default;
