// @ali/4ever-bi/dist-node/plugins/video/mobile/video.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everFactory = require("@ali/4ever-factory");

const _createElement = /*#__PURE__*/_react.default.createElement;
const {
  MobileEmbedCard,
  VideoPlayer
} = _everBamboo.EmbedPlugin;

class Video extends _react.PureComponent {
  constructor(...args) {
    super(...args);
    this.isClicked = false;
    this.ref = /*#__PURE__*/_react.default.createRef();
    this.previewRef = /*#__PURE__*/_react.default.createRef();
    this.interval = null;

    this.getNodeData = () => {
      const {
        node
      } = this.props;
      const {
        data
      } = node || {};
      return data;
    };

    this.onMediaLoadError = () => {
      const {
        src
      } = this.getNodeData();
      const {
        onMediaLoadError
      } = this.props;

      if (onMediaLoadError) {
        onMediaLoadError(src);
      }
    };

    this.onMediaLoadSuccess = () => {
      const {
        src
      } = this.getNodeData();
      const {
        onMediaLoadSuccess
      } = this.props;

      if (onMediaLoadSuccess) {
        onMediaLoadSuccess(src);
      }
    };

    this.onDownloadFile = () => {
      const {
        src,
        name,
        type,
        size
      } = this.getNodeData();
      const {
        onDownload
      } = this.props;

      if (onDownload) {
        onDownload(src, {
          name,
          type,
          size
        });
      }
    };
  }

  renderPreview() {
    const {
      isSelected,
      node,
      setTempPosterUrl,
      onMediaLoadSuccess,
      transformVideoURL,
      renderVideoMask,
      onDownload,
      allowDownload,
      locale,
      allowFullscreen
    } = this.props;
    return /*#__PURE__*/_createElement(VideoPlayer, {
      transformVideoURL: transformVideoURL,
      onMediaLoadSuccess: onMediaLoadSuccess,
      onMediaLoadError: this.onMediaLoadError,
      setTempPosterUrl: setTempPosterUrl,
      node: node,
      isSelected: isSelected,
      renderVideoMask: renderVideoMask,
      onDownload: this.onDownloadFile,
      isMobile: true,
      allowDownload: onDownload && allowDownload,
      locale: locale,
      allowFullscreen: allowFullscreen
    });
  }

  renderCard() {
    const {
      name,
      src,
      type,
      size
    } = this.getNodeData();
    const {
      onPreview
    } = this.props;
    return /*#__PURE__*/_createElement(MobileEmbedCard, {
      name: name,
      onPreview: onPreview,
      src: src,
      type: type,
      size: size
    });
  }

  renderInlineToolbar() {
    const {
      controller,
      node,
      locale
    } = this.props;
    return /*#__PURE__*/_createElement(_everFactory.EmbedInlineToolbar, {
      controller: controller,
      node: node,
      locale: locale,
      triggerNode: this.ref.current
    });
  }

  render() {
    const {
      viewType
    } = this.getNodeData();
    const {
      attributes,
      isSelected
    } = this.props;
    return /*#__PURE__*/_createElement(_react.default.Fragment, null, isSelected ? this.renderInlineToolbar() : null, /*#__PURE__*/_createElement("div", (0, _extends2.default)({}, attributes, {
      ref: this.ref
    }), viewType === 'preview' ? this.renderPreview() : this.renderCard()));
  }

}

const Container = props => {
  const scrollableContainer = (0, _everBamboo.useScrollableContainer)();
  return /*#__PURE__*/_createElement(Video, (0, _extends2.default)({}, props, {
    scrollableContainer: scrollableContainer
  }));
};

var _default = Container;
exports.default = _default;
