// @ali/code-markdown/esm/plugins/richText.js
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import * as React from 'react';
import { PluginComponent } from './Plugin';

var MdChangeTypePlugin = /*#__PURE__*/function (_PluginComponent) {
  _inheritsLoose(MdChangeTypePlugin, _PluginComponent);

  function MdChangeTypePlugin(props) {
    var _this;

    _this = _PluginComponent.call(this, props) || this;
    _this.handleClick = _this.handleClick.bind(_assertThisInitialized(_this));
    return _this;
  }

  var _proto = MdChangeTypePlugin.prototype;

  _proto.handleClick = function handleClick() {
    var _this$editorConfig;

    (_this$editorConfig = this.editorConfig) === null || _this$editorConfig === void 0 ? void 0 : _this$editorConfig.richText();
  };

  _proto.render = function render() {
    var _this$editorConfig2;

    if (((_this$editorConfig2 = this.editorConfig) === null || _this$editorConfig2 === void 0 ? void 0 : _this$editorConfig2.richText) === undefined) {
      return /*#__PURE__*/React.createElement(React.Fragment, null);
    }

    return /*#__PURE__*/React.createElement("span", {
      className: "button button-type-clear",
      title: '切换富文本',
      onClick: this.handleClick
    }, "\u5BCC\u6587\u672C");
  };

  return MdChangeTypePlugin;
}(PluginComponent);

MdChangeTypePlugin.pluginName = 'clear';
export { MdChangeTypePlugin as default };
