// @ali/4ever-bi/dist-web/plugins/embed/components/styled.js
import styled from 'styled-components';
export var ContentDesc = /*#__PURE__*/styled.div(["font-size:12px;color:", ";margin-right:16px;"], function (props) {
  return props.mode === 'dark' ? '#F2F4F5' : '#7A8189';
});
export var EmbedFilePreviewWrapper = /*#__PURE__*/styled.div(["position:relative;width:100%;display:flex;flex-direction:column;border:1px solid ", ";border-radius:6px;transition:height .6s;"], function (props) {
  return props.isSelected ? '#D9DBDD' : '#ECEDEF';
});
export var PreviewContentWrapper = /*#__PURE__*/styled.div(["overflow:auto;flex:1;position:relative;justify-content:center;align-items:center;display:flex;background-color:#EDEDED;border-bottom-left-radius:6px;border-bottom-right-radius:6px;"]);
export var CardView = /*#__PURE__*/styled.span(["flex:1;cursor:pointer;:hover{color:#3296fa;}"]);
export var PreviewView = /*#__PURE__*/styled.span(["flex:1;cursor:pointer;:hover{color:#3296fa;}"]);
export var Delete = /*#__PURE__*/styled.span(["flex:1;cursor:pointer;:hover{color:#3296fa;}"]);
export var Copy = /*#__PURE__*/styled.span(["flex:1;cursor:pointer;:hover{color:#3296fa;}"]);
export var Rotation = /*#__PURE__*/styled.span(["flex:1;cursor:pointer;:hover{color:#3296fa;}"]);
export var EmbedFilePreviewMask = /*#__PURE__*/styled.div(["position:absolute;left:0;right:0;top:0;bottom:0;z-index:2;"]);
export var ProgressBar = /*#__PURE__*/styled.div(["width:100%;height:4px;background-color:#D9DBDD;border-radius:4px;"]);
export var ProgressBarInner = /*#__PURE__*/styled.div(["background-color:#0089FF;width:", ";height:100%;transition:width .6s;border-radius:4px;"], function (props) {
  return props.progress + "%";
});
export var ErrorDesc = /*#__PURE__*/styled.div(["font-size:12px;color:#D0021B;"]);
export var PreviewError = /*#__PURE__*/styled.div(["text-align:center;"]);
export var PreviewErrorText = /*#__PURE__*/styled.div(["font-size:13px;color:#414c56;display:inline-block;"]);
export var PreviewErrorRetry = /*#__PURE__*/styled.span(["color:#3A454F;font-weight:bold;cursor:pointer;"]);
