// @ali/we-design-next/dist-node/basic/split-button/SplitButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SplitButton = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _common = require("../../common");
var _popover = require("../../display/popover");
var _styled = require("./styled");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const _createElement = /*#__PURE__*/_react.default.createElement;
const SplitButton = /*#__PURE__*/(0, _react.memo)( /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    secondary = /*#__PURE__*/_createElement(_react.default.Fragment, null),
    children,
    icon,
    secondaryVisible: _secondaryVisible,
    onSecondaryVisibleChange: onSecondaryVisibleChange,
    ...buttonProps
  } = props;
  const [secondaryVisible, setSecondaryVisible] = (0, _common.useDefaultProp)(false, _secondaryVisible, onSecondaryVisibleChange);
  const {
    type = 'normal',
    size = 'normal'
  } = buttonProps;
  return /*#__PURE__*/_createElement(_popover.Popover, {
    content: secondary,
    visible: secondaryVisible,
    onVisibleChange: setSecondaryVisible,
    placement: "bottomRight",
    placementOffset: [4, 0]
  }, /*#__PURE__*/_createElement(_styled.Button, buttonProps, /*#__PURE__*/_createElement(_styled.Main, (0, _extends2.default)({
    size: size,
    btnType: type
  }, (0, _common.proxyRestProps)(buttonProps, 'main')), icon, children), /*#__PURE__*/_createElement(_styled.Secondary, (0, _extends2.default)({
    size: size,
    btnType: type,
    onClick: e => {
      setSecondaryVisible(!secondaryVisible);
      e.stopPropagation();
    }
  }, (0, _common.proxyRestProps)(buttonProps, 'secondary')), /*#__PURE__*/_createElement(_common.ArrowUpDown12, {
    dir: secondaryVisible ? 'up' : 'down'
  }))));
}));
exports.SplitButton = SplitButton;
SplitButton.displayName = 'SplitButton';
