// @ali/4ever-bi/dist-node/plugins/image/components/mobileImageButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MobileImage = MobileImage;
exports.MobileBigImage = MobileBigImage;

var _react = _interopRequireDefault(require("react"));

var _actions = require("../actions");

const _createElement = /*#__PURE__*/_react.default.createElement;

const isButtonDisabled = controller => {
  const {
    document,
    selection
  } = controller.value;
  const isDisabled = document.getClosest(selection.anchor.key, n => controller.query('isVoid', n));
  return Boolean(isDisabled);
};

function MobileImage(props) {
  const {
    controller,
    Wrapper,
    item,
    locale
  } = props;

  const handleClick = () => {
    controller.run('onCangjieBlur');
    controller.run('onAction', (0, _actions.selectImage)());
  };

  return /*#__PURE__*/_createElement(Wrapper, {
    item: item,
    type: "line",
    icon: "image",
    lineText: locale?.quickInsertTitle,
    isDisabled: isButtonDisabled(controller),
    onClick: handleClick
  });
}

function MobileBigImage(props) {
  const {
    controller,
    Wrapper,
    item
  } = props;

  const handleClick = () => {
    controller.run('onCangjieBlur');
    controller.run('onAction', (0, _actions.selectImage)());
  };

  return /*#__PURE__*/_createElement(Wrapper, {
    item: item,
    type: "big",
    icon: "image",
    isDisabled: isButtonDisabled(controller),
    onClick: handleClick
  });
}
