// @ali/4ever-bi/dist-node/plugins/image/components/recommendStyleItems.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RecommendStyleMenu = exports.RecommendStyleMenuType = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _classnames = _interopRequireDefault(require("classnames"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everLogger = _interopRequireDefault(require("@ali/4ever-logger"));

var _isStyleActive = require("../utils/isStyleActive");

var _everUtils = require("@ali/4ever-utils");

var _config = require("../config");

const _createElement = /*#__PURE__*/_react.default.createElement;
let RecommendStyleMenuType;
exports.RecommendStyleMenuType = RecommendStyleMenuType;

(function (RecommendStyleMenuType) {
  RecommendStyleMenuType["SMALL"] = "small";
  RecommendStyleMenuType["LARGE"] = "large";
})(RecommendStyleMenuType || (exports.RecommendStyleMenuType = RecommendStyleMenuType = {}));

const IMAGE_EXAMPLE_URL = 'https://img.alicdn.com/imgextra/i3/O1CN016CBXo01OH1jCj0CB4_!!6000000001679-2-tps-224-160.png';
const EXAMPLE_WIDTH = 224;
const EXAMPLE_HEIGHT = 160;
const EXAMPLE_RATIO = EXAMPLE_HEIGHT / EXAMPLE_WIDTH;
const IMAGE_STYLE_MENU_SMALL = 'image-style-menu-small';
const IMAGE_STYLE_MENU_ITEM_SMALL = 'image-style-menu-item-small';
const IMAGE_STYLE_ITEM_SELECTED = 'image-style-item-selected';
const IMAGE_MARGIN_SMALL = 6;
const IMAGE_MARGIN_NORMAL = 8;

const ItemImage = /*#__PURE__*/_styledComponents.default.div(["background-image:url('", "'),linear-gradient(#FAFAFB,#F2F4F5);height:100%;box-sizing:border-box;background-size:cover;cursor:pointer;"], IMAGE_EXAMPLE_URL);

const ItemImageWrapper = /*#__PURE__*/_styledComponents.default.div(["position:absolute;top:", ";right:", ";width:", ";height:", ";"], p => `${p.margin}px`, p => `${p.margin}px`, p => `calc(100% - ${p.margin * 2}px)`, p => `calc(100% - ${p.margin * 2}px)`);

const ItemImageBorderWrapper = /*#__PURE__*/_styledComponents.default.div(["width:100%;height:100%;overflow:hidden;"]);

const MenuWrapper = /*#__PURE__*/_styledComponents.default.div(["display:grid;grid-template-columns:repeat(3,80px);grid-row-gap:9px;grid-column-gap:3px;justify-content:space-around;&.", "{grid-template-columns:repeat(3,1fr);grid-row-gap:6px;justify-content:space-between;}"], IMAGE_STYLE_MENU_SMALL);

const ItemImageStatusWrapper = /*#__PURE__*/_styledComponents.default.div(["position:relative;display:flex;align-items:center;justify-content:center;padding-top:calc(calc(100% * ", ") + ", "px);border-radius:8px;background:transparent;transition:background 0.3s ease-in-out;&:hover{background:rgba(126,134,142,0.12);}&.", "{background:rgba(0,137,255,0.12);}&.", "{padding-top:calc(calc(100% * ", ") + ", "px);border-radius:5px;}"], EXAMPLE_RATIO, (1 - EXAMPLE_RATIO) * IMAGE_MARGIN_NORMAL * 2, IMAGE_STYLE_ITEM_SELECTED, IMAGE_STYLE_MENU_ITEM_SMALL, EXAMPLE_RATIO, (1 - EXAMPLE_RATIO) * IMAGE_MARGIN_SMALL * 2);

const RecommendItem = props => {
  const {
    isActive,
    imageData,
    demoData,
    onClick,
    size,
    index
  } = props;
  const {
    outline = {},
    radius,
    shadow
  } = demoData;
  const {
    width: outlineWidth
  } = outline;
  const loggerKey = `${size === RecommendStyleMenuType.SMALL ? 'image-inline-toolbar-' : 'image-sidebar-'}${index}`;

  const imageWrapperStyle = _react.default.useMemo(() => {
    const border = _everBamboo.ImagePlugin.getBorder(outline);

    return {
      border,
      boxShadow: shadow,
      borderRadius: radius || 0
    };
  }, [outlineWidth, shadow, radius]);

  const handleClick = _react.default.useCallback(e => {
    e.preventDefault();
    onClick(imageData);

    _everLogger.default.sum(loggerKey);
  }, [onClick, imageData, loggerKey]);

  return /*#__PURE__*/_createElement(ItemImageStatusWrapper, {
    className: (0, _classnames.default)({
      [IMAGE_STYLE_ITEM_SELECTED]: isActive,
      [IMAGE_STYLE_MENU_ITEM_SMALL]: size === RecommendStyleMenuType.SMALL
    }),
    "data-testid": `image-style-option-${index}`
  }, /*#__PURE__*/_createElement(ItemImageWrapper, {
    margin: size === RecommendStyleMenuType.SMALL ? IMAGE_MARGIN_SMALL : IMAGE_MARGIN_NORMAL
  }, /*#__PURE__*/_createElement(ItemImageBorderWrapper, {
    style: imageWrapperStyle
  }, /*#__PURE__*/_createElement(ItemImage, {
    onClick: handleClick,
    onMouseDown: _everUtils.blockEvent,
    "data-testid": `image-recommend-style-${index}`,
    "data-role": `${_everBamboo.PluginRoles.imageRecommendStyleItem}${index}`
  }))));
};

const RecommendStyleMenu = props => {
  const {
    size = RecommendStyleMenuType.LARGE,
    data,
    onClick
  } = props;

  const items = _react.default.useMemo(_config.getRecommendStyleItems, []);

  return /*#__PURE__*/_createElement(MenuWrapper, {
    className: (0, _classnames.default)({
      [IMAGE_STYLE_MENU_SMALL]: size === RecommendStyleMenuType.SMALL
    })
  }, items.map((item, index) => /*#__PURE__*/_createElement(RecommendItem, {
    imageData: item.data,
    demoData: item.demo,
    isActive: (0, _isStyleActive.isStyleActive)(item.data, data || null),
    onClick: onClick,
    size: size,
    key: `image_recommend_style_item_${index}`,
    index: index
  })));
};

exports.RecommendStyleMenu = RecommendStyleMenu;
