// @ali/we-ui/dist-web/components/table-style/styled.js
/* eslint-disable arrow-body-style */

/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/explicit-function-return-type */
import styled from 'styled-components';
import { withTheme } from "../theme";
export var imgPadding = 1;
export var imgMargin = 1;
export var imgBorder = 1;
export var menuImgMargin = {
  upDown: 4,
  leftRight: 6
};
export var smallMode = {
  upper: 26,
  down: 38
};
export var normalMode = {
  upper: 38,
  down: 38
};
export var btnWidth = 12;
export var IconWrapper = withTheme(styled.div(["width:", ";height:", ";cursor:pointer;display:flex;justify-content:center;align-items:center;border-radius:", ";background-color:", ";&:hover{background-color:#eee;}&:active{background-color:#ddd;}.weicon{color:", ";&.disabled{color:", ";cursor:not-allowed;}}"], function (p) {
  return p.isDown ? '26px' : btnWidth + "px";
}, function (p) {
  return p.isDown ? '8px' : btnWidth + "px";
}, function (p) {
  return p.isDown ? '2px 2px 0 0' : '50%';
}, function (p) {
  return p.active ? '#eee' : '';
}, function (p) {
  return p.theme.ui.commonBaseColor3;
}, function (p) {
  return p.theme.ui.commonBaseColor4;
}));
export var TableStyleItemWrap = withTheme(styled.div(["padding:", ";margin:", ";border:", "px solid transparent;width:", ";border-radius:2px;&:hover{border:", ";}&.actived{border:", "px solid #0089FF;}"], function (p) {
  return p.size === 'small' ? imgPadding - 1 + "px" : imgPadding + "px";
}, function (p) {
  if (p.footer) {
    return menuImgMargin.upDown + "px " + menuImgMargin.leftRight + "px";
  } else {
    return "0px " + imgMargin + "px";
  }
}, imgBorder, function (p) {
  if (p.footer === true) {
    return p.mode.down + "px";
  } else {
    return p.mode.upper + "px";
  }
}, function (p) {
  return imgBorder + "px solid " + p.theme.ui.commonBaseColor4;
}, imgBorder));
export var TableStyleItem = withTheme(styled.img(["width:100%;cursor:pointer;display:block;"]));
export var TableStyleSet = withTheme(styled.div(["display:flex;justify-content:flex-start;"]));
export var TableStyleDown = /*#__PURE__*/styled.div(["display:flex;justify-content:center;"]);
export var TableStyleWrapper = /*#__PURE__*/styled.div(["width:", ";padding-top:", ";border:", ";border-radius:", ";"], function (p) {
  return p.width + "px";
}, function (p) {
  return p.mode === 'normal' ? '8px' : '';
}, function (p) {
  return p.mode === 'normal' ? '1px solid rgba(17,31,44,0.08)' : '';
}, function (p) {
  return p.mode === 'normal' ? '4px' : '';
});
export var TableStyleUpper = /*#__PURE__*/styled.div(["display:flex;align-items:center;justify-content:space-between;"]);
export var OverlayWrapper = /*#__PURE__*/styled.div(["padding:10px 10px;"]);
