// @ali/we-ui/dist-web/components/table-style/TableStyle.js
/* eslint-disable no-param-reassign */

/* eslint-disable max-lines-per-function */
import React, { useMemo, useState } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { TableStyleItem, TableStyleSet, TableStyleItemWrap, TableStyleWrapper, TableStyleUpper, TableStyleDown, OverlayWrapper, imgPadding, imgMargin, imgBorder, smallMode, normalMode, btnWidth, IconWrapper } from "./styled";
import useControlledValue from "../../hooks/useControlledValue";
import { Dropdown } from "../dropdown";
import { Tabs, TabPane } from "../tabs";
import { ArrowUpTriangleNormal, ArrowDownTriangleNormal } from "../icons";
import { classnames } from "../../utils";

function rotateArrow(angle) {
  return {
    flex: 'none',
    transform: "rotate(" + angle + "deg)",
    width: 8,
    height: 8
  };
}

var styleList2D = [];
var menuStyleList2D = [];
var STYLE_LIST = [{
  name: '3-1',
  url: 'https://img.alicdn.com/tfs/TB1IZK.e5DsXe8jSZR0XXXK6FXa-38-28.svg'
}, {
  name: '3-2',
  url: 'https://img.alicdn.com/tfs/TB11ecLjaNj0u4jSZFyXXXgMVXa-38-28.svg'
}, {
  name: '3-3',
  url: 'https://img.alicdn.com/tfs/TB1LSH9SAL0gK0jSZFtXXXQCXXa-38-28.svg'
}, {
  name: '3-4',
  url: 'https://img.alicdn.com/tfs/TB1o9gZdZieb18jSZFvXXaI3FXa-38-28.svg'
}, {
  name: '3-5',
  url: 'https://img.alicdn.com/tfs/TB1aAYxg9R26e4jSZFEXXbwuXXa-38-28.svg'
}, {
  name: '3-6',
  url: 'https://img.alicdn.com/tfs/TB1sA7hSpT7gK0jSZFpXXaTkpXa-38-28.svg'
}, {
  name: '3-7',
  url: 'https://img.alicdn.com/tfs/TB1jiUAetTfau8jSZFwXXX1mVXa-38-28.svg'
}, {
  name: '4-1',
  url: 'https://img.alicdn.com/tfs/TB16_wfSuH2gK0jSZJnXXaT1FXa-38-28.svg'
}, {
  name: '4-2',
  url: 'https://img.alicdn.com/tfs/TB1X84Ge9slXu8jSZFuXXXg7FXa-38-28.svg'
}, {
  name: '4-3',
  url: 'https://img.alicdn.com/tfs/TB1OcIvSpP7gK0jSZFjXXc5aXXa-38-28.svg'
}, {
  name: '4-4',
  url: 'https://img.alicdn.com/tfs/TB1oiUAetTfau8jSZFwXXX1mVXa-38-28.svg'
}, {
  name: '4-5',
  url: 'https://img.alicdn.com/tfs/TB1IQZhSqL7gK0jSZFBXXXZZpXa-38-28.svg'
}, {
  name: '4-6',
  url: 'https://img.alicdn.com/tfs/TB1jkkiSAL0gK0jSZFAXXcA9pXa-38-28.svg'
}, {
  name: '4-7',
  url: 'https://img.alicdn.com/tfs/TB1MDEAi8Bh1e4jSZFhXXcC9VXa-38-28.svg'
}, {
  name: '5-1',
  url: 'https://img.alicdn.com/tfs/TB1hB4Ge9slXu8jSZFuXXXg7FXa-38-28.svg'
}, {
  name: '5-2',
  url: 'https://img.alicdn.com/tfs/TB1jQYxg9R26e4jSZFEXXbwuXXa-38-28.svg'
}, {
  name: '5-3',
  url: 'https://img.alicdn.com/tfs/TB1VMz_gzMZ7e4jSZFOXXX7epXa-38-28.svg'
}, {
  name: '5-4',
  url: 'https://img.alicdn.com/tfs/TB1MeRhi_M11u4jSZPxXXahcXXa-38-28.svg'
}, {
  name: '5-6',
  url: 'https://img.alicdn.com/tfs/TB1ZKIcSxD1gK0jSZFsXXbldVXa-38-28.svg'
}, {
  name: '5-5',
  url: 'https://img.alicdn.com/tfs/TB1TuUbSEz1gK0jSZLeXXb9kVXa-38-28.svg'
}, {
  name: '5-7',
  url: 'https://img.alicdn.com/tfs/TB1OTEAi8Bh1e4jSZFhXXcC9VXa-38-28.svg'
}, {
  name: '1-1',
  url: 'https://img.alicdn.com/tfs/TB1WcVCgwgP7K4jSZFqXXamhVXa-38-28.svg'
}, {
  name: '1-2',
  url: 'https://img.alicdn.com/tfs/TB1KokmSAY2gK0jSZFgXXc5OFXa-38-28.svg'
}, {
  name: '1-3',
  url: 'https://img.alicdn.com/tfs/TB1CTJ0dhvbeK8jSZPfXXariXXa-38-28.svg'
}, {
  name: '1-4',
  url: 'https://img.alicdn.com/tfs/TB18dIcSvb2gK0jSZK9XXaEgFXa-38-28.svg'
}, {
  name: '1-5',
  url: 'https://img.alicdn.com/tfs/TB1_iQAetTfau8jSZFwXXX1mVXa-38-28.svg'
}, {
  name: '1-6',
  url: 'https://img.alicdn.com/tfs/TB1qdkXSpY7gK0jSZKzXXaikpXa-38-28.svg'
}, {
  name: '1-7',
  url: 'https://img.alicdn.com/tfs/TB1DKUbSEz1gK0jSZLeXXb9kVXa-38-28.svg'
}, {
  name: '2-1',
  url: 'https://img.alicdn.com/tfs/TB16._3fAcx_u4jSZFlXXXnUFXa-38-28.svg'
}, {
  name: '2-2',
  url: 'https://img.alicdn.com/tfs/TB1GDJ0dhvbeK8jSZPfXXariXXa-38-28.svg'
}, {
  name: '2-3',
  url: 'https://img.alicdn.com/tfs/TB1v_EAi8Bh1e4jSZFhXXcC9VXa-38-28.svg'
}, {
  name: '2-4',
  url: 'https://img.alicdn.com/tfs/TB1gHwhSEH1gK0jSZSyXXXtlpXa-38-28.svg'
}, {
  name: '2-5',
  url: 'https://img.alicdn.com/tfs/TB1XtMcSvb2gK0jSZK9XXaEgFXa-38-28.svg'
}, {
  name: '2-6',
  url: 'https://img.alicdn.com/tfs/TB12DwfSuH2gK0jSZJnXXaT1FXa-38-28.svg'
}, {
  name: '2-7',
  url: 'https://img.alicdn.com/tfs/TB1_7giSAL0gK0jSZFAXXcA9pXa-38-28.svg'
}];

function convert2D(arr, pageSize) {
  var res = [];

  for (var i = 0; i < arr.length; i += pageSize) {
    var temp = arr.slice(i, i + pageSize);
    res.push(temp);
  }

  return res;
}

var currentImgPadding = imgPadding;
var TableStyle = /*#__PURE__*/React.memo(function (props) {
  var onChange = props.onChange,
      _props$size = props.size,
      size = _props$size === void 0 ? 'normal' : _props$size,
      _props$styleList = props.styleList,
      styleList = _props$styleList === void 0 ? STYLE_LIST : _props$styleList,
      _props$pageSize = props.pageSize,
      pageSize = _props$pageSize === void 0 ? 4 : _props$pageSize,
      _props$menuPageSize = props.menuPageSize,
      menuPageSize = _props$menuPageSize === void 0 ? 7 : _props$menuPageSize;
  currentImgPadding = size === 'normal' ? imgPadding : imgPadding - 1;
  var currentMode = size === 'small' ? smallMode : normalMode;
  var wrapperWidth = btnWidth * 2 + (currentMode.upper + currentImgPadding * 2 + imgMargin * 2 + imgBorder * 2) * pageSize;
  styleList2D = useMemo(function () {
    return convert2D(styleList, pageSize);
  }, [pageSize, styleList]);
  menuStyleList2D = useMemo(function () {
    return convert2D(styleList, menuPageSize);
  }, [menuPageSize, styleList]);
  var maxPage = styleList2D.length - 1;

  var _useControlledValue = useControlledValue(props.value, props.defaultValue || styleList[0].name),
      value = _useControlledValue[0],
      setValue = _useControlledValue[1];

  var handleClick = function handleClick(item) {
    if (item !== value && onChange) {
      onChange(item);
    }

    setValue(item);
    setOverlayVisible(false);
    var idx = styleList.findIndex(function (ele) {
      return ele.name === item;
    });
    var rowIndex = Math.floor(idx / pageSize);

    if (rowIndex !== undefined) {
      switchPage(rowIndex);
    }
  };

  var _useState = useState(0),
      currentTab = _useState[0],
      setCurrentTab = _useState[1];

  var _useState2 = useState(false),
      showArrowDown = _useState2[0],
      setShowArrowDown = _useState2[1];

  var _useState3 = useState(false),
      overlayVisible = _useState3[0],
      setOverlayVisible = _useState3[1];

  var switchPage = function switchPage(nextPage) {
    if (nextPage >= 0 && nextPage <= maxPage) {
      setCurrentTab(nextPage);
    }
  };

  var getOverlay = function getOverlay() {
    return /*#__PURE__*/_createElement(OverlayWrapper, null, menuStyleList2D.map(function (row, rIndex) {
      return /*#__PURE__*/_createElement(TableStyleSet, {
        key: "footer-" + rIndex
      }, row.map(function (item, cIndex) {
        return /*#__PURE__*/_createElement(TableStyleItemWrap, {
          key: "footer-" + item.name,
          onClick: function onClick() {
            return handleClick(item.name);
          },
          footer: true,
          mode: currentMode,
          className: item.name === value ? 'actived' : ''
        }, /*#__PURE__*/_createElement(TableStyleItem, {
          src: menuStyleList2D[rIndex][cIndex].url
        }));
      }));
    }));
  };

  return /*#__PURE__*/_createElement(TableStyleWrapper, {
    onMouseEnter: function onMouseEnter() {
      return setShowArrowDown(true);
    },
    onMouseLeave: function onMouseLeave() {
      return setShowArrowDown(false);
    },
    width: wrapperWidth,
    mode: size
  }, /*#__PURE__*/_createElement(TableStyleUpper, null, /*#__PURE__*/_createElement(IconWrapper, {
    onClick: function onClick() {
      return switchPage(currentTab - 1);
    },
    "data-testid": "arrow-increase"
  }, /*#__PURE__*/_createElement(ArrowUpTriangleNormal, {
    style: rotateArrow(-90),
    className: classnames({
      disabled: currentTab === 0
    })
  })), /*#__PURE__*/_createElement(Tabs, {
    activeKey: String(currentTab),
    hasTabBar: false,
    style: {
      flex: 1
    }
  }, styleList2D.map(function (row, rowIndex) {
    return /*#__PURE__*/_createElement(TabPane, {
      key: String(rowIndex),
      tab: String(rowIndex)
    }, /*#__PURE__*/_createElement(TableStyleSet, null, row.map(function (item, colIndex) {
      return /*#__PURE__*/_createElement(TableStyleItemWrap, {
        key: item.name,
        onClick: function onClick() {
          return handleClick(item.name);
        },
        mode: currentMode,
        size: size,
        className: item.name === value ? 'actived' : ''
      }, /*#__PURE__*/_createElement(TableStyleItem, {
        src: styleList2D[rowIndex][colIndex].url
      }));
    })));
  })), /*#__PURE__*/_createElement(IconWrapper, {
    onClick: function onClick() {
      return switchPage(currentTab + 1);
    }
  }, /*#__PURE__*/_createElement(ArrowUpTriangleNormal, {
    style: rotateArrow(90),
    className: classnames({
      disabled: currentTab === maxPage
    })
  }))), /*#__PURE__*/_createElement(TableStyleDown, null, /*#__PURE__*/_createElement(Dropdown, {
    overlay: getOverlay(),
    onVisibleChange: function onVisibleChange(vi) {
      return setOverlayVisible(vi);
    },
    placement: "bottom",
    visible: overlayVisible
  }, /*#__PURE__*/_createElement(IconWrapper, {
    isDown: true,
    onClick: function onClick() {
      return setOverlayVisible(!overlayVisible);
    },
    width: "26px",
    active: overlayVisible
  }, /*#__PURE__*/_createElement(ArrowDownTriangleNormal, {
    style: showArrowDown || overlayVisible ? {
      width: 8,
      height: 8
    } : {
      visibility: 'hidden',
      width: 8,
      height: 8
    }
  })))));
});
TableStyle.displayName = 'TableStyle';
TableStyle.defaultProps = {
  pageSize: 4,
  size: 'normal',
  menuPageSize: 7
};
export { TableStyle };
