// @ali/we-design-next/dist-web/navigator/toolbar/toolbarHoverTip/index.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Popover } from "../../../display/popover";
import { GuidePanel } from "../../../office/guide-panel";
import { TooltipContent, tooltipOverlayRestStyle } from "../../../display/tooltip/Tooltip";
var placementOffset = [4, 0];
export var ToolbarHoverTip = /*#__PURE__*/React.memo(function (props) {
  var children = props.children,
    disabled = props.disabled,
    tooltip = props.tooltip,
    overlayVisible = props.overlayVisible;
  var contentNode;
  var overlayStyle;
  var placement = 'bottom';
  if (!overlayVisible && tooltip) {
    if (typeof tooltip === 'object' && 'title' in tooltip) {
      var title = tooltip.title,
        description = tooltip.description,
        shortCut = tooltip.shortCut;
      if (disabled) {
        if (!('image' in tooltip)) {
          contentNode = /*#__PURE__*/_createElement(TooltipContent, {
            title: title,
            description: description
          });
          overlayStyle = tooltipOverlayRestStyle;
        }
      } else if ('image' in tooltip) {
        contentNode = /*#__PURE__*/_createElement(GuidePanel, {
          title: title + " " + (shortCut ? " (" + shortCut + ")" : ''),
          description: description,
          image: tooltip.image
        });
        placement = 'bottomLeft';
      } else {
        contentNode = /*#__PURE__*/_createElement(TooltipContent, {
          title: title + " " + (shortCut ? " (" + shortCut + ")" : ''),
          description: description
        });
        overlayStyle = tooltipOverlayRestStyle;
      }
    } else {
      contentNode = /*#__PURE__*/_createElement(TooltipContent, {
        title: tooltip
      });
      overlayStyle = tooltipOverlayRestStyle;
    }
  }

  // @ts-ignore
  return /*#__PURE__*/_createElement(Popover, {
    trigger: "hover",
    placement: placement,
    content: contentNode,
    style: overlayStyle,
    placementOffset: placementOffset
    // visible={!!contentNode}
  }, children);
});
