// @ali/4ever-plugin-table/dist-web/bi/components/styled.js
import styled from 'styled-components';
import { INSERT_INDICATOR_LINE_SIZE, THEME } from "../constants";
import { SCROLL_BAR_HEIGHT } from "../../utils/constants";
export var TableWrapper = /*#__PURE__*/styled.div(["position:relative;", ""], function (p) {
  return p.visible === false ? 'display: none' : '';
});
export var RealTable = /*#__PURE__*/styled.table(["table-layout:fixed;border-collapse:collapse;position:relative;"]);
export var SelectionRect = /*#__PURE__*/styled.div(["position:absolute;display:none;background-color:", ";z-index:1;pointer-events:none;outline-offset:-2px;border:1px solid ", ";"], function (p) {
  return p.isHighlight ? THEME.red0 : THEME.blue0;
}, function (p) {
  return p.isHighlight ? THEME.red1 : THEME.blue3;
});
export var TableResizeController = /*#__PURE__*/styled.div(["position:absolute;"]);
export var TableResizeControlPoint = /*#__PURE__*/styled.div(["z-index:3;position:absolute;width:7px;height:7px;cursor:nw-resize;cursor:nwse-resize;background-color:white;border:1px solid rgba(23,26,29,0.24);left:-4px;top:-4px;"]);
export var TableResizeControlLines = /*#__PURE__*/styled.div(["display:none;"]);
export var TableResizeControlBottomLine = /*#__PURE__*/styled.div(["z-index:2;position:absolute;width:100px;height:1px;background:#0089FF;right:0;bottom:0;background:linear-gradient(to left,rgba(0,137,255,1),rgba(0,137,255,0));mask-image:linear-gradient(to left,#000 6px,transparent 6px);mask-repeat:repeat-x;mask-position:0 0;mask-size:8px 2px;"]);
export var TableResizeControlRightLine = /*#__PURE__*/styled.div(["z-index:2;position:absolute;height:100px;width:1px;background:#0089FF;right:0;bottom:0;background:linear-gradient(to top,rgba(0,137,255,1),rgba(0,137,255,0));mask-image:linear-gradient(to top,#000 6px,transparent 6px);mask-repeat:repeat-y;mask-position:0 0;mask-size:2px 8px;"]);
export var ColResizer = /*#__PURE__*/styled.div(["display:none;position:absolute;bottom:", "px;width:0;border-left:1px solid #3296FA;z-index:30;cursor:col-resize;"], SCROLL_BAR_HEIGHT);
export var RowResizer = /*#__PURE__*/styled.div(["display:none;position:absolute;height:0;border-top:1px solid #3296FA;z-index:5;cursor:row-resize;"]);
var INDICATOR_HEAD_HEIGHT = 4;
var INDICATOR_COLOR = '#0089FF';
export var ColIndicator = /*#__PURE__*/styled.div(["position:absolute;width:0;border-left:", "px solid ", ";cursor:col-resize;z-index:30;opacity:0;transition:opacity ease-in 0.25s;pointer-events:none;"], INSERT_INDICATOR_LINE_SIZE, INDICATOR_COLOR);
export var ColIndicatorHeader = /*#__PURE__*/styled.div(["position:absolute;width:0;border-left:", "px solid ", ";height:", "px;background-color:", ";top:-", "px;left:-", "px;pointer-events:none;"], INSERT_INDICATOR_LINE_SIZE, INDICATOR_COLOR, function (p) {
  return INDICATOR_HEAD_HEIGHT / p.scale;
}, THEME.rowColIndicatorColor, function (p) {
  return INDICATOR_HEAD_HEIGHT / p.scale;
}, INSERT_INDICATOR_LINE_SIZE);
export var RowIndicator = /*#__PURE__*/styled.div(["position:absolute;height:0;border-top:", "px solid ", ";transform:translateY(-1px);cursor:row-resize;z-index:99;opacity:0;transition:opacity ease-in 0.25s;pointer-events:none;"], INSERT_INDICATOR_LINE_SIZE, INDICATOR_COLOR);
export var RowIndicatorHeader = /*#__PURE__*/styled.div(["position:absolute;height:0;width:", "px;background-color:", ";left:-", "px;top:-", "px;pointer-events:none;"], function (p) {
  return INDICATOR_HEAD_HEIGHT / p.scale;
}, THEME.rowColIndicatorColor, function (p) {
  return (INDICATOR_HEAD_HEIGHT - 1) / p.scale;
}, INSERT_INDICATOR_LINE_SIZE);
export var TableSideShadow = /*#__PURE__*/styled.div(["position:absolute;bottom:", "px;width:4px;height:100%;opacity:.5;pointer-events:none;"], function (p) {
  return SCROLL_BAR_HEIGHT / (p.scale || 1);
});
export var TableLeftSideShadow = /*#__PURE__*/styled(TableSideShadow)(["left:0;background:linear-gradient(270deg,rgba(99,114,130,0),rgba(99,114,130,.16));"]);
export var TableRightSideShadow = /*#__PURE__*/styled(TableSideShadow)(["right:0;background:linear-gradient(90deg,rgba(99,114,130,0),rgba(99,114,130,.16));"]);
export var Menu = /*#__PURE__*/styled.ul(["position:absolute;border-radius:4px;box-shadow:0 2px 8px rgba(0,0,0,0.15);margin:0;padding-left:0;background-color:#fff;list-style:none;z-index:3;"]);
export var MenuItem = /*#__PURE__*/styled.li(["height:32px;line-height:32px;word-break:keep-all;white-space:nowrap;font-size:12px;padding:0 12px;color:", ";cursor:", ";:hover{background-color:", ";}"], function (_ref) {
  var disabled = _ref.disabled;
  return disabled ? 'rgba(0, 0, 0, 0.25)' : 'rgb(89, 89, 89)';
}, function (_ref2) {
  var disabled = _ref2.disabled;
  return disabled ? 'not-allowed' : 'pointer';
}, function (_ref3) {
  var disabled = _ref3.disabled;
  return disabled ? 'transparent' : '#f5f6f7';
});
export var MenuDivider = /*#__PURE__*/styled.li(["height:1px;margin:1px 0;overflow:hidden;background-color:#e8e8e8;"]);
export var DEFAULT_BUTTON_SIZE = 20;
export var ButtonWrapper = /*#__PURE__*/styled.div(["display:flex;align-items:center;justify-content:center;width:", "px;height:", "px;font-size:", "px;line-height:", "px;color:black;border-radius:", "px;background-color:white;border:1px solid rgba(17,31,44,0.08);box-shadow:0px 0px 1px rgba(0,0,0,0.16),0px 1px 4px rgba(126,134,142,0.18);box-sizing:border-box;"], function (p) {
  return (p.size || DEFAULT_BUTTON_SIZE) / p.scale;
}, function (p) {
  return (p.size || DEFAULT_BUTTON_SIZE) / p.scale;
}, function (p) {
  return 20 / p.scale;
}, function (p) {
  return 20 / p.scale;
}, function (p) {
  return 4 / p.scale;
});
export var IconWrapper = /*#__PURE__*/styled.span(["display:inline-flex;align-items:center;justify-content:center;"]);
export var Wrapper = /*#__PURE__*/styled.div(["position:relative;"]);
export var TableSelectionRect = /*#__PURE__*/styled.div(["position:absolute;z-index:100;pointer-events:none;outline-offset:-2px;box-sizing:border-box;border:1px solid ", ";"], THEME.blue3);
