// @ali/4ever-bi/dist-node/plugins/onlineVideo/components/onlineVideoButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = OnlineVideoButton;

var React = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _actions = require("../actions");

const _createElement = /*#__PURE__*/React.createElement;
;

var _ref = /*#__PURE__*/_createElement(_weDesign.VedioNormal, null);

function OnlineVideoButton(props) {
  const {
    controller,
    locale
  } = props;
  const isDisabled = controller.query('isSelectionInTable', {
    excludeColumns: true
  }) || controller.query('isSelectionInImage') || controller.query('isSelectionInLinkCard');

  const handleClick = () => {
    if (isDisabled) {
      return;
    }

    controller.run('onAction', (0, _actions.insertOnlineVideo)());
  };

  return /*#__PURE__*/_createElement(_weDesign.Button, {
    title: locale.buttonTitle,
    bordered: false,
    testid: "bi-toolbar-onlinevideo",
    icon: _ref,
    onClick: handleClick
  });
}
