// @ali/4ever-bi/dist-node/plugins/onlineVideo/components/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WrapTooltip = exports.Delete = exports.Toolbar = exports.YoukuPlayer = exports.YoukuPlayerMask = exports.YoukuPlayerPlaceholder = exports.YoukuPlayerWrapper = exports.VideoWrapper = exports.Button = exports.Input = exports.IconWrapper = exports.InsertWrapper = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

const InsertWrapper = /*#__PURE__*/_styledComponents.default.div(["position:relative;box-shadow:1px 3px 8px 0 rgba(0,0,0,0.1);border-radius:6px;border:1px solid rgba(17,31,44,0.08);height:42px;width:390px;background:#fff;display:flex;align-items:center;justify-content:center;"]);

exports.InsertWrapper = InsertWrapper;

const IconWrapper = /*#__PURE__*/_styledComponents.default.div(["padding-left:8px;height:40px;line-height:40px;display:flex;align-items:center;.weicon{margin-top:2px;}"]);

exports.IconWrapper = IconWrapper;

const Input = /*#__PURE__*/_styledComponents.default.input(["height:40px;line-height:40px;display:inline-block;border:none;outline:none;flex:auto;padding:0 10px;text-overflow:ellipsis;white-space:nowrap;overflow:hidden;"]);

exports.Input = Input;

const Button = /*#__PURE__*/_styledComponents.default.button(["outline:none;background:#3296FA;padding:0 5px;height:32px;line-height:32px;color:#fff;border-radius:18px;border:none;cursor:pointer;margin-right:8px;"]);

exports.Button = Button;

const VideoWrapper = /*#__PURE__*/_styledComponents.default.div(["width:100%;height:", ";border:1px solid #ECEDEF;border-radius:6px;video{width:100%;height:100%;background:#000;outline:none;border-radius:6px;}"], props => props.isMobile === true ? '246px' : '436px');

exports.VideoWrapper = VideoWrapper;

const YoukuPlayerWrapper = /*#__PURE__*/_styledComponents.default.div(["position:relative;width:100%;height:", ";"], props => props.isMobile === true ? '246px' : '436px');

exports.YoukuPlayerWrapper = YoukuPlayerWrapper;

const YoukuPlayerPlaceholder = /*#__PURE__*/_styledComponents.default.div(["width:100%;height:45px;line-height:45px;background:#F7F8F9;border-radius:10px;background-color:#F7F8F9;text-align:center;span{font-size:12px;color:rgba(17,31,44,0.4);}"]);

exports.YoukuPlayerPlaceholder = YoukuPlayerPlaceholder;

const YoukuPlayerMask = /*#__PURE__*/_styledComponents.default.div(["position:absolute;top:0;right:0;left:0;bottom:0;z-index:2;"]);

exports.YoukuPlayerMask = YoukuPlayerMask;

const YoukuPlayer = /*#__PURE__*/_styledComponents.default.iframe(["position:absolute;top:0;right:0;left:0;bottom:0;z-index:1;width:100%;height:100%;border:none;border-radius:6px;"]);

exports.YoukuPlayer = YoukuPlayer;

const Toolbar = /*#__PURE__*/_styledComponents.default.div(["span:last-child .split{display:none;}"]);

exports.Toolbar = Toolbar;

const Delete = /*#__PURE__*/_styledComponents.default.span(["flex:1;cursor:pointer;:hover{color:#3296fa;}"]);

exports.Delete = Delete;

const WrapTooltip = /*#__PURE__*/_styledComponents.default.span(["position:relative;"]);

exports.WrapTooltip = WrapTooltip;
