// @ali/we-toolbar/dist-web/biz/splitButton/BorderSet/index.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _ToolbarBorderOuterNo, _ToolbarBorderInnerNo, _ToolbarBorderAllNorm, _ToolbarBorderLeftNor, _ToolbarBorderRightNo, _ToolbarBorderTopNorm, _ToolbarBorderBottomN, _ToolbarBorderNoneNor, _ToolbarBorderHorizon, _ToolbarBorderVertica, _ToolbarBorderNoneNor2;

import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { IconButtonDefaultProps } from "../../../utils/defaultProps";
import SplitButton from "../../../common/splitButton";
import { CommonLineSelect, ToolbarBorderAllNormal, ToolbarBorderBottomNormal, ToolbarBorderHorizonNormal, ToolbarBorderLeftNormal, ToolbarBorderNoneNormal, ToolbarBorderOuterNormal, ToolbarBorderInnerNormal, ToolbarBorderTopNormal, ToolbarBorderRightNormal, ToolbarBorderVerticalNormal } from '@ali/we-design';
import { TranslateComp } from '@ali/we-util';
import { BorderTypePanel, BorderType } from "./BorderTypePanel";
import { FillColorPicker } from "../colorPicker/FillColorPicker";
import { BorderSetWrapper, LineTypeWrapper, ContentWrapper, FooterWrapper, IconWrapper, TextSpan, FillColorWrapper, LineSelectWrapper, LabelWrapper, FooterInner } from "./styled";
import { classnames, isWhite } from "../../../utils";
var DEFAULT_BORDER_COLOR = '#000000';
var ICON_DEFAULT_COLOR = 'rgba(58, 60, 63, 1)';
export var BorderSet = /*#__PURE__*/function (_React$PureComponent) {
  _inheritsLoose(BorderSet, _React$PureComponent);

  function BorderSet(props) {
    var _this;

    _this = _React$PureComponent.call(this, props) || this;

    _this.handleClick = function (e) {
      // 阻止冒泡
      e.stopPropagation();
    };

    _this.handleVisibleChange = function (visible) {
      _this.setState({
        dropdownVisible: visible
      });
    };

    _this.handleClickLeft = function () {
      var onClick = _this.props.onClick;

      if (onClick) {
        var _this$state = _this.state,
            innerColor = _this$state.innerColor,
            innerLineType = _this$state.innerLineType,
            innerBorderType = _this$state.innerBorderType;
        var data = {
          borderColor: innerColor,
          borderType: innerBorderType,
          borderStyle: innerLineType
        };
        onClick(data);
      }
    };

    _this.handlePropertyChange = function (borderProps) {
      var onChange = _this.props.onChange;

      if (onChange) {
        var _this$state2 = _this.state,
            innerColor = _this$state2.innerColor,
            innerLineType = _this$state2.innerLineType,
            innerBorderType = _this$state2.innerBorderType;
        var data = {
          borderColor: innerColor,
          borderType: innerBorderType,
          borderStyle: innerLineType
        };

        var _allProps = _extends({}, data, borderProps);

        onChange(_allProps, borderProps);
      }
    };

    _this.handleColorChange = function (color) {
      _this.setState({
        innerColor: color
      });

      _this.handlePropertyChange({
        borderColor: color
      });

      _this.setState({
        dropdownVisible: true
      });
    };

    _this.handleLineChange = function (type) {
      _this.setState({
        innerLineType: type
      });

      _this.handlePropertyChange({
        borderStyle: type
      });

      _this.setState({
        dropdownVisible: true
      });
    };

    _this.handleBorderChange = function (type) {
      _this.setState({
        innerBorderType: type
      });

      _this.handlePropertyChange({
        borderType: type
      });
    };

    _this.handleColorPickerVisible = function (visible) {
      _this.setState({
        colorPickerVisible: visible
      });
    };

    _this.handleLineSelectVisible = function (e) {
      _this.setState({
        lineSelectVisible: e
      });
    };

    _this.handleBorderNoneClick = function () {
      return _this.handleBorderChange(BorderType.none);
    };

    _this.getOverlayContent = function () {
      var _this$props = _this.props,
          borderIconSize = _this$props.borderIconSize,
          lineTypes = _this$props.borderStyles,
          color = _this$props.color,
          locale = _this$props.locale,
          showToggleSelect = _this$props.showToggleSelect,
          onSelectValueChange = _this$props.onSelectValueChange,
          showGradient = _this$props.showGradient,
          getColorPanelMode = _this$props.getColorPanelMode,
          doubleGapRowIndex = _this$props.doubleGapRowIndex,
          zIndex = _this$props.zIndex;
      var _this$state3 = _this.state,
          currentColor = _this$state3.innerColor,
          innerBorderType = _this$state3.innerBorderType,
          innerLineType = _this$state3.innerLineType,
          colorPickerVisible = _this$state3.colorPickerVisible,
          lineSelectVisible = _this$state3.lineSelectVisible;
      var activeColor = color || currentColor;
      var inputStyle = {
        width: 63,
        height: 22
      };
      var iconStyle = {
        width: borderIconSize,
        height: borderIconSize,
        boxSizing: 'border-box',
        padding: '1px 1px'
      };
      return /*#__PURE__*/_createElement(BorderSetWrapper, {
        onMouseDown: _this.handleClick
      }, /*#__PURE__*/_createElement(ContentWrapper, null, /*#__PURE__*/_createElement(BorderTypePanel, {
        borderType: innerBorderType,
        borderIconSize: borderIconSize,
        onBorderTypeChange: _this.handleBorderChange,
        borderSelectedColor: activeColor
      }), /*#__PURE__*/_createElement(LineTypeWrapper, null, /*#__PURE__*/_createElement(FillColorWrapper, null, /*#__PURE__*/_createElement(LabelWrapper, null, /*#__PURE__*/_createElement(TranslateComp, {
        locale: locale,
        textKey: 'we_toolbar_border_set_color'
      })), /*#__PURE__*/_createElement(FillColorPicker, {
        active: colorPickerVisible,
        onActiveChange: _this.handleColorPickerVisible,
        color: activeColor,
        automaticColor: DEFAULT_BORDER_COLOR,
        onColorPicker: _this.handleColorChange,
        size: "small",
        locale: locale,
        showToggleSelect: showToggleSelect,
        getColorPanelMode: getColorPanelMode,
        showGradient: showGradient,
        onSelectValueChange: onSelectValueChange,
        doubleGapRowIndex: doubleGapRowIndex,
        zIndex: zIndex
      })), /*#__PURE__*/_createElement(LineSelectWrapper, null, /*#__PURE__*/_createElement(LabelWrapper, null, /*#__PURE__*/_createElement(TranslateComp, {
        locale: locale,
        textKey: 'we_toolbar_border_set_line_type'
      })), /*#__PURE__*/_createElement(CommonLineSelect, {
        lineTypeArray: lineTypes,
        length: 80,
        style: inputStyle,
        lineType: innerLineType // 线型颜色不需要跟随变化
        ,
        color: "#000000",
        onLineTypeChange: _this.handleLineChange,
        visible: lineSelectVisible,
        onVisibleChange: _this.handleLineSelectVisible,
        zIndex: zIndex
      })))), /*#__PURE__*/_createElement(FooterWrapper, null, /*#__PURE__*/_createElement(FooterInner, {
        className: classnames({
          active: innerBorderType === BorderType.none
        }),
        onClick: _this.handleBorderNoneClick
      }, /*#__PURE__*/_createElement(IconWrapper, {
        style: {
          color: ICON_DEFAULT_COLOR,
          backgroundColor: 'transparent'
        }
      }, /*#__PURE__*/_createElement(ToolbarBorderNoneNormal, {
        style: iconStyle
      })), /*#__PURE__*/_createElement(TextSpan, null, /*#__PURE__*/_createElement(TranslateComp, {
        locale: locale,
        textKey: 'we_toolbar_border_none'
      })))));
    };

    _this.renderShowIcon = function () {
      var innerBorderType = _this.state.innerBorderType;

      switch (innerBorderType) {
        case 'outer':
          return _ToolbarBorderOuterNo || (_ToolbarBorderOuterNo = /*#__PURE__*/_createElement(ToolbarBorderOuterNormal, null));

        case 'inner':
          return _ToolbarBorderInnerNo || (_ToolbarBorderInnerNo = /*#__PURE__*/_createElement(ToolbarBorderInnerNormal, null));

        case 'all':
          return _ToolbarBorderAllNorm || (_ToolbarBorderAllNorm = /*#__PURE__*/_createElement(ToolbarBorderAllNormal, null));

        case 'left':
          return _ToolbarBorderLeftNor || (_ToolbarBorderLeftNor = /*#__PURE__*/_createElement(ToolbarBorderLeftNormal, null));

        case 'right':
          return _ToolbarBorderRightNo || (_ToolbarBorderRightNo = /*#__PURE__*/_createElement(ToolbarBorderRightNormal, null));

        case 'top':
          return _ToolbarBorderTopNorm || (_ToolbarBorderTopNorm = /*#__PURE__*/_createElement(ToolbarBorderTopNormal, null));

        case 'bottom':
          return _ToolbarBorderBottomN || (_ToolbarBorderBottomN = /*#__PURE__*/_createElement(ToolbarBorderBottomNormal, null));

        case 'none':
          return _ToolbarBorderNoneNor || (_ToolbarBorderNoneNor = /*#__PURE__*/_createElement(ToolbarBorderNoneNormal, null));

        case 'horizon':
          return _ToolbarBorderHorizon || (_ToolbarBorderHorizon = /*#__PURE__*/_createElement(ToolbarBorderHorizonNormal, null));

        case 'vertical':
          return _ToolbarBorderVertica || (_ToolbarBorderVertica = /*#__PURE__*/_createElement(ToolbarBorderVerticalNormal, null));

        default:
          return _ToolbarBorderNoneNor2 || (_ToolbarBorderNoneNor2 = /*#__PURE__*/_createElement(ToolbarBorderNoneNormal, null));
      }
    };

    var _innerColor = props.color ? props.color : DEFAULT_BORDER_COLOR;

    var _innerBorderType = props.borderType || BorderType.all;

    var _innerLineType = props.defaultBorderStyle || 'single';

    _this.state = {
      innerColor: _innerColor,
      innerBorderType: _innerBorderType,
      innerLineType: _innerLineType,
      dropdownVisible: false,
      colorPickerVisible: false,
      lineSelectVisible: false
    };
    return _this;
  }

  var _proto = BorderSet.prototype;

  _proto.render = function render() {
    var _this$state4 = this.state,
        dropdownVisible = _this$state4.dropdownVisible,
        colorPickerVisible = _this$state4.colorPickerVisible,
        lineSelectVisible = _this$state4.lineSelectVisible,
        color = _this$state4.innerColor,
        innerBorderType = _this$state4.innerBorderType;
    var _this$props2 = this.props,
        disabled = _this$props2.disabled,
        title = _this$props2.title,
        active = _this$props2.active,
        showTitle = _this$props2.showTitle,
        size = _this$props2.size,
        disabledTip = _this$props2.disabledTip,
        testid = _this$props2.testid;
    var overlayVisible = dropdownVisible || colorPickerVisible || lineSelectVisible;
    var pathColors = innerBorderType === 'none' ? [] : [color || ICON_DEFAULT_COLOR, // 当选中颜色为白色时，选中边框的外边缘为黑色，否则选中边框外边缘与内部颜色一致
    !color || isWhite(color) ? ICON_DEFAULT_COLOR : 'transparent'];
    return /*#__PURE__*/_createElement(SplitButton, _extends({}, this.props, {
      overlay: overlayVisible ? this.getOverlayContent() : /*#__PURE__*/_createElement(React.Fragment, null),
      active: overlayVisible,
      onClickLeft: this.handleClickLeft,
      onActiveChange: this.handleVisibleChange,
      disabled: disabled,
      testid: testid || 'BorderSet',
      title: title,
      showTitle: showTitle,
      size: size,
      activeLeft: active,
      icon: /*#__PURE__*/React.cloneElement(this.renderShowIcon(), {
        pathColors: pathColors
      }),
      disabledTip: disabledTip
    }));
  };

  return BorderSet;
}(React.PureComponent);
BorderSet.defaultProps = _extends({
  borderIconSize: 20
}, IconButtonDefaultProps);
