// @ali/we-toolbar/dist-web/biz/splitButton/BorderSet/styled.js
import styled from 'styled-components';
import { withTheme, theme } from '@ali/we-design';
export var BorderSetWrapper = /*#__PURE__*/styled.div(["display:flex;flex-direction:column;justify-content:center;align-items:center;padding:14px 14px 9px 14px;align-items:baseline;user-select:none;"]);
export var BorderTypeWrapper = withTheme(styled.div(["display:grid;grid-template-columns:repeat(3,1fr);grid-gap:2px;justify-items:center;align-items:center;border:1px solid ", ";border-radius:", ";"], function (p) {
  return p.theme.ui.commonLineLightColor;
}, function (p) {
  return p.theme.ui.radius;
}));
export var LineTypeWrapper = /*#__PURE__*/styled.div(["display:flex;flex-direction:column;justify-content:flex-start;height:100px;width:100px;margin-left:16px;margin-right:2px;"]);
export var IconWrapper = withTheme(styled.span(["box-sizing:border-box;border-radius:4px;width:28px;height:28px;background-color:", ";cursor:pointer;display:flex;align-items:center;justify-content:center;&:hover{background-color:", ";}"], function (p) {
  return p.active && p.theme.ui.buttonNormalActiveBgColor;
}, function (p) {
  return p.theme.ui.buttonNormalHoverBgColor;
}));
export var ContentWrapper = /*#__PURE__*/styled.div(["display:flex;flex-direction:row;align-items:center;padding-bottom:14px;"]);
export var FooterWrapper = withTheme(styled.div(["width:100%;padding-top:5px;border-top:1px solid ", ";"], function (p) {
  return p.theme.ui.commonLineLightColor;
}));
export var FooterInner = withTheme(styled.div(["display:flex;flex-direction:row;justify-content:start;align-items:center;width:100%;border-radius:8px;&:hover{background-color:", ";}&.active{background-color:", ";}"], function (p) {
  return p.theme.ui.buttonNormalHoverBgColor;
}, function (_ref) {
  var ui = _ref.theme.ui;
  return ui.buttonNormalActiveBgColor;
}));
export var TextSpan = /*#__PURE__*/styled.span(["font-size:12px;line-height:17px;margin-left:3px;color:", ";cursor:pointer;"], theme.colorText5);
export var FillColorWrapper = /*#__PURE__*/styled.div(["display:flex;flex-direction:row;justify-content:flex-start;align-items:center;margin-top:2px;"]);
export var LineSelectWrapper = /*#__PURE__*/styled.div(["display:flex;flex-direction:row;justify-content:flex-start;align-items:center;margin:12px 0;"]);
export var ThicknessInputWrapper = /*#__PURE__*/styled.div(["display:flex;flex-direction:row;justify-content:flex-start;align-items:center;"]);
export var LabelWrapper = /*#__PURE__*/styled.div(["font-size:12px;line-height:17px;color:", ";width:37px;"], theme.colorText5);
