// @ali/4ever-plugin-vote-card/dist-node/utils/components/result.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _dayjs = _interopRequireDefault(require("dayjs"));

var _actions = require("../actions");

var _types = require("../types");

var _headerToolbar = _interopRequireDefault(require("./headerToolbar"));

var _styled = require("./styled");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement("i", null);

const Result = props => {
  const {
    node,
    locale,
    controller,
    isSelected,
    data,
    readOnly,
    isPublish
  } = props;
  const handleReVote = (0, _react.useCallback)(() => {
    controller.run('onAction', (0, _actions.votingActive)(node));
  }, []); // 我是否有投过票

  const userOptionTicketed = (0, _react.useMemo)(() => !!data.optionInfoList?.find(item => item.userOptionTicket !== 0), [data]);
  return /*#__PURE__*/_createElement(_styled.ContainerWrapper, {
    isSelected: isSelected,
    "data-role": "vote",
    "data-testid": "card-vote-container-result"
  }, /*#__PURE__*/_createElement(_styled.ResultWrapper, null, !controller.readOnly && !readOnly && !isPublish && /*#__PURE__*/_createElement(_headerToolbar.default, {
    controller: controller,
    node: node
  }), /*#__PURE__*/_createElement(_styled.TitleReaderWrapper, null, /*#__PURE__*/_createElement("span", null, data.voteType === _types.EnumVoteType.RADIO ? locale?.radio : locale?.checkbox), /*#__PURE__*/_createElement("span", {
    "data-testid": "card-vote-result-title"
  }, data.title)), /*#__PURE__*/_createElement("div", {
    className: "result-options"
  }, data.optionInfoList?.map(item => /*#__PURE__*/_createElement("div", {
    className: (0, _classnames.default)({
      checked: item.userOptionTicket !== 0
    })
  }, /*#__PURE__*/_createElement("span", null, item.content), /*#__PURE__*/_createElement("span", null, `${String(item.optionTicket)} ${locale?.unit}`, _ref, !item.optionTicket || !data.totalVoteCount ? 0 : ((item.optionTicket ?? 0) / (data.totalVoteCount ?? 0) * 100).toFixed(2), "%"), /*#__PURE__*/_createElement("div", null, /*#__PURE__*/_createElement("div", {
    style: {
      width: `${!item.optionTicket || !data.totalVoteCount ? 0 : ((item.optionTicket ?? 0) / (data.totalVoteCount ?? 0) * 100).toFixed(2)}%`
    }
  }))))), /*#__PURE__*/_createElement("div", {
    className: "result-footer"
  }, /*#__PURE__*/_createElement("span", null, `${String(data.participateUserCount ?? 0)}${locale?.total}`), /*#__PURE__*/_createElement("span", null, (0, _dayjs.default)(data.endTime).format('YYYY-MM-DD HH:mm:ss'), " ", locale?.expired), !data.hasEnded && !isPublish && /*#__PURE__*/_createElement("span", {
    onClick: handleReVote
  }, userOptionTicketed ? locale?.reVote : locale?.submit))));
};

var _default = Result;
exports.default = _default;
