// @ali/4ever-plugin-history/dist-node/bi/components/mobileButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MobileUndoButton = MobileUndoButton;
exports.MobileRedoButton = MobileRedoButton;

var _react = _interopRequireDefault(require("react"));

const _createElement = /*#__PURE__*/_react.default.createElement;

function MobileUndoButton(props) {
  const {
    controller,
    Wrapper,
    item
  } = props;
  const isDisabled = !controller.query('canUndo');

  const handleClick = _react.default.useCallback(() => controller.dispatch('undo'), [controller]);

  return /*#__PURE__*/_createElement(Wrapper, {
    item: item,
    type: "big",
    icon: "undo",
    isDisabled: isDisabled,
    onClick: handleClick
  });
}

function MobileRedoButton(props) {
  const {
    controller,
    Wrapper,
    item
  } = props;
  const isDisabled = !controller.query('canRedo');

  const handleClick = _react.default.useCallback(() => controller.dispatch('redo'), [controller]);

  return /*#__PURE__*/_createElement(Wrapper, {
    item: item,
    type: "big",
    icon: "redo",
    isDisabled: isDisabled,
    onClick: handleClick
  });
}
