// @ali/4ever-plugin-history/dist-node/bi/components/undoButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = UndoButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _everCangjie = require("@ali/4ever-cangjie");

var _role = require("../../utils/role");

const _createElement = /*#__PURE__*/React.createElement;
const SHORTCUTKEY = _everCangjie.environment.IS_MAC ? '⌘+Z' : 'Ctrl+Z';

function UndoButton(props) {
  const {
    controller,
    locale,
    ...rest
  } = props;
  const isDisabled = !controller.query('canUndo');
  const handleClick = React.useCallback(() => controller.dispatch('undo'), [controller]);
  return /*#__PURE__*/_createElement(_weToolbar.UndoButton, (0, _extends2.default)({}, rest, {
    testid: "bi-toolbar-undo",
    disabled: isDisabled,
    onClick: handleClick,
    shortcutText: SHORTCUTKEY,
    role: _role.HistoryRole.undo
  }));
}
