// @ali/we-design/dist-web/components/table-set/TablePanel.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React, { PureComponent } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import warning from 'warning';
import { isMobile } from "../../utils/dom";
import { Table, Title, TableWrapper } from "./styled";
export var TablePanel = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(TablePanel, _PureComponent);

  // eslint-disable-next-line react/static-property-placement
  // eslint-disable-next-line react/state-in-constructor
  function TablePanel(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.state = {
      currentRow: 0,
      currentCol: 0
    };

    _this.onMouseLeave = function () {
      _this.setState({
        currentRow: 0,
        currentCol: 0
      });
    };

    _this.pick = function () {
      var _this$state = _this.state,
          currentRow = _this$state.currentRow,
          currentCol = _this$state.currentCol; // eslint-disable-next-line react/destructuring-assignment

      _this.props.onPick == null ? void 0 : _this.props.onPick(currentRow, currentCol);
    };

    _this.state = {
      currentRow: props.selectedRow || 0,
      currentCol: props.selectedCol || 0
    };
    return _this;
  }

  var _proto = TablePanel.prototype;

  _proto.render = function render() {
    var _this2 = this;

    var _this$props = this.props,
        row = _this$props.row,
        col = _this$props.col,
        selectedCol = _this$props.selectedCol,
        selectedRow = _this$props.selectedRow,
        onChange = _this$props.onChange;
    var currentRow = selectedRow || this.state.currentRow;
    var currentCol = selectedCol || this.state.currentCol;

    if (!row || !col) {
      warning('Please define the row and col properties.');
      return null;
    }

    var vRows = [];

    var _loop = function _loop(r) {
      var vCols = [];

      var _loop2 = function _loop2(c) {
        var props = {
          key: "row-" + r + "::col-" + c,
          className: r < currentRow && c < currentCol ? 'active' : '',
          onMouseEnter: function onMouseEnter() {
            _this2.setState({
              currentRow: r + 1,
              currentCol: c + 1
            });

            onChange == null ? void 0 : onChange(r + 1, c + 1);
          }
        };
        vCols.push( /*#__PURE__*/_createElement("td", props));
      };

      for (var c = 0; c < col; c += 1) {
        _loop2(c);
      }

      vRows.push( /*#__PURE__*/_createElement("tr", {
        key: "row-" + r
      }, vCols));
    };

    for (var r = 0; r < row; r += 1) {
      _loop(r);
    }

    return /*#__PURE__*/_createElement(TableWrapper, null, !isMobile && /*#__PURE__*/_createElement(Title, null, currentRow + " \xD7 " + currentCol), /*#__PURE__*/_createElement(Table, {
      onMouseLeave: this.onMouseLeave,
      onClick: this.pick,
      "data-testid": "table-TableSet"
    }, /*#__PURE__*/_createElement("tbody", null, vRows)));
  };

  return TablePanel;
}(PureComponent);
TablePanel.defaultProps = {
  row: 10,
  col: 10,
  selectedRow: 0,
  selectedCol: 0
};
