// @ali/4ever-bi/dist-node/plugins/vertAlign/components/subButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SubButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _everBamboo = require("@ali/4ever-bamboo");

var _everComponent = require("@ali/4ever-component");

var _actions = require("../actions");

var _utils = require("../utils");

const _createElement = /*#__PURE__*/React.createElement;
;
;

function SubButton(props) {
  const {
    controller,
    locale,
    ...rest
  } = props;
  const isActive = controller.query('isSub');
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInMention') || controller.query('isSelectionInLink') || controller.query('isSelectionInListSymbol') || controller.query('isSelectionInLinkCard');
  const handleClick = React.useCallback(() => controller.run('onAction', (0, _actions.toggleSub)()), [controller]);
  const title = locale?.subButtonTitle || '';
  const shortCut = _utils.SHORTCUT.sub;
  const description = `Markdown: ~${locale?.text || 'Text'}~ + ${locale?.space || 'Space'}`;
  return /*#__PURE__*/_createElement(_weToolbar.SubscriptButton, (0, _extends2.default)({}, rest, {
    shortcutText: _utils.SHORTCUT.sub,
    testid: "bi-toolbar-sub",
    active: isActive,
    disabled: !!isDisabled,
    onClick: handleClick,
    role: _everBamboo.PluginRoles.vertAlignSub,
    tooltip: /*#__PURE__*/_createElement(_everComponent.IconButtonShortcut, {
      title: title,
      shortcut: shortCut,
      description: description
    })
  }));
}
