// @ali/4ever-bi/dist-node/plugins/vertAlign/components/supButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SupButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _everBamboo = require("@ali/4ever-bamboo");

var _everComponent = require("@ali/4ever-component");

var _actions = require("../actions");

var _utils = require("../utils");

const _createElement = /*#__PURE__*/React.createElement;
;
;

function SupButton(props) {
  const {
    controller,
    locale,
    ...rest
  } = props;
  const isActive = controller.query('isSup');
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInMention') || controller.query('isSelectionInLink') || controller.query('isSelectionInListSymbol') || controller.query('isSelectionInLinkCard');
  const handleClick = React.useCallback(() => controller.run('onAction', (0, _actions.toggleSup)()), [controller]);
  const title = locale?.supButtonTitle || '';
  const shortCut = _utils.SHORTCUT.sup;
  const description = `Markdown: ^${locale?.text || 'Text'}^ + ${locale?.space || 'Space'}`;
  return /*#__PURE__*/_createElement(_weToolbar.SuperscriptButton, (0, _extends2.default)({}, rest, {
    shortcutText: _utils.SHORTCUT.sup,
    testid: "bi-toolbar-sup",
    active: isActive,
    disabled: !!isDisabled,
    onClick: handleClick,
    role: _everBamboo.PluginRoles.vertAlignSup,
    tooltip: /*#__PURE__*/_createElement(_everComponent.IconButtonShortcut, {
      title: title,
      shortcut: shortCut,
      description: description
    })
  }));
}
