// @ali/4ever-bi/dist-node/plugins/letterSpacing/components/letterSpacingMenu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = LetterSpacingMenu;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everMo = require("@ali/4ever-mo");

var _weDesign = require("@ali/we-design");

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../actions");

var _style = require("./style");

const _createElement = /*#__PURE__*/React.createElement;
const MIN = -10;
const MAX = 10;
const STEP = 0.5;

function canSetMark(controller) {
  const {
    isCollapsedAtEdge,
    isExpanded
  } = controller.value.selection;
  if (isCollapsedAtEdge || isExpanded) return true;
  const {
    startBlock,
    startInline
  } = controller.value;

  if (startInline && !_everMo.Link.isLink(startInline)) {
    // 仅支持 link
    return false;
  }

  if (_everMo.Paragraph.isParagraph(startBlock) || _everMo.Heading.isHeading(startBlock)) {
    // 仅支持文字段落
    return true;
  }

  return false;
}

var _ref = /*#__PURE__*/_createElement(_weDesign.MinusBetaNormal, null);

var _ref2 = /*#__PURE__*/_createElement(_weDesign.PlusBetaNormal, null);

function LetterSpacingMenu(props) {
  const {
    menuItem,
    controller
  } = props;
  const {
    title,
    ...rest
  } = menuItem;
  const isDisabled = !canSetMark(controller);
  const letterSpacing = controller.query('getLetterSpacing');

  const handleMinusClick = () => {
    const spacing = letterSpacing - STEP;
    if (spacing < MIN) return;
    controller.run('onAction', (0, _actions.setLetterSpacing)(spacing));
  };

  const handlePlusClick = () => {
    const spacing = letterSpacing + STEP;
    if (spacing > MAX) return;
    controller.run('onAction', (0, _actions.setLetterSpacing)(spacing));
  };

  return /*#__PURE__*/_createElement(_weDesign.SubMenu, (0, _extends2.default)({}, rest, {
    title: title,
    disabled: isDisabled
  }), /*#__PURE__*/_createElement(_style.SpacingSet, null, /*#__PURE__*/_createElement(_weDesign.Button, {
    icon: _ref,
    style: _style.commonButtonStyle,
    bordered: false,
    onClick: handleMinusClick,
    role: _everBamboo.PluginRoles.letterSpacingMinus
  }), /*#__PURE__*/_createElement(_style.SpacingValue, null, letterSpacing, "pt"), /*#__PURE__*/_createElement(_weDesign.Button, {
    icon: _ref2,
    style: _style.commonButtonStyle,
    bordered: false,
    onClick: handlePlusClick,
    role: _everBamboo.PluginRoles.letterSpacingPlus
  })));
}
