// @ali/4ever-utils/dist-web/changeNullSelectionMark.js
import { Commands, Mark } from '@ali/4ever-cangjie';
export default function changeNullSelectionMark(controller, markType, currentMark, ranges, document) {
  var selection = controller.value.selection;
  var documentMarks = document.getMarksAtRange(ranges).filter(function (mark) {
    return mark.type === markType;
  });
  /* 删除document的mark */

  if (documentMarks.length) {
    documentMarks.forEach(function (mark) {
      controller.command(Commands.removeMark, mark);
    });
  }
  /* 删除selection的mark */


  if (selection.marks) {
    var marks = selection.marks;
    marks.forEach(function (mark) {
      if (mark.type === markType) {
        controller.command(Commands.removeMark, mark);
      }
    });
  }

  if (markType === 'fonts') {
    controller.command(Commands.addMark, Mark.create({
      type: markType,
      data: {
        ascii: currentMark,
        hAnsi: currentMark,
        cs: currentMark,
        eastAsia: currentMark
      }
    }));
  } else if (markType === 'sz') {
    controller.command(Commands.addMark, Mark.create({
      type: markType,
      data: {
        value: currentMark,
        szUnit: 'pt'
      }
    }));
  } else if (markType === 'highlight') {
    if (currentMark && typeof currentMark === 'string') {
      controller.command(Commands.addMark, Mark.create({
        type: markType,
        data: {
          value: currentMark
        }
      }));
    }
  } else {
    controller.command(Commands.addMark, Mark.create({
      type: markType,
      data: {
        value: currentMark
      }
    }));
  }

  return controller;
}
