// @ali/4ever-utils/dist-web/cssUtils.js
import _extends from "@babel/runtime/helpers/extends";
import { TinyColor } from '@ctrl/tinycolor';

function assertNever(x) {
  console.warn('Unexpected value:', x);
}
/**
 * 效验是否符合 css 长度单位标准
 * @param str
 */


export function isCssLength(str) {
  var length = /^(\+|-)?([0-9]*\.)?[0-9]+(em|ex|ch|rem|vh|vw|vmin|vmax|px|mm|cm|in|pt|pc|%)$/i;
  var zero = /^(\+|-)?(0*\.)?0+$/;
  return length.test(str) || zero.test(str);
}
/**
 * 效验是否服务 css width 值标准
 * @param str
 */

export function isWidth(str) {
  return isCssLength(str) || str === 'thin' || str === 'medium' || str === 'thick';
}
/**
 * 效验是否是 css border style
 * @param str
 */

export function isBorderStyle(str) {
  var value = /^(none|hidden|dotted|dashed|solid|double|groove|ridge|inset|outset)$/i;
  return value.test(str);
}

function checkCssBorder(val) {
  if (new TinyColor(val).isValid) return 'color';
  if (isWidth(val)) return 'width';
  if (isBorderStyle(val)) return 'style';
  return null;
}

export function cssBorderParse(str) {
  var value = str.split(' ');
  var ret = {};
  value.forEach(function (v) {
    var prop = checkCssBorder(v);

    if (prop) {
      ret[prop] = v;
    }
  });
  return ret;
}

function isValidBorder(value) {
  if (value === 'single') return true;
  if (value === 'dashed') return true;
  return false;
}

export function convertBorderToData(parsed) {
  var sz = 1;

  if (parsed.width) {
    var w = parseFloat(parseCSSAbsoluteLength(parsed.width));
    sz = Math.round(w) || Math.ceil(w);
  }

  return {
    val: isValidBorder(parsed.style) ? parsed.style : 'single',
    sz: sz,
    color: "#" + new TinyColor(parsed.color).toHex()
  };
}
/**
 * 解析 css 的 background 属性（目前只支持 color）
 * @param value
 */

export function praseCSSBackground(value) {
  var values = value.split(' ');
  return values.reduce(function (prev, val) {
    if (new TinyColor(val).isValid) {
      return _extends({}, prev, {
        backgroundColor: val
      });
    }

    return prev;
  }, {});
}

/**
 * 解析 css 边距单位
 * @param value
 * @see https://developer.mozilla.org/en-US/docs/Web/CSS/padding
 */
export function parseCSSSides(value) {
  // 解析前先去除多余的空白符
  var sides = value.replace(/\s+/g, ' ').split(' ');

  if (sides.length === 1) {
    var side = sides[0];
    return {
      top: side,
      right: side,
      bottom: side,
      left: side
    };
  }

  if (sides.length === 2) {
    var vertical = sides[0],
        horizontal = sides[1];
    return {
      top: vertical,
      right: horizontal,
      bottom: vertical,
      left: horizontal
    };
  }

  if (sides.length === 3) {
    var top = sides[0],
        _horizontal = sides[1],
        bottom = sides[2];
    return {
      top: top,
      right: _horizontal,
      bottom: bottom,
      left: _horizontal
    };
  }

  if (sides.length === 4) {
    var _top = sides[0],
        right = sides[1],
        _bottom = sides[2],
        left = sides[3];
    return {
      top: _top,
      right: right,
      bottom: _bottom,
      left: left
    };
  }

  return {
    top: '',
    right: '',
    bottom: '',
    left: ''
  };
}
/**
 * 解析 css margin 参数
 * @param value
 */

export function parseCSSMargin(value) {
  var sides = parseCSSSides(value);
  var top = sides.top,
      right = sides.right,
      bottom = sides.bottom,
      left = sides.left;
  return {
    marginTop: top,
    marginRight: right,
    marginBottom: bottom,
    marginLeft: left
  };
}
/**
 * 解析 css padding 参数
 * @param value
 */

export function parseCSSPadding(value) {
  var sides = parseCSSSides(value);
  var top = sides.top,
      right = sides.right,
      bottom = sides.bottom,
      left = sides.left;
  return {
    paddingTop: top,
    paddingRight: right,
    paddingBottom: bottom,
    paddingLeft: left
  };
}
/**
 * 效验一个值是否符合数字规范
 * @param num 效验值
 * @see https://stackoverflow.com/questions/175739/built-in-way-in-javascript-to-check-if-a-string-is-a-valid-number
 * @description 注意一下 lodash 的 isNaN 和浏览器 API 有点不同，12px 这个值在 lodash 中会认为不是 NaN
 */

export function isNumeric(num) {
  return !isNaN(Number(num));
}

function inchesToUnit(num, to) {
  switch (to) {
    case 'cm':
      return num * 2.54;

    case 'mm':
      return num * 10 * 2.54;

    case 'Q':
      return num * 40 * 2.54;

    case 'in':
      return num;

    case 'pc':
      return num * 6;

    case 'pt':
      return num * 72;

    case 'px':
      return num * 96;

    default:
      assertNever(to);
      break;
  }

  return num;
}

function unitToInches(num, from) {
  switch (from) {
    case 'cm':
      return num / 2.54;

    case 'mm':
      return num / 10 / 2.54;

    case 'Q':
      return num / 40 / 2.54;

    case 'in':
      return num;

    case 'pc':
      return num / 6;

    case 'pt':
      return num / 72;

    case 'px':
      return num / 96;

    default:
      assertNever(from);
      break;
  }

  return num;
}

var ABS_L_REG = /^(\d*\.?\d*)(cm|mm|Q|in|pc|pt|px)$/;
/**
 * 测试一个值是否是绝对单位
 * @param value
 */

export function isCSSAbsoluteLength(value) {
  var length = isNumeric(value || '') ? value + "px" : "" + value;
  return ABS_L_REG.test("" + length);
}
/**
 * 解析 CSS 绝对长度单位，默认的目标单位是 px，value 值若为 number 则断言为 px 单位
 * @param value 原始值
 * @param unit 目标单位
 * @see https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#Lengths
 * @example 1in = 2.54cm = 96px
 *
 * Unit |  Name                 | Equivalent to
 *  cm  |  Centimeters          | 1cm = 96px/2.54
 *  mm  |  Millimeters          | 1mm = 1/10th of 1cm
 *  Q   |  Quarter-millimeters  | 1Q  = 1/40th of 1cm
 *  in  |  Inches               | 1in = 2.54cm = 96px
 *  pc  |  Picas                | 1pc = 1/6th of 1in
 *  pt  |  Points               | 1pt = 1/72th of 1in
 *  px  |  Pixels               | 1px = 1/96th of 1in
 */

export function parseCSSAbsoluteLength(value, unit) {
  if (unit === void 0) {
    unit = 'px';
  }

  var length = isNumeric(value) ? value + "px" : "" + value;
  var match = length.match(ABS_L_REG);

  if (match) {
    // [原始值, 数字组, 单位组]
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    var _ref = match,
        _ = _ref[0],
        n = _ref[1],
        u = _ref[2];
    var num = parseFloat(n);

    if (!isNaN(num)) {
      return inchesToUnit(unitToInches(num, u), unit) + unit;
    }
  }

  return 0 + unit;
}
var PERCENT_REG = /^(\d*\.?\d*)(%)$/;
/**
 * 测试一个值是否是 css 百分比单位
 * @param value
 */

export function isCSSPercent(value) {
  return PERCENT_REG.test(value);
}
/**
 * 解析 css 百分比单位
 * @param value
 */

export function parseCSSPercent(value) {
  var match = value.match(PERCENT_REG);

  if (match) {
    // [原始值, 数字组, 百分号]
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    var _ref2 = match,
        _ = _ref2[0],
        n = _ref2[1];
    var num = parseFloat(n);

    if (!isNaN(num)) {
      return num / 100;
    }
  }

  return NaN;
}
