// @ali/4ever-utils/dist-web/getInnerWidth.js
import { domUtils } from '@ali/4ever-cangjie';
export var isList = function isList(node) {
  var _node$data;

  return !!(node != null && (_node$data = node.data) != null && _node$data.list);
};
export var isOrderedList = function isOrderedList(node) {
  var _node$data2, _node$data2$list;

  return !!((_node$data2 = node.data) != null && (_node$data2$list = _node$data2.list) != null && _node$data2$list.isOrdered);
};
export var isTaskList = function isTaskList(node) {
  var _node$data3;

  return ((_node$data3 = node.data) == null ? void 0 : _node$data3.list) && !node.data.list.isOrdered && node.data.list.isTaskList;
};
export var isUnorderedList = function isUnorderedList(node) {
  var _node$data4;

  return ((_node$data4 = node.data) == null ? void 0 : _node$data4.list) && !isOrderedList(node) && !isTaskList(node);
};
export function getInnerWidth(node, root) {
  if (!node) {
    return null;
  }

  var domNode = null; // HACK: 排版模式通过key-0获取dom元素

  try {
    domNode = domUtils.findDOMNode(node.key, root);
  } catch (e) {// do nothing
  }

  if (!domNode) {
    try {
      domNode = domUtils.findDOMNode(node.key + "-0", root);
    } catch (e) {// do nothing
    }
  }

  if (!domNode) {
    return null;
  }

  return getInnerWidthFromDOM(domNode);
} // 获取内部撑开的宽度，主要处理依赖父元素宽度切默认宽度是 100% 的节点(如图片、表格)

export function getInnerFullWidth(node, document, root) {
  if (!node) return null;
  var innerWidth = getInnerWidth(node, root);

  if (isOrderedList(node) || isUnorderedList(node)) {
    var parentNode = document.getParent(node.key);

    if (parentNode) {
      var parentWidth = getInnerWidth(parentNode, root); // 列表节点在无内容时宽度为零，则宽度以它的父节点为准
      // 内容为空时，由于 placeholder 存在，所以 inner width 不为 0，但是未被完全撑开

      if (innerWidth === 0 || (innerWidth || 0) < (parentWidth || 0)) {
        innerWidth = parentWidth;
      }
    }

    return innerWidth;
  }

  if (innerWidth === 0) {
    var _parentNode = document.getParent(node.key);

    if (!_parentNode) return null;
    innerWidth = getInnerWidth(_parentNode, root);
  }

  return innerWidth;
}
export function getInnerWidthFromDOM(domNode) {
  // TODO: 排版模式下 document 节点的宽度计算偏大，会导致空列表中插入的图片尺寸偏大溢出
  if (domNode.getAttribute('data-type') === 'page') {
    domNode = domNode.childNodes[0];
  }

  var computedStyle = window.getComputedStyle(domNode, null);
  var offsetWidth = domNode.offsetWidth;
  var paddingLeft = parseInt(computedStyle["padding-left"], 10) || 0;
  var paddingRight = parseInt(computedStyle["padding-right"], 10) || 0;
  var width = offsetWidth - paddingLeft - paddingRight;
  return width;
}
