// @ali/4ever-utils/dist-web/getRectOfNodeFirstText.js
import { Block } from '@ali/4ever-cangjie';
import { getClosestDom } from "./getClosestDom";
var HEADING_TYPE = new Set(['heading-1', 'heading-2', 'heading-3', 'heading-4', 'heading-5', 'heading-6']);

function isHeading(node) {
  return Block.isBlock(node) && HEADING_TYPE.has(node.type);
}

function isParagraph(node) {
  return Block.isBlock(node) && node.type === 'paragraph';
}
/**
 * 获取 node 节点中第一个文本节点位置，找不到则返回 node 位置
 * @param node
 */


export default function getRectOfNodeFirstText(node) {
  var rect;

  try {
    var nodeDom = getClosestDom(node);
    if (!nodeDom) return rect;
    var nodeDomRect = nodeDom.getBoundingClientRect();

    if (!nodeDomRect) {
      return null;
    }

    rect = {
      top: nodeDomRect.top,
      left: nodeDomRect.left
    };

    if (isHeading(node) || isParagraph(node)) {
      // 找到第一个文字子节点 dom
      var firstTextChild = nodeDom.querySelector('[data-cangjie-leaf]');

      if (firstTextChild) {
        var range = document.createRange();
        range.setStart(firstTextChild, 0);
        range.setEnd(firstTextChild, 1);
        var textRect = range.getClientRects()[0];
        rect = {
          top: textRect.top,
          left: nodeDomRect.left,
          height: textRect.height
        };
      }
    }
  } catch (e) {
    console.error(e);
  }

  return rect;
}
