// @ali/4ever-utils/dist-web/index.js
// 和 mo 有相互嵌套引用的关系，在 mo 的 import 时就有调用，需要保持在前面
export { default as createLinkify, matchesLink } from "./linkify";
export * from "./objectUtils";
export { loadImg } from "./loadImg";
export { default as sanitizeUrl } from "./sanitizeUrl";
export { default as getRectOfNodeFirstText } from "./getRectOfNodeFirstText";
export { default as inspectScroller } from "./inspectScroller";
export * from "./getClosestDom";
export * from "./calcRotation";
export * from "./getClosestContainerBlock";
export * from "./getInnerWidth";
export * from "./getInnerWidth";
export * from "./generateStyle";
export * from "./idleCallback";
export * from "./relativePosition";
export * from "./hr/mapping";
export * from "./idleCallback";
export * from "./cssUtils";
export * from "./generateStyle";
export * from "./aspect/latency";
export * from "./pluginsUtil";
export * from "./sectionDataUtils";
export * from "./jsonml";
export * from "./matcher";
export * from "./sortPlugins";
export * from "./heading";
export * from "./regenerateUuid";
export * from "./indent"; // theme from 'mo/theme'

export * from "./theme"; // utils from 'mo/utils'

export { default as throttleByIdleCallback } from "./utils/throttleByIdleCallback";
export { default as copyToClipboard } from "./utils/copyToClipboard";
export { default as UnitConvert } from "./utils/unitConvert";
export { default as isFirstPart } from "./utils/isFirstPart";
export { addDefaultUnit, transformThemeObjToCssString } from "./utils/unit";
export { isNaN, ptToPx, unitsConverter } from "./utils/toUnits";
export { sample, Sample } from "./utils/sample";
export { default as SimpleMapCache } from "./utils/simpleMapCache";
export { default as fontRenderFallback } from "./utils/fontRenderFallback";
export { default as ShdUtil } from "./utils/formats/shd";
export { default as UnderlineUtil } from "./utils/formats/underline";
export { default as calculatedPluginSize } from "./utils/calculatedPluginSize";
export { SpacingUtil } from "./utils/formats/spacing";
export { IndUtil, DEFAULT_INDENT, IndentTrigger, IndentType } from "./utils/formats/ind";
export { SzUtil } from "./utils/formats/sz";
export { CommonFormatUtil } from "./utils/formats/common";
export { default as GradientUtil, isGradientColor, transferGradientColor, shouldDegradeGradient, setGradientStyle, GRADIENT_COLOR_STYLE, GRADIENT_STYLE_FLAG } from "./utils/formats/gradientColor";
export { default as Measure, CJK_REG } from "./utils/measure";
export { default as isEmptyParagraph } from "./utils/isEmptyParagraph";
export { isEmojiClsName } from "./utils/isEmojiClsName"; // utils from 'bamboo/utils or bamboo/plugins/xx but is not plugin'

export { zIndex } from "./zIndex";
export { default as HoverBlock } from "./hoverBlock";
export { isSingleTableCell, unwrapSingleTableCell } from "./utils/table/singleTableCell"; // utils from 'zhi/utils'

export { getData } from "./utils/getData";
export { isEmptyValue } from "./utils/isEmptyValue"; // utils from 'bi/utils'

export { default as addMarkAtCurrentRange, addMarkAtRange } from "./utils/editor/addMarkAtCurrentRange";
export { SELECTION_TRIGGER, DEFAULT_PAGE_WIDTH } from "./utils/editor/constants";
export { default as generateRangesFromTableSelection } from "./utils/editor/generateRangesFromTableSelection";
export { getBlocksFromEditor } from "./utils/editor/getBlocksFromEditor";
export { getRangesFromEditor } from "./utils/editor/getRangesFromEditor";
export { removeMarkAtRange, removeMarkAtCurrentRange } from "./utils/editor/removeMarkAtCurrentRange";
export { blockEvent } from "./utils/blockEvent";
export { createStore } from "./utils/createStore";
export { LEGAL_EVENT, getCustomEvent, dispatchCustomEvent } from "./utils/customEventUtils";
export { isEqualWithDefault } from "./utils/isEqualWithDefault";
export { getKeyword as getQuickInsertKeyword } from "./getQuickInsertKeyword";
export { APPLICATION_CANGJIE_FRAGMENT, APPLICATION_SLATE_FRAGMENT, APPLICATION_DING_SHEET, APPLICATION_CANGJIE_DRAG_FRAGMENT, APPLICATION_GOOGLE_DOCS_CLIP, APPLICATION_GOOGLE_SHEET_CLIP, APPLICATION_VSCODE, TEXT_PLAIN, TEXT_HTML, TEXT_LINK, TEXT_RTF } from "./utils/mime-types";
import * as _mimeTypes from "./utils/mime-types";
export { _mimeTypes as mimeTypes };
export { isPlaceholderVisible } from "./utils/placeholder";
export { toHash } from "./utils/toHash";
export { uniqueId } from "./utils/uniqueId";
export * from "./utils/modelData";
export { default as paragraphTags } from "./utils/paragraphTags"; // utils from bi/src/plugins/common

export { default as changeNullSelectionMark } from "./changeNullSelectionMark";
export { default as generateKey } from "./generateKey";
export { default as getClosestParagraph } from "./getClosestParagraph";
export { default as getRandomId } from "./getRandomId";
export { default as hasMark } from "./hasMark";
export { default as isCollapsedInEmptyText } from "./isCollapsedInEmptyText";
export { default as isInsertInline } from "./isInsertInline";
export { default as createSpaceHandler } from "./createSpaceHandler";
export { re, getMatches, getOffset, getMarkTextMatchInfo } from "./biUtils";
export { getPointAndBlockOnSelection, isSelectionAtSameBlock, isAtStartOfParagraph, isAtEndOfParagraph } from "./isParagraph"; // from bi/components/toobar/utils

export { isSelectedBlock } from "./isSelectedBlock"; // actions 待插件迁移完成删除

export * from "./pluginTmp/actions";
export { default as PerfResizeObserver } from "./perfResizeObserver"; // 获取字符串模糊查询的正则

export { getFuzzyQuery, fuzzySearch } from "./fuzzyQuery";
