// @ali/4ever-utils/dist-web/isParagraph.js
/**
 * @description 获取选区开始与结束位置信息
 */
export function getPointAndBlockOnSelection(controller) {
  var _controller$value = controller.value,
      rangeSelection = _controller$value.rangeSelection,
      document = _controller$value.document;
  var startPoint = rangeSelection.getStart(document);
  var startBlock = document.getNode(startPoint.key);
  var endPoint = rangeSelection.getEnd(document);
  var endBlock = document.getNode(endPoint.key);
  return {
    startPoint: startPoint,
    startBlock: startBlock,
    endPoint: endPoint,
    endBlock: endBlock
  };
}
/**
 * @description 获取选区是否在同个节点上
 */

export function isSelectionAtSameBlock(controller) {
  var _getPointAndBlockOnSe = getPointAndBlockOnSelection(controller),
      startBlock = _getPointAndBlockOnSe.startBlock,
      endBlock = _getPointAndBlockOnSe.endBlock;

  return startBlock === endBlock;
}
/**
 * @description 获取选区是否在段落开始位置
 */

export function isAtStartOfParagraph(controller) {
  var _getPointAndBlockOnSe2 = getPointAndBlockOnSelection(controller),
      startBlock = _getPointAndBlockOnSe2.startBlock,
      startPoint = _getPointAndBlockOnSe2.startPoint;

  return !!(startBlock && startPoint.isAtStartOfNode(startBlock));
}
/**
 * @description 获取选区是否在段落结束位置
 */

export function isAtEndOfParagraph(controller) {
  var _getPointAndBlockOnSe3 = getPointAndBlockOnSelection(controller),
      endBlock = _getPointAndBlockOnSe3.endBlock,
      endPoint = _getPointAndBlockOnSe3.endPoint;

  return !!(endBlock && endPoint.isAtEndOfNode(endBlock));
}
