// @ali/4ever-utils/dist-web/relativePosition.js
import _extends from "@babel/runtime/helpers/extends";
export function getBoundingRelativeRect(el, refEl) {
  var ref = refEl || document.body;
  var elRect = el.getBoundingClientRect();
  var refElRect = ref.getBoundingClientRect();
  var left = elRect.left - refElRect.left;
  var top = elRect.top - refElRect.top;
  var width = elRect.width,
      height = elRect.height;
  return {
    left: left,
    top: top,
    right: left + width,
    bottom: top + height,
    width: width,
    height: height
  };
}
export function getRelativeMouseEvent(event, refEl) {
  var clientX = event.clientX,
      clientY = event.clientY;
  var refElRect = refEl.getBoundingClientRect();
  var relativeX = clientX - refElRect.left;
  var relativeY = clientY - refElRect.top;
  return _extends({}, event, {
    clientX: clientX,
    clientY: clientY,
    relativeX: relativeX,
    relativeY: relativeY
  });
}
export function getRelativeTouchEvent(touch, refEl) {
  var clientX = touch.clientX,
      clientY = touch.clientY;
  var refElRect = refEl.getBoundingClientRect();
  var relativeX = clientX - refElRect.left;
  var relativeY = clientY - refElRect.top;
  return {
    clientX: clientX,
    clientY: clientY,
    relativeX: relativeX,
    relativeY: relativeY
  };
}
