// @ali/4ever-utils/dist-web/sectionDataUtils.js
import _extends from "@babel/runtime/helpers/extends";

/**
 * 纸张的属性
 */
var defaultSectPr = {
  pgSz: {
    h: 1122.67,
    w: 793.33,
    orient: 'portrait'
  },
  pgMar: {
    top: 96,
    bottom: 96,
    left: 72,
    right: 72,
    header: 56.73,
    footer: 66.13,
    gutter: 0
  }
};
/**
 * Pi-1.0 使用
 * @returns
 * @deprecated
 */

export var getSectPrFromJson = function getSectPrFromJson(data) {
  if (!data.sectPr) return defaultSectPr;
  var pgMar = defaultSectPr.pgMar;
  var pgSz = defaultSectPr.pgSz;

  var _ref = data.sectPr.pgSz || {},
      _ref$w = _ref.w,
      w = _ref$w === void 0 ? pgSz.w : _ref$w,
      _ref$h = _ref.h,
      h = _ref$h === void 0 ? pgSz.h : _ref$h,
      _ref$orient = _ref.orient,
      orient = _ref$orient === void 0 ? pgSz.orient : _ref$orient;

  var _ref2 = data.sectPr.pgMar || {},
      _ref2$header = _ref2.header,
      header = _ref2$header === void 0 ? pgMar.header : _ref2$header,
      _ref2$footer = _ref2.footer,
      footer = _ref2$footer === void 0 ? pgMar.footer : _ref2$footer,
      _ref2$top = _ref2.top,
      top = _ref2$top === void 0 ? pgMar.top : _ref2$top,
      _ref2$bottom = _ref2.bottom,
      bottom = _ref2$bottom === void 0 ? pgMar.bottom : _ref2$bottom,
      _ref2$left = _ref2.left,
      left = _ref2$left === void 0 ? pgMar.left : _ref2$left,
      _ref2$right = _ref2.right,
      right = _ref2$right === void 0 ? pgMar.right : _ref2$right,
      _ref2$gutter = _ref2.gutter,
      gutter = _ref2$gutter === void 0 ? pgMar.gutter : _ref2$gutter;

  var defaultFooterId;
  var defaultHeaderId;

  if (data.sectPr.footerRefs) {
    defaultFooterId = data.sectPr.footerRefs.defaultRefId;
  }

  if (data.sectPr.headerRefs) {
    defaultHeaderId = data.sectPr.headerRefs.defaultRefId;
  }

  return {
    defaultHeaderId: defaultHeaderId,
    defaultFooterId: defaultFooterId,
    pgMar: {
      header: header,
      footer: footer,
      top: top,
      bottom: bottom,
      left: left,
      right: right,
      gutter: gutter
    },
    pgSz: {
      w: w,
      h: h,
      orient: orient
    }
  };
}; // Get sectPr from document.data

export var getSectPr = function getSectPr(documentData, field) {
  if (field === void 0) {
    field = '';
  }

  var sectPr = documentData.sectPr;

  if (!sectPr) {
    if (field && defaultSectPr[field]) {
      return defaultSectPr[field];
    }

    return defaultSectPr;
  }

  var pgSz = defaultSectPr.pgSz,
      pgMar = defaultSectPr.pgMar;

  if (sectPr.pgSz) {
    var _sectPr$pgSz = sectPr.pgSz,
        _sectPr$pgSz$w = _sectPr$pgSz.w,
        w = _sectPr$pgSz$w === void 0 ? pgSz.w : _sectPr$pgSz$w,
        _sectPr$pgSz$h = _sectPr$pgSz.h,
        h = _sectPr$pgSz$h === void 0 ? pgSz.h : _sectPr$pgSz$h,
        _sectPr$pgSz$orient = _sectPr$pgSz.orient,
        orient = _sectPr$pgSz$orient === void 0 ? pgSz.orient : _sectPr$pgSz$orient;
    pgSz = {
      w: w,
      h: h,
      orient: orient
    };
  }

  if (sectPr.pgMar) {
    var _sectPr$pgMar = sectPr.pgMar,
        _sectPr$pgMar$top = _sectPr$pgMar.top,
        top = _sectPr$pgMar$top === void 0 ? pgMar.top : _sectPr$pgMar$top,
        _sectPr$pgMar$bottom = _sectPr$pgMar.bottom,
        bottom = _sectPr$pgMar$bottom === void 0 ? pgMar.bottom : _sectPr$pgMar$bottom,
        _sectPr$pgMar$left = _sectPr$pgMar.left,
        left = _sectPr$pgMar$left === void 0 ? pgMar.left : _sectPr$pgMar$left,
        _sectPr$pgMar$right = _sectPr$pgMar.right,
        right = _sectPr$pgMar$right === void 0 ? pgMar.right : _sectPr$pgMar$right,
        _sectPr$pgMar$header = _sectPr$pgMar.header,
        header = _sectPr$pgMar$header === void 0 ? pgMar.header : _sectPr$pgMar$header,
        _sectPr$pgMar$footer = _sectPr$pgMar.footer,
        footer = _sectPr$pgMar$footer === void 0 ? pgMar.footer : _sectPr$pgMar$footer,
        _sectPr$pgMar$gutter = _sectPr$pgMar.gutter,
        gutter = _sectPr$pgMar$gutter === void 0 ? pgMar.gutter : _sectPr$pgMar$gutter;
    pgMar = {
      header: header,
      footer: footer,
      top: top,
      bottom: bottom,
      left: left,
      right: right,
      gutter: gutter
    };
  }

  var result = _extends({}, sectPr, {
    pgSz: pgSz,
    pgMar: pgMar
  });

  if (field && result[field]) {
    return result[field];
  }

  return result;
};
export function getMaxPageWidth(document) {
  var sectPr = getSectPr(document.data);
  var sectPrPgSz = sectPr.pgSz;
  var sectPrPageWidth = sectPrPgSz.w;
  return document.nodes.filter(function (node) {
    return node && node.type === 'sectionBr';
  }).reduce(function (maxWidth, node) {
    var _node$data$pgSz;

    if ((_node$data$pgSz = node.data.pgSz) != null && _node$data$pgSz.w) {
      return Math.max(node.data.pgSz.w, maxWidth);
    }

    return maxWidth;
  }, sectPrPageWidth);
} // FIX: 重命名 getMaxPageWidthSection 导致 we-word 中无法找到该方法 to @白昱

export var getMaxPageWidthSection = getMaxPageWidth;
