// @ali/4ever-bi/dist-node/plugins/sticker/components/next/stickerButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createStickerButton;

var _react = _interopRequireDefault(require("react"));

var _actions = require("../../actions");

var _constants = require("../../constants");

var _everComponent = require("@ali/4ever-component");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement(_everComponent.IconEmoji, null);

function createStickerButton(props) {
  const {
    controller,
    locale = {},
    icon,
    title,
    tooltip,
    badgeDot,
    disabled
  } = props;

  const handleClick = () => {
    controller.run('onAction', (0, _actions.insertSticker)(_constants.STICKER_PLACEHOLDER_DATA));
    controller.run('onAction', (0, _actions.setStickerPanelVisible)({
      visible: true
    }));
  };

  const isDisabled = disabled || controller.query('isDisableSticker') || controller.query('isSelectionInImage') || controller.query('isSelectionInLinkCard');
  const buttonConfig = {
    icon: icon || _ref,
    onClick: handleClick,
    title: title || locale?.quickInsertTitle,
    tooltip: {
      title: title || locale?.quickInsertTitle,
      ...tooltip
    },
    badgeDot,
    disabled: isDisabled,
    testid: "bi-toolbar-sticker"
  };
  return buttonConfig;
}
