// @builder/vite-service/lib/start.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.viteStart = void 0;
var chalk_1 = require("chalk");
var vite_1 = require("vite");
var wp2vite_1 = require("./wp2vite");
function viteStart(context) {
    return __awaiter(this, void 0, void 0, function () {
        var applyHook, command, commandArgs, configArr, devConfig, errorMsg, devServer, err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    applyHook = context.applyHook, command = context.command, commandArgs = context.commandArgs;
                    configArr = context.getWebpackConfig();
                    devConfig = configArr.length > 0 ? (0, wp2vite_1.wp2vite)(context) : {};
                    return [4 /*yield*/, applyHook("before.".concat(command, ".load"), {
                            args: commandArgs,
                            webpackConfig: configArr,
                            viteConfig: devConfig,
                        })];
                case 1:
                    _a.sent();
                    if (!!configArr.length) return [3 /*break*/, 3];
                    errorMsg = 'No config found.';
                    return [4 /*yield*/, applyHook('error', { err: new Error(errorMsg) })];
                case 2:
                    _a.sent();
                    return [2 /*return*/];
                case 3: return [4 /*yield*/, applyHook("before.".concat(command, ".run"), {
                        args: commandArgs,
                        config: devConfig,
                    })];
                case 4:
                    _a.sent();
                    _a.label = 5;
                case 5:
                    _a.trys.push([5, 8, , 10]);
                    return [4 /*yield*/, (0, vite_1.createServer)(devConfig)];
                case 6:
                    devServer = _a.sent();
                    return [4 /*yield*/, devServer.listen()];
                case 7:
                    _a.sent();
                    applyHook("after.".concat(command, ".devServer"), {
                        devServer: devServer,
                        url: generateDevServerUrl(devConfig),
                    });
                    devServer.printUrls();
                    return [2 /*return*/, devServer];
                case 8:
                    err_1 = _a.sent();
                    console.error('Vite', chalk_1.default.red('Failed to start vite server.'));
                    return [4 /*yield*/, applyHook('error', { err: err_1 })];
                case 9:
                    _a.sent();
                    throw err_1;
                case 10: return [2 /*return*/];
            }
        });
    });
}
exports.viteStart = viteStart;
function generateDevServerUrl(devConfig) {
    var _a = devConfig.server, port = _a.port, https = _a.https, host = _a.host;
    var protocol = https ? 'https' : 'http';
    return "".concat(protocol, "://").concat(resolveHostname(host), ":").concat(port, "/");
}
// Reference: https://github.com/vitejs/vite/blob/7e3e84e1b733f4cb0cba3bd69f28a5671b52261c/packages/vite/src/node/utils.ts#L580
function resolveHostname(host) {
    if (host === '0.0.0.0' || host === '::' || host === '127.0.0.1') {
        return 'localhost';
    }
    return host;
}
