// @ali/4ever-plugin-paragraph/dist-web/bi/queries/getParagraphData.js
import _extends from "@babel/runtime/helpers/extends";
import { getBlocksFromEditor, SpacingUtil, IndUtil, ShdUtil } from '@ali/4ever-utils';
var defaultSpacing = {
  before: 0,
  after: 0,
  line: 1
};
var defaultInd = {
  left: 0,
  right: 0
};
var defaultShd = {
  val: 'clear',
  color: 'auto',
  fill: 'auto'
};

function createGetSpacingFromBlock(defaultProperties) {
  return function getSpacingFromBlock(block) {
    var mergedSpacing = {}; // 拿到转换后的 spacing

    var spacing = SpacingUtil.fromBlock(block);
    mergedSpacing = _extends({}, defaultSpacing, defaultProperties, spacing);
    return mergedSpacing;
  };
}

function getIndFromBlock(block) {
  var mergedInd = {}; // 拿到转换后的 spacing

  var ind = IndUtil.fromBlock(block);
  mergedInd = _extends({}, defaultInd, ind);
  return mergedInd;
}

export function getShdFromBlock(block) {
  var mergedShd = {}; // 拿到转换后的 shd

  var shd = ShdUtil.fromBlock(block);
  mergedShd = _extends({}, defaultShd, shd);
  return mergedShd;
}
/**
 * 提取公共属性
 * 这是个 n * m 复杂度的计算
 * 3000个段落大概是 5ms
 * TODO: 或许有人有更好的解法？
 */

function getCommonProperties(items, processor) {
  var first = {};
  if (!items.length) return first;
  first = processor(items[0]);
  if (items.length === 1) return first;
  var dirtyMap = {};

  for (var i = 1; i < items.length; i++) {
    var nextSpacing = processor(items[i]);
    var keys = Object.keys(nextSpacing);

    for (var j = 0; j < keys.length; j++) {
      var key = keys[j];
      if (dirtyMap[key]) continue;

      if (!(key in dirtyMap)) {
        dirtyMap[key] = false;
      }

      if (!dirtyMap[key] && nextSpacing[key] !== first[key]) {
        // 标记不一致
        dirtyMap[key] = true;
      }
    }
  }

  Object.keys(dirtyMap).forEach(function (dirtyKey) {
    if (dirtyMap[dirtyKey]) {
      // 发现了不一样的视为未发现公共属性
      delete first[dirtyKey];
    }
  });
  Object.keys(first).forEach(function (key) {
    if (first[key] !== undefined && !(key in dirtyMap)) {
      // 从未发现后续 block 有此属性也视为未发现公共属性
      delete first[key];
    }
  });
  return first;
}

export function getCommonSpacing(controller, defaultProperties) {
  var blocks = getBlocksFromEditor(controller);
  return getCommonProperties(blocks, createGetSpacingFromBlock(defaultProperties));
}
export function getCommonInd(controller) {
  var blocks = getBlocksFromEditor(controller);
  return getCommonProperties(blocks, getIndFromBlock);
}
export function getCommonShd(controller) {
  var blocks = getBlocksFromEditor(controller);
  return getCommonProperties(blocks, getShdFromBlock);
}
