// @ali/we-toolbar/dist-web/biz/galleryContainer/fontSizeMaxSelect/FontSizeButton.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { classnames } from '@ali/we-design';
import { DesWrap, GalleryButton, TextWrap, TitleWrap } from "./style";

var FontSizeButton = function FontSizeButton(_ref) {
  var option = _ref.option,
      value = _ref.value,
      minFontSize = _ref.minFontSize,
      maxFontSize = _ref.maxFontSize,
      defaultFontFamily = _ref.defaultFontFamily;
  return /*#__PURE__*/_createElement(GalleryButton, {
    key: option.type,
    value: option.type,
    testid: option.type,
    active: option.type === value,
    className: classnames({
      active: option.type === value
    })
  }, /*#__PURE__*/_createElement(TitleWrap, {
    type: option.type,
    style: {
      fontSize: Math.min(Math.max(option["default"] * 0.75, minFontSize), maxFontSize) + "pt",
      fontWeight: option.fontWeight,
      lineHeight: '29px',
      height: '29px',
      textOverflow: 'clip',
      maxWidth: '65px',
      overflow: 'hidden'
    }
  }, option.title), /*#__PURE__*/_createElement(DesWrap, null, /*#__PURE__*/_createElement(TextWrap, null, option.title === 'AaBbCcDdEe' ? option.des : defaultFontFamily + " " + (option.type === 'blockquote' ? '-' : option["default"]))));
};

export default FontSizeButton;
