// @ali/4ever-zhi/dist-node/plugins/headingGroup/handlers/createRenderNode.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _everCangjie = require("@ali/4ever-cangjie");

var _everMo = require("@ali/4ever-mo");

var _everBamboo = require("@ali/4ever-bamboo");

var _group = _interopRequireDefault(require("../components/group"));

const _createElement = /*#__PURE__*/React.createElement;
const {
  utils
} = _everBamboo.HeadingGroupPlugin;
const {
  Heading1,
  Heading2,
  Heading3,
  Heading4,
  Heading5,
  Heading6
} = _everMo.Heading;
const headingList = [Heading1, Heading2, Heading3, Heading4, Heading5, Heading6];

function createRenderNode(config) {
  const {
    level = 3
  } = config;
  const hCondition = utils.condition(...headingList.slice(0, level).map(H => {
    return node => {
      return _everCangjie.Block.isBlock(node) && node.type === utils.generateGroupType(H.TYPE);
    };
  }));
  return function renderNode(props, controller, next) {
    const {
      node,
      attributes,
      children
    } = props;

    if (hCondition.some(node)) {
      return /*#__PURE__*/_createElement(_group.default, {
        node: node,
        attributes: attributes,
        controller: controller,
        config: config
      }, children());
    }

    return next();
  };
}

var _default = createRenderNode;
exports.default = _default;
