// @ali/4ever-bi/dist-web/plugins/comment/index.js
import { CommentPlugin } from '@ali/4ever-bamboo';
import toolbar from "./toolbar";
import createRenderToolbar from "./createRenderToolbar";
import createRenderHoverToolbar from "./createRenderHoverToolbar";
var hooks = CommentPlugin.hooks,
    types = CommentPlugin.types,
    actions = CommentPlugin.actions,
    commands = CommentPlugin.commands,
    queries = CommentPlugin.queries;
var lazyRenderStrategy = queries.lazyRenderStrategy;

function createCommentPlugin(allConfigs) {
  var createContextMenu = hooks.createContextMenu,
      createOnKeyDown = hooks.createOnKeyDown,
      createOnMouseMove = hooks.createOnMouseMove,
      createOnCangjieInput = hooks.createOnCangjieInput,
      createOnAction = hooks.createOnAction,
      createOnChange = hooks.createOnChange,
      createRenderMark = hooks.createRenderMark,
      createRenderNode = hooks.createRenderNode,
      createRenderContent = hooks.createRenderContent,
      createRenderEditable = hooks.createRenderEditable,
      createOnCangjieSelect = hooks.createOnCangjieSelect,
      createOnCangjieCopyOrCut = hooks.createOnCangjieCopyOrCut;
  var configs = allConfigs.comment || {};
  return {
    queries: {
      lazyRenderStrategy: lazyRenderStrategy
    },
    contextMenu: createContextMenu(configs),
    onKeyDown: createOnKeyDown(configs),
    onMouseMove: createOnMouseMove(configs),
    onCangjieInput: createOnCangjieInput(configs),
    onAction: createOnAction(configs),
    onChange: createOnChange(configs),
    renderMark: createRenderMark(configs),
    renderNode: createRenderNode(configs),
    renderContent: createRenderContent(configs),
    renderEditable: createRenderEditable(configs),
    onCangjieSelect: createOnCangjieSelect(configs),
    onCangjieCopy: createOnCangjieCopyOrCut(),
    onCangjieCut: createOnCangjieCopyOrCut(),
    renderToolbar: createRenderToolbar(configs.locale || {}),
    renderHoverToolbar: createRenderHoverToolbar(configs),
    commands: commands,
    toolbar: toolbar(configs)
  };
}

createCommentPlugin.types = types;
createCommentPlugin.actions = actions;
createCommentPlugin.commands = commands;
createCommentPlugin.queries = queries;
export default createCommentPlugin;
