// @ali/4ever-plugin-hr/dist-node/mo/serializer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createSerializer;
exports.createServerSerializationPlugin = createServerSerializationPlugin;

var _htmlToJsonML = _interopRequireDefault(require("./htmlToJsonML"));

var _jsonMLToValue = _interopRequireDefault(require("./jsonMLToValue"));

var _valueToJsonML = _interopRequireDefault(require("./valueToJsonML"));

var _jsonMLToHTML = _interopRequireDefault(require("./jsonMLToHTML"));

var _markdownToHTML = _interopRequireDefault(require("./markdownToHTML"));

var _htmlToMarkdown = _interopRequireDefault(require("./htmlToMarkdown"));

function createSerializer() {
  return {
    htmlToJsonML: _htmlToJsonML.default,
    jsonMLToValue: _jsonMLToValue.default,
    valueToJsonML: _valueToJsonML.default,
    jsonMLToHTML: _jsonMLToHTML.default,
    markdownToHTML: _markdownToHTML.default,
    htmlToMarkdown: _htmlToMarkdown.default
  };
}

function createServerSerializationPlugin() {
  return {
    valueToJsonML: _valueToJsonML.default,
    jsonMLToValue: _jsonMLToValue.default
  };
}
