// @ali/we-ui/dist-web/components/select/SelectButton.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { InputStyled } from "./styled";

var stopPropagation = function stopPropagation(e) {
  e.stopPropagation();
};

var getValueString = function getValueString(value) {
  return typeof value === 'undefined' ? '' : "" + value;
};

export function SelectButton(props) {
  var locale = props.locale,
      renderTrigger = props.renderTrigger,
      value = props.value,
      children = props.children,
      mode = props.mode,
      visible = props.visible,
      keyword = props.keyword,
      isInputFocused = props.isInputFocused,
      onFocus = props.onFocus,
      onBlur = props.onBlur,
      onEnter = props.onEnter,
      onChange = props.onChange,
      inputRef = props.inputRef,
      testid = props.testid; // 自定义显示区函数优先级最高

  if (typeof renderTrigger === 'function') {
    return renderTrigger(value);
  } // 支持输入的模式


  if (mode === 'search' || mode === 'input') {
    return /*#__PURE__*/_createElement(InputStyled, {
      ref: inputRef,
      placeholder: getValueString(value),
      locale: locale,
      value: "" + (!isInputFocused || !visible ? getValueString(value) : keyword),
      autoComplete: "off",
      readOnly: !visible,
      onFocus: onFocus,
      onChange: onChange,
      onBlur: onBlur,
      onEnter: onEnter,
      onMouseDown: visible ? stopPropagation : undefined,
      onClick: visible ? stopPropagation : undefined,
      "data-testid": testid && "input-" + testid
    });
  }

  return children;
}
