// @ali/we-ui/dist-web/components/select/styled.js
import styled, { css } from 'styled-components';
import { Input } from "../input";
import { resetComponent, Ellipsis } from "../../utils/styled.mixin";
import { withTheme } from "../theme";
var iconStyle = /*#__PURE__*/css(["text-transform:none;text-rendering:optimizeLegibility;line-height:0;text-align:center;"]);
export var SelectWrap = withTheme(styled.div(["", " display:inline-block;user-select:none;height:", ";padding:", ";border-radius:", ";transition:all .1s ease-in-out;cursor:pointer;outline:none;&.small{height:", ";}&.large{height:", ";padding:", ";}&:hover{background-color:", ";}&:active,&.actived{background-color:", ";outline:none;}&.bordered{border:1px solid ", ";}&.errored{border:1px solid #FF523C;}&.disabled{cursor:not-allowed;color:", ";background-color:transparent;& .weicon{color:", ";}& .select-arrow{color:", ";;}}"], resetComponent, function (_ref) {
  var theme = _ref.theme;
  return theme.ui.elementHeight;
}, function (_ref2) {
  var showArrow = _ref2.showArrow;
  return showArrow && '0 4px 0 6px';
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.radius;
}, function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.elementHeightSmall;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.ui.elementHeightLarge;
}, function (_ref6) {
  var showArrow = _ref6.showArrow;
  return showArrow && '0 12px 0 12px';
}, function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.commonBgHoverColor;
}, function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.commonBgPressColor;
}, function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.commonLineLightColor;
}, function (_ref10) {
  var theme = _ref10.theme;
  return theme.ui.disabledColor;
}, function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.disabledColor;
}, function (_ref12) {
  var theme = _ref12.theme;
  return theme.ui.disabledColor;
}));
export var TriggerWrap = withTheme(styled.div(["position:relative;height:100%;display:flex;align-items:center;justify-content:space-between;.weicon{display:block;color:", ";}.select-arrow{", " color:", ";flex-shrink:0;flex-grow:0;font-size:12px;transform:rotate(0deg) scale(0.67);transform-origin:center center;margin-left:2px;&.opened{transform:rotate(180deg) scale(0.67);}&.hidden{display:none;}}"], function (_ref13) {
  var theme = _ref13.theme;
  return theme.ui.commonActionIconColor;
}, iconStyle, function (_ref14) {
  var theme = _ref14.theme;
  return theme.ui.commonBaseColor3;
}));
export var TriggerContent = withTheme(styled.div(["", ";flex-grow:1;flex-shrink:1;"], Ellipsis));
export var OverlayWrap = withTheme(styled.div(["max-height:400px;overflow-y:auto;padding:", " 0;"], function (_ref15) {
  var theme = _ref15.theme;
  return theme.ui.padding;
}));
export var SelectOptionWrap = withTheme(styled.div(["", " position:relative;display:flex;justify-content:space-between;align-items:center;padding:", " 20px;margin:0 ", ";height:100%;min-width:100px;min-height:32px;font-size:13px;line-height:20px;white-space:nowrap;cursor:pointer;user-select:none;border-radius:", ";&:hover,&.hovered{background-color:", ";}&.selected{.check-icon{display:inline-block;}}&.disabled{color:", ";background-color:transparent;cursor:not-allowed;}.check-icon{", " color:", ";font-size:12px;display:none;color:#3296fa;position:absolute;left:4px;top:50%;transform:translateY(-50%);}"], resetComponent, function (_ref16) {
  var theme = _ref16.theme;
  return theme.ui.paddingSmall;
}, function (_ref17) {
  var theme = _ref17.theme;
  return theme.ui.marginMedium;
}, function (_ref18) {
  var theme = _ref18.theme;
  return theme.ui.radiusLarge;
}, function (_ref19) {
  var theme = _ref19.theme;
  return theme.ui.backgroundColor;
}, function (_ref20) {
  var theme = _ref20.theme;
  return theme.ui.disabledColor;
}, iconStyle, function (_ref21) {
  var theme = _ref21.theme;
  return theme.ui.primaryColor;
}));
export var InputStyled = withTheme(styled(Input)(["border:0;background-color:transparent;color:inherit;font-size:inherit;line-height:inherit;padding-left:0;padding-right:0;& > input{font-size:inherit;color:inherit;line-height:inherit;}"]));
export var EmptyWrap = withTheme(styled.div(["padding:60px 24px;"]));
export var OptionGroupWrap = withTheme(styled.div(["position:relative;&:after{display:block;content:' ';height:1px;background-color:", ";width:80%;margin:6px auto;}&:before{display:block;content:' ';height:1px;background-color:", ";width:80%;margin:6px auto;}&:nth-child(1){&:before{display:none;}}&:last-child{&:after{display:none;}}~ .select-group-wrapper{&:before{display:none;}}"], function (_ref22) {
  var theme = _ref22.theme;
  return theme.ui.commonLineLightColor;
}, function (_ref23) {
  var theme = _ref23.theme;
  return theme.ui.commonLineLightColor;
}));
