// @ali/4ever-bi/dist-web/plugins/hoverToolbar/index.js
import _extends from "@babel/runtime/helpers/extends";
import createRenderEditable from "./createRenderEditable";
import createOnMouseMove from "./createOnMouseMove";
import { HoverCapture } from "./utils/hoverCapture";
import createRenderHoverToolbar from "./createRenderHoverToolbar";
import createOnAction from "./handles/onAction";
import createOnCangjieInput from "./handles/onCangjieInput";
import createOnContextMenu from "./createOnContextMenu";
import createOnQuery from "./createOnQuery";
import { DEFAULT_LAYOUTS } from "./constants";

function createHoverToolbarPlugin(allConfigs) {
  var config = allConfigs.hoverToolbar;
  var hoverCapture = new HoverCapture();

  var hoverToolbarConfig = _extends({}, config, {
    layouts: config.layouts || DEFAULT_LAYOUTS
  });

  return {
    onMouseMove: createOnMouseMove(hoverCapture, hoverToolbarConfig),
    onContextMenu: createOnContextMenu(hoverCapture, hoverToolbarConfig),
    renderEditable: createRenderEditable(hoverCapture, hoverToolbarConfig),
    renderHoverToolbar: createRenderHoverToolbar(hoverToolbarConfig),
    onAction: createOnAction(),
    onCangjieInput: createOnCangjieInput(hoverToolbarConfig, hoverCapture),
    onQuery: createOnQuery(hoverToolbarConfig)
  };
}

export default createHoverToolbarPlugin;
