// @ali/4ever-bi/dist-node/plugins/heading/components/createHeadingMenu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.heading4Inline = exports.heading3Inline = exports.heading2Inline = exports.heading1Inline = exports.paragraph = exports.heading4 = exports.heading3 = exports.heading2 = exports.heading1 = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _everMo = require("@ali/4ever-mo");

var _actions = require("../actions");

var _isSelectionInHeading = require("../utils/isSelectionInHeading");

const _createElement = /*#__PURE__*/React.createElement;

function createHeadingMenu(Component, headingType) {
  return function HeadingMenu(props) {
    const {
      menuItem,
      controller
    } = props;
    const {
      action,
      title,
      ...rest
    } = menuItem;
    const isDisabled = controller.query('isSelectionInLinkCard') || controller.query('isSelectionInSubjectHeading') || action !== 'format' && headingType === _everMo.Paragraph.TYPE && (0, _isSelectionInHeading.isSelectionInOnlyHeading)(controller.value, _everMo.Paragraph.TYPE);

    const handleClick = () => {
      if (action === 'format') {
        return controller.run('onAction', (0, _actions.transformHeading)(headingType));
      }

      controller.run('onAction', (0, _actions.setHeading)(headingType));
    };

    return /*#__PURE__*/_createElement(Component, (0, _extends2.default)({
      disabled: isDisabled
    }, rest, {
      onClick: handleClick,
      role: headingType
    }), title);
  };
}

const heading1 = createHeadingMenu(_weDesign.MenuItem, _everMo.Heading.Heading1.TYPE);
exports.heading1 = heading1;
const heading2 = createHeadingMenu(_weDesign.MenuItem, _everMo.Heading.Heading2.TYPE);
exports.heading2 = heading2;
const heading3 = createHeadingMenu(_weDesign.MenuItem, _everMo.Heading.Heading3.TYPE);
exports.heading3 = heading3;
const heading4 = createHeadingMenu(_weDesign.MenuItem, _everMo.Heading.Heading4.TYPE);
exports.heading4 = heading4;
const paragraph = createHeadingMenu(_weDesign.MenuItem, 'paragraph');
exports.paragraph = paragraph;
const heading1Inline = createHeadingMenu(_weDesign.InlineMenuItem, _everMo.Heading.Heading1.TYPE);
exports.heading1Inline = heading1Inline;
const heading2Inline = createHeadingMenu(_weDesign.InlineMenuItem, _everMo.Heading.Heading2.TYPE);
exports.heading2Inline = heading2Inline;
const heading3Inline = createHeadingMenu(_weDesign.InlineMenuItem, _everMo.Heading.Heading3.TYPE);
exports.heading3Inline = heading3Inline;
const heading4Inline = createHeadingMenu(_weDesign.InlineMenuItem, _everMo.Heading.Heading4.TYPE);
exports.heading4Inline = heading4Inline;
