// @ali/4ever-bi/dist-node/plugins/heading/components/createHeading.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createHeading;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _everBamboo = require("@ali/4ever-bamboo");

const _createElement = /*#__PURE__*/React.createElement;

function createHeading(Paragraph) {
  var _class, _temp;

  return _temp = _class = class Heading extends React.Component {
    componentDidMount() {
      const {
        onHeadingChange
      } = this.context;

      if (onHeadingChange) {
        onHeadingChange();
      }
    }

    componentDidUpdate(prevProps) {
      const prevNode = prevProps.node;
      const {
        node: currentNode
      } = this.props;
      const {
        onHeadingChange
      } = this.context;

      if (onHeadingChange && prevNode !== currentNode) {
        onHeadingChange();
      }
    }

    componentWillUnmount() {
      const {
        onHeadingChange
      } = this.context;

      if (onHeadingChange) {
        onHeadingChange();
      }
    }

    render() {
      const {
        node
      } = this.props;
      return /*#__PURE__*/_createElement(Paragraph, (0, _extends2.default)({}, this.props, {
        Component: (0, _everBamboo.styledComponents)(node.type),
        id: (0, _everBamboo.createHeadingId)(node)
      }));
    }

  }, _class.contextTypes = {
    onHeadingChange: PropTypes.func
  }, _temp;
}
