// @ali/4ever-bi/dist-node/plugins/heading/components/headingSimpleButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = HeadingSimpleButton;
exports.isDisabled = void 0;

var React = _interopRequireWildcard(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../actions");

var _everUtils = require("@ali/4ever-utils");

const _createElement = /*#__PURE__*/React.createElement;
;
const SUPPORT_HEADING = {
  paragraph: true,
  'heading-1': true,
  'heading-2': true,
  'heading-3': true,
  'heading-4': true,
  blockquote: true
};

const isDisabled = controller => {
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInListSymbol') || controller.query('isSelectionInLinkCard') // fix: https://aone.alibaba-inc.com/v2/bug/35749970
  || controller.query('isSelectionInSubjectHeading');
  return Boolean(isDisabled);
};

exports.isDisabled = isDisabled;

function HeadingSimpleButton(props) {
  const {
    controller,
    locale,
    size
  } = props;
  const heading = controller.query('getBlockquoteState') ? 'blockquote' : controller.query('getHeading');
  const handleChange = React.useCallback(value => {
    if (value === 'blockquote') {
      controller.run('onAction', (0, _actions.setHeading)('paragraph'));
      controller.run('onAction', (0, _everUtils.enableBlockquote)());
    } else {
      controller.run('onAction', (0, _actions.setHeading)(value));
      controller.run('onAction', (0, _everUtils.disableBlockquote)());
    }
  }, [controller]);
  const HeadingSimpleSelect = React.useMemo(() => _weToolbar.HeadingSimpleSelect, []);
  return /*#__PURE__*/_createElement(HeadingSimpleSelect, {
    size: size,
    onChange: handleChange,
    value: SUPPORT_HEADING[heading] ? heading : 'paragraph',
    testid: "bi-toolbar-simple-heading",
    disabled: isDisabled(controller),
    tooltip: locale.buttonTitle,
    role: _everBamboo.PluginRoles.headingSelect
  });
}
