// @ali/4ever-bi/dist-node/plugins/heading/components/headingButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = HeadingButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _everBamboo = require("@ali/4ever-bamboo");

var _everComponent = require("@ali/4ever-component");

var _actions = require("../actions");

var _getTipsByHeading = _interopRequireDefault(require("../utils/getTipsByHeading"));

const _createElement = /*#__PURE__*/React.createElement;
;
const TextTypeSelect = /*#__PURE__*/React.memo(_weToolbar.TextTypeSelect);

function HeadingButton(props) {
  const {
    controller,
    locale,
    toolbarMode,
    ...rest
  } = props;
  const heading = controller.query('getHeading') || '';
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInListSymbol') || controller.query('isSelectionInLinkCard');
  const isDouble = toolbarMode === _everComponent.NewToolbarMode.double;
  const handleChange = React.useCallback(headingStr => {
    controller.run('onAction', (0, _actions.setHeading)(headingStr));
  }, [controller]);
  const title = (0, _getTipsByHeading.default)(heading, locale);
  const titleTips = React.useMemo(() => ({
    paragraph: {
      title: (0, _getTipsByHeading.default)('paragraph', locale, true),
      placement: 'right'
    },
    'heading-1': {
      title: (0, _getTipsByHeading.default)('heading-1', locale, true),
      placement: 'right'
    },
    'heading-2': {
      title: (0, _getTipsByHeading.default)('heading-2', locale, true),
      placement: 'right'
    },
    'heading-3': {
      title: (0, _getTipsByHeading.default)('heading-3', locale, true),
      placement: 'right'
    },
    'heading-4': {
      title: (0, _getTipsByHeading.default)('heading-4', locale, true),
      placement: 'right'
    }
  }), [locale]);
  const realWidth = 66;
  return /*#__PURE__*/_createElement(TextTypeSelect, (0, _extends2.default)({}, rest, {
    testid: "bi-toolbar-heading",
    disabled: !!isDisabled,
    value: heading,
    onChange: handleChange,
    titles: titleTips,
    bordered: isDouble,
    width: realWidth,
    shortcutText: title,
    role: _everBamboo.PluginRoles.headingSelect
  }));
}
