// @ali/4ever-bi/dist-node/plugins/heading/components/headingWrapper.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _everBamboo = require("@ali/4ever-bamboo");

class HeadingWrapper extends React.Component {
  constructor(...args) {
    super(...args);
    this.handleHeadingChange = (0, _everBamboo.throttleByIdleCallback)(() => {
      const {
        controller,
        onHeadingsChange
      } = this.props;
      const dataValue = controller.query('getDataValue');
      onHeadingsChange((0, _everBamboo.getHeadings)(dataValue || controller.value));
    });
  }

  getChildContext() {
    return {
      onHeadingChange: this.handleHeadingChange
    };
  }

  componentWillUnmount() {
    if (this.handleHeadingChange) {
      this.handleHeadingChange.cancel();
      this.handleHeadingChange = null;
    }
  }

  render() {
    const {
      children
    } = this.props;
    return children;
  }

}

exports.default = HeadingWrapper;
HeadingWrapper.childContextTypes = {
  onHeadingChange: PropTypes.func
};
;
