// @ali/4ever-bi/dist-node/plugins/heading/components/mobileHeadingButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = mobileHeadingButton;

var React = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../actions");

const _createElement = /*#__PURE__*/React.createElement;

function mobileHeadingButton(heading) {
  return function HeadingButton(props) {
    const {
      controller
    } = props;
    const {
      value
    } = controller;
    const {
      isCollapsedAtEdge
    } = controller.value.selection;
    const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInLinkCard');

    const handleClick = () => controller.run('onAction', (0, _actions.setHeading)(heading));

    let level = '';
    const levelMatch = heading?.match(/\d+/);

    if (levelMatch && levelMatch[0]) {
      level = levelMatch[0];
    }

    return /*#__PURE__*/_createElement(_weDesign.Button, {
      testid: "bi-toolbar-heading",
      icon: `H${level}`,
      bordered: false,
      disabled: isDisabled,
      onClick: handleClick,
      role: _everBamboo.PluginRoles.heading
    });
  };
}
