// @ali/4ever-plugin-im-tag/dist-web/utils/styled.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { ArrowRightNormal } from '@ali/we-design';
import { IMTAG_MIN_CONTAINER_WIDTH, IMTAG_DEFAULT_HEIGHT } from "./constants";
;
var ContainerWrapper = /*#__PURE__*/styled.span(["font-size:14px;cursor:pointer;display:inline-flex;border-radius:12px;border:1px solid rgba(80,116,154,0.12);cursor:auto;width:", ";height:", "px;line-height:", "px;&:hover{background-color:rgba(17,31,44,0.08);border:1px solid rgba(79,119,172,0.12);}"], function (props) {
  return props.width;
}, function (props) {
  return props.height;
}, function (props) {
  return props.height;
});
var ContentWrapper = /*#__PURE__*/styled.span(["margin:0 7px;color:#50749A;letter-spacing:0;cursor:pointer;display:flex;line-height:calc(", "px - 2px);"], function (props) {
  return props.height;
});
export var TextWrapper = /*#__PURE__*/styled.span(["width:", ";display:inline-block;letter-spacing:0;white-space:nowrap;font-size:14px;overflow:hidden;text-overflow:ellipsis;flex-grow:1;"], function (props) {
  return props.width ? props.width - IMTAG_MIN_CONTAINER_WIDTH + 'px' : 'auto';
});
var IconWrapper = /*#__PURE__*/styled.span(["width:14px;text-align:center;display:inline-flex;& > span{align-self:center;}"]);

var _ref = /*#__PURE__*/_createElement(IconWrapper, null, /*#__PURE__*/_createElement(ArrowRightNormal, null));

export var ImTagContainer = function ImTagContainer(props) {
  var height = props.height,
      width = props.width,
      text = props.text,
      onClick = props.onClick,
      onMouseDown = props.onMouseDown;
  return /*#__PURE__*/_createElement(ContainerWrapper, {
    height: height || IMTAG_DEFAULT_HEIGHT,
    width: width + 'px',
    onClick: onClick,
    onMouseDown: onMouseDown
  }, /*#__PURE__*/_createElement(ContentWrapper, {
    height: height || IMTAG_DEFAULT_HEIGHT
  }, /*#__PURE__*/_createElement(TextWrapper, {
    width: width
  }, text), _ref));
};
