// @ali/4ever-factory/dist-web/index.js
/** card */
export { default as createUnknownCardPlugin } from "./cardFactory/createUnknownCardPlugin";
export { default as createUnknownCardMobilePlugin, Placeholder } from "./cardFactory/createUnknownCardMobilePlugin";
export { removeCard, setCardData, setCardInjection, insertCard, cardFactory } from "./cardFactory";
export { CARD_SET_INJECTION } from "./cardFactory/actions";
export { cardFactory as mobileCardFactory } from "./cardFactory/mobileCardFactory";
export { insertCardCommand, setCardInjectionCommand } from "./cardFactory/commands";

/** tag */
export { default as createUnknownTagPlugin } from "./tagFactory/createUnknownTagPlugin";
export { default as createUnknownTagMobilePlugin } from "./tagFactory/createUnknownTagPlugin";
export { insertTag, setTagData, removeTag, tagFactory } from "./tagFactory";
export { TAG_INSERT } from "./tagFactory/actions";

/** embed */
export { FILE_SELECT, FILE_REMOVE, FILE_UPDATE, IMPORT_MEMO, updateFile, selectFile, importMemo } from "./embedFactory/actions";
export { EmbedInlineToolbar, mobileEmbedPluginFactory, embedPluginFactory, UploadOrFailedCard } from "./embedFactory";
export { default as insertFile } from "./embedFactory/commands/insertFile";
export { default as uploadFile } from "./embedFactory/handlers/uploadFile";
/** group */

export { groupFactory } from "./groupFactory";
export { toggleFold } from "./groupFactory/commands";
/** zhi */

export { cardFactory as zhiCardFactory } from "./zhi/cardFactory";
export { tagFactory as zhiTagFactory } from "./zhi/tagFactory";
export { groupFactory as zhiGroupFactory } from "./zhi/groupFactory";
export { default as createUnknownCardZhiPlugin } from "./zhi/cardFactory/createUnknownCardPlugin";
export { default as createUnknownTagZhiPlugin } from "./zhi/tagFactory/createUnknownTagPlugin";
