// @ali/4ever-plugin-image/dist-web/utils/img.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import Image from "./components/image";
export default /*#__PURE__*/React.forwardRef(function (props, ref) {
  var node = props.node,
      controller = props.controller,
      imageConfig = props.imageConfig,
      isTransferring = props.isTransferring,
      uploadPlaceholder = props.uploadPlaceholder,
      uploadProgress = props.uploadProgress,
      onLoad = props.onLoad,
      rest = _objectWithoutPropertiesLoose(props, ["node", "controller", "imageConfig", "isTransferring", "uploadPlaceholder", "uploadProgress", "onLoad"]);

  return /*#__PURE__*/_createElement(Image, _extends({}, rest, {
    ref: ref,
    node: node,
    controller: controller,
    imageConfig: imageConfig,
    isTransferring: isTransferring,
    uploadPlaceholder: uploadPlaceholder,
    uploadProgress: uploadProgress,
    onLoad: onLoad
  }));
});
