// @ali/4ever-plugin-image/dist-web/utils/role.js
export var ImageRole;

(function (ImageRole) {
  ImageRole["image"] = "image";
  ImageRole["imageBorder"] = "imageBorder";
  ImageRole["imageReset"] = "imageReset";
  ImageRole["imageClearStyle"] = "imageClearStyle";
  ImageRole["imageOptions"] = "imageOptions";
  ImageRole["imageRecommendStyle"] = "imageRecommendStyle";
  ImageRole["imageRecommendStyleItem"] = "imageRecommendStyleItem";
  ImageRole["imageCropper"] = "imageCropper";
  ImageRole["imageResizer"] = "imageResizer";
  ImageRole["imageRotator"] = "imageRotator";
  ImageRole["imageSingleRotator"] = "imageSingleRotator";
  ImageRole["imageInlineToolbarCrop"] = "imageInlineToolbarCrop";
  ImageRole["imageInlineToolbarDownload"] = "imageInlineToolbarDownload";
  ImageRole["imageInlineToolbarPreview"] = "imageInlineToolbarPreview";
  ImageRole["imageInlineToolbarRotateReverse"] = "imageInlineToolbarRotateReverse";
  ImageRole["imageInlineToolbarRotate"] = "imageInlineToolbarRotate";
  ImageRole["imageInlineToolbarSingleRotateReverse"] = "imageInlineToolbarSingleRotateReverse";
  ImageRole["imageInlineToolbarMore"] = "imageInlineToolbarMore";
  ImageRole["imageMenuOptions"] = "imageMenuOptions";
  ImageRole["imageMenuDownload"] = "imageMenuDownload";
  ImageRole["imageSidebarWidthInput"] = "imageSidebarWidthInput";
  ImageRole["imageSidebarHeightInput"] = "imageSidebarHeightInput";
  ImageRole["imageSidebarRotationInput"] = "imageSidebarRotationInput";
  ImageRole["imageSidebarReverseRotate90"] = "imageSidebarReverseRotate90";
  ImageRole["imageSidebarRotate90"] = "imageSidebarRotate90";
  ImageRole["imageSidebarRadiusInput"] = "imageSidebarRadiusInput";
  ImageRole["imageSidebarRestore"] = "imageSidebarRestore";
})(ImageRole || (ImageRole = {}));
