// @ali/4ever-zhi/dist-node/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  createZhiPlugins: true,
  createMobileZhiPlugins: true,
  createCustomizedZhiPlugins: true,
  createMobileCustomizedZhiPlugins: true,
  plugins: true,
  mobilePlugins: true,
  Reader: true,
  locale_en_US: true,
  locale_ja_JP: true,
  locale_zh_CN: true,
  locale_zh_HK: true,
  locale_zh_TW: true,
  MarksPlugin: true,
  MobileMarksPlugin: true,
  AudioPlugin: true,
  MobileAudioPlugin: true,
  BlockquotePlugin: true,
  MobileBlockquotePlugin: true,
  ClipboardPlugin: true,
  MobileClipboardPlugin: true,
  FormulaPlugin: true,
  MobileCodePlugin: true,
  CodePlugin: true,
  InlineCodePlugin: true,
  ColumnsPlugin: true,
  MobileColumnsPlugin: true,
  CommentPlugin: true,
  MobileCommentPlugin: true,
  ContextMenuPlugin: true,
  EmbedPlugin: true,
  MobileEmbedPlugin: true,
  GroupTagPlugin: true,
  MobileGroupTagPlugin: true,
  HeadingPlugin: true,
  MobileHeadingPlugin: true,
  HrPlugin: true,
  MobileHrPlugin: true,
  ImagePlugin: true,
  MobileImagePlugin: true,
  LinkPlugin: true,
  LinkCardPlugin: true,
  MobileLinkPlugin: true,
  MobileLinkCardPlugin: true,
  ListPlugin: true,
  MobileListPlugin: true,
  VideoPlugin: true,
  MobileVideoPlugin: true,
  MentionPlugin: true,
  MobileMentionPlugin: true,
  OnlineVideoPlugin: true,
  MobileOnlineVideoPlugin: true,
  ParagraphPlugin: true,
  MobileParagraphPlugin: true,
  PlaceholderPlugin: true,
  MobilePlaceholderPlugin: true,
  ReferPlugin: true,
  MobileReferPlugin: true,
  TablePlugin: true,
  MobileTablePlugin: true,
  StickerPlugin: true,
  MobileStickerPlugin: true,
  ImTagPlugin: true,
  MobileImTagPlugin: true,
  CalendarCardPlugin: true,
  VoteCardPlugin: true,
  FindAndReplacePlugin: true,
  ProjectCardPlugin: true,
  MobileCalendarCardPlugin: true,
  URLCardPlugin: true,
  CalloutPlugin: true,
  HeadingGroupPlugin: true,
  HeadingGroupNewPlugin: true,
  ListGroupPlugin: true,
  SubjectPlugin: true,
  cardFactory: true,
  groupFactory: true,
  tagFactory: true
};
Object.defineProperty(exports, "createZhiPlugins", {
  enumerable: true,
  get: function () {
    return _createZhiPlugins.default;
  }
});
Object.defineProperty(exports, "createMobileZhiPlugins", {
  enumerable: true,
  get: function () {
    return _createMobileZhiPlugins.default;
  }
});
Object.defineProperty(exports, "createCustomizedZhiPlugins", {
  enumerable: true,
  get: function () {
    return _createCustomizedZhiPlugins.default;
  }
});
Object.defineProperty(exports, "createMobileCustomizedZhiPlugins", {
  enumerable: true,
  get: function () {
    return _createCustomizedZhiPlugins.default;
  }
});
Object.defineProperty(exports, "plugins", {
  enumerable: true,
  get: function () {
    return _plugins.default;
  }
});
Object.defineProperty(exports, "mobilePlugins", {
  enumerable: true,
  get: function () {
    return _mobile.default;
  }
});
Object.defineProperty(exports, "Reader", {
  enumerable: true,
  get: function () {
    return _reader.default;
  }
});
Object.defineProperty(exports, "locale_en_US", {
  enumerable: true,
  get: function () {
    return _en_US.default;
  }
});
Object.defineProperty(exports, "locale_ja_JP", {
  enumerable: true,
  get: function () {
    return _ja_JP.default;
  }
});
Object.defineProperty(exports, "locale_zh_CN", {
  enumerable: true,
  get: function () {
    return _zh_CN.default;
  }
});
Object.defineProperty(exports, "locale_zh_HK", {
  enumerable: true,
  get: function () {
    return _zh_HK.default;
  }
});
Object.defineProperty(exports, "locale_zh_TW", {
  enumerable: true,
  get: function () {
    return _zh_TW.default;
  }
});
Object.defineProperty(exports, "MarksPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.zhiFactory;
  }
});
Object.defineProperty(exports, "MobileMarksPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.mobileZhiFactory;
  }
});
Object.defineProperty(exports, "AudioPlugin", {
  enumerable: true,
  get: function () {
    return _audio.default;
  }
});
Object.defineProperty(exports, "MobileAudioPlugin", {
  enumerable: true,
  get: function () {
    return _mobile2.default;
  }
});
Object.defineProperty(exports, "BlockquotePlugin", {
  enumerable: true,
  get: function () {
    return _blockquote.default;
  }
});
Object.defineProperty(exports, "MobileBlockquotePlugin", {
  enumerable: true,
  get: function () {
    return _blockquote.default;
  }
});
Object.defineProperty(exports, "ClipboardPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginClipboard.zhiFactory;
  }
});
Object.defineProperty(exports, "MobileClipboardPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginClipboard.mobileZhiFactory;
  }
});
Object.defineProperty(exports, "FormulaPlugin", {
  enumerable: true,
  get: function () {
    return _formula.default;
  }
});
Object.defineProperty(exports, "MobileCodePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginCode.mobileZhiFactory;
  }
});
Object.defineProperty(exports, "CodePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginCode.zhiFactory;
  }
});
Object.defineProperty(exports, "InlineCodePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginInlineCode.zhiFactory;
  }
});
Object.defineProperty(exports, "ColumnsPlugin", {
  enumerable: true,
  get: function () {
    return _columns.default;
  }
});
Object.defineProperty(exports, "MobileColumnsPlugin", {
  enumerable: true,
  get: function () {
    return _mobile3.default;
  }
});
Object.defineProperty(exports, "CommentPlugin", {
  enumerable: true,
  get: function () {
    return _comment.default;
  }
});
Object.defineProperty(exports, "MobileCommentPlugin", {
  enumerable: true,
  get: function () {
    return _mobile4.default;
  }
});
Object.defineProperty(exports, "ContextMenuPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginPcContexMenu.zhiFactory;
  }
});
Object.defineProperty(exports, "EmbedPlugin", {
  enumerable: true,
  get: function () {
    return _embed.default;
  }
});
Object.defineProperty(exports, "MobileEmbedPlugin", {
  enumerable: true,
  get: function () {
    return _mobile5.default;
  }
});
Object.defineProperty(exports, "GroupTagPlugin", {
  enumerable: true,
  get: function () {
    return _groupTag.default;
  }
});
Object.defineProperty(exports, "MobileGroupTagPlugin", {
  enumerable: true,
  get: function () {
    return _groupTag.default;
  }
});
Object.defineProperty(exports, "HeadingPlugin", {
  enumerable: true,
  get: function () {
    return _heading.default;
  }
});
Object.defineProperty(exports, "MobileHeadingPlugin", {
  enumerable: true,
  get: function () {
    return _heading.default;
  }
});
Object.defineProperty(exports, "HrPlugin", {
  enumerable: true,
  get: function () {
    return _hr.default;
  }
});
Object.defineProperty(exports, "MobileHrPlugin", {
  enumerable: true,
  get: function () {
    return _hr.default;
  }
});
Object.defineProperty(exports, "ImagePlugin", {
  enumerable: true,
  get: function () {
    return _image.default;
  }
});
Object.defineProperty(exports, "MobileImagePlugin", {
  enumerable: true,
  get: function () {
    return _mobile6.default;
  }
});
Object.defineProperty(exports, "LinkPlugin", {
  enumerable: true,
  get: function () {
    return _link.default;
  }
});
Object.defineProperty(exports, "LinkCardPlugin", {
  enumerable: true,
  get: function () {
    return _index.default;
  }
});
Object.defineProperty(exports, "MobileLinkPlugin", {
  enumerable: true,
  get: function () {
    return _mobile7.default;
  }
});
Object.defineProperty(exports, "MobileLinkCardPlugin", {
  enumerable: true,
  get: function () {
    return _mobile8.default;
  }
});
Object.defineProperty(exports, "ListPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginList.zhiFactory;
  }
});
Object.defineProperty(exports, "MobileListPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginList.mobileZhiFactory;
  }
});
Object.defineProperty(exports, "VideoPlugin", {
  enumerable: true,
  get: function () {
    return _video.default;
  }
});
Object.defineProperty(exports, "MobileVideoPlugin", {
  enumerable: true,
  get: function () {
    return _mobile9.default;
  }
});
Object.defineProperty(exports, "MentionPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginMention.zhiFactory;
  }
});
Object.defineProperty(exports, "MobileMentionPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginMention.mobileZhiFactory;
  }
});
Object.defineProperty(exports, "OnlineVideoPlugin", {
  enumerable: true,
  get: function () {
    return _onlineVideo.default;
  }
});
Object.defineProperty(exports, "MobileOnlineVideoPlugin", {
  enumerable: true,
  get: function () {
    return _onlineVideo.default;
  }
});
Object.defineProperty(exports, "ParagraphPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginParagraph.zhiFactory;
  }
});
Object.defineProperty(exports, "MobileParagraphPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginParagraph.mobileZhiFactory;
  }
});
Object.defineProperty(exports, "PlaceholderPlugin", {
  enumerable: true,
  get: function () {
    return _placeholder.default;
  }
});
Object.defineProperty(exports, "MobilePlaceholderPlugin", {
  enumerable: true,
  get: function () {
    return _placeholder.default;
  }
});
Object.defineProperty(exports, "ReferPlugin", {
  enumerable: true,
  get: function () {
    return _refer.default;
  }
});
Object.defineProperty(exports, "MobileReferPlugin", {
  enumerable: true,
  get: function () {
    return _refer.default;
  }
});
Object.defineProperty(exports, "TablePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginTable.zhiFactory;
  }
});
Object.defineProperty(exports, "MobileTablePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginTable.mobileZhiFactory;
  }
});
Object.defineProperty(exports, "StickerPlugin", {
  enumerable: true,
  get: function () {
    return _sticker.default;
  }
});
Object.defineProperty(exports, "MobileStickerPlugin", {
  enumerable: true,
  get: function () {
    return _sticker.default;
  }
});
Object.defineProperty(exports, "ImTagPlugin", {
  enumerable: true,
  get: function () {
    return _imTag.default;
  }
});
Object.defineProperty(exports, "MobileImTagPlugin", {
  enumerable: true,
  get: function () {
    return _imTag.default;
  }
});
Object.defineProperty(exports, "CalendarCardPlugin", {
  enumerable: true,
  get: function () {
    return _calendarCard.default;
  }
});
Object.defineProperty(exports, "VoteCardPlugin", {
  enumerable: true,
  get: function () {
    return _voteCard.default;
  }
});
Object.defineProperty(exports, "FindAndReplacePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginFindAndReplace.createFindAndReplacePlugin;
  }
});
Object.defineProperty(exports, "ProjectCardPlugin", {
  enumerable: true,
  get: function () {
    return _projectCard.default;
  }
});
Object.defineProperty(exports, "MobileCalendarCardPlugin", {
  enumerable: true,
  get: function () {
    return _mobile10.default;
  }
});
Object.defineProperty(exports, "URLCardPlugin", {
  enumerable: true,
  get: function () {
    return _urlCard.default;
  }
});
Object.defineProperty(exports, "CalloutPlugin", {
  enumerable: true,
  get: function () {
    return _callout.default;
  }
});
Object.defineProperty(exports, "HeadingGroupPlugin", {
  enumerable: true,
  get: function () {
    return _headingGroup.default;
  }
});
Object.defineProperty(exports, "HeadingGroupNewPlugin", {
  enumerable: true,
  get: function () {
    return _headingGroupNew.default;
  }
});
Object.defineProperty(exports, "ListGroupPlugin", {
  enumerable: true,
  get: function () {
    return _listGroup.default;
  }
});
Object.defineProperty(exports, "SubjectPlugin", {
  enumerable: true,
  get: function () {
    return _subject.default;
  }
});
Object.defineProperty(exports, "cardFactory", {
  enumerable: true,
  get: function () {
    return _everFactory.zhiCardFactory;
  }
});
Object.defineProperty(exports, "groupFactory", {
  enumerable: true,
  get: function () {
    return _everFactory.zhiGroupFactory;
  }
});
Object.defineProperty(exports, "tagFactory", {
  enumerable: true,
  get: function () {
    return _everFactory.zhiTagFactory;
  }
});

var _createZhiPlugins = _interopRequireDefault(require("./createZhiPlugins"));

var _createMobileZhiPlugins = _interopRequireDefault(require("./createMobileZhiPlugins"));

var _createCustomizedZhiPlugins = _interopRequireDefault(require("./createCustomizedZhiPlugins"));

var _plugins = _interopRequireDefault(require("./plugins"));

var _mobile = _interopRequireDefault(require("./plugins/mobile"));

var _reader = _interopRequireDefault(require("./reader"));

var _en_US = _interopRequireDefault(require("./locales/en_US"));

var _ja_JP = _interopRequireDefault(require("./locales/ja_JP"));

var _zh_CN = _interopRequireDefault(require("./locales/zh_CN"));

var _zh_HK = _interopRequireDefault(require("./locales/zh_HK"));

var _zh_TW = _interopRequireDefault(require("./locales/zh_TW"));

var _everPluginMarks = require("@ali/4ever-plugin-marks");

var _audio = _interopRequireDefault(require("./plugins/audio"));

var _mobile2 = _interopRequireDefault(require("./plugins/audio/mobile"));

var _blockquote = _interopRequireDefault(require("./plugins/blockquote"));

var _everPluginClipboard = require("@ali/4ever-plugin-clipboard");

var _formula = _interopRequireDefault(require("./plugins/formula"));

var _everPluginCode = require("@ali/4ever-plugin-code");

var _everPluginInlineCode = require("@ali/4ever-plugin-inline-code");

var _columns = _interopRequireDefault(require("./plugins/columns"));

var _mobile3 = _interopRequireDefault(require("./plugins/columns/mobile"));

var _comment = _interopRequireDefault(require("./plugins/comment"));

var _mobile4 = _interopRequireDefault(require("./plugins/comment/mobile"));

var _everPluginPcContexMenu = require("@ali/4ever-plugin-pc-contex-menu");

var _embed = _interopRequireDefault(require("./plugins/embed"));

var _mobile5 = _interopRequireDefault(require("./plugins/embed/mobile"));

var _groupTag = _interopRequireDefault(require("./plugins/groupTag"));

var _heading = _interopRequireDefault(require("./plugins/heading"));

var _hr = _interopRequireDefault(require("./plugins/hr"));

var _image = _interopRequireDefault(require("./plugins/image"));

var _mobile6 = _interopRequireDefault(require("./plugins/image/mobile"));

var _link = _interopRequireDefault(require("./plugins/link"));

var _index = _interopRequireDefault(require("./plugins/link/linkCard/index"));

var _mobile7 = _interopRequireDefault(require("./plugins/link/mobile"));

var _mobile8 = _interopRequireDefault(require("./plugins/link/linkCard/mobile"));

var _everPluginList = require("@ali/4ever-plugin-list");

var _video = _interopRequireDefault(require("./plugins/video"));

var _mobile9 = _interopRequireDefault(require("./plugins/video/mobile"));

var _everPluginMention = require("@ali/4ever-plugin-mention");

var _onlineVideo = _interopRequireDefault(require("./plugins/onlineVideo"));

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _placeholder = _interopRequireDefault(require("./plugins/placeholder"));

var _refer = _interopRequireDefault(require("./plugins/refer"));

var _everPluginTable = require("@ali/4ever-plugin-table");

var _sticker = _interopRequireDefault(require("./plugins/sticker"));

var _imTag = _interopRequireDefault(require("./plugins/imTag"));

var _calendarCard = _interopRequireDefault(require("./plugins/calendarCard"));

var _voteCard = _interopRequireDefault(require("./plugins/voteCard"));

var _everPluginFindAndReplace = require("@ali/4ever-plugin-find-and-replace");

var _projectCard = _interopRequireDefault(require("./plugins/projectCard"));

var _mobile10 = _interopRequireDefault(require("./plugins/calendarCard/mobile"));

var _urlCard = _interopRequireDefault(require("./plugins/urlCard"));

var _callout = _interopRequireDefault(require("./plugins/callout"));

var _headingGroup = _interopRequireDefault(require("./plugins/headingGroup"));

var _headingGroupNew = _interopRequireDefault(require("./plugins/headingGroupNew"));

var _listGroup = _interopRequireDefault(require("./plugins/listGroup"));

var _subject = _interopRequireDefault(require("./plugins/subject"));

var _everFactory = require("@ali/4ever-factory");

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
