// @ali/4ever-component/dist-node/ToolbarNext/basicComponents/buttons/AlignButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AlignLeftButton = AlignLeftButton;
exports.AlignRightButton = AlignRightButton;
exports.AlignCenterButton = AlignCenterButton;
exports.AlignJustifyButton = AlignJustifyButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _Basics = require("./Basics");

var _icons = require("../icons");

const _createElement = /*#__PURE__*/React.createElement;

var _ref = /*#__PURE__*/_createElement(_icons.IconAlignLeft, null);

function AlignLeftButton(props) {
  const {
    icon = _ref,
    ...rest
  } = props;
  return /*#__PURE__*/_createElement(_Basics.IconButton, (0, _extends2.default)({
    testid: "align_left",
    icon: icon
  }, rest));
}

var _ref2 = /*#__PURE__*/_createElement(_icons.IconAlignRight, null);

function AlignRightButton(props) {
  const {
    icon = _ref2,
    ...rest
  } = props;
  return /*#__PURE__*/_createElement(_Basics.IconButton, (0, _extends2.default)({
    testid: "align_right",
    icon: icon
  }, rest));
}

var _ref3 = /*#__PURE__*/_createElement(_icons.IconAlignCenter, null);

function AlignCenterButton(props) {
  const {
    icon = _ref3,
    ...rest
  } = props;
  return /*#__PURE__*/_createElement(_Basics.IconButton, (0, _extends2.default)({
    testid: "align_center",
    icon: icon
  }, rest));
}

var _ref4 = /*#__PURE__*/_createElement(_icons.IconAlignJustify, null);

function AlignJustifyButton(props) {
  const {
    icon = _ref4,
    ...rest
  } = props;
  return /*#__PURE__*/_createElement(_Basics.IconButton, (0, _extends2.default)({
    testid: "align_justify",
    icon: icon
  }, rest));
}
