// @ali/4ever-component/dist-node/ToolbarNext/basicComponents/buttons/Basics.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "IconButton", {
  enumerable: true,
  get: function () {
    return _weDesignNext.TBIconButton;
  }
});
Object.defineProperty(exports, "IconSelect", {
  enumerable: true,
  get: function () {
    return _weDesignNext.TBIconSelect;
  }
});
Object.defineProperty(exports, "ItemSlot", {
  enumerable: true,
  get: function () {
    return _weDesignNext.TBItemSlot;
  }
});
Object.defineProperty(exports, "SplitButton", {
  enumerable: true,
  get: function () {
    return _weDesignNext.TBSplitButton;
  }
});
Object.defineProperty(exports, "TextSelect", {
  enumerable: true,
  get: function () {
    return _weDesignNext.TBTextSelect;
  }
});
Object.defineProperty(exports, "IconMenu", {
  enumerable: true,
  get: function () {
    return _weDesignNext.TBIconMenu;
  }
});
Object.defineProperty(exports, "ColorPicker", {
  enumerable: true,
  get: function () {
    return _weDesignNext.ColorPicker;
  }
});
Object.defineProperty(exports, "Menu", {
  enumerable: true,
  get: function () {
    return _weDesignNext.Menu;
  }
});
Object.defineProperty(exports, "List", {
  enumerable: true,
  get: function () {
    return _weDesignNext.List;
  }
});
Object.defineProperty(exports, "Tooltip", {
  enumerable: true,
  get: function () {
    return _weDesignNext.Tooltip;
  }
});
Object.defineProperty(exports, "FontFamilySelect", {
  enumerable: true,
  get: function () {
    return _weDesignNext.TBFontFamilySelect;
  }
});
Object.defineProperty(exports, "BorderPicker", {
  enumerable: true,
  get: function () {
    return _weDesignNext.BorderPicker;
  }
});
exports.FontSizeSelectSingleLineDefaultWidth = exports.TextTypeSingleLineDefaultWidth = exports.IconSelectSingleLineDefaultWidth = exports.SlotType = exports.SelectOption = void 0;

var _weDesignNext = require("@ali/we-design-next");

const SelectOption = _weDesignNext.Select.Option;
exports.SelectOption = SelectOption;
let SlotType;
exports.SlotType = SlotType;

(function (SlotType) {
  SlotType["iconButton"] = "iconButton";
  SlotType["iconMenu"] = "iconMenu";
  SlotType["splitButton"] = "splitButton";
  SlotType["iconSelect"] = "iconSelect";
  SlotType["textSelect"] = "textSelect";
  SlotType["custom"] = "custom";
})(SlotType || (exports.SlotType = SlotType = {}));

const IconSelectSingleLineDefaultWidth = 37;
exports.IconSelectSingleLineDefaultWidth = IconSelectSingleLineDefaultWidth;
const TextTypeSingleLineDefaultWidth = 70;
exports.TextTypeSingleLineDefaultWidth = TextTypeSingleLineDefaultWidth;
const FontSizeSelectSingleLineDefaultWidth = 49;
exports.FontSizeSelectSingleLineDefaultWidth = FontSizeSelectSingleLineDefaultWidth;
