// @ali/4ever-component/dist-node/ToolbarNext/basicComponents/buttons/FontSizeSelect.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FontSizeSelect = exports.szList = void 0;

var _react = _interopRequireWildcard(require("react"));

var _Basics = require("./Basics");

const _createElement = /*#__PURE__*/_react.default.createElement;
const szList = [9, 10, 11, 12, 14, 16, 18, 20, 22, 24, 26, 28, 36, 42, 48, 72];
exports.szList = szList;

const FontSizeSelect = props => {
  const {
    onChange,
    disabled,
    testid,
    value = 12,
    role,
    tooltip,
    options
  } = props;
  const szOptions = (0, _react.useMemo)(() => {
    const optionsInNumber = options || szList;
    return optionsInNumber.map(option => ({
      key: String(option),
      title: String(option),
      testid: `option-${option}`
    }));
  }, [options]);
  const handleChange = (0, _react.useCallback)(value => {
    try {
      const num = Number(value);

      if (!Number.isNaN(num)) {
        if (onChange) {
          onChange(num);
        }
      }
    } catch {}
  }, [onChange]);
  return /*#__PURE__*/_createElement(_Basics.TextSelect, {
    value: value > 0 ? String(value) : '0',
    disabled: disabled,
    role: role,
    width: _Basics.FontSizeSelectSingleLineDefaultWidth,
    options: szOptions,
    onChange: handleChange,
    mode: 'input',
    testid: testid || 'FontSizeSelect',
    tooltip: tooltip,
    "data-role": role || 'font_size_select'
  });
};

exports.FontSizeSelect = FontSizeSelect;
