// @ali/4ever-component/dist-node/ToolbarNext/basicComponents/buttons/HorizontalAlignSelect.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HorizontalAlignSelect = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _icons = require("../icons");

var _Basics = require("./Basics");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement(_icons.IconAlignLeft, null);

var _ref2 = /*#__PURE__*/_createElement(_icons.IconAlignCenter, null);

var _ref3 = /*#__PURE__*/_createElement(_icons.IconAlignRight, null);

var _ref4 = /*#__PURE__*/_createElement(_icons.IconAlignJustify, null);

const HorizontalAlignSelect = props => {
  const {
    onChange,
    options,
    role,
    value,
    tooltip,
    ...rest
  } = props; // 垂直对齐的选项

  const normalizeOptions = (0, _react.useMemo)(() => [{
    title: options.left.title,
    value: 'left',
    key: 'left',
    icon: options.left.icon || _ref,
    testid: 'option-left'
  }, {
    title: options.center.title,
    value: 'center',
    key: 'center',
    icon: options.center.icon || _ref2,
    testid: 'option-center'
  }, {
    title: options.right.title,
    value: 'right',
    key: 'right',
    icon: options.right.icon || _ref3,
    testid: 'option-right'
  }, {
    title: options.justify.title,
    value: 'justify',
    key: 'justify',
    icon: options.justify.icon || _ref4,
    testid: 'option-justify'
  }], [options]);
  const handleChange = (0, _react.useCallback)(e => {
    if (onChange) {
      onChange(e);
    }
  }, [onChange]);
  return /*#__PURE__*/_createElement(_Basics.IconSelect, (0, _extends2.default)({
    value: value,
    onChange: handleChange,
    testid: "HorizontalAlignSelect",
    tooltip: tooltip,
    role: role,
    options: normalizeOptions,
    width: _Basics.IconSelectSingleLineDefaultWidth
  }, rest));
};

exports.HorizontalAlignSelect = HorizontalAlignSelect;
