// @ali/4ever-component/dist-node/ToolbarNext/basicComponents/buttons/ListSet.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListSet = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Basics = require("./Basics");

var _icons = require("../icons");

var _classnames = _interopRequireDefault(require("classnames"));

const _createElement = /*#__PURE__*/_react.default.createElement;
const DEFAULT_BULLET_LIST_ARRAY = ['SCIR_ECIR_SREC', 'SFDI_SRIG_SREC', 'EREC_EREC_EREC', 'SARR_SDIO_SCIR', 'SSTA_ECIR_SREC', 'SRIG_ECIR_SREC'];
const DEFAULT_STICKER_LIST_ARRAY = ['VACATION_TEA_MILKTEA', 'RIGHT_NAIL_LOVE', 'FIRE_ROCKET_NAIL'];
const DEFAULT_NUMBER_LIST_ARRAY = ['DEC_LEN_LROM_P', 'DEC_LEN_LROM_H', 'DEC_DEC_DEC_P', 'UEN_LEN_LROM_P', 'UROM_UEN_DEC_P', 'DDEC_LEN_LROM_P', 'ZH_DEC_LEN_D', 'ZH_DEC_LEN_A', 'ZH_ZH_DEC_D'];
const LIST_STYLE_MAP = {
  SCIR_ECIR_SREC: 'https://img.alicdn.com/tfs/TB1mWdYbAcx_u4jSZFlXXXnUFXa-78-78.svg',
  SFDI_SRIG_SREC: 'https://img.alicdn.com/tfs/TB1hpepOpP7gK0jSZFjXXc5aXXa-78-78.svg',
  EREC_EREC_EREC: 'https://img.alicdn.com/tfs/TB1cJR4czMZ7e4jSZFOXXX7epXa-78-78.svg',
  SARR_SDIO_SCIR: 'https://img.alicdn.com/tfs/TB1aaWcOEY1gK0jSZFMXXaWcVXa-78-78.svg',
  SSTA_ECIR_SREC: 'https://img.alicdn.com/tfs/TB1VoX8OxD1gK0jSZFsXXbldVXa-78-78.svg',
  SRIG_ECIR_SREC: 'https://img.alicdn.com/tfs/TB1cqp4OAL0gK0jSZFtXXXQCXXa-78-78.svg',
  VACATION_TEA_MILKTEA: 'https://img.alicdn.com/imgextra/i4/O1CN01iGy3OC23bXk0A5V99_!!6000000007274-2-tps-240-240.png',
  RIGHT_NAIL_LOVE: 'https://img.alicdn.com/imgextra/i4/O1CN01rYMRza1OM46kvi2Yx_!!6000000001690-2-tps-240-240.png',
  FIRE_ROCKET_NAIL: 'https://img.alicdn.com/imgextra/i3/O1CN01QfDOFq1R49SyPKcpW_!!6000000002057-2-tps-240-240.png',
  DEC_LEN_LROM_P: 'https://img.alicdn.com/tfs/TB19lt9OEz1gK0jSZLeXXb9kVXa-78-78.svg',
  DEC_LEN_LROM_H: 'https://img.alicdn.com/tfs/TB18HydOAL0gK0jSZFAXXcA9pXa-78-78.svg',
  DEC_DEC_DEC_P: 'https://img.alicdn.com/tfs/TB1gHycOqL7gK0jSZFBXXXZZpXa-78-78.svg',
  UEN_LEN_LROM_P: 'https://img.alicdn.com/tfs/TB1VaOnOBr0gK0jSZFnXXbRRXXa-78-78.svg',
  UROM_UEN_DEC_P: 'https://img.alicdn.com/tfs/TB1xcqhOAY2gK0jSZFgXXc5OFXa-78-78.svg',
  DDEC_LEN_LROM_P: 'https://img.alicdn.com/tfs/TB1yjive8Bh1e4jSZFhXXcC9VXa-78-78.svg',
  ZH_DEC_LEN_D: 'https://img.alicdn.com/tfs/TB1bFQZ0aL7gK0jSZFBXXXZZpXa-82-82.svg',
  ZH_DEC_LEN_A: 'https://img.alicdn.com/tfs/TB11.iDj4vbeK8jSZPfXXariXXa-82-82.svg',
  ZH_ZH_DEC_D: 'https://img.alicdn.com/imgextra/i4/O1CN01XYORsY1OjQIxXuZxz_!!6000000001741-55-tps-82-82.svg'
};

const OverlayContent = /*#__PURE__*/_styledComponents.default.div(["width:272px;display:flex;flex-wrap:wrap;justify-content:flex-start;padding:4px;"]);

const ListStyleItemWrap = /*#__PURE__*/_styledComponents.default.div(["user-select:none;"]);

const ListStyleItem = /*#__PURE__*/_styledComponents.default.img(["width:80px;height:80px;margin:4px;border:1px solid rgba(17,31,44,0.12);border-radius:2px;cursor:pointer;image-rendering:pixelated;&:hover{border:1px solid #0089FF;}&.actived{border:2px solid #0089FF;}"]);

function getListRole(mode, key) {
  const prefix = mode === 'ul' ? 'listUnordered' : 'listOrdered';
  return `${prefix}_${key}`;
}

function StopEvent(event) {
  event.stopPropagation();
}

var _ref = /*#__PURE__*/_createElement(_icons.IconUnorderList, null);

var _ref2 = /*#__PURE__*/_createElement(_icons.IconOrderList, null);

const ListSet = props => {
  const {
    mode = 'ul',
    defaultListStyle,
    onChange,
    onClick,
    disabled,
    icon,
    enableSticker,
    active,
    ...rest
  } = props;
  const listStyleArray = (0, _react.useMemo)(() => {
    if (mode !== 'ul') {
      return DEFAULT_NUMBER_LIST_ARRAY;
    } else {
      return enableSticker ? DEFAULT_BULLET_LIST_ARRAY.concat(DEFAULT_STICKER_LIST_ARRAY) : DEFAULT_BULLET_LIST_ARRAY;
    }
  }, [mode]);
  const [value, setValue] = (0, _react.useState)(defaultListStyle || listStyleArray[0]);
  const [overlayVisible, setOverlayVisible] = (0, _react.useState)(false);
  const handleChange = (0, _react.useCallback)(e => {
    if (onChange) {
      onChange(e);
    }

    setValue(e);
    setOverlayVisible(false);
  }, [onChange, setValue]);
  const handleVisibleChange = (0, _react.useCallback)(visible => {
    setOverlayVisible(visible);
  }, []);
  const handleClickLeft = (0, _react.useCallback)(() => {
    if (onClick) {
      onClick(value);
    }
  }, [onClick, value]);
  const getOverlay = (0, _react.useMemo)(() => {
    return /*#__PURE__*/_createElement(OverlayContent, {
      "data-testid": "overlay-ListSet",
      onClick: StopEvent
    }, listStyleArray.map(item => {
      return /*#__PURE__*/_createElement(ListStyleItemWrap, {
        key: item,
        onClick: () => handleChange(item),
        "data-role": getListRole(mode, item),
        role: "button"
      }, /*#__PURE__*/_createElement(ListStyleItem, {
        src: LIST_STYLE_MAP[item],
        alt: "list-style-item",
        className: (0, _classnames.default)({
          actived: item === value
        }),
        "data-testid": `option-${item}`
      }));
    }));
  }, [value]);
  const renderTriggerIcon = (0, _react.useMemo)(() => {
    if (icon) {
      return icon;
    }

    return mode === 'ul' ? _ref : _ref2;
  }, [mode, icon]);
  return /*#__PURE__*/_createElement(_Basics.SplitButton, (0, _extends2.default)({
    overlay: getOverlay,
    disabled: disabled,
    onClick: handleClickLeft,
    overlayVisible: overlayVisible,
    onVisibleChange: handleVisibleChange,
    icon: renderTriggerIcon,
    active: active || overlayVisible
  }, rest));
};

exports.ListSet = ListSet;
