// @ali/4ever-component/dist-node/ToolbarNext/basicComponents/buttons/TranslationSelect.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TranslationSelect = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _icons = require("../icons");

var _react = _interopRequireWildcard(require("react"));

var _Basics = require("./Basics");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement(_icons.IconTranslate, null);

const TranslationSelect = props => {
  const {
    onChange,
    testid,
    menuItems,
    ...restProps
  } = props;

  const [overlayVisible, setOverlayVisible] = _react.default.useState(false);

  const items = (0, _react.useMemo)(() => {
    const options = menuItems.map(({
      key,
      title
    }) => ({
      key: String(key),
      title: String(title),
      testid: `option-${key}`
    }));
    return options;
  }, [menuItems]);
  const handleSelect = (0, _react.useCallback)(([v]) => {
    onChange && onChange(v);
    setOverlayVisible(false);
  }, [onChange]);
  return /*#__PURE__*/_createElement(_Basics.IconMenu, (0, _extends2.default)({
    icon: _ref,
    overlay: /*#__PURE__*/_createElement(_Basics.Menu, {
      "data-testid": `${testid}-options`,
      items: items,
      onSelect: handleSelect
    }),
    overlayVisible: overlayVisible,
    onVisibleChange: setOverlayVisible,
    testid: testid
  }, restProps));
};

exports.TranslationSelect = TranslationSelect;
