// @ali/4ever-component/dist-node/ToolbarNext/basicComponents/buttons/VerticalAlignSelect.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VerticalAlignSelect = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _icons = require("../icons");

var _Basics = require("./Basics");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement(_icons.IconVerticalAlignTop, null);

var _ref2 = /*#__PURE__*/_createElement(_icons.IconVerticalAlignMiddle, null);

var _ref3 = /*#__PURE__*/_createElement(_icons.IconVerticalAlignBottom, null);

const VerticalAlignSelect = props => {
  const {
    onChange,
    options,
    value,
    ...rest
  } = props; // 垂直对齐的选项

  const normalizeOptions = (0, _react.useMemo)(() => [{
    title: options.top.title,
    value: 'top',
    key: 'top',
    icon: options.top.icon || _ref,
    testid: 'option-top'
  }, {
    title: options.middle.title,
    value: 'middle',
    key: 'middle',
    icon: options.middle.icon || _ref2,
    testid: 'option-middle'
  }, {
    title: options.bottom.title,
    value: 'bottom',
    key: 'bottom',
    icon: options.bottom.icon || _ref3,
    testid: 'option-bottom'
  }], [options]);
  const handleChange = (0, _react.useCallback)(e => {
    if (onChange) {
      onChange(e);
    }
  }, [onChange]);
  return /*#__PURE__*/_createElement(_Basics.IconSelect, (0, _extends2.default)({
    value: value,
    onChange: handleChange,
    testid: "VerticalAlignSelect",
    options: normalizeOptions // width={IconSelectSingleLineDefaultWidth}

  }, rest));
};

exports.VerticalAlignSelect = VerticalAlignSelect;
