// @ali/4ever-plugin-im-tag/dist-node/utils/bamboo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "IMTAG_DEFAULT_HEIGHT", {
  enumerable: true,
  get: function () {
    return _constants.IMTAG_DEFAULT_HEIGHT;
  }
});
Object.defineProperty(exports, "IMTAG_MAX_WIDTH", {
  enumerable: true,
  get: function () {
    return _constants.IMTAG_MAX_WIDTH;
  }
});
Object.defineProperty(exports, "IMTAG_MEASURE_STRING", {
  enumerable: true,
  get: function () {
    return _constants.IMTAG_MEASURE_STRING;
  }
});
Object.defineProperty(exports, "IMTAG_MIN_CONTAINER_WIDTH", {
  enumerable: true,
  get: function () {
    return _constants.IMTAG_MIN_CONTAINER_WIDTH;
  }
});
Object.defineProperty(exports, "IMTAG_INIT_WIDTH", {
  enumerable: true,
  get: function () {
    return _constants.IMTAG_INIT_WIDTH;
  }
});
Object.defineProperty(exports, "TextWrapper", {
  enumerable: true,
  get: function () {
    return _styled.TextWrapper;
  }
});
Object.defineProperty(exports, "ImTagContainer", {
  enumerable: true,
  get: function () {
    return _styled.ImTagContainer;
  }
});

var _constants = require("./constants");

var _styled = require("./styled");
