// @ali/4ever-plugin-im-tag/dist-node/utils/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImTagContainer = exports.TextWrapper = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _weDesign = require("@ali/we-design");

var _constants = require("./constants");

const _createElement = /*#__PURE__*/_react.default.createElement;
;

const ContainerWrapper = /*#__PURE__*/_styledComponents.default.span(["font-size:14px;cursor:pointer;display:inline-flex;border-radius:12px;border:1px solid rgba(80,116,154,0.12);cursor:auto;width:", ";height:", "px;line-height:", "px;&:hover{background-color:rgba(17,31,44,0.08);border:1px solid rgba(79,119,172,0.12);}"], props => props.width, props => props.height, props => props.height);

const ContentWrapper = /*#__PURE__*/_styledComponents.default.span(["margin:0 7px;color:#50749A;letter-spacing:0;cursor:pointer;display:flex;line-height:calc(", "px - 2px);"], props => props.height);

const TextWrapper = /*#__PURE__*/_styledComponents.default.span(["width:", ";display:inline-block;letter-spacing:0;white-space:nowrap;font-size:14px;overflow:hidden;text-overflow:ellipsis;flex-grow:1;"], props => props.width ? props.width - _constants.IMTAG_MIN_CONTAINER_WIDTH + 'px' : 'auto');

exports.TextWrapper = TextWrapper;

const IconWrapper = /*#__PURE__*/_styledComponents.default.span(["width:14px;text-align:center;display:inline-flex;& > span{align-self:center;}"]);

var _ref = /*#__PURE__*/_createElement(IconWrapper, null, /*#__PURE__*/_createElement(_weDesign.ArrowRightNormal, null));

const ImTagContainer = props => {
  const {
    height,
    width,
    text,
    onClick,
    onMouseDown
  } = props;
  return /*#__PURE__*/_createElement(ContainerWrapper, {
    height: height || _constants.IMTAG_DEFAULT_HEIGHT,
    width: width + 'px',
    onClick: onClick,
    onMouseDown: onMouseDown
  }, /*#__PURE__*/_createElement(ContentWrapper, {
    height: height || _constants.IMTAG_DEFAULT_HEIGHT
  }, /*#__PURE__*/_createElement(TextWrapper, {
    width: width
  }, text), _ref));
};

exports.ImTagContainer = ImTagContainer;
