// @ali/we-project-card-placeholder/dist-node/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var styled = require('styled-components');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var styled__default = /*#__PURE__*/_interopDefaultLegacy(styled);

const Root = styled__default['default'].div`
  height: 52px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  background-color: #F1F3F6;
  border-radius: 12px 12px 0 0;
`;
const Left = styled__default['default'].div`
  flex: 1 1 auto;
  display: flex;
  flex-direction: row;
  align-items: center;
`;
const Right = styled__default['default'].div`
  flex: 1 1 auto;
  display: flex;
  flex-direction: row-reverse;
  align-items: center;
  padding-right: 16px;
`;
const ToolbarImg = styled__default['default'].img`
  margin: 12px;
  width: 20px;
  height: 20px;
`;
const ToolbarName = styled__default['default'].a`
  color:#111F2C;
  text-decoration: none;
  font-weight: 500;
  font-size: 16px;
  line-height: 20px;
  ${truncate('170px')}

  &:visited {color:#111F2C;}
  &:hover {
    color:#5B7DED;
    text-decoration: underline;
  }
  &:active {color:#111F2C;}
`;
function truncate(width) {
  return `
    max-width: ${width};
    word-break: keep-all;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  `;
}

const _createElement = /*#__PURE__*/React__default['default'].createElement;

var _ref = /*#__PURE__*/_createElement(ToolbarImg, {
  src: "https://img.alicdn.com/tfs/TB1sNayNfb2gK0jSZK9XXaEgFXa-52-48.png"
});

function Toolbar(props) {
  return /*#__PURE__*/_createElement(Root, null, /*#__PURE__*/_createElement(Left, null, _ref, /*#__PURE__*/_createElement(ToolbarName, {
    href: props.url
  }, props.name)), /*#__PURE__*/_createElement(Right, null, props.children));
}

const _createElement$1 = /*#__PURE__*/React__default['default'].createElement;
const Container = styled__default['default'].div`
  flex: 0;
  display: flex;
  flex-direction: column;
  border-radius: 12px;
  min-height: 200px;
  background-color: #F1F3F6;
`;
const Content = styled__default['default'].div`
  flex: 1 1 auto;
  display: flex;
  background-color: white;
  border-radius: 12px;
  border: 1px solid #DDDDDD;
`;
function CardContainer(props) {
  return /*#__PURE__*/_createElement$1(Container, {
    style: { ...props.style
    }
  }, /*#__PURE__*/_createElement$1(Toolbar, {
    name: props.name,
    url: props.url
  }), /*#__PURE__*/_createElement$1(Content, {
    style: { ...props.contentStyle
    }
  }, props.children));
}

const _createElement$2 = /*#__PURE__*/React__default['default'].createElement;
const Container$1 = styled__default['default'].div`
  flex: 1 1 auto;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;
const Title = styled__default['default'].div`
  font-size: 28px;
  margin-top: 32px;
  margin-bottom: 28px;
`;
const Description = styled__default['default'].span`
  font-size: 12px;
  ${truncate('300px')}
`;
const ProjectInfoCard = (props = {
  description: '创建者：未知',
  title: '未加入当前项目'
}) => {
  return /*#__PURE__*/_createElement$2(Container$1, null, /*#__PURE__*/_createElement$2(Title, null, props.title || ''), /*#__PURE__*/_createElement$2(Description, {
    title: props.description || ''
  }, props.description || ''));
};

const _createElement$3 = /*#__PURE__*/React__default['default'].createElement;
const Printable = props => {
  return /*#__PURE__*/_createElement$3(CardContainer, {
    style: props.style,
    name: props.name,
    url: props.url
  }, /*#__PURE__*/_createElement$3(ProjectInfoCard, {
    title: props.title
  }));
};

exports.PrintableCard = Printable;
