// @ali/4ever-plugin-table/dist-node/bi/utils/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "adjustSelection", {
  enumerable: true,
  get: function () {
    return _adjustSelection.default;
  }
});
Object.defineProperty(exports, "createEmptyParagraph", {
  enumerable: true,
  get: function () {
    return _createEmptyParagraph.default;
  }
});
Object.defineProperty(exports, "createEmptyTableCell", {
  enumerable: true,
  get: function () {
    return _createEmptyTableCell.default;
  }
});
Object.defineProperty(exports, "createTableRow", {
  enumerable: true,
  get: function () {
    return _createTableRow.default;
  }
});
Object.defineProperty(exports, "createTableData", {
  enumerable: true,
  get: function () {
    return _createTableData.default;
  }
});
Object.defineProperty(exports, "createTableSelectionForFocusedCell", {
  enumerable: true,
  get: function () {
    return _createTableSelectionForFocusedCell.default;
  }
});
Object.defineProperty(exports, "generateClipTableFromTableSelection", {
  enumerable: true,
  get: function () {
    return _generateClipTableFromTableSelection.default;
  }
});
Object.defineProperty(exports, "generateRangesFromTableSelection", {
  enumerable: true,
  get: function () {
    return _generateRangesFromTableSelection.default;
  }
});
Object.defineProperty(exports, "getMirrorCellOfSameBorder", {
  enumerable: true,
  get: function () {
    return _getMirrorCellOfSameBorder.default;
  }
});
Object.defineProperty(exports, "getPositionOfCell", {
  enumerable: true,
  get: function () {
    return _getPositionOfCell.default;
  }
});
Object.defineProperty(exports, "getPositionOfRow", {
  enumerable: true,
  get: function () {
    return _getPositionOfRow.default;
  }
});
Object.defineProperty(exports, "getRealNodeInTable", {
  enumerable: true,
  get: function () {
    return _getRealNodeInTable.default;
  }
});
Object.defineProperty(exports, "isEmptyTable", {
  enumerable: true,
  get: function () {
    return _isEmptyTable.default;
  }
});
Object.defineProperty(exports, "isMoreThanNCell", {
  enumerable: true,
  get: function () {
    return _isMoreThanNCell.default;
  }
});
Object.defineProperty(exports, "isSelectionAcrossTable", {
  enumerable: true,
  get: function () {
    return _isSelectionAcrossTable.default;
  }
});
Object.defineProperty(exports, "isSelectionInTable", {
  enumerable: true,
  get: function () {
    return _isSelectionInTable.default;
  }
});
Object.defineProperty(exports, "isSelectionInTableCell", {
  enumerable: true,
  get: function () {
    return _isSelectionInTableCell.default;
  }
});
Object.defineProperty(exports, "isSelectionInTableEdge", {
  enumerable: true,
  get: function () {
    return _isSelectionInTableEdge.default;
  }
});
Object.defineProperty(exports, "resizeTable", {
  enumerable: true,
  get: function () {
    return _resizeTable.default;
  }
});
Object.defineProperty(exports, "traverseAffectedColCells", {
  enumerable: true,
  get: function () {
    return _traverseAffectedColCells.default;
  }
});
Object.defineProperty(exports, "traverseAffectedRowCells", {
  enumerable: true,
  get: function () {
    return _traverseAffectedRowCells.default;
  }
});
Object.defineProperty(exports, "traverseSingleColCells", {
  enumerable: true,
  get: function () {
    return _traverseSingleColCells.default;
  }
});
Object.defineProperty(exports, "traverseSingleRowCells", {
  enumerable: true,
  get: function () {
    return _traverseSingleRowCells.default;
  }
});
Object.defineProperty(exports, "hasIntersectionObserver", {
  enumerable: true,
  get: function () {
    return _hasIntersectionObserver.default;
  }
});
Object.defineProperty(exports, "getTable", {
  enumerable: true,
  get: function () {
    return _getTable.default;
  }
});
Object.defineProperty(exports, "getTableSelection", {
  enumerable: true,
  get: function () {
    return _getTableSelection.default;
  }
});
Object.defineProperty(exports, "handleCangjiePaste", {
  enumerable: true,
  get: function () {
    return _handleCangjiePaste.handleCangjiePaste;
  }
});

var _adjustSelection = _interopRequireDefault(require("./adjustSelection"));

var _createEmptyParagraph = _interopRequireDefault(require("./createEmptyParagraph"));

var _createEmptyTableCell = _interopRequireDefault(require("./createEmptyTableCell"));

var _createTableRow = _interopRequireDefault(require("./createTableRow"));

var _createTableData = _interopRequireDefault(require("./createTableData"));

var _createTableSelectionForFocusedCell = _interopRequireDefault(require("./createTableSelectionForFocusedCell"));

var _generateClipTableFromTableSelection = _interopRequireDefault(require("./generateClipTableFromTableSelection"));

var _generateRangesFromTableSelection = _interopRequireDefault(require("./generateRangesFromTableSelection"));

var _getMirrorCellOfSameBorder = _interopRequireDefault(require("./getMirrorCellOfSameBorder"));

var _getPositionOfCell = _interopRequireDefault(require("./getPositionOfCell"));

var _getPositionOfRow = _interopRequireDefault(require("./getPositionOfRow"));

var _getRealNodeInTable = _interopRequireDefault(require("./getRealNodeInTable"));

var _isEmptyTable = _interopRequireDefault(require("./isEmptyTable"));

var _isMoreThanNCell = _interopRequireDefault(require("./isMoreThanNCell"));

var _isSelectionAcrossTable = _interopRequireDefault(require("./isSelectionAcrossTable"));

var _isSelectionInTable = _interopRequireDefault(require("./isSelectionInTable"));

var _isSelectionInTableCell = _interopRequireDefault(require("./isSelectionInTableCell"));

var _isSelectionInTableEdge = _interopRequireDefault(require("./isSelectionInTableEdge"));

var _resizeTable = _interopRequireDefault(require("./resizeTable"));

var _traverseAffectedColCells = _interopRequireDefault(require("./traverseAffectedColCells"));

var _traverseAffectedRowCells = _interopRequireDefault(require("./traverseAffectedRowCells"));

var _traverseSingleColCells = _interopRequireDefault(require("./traverseSingleColCells"));

var _traverseSingleRowCells = _interopRequireDefault(require("./traverseSingleRowCells"));

var _hasIntersectionObserver = _interopRequireDefault(require("./hasIntersectionObserver"));

var _getTable = _interopRequireDefault(require("./getTable"));

var _getTableSelection = _interopRequireDefault(require("./getTableSelection"));

var _handleCangjiePaste = require("./handleCangjiePaste");
