// @ali/4ever-bi/dist-web/plugins/sticker/mobile/index.js
import _extends from "@babel/runtime/helpers/extends";
import { StickerPlugin } from '@ali/4ever-bamboo';
import createRenderNode from "../renderNode";
import onCangjieSelect from "../handlers/onCangjieSelect";
import onAction from "../handlers/onAction";
import onKeyDown from "../handlers/onKeyDown";
import { DEFAULT_LANG, STICKER_NODE_TYPE } from "../constants";
import { moFactory } from '@ali/4ever-plugin-sticker';
var DEFAULT_STICKERS = StickerPlugin.DEFAULT_STICKERS,
    Sticker = StickerPlugin.Sticker;

function createStickerPlugin(config) {
  var stickerConfig = _extends({
    lang: DEFAULT_LANG,
    customSticker: [DEFAULT_STICKERS]
  }, config.sticker);

  var sticker = new Sticker(stickerConfig);
  var params = {
    stickerConfig: stickerConfig,
    sticker: sticker
  };
  return _extends({}, moFactory({}), {
    // @ts-ignore
    renderNode: createRenderNode(params),
    onAction: onAction,
    onKeyDown: onKeyDown,
    onCangjieSelect: onCangjieSelect,
    models: {
      klass: 'inline',
      type: STICKER_NODE_TYPE
    }
  });
}

createStickerPlugin.defaultSticker = DEFAULT_STICKERS;
export default createStickerPlugin;
