// @ali/4ever-logger/dist-node/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var WeLogger = require('@ali/we-logger-main');
var DebugAppender = require('@ali/we-logger-appender-debug');
var packageJSON = require('@ali/4ever-logger/package.json');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var WeLogger__default = /*#__PURE__*/_interopDefaultLegacy(WeLogger);
var DebugAppender__default = /*#__PURE__*/_interopDefaultLegacy(DebugAppender);
var packageJSON__default = /*#__PURE__*/_interopDefaultLegacy(packageJSON);

const {
  version
} = packageJSON__default['default'];
/**
 * @see https://arms.console.aliyun.com/retcode?pid=en91t6prdd%4081db28b43ff68d6#/index
 */

const PID = 'en91t6prdd@81db28b43ff68d6';
/**
 * 判断是否是本地环境
 */

const IS_LOCAL = process.env.NODE_ENV === 'development';
const IS_TEST = process.env.NODE_ENV === 'test';

(function (EnvEnum) {
  EnvEnum["DAILY"] = "daily";
  EnvEnum["PRE"] = "pre";
  EnvEnum["PROD"] = "prod";
  EnvEnum["LOCAL"] = "local";
  EnvEnum["UNKNOWN"] = "unknown";
})(exports.EnvEnum || (exports.EnvEnum = {}));

const substr2EnvMap = new Map([['pre-', exports.EnvEnum.PRE], ['daily-', exports.EnvEnum.DAILY]]);
const getEnvironment = () => {
  const {
    location: {
      hostname
    }
  } = window;

  for (const [substr, env] of substr2EnvMap) {
    if (hostname.startsWith(substr)) {
      return env;
    }
  }

  return exports.EnvEnum.PROD;
};
/**
 * 默认需要一个 debug appender, 外层业务也可以主动移除
 */

const debugAppender = new DebugAppender__default['default']('EDITOR:LOGGER');
/**
 * 默认配置好一个 arms appender, 需要外层业务手动注入才能够生效
 * @see https://arms.console.aliyun.com/retcode?pid=en91t6prdd%4081db28b43ff68d6#/index
 */

const getArmsAppender = function () {
  if (IS_TEST) {
    return null;
  } // 单测环境下 Arms 会在 Console 抛错，这里做一个优化处理
  // eslint-disable-next-line no-underscore-dangle


  const _ArmsAppender = require('@ali/we-logger-appender-arms').default;

  return new _ArmsAppender({
    release: version,
    pid: PID,
    appType: 'web',
    // SDK 层不监听 hashchange
    enableSPA: false,
    // SDK 层不监听 HTTP
    disableHook: true,
    // 本地开发环境不上报
    disabled: IS_LOCAL,
    environment: getEnvironment()
  });
};

class Logger extends WeLogger__default['default'] {
  constructor(...args) {
    super(...args);
    this.armsAppender = null;
    this.sentPV = false;
    this.bizName = location.host;
  }

  setEnabled(flag = true) {
    if (flag) {
      if (!this.armsAppender) {
        this.armsAppender = getArmsAppender();
        this.armsAppender && this.append(this.armsAppender);
      }

      if (!this.sentPV && this.armsAppender) {
        // Note: page's value
        this.armsAppender.getArms().setPage(location.href);
        this.armsAppender.sum(`BIZ: ${this.bizName}`);
        this.sentPV = true;
      }
    } else if (this.armsAppender) {
      this.remove(this.armsAppender);
      this.armsAppender = null;
    }
  }

  setBizName(name) {
    this.bizName = name;
    return this;
  }

}

const logger = new Logger().append(debugAppender);

exports.debugAppender = debugAppender;
exports.default = logger;
exports.getArmsAppender = getArmsAppender;
exports.getEnvironment = getEnvironment;
