// @ali/we-toolbar/dist-web/biz/splitButton/colorPicker/ColorPalette.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _SomecolorNormal;

import React, { PureComponent } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { ArrowRightNormal, SomecolorNormal, Select, SelectOption } from '@ali/we-design';
import { TranslateComp } from '@ali/we-util';
import { colorDataNormal, colorDataArt, colorDataArtWithGradient } from "./colorData";
import { LineGroupNormal, LineGroupArt, PaletteWrapper, CustomGroup, CustomPicker, CustomText, DefaultGroup, LatestColorWrap, LatestColorText, CustomGroupInnerWrap, MainColorGroup, Header } from "./styled";
import { classnames } from "../../../utils";
import { ColorBlockItem } from "./ColorBlockItem";
import SketchPicker from "./sketch";
import { colorUtil } from "./colorUtil";
import useClickOuterRef from "../../../hooks/useClickOuterRef";
export var EColorPanelMode;

(function (EColorPanelMode) {
  EColorPanelMode["normal"] = "normal";
  EColorPanelMode["art"] = "art";
})(EColorPanelMode || (EColorPanelMode = {}));

var decimalToHex = function decimalToHex(alpha) {
  return alpha === 0 ? '00' : Math.round(255 * alpha).toString(16).padStart(2, '0');
};

// 双倍间距
var doubleLineStyle = {
  paddingBottom: '10px'
};
var CUSTOMER_CONFIG_KEY = 'color-picker-custom-config';
var MAX_LATEST_COLOR = 10;

var ClickOuterRef = function ClickOuterRef(props) {
  var handler = props.handler,
      on = props.on,
      box = props.box,
      children = props.children;
  var ref = useClickOuterRef(handler, on, 'mousedown', box);
  return children(ref);
};

export var ColorPalette = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(ColorPalette, _PureComponent);

  // eslint-disable-next-line react/static-property-placement
  function ColorPalette(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.refPalette = /*#__PURE__*/React.createRef();
    _this.moreButtonRef = /*#__PURE__*/React.createRef();
    _this.timer = 0;

    _this.handlePick = function (color) {
      var onPick = _this.props.onPick;
      onPick(color);
    };

    _this.handleCustomClick = function (e) {
      // 阻止 更多颜色面板的 点击事件冒泡
      e.preventDefault();
      e.stopPropagation();
    };

    _this.showHandleCustomPicker = function (e) {
      var picker = _this.props.picker;

      if (!picker) {
        return;
      }

      if (_this.timer) window.clearTimeout(_this.timer);
      var target = e.target; // 确认按钮不阻止事件流

      if (!target.id || target.id !== CUSTOMER_CONFIG_KEY) {
        e.preventDefault();
        e.stopPropagation();
      }

      var showPicker = _this.state.showPicker;
      if (showPicker) return;

      _this.changePickerAlignment();

      _this.setState({
        showPicker: true
      });
    };

    _this.hideHandleCustomPicker = function () {
      _this.timer = window.setTimeout(function () {
        var customColor = _this.state.customColor;

        _this.addRecentColor(customColor);

        _this.setState({
          showPicker: false
        });
      }, 200);
    };

    _this.changePickerAlignment = function () {
      if (!_this.refPalette.current) {
        return;
      }

      var _this$refPalette$curr = _this.refPalette.current.getBoundingClientRect(),
          left = _this$refPalette$curr.left,
          width = _this$refPalette$curr.width;

      var right = window.innerWidth - left - width; // SketchPicker 的宽度为 200px

      if (right <= 210) {
        _this.setState({
          isPickerAlignRight: false
        });
      } else {
        _this.setState({
          isPickerAlignRight: true
        });
      }
    };

    _this.getColorStringColorObj = function (colorObj) {
      var result = colorObj.hex; // 这里可能是 "transparent"

      if (result.startsWith('#')) {
        if (colorObj.rgb && colorObj.rgb.a !== 1) {
          result = ("" + colorObj.hex + decimalToHex(colorObj.rgb.a || 0)).toLowerCase();
        }
      }

      return result;
    };

    _this.handleCustomPickerChange = function (colorObj) {
      var _this$props = _this.props,
          onCustomColorPick = _this$props.onCustomColorPick,
          onPick = _this$props.onPick,
          onCustomColorAfterPick = _this$props.onCustomColorAfterPick;

      var color = _this.getColorStringColorObj(colorObj);

      _this.setState({
        customColor: color
      }); // 选中颜色变化时立即应用


      if (onCustomColorPick) {
        onCustomColorPick(color);
      } else if (!onCustomColorAfterPick) {
        onPick(color);
      }
    };

    _this.handleCustomPickerAfterChange = function (colorObj) {
      var onCustomColorAfterPick = _this.props.onCustomColorAfterPick;

      var color = _this.getColorStringColorObj(colorObj);

      if (onCustomColorAfterPick) {
        onCustomColorAfterPick(color);
      }
    };

    _this.handleConfirmCustomColor = function () {
      var customColor = _this.state.customColor;

      _this.addRecentColor(customColor);

      _this.setState({
        customColor: '',
        showPicker: false
      });
    };

    _this.addRecentColor = function (color) {
      if (!color) return;

      _this.setState(function (state) {
        var colors = [].concat(state.recentColors);
        var findIdx = colors.indexOf(color);

        if (findIdx >= 0) {
          var _colors;

          // 调整已有元素的位置
          var deleteColor = colors.splice(findIdx, 1);

          (_colors = colors).unshift.apply(_colors, deleteColor);
        } else {
          colors.unshift(color);
        }

        colors = colors.slice(0, MAX_LATEST_COLOR);
        colorUtil.saveRecentColorsToLocalStorage(colors);
        return {
          recentColors: colors
        };
      });
    };

    _this.handleSelectChange = function (mode) {
      _this.setState({
        mode: mode
      });

      var onSelectValueChange = _this.props.onSelectValueChange;

      if (typeof onSelectValueChange === 'function') {
        onSelectValueChange(mode);
      }
    };

    _this.getColorData = function () {
      var mode = _this.state.mode;
      var showGradient = _this.props.showGradient;

      if (mode === 'art') {
        return showGradient ? colorDataArtWithGradient : colorDataArt;
      }

      return colorDataNormal;
    };

    _this.renderCustomerColor = function () {
      var _this$props2 = _this.props,
          clearColor = _this$props2.clearColor,
          color = _this$props2.color,
          locale = _this$props2.locale,
          showRecentColors = _this$props2.showRecentColors,
          colorBlockText = _this$props2.colorBlockText,
          customColorBlockTextColor = _this$props2.customColorBlockTextColor;
      var _this$state = _this.state,
          isPickerAlignRight = _this$state.isPickerAlignRight,
          customColor = _this$state.customColor,
          showPicker = _this$state.showPicker,
          customerColors = _this$state.recentColors,
          mode = _this$state.mode;
      var colorCount = Math.min(customerColors.length, MAX_LATEST_COLOR);
      var limitedColors = customerColors.slice(0, colorCount);
      var LineGroup = mode === 'normal' ? LineGroupNormal : LineGroupArt;
      return /*#__PURE__*/_createElement(React.Fragment, null, !colorCount || !showRecentColors ? null : /*#__PURE__*/_createElement(LatestColorWrap, null, /*#__PURE__*/_createElement(LatestColorText, null, /*#__PURE__*/_createElement(TranslateComp, {
        locale: locale,
        textKey: 'we_toolbar_color_picker_latest'
      })), /*#__PURE__*/_createElement(LineGroup, {
        key: "latestColors-" + limitedColors[0]
      }, limitedColors.map(function (curColor, index) {
        return /*#__PURE__*/_createElement(ColorBlockItem, {
          key: curColor + "-" + index,
          colorValue: curColor,
          clearColor: clearColor,
          index: index,
          active: color === curColor,
          onColorPick: _this.handlePick,
          colorBlockText: colorBlockText,
          setColorBlockTextColor: customColorBlockTextColor,
          mode: mode
        });
      }))), /*#__PURE__*/_createElement(CustomGroup, {
        onMouseEnter: _this.showHandleCustomPicker,
        onMouseLeave: _this.hideHandleCustomPicker,
        ref: _this.moreButtonRef
      }, /*#__PURE__*/_createElement(CustomGroupInnerWrap, {
        className: classnames({
          actived: showPicker
        })
      }, /*#__PURE__*/_createElement("div", {
        style: {
          display: 'flex',
          alignItems: 'center'
        }
      }, _SomecolorNormal || (_SomecolorNormal = /*#__PURE__*/_createElement(SomecolorNormal, null)), /*#__PURE__*/_createElement(CustomText, null, /*#__PURE__*/_createElement(TranslateComp, {
        locale: locale,
        textKey: 'we_toolbar_color_picker_more'
      })), showPicker &&
      /*#__PURE__*/
      // 点击 更多颜色 面板之外的区域关闭面板，并将颜色添加到 最近颜色 中
      _createElement(ClickOuterRef, {
        handler: _this.handleConfirmCustomColor,
        on: showPicker,
        box: _this.moreButtonRef.current
      }, function (customPickerRef) {
        return /*#__PURE__*/_createElement(CustomPicker, {
          rightSide: isPickerAlignRight,
          onClick: _this.handleCustomClick,
          ref: customPickerRef
        }, /*#__PURE__*/_createElement(SketchPicker, {
          color: customColor,
          onChange: _this.handleCustomPickerChange,
          onChangeComplete: _this.handleCustomPickerAfterChange
        }));
      })), /*#__PURE__*/_createElement(ArrowRightNormal, {
        style: {
          color: 'rgba(17, 31, 44, 0.4)'
        }
      }))));
    };

    var getColorPanelMode = _this.props.getColorPanelMode;
    _this.state = {
      customColor: '',
      showPicker: false,
      isPickerAlignRight: true,
      recentColors: colorUtil.getRecentColorsFromLocalStorage(),
      mode: typeof getColorPanelMode === 'function' ? getColorPanelMode() || EColorPanelMode.normal : EColorPanelMode.normal
    };
    return _this;
  } // 选中颜色之后，数值回调


  var _proto = ColorPalette.prototype;

  _proto.componentWillUnmount = function componentWillUnmount() {
    if (this.timer) window.clearTimeout(this.timer);
  };

  _proto.render = function render() {
    var _this2 = this;

    var _this$props3 = this.props,
        color = _this$props3.color,
        clearColor = _this$props3.clearColor,
        automaticColor = _this$props3.automaticColor,
        picker = _this$props3.picker,
        locale = _this$props3.locale,
        quickSetText = _this$props3.quickSetText,
        showAutomatic = _this$props3.showAutomatic,
        colorBlockText = _this$props3.colorBlockText,
        customColorPalette = _this$props3.customColorPalette,
        customColorBlockTextColor = _this$props3.customColorBlockTextColor,
        showToggleSelect = _this$props3.showToggleSelect,
        _this$props3$doubleGa = _this$props3.doubleGapRowIndex,
        doubleGapRowIndex = _this$props3$doubleGa === void 0 ? 0 : _this$props3$doubleGa;
    var mode = this.state.mode;
    var doubleGapRowIndexWithMode = mode === 'normal' ? 0 : doubleGapRowIndex;
    var LineGroup = mode === 'normal' ? LineGroupNormal : LineGroupArt;
    var colors = this.getColorData();
    return /*#__PURE__*/_createElement(PaletteWrapper, {
      ref: this.refPalette
    }, /*#__PURE__*/_createElement(Header, null, showAutomatic && /*#__PURE__*/_createElement(DefaultGroup, {
      onClick: function onClick() {
        return _this2.handlePick(automaticColor);
      }
    }, /*#__PURE__*/_createElement(ColorBlockItem, {
      colorValue: automaticColor,
      clearColor: clearColor,
      index: 1,
      active: color === automaticColor,
      onColorPick: this.handlePick,
      colorBlockText: colorBlockText,
      setColorBlockTextColor: customColorBlockTextColor
    }), /*#__PURE__*/_createElement(CustomText, null, quickSetText || /*#__PURE__*/_createElement(TranslateComp, {
      locale: locale,
      textKey: 'we_toolbar_color_picker_default'
    }))), showToggleSelect && /*#__PURE__*/_createElement(Select, {
      defaultValue: mode,
      style: {
        height: 20,
        fontSize: 12,
        lineHeight: 20
      },
      onChange: this.handleSelectChange,
      getContainer: function getContainer() {
        return _this2.refPalette.current || document.body;
      },
      bordered: true,
      testid: "color_panel_mode_select"
    }, /*#__PURE__*/_createElement(SelectOption, {
      value: EColorPanelMode.normal,
      testid: "color_panel_normal_mode"
    }, /*#__PURE__*/_createElement(TranslateComp, {
      locale: locale,
      textKey: 'we_toolbar_color_normal_mode'
    })), /*#__PURE__*/_createElement(SelectOption, {
      value: EColorPanelMode.art,
      testid: "color_panel_art_mode"
    }, /*#__PURE__*/_createElement(TranslateComp, {
      locale: locale,
      textKey: 'we_toolbar_color_art_mode'
    })))), /*#__PURE__*/_createElement(MainColorGroup, null, colors.map(function (data, index) {
      var groupProps = {
        key: index + "-" + data[0],
        style: index === doubleGapRowIndexWithMode ? doubleLineStyle : undefined
      }; // 渲染一行色块

      return /*#__PURE__*/_createElement(LineGroup, groupProps, data.map(function (curColor, colIndex) {
        return /*#__PURE__*/_createElement(ColorBlockItem, {
          key: curColor + "-" + colIndex,
          colorValue: curColor,
          clearColor: clearColor,
          index: colIndex,
          active: color === curColor,
          onColorPick: _this2.handlePick,
          colorBlockText: colorBlockText,
          setColorBlockTextColor: customColorBlockTextColor,
          mode: mode
        });
      }));
    })), picker ? this.renderCustomerColor() : null, customColorPalette);
  };

  return ColorPalette;
}(PureComponent);
ColorPalette.defaultProps = {
  color: 'black',
  onPick: function onPick() {},
  canClear: false,
  automaticColor: '',
  latestColors: [],
  picker: true,
  showRecentColors: true,
  showAutomatic: true
};
