// @ali/we-toolbar/dist-web/biz/splitButton/colorPicker/styled.js
import { withTheme } from '@ali/we-design';
import SplitButton from "../../../common/splitButton";
import styled, { css } from 'styled-components';
import { colorUtil } from "./colorUtil";
export var Divider = /*#__PURE__*/css(["content:'';display:block;position:absolute;background-color:rgba(17,31,44,0.08);height:1px;width:100%;"]);
var groupCommon = /*#__PURE__*/css(["margin:0 ", ";margin-bottom:10px;padding-bottom:10px;position:relative;&:after{", " bottom:0;}"], function (_ref) {
  var theme = _ref.theme;
  return theme.ui.marginMedium;
}, Divider);
export var ColorBoxItem = /*#__PURE__*/styled.span(["position:relative;box-sizing:border-box;width:20px;height:20px;display:inline-block;border-radius:3px 3px;border:1px solid transparent;flex:0 0 auto;cursor:pointer;background-color:#fff;&:hover{padding:2px 2px;border:1px solid #d9d9d9;box-shadow:0 1px 2px rgba(0,0,0,0.12);}"]);
export var ColorBlock = withTheme(styled.span(["position:relative;box-sizing:border-box;width:100%;height:100%;display:block;border:1px solid transparent;svg{position:absolute;top:-1px;left:1px;width:12px;height:12px;}&.isHover{svg{top:-3px;left:-1px;}}.colorBlockText{position:absolute;left:0;right:0;bottom:0;top:0;display:flex;justify-content:center;align-items:center;font-size:12px;}"]));
export var ClearItem = /*#__PURE__*/styled(ColorBoxItem)(["&:after{content:'';display:block;position:absolute;top:8px;left:-2px;width:22px;height:0;border-bottom:2px solid #ff5151;transform:rotate(45deg);}&.isHover{&:after{top:8px;left:1px;width:16px;}}"]);
export var MainColorGroup = withTheme(styled.div(["", ";"], groupCommon));
export var LineGroupNormal = withTheme(styled.div(["position:relative;display:flex;width:100%;height:auto;"]));
export var LineGroupArt = withTheme(styled.div(["display:grid;grid-template-columns:repeat(10,20px);grid-template-rows:20px;gap:2px;margin-bottom:2px;"]));
export var PaletteWrapper = /*#__PURE__*/styled.div(["box-sizing:border-box;margin:12px 12px;border-radius:4px;user-select:none;"]);
export var CustomGroup = withTheme(styled.div(["height:auto;cursor:pointer;.sketch-picker{*,*::before,*::after{box-sizing:content-box;}input{border-radius:6px;&:focus{outline:none;}}}"]));
export var CustomGroupInnerWrap = withTheme(styled.div(["display:flex;flex-direction:row;justify-content:space-between;align-items:center;position:relative;border-radius:4px;margin-top:", ";padding:0 ", ";height:36px;&:hover,&.actived{background-color:", ";}"], function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.marginTiny;
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.paddingMedium;
}, function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.commonBgHoverColor;
}));
export var CustomPicker = /*#__PURE__*/styled.div(["position:absolute;top:-233px;right:", ";"], function (p) {
  return p.rightSide ? '-236px' : '256px';
});
export var CustomText = /*#__PURE__*/styled.div(["height:20px;line-height:20px;font-size:12px;padding-left:4px;flex:0 0 auto;display:inline-block;"]);
export var CustomGroupBtn = /*#__PURE__*/styled.span(["position:absolute;right:2px;top:8px;width:auto;line-height:24px;border-radius:14px;color:#fff;background-image:linear-gradient(179deg,#237ff5 1%,#009aff 100%);padding:0 12px;font-size:12px;"]);
export var DefaultGroup = withTheme(styled.div(["display:flex;position:relative;height:auto;cursor:pointer;"]));
export var SplitButtonWrapper = /*#__PURE__*/styled.div(["line-height:0;position:relative;"]);
export var FillColorArea = /*#__PURE__*/styled.div(["height:2px;background-color:", ";"], function (p) {
  return p.color;
});
export var LatestColorWrap = withTheme(styled.div(["", ";display:flex;flex-direction:column;"], groupCommon));
export var LatestColorText = withTheme(styled.div(["padding-bottom:8px;font-size:12px;"]));
export var ColorPaletteWrapper = /*#__PURE__*/styled.div(["padding:4px 0;"]);
export var FullFillColorArea = withTheme(styled.div(["width:", "px;height:", "px;margin:", ";background-color:", ";border-radius:2px;border:", ";"], function (_ref5) {
  var ui = _ref5.theme.ui;
  return ui.fillColorAreaWidth;
}, function (_ref6) {
  var ui = _ref6.theme.ui;
  return ui.fillColorAreaHeight;
}, function (_ref7) {
  var ui = _ref7.theme.ui;
  return ui.fillColorAreaMargin;
}, function (p) {
  return p.color;
}, function (_ref8) {
  var isWhite = _ref8.isWhite,
      ui = _ref8.theme.ui;
  return isWhite ? "1px solid " + ui.colorLine1 : 'none';
}));
export var FillColorPickerButton = withTheme(styled(SplitButton)(["border:1px solid ", ";border-radius:", ";.splitbutton-right .arrow-icon{font-size:12px;}"], function (_ref9) {
  var ui = _ref9.theme.ui;
  return ui.commonLineLightColor;
}, function (_ref10) {
  var ui = _ref10.theme.ui;
  return ui.radius;
}));
export var Header = withTheme(styled.div(["", ";display:flex;justify-content:space-between;"], groupCommon));
export var ColorBoxWrapper = /*#__PURE__*/styled.div(["display:flex;justify-content:center;align-items:center;position:relative;box-sizing:border-box;border-radius:4px;cursor:pointer;", ";", ";&:hover{border:2px solid #0089ff;}"], function (_ref11) {
  var color = _ref11.color,
      isClear = _ref11.isClear;

  if (color === '#FFFFFF' || isClear) {
    return css(["border:2px solid rgba(126,134,142,0.16);"]);
  }
}, function (_ref12) {
  var color = _ref12.color;

  if (colorUtil.isGradientColor(color)) {
    var _colorUtil$transferGr = colorUtil.transferGradientColor(color),
        startColor = _colorUtil$transferGr[0],
        endColor = _colorUtil$transferGr[1];

    return css(["background-image:linear-gradient(90deg,", ",", ");"], startColor, endColor);
  } else {
    return css(["background:", ";"], color);
  }
});
