// @ali/4ever-component/dist-node/Placeholder/inlinePlaceholder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InlinePlaceholder = exports.InlinePlaceholderWrapper = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _everUtils = require("@ali/4ever-utils");

const _createElement = /*#__PURE__*/_react.default.createElement;

const InlinePlaceholderWrapper = /*#__PURE__*/_styledComponents.default.span(["", " color:#A0A5AB;background-color:#F2F4F5;border-style:solid;border-color:", ";white-space:pre;"], (0, _everUtils.transformThemeObjToCssString)(_everUtils.theme.inlinePlaceholder), props => props.isSelected ? '#D9DBDD' : '#ECEDEF');

exports.InlinePlaceholderWrapper = InlinePlaceholderWrapper;

class InlinePlaceholder extends _react.default.PureComponent {
  constructor(...args) {
    super(...args);

    this.renderText = () => {
      const {
        renderText,
        node
      } = this.props;
      const {
        name
      } = node.data;
      let placeholderText = '';

      if (typeof renderText === 'function') {
        placeholderText = renderText(name);
      }

      return placeholderText || `${name || ''}暂无法解析`;
    };
  }

  render() {
    const {
      attributes,
      isSelected
    } = this.props;
    const text = this.renderText();
    return /*#__PURE__*/_createElement(InlinePlaceholderWrapper, (0, _extends2.default)({}, attributes, {
      isSelected: isSelected,
      title: text
    }), text);
  }

}

exports.InlinePlaceholder = InlinePlaceholder;
