// @ali/4ever-component/dist-node/Placeholder/PlaceholderInlineToolbar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _inlineToolbar = require("../Toolbar/inlineToolbar");

const _createElement = /*#__PURE__*/React.createElement;

var _ref2 = /*#__PURE__*/_createElement(_weDesign.DeleteBetaNormal, null);

const PlaceholderInlineToolbar = props => {
  const onDeletePlaceholder = e => {
    e.preventDefault();
    e.stopPropagation();
    props.handleRemovePlaceholder();
  };

  var _ref = /*#__PURE__*/_createElement(_weDesign.Button, {
    testid: "ph-delete",
    onClick: onDeletePlaceholder,
    bordered: false,
    icon: _ref2
  });

  const DelPlaceholderButton = () => {
    return _ref;
  };

  const customToolButtons = {
    'placeholder.del': DelPlaceholderButton
  };
  const layout = [{
    toolButtons: [{
      name: 'placeholder.del'
    }]
  }];
  return /*#__PURE__*/_createElement(_inlineToolbar.InlineToolbar, {
    visible: true,
    triggerNode: props.triggerNode || undefined,
    customToolButtons: customToolButtons,
    inlineLayout: layout
  });
};

var _default = /*#__PURE__*/React.memo(PlaceholderInlineToolbar);

exports.default = _default;
