// @ali/4ever-plugin-find-and-replace/dist-web/common/actions.js
export var FIND_AND_REPLACE_TOGGLE = Symbol('FIND_AND_REPLACE_TOGGLE');
export var REPLACE_TEXT_AT_DECORATION = Symbol('REPLACE_TEXT_AT_DECORATION');
export var UPDATE_FIND_TEXT = Symbol('UPDATE_FIND_TEXT');
export var UPDATE_FIND_FRAGMENT = Symbol('UPDATE_FIND_FRAGMENT');
export var UPDATE_SEARCH_RESULT = Symbol('UPDATE_SEARCH_RESULT');
export function toggleFindAndReplace(visible) {
  return {
    type: FIND_AND_REPLACE_TOGGLE,
    payload: {
      visible: visible
    }
  };
}
export function replaceText(focus, text) {
  return {
    type: REPLACE_TEXT_AT_DECORATION,
    payload: {
      focus: focus,
      text: text
    }
  };
}
export function updateFindText(findText) {
  return {
    type: UPDATE_FIND_TEXT,
    payload: {
      findText: findText
    }
  };
}
export function updateFindFragment(findFragment) {
  return {
    type: UPDATE_FIND_FRAGMENT,
    payload: {
      findFragment: findFragment
    }
  };
}
export function updateSearchResult(focus) {
  return {
    type: UPDATE_SEARCH_RESULT,
    payload: {
      focus: focus
    }
  };
}
