// @ali/4ever-bi/dist-node/plugins/hr/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _renderNode = _interopRequireDefault(require("./renderNode"));

var _actions = require("./actions");

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _insertHr = _interopRequireDefault(require("./commands/insertHr"));

var _quickInsertConfig = _interopRequireDefault(require("./quickInsertConfig"));

var _queries = require("./queries");

var utils = _interopRequireWildcard(require("./utils"));

var _createRenderHoverToolbar = _interopRequireDefault(require("./createRenderHoverToolbar"));

var _everPluginHr = require("@ali/4ever-plugin-hr");

const commands = {
  insertHr: _insertHr.default
};

function createHrPlugin(config) {
  const hrConfig = config.hr || {};
  return { ...(0, _everPluginHr.moFactory)({}),
    toolbar: (0, _toolbar.default)(hrConfig),
    menubar: (0, _menubar.default)(),
    renderNode: _renderNode.default,
    actions: _actions.actions,
    onAction: _onAction.default,
    queries: _queries.queries,
    onKeyDown: _onKeyDown.default,
    commands,
    quickInsert: _quickInsertConfig.default,
    models: {
      klass: 'block',
      type: 'hr'
    },
    renderHoverToolbar: (0, _createRenderHoverToolbar.default)(hrConfig)
  };
}

createHrPlugin.actions = _actions.DEPRECATED_ACTIONS;
createHrPlugin.commands = commands;
createHrPlugin.utils = utils;
var _default = createHrPlugin;
exports.default = _default;
