// @ali/4ever-plugin-list/dist-node/bi/components/next/orderedListButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = OrderedListButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _everCangjie = require("@ali/4ever-cangjie");

var _everComponent = require("@ali/4ever-component");

var _isOrderedList = _interopRequireDefault(require("../../queries/isOrderedList"));

var _utils = require("../../utils");

var _actions = require("../../actions");

var _role = require("../../../utils/role");

var _listStyleMapping = require("../../../utils/listStyleMapping");

const _createElement = /*#__PURE__*/_react.default.createElement;
const isMobile = _everCangjie.environment.IS_MOBILE;

function OrderedListButton(props) {
  const {
    controller,
    locale,
    ...rest
  } = props;
  const listStyleType = (0, _utils.getListStyleType)(controller);
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInHr') || // fix: https://aone.alibaba-inc.com/v2/project/1090534/bug/35667506
  controller.query('isSelectionInSubjectHeading');
  const isActive = (0, _isOrderedList.default)(controller);
  const handleClick = (0, _react.useCallback)(listType => {
    controller.run('onAction', (0, _actions.toggleList)('olist', isMobile || _listStyleMapping.DEFAULT_NUMBER_LIST_ARRAY[listType] === undefined ? _listStyleMapping.DEFAULT_NUMBER_LIST_ARRAY[0] : listType));
  }, [controller]);

  const tooltip = _react.default.useMemo(() => ({
    title: locale.orderedList,
    shortCut: _utils.OLLIST_HOT_KEY_TIP,
    description: 'Markdown: 1. ' + locale?.space || 'Space'
  }), []);

  return /*#__PURE__*/_createElement(_everComponent.ListSet, (0, _extends2.default)({
    mode: "ol",
    active: isActive,
    testid: "bi-toolbar-ollist",
    disabled: !!isDisabled,
    defaultListStyle: listStyleType,
    onChange: handleClick,
    onClick: handleClick,
    role: _role.ListRole.listOrdered,
    title: locale.orderedList,
    tooltip: tooltip
  }, rest));
}
