// @ali/4ever-plugin-list/dist-node/bi/components/next/taskListButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TaskListButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everComponent = require("@ali/4ever-component");

var _isTaskList = _interopRequireDefault(require("../../queries/isTaskList"));

var _actions = require("../../actions");

var _utils = require("../../utils");

var _role = require("../../../utils/role");

const _createElement = /*#__PURE__*/React.createElement;

const isDisabled = controller => {
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInHr') || // fix: https://aone.alibaba-inc.com/v2/project/1090534/bug/35667506
  controller.query('isSelectionInSubjectHeading');
  return Boolean(isDisabled);
};

function TaskListButton(props) {
  const {
    controller,
    locale,
    disabled,
    tooltip: tooltipProp,
    ...rest
  } = props;
  const isActive = (0, _isTaskList.default)(controller);
  const handleClick = React.useCallback(() => controller.run('onAction', (0, _actions.toggleList)('tlist')), [controller]);
  const tooltip = React.useMemo(() => ({
    title: locale?.taskList,
    shortCut: _utils.TASKLIST_HOT_KEY_TIP,
    description: ' Markdown: [] ' + (locale?.space || 'Space'),
    ...tooltipProp
  }), [tooltipProp]);
  return /*#__PURE__*/_createElement(_everComponent.TaskListButton, (0, _extends2.default)({
    testid: "bi-toolbar-tasklist" // @ts-ignore 组件不存在 bordered 属性
    ,
    bordered: false,
    active: isActive,
    disabled: disabled || isDisabled(controller),
    onClick: handleClick,
    role: _role.ListRole.listTask,
    title: locale?.taskList,
    tooltip: tooltip
  }, rest));
}
