// @ali/4ever-plugin-embed/dist-node/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.default;
  }
});
Object.defineProperty(exports, "Embed", {
  enumerable: true,
  get: function () {
    return _models.default;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.createServerSerializationPlugin;
  }
});
Object.defineProperty(exports, "FileIconMap", {
  enumerable: true,
  get: function () {
    return _fileTypeHelper.FileIconMap;
  }
});
Object.defineProperty(exports, "FileTypeMap", {
  enumerable: true,
  get: function () {
    return _fileTypeHelper.FileTypeMap;
  }
});
Object.defineProperty(exports, "getFileTypeForRead", {
  enumerable: true,
  get: function () {
    return _fileTypeHelper.getFileTypeForRead;
  }
});
Object.defineProperty(exports, "CanPreviewFileTypes", {
  enumerable: true,
  get: function () {
    return _fileTypeHelper.CanPreviewFileTypes;
  }
});
Object.defineProperty(exports, "EmbedFailedCard", {
  enumerable: true,
  get: function () {
    return _embedFailedCard.default;
  }
});
Object.defineProperty(exports, "EmbedUploadProgressCard", {
  enumerable: true,
  get: function () {
    return _embedUploadProgressCard.default;
  }
});
Object.defineProperty(exports, "UploadErrorType", {
  enumerable: true,
  get: function () {
    return _types.UploadErrorType;
  }
});
exports.OldBambooPlugin = void 0;

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;

var _index = _interopRequireDefault(require("./mo/index"));

var _models = _interopRequireDefault(require("./mo/models"));

var _index2 = _interopRequireWildcard(require("./mo/serializer/index"));

var _fileTypeHelper = require("./utils/fileTypeHelper");

var _embedFailedCard = _interopRequireDefault(require("./utils/components/common/embedCard/embedFailedCard"));

var _embedUploadProgressCard = _interopRequireDefault(require("./utils/components/common/embedCard/embedUploadProgressCard"));

var _types = require("./utils/types");
