// @ali/4ever-plugin-node-select/dist-node/bi/createGetNodeSelectToolbarItems.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createGetNodeSelectToolbarItems = void 0;

var _react = _interopRequireDefault(require("react"));

var _everComponent = require("@ali/4ever-component");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement(_everComponent.MobileIconCopy, {
  color: "rgb(92, 113, 221)"
});

var _ref2 = /*#__PURE__*/_createElement(_everComponent.MobileIconCut, {
  color: "rgb(75, 173, 112)"
});

var _ref3 = /*#__PURE__*/_createElement(_everComponent.MobileIconDelete, {
  color: "rgb(228, 117, 117)"
});

const createGetNodeSelectToolbarItems = locale => {
  return (props, _, next) => {
    const copyItem = {
      key: 'copy',
      icon: _ref,
      action: 'copySelectedNodes',
      priority: 3,
      title: locale.copy || 'Copy'
    };
    const cutItem = {
      key: 'cut',
      icon: _ref2,
      action: 'cutSelectedNodes',
      priority: 2,
      title: locale.cut || 'Cut'
    };
    const deleteItem = {
      key: 'delete',
      icon: _ref3,
      action: 'deleteSelectedNodes',
      priority: 4,
      title: locale.delete || 'Delete'
    };
    const res = [copyItem, cutItem, deleteItem];
    const nextItems = next();
    return [...res, ...(Array.isArray(nextItems) ? nextItems : [])];
  };
};

exports.createGetNodeSelectToolbarItems = createGetNodeSelectToolbarItems;
