// @ali/4ever-plugin-draggable/dist-node/bi/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.biFactory = void 0;

var _renderContent = _interopRequireDefault(require("./renderContent"));

var _actions = require("./actions");

var _createRenderNode = _interopRequireDefault(require("./createRenderNode"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _dragHander = _interopRequireDefault(require("./components/dragHander"));

var _leftToolbarHandlers = require("./leftToolbarHandlers");

var _queries = require("./queries");

var _meta = require("../utils/meta");

var _commands = require("./commands");

const biFactory = configs => {
  return {
    key: _meta.key,
    renderContent: (0, _renderContent.default)(configs, _leftToolbarHandlers.handleDragChange),
    renderNode: (0, _createRenderNode.default)(configs),
    queries: { ..._queries.queries,

      /**
       * TODO: 获取 configs 不应该是个 query，待确定历史原因
       */
      getDraggableConfig: () => {
        return configs.draggable;
      }
    },
    onAction: _onAction.default,
    actions: _actions.actions,
    isActions: _actions.isActions,
    commands: _commands.commands,
    models: [],
    toolbar: (0, _toolbar.default)(configs),
    leftToolbar: (0, _dragHander.default)(configs),
    leftToolbarHandlers: _leftToolbarHandlers.leftToolbarHandlers
  };
};

exports.biFactory = biFactory;
