// @ali/4ever-plugin-table/dist-node/bi/components/tableSelection/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _fastDeepEqual = _interopRequireDefault(require("fast-deep-equal"));

var _everCangjie = require("@ali/4ever-cangjie");

var _constants = require("../../constants");

var _selectionRect = _interopRequireDefault(require("./selectionRect"));

var _usePixelColsWidth = require("../../../utils/hooks/usePixelColsWidth");

const _createElement = /*#__PURE__*/_react.default.createElement;

const TableSelection = props => {
  const {
    node,
    controller,
    selection: propSelection,
    isHighlight,
    tableRef,
    scale = 1,
    getTableSelection,
    forceUpdate
  } = props;
  const [colsWidth] = (0, _usePixelColsWidth.usePixelColsWidth)();
  const [selection, setSelection] = (0, _react.useState)(propSelection);
  (0, _react.useEffect)(() => {
    setSelection(propSelection);
  }, [propSelection]);
  (0, _everCangjie.useSelectionData)(controller.selectionData$, null, () => {
    const tempSelection = getTableSelection();

    if (!(0, _fastDeepEqual.default)(selection, tempSelection)) {
      setSelection(tempSelection);
    }
  }, []);

  const onHots = _react.default.useCallback(hots => {
    // PERF: 非当前表格热区不处理
    if (hots.some(hot => node.key === hot.node.key || node.hasNode(hot.node.key))) {
      const tableSelection = getTableSelection();
      setSelection(tableSelection);
    }
  }, [node, getTableSelection]);

  (0, _everCangjie.useSelectingHots)(controller, onHots);
  return /*#__PURE__*/_createElement(_selectionRect.default, {
    testid: "table-selection-wrapper",
    table: node,
    tableRef: tableRef,
    tableSelection: selection,
    scale: scale,
    colsWidth: colsWidth,
    style: {
      backgroundColor: isHighlight ? _constants.THEME.red0 : _constants.THEME.blue0,
      borderColor: isHighlight ? _constants.THEME.red1 : _constants.THEME.blue3
    },
    forceUpdate: forceUpdate
  });
};

var _default = TableSelection;
exports.default = _default;
