// @ali/4ever-plugin-table/dist-node/bi/components/tableSelection/tableHoverRect.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _selectionRect = _interopRequireDefault(require("./selectionRect"));

var _usePixelColsWidth = require("../../../utils/hooks/usePixelColsWidth");

const _createElement = /*#__PURE__*/React.createElement;

const TableHoverRect = props => {
  const {
    node,
    tableRef,
    hoverSelection,
    scale
  } = props;
  const [colsWidth] = (0, _usePixelColsWidth.usePixelColsWidth)();
  return /*#__PURE__*/_createElement(_selectionRect.default, {
    testid: "table-hover-rect-wrapper",
    table: node,
    tableRef: tableRef,
    tableSelection: hoverSelection,
    scale: scale,
    colsWidth: colsWidth
  });
};

var _default = TableHoverRect;
exports.default = _default;
