// @ali/4ever-bi/dist-node/plugins/columns/components/columnsMenuLegacy.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getMenu;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _weDesign = require("@ali/we-design");

const _createElement = /*#__PURE__*/React.createElement;
const {
  insertColumns
} = _everBamboo.ColumnsPlugin.actions;

function getMenu(size, configs) {
  return function ColumnsMenu(props) {
    const {
      menuItem,
      controller
    } = props;
    const {
      title,
      ...rest
    } = menuItem;
    const {
      isCollapsedAtEdge
    } = controller.value.selection;

    const handleClick = () => {
      controller.run('onAction', insertColumns({
        size,
        text: ''
      }));
    };

    return /*#__PURE__*/_createElement(_weDesign.MenuItem, (0, _extends2.default)({}, rest, {
      disabled: isCollapsedAtEdge,
      onClick: handleClick,
      role: `${_everBamboo.PluginRoles.columns}${size}`
    }), title || '');
  };
}
