// @ali/4ever-bi/dist-node/plugins/columns/components/columnsMenu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _weDesign = require("@ali/we-design");

var _everUtils = require("@ali/4ever-utils");

var _ColumnPicker = _interopRequireDefault(require("./ColumnPicker"));

const _createElement = /*#__PURE__*/_react.default.createElement;
const {
  insertColumns
} = _everBamboo.ColumnsPlugin.actions;

const createColumnsMenu = configs => {
  return function ColumnsMenu(props) {
    const [count, setCount] = (0, _react.useState)(2);
    const {
      menuItem,
      controller
    } = props;
    const {
      title,
      menuKey,
      ...rest
    } = menuItem;
    const {
      isCollapsedAtEdge
    } = controller.value.selection;
    const handleClick = (0, _react.useCallback)((_, event) => {
      if (!event) {
        // 回车触发
        // menuItem 有bug, 这里必须在 setCount 里才能拿到最新的 count
        setCount(count => {
          controller.run('onAction', insertColumns({
            size: count,
            text: ''
          }));
          return count;
        });
      } else {
        // click 触发
        controller.run('onAction', insertColumns({
          size: count,
          text: ''
        }));
      }

      event && event.stopPropagation();
    }, [count]);
    const handleDefaultClick = (0, _react.useCallback)(event => {
      controller.run('onAction', insertColumns({
        size: 2,
        text: ''
      }));
      controller.run('onAction', (0, _everUtils.removeNewlineTriggerText)(false));
      event && event.stopPropagation();
    }, [controller]);
    const columnText = configs?.locale?.insertText2;
    return /*#__PURE__*/_createElement("div", {
      onClick: handleDefaultClick
    }, /*#__PURE__*/_createElement(_weDesign.SubMenu, (0, _extends2.default)({}, rest, {
      disabled: isCollapsedAtEdge,
      title: title || '',
      overlayStyle: {
        padding: 0
      },
      overlayClassName: "menu-columns-overlay"
    }), /*#__PURE__*/_createElement(_weDesign.MenuItem, {
      menuKey: menuKey || 'columnsPicker',
      hoverable: false,
      role: _everBamboo.PluginRoles.columns,
      style: {
        height: 'auto',
        padding: '12px'
      },
      onClick: handleClick
    }, /*#__PURE__*/_createElement(_ColumnPicker.default, {
      activeCount: count,
      onChange: setCount,
      columnText: columnText
    }))));
  };
};

var _default = createColumnsMenu;
exports.default = _default;
