// @ali/4ever-plugin-clipboard/dist-web/bi/handlers/createOnEditable.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import ClipboardPad from "../../utils/ClipboardPad";
import LazyPastePanel from "../components/lazyPastePanel";

var createOnEditable = function createOnEditable(clipboard) {
  var mounted = true;
  return function (props, controller, next) {
    var customPanel = controller.value.data.customPanel;
    var clipboardPad = ClipboardPad.getCachedClipboardPad(clipboard == null ? void 0 : clipboard.key);

    if (mounted) {
      clipboardPad == null ? void 0 : clipboardPad.initPluginModels(controller);
      mounted = false;
    }

    return /*#__PURE__*/_createElement(React.Fragment, null, next(), /*#__PURE__*/_createElement(LazyPastePanel, {
      config: clipboard,
      customPanel: customPanel,
      controller: controller,
      clipboardPad: clipboardPad,
      locale: clipboard.locale,
      beforeInsertMarkdown: clipboard.beforeInsertMarkdown,
      afterInsertMarkdown: clipboard.afterInsertMarkdown,
      getContainer: clipboard.getContainer
    }));
  };
};

export default createOnEditable;
