// @ali/we-project-card-placeholder/dist-web/index.js
import React from 'react';
import _extends from '@babel/runtime/helpers/extends';
import _taggedTemplateLiteralLoose from '@babel/runtime/helpers/taggedTemplateLiteralLoose';
import styled from 'styled-components';

function _templateObject5() {
  var data = _taggedTemplateLiteralLoose(["\n  color:#111F2C;\n  text-decoration: none;\n  font-weight: 500;\n  font-size: 16px;\n  line-height: 20px;\n  ", "\n\n  &:visited {color:#111F2C;}\n  &:hover {\n    color:#5B7DED;\n    text-decoration: underline;\n  }\n  &:active {color:#111F2C;}\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteralLoose(["\n  margin: 12px;\n  width: 20px;\n  height: 20px;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteralLoose(["\n  flex: 1 1 auto;\n  display: flex;\n  flex-direction: row-reverse;\n  align-items: center;\n  padding-right: 16px;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose(["\n  flex: 1 1 auto;\n  display: flex;\n  flex-direction: row;\n  align-items: center;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n  height: 52px;\n  display: flex;\n  flex-direction: row;\n  align-items: center;\n  justify-content: space-between;\n  background-color: #F1F3F6;\n  border-radius: 12px 12px 0 0;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}
var Root = styled.div(_templateObject());
var Left = styled.div(_templateObject2());
var Right = styled.div(_templateObject3());
var ToolbarImg = styled.img(_templateObject4());
var ToolbarName = styled.a(_templateObject5(), truncate('170px'));
function truncate(width) {
  return "\n    max-width: " + width + ";\n    word-break: keep-all;\n    white-space: nowrap;\n    overflow: hidden;\n    text-overflow: ellipsis;\n  ";
}

var _createElement = /*#__PURE__*/React.createElement;

var _ref = /*#__PURE__*/_createElement(ToolbarImg, {
  src: "https://img.alicdn.com/tfs/TB1sNayNfb2gK0jSZK9XXaEgFXa-52-48.png"
});

function Toolbar(props) {
  return /*#__PURE__*/_createElement(Root, null, /*#__PURE__*/_createElement(Left, null, _ref, /*#__PURE__*/_createElement(ToolbarName, {
    href: props.url
  }, props.name)), /*#__PURE__*/_createElement(Right, null, props.children));
}

function _templateObject2$1() {
  var data = _taggedTemplateLiteralLoose(["\n  flex: 1 1 auto;\n  display: flex;\n  background-color: white;\n  border-radius: 12px;\n  border: 1px solid #DDDDDD;\n"]);

  _templateObject2$1 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$1() {
  var data = _taggedTemplateLiteralLoose(["\n  flex: 0;\n  display: flex;\n  flex-direction: column;\n  border-radius: 12px;\n  min-height: 200px;\n  background-color: #F1F3F6;\n"]);

  _templateObject$1 = function _templateObject() {
    return data;
  };

  return data;
}
var _createElement$1 = /*#__PURE__*/React.createElement;
var Container = styled.div(_templateObject$1());
var Content = styled.div(_templateObject2$1());
function CardContainer(props) {
  return /*#__PURE__*/_createElement$1(Container, {
    style: _extends({}, props.style)
  }, /*#__PURE__*/_createElement$1(Toolbar, {
    name: props.name,
    url: props.url
  }), /*#__PURE__*/_createElement$1(Content, {
    style: _extends({}, props.contentStyle)
  }, props.children));
}

function _templateObject3$1() {
  var data = _taggedTemplateLiteralLoose(["\n  font-size: 12px;\n  ", "\n"]);

  _templateObject3$1 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2$2() {
  var data = _taggedTemplateLiteralLoose(["\n  font-size: 28px;\n  margin-top: 32px;\n  margin-bottom: 28px;\n"]);

  _templateObject2$2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$2() {
  var data = _taggedTemplateLiteralLoose(["\n  flex: 1 1 auto;\n  display: flex;\n  flex-direction: column;\n  align-items: center;\n  justify-content: center;\n"]);

  _templateObject$2 = function _templateObject() {
    return data;
  };

  return data;
}
var _createElement$2 = /*#__PURE__*/React.createElement;
var Container$1 = styled.div(_templateObject$2());
var Title = styled.div(_templateObject2$2());
var Description = styled.span(_templateObject3$1(), truncate('300px'));
var ProjectInfoCard = function ProjectInfoCard(props) {
  if (props === void 0) {
    props = {
      description: '创建者：未知',
      title: '未加入当前项目'
    };
  }

  return /*#__PURE__*/_createElement$2(Container$1, null, /*#__PURE__*/_createElement$2(Title, null, props.title || ''), /*#__PURE__*/_createElement$2(Description, {
    title: props.description || ''
  }, props.description || ''));
};

var _createElement$3 = /*#__PURE__*/React.createElement;
var Printable = function Printable(props) {
  return /*#__PURE__*/_createElement$3(CardContainer, {
    style: props.style,
    name: props.name,
    url: props.url
  }, /*#__PURE__*/_createElement$3(ProjectInfoCard, {
    title: props.title
  }));
};

export { Printable as PrintableCard };
