// @ali/4ever-bamboo/dist-web/index.js
import { Ul, Ol, Li } from '@ali/4ever-plugin-list';
import { H1, H2, H3, H4, H5, H6 } from '@ali/4ever-plugin-heading';
import { P } from '@ali/4ever-plugin-paragraph';
import { TD } from '@ali/4ever-plugin-table';
import { BlockquoteWrapper } from '@ali/4ever-plugin-blockquote'; // reexport

export { BambooPlaceholder as Placeholder, PlaceholderWrapper, InlinePlaceholder, ErrorBoundary, Select, InputNumberItem, CustomInputNumberItem } from '@ali/4ever-component';
export { useScrollableContainer, useScrollableContent } from '@ali/4ever-cangjie';
export { useContainerWidthResize } from '@ali/4ever-plugin-table';
export * from "./plugins";
export { getHeadings, createHeadingId } from '@ali/4ever-plugin-heading';
export { marksToStyle as convertMarks } from '@ali/4ever-plugin-marks';
export { isPureTableNode, isTableNode } from '@ali/4ever-plugin-table';
export { throttleByIdleCallback, copyToClipboard, UnitConvert, isFirstPart, addDefaultUnit, transformThemeObjToCssString, isNaN, ptToPx, unitsConverter, sample, Sample, SimpleMapCache, fontRenderFallback, ShdUtil, UnderlineUtil, calculatedPluginSize, SpacingUtil, IndUtil, SzUtil, CommonFormatUtil, GradientUtil, isEmptyParagraph, zIndex, isSingleTableCell, unwrapSingleTableCell, fontSize, lineHeight, blockMarginStyleInContainer, theme } from '@ali/4ever-utils';
export var List = {
  Ul: Ul,
  Ol: Ol,
  Li: Li
};
export { H1, H2, H3, H4, H5, H6, P, TD, BlockquoteWrapper };
var typeToComponent = {
  'table-cell': TD,
  'heading-1': H1,
  'heading-2': H2,
  'heading-3': H3,
  'heading-4': H4,
  'heading-5': H5,
  'heading-6': H6,
  paragraph: P
};
export function styledComponents(type) {
  return typeToComponent[type];
}
