// @ali/we-design/dist-web/components/preview-image-mobile/utils.js
export var preventDefault = function preventDefault(e) {
  e.preventDefault();
};
export var CustomEventEmitter = /*#__PURE__*/function () {
  function CustomEventEmitter() {
    this.eventHandlersMap = new Map();
  }

  var _proto = CustomEventEmitter.prototype;

  _proto.on = function on(event, handler) {
    if (this.eventHandlersMap.has(event)) {
      var handlerArray = this.eventHandlersMap.get(event);
      handlerArray.push(handler);
    } else {
      this.eventHandlersMap.set(event, [handler]);
    }
  };

  _proto.off = function off(event, handler) {
    if (this.eventHandlersMap.has(event)) {
      var handlerArray = this.eventHandlersMap.get(event);
      var handlerIndex = handlerArray.indexOf(handler);

      if (handlerIndex > -1) {
        handlerArray.splice(handlerIndex, 1);
      }
    }
  };

  _proto.emit = function emit(event) {
    for (var _len = arguments.length, params = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      params[_key - 1] = arguments[_key];
    }

    var handlerArray = this.eventHandlersMap.get(event);
    handlerArray == null ? void 0 : handlerArray.forEach(function (handler) {
      handler.apply(void 0, params);
    });
  };

  return CustomEventEmitter;
}();
