// @ali/4ever-plugin-list/dist-web/bi/createContextMenu.js
import { getCurrentItem } from "./utils";
import { restartNumbering, continueNumbering } from "./actions";
export default function createContextMenu() {
  return function contextMenuPlugin(pluginConfig) {
    var locale = pluginConfig.locale || {};
    return function contextMenu(controller, next) {
      var menus = [];
      var selection = controller.value.selection;

      if (selection != null && selection.isCollapsed) {
        var _currentItem$data;

        var currentItem = getCurrentItem(controller)[0];
        var listConfig = currentItem == null ? void 0 : (_currentItem$data = currentItem.data) == null ? void 0 : _currentItem$data.list;

        if (listConfig && listConfig.isOrdered) {
          // 有序列表才有下列选项
          menus.push({
            name: locale.restartNumbering,
            action: restartNumbering(),
            key: 'list-restart-numbering',
            options: {
              mobile: false,
              group: 2
            }
          }, {
            name: locale.continueNumbering,
            action: continueNumbering(),
            key: 'list-continue-numbering',
            options: {
              mobile: false,
              group: 2
            }
          });
        }
      }

      var others = next() || [];
      return menus.concat(others);
    };
  };
}
