// @ali/4ever-zhi/dist-web/plugins/callout/handlers/createRenderNode.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { ColorBlocks, Container } from '@ali/4ever-mo';
import Callout from "../components/callout";

function createRenderNode(param) {
  var _ref;

  return _ref = {
    'callout-group': function calloutGroup(props, controller) {
      var sticker = param.sticker;
      var attributes = props.attributes,
          children = props.children;
      var node = props.node;
      return /*#__PURE__*/_createElement(Callout, {
        controller: controller,
        node: node,
        attributes: attributes,
        sticker: sticker
      }, children());
    }
  }, _ref[Container.TYPE] = function (props, controller) {
    var sticker = param.sticker;
    var attributes = props.attributes,
        children = props.children;
    var node = props.node;

    if (node.data.subType !== ColorBlocks.SUB_TYPE) {
      return /*#__PURE__*/_createElement(React.Fragment, null);
    }

    return /*#__PURE__*/_createElement(Callout, {
      controller: controller,
      node: node,
      attributes: attributes,
      sticker: sticker
    }, children());
  }, _ref;
}

export default createRenderNode;
