// @ali/we-ui/dist-web/components/button/Button.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { classnames } from "../../utils";
import { ButtonWrap, ButtonChildrenWrap, ButtonInner } from "./styled";
import { Tooltip } from "../tooltip";
var Button = /*#__PURE__*/React.memo( /*#__PURE__*/React.forwardRef(function (props, ref) {
  var style = props.style,
      children = props.children,
      className = props.className,
      onClick = props.onClick,
      shape = props.shape,
      size = props.size,
      type = props.type,
      onDoubleClick = props.onDoubleClick,
      onMouseDown = props.onMouseDown,
      bordered = props.bordered,
      icon = props.icon,
      active = props.active,
      disabled = props.disabled,
      tabIndex = props.tabIndex,
      title = props.title,
      testid = props.testid,
      restProps = _objectWithoutPropertiesLoose(props, ["style", "children", "className", "onClick", "shape", "size", "type", "onDoubleClick", "onMouseDown", "bordered", "icon", "active", "disabled", "tabIndex", "title", "testid"]);

  var buttonNode = /*#__PURE__*/_createElement(ButtonWrap, _extends({}, restProps, {
    className: classnames(className, {
      active: active
    }),
    shape: shape,
    style: style,
    size: size,
    type: type,
    ref: ref,
    onClick: disabled ? undefined : onClick,
    onDoubleClick: disabled ? undefined : onDoubleClick,
    onMouseDown: disabled ? undefined : onMouseDown,
    bordered: bordered,
    disabled: disabled,
    onlyIcon: !!icon && !children,
    tabIndex: tabIndex === '-1' ? undefined : tabIndex || '0',
    "data-testid": testid ? "button-" + testid : undefined
  }), /*#__PURE__*/_createElement(ButtonInner, null, icon, /*#__PURE__*/_createElement(ButtonChildrenWrap, null, children)));

  return title ? /*#__PURE__*/_createElement(Tooltip, {
    title: title
  }, buttonNode) : buttonNode;
}));
Button.defaultProps = {
  shape: 'rect',
  size: 'normal',
  type: 'normal',
  bordered: true,
  icon: null,
  active: false,
  disabled: false
};
export { Button };
