// @ali/we-ui/dist-web/components/button/styled.js
/* eslint-disable no-nested-ternary */

/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/explicit-function-return-type */
// /* eslint-disable @typescript-eslint/explicit-function-return-type */
import styled, { css } from 'styled-components';
import { withTheme } from "../theme";
import { resetComponent, Ellipsis } from "../../utils/styled.mixin";
var commonButton = /*#__PURE__*/css(["", " display:inline-block;width:fit-content;border:1px solid ", ";outline:0;height:", ";", ";min-width:", ";padding:", ";cursor:", ";transition:all .2s ease-out;user-select:none;"], resetComponent, function (_ref) {
  var theme = _ref.theme;
  return theme.ui.commonLineLightColor;
}, function (_ref2) {
  var theme = _ref2.theme,
      size = _ref2.size;
  var sizeString = '';
  if (size === 'small') sizeString = 'Small';
  if (size === 'large') sizeString = 'Large';
  return theme.ui["elementHeight" + sizeString];
}, function (_ref3) {
  var theme = _ref3.theme,
      onlyIcon = _ref3.onlyIcon,
      size = _ref3.size;
  var isRect = onlyIcon;
  var width = theme.ui["elementHeight" + (size === 'small' ? 'Small' : size === 'large' ? 'Large' : '')];
  return isRect ? "\n      width: " + width + ";\n    " : '';
}, function (_ref4) {
  var theme = _ref4.theme,
      size = _ref4.size,
      onlyIcon = _ref4.onlyIcon;
  var sizeString = '';
  if (onlyIcon) return 'auto';
  if (size === 'small') sizeString = 'Small';
  if (size === 'large') sizeString = 'Large';
  return theme.ui["elementHeight" + sizeString];
}, function (_ref5) {
  var theme = _ref5.theme,
      onlyIcon = _ref5.onlyIcon;
  if (onlyIcon) return '0';
  return "0 " + theme.ui.paddingLarge;
}, function (_ref6) {
  var disabled = _ref6.disabled;
  return disabled ? 'not-allowed' : 'pointer';
});
var roundButton = /*#__PURE__*/css(["border-radius:", ";"], function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.elementHeightSmall;
});
var rectButton = /*#__PURE__*/css(["border-radius:", ";"], function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.radius;
});
var circleButton = /*#__PURE__*/css(["border-radius:50%;"]);
var primaryButton = /*#__PURE__*/css(["background-color:", ";border-color:", ";color:", ";", ";"], function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.primaryColor;
}, function (_ref10) {
  var theme = _ref10.theme;
  return theme.ui.primaryColor;
}, function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.colorGray1;
}, function (_ref12) {
  var theme = _ref12.theme,
      disabled = _ref12.disabled;

  if (!disabled) {
    return "\n        &:hover, &:focus {\n          border-color: " + theme.ui.primaryHoverColor + ";\n          background-color: " + theme.ui.primaryHoverColor + ";\n        }\n\n        &:active, &.active {\n          border-color: " + theme.ui.primaryPressColor + ";\n          background-color: " + theme.ui.primaryPressColor + ";\n        }\n      ";
  }

  return "\n      border-color: " + theme.ui.primaryDisabledColor + ";\n      background-color: " + theme.ui.primaryDisabledColor + ";\n    ";
});
var normalButton = /*#__PURE__*/css(["border-color:", ";", ";"], function (_ref13) {
  var theme = _ref13.theme;
  return theme.ui.borderColor;
}, function (_ref14) {
  var theme = _ref14.theme,
      disabled = _ref14.disabled;

  if (!disabled) {
    return "\n        &:hover, &:focus {\n          background-color: " + theme.ui.buttonNormalHoverBgColor + ";\n        }\n\n        &.toolBoxHover {\n          background-color: " + theme.ui.buttonNormalHoverBgColor + ";\n        }\n\n        &:active {\n          background-color: " + theme.ui.buttonNormalClickBgColor + ";\n        }\n\n        &.active {\n          background-color: " + theme.ui.buttonNormalActiveBgColor + ";\n        }\n      ";
  }

  return "\n      border-color: " + theme.ui.buttonNormalDisabledBorderColor + ";\n      background-color: transparent;\n      color: " + theme.ui.disabledColor + ";\n    ";
});
var dangerButton = /*#__PURE__*/css(["background-color:", ";border-color:", ";color:", ";", ";"], function (_ref15) {
  var theme = _ref15.theme;
  return theme.ui.buttonDangerBgColor;
}, function (_ref16) {
  var theme = _ref16.theme;
  return theme.ui.buttonDangerBgColor;
}, function (_ref17) {
  var theme = _ref17.theme;
  return theme.ui.colorGray1;
}, function (_ref18) {
  var theme = _ref18.theme,
      disabled = _ref18.disabled;

  if (!disabled) {
    return "\n        &:hover, &:focus {\n          background-color: " + theme.ui.buttonDangerHoverBgColor + ";\n        }\n\n        &:active, &.active {\n          background-color: " + theme.ui.buttonDangerActiveBgColor + ";\n        }\n      ";
  }

  return "\n      border-color: " + theme.ui.buttonDangerDisabledBgColor + ";\n      background-color: " + theme.ui.buttonDangerDisabledBgColor + ";\n    ";
});
var ghostButton = /*#__PURE__*/css(["background-color:transparent;border-color:", ";color:", ";", ";"], function (_ref19) {
  var theme = _ref19.theme;
  return theme.ui.colorGray1;
}, function (_ref20) {
  var theme = _ref20.theme;
  return theme.ui.colorGray1;
}, function (_ref21) {
  var theme = _ref21.theme,
      disabled = _ref21.disabled;

  if (!disabled) {
    return "\n        &:hover, &:focus {\n          background-color: " + theme.ui.buttonGhostHoverBgColor + ";\n          border-color: " + theme.ui.buttonGhostHoverBorderColor + ";\n        }\n\n        &:active, &.active {\n          background-color: " + theme.ui.buttonGhostActiveBgColor + ";\n          border-color: " + theme.ui.buttonGhostActiveBorderColor + ";\n        }\n      ";
  }

  return "\n      border-color: " + theme.ui.buttonGhostDisabledBorderColor + ";\n      color: " + theme.ui.buttonGhostDisabledBorderColor + ";\n    ";
});
var textButton = /*#__PURE__*/css(["background-color:transparent;border-color:transparent;color:", ";padding:0;", ";"], function (_ref22) {
  var theme = _ref22.theme;
  return theme.ui.primaryColor;
}, function (_ref23) {
  var theme = _ref23.theme,
      disabled = _ref23.disabled;

  if (!disabled) {
    return "\n        &:hover, &:focus {\n          color: " + theme.ui.buttonPrimaryHoverBgColor + ";\n        }\n\n        &:active, &.active {\n          color: " + theme.ui.buttonPrimaryActiveBgColor + ";\n        }\n      ";
  }

  return "\n      color: " + theme.ui.buttonPrimaryDisabledBgColor + ";\n    ";
});
export var ButtonInner = /*#__PURE__*/styled.div(["box-sizing:border-box;height:100%;display:flex;align-items:center;justify-content:center;", ""], Ellipsis);
export var ButtonChildrenWrap = /*#__PURE__*/styled.div(["max-width:100%;", ""], Ellipsis);
export var ButtonWrap = withTheme(styled.div(["", " ", " ", " ", " ", " ", " ", " ", " ", " ", " ", " ", ""], resetComponent, commonButton, function (_ref24) {
  var disabled = _ref24.disabled,
      theme = _ref24.theme;
  return disabled ? '' : ".weicon {color: " + theme.ui.commonActionIconColor + ";}";
}, function (_ref25) {
  var shape = _ref25.shape;
  return shape === 'round' && roundButton;
}, function (_ref26) {
  var shape = _ref26.shape;
  return shape === 'rect' && rectButton;
}, function (_ref27) {
  var shape = _ref27.shape;
  return shape === 'circle' && circleButton;
}, function (_ref28) {
  var type = _ref28.type;
  return type === 'primary' && primaryButton;
}, function (_ref29) {
  var type = _ref29.type;
  return type === 'normal' && normalButton;
}, function (_ref30) {
  var type = _ref30.type;
  return type === 'danger' && dangerButton;
}, function (_ref31) {
  var type = _ref31.type;
  return type === 'ghost' && ghostButton;
}, function (_ref32) {
  var type = _ref32.type;
  return type === 'text' && textButton;
}, function (_ref33) {
  var bordered = _ref33.bordered;
  return !bordered && "\n    border-color: transparent !important;\n  ";
}));
export var SplitButtonWrap = withTheme(styled.div(["", " ", " ", " border:", ";border-radius:", "};padding:0;overflow:hidden;", ";&.leftSelected .splitbutton-left{background-color:", ";}&.leftSelected .splitbutton-right{background-color:", ";}"], resetComponent, commonButton, function (_ref34) {
  var disabled = _ref34.disabled,
      theme = _ref34.theme;
  return disabled ? '' : ".weicon {color: " + theme.ui.commonActionIconColor + ";}";
}, function (_ref35) {
  var bordered = _ref35.bordered,
      theme = _ref35.theme;
  return bordered ? "1px solid " + theme.ui.commonLineLightColor : 'none';
}, function (_ref36) {
  var theme = _ref36.theme;
  return theme.ui.radius;
}, function (_ref37) {
  var theme = _ref37.theme,
      disabled = _ref37.disabled;

  if (!disabled) {
    return "\n        &:hover, &.active {\n          background-color: " + theme.ui.splitButtonBgColor + ";\n        }\n      ";
  }

  return "\n      cursor: not-allowed;\n      background-color: transparent;\n      color: " + theme.ui.disabledColor + ";\n\n      // \u63D0\u5347\u6B64\u89C4\u5219\u7684\u4F18\u5148\u7EA7\n      .arrow-icon.arrow-icon {\n        color: " + theme.ui.disabledColor + ";\n      }\n    ";
}, function (_ref38) {
  var theme = _ref38.theme;
  return theme.ui.splitButtonLeftActiveBgColor;
}, function (_ref39) {
  var theme = _ref39.theme;
  return theme.ui.splitButtonBgColor;
}));
export var SplitButtonInner = /*#__PURE__*/styled.div(["box-sizing:border-box;height:100%;display:flex;align-items:stretch;justify-content:center;", ""], Ellipsis);
export var SplitButtonLeft = withTheme(styled.div(["display:flex;align-items:center;justify-content:center;padding:0 4px 0 4px;margin:2px 0 2px 0;transition:background-color 0.1s ease-in-out;", ""], function (_ref40) {
  var disabled = _ref40.disabled,
      theme = _ref40.theme;

  if (!disabled) {
    return "\n          &:hover {\n            padding-right: 3px;\n            border-right: 1px solid " + theme.ui.commonLineLightColor + ";\n          }\n          &:active {\n            background-color: " + theme.ui.splitButtonLeftActiveBgColor + ";\n          }\n        ";
  }

  return 'opacity:0.5';
}));
export var SplitButtonRight = withTheme(styled.div(["display:flex;align-items:center;justify-content:center;width:10px;padding:0 4px 0 1px;box-sizing:content-box;", " .arrow-icon{transform:", ";font-size:12px;color:", ";}"], function (_ref41) {
  var disabled = _ref41.disabled,
      theme = _ref41.theme,
      active = _ref41.active;

  if (!disabled) {
    if (active) {
      return "background-color: " + theme.ui.splitButtonActiveBgColor + ";";
    }

    return "\n          &:hover {\n            background-color: " + theme.ui.splitButtonHoverBgColor + ";\n          }\n        ";
  }

  return '';
}, function (_ref42) {
  var active = _ref42.active;
  return active ? 'rotate(180deg) scale(.67)' : 'rotate(0) scale(.67)';
}, function (_ref43) {
  var theme = _ref43.theme;
  return theme.ui.commonBaseColor3;
}));
