// @ali/4ever-plugin-tag/dist-node/mo/serializer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createSerializer;
exports.createServerSerializationPlugin = createServerSerializationPlugin;

var _jsonMLToValue = _interopRequireDefault(require("./jsonMLToValue"));

var _valueToJsonML = _interopRequireDefault(require("./valueToJsonML"));

var _valueToText = _interopRequireDefault(require("./valueToText"));

var _jsonMLToHTML = _interopRequireDefault(require("./jsonMLToHTML"));

var _im = _interopRequireDefault(require("./jsonMLToHTML/im"));

var _group = _interopRequireDefault(require("./jsonMLToHTML/group"));

const convertPlugins = {
  im: _im.default,
  group: _group.default
};

function createSerializer(configs) {
  return {
    jsonMLToHTML: (0, _jsonMLToHTML.default)(convertPlugins)(configs),
    jsonMLToValue: _jsonMLToValue.default,
    valueToJsonML: _valueToJsonML.default,
    valueToText: _valueToText.default
  };
}

function createServerSerializationPlugin() {
  return {
    valueToJsonML: _valueToJsonML.default,
    jsonMLToValue: _jsonMLToValue.default
  };
}
