// @ali/4ever-plugin-heading/dist-node/utils/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HeadingRole = exports.HeadingTypes = void 0;
let HeadingTypes;
exports.HeadingTypes = HeadingTypes;

(function (HeadingTypes) {
  HeadingTypes["Heading1"] = "heading-1";
  HeadingTypes["Heading2"] = "heading-2";
  HeadingTypes["Heading3"] = "heading-3";
  HeadingTypes["Heading4"] = "heading-4";
  HeadingTypes["Heading5"] = "heading-5";
  HeadingTypes["Heading6"] = "heading-6";
})(HeadingTypes || (exports.HeadingTypes = HeadingTypes = {}));

let HeadingRole;
exports.HeadingRole = HeadingRole;

(function (HeadingRole) {
  HeadingRole["heading"] = "heading";
  HeadingRole["headingSelect"] = "headingSelect";
  HeadingRole["headingGroup"] = "headingGroup";
  HeadingRole["heading1"] = "heading-1";
  HeadingRole["heading2"] = "heading-2";
  HeadingRole["heading3"] = "heading-3";
  HeadingRole["heading4"] = "heading-4";
  HeadingRole["heading5"] = "heading-5";
  HeadingRole["heading6"] = "heading-6";
})(HeadingRole || (exports.HeadingRole = HeadingRole = {}));
