// @ali/4ever-plugin-tag/dist-node/mo/serializer/jsonMLToHTML/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createJsonMLToHTMLFactory;

var React = _interopRequireWildcard(require("react"));

var _everUtils = require("@ali/4ever-utils");

const _createElement = /*#__PURE__*/React.createElement;

function createJsonMLToHTMLFactory(plugins) {
  return function createJsonMLToHTML(configs) {
    return {
      name: 'tag',
      match: node => {
        const tag = (0, _everUtils.getTagName)(node);
        const attrs = (0, _everUtils.getAttributes)(node);
        return tag === 'tag' && !!attrs && attrs.tagType;
      },
      convert: node => {
        // @ts-ignore
        const attrs = (0, _everUtils.getAttributes)(node) || {};
        const {
          metadata,
          tagType
        } = attrs;
        const {
          imTitle
        } = metadata || {
          imTitle: ''
        };
        const targetConvert = plugins[tagType];
        if (targetConvert) return targetConvert(attrs, configs);
        return /*#__PURE__*/_createElement(React.Fragment, null, /*#__PURE__*/_createElement("span", {
          "data-type": "tag"
        }, imTitle));
      }
    };
  };
}
