// @ali/4ever-bi/dist-node/plugins/calendarCard/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everFactory = require("@ali/4ever-factory");

var _zh_CN = _interopRequireDefault(require("../../locales/zh_CN"));

var _createOnCangjiePaste = _interopRequireDefault(require("./handlers/createOnCangjiePaste"));

var _createRenderNode = _interopRequireDefault(require("./createRenderNode"));

var _createMenu = _interopRequireDefault(require("./handlers/createMenu"));

const createPlugin = bizConfig => {
  const calendarCardConfig = bizConfig.calendarCard || {
    locale: { ..._zh_CN.default.calendarCard.locale
    }
  };
  const renderNode = (0, _createRenderNode.default)(calendarCardConfig);
  return {
    cardType: 'calendar',
    cardStyle: {
      minWidth: 300
    },
    renderNode,
    onCangjiePaste: (0, _createOnCangjiePaste.default)(bizConfig),
    menubar: {
      menu: (0, _createMenu.default)(calendarCardConfig)
    }
  };
};

const createBiCalendarCardPlugin = (0, _everFactory.cardFactory)(createPlugin);
var _default = createBiCalendarCardPlugin;
exports.default = _default;
