// @ali/4ever-plugin-clipboard/dist-web/utils/constants.js
// 匹配目标格式粘贴
var MATCH_FORMAT = 'matchFormat'; // 保留源格式粘贴

var SOURCE_FROMAT = 'sourceFormat'; // 纯文本格式粘贴

var TEXT_FORMAT = 'textFormat';
var CARET_HEIGHT = 20;
var PASTE_DECORATION = 'paste-decoration';
var ICON_CHECK_STYLE = {
  fontSize: 12,
  color: '#007AFF'
};
var ICON_PASTE_STYLE = {
  fontSize: 18,
  color: '#111F2C',
  opacity: 0.72
};
var DROPDOWN_STYLE = {
  borderRadius: 4,
  border: '0.5px solid rgba(25,31,37,0.08)',
  boxShadow: '0 10px 20px 0 rgba(0,0,0,0.2)',
  fontSize: '14px'
};
var SPLIT_BUTTON_STYLE = {
  width: '100%',
  height: '100%'
};
var PASTE_MODES = {
  sourceFormat: SOURCE_FROMAT,
  matchFormat: MATCH_FORMAT,
  textFormat: TEXT_FORMAT,
  markdownToDoc: 'markdownToDoc'
};
var DEFAULT_PASTE_MODES = [PASTE_MODES.sourceFormat, PASTE_MODES.matchFormat, PASTE_MODES.textFormat];
var ClipboardErrorEnum;

(function (ClipboardErrorEnum) {
  ClipboardErrorEnum["COPY_FAILED"] = "COPY_FAILED";
  ClipboardErrorEnum["CUT_FAILED"] = "CUT_FAILED";
  ClipboardErrorEnum["PASTE_FAILED"] = "PASTE_FAILED";
  ClipboardErrorEnum["PASTE_TIME_OUT"] = "PASTE_TIME_OUT";
})(ClipboardErrorEnum || (ClipboardErrorEnum = {}));

/**
 *  1. 首字母为 # 开头，
 *  2. 首字母为 * + 空格
 *  3. 首字母为 --- + \n
 *  4. 首字母为 数字 + . + 空格
 *  满足以上任一情况，则识别为 markdown 语法
 * */
var MARKDOWN_GRAMMER_REGEXP = /^#|^(\* )|^(\- )|^(\-{3}\n|^(\d+\. ))/;
var FILE_IMG_TYPE = 'image/png';
export { CARET_HEIGHT, DROPDOWN_STYLE, MATCH_FORMAT, ICON_CHECK_STYLE, ICON_PASTE_STYLE, DEFAULT_PASTE_MODES, PASTE_DECORATION, SOURCE_FROMAT, SPLIT_BUTTON_STYLE, TEXT_FORMAT, PASTE_MODES, MARKDOWN_GRAMMER_REGEXP, FILE_IMG_TYPE, ClipboardErrorEnum };
