// @ali/we-design/dist-web/components/segment/tabs.js
import _extends from "@babel/runtime/helpers/extends";
import React, { memo } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { Tabs } from "../tabs";
var SegmentTabsWrapper = /*#__PURE__*/styled.div([".nav-bar-divider{display:none;}.segment-tab-pane{align-items:center;border-radius:4px;height:28px;box-sizing:border-box;background-color:rgba(0,0,0,0.16);&-wrapper{padding:0;flex:1;margin:0;&.actived:first-child > div::after{right:-3px;}}&-item{font-size:13px;color:#171a1d;text-align:center;position:relative;box-sizing:border-box;border-radius:4px;height:24px;line-height:24px;margin:2px;font-weight:normal;&.capsule{padding:0;background-color:unset;}&.actived{background-color:#fff;box-shadow:0px 0px 1px rgba(0,0,0,0.16),0px 1px 4px rgba(23,26,29,0.16);}&::after{content:'';position:absolute;top:5px;height:14px;width:1px;right:-2.5px;background-color:rgba(126,134,142,0.16);}}}"]);
var tabBarStyle = {
  padding: 0
};
var tabStyle = {
  alignItems: 'center'
};
export var SegmentTabs = /*#__PURE__*/memo(function (props) {
  return /*#__PURE__*/_createElement(SegmentTabsWrapper, null, /*#__PURE__*/_createElement(Tabs, _extends({
    animated: false,
    testid: "SegmentTabs"
  }, props, {
    tabBarStyle: tabBarStyle,
    type: "capsule",
    style: tabStyle,
    tabBarItemClassName: "segment-tab-pane"
  })));
});
