// @ali/4ever-bi/dist-web/plugins/vertAlign/components/subMenu.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { MenuItem } from '@ali/we-design';
import { PluginRoles } from '@ali/4ever-bamboo';
;
;
import { toggleSub } from "../actions"; // 目前 menu 不支持样式上的禁用，待统一调整

export default function SubMenu(props) {
  var menuItem = props.menuItem,
      controller = props.controller;

  var title = menuItem.title,
      rest = _objectWithoutPropertiesLoose(menuItem, ["title"]);

  var isActive = controller.query('isSub');
  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInMention') || controller.query('isSelectionInLink') || controller.query('isSelectionInListSymbol');

  var handleClick = function handleClick() {
    return controller.run('onAction', toggleSub());
  };

  return /*#__PURE__*/_createElement(MenuItem, _extends({}, rest, {
    onClick: handleClick,
    selected: isActive,
    disabled: isDisabled,
    role: PluginRoles.vertAlignSub
  }), title);
}
