// @ali/4ever-bi/dist-node/plugins/callout/mobile/createRenderNode.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _everMo = require("@ali/4ever-mo");

var _calloutMobile = _interopRequireDefault(require("../components/calloutMobile"));

const _createElement = /*#__PURE__*/React.createElement;

function createRenderNode(param) {
  return {
    [_everMo.Container.TYPE]: (props, controller) => {
      const {
        sticker,
        locale
      } = param;
      const {
        attributes,
        children
      } = props;
      const {
        node
      } = props;

      if (node.data.subType !== _everMo.ColorBlocks.SUB_TYPE) {
        return /*#__PURE__*/_createElement(React.Fragment, null);
      }

      return /*#__PURE__*/_createElement(_calloutMobile.default, {
        controller: controller,
        locale: locale,
        node: node,
        attributes: attributes,
        sticker: sticker
      }, children());
    }
  };
}

var _default = createRenderNode;
exports.default = _default;
