// @ali/4ever-utils/dist-web/theme/index.js
import _extends from "@babel/runtime/helpers/extends";
import globalStyle, { defaultKeyTypographyStyle } from "./globalStyle";
import heading from "./heading";
import paragraph from "./paragraph";
import code from "./code";
import embed from "./embed";
import onlineVideo from "./onlineVideo";
import hr from "./hr";
import { default as link, PLACEHOLDER_LINK_PADDING_HORIZONTAL, PLACEHOLDER_LINK_PADDING_VERTICAL } from "./link";
import placeholder from "./placeHolder";
import table from "./table";
import blockquote from "./blockquote";
import mention from "./mention";
import constant from "./constant";
import refer from "./refer";
import indent from "./indent";
import inlineCode from "./inlineCode";
import { fontSize, lineHeight, blockMarginStyleInContainer } from "./basic";
import spacing, { defaultCompactSpacing, defaultSpacing } from "./spacing";
import DEFAULT_SHD_STYLES from "./shd";
export { defaultSpacing, defaultCompactSpacing, fontSize, lineHeight, blockMarginStyleInContainer, mention as mentionTheme, spacing, code as codeTheme, inlineCode as inlineCodeTheme, link as linkTheme, DEFAULT_SHD_STYLES, indent };
export var theme = _extends({}, heading, table, placeholder, {
  paragraph: paragraph,
  code: code,
  embed: embed,
  inlineCode: inlineCode,
  onlineVideo: onlineVideo,
  hr: hr,
  link: link,
  PLACEHOLDER_LINK_PADDING_HORIZONTAL: PLACEHOLDER_LINK_PADDING_HORIZONTAL,
  PLACEHOLDER_LINK_PADDING_VERTICAL: PLACEHOLDER_LINK_PADDING_VERTICAL,
  blockquote: blockquote,
  mention: mention,
  globalStyle: globalStyle,
  constant: constant,
  refer: refer,
  fontSize: fontSize,
  lineHeight: lineHeight,
  defaultKeyTypographyStyle: defaultKeyTypographyStyle,
  spacing: spacing,
  DEFAULT_SHD_STYLES: DEFAULT_SHD_STYLES,
  indent: indent
});
