// @ali/4ever-bi/dist-node/plugins/link/commands/updatePureLink.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = updatePureLink;

var _everCangjie = require("@ali/4ever-cangjie");

var _removePureLink = _interopRequireDefault(require("./removePureLink"));

function updatePureLink(controller, node, offset, text, href, srcHref) {
  // 存在文本或者链接变化时，将 pure link 转换为 text link
  if (srcHref !== href || text) {
    const nodeText = text.length ? text : href;

    const point = _everCangjie.TextPoint.create({
      key: node.key,
      offset: offset
    });

    const selection = _everCangjie.Selection.create({
      anchor: point,
      focus: point
    });

    return (0, _removePureLink.default)(controller, node, offset, srcHref).command(_everCangjie.Commands.select, selection).command(_everCangjie.Commands.insertInline, {
      type: 'link',
      data: {
        href
      },
      nodes: [_everCangjie.Text.create(nodeText)]
    });
  }

  return controller;
}
