// @ali/4ever-bi/dist-node/plugins/underline/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _toggleUnderlineMark = _interopRequireDefault(require("./commands/toggleUnderlineMark"));

var _isUnderline = _interopRequireDefault(require("./queries/isUnderline"));

var actions = _interopRequireWildcard(require("./actions"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _onCangjieInput = _interopRequireDefault(require("./handlers/onCangjieInput"));

var _createRenderToolbar = _interopRequireDefault(require("./createRenderToolbar"));

var _everPluginUnderline = require("@ali/4ever-plugin-underline");

const commands = {
  toggleUnderline: _toggleUnderlineMark.default
};
const queries = {
  isUnderline: _isUnderline.default
};

function createUnderlinePlugin(config) {
  const underlineConfig = config.underline || {};
  return { ...(0, _everPluginUnderline.moFactory)({}),
    menubar: (0, _menubar.default)(),
    toolbar: (0, _toolbar.default)(underlineConfig),
    renderToolbar: (0, _createRenderToolbar.default)(underlineConfig.locale),
    commands,
    queries,
    actions,
    onAction: _onAction.default,
    onKeyDown: _onKeyDown.default,
    onCangjieInput: (0, _onCangjieInput.default)(),
    models: {
      klass: 'mark',
      type: 'underline'
    }
  };
}

createUnderlinePlugin.actions = actions;
createUnderlinePlugin.commands = commands;
createUnderlinePlugin.queries = queries;
var _default = createUnderlinePlugin;
exports.default = _default;
