// @builder/vite-service/lib/plugins/html.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.htmlPlugin = void 0;
var lodash_1 = require("lodash");
var fs = require("fs-extra");
var path = require("path");
var cheerio_1 = require("cheerio");
var app_helpers_1 = require("@builder/app-helpers");
var getHtmlContent = function (_a) {
    var template = _a.template, entry = _a.entry, templateParameters = _a.templateParameters;
    var html = fs.readFileSync(template, 'utf-8');
    if (templateParameters) {
        var compiled = (0, lodash_1.template)(html);
        try {
            html = compiled(templateParameters);
        }
        catch (e) {
            console.log('');
        }
    }
    var $ = cheerio_1.default.load(html);
    $('body').append("<script type=\"module\" src=\"".concat(entry, "\" />"));
    return $.html({});
};
var htmlPlugin = function (_a) {
    var filename = _a.filename, template = _a.template, entry = _a.entry, rootDir = _a.rootDir, _b = _a.templateParameters, templateParameters = _b === void 0 ? {} : _b, ssr = _a.ssr, command = _a.command;
    var pageName = filename.replace('.html', '');
    var getEntry = function () {
        var entryPath = entry;
        if (entry.includes(rootDir)) {
            entryPath = path.relative(rootDir, entry);
        }
        return "/".concat(entryPath);
    };
    var html = getHtmlContent({
        entry: getEntry(),
        template: template,
        templateParameters: templateParameters,
    });
    // vite will get relative path by `path.posix.relative(config.root, id)`
    // path.posix.relative will get error path when pass relative path of index html
    var absoluteHtmlPath = (0, app_helpers_1.formatPath)(path.join(rootDir, filename));
    var plugin = {
        name: "vite-plugin-html-".concat(pageName),
        enforce: 'pre',
        config: function (cfg) {
            cfg.build = (0, lodash_1.set)(cfg.build, "rollupOptions.input.".concat(pageName), absoluteHtmlPath);
        },
        resolveId: function (id) {
            if (id.includes('.html')) {
                // win32 场景下在不进行 formatPath，返回的 id 类似 D:\workspace\project\index.html
                // 代码 https://github.com/vitejs/vite/blob/3ceffcca66311f9a7d71612a596b84888c3f843b/packages/vite/src/node/plugins/html.ts#L472 处理后得到 ../../D:\workspace\project\index.html 导致报错，其中 config.root 为 D:/workspace/project
                return (0, app_helpers_1.formatPath)(id);
            }
            return null;
        },
        load: function (id) {
            if ((0, app_helpers_1.formatPath)(id) === absoluteHtmlPath) {
                return html;
            }
            return null;
        },
        configureServer: function (server) {
            var _this = this;
            if (!ssr) {
                return function () {
                    server.middlewares.use(function (req, res, next) { return __awaiter(_this, void 0, void 0, function () {
                        var _a, _b, e_1;
                        var _c;
                        return __generator(this, function (_d) {
                            switch (_d.label) {
                                case 0:
                                    if (!((_c = req.url) === null || _c === void 0 ? void 0 : _c.endsWith('.html')) && req.url !== '/') {
                                        return [2 /*return*/, next()];
                                    }
                                    if (!(req.url === "/".concat(filename))) return [3 /*break*/, 4];
                                    _d.label = 1;
                                case 1:
                                    _d.trys.push([1, 3, , 4]);
                                    res.setHeader('Content-Type', 'text/html');
                                    _b = (_a = res).end;
                                    return [4 /*yield*/, server.transformIndexHtml(req.url, html)];
                                case 2:
                                    _b.apply(_a, [_d.sent()]);
                                    return [3 /*break*/, 4];
                                case 3:
                                    e_1 = _d.sent();
                                    return [2 /*return*/, next(e_1)];
                                case 4:
                                    next();
                                    return [2 /*return*/];
                            }
                        });
                    }); });
                };
            }
        }
    };
    // ssr 在 dev 阶段由中间件进行 html 返回
    if (ssr && command === 'start') {
        plugin.transformIndexHtml = function () {
            return html;
        };
    }
    return plugin;
};
exports.htmlPlugin = htmlPlugin;
