// @ali/4ever-plugin-calendar-card/dist-node/utils/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RECEIVER_MAX_WIDTH = exports.MORE_INDO_EXTRA = exports.RECEIVER_WRAPPER_MARGIN = exports.NAME_WRAPPER_PADDING = exports.AVATAR_WRAPPER_SZ = exports.DEFAULT_ICON_STYLE = exports.RECEIVER_MARGIN = exports.CALENDAR_CARD_MAX_RECEIVER_WIDTH = exports.CALENDAR_CARD_SINGLE_LINE_HEIGHT = exports.CALENDAR_CARD_PARAGRAPH_MARGIN_TOP = exports.CALENDAR_CARD_EXTRA_ICON_WIDTH = exports.CALENDAR_CARD_ICON_WIDTH = exports.CALENDAR_CARD_PADDING_LEN = exports.CALENDAR_CARD_MAX_HEIGHT = exports.CALENDAR_CARD_MIN_HEIGHT = void 0;
// 最小高度
const CALENDAR_CARD_MIN_HEIGHT = 129;
exports.CALENDAR_CARD_MIN_HEIGHT = CALENDAR_CARD_MIN_HEIGHT;
const CALENDAR_CARD_MAX_HEIGHT = 318; // padding都是12

exports.CALENDAR_CARD_MAX_HEIGHT = CALENDAR_CARD_MAX_HEIGHT;
const CALENDAR_CARD_PADDING_LEN = 12; // icon 的宽度

exports.CALENDAR_CARD_PADDING_LEN = CALENDAR_CARD_PADDING_LEN;
const CALENDAR_CARD_ICON_WIDTH = 26; // extra icon 的宽度

exports.CALENDAR_CARD_ICON_WIDTH = CALENDAR_CARD_ICON_WIDTH;
const CALENDAR_CARD_EXTRA_ICON_WIDTH = 30;
exports.CALENDAR_CARD_EXTRA_ICON_WIDTH = CALENDAR_CARD_EXTRA_ICON_WIDTH;
const CALENDAR_CARD_PARAGRAPH_MARGIN_TOP = 7;
exports.CALENDAR_CARD_PARAGRAPH_MARGIN_TOP = CALENDAR_CARD_PARAGRAPH_MARGIN_TOP;
const CALENDAR_CARD_SINGLE_LINE_HEIGHT = 29; // 移动端增加左右 20px 边距，这

exports.CALENDAR_CARD_SINGLE_LINE_HEIGHT = CALENDAR_CARD_SINGLE_LINE_HEIGHT;
const CALENDAR_CARD_MAX_RECEIVER_WIDTH = 225;
exports.CALENDAR_CARD_MAX_RECEIVER_WIDTH = CALENDAR_CARD_MAX_RECEIVER_WIDTH;
const RECEIVER_MARGIN = 5;
exports.RECEIVER_MARGIN = RECEIVER_MARGIN;
const DEFAULT_ICON_STYLE = {
  width: '20px',
  height: '20px'
}; // 接收者头像宽度

exports.DEFAULT_ICON_STYLE = DEFAULT_ICON_STYLE;
const AVATAR_WRAPPER_SZ = 24; // 名字容器 padding

exports.AVATAR_WRAPPER_SZ = AVATAR_WRAPPER_SZ;
const NAME_WRAPPER_PADDING = 4; // 接受者容器 margin

exports.NAME_WRAPPER_PADDING = NAME_WRAPPER_PADDING;
const RECEIVER_WRAPPER_MARGIN = 4; // 更多参会人信息额外宽度，16 是 ">" 的svg 宽度

exports.RECEIVER_WRAPPER_MARGIN = RECEIVER_WRAPPER_MARGIN;
const MORE_INDO_EXTRA = RECEIVER_WRAPPER_MARGIN + NAME_WRAPPER_PADDING + 16; // 接收者信息渲染的最大宽度

exports.MORE_INDO_EXTRA = MORE_INDO_EXTRA;
const RECEIVER_MAX_WIDTH = 229;
exports.RECEIVER_MAX_WIDTH = RECEIVER_MAX_WIDTH;
