// @ali/4ever-plugin-calendar-card/dist-node/utils/type.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarType = exports.AttachmentTypeEnum = exports.ReceiverTypeEnum = void 0;
// 1-未响应，2-已接受，10-已拒绝, 20-待定；目前仅显示未响应跟已拒绝人员
let ReceiverTypeEnum;
exports.ReceiverTypeEnum = ReceiverTypeEnum;

(function (ReceiverTypeEnum) {
  ReceiverTypeEnum[ReceiverTypeEnum["noresponse"] = 1] = "noresponse";
  ReceiverTypeEnum[ReceiverTypeEnum["refused"] = 10] = "refused";
})(ReceiverTypeEnum || (exports.ReceiverTypeEnum = ReceiverTypeEnum = {}));

let AttachmentTypeEnum;
exports.AttachmentTypeEnum = AttachmentTypeEnum;

(function (AttachmentTypeEnum) {
  AttachmentTypeEnum[AttachmentTypeEnum["text"] = 1] = "text";
  AttachmentTypeEnum[AttachmentTypeEnum["audio"] = 2] = "audio";
  AttachmentTypeEnum[AttachmentTypeEnum["image"] = 3] = "image";
  AttachmentTypeEnum[AttachmentTypeEnum["file"] = 4] = "file";
  AttachmentTypeEnum[AttachmentTypeEnum["link"] = 5] = "link";
  AttachmentTypeEnum[AttachmentTypeEnum["redenvelop"] = 6] = "redenvelop";
  AttachmentTypeEnum[AttachmentTypeEnum["dingfile"] = 7] = "dingfile";
})(AttachmentTypeEnum || (exports.AttachmentTypeEnum = AttachmentTypeEnum = {}));

let CalendarType; // 接受者列表

exports.CalendarType = CalendarType;

(function (CalendarType) {
  CalendarType["SHANHUI"] = "shanhui";
})(CalendarType || (exports.CalendarType = CalendarType = {}));
