// @ali/we-ui/dist-web/components/lines/commonLine.js
import _taggedTemplateLiteralLoose from "@babel/runtime/helpers/taggedTemplateLiteralLoose";

function _templateObject14() {
  var data = _taggedTemplateLiteralLoose(["e7f9"], ["\\e7f9"]);

  _templateObject14 = function _templateObject14() {
    return data;
  };

  return data;
}

function _templateObject13() {
  var data = _taggedTemplateLiteralLoose(["e7f2"], ["\\e7f2"]);

  _templateObject13 = function _templateObject13() {
    return data;
  };

  return data;
}

function _templateObject12() {
  var data = _taggedTemplateLiteralLoose(["e7f1"], ["\\e7f1"]);

  _templateObject12 = function _templateObject12() {
    return data;
  };

  return data;
}

function _templateObject11() {
  var data = _taggedTemplateLiteralLoose(["e7e7"], ["\\e7e7"]);

  _templateObject11 = function _templateObject11() {
    return data;
  };

  return data;
}

function _templateObject10() {
  var data = _taggedTemplateLiteralLoose(["e7ef"], ["\\e7ef"]);

  _templateObject10 = function _templateObject10() {
    return data;
  };

  return data;
}

function _templateObject9() {
  var data = _taggedTemplateLiteralLoose(["e7f0"], ["\\e7f0"]);

  _templateObject9 = function _templateObject9() {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = _taggedTemplateLiteralLoose(["e7f3"], ["\\e7f3"]);

  _templateObject8 = function _templateObject8() {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = _taggedTemplateLiteralLoose(["e7ec"], ["\\e7ec"]);

  _templateObject7 = function _templateObject7() {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = _taggedTemplateLiteralLoose(["e7ea"], ["\\e7ea"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = _taggedTemplateLiteralLoose(["e7ed"], ["\\e7ed"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteralLoose(["e7f6"], ["\\e7f6"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteralLoose(["e7f4"], ["\\e7f4"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose(["e7f8"], ["\\e7f8"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["e7f5"], ["\\e7f5"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Wrapper, Line, Single } from "./styled";
var defaultLineColor = '#000';
export var lineWidthByBuiltIn = [1, 2, 4, 6];
export var lineWidthByCustom = [4, 6]; // 线段属性

export var commonLineMap = {
  single: {
    content: '',
    repeat: 0,
    sizeList: lineWidthByBuiltIn,
    defaultSize: 2,
    unitHeight: 6
  },
  dotted: {
    content: String.raw(_templateObject()),
    repeat: 800,
    sizeList: lineWidthByBuiltIn,
    defaultSize: 2,
    unitHeight: 16 / 3
  },
  dashed: {
    content: String.raw(_templateObject2()),
    repeat: 400,
    sizeList: lineWidthByBuiltIn,
    defaultSize: 2,
    unitHeight: 16 / 2
  },
  dotDash: {
    content: String.raw(_templateObject3()),
    repeat: 400,
    sizeList: lineWidthByBuiltIn,
    defaultSize: 2,
    unitHeight: 16
  },
  dotDotDash: {
    content: String.raw(_templateObject4()),
    repeat: 400,
    sizeList: lineWidthByBuiltIn,
    defaultSize: 2,
    unitHeight: 16
  },
  "double": {
    content: String.raw(_templateObject5()),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 4
  },
  thickThinSmallGap: {
    content: String.raw(_templateObject6()),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 16 / 6
  },
  thinThickThinMediumGap: {
    content: String.raw(_templateObject7()),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 16 / 6
  },
  dashStroked: {
    content: String.raw(_templateObject8()),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 2,
    unitHeight: 16 / 7
  },
  wave: {
    content: String.raw(_templateObject9()),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 10,
    scale: 0.2
  },
  doubleWave: {
    content: String.raw(_templateObject10()),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 6,
    scale: 0.2
  },
  dashDotStroked: {
    content: String.raw(_templateObject11()),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 16 / 5
  },
  widthDoubleWave: {
    content: String.raw(_templateObject12()),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 10,
    scale: 0.2
  },
  widthWave: {
    content: String.raw(_templateObject13()),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 10,
    scale: 0.2
  },
  roundDot: {
    content: String.raw(_templateObject14()),
    repeat: 400,
    sizeList: lineWidthByCustom,
    defaultSize: 4,
    unitHeight: 16 / 3
  }
};

function repeatContent(content, times) {
  if (!content || !times) return '';
  var res = content.repeat(times);
  return res;
}

var ICONFONT_CSS = "\n@font-face {\n  font-family: 'commonLineIconfont';  /* project id 1679878 */\n  src: url('//at.alicdn.com/t/font_1679878_3m8w8kt4k7i.eot');\n  src: url('//at.alicdn.com/t/font_1679878_3m8w8kt4k7i.eot?#iefix') format('embedded-opentype'),\n  url('//at.alicdn.com/t/font_1679878_3m8w8kt4k7i.woff2') format('woff2'),\n  url('//at.alicdn.com/t/font_1679878_3m8w8kt4k7i.woff') format('woff'),\n  url('//at.alicdn.com/t/font_1679878_3m8w8kt4k7i.ttf') format('truetype'),\n  url('//at.alicdn.com/t/font_1679878_3m8w8kt4k7i.svg#iconfont') format('svg');\n}\n";
var ICONFONT_ID = 'common-line-font-styles';

var inject = function inject() {
  var styleSheet = document.getElementById(ICONFONT_ID);

  if (!styleSheet) {
    styleSheet = document.createElement('style');
    styleSheet.setAttribute('type', 'text/css');
    styleSheet.id = ICONFONT_ID;
    styleSheet.innerHTML = ICONFONT_CSS;
    var head = document.getElementsByTagName('head')[0];

    if (head) {
      head.appendChild(styleSheet);
    }
  }
};

export var CommonLine = function CommonLine(props) {
  var size = props.size,
      length = props.length,
      color = props.color,
      lineStyle = props.lineType; // 注入全局 iconfont 样式

  inject();
  var finalColor = color; // 解决css color 无auto的问题

  if (color === undefined || color === '' || color === 'auto') {
    finalColor = defaultLineColor;
  }

  var finalLineStyle = lineStyle; // 加入最小单位高度, 用于做transform的scale比例计算

  var unitHeight = 8;
  var scale = 1;
  var selectedLine = commonLineMap[lineStyle];
  var content = commonLineMap.single && commonLineMap.single.content;

  if (selectedLine) {
    content = repeatContent(selectedLine.content, selectedLine.repeat);
    unitHeight = selectedLine.unitHeight;
    scale = selectedLine.scale || 1;
  } else {
    // 如果未找到, 则single 代替
    finalLineStyle = 'single';
  }
  /* eslint-disable */


  return /*#__PURE__*/_createElement(Wrapper, null, finalLineStyle === 'single' ? /*#__PURE__*/_createElement(Single, {
    length: length,
    bgColor: finalColor,
    size: size,
    color: finalColor
  }) : /*#__PURE__*/_createElement(Line, {
    length: length,
    lineContent: content,
    size: size,
    color: color,
    unitHeight: unitHeight,
    scale: scale
  }));
};
CommonLine.defaultProps = {
  lineType: 'single',
  size: 1,
  // 默认值 兼容之前hr颜色和厚度方案
  color: defaultLineColor
}; // export default CommonLine;
