// @ali/we-ui/dist-web/components/lines/styled.js
import styled from 'styled-components';
import { SelectOption } from "../select";
export var LineBox = /*#__PURE__*/styled.div(["overflow:hidden;white-space:nowrap;display:flex;height:24px;align-items:center;"]);
export var Wrapper = /*#__PURE__*/styled.div(["&>*{margin:6px 0;}"]);
export var Line = /*#__PURE__*/styled.hr(["border:none;line-height:", ";overflow:hidden;position:relative;letter-spacing:0;width:", ";height:", ";background-color:", ";::before{font-size:", ";font-family:\"commonLineIconfont\" !important;content:\"", "\";color:", ";white-space:nowrap;font-style:normal;transform:", ";transform-origin:left;position:absolute;}"], function (props) {
  return props.size ? props.size + "px" : 'inherit';
}, function (props) {
  return props.length ? props.length + "px" : 'inherit';
}, function (props) {
  return props.size ? props.size + "px" : 'inherit';
}, function (props) {
  return props.bgColor ? props.bgColor : 'transparent';
}, function (props) {
  return props.size ? props.unitHeight * props.size + "px" : 'auto';
}, function (props) {
  return props.lineContent ? props.lineContent : '';
}, function (props) {
  return props.color ? props.color : 'inherit';
}, function (props) {
  return props.scale > 0 ? "scale(" + props.scale + ")" : 'scale(1)';
});
export var Single = /*#__PURE__*/styled.hr(["border:none;line-height:", ";overflow:hidden;width:", ";height:", ";background-color:", ";"], function (props) {
  return props.size ? props.size + "px" : 'inherit';
}, function (props) {
  return props.length ? props.length + "px" : 'inherit';
}, function (props) {
  return props.size ? props.size + "px" : 'inherit';
}, function (props) {
  return props.bgColor ? props.bgColor : 'inherit';
});
export var OptionStyled = /*#__PURE__*/styled(SelectOption)(["&.with-text{display:flex;}"]);
export var TextSpan = /*#__PURE__*/styled.span(["box-sizing:border-box;width:25px;margin-right:0;"]);
export var TriggerStyled = /*#__PURE__*/styled.div(["display:flex;font-size:12px;"]);
