// @ali/4ever-plugin-table/dist-node/zhi/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mobileZhiFactory = exports.zhiFactory = exports.createZhiFactory = void 0;

var _index = require("../mo/index");

var _createRenderNode = require("./handlers/createRenderNode");

var _queries = _interopRequireDefault(require("./queries"));

var _getNodeStyle = _interopRequireDefault(require("../common/handlers/getNodeStyle"));

var _onAction = _interopRequireDefault(require("../utils/handlers/onAction"));

var _shouldNodeComponentUpdate = _interopRequireDefault(require("../common/handlers/shouldNodeComponentUpdate"));

var _createOnCangjieCopy = _interopRequireDefault(require("./handlers/createOnCangjieCopy"));

const createZhiFactory = isMobile => {
  return configs => {
    const tableConfig = configs.table || {};
    return { ...(0, _index.moFactory)({}),
      queries: _queries.default,
      renderNode: (0, _createRenderNode.createRenderNode)(tableConfig, isMobile),
      getNodeStyle: _getNodeStyle.default,
      onAction: _onAction.default,
      shouldNodeComponentUpdate: _shouldNodeComponentUpdate.default,
      onCangjieCopy: (0, _createOnCangjieCopy.default)(isMobile)
    };
  };
};

exports.createZhiFactory = createZhiFactory;
const zhiFactory = createZhiFactory();
exports.zhiFactory = zhiFactory;
const mobileZhiFactory = createZhiFactory(true);
exports.mobileZhiFactory = mobileZhiFactory;
