// @ali/4ever-plugin-code/dist-web/zhi/index.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { getData } from '@ali/4ever-utils';
import moFactory from "../mo/index";
import lazyRenderStrategy from "../common/queries/lazyRenderStrategy";
import CodeComponent from "../components/ZhiCode";
import { DEFAULT_THEME } from "../utils/constants/themes";
import { DEFAULT_LANGUAGE } from "../utils/constants/languages";
export var zhiFactory = function zhiFactory(configs) {
  return _extends({}, moFactory({}), {
    queries: {
      lazyRenderStrategy: lazyRenderStrategy
    },
    renderNode: function renderNode(props, controller, next) {
      var node = props.node,
          attributes = props.attributes;

      var ref = attributes.ref,
          rest = _objectWithoutPropertiesLoose(attributes, ["ref"]);

      var _ref = node,
          type = _ref.type,
          text = _ref.text,
          data = _ref.data;

      if (type !== 'code') {
        return next();
      }

      var _ref2 = configs.code || {},
          _ref2$print = _ref2.print,
          print = _ref2$print === void 0 ? false : _ref2$print;

      return /*#__PURE__*/_createElement(CodeComponent, _extends({}, rest, {
        controller: controller,
        code: text,
        node: node,
        theme: getData(data, 'theme', DEFAULT_THEME),
        syntax: getData(data, 'syntax', DEFAULT_LANGUAGE),
        height: getData(data, 'height', undefined),
        lineWrapping: !!getData(data, 'wrap', false),
        print: print
      }));
    }
  });
};
