// @ali/4ever-plugin-code/dist-web/zhi/mobile.js
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { getData } from '@ali/4ever-utils';
import moFactory from "../mo/index";
import lazyRenderStrategy from "../common/queries/lazyRenderStrategy";
import SimpleCode from "../components/SimpleCode";
export var mobileZhiFactory = function mobileZhiFactory(configs) {
  return _extends({}, moFactory({}), {
    queries: {
      lazyRenderStrategy: lazyRenderStrategy
    },
    renderNode: function renderNode(props, _, next) {
      var node = props.node,
          attributes = props.attributes;
      var _ref = node,
          type = _ref.type,
          text = _ref.text,
          data = _ref.data;

      if (type !== 'code') {
        return next();
      }

      var _ref2 = configs.code || {},
          _ref2$print = _ref2.print,
          print = _ref2$print === void 0 ? false : _ref2$print;

      return /*#__PURE__*/_createElement(SimpleCode, _extends({
        node: node
      }, attributes, {
        code: text,
        theme: getData(data, 'theme', 'default'),
        height: !print ? getData(data, 'height', undefined) : undefined,
        lineWrapping: !!getData(data, 'wrap', false) || print,
        print: print
      }));
    }
  });
};
