// @ali/4ever-bi/dist-web/plugins/hoverToolbar/utils/hoverCapture.js
export var HoverCapture = /*#__PURE__*/function () {
  function HoverCapture() {
    this.callbacks = [];
    this.prev = void 0;
    this.disabled = false;
  }

  var _proto = HoverCapture.prototype;

  _proto.on = function on(c) {
    this.callbacks.push(c);
  };

  _proto.off = function off(c) {
    this.callbacks = this.callbacks.filter(function (cb) {
      return cb !== c;
    });
  };

  _proto.disableTrigger = function disableTrigger() {
    this.disabled = true;
  };

  _proto.enableTrigger = function enableTrigger() {
    this.disabled = false;
  };

  _proto.trigger = function trigger(msg) {
    if (msg === void 0) {
      msg = {};
    }

    if (this.prev && this.prev === msg.node || this.disabled) {
      return;
    }

    this.prev = msg.node;
    this.callbacks.forEach(function (c) {
      c(msg);
    });
  };

  _proto.forceClose = function forceClose() {
    this.trigger({});
  };

  _proto.getCurrentNode = function getCurrentNode() {
    return this.prev;
  };

  return HoverCapture;
}();
export var hoverCapture = new HoverCapture();
