// @ali/4ever-component/dist-node/leftToolbar/reducer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.preventMouseMove = preventMouseMove;
exports.clearPrevent = clearPrevent;
exports.updateHoverBlock = updateHoverBlock;
exports.default = leftToolbarReducer;
const PREVENT_MOUSE_MOVE = Symbol('PREVENT_MOUSE_MOVE');
const CLEAR_PREVENT = Symbol('CLEAR_PREVENT');
const UPDATE_HOVER_BLOCK = Symbol('UPDATE_HOVER_BLOCK');

function preventMouseMove() {
  return {
    type: PREVENT_MOUSE_MOVE
  };
}

function clearPrevent() {
  return {
    type: CLEAR_PREVENT
  };
}

function updateHoverBlock(payload) {
  return {
    type: UPDATE_HOVER_BLOCK,
    payload
  };
}

function leftToolbarReducer(data, action) {
  const {
    type,
    payload
  } = action;
  const {
    block,
    ...rest
  } = data;

  switch (type) {
    case PREVENT_MOUSE_MOVE:
      return { ...data,
        preventMouseMove: true
      };

    case UPDATE_HOVER_BLOCK:
      if (!payload) {
        return rest;
      }

      return { ...rest,
        block: payload
      };

    case CLEAR_PREVENT:
      return { ...data,
        preventMouseMove: false
      };

    default:
      return data;
  }
}
