// @ali/4ever-plugin-paragraph/dist-node/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.canSetParagraphAttr = canSetParagraphAttr;
exports.DEFAULT_SPACING = exports.generalParagraphType = exports.DEFAULT_INDENT_FULL = void 0;

var _everUtils = require("@ali/4ever-utils");

var _meta = require("../mo/meta");

// TOOD: 与 heading 解耦
const NODE_CAN_SET_SPACING = {
  paragraph: true,
  'heading-1': true,
  'heading-2': true,
  'heading-3': true,
  'heading-4': true,
  'heading-5': true,
  'heading-6': true,
  hr: true,
  code: true,
  embed: true,
  card: true
};

function canSetParagraphAttr(type) {
  return !!NODE_CAN_SET_SPACING[type];
}

const generalParagraphType = {};
exports.generalParagraphType = generalParagraphType;
[_meta.ELEMENT_TYPE, 'heading-1', 'heading-2', 'heading-3', 'heading-4', 'heading-5', 'heading-6'].forEach(type => {
  generalParagraphType[type] = true;
});
const DEFAULT_SPACING = {
  line: 1,
  lineRule: 'auto',
  before: 0,
  beforeLines: 0,
  beforeAutospacing: false,
  after: 0,
  afterLines: 0,
  afterAutospacing: false
};
exports.DEFAULT_SPACING = DEFAULT_SPACING;
const DEFAULT_INDENT_FULL = { ..._everUtils.DEFAULT_INDENT,
  right: 0,
  rightChars: 0,
  start: 0,
  end: 0
};
exports.DEFAULT_INDENT_FULL = DEFAULT_INDENT_FULL;
