// @ali/4ever-plugin-list-group/dist-node/utils/handlers/createRenderNode.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _everUtils = require("@ali/4ever-utils");

var _everPluginGroupFactory = require("@ali/4ever-plugin-group-factory");

var _groupBackground = _interopRequireDefault(require("../components/groupBackground"));

var _multibar = _interopRequireDefault(require("../components/multibar"));

var _utils = require("../utils");

var _constants = require("../constants");

var _role = require("../role");

const _createElement = /*#__PURE__*/_react.default.createElement;

const renderBackground = (width, backgroundColor) => {
  return backgroundColor ? /*#__PURE__*/_createElement(_groupBackground.default, {
    width: width,
    backgroundColor: backgroundColor
  }) : null;
};

function createRenderNode(config) {
  return {
    [_constants.LIST_GROUP_TYPE]: (props, controller) => {
      const {
        attributes,
        isSelected,
        children,
        visible
      } = props;
      const node = props.node;
      const {
        shd
      } = node.data;
      const {
        symbolWidth,
        left
      } = (0, _utils.calcMultibarPosition)(node);
      const foldBoxLeft = (0, _utils.calcFoldBoxPosition)(node);

      const backgroundColor = _everUtils.ShdUtil.transformShd(shd);

      const backgroundWidth = (0, _utils.calcBackgroundWidth)(node, left, symbolWidth);
      const attrs = (0, _utils.getAttributes)(attributes, node, controller);
      return /*#__PURE__*/_createElement(_everPluginGroupFactory.Group, {
        node: node,
        attributes: attrs,
        controller: controller,
        isSelected: isSelected,
        renderExtra: /*#__PURE__*/_createElement(_multibar.default, {
          node: node
        }),
        renderBackground: renderBackground(backgroundWidth, backgroundColor),
        foldBoxLeft: foldBoxLeft,
        role: _role.ListGroupRole.listGroup,
        visible: visible
      }, children());
    }
  };
}

var _default = createRenderNode;
exports.default = _default;
