// @ali/4ever-plugin-table/dist-web/bi/components/toolbarButtons/tableButton.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Button, TableLightNormal, TableNormal } from '@ali/we-design';
import { TableRole as PluginRoles } from "../../../utils/role";
import { insertTable } from "../../actions";
import { ToolbarMode } from '@ali/4ever-component';

var _ref = /*#__PURE__*/_createElement(TableNormal, null);

var _ref2 = /*#__PURE__*/_createElement(TableLightNormal, null);

export default function TableButton(props) {
  var controller = props.controller,
      locale = props.locale,
      toolbarMode = props.toolbarMode;
  var isDisabled = !!controller.query('canInsertTable');
  var handleClick = React.useCallback(function () {
    return controller.run('onAction', insertTable(3, 3));
  }, [controller]);
  return /*#__PURE__*/_createElement(Button, {
    title: locale.buttonTitle,
    bordered: false,
    className: "bi-toolbar-table",
    icon: toolbarMode === ToolbarMode.single ? _ref : _ref2,
    disabled: isDisabled,
    onClick: handleClick,
    role: PluginRoles.table
  });
}
