// @ali/4ever-plugin-table/dist-web/bi/components/toolbarButtons/tableDeleteButton.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { IconButton } from '@ali/we-toolbar';
import { ToolbarTableDeleteNormal } from '@ali/we-design';
import { TableRole as PluginRoles } from "../../../utils/role";
import getTableSelection from "../../utils/getTableSelection";
import { deleteTable, deleteTableCols, deleteTableRows } from "../../actions";
import { isTable } from "../../types";

var _ref = /*#__PURE__*/_createElement(ToolbarTableDeleteNormal, null);

export default function TableDeleteButton(props) {
  var controller = props.controller,
      locale = props.locale;
  var selection = getTableSelection(controller);
  if (!selection) return null;
  var node = controller.value.document.getNode(selection == null ? void 0 : selection.key);
  if (!(node && isTable(node))) return null;
  var tableRowCount = node.nodes.length - 1;
  var tableColCount = (node.data.colsWidth || []).length - 1;
  var startRowIndex = selection.startRowIndex,
      startColIndex = selection.startColIndex,
      endRowIndex = selection.endRowIndex,
      endColIndex = selection.endColIndex;
  var isSelectWholeRow = startColIndex === 0 && endColIndex >= tableColCount;
  var isSelectWholeCol = startRowIndex === 0 && endRowIndex >= tableRowCount;
  var isSelectWholeTable = isSelectWholeCol && isSelectWholeRow;

  if (!isSelectWholeRow && !isSelectWholeCol) {
    return null;
  }

  var handleMouseDown = function handleMouseDown() {
    if (isSelectWholeTable) {
      controller.run('onAction', deleteTable(node));
    } else if (isSelectWholeRow) {
      controller.run('onAction', deleteTableRows(node));
    } else if (isSelectWholeCol) {
      controller.run('onAction', deleteTableCols(node));
    }
  };

  return /*#__PURE__*/_createElement(IconButton, {
    icon: _ref,
    title: locale == null ? void 0 : locale.contextMenuDelete,
    tooltip: locale == null ? void 0 : locale.contextMenuDelete,
    testid: "bi-toolbar-table-delete",
    onClick: handleMouseDown,
    role: PluginRoles.tableDelete
  });
}

TableDeleteButton.needRender = function (controller) {
  var selection = getTableSelection(controller);
  if (!selection) return false;
  var node = controller.value.document.getNode(selection.key);
  if (!(node && isTable(node))) return false;
  var tableRowCount = node.nodes.length - 1;
  var tableColCount = (node.data.colsWidth || []).length - 1;
  var startRowIndex = selection.startRowIndex,
      startColIndex = selection.startColIndex,
      endRowIndex = selection.endRowIndex,
      endColIndex = selection.endColIndex;
  var isSelectWholeRow = startColIndex === 0 && endColIndex >= tableColCount;
  var isSelectWholeCol = startRowIndex === 0 && endRowIndex >= tableRowCount;

  if (!isSelectWholeRow && !isSelectWholeCol) {
    return false;
  }

  return true;
};
