// @ali/4ever-plugin-table/dist-web/bi/components/toolbarButtons/tableSimpleSplitButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { IconButton } from '@ali/we-toolbar';
import { ToolbarSplitCellNormal } from '@ali/we-design';
import { TableRole as PluginRoles } from "../../../utils/role";
import { splitTableCell } from "../../actions";
import getTable from "../../utils/getTable";
import getTableSelection from "../../utils/getTableSelection";

var _ref = /*#__PURE__*/_createElement(ToolbarSplitCellNormal, null);

export default function TableSimpleSplitButton(props) {
  var controller = props.controller,
      locale = props.locale,
      rest = _objectWithoutPropertiesLoose(props, ["controller", "locale"]);

  var table = getTable(controller.value);
  var selection = getTableSelection(controller);
  var isSelectionSupportSplit = React.useMemo(function () {
    return !!table && controller.query('isSelectionSupportSplit', {
      visible: true,
      node: table
    }); // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [table, selection, controller]);
  var isSelectionSupportMerge = React.useMemo(function () {
    return !!table && controller.query('isSelectionSupportMerge', {
      visible: true,
      node: table
    }); // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [table, selection, controller]);
  var title = locale.contextMenuSplit;

  var handleClick = function handleClick() {
    controller.run('onAction', splitTableCell(table));
  }; // MergeButton SplitButton 二选一逻辑，非常业务化
  // 但是条件又属于与表格组件逻辑强关联
  // 目前除了在此处直接蒸发，没有其他方式实现


  if (isSelectionSupportMerge || !isSelectionSupportSplit) {
    return null;
  }

  return /*#__PURE__*/_createElement(IconButton, _extends({}, rest, {
    icon: _ref,
    title: title,
    tooltip: title,
    testid: "bi-toolbar-tablesimplesplit",
    onClick: handleClick,
    disabled: !isSelectionSupportSplit,
    role: PluginRoles.tableSplit
  }));
}
