// @ali/4ever-plugin-code/dist-node/mo/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

var _serializer = _interopRequireDefault(require("./serializer"));

var _meta = require("./meta");

const factory = () => ({
  key: 'code',
  schemas: {
    type: _meta.ELEMENT_TYPE,
    isBlock: true,
    isEdgeSelectable: true
  },
  normalizeNode: (node, controller, next) => {
    if (_everCangjie.Block.isBlock(node) && node.type === 'code') {
      if (node.nodes.length !== 1 || !_everCangjie.Text.isText(node.nodes[0]) || node.nodes[0].leaves.length !== 1) {
        return () => {
          const newLeaf = _everCangjie.Text.create(node.text); // 删除之前的节点


          const {
            length
          } = node.nodes;

          for (let i = 0; i < length; i += 1) {
            const child = node.nodes[i];
            controller.command(_everCangjie.Commands.removeNodeByKey, child.key);
          } // 插入 text 节点


          controller.command(_everCangjie.Commands.insertNodeByKey, node.key, 0, newLeaf);
        };
      }
    }

    return next();
  },
  serializer: (0, _serializer.default)(),
  models: {
    klass: 'block',
    type: _meta.ELEMENT_TYPE,
    data: {
      ind: [undefined, Object],
      blockquote: [undefined, Boolean],
      spacing: [undefined, Object],
      jc: [undefined, String],
      calloutPr: [undefined, Object],
      subjectPr: [undefined, Object],
      uuid: [undefined, String],
      shd: [undefined, Object]
    }
  }
});

var _default = factory;
exports.default = _default;
