// @ali/4ever-plugin-mention/dist-node/bi/actions/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.deactiveMention = deactiveMention;
exports.insertMention = insertMention;
exports.insertMentions = insertMentions;
exports.cancelMentionSelect = cancelMentionSelect;
exports.insertPlaceholder = insertPlaceholder;
exports.exitPlaceholder = exitPlaceholder;
exports.cancelPlaceholder = cancelPlaceholder;
Object.defineProperty(exports, "MENTION_ACTIVE", {
  enumerable: true,
  get: function () {
    return _activeMention.MENTION_ACTIVE;
  }
});
Object.defineProperty(exports, "activeMention", {
  enumerable: true,
  get: function () {
    return _activeMention.activeMention;
  }
});
Object.defineProperty(exports, "MENTION_SELECT", {
  enumerable: true,
  get: function () {
    return _selectMention.MENTION_SELECT;
  }
});
Object.defineProperty(exports, "selectMention", {
  enumerable: true,
  get: function () {
    return _selectMention.selectMention;
  }
});
Object.defineProperty(exports, "MENTION_PLACEHOLDER_ACTIVE", {
  enumerable: true,
  get: function () {
    return _activePlaceholder.MENTION_PLACEHOLDER_ACTIVE;
  }
});
Object.defineProperty(exports, "activePlaceholder", {
  enumerable: true,
  get: function () {
    return _activePlaceholder.activePlaceholder;
  }
});
exports.isActions = exports.actions = exports.MENTION_PLACEHOLDER_CANCEL = exports.MENTION_PLACEHOLDER_EXIT = exports.MENTION_PLACEHOLDER_INSERT = exports.MENTION_CANCEL_SELECT = exports.MENTION_MULTIPLE_INSERT = exports.MENTION_INSERT = exports.MENTION_DEACTIVE = void 0;

var _activeMention = require("./activeMention");

var _selectMention = require("./selectMention");

var _activePlaceholder = require("./activePlaceholder");

const MENTION_DEACTIVE = Symbol('MENTION_DEACTIVE');
exports.MENTION_DEACTIVE = MENTION_DEACTIVE;
const MENTION_INSERT = Symbol('MENTION_INSERT');
exports.MENTION_INSERT = MENTION_INSERT;
const MENTION_MULTIPLE_INSERT = Symbol('MENTION_MULTIPLE_INSERT');
exports.MENTION_MULTIPLE_INSERT = MENTION_MULTIPLE_INSERT;
const MENTION_CANCEL_SELECT = Symbol('MENTION_CANCEL_SELECT');
exports.MENTION_CANCEL_SELECT = MENTION_CANCEL_SELECT;
const MENTION_PLACEHOLDER_INSERT = Symbol('MENTION_PLACEHOLDER_INSERT');
exports.MENTION_PLACEHOLDER_INSERT = MENTION_PLACEHOLDER_INSERT;
const MENTION_PLACEHOLDER_EXIT = Symbol('MENTION_PLACEHOLDER_EXIT');
exports.MENTION_PLACEHOLDER_EXIT = MENTION_PLACEHOLDER_EXIT;
const MENTION_PLACEHOLDER_CANCEL = Symbol('MENTION_PLACEHOLDER_CANCEL');
exports.MENTION_PLACEHOLDER_CANCEL = MENTION_PLACEHOLDER_CANCEL;

function deactiveMention() {
  return {
    type: MENTION_DEACTIVE,
    payload: {}
  };
}

function insertMention(suggestion, at) {
  return {
    type: MENTION_INSERT,
    payload: {
      suggestion,
      at
    }
  };
}

function insertMentions(suggestions) {
  return {
    type: MENTION_MULTIPLE_INSERT,
    payload: {
      suggestions
    }
  };
}

function cancelMentionSelect() {
  return {
    type: MENTION_CANCEL_SELECT,
    payload: {}
  };
}

function insertPlaceholder() {
  return {
    type: MENTION_PLACEHOLDER_INSERT,
    payload: {}
  };
}

function exitPlaceholder(text, node) {
  return {
    type: MENTION_PLACEHOLDER_EXIT,
    payload: {
      text,
      node
    }
  };
}

function cancelPlaceholder(key) {
  return {
    type: MENTION_PLACEHOLDER_CANCEL,
    payload: {
      key
    }
  };
}

const actions = {
  activeMention: _activeMention.activeMention,
  deactiveMention,
  insertMention,
  selectMention: _selectMention.selectMention,
  insertMentions,
  cancelMentionSelect,
  insertPlaceholder,
  exitPlaceholder,
  cancelPlaceholder,
  activePlaceholder: _activePlaceholder.activePlaceholder
};
exports.actions = actions;
const isActions = {
  isActiveMention: _activeMention.isActiveMention,
  isSelectMention: _selectMention.isSelectMention,
  isActivePlaceholder: _activePlaceholder.isActivePlaceholder
};
exports.isActions = isActions;
