// @ali/4ever-plugin-heading-group/dist-web/utils/components/title.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import classnames from 'classnames';
import { domUtils, useZoom } from '@ali/4ever-cangjie';
var FoldBox = /*#__PURE__*/styled.div(["opacity:0;position:absolute;width:20px;height:20px;left:", "px;margin:auto;cursor:default;z-index:2;border-radius:4px;&.disabled{display:none;}&.visible{opacity:1;}&:hover{background:rgba(25,31,37,0.08);}"], function (props) {
  return props.foldBoxLeft || 0 - 20;
});
var TitleBox = /*#__PURE__*/styled.div(["position:relative;"]);
var ArrowIconBox = /*#__PURE__*/styled.span(["display:flex;justify-content:center;align-items:center;width:", "px;height:", "px;& svg{transform:rotateZ(", "deg);transition:transform 0.2s ease-in-out;}"], function (props) {
  return props.size;
}, function (props) {
  return props.size;
}, function (props) {
  return props.rotate;
});

var arrow = /*#__PURE__*/_createElement("svg", {
  width: "10",
  height: "7",
  viewBox: "0 0 10 7",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/_createElement("path", {
  d: "M4.60369 5.98479C4.80384 6.24499 5.19616 6.24499 5.39631 5.98479L9.38088 0.804856C9.63381 0.476046 9.39941 0 8.98457 0L1.01543 0C0.600595 0 0.366189 0.476046 0.619119 0.804855L4.60369 5.98479Z",
  fill: "#171A1D",
  fillOpacity: "0.4"
}));

/**
 * 分组标题组件
 * @param props
 * @returns
 */
var Title = function Title(props) {
  var style = props.style,
      className = props.className,
      fold = props.fold,
      onChange = props.onChange,
      dataName = props.dataName,
      disabled = props.disabled,
      children = props.children,
      foldBoxLeft = props.foldBoxLeft,
      role = props.role;
  var foldBtnRef = React.useRef(null);
  var titleRef = React.useRef(null);
  var zoom = useZoom();

  var _React$useState = React.useState(false),
      hovered = _React$useState[0],
      setHovered = _React$useState[1];

  var handleMouseEnterTitle = React.useCallback(function () {
    return setHovered(true);
  }, []);
  var handleMouseLeaveTitle = React.useCallback(function () {
    return setHovered(false);
  }, []);
  var visible = hovered || fold;
  /**
   * 点击标题折叠
   */

  var handleClickFold = React.useCallback(function (e) {
    e.preventDefault();
    e.stopPropagation();

    if (typeof onChange === 'function') {
      onChange(!fold);
    }
  }, [fold, onChange]);
  React.useLayoutEffect(function () {
    if (!visible) {
      return;
    }

    var title = titleRef.current;
    var foldBtn = foldBtnRef.current;

    if (!title || !foldBtn) {
      return;
    }

    var line = title.querySelector('[data-name="heading-group-title-content"]') || title;

    var _domUtils$splitRows = domUtils.splitRows(line, true),
        row = _domUtils$splitRows[0];

    var rowHeight = row.height;
    var foldBtnTop = Math.round((rowHeight / zoom - foldBtn.offsetHeight) / 2) + "px";

    if (getComputedStyle(foldBtn).top !== foldBtnTop) {
      // 与第一行保持居中对齐
      foldBtn.style.top = foldBtnTop;
    }
  });
  var rotate = fold ? -90 : 0;
  var foldClassName = classnames({
    disabled: disabled,
    visible: visible
  });
  return /*#__PURE__*/_createElement(TitleBox, {
    ref: titleRef,
    style: style,
    className: className,
    "data-name": dataName,
    onMouseEnter: handleMouseEnterTitle,
    onMouseLeave: handleMouseLeaveTitle
  }, /*#__PURE__*/_createElement(FoldBox, {
    ref: foldBtnRef,
    foldBoxLeft: foldBoxLeft,
    "data-name": "group-fold-btn",
    className: foldClassName,
    onClick: handleClickFold,
    "data-role": role
  }, /*#__PURE__*/_createElement(ArrowIconBox, {
    size: 20,
    rotate: rotate
  }, arrow)), children);
};

export default Title;
