// @ali/4ever-bi/dist-web/plugins/hr/components/getHrMenuList.js
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { CommonLine, MenuItem, SubMenu, InlineMenuItem } from '@ali/we-design';
import { HrPlugin } from '@ali/4ever-bamboo';
import { insertHr } from "../actions";
import { DEFAULT_HR_COLOR } from "../constants";
var hrLines = HrPlugin.hrLines;
export var HrMenu = function HrMenu(props) {
  var menuItem = props.menuItem,
      controller = props.controller;

  var handleClick = function handleClick(key) {
    var lineType = key.split('_')[1];
    var line = hrLines.find(function (item) {
      return item.lineType === lineType;
    });

    if (line) {
      controller.run('onAction', insertHr({
        type: line.lineType,
        color: DEFAULT_HR_COLOR,
        sz: line.size
      }));
    }
  };

  var lines = hrLines.map(function (line) {
    return /*#__PURE__*/_createElement(MenuItem, {
      testid: "hr_" + line.lineType,
      menuKey: "hr_" + line.lineType,
      onClick: handleClick,
      key: line.lineType
    }, /*#__PURE__*/_createElement(CommonLine, {
      lineType: line.lineType,
      length: 134,
      size: line.size
    }));
  });
  return /*#__PURE__*/_createElement(SubMenu, _extends({}, menuItem, {
    overlayClassName: "menu-hr-overlay",
    title: menuItem.title || '',
    maxHeight: "600"
  }), lines);
};
export var HrInlineMenu = function HrInlineMenu(props) {
  var menuItem = props.menuItem,
      controller = props.controller;

  var _ref = menuItem,
      rest = _extends({}, _ref);

  var handleClick = function handleClick() {
    var line = hrLines[0];
    controller.run('onAction', insertHr({
      type: line.lineType,
      color: DEFAULT_HR_COLOR,
      sz: line.size
    }));
  };

  return /*#__PURE__*/_createElement(InlineMenuItem, _extends({}, rest, {
    onClick: handleClick
  }));
};
