// @ali/4ever-bi/dist-web/plugins/hr/components/hr.js
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { environment, Controller } from '@ali/4ever-cangjie';
import { CommonLine } from '@ali/we-design';
import updateHr from "../commands/updateHr";
import HrControl from "./hrControl";
import { InlineToolbar } from "../../../index";
var hrLayout = [{
  toolButtons: [{
    name: 'hr.color'
  }, {
    name: 'hr.style'
  }, {
    name: 'hr.weight'
  }]
}];

var Hr = function Hr(props) {
  var isSelected = props.isSelected,
      payload = props.payload,
      attributes = props.attributes;
  var controller = Controller.useController();
  var disableInlineToolbar = controller.query('hasHoverToolbar', 'hr');
  return /*#__PURE__*/_createElement(InlineToolbar, {
    visible: !disableInlineToolbar && !environment.IS_MOBILE && isSelected,
    inlineLayout: hrLayout
  }, /*#__PURE__*/_createElement(HrControl, _extends({
    attributes: attributes,
    onHrChange: updateHr,
    isSelected: isSelected
  }, props), /*#__PURE__*/_createElement(CommonLine, payload)));
};

export default Hr;
