// @ali/4ever-bi/dist-web/plugins/hr/components/quickInsertPanel.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { CommonLine, commonLineMap, theme } from '@ali/we-design';
import { insertHr } from "../actions";
var LineWrapper = /*#__PURE__*/styled.div(["width:100%;padding:10px;&:hover{background-color:", ";}"], function () {
  return theme.colorFill4;
});
var Line = /*#__PURE__*/React.memo(function (props) {
  var lineStyle = props.lineStyle,
      onClick = props.onClick;
  var defaultSize = commonLineMap[lineStyle].defaultSize;
  var handleClick = React.useCallback(function (style, width) {
    onClick(style, width);
  }, [onClick]);
  return /*#__PURE__*/_createElement(LineWrapper, {
    onClick: function onClick() {
      return handleClick(lineStyle, defaultSize);
    }
  }, /*#__PURE__*/_createElement(CommonLine, {
    color: "black",
    size: defaultSize,
    lineType: lineStyle
  }));
});

var QuickInsertPanel = function QuickInsertPanel(props) {
  var controller = props.controller,
      close = props.close;
  var handleInsert = React.useCallback(function (lineStyle, width) {
    close();
    controller.run('onAction', insertHr({
      type: lineStyle,
      color: 'black',
      sz: width
    }));
  }, [close, controller]);
  return /*#__PURE__*/_createElement("div", null, Object.keys(commonLineMap).map(function (lineStyle) {
    return /*#__PURE__*/_createElement(Line, {
      key: lineStyle,
      onClick: handleInsert,
      lineStyle: lineStyle
    });
  }));
};

export default QuickInsertPanel;
