// @ali/4ever-utils/dist-node/zIndex/z-index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PI_COLUMN_Z_INDEX = exports.PI_HF_EDITING = exports.PI_HF_Z_INDEX = exports.PI_BODY_Z_INDEX = exports.PI_HF_BAR_Z_INDEX = exports.PI_FLOAT_TEXT_FRAME_Z_INDEX = exports.PI_FLOAT_IMAGE_Z_INDEX = exports.BI_IMAGE_RESIZE_Z_INDEX = exports.PI_PG_BORDER_Z_INDEX = exports.BI_FLOAT_TOOLBAR_Z_INDEX = void 0;

/** ***********************************************************
 *      此文件用于排列编辑器层的整体 z-index 优先级
 ************************************************************ */

/**
 * 极简模式浮动工具栏
 */
const BI_FLOAT_TOOLBAR_Z_INDEX = 999;
/**
 * 页面边框
 */

exports.BI_FLOAT_TOOLBAR_Z_INDEX = BI_FLOAT_TOOLBAR_Z_INDEX;
const PI_PG_BORDER_Z_INDEX = 2;
/**
 * 图片拖动的工具浮层
 */

exports.PI_PG_BORDER_Z_INDEX = PI_PG_BORDER_Z_INDEX;
const BI_IMAGE_RESIZE_Z_INDEX = 9;
/**
 * 图文混排浮动图片
 */

exports.BI_IMAGE_RESIZE_Z_INDEX = BI_IMAGE_RESIZE_Z_INDEX;
const PI_FLOAT_IMAGE_Z_INDEX = 8;
/**
 * TextFrame（页码/首字下沉）
 */

exports.PI_FLOAT_IMAGE_Z_INDEX = PI_FLOAT_IMAGE_Z_INDEX;
const PI_FLOAT_TEXT_FRAME_Z_INDEX = 5;
/**
 * Header/Footer 编辑栏
 */

exports.PI_FLOAT_TEXT_FRAME_Z_INDEX = PI_FLOAT_TEXT_FRAME_Z_INDEX;
const PI_HF_BAR_Z_INDEX = 2;
/**
 * 分页模式 body
 */

exports.PI_HF_BAR_Z_INDEX = PI_HF_BAR_Z_INDEX;
const PI_BODY_Z_INDEX = 2;
/**
 * 分页模式的 header/footer
 */

exports.PI_BODY_Z_INDEX = PI_BODY_Z_INDEX;
const PI_HF_Z_INDEX = 1;
/**
 * 分页模式 header/footer 聚焦状态
 */

exports.PI_HF_Z_INDEX = PI_HF_Z_INDEX;
const PI_HF_EDITING = 9;
/**
 * 分页模式的 column
 */

exports.PI_HF_EDITING = PI_HF_EDITING;
const PI_COLUMN_Z_INDEX = 1;
exports.PI_COLUMN_Z_INDEX = PI_COLUMN_Z_INDEX;
