// @ali/4ever-plugin-mention/dist-web/bi/index.js
import _extends from "@babel/runtime/helpers/extends";
import { moFactory } from "../mo/index";
import toolbar from "./toolbar";
import menubar from "./menubar";
import renderMention from "./renderNode";
import renderMark from "./renderMark";
import onCangjieInput from "./handlers/onCangjieInput";
import onKeyDown from "./handlers/onKeyDown";
import onSelect from "./handlers/onSelect";
import createOnCangjiePaste from "./handlers/onCangjiePaste";
import createOnAction from "./handlers/createOnAction";
import { queries } from "./queries";
import { actions, isActions } from "./actions";
import createRenderToolbar from "./createRenderToolbar";
var DEFAULT_THEME = {
  suggestionWidth: '200px',
  suggestionMaxHeight: '250px'
};
export var biFactory = function biFactory(configs) {
  var _configs$mention;

  var mentionConfig = _extends({
    theme: _extends({}, DEFAULT_THEME, (_configs$mention = configs.mention) == null ? void 0 : _configs$mention.popupOptions)
  }, configs.mention);

  var mentionId = mentionConfig.getMentionId ? mentionConfig.getMentionId() : '';
  return _extends({}, moFactory({}), {
    toolbar: toolbar(mentionConfig),
    renderToolbar: createRenderToolbar(mentionConfig.locale),
    renderNode: renderMention(mentionId, mentionConfig),
    renderMark: renderMark(mentionConfig),
    onKeyDown: onKeyDown,
    menubar: menubar,
    onCangjieInput: onCangjieInput,
    onCangjiePaste: createOnCangjiePaste(configs),
    onSelect: onSelect,
    onAction: createOnAction(mentionConfig),
    locale: mentionConfig.locale || {},
    queries: queries,
    actions: actions,
    isActions: isActions
  });
};
