// @ali/4ever-bi/dist-web/plugins/hr/commands/insertHr.js
import { Commands } from '@ali/4ever-cangjie';
import { generateKey } from '@ali/4ever-utils';
import updateHr from "./updateHr";
import { DEFAULT_HR_COLOR } from "../constants";
import { shouldDeleteParagraph } from "../utils";
export default (function (controller, payload) {
  if (payload === void 0) {
    payload = {};
  }

  var value = controller.value;
  var focusBlock = value.focusBlock;

  if ((focusBlock == null ? void 0 : focusBlock.type) === 'hr') {
    return updateHr(controller, focusBlock, payload);
  }

  var _payload = payload,
      color = _payload.color,
      type = _payload.type,
      sz = _payload.sz;
  var hrColor = color != null ? color : DEFAULT_HR_COLOR;
  controller.command(Commands.insertBlock, {
    type: 'hr',
    data: {
      color: hrColor,
      type: type,
      width: '100%',
      // 默认占满一行
      sz: sz || 2,
      // size 默认为 2
      id: generateKey()
    }
  }).command(Commands.insertBlock, 'paragraph'); // 插入前是空行则删除

  if (shouldDeleteParagraph(focusBlock)) {
    controller.command(Commands.removeNodeByKey, focusBlock.key);
  }

  return controller;
});
