// @ali/4ever-component/dist-node/MobileToolbar/layout/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mobileLayout = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _font = _interopRequireDefault(require("./font"));

var _insert = _interopRequireDefault(require("./insert"));

var _align = _interopRequireDefault(require("./align"));

const _createElement = /*#__PURE__*/React.createElement;

function getButton(focusIcon, icon) {
  return ({
    Wrapper,
    panelShowing,
    ...rest
  }) => /*#__PURE__*/_createElement(Wrapper, (0, _extends2.default)({
    type: "big"
  }, rest, {
    isActive: panelShowing,
    icon: panelShowing ? focusIcon : icon
  }));
}

const mobileLayout = [{
  key: 'group1',
  items: [{
    name: 'insert',
    Component: getButton('plus', 'plus'),
    panel: {
      title: 'Insert',
      content: _insert.default
    }
  }, {
    name: 'font',
    Component: getButton('font', 'font'),
    panel: {
      title: 'Fonts',
      content: _font.default
    }
  }, {
    name: 'align',
    Component: getButton('alignLeft', 'alignLeft'),
    panel: {
      title: 'Align',
      content: _align.default
    }
  }]
}, {
  key: 'group2',
  items: ['image.mobileBig', 'list.mobileTask', 'mention.mobile', 'history.undoMobile', 'history.redoMobile']
}, {
  key: 'fixed',
  fixedOnRight: true,
  items: ['keyboard']
}];
exports.mobileLayout = mobileLayout;
