// @ali/4ever-bi/dist-node/plugins/fonts/components/next/fontsButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = FontsButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everCangjie = require("@ali/4ever-cangjie");

var _everBamboo = require("@ali/4ever-bamboo");

var _everComponent = require("@ali/4ever-component");

var _actions = require("../../actions");

var _getActiveFontsMark = require("../../utils/getActiveFontsMark");

var _getFontsTip = _interopRequireDefault(require("../../utils/getFontsTip"));

const _createElement = /*#__PURE__*/React.createElement;
;

function FontsButton(props) {
  const {
    controller,
    locale,
    width = 133,
    ...rest
  } = props;
  const docStyle = (0, _everCangjie.useStyle)();
  const fonts = controller.query('getFonts', docStyle?.docDefaults?.data?.rPr?.fonts);
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInLinkCard');
  const handleChange = React.useCallback(fontsValue => controller.run('onAction', (0, _actions.setFonts)(fontsValue)), [controller]);
  let title = locale?.buttonTitle || '';

  if (fonts && fonts !== _getActiveFontsMark.defaultFonts) {
    // 提示显示当前字体
    title += ` ${(0, _getFontsTip.default)(fonts, locale)}`;
  }

  return /*#__PURE__*/_createElement(_everComponent.FontFamilySelect, (0, _extends2.default)({
    testid: "bi-toolbar-fonts",
    disabled: !!isDisabled,
    value: fonts || locale?.buttonTitle,
    onChange: handleChange,
    role: _everBamboo.PluginRoles.fontFamilySelect,
    width: width,
    tooltip: title || 'Font Family'
  }, rest));
}
