// @ali/4ever-bi/dist-node/plugins/clearStyle/components/next/createClearStyleButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createClearStyleButton;
exports.CLEAR_STYLE_HOT_KEY = void 0;

var React = _interopRequireWildcard(require("react"));

var _everCangjie = require("@ali/4ever-cangjie");

var _everComponent = require("@ali/4ever-component");

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../../actions");

const _createElement = /*#__PURE__*/React.createElement;
const MOD = _everCangjie.environment.IS_MAC ? '⌘' : 'Ctrl';
const CLEAR_STYLE_HOT_KEY = `${MOD}+\\`;
exports.CLEAR_STYLE_HOT_KEY = CLEAR_STYLE_HOT_KEY;

var _ref = /*#__PURE__*/_createElement(_everComponent.IconEraser, null);

function createClearStyleButton(props) {
  const {
    controller,
    locale,
    icon,
    title,
    tooltip: tooltipProp,
    badgeDot,
    disabled
  } = props;

  const handleClick = () => {
    controller.run('onAction', (0, _actions.clearStyles)());
  };

  const tooltip = React.useMemo(() => ({
    title: title || locale?.buttonTitle,
    shortCut: CLEAR_STYLE_HOT_KEY,
    ...tooltipProp
  }), [title, tooltipProp]);
  const buttonConfig = {
    icon: icon || _ref,
    onClick: handleClick,
    title: title || locale?.buttonTitle,
    tooltip: tooltip,
    badgeDot,
    disabled,
    testid: 'bi-toolbar-clearstyle',
    role: _everBamboo.PluginRoles.clear
  };
  return buttonConfig;
}
