// @ali/we-design-next/dist-node/display/tooltip/Tooltip.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tooltipOverlayRestStyle = exports.TooltipContent = exports.Tooltip = void 0;
var _react = _interopRequireDefault(require("react"));
var _Popover = require("../popover/Popover");
var _styled = require("./styled");
const _createElement = /*#__PURE__*/_react.default.createElement;
const tooltipOverlayRestStyle = {
  border: 'unset',
  boxShadow: 'unset',
  backgroundColor: 'transparent'
};
exports.tooltipOverlayRestStyle = tooltipOverlayRestStyle;
const TooltipContent = /*#__PURE__*/_react.default.memo(({
  title,
  description
}) => {
  return /*#__PURE__*/_createElement(_styled.TooltipWrapper, null, title && /*#__PURE__*/_createElement(_styled.TooltipTitle, null, title), description && /*#__PURE__*/_createElement(_styled.TooltipDescription, null, description));
});
exports.TooltipContent = TooltipContent;
const Tooltip = /*#__PURE__*/_react.default.memo(props => {
  const {
    children,
    title,
    description,
    placement,
    getContainer
  } = props;
  return /*#__PURE__*/_createElement(_Popover.Popover, {
    content: /*#__PURE__*/_createElement(TooltipContent, {
      title: title,
      description: description
    }),
    style: tooltipOverlayRestStyle,
    placement: placement,
    placementOffset: [4, 0],
    trigger: "hover",
    getContainer: getContainer,
    animation: true
  }, children);
});
exports.Tooltip = Tooltip;
