// @ali/4ever-bi/dist-web/plugins/clearStyle/components/clearStyleButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { environment } from '@ali/4ever-cangjie';
import { EraserButton } from '@ali/we-toolbar';
import { PluginRoles } from '@ali/4ever-bamboo';
import { clearStyles } from "../actions";
var MOD = environment.IS_MAC ? '⌘' : 'Ctrl';
export var CLEAR_STYLE_HOT_KEY = MOD + "+\\";
export default function ClearStyleButton(props) {
  var controller = props.controller,
      locale = props.locale,
      rest = _objectWithoutPropertiesLoose(props, ["controller", "locale"]);

  var handleClick = React.useCallback(function () {
    controller.run('onAction', clearStyles());
  }, [controller]);
  return /*#__PURE__*/_createElement(EraserButton, _extends({}, rest, {
    onClick: handleClick,
    shortcutText: CLEAR_STYLE_HOT_KEY,
    testid: 'bi-toolbar-clearstyle',
    role: PluginRoles.clear
  }));
}
