// @ali/4ever-bi/dist-web/plugins/clearStyle/components/clearStyleButtonMobile.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { clearStyles } from "../actions";
export function MobileClearButton(props) {
  var controller = props.controller,
      Wrapper = props.Wrapper,
      item = props.item,
      locale = props.locale;
  var handleClick = React.useCallback(function () {
    controller.run('onAction', clearStyles());
  }, [controller]);
  return /*#__PURE__*/_createElement(Wrapper, {
    type: 'horiz',
    item: item,
    width: 2.1,
    onClick: handleClick
  }, /*#__PURE__*/_createElement("div", null, locale.buttonTitle));
}
export function MobileClearButtonBig(props) {
  var controller = props.controller,
      Wrapper = props.Wrapper,
      item = props.item;
  var handleClick = React.useCallback(function () {
    controller.run('onAction', clearStyles());
  }, [controller]);
  return /*#__PURE__*/_createElement(Wrapper, {
    type: 'big',
    icon: "clearStyle",
    item: item,
    onClick: handleClick
  });
}
