// @ali/we-design-next/dist-node/form/input-number/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NumberControlWrap = exports.ArrowUpIcon = exports.ArrowDownIcon = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _common = require("../../common");
var _colors = require("../../common/consts/colors");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const _createElement = /*#__PURE__*/_react.default.createElement;
const IconSvg = /*#__PURE__*/_styledComponents.default.svg(["cursor:auto;stroke-opacity:0.6;&:active{stroke-opacity:0.24;}flex-grow:1;", " ", " &:hover{background-color:rgba(126,134,142,0.12);}&:active{background-color:", ";}"], ({
  disabled
}) => disabled ? `color: ${_colors.COLORS.LIGHT.LEVEL4_BASE_COLOR};` : null, _common.CSS_CURSOR_DISABLE, _colors.COLORS.COMMON.OVERLAY_COLOR);
const ArrowUpIcon = /*#__PURE__*/(0, _react.memo)(props => {
  const {
    disabled,
    ...rest
  } = props;
  return /*#__PURE__*/_createElement(IconSvg, (0, _extends2.default)({
    width: "12",
    height: "12",
    viewBox: "0 0 12 12",
    fill: "none",
    disabled: disabled
  }, rest), /*#__PURE__*/_createElement("path", {
    d: "M9 7L6.04273 4L3 7",
    stroke: _colors.COLORS.LIGHT.LEVEL1_BASE_COLOR
  }));
});
exports.ArrowUpIcon = ArrowUpIcon;
ArrowUpIcon.displayName = 'InputNumberArrowUpIcon';
const ArrowDownIcon = /*#__PURE__*/(0, _react.memo)(props => {
  const {
    disabled,
    ...rest
  } = props;
  return /*#__PURE__*/_createElement(IconSvg, (0, _extends2.default)({
    width: "12",
    height: "12",
    viewBox: "0 0 12 12",
    fill: "none",
    disabled: disabled
  }, rest), /*#__PURE__*/_createElement("path", {
    d: "M3 5L5.95727 8L9 5",
    stroke: _colors.COLORS.LIGHT.LEVEL1_BASE_COLOR
  }));
});
exports.ArrowDownIcon = ArrowDownIcon;
ArrowDownIcon.displayName = 'InputNumberArrowDownIcon';
const NumberControlWrap = /*#__PURE__*/_styledComponents.default.div(["display:flex;flex-direction:column;background-color:#f2f2f6;border-left:1px solid rgb(205,205,205);height:100%;"]);
exports.NumberControlWrap = NumberControlWrap;
