// @ali/4ever-plugin-mention/dist-web/bi/actions/index.js
import { MENTION_ACTIVE, activeMention, isActiveMention } from "./activeMention";
import { MENTION_SELECT, selectMention, isSelectMention } from "./selectMention";
import { MENTION_PLACEHOLDER_ACTIVE, activePlaceholder, isActivePlaceholder } from "./activePlaceholder";
export { MENTION_ACTIVE, activeMention, MENTION_SELECT, selectMention, MENTION_PLACEHOLDER_ACTIVE, activePlaceholder };
export var MENTION_DEACTIVE = Symbol('MENTION_DEACTIVE');
export var MENTION_INSERT = Symbol('MENTION_INSERT');
export var MENTION_MULTIPLE_INSERT = Symbol('MENTION_MULTIPLE_INSERT');
export var MENTION_CANCEL_SELECT = Symbol('MENTION_CANCEL_SELECT');
export var MENTION_PLACEHOLDER_INSERT = Symbol('MENTION_PLACEHOLDER_INSERT');
export var MENTION_PLACEHOLDER_EXIT = Symbol('MENTION_PLACEHOLDER_EXIT');
export var MENTION_PLACEHOLDER_CANCEL = Symbol('MENTION_PLACEHOLDER_CANCEL');
export function deactiveMention() {
  return {
    type: MENTION_DEACTIVE,
    payload: {}
  };
}
export function insertMention(suggestion, at) {
  return {
    type: MENTION_INSERT,
    payload: {
      suggestion: suggestion,
      at: at
    }
  };
}
export function insertMentions(suggestions) {
  return {
    type: MENTION_MULTIPLE_INSERT,
    payload: {
      suggestions: suggestions
    }
  };
}
export function cancelMentionSelect() {
  return {
    type: MENTION_CANCEL_SELECT,
    payload: {}
  };
}
export function insertPlaceholder() {
  return {
    type: MENTION_PLACEHOLDER_INSERT,
    payload: {}
  };
}
export function exitPlaceholder(text, node) {
  return {
    type: MENTION_PLACEHOLDER_EXIT,
    payload: {
      text: text,
      node: node
    }
  };
}
export function cancelPlaceholder(key) {
  return {
    type: MENTION_PLACEHOLDER_CANCEL,
    payload: {
      key: key
    }
  };
}
export var actions = {
  activeMention: activeMention,
  deactiveMention: deactiveMention,
  insertMention: insertMention,
  selectMention: selectMention,
  insertMentions: insertMentions,
  cancelMentionSelect: cancelMentionSelect,
  insertPlaceholder: insertPlaceholder,
  exitPlaceholder: exitPlaceholder,
  cancelPlaceholder: cancelPlaceholder,
  activePlaceholder: activePlaceholder
};
export var isActions = {
  isActiveMention: isActiveMention,
  isSelectMention: isSelectMention,
  isActivePlaceholder: isActivePlaceholder
};
