// @ali/4ever-plugin-subject/dist-web/utils/utils.js
import _extends from "@babel/runtime/helpers/extends";
import { Paragraph } from '@ali/4ever-plugin-paragraph';
import { Text, Block } from '@ali/4ever-cangjie';

/**
 * 判断节点是否是议题标题
 * @param node
 * @returns
 */
export function isSubjectHeading(node) {
  return Block.isBlock(node) && node.type === 'heading-1' && !!node.data.subjectPr;
}
/**
 * 判断节点是否是议题组
 */

export function isSubject(node) {
  return Block.isBlock(node) && node.type === 'subject-group';
}
export function createSubject(props) {
  return Block.create(_extends({}, props, {
    type: 'subject-group'
  }));
}
/**
 * 判断节点是否是空节点
 * @param node
 * @returns
 */

export function isEmptyNode(node) {
  return node.nodes.length === 0 || node.nodes.length === 1 && Paragraph.isParagraph(node) && Text.isEmptyText(node.nodes[0]);
}
export function isNormalParagraph(node) {
  return Paragraph.isParagraph(node) && !node.data.list;
}
/**
 * 简单生成 uniqueId
 * @see https://stackoverflow.com/questions/105034/how-to-create-a-guid-uuid#answer-44078785
 * @returns
 */

export function uniqueId() {
  return Date.now().toString(36) + Math.random().toString(36).substring(2);
}
