// @ali/4ever-plugin-mention/dist-node/zhi/mention.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Mention;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _everPluginMarks = require("@ali/4ever-plugin-marks");

var _placeholder = require("../components/placeholder");

var _mention = _interopRequireDefault(require("../components/mention"));

var _styled = require("../components/styled");

const _createElement = /*#__PURE__*/_react.default.createElement;

function Mention(props) {
  const {
    node,
    onMentionClick,
    onMentionHover,
    getCurrentUser,
    locale,
    ...rest
  } = props;
  const {
    name,
    login,
    id,
    isPlaceholder
  } = node.data;
  const text = node.getFirstText();
  const leaf = text && text.leaves[0];
  const marks = leaf ? leaf.marks : [];
  const style = (0, _everPluginMarks.marksToStyle)(marks);
  const isSelf = getCurrentUser ? getCurrentUser() === login : false;
  const handleMentionClick = (0, _react.useCallback)(event => {
    event.preventDefault();
    event.stopPropagation();

    if (onMentionClick) {
      // @ts-ignore
      onMentionClick(node.data);
    }
  }, [node.data, onMentionClick]);
  const handleMentionHover = (0, _react.useCallback)(event => {
    if (onMentionHover) {
      // @ts-ignore
      onMentionHover(node.data);
    }
  }, [node.data, onMentionHover]);

  if (isPlaceholder) {
    return /*#__PURE__*/_createElement(_placeholder.BtnWrapper, {
      isReader: true
    }, "@ ", locale.mentionPlaceholder);
  }

  return /*#__PURE__*/_createElement(_mention.default, (0, _extends2.default)({}, rest, {
    // @ts-ignore
    id: id,
    "data-type": node.type,
    style: style,
    onClick: handleMentionClick,
    onMouseEnter: handleMentionHover,
    isSelf: isSelf,
    node: node
  }), "@", /*#__PURE__*/_createElement(_styled.MentionNameWrapper, null, name));
}
