// @ali/4ever-bi/dist-node/actionTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _actions = require("./plugins/alignment/actions");

var _actions2 = require("./plugins/blockquote/actions");

var _actions3 = require("./plugins/clearStyle/actions");

var _everPluginCode = require("@ali/4ever-plugin-code");

var _everPluginColor = require("@ali/4ever-plugin-color");

var _everPluginClipboard = require("@ali/4ever-plugin-clipboard");

var _everPluginMarks = require("@ali/4ever-plugin-marks");

var _actions4 = require("./plugins/fonts/actions");

var _actions5 = require("./plugins/formatPainter/actions");

var _everPluginHeading = require("@ali/4ever-plugin-heading");

var _everPluginHighlight = require("@ali/4ever-plugin-highlight");

var _everPluginHistory = require("@ali/4ever-plugin-history");

var _actions6 = require("./plugins/hr/actions");

var _actions7 = require("./plugins/image/actions");

var _actions8 = require("./plugins/indent/actions");

var _actions9 = require("./plugins/lineheight/actions");

var _actions10 = require("./plugins/link/actions");

var _everFactory = require("@ali/4ever-factory");

var _actions11 = require("./plugins/newlineGuide/actions");

var _everPluginList = require("@ali/4ever-plugin-list");

var _everPluginMention = require("@ali/4ever-plugin-mention");

var _actions12 = require("./plugins/onlineVideo/actions");

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _actions13 = require("./plugins/placeholder/actions");

var _actions14 = require("./plugins/quickInsert/actions");

var _actions15 = require("./plugins/sz/actions");

var _everPluginTable = require("@ali/4ever-plugin-table");

var _actions16 = require("./plugins/formula/actions");

var _everPluginDraggable = require("@ali/4ever-plugin-draggable");

var _actions17 = require("./plugins/hoverBlock/actions");

var _actions18 = require("./plugins/translate/actions");

var _actions19 = require("./plugins/underline/actions");

var _actions20 = require("./plugins/vertAlign/actions");

var _actions21 = require("./plugins/refer/actions");

var _actions22 = require("./plugins/video/actions");

var _actions23 = require("./plugins/eventTracking/actions");

var _everComponent = require("@ali/4ever-component");

const {
  LIST_REMOVE_LIST,
  LIST_SET_CHECK_STATUS,
  LIST_TOGGLE_LIST,
  LIST_SELECT_SYMBOL,
  CLEAR_RANGE_SELECTION,
  SELECT_LIST_SELECTIONS,
  CLEAR_LIST_SELECTIONS
} = _everPluginList.OldActions;
const {
  MENTION_ACTIVE,
  MENTION_DEACTIVE,
  MENTION_INSERT,
  MENTION_SELECT,
  MENTION_MULTIPLE_INSERT,
  MENTION_CANCEL_SELECT,
  MENTION_PLACEHOLDER_INSERT,
  MENTION_PLACEHOLDER_ACTIVE
} = _everPluginMention.OldBiActions;
const {
  CLEAR_DROPHOLDER,
  UPDATE_DRAGHANDER,
  UPDATE_DRAGHIGHLIGHT,
  REMOVE_DRAGHIGHLIGHT_SINGLE,
  UPDATE_DARGFRAGMENT,
  DRAG_MENU_INSERT_AFTER,
  DRAG_MENU_INSERT_BEFORE,
  DRAG_MENU_CUT_OR_COPY,
  DRAG_MENU_DELETE,
  MOUNT_LINE_HEADER_TOOLBAR
} = _everPluginDraggable.oldActions;
var _default = {
  UPDATE_DRAGHANDER,
  UPDATE_DRAGHIGHLIGHT,
  UPDATE_DARGFRAGMENT,
  REMOVE_DRAGHIGHLIGHT_SINGLE,
  UPDATE_HOVER_HIGHLIGHT: _actions17.UPDATE_HOVER_HIGHLIGHT,
  REMOVE_HOVER_HIGHLIGHT: _actions17.REMOVE_HOVER_HIGHLIGHT,
  NEWLINE_SET_DATA: _actions11.NEWLINE_SET_DATA,
  FORMULA_SET_INJECTION: _actions16.FORMULA_SET_INJECTION,
  TAG_INSERT: _everFactory.TAG_INSERT,
  CARD_SET_INJECTION: _everFactory.CARD_SET_INJECTION,
  CLEAR_DROPHOLDER,
  CLIPBOARD_SET_DECORATIONS: _everPluginClipboard.CLIPBOARD_SET_DECORATIONS,
  CLIPBOARD_HIDE_PASTEPANEL: _everPluginClipboard.CLIPBOARD_HIDE_PASTEPANEL,
  ALIGNMENT_SET: _actions.ALIGNMENT_SET,
  ALBUM_SELECT: _actions22.ALBUM_SELECT,
  BLOCKQUOTE_TOGGLE: _actions2.BLOCKQUOTE_TOGGLE,
  BLOCKQUOTE_ENABLE: _actions2.BLOCKQUOTE_ENABLE,
  BLOCKQUOTE_DISABLE: _actions2.BLOCKQUOTE_DISABLE,
  CLEAR_STYLES: _actions3.CLEAR_STYLES,
  CODE_BLUR: _everPluginCode.CODE_BLUR,
  CODE_CHANGE: _everPluginCode.CODE_CHANGE,
  CODE_FOCUS_TO_CODE_EDITOR: _everPluginCode.CODE_FOCUS_TO_CODE_EDITOR,
  CODE_FOCUS_TO_TEXT_EDITOR: _everPluginCode.CODE_FOCUS_TO_TEXT_EDITOR,
  CODE_REMOVE: _everPluginCode.CODE_REMOVE,
  CODE_INSERT: _everPluginCode.CODE_INSERT,
  CODE_CHANGE_DATA: _everPluginCode.CODE_CHANGE_DATA,
  COLOR_SET: _everPluginColor.COLOR_SET,
  MARK_TOGGLE: _everPluginMarks.MARK_TOGGLE,
  FILE_SELECT: _everFactory.FILE_SELECT,
  FILE_REMOVE: _everFactory.FILE_REMOVE,
  FILE_UPDATE: _everFactory.FILE_UPDATE,
  IMPORT_MEMO: _everFactory.IMPORT_MEMO,
  FONTS_SET: _actions4.FONTS_SET,
  FORMAT_PAINTER_COPY: _actions5.FORMAT_PAINTER_COPY,
  FORMAT_PAINTER_CLEAR_CLIPBOARD: _actions5.FORMAT_PAINTER_CLEAR_CLIPBOARD,
  HEADING_SET: _everPluginHeading.HEADING_SET,
  HIGHLIGHT_SET: _everPluginHighlight.HIGHLIGHT_SET,
  HISTORY_REDO: _everPluginHistory.HISTORY_REDO,
  HISTORY_UNDO: _everPluginHistory.HISTORY_UNDO,
  HR_INSERT: _actions6.HR_INSERT,
  IMAGE_SELECT: _actions7.IMAGE_SELECT,
  IMAGE_UPDATE: _actions7.IMAGE_UPDATE,
  IMAGE_REMOVE: _actions7.IMAGE_REMOVE,
  INDENT_INCREASE: _actions8.INDENT_INCREASE,
  INDENT_DECREASE: _actions8.INDENT_DECREASE,
  LINEHEIGHT_SET: _actions9.LINEHEIGHT_SET,
  LINK_FOCUS_EDITOR: _actions10.LINK_FOCUS_EDITOR,
  LINK_HIDE_EDITOR: _actions10.LINK_HIDE_EDITOR,
  LINK_REMOVE: _actions10.LINK_REMOVE,
  LINK_REMOVE_PURE: _actions10.LINK_REMOVE_PURE,
  LINK_SWITCH_DISPLAYTYPE: _actions10.LINK_SWITCH_DISPLAYTYPE,
  LINK_SWITCH_LINK_TO_CARD: _actions10.LINK_SWITCH_LINK_TO_CARD,
  LINK_EDIT_LINK_CARD: _actions10.LINK_EDIT_LINK_CARD,
  LINK_HOVER_LINK_CARD: _actions10.LINK_HOVER_LINK_CARD,
  LINK_CLICK_PLACEHOLDER: _actions10.LINK_CLICK_PLACEHOLDER,
  LINK_UNWRAP: _actions10.LINK_UNWRAP,
  LINK_UPDATE: _actions10.LINK_UPDATE,
  LINK_UPDATE_PURE: _actions10.LINK_UPDATE_PURE,
  LINK_WRAP: _actions10.LINK_WRAP,
  LINK_UNWRAP_CARD: _actions10.LINK_UNWRAP_CARD,
  LINK_UNWRAP_PURE_LINK: _actions10.LINK_UNWRAP_PURE_LINK,
  LINK_INJECT_CARDINFO: _actions10.LINK_INJECT_CARDINFO,
  LINK_REMOVE_INJECTED_CARDINFO: _actions10.LINK_REMOVE_INJECTED_CARDINFO,
  LINK_COPY_LINK: _actions10.LINK_COPY_LINK,
  LIST_REMOVE_LIST,
  LIST_SET_CHECK_STATUS,
  LIST_TOGGLE_LIST,
  LIST_SELECT_SYMBOL,
  CLEAR_RANGE_SELECTION,
  SELECT_LIST_SELECTIONS,
  CLEAR_LIST_SELECTIONS,
  MENTION_ACTIVE,
  MENTION_DEACTIVE,
  MENTION_INSERT,
  MENTION_SELECT,
  MENTION_MULTIPLE_INSERT,
  MENTION_CANCEL_SELECT,
  MENTION_PLACEHOLDER_INSERT,
  MENTION_PLACEHOLDER_ACTIVE,
  ONLINEVIDEO_INSERT: _actions12.ONLINEVIDEO_INSERT,
  ONLINEVIDEO_UPDATE: _actions12.ONLINEVIDEO_UPDATE,
  ONLINEVIDEO_REMOVE: _actions12.ONLINEVIDEO_REMOVE,
  ONLINEVIDEO_FOCUS_EDITOR: _actions12.ONLINEVIDEO_FOCUS_EDITOR,
  ONLINEVIDEO_SELECT: _actions12.ONLINEVIDEO_SELECT,
  PARAGRAPH_SET_SPACING: _everPluginParagraph.PARAGRAPH_SET_SPACING,
  PLACEHOLDER_REMOVE: _actions13.PLACEHOLDER_REMOVE,
  QUICK_INSERT_INSERT: _actions14.QUICK_INSERT_INSERT,
  QUICK_INSERT_REMOVE: _actions14.QUICK_INSERT_REMOVE,
  QUICK_INSERT_RESET: _actions14.QUICK_INSERT_RESET,
  QUICK_INSERT_SET: _actions14.QUICK_INSERT_SET,
  QUICK_INSERT_SHOW: _actions14.QUICK_INSERT_SHOW,
  SZ_SET: _actions15.SZ_SET,
  IMAGE_SET_INJECTION: _actions7.IMAGE_SET_INJECTION,
  ..._everPluginTable.biActionTypes,
  TRANSLATE_INSERT: _actions18.TRANSLATE_INSERT,
  TRANSLATE_REPLACE: _actions18.TRANSLATE_REPLACE,
  UNDERLINE_TOGGLE: _actions19.UNDERLINE_TOGGLE,
  VERT_ALIGN_TOGGLE_SUB: _actions20.VERT_ALIGN_TOGGLE_SUB,
  VERT_ALIGN_TOGGLE_SUP: _actions20.VERT_ALIGN_TOGGLE_SUP,
  VIDEO_SELECT: _actions22.VIDEO_SELECT,
  DINGPAN_SELECT: _actions21.DINGPAN_SELECT,
  DINGPAN_UPDATE: _actions21.DINGPAN_UPDATE,
  DINGPAN_REMOVE: _actions21.DINGPAN_REMOVE,
  EVENT_TRACKING: _actions23.EVENT_TRACKING,
  DRAG_MENU_INSERT_AFTER,
  DRAG_MENU_INSERT_BEFORE,
  DRAG_MENU_CUT_OR_COPY,
  DRAG_MENU_DELETE,
  MOUNT_PLUS_TOOLBAR: _actions11.MOUNT_PLUS_TOOLBAR,
  MOUNT_SLASH_TOOLBAR: _actions11.MOUNT_SLASH_TOOLBAR,
  MOUNT_LINE_HEADER_TOOLBAR,
  MOUNT_SELECTION_TOOLBAR: _everComponent.MOUNT_SELECTION_TOOLBAR
};
exports.default = _default;
