// @ali/4ever-bi/dist-node/createCustomizedMobileBiPlugins.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createCustomizedBiPlugins;

var _everCangjie = require("@ali/4ever-cangjie");

var _everPluginException = require("@ali/4ever-plugin-exception");

var _everUtils = require("@ali/4ever-utils");

var _autoClose = _interopRequireDefault(require("./plugins/autoClose"));

var _everFactory = require("@ali/4ever-factory");

var _everPluginBasic = require("@ali/4ever-plugin-basic");

var _mobile = _interopRequireDefault(require("./plugins/clipboard/mobile"));

var _mobile2 = _interopRequireDefault(require("./plugins/history/mobile"));

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _everPluginContainer = require("@ali/4ever-plugin-container");

/**
 * @packageDocumentation
 * @module Editor(Bi)
 */
function createCustomizedBiPlugins(plugins) {
  const paragraphPlugin = plugins.paragraph || _everPluginParagraph.mobileBiFactory;
  const {
    unknownCard = _everFactory.createUnknownCardMobilePlugin,
    unknownTag = _everFactory.createUnknownTagMobilePlugin,
    ...rest
  } = plugins;
  const allPlugins = {
    exception: _everPluginException.mobileBiFactory,
    ...rest,
    container: _everPluginContainer.biFactory,
    unknownCard,
    unknownTag,
    autoClose: _autoClose.default,
    clipboard: _mobile.default,
    history: _mobile2.default,
    paragraph: paragraphPlugin,
    basic: _everPluginBasic.biFactory
  };
  return function createBiPlugins(configs = {}) {
    const array = [];
    const hash = {}; // @ts-ignore

    (0, _everUtils.sortPlugins)(allPlugins).forEach(({
      key,
      factory
    }) => {
      const plugin = factory(configs, hash);
      array.push(plugin);
      hash[key] = plugin;
    });
    const schema = (0, _everCangjie.createSchema)(array);
    array.push({
      schema
    });
    return {
      array,
      hash,
      lang: configs.lang || 'zh-CN'
    };
  };
}
