// @ali/4ever-bi/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Provider: true,
  Content: true,
  createBiPlugins: true,
  createCustomizedBiPlugins: true,
  actionTypes: true,
  plugins: true,
  locale_en_US: true,
  locale_ja_JP: true,
  locale_zh_CN: true,
  locale_zh_HK: true,
  locale_zh_TW: true,
  Resizable: true,
  MenuTitleWithNewBadge: true,
  Toolbar: true,
  InlineToolbar: true,
  ToolbarConfig: true,
  ToolBoxLayoutMode: true,
  HeaderToolbarMode: true,
  ToolbarMode: true,
  Sidebar: true,
  NewToolbar: true,
  NewToolbarConfig: true,
  SimpleToolbar: true,
  defaultSingleLayout: true,
  defaultDisableTargets: true,
  defaultDoubleLayout: true,
  HeadingToolbar: true,
  MobileToolbar: true,
  mobileLayout: true,
  MobileToolbarConstant: true,
  LeftToolbar: true,
  SelectionBar: true,
  BiMenu: true,
  insertTag: true,
  setTagData: true,
  removeTag: true,
  tagFactory: true,
  removeCard: true,
  setCardData: true,
  setCardInjection: true,
  insertCard: true,
  cardFactory: true,
  mobileCardFactory: true,
  groupFactory: true,
  insertTemplateButton: true,
  TemplateButton: true,
  createMobileBiPlugins: true,
  mobilePlugins: true,
  AlignmentPlugin: true,
  MobileAlignmentPlugin: true,
  AudioPlugin: true,
  MobileAudioPlugin: true,
  BasicPlugin: true,
  BasicIsActions: true,
  BlockquotePlugin: true,
  MobileBlockquotePlugin: true,
  BoldPlugin: true,
  MobileBoldPlugin: true,
  ClearStylePlugin: true,
  MobileClearStylePlugin: true,
  ClipBoardPlugin: true,
  MobileClipBoardPlugin: true,
  CodePlugin: true,
  MobileCodePlugin: true,
  InlineCodePlugin: true,
  ColumnsPlugin: true,
  CommentPlugin: true,
  MobileCommentPlugin: true,
  CommonPlugin: true,
  CalendarCardPlugin: true,
  MobileCalendarCardPlugin: true,
  VoteCardPlugin: true,
  MobileVoteCardPlugin: true,
  FindAndReplacePlugin: true,
  ImTagPlugin: true,
  ColorPlugin: true,
  ContextMenuPlugin: true,
  EmbedPlugin: true,
  MobileEmbedPlugin: true,
  FontsPlugin: true,
  FormatPainterPlugin: true,
  HeadingPlugin: true,
  HighlightPlugin: true,
  HrPlugin: true,
  HistoryPlugin: true,
  MobileHistoryPlugin: true,
  ImagePlugin: true,
  MobileImagePlugin: true,
  IndentPlugin: true,
  ItalicPlugin: true,
  LineHeightPlugin: true,
  LinkPlugin: true,
  LinkCardPlugin: true,
  MobileLinkPlugin: true,
  MobileLinkCardPlugin: true,
  ListPlugin: true,
  MarksPlugin: true,
  MentionPlugin: true,
  MobileMentionPlugin: true,
  OnlineVideoPlugin: true,
  MobileOnlineVideoPlugin: true,
  ParagraphPlugin: true,
  PlaceholderPlugin: true,
  QuickInsertPlugin: true,
  NewlineGuidePlugin: true,
  ReferPlugin: true,
  StrikePlugin: true,
  SzPlugin: true,
  TablePlugin: true,
  MobileTablePlugin: true,
  TableActions: true,
  TranslatePlugin: true,
  UnderlinePlugin: true,
  VertAlignPlugin: true,
  VideoPlugin: true,
  DraggablePlugin: true,
  FormulaPlugin: true,
  StickerPlugin: true,
  MobileVideoPlugin: true,
  EventTrackingPlugin: true,
  SubjectPlugin: true,
  HeadingGroupPlugin: true,
  HeadingGroupNewPlugin: true,
  ListGroupPlugin: true,
  CalloutPlugin: true,
  QuickPreviewPlugin: true,
  LaserPointerPlugin: true,
  LeftToolbarPlugin: true,
  LaserPointerRole: true,
  LaserPointerActions: true,
  LaserPointerUtils: true,
  PrefPlugin: true,
  createCustomizedMobileBiPlugins: true,
  HoverToolbarPlugin: true,
  HoverBlockPlugin: true
};
Object.defineProperty(exports, "Provider", {
  enumerable: true,
  get: function () {
    return _provider.default;
  }
});
Object.defineProperty(exports, "Content", {
  enumerable: true,
  get: function () {
    return _content.default;
  }
});
Object.defineProperty(exports, "createBiPlugins", {
  enumerable: true,
  get: function () {
    return _createBiPlugins.default;
  }
});
Object.defineProperty(exports, "createCustomizedBiPlugins", {
  enumerable: true,
  get: function () {
    return _createCustomizedBiPlugins.default;
  }
});
Object.defineProperty(exports, "actionTypes", {
  enumerable: true,
  get: function () {
    return _actionTypes.default;
  }
});
Object.defineProperty(exports, "plugins", {
  enumerable: true,
  get: function () {
    return _plugins.default;
  }
});
Object.defineProperty(exports, "locale_en_US", {
  enumerable: true,
  get: function () {
    return _en_US.default;
  }
});
Object.defineProperty(exports, "locale_ja_JP", {
  enumerable: true,
  get: function () {
    return _ja_JP.default;
  }
});
Object.defineProperty(exports, "locale_zh_CN", {
  enumerable: true,
  get: function () {
    return _zh_CN.default;
  }
});
Object.defineProperty(exports, "locale_zh_HK", {
  enumerable: true,
  get: function () {
    return _zh_HK.default;
  }
});
Object.defineProperty(exports, "locale_zh_TW", {
  enumerable: true,
  get: function () {
    return _zh_TW.default;
  }
});
Object.defineProperty(exports, "Resizable", {
  enumerable: true,
  get: function () {
    return _everComponent.Resizable;
  }
});
Object.defineProperty(exports, "MenuTitleWithNewBadge", {
  enumerable: true,
  get: function () {
    return _everComponent.MenuTitleWithNewBadge;
  }
});
Object.defineProperty(exports, "Toolbar", {
  enumerable: true,
  get: function () {
    return _everComponent.Toolbar;
  }
});
Object.defineProperty(exports, "InlineToolbar", {
  enumerable: true,
  get: function () {
    return _everComponent.InlineToolbar;
  }
});
Object.defineProperty(exports, "ToolbarConfig", {
  enumerable: true,
  get: function () {
    return _everComponent.ToolbarConfig;
  }
});
Object.defineProperty(exports, "ToolBoxLayoutMode", {
  enumerable: true,
  get: function () {
    return _everComponent.ToolBoxLayoutMode;
  }
});
Object.defineProperty(exports, "HeaderToolbarMode", {
  enumerable: true,
  get: function () {
    return _everComponent.HeaderToolbarMode;
  }
});
Object.defineProperty(exports, "ToolbarMode", {
  enumerable: true,
  get: function () {
    return _everComponent.ToolbarMode;
  }
});
Object.defineProperty(exports, "Sidebar", {
  enumerable: true,
  get: function () {
    return _everComponent.Sidebar;
  }
});
Object.defineProperty(exports, "NewToolbar", {
  enumerable: true,
  get: function () {
    return _everComponent.NewToolbar;
  }
});
Object.defineProperty(exports, "NewToolbarConfig", {
  enumerable: true,
  get: function () {
    return _everComponent.NewToolbarConfig;
  }
});
Object.defineProperty(exports, "SimpleToolbar", {
  enumerable: true,
  get: function () {
    return _everComponent.SimpleToolbar;
  }
});
Object.defineProperty(exports, "defaultSingleLayout", {
  enumerable: true,
  get: function () {
    return _everComponent.defaultSingleLayout;
  }
});
Object.defineProperty(exports, "defaultDisableTargets", {
  enumerable: true,
  get: function () {
    return _everComponent.defaultDisableTargets;
  }
});
Object.defineProperty(exports, "defaultDoubleLayout", {
  enumerable: true,
  get: function () {
    return _everComponent.defaultDoubleLayout;
  }
});
Object.defineProperty(exports, "HeadingToolbar", {
  enumerable: true,
  get: function () {
    return _everComponent.HeadingToolbar;
  }
});
Object.defineProperty(exports, "MobileToolbar", {
  enumerable: true,
  get: function () {
    return _everComponent.MobileToolbar;
  }
});
Object.defineProperty(exports, "mobileLayout", {
  enumerable: true,
  get: function () {
    return _everComponent.mobileLayout;
  }
});
Object.defineProperty(exports, "MobileToolbarConstant", {
  enumerable: true,
  get: function () {
    return _everComponent.MobileToolbarConstant;
  }
});
Object.defineProperty(exports, "LeftToolbar", {
  enumerable: true,
  get: function () {
    return _everComponent.LeftToolbar;
  }
});
Object.defineProperty(exports, "SelectionBar", {
  enumerable: true,
  get: function () {
    return _everComponent.Trigger;
  }
});
Object.defineProperty(exports, "insertTag", {
  enumerable: true,
  get: function () {
    return _everFactory.insertTag;
  }
});
Object.defineProperty(exports, "setTagData", {
  enumerable: true,
  get: function () {
    return _everFactory.setTagData;
  }
});
Object.defineProperty(exports, "removeTag", {
  enumerable: true,
  get: function () {
    return _everFactory.removeTag;
  }
});
Object.defineProperty(exports, "tagFactory", {
  enumerable: true,
  get: function () {
    return _everFactory.tagFactory;
  }
});
Object.defineProperty(exports, "removeCard", {
  enumerable: true,
  get: function () {
    return _everFactory.removeCard;
  }
});
Object.defineProperty(exports, "setCardData", {
  enumerable: true,
  get: function () {
    return _everFactory.setCardData;
  }
});
Object.defineProperty(exports, "setCardInjection", {
  enumerable: true,
  get: function () {
    return _everFactory.setCardInjection;
  }
});
Object.defineProperty(exports, "insertCard", {
  enumerable: true,
  get: function () {
    return _everFactory.insertCard;
  }
});
Object.defineProperty(exports, "cardFactory", {
  enumerable: true,
  get: function () {
    return _everFactory.cardFactory;
  }
});
Object.defineProperty(exports, "mobileCardFactory", {
  enumerable: true,
  get: function () {
    return _everFactory.mobileCardFactory;
  }
});
Object.defineProperty(exports, "groupFactory", {
  enumerable: true,
  get: function () {
    return _everFactory.groupFactory;
  }
});
Object.defineProperty(exports, "insertTemplateButton", {
  enumerable: true,
  get: function () {
    return _everPluginTemplateButton.insertTemplateButton;
  }
});
Object.defineProperty(exports, "TemplateButton", {
  enumerable: true,
  get: function () {
    return _everPluginTemplateButton.biFactory;
  }
});
Object.defineProperty(exports, "createMobileBiPlugins", {
  enumerable: true,
  get: function () {
    return _createMobileBiPlugins.default;
  }
});
Object.defineProperty(exports, "mobilePlugins", {
  enumerable: true,
  get: function () {
    return _mobile.default;
  }
});
Object.defineProperty(exports, "AlignmentPlugin", {
  enumerable: true,
  get: function () {
    return _alignment.default;
  }
});
Object.defineProperty(exports, "MobileAlignmentPlugin", {
  enumerable: true,
  get: function () {
    return _mobile2.default;
  }
});
Object.defineProperty(exports, "AudioPlugin", {
  enumerable: true,
  get: function () {
    return _audio.default;
  }
});
Object.defineProperty(exports, "MobileAudioPlugin", {
  enumerable: true,
  get: function () {
    return _mobile3.default;
  }
});
Object.defineProperty(exports, "BasicPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginBasic.biFactory;
  }
});
Object.defineProperty(exports, "BasicIsActions", {
  enumerable: true,
  get: function () {
    return _everPluginBasic.isActions;
  }
});
Object.defineProperty(exports, "BlockquotePlugin", {
  enumerable: true,
  get: function () {
    return _blockquote.default;
  }
});
Object.defineProperty(exports, "MobileBlockquotePlugin", {
  enumerable: true,
  get: function () {
    return _mobile4.default;
  }
});
Object.defineProperty(exports, "BoldPlugin", {
  enumerable: true,
  get: function () {
    return _bold.default;
  }
});
Object.defineProperty(exports, "MobileBoldPlugin", {
  enumerable: true,
  get: function () {
    return _mobile5.default;
  }
});
Object.defineProperty(exports, "ClearStylePlugin", {
  enumerable: true,
  get: function () {
    return _clearStyle.default;
  }
});
Object.defineProperty(exports, "MobileClearStylePlugin", {
  enumerable: true,
  get: function () {
    return _mobile6.default;
  }
});
Object.defineProperty(exports, "ClipBoardPlugin", {
  enumerable: true,
  get: function () {
    return _clipboard.default;
  }
});
Object.defineProperty(exports, "MobileClipBoardPlugin", {
  enumerable: true,
  get: function () {
    return _mobile7.default;
  }
});
Object.defineProperty(exports, "CodePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginCode.biFactory;
  }
});
Object.defineProperty(exports, "MobileCodePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginCode.mobileBiFactory;
  }
});
Object.defineProperty(exports, "InlineCodePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginInlineCode.biFactory;
  }
});
Object.defineProperty(exports, "ColumnsPlugin", {
  enumerable: true,
  get: function () {
    return _columns.default;
  }
});
Object.defineProperty(exports, "CommentPlugin", {
  enumerable: true,
  get: function () {
    return _comment.default;
  }
});
Object.defineProperty(exports, "MobileCommentPlugin", {
  enumerable: true,
  get: function () {
    return _mobile8.default;
  }
});
Object.defineProperty(exports, "CommonPlugin", {
  enumerable: true,
  get: function () {
    return _commonPlugin.CommonPlugin;
  }
});
Object.defineProperty(exports, "CalendarCardPlugin", {
  enumerable: true,
  get: function () {
    return _calendarCard.default;
  }
});
Object.defineProperty(exports, "MobileCalendarCardPlugin", {
  enumerable: true,
  get: function () {
    return _mobile9.default;
  }
});
Object.defineProperty(exports, "VoteCardPlugin", {
  enumerable: true,
  get: function () {
    return _voteCard.default;
  }
});
Object.defineProperty(exports, "MobileVoteCardPlugin", {
  enumerable: true,
  get: function () {
    return _mobile10.default;
  }
});
Object.defineProperty(exports, "FindAndReplacePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginFindAndReplace.createFindAndReplacePlugin;
  }
});
Object.defineProperty(exports, "ImTagPlugin", {
  enumerable: true,
  get: function () {
    return _imTag.default;
  }
});
Object.defineProperty(exports, "ColorPlugin", {
  enumerable: true,
  get: function () {
    return _color.default;
  }
});
Object.defineProperty(exports, "ContextMenuPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginPcContexMenu.biFactory;
  }
});
Object.defineProperty(exports, "EmbedPlugin", {
  enumerable: true,
  get: function () {
    return _embed.default;
  }
});
Object.defineProperty(exports, "MobileEmbedPlugin", {
  enumerable: true,
  get: function () {
    return _mobile11.default;
  }
});
Object.defineProperty(exports, "FontsPlugin", {
  enumerable: true,
  get: function () {
    return _fonts.default;
  }
});
Object.defineProperty(exports, "FormatPainterPlugin", {
  enumerable: true,
  get: function () {
    return _formatPainter.default;
  }
});
Object.defineProperty(exports, "HeadingPlugin", {
  enumerable: true,
  get: function () {
    return _heading.default;
  }
});
Object.defineProperty(exports, "HighlightPlugin", {
  enumerable: true,
  get: function () {
    return _highlight.default;
  }
});
Object.defineProperty(exports, "HrPlugin", {
  enumerable: true,
  get: function () {
    return _hr.default;
  }
});
Object.defineProperty(exports, "HistoryPlugin", {
  enumerable: true,
  get: function () {
    return _history.default;
  }
});
Object.defineProperty(exports, "MobileHistoryPlugin", {
  enumerable: true,
  get: function () {
    return _mobile12.default;
  }
});
Object.defineProperty(exports, "ImagePlugin", {
  enumerable: true,
  get: function () {
    return _image.default;
  }
});
Object.defineProperty(exports, "MobileImagePlugin", {
  enumerable: true,
  get: function () {
    return _mobile13.default;
  }
});
Object.defineProperty(exports, "IndentPlugin", {
  enumerable: true,
  get: function () {
    return _indent.default;
  }
});
Object.defineProperty(exports, "ItalicPlugin", {
  enumerable: true,
  get: function () {
    return _italic.default;
  }
});
Object.defineProperty(exports, "LineHeightPlugin", {
  enumerable: true,
  get: function () {
    return _lineheight.default;
  }
});
Object.defineProperty(exports, "LinkPlugin", {
  enumerable: true,
  get: function () {
    return _link.default;
  }
});
Object.defineProperty(exports, "LinkCardPlugin", {
  enumerable: true,
  get: function () {
    return _linkCard.default;
  }
});
Object.defineProperty(exports, "MobileLinkPlugin", {
  enumerable: true,
  get: function () {
    return _mobile14.default;
  }
});
Object.defineProperty(exports, "MobileLinkCardPlugin", {
  enumerable: true,
  get: function () {
    return _mobile15.default;
  }
});
Object.defineProperty(exports, "ListPlugin", {
  enumerable: true,
  get: function () {
    return _list.default;
  }
});
Object.defineProperty(exports, "MarksPlugin", {
  enumerable: true,
  get: function () {
    return _marks.default;
  }
});
Object.defineProperty(exports, "MentionPlugin", {
  enumerable: true,
  get: function () {
    return _mention.default;
  }
});
Object.defineProperty(exports, "MobileMentionPlugin", {
  enumerable: true,
  get: function () {
    return _mobile16.default;
  }
});
Object.defineProperty(exports, "OnlineVideoPlugin", {
  enumerable: true,
  get: function () {
    return _onlineVideo.default;
  }
});
Object.defineProperty(exports, "MobileOnlineVideoPlugin", {
  enumerable: true,
  get: function () {
    return _mobile17.default;
  }
});
Object.defineProperty(exports, "ParagraphPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginParagraph.biFactory;
  }
});
Object.defineProperty(exports, "PlaceholderPlugin", {
  enumerable: true,
  get: function () {
    return _placeholder.default;
  }
});
Object.defineProperty(exports, "QuickInsertPlugin", {
  enumerable: true,
  get: function () {
    return _quickInsert.default;
  }
});
Object.defineProperty(exports, "NewlineGuidePlugin", {
  enumerable: true,
  get: function () {
    return _newlineGuide.default;
  }
});
Object.defineProperty(exports, "ReferPlugin", {
  enumerable: true,
  get: function () {
    return _refer.default;
  }
});
Object.defineProperty(exports, "StrikePlugin", {
  enumerable: true,
  get: function () {
    return _strike.default;
  }
});
Object.defineProperty(exports, "SzPlugin", {
  enumerable: true,
  get: function () {
    return _sz.default;
  }
});
Object.defineProperty(exports, "TablePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginTable.biFactory;
  }
});
Object.defineProperty(exports, "MobileTablePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginTable.mobileBiFactory;
  }
});
Object.defineProperty(exports, "TableActions", {
  enumerable: true,
  get: function () {
    return _everPluginTable.biActionTypes;
  }
});
Object.defineProperty(exports, "TranslatePlugin", {
  enumerable: true,
  get: function () {
    return _translate.default;
  }
});
Object.defineProperty(exports, "UnderlinePlugin", {
  enumerable: true,
  get: function () {
    return _underline.default;
  }
});
Object.defineProperty(exports, "VertAlignPlugin", {
  enumerable: true,
  get: function () {
    return _vertAlign.default;
  }
});
Object.defineProperty(exports, "VideoPlugin", {
  enumerable: true,
  get: function () {
    return _video.default;
  }
});
Object.defineProperty(exports, "DraggablePlugin", {
  enumerable: true,
  get: function () {
    return _draggable.default;
  }
});
Object.defineProperty(exports, "FormulaPlugin", {
  enumerable: true,
  get: function () {
    return _formula.default;
  }
});
Object.defineProperty(exports, "StickerPlugin", {
  enumerable: true,
  get: function () {
    return _sticker.default;
  }
});
Object.defineProperty(exports, "MobileVideoPlugin", {
  enumerable: true,
  get: function () {
    return _mobile18.default;
  }
});
Object.defineProperty(exports, "EventTrackingPlugin", {
  enumerable: true,
  get: function () {
    return _eventTracking.default;
  }
});
Object.defineProperty(exports, "SubjectPlugin", {
  enumerable: true,
  get: function () {
    return _subject.default;
  }
});
Object.defineProperty(exports, "HeadingGroupPlugin", {
  enumerable: true,
  get: function () {
    return _headingGroup.default;
  }
});
Object.defineProperty(exports, "HeadingGroupNewPlugin", {
  enumerable: true,
  get: function () {
    return _headingGroupNew.default;
  }
});
Object.defineProperty(exports, "ListGroupPlugin", {
  enumerable: true,
  get: function () {
    return _listGroup.default;
  }
});
Object.defineProperty(exports, "CalloutPlugin", {
  enumerable: true,
  get: function () {
    return _callout.default;
  }
});
Object.defineProperty(exports, "QuickPreviewPlugin", {
  enumerable: true,
  get: function () {
    return _quickPreview.default;
  }
});
Object.defineProperty(exports, "LaserPointerPlugin", {
  enumerable: true,
  get: function () {
    return _laserPointer.default;
  }
});
Object.defineProperty(exports, "LeftToolbarPlugin", {
  enumerable: true,
  get: function () {
    return _leftToolbar.default;
  }
});
Object.defineProperty(exports, "LaserPointerRole", {
  enumerable: true,
  get: function () {
    return _types2.LaserPointerRole;
  }
});
Object.defineProperty(exports, "PrefPlugin", {
  enumerable: true,
  get: function () {
    return _pref.default;
  }
});
Object.defineProperty(exports, "createCustomizedMobileBiPlugins", {
  enumerable: true,
  get: function () {
    return _createCustomizedMobileBiPlugins.default;
  }
});
Object.defineProperty(exports, "HoverToolbarPlugin", {
  enumerable: true,
  get: function () {
    return _hoverToolbar.default;
  }
});
Object.defineProperty(exports, "HoverBlockPlugin", {
  enumerable: true,
  get: function () {
    return _hoverBlock.default;
  }
});
exports.LaserPointerUtils = exports.LaserPointerActions = exports.BiMenu = void 0;

var _provider = _interopRequireDefault(require("./provider"));

var _content = _interopRequireWildcard(require("./content"));

Object.keys(_content).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _content[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _content[key];
    }
  });
});

var _createBiPlugins = _interopRequireDefault(require("./createBiPlugins"));

var _createCustomizedBiPlugins = _interopRequireWildcard(require("./createCustomizedBiPlugins"));

Object.keys(_createCustomizedBiPlugins).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _createCustomizedBiPlugins[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _createCustomizedBiPlugins[key];
    }
  });
});

var _actionTypes = _interopRequireDefault(require("./actionTypes"));

var _plugins = _interopRequireDefault(require("./plugins"));

var _en_US = _interopRequireDefault(require("./locales/en_US"));

var _ja_JP = _interopRequireDefault(require("./locales/ja_JP"));

var _zh_CN = _interopRequireDefault(require("./locales/zh_CN"));

var _zh_HK = _interopRequireDefault(require("./locales/zh_HK"));

var _zh_TW = _interopRequireDefault(require("./locales/zh_TW"));

var _everComponent = require("@ali/4ever-component");

var _pluginsContext = require("./pluginsContext");

Object.keys(_pluginsContext).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _pluginsContext[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _pluginsContext[key];
    }
  });
});

var _BiMenu = _interopRequireWildcard(require("./menu"));

exports.BiMenu = _BiMenu;
Object.keys(_BiMenu).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _BiMenu[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _BiMenu[key];
    }
  });
});

var _everFactory = require("@ali/4ever-factory");

var _everPluginTemplateButton = require("@ali/4ever-plugin-template-button");

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});

var _createMobileBiPlugins = _interopRequireDefault(require("./createMobileBiPlugins"));

var _mobile = _interopRequireDefault(require("./plugins/mobile"));

var _alignment = _interopRequireDefault(require("./plugins/alignment"));

var _mobile2 = _interopRequireDefault(require("./plugins/alignment/mobile"));

var _audio = _interopRequireDefault(require("./plugins/audio"));

var _mobile3 = _interopRequireDefault(require("./plugins/audio/mobile"));

var _everPluginBasic = require("@ali/4ever-plugin-basic");

var _blockquote = _interopRequireDefault(require("./plugins/blockquote"));

var _mobile4 = _interopRequireDefault(require("./plugins/blockquote/mobile"));

var _bold = _interopRequireDefault(require("./plugins/bold"));

var _mobile5 = _interopRequireDefault(require("./plugins/bold/mobile"));

var _clearStyle = _interopRequireDefault(require("./plugins/clearStyle"));

var _mobile6 = _interopRequireDefault(require("./plugins/clearStyle/mobile"));

var _clipboard = _interopRequireDefault(require("./plugins/clipboard"));

var _mobile7 = _interopRequireDefault(require("./plugins/clipboard/mobile"));

var _everPluginCode = require("@ali/4ever-plugin-code");

var _everPluginInlineCode = require("@ali/4ever-plugin-inline-code");

var _columns = _interopRequireDefault(require("./plugins/columns"));

var _comment = _interopRequireDefault(require("./plugins/comment"));

var _mobile8 = _interopRequireDefault(require("./plugins/comment/mobile"));

var _commonPlugin = require("./plugins/common/commonPlugin");

var _calendarCard = _interopRequireDefault(require("./plugins/calendarCard"));

var _mobile9 = _interopRequireDefault(require("./plugins/calendarCard/mobile"));

var _voteCard = _interopRequireDefault(require("./plugins/voteCard"));

var _mobile10 = _interopRequireDefault(require("./plugins/voteCard/mobile"));

var _everPluginFindAndReplace = require("@ali/4ever-plugin-find-and-replace");

var _imTag = _interopRequireDefault(require("./plugins/imTag"));

var _color = _interopRequireDefault(require("./plugins/color"));

var _everPluginPcContexMenu = require("@ali/4ever-plugin-pc-contex-menu");

var _embed = _interopRequireDefault(require("./plugins/embed"));

var _mobile11 = _interopRequireDefault(require("./plugins/embed/mobile"));

var _fonts = _interopRequireDefault(require("./plugins/fonts"));

var _formatPainter = _interopRequireDefault(require("./plugins/formatPainter"));

var _heading = _interopRequireDefault(require("./plugins/heading"));

var _highlight = _interopRequireDefault(require("./plugins/highlight"));

var _hr = _interopRequireDefault(require("./plugins/hr"));

var _history = _interopRequireDefault(require("./plugins/history"));

var _mobile12 = _interopRequireDefault(require("./plugins/history/mobile"));

var _image = _interopRequireDefault(require("./plugins/image"));

var _mobile13 = _interopRequireDefault(require("./plugins/image/mobile"));

var _indent = _interopRequireDefault(require("./plugins/indent"));

var _italic = _interopRequireDefault(require("./plugins/italic"));

var _lineheight = _interopRequireDefault(require("./plugins/lineheight"));

var _link = _interopRequireDefault(require("./plugins/link"));

var _linkCard = _interopRequireDefault(require("./plugins/link/linkCard"));

var _mobile14 = _interopRequireDefault(require("./plugins/link/mobile"));

var _mobile15 = _interopRequireDefault(require("./plugins/link/linkCard/mobile"));

var _list = _interopRequireDefault(require("./plugins/list"));

var _marks = _interopRequireDefault(require("./plugins/marks"));

var _mention = _interopRequireDefault(require("./plugins/mention"));

var _mobile16 = _interopRequireDefault(require("./plugins/mention/mobile"));

var _onlineVideo = _interopRequireDefault(require("./plugins/onlineVideo"));

var _mobile17 = _interopRequireDefault(require("./plugins/onlineVideo/mobile"));

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _placeholder = _interopRequireDefault(require("./plugins/placeholder"));

var _quickInsert = _interopRequireDefault(require("./plugins/quickInsert"));

var _newlineGuide = _interopRequireDefault(require("./plugins/newlineGuide"));

var _refer = _interopRequireDefault(require("./plugins/refer"));

var _strike = _interopRequireDefault(require("./plugins/strike"));

var _sz = _interopRequireDefault(require("./plugins/sz"));

var _everPluginTable = require("@ali/4ever-plugin-table");

var _translate = _interopRequireDefault(require("./plugins/translate"));

var _underline = _interopRequireDefault(require("./plugins/underline"));

var _vertAlign = _interopRequireDefault(require("./plugins/vertAlign"));

var _video = _interopRequireDefault(require("./plugins/video"));

var _draggable = _interopRequireDefault(require("./plugins/draggable"));

var _formula = _interopRequireDefault(require("./plugins/formula"));

var _sticker = _interopRequireDefault(require("./plugins/sticker"));

var _mobile18 = _interopRequireDefault(require("./plugins/video/mobile"));

var _eventTracking = _interopRequireDefault(require("./plugins/eventTracking"));

var _subject = _interopRequireDefault(require("./plugins/subject"));

var _headingGroup = _interopRequireDefault(require("./plugins/headingGroup"));

var _headingGroupNew = _interopRequireDefault(require("./plugins/headingGroupNew"));

var _listGroup = _interopRequireDefault(require("./plugins/listGroup"));

var _callout = _interopRequireDefault(require("./plugins/callout"));

var _quickPreview = _interopRequireDefault(require("./plugins/quickPreview"));

var _laserPointer = _interopRequireDefault(require("./plugins/laserPointer"));

var _leftToolbar = _interopRequireDefault(require("./plugins/leftToolbar"));

var _types2 = require("./plugins/laserPointer/types");

var _LaserPointerActions = _interopRequireWildcard(require("./plugins/laserPointer/actions"));

exports.LaserPointerActions = _LaserPointerActions;

var _LaserPointerUtils = _interopRequireWildcard(require("./plugins/laserPointer/utils"));

exports.LaserPointerUtils = _LaserPointerUtils;

var _pref = _interopRequireDefault(require("./plugins/pref"));

var _createCustomizedMobileBiPlugins = _interopRequireDefault(require("./createCustomizedMobileBiPlugins"));

var _hoverToolbar = _interopRequireDefault(require("./plugins/hoverToolbar"));

var _hoverBlock = _interopRequireDefault(require("./plugins/hoverBlock"));
