// @ali/4ever-bi/dist-node/menu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "TYPE_GROUP", {
  enumerable: true,
  get: function () {
    return _everComponent.TYPE_GROUP;
  }
});
Object.defineProperty(exports, "TYPE_SUB_MENU", {
  enumerable: true,
  get: function () {
    return _everComponent.TYPE_SUB_MENU;
  }
});
Object.defineProperty(exports, "TYPE_MENU_ITEM", {
  enumerable: true,
  get: function () {
    return _everComponent.TYPE_MENU_ITEM;
  }
});
Object.defineProperty(exports, "TYPE_INLINE_MENU_GROUP", {
  enumerable: true,
  get: function () {
    return _everComponent.TYPE_INLINE_MENU_GROUP;
  }
});
Object.defineProperty(exports, "TYPE_INLINE_MENU_WRAPPER", {
  enumerable: true,
  get: function () {
    return _everComponent.TYPE_INLINE_MENU_WRAPPER;
  }
});
Object.defineProperty(exports, "TYPE_INLINE_MENU_ITEM", {
  enumerable: true,
  get: function () {
    return _everComponent.TYPE_INLINE_MENU_ITEM;
  }
});
Object.defineProperty(exports, "TYPE_QUERY_ONLY_INLINE_MENU_ITEM", {
  enumerable: true,
  get: function () {
    return _everComponent.TYPE_QUERY_ONLY_INLINE_MENU_ITEM;
  }
});
Object.defineProperty(exports, "Menu", {
  enumerable: true,
  get: function () {
    return _everComponent.Menu;
  }
});
Object.defineProperty(exports, "MenuBar", {
  enumerable: true,
  get: function () {
    return _everComponent.MenuBar;
  }
});

var _everComponent = require("@ali/4ever-component");
