// @ali/4ever-zhi/dist-web/plugins/calendarCard/shanhui/FlashCalendarCardContainer.js
import { CalendarCardPlugin } from '@ali/4ever-bamboo';
import { memoize } from 'lodash-es';
import React, { useCallback, useEffect, useRef, useState } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { JumpToCalendarEventType } from "../type";
var TimeTextParagraph = CalendarCardPlugin.TimeTextParagraph,
    PlaceTextParagraph = CalendarCardPlugin.PlaceTextParagraph,
    ReceiversParagraph = CalendarCardPlugin.ReceiversParagraph,
    ContainerWrapper = CalendarCardPlugin.ContainerWrapper,
    BodyWrapper = CalendarCardPlugin.BodyWrapper,
    TimePlaceholder = CalendarCardPlugin.TimePlaceholder; // 当mousedown与click之间的间隔少于150ms时认为是点击点击而不是拖动选中

var clickTimeout = 150;
export default function FlashCalendarCardContainer(props) {
  var locale = props.locale,
      metadata = props.metadata,
      jumpToCalendar = props.jumpToCalendar,
      openProfile = props.openProfile;
  var time = metadata.time,
      place = metadata.place,
      redirectUrl = metadata.redirectUrl,
      isPlaceholder = metadata.isPlaceholder;
  var jumpToCalendarFactory = useCallback(memoize(function (eventType) {
    return function (event) {
      event.preventDefault();
      event.stopPropagation();
      if (!jumpToCalendar) return;
      if (isPlaceholder) return;
      jumpToCalendar(redirectUrl, eventType);
    };
  }), [jumpToCalendar]);

  var _useState = useState({}),
      style = _useState[0],
      setStyle = _useState[1];

  useEffect(function () {
    setStyle({
      color: isPlaceholder ? 'rgba(17,31,44,0.4)' : '',
      height: 'auto',
      width: 'auto',
      border: '1px solid rgba(17,31,44,0)'
    });
  }, [isPlaceholder]);
  var clickTimeRef = useRef();
  var handleMouseDown = useCallback(function (event) {
    clickTimeRef.current = new Date().getTime();
    event.stopPropagation();
  }, []);
  var handleClick = useCallback(function (event) {
    if (!clickTimeRef.current || new Date().getTime() - clickTimeRef.current > clickTimeout) {
      event.stopPropagation();
    }

    clickTimeRef.current = undefined;
  }, []);
  return /*#__PURE__*/_createElement(ContainerWrapper, {
    className: "flash-calendar-container",
    onMouseDown: handleMouseDown,
    onClick: handleClick,
    style: style,
    backgroundColor: isPlaceholder ? '#F9FAFC' : '#F0F7FF'
  }, /*#__PURE__*/_createElement(BodyWrapper, null, /*#__PURE__*/_createElement(TimeTextParagraph, null, isPlaceholder ? /*#__PURE__*/_createElement(TimePlaceholder, null, time) : time), /*#__PURE__*/_createElement(PlaceTextParagraph, null, place), /*#__PURE__*/_createElement(ReceiversParagraph, {
    locale: locale,
    metadata: metadata,
    openProfile: openProfile,
    jumpToCalendar: jumpToCalendarFactory(JumpToCalendarEventType.RECEIVER)
  })));
}
