// @ali/4ever-zhi/dist-web/plugins/headingGroupNew/index.js
import _extends from "@babel/runtime/helpers/extends";
import { GroupFactory, HeadingGroupPlugin as BambooHeadingGroupPlugin } from '@ali/4ever-bamboo';
import { Heading } from '@ali/4ever-mo';
import createRenderGroup from "./handlers/createRenderGroup";
import createGenerateGroup from "./handlers/createGenerateGroup";
var createRenderMark = GroupFactory.createRenderMark;
var Heading1 = Heading.Heading1,
    Heading2 = Heading.Heading2,
    Heading3 = Heading.Heading3;

var HeadingGroupPlugin = function HeadingGroupPlugin(config) {
  var _config$headingGroup = config.headingGroup,
      headingGroup = _config$headingGroup === void 0 ? {} : _config$headingGroup;
  var _headingGroup$level = headingGroup.level,
      level = _headingGroup$level === void 0 ? 3 : _headingGroup$level;
  return {
    renderGroup: createRenderGroup(headingGroup),
    generateGroup: createGenerateGroup(headingGroup),
    renderMark: createRenderMark({
      autoUnfold: headingGroup.autoUnfold,
      inject: true
    }),
    queries: _extends({
      getHeadingGroupLevel: function getHeadingGroupLevel() {
        return level;
      },
      isHeadingGroupAvailable: function isHeadingGroupAvailable() {
        return true;
      },
      isElementInGroup: function isElementInGroup(controller, element, group) {
        // 分组 2 只能存在于分组 1 之中
        if (Heading1.isHeading1(element)) {
          return false;
        } else if (Heading2.isHeading2(element)) {
          if (group.type === 'heading1-group') {
            return true;
          }

          return false;
        } else if (Heading3.isHeading3(element)) {
          if (group.type === 'heading1-group' || group.type === 'heading2-group') {
            return true;
          }

          return false;
        }

        return true;
      }
    }, BambooHeadingGroupPlugin.queries)
  };
}; // @ts-ignore


HeadingGroupPlugin.actions = BambooHeadingGroupPlugin.actions; // @ts-ignore

HeadingGroupPlugin.queries = BambooHeadingGroupPlugin.queries;
export default HeadingGroupPlugin;
