// @ali/4ever-factory/dist-node/zhi/groupFactory/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.groupFactory = groupFactory;

var _everPluginGroupFactory = require("@ali/4ever-plugin-group-factory");

var _everPluginHeadingGroup = require("@ali/4ever-plugin-heading-group");

var _createOnAction = _interopRequireDefault(require("./handlers/createOnAction"));

var _createOnCangjieSelect = _interopRequireDefault(require("./handlers/createOnCangjieSelect"));

var factoryCommands = _interopRequireWildcard(require("./commands"));

function groupFactory(createPlugin) {
  return function createGroupPlugin(zhiConfig) {
    const myPlugin = createPlugin(zhiConfig);
    const {
      commands,
      onAction,
      queries,
      groupNodes,
      renderNode,
      renderMark,
      groupType
    } = myPlugin;
    const config = zhiConfig[groupType] || {};
    return {
      shouldNodeComponentUpdate: (0, _everPluginGroupFactory.createShouldComponentUpdate)(),
      renderNode: renderNode || (0, _everPluginGroupFactory.createRenderNode)(groupType),
      renderMark: renderMark || (0, _everPluginHeadingGroup.createRenderMark)(config),
      groupNodes,
      onAction: (0, _createOnAction.default)(onAction),
      onCangjieSelect: (0, _createOnCangjieSelect.default)(config),
      commands: { ...factoryCommands,
        ...commands
      },
      queries
    };
  };
}
