// @ali/we-design-next/dist-web/display/modal/methods.js
import _extends from "@babel/runtime/helpers/extends";
var _div;
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import ReactDOM from 'react-dom';
import { Button } from "../../basic/button";
import { Modal } from "./Modal";
import { ModalFooter } from "./ModalFooter";
import { H1, Text } from "../../basic/text";
var iconInfo = /*#__PURE__*/_createElement("svg", {
  width: "44",
  height: "44",
  viewBox: "0 0 44 44",
  fill: "none",
  stroke: "none",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/_createElement("circle", {
  cx: "22",
  cy: "22",
  r: "20",
  fill: "#0089FF"
}), /*#__PURE__*/_createElement("path", {
  d: "M20 22C20 20.8954 20.8954 20 22 20C23.1046 20 24 20.8954 24 22V32C24 33.1046 23.1046 34 22 34C20.8954 34 20 33.1046 20 32V22Z",
  fill: "white"
}), /*#__PURE__*/_createElement("circle", {
  cx: "22",
  cy: "14",
  r: "3",
  fill: "white"
}));
var iconWarning = /*#__PURE__*/_createElement("svg", {
  width: "44",
  height: "44",
  viewBox: "0 0 44 44",
  fill: "none",
  stroke: "none",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/_createElement("circle", {
  cx: "22",
  cy: "22",
  r: "20",
  fill: "#FF9200"
}), /*#__PURE__*/_createElement("path", {
  d: "M20 13C20 11.8954 20.8954 11 22 11C23.1046 11 24 11.8954 24 13V23C24 24.1046 23.1046 25 22 25C20.8954 25 20 24.1046 20 23V13Z",
  fill: "white"
}), /*#__PURE__*/_createElement("circle", {
  cx: "22",
  cy: "31",
  r: "3",
  fill: "white"
}));
var iconSuccess = /*#__PURE__*/_createElement("svg", {
  width: "44",
  height: "44",
  viewBox: "0 0 44 44",
  fill: "none",
  stroke: "none",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/_createElement("circle", {
  cx: "22",
  cy: "22",
  r: "20",
  fill: "#00B042"
}), /*#__PURE__*/_createElement("path", {
  d: "M31 17L20 28L14 22",
  stroke: "white",
  "stroke-width": "4",
  "stroke-linecap": "round",
  "stroke-linejoin": "round"
}));
var ModalMethod = /*#__PURE__*/React.memo(function (props) {
  var type = props.type,
    message = props.message,
    title = props.title,
    _props$options = props.options,
    options = _props$options === void 0 ? [] : _props$options,
    confirmText = props.confirmText,
    cancelText = props.cancelText,
    onConfirm = props.onConfirm,
    onCancel = props.onCancel,
    mask = props.mask;
  return /*#__PURE__*/_createElement(Modal, {
    style: {
      width: 360
    },
    visible: true,
    header: null,
    content: /*#__PURE__*/_createElement("div", {
      style: {
        display: 'flex'
      }
    }, /*#__PURE__*/_createElement("div", {
      style: {
        width: '40px',
        marginTop: 10
      }
    }, type === 'info' || type === 'confirm' ? iconInfo : type === 'success' ? iconSuccess : type === 'warning' ? iconWarning : null), /*#__PURE__*/_createElement("div", {
      style: {
        paddingLeft: '16px'
      }
    }, title ? /*#__PURE__*/_createElement(H1, null, title) : null, /*#__PURE__*/_createElement(Text, null, message), options.length > 0 ? /*#__PURE__*/_createElement("div", null, options.map(function (opt) {
      return _div || (_div = /*#__PURE__*/_createElement("div", null));
    })) : null)),
    footer: /*#__PURE__*/_createElement(ModalFooter, {
      cancel: type === 'confirm' || type === 'warning' ? /*#__PURE__*/_createElement(Button, {
        children: cancelText || undefined,
        onClick: function onClick() {
          onCancel && onCancel();
          destory();
        }
      }) : null,
      confirm: /*#__PURE__*/_createElement(Button, {
        children: confirmText || undefined,
        onClick: function onClick() {
          onConfirm && onConfirm();
          destory();
        }
      })
    }),
    mask: mask
  });
});
var method = function method(options) {
  ReactDOM.render( /*#__PURE__*/_createElement(ModalMethod, options), ensureContainer());
};
var destory = function destory() {
  ReactDOM.render( /*#__PURE__*/_createElement(React.Fragment, null), ensureContainer());
};
var ensureContainer = function ensureContainer() {
  var id = 'wdn-modal-method-root';
  var divEle = document.querySelector("div#" + id);
  if (!divEle) {
    var ele = document.createElement('div');
    ele.id = id;
    ele.style.position = 'absolute';
    ele.style.width = '100%';
    ele.style.top = '0px';
    ele.style.left = '0px';
    document.body.appendChild(ele);
  }
  return document.querySelector("div#" + id);
};
export var info = function info(message, options) {
  if (options === void 0) {
    options = {};
  }
  return method(_extends({
    type: 'info',
    message: message
  }, options));
};
export var warn = function warn(message, options) {
  if (options === void 0) {
    options = {};
  }
  return method(_extends({
    type: 'warning',
    message: message
  }, options));
};
export var success = function success(message, options) {
  if (options === void 0) {
    options = {};
  }
  return method(_extends({
    type: 'success',
    message: message
  }, options));
};
export var confirm = function confirm(message, options) {
  if (options === void 0) {
    options = {};
  }
  return method(_extends({
    type: 'confirm',
    message: message
  }, options));
};
