// @ali/we-design-next/dist-web/display/modal/styled.js
import styled from 'styled-components';
import { CSS_H1, CSS_TEXT } from "../../basic/text/styled";
import { COLORS, CSS_BOX_SHADOWS } from "../../common";
export var MoadlRoot = /*#__PURE__*/styled.div(["box-sizing:border-box;"]);
export var ModalContainer = /*#__PURE__*/styled.div(["", " background-color:", ";box-sizing:border-box;position:fixed;left:0;top:0;max-width:calc(100vw - 32px);min-width:300px;margin:0 auto;border:0;border-radius:4px;padding:16px;transform:translate(-9999px,-9999px);z-index:600;"], CSS_BOX_SHADOWS.L, COLORS.LIGHT.DAVINCI_BG_NEW);
export var ModalMask = /*#__PURE__*/styled.div(["display:block;position:fixed;top:0;right:0;bottom:0;left:0;z-index:500;height:100%;background-color:", ";"], function (_ref) {
  var isTransparent = _ref.isTransparent;
  return isTransparent ? COLORS.TRANSPARENT : 'rgba(0, 0, 0, 0.4)';
});
export var ModalHeader = /*#__PURE__*/styled.div(["", " display:flex;align-items:center;font-weight:500;user-select:none;line-height:16px;"], CSS_H1);
export var ModalHeaderTitle = /*#__PURE__*/styled.div(["flex-grow:1;padding-bottom:16px;", ""], function (_ref2) {
  var move = _ref2.move;
  return move ? "cursor: move;" : null;
});
export var ModalHeaderClose = /*#__PURE__*/styled.div(["margin-bottom:16px;& > .wdn-icon{border-radius:4px;transition:background 0.2s ease-out;}& > .wdn-icon:hover{background:rgba(228,230,233,1);}"]);
export var ModalContent = /*#__PURE__*/styled.div(["min-height:100px;", ""], CSS_TEXT);
export var ModalFooter = /*#__PURE__*/styled.div(["padding-top:16px;display:flex;justify-content:flex-end;align-items:center;& > .wdn-button{margin-left:12px;vertical-align:middle;}"]);
export var ModalFooterAction = /*#__PURE__*/styled.div(["flex-grow:1;"]);
