// @ali/4ever-plugin-embed/dist-web/utils/components/common/embedPreviewHeader/index.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { formatSize, getIcon } from "../../../fileTypeHelper";
import { PreviewHeaderWrapper, CardIconWrapper, CardContentWrapper, CardContentInnerWrapper, ContentHeader, ContentDesc } from "./styled";

function EmbedPreviewHeader(props) {
  var name = props.name,
      size = props.size,
      type = props.type,
      mode = props.mode,
      isMobile = props.isMobile,
      renderExtra = props.renderExtra;
  return /*#__PURE__*/_createElement(PreviewHeaderWrapper, {
    mode: mode
  }, /*#__PURE__*/_createElement(CardIconWrapper, null, /*#__PURE__*/_createElement("img", {
    src: getIcon(type),
    alt: "icon",
    style: isMobile ? {
      width: 23,
      height: 24
    } : {
      width: 16,
      height: 23
    }
  })), /*#__PURE__*/_createElement(CardContentWrapper, null, /*#__PURE__*/_createElement(CardContentInnerWrapper, null, /*#__PURE__*/_createElement(ContentHeader, {
    mode: mode
  }, name), /*#__PURE__*/_createElement(ContentDesc, {
    mode: mode
  }, formatSize(size)))), renderExtra && renderExtra());
}

export default EmbedPreviewHeader;
