// @ali/4ever-plugin-embed/dist-web/utils/components/common/embedPreviewHeader/styled.js
import styled from 'styled-components';
export var PreviewHeaderWrapper = /*#__PURE__*/styled.div(["width:100%;background-color:", ";padding:0 16px;height:32px;display:flex;border-top-left-radius:6px;border-top-right-radius:6px;line-height:23px;box-sizing:border-box;align-items:center;"], function (props) {
  return props.mode === 'dark' ? '#111F2C' : '#F6F8F9';
});
export var CardIconWrapper = /*#__PURE__*/styled.div(["margin-right:16px;flex:0;line-height:100%;img{vertical-align:initial;}"]);
export var CardContentWrapper = /*#__PURE__*/styled.div(["flex:1;min-width:0;"]);
export var CardContentInnerWrapper = /*#__PURE__*/styled.div(["display:flex;margin-top:'';"]);
export var ContentHeader = /*#__PURE__*/styled.div(["font-size:14px;color:", ";flex:1;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;"], function (props) {
  return props.mode === 'dark' ? '#F2F4F5' : '#111F2C';
});
export var ContentDesc = /*#__PURE__*/styled.div(["font-size:12px;color:", ";margin-right:16px;"], function (props) {
  return props.mode === 'dark' ? '#F2F4F5' : '#7A8189';
});
