// @ali/4ever-bi/dist-node/plugins/comment/components/next/createCommentButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createCommentButton;

var React = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everComponent = require("@ali/4ever-component");

var _everCangjie = require("@ali/4ever-cangjie");

const _createElement = /*#__PURE__*/React.createElement;
const MOD = _everCangjie.environment.IS_MAC ? '⌘' : 'Ctrl';

var _ref = /*#__PURE__*/_createElement(_everComponent.IconComment, null);

function createCommentButton(props) {
  const {
    controller,
    locale,
    icon,
    title,
    tooltip,
    badgeDot,
    disabled
  } = props;

  const handleClick = () => {
    controller.run('onAction', {
      type: _everBamboo.CommentPlugin.actions.ACTION_COMMENT_CLICK
    });
  };

  const isDisabled = disabled || !controller.query(_everBamboo.CommentPlugin.queries.canComment);
  const defaultTooltip = React.useMemo(() => ({
    title: locale?.comment,
    shortCut: `${MOD}+Shift+m`
  }), []);
  const buttonConfig = {
    icon: icon || _ref,
    onClick: handleClick,
    title: title || locale?.title,
    tooltip: tooltip || defaultTooltip,
    badgeDot,
    disabled: isDisabled,
    role: _everBamboo.PluginRoles.comment
  };
  return buttonConfig;
}
