// @ali/4ever-component/dist-web/MobileToolbar/constant.js
export var BOTTOM_BAR_SIZE = 44;
export var BOTOTM_PANEL_HEIGHT = 400;
export var Z_INDEX = 1900;
export var PANEL_HEADER_SIZE = 56; // icon 类的 button 宽度

export var ICON_BUTOTN_WIDTH = 68; // 无渲染内容的分割，用户区别两个连续的按钮，做 round border 的效果

export var DIVIDER_NONE = 'mobileToolbar_tinySeperator'; // 新行

export var NEW_LINE = 'mobileToolbar_newLine'; // 小分割（约 12px）

export var DIVIDER = 'mobileToolbar_divider'; // 小分割（竖向）

export var DIVIDER_VERT = 'mobileToolbar_divider_vert'; // 空白的 icon，用于最后一行的左对齐

export var ALIGN_PLACEHOLDER = 'mobileToolbar_emptyIcon';
export var THREED_TRAMSFORM = 'translate3d(0, 0, 0)'; // 浮出二级菜单的最小左侧距离

export var SUBLINE_MIN_LEFT = 24; // 浮出二级菜单和底部的 margin

export var SUBLINE_MARGIN_BOTTOM = 8; // name -> itemKey。移除 [. 空格 -]，转小写

export var itemToKey = function itemToKey(item) {
  var name = typeof item === 'object' ? item.name : item;
  return String(name).replace(/ |\.|-/g, '_').toLowerCase();
};
export default {
  DIVIDER: DIVIDER,
  DIVIDER_NONE: DIVIDER_NONE,
  DIVIDER_VERT: DIVIDER_VERT,
  NEW_LINE: NEW_LINE
};
