// @ali/4ever-plugin-table/dist-node/bi/components/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TableSelectionRect = exports.Wrapper = exports.IconWrapper = exports.ButtonWrapper = exports.DEFAULT_BUTTON_SIZE = exports.MenuDivider = exports.MenuItem = exports.Menu = exports.TableRightSideShadow = exports.TableLeftSideShadow = exports.TableSideShadow = exports.RowIndicatorHeader = exports.RowIndicator = exports.ColIndicatorHeader = exports.ColIndicator = exports.RowResizer = exports.ColResizer = exports.TableResizeControlRightLine = exports.TableResizeControlBottomLine = exports.TableResizeControlLines = exports.TableResizeControlPoint = exports.TableResizeController = exports.SelectionRect = exports.RealTable = exports.TableWrapper = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constants = require("../constants");

var _constants2 = require("../../utils/constants");

const TableWrapper = /*#__PURE__*/_styledComponents.default.div(["position:relative;", ""], p => p.visible === false ? 'display: none' : '');

exports.TableWrapper = TableWrapper;

const RealTable = /*#__PURE__*/_styledComponents.default.table(["table-layout:fixed;border-collapse:collapse;position:relative;"]);

exports.RealTable = RealTable;

const SelectionRect = /*#__PURE__*/_styledComponents.default.div(["position:absolute;display:none;background-color:", ";z-index:1;pointer-events:none;outline-offset:-2px;border:1px solid ", ";"], p => p.isHighlight ? _constants.THEME.red0 : _constants.THEME.blue0, p => p.isHighlight ? _constants.THEME.red1 : _constants.THEME.blue3);

exports.SelectionRect = SelectionRect;

const TableResizeController = /*#__PURE__*/_styledComponents.default.div(["position:absolute;"]);

exports.TableResizeController = TableResizeController;

const TableResizeControlPoint = /*#__PURE__*/_styledComponents.default.div(["z-index:3;position:absolute;width:7px;height:7px;cursor:nw-resize;cursor:nwse-resize;background-color:white;border:1px solid rgba(23,26,29,0.24);left:-4px;top:-4px;"]);

exports.TableResizeControlPoint = TableResizeControlPoint;

const TableResizeControlLines = /*#__PURE__*/_styledComponents.default.div(["display:none;"]);

exports.TableResizeControlLines = TableResizeControlLines;

const TableResizeControlBottomLine = /*#__PURE__*/_styledComponents.default.div(["z-index:2;position:absolute;width:100px;height:1px;background:#0089FF;right:0;bottom:0;background:linear-gradient(to left,rgba(0,137,255,1),rgba(0,137,255,0));mask-image:linear-gradient(to left,#000 6px,transparent 6px);mask-repeat:repeat-x;mask-position:0 0;mask-size:8px 2px;"]);

exports.TableResizeControlBottomLine = TableResizeControlBottomLine;

const TableResizeControlRightLine = /*#__PURE__*/_styledComponents.default.div(["z-index:2;position:absolute;height:100px;width:1px;background:#0089FF;right:0;bottom:0;background:linear-gradient(to top,rgba(0,137,255,1),rgba(0,137,255,0));mask-image:linear-gradient(to top,#000 6px,transparent 6px);mask-repeat:repeat-y;mask-position:0 0;mask-size:2px 8px;"]);

exports.TableResizeControlRightLine = TableResizeControlRightLine;

const ColResizer = /*#__PURE__*/_styledComponents.default.div(["display:none;position:absolute;bottom:", "px;width:0;border-left:1px solid #3296FA;z-index:30;cursor:col-resize;"], _constants2.SCROLL_BAR_HEIGHT);

exports.ColResizer = ColResizer;

const RowResizer = /*#__PURE__*/_styledComponents.default.div(["display:none;position:absolute;height:0;border-top:1px solid #3296FA;z-index:5;cursor:row-resize;"]);

exports.RowResizer = RowResizer;
const INDICATOR_HEAD_HEIGHT = 4;
const INDICATOR_COLOR = '#0089FF';

const ColIndicator = /*#__PURE__*/_styledComponents.default.div(["position:absolute;width:0;border-left:", "px solid ", ";cursor:col-resize;z-index:30;opacity:0;transition:opacity ease-in 0.25s;pointer-events:none;"], _constants.INSERT_INDICATOR_LINE_SIZE, INDICATOR_COLOR);

exports.ColIndicator = ColIndicator;

const ColIndicatorHeader = /*#__PURE__*/_styledComponents.default.div(["position:absolute;width:0;border-left:", "px solid ", ";height:", "px;background-color:", ";top:-", "px;left:-", "px;pointer-events:none;"], _constants.INSERT_INDICATOR_LINE_SIZE, INDICATOR_COLOR, p => INDICATOR_HEAD_HEIGHT / p.scale, _constants.THEME.rowColIndicatorColor, p => INDICATOR_HEAD_HEIGHT / p.scale, _constants.INSERT_INDICATOR_LINE_SIZE);

exports.ColIndicatorHeader = ColIndicatorHeader;

const RowIndicator = /*#__PURE__*/_styledComponents.default.div(["position:absolute;height:0;border-top:", "px solid ", ";transform:translateY(-1px);cursor:row-resize;z-index:99;opacity:0;transition:opacity ease-in 0.25s;pointer-events:none;"], _constants.INSERT_INDICATOR_LINE_SIZE, INDICATOR_COLOR);

exports.RowIndicator = RowIndicator;

const RowIndicatorHeader = /*#__PURE__*/_styledComponents.default.div(["position:absolute;height:0;width:", "px;background-color:", ";left:-", "px;top:-", "px;pointer-events:none;"], p => INDICATOR_HEAD_HEIGHT / p.scale, _constants.THEME.rowColIndicatorColor, p => (INDICATOR_HEAD_HEIGHT - 1) / p.scale, _constants.INSERT_INDICATOR_LINE_SIZE);

exports.RowIndicatorHeader = RowIndicatorHeader;

const TableSideShadow = /*#__PURE__*/_styledComponents.default.div(["position:absolute;bottom:", "px;width:4px;height:100%;opacity:.5;pointer-events:none;"], p => _constants2.SCROLL_BAR_HEIGHT / (p.scale || 1));

exports.TableSideShadow = TableSideShadow;
const TableLeftSideShadow = /*#__PURE__*/(0, _styledComponents.default)(TableSideShadow)(["left:0;background:linear-gradient(270deg,rgba(99,114,130,0),rgba(99,114,130,.16));"]);
exports.TableLeftSideShadow = TableLeftSideShadow;
const TableRightSideShadow = /*#__PURE__*/(0, _styledComponents.default)(TableSideShadow)(["right:0;background:linear-gradient(90deg,rgba(99,114,130,0),rgba(99,114,130,.16));"]);
exports.TableRightSideShadow = TableRightSideShadow;

const Menu = /*#__PURE__*/_styledComponents.default.ul(["position:absolute;border-radius:4px;box-shadow:0 2px 8px rgba(0,0,0,0.15);margin:0;padding-left:0;background-color:#fff;list-style:none;z-index:3;"]);

exports.Menu = Menu;

const MenuItem = /*#__PURE__*/_styledComponents.default.li(["height:32px;line-height:32px;word-break:keep-all;white-space:nowrap;font-size:12px;padding:0 12px;color:", ";cursor:", ";:hover{background-color:", ";}"], ({
  disabled
}) => disabled ? 'rgba(0, 0, 0, 0.25)' : 'rgb(89, 89, 89)', ({
  disabled
}) => disabled ? 'not-allowed' : 'pointer', ({
  disabled
}) => disabled ? 'transparent' : '#f5f6f7');

exports.MenuItem = MenuItem;

const MenuDivider = /*#__PURE__*/_styledComponents.default.li(["height:1px;margin:1px 0;overflow:hidden;background-color:#e8e8e8;"]);

exports.MenuDivider = MenuDivider;
const DEFAULT_BUTTON_SIZE = 20;
exports.DEFAULT_BUTTON_SIZE = DEFAULT_BUTTON_SIZE;

const ButtonWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;align-items:center;justify-content:center;width:", "px;height:", "px;font-size:", "px;line-height:", "px;color:black;border-radius:", "px;background-color:white;border:1px solid rgba(17,31,44,0.08);box-shadow:0px 0px 1px rgba(0,0,0,0.16),0px 1px 4px rgba(126,134,142,0.18);box-sizing:border-box;"], p => (p.size || DEFAULT_BUTTON_SIZE) / p.scale, p => (p.size || DEFAULT_BUTTON_SIZE) / p.scale, p => 20 / p.scale, p => 20 / p.scale, p => 4 / p.scale);

exports.ButtonWrapper = ButtonWrapper;

const IconWrapper = /*#__PURE__*/_styledComponents.default.span(["display:inline-flex;align-items:center;justify-content:center;"]);

exports.IconWrapper = IconWrapper;

const Wrapper = /*#__PURE__*/_styledComponents.default.div(["position:relative;"]);

exports.Wrapper = Wrapper;

const TableSelectionRect = /*#__PURE__*/_styledComponents.default.div(["position:absolute;z-index:100;pointer-events:none;outline-offset:-2px;box-sizing:border-box;border:1px solid ", ";"], _constants.THEME.blue3);

exports.TableSelectionRect = TableSelectionRect;
