// @ali/we-toolbar/dist-web/common/onIdle/idle-callback-polyfills.js
// @ts-nocheck

/*
 * Copyright 2018 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var supportsRequestIdleCallback_ = typeof requestIdleCallback === 'function';
/**
 * A minimal shim of the native IdleDeadline class.
 */

var IdleDeadline = /*#__PURE__*/function () {
  /** @param {number} initTime */
  function IdleDeadline(initTime) {
    this.didTimeout = false;
    this.initTime_ = initTime;
  }
  /** @return {boolean} */


  var _proto = IdleDeadline.prototype;

  /** @return {number} */
  _proto.timeRemaining = function timeRemaining() {
    return Math.max(0, 50 - (Date.now() - this.initTime_));
  };

  return IdleDeadline;
}();
/**
 * A minimal shim for the requestIdleCallback function. This accepts a
 * callback function and runs it at the next idle period, passing in an
 * object with a `timeRemaining()` method.
 * @private
 * @param {!Function} callback
 * @return {number}
 */


var requestIdleCallbackShim = function requestIdleCallbackShim(callback) {
  var deadline = new IdleDeadline(Date.now());
  return setTimeout(function () {
    return callback(deadline);
  }, 0);
};
/**
 * A minimal shim for the  cancelIdleCallback function. This accepts a
 * handle identifying the idle callback to cancel.
 * @private
 * @param {number|null} handle
 */


var cancelIdleCallbackShim = function cancelIdleCallbackShim(handle) {
  clearTimeout(handle);
};
/**
 * The native `requestIdleCallback()` function or `cancelIdleCallbackShim()`
 *.if the browser doesn't support it.
 * @param {!Function} callback
 * @return {number}
 */


export var rIC = supportsRequestIdleCallback_ ? requestIdleCallback : requestIdleCallbackShim;
/**
 * The native `cancelIdleCallback()` function or `cancelIdleCallbackShim()`
 * if the browser doesn't support it.
 * @param {number} handle
 */

export var cIC = supportsRequestIdleCallback_ ? cancelIdleCallback : cancelIdleCallbackShim;
