// @ali/we-design-next/dist-node/form/date-picker/panels/TimePanel/TimeColumn.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateUnits = exports.TimeColumn = void 0;
var _react = _interopRequireWildcard(require("react"));
var _common = require("../../../../common");
var _styled = require("./styled");
var _dateUtil = require("../../date-util");
var _Animation = _interopRequireDefault(require("./Animation"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const _createElement = /*#__PURE__*/_react.default.createElement;
const generateUnits = (start, end, step = 1) => {
  const units = [];
  for (let i = start; i <= end; i += step) {
    units.push({
      label: (0, _dateUtil.leftPad)(i, 2),
      value: i
    });
  }
  return units;
};
exports.generateUnits = generateUnits;
const TimeColumn = /*#__PURE__*/_react.default.memo(props => {
  const {
    value = 0,
    units,
    onSelect,
    title,
    showNow
  } = props;
  const ulRef = (0, _react.useRef)(null);
  (0, _react.useLayoutEffect)(() => {
    if (!ulRef.current) {
      return;
    }
    const targetValue = _styled.LI_HEIGHT * value;
    const currentValue = ulRef.current.scrollTop;
    const animation = new _Animation.default({
      duration: 240,
      progress(p) {
        if (ulRef.current) {
          ulRef.current.scrollTop = p * (targetValue - currentValue) + currentValue;
        }
      },
      done() {
        if (ulRef.current) {
          ulRef.current.scrollTop = targetValue;
        }
      }
    }).start();
    return () => {
      animation.clear();
    };
  }, [value]);
  (0, _react.useEffect)(() => {
    const dom = ulRef && ulRef.current;
    const stopPropagation = e => e.stopPropagation();
    dom === null || dom === void 0 ? void 0 : dom.addEventListener('mousewheel', stopPropagation);
    return () => dom === null || dom === void 0 ? void 0 : dom.removeEventListener('mousewheel', stopPropagation);
  }, []);
  return /*#__PURE__*/_createElement(_styled.UlWrapper, {
    ref: ulRef,
    showNow: showNow
  }, title ? /*#__PURE__*/_createElement(_styled.TitleLiWrapper, null, /*#__PURE__*/_createElement(_styled.TitleWrapper, null, title)) : null, units.map(unit => /*#__PURE__*/_createElement(_styled.LiWrapper, {
    key: unit.value,
    onClick: () => onSelect === null || onSelect === void 0 ? void 0 : onSelect(unit.value),
    className: (0, _common.mergeClassName)(value === unit.value ? "selected" : undefined)
  }, /*#__PURE__*/_createElement(_styled.NumberWrapper, null, unit.label))));
});
exports.TimeColumn = TimeColumn;
