// @ali/4ever-component/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "idleRender", {
  enumerable: true,
  get: function () {
    return _idleRender.idleRender;
  }
});
Object.defineProperty(exports, "SimpleLazyLoad", {
  enumerable: true,
  get: function () {
    return _SimpleLazyLoad.default;
  }
});
Object.defineProperty(exports, "Portal", {
  enumerable: true,
  get: function () {
    return _index.Portal;
  }
});
Object.defineProperty(exports, "ErrorBoundary", {
  enumerable: true,
  get: function () {
    return _ErrorBoundary.ErrorBoundary;
  }
});
Object.defineProperty(exports, "InlinePlaceholder", {
  enumerable: true,
  get: function () {
    return _inlinePlaceholder.InlinePlaceholder;
  }
});
Object.defineProperty(exports, "InlinePlaceholderWrapper", {
  enumerable: true,
  get: function () {
    return _inlinePlaceholder.InlinePlaceholderWrapper;
  }
});
Object.defineProperty(exports, "BambooPlaceholder", {
  enumerable: true,
  get: function () {
    return _bambooPlaceholder.BambooPlaceholder;
  }
});
Object.defineProperty(exports, "PlaceholderWrapper", {
  enumerable: true,
  get: function () {
    return _bambooPlaceholder.PlaceholderWrapper;
  }
});
Object.defineProperty(exports, "Placeholder", {
  enumerable: true,
  get: function () {
    return _placeholder.Placeholder;
  }
});
Object.defineProperty(exports, "Select", {
  enumerable: true,
  get: function () {
    return _index2.Select;
  }
});
Object.defineProperty(exports, "InputNumberItem", {
  enumerable: true,
  get: function () {
    return _index2.InputNumberItem;
  }
});
Object.defineProperty(exports, "CustomInputNumberItem", {
  enumerable: true,
  get: function () {
    return _index2.CustomInputNumberItem;
  }
});
Object.defineProperty(exports, "Sidebar", {
  enumerable: true,
  get: function () {
    return _index2.Sidebar;
  }
});
Object.defineProperty(exports, "SidebarItems", {
  enumerable: true,
  get: function () {
    return _index2.SidebarItems;
  }
});
Object.defineProperty(exports, "QuickInsertButton", {
  enumerable: true,
  get: function () {
    return _button.default;
  }
});
Object.defineProperty(exports, "QuickInsertCommonPanel", {
  enumerable: true,
  get: function () {
    return _commonPanel.default;
  }
});
Object.defineProperty(exports, "Resizable", {
  enumerable: true,
  get: function () {
    return _Resizable.default;
  }
});
Object.defineProperty(exports, "PluginsContext", {
  enumerable: true,
  get: function () {
    return _PluginsContext.PluginsContext;
  }
});
Object.defineProperty(exports, "usePlugins", {
  enumerable: true,
  get: function () {
    return _PluginsContext.usePlugins;
  }
});
Object.defineProperty(exports, "connectPlugins", {
  enumerable: true,
  get: function () {
    return _PluginsContext.connectPlugins;
  }
});
Object.defineProperty(exports, "MenuTitleWithNewBadge", {
  enumerable: true,
  get: function () {
    return _MenuTitleWithNewBadge.MenuTitleWithNewBadge;
  }
});
Object.defineProperty(exports, "ToolBoxLayoutMode", {
  enumerable: true,
  get: function () {
    return _interface.ToolBoxLayoutMode;
  }
});
Object.defineProperty(exports, "ToolbarMode", {
  enumerable: true,
  get: function () {
    return _interface.ToolbarMode;
  }
});
Object.defineProperty(exports, "Toolbar", {
  enumerable: true,
  get: function () {
    return _Toolbar.Toolbar;
  }
});
Object.defineProperty(exports, "InlineToolbar", {
  enumerable: true,
  get: function () {
    return _inlineToolbar.InlineToolbar;
  }
});
Object.defineProperty(exports, "ToolbarConfig", {
  enumerable: true,
  get: function () {
    return _config.default;
  }
});
Object.defineProperty(exports, "MenuBar", {
  enumerable: true,
  get: function () {
    return _menubar.default;
  }
});
Object.defineProperty(exports, "Menu", {
  enumerable: true,
  get: function () {
    return _menu.default;
  }
});
Object.defineProperty(exports, "TD", {
  enumerable: true,
  get: function () {
    return _Td.TD;
  }
});
Object.defineProperty(exports, "TYPE_GROUP", {
  enumerable: true,
  get: function () {
    return _interface2.TYPE_GROUP;
  }
});
Object.defineProperty(exports, "TYPE_SUB_MENU", {
  enumerable: true,
  get: function () {
    return _interface2.TYPE_SUB_MENU;
  }
});
Object.defineProperty(exports, "TYPE_MENU_ITEM", {
  enumerable: true,
  get: function () {
    return _interface2.TYPE_MENU_ITEM;
  }
});
Object.defineProperty(exports, "TYPE_INLINE_MENU_GROUP", {
  enumerable: true,
  get: function () {
    return _interface2.TYPE_INLINE_MENU_GROUP;
  }
});
Object.defineProperty(exports, "TYPE_INLINE_MENU_WRAPPER", {
  enumerable: true,
  get: function () {
    return _interface2.TYPE_INLINE_MENU_WRAPPER;
  }
});
Object.defineProperty(exports, "TYPE_INLINE_MENU_ITEM", {
  enumerable: true,
  get: function () {
    return _interface2.TYPE_INLINE_MENU_ITEM;
  }
});
Object.defineProperty(exports, "TYPE_QUERY_ONLY_INLINE_MENU_ITEM", {
  enumerable: true,
  get: function () {
    return _interface2.TYPE_QUERY_ONLY_INLINE_MENU_ITEM;
  }
});
Object.defineProperty(exports, "NewToolbar", {
  enumerable: true,
  get: function () {
    return _NewToolbar.NewToolbar;
  }
});
Object.defineProperty(exports, "NewToolbarConfig", {
  enumerable: true,
  get: function () {
    return _NewToolbar.NewToolbarConfig;
  }
});
Object.defineProperty(exports, "NewToolbarMode", {
  enumerable: true,
  get: function () {
    return _NewToolbar.NewToolbarMode;
  }
});
Object.defineProperty(exports, "SimpleToolbar", {
  enumerable: true,
  get: function () {
    return _NewToolbar.SimpleToolbar;
  }
});
Object.defineProperty(exports, "AddLinkButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.AddLinkButton;
  }
});
Object.defineProperty(exports, "AlignCenterButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.AlignCenterButton;
  }
});
Object.defineProperty(exports, "AlignJustifyButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.AlignJustifyButton;
  }
});
Object.defineProperty(exports, "AlignLeftButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.AlignLeftButton;
  }
});
Object.defineProperty(exports, "AlignRightButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.AlignRightButton;
  }
});
Object.defineProperty(exports, "BaseColorPicker", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.BaseColorPicker;
  }
});
Object.defineProperty(exports, "BrushButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.BrushButton;
  }
});
Object.defineProperty(exports, "ColorPicker", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.ColorPicker;
  }
});
Object.defineProperty(exports, "ComplexColorIconMenu", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.ComplexColorIconMenu;
  }
});
Object.defineProperty(exports, "DeleteLinkButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.DeleteLinkButton;
  }
});
Object.defineProperty(exports, "FontColorPicker", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.FontColorPicker;
  }
});
Object.defineProperty(exports, "FontFamilySelect", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.FontFamilySelect;
  }
});
Object.defineProperty(exports, "FontSizeSelect", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.FontSizeSelect;
  }
});
Object.defineProperty(exports, "FontSizeSelectSingleLineDefaultWidth", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.FontSizeSelectSingleLineDefaultWidth;
  }
});
Object.defineProperty(exports, "FontSizeMaxSelect", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.FontSizeMaxSelect;
  }
});
Object.defineProperty(exports, "HeaderToolbarMode", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.HeaderToolbarMode;
  }
});
Object.defineProperty(exports, "HighlightColorPicker", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.HighlightColorPicker;
  }
});
Object.defineProperty(exports, "HorizontalAlignSelect", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.HorizontalAlignSelect;
  }
});
Object.defineProperty(exports, "ItemKey", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.ItemKey;
  }
});
Object.defineProperty(exports, "IconButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconButton;
  }
});
Object.defineProperty(exports, "ItemSlot", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.ItemSlot;
  }
});
Object.defineProperty(exports, "ItemState", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.ItemState;
  }
});
Object.defineProperty(exports, "IconSelectSingleLineDefaultWidth", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconSelectSingleLineDefaultWidth;
  }
});
Object.defineProperty(exports, "LineHeightSelect", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.LineHeightSelect;
  }
});
Object.defineProperty(exports, "ListSet", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.ListSet;
  }
});
Object.defineProperty(exports, "RedoButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.RedoButton;
  }
});
Object.defineProperty(exports, "SearchButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.SearchButton;
  }
});
Object.defineProperty(exports, "SlotType", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.SlotType;
  }
});
Object.defineProperty(exports, "SplitButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.SplitButton;
  }
});
Object.defineProperty(exports, "SubscriptButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.SubscriptButton;
  }
});
Object.defineProperty(exports, "SuperscriptButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.SuperscriptButton;
  }
});
Object.defineProperty(exports, "ScriptSplit", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.ScriptSplit;
  }
});
Object.defineProperty(exports, "TaskListButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.TaskListButton;
  }
});
Object.defineProperty(exports, "TextSelect", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.TextSelect;
  }
});
Object.defineProperty(exports, "IconMenu", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconMenu;
  }
});
Object.defineProperty(exports, "TextTypeSelect", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.TextTypeSelect;
  }
});
Object.defineProperty(exports, "TextTypeSingleLineDefaultWidth", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.TextTypeSingleLineDefaultWidth;
  }
});
Object.defineProperty(exports, "TableFillColorPicker", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.TableFillColorPicker;
  }
});
Object.defineProperty(exports, "UnderlineButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.UnderlineButton;
  }
});
Object.defineProperty(exports, "UndoButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.UndoButton;
  }
});
Object.defineProperty(exports, "BorderSet", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.BorderSet;
  }
});
Object.defineProperty(exports, "IconSelect", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconSelect;
  }
});
Object.defineProperty(exports, "TableMergeButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.TableMergeButton;
  }
});
Object.defineProperty(exports, "TableThemeButton", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.TableThemeButton;
  }
});
Object.defineProperty(exports, "VerticalAlignSelect", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.VerticalAlignSelect;
  }
});
Object.defineProperty(exports, "defaultSingleLayout", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.defaultSingleLayout;
  }
});
Object.defineProperty(exports, "defaultDisableTargets", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.defaultDisableTargets;
  }
});
Object.defineProperty(exports, "defaultDoubleLayout", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.defaultDoubleLayout;
  }
});
Object.defineProperty(exports, "HeadingToolbar", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.HeadingToolbar;
  }
});
Object.defineProperty(exports, "TranslationSelect", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.TranslationSelect;
  }
});
Object.defineProperty(exports, "COLORS", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.COLORS;
  }
});
Object.defineProperty(exports, "IconPalette", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconPalette;
  }
});
Object.defineProperty(exports, "IconSetting", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconSetting;
  }
});
Object.defineProperty(exports, "IconCutPic", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconCutPic;
  }
});
Object.defineProperty(exports, "IconRotateLeftLineThin", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconRotateLeftLineThin;
  }
});
Object.defineProperty(exports, "IconEdit", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconEdit;
  }
});
Object.defineProperty(exports, "IconLineType", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconLineType;
  }
});
Object.defineProperty(exports, "IconLineWeight", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconLineWeight;
  }
});
Object.defineProperty(exports, "IconAlignCenter", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconAlignCenter;
  }
});
Object.defineProperty(exports, "IconAlignLeft", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconAlignLeft;
  }
});
Object.defineProperty(exports, "IconAlignRight", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconAlignRight;
  }
});
Object.defineProperty(exports, "IconAlignJustify", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconAlignJustify;
  }
});
Object.defineProperty(exports, "IconEraser", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconEraser;
  }
});
Object.defineProperty(exports, "IconAutoWrap", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconAutoWrap;
  }
});
Object.defineProperty(exports, "IconOverflow", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconOverflow;
  }
});
Object.defineProperty(exports, "IconCode", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconCode;
  }
});
Object.defineProperty(exports, "IconFontColor", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconFontColor;
  }
});
Object.defineProperty(exports, "IconComment", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconComment;
  }
});
Object.defineProperty(exports, "IconBold", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconBold;
  }
});
Object.defineProperty(exports, "IconItalic", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconItalic;
  }
});
Object.defineProperty(exports, "IconStrikethrough", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconStrikethrough;
  }
});
Object.defineProperty(exports, "IconBrush", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconBrush;
  }
});
Object.defineProperty(exports, "IconPen", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconPen;
  }
});
Object.defineProperty(exports, "IconLineHeight", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconLineHeight;
  }
});
Object.defineProperty(exports, "IconLink", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconLink;
  }
});
Object.defineProperty(exports, "IconDeleteLink", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconDeleteLink;
  }
});
Object.defineProperty(exports, "IconOrderList", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconOrderList;
  }
});
Object.defineProperty(exports, "IconUnorderList", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconUnorderList;
  }
});
Object.defineProperty(exports, "IconUndo", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconUndo;
  }
});
Object.defineProperty(exports, "IconRedo", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconRedo;
  }
});
Object.defineProperty(exports, "IconSearch", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconSearch;
  }
});
Object.defineProperty(exports, "IconTodoList", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconTodoList;
  }
});
Object.defineProperty(exports, "IconUnderline", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconUnderline;
  }
});
Object.defineProperty(exports, "IconSubscript", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconSubscript;
  }
});
Object.defineProperty(exports, "IconSuperscript", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconSuperscript;
  }
});
Object.defineProperty(exports, "IconMore", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconMore;
  }
});
Object.defineProperty(exports, "IconCopy", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconCopy;
  }
});
Object.defineProperty(exports, "IconCut", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconCut;
  }
});
Object.defineProperty(exports, "IconPaste", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconPaste;
  }
});
Object.defineProperty(exports, "IconDelete", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconDelete;
  }
});
Object.defineProperty(exports, "IconDownload", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconDownload;
  }
});
Object.defineProperty(exports, "IconPicture", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconPicture;
  }
});
Object.defineProperty(exports, "IconReduceIndent", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconReduceIndent;
  }
});
Object.defineProperty(exports, "IconIncreaseIndent", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconIncreaseIndent;
  }
});
Object.defineProperty(exports, "IconMention", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconMention;
  }
});
Object.defineProperty(exports, "IconEmoji", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconEmoji;
  }
});
Object.defineProperty(exports, "IconFont", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconFont;
  }
});
Object.defineProperty(exports, "IconReduceFont", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconReduceFont;
  }
});
Object.defineProperty(exports, "IconTable", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconTable;
  }
});
Object.defineProperty(exports, "IconTranslate", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconTranslate;
  }
});
Object.defineProperty(exports, "IconMergeCell", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconMergeCell;
  }
});
Object.defineProperty(exports, "IconSplitCell", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconSplitCell;
  }
});
Object.defineProperty(exports, "IconCellFillColor", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconCellFillColor;
  }
});
Object.defineProperty(exports, "IconVerticalAlignTop", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconVerticalAlignTop;
  }
});
Object.defineProperty(exports, "IconVerticalAlignMiddle", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconVerticalAlignMiddle;
  }
});
Object.defineProperty(exports, "IconVerticalAlignBottom", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconVerticalAlignBottom;
  }
});
Object.defineProperty(exports, "IconBorderAll", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconBorderAll;
  }
});
Object.defineProperty(exports, "IconBorderBottom", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconBorderBottom;
  }
});
Object.defineProperty(exports, "IconBorderInner", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconBorderInner;
  }
});
Object.defineProperty(exports, "IconBorderInnerHorizon", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconBorderInnerHorizon;
  }
});
Object.defineProperty(exports, "IconBorderInnerVertical", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconBorderInnerVertical;
  }
});
Object.defineProperty(exports, "IconBorderLeft", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconBorderLeft;
  }
});
Object.defineProperty(exports, "IconBorderNone", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconBorderNone;
  }
});
Object.defineProperty(exports, "IconBorderOuter", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconBorderOuter;
  }
});
Object.defineProperty(exports, "IconBorderRight", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconBorderRight;
  }
});
Object.defineProperty(exports, "IconBorderTop", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconBorderTop;
  }
});
Object.defineProperty(exports, "IconSquareBrackets", {
  enumerable: true,
  get: function () {
    return _ToolbarNext.IconSquareBrackets;
  }
});
Object.defineProperty(exports, "hoverCSS", {
  enumerable: true,
  get: function () {
    return _HoverBlock.hoverCSS;
  }
});
Object.defineProperty(exports, "useHover", {
  enumerable: true,
  get: function () {
    return _HoverBlock.useHover;
  }
});
Object.defineProperty(exports, "ActiveInteractionProvider", {
  enumerable: true,
  get: function () {
    return _HoverBlock.ActiveInteractionProvider;
  }
});
Object.defineProperty(exports, "ActiveInteractionHooks", {
  enumerable: true,
  get: function () {
    return _HoverBlock.ActiveInteractionHooks;
  }
});
Object.defineProperty(exports, "HoverToolbarContexts", {
  enumerable: true,
  get: function () {
    return _HoverToolbar.HoverToolbarContexts;
  }
});
Object.defineProperty(exports, "HoverToolbarHooks", {
  enumerable: true,
  get: function () {
    return _HoverToolbar.HoverToolbarHooks;
  }
});
Object.defineProperty(exports, "MobileIconComment", {
  enumerable: true,
  get: function () {
    return _MobileIcons.MobileIconComment;
  }
});
Object.defineProperty(exports, "MobileIconCopy", {
  enumerable: true,
  get: function () {
    return _MobileIcons.MobileIconCopy;
  }
});
Object.defineProperty(exports, "MobileIconCheck", {
  enumerable: true,
  get: function () {
    return _MobileIcons.MobileIconCheck;
  }
});
Object.defineProperty(exports, "MobileIconCut", {
  enumerable: true,
  get: function () {
    return _MobileIcons.MobileIconCut;
  }
});
Object.defineProperty(exports, "MobileIconDelete", {
  enumerable: true,
  get: function () {
    return _MobileIcons.MobileIconDelete;
  }
});
Object.defineProperty(exports, "MobileIconBack", {
  enumerable: true,
  get: function () {
    return _MobileIcons.MobileIconBack;
  }
});
Object.defineProperty(exports, "IconButtonShortcut", {
  enumerable: true,
  get: function () {
    return _OldIconButtonShortcut.IconButtonShortcut;
  }
});
Object.defineProperty(exports, "MobileToolbar", {
  enumerable: true,
  get: function () {
    return _index3.MobileToolbar;
  }
});
Object.defineProperty(exports, "mobileLayout", {
  enumerable: true,
  get: function () {
    return _index3.mobileLayout;
  }
});
Object.defineProperty(exports, "MobileToolbarConstant", {
  enumerable: true,
  get: function () {
    return _index3.MobileToolbarConstant;
  }
});
Object.defineProperty(exports, "SelectionBar", {
  enumerable: true,
  get: function () {
    return _SelectionBar.SelectionBar;
  }
});
Object.defineProperty(exports, "SELECTION_BAR_HEIGHT", {
  enumerable: true,
  get: function () {
    return _SelectionBar.SELECTION_BAR_HEIGHT;
  }
});
Object.defineProperty(exports, "useSelectionBarContext", {
  enumerable: true,
  get: function () {
    return _SelectionBar.useSelectionBarContext;
  }
});
Object.defineProperty(exports, "SelectionBarProvider", {
  enumerable: true,
  get: function () {
    return _SelectionBar.SelectionBarProvider;
  }
});
Object.defineProperty(exports, "MOUNT_SELECTION_TOOLBAR", {
  enumerable: true,
  get: function () {
    return _SelectionBar.MOUNT_SELECTION_TOOLBAR;
  }
});
Object.defineProperty(exports, "Trigger", {
  enumerable: true,
  get: function () {
    return _SelectionBar.Trigger;
  }
});
Object.defineProperty(exports, "BasicPortal", {
  enumerable: true,
  get: function () {
    return _SelectionBar.BasicPortal;
  }
});
Object.defineProperty(exports, "LeftToolbar", {
  enumerable: true,
  get: function () {
    return _leftToolbar.LeftToolbar;
  }
});
Object.defineProperty(exports, "useLeftToolbar", {
  enumerable: true,
  get: function () {
    return _leftToolbar.useLeftToolbar;
  }
});
Object.defineProperty(exports, "getParentByAttributes", {
  enumerable: true,
  get: function () {
    return _leftToolbar.getParentByAttributes;
  }
});
Object.defineProperty(exports, "useDefaultSz", {
  enumerable: true,
  get: function () {
    return _useDefaultSz.default;
  }
});
Object.defineProperty(exports, "getDocDefaultSz", {
  enumerable: true,
  get: function () {
    return _useDefaultSz.getDocDefaultSz;
  }
});
Object.defineProperty(exports, "useDelayedState", {
  enumerable: true,
  get: function () {
    return _useDelayedState.default;
  }
});
Object.defineProperty(exports, "useOnClickOutside", {
  enumerable: true,
  get: function () {
    return _useOnClickOutside.default;
  }
});

var _idleRender = require("./idleRender");

var _SimpleLazyLoad = _interopRequireDefault(require("./idleRender/SimpleLazyLoad"));

var _index = require("./Portal/index");

var _ErrorBoundary = require("./ErrorBoundary");

var _inlinePlaceholder = require("./Placeholder/inlinePlaceholder");

var _bambooPlaceholder = require("./Placeholder/bambooPlaceholder");

var _placeholder = require("./Placeholder/placeholder");

var _index2 = require("./Sidebar/index");

var _button = _interopRequireDefault(require("./QuickInsert/button"));

var _commonPanel = _interopRequireDefault(require("./QuickInsert/commonPanel"));

var _Resizable = _interopRequireDefault(require("./Resizable"));

var _PluginsContext = require("./PluginsContext");

var _MenuTitleWithNewBadge = require("./MenuTitleWithNewBadge");

var _interface = require("./Toolbar/interface");

var _Toolbar = require("./Toolbar");

var _inlineToolbar = require("./Toolbar/inlineToolbar");

var _config = _interopRequireDefault(require("./Toolbar/config"));

var _menubar = _interopRequireDefault(require("./Menu/menubar"));

var _menu = _interopRequireDefault(require("./Menu/menu"));

var _Td = require("./Td");

var _interface2 = require("./Menu/interface");

var _NewToolbar = require("./NewToolbar");

var _ToolbarNext = require("./ToolbarNext");

var _HoverBlock = require("./HoverBlock");

var _HoverToolbar = require("./HoverToolbar");

var _MobileIcons = require("./MobileIcons");

var _OldIconButtonShortcut = require("./OldIconButtonShortcut");

var _index3 = require("./MobileToolbar/index");

var _SelectionBar = require("./SelectionBar");

var _leftToolbar = require("./leftToolbar");

var _useDefaultSz = _interopRequireWildcard(require("./hooks/useDefaultSz"));

var _useDelayedState = _interopRequireDefault(require("./hooks/useDelayedState"));

var _useOnClickOutside = _interopRequireDefault(require("./hooks/useOnClickOutside"));
