// @ali/4ever-plugin-group-factory/dist-node/utils/bamboo.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createRenderMark", {
  enumerable: true,
  get: function () {
    return _everPluginHeadingGroup.createRenderMark;
  }
});
Object.defineProperty(exports, "createCommentShow", {
  enumerable: true,
  get: function () {
    return _everPluginHeadingGroup.createCommentShow;
  }
});
Object.defineProperty(exports, "createAutoUnfold", {
  enumerable: true,
  get: function () {
    return _everPluginHeadingGroup.createAutoUnfold;
  }
});
Object.defineProperty(exports, "createOnCangjieSelect", {
  enumerable: true,
  get: function () {
    return _createOnCangjieSelect.default;
  }
});
Object.defineProperty(exports, "createRenderNode", {
  enumerable: true,
  get: function () {
    return _createRenderNode.default;
  }
});
Object.defineProperty(exports, "createShouldComponentUpdate", {
  enumerable: true,
  get: function () {
    return _createShouldComponentUpdate.default;
  }
});
Object.defineProperty(exports, "createOnKeyDown", {
  enumerable: true,
  get: function () {
    return _createOnKeyDown.default;
  }
});
Object.defineProperty(exports, "Group", {
  enumerable: true,
  get: function () {
    return _group.default;
  }
});
Object.defineProperty(exports, "GroupNew", {
  enumerable: true,
  get: function () {
    return _groupNew.default;
  }
});
Object.defineProperty(exports, "isSelectionInFoldArea", {
  enumerable: true,
  get: function () {
    return _helpers.isSelectionInFoldArea;
  }
});
exports.actions = void 0;

var _everPluginHeadingGroup = require("@ali/4ever-plugin-heading-group");

var _createOnCangjieSelect = _interopRequireDefault(require("./handlers/createOnCangjieSelect"));

var _createRenderNode = _interopRequireDefault(require("./handlers/createRenderNode"));

var _createShouldComponentUpdate = _interopRequireDefault(require("./handlers/createShouldComponentUpdate"));

var _createOnKeyDown = _interopRequireDefault(require("./handlers/createOnKeyDown"));

var _group = _interopRequireDefault(require("./components/group"));

var _groupNew = _interopRequireDefault(require("./components/groupNew"));

var actions = _interopRequireWildcard(require("./actions"));

exports.actions = actions;

var _helpers = require("./helpers");
