// @ali/4ever-utils/dist-node/hr/mapping.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseJsonMLToValue = exports.parseValueToJsonML = exports.MAPPING = void 0;
// 处理字段映射相关
const MAPPING = [{
  jsonML: 'width',
  value: 'length'
}, {
  jsonML: 'type',
  value: 'lineType'
}, {
  jsonML: 'sz',
  value: 'size'
}, {
  jsonML: 'color',
  value: 'color'
}];
exports.MAPPING = MAPPING;

/**
 *
 * @param data slate value data
 * @param map 映射对象
 */
const parseValueToJsonML = (data, map) => {
  const newObj = {
    width: null,
    type: null,
    sz: null,
    color: null
  };
  if (!data) return newObj;
  map.forEach(item => {
    newObj[item.jsonML] = data[item.value];
  });
  return newObj;
};
/**
 *
 * @param data jsonML data
 * @param map  映射对象
 */


exports.parseValueToJsonML = parseValueToJsonML;

const parseJsonMLToValue = (data, map) => {
  const newObj = {
    length: null,
    lineType: null,
    size: null,
    color: null
  };
  if (!data) return newObj;
  map.forEach(item => {
    newObj[item.value] = data[item.jsonML];
  });
  return newObj;
};

exports.parseJsonMLToValue = parseJsonMLToValue;
