// @ali/4ever-zhi/dist-node/plugins/image/mobile/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createImagePlugin;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _lodash = require("lodash");

var _everBamboo = require("@ali/4ever-bamboo");

var _createRenderEditable = _interopRequireDefault(require("../createRenderEditable"));

var _everPluginImage = require("@ali/4ever-plugin-image");

const _createElement = /*#__PURE__*/_react.default.createElement;

function createImagePlugin(config) {
  const imageConfig = config.image || {};
  return { ...(0, _everPluginImage.moFactory)({}),

    renderNode(props, controller, next) {
      const {
        node
      } = props; // @ts-ignore

      if (node.type !== 'image') {
        return next();
      }

      if ((0, _lodash.isEmpty)(node.data)) {
        return /*#__PURE__*/_createElement(_everBamboo.ImagePlugin.InlinePlaceholder, null, imageConfig?.locale?.placeholder || 'Image');
      }

      const {
        attributes
      } = props;
      return /*#__PURE__*/_createElement(_everBamboo.ImagePlugin.MobileImage, (0, _extends2.default)({}, attributes, {
        // @ts-ignore
        node: node,
        controller: controller // @ts-ignore
        ,
        imageConfig: imageConfig,
        readonly: true
      }));
    },

    renderEditable: (0, _createRenderEditable.default)(imageConfig)
  };
}
