// @ali/we-design/dist-web/components/theme/components/button.js
export default {
  // 普通按钮
  buttonNormalHoverBgColor: 'rgba(17, 31, 44, 0.08)',
  buttonNormalHoverBorderColor: 'rgba(0, 106, 254, 0.48)',
  buttonNormalHoverTextColor: 'rgba(46,133,255,1)',
  buttonNormalActiveBgColor: 'rgba(17, 31, 44, 0.12)',
  // 激活状态
  buttonNormalClickBgColor: '#dee0e1',
  // 点击过程
  buttonNormalClickBgColor2: 'rgba(0, 106, 254, 0.12)',
  // 点击过程
  buttonNormalClickBorderColor: 'rgba(0, 106, 254, 0.48)',
  buttonNormalClickTextColor: 'rgba(0,79,198,1)',
  buttonNormalDisabledBorderColor: 'rgba(17, 31, 44, 0.08)',
  // buttonNormalDisabledBgColor: colorGray1, // 略
  // 警告按钮
  buttonDangerBgColor: '#FB602D',
  buttonDangerHoverBgColor: 'rgba(251, 96, 45, 0.88)',
  buttonDangerActiveBgColor: '#DE582C',
  buttonDangerDisabledBgColor: '#FDBFAB',
  // 幽灵按钮
  buttonGhostHoverBgColor: 'rgba(255, 255, 255, 0.12)',
  buttonGhostHoverBorderColor: 'rgba(255, 255, 255, 0.56)',
  buttonGhostActiveBorderColor: 'rgba(255, 255, 255, 0.56)',
  buttonGhostActiveBgColor: 'rgba(17, 31, 44, 0.12)',
  buttonGhostDisabledBorderColor: 'rgba(255, 255, 255, 0.5)',
  // 分割按钮
  splitButtonBgColor: 'rgba(17, 31, 44, 0.08)',
  splitButtonLeftActiveBgColor: 'rgba(17, 31, 44, 0.12)',
  splitButtonActiveBgColor: 'rgba(17, 31, 44, 0.16)',
  splitButtonHoverBgColor: 'rgba(17, 31, 44, 0.12)'
};
