// @ali/4ever-component/dist-web/ToolbarNext/basicComponents/buttons/fontSizeMaxSelect/TabBarTabsNode.js
/* eslint-disable max-lines-per-function */

/* eslint-disable @typescript-eslint/explicit-function-return-type */
import React, { useRef } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import warning from 'tiny-warning';
import classnames from 'classnames';
import { TabBarTab, TabBarWrapper } from "./style";

var TabBarTabsNode = function TabBarTabsNode(props) {
  var saveRef = props.saveRef,
      panes = props.panes,
      activeKey = props.activeKey,
      testid = props.testid;
  var rst = []; // 下面两个ref是为了缓存函数引用，根据key进行缓存
  // eslint-disable-next-line @typescript-eslint/no-explicit-any

  var onTabClicks = useRef({}); // eslint-disable-next-line @typescript-eslint/no-explicit-any

  React.Children.forEach(panes, function (child) {
    if (!child) {
      return;
    }

    var key = child.key;
    var isActived = activeKey === key;

    if (!onTabClicks.current[key] || onTabClicks.current[key] !== props.onTabClick) {
      onTabClicks.current[key] = function (e) {
        if (props.onTabClick) {
          props.onTabClick(key, e);
        }
      };
    }

    warning('tab' in child.props, 'There must be `tab` property on children of Tabs.');
    var id = props.id ? key + "-" + props.id : key;

    var node = /*#__PURE__*/_createElement(TabBarTab, {
      role: "tab",
      "aria-disabled": child.props.disabled ? 'true' : 'false',
      "aria-selected": isActived ? 'true' : 'false',
      "aria-controls": "tabpane-" + id,
      tabIndex: isActived ? 0 : -1,
      key: key,
      id: "tab-" + id,
      className: classnames({
        disabled: !!child.props.disabled,
        actived: isActived
      }),
      ref: isActived ? saveRef('activeTab') : saveRef("tab_" + key),
      onClick: !isActived && !child.props.disabled ? onTabClicks.current[key] : undefined,
      "data-testid": testid + "_" + id
    }, child.props.tab);

    rst.push(node);
  });
  return /*#__PURE__*/_createElement(TabBarWrapper, null, rst);
};

export default TabBarTabsNode;
