// @ali/4ever-component/dist-web/ToolbarNext/basicComponents/buttons/fontSizeMaxSelect/Tabs.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { TabsWrap } from "./style";
import { getDataAttr, omit } from "./utils";
import TabBar from "./TabBar";

function getDefaultActiveKey(props) {
  var activeKey;
  React.Children.forEach(props.children, function (child) {
    if (child && !activeKey && !child.props.disabled) {
      activeKey = child.key;
    }
  });
  return activeKey;
}

function activeKeyIsValid(props, key) {
  var keys = React.Children.map(props.children, function (child) {
    return child && child.key;
  });
  return keys.indexOf(key) >= 0;
}

export var Tabs = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(Tabs, _React$Component);

  function Tabs(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;

    _this.onTabClick = function (activeKey, e) {
      var onTabClick = _this.props.onTabClick;

      if (onTabClick) {
        onTabClick(activeKey, e);
      }

      _this.setActiveKey(activeKey);
    };

    _this.onScroll = function (_ref) {
      var target = _ref.target,
          currentTarget = _ref.currentTarget;

      if (target === currentTarget && target.scrollLeft > 0) {
        target.scrollLeft = 0;
      }
    };

    _this.setActiveKey = function (nextActiveKey) {
      var _this$props = _this.props,
          activeKey = _this$props.activeKey,
          onChange = _this$props.onChange;

      if (activeKey !== nextActiveKey) {
        if (!activeKey) {
          _this.setState({
            activeKey: nextActiveKey
          });
        }

        if (onChange) {
          onChange(nextActiveKey);
        }
      }
    };

    _this.getNextActiveKey = function (next) {
      var activeKey = _this.state.activeKey;
      var children = [];
      React.Children.forEach(_this.props.children, function (c) {
        if (c && !c.props.disabled) {
          if (next) {
            children.push(c);
          } else {
            children.unshift(c);
          }
        }
      });
      var length = children.length;
      var ret = length && children[0].key;
      children.forEach(function (child, i) {
        if (child.key === activeKey) {
          if (i === length - 1) {
            ret = children[0].key;
          } else {
            ret = children[i + 1].key;
          }
        }
      });
      return ret;
    };

    var _activeKey;

    if (props.activeKey) {
      _activeKey = props.activeKey;
    } else if (props.defaultActiveKey) {
      _activeKey = props.defaultActiveKey;
    } else {
      _activeKey = getDefaultActiveKey(props);
    }

    _this.state = {
      activeKey: _activeKey
    };
    return _this;
  }

  Tabs.getDerivedStateFromProps = function getDerivedStateFromProps(props, state) {
    var newState = {};

    if (props.activeKey) {
      newState.activeKey = props.activeKey;
    } else if (!activeKeyIsValid(props, state.activeKey)) {
      newState.activeKey = getDefaultActiveKey(props);
    }

    if (Object.keys(newState).length > 0) {
      return newState;
    }

    return null;
  };

  var _proto = Tabs.prototype;

  _proto.render = function render() {
    var props = this.props,
        state = this.state;

    var children = props.children,
        restProps = _objectWithoutPropertiesLoose(props, ["children"]);

    return /*#__PURE__*/_createElement(TabsWrap, _extends({
      style: props.style
    }, getDataAttr(restProps), {
      onScroll: this.onScroll
    }), /*#__PURE__*/_createElement(TabBar, _extends({}, omit(props, ['className']), {
      panes: children,
      style: props.style,
      activeKey: state.activeKey,
      onTabClick: this.onTabClick,
      testid: props.testid
    })));
  };

  return Tabs;
}(React.Component);
Tabs.defaultProps = {
  onChange: function onChange() {},
  children: null,
  style: {},
  destroyInactiveTabPane: false
};
