// @ali/4ever-plugin-mention/dist-node/bi/components/mentionButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = MentionButton;

var React = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _isSelectionInMention = _interopRequireDefault(require("../utils/isSelectionInMention"));

var _actions = require("../actions");

var _role = require("../../utils/role");

const _createElement = /*#__PURE__*/React.createElement;

const isDisabled = controller => {
  const {
    value
  } = controller;
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const isDisabled = isCollapsedAtEdge || (0, _isSelectionInMention.default)(value) || controller.query('isSelectionInLink') || controller.query('isSelectionInImage') || controller.query('isSelectionInLinkCard') || controller.query('isSelectionInHr');
  return Boolean(isDisabled);
};

var _ref = /*#__PURE__*/_createElement(_weDesign.MentionBetaNormal, null);

function MentionButton(props) {
  const {
    controller,
    locale
  } = props;

  const handleClick = () => controller.run('onAction', (0, _actions.activeMention)());

  return /*#__PURE__*/_createElement(_weDesign.Button, {
    title: locale.buttonTitle,
    bordered: false,
    testid: "bi-toolbar-mention",
    icon: _ref,
    disabled: isDisabled(controller),
    onClick: handleClick,
    role: _role.MentionRole.mention
  });
}
