// @ali/4ever-plugin-mention/dist-node/bi/components/mentionButtonMobile.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MobileMentionButton = MobileMentionButton;
exports.MobileMentionInsertButton = MobileMentionInsertButton;

var React = _interopRequireWildcard(require("react"));

var _actions = require("../actions");

const _createElement = /*#__PURE__*/React.createElement;

const isDisabled = controller => {
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  return isCollapsedAtEdge;
};

function MobileMentionButton(props) {
  const {
    controller,
    Wrapper,
    item
  } = props;

  const handleClick = () => controller.run('onAction', (0, _actions.selectMention)());

  return /*#__PURE__*/_createElement(Wrapper, {
    type: "big",
    icon: "at",
    item: item,
    isDisabled: isDisabled(controller),
    onClick: handleClick
  });
}

function MobileMentionInsertButton(props) {
  const {
    controller,
    Wrapper,
    item,
    locale
  } = props;

  const handleClick = () => controller.run('onAction', (0, _actions.selectMention)());

  return /*#__PURE__*/_createElement(Wrapper, {
    type: "line",
    icon: "at",
    item: item,
    isDisabled: isDisabled(controller),
    lineText: locale?.buttonTitle,
    onClick: handleClick
  });
}
