// @ali/4ever-plugin-mention/dist-node/bi/components/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PlaceholderIcon = exports.PlaceholderTextWrapper = exports.InputSpanWrapper = exports.InputWrapper = exports.InsertWrapper = exports.PlaceholderBtnWrapper = exports.PlaceholderInputWrapper = exports.LoadingWrapper = exports.SuggestionsWrapper = exports.ItemName = exports.ItemText = exports.Avatar = exports.MentionItemContainer = exports.MentionItem = exports.Toolbar = exports.MentionAt = exports.Mention = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _weDesign = require("@ali/we-design");

const Mention = /*#__PURE__*/_styledComponents.default.div(["position:relative;padding:0;margin:0;width:", ";max-height:", ";height:100%;overflow:auto;div:first-child{overflow-y:auto;}"], props => props.theme.suggestionWidth, props => props.theme.suggestionMaxHeight);

exports.Mention = Mention;

const MentionAt = /*#__PURE__*/_styledComponents.default.span(["position:relative;[data-cangjie-leaf]{color:", ";}"], props => props.active ? _weDesign.theme.primaryColor : 'inherit');

exports.MentionAt = MentionAt;

const Toolbar = /*#__PURE__*/_styledComponents.default.div(["position:absolute;opacity:", ";visibility:", ";min-width:200px;line-height:28px;display:flex;flex-direction:row;font-size:14px;font-weight:normal;text-indent:0;top:auto;bottom:-46px;left:-1px;right:auto;background:", ";border-radius:8px;height:40px;border:1px solid ", ";box-shadow:", ";padding:5px 3px;z-index:1;padding:0 0;height:auto;bottom:auto;overflow:auto;"], props => props.active ? 1 : 0, props => props.active ? 'visible' : 'hidden', _weDesign.theme.colorGray1, _weDesign.theme.borderColor, _weDesign.theme.boxShadow);

exports.Toolbar = Toolbar;

const MentionItem = /*#__PURE__*/_styledComponents.default.div(["padding:8px 8px 6px 8px;cursor:pointer;background:", ";"], props => props.actived ? _weDesign.theme.colorGray3 : '');

exports.MentionItem = MentionItem;

const MentionItemContainer = /*#__PURE__*/_styledComponents.default.div(["display:flex;"]);

exports.MentionItemContainer = MentionItemContainer;

const Avatar = /*#__PURE__*/_styledComponents.default.div(["width:36px;height:36px;display:block;padding:3px 3px;img{width:100%;height:100%;border-radius:18px;}"]);

exports.Avatar = Avatar;

const ItemText = /*#__PURE__*/_styledComponents.default.div(["flex:1;display:block;padding-left:8px;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;line-height:36px;"]);

exports.ItemText = ItemText;

const ItemName = /*#__PURE__*/_styledComponents.default.span(["color:", ";"], _weDesign.theme.colorText4);

exports.ItemName = ItemName;

const SuggestionsWrapper = /*#__PURE__*/_styledComponents.default.div(["min-height:323px;height:100%;width:", ";max-height:", ";display:flex;flex-direction:column;& .mention-suggestion-list{::-webkit-scrollbar{background-color:transparent;width:8px;}::-webkit-scrollbar-thumb{background:rgba(194,194,194,1);border-radius:10px;}::-webkit-scrollbar-thumb:hover{background:rgba(0,0,0,0.45);}::-webkit-scrollbar-track{background:transparent;}}"], ({
  theme: {
    suggestionWidth
  }
}) => typeof suggestionWidth === 'string' ? suggestionWidth : `${suggestionWidth}px`, ({
  theme: {
    suggestionMaxHeight
  }
}) => typeof suggestionMaxHeight === 'string' ? suggestionMaxHeight : `${suggestionMaxHeight}px`);

exports.SuggestionsWrapper = SuggestionsWrapper;

const LoadingWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;justify-content:center;align-items:center;height:", "px;width:", "px;color:", ";font-size:36px;"], props => props.height, props => props.width, _weDesign.theme.colorBlue1);

exports.LoadingWrapper = LoadingWrapper;

const PlaceholderInputWrapper = /*#__PURE__*/_styledComponents.default.div(["user-select:none;"]);

exports.PlaceholderInputWrapper = PlaceholderInputWrapper;

const PlaceholderBtnWrapper = /*#__PURE__*/_styledComponents.default.div(["color:rgba(23,26,29,0.4);padding:3px;border-radius:6px;cursor:pointer;background:", ";&:hover{background:rgba(126,134,142,0.12);}&:active{background:rgba(126,134,142,0.16);}"], props => props.isSelected ? _weDesign.theme.colorGray4 : '');

exports.PlaceholderBtnWrapper = PlaceholderBtnWrapper;

const InsertWrapper = /*#__PURE__*/_styledComponents.default.div(["position:absolute;display:flex;flex-direction:column;background-color:#ffffff;border:1px solid ", ";box-shadow:", ";border-radius:6px;height:375px;width:335px;"], _weDesign.theme.colorLine1, _weDesign.theme.boxShadow);

exports.InsertWrapper = InsertWrapper;

const InputWrapper = /*#__PURE__*/_styledComponents.default.div(["user-select:none;max-width:400px;overflow-x:scroll;&::-webkit-scrollbar{display:none;}& .inline-placeholder{text-overflow:ellipsis;white-space:nowrap;display:inline-block;}"]);

exports.InputWrapper = InputWrapper;

const InputSpanWrapper = /*#__PURE__*/_styledComponents.default.span(["border:none;outline:none;caret-color:#0089FF;color:#0089FF;user-select:text;"]);

exports.InputSpanWrapper = InputSpanWrapper;

const PlaceholderTextWrapper = /*#__PURE__*/_styledComponents.default.span(["color:rgba(23,26,29,0.4);"]);

exports.PlaceholderTextWrapper = PlaceholderTextWrapper;

const PlaceholderIcon = /*#__PURE__*/_styledComponents.default.span(["display:inline-block;color:#0089FF;user-select:text;border:none;outline:none;"]);

exports.PlaceholderIcon = PlaceholderIcon;
