// @ali/4ever-plugin-mention/dist-node/bi/components/suggestionItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _lodash = require("lodash");

var _styled = require("./styled");

const _createElement = /*#__PURE__*/React.createElement;

const defaultRender = suggestion => /*#__PURE__*/_createElement(_styled.MentionItemContainer, null, /*#__PURE__*/_createElement(_styled.Avatar, null, /*#__PURE__*/_createElement("img", {
  src: suggestion.avatarUrl,
  alt: ""
})), /*#__PURE__*/_createElement(_styled.ItemText, null, /*#__PURE__*/_createElement(_styled.ItemName, null, suggestion.name)));

const MentionSuggestionItem = props => {
  const {
    index,
    style = {},
    data
  } = props;
  const {
    suggestions,
    selectedIndex,
    renderItem = defaultRender,
    onSelect = _lodash.noop,
    onSelectedIndexChange
  } = data;
  const handleMouseEnter = React.useCallback(() => {
    onSelectedIndexChange(index);
  }, [index, onSelectedIndexChange]);
  const handleMouseDown = React.useCallback(event => {
    event.preventDefault();
    event.stopPropagation();
    onSelect(index);
  }, [index, onSelect]);
  const actived = selectedIndex === index;
  return /*#__PURE__*/_createElement(_styled.MentionItem, {
    className: `mention-suggestion-item ${actived ? 'actived' : ''}`,
    "data-testid": "mention-suggestion-item",
    "data-type": "mention-suggestion-item",
    style: style,
    onMouseDown: handleMouseDown,
    onMouseEnter: handleMouseEnter,
    actived: actived
  }, renderItem(suggestions[index]));
};

var _default = /*#__PURE__*/React.memo(MentionSuggestionItem);

exports.default = _default;
