// @ali/4ever-plugin-mention/dist-node/bi/components/suggestionForQuickInsert.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactWindow = require("react-window");

var _reactVirtualizedAutoSizer = _interopRequireDefault(require("react-virtualized-auto-sizer"));

var _lodash = require("lodash");

var _weDesign = require("@ali/we-design");

var _isHotkey = _interopRequireDefault(require("is-hotkey"));

var _actions = require("../actions");

var _suggestionItem = _interopRequireDefault(require("./suggestionItem"));

var _styled = require("./styled");

var _everComponent = require("@ali/4ever-component");

const _createElement = /*#__PURE__*/_react.default.createElement;
const isUpHotKey = (0, _isHotkey.default)('up');
const isDownHotKey = (0, _isHotkey.default)('down');
const isLeftHotKey = (0, _isHotkey.default)('left');
const isRightHotKey = (0, _isHotkey.default)('right');

var _ref = /*#__PURE__*/_createElement(_weDesign.LoadingBetaNormal, null);

const MentionSuggestion = props => {
  const {
    controller,
    text,
    searchSuggestions,
    renderItem,
    renderHeader,
    renderFooter,
    renderLoading,
    renderEmpty,
    offsetBottom = 0,
    loadMoreSuggestions,
    itemHeight = 50,
    at
  } = props;

  const [suggestions, setSuggestions] = _react.default.useState([]);

  const [errData, setErrData] = _react.default.useState({});

  const [selectedIndex, setSelectedIndex] = _react.default.useState(0);

  const searching = _react.default.useRef(false);

  const [delayedSearching, setDelayedSearching] = (0, _everComponent.useDelayedState)(false);

  const listRef = _react.default.useRef(null);

  const lastItem = _react.default.useRef(null);

  const editorRef = _react.default.useRef(controller);

  const isEmptyText = text.trim().length === 0;
  const isEmptySuggestions = suggestions.length === 0;

  const handleSelect = _react.default.useCallback(index => {
    const suggestion = suggestions[(0, _lodash.isNil)(index) ? selectedIndex : index];
    editorRef.current.run('onAction', (0, _actions.insertMention)(suggestion, at));
  }, [suggestions, selectedIndex]);

  const handleScroll = _react.default.useCallback(({
    scrollOffset
  }) => {
    if (!loadMoreSuggestions || !listRef.current) {
      return;
    }
    /* 当倒数第 offsetBottom + 1 个元素进入可视区域时触发 loadMoreSuggestions */


    const startOffset = (suggestions.length - offsetBottom - 2) * itemHeight - listRef.current.props.height;

    if (startOffset > 0 && startOffset <= scrollOffset && lastItem.current !== suggestions.length) {
      lastItem.current = suggestions.length;
      loadMoreSuggestions(text, suggestions.length, setSuggestions, setErrData);
    }
  }, [loadMoreSuggestions, suggestions.length, offsetBottom, itemHeight, text]);

  const itemData = _react.default.useMemo(() => ({
    suggestions,
    onSelect: handleSelect,
    renderItem,
    selectedIndex,
    onSelectedIndexChange: setSelectedIndex
  }), [suggestions, renderItem, selectedIndex, handleSelect]);

  _react.default.useEffect(() => {
    editorRef.current = controller;
  }, [controller]); // listen up & down


  _react.default.useEffect(() => {
    const handleMove = event => {
      if (event.isComposing || searching.current) {
        return;
      }

      if (isRightHotKey(event) || isDownHotKey(event)) {
        event.preventDefault();
        setSelectedIndex(index => (index + 1) % suggestions.length);
      }

      if (isLeftHotKey(event) || isUpHotKey(event)) {
        event.preventDefault();
        setSelectedIndex(index => index === 0 ? suggestions.length - 1 : index - 1);
      }
    };

    document.addEventListener('keydown', handleMove);
    return () => {
      document.removeEventListener('keydown', handleMove);
    };
  }, [suggestions]);

  _react.default.useEffect(() => {
    const handleKeyDown = event => {
      if (!event.key) {
        return;
      }

      const key = event.key.toLowerCase();

      if (event.isComposing || searching.current) {
        return;
      }

      if (key === 'enter') {
        event.preventDefault();
        event.stopPropagation();

        if (delayedSearching) {
          return;
        }

        handleSelect();
        return editorRef.current.run('onAction', (0, _actions.deactiveMention)());
      }

      if (key === ' ' && event.keyCode === 32) {
        if (!isEmptyText && !isEmptySuggestions) {
          event.preventDefault();
          event.stopPropagation();
          handleSelect(0);
        } else {
          return editorRef.current.run('onAction', (0, _actions.deactiveMention)());
        }
      }
    };

    document.addEventListener('keydown', handleKeyDown);
    return () => {
      document.removeEventListener('keydown', handleKeyDown);
    };
  }, [handleSelect, isEmptySuggestions, isEmptyText, delayedSearching]); // reset loading when search done


  _react.default.useEffect(() => {
    setDelayedSearching(false);
    searching.current = false;
  }, [setDelayedSearching, suggestions]); // do search when text changed


  _react.default.useEffect(() => {
    // @ts-ignore value 没有 isComposing 属性
    if (editorRef.current.value.isComposing) {
      return;
    }

    if (!searchSuggestions) {
      // eslint-disable-next-line no-console
      console.warn('Missing `searchSuggestions`');
      return;
    }

    setDelayedSearching(true, 200);
    searching.current = true;
    searchSuggestions(text, setSuggestions, setErrData);
  }, [text, searchSuggestions, setDelayedSearching]); // scroll to selected index


  _react.default.useEffect(() => {
    if (listRef.current) {
      listRef.current.scrollToItem(selectedIndex);
    }
  }, [selectedIndex]);

  return /*#__PURE__*/_createElement(_styled.SuggestionsWrapper, null, renderHeader && renderHeader(), /*#__PURE__*/_createElement("div", {
    style: {
      flex: 1
    }
  }, /*#__PURE__*/_createElement(_reactVirtualizedAutoSizer.default, null, ({
    height,
    width
  }) => delayedSearching ? /*#__PURE__*/_createElement(_styled.LoadingWrapper, {
    "data-testid": "mention-suggestion-loading",
    height: height,
    width: width
  }, renderLoading ? renderLoading() : _ref) : /*#__PURE__*/_createElement(_react.default.Fragment, null, errData.isServiceError ? /*#__PURE__*/_createElement(_weDesign.SpinErrorCard, {
    style: {
      width: `${width}px`,
      height: `${height}px`,
      border: 'none'
    },
    tip: errData.serviceErrorMsg
  }) : suggestions.length === 0 && renderEmpty && renderEmpty(), suggestions.length > 0 && /*#__PURE__*/_createElement(_reactWindow.FixedSizeList, {
    className: "mention-suggestion-list",
    ref: listRef,
    height: height,
    itemCount: suggestions.length,
    itemSize: itemHeight,
    itemData: itemData,
    width: width,
    onScroll: handleScroll
  }, _suggestionItem.default)))), renderFooter && renderFooter());
};

var _default = /*#__PURE__*/_react.default.memo(MentionSuggestion);

exports.default = _default;
