// @ali/4ever-bi/dist-web/plugins/video/mobile/createRenderNode.js
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import Video from "./video";
export default function createRenderNode(config) {
  var videoConfig = config.video || {};
  var _videoConfig$locale = videoConfig.locale,
      videoLocaleConfig = _videoConfig$locale === void 0 ? {} : _videoConfig$locale,
      downloadFile = videoConfig.downloadFile,
      onMediaLoadSuccess = videoConfig.onMediaLoadSuccess,
      onMediaLoadError = videoConfig.onMediaLoadError,
      transformVideoURL = videoConfig.transformVideoURL,
      setTempPosterUrl = videoConfig.setTempPosterUrl,
      renderVideoMask = videoConfig.renderVideoMask,
      allowDownload = videoConfig.allowDownload,
      onPreview = videoConfig.onPreview,
      allowFullscreen = videoConfig.allowFullscreen;
  return {
    embed: function embed(props, controller, next) {
      var node = props.node;

      if (node.data.type === 'video' && node.data.src) {
        return /*#__PURE__*/_createElement(Video, _extends({}, props, {
          controller: controller,
          onDownload: downloadFile,
          locale: videoLocaleConfig,
          onMediaLoadSuccess: onMediaLoadSuccess,
          onMediaLoadError: onMediaLoadError,
          transformVideoURL: transformVideoURL,
          setTempPosterUrl: setTempPosterUrl,
          renderVideoMask: renderVideoMask,
          allowDownload: allowDownload,
          onPreview: onPreview,
          allowFullscreen: allowFullscreen
        }));
      }

      return next();
    }
  };
}
