// @ali/4ever-bi/dist-web/plugins/video/mobile/video.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React, { PureComponent } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { EmbedPlugin, useScrollableContainer } from '@ali/4ever-bamboo';
import { EmbedInlineToolbar as VideoInlineToolbar } from '@ali/4ever-factory';
var MobileEmbedCard = EmbedPlugin.MobileEmbedCard,
    VideoPlayer = EmbedPlugin.VideoPlayer;

var Video = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Video, _PureComponent);

  function Video() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.isClicked = false;
    _this.ref = /*#__PURE__*/React.createRef();
    _this.previewRef = /*#__PURE__*/React.createRef();
    _this.interval = null;

    _this.getNodeData = function () {
      var node = _this.props.node;

      var _ref = node || {},
          data = _ref.data;

      return data;
    };

    _this.onMediaLoadError = function () {
      var _this$getNodeData = _this.getNodeData(),
          src = _this$getNodeData.src;

      var onMediaLoadError = _this.props.onMediaLoadError;

      if (onMediaLoadError) {
        onMediaLoadError(src);
      }
    };

    _this.onMediaLoadSuccess = function () {
      var _this$getNodeData2 = _this.getNodeData(),
          src = _this$getNodeData2.src;

      var onMediaLoadSuccess = _this.props.onMediaLoadSuccess;

      if (onMediaLoadSuccess) {
        onMediaLoadSuccess(src);
      }
    };

    _this.onDownloadFile = function () {
      var _this$getNodeData3 = _this.getNodeData(),
          src = _this$getNodeData3.src,
          name = _this$getNodeData3.name,
          type = _this$getNodeData3.type,
          size = _this$getNodeData3.size;

      var onDownload = _this.props.onDownload;

      if (onDownload) {
        onDownload(src, {
          name: name,
          type: type,
          size: size
        });
      }
    };

    return _this;
  }

  var _proto = Video.prototype;

  _proto.renderPreview = function renderPreview() {
    var _this$props = this.props,
        isSelected = _this$props.isSelected,
        node = _this$props.node,
        setTempPosterUrl = _this$props.setTempPosterUrl,
        onMediaLoadSuccess = _this$props.onMediaLoadSuccess,
        transformVideoURL = _this$props.transformVideoURL,
        renderVideoMask = _this$props.renderVideoMask,
        onDownload = _this$props.onDownload,
        allowDownload = _this$props.allowDownload,
        locale = _this$props.locale,
        allowFullscreen = _this$props.allowFullscreen;
    return /*#__PURE__*/_createElement(VideoPlayer, {
      transformVideoURL: transformVideoURL,
      onMediaLoadSuccess: onMediaLoadSuccess,
      onMediaLoadError: this.onMediaLoadError,
      setTempPosterUrl: setTempPosterUrl,
      node: node,
      isSelected: isSelected,
      renderVideoMask: renderVideoMask,
      onDownload: this.onDownloadFile,
      isMobile: true,
      allowDownload: onDownload && allowDownload,
      locale: locale,
      allowFullscreen: allowFullscreen
    });
  };

  _proto.renderCard = function renderCard() {
    var _this$getNodeData4 = this.getNodeData(),
        name = _this$getNodeData4.name,
        src = _this$getNodeData4.src,
        type = _this$getNodeData4.type,
        size = _this$getNodeData4.size;

    var onPreview = this.props.onPreview;
    return /*#__PURE__*/_createElement(MobileEmbedCard, {
      name: name,
      onPreview: onPreview,
      src: src,
      type: type,
      size: size
    });
  };

  _proto.renderInlineToolbar = function renderInlineToolbar() {
    var _this$props2 = this.props,
        controller = _this$props2.controller,
        node = _this$props2.node,
        locale = _this$props2.locale;
    return /*#__PURE__*/_createElement(VideoInlineToolbar, {
      controller: controller,
      node: node,
      locale: locale,
      triggerNode: this.ref.current
    });
  };

  _proto.render = function render() {
    var _this$getNodeData5 = this.getNodeData(),
        viewType = _this$getNodeData5.viewType;

    var _this$props3 = this.props,
        attributes = _this$props3.attributes,
        isSelected = _this$props3.isSelected;
    return /*#__PURE__*/_createElement(React.Fragment, null, isSelected ? this.renderInlineToolbar() : null, /*#__PURE__*/_createElement("div", _extends({}, attributes, {
      ref: this.ref
    }), viewType === 'preview' ? this.renderPreview() : this.renderCard()));
  };

  return Video;
}(PureComponent);

var Container = function Container(props) {
  var scrollableContainer = useScrollableContainer();
  return /*#__PURE__*/_createElement(Video, _extends({}, props, {
    scrollableContainer: scrollableContainer
  }));
};

export default Container;
