// @ali/4ever-bi/dist-node/plugins/comment/createRenderHoverToolbar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderHoverToolbar;

var React = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everComponent = require("@ali/4ever-component");

const _createElement = /*#__PURE__*/React.createElement;

var _ref = /*#__PURE__*/_createElement(_everComponent.IconComment, null);

function createRenderHoverToolbar(config) {
  const {
    locale = {},
    isDisabled
  } = config;
  return (props, controller, getNode) => {
    const {
      buttonName
    } = props;

    switch (buttonName) {
      case 'default':
        {
          const disabled = isDisabled && isDisabled();
          return !disabled ? {
            icon: _ref,
            title: locale?.addComment,
            tooltip: locale?.addComment,
            onClick: () => {
              const node = getNode();
              if (!node) return;
              controller.command('selectNode', node);
              controller.run('onAction', {
                type: _everBamboo.CommentPlugin.actions.ACTION_COMMENT_CLICK
              });
            },
            role: _everBamboo.PluginRoles.comment
          } : null;
        }

      default:
        return null;
    }
  };
}
