// @ali/4ever-bi/dist-node/plugins/comment/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everBamboo = require("@ali/4ever-bamboo");

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _createRenderToolbar = _interopRequireDefault(require("./createRenderToolbar"));

var _createRenderHoverToolbar = _interopRequireDefault(require("./createRenderHoverToolbar"));

const {
  hooks,
  types,
  actions,
  commands,
  queries
} = _everBamboo.CommentPlugin;
const {
  lazyRenderStrategy
} = queries;

function createCommentPlugin(allConfigs) {
  const {
    createContextMenu,
    createOnKeyDown,
    createOnMouseMove,
    createOnCangjieInput,
    createOnAction,
    createOnChange,
    createRenderMark,
    createRenderNode,
    createRenderContent,
    createRenderEditable,
    createOnCangjieSelect,
    createOnCangjieCopyOrCut
  } = hooks;
  const configs = allConfigs.comment || {};
  return {
    queries: {
      lazyRenderStrategy
    },
    contextMenu: createContextMenu(configs),
    onKeyDown: createOnKeyDown(configs),
    onMouseMove: createOnMouseMove(configs),
    onCangjieInput: createOnCangjieInput(configs),
    onAction: createOnAction(configs),
    onChange: createOnChange(configs),
    renderMark: createRenderMark(configs),
    renderNode: createRenderNode(configs),
    renderContent: createRenderContent(configs),
    renderEditable: createRenderEditable(configs),
    onCangjieSelect: createOnCangjieSelect(configs),
    onCangjieCopy: createOnCangjieCopyOrCut(),
    onCangjieCut: createOnCangjieCopyOrCut(),
    renderToolbar: (0, _createRenderToolbar.default)(configs.locale || {}),
    renderHoverToolbar: (0, _createRenderHoverToolbar.default)(configs),
    commands,
    toolbar: (0, _toolbar.default)(configs)
  };
}

createCommentPlugin.types = types;
createCommentPlugin.actions = actions;
createCommentPlugin.commands = commands;
createCommentPlugin.queries = queries;
var _default = createCommentPlugin;
exports.default = _default;
