// @ali/4ever-plugin-list/dist-web/utils/constant.js
// 罗马类型
var ROMAN_FORMAT = ['upperRoman', 'lowerRoman']; // 罗马项目符号额外增加缩进值

var ROMAN_EXTRA_INDENT = 20; // 列表项目符号到内容之间的距离

var SYMBOL_CONTENT_SPACING = 8; // 项目符号默认字体

var SYMBOL_FONT = 'Arial'; // 列表右侧的按钮区域宽度

var TASK_EXTRA_BUTTONS_WIDTH = 120; // 列表看不见的左侧 border 宽度

var TASK_BORDER_LEFT_WIDTH = 10; // 代办 checbox 宽度

var TASK_BOX_WIDTH = 16; // 代办列表的 padding

var TASK_PADDING = 2; // 项目符号默认大小

var SYMBOL_SZ = 11;
var SYMBOL_STYLE = {// sz: SYMBOL_SZ,
};
var STICKER_REG = /^\[(.+)\]$/; // ● ■ 不是居中的 unicode 符号，需要使用 magic number 调整

var SYMBOL_MAGIC_OFFSET_Y_MAP = {
  "\u25CF": 4.5,
  "\u25CB": 4.5,
  "\u25A0": 12,
  "\u2605": 4.5
};
var SYMBOL_MAGIC_OFFSET_X_MAP = {
  "\u25C6": 1,
  "\u2605": 1
};
var SYMBOL_MAGIC_SIZE_MAP = {
  "\u25A0": 2,
  "\u25C6": 6,
  "\u2605": 2
};
export { ROMAN_EXTRA_INDENT, ROMAN_FORMAT, SYMBOL_CONTENT_SPACING, SYMBOL_FONT, SYMBOL_STYLE, SYMBOL_SZ, TASK_EXTRA_BUTTONS_WIDTH, TASK_PADDING, TASK_BORDER_LEFT_WIDTH, TASK_BOX_WIDTH, STICKER_REG, SYMBOL_MAGIC_OFFSET_Y_MAP, SYMBOL_MAGIC_OFFSET_X_MAP, SYMBOL_MAGIC_SIZE_MAP };
