// @ali/4ever-plugin-list/dist-web/utils/listStyleMapping.js
/**
 * 无序命名规则
 * 例：SCIR_ECIR_SREC, 形状 + 形状 + 形状
 *
 * 首字母：
 * S -> solid (■)
 * E -> empty (❏)
 *
 * 形状：
 * ● -> CIR -> circle
 * ❏ -> REC -> rect
 * ❖ -> FDI -> four diomand
 * ◆ -> DIO -> diomand
 * ➔ -> ARR -> arrow
 * ➢ -> RIG -> right
 * */
export var symbol_text_microsoft = {
  Ø: "\u27A2",
  l: "\u25CF",
  u: "\u25C6",
  n: "\u25A0",
  '+': '+',
  '-': '-'
};
export var symbol_type_miscrosoft = {
  Ø: 'SRIG_ECIR_SREC',
  l: 'SCIR_ECIR_SREC',
  u: 'SARR_SDIO_SCIR',
  n: 'SCIR_ECIR_SREC',
  '+': 'MINUS_MINUS_MINUS',
  '-': 'PLUS_PLUS_PLUS'
};
export var DEFAULT_BULLET_LIST_ARRAY;
/**
 * 有序命名规则
 * 例： DEC_LEN_LROM_P ，符号类型 + 符号类型 + 符号类型 + 首个符号格式
 *
 * 符号类型：
 * DEC -> decimal
 * DDEC -> decimalZero
 * LEN -> lowerLetter
 * UEN -> upperLetter
 * LROM -> lowerRoman
 * UROM -> upperRoman
 * ZH -> chineseCountingThousand
 *
 * 符号格式：
 * P（point) -> xxx.
 * H（half)  -> xxx)
 * A（all)   -> (xxx)
 * D (don)   -> xxx、
 * */

(function (DEFAULT_BULLET_LIST_ARRAY) {
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["SCIR_ECIR_SREC"] = 0] = "SCIR_ECIR_SREC";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["SFDI_SRIG_SREC"] = 1] = "SFDI_SRIG_SREC";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["EREC_EREC_EREC"] = 2] = "EREC_EREC_EREC";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["SARR_SDIO_SCIR"] = 3] = "SARR_SDIO_SCIR";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["SSTA_ECIR_SREC"] = 4] = "SSTA_ECIR_SREC";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["SRIG_ECIR_SREC"] = 5] = "SRIG_ECIR_SREC";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["MINUS_MINUS_MINUS"] = 6] = "MINUS_MINUS_MINUS";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["PLUS_PLUS_PLUS"] = 7] = "PLUS_PLUS_PLUS";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["VACATION_TEA_MILKTEA"] = 8] = "VACATION_TEA_MILKTEA";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["RIGHT_NAIL_LOVE"] = 9] = "RIGHT_NAIL_LOVE";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["FIRE_ROCKET_NAIL"] = 10] = "FIRE_ROCKET_NAIL";
})(DEFAULT_BULLET_LIST_ARRAY || (DEFAULT_BULLET_LIST_ARRAY = {}));

export var DEFAULT_NUMBER_LIST_ARRAY;
/**
 * unicode映射表
 * ● -> \u25cf
 * ○ -> \u25cb
 * ■ -> \u25a0
 * ❖ -> \u2756
 * ❏ -> \u274f
 * ➔ -> \u2794
 * ◆  -> \u25c6
 * ★ -> \u2605
 * ➢ -> \u27a2
 * -  -> \u2012
 * +  ->  +
 * */

(function (DEFAULT_NUMBER_LIST_ARRAY) {
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["DEC_LEN_LROM_P"] = 0] = "DEC_LEN_LROM_P";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["DEC_LEN_LROM_H"] = 1] = "DEC_LEN_LROM_H";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["DEC_DEC_DEC_P"] = 2] = "DEC_DEC_DEC_P";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["UEN_LEN_LROM_P"] = 3] = "UEN_LEN_LROM_P";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["UROM_UEN_DEC_P"] = 4] = "UROM_UEN_DEC_P";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["DDEC_LEN_LROM_P"] = 5] = "DDEC_LEN_LROM_P";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["ZH_DEC_LEN_D"] = 6] = "ZH_DEC_LEN_D";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["ZH_DEC_LEN_A"] = 7] = "ZH_DEC_LEN_A";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["LEN_LROM_DEC_P"] = 8] = "LEN_LROM_DEC_P";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["DEC_LEN_LROM_A"] = 9] = "DEC_LEN_LROM_A";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["LEN_LROM_DEC_H"] = 10] = "LEN_LROM_DEC_H";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["DEC_LEN_LROM_D"] = 11] = "DEC_LEN_LROM_D";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["ZH_LEN_LROM_A"] = 12] = "ZH_LEN_LROM_A";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["ZH_LEN_LROM_D"] = 13] = "ZH_LEN_LROM_D";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["ZH_ZH_DEC_D"] = 14] = "ZH_ZH_DEC_D";
})(DEFAULT_NUMBER_LIST_ARRAY || (DEFAULT_NUMBER_LIST_ARRAY = {}));

var SCIR_ECIR_SREC = [['bullet', "\u25CF", 'left'], ['bullet', "\u25CB", 'left'], ['bullet', "\u25A0", 'left']];
var SFDI_SRIG_SREC = [['bullet', "\u2756", 'left'], ['bullet', "\u27A2", 'left'], ['bullet', "\u25A0", 'left']];
var EREC_EREC_EREC = [['bullet', "\u274F", 'left'], ['bullet', "\u274F", 'left'], ['bullet', "\u274F", 'left']];
var SARR_SDIO_SCIR = [['bullet', "\u2794", 'left'], ['bullet', "\u25C6", 'left'], ['bullet', "\u25CF", 'left']];
var SSTA_ECIR_SREC = [['bullet', "\u2605", 'left'], ['bullet', "\u25CB", 'left'], ['bullet', "\u25A0", 'left']];
var SRIG_ECIR_SREC = [['bullet', "\u27A2", 'left'], ['bullet', "\u25CB", 'left'], ['bullet', "\u25A0", 'left']];
var MINUS_MINUS_MINUS = [['bullet', "\u2012", 'left'], ['bullet', "\u2012", 'left'], ['bullet', "\u2012", 'left']];
var PLUS_PLUS_PLUS = [['bullet', '+', 'left'], ['bullet', '+', 'left'], ['bullet', '+', 'left']];
var VACATION_TEA_MILKTEA = [['bullet', '[休假]', 'left'], ['bullet', '[茶]', 'left'], ['bullet', '[奶茶]', 'left']];
var RIGHT_NAIL_LOVE = [['bullet', '[向右]', 'left'], ['bullet', '[钉子]', 'left'], ['bullet', '[爱心]', 'left']];
var FIRE_ROCKET_NAIL = [['bullet', '[火]', 'left'], ['bullet', '[火箭]', 'left'], ['bullet', '[钉子]', 'left']]; // ordered list

var DEC_LEN_LROM_P = [['decimal', '%1.', 'left'], ['lowerLetter', '%2.', 'left'], ['lowerRoman', '%3.', 'left']];
var DEC_LEN_LROM_H = [['decimal', '%1)', 'left'], ['lowerLetter', '%2)', 'left'], ['lowerRoman', '%3)', 'left']];
var DEC_DEC_DEC_P = [['decimal', '%1.', 'left'], ['decimal', '%1.%2', 'left'], ['decimal', '%1.%2.%3', 'left']];
var UEN_LEN_LROM_P = [['upperLetter', '%1.', 'left'], ['lowerLetter', '%2.', 'left'], ['lowerRoman', '%3.', 'left']];
var UROM_UEN_DEC_P = [['upperRoman', '%1.', 'left'], ['upperLetter', '%2.', 'left'], ['decimal', '%3.', 'left']];
var DDEC_LEN_LROM_P = [['decimalZero', '%1.', 'left'], ['lowerLetter', '%2.', 'left'], ['lowerRoman', '%3.', 'left']];
var ZH_LEN_LROM_D = [['chineseCountingThousand', '%1、', 'left'], ['lowerLetter', '%2)', 'left'], ['lowerRoman', '%3.', 'left']];
var ZH_LEN_LROM_A = [['chineseCountingThousand', '(%1)', 'left'], ['lowerLetter', '%2)', 'left'], ['lowerRoman', '%3.', 'left']];
var ZH_DEC_LEN_D = [['chineseCountingThousand', '%1、', 'left'], ['decimal', '%2.', 'left'], ['lowerLetter', '%3.', 'left']];
var ZH_DEC_LEN_A = [['chineseCountingThousand', '(%1)', 'left'], ['decimal', '%2.', 'left'], ['lowerLetter', '%3.', 'left']];
var ZH_ZH_DEC_D = [['chineseCountingThousand', '%1、', 'left'], ['chineseCountingThousand', '(%1)', 'left'], ['decimal', '%3.', 'left']];
var LEN_LROM_DEC_P = [['lowerLetter', '%2.', 'left'], ['lowerRoman', '%3.', 'left'], ['decimal', '%1.', 'left']];
var DEC_LEN_LROM_A = [['decimal', '(%1)', 'left'], ['lowerLetter', '%2)', 'left'], ['lowerRoman', '%3)', 'left']];
var LEN_LROM_DEC_H = [['lowerLetter', '%1)', 'left'], ['lowerRoman', '%2)', 'left'], ['decimal', '%3)', 'left']];
var DEC_LEN_LROM_D = [['decimal', '%1、', 'left'], ['lowerLetter', '%2.', 'left'], ['lowerRoman', '%3.', 'left']];
export var mapping = {
  SCIR_ECIR_SREC: SCIR_ECIR_SREC,
  SFDI_SRIG_SREC: SFDI_SRIG_SREC,
  EREC_EREC_EREC: EREC_EREC_EREC,
  SARR_SDIO_SCIR: SARR_SDIO_SCIR,
  SSTA_ECIR_SREC: SSTA_ECIR_SREC,
  SRIG_ECIR_SREC: SRIG_ECIR_SREC,
  MINUS_MINUS_MINUS: MINUS_MINUS_MINUS,
  PLUS_PLUS_PLUS: PLUS_PLUS_PLUS,
  VACATION_TEA_MILKTEA: VACATION_TEA_MILKTEA,
  RIGHT_NAIL_LOVE: RIGHT_NAIL_LOVE,
  FIRE_ROCKET_NAIL: FIRE_ROCKET_NAIL,
  DEC_LEN_LROM_P: DEC_LEN_LROM_P,
  DEC_LEN_LROM_H: DEC_LEN_LROM_H,
  DEC_LEN_LROM_A: DEC_LEN_LROM_A,
  DEC_DEC_DEC_P: DEC_DEC_DEC_P,
  UEN_LEN_LROM_P: UEN_LEN_LROM_P,
  UROM_UEN_DEC_P: UROM_UEN_DEC_P,
  DDEC_LEN_LROM_P: DDEC_LEN_LROM_P,
  ZH_LEN_LROM_D: ZH_LEN_LROM_D,
  ZH_LEN_LROM_A: ZH_LEN_LROM_A,
  ZH_DEC_LEN_D: ZH_DEC_LEN_D,
  ZH_DEC_LEN_A: ZH_DEC_LEN_A,
  ZH_ZH_DEC_D: ZH_ZH_DEC_D,
  LEN_LROM_DEC_P: LEN_LROM_DEC_P,
  LEN_LROM_DEC_H: LEN_LROM_DEC_H,
  DEC_LEN_LROM_D: DEC_LEN_LROM_D
};
