// @ali/we-design/dist-web/components/color-panel/constants.js
import { EColorType } from "./types";
var COLOR_1 = ['#222328', '#9B9A97', '#64473A', '#D9730D', '#DFAB01', '#0F7B6C', '#408BFB', '#6940A5', '#AD1A72', '#E03E3E'];
var COLOR_2 = ['#FFFFFF', '#F5F6F7', '#F5F1F0', '#FEF6E9', '#FFFAE5', '#ECF9F0', '#E8F2FE', '#F0EBF7', '#FCEFF3', '#FEF3F3'];
var COLOR_3 = ['#858689', '#E6E7E7', '#D7CCC8', '#F9DDB2', '#FFF8B9', '#D8ECE7', '#B7DCFB', '#D5C8EE', '#F4B9CD', '#F2BABA'];
var COLOR_4 = ['#646568', '#CBCCCD', '#C0A9A1', '#FFA726', '#FFEE58', '#B0D4CC', '#72B7EF', '#9C79D9', '#EB6793', '#EE7268'];
var COLOR_5 = ['#434448', '#B6B8B9', '#9F786B', '#FB8C00', '#FDD835', '#5DBEAC', '#0089FF', '#6C3BCF', '#D81B60', '#E53935'];
var COLOR_6 = ['#222328', '#979A9B', '#704F43', '#D16A15', '#DFB600', '#16816D', '#1564DC', '#44279B', '#A41D57', '#BB2626'];
export var BORDER_COLOR = ['#37352F', '#979A9B', '#937264', '#FFA344', '#FFDC49', '#4DAB9A', '#54A1E8', '#9A6DD7', '#E255A1', '#FF7369'];
export var FONT_SIMPLE_COLOR = COLOR_1;
export var FONT_GRADIENT_COLOR = ['#28B4BE__#3935DD', '#2449FE__#CA4BA7', '#FF7100__#F300AD', '#B16A4D__#D5A638', '#1890FF__#4FF3E0', '#F3AA00__#E43E29', '#EB47FF__#F95101', '#10CCF7__#AB44FB', '#F55B2D__#510101'];
export var FONT_COLOR = [].concat(FONT_SIMPLE_COLOR, ['#FFFFFF'], FONT_GRADIENT_COLOR); // 当存在多个颜色面板时，背景面板有 3 行

export var BACKGROUND_COLOR = [].concat(COLOR_2, COLOR_3, COLOR_4); // 当只存在背景面板时，背景面板有 6 行

export var BACKGROUND_COLOR_ALL = [].concat(COLOR_1, COLOR_2, COLOR_3, COLOR_4, COLOR_5, COLOR_6);
export var NO_COLOR = 'NO_COLOR';
export var DEFAULT_FONT_COLOR = '#404040';
export var DEFAULT_BG_COLOR = ''; // 当字体颜色为「默认值或色板的第一个值」时，可在「背景色为深色时」向用户推荐白色字体颜色

export var COULD_SHOW_TIP_COLORS = [DEFAULT_FONT_COLOR, COLOR_1[0]];
export var COLOR_SELECTED_KEY = 'color';
export var BORDER_SELECTED_KEY = 'borderColor';
export var BG_SELECTED_KEY = 'backgroundColor';
export var DEFAULT_PANEL_LAYOUT = [{
  title: '字体颜色',
  selectedKey: COLOR_SELECTED_KEY,
  colorType: EColorType.font,
  colors: FONT_COLOR,
  defaultColor: DEFAULT_FONT_COLOR,
  showMoreIcon: false,
  enableCustomColor: true
}, {
  title: '边框色',
  selectedKey: BORDER_SELECTED_KEY,
  colorType: EColorType.border,
  colors: BORDER_COLOR,
  defaultColor: '',
  showMoreIcon: false,
  enableCustomColor: true
}, {
  title: '背景颜色',
  selectedKey: BG_SELECTED_KEY,
  colorType: EColorType.background,
  colors: BACKGROUND_COLOR,
  defaultColor: '',
  showMoreIcon: false,
  enableCustomColor: true
}];
export var DEFAULT_PREVIEW_STATE = {
  isShow: false,
  shouldShowTip: false,
  color: DEFAULT_FONT_COLOR,
  backgroundColor: DEFAULT_BG_COLOR
};
