// @ali/we-design/dist-web/components/color-panel/styled.js
import styled, { css } from 'styled-components';
import { isMobile } from "../../utils/dom"; // colorSelector 显示「更多颜色」时右边距为 18， 不显示时为 10

export var PickerWrapper = /*#__PURE__*/styled.div(["padding:", ";box-sizing:border-box;", ""], function (props) {
  if (isMobile) {
    return '0';
  }

  return "10px " + (props.showMoreColorIcon ? 18 : 10) + "px 0px 10px";
}, isMobile ? 'flex: 1' : '');
export var SelectorWrapper = /*#__PURE__*/styled.div(["display:relative;padding-bottom:12px;"]);
export var SelectorTitle = /*#__PURE__*/styled.div(["font-size:12px;color:rgba(23,26,29,0.6);user-select:none;", ""], isMobile ? 'color: rgba(23,26,29,0.4);' : '');
export var ResetButton = /*#__PURE__*/styled.div(["box-sizing:border-box;width:100%;height:32px;border-top:1px solid rgba(126,134,142,0.16);text-align:center;cursor:pointer;span{font-size:13px;line-height:32px;color:#0089ff;}"]);
export var CheckboxWrapper = /*#__PURE__*/styled.div(["display:flex;width:100%;justify-content:space-between;padding-bottom:12px;"]);
export var CheckboxTitle = /*#__PURE__*/styled.div(["font-size:12px;line-height:20px;color:rgba(23,26,29,0.6);"]);
export var SelectorHeader = /*#__PURE__*/styled.div(["display:flex;justify-content:space-between;align-items:center;line-height:17px;padding-bottom:5px;", ""], isMobile ? "\n    padding-top: 4px;\n    padding-bottom: 10px;\n  " : '');
export var SelectorIconWrapper = /*#__PURE__*/styled.div(["display:flex;justify-content:center;align-items:center;cursor:pointer;width:22px;border-radius:4px;", " ", ""], !isMobile ? "\n    &:hover {\n      background-color: rgba(25, 31, 37, 0.16);\n    }\n  " : '', function (_ref) {
  var active = _ref.active;
  return active && css(["background-color:rgba(25,31,37,0.12);"]);
});
export var CustomColorPicker = /*#__PURE__*/styled.div(["position:absolute;top:0;.sketch-picker{*,*::before,*::after{box-sizing:content-box;}input{border-radius:6px;&:focus{outline:none;}}}"]);
