// @ali/4ever-plugin-table/dist-node/bi/createRenderNode.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderNode;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _lodash = require("lodash");

var _models = _interopRequireDefault(require("../mo/models"));

var _tableRow = _interopRequireDefault(require("../mo/models/tableRow"));

var _tableCell = _interopRequireDefault(require("../mo/models/tableCell"));

var _table = _interopRequireDefault(require("./components/table"));

var _tableRow2 = _interopRequireDefault(require("./components/tableRow"));

var _tableCell2 = _interopRequireDefault(require("./components/tableCell"));

var _styled = require("./components/styled");

var _constants = require("../utils/constants");

const _createElement = /*#__PURE__*/React.createElement;

function createRenderNode(config) {
  const tableConfig = config.table || {};
  const draggableConfig = config.draggable;
  const tableLocaleConfig = tableConfig.locale || {}; // @ts-ignore

  return {
    [_models.default.TYPE]: props => {
      const {
        node,
        visible
      } = props;

      if (!_models.default.isValidTable(node)) {
        return /*#__PURE__*/_createElement(React.Fragment, null);
      }

      const wrapperStyle = {
        paddingTop: _constants.SPACING - _constants.TOOLBAR_ITEM_SIZE,
        paddingBottom: _constants.SPACING - _constants.SCROLL_BAR_HEIGHT
      };
      return /*#__PURE__*/_createElement(_styled.TableWrapper, (0, _extends2.default)({}, props.attributes, {
        visible: visible,
        style: wrapperStyle
      }), /*#__PURE__*/_createElement(_table.default, (0, _extends2.default)({}, (0, _lodash.omit)(props, 'attributes'), {
        locale: tableLocaleConfig,
        tableConfig: tableConfig,
        draggableConfig: draggableConfig
      })));
    },
    [_tableRow.default.TYPE]: props => {
      return /*#__PURE__*/_createElement(_tableRow2.default, props);
    },
    [_tableCell.default.TYPE]: props => {
      return /*#__PURE__*/_createElement(_tableCell2.default, props);
    }
  };
}
