// @ali/4ever-plugin-table/dist-node/bi/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "mobileBiFactory", {
  enumerable: true,
  get: function () {
    return _mobile.mobileBiFactory;
  }
});
exports.biFactory = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _createOnAction = _interopRequireDefault(require("./handlers/createOnAction"));

var _onCangjieInput = _interopRequireDefault(require("./handlers/onCangjieInput"));

var _createOnKeyDown = _interopRequireDefault(require("./handlers/createOnKeyDown"));

var _onCangjiePaste = _interopRequireDefault(require("./handlers/onCangjiePaste"));

var _quickInsertConfig = _interopRequireDefault(require("./quickInsertConfig"));

var _queries = _interopRequireDefault(require("./queries"));

var _createOnCangjieCopy = _interopRequireDefault(require("./handlers/createOnCangjieCopy"));

var _createOnCangjieCut = _interopRequireDefault(require("./handlers/createOnCangjieCut"));

var _onMouseDown = _interopRequireDefault(require("./handlers/onMouseDown"));

var _createContextMenu = _interopRequireDefault(require("./createContextMenu"));

var _createRenderNode = _interopRequireDefault(require("./createRenderNode"));

var _createRenderToolbar = _interopRequireDefault(require("./createRenderToolbar"));

var _shouldNodeComponentUpdate = _interopRequireDefault(require("../common/handlers/shouldNodeComponentUpdate"));

var _selectionTrigger = _interopRequireDefault(require("../utils/utils/selectionTrigger"));

var _getNodeStyle = _interopRequireDefault(require("../common/handlers/getNodeStyle"));

var _mo = require("../mo");

var _actions = require("./actions");

var commands = _interopRequireWildcard(require("./commands"));

var _mobile = require("./mobile");

const contextMenu = (0, _createContextMenu.default)();

const biFactory = configs => {
  const tableConfig = configs.table || {};
  const {
    clipboard = {}
  } = configs;
  return { ...(0, _mo.moFactory)({}),
    toolbar: (0, _toolbar.default)(tableConfig),
    renderToolbar: (0, _createRenderToolbar.default)(tableConfig),
    menubar: (0, _menubar.default)(),
    renderNode: (0, _createRenderNode.default)(configs),
    shouldNodeComponentUpdate: _shouldNodeComponentUpdate.default,
    commands,
    actions: _actions.actions,
    isActions: _actions.isActions,
    onAction: (0, _createOnAction.default)(tableConfig),
    onCangjieInput: _onCangjieInput.default,
    onKeyDown: (0, _createOnKeyDown.default)(tableConfig),
    onCangjieCopy: (0, _createOnCangjieCopy.default)(),
    onCangjieSelect: (event, _, next) => {
      _selectionTrigger.default.set(event.detail.trigger);

      return next();
    },
    onCangjieCut: (0, _createOnCangjieCut.default)(),
    onCangjiePaste: (0, _onCangjiePaste.default)(clipboard, tableConfig),
    onMouseDown: _onMouseDown.default,
    quickInsert: _quickInsertConfig.default,
    queries: _queries.default,
    getNodeStyle: _getNodeStyle.default,
    contextMenu: contextMenu(tableConfig),
    models: [{
      klass: 'block',
      type: 'table'
    }, {
      klass: 'block',
      type: 'table-cell'
    }, {
      klass: 'block',
      type: 'table-row'
    }]
  };
};

exports.biFactory = biFactory;
