// @ali/we-design-next/dist-web/basic/list/styled.js
import styled from "styled-components";
import { COLORS, CSS_SCROLLBAR, CSS_CURSOR_DISABLE } from "../../common";
import { TEXT_CSS } from "../text";
export var ListWrap = /*#__PURE__*/styled.div(["", " outline:none;min-width:150px;overflow-x:hidden;overflow-y:overlay;"], CSS_SCROLLBAR);
export var ListDivider = /*#__PURE__*/styled.div(["height:1px;margin:4px;background-color:", ";"], COLORS.OTHERS.DIVIDER);
export var ListItemWrap = /*#__PURE__*/styled.div(["", " ", " display:flex;align-items:center;user-select:none;min-height:34px;padding:0 8px;margin:4px;border-radius:4px;", " color:", ";"], TEXT_CSS.TEXT, CSS_CURSOR_DISABLE, function (_ref) {
  var isActive = _ref.isActive;
  return isActive ? "background: " + COLORS.OTHERS.DAO_HOVER_BG_COLOR + ";" : null;
}, function (_ref2) {
  var disabled = _ref2.disabled;
  return disabled ? COLORS.LIGHT.LEVEL4_BASE_COLOR : 'unset';
});
export var ListItemPrefix = /*#__PURE__*/styled.div(["flex-grow:0;margin-right:4px;display:flex;align-items:center;min-width:20px;"]);
export var ListItemSuffix = /*#__PURE__*/styled.div(["flex-grow:0;margin-left:16px;display:flex;align-items:center;min-width:20px;"]);
export var ListItemContent = /*#__PURE__*/styled.div(["flex-grow:1;word-break:keep-all;"]);
export var ListItemGroupTitle = /*#__PURE__*/styled.div(["", " color:", ";line-height:36px;height:36px;padding-left:12px;user-select:none;"], TEXT_CSS.TEXT_SUB, COLORS.LIGHT.LEVEL2_BASE_COLOR);
