// @ali/4ever-plugin-clipboard/dist-node/utils/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ClipboardErrorEnum = exports.FILE_IMG_TYPE = exports.MARKDOWN_GRAMMER_REGEXP = exports.PASTE_MODES = exports.TEXT_FORMAT = exports.SPLIT_BUTTON_STYLE = exports.SOURCE_FROMAT = exports.PASTE_DECORATION = exports.DEFAULT_PASTE_MODES = exports.ICON_PASTE_STYLE = exports.ICON_CHECK_STYLE = exports.MATCH_FORMAT = exports.DROPDOWN_STYLE = exports.CARET_HEIGHT = void 0;
// 匹配目标格式粘贴
const MATCH_FORMAT = 'matchFormat'; // 保留源格式粘贴

exports.MATCH_FORMAT = MATCH_FORMAT;
const SOURCE_FROMAT = 'sourceFormat'; // 纯文本格式粘贴

exports.SOURCE_FROMAT = SOURCE_FROMAT;
const TEXT_FORMAT = 'textFormat';
exports.TEXT_FORMAT = TEXT_FORMAT;
const CARET_HEIGHT = 20;
exports.CARET_HEIGHT = CARET_HEIGHT;
const PASTE_DECORATION = 'paste-decoration';
exports.PASTE_DECORATION = PASTE_DECORATION;
const ICON_CHECK_STYLE = {
  fontSize: 12,
  color: '#007AFF'
};
exports.ICON_CHECK_STYLE = ICON_CHECK_STYLE;
const ICON_PASTE_STYLE = {
  fontSize: 18,
  color: '#111F2C',
  opacity: 0.72
};
exports.ICON_PASTE_STYLE = ICON_PASTE_STYLE;
const DROPDOWN_STYLE = {
  borderRadius: 4,
  border: '0.5px solid rgba(25,31,37,0.08)',
  boxShadow: '0 10px 20px 0 rgba(0,0,0,0.2)',
  fontSize: '14px'
};
exports.DROPDOWN_STYLE = DROPDOWN_STYLE;
const SPLIT_BUTTON_STYLE = {
  width: '100%',
  height: '100%'
};
exports.SPLIT_BUTTON_STYLE = SPLIT_BUTTON_STYLE;
const PASTE_MODES = {
  sourceFormat: SOURCE_FROMAT,
  matchFormat: MATCH_FORMAT,
  textFormat: TEXT_FORMAT,
  markdownToDoc: 'markdownToDoc'
};
exports.PASTE_MODES = PASTE_MODES;
const DEFAULT_PASTE_MODES = [PASTE_MODES.sourceFormat, PASTE_MODES.matchFormat, PASTE_MODES.textFormat];
exports.DEFAULT_PASTE_MODES = DEFAULT_PASTE_MODES;
var ClipboardErrorEnum;
exports.ClipboardErrorEnum = ClipboardErrorEnum;

(function (ClipboardErrorEnum) {
  ClipboardErrorEnum["COPY_FAILED"] = "COPY_FAILED";
  ClipboardErrorEnum["CUT_FAILED"] = "CUT_FAILED";
  ClipboardErrorEnum["PASTE_FAILED"] = "PASTE_FAILED";
  ClipboardErrorEnum["PASTE_TIME_OUT"] = "PASTE_TIME_OUT";
})(ClipboardErrorEnum || (exports.ClipboardErrorEnum = ClipboardErrorEnum = {}));

/**
 *  1. 首字母为 # 开头，
 *  2. 首字母为 * + 空格
 *  3. 首字母为 --- + \n
 *  4. 首字母为 数字 + . + 空格
 *  满足以上任一情况，则识别为 markdown 语法
 * */
const MARKDOWN_GRAMMER_REGEXP = /^#|^(\* )|^(\- )|^(\-{3}\n|^(\d+\. ))/;
exports.MARKDOWN_GRAMMER_REGEXP = MARKDOWN_GRAMMER_REGEXP;
const FILE_IMG_TYPE = 'image/png';
exports.FILE_IMG_TYPE = FILE_IMG_TYPE;
