// @ali/4ever-plugin-clipboard/dist-node/utils/tester.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.hasTextHtml = hasTextHtml;
exports.hasImage = hasImage;
exports.getSizeofClipboardData = getSizeofClipboardData;
exports.isDingDocs = isDingDocs;
exports.isDingSheet = isDingSheet;
exports.isMicrosoftOffice = isMicrosoftOffice;
exports.isMicrosoftWordOrExcel = isMicrosoftWordOrExcel;
exports.isMicrosoftWord = isMicrosoftWord;
exports.isMicrosoftExcel = isMicrosoftExcel;
exports.isMicrosoftPowerPoint = isMicrosoftPowerPoint;
exports.isGoogleDocs = isGoogleDocs;
exports.isGoogleSheet = isGoogleSheet;
exports.isShimoDocs = isShimoDocs;
exports.isYuqueDocs = isYuqueDocs;
exports.isVsCode = isVsCode;

var _everCangjie = require("@ali/4ever-cangjie");

var _everUtils = require("@ali/4ever-utils");

function hasTextHtml(data) {
  return data.types.includes(_everUtils.mimeTypes.TEXT_HTML);
}

function hasImage(data) {
  // office 会注入一个 image 的数据
  if (isMicrosoftOffice(data)) {
    return false;
  } // files 不存在


  if (!data.types.includes('Files')) {
    return false;
  }

  for (const item of data.items) {
    if (item.type.startsWith('image/')) {
      return true;
    }
  }

  return false;
}

function getSizeofClipboardData(data, format) {
  let content = '';

  if (format) {
    content = data.getData(format);
  } else {
    content = data.getData(_everUtils.mimeTypes.TEXT_HTML) || data.getData(_everUtils.mimeTypes.APPLICATION_CANGJIE_FRAGMENT) || data.getData(_everUtils.mimeTypes.APPLICATION_SLATE_FRAGMENT) || data.getData(_everUtils.mimeTypes.APPLICATION_DING_SHEET) || data.getData(_everUtils.mimeTypes.APPLICATION_GOOGLE_DOCS_CLIP) || data.getData(_everUtils.mimeTypes.TEXT_PLAIN);
  }

  const blob = new Blob([content]);
  return blob.size;
}

function isDingDocs(data) {
  // 仓颉编辑器
  if (data.types.includes(_everUtils.mimeTypes.APPLICATION_CANGJIE_FRAGMENT)) return true; // Slate 编辑器

  if (data.types.includes(_everUtils.mimeTypes.APPLICATION_SLATE_FRAGMENT)) return true; // 仓颉编辑器，有挂载 ASL 的 HTML

  if (data.types.includes(_everUtils.mimeTypes.TEXT_HTML)) {
    const html = data.getData(_everUtils.mimeTypes.TEXT_HTML);
    const domparser = new DOMParser();
    const root = domparser.parseFromString(html, 'text/html');
    const clipboardDataNode = root.querySelector(`[${_everCangjie.DATA_CLIPBORAD_CANGJIE}]`);

    if (clipboardDataNode && root.body.firstElementChild === clipboardDataNode) {
      return true;
    }
  }

  return false;
}

function isDingSheet(data) {
  return data.types.includes(_everUtils.mimeTypes.APPLICATION_DING_SHEET);
}

function isMicrosoftOffice(data) {
  return isMicrosoftWordOrExcel(data) || isMicrosoftPowerPoint(data);
}

function isMicrosoftWordOrExcel(data) {
  const html = data.getData(_everUtils.mimeTypes.TEXT_HTML);
  return Boolean(html?.includes('xmlns:o="urn:schemas-microsoft-com:office:office"'));
}

function isMicrosoftWord(data) {
  const html = data.getData(_everUtils.mimeTypes.TEXT_HTML);
  return Boolean(html?.includes('xmlns:w="urn:schemas-microsoft-com:office:word"'));
}

function isMicrosoftExcel(data) {
  const html = data.getData(_everUtils.mimeTypes.TEXT_HTML);
  return Boolean(html?.includes('xmlns:x="urn:schemas-microsoft-com:office:excel"'));
}

function isMicrosoftPowerPoint(data) {
  const html = data.getData(_everUtils.mimeTypes.TEXT_HTML);
  return Boolean(html?.includes('<meta name=ProgId content=PowerPoint.Slide>'));
}

function isGoogleDocs(data) {
  return data.types.includes(_everUtils.mimeTypes.APPLICATION_GOOGLE_DOCS_CLIP);
}

function isGoogleSheet(data) {
  const html = data.getData(_everUtils.mimeTypes.TEXT_HTML);
  return Boolean(html?.includes('<google-sheets-html-origin>'));
}

function isShimoDocs(data) {
  const html = data.getData(_everUtils.mimeTypes.TEXT_HTML);
  return Boolean(html?.includes('data-shimo-docs'));
}

function isYuqueDocs(data) {
  const html = data.getData(_everUtils.mimeTypes.TEXT_HTML);
  return Boolean(html?.includes('<meta name="source" content="lake"/>'));
}

function isVsCode(data) {
  return data.types.includes(_everUtils.mimeTypes.APPLICATION_VSCODE);
}
