// @ali/4ever-plugin-table/dist-node/mo/serializer/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createSerializer;
exports.createServerSerializationPlugin = createServerSerializationPlugin;

var _jsonMLToValue = _interopRequireDefault(require("./jsonMLToValue"));

var _valueToJsonML = _interopRequireDefault(require("./valueToJsonML"));

var _htmlToJsonML = _interopRequireDefault(require("./htmlToJsonML"));

var _jsonMLToHTML = _interopRequireDefault(require("./jsonMLToHTML"));

var _valueToText = _interopRequireDefault(require("./valueToText"));

var _markdownToHTML = _interopRequireDefault(require("./markdownToHTML"));

var _htmlToMarkdown = _interopRequireDefault(require("./htmlToMarkdown"));

var constants = _interopRequireWildcard(require("./constants"));

function createSerializer() {
  return {
    jsonMLToValue: _jsonMLToValue.default,
    valueToJsonML: _valueToJsonML.default,
    htmlToJsonML: _htmlToJsonML.default,
    jsonMLToHTML: _jsonMLToHTML.default,
    valueToText: _valueToText.default,
    markdownToHTML: _markdownToHTML.default,
    htmlToMarkdown: _htmlToMarkdown.default
  };
} // 兼容原 mo 里的 table......


createSerializer.constants = constants;

function createServerSerializationPlugin() {
  return {
    valueToJsonML: _valueToJsonML.default,
    jsonMLToValue: _jsonMLToValue.default
  };
}
