// @ali/4ever-plugin-table/dist-node/mo/serializer/jsonMLToHTML.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everUtils = require("@ali/4ever-utils");

var _styles = require("./styles");

var _constants = require("./constants");

const _createElement = /*#__PURE__*/React.createElement;

const TableCell = props => {
  const {
    node,
    children,
    ...rest
  } = props;
  const attrs = (0, _everUtils.getAttributes)(node) || {};
  const {
    colSpan = 1,
    rowSpan = 1,
    hidden
  } = attrs; // 被合并掉的单元格不输出，在 Word 中如果输出 hidden 单元格则会异常

  if (hidden) {
    return /*#__PURE__*/_createElement(React.Fragment, null);
  }

  const style = (0, _styles.generateTableCellStyle)(attrs);
  return /*#__PURE__*/_createElement("td", (0, _extends2.default)({}, rest, {
    colSpan: colSpan,
    rowSpan: rowSpan,
    style: style
  }), children);
};

const TableRow = props => {
  const {
    node,
    children,
    ...rest
  } = props;
  const {
    h
  } = (0, _everUtils.getAttributes)(node) || {};
  const style = {};
  const text = (0, _everUtils.getText)(node); // 无内容的行在 Word 中会塌陷，这里使用一个宽度让其能够撑开

  if (text.length === 0) {
    style.height = _styles.DEFAULT_TABLE_ROW_HEIGHT;
  }

  if (h) {
    style.height = h;
  }

  return /*#__PURE__*/_createElement("tr", (0, _extends2.default)({}, rest, {
    style: style
  }), children);
};

const Table = props => {
  const {
    node,
    children,
    ...rest
  } = props;
  const attrs = (0, _everUtils.getAttributes)(node);
  const {
    colsWidth = [],
    tblW
  } = attrs; // 宽度自适应表格将列宽单位由百分比转为 px

  let pixelColsWidth = colsWidth;

  if (tblW?.type === 'pct') {
    const totalColsWidth = colsWidth.reduce((acc, w) => acc + w);
    pixelColsWidth = [];
    colsWidth.reduce((acc, w, index) => {
      if (index === colsWidth.length - 1) {
        return pixelColsWidth.push(Math.max(_constants.DEFAULT_PAGE_WIDTH - acc), _constants.MIN_COL_WIDTH);
      }

      const pixelW = Math.round(w / totalColsWidth * _constants.DEFAULT_PAGE_WIDTH);
      pixelColsWidth.push(pixelW);
      return acc + pixelW;
    }, 0);
  }

  return /*#__PURE__*/_createElement("table", (0, _extends2.default)({}, rest, {
    style: (0, _styles.generateTableStyle)(attrs)
  }), /*#__PURE__*/_createElement("colgroup", null, pixelColsWidth.map((width, index) => /*#__PURE__*/_createElement("col", {
    key: index,
    width: width
  }))), /*#__PURE__*/_createElement("tbody", null, children));
};

const tagToComponent = {
  table: Table,
  tr: TableRow,
  tc: TableCell,
  td: TableCell
};
var _default = {
  name: 'table',
  match: node => {
    const tagName = (0, _everUtils.getTagName)(node);
    return !!tagToComponent[tagName];
  },
  convert: (node, path, convert) => {
    const tagName = (0, _everUtils.getTagName)(node);
    const Component = tagToComponent[tagName];
    const children = (0, _everUtils.getChildren)(node).map((child, index) => convert(child, [...path, index]));
    return /*#__PURE__*/_createElement(Component, {
      node: node,
      key: path.toString()
    }, children);
  }
};
exports.default = _default;
