// @ali/4ever-bi/dist-web/plugins/underline/index.js
import _extends from "@babel/runtime/helpers/extends";
import toolbar from "./toolbar";
import menubar from "./menubar";
import toggleUnderline from "./commands/toggleUnderlineMark";
import isUnderline from "./queries/isUnderline";
import * as actions from "./actions";
import onAction from "./handlers/onAction";
import onKeyDown from "./handlers/onKeyDown";
import createOnCangjieInput from "./handlers/onCangjieInput";
import createRenderToolbar from "./createRenderToolbar";
import { moFactory } from '@ali/4ever-plugin-underline';
var commands = {
  toggleUnderline: toggleUnderline
};
var queries = {
  isUnderline: isUnderline
};

function createUnderlinePlugin(config) {
  var underlineConfig = config.underline || {};
  return _extends({}, moFactory({}), {
    menubar: menubar(),
    toolbar: toolbar(underlineConfig),
    renderToolbar: createRenderToolbar(underlineConfig.locale),
    commands: commands,
    queries: queries,
    actions: actions,
    onAction: onAction,
    onKeyDown: onKeyDown,
    onCangjieInput: createOnCangjieInput(),
    models: {
      klass: 'mark',
      type: 'underline'
    }
  });
}

createUnderlinePlugin.actions = actions;
createUnderlinePlugin.commands = commands;
createUnderlinePlugin.queries = queries;
export default createUnderlinePlugin;
