// @ali/4ever-plugin-draggable/dist-web/bi/components/styled.js
import _taggedTemplateLiteralLoose from "@babel/runtime/helpers/taggedTemplateLiteralLoose";

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n  .draggable-highlight:not(.disable-draggable-highlight) {\n    border-radius: 4px;\n    opacity: 0.99;\n    // \u7F13\u52A8\n    transition: background 100ms ease;\n    background-color: rgba(0, 137, 255, 0.12) !important;\n\n    &.shadow-highlight {\n      border-radius: 8px;\n      box-shadow: 0px 0px 0px 3px ", ";\n    }\n  }\n\n  .draggable-highlight-table {\n    border-radius: 4px;\n    opacity: 0.99;\n    // \u7F13\u52A8\n    transition: background 100ms ease;\n  }\n\n  [data-draggable] {\n    transition: backgroundColor 300ms ease-out;\n    &.in-blockquote {\n      [data-draggable-before] {\n        width: 45px;\n      }\n    }\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled, { createGlobalStyle } from 'styled-components';
import { theme } from '@ali/we-design';
/**
 * 块级元素添加背景，行内元素不添加背景
 * 添加透明度是为了生成缩略图时不粘连相邻元素
 */

export var DragHighlightStyle = /*#__PURE__*/createGlobalStyle(_templateObject(), theme.commonBgHoverColor);
export var dragDirectlyStyle = {
  cursor: 'initial',
  transform: 'translateZ(0)'
};
export var DragHand = /*#__PURE__*/styled.div(["display:flex;align-items:center;justify-content:center;cursor:", ";line-height:initial;transition:background-color 100ms ease;background-color:#fff;border:1px solid rgba(126,134,142,0.24);border-radius:50%;width:", "px;height:", "px;svg{color:", ";}", " &:hover,&[data-is-hover=true]{background-color:rgba(50,96,201,0.06);}&:active{background-color:rgba(50,96,201,0.09);}"], function (_ref) {
  var draggable = _ref.draggable;
  return draggable ? 'grab' : 'pointer';
}, function (p) {
  return 19 / p.scale;
}, function (p) {
  return 19 / p.scale;
}, theme.colorText3, function (_ref2) {
  var active = _ref2.active;
  return active && "background-color: #EAECED;";
});
export var DropHolderWrapper = /*#__PURE__*/styled.div(["background-color:rgba(0,137,255,0.48);border-radius:4px;position:fixed;top:-9999px;pointer-events:none;z-index:999;"]);
export var DragTip = /*#__PURE__*/styled.div(["display:none;position:absolute;padding:6px 12px;border-radius:16px;font-size:12px;line-height:20px;background-color:rgba(17,31,44,0.85);color:#ffffff;text-align:center;user-select:none;line-height:1.2;white-space:nowrap;top:27px;left:50%;transform:translateX(-50%);"]);
export var DragIconWrapper = /*#__PURE__*/styled.div(["width:", "px;height:", "px;border-radius:", "px;svg{width:", "px;height:", "px;}"], function (p) {
  return 20 / p.scale;
}, function (p) {
  return 20 / p.scale;
}, function (p) {
  return 4 / p.scale;
}, function (p) {
  return 20 / p.scale;
}, function (p) {
  return 20 / p.scale;
});
export var TooltipText = /*#__PURE__*/styled.div(["color:rgba(255,255,255,.56);display:flex;"]);
export var PrimaryText = /*#__PURE__*/styled.div(["margin-right:4px;color:rgba(238,242,248,1);"]);
export var DragIcon = /*#__PURE__*/_createElement("svg", {
  width: "7",
  height: "7",
  viewBox: "0 0 7 7",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/_createElement("path", {
  d: "M7 1H0V0H7V1ZM7 4H0V3H7V4ZM0 7H7V6H0V7Z",
  fill: "#171A1D",
  fillOpacity: "0.6"
}));
