// @ali/we-ui/dist-web/components/date-picker/styled.js
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/explicit-function-return-type */
import styled from 'styled-components';
import { withTheme } from "../theme";
import { Ellipsis } from "../../utils/styled.mixin";
/**
 * ================ 控制器相关 start ====================
 */

export var ControlWrap = withTheme(styled.div(["margin-bottom:", ";user-select:none;color:rgba(0,0,0,0.85);"], function (_ref) {
  var theme = _ref.theme;
  return theme.ui.marginSmall;
}));
export var ControlRoot = withTheme(styled.div(["display:flex;align-items:center;"]));
export var ControlPrevWrap = withTheme(styled.div(["display:flex;align-items:center;"]));
export var ControlNextWrap = withTheme(styled.div(["display:flex;align-items:center;"]));
export var ControlMain = withTheme(styled.div(["width:0;flex:1 auto;"]));
export var ArrowWrap = withTheme(styled.div(["position:relative;color:", ";cursor:pointer;text-align:center;line-height:1;&:hover{color:", ";}.arrow-icon{font-size:", ";}"], function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.datePickerArrowColor;
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.datePickerArrowHoverColor;
}, function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.fontSizeH3;
}));
export var PanelControlWrap = withTheme(styled.div(["text-align:center;", " color:rgba(0,0,0,0.85);&.clickable{&:hover{color:", ";cursor:pointer;}}"], Ellipsis, function (_ref5) {
  var theme = _ref5.theme;
  return theme.ui.primaryColor;
}));
/**
 * ================ 控制器相关 end ====================
 */

/**
 * ================ 月份面板相关 start ====================
 */

export var MonthPanelWrap = withTheme(styled.div(["box-sizing:border-box;*{box-sizing:border-box;}"]));
export var MonthPanelBody = withTheme(styled.div(["display:flex;flex-wrap:wrap;"]));
export var MonthPanelItem = withTheme(styled.div(["width:33%;padding:", " ", ";span{display:block;text-align:center;", " padding:", " 0;border-radius:", ";&:hover{background-color:", ";cursor:pointer;}&.actived{color:", ";background-color:", ";cursor:default;}&.isDisabled{color:", ";background-color:", ";cursor:not-allowed;}}"], function (_ref6) {
  var theme = _ref6.theme;
  return theme.ui.padding;
}, function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.paddingLarge;
}, Ellipsis, function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.paddingTiny;
}, function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.elementHeight;
}, function (_ref10) {
  var theme = _ref10.theme;
  return theme.ui.backgroundColor;
}, function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.colorGray1;
}, function (_ref12) {
  var theme = _ref12.theme;
  return theme.ui.primaryColor;
}, function (_ref13) {
  var theme = _ref13.theme;
  return theme.ui.disabledColor;
}, function (_ref14) {
  var theme = _ref14.theme;
  return theme.ui.disabledBackgroundColor;
}));
/**
* ================ 月份面板相关 end ====================
*/

/**
 * ================ 年份面板相关 start ====================
 */

export var YearPanelWrap = withTheme(styled.div(["box-sizing:border-box;*{box-sizing:border-box;}"]));
export var YearPanelBody = withTheme(styled.div(["display:flex;flex-wrap:wrap;"]));
export var YearPanelItem = withTheme(styled.div(["width:33%;padding:", " ", ";span{display:block;text-align:center;", " padding:", " 0;border-radius:", ";&:hover{background-color:", ";cursor:pointer;}&.actived{color:", ";background-color:", ";cursor:default;}&.isDisabled{color:", ";background-color:", ";cursor:not-allowed;}}"], function (_ref15) {
  var theme = _ref15.theme;
  return theme.ui.padding;
}, function (_ref16) {
  var theme = _ref16.theme;
  return theme.ui.paddingLarge;
}, Ellipsis, function (_ref17) {
  var theme = _ref17.theme;
  return theme.ui.paddingTiny;
}, function (_ref18) {
  var theme = _ref18.theme;
  return theme.ui.elementHeight;
}, function (_ref19) {
  var theme = _ref19.theme;
  return theme.ui.backgroundColor;
}, function (_ref20) {
  var theme = _ref20.theme;
  return theme.ui.colorGray1;
}, function (_ref21) {
  var theme = _ref21.theme;
  return theme.ui.primaryColor;
}, function (_ref22) {
  var theme = _ref22.theme;
  return theme.ui.disabledColor;
}, function (_ref23) {
  var theme = _ref23.theme;
  return theme.ui.disabledBackgroundColor;
}));
/**
* ================ 年份面板相关 end ====================
*/

/**
 * ================ 日期面板相关 start ====================
 */

export var DatePanelWrap = withTheme(styled.div(["box-sizing:border-box;font-size:", ";color:", ";*{box-sizing:border-box;}"], function (_ref24) {
  var theme = _ref24.theme;
  return theme.ui.fontSize;
}, function (_ref25) {
  var theme = _ref25.theme;
  return theme.ui.textColor;
}));
export var CanlendarTable = withTheme(styled.table(["border-collapse:collapse;border-spacing:0;width:100%;table-layout:fixed;border:none;user-select:none;"]));
export var CanlendarWeekday = withTheme(styled.div(["margin-bottom:4px;color:rgba(0,0,0,0.85);"]));
export var CanlendarDate = withTheme(styled.div(["width:24px;height:24px;line-height:22px;border:1px solid transparent;margin:3px auto;border-radius:12px;text-align:center;color:", ";&.notCurrentMonth{color:", ";}&.isToday{position:relative;color:", ";border-color:", ";&:hover{background-color:transparent;border-color:", ";}}&:hover{background-color:", ";border-color:", ";cursor:pointer;}&.selected{color:", ";background-color:", " !important;border-color:", ";&:before{display:none;}}&.isDisabled{color:", ";background-color:", " !important;cursor:not-allowed;}"], function (_ref26) {
  var theme = _ref26.theme;
  return theme.ui.colorText5;
}, function (_ref27) {
  var theme = _ref27.theme;
  return theme.ui.colorText3;
}, function (_ref28) {
  var theme = _ref28.theme;
  return theme.ui.primaryColor;
}, function (_ref29) {
  var theme = _ref29.theme;
  return theme.ui.primaryColor;
}, function (_ref30) {
  var theme = _ref30.theme;
  return theme.ui.primaryColor;
}, function (_ref31) {
  var theme = _ref31.theme;
  return theme.ui.backgroundColor;
}, function (_ref32) {
  var theme = _ref32.theme;
  return theme.ui.backgroundColor;
}, function (_ref33) {
  var theme = _ref33.theme;
  return theme.ui.colorGray1;
}, function (_ref34) {
  var theme = _ref34.theme;
  return theme.ui.primaryColor;
}, function (_ref35) {
  var theme = _ref35.theme;
  return theme.ui.primaryColor;
}, function (_ref36) {
  var theme = _ref36.theme;
  return theme.ui.disabledColor;
}, function (_ref37) {
  var theme = _ref37.theme;
  return theme.ui.disabledBackgroundColor;
}));
export var DatePanelBody = withTheme(styled.div(["display:flex;flex-wrap:wrap;"]));
export var DatePanelItem = withTheme(styled.div(["width:33%;padding:", " ", ";span{display:block;text-align:center;", " padding:", " 0;border-radius:", ";&:hover{background-color:", ";cursor:pointer;}&.actived{color:", ";background-color:", ";cursor:default;}}"], function (_ref38) {
  var theme = _ref38.theme;
  return theme.ui.padding;
}, function (_ref39) {
  var theme = _ref39.theme;
  return theme.ui.paddingLarge;
}, Ellipsis, function (_ref40) {
  var theme = _ref40.theme;
  return theme.ui.paddingSmall;
}, function (_ref41) {
  var theme = _ref41.theme;
  return theme.ui.paddingSmall;
}, function (_ref42) {
  var theme = _ref42.theme;
  return theme.ui.primaryColor1;
}, function (_ref43) {
  var theme = _ref43.theme;
  return theme.ui.colorGray1;
}, function (_ref44) {
  var theme = _ref44.theme;
  return theme.ui.primaryColor;
}));
/**
* ================ 日期面板相关 end ====================
*/

/**
* ================ DatePicker相关 start ====================
*/

export var DatePickerTrigger = withTheme(styled.div(["position:relative;box-sizing:border-box;display:inline-block;width:100%;min-width:100px;height:", ";line-height:", ";border:1px solid ", ";border-radius:", ";transition:all .24s ease;padding:0 ", ";font-size:0 ", ";user-select:none;&:hover{border-color:", ";cursor:pointer;}.calendar-icon{margin-left:", ";color:", ";}&.placeholder{color:", ";}&.disabled{color:", ";background-color:", ";border-color:", ";cursor:not-allowed;}&.allowClear{.clear-icon{position:absolute;color:", ";right:", ";display:none;}&:hover{.calendar-icon{opacity:0;}.clear-icon{display:inline-block;}}}*{box-sizing:border-box;}"], function (_ref45) {
  var theme = _ref45.theme;
  return theme.ui.elementHeight;
}, function (_ref46) {
  var theme = _ref46.theme;
  return parseInt(theme.ui.elementHeight, 10) - 2 + "px";
}, function (_ref47) {
  var theme = _ref47.theme;
  return theme.ui.commonLineLightColor;
}, function (_ref48) {
  var theme = _ref48.theme;
  return theme.ui.radius;
}, function (_ref49) {
  var theme = _ref49.theme;
  return theme.ui.padding;
}, function (_ref50) {
  var theme = _ref50.theme;
  return theme.ui.fontSize;
}, function (_ref51) {
  var theme = _ref51.theme;
  return theme.ui.primaryColor;
}, function (_ref52) {
  var theme = _ref52.theme;
  return theme.ui.marginSmall;
}, function (_ref53) {
  var theme = _ref53.theme;
  return theme.ui.colorText2;
}, function (_ref54) {
  var theme = _ref54.theme;
  return theme.ui.colorText2;
}, function (_ref55) {
  var theme = _ref55.theme;
  return theme.ui.disabledColor;
}, function (_ref56) {
  var theme = _ref56.theme;
  return theme.ui.disabledBackgroundColor;
}, function (_ref57) {
  var theme = _ref57.theme;
  return theme.ui.disabledBorderColor;
}, function (_ref58) {
  var theme = _ref58.theme;
  return theme.ui.colorText2;
}, function (_ref59) {
  var theme = _ref59.theme;
  return theme.ui.padding;
}));
export var DatePickerTriggerWrap = withTheme(styled.div(["display:flex;align-items:center;.text{flex:1 auto;width:0;", "}"], Ellipsis));
export var DatePickerOverlay = withTheme(styled.div(["box-sizing:border-box;width:300px;font-size:", ";*{box-sizing:border-box;}"], function (_ref60) {
  var theme = _ref60.theme;
  return theme.ui.fontSize;
}));
export var DatePickerOverlayInputs = withTheme(styled.div(["display:flex;padding:", ";.datetime-input{margin-right:", ";&:last-child{margin-right:0;}}"], function (_ref61) {
  var theme = _ref61.theme;
  return theme.ui.padding;
}, function (_ref62) {
  var theme = _ref62.theme;
  return theme.ui.margin;
}));
export var DatePickerOverlayPanel = withTheme(styled.div(["padding:", ";"], function (_ref63) {
  var theme = _ref63.theme;
  return theme.ui.padding;
}));
export var DatePickerOverlayFooter = withTheme(styled.div(["padding:", ";border-top:1px solid ", ";display:flex;justify-content:flex-end;align-items:center;.time-button{color:", ";margin-right:", ";cursor:pointer;&.disabled{color:", ";cursor:not-allowed;}}"], function (_ref64) {
  var theme = _ref64.theme;
  return theme.ui.padding;
}, function (_ref65) {
  var theme = _ref65.theme;
  return theme.ui.borderColor;
}, function (_ref66) {
  var theme = _ref66.theme;
  return theme.ui.primaryColor;
}, function (_ref67) {
  var theme = _ref67.theme;
  return theme.ui.margin;
}, function (_ref68) {
  var theme = _ref68.theme;
  return theme.ui.disabledColor;
}));
/**
* ================ DatePicker相关 end ====================
*/
