// @ali/4ever-utils/dist-node/changeNullSelectionMark.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = changeNullSelectionMark;

var _everCangjie = require("@ali/4ever-cangjie");

function changeNullSelectionMark(controller, markType, currentMark, ranges, document) {
  const {
    selection
  } = controller.value;
  const documentMarks = document.getMarksAtRange(ranges).filter(mark => mark.type === markType);
  /* 删除document的mark */

  if (documentMarks.length) {
    documentMarks.forEach(mark => {
      controller.command(_everCangjie.Commands.removeMark, mark);
    });
  }
  /* 删除selection的mark */


  if (selection.marks) {
    const marks = selection.marks;
    marks.forEach(mark => {
      if (mark.type === markType) {
        controller.command(_everCangjie.Commands.removeMark, mark);
      }
    });
  }

  if (markType === 'fonts') {
    controller.command(_everCangjie.Commands.addMark, _everCangjie.Mark.create({
      type: markType,
      data: {
        ascii: currentMark,
        hAnsi: currentMark,
        cs: currentMark,
        eastAsia: currentMark
      }
    }));
  } else if (markType === 'sz') {
    controller.command(_everCangjie.Commands.addMark, _everCangjie.Mark.create({
      type: markType,
      data: {
        value: currentMark,
        szUnit: 'pt'
      }
    }));
  } else if (markType === 'highlight') {
    if (currentMark && typeof currentMark === 'string') {
      controller.command(_everCangjie.Commands.addMark, _everCangjie.Mark.create({
        type: markType,
        data: {
          value: currentMark
        }
      }));
    }
  } else {
    controller.command(_everCangjie.Commands.addMark, _everCangjie.Mark.create({
      type: markType,
      data: {
        value: currentMark
      }
    }));
  }

  return controller;
}
