// @ali/4ever-utils/dist-node/getClosestContainerBlock.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getClosestContainerBlock = getClosestContainerBlock;
exports.getClosestContainerBlockFromViewModel = getClosestContainerBlockFromViewModel;
exports.getClosestContainerBlockFromNative = getClosestContainerBlockFromNative;

var _objectUtils = require("./objectUtils");

const containerTypeMap = (0, _objectUtils.createDict)({
  paragraph: true,
  'heading-1': true,
  'heading-2': true,
  'heading-3': true,
  'heading-4': true,
  'heading-5': true,
  'heading-6': true,
  'table-cell': true,
  page: true,
  section: true
}); // viewModel下可以当作容器节点的 data-type

const viewModelContainerTypeMap = (0, _objectUtils.createDict)({ ...containerTypeMap,
  'paragraph-group': true,
  'hey-group': true,
  'callout-group': true,
  'subject-group': true,
  'heading-1-group': true,
  'heading-2-group': true,
  'heading-3-group': true,
  'heading-4-group': true,
  'heading-5-group': true,
  'heading-6-group': true
});

function getClosestContainerBlock(value, node) {
  if (!value) return null;
  const {
    document,
    startText
  } = value;

  if (!node) {
    if (!startText) {
      return null;
    } // @ts-ignore


    node = startText;
  }

  const closest = document.getClosest(node.key, n => containerTypeMap[n.type]); // 这个node确实归属这个document

  if (closest === null && document.getNode(node.key)) {
    return document;
  }

  return closest?.isElement() ? closest : null;
}
/**
 * 基于controller的view来寻找node的最近容器节点
 * @param document
 * @param node
 * @returns
 */


function getClosestContainerBlockFromViewModel(document, node) {
  if (!document) return null;
  const closest = document.getClosest(node.key, n => viewModelContainerTypeMap[n.type]); // 如果找不到node的包裹节点，就返回document这个最外层节点

  if (closest === null && document.getNode(node.key)) {
    return document;
  }

  return closest?.isElement() ? closest : null;
}

function getClosestContainerBlockFromNative(elementNode, selector) {
  if (!(elementNode && elementNode.nodeType === Node.ELEMENT_NODE)) {
    const selection = window.getSelection();
    const anchorNode = selection?.anchorNode;

    if (!anchorNode) {
      return null;
    } // @ts-ignore


    if (anchorNode.nodeType !== Node.ELEMENT_NODE) {
      elementNode = anchorNode.parentElement;
    } else {
      elementNode = anchorNode;
    } // @ts-ignore


    if (elementNode.nodeType !== Node.ELEMENT_NODE) {
      return null;
    }
  }

  if (!selector) {
    selector = '[data-type=page],[data-type=paragraph],[data-type=heading-1],[data-type=heading-2],[data-type=heading-3],[data-type=heading-4],[data-type=heading-5],[data-type=heading-6],[data-type=table-cell],[data-cangjie-content=true]';
  }

  return elementNode?.closest(selector);
}
