// @ali/4ever-utils/dist-node/getRectOfNodeFirstText.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getRectOfNodeFirstText;

var _everCangjie = require("@ali/4ever-cangjie");

var _getClosestDom = require("./getClosestDom");

const HEADING_TYPE = new Set(['heading-1', 'heading-2', 'heading-3', 'heading-4', 'heading-5', 'heading-6']);

function isHeading(node) {
  return _everCangjie.Block.isBlock(node) && HEADING_TYPE.has(node.type);
}

function isParagraph(node) {
  return _everCangjie.Block.isBlock(node) && node.type === 'paragraph';
}
/**
 * 获取 node 节点中第一个文本节点位置，找不到则返回 node 位置
 * @param node
 */


function getRectOfNodeFirstText(node) {
  let rect;

  try {
    const nodeDom = (0, _getClosestDom.getClosestDom)(node);
    if (!nodeDom) return rect;
    const nodeDomRect = nodeDom.getBoundingClientRect();

    if (!nodeDomRect) {
      return null;
    }

    rect = {
      top: nodeDomRect.top,
      left: nodeDomRect.left
    };

    if (isHeading(node) || isParagraph(node)) {
      // 找到第一个文字子节点 dom
      const firstTextChild = nodeDom.querySelector('[data-cangjie-leaf]');

      if (firstTextChild) {
        const range = document.createRange();
        range.setStart(firstTextChild, 0);
        range.setEnd(firstTextChild, 1);
        const textRect = range.getClientRects()[0];
        rect = {
          top: textRect.top,
          left: nodeDomRect.left,
          height: textRect.height
        };
      }
    }
  } catch (e) {
    console.error(e);
  }

  return rect;
}
