// @ali/4ever-utils/dist-node/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  createLinkify: true,
  matchesLink: true,
  loadImg: true,
  sanitizeUrl: true,
  getRectOfNodeFirstText: true,
  inspectScroller: true,
  throttleByIdleCallback: true,
  copyToClipboard: true,
  UnitConvert: true,
  isFirstPart: true,
  addDefaultUnit: true,
  transformThemeObjToCssString: true,
  isNaN: true,
  ptToPx: true,
  unitsConverter: true,
  sample: true,
  Sample: true,
  SimpleMapCache: true,
  fontRenderFallback: true,
  ShdUtil: true,
  UnderlineUtil: true,
  calculatedPluginSize: true,
  SpacingUtil: true,
  IndUtil: true,
  DEFAULT_INDENT: true,
  IndentTrigger: true,
  IndentType: true,
  SzUtil: true,
  CommonFormatUtil: true,
  GradientUtil: true,
  isGradientColor: true,
  transferGradientColor: true,
  shouldDegradeGradient: true,
  setGradientStyle: true,
  GRADIENT_COLOR_STYLE: true,
  GRADIENT_STYLE_FLAG: true,
  Measure: true,
  CJK_REG: true,
  isEmptyParagraph: true,
  isEmojiClsName: true,
  zIndex: true,
  HoverBlock: true,
  isSingleTableCell: true,
  unwrapSingleTableCell: true,
  getData: true,
  isEmptyValue: true,
  addMarkAtCurrentRange: true,
  addMarkAtRange: true,
  SELECTION_TRIGGER: true,
  DEFAULT_PAGE_WIDTH: true,
  generateRangesFromTableSelection: true,
  getBlocksFromEditor: true,
  getRangesFromEditor: true,
  removeMarkAtRange: true,
  removeMarkAtCurrentRange: true,
  blockEvent: true,
  createStore: true,
  LEGAL_EVENT: true,
  getCustomEvent: true,
  dispatchCustomEvent: true,
  isEqualWithDefault: true,
  getQuickInsertKeyword: true,
  APPLICATION_CANGJIE_FRAGMENT: true,
  APPLICATION_SLATE_FRAGMENT: true,
  APPLICATION_DING_SHEET: true,
  APPLICATION_CANGJIE_DRAG_FRAGMENT: true,
  APPLICATION_GOOGLE_DOCS_CLIP: true,
  APPLICATION_GOOGLE_SHEET_CLIP: true,
  APPLICATION_VSCODE: true,
  TEXT_PLAIN: true,
  TEXT_HTML: true,
  TEXT_LINK: true,
  TEXT_RTF: true,
  mimeTypes: true,
  isPlaceholderVisible: true,
  toHash: true,
  uniqueId: true,
  paragraphTags: true,
  changeNullSelectionMark: true,
  generateKey: true,
  getClosestParagraph: true,
  getRandomId: true,
  hasMark: true,
  isCollapsedInEmptyText: true,
  isInsertInline: true,
  createSpaceHandler: true,
  re: true,
  getMatches: true,
  getOffset: true,
  getMarkTextMatchInfo: true,
  getPointAndBlockOnSelection: true,
  isSelectionAtSameBlock: true,
  isAtStartOfParagraph: true,
  isAtEndOfParagraph: true,
  isSelectedBlock: true,
  PerfResizeObserver: true,
  getFuzzyQuery: true,
  fuzzySearch: true
};
Object.defineProperty(exports, "createLinkify", {
  enumerable: true,
  get: function () {
    return _linkify.default;
  }
});
Object.defineProperty(exports, "matchesLink", {
  enumerable: true,
  get: function () {
    return _linkify.matchesLink;
  }
});
Object.defineProperty(exports, "loadImg", {
  enumerable: true,
  get: function () {
    return _loadImg.loadImg;
  }
});
Object.defineProperty(exports, "sanitizeUrl", {
  enumerable: true,
  get: function () {
    return _sanitizeUrl.default;
  }
});
Object.defineProperty(exports, "getRectOfNodeFirstText", {
  enumerable: true,
  get: function () {
    return _getRectOfNodeFirstText.default;
  }
});
Object.defineProperty(exports, "inspectScroller", {
  enumerable: true,
  get: function () {
    return _inspectScroller.default;
  }
});
Object.defineProperty(exports, "throttleByIdleCallback", {
  enumerable: true,
  get: function () {
    return _throttleByIdleCallback.default;
  }
});
Object.defineProperty(exports, "copyToClipboard", {
  enumerable: true,
  get: function () {
    return _copyToClipboard.default;
  }
});
Object.defineProperty(exports, "UnitConvert", {
  enumerable: true,
  get: function () {
    return _unitConvert.default;
  }
});
Object.defineProperty(exports, "isFirstPart", {
  enumerable: true,
  get: function () {
    return _isFirstPart.default;
  }
});
Object.defineProperty(exports, "addDefaultUnit", {
  enumerable: true,
  get: function () {
    return _unit.addDefaultUnit;
  }
});
Object.defineProperty(exports, "transformThemeObjToCssString", {
  enumerable: true,
  get: function () {
    return _unit.transformThemeObjToCssString;
  }
});
Object.defineProperty(exports, "isNaN", {
  enumerable: true,
  get: function () {
    return _toUnits.isNaN;
  }
});
Object.defineProperty(exports, "ptToPx", {
  enumerable: true,
  get: function () {
    return _toUnits.ptToPx;
  }
});
Object.defineProperty(exports, "unitsConverter", {
  enumerable: true,
  get: function () {
    return _toUnits.unitsConverter;
  }
});
Object.defineProperty(exports, "sample", {
  enumerable: true,
  get: function () {
    return _sample.sample;
  }
});
Object.defineProperty(exports, "Sample", {
  enumerable: true,
  get: function () {
    return _sample.Sample;
  }
});
Object.defineProperty(exports, "SimpleMapCache", {
  enumerable: true,
  get: function () {
    return _simpleMapCache.default;
  }
});
Object.defineProperty(exports, "fontRenderFallback", {
  enumerable: true,
  get: function () {
    return _fontRenderFallback.default;
  }
});
Object.defineProperty(exports, "ShdUtil", {
  enumerable: true,
  get: function () {
    return _shd.default;
  }
});
Object.defineProperty(exports, "UnderlineUtil", {
  enumerable: true,
  get: function () {
    return _underline.default;
  }
});
Object.defineProperty(exports, "calculatedPluginSize", {
  enumerable: true,
  get: function () {
    return _calculatedPluginSize.default;
  }
});
Object.defineProperty(exports, "SpacingUtil", {
  enumerable: true,
  get: function () {
    return _spacing.SpacingUtil;
  }
});
Object.defineProperty(exports, "IndUtil", {
  enumerable: true,
  get: function () {
    return _ind.IndUtil;
  }
});
Object.defineProperty(exports, "DEFAULT_INDENT", {
  enumerable: true,
  get: function () {
    return _ind.DEFAULT_INDENT;
  }
});
Object.defineProperty(exports, "IndentTrigger", {
  enumerable: true,
  get: function () {
    return _ind.IndentTrigger;
  }
});
Object.defineProperty(exports, "IndentType", {
  enumerable: true,
  get: function () {
    return _ind.IndentType;
  }
});
Object.defineProperty(exports, "SzUtil", {
  enumerable: true,
  get: function () {
    return _sz.SzUtil;
  }
});
Object.defineProperty(exports, "CommonFormatUtil", {
  enumerable: true,
  get: function () {
    return _common.CommonFormatUtil;
  }
});
Object.defineProperty(exports, "GradientUtil", {
  enumerable: true,
  get: function () {
    return _gradientColor.default;
  }
});
Object.defineProperty(exports, "isGradientColor", {
  enumerable: true,
  get: function () {
    return _gradientColor.isGradientColor;
  }
});
Object.defineProperty(exports, "transferGradientColor", {
  enumerable: true,
  get: function () {
    return _gradientColor.transferGradientColor;
  }
});
Object.defineProperty(exports, "shouldDegradeGradient", {
  enumerable: true,
  get: function () {
    return _gradientColor.shouldDegradeGradient;
  }
});
Object.defineProperty(exports, "setGradientStyle", {
  enumerable: true,
  get: function () {
    return _gradientColor.setGradientStyle;
  }
});
Object.defineProperty(exports, "GRADIENT_COLOR_STYLE", {
  enumerable: true,
  get: function () {
    return _gradientColor.GRADIENT_COLOR_STYLE;
  }
});
Object.defineProperty(exports, "GRADIENT_STYLE_FLAG", {
  enumerable: true,
  get: function () {
    return _gradientColor.GRADIENT_STYLE_FLAG;
  }
});
Object.defineProperty(exports, "Measure", {
  enumerable: true,
  get: function () {
    return _measure.default;
  }
});
Object.defineProperty(exports, "CJK_REG", {
  enumerable: true,
  get: function () {
    return _measure.CJK_REG;
  }
});
Object.defineProperty(exports, "isEmptyParagraph", {
  enumerable: true,
  get: function () {
    return _isEmptyParagraph.default;
  }
});
Object.defineProperty(exports, "isEmojiClsName", {
  enumerable: true,
  get: function () {
    return _isEmojiClsName.isEmojiClsName;
  }
});
Object.defineProperty(exports, "zIndex", {
  enumerable: true,
  get: function () {
    return _zIndex.zIndex;
  }
});
Object.defineProperty(exports, "HoverBlock", {
  enumerable: true,
  get: function () {
    return _hoverBlock.default;
  }
});
Object.defineProperty(exports, "isSingleTableCell", {
  enumerable: true,
  get: function () {
    return _singleTableCell.isSingleTableCell;
  }
});
Object.defineProperty(exports, "unwrapSingleTableCell", {
  enumerable: true,
  get: function () {
    return _singleTableCell.unwrapSingleTableCell;
  }
});
Object.defineProperty(exports, "getData", {
  enumerable: true,
  get: function () {
    return _getData.getData;
  }
});
Object.defineProperty(exports, "isEmptyValue", {
  enumerable: true,
  get: function () {
    return _isEmptyValue.isEmptyValue;
  }
});
Object.defineProperty(exports, "addMarkAtCurrentRange", {
  enumerable: true,
  get: function () {
    return _addMarkAtCurrentRange.default;
  }
});
Object.defineProperty(exports, "addMarkAtRange", {
  enumerable: true,
  get: function () {
    return _addMarkAtCurrentRange.addMarkAtRange;
  }
});
Object.defineProperty(exports, "SELECTION_TRIGGER", {
  enumerable: true,
  get: function () {
    return _constants.SELECTION_TRIGGER;
  }
});
Object.defineProperty(exports, "DEFAULT_PAGE_WIDTH", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_PAGE_WIDTH;
  }
});
Object.defineProperty(exports, "generateRangesFromTableSelection", {
  enumerable: true,
  get: function () {
    return _generateRangesFromTableSelection.default;
  }
});
Object.defineProperty(exports, "getBlocksFromEditor", {
  enumerable: true,
  get: function () {
    return _getBlocksFromEditor.getBlocksFromEditor;
  }
});
Object.defineProperty(exports, "getRangesFromEditor", {
  enumerable: true,
  get: function () {
    return _getRangesFromEditor.getRangesFromEditor;
  }
});
Object.defineProperty(exports, "removeMarkAtRange", {
  enumerable: true,
  get: function () {
    return _removeMarkAtCurrentRange.removeMarkAtRange;
  }
});
Object.defineProperty(exports, "removeMarkAtCurrentRange", {
  enumerable: true,
  get: function () {
    return _removeMarkAtCurrentRange.removeMarkAtCurrentRange;
  }
});
Object.defineProperty(exports, "blockEvent", {
  enumerable: true,
  get: function () {
    return _blockEvent.blockEvent;
  }
});
Object.defineProperty(exports, "createStore", {
  enumerable: true,
  get: function () {
    return _createStore.createStore;
  }
});
Object.defineProperty(exports, "LEGAL_EVENT", {
  enumerable: true,
  get: function () {
    return _customEventUtils.LEGAL_EVENT;
  }
});
Object.defineProperty(exports, "getCustomEvent", {
  enumerable: true,
  get: function () {
    return _customEventUtils.getCustomEvent;
  }
});
Object.defineProperty(exports, "dispatchCustomEvent", {
  enumerable: true,
  get: function () {
    return _customEventUtils.dispatchCustomEvent;
  }
});
Object.defineProperty(exports, "isEqualWithDefault", {
  enumerable: true,
  get: function () {
    return _isEqualWithDefault.isEqualWithDefault;
  }
});
Object.defineProperty(exports, "getQuickInsertKeyword", {
  enumerable: true,
  get: function () {
    return _getQuickInsertKeyword.getKeyword;
  }
});
Object.defineProperty(exports, "APPLICATION_CANGJIE_FRAGMENT", {
  enumerable: true,
  get: function () {
    return _mimeTypes.APPLICATION_CANGJIE_FRAGMENT;
  }
});
Object.defineProperty(exports, "APPLICATION_SLATE_FRAGMENT", {
  enumerable: true,
  get: function () {
    return _mimeTypes.APPLICATION_SLATE_FRAGMENT;
  }
});
Object.defineProperty(exports, "APPLICATION_DING_SHEET", {
  enumerable: true,
  get: function () {
    return _mimeTypes.APPLICATION_DING_SHEET;
  }
});
Object.defineProperty(exports, "APPLICATION_CANGJIE_DRAG_FRAGMENT", {
  enumerable: true,
  get: function () {
    return _mimeTypes.APPLICATION_CANGJIE_DRAG_FRAGMENT;
  }
});
Object.defineProperty(exports, "APPLICATION_GOOGLE_DOCS_CLIP", {
  enumerable: true,
  get: function () {
    return _mimeTypes.APPLICATION_GOOGLE_DOCS_CLIP;
  }
});
Object.defineProperty(exports, "APPLICATION_GOOGLE_SHEET_CLIP", {
  enumerable: true,
  get: function () {
    return _mimeTypes.APPLICATION_GOOGLE_SHEET_CLIP;
  }
});
Object.defineProperty(exports, "APPLICATION_VSCODE", {
  enumerable: true,
  get: function () {
    return _mimeTypes.APPLICATION_VSCODE;
  }
});
Object.defineProperty(exports, "TEXT_PLAIN", {
  enumerable: true,
  get: function () {
    return _mimeTypes.TEXT_PLAIN;
  }
});
Object.defineProperty(exports, "TEXT_HTML", {
  enumerable: true,
  get: function () {
    return _mimeTypes.TEXT_HTML;
  }
});
Object.defineProperty(exports, "TEXT_LINK", {
  enumerable: true,
  get: function () {
    return _mimeTypes.TEXT_LINK;
  }
});
Object.defineProperty(exports, "TEXT_RTF", {
  enumerable: true,
  get: function () {
    return _mimeTypes.TEXT_RTF;
  }
});
Object.defineProperty(exports, "isPlaceholderVisible", {
  enumerable: true,
  get: function () {
    return _placeholder.isPlaceholderVisible;
  }
});
Object.defineProperty(exports, "toHash", {
  enumerable: true,
  get: function () {
    return _toHash.toHash;
  }
});
Object.defineProperty(exports, "uniqueId", {
  enumerable: true,
  get: function () {
    return _uniqueId.uniqueId;
  }
});
Object.defineProperty(exports, "paragraphTags", {
  enumerable: true,
  get: function () {
    return _paragraphTags.default;
  }
});
Object.defineProperty(exports, "changeNullSelectionMark", {
  enumerable: true,
  get: function () {
    return _changeNullSelectionMark.default;
  }
});
Object.defineProperty(exports, "generateKey", {
  enumerable: true,
  get: function () {
    return _generateKey.default;
  }
});
Object.defineProperty(exports, "getClosestParagraph", {
  enumerable: true,
  get: function () {
    return _getClosestParagraph.default;
  }
});
Object.defineProperty(exports, "getRandomId", {
  enumerable: true,
  get: function () {
    return _getRandomId.default;
  }
});
Object.defineProperty(exports, "hasMark", {
  enumerable: true,
  get: function () {
    return _hasMark.default;
  }
});
Object.defineProperty(exports, "isCollapsedInEmptyText", {
  enumerable: true,
  get: function () {
    return _isCollapsedInEmptyText.default;
  }
});
Object.defineProperty(exports, "isInsertInline", {
  enumerable: true,
  get: function () {
    return _isInsertInline.default;
  }
});
Object.defineProperty(exports, "createSpaceHandler", {
  enumerable: true,
  get: function () {
    return _createSpaceHandler.default;
  }
});
Object.defineProperty(exports, "re", {
  enumerable: true,
  get: function () {
    return _biUtils.re;
  }
});
Object.defineProperty(exports, "getMatches", {
  enumerable: true,
  get: function () {
    return _biUtils.getMatches;
  }
});
Object.defineProperty(exports, "getOffset", {
  enumerable: true,
  get: function () {
    return _biUtils.getOffset;
  }
});
Object.defineProperty(exports, "getMarkTextMatchInfo", {
  enumerable: true,
  get: function () {
    return _biUtils.getMarkTextMatchInfo;
  }
});
Object.defineProperty(exports, "getPointAndBlockOnSelection", {
  enumerable: true,
  get: function () {
    return _isParagraph.getPointAndBlockOnSelection;
  }
});
Object.defineProperty(exports, "isSelectionAtSameBlock", {
  enumerable: true,
  get: function () {
    return _isParagraph.isSelectionAtSameBlock;
  }
});
Object.defineProperty(exports, "isAtStartOfParagraph", {
  enumerable: true,
  get: function () {
    return _isParagraph.isAtStartOfParagraph;
  }
});
Object.defineProperty(exports, "isAtEndOfParagraph", {
  enumerable: true,
  get: function () {
    return _isParagraph.isAtEndOfParagraph;
  }
});
Object.defineProperty(exports, "isSelectedBlock", {
  enumerable: true,
  get: function () {
    return _isSelectedBlock.isSelectedBlock;
  }
});
Object.defineProperty(exports, "PerfResizeObserver", {
  enumerable: true,
  get: function () {
    return _perfResizeObserver.default;
  }
});
Object.defineProperty(exports, "getFuzzyQuery", {
  enumerable: true,
  get: function () {
    return _fuzzyQuery.getFuzzyQuery;
  }
});
Object.defineProperty(exports, "fuzzySearch", {
  enumerable: true,
  get: function () {
    return _fuzzyQuery.fuzzySearch;
  }
});
exports.mimeTypes = void 0;

var _linkify = _interopRequireWildcard(require("./linkify"));

var _objectUtils = require("./objectUtils");

Object.keys(_objectUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _objectUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _objectUtils[key];
    }
  });
});

var _loadImg = require("./loadImg");

var _sanitizeUrl = _interopRequireDefault(require("./sanitizeUrl"));

var _getRectOfNodeFirstText = _interopRequireDefault(require("./getRectOfNodeFirstText"));

var _inspectScroller = _interopRequireDefault(require("./inspectScroller"));

var _getClosestDom = require("./getClosestDom");

Object.keys(_getClosestDom).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getClosestDom[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getClosestDom[key];
    }
  });
});

var _calcRotation = require("./calcRotation");

Object.keys(_calcRotation).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _calcRotation[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _calcRotation[key];
    }
  });
});

var _getClosestContainerBlock = require("./getClosestContainerBlock");

Object.keys(_getClosestContainerBlock).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getClosestContainerBlock[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getClosestContainerBlock[key];
    }
  });
});

var _getInnerWidth = require("./getInnerWidth");

Object.keys(_getInnerWidth).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getInnerWidth[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getInnerWidth[key];
    }
  });
});

var _generateStyle = require("./generateStyle");

Object.keys(_generateStyle).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _generateStyle[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _generateStyle[key];
    }
  });
});

var _idleCallback = require("./idleCallback");

Object.keys(_idleCallback).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _idleCallback[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _idleCallback[key];
    }
  });
});

var _relativePosition = require("./relativePosition");

Object.keys(_relativePosition).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _relativePosition[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _relativePosition[key];
    }
  });
});

var _mapping = require("./hr/mapping");

Object.keys(_mapping).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mapping[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _mapping[key];
    }
  });
});

var _cssUtils = require("./cssUtils");

Object.keys(_cssUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _cssUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _cssUtils[key];
    }
  });
});

var _latency = require("./aspect/latency");

Object.keys(_latency).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _latency[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _latency[key];
    }
  });
});

var _pluginsUtil = require("./pluginsUtil");

Object.keys(_pluginsUtil).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _pluginsUtil[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _pluginsUtil[key];
    }
  });
});

var _sectionDataUtils = require("./sectionDataUtils");

Object.keys(_sectionDataUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _sectionDataUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _sectionDataUtils[key];
    }
  });
});

var _jsonml = require("./jsonml");

Object.keys(_jsonml).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _jsonml[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _jsonml[key];
    }
  });
});

var _matcher = require("./matcher");

Object.keys(_matcher).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _matcher[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _matcher[key];
    }
  });
});

var _sortPlugins = require("./sortPlugins");

Object.keys(_sortPlugins).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _sortPlugins[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _sortPlugins[key];
    }
  });
});

var _heading = require("./heading");

Object.keys(_heading).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _heading[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _heading[key];
    }
  });
});

var _regenerateUuid = require("./regenerateUuid");

Object.keys(_regenerateUuid).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _regenerateUuid[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _regenerateUuid[key];
    }
  });
});

var _indent = require("./indent");

Object.keys(_indent).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _indent[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _indent[key];
    }
  });
});

var _theme = require("./theme");

Object.keys(_theme).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _theme[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _theme[key];
    }
  });
});

var _throttleByIdleCallback = _interopRequireDefault(require("./utils/throttleByIdleCallback"));

var _copyToClipboard = _interopRequireDefault(require("./utils/copyToClipboard"));

var _unitConvert = _interopRequireDefault(require("./utils/unitConvert"));

var _isFirstPart = _interopRequireDefault(require("./utils/isFirstPart"));

var _unit = require("./utils/unit");

var _toUnits = require("./utils/toUnits");

var _sample = require("./utils/sample");

var _simpleMapCache = _interopRequireDefault(require("./utils/simpleMapCache"));

var _fontRenderFallback = _interopRequireDefault(require("./utils/fontRenderFallback"));

var _shd = _interopRequireDefault(require("./utils/formats/shd"));

var _underline = _interopRequireDefault(require("./utils/formats/underline"));

var _calculatedPluginSize = _interopRequireDefault(require("./utils/calculatedPluginSize"));

var _spacing = require("./utils/formats/spacing");

var _ind = require("./utils/formats/ind");

var _sz = require("./utils/formats/sz");

var _common = require("./utils/formats/common");

var _gradientColor = _interopRequireWildcard(require("./utils/formats/gradientColor"));

var _measure = _interopRequireWildcard(require("./utils/measure"));

var _isEmptyParagraph = _interopRequireDefault(require("./utils/isEmptyParagraph"));

var _isEmojiClsName = require("./utils/isEmojiClsName");

var _zIndex = require("./zIndex");

var _hoverBlock = _interopRequireDefault(require("./hoverBlock"));

var _singleTableCell = require("./utils/table/singleTableCell");

var _getData = require("./utils/getData");

var _isEmptyValue = require("./utils/isEmptyValue");

var _addMarkAtCurrentRange = _interopRequireWildcard(require("./utils/editor/addMarkAtCurrentRange"));

var _constants = require("./utils/editor/constants");

var _generateRangesFromTableSelection = _interopRequireDefault(require("./utils/editor/generateRangesFromTableSelection"));

var _getBlocksFromEditor = require("./utils/editor/getBlocksFromEditor");

var _getRangesFromEditor = require("./utils/editor/getRangesFromEditor");

var _removeMarkAtCurrentRange = require("./utils/editor/removeMarkAtCurrentRange");

var _blockEvent = require("./utils/blockEvent");

var _createStore = require("./utils/createStore");

var _customEventUtils = require("./utils/customEventUtils");

var _isEqualWithDefault = require("./utils/isEqualWithDefault");

var _getQuickInsertKeyword = require("./getQuickInsertKeyword");

var _mimeTypes = _interopRequireWildcard(require("./utils/mime-types"));

exports.mimeTypes = _mimeTypes;

var _placeholder = require("./utils/placeholder");

var _toHash = require("./utils/toHash");

var _uniqueId = require("./utils/uniqueId");

var _modelData = require("./utils/modelData");

Object.keys(_modelData).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _modelData[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _modelData[key];
    }
  });
});

var _paragraphTags = _interopRequireDefault(require("./utils/paragraphTags"));

var _changeNullSelectionMark = _interopRequireDefault(require("./changeNullSelectionMark"));

var _generateKey = _interopRequireDefault(require("./generateKey"));

var _getClosestParagraph = _interopRequireDefault(require("./getClosestParagraph"));

var _getRandomId = _interopRequireDefault(require("./getRandomId"));

var _hasMark = _interopRequireDefault(require("./hasMark"));

var _isCollapsedInEmptyText = _interopRequireDefault(require("./isCollapsedInEmptyText"));

var _isInsertInline = _interopRequireDefault(require("./isInsertInline"));

var _createSpaceHandler = _interopRequireDefault(require("./createSpaceHandler"));

var _biUtils = require("./biUtils");

var _isParagraph = require("./isParagraph");

var _isSelectedBlock = require("./isSelectedBlock");

var _actions = require("./pluginTmp/actions");

Object.keys(_actions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _actions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _actions[key];
    }
  });
});

var _perfResizeObserver = _interopRequireDefault(require("./perfResizeObserver"));

var _fuzzyQuery = require("./fuzzyQuery");
