// @ali/4ever-utils/dist-node/perfResizeObserver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _resizeObserverPolyfill = _interopRequireDefault(require("resize-observer-polyfill"));

class PerfResizeObserver {
  constructor() {
    this.observer = void 0;
    this.weakMap = void 0;

    this.resizeCallback = entries => {
      for (const entry of entries) {
        const cb = this.weakMap.get(entry.target);

        if (cb) {
          cb(entry.target);
        }
      }
    };

    this.observe = (target, cb, options) => {
      if (this.weakMap.has(target)) {
        this.observer.unobserve(target);
      }

      this.weakMap.set(target, cb); // @ts-ignore polyfill typing

      this.observer.observe(target, options);
    };

    this.unobserve = target => {
      this.weakMap.delete(target);
      this.observer.unobserve(target);
    };

    this.observer = new _resizeObserverPolyfill.default(this.resizeCallback);
    this.weakMap = new WeakMap();
  }
  /**
   *  消费同一个实例
   */


}

PerfResizeObserver._resizeObserver = new PerfResizeObserver();

PerfResizeObserver.create = () => PerfResizeObserver._resizeObserver;

var _default = PerfResizeObserver;
exports.default = _default;
