// @ali/4ever-utils/dist-node/sectionDataUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getMaxPageWidth = getMaxPageWidth;
exports.getMaxPageWidthSection = exports.getSectPr = exports.getSectPrFromJson = void 0;

/**
 * 纸张的属性
 */
const defaultSectPr = {
  pgSz: {
    h: 1122.67,
    w: 793.33,
    orient: 'portrait'
  },
  pgMar: {
    top: 96,
    bottom: 96,
    left: 72,
    right: 72,
    header: 56.73,
    footer: 66.13,
    gutter: 0
  }
};
/**
 * Pi-1.0 使用
 * @returns
 * @deprecated
 */

const getSectPrFromJson = data => {
  if (!data.sectPr) return defaultSectPr;
  const {
    pgMar
  } = defaultSectPr;
  const {
    pgSz
  } = defaultSectPr;
  const {
    w = pgSz.w,
    h = pgSz.h,
    orient = pgSz.orient
  } = data.sectPr.pgSz || {};
  const {
    header = pgMar.header,
    footer = pgMar.footer,
    top = pgMar.top,
    bottom = pgMar.bottom,
    left = pgMar.left,
    right = pgMar.right,
    gutter = pgMar.gutter
  } = data.sectPr.pgMar || {};
  let defaultFooterId;
  let defaultHeaderId;

  if (data.sectPr.footerRefs) {
    defaultFooterId = data.sectPr.footerRefs.defaultRefId;
  }

  if (data.sectPr.headerRefs) {
    defaultHeaderId = data.sectPr.headerRefs.defaultRefId;
  }

  return {
    defaultHeaderId,
    defaultFooterId,
    pgMar: {
      header,
      footer,
      top,
      bottom,
      left,
      right,
      gutter
    },
    pgSz: {
      w,
      h,
      orient
    }
  };
}; // Get sectPr from document.data


exports.getSectPrFromJson = getSectPrFromJson;

const getSectPr = (documentData, field = '') => {
  const {
    sectPr
  } = documentData;

  if (!sectPr) {
    if (field && defaultSectPr[field]) {
      return defaultSectPr[field];
    }

    return defaultSectPr;
  }

  let {
    pgSz,
    pgMar
  } = defaultSectPr;

  if (sectPr.pgSz) {
    const {
      w = pgSz.w,
      h = pgSz.h,
      orient = pgSz.orient
    } = sectPr.pgSz;
    pgSz = {
      w,
      h,
      orient
    };
  }

  if (sectPr.pgMar) {
    const {
      top = pgMar.top,
      bottom = pgMar.bottom,
      left = pgMar.left,
      right = pgMar.right,
      header = pgMar.header,
      footer = pgMar.footer,
      gutter = pgMar.gutter
    } = sectPr.pgMar;
    pgMar = {
      header,
      footer,
      top,
      bottom,
      left,
      right,
      gutter
    };
  }

  const result = { ...sectPr,
    pgSz,
    pgMar
  };

  if (field && result[field]) {
    return result[field];
  }

  return result;
};

exports.getSectPr = getSectPr;

function getMaxPageWidth(document) {
  const sectPr = getSectPr(document.data);
  const sectPrPgSz = sectPr.pgSz;
  const sectPrPageWidth = sectPrPgSz.w;
  return document.nodes.filter(node => {
    return node && node.type === 'sectionBr';
  }).reduce((maxWidth, node) => {
    if (node.data.pgSz?.w) {
      return Math.max(node.data.pgSz.w, maxWidth);
    }

    return maxWidth;
  }, sectPrPageWidth);
} // FIX: 重命名 getMaxPageWidthSection 导致 we-word 中无法找到该方法 to @白昱


const getMaxPageWidthSection = getMaxPageWidth;
exports.getMaxPageWidthSection = getMaxPageWidthSection;
