// @ali/we-design-next/dist-web/office/title-cover/TitleCover.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
var _excluded = ["cover", "renderError"];
function _createForOfIteratorHelperLoose(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (it) return (it = it.call(o)).next.bind(it); if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; return function () { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
import React, { useEffect, useImperativeHandle, useRef, useState } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { useTranslate } from "../../locale";
import { mergeRestProps } from "../../common/react/mergeRestProps";
import { useEventCallback } from "../../common";
import { Cover } from "./components/cover";
import { RepositionTips, TitleCoverWrapper } from "./style";
import { getSafePosition } from "./utils";
import { DEFAULT_POSITION } from "./constants";
export var TitleCover = /*#__PURE__*/React.memo( /*#__PURE__*/React.forwardRef(function (props, ref) {
  var cover = props.cover,
    renderError = props.renderError,
    rest = _objectWithoutPropertiesLoose(props, _excluded);
  var src = cover.src,
    thumbSrc = cover.thumbSrc,
    _cover$transform = cover.transform,
    transform = _cover$transform === void 0 ? {} : _cover$transform;
  var getCover = useEventCallback(function () {
    return cover;
  });
  var dataPosition = transform.offsetY;
  var _useState = useState(DEFAULT_POSITION),
    position = _useState[0],
    setPosition = _useState[1];
  var imageRef = useRef(null);
  var _useState2 = useState(false),
    isReposition = _useState2[0],
    setIsReposition = _useState2[1];
  var onChangeFnsRef = useRef([]);
  useImperativeHandle(ref, function () {
    return {
      onChange: function onChange(fn) {
        onChangeFnsRef.current.push(fn);
        return function () {
          var i = onChangeFnsRef.current.indexOf(fn);
          if (i !== -1) {
            onChangeFnsRef.current.splice(i, 1);
          }
        };
      },
      setRenderPosition: function setRenderPosition(renderP) {
        setPosition(renderP);
      },
      startReposition: function startReposition() {
        setIsReposition(true);
      },
      endReposition: function endReposition() {
        setIsReposition(false);
      },
      getImageRef: function getImageRef() {
        return imageRef;
      },
      getCover: getCover
    };
  });
  useEffect(function () {
    setPosition(getSafePosition(dataPosition));
  }, [dataPosition]);
  useEffect(function () {
    for (var _iterator = _createForOfIteratorHelperLoose(onChangeFnsRef.current), _step; !(_step = _iterator()).done;) {
      var _fn = _step.value;
      _fn({
        isReposition: isReposition,
        position: position
      });
    }
  }, [isReposition, position]);
  var safePosition = getSafePosition(position);
  var t = useTranslate();
  var style = {
    cursor: isReposition ? "move" : "auto"
  };
  return /*#__PURE__*/_createElement(TitleCoverWrapper, mergeRestProps(rest, 'wdn-title-cover', style), /*#__PURE__*/_createElement(Cover, {
    ref: imageRef,
    src: src,
    thumbSrc: thumbSrc,
    position: safePosition,
    renderError: renderError
  }), /*#__PURE__*/_createElement(RepositionTips, {
    visible: isReposition
  }, t("wdn_title_cover_drag_reposition")));
}));
