// @ali/we-design-next/dist-node/office/border-picker/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BorderPickerWrapper = exports.BorderPickerStyleWrap = exports.BorderPickerMain = exports.BorderPickerLayoutWrapper = exports.BorderPickerLayoutIconWrapper = exports.BorderPickerFooter = exports.BorderPickerColorWrap = exports.BorderPickerColorRect = exports.BorderPickerColorButton = void 0;
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _common = require("../../common");
const BorderPickerWrapper = /*#__PURE__*/_styledComponents.default.div([""]);
exports.BorderPickerWrapper = BorderPickerWrapper;
const BorderPickerMain = /*#__PURE__*/_styledComponents.default.div(["display:flex;margin-bottom:8px;"]);
exports.BorderPickerMain = BorderPickerMain;
const BorderPickerLayoutWrapper = /*#__PURE__*/_styledComponents.default.div(["", " width:100px;height:100px;display:flex;flex-wrap:wrap;border-radius:4px;border:1px solid ", ";"], _common.CSS_BORDER_BOX, _common.COLORS.COMMON.OVERLAY_COLOR);
exports.BorderPickerLayoutWrapper = BorderPickerLayoutWrapper;
const BorderPickerLayoutIconWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;align-items:center;justify-content:center;width:28px;height:28px;margin:2px;text-align:center;border-radius:4px;&:hover{background:", ";}", ""], _common.COLORS.COMMON.LINE_HARD_COLOR, ({
  active
}) => active ? `background: ${_common.COLORS.COMMON.LINE_HARD_COLOR};` : null);
exports.BorderPickerLayoutIconWrapper = BorderPickerLayoutIconWrapper;
const BorderPickerColorWrap = /*#__PURE__*/_styledComponents.default.div(["display:flex;align-items:center;"]);
exports.BorderPickerColorWrap = BorderPickerColorWrap;
const BorderPickerColorButton = /*#__PURE__*/_styledComponents.default.div(["", " padding:6px 8px;border:1px solid ", ";border-radius:4px;&:hover{border-color:", ";}"], _common.CSS_BORDER_BOX, _common.COLORS.COMMON.OVERLAY_COLOR, _common.COLORS.COMMON.BLUE1_HOVER_COLOR);
exports.BorderPickerColorButton = BorderPickerColorButton;
const BorderPickerColorRect = /*#__PURE__*/_styledComponents.default.div(["", " width:30px;height:14px;"], _common.CSS_BORDER_BOX);
exports.BorderPickerColorRect = BorderPickerColorRect;
const BorderPickerStyleWrap = /*#__PURE__*/_styledComponents.default.div(["margin-top:8px;display:flex;align-items:center;"]);
exports.BorderPickerStyleWrap = BorderPickerStyleWrap;
const BorderPickerFooter = /*#__PURE__*/(0, _styledComponents.default)(_common.RowItem)(["&:hover{background:", ";}"], _common.COLORS.OTHERS.DAO_HOVER_BG_COLOR);
exports.BorderPickerFooter = BorderPickerFooter;
