// @ali/4ever-plugin-table/dist-node/bi/commands/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "deleteCols", {
  enumerable: true,
  get: function () {
    return _deleteCols.default;
  }
});
Object.defineProperty(exports, "deleteRows", {
  enumerable: true,
  get: function () {
    return _deleteRows.default;
  }
});
Object.defineProperty(exports, "deleteTable", {
  enumerable: true,
  get: function () {
    return _deleteTable.default;
  }
});
Object.defineProperty(exports, "emptyTableSelection", {
  enumerable: true,
  get: function () {
    return _emptyTableSelection.default;
  }
});
Object.defineProperty(exports, "insertCol", {
  enumerable: true,
  get: function () {
    return _insertCol.default;
  }
});
Object.defineProperty(exports, "insertRow", {
  enumerable: true,
  get: function () {
    return _insertRow.default;
  }
});
Object.defineProperty(exports, "insertRowBySelections", {
  enumerable: true,
  get: function () {
    return _insertRowBySelections.default;
  }
});
Object.defineProperty(exports, "insertTable", {
  enumerable: true,
  get: function () {
    return _insertTable.default;
  }
});
Object.defineProperty(exports, "mergeCells", {
  enumerable: true,
  get: function () {
    return _mergeCells.default;
  }
});
Object.defineProperty(exports, "moveCols", {
  enumerable: true,
  get: function () {
    return _moveCols.default;
  }
});
Object.defineProperty(exports, "moveRows", {
  enumerable: true,
  get: function () {
    return _moveRows.default;
  }
});
Object.defineProperty(exports, "resizeColsWidth", {
  enumerable: true,
  get: function () {
    return _resizeColsWidth.default;
  }
});
Object.defineProperty(exports, "resizeRowsHeight", {
  enumerable: true,
  get: function () {
    return _resizeRowsHeight.default;
  }
});
Object.defineProperty(exports, "resizeTableHeight", {
  enumerable: true,
  get: function () {
    return _resizeTableHeight.default;
  }
});
Object.defineProperty(exports, "resizeTableToAutofitWidth", {
  enumerable: true,
  get: function () {
    return _resizeTableToAutofitWidth.default;
  }
});
Object.defineProperty(exports, "resizeTableWidth", {
  enumerable: true,
  get: function () {
    return _resizeTableWidth.default;
  }
});
Object.defineProperty(exports, "resizeTableWidthFitContainer", {
  enumerable: true,
  get: function () {
    return _resizeTableWidthFitContainer.default;
  }
});
Object.defineProperty(exports, "selectCols", {
  enumerable: true,
  get: function () {
    return _selectCols.default;
  }
});
Object.defineProperty(exports, "selectRows", {
  enumerable: true,
  get: function () {
    return _selectRows.default;
  }
});
Object.defineProperty(exports, "selectTableSelection", {
  enumerable: true,
  get: function () {
    return _selectTableSelection.default;
  }
});
Object.defineProperty(exports, "selectWholeTable", {
  enumerable: true,
  get: function () {
    return _selectWholeTable.default;
  }
});
Object.defineProperty(exports, "setCellsData", {
  enumerable: true,
  get: function () {
    return _setCellsData.default;
  }
});
Object.defineProperty(exports, "setTableAutofitWidth", {
  enumerable: true,
  get: function () {
    return _setTableAutofitWidth.default;
  }
});
Object.defineProperty(exports, "setTableBorder", {
  enumerable: true,
  get: function () {
    return _setTableBorder.default;
  }
});
Object.defineProperty(exports, "splitCell", {
  enumerable: true,
  get: function () {
    return _splitCell.default;
  }
});
Object.defineProperty(exports, "toggleColHeader", {
  enumerable: true,
  get: function () {
    return _toggleColHeader.default;
  }
});
Object.defineProperty(exports, "toggleRowHeader", {
  enumerable: true,
  get: function () {
    return _toggleRowHeader.default;
  }
});
Object.defineProperty(exports, "updateTableStyle", {
  enumerable: true,
  get: function () {
    return _updateTableStyle.default;
  }
});

var _deleteCols = _interopRequireDefault(require("./deleteCols"));

var _deleteRows = _interopRequireDefault(require("./deleteRows"));

var _deleteTable = _interopRequireDefault(require("./deleteTable"));

var _emptyTableSelection = _interopRequireDefault(require("./emptyTableSelection"));

var _insertCol = _interopRequireDefault(require("./insertCol"));

var _insertRow = _interopRequireDefault(require("./insertRow"));

var _insertRowBySelections = _interopRequireDefault(require("./insertRowBySelections"));

var _insertTable = _interopRequireDefault(require("./insertTable"));

var _mergeCells = _interopRequireDefault(require("./mergeCells"));

var _moveCols = _interopRequireDefault(require("./moveCols"));

var _moveRows = _interopRequireDefault(require("./moveRows"));

var _resizeColsWidth = _interopRequireDefault(require("./resizeColsWidth"));

var _resizeRowsHeight = _interopRequireDefault(require("./resizeRowsHeight"));

var _resizeTableHeight = _interopRequireDefault(require("./resizeTableHeight"));

var _resizeTableToAutofitWidth = _interopRequireDefault(require("./resizeTableToAutofitWidth"));

var _resizeTableWidth = _interopRequireDefault(require("./resizeTableWidth"));

var _resizeTableWidthFitContainer = _interopRequireDefault(require("./resizeTableWidthFitContainer"));

var _selectCols = _interopRequireDefault(require("./selectCols"));

var _selectRows = _interopRequireDefault(require("./selectRows"));

var _selectTableSelection = _interopRequireDefault(require("./selectTableSelection"));

var _selectWholeTable = _interopRequireDefault(require("./selectWholeTable"));

var _setCellsData = _interopRequireDefault(require("./setCellsData"));

var _setTableAutofitWidth = _interopRequireDefault(require("./setTableAutofitWidth"));

var _setTableBorder = _interopRequireDefault(require("./setTableBorder"));

var _splitCell = _interopRequireDefault(require("./splitCell"));

var _toggleColHeader = _interopRequireDefault(require("./toggleColHeader"));

var _toggleRowHeader = _interopRequireDefault(require("./toggleRowHeader"));

var _updateTableStyle = _interopRequireDefault(require("./updateTableStyle"));
