// @ali/4ever-plugin-table/dist-node/bi/commands/insertTable.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = insertTable;

var _everCangjie = require("@ali/4ever-cangjie");

var _createTable = _interopRequireDefault(require("../utils/createTable"));

var _createTableData = _interopRequireDefault(require("../utils/createTableData"));

var _everPluginBasic = require("@ali/4ever-plugin-basic");

var _actions = require("../actions");

function insertTable(controller, payload) {
  const {
    rowSize,
    colSize,
    containerWidth,
    defaultAutofit,
    defaultHeader,
    setTableHeaderStyle
  } = payload;
  const tableData = (0, _createTableData.default)(controller, {
    colSize,
    containerWidth,
    defaultAutofit
  });
  const table = (0, _createTable.default)({
    rowSize,
    colSize,
    data: tableData
  });

  const point = _everCangjie.TextPoint.create({
    key: table.getFirstNode().key,
    offset: 0
  });

  const newSelection = _everCangjie.Selection.create({
    anchor: point,
    focus: point
  });

  if (defaultHeader) {
    // NOTE: 需要先执行 setTableHeaderStyle
    setTableHeaderStyle?.();
  }

  controller.run('onAction', _everPluginBasic.biActions.createInsertBlockWithoutExtraBlankAction({
    node: table
  }));
  controller.command(_everCangjie.Commands.select, newSelection); // 自动定位到第一个单元格

  if (defaultHeader) {
    const newTable = controller.value.document.getNode(table.key);
    controller.run('onAction', (0, _actions.setTableRowHeader)(newTable));
  }

  return controller;
}
