// @ali/4ever-plugin-table/dist-node/bi/commands/updateTableStyle.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = updateTableStyle;

var _everCangjie = require("@ali/4ever-cangjie");

var _models = _interopRequireDefault(require("../../mo/models"));

var _constants = require("../constants");

const DEFAULT_TBLLOOK = {
  firstRow: 0,
  firstColumn: 0,
  lastRow: 0,
  lastColumn: 0
};

function updateStyleId(controller, table) {
  const newTable = controller.value.document.getNode(table.key);
  if (!newTable) return controller;
  const tableData = { ...(newTable.data || {})
  };
  const {
    styleId
  } = tableData;
  const shouldRemoveStyleId = !_models.default.isRowHeader(newTable) && !_models.default.isColumnHeader(newTable);

  if (shouldRemoveStyleId && styleId) {
    delete tableData.styleId;
    return controller.command(_everCangjie.Commands.setNodeByKey, newTable.key, {
      data: tableData
    });
  }

  if (!shouldRemoveStyleId && !styleId) {
    tableData.styleId = _constants.TABLE_HEADER_STYLE_ID;
    return controller.command(_everCangjie.Commands.setNodeByKey, newTable.key, {
      data: tableData
    });
  }

  return controller;
}

function updateTblLook(controller, table) {
  const newTable = controller.value.document.getNode(table.key);
  if (!newTable) return controller;
  const tableData = { ...(newTable.data || {})
  };
  const newTblLook = { ...DEFAULT_TBLLOOK
  };

  const isRowHeader = _models.default.isRowHeader(newTable);

  const isColumnHeader = _models.default.isColumnHeader(newTable);

  newTblLook.firstRow = isRowHeader ? 1 : 0;
  newTblLook.firstColumn = isColumnHeader ? 1 : 0;
  return controller.command(_everCangjie.Commands.setNodeByKey, newTable.key, {
    data: { ...tableData,
      tblLook: newTblLook
    }
  });
}

function updateTableStyle(controller, table) {
  updateStyleId(controller, table);
  updateTblLook(controller, table);
  return controller;
}
