// @ali/we-ui/dist-web/components/avatar/styled.js
import styled, { css } from 'styled-components';
import { resetComponent, optionWrapStyle } from "../../utils/styled.mixin";
import { withTheme } from "../theme";
var directionCss = /*#__PURE__*/css(["", ""], function (_ref) {
  var direction = _ref.direction;

  if (direction === 'rtl') {
    return 'flex-direction: row-reverse;';
  }

  return '';
});
var avatarWidth = /*#__PURE__*/css(["width:", ";"], function (_ref2) {
  var theme = _ref2.theme,
      size = _ref2.size;
  var sizeString = '';
  if (size === 'small') sizeString = 'Small';
  if (size === 'large') sizeString = 'Large';
  return theme.ui["elementHeight" + sizeString];
});
var avatarHeight = /*#__PURE__*/css(["height:", ";"], function (_ref3) {
  var theme = _ref3.theme,
      size = _ref3.size;
  var sizeString = '';
  if (size === 'small') sizeString = 'Small';
  if (size === 'large') sizeString = 'Large';
  return theme.ui["elementHeight" + sizeString];
});
export var AvatarWrap = withTheme(styled.div(["", " display:inline-block;position:relative;", ";", ";cursor:pointer;"], resetComponent, avatarWidth, avatarHeight));
export var AvatarInner = withTheme(styled.div(["box-sizing:border-box;position:relative;width:100%;height:100%;border-radius:50%;overflow:hidden;border:1px solid ", ";&:hover,&:active{border:1px solid ", ";}"], function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.colorGray1;
}, function (_ref5) {
  var color = _ref5.color,
      theme = _ref5.theme;
  return typeof color === 'string' ? color : theme.ui.colorGray1;
}));
var divider = /*#__PURE__*/css(["content:'';display:block;position:absolute;background-color:", ";"], function (_ref6) {
  var theme = _ref6.theme;
  return theme.ui.colorGray1;
});
var leftDivider = /*#__PURE__*/css(["", " left:0;top:0;bottom:0;width:1px;"], divider);
var rightDivider = /*#__PURE__*/css(["", " right:0;top:0;bottom:0;width:1px;"], divider);
var topDivider = /*#__PURE__*/css(["", " right:0;left:0;top:0;height:1px;"], divider);
export var AvatarImage = withTheme(styled.div(["position:absolute;left:0;top:0;width:100%;height:100%;background-image:url(", ");background-size:cover;background-position:center center;&.avatar_1_1{border-radius:50%;}&.avatar_2_1,&.avatar_3_1{transform:translate(-50%,0);}&.avatar_2_2{transform:translate(50%,0);&:after{", "}}&.avatar_3_2{transform:translate(50%,-50%);&:after{", "}}&.avatar_3_3{transform:translate(50%,50%);&:after{", "}&:before{", "}}&.avatar_4_1{transform:translate(-50%,-50%);&:after{", "}}&.avatar_4_2{transform:translate(-50%,50%);&:after{", "}&:before{", "}}&.avatar_4_3{transform:translate(50%,-50%);}&.avatar_4_4{transform:translate(50%,50%);&:after{", "}}"], function (_ref7) {
  var url = _ref7.url;
  return url;
}, leftDivider, leftDivider, leftDivider, topDivider, rightDivider, rightDivider, topDivider, topDivider));
export var AvatarCorner = withTheme(styled.div(["position:absolute;bottom:0;right:0;width:8px;height:8px;border-radius:50%;overflow:hidden;background-color:", ";border:1px solid ", ";box-sizing:border-box;"], function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.colorGray1;
}, function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.colorGray1;
}));
export var AvatarColor = withTheme(styled.div(["width:100%;height:100%;background-color:", ";"], function (_ref10) {
  var backgroundColor = _ref10.backgroundColor;
  return backgroundColor;
}));
export var NameWrap = withTheme(styled.div(["width:100%;height:100%;background-color:", ";color:", ";display:flex;justify-content:center;align-items:center;font-size:14px;"], function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.primaryColor;
}, function (_ref12) {
  var theme = _ref12.theme;
  return theme.ui.colorGray1;
}));
export var TooltipWrap = /*#__PURE__*/styled.div(["display:flex;align-items:center;.status-icon{font-size:12px;}"]);
export var TextSpan = /*#__PURE__*/styled.span(["margin:0 3px 0 7px;"]);
export var AvatarGroupWrap = withTheme(styled.div(["display:inline-block;"]));
export var GroupAvatarList = withTheme(styled.ul(["", ";", ";width:100%;", ";display:flex;position:relative;"], resetComponent, directionCss, avatarHeight));
export var GroupAvatarItem = withTheme(styled.li(["", ";", ";position:relative;flex:1 auto;&.more{border:0;}&.more:hover .more-item-span{background-color:#C6C9CD;}"], resetComponent, avatarHeight));
export var GroupMoreItem = withTheme(styled.div(["box-sizing:border-box;", ";", ";background-color:#E2E4E5;border-radius:50%;border:1px solid ", ";display:flex;justify-content:center;align-items:center;cursor:pointer;i{display:block;background-color:", ";width:2px;height:2px;margin-right:2px;border-radius:50%;&:last-child{margin-right:0;}}"], avatarHeight, avatarWidth, function (_ref13) {
  var theme = _ref13.theme;
  return theme.ui.colorGray1;
}, function (_ref14) {
  var theme = _ref14.theme;
  return theme.ui.textColor;
}));
export var GroupMoreItemSpan = withTheme(styled.span(["font-size:12px;transform:scale(.75);"]));
export var OverlayWrap = withTheme(styled.div(["padding:10px 0;min-width:100px;"]));
export var CountWrap = withTheme(styled.div(["font-size:", ";color:", ";line-height:2;padding-left:6px;"], function (_ref15) {
  var theme = _ref15.theme;
  return theme.ui.fontSizeSmall;
}, function (_ref16) {
  var theme = _ref16.theme;
  return theme.ui.commonBaseColor3;
}));
export var OverlayAvatarItem = withTheme(styled.div(["", " justify-content:space-between;margin:0 6px;padding:6px;.status-icon{font-size:14px;}"], optionWrapStyle));
export var OverlayAvatarName = withTheme(styled.div(["margin-left:7px;padding-right:16px;"]));
export var ListWrap = withTheme(styled.div(["width:100%;overflow-y:auto;"]));
export var OverlayAvatarWrap = withTheme(styled.div(["display:flex;align-items:center;"]));
export var OverlayAvatarStatus = withTheme(styled.span(["margin-left:auto;color:", ";"], function (_ref17) {
  var theme = _ref17.theme;
  return theme.ui.commonBaseColor3;
}));
