// @ali/4ever-bi/dist-node/plugins/link/mobile/contextMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../actions");

var _zh_CN = _interopRequireDefault(require("../../../locales/zh_CN"));

var _default = configs => {
  return function contextMenu(controller, next) {
    const link = controller.query('getSelectedLinkUrl');

    if (!link) {
      return next();
    }

    const title = configs?.locale.openLinkMenu || _zh_CN.default.link.locale.openLinkMenu;
    const item = {
      name: title,
      action: {
        type: _actions.LINK_OPEN,
        payload: {
          href: link
        }
      },
      key: 'openLink',
      icon: 'https://gw.alicdn.com/imgextra/i3/O1CN01PvOSYX1Y9NJ9rGzkZ_!!6000000003016-55-tps-24-24.svg',
      options: {
        group: 100,
        mobile: true
      },
      role: _everBamboo.PluginRoles.link
    }; // 非只读的预览模式，显示 open link 按钮

    if (controller.query('isQuickPreview')) {
      return [item];
    }

    return [item, ...next()];
  };
};

exports.default = _default;
