// @ali/4ever-utils/dist-node/pluginTmp/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  SET_NODE_UUID: true,
  setNodeUUID: true,
  TOGGLE_FOLD: true,
  toggleFold: true
};
exports.setNodeUUID = setNodeUUID;
exports.toggleFold = toggleFold;
exports.TOGGLE_FOLD = exports.SET_NODE_UUID = void 0;

var _newlineGuide = require("./newlineGuide");

Object.keys(_newlineGuide).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _newlineGuide[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _newlineGuide[key];
    }
  });
});

var _blockquote = require("./blockquote");

Object.keys(_blockquote).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _blockquote[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _blockquote[key];
    }
  });
});

/** uuid */
const SET_NODE_UUID = Symbol('SET_NODE_UUID');
exports.SET_NODE_UUID = SET_NODE_UUID;

function setNodeUUID(node, uuid) {
  return {
    type: SET_NODE_UUID,
    payload: {
      node,
      uuid
    }
  };
}
/** newlineguide */


const TOGGLE_FOLD = Symbol('TOGGLE_FOLD');
exports.TOGGLE_FOLD = TOGGLE_FOLD;

function toggleFold(payload) {
  return {
    type: TOGGLE_FOLD,
    payload
  };
}
