// @ali/4ever-utils/dist-node/pluginTmp/blockquote.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toggleBlockquote = toggleBlockquote;
exports.transformBlockquote = transformBlockquote;
exports.enableBlockquote = enableBlockquote;
exports.disableBlockquote = disableBlockquote;
exports.BLOCKQUOTE_DISABLE = exports.BLOCKQUOTE_TRANSFORM = exports.BLOCKQUOTE_ENABLE = exports.BLOCKQUOTE_TOGGLE = void 0;
const BLOCKQUOTE_TOGGLE = Symbol('BLOCKQUOTE_TOGGLE');
exports.BLOCKQUOTE_TOGGLE = BLOCKQUOTE_TOGGLE;
const BLOCKQUOTE_ENABLE = Symbol('BLOCKQUOTE_ENABLE');
exports.BLOCKQUOTE_ENABLE = BLOCKQUOTE_ENABLE;
const BLOCKQUOTE_TRANSFORM = Symbol('BLOCKQUOTE_TRANSFORM');
exports.BLOCKQUOTE_TRANSFORM = BLOCKQUOTE_TRANSFORM;
const BLOCKQUOTE_DISABLE = Symbol('BLOCKQUOTE_DISABLE');
exports.BLOCKQUOTE_DISABLE = BLOCKQUOTE_DISABLE;

function toggleBlockquote() {
  return {
    type: BLOCKQUOTE_TOGGLE
  };
}

function transformBlockquote() {
  return {
    type: BLOCKQUOTE_TRANSFORM
  };
}

function enableBlockquote() {
  return {
    type: BLOCKQUOTE_ENABLE
  };
}

function disableBlockquote() {
  return {
    type: BLOCKQUOTE_DISABLE
  };
}
