// @ali/4ever-utils/dist-node/pluginTmp/newlineGuide.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.removeNewlineTriggerText = removeNewlineTriggerText;
exports.removeNewlineData = removeNewlineData;
exports.setNewlineData = setNewlineData;
exports.mountInsertMenuFromLeftToolbar = mountInsertMenuFromLeftToolbar;
exports.mountInsertMenu = mountInsertMenu;
exports.MOUNT_SLASH_TOOLBAR = exports.MOUNT_PLUS_TOOLBAR = exports.NEWLINE_SET_DATA = exports.NEWLINE_REMOVE_DATA = exports.NEWLINE_REMOVE_TRIGGER_TEXT = void 0;
const NEWLINE_REMOVE_TRIGGER_TEXT = Symbol('NEWLINE_REMOVE_TRIGGER_TEXT');
exports.NEWLINE_REMOVE_TRIGGER_TEXT = NEWLINE_REMOVE_TRIGGER_TEXT;
const NEWLINE_REMOVE_DATA = Symbol('NEWLINE_REMOVE_DATA');
exports.NEWLINE_REMOVE_DATA = NEWLINE_REMOVE_DATA;
const NEWLINE_SET_DATA = Symbol('NEWLINE_SET_DATA');
exports.NEWLINE_SET_DATA = NEWLINE_SET_DATA;
const MOUNT_PLUS_TOOLBAR = Symbol('MOUNT_PLUS_TOOLBAR');
exports.MOUNT_PLUS_TOOLBAR = MOUNT_PLUS_TOOLBAR;
const MOUNT_SLASH_TOOLBAR = Symbol('MOUNT_SLASH_TOOLBAR');
exports.MOUNT_SLASH_TOOLBAR = MOUNT_SLASH_TOOLBAR;

function removeNewlineTriggerText(needRemoveEmptyParagraph) {
  return {
    type: NEWLINE_REMOVE_TRIGGER_TEXT,
    payload: {
      needRemoveEmptyParagraph
    }
  };
}

function removeNewlineData() {
  return {
    type: NEWLINE_REMOVE_DATA
  };
}

function setNewlineData(payload) {
  return {
    type: NEWLINE_SET_DATA,
    payload
  };
}

function mountInsertMenuFromLeftToolbar() {
  return {
    type: MOUNT_PLUS_TOOLBAR
  };
}

function mountInsertMenu() {
  return {
    type: MOUNT_SLASH_TOOLBAR
  };
}
