// @ali/4ever-plugin-image/dist-node/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.default;
  }
});
Object.defineProperty(exports, "Image", {
  enumerable: true,
  get: function () {
    return _models.default;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.createServerSerializationPlugin;
  }
});
Object.defineProperty(exports, "isSupportImageOrientation", {
  enumerable: true,
  get: function () {
    return _imageOrientation.isNativeSupport;
  }
});
exports.OldBambooPlugin = void 0;

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;

var _index = _interopRequireDefault(require("./mo/index"));

var _models = _interopRequireDefault(require("./mo/models"));

var _index2 = _interopRequireWildcard(require("./mo/serializer/index"));

var _imageOrientation = require("./utils/utils/imageOrientation");
