// @ali/4ever-plugin-table/dist-web/bi/queries/index.js
import canInsertTable from "./canInsertTable";
import getAllRowsHeight from "./getAllRowsHeight";
import getAllRowsDomRect from "./getAllRowsDomRect";
import getAllRowsNodes from "./getAllRowsNodes";
import getSelectionFromTableSelection from "./getSelectionFromTableSelection";
import getTableSelection from "./getTableSelection";
import getRowsHeight from "./getRowsHeight";
import hasTableSelection from "./hasTableSelection";
import isAllSelectedCellsComplete from "./isAllSelectedCellsComplete";
import isHideSelection from "./isHideSelection";
import isSelectionSupportMerge from "./isSelectionSupportMerge";
import isSelectionSupportSplit from "./isSelectionSupportSplit";
import isSelectWholeCol from "./isSelectWholeCol";
import isSelectWholeRow from "./isSelectWholeRow";
import isSelectWholeTable from "./isSelectWholeTable";
import isTableFromSameOrigin from "./isTableFromSameOrigin";
import isTableSupportResize from "./isTableSupportResize";
import tableSelection from "./tableSelection";
import viewTableSelections from "./viewTableSelections";
import getLeftToolbarOffset from "./getLeftToolbarOffset";
import getRowIndexByClientY from "./getRowIndexByClientY";
import shouldSelectAll from "./shouldSelectAll";
import isHideDragButton from "./isHideDragButton";
import { isPureTableSelection } from "./isPureTableSelection";
import { isSelectionInTable } from "./isSelectionInTable";
import { isSelectionInTableCell } from "./isSelectionInTableCell";
import { isSelectionInTableEdge } from "./isSelectionInTableEdge";
import { getSelectionTheme as getTableSelectionTheme } from "./getSelectionTheme";
import disableLeftToolbar from "./disableLeftToolbar";
import calcTableSelectionPos from "./calcTableSelectionPos";
import { default as commonQueries } from "../../common/queries";
import { getTableConstants } from "../../common/queries/getTableConstants";
var calcMinTableColWidth = commonQueries.calcMinTableColWidth,
    getSelectionContainers = commonQueries.getSelectionContainers,
    isNodeInTable = commonQueries.isNodeInTable,
    isTable = commonQueries.isTable,
    isTableCell = commonQueries.isTableCell,
    isTableRow = commonQueries.isTableRow,
    isTableNode = commonQueries.isTableNode;
var queries = {
  calcMinTableColWidth: calcMinTableColWidth,
  calcTableSelectionPos: calcTableSelectionPos,
  canInsertTable: canInsertTable,
  getAllRowsDomRect: getAllRowsDomRect,
  getAllRowsHeight: getAllRowsHeight,
  getAllRowsNodes: getAllRowsNodes,
  getSelectionFromTableSelection: getSelectionFromTableSelection,
  getTableSelection: getTableSelection,
  getRowsHeight: getRowsHeight,
  hasTableSelection: hasTableSelection,
  isAllSelectedCellsComplete: isAllSelectedCellsComplete,
  isPureTableSelection: isPureTableSelection,
  isHideSelection: isHideSelection,
  isSelectionSupportMerge: isSelectionSupportMerge,
  isSelectionSupportSplit: isSelectionSupportSplit,
  isSelectWholeCol: isSelectWholeCol,
  isSelectWholeRow: isSelectWholeRow,
  isSelectWholeTable: isSelectWholeTable,
  isSelectionInTable: isSelectionInTable,
  isSelectionInTableCell: isSelectionInTableCell,
  isSelectionInTableEdge: isSelectionInTableEdge,
  isTableFromSameOrigin: isTableFromSameOrigin,
  isTableSupportResize: isTableSupportResize,
  tableSelection: tableSelection,
  viewTableSelections: viewTableSelections,
  getLeftToolbarOffset: getLeftToolbarOffset,
  getLeftToolbarOffsetWithFocus: getLeftToolbarOffset,
  getRowIndexByClientY: getRowIndexByClientY,
  getSelectionContainers: getSelectionContainers,
  shouldSelectAll: shouldSelectAll,
  isNodeInTable: isNodeInTable,
  isTableNode: isTableNode,
  isHideDragButton: isHideDragButton,
  disableLeftToolbar: disableLeftToolbar,
  isTable: isTable,
  isTableRow: isTableRow,
  isTableCell: isTableCell,
  getTableSelectionTheme: getTableSelectionTheme,
  getTableConstants: getTableConstants
};
export default queries;
