// @ali/4ever-plugin-code/dist-node/zhi/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.zhiFactory = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everUtils = require("@ali/4ever-utils");

var _index = _interopRequireDefault(require("../mo/index"));

var _lazyRenderStrategy = _interopRequireDefault(require("../common/queries/lazyRenderStrategy"));

var _ZhiCode = _interopRequireDefault(require("../components/ZhiCode"));

var _themes = require("../utils/constants/themes");

var _languages = require("../utils/constants/languages");

const _createElement = /*#__PURE__*/React.createElement;

const zhiFactory = configs => {
  return { ...(0, _index.default)({}),
    queries: {
      lazyRenderStrategy: _lazyRenderStrategy.default
    },

    renderNode(props, controller, next) {
      const {
        node,
        attributes
      } = props;
      const {
        ref,
        ...rest
      } = attributes;
      const {
        type,
        text,
        data
      } = node;

      if (type !== 'code') {
        return next();
      }

      const {
        print = false
      } = configs.code || {};
      return /*#__PURE__*/_createElement(_ZhiCode.default, (0, _extends2.default)({}, rest, {
        controller: controller,
        code: text,
        node: node,
        theme: (0, _everUtils.getData)(data, 'theme', _themes.DEFAULT_THEME),
        syntax: (0, _everUtils.getData)(data, 'syntax', _languages.DEFAULT_LANGUAGE),
        height: (0, _everUtils.getData)(data, 'height', undefined),
        lineWrapping: !!(0, _everUtils.getData)(data, 'wrap', false),
        print: print
      }));
    }

  };
};

exports.zhiFactory = zhiFactory;
