// @ali/4ever-plugin-code/dist-node/zhi/mobile.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mobileZhiFactory = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everUtils = require("@ali/4ever-utils");

var _index = _interopRequireDefault(require("../mo/index"));

var _lazyRenderStrategy = _interopRequireDefault(require("../common/queries/lazyRenderStrategy"));

var _SimpleCode = _interopRequireDefault(require("../components/SimpleCode"));

const _createElement = /*#__PURE__*/React.createElement;

const mobileZhiFactory = configs => {
  return { ...(0, _index.default)({}),
    queries: {
      lazyRenderStrategy: _lazyRenderStrategy.default
    },

    renderNode(props, _, next) {
      const {
        node,
        attributes
      } = props;
      const {
        type,
        text,
        data
      } = node;

      if (type !== 'code') {
        return next();
      }

      const {
        print = false
      } = configs.code || {};
      return /*#__PURE__*/_createElement(_SimpleCode.default, (0, _extends2.default)({
        node: node
      }, attributes, {
        code: text,
        theme: (0, _everUtils.getData)(data, 'theme', 'default'),
        height: !print ? (0, _everUtils.getData)(data, 'height', undefined) : undefined,
        lineWrapping: !!(0, _everUtils.getData)(data, 'wrap', false) || print,
        print: print
      }));
    }

  };
};

exports.mobileZhiFactory = mobileZhiFactory;
