// @ali/4ever-bi/dist-web/plugins/translate/components/next/translateButton.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { IconTranslate } from '@ali/4ever-component';
import { PluginRoles } from '@ali/4ever-bamboo';
;

var _ref = /*#__PURE__*/_createElement(IconTranslate, null);

export default function createStickerButton(props) {
  var controller = props.controller,
      locale = props.locale,
      icon = props.icon,
      title = props.title,
      tooltip = props.tooltip,
      badgeDot = props.badgeDot,
      disabled = props.disabled,
      config = props.config;
  var _config$triggerTransl = config.triggerTranslateSideMenu,
      triggerTranslateSideMenu = _config$triggerTransl === void 0 ? function () {} : _config$triggerTransl;

  var handleClick = function handleClick() {
    triggerTranslateSideMenu();
  };

  var isDisabled = disabled || controller.query('isSelectionInListSymbol') || false;
  var buttonConfig = {
    icon: icon || _ref,
    onClick: handleClick,
    title: title || (locale == null ? void 0 : locale.title),
    tooltip: _extends({
      title: title || (locale == null ? void 0 : locale.title) || ''
    }, tooltip),
    badgeDot: badgeDot,
    testid: "translate_button",
    role: PluginRoles.translate,
    disabled: isDisabled
  };
  return buttonConfig;
}
