// @ali/we-design/dist-web/utils/Animation.js
var Animation = /*#__PURE__*/function () {
  function Animation(options) {
    var _this = this;

    this.timer = 0;
    this.startTime = 0;
    this.duration = void 0;
    this.progress = void 0;
    this.done = void 0;

    this.handleProgress = function () {
      var p = (Date.now() - _this.startTime) / _this.duration;

      if (p >= 1) {
        _this.done(1);
      } else {
        _this.progress(p);

        _this.timer = requestAnimationFrame(_this.handleProgress);
      }
    };

    this.duration = options.duration;
    this.progress = options.progress;
    this.done = options.done;
  }

  var _proto = Animation.prototype;

  _proto.start = function start() {
    this.clear();
    this.startTime = Date.now();
    this.timer = requestAnimationFrame(this.handleProgress);
    return this;
  };

  _proto.clear = function clear() {
    if (this.timer) {
      cancelAnimationFrame(this.timer);
      this.timer = 0;
    }

    return this;
  };

  return Animation;
}();

export { Animation as default };
