// @ali/we-design/dist-web/utils/hotKey.js
export var isEscHotKey = function isEscHotKey(event) {
  var key = event.key,
      keyCode = event.keyCode;

  if (!key || !keyCode) {
    return false;
  }

  if (key.toLowerCase() === 'escape' || keyCode === 27) {
    return true;
  }

  return false;
};
export var isEnterHotKey = function isEnterHotKey(event) {
  var key = event.key,
      keyCode = event.keyCode;

  if (!key || !keyCode) {
    return false;
  }

  if (key.toLowerCase() === 'enter' || keyCode === 13) {
    return true;
  }

  return false;
};
export var isShiftEnterHotKey = function isShiftEnterHotKey(event) {
  var key = event.key,
      keyCode = event.keyCode;

  if (!key || !keyCode) {
    return false;
  }

  if (event.shiftKey && (key.toLowerCase() === 'enter' || keyCode === 13)) {
    return true;
  }

  return false;
};
export var isUpHotKey = function isUpHotKey(event) {
  var key = event.key,
      keyCode = event.keyCode;

  if (!key || !keyCode) {
    return false;
  }

  if (key.toLowerCase() === 'up arrow' || keyCode === 38) {
    return true;
  }

  return false;
};
export var isDownHotKey = function isDownHotKey(event) {
  var key = event.key,
      keyCode = event.keyCode;

  if (!key || !keyCode) {
    return false;
  }

  if (key.toLowerCase() === 'down arrow' || keyCode === 40) {
    return true;
  }

  return false;
};
export var isComposing = function isComposing(event) {
  var _event$nativeEvent;

  return (_event$nativeEvent = event.nativeEvent) == null ? void 0 : _event$nativeEvent.isComposing;
};
