// @ali/4ever-plugin-table/dist-node/bi/mobile/components/floatToolbar/basic.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BasicFloatToolbar = void 0;

var _react = _interopRequireDefault(require("react"));

var _weDesignNext = require("@ali/we-design-next");

const _createElement = /*#__PURE__*/_react.default.createElement;
const {
  Row,
  Divider,
  Slot
} = _weDesignNext.FloatToolbarMobile;

const BasicFloatToolbar = props => {
  const {
    toolbarLayout,
    style
  } = props;
  const shouldRender = toolbarLayout.length > 0;
  return shouldRender ? /*#__PURE__*/_createElement(_weDesignNext.FloatToolbarMobile, {
    style: {
      width: 'fit-content',
      zIndex: 10000,
      position: 'fixed',
      ...style
    }
  }, toolbarLayout.map((configs, i) => /*#__PURE__*/_createElement(_react.default.Fragment, null, /*#__PURE__*/_createElement(Row, {
    key: `row_${i}`
  }, configs.map(config => /*#__PURE__*/_createElement(Slot, {
    key: config.key,
    title: config.title,
    icon: config.icon,
    onClick: config.onClick
  }))), i !== toolbarLayout.length - 1 && /*#__PURE__*/_createElement(Divider, {
    key: `divider_${i}`
  })))) : null;
};

exports.BasicFloatToolbar = BasicFloatToolbar;
