// @ali/4ever-plugin-list/dist-node/bi/regexp/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.STICKER_CODE_REG = exports.ZH_LET_HALF_BRACKET = exports.EN_LET_HALF_BRACKET = exports.ZH_NUM_HALF_BRACKET = exports.EN_NUM_HALF_BRACKET = exports.ZH_NUM_BRACKET = exports.EN_NUM_BRACKET = exports.NUMBER_DON = exports.NUMBER_POINT = exports.ARABIC_REG = exports.UPPER_LETTER_REG = exports.LOWER_LETTER_REG = exports.STATS_WITH_a = exports.STATS_WITH_A = exports.STATS_WITH_1 = void 0;
// 以1. | a. | A. 开始，紧跟字段是中文或英文
const STATS_WITH_1 = /^1\.[\u4e00-\u9fa5|a-zA-Z]+/;
exports.STATS_WITH_1 = STATS_WITH_1;
const STATS_WITH_A = /^A\.[\u4e00-\u9fa5|a-zA-Z]+/;
exports.STATS_WITH_A = STATS_WITH_A;
const STATS_WITH_a = /^a\.[\u4e00-\u9fa5|a-zA-Z]+/;
exports.STATS_WITH_a = STATS_WITH_a;
const LOWER_LETTER_REG = /^[a-z]+[.|。]$/;
exports.LOWER_LETTER_REG = LOWER_LETTER_REG;
const UPPER_LETTER_REG = /^[A-Z]+[.|。]$/;
exports.UPPER_LETTER_REG = UPPER_LETTER_REG;
const ARABIC_REG = /^\d+[.|。｜、]$/;
exports.ARABIC_REG = ARABIC_REG;
const NUMBER_POINT = /^\w+[.|。]$/;
exports.NUMBER_POINT = NUMBER_POINT;
const NUMBER_DON = /^\d+、$/;
exports.NUMBER_DON = NUMBER_DON;
const EN_NUM_BRACKET = /^\(\d+\)$/;
exports.EN_NUM_BRACKET = EN_NUM_BRACKET;
const ZH_NUM_BRACKET = /^（\d+）$/;
exports.ZH_NUM_BRACKET = ZH_NUM_BRACKET;
const EN_NUM_HALF_BRACKET = /^\d+\)$/;
exports.EN_NUM_HALF_BRACKET = EN_NUM_HALF_BRACKET;
const ZH_NUM_HALF_BRACKET = /^\d+）$/;
exports.ZH_NUM_HALF_BRACKET = ZH_NUM_HALF_BRACKET;
const EN_LET_HALF_BRACKET = /^[a-z]+\)$/;
exports.EN_LET_HALF_BRACKET = EN_LET_HALF_BRACKET;
const ZH_LET_HALF_BRACKET = /^[a-z]+）$/;
exports.ZH_LET_HALF_BRACKET = ZH_LET_HALF_BRACKET;
const STICKER_CODE_REG = /^\[.*\]$/;
exports.STICKER_CODE_REG = STICKER_CODE_REG;
