// @ali/we-design/dist-web/components/full-screen/index.js
import _extends from "@babel/runtime/helpers/extends";

var _div;

import React, { useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled, { keyframes } from 'styled-components';
import { SmallScreenNormal } from '@ali/we-icon';
import CSSTransitionGroup from 'react-transition-group/CSSTransitionGroup';
import { PortalComponent } from "../portal-component";
var enterTime = 250;
var leaveTime = 250;
export var FullScreen = /*#__PURE__*/React.memo(function (props) {
  var zIndex = props.zIndex,
      style = props.style,
      logo = props.logo,
      title = props.title,
      headerContent = props.headerContent,
      extra = props.extra,
      children = props.children,
      visible = props.visible,
      onVisibleChange = props.onVisibleChange,
      testid = props.testid,
      getContainer = props.getContainer,
      headerStyle = props.headerStyle,
      bodyStyle = props.bodyStyle;
  var stopEvent = useCallback(function (e) {
    e.stopPropagation(); // e.preventDefault();  如果有这个 会影响 编辑区域的 input 选中等行为
  }, []);
  return /*#__PURE__*/_createElement(PortalComponent, {
    visible: !!visible,
    getContainer: getContainer
  }, /*#__PURE__*/_createElement(FullScreenWrap, {
    "data-testid": testid
  }, /*#__PURE__*/_createElement(CSSTransitionGroup, {
    transitionName: "full-screen-transition",
    transitionEnter: true,
    transitionEnterTimeout: enterTime,
    transitionLeave: true,
    transitionLeaveTimeout: leaveTime
  }, visible && /*#__PURE__*/_createElement("div", {
    className: "full-screen-box",
    style: _extends({
      zIndex: zIndex
    }, style),
    onClick: stopEvent,
    onMouseDown: stopEvent
  }, /*#__PURE__*/_createElement("div", {
    className: "full-screen-box-header",
    style: headerStyle
  }, /*#__PURE__*/_createElement("div", {
    className: "full-screen-box-header-wrap"
  }, /*#__PURE__*/_createElement("div", {
    className: "full-screen-box-header-left"
  }, logo && /*#__PURE__*/_createElement("div", {
    className: "full-screen-box-header-logo"
  }, logo), /*#__PURE__*/_createElement("div", {
    className: "full-screen-box-header-title"
  }, title), headerContent), /*#__PURE__*/_createElement("div", {
    className: "full-screen-box-header-right"
  }, extra)), /*#__PURE__*/_createElement("div", {
    className: "full-screen-box-header-close",
    "data-testid": (testid || '') + "_close",
    onClick: function onClick() {
      return onVisibleChange && onVisibleChange(false);
    }
  }, _div || (_div = /*#__PURE__*/_createElement("div", {
    className: "full-screen-box-header-close-icon-button"
  }, /*#__PURE__*/_createElement(SmallScreenNormal, null))))), /*#__PURE__*/_createElement("div", {
    className: "full-screen-box-body",
    style: bodyStyle
  }, children)))));
});
FullScreen.defaultProps = {
  zIndex: 999
};
var scaleUpCenter = /*#__PURE__*/keyframes(["0%{transform:translateZ(-80px);opacity:0;}100%{transform:translateZ(0);opacity:1;}"]);
var scaleDownCenter = /*#__PURE__*/keyframes(["0%{transform:translateZ(0);opacity:1;}100%{transform:translateZ(-80px);opacity:0;}"]);
var FullScreenWrap = /*#__PURE__*/styled.div([".full-screen-box{position:fixed;top:0;left:0;right:0;bottom:0;background-color:#f5f6f8;display:flex;flex-direction:column;.full-screen-box-header{height:48px;border-bottom:1px solid rgba(126,134,142,0.16);display:flex;padding:8px;background-color:#fff;box-sizing:border-box;.full-screen-box-header-close{font-size:20px;padding:0 20px 0 12px;border-left:1px solid rgba(126,134,142,0.16);display:flex;align-items:center;.full-screen-box-header-close-icon-button{width:28px;height:28px;display:flex;align-items:center;justify-content:center;border-radius:4px;&:hover{background-color:rgba(25,31,37,0.08);}&:active{background-color:rgba(25,31,37,0.16);}.weicon{font-size:20px;}}}.full-screen-box-header-wrap{display:flex;flex:1;padding-left:6px;.full-screen-box-header-left{flex:1;display:flex;align-items:center;.full-screen-box-header-logo{margin-right:8px;}.full-screen-box-header-title{font-size:14px;color:#171a1d;margin-right:9px;font-weight:600;}}.full-screen-box-header-right{display:flex;align-items:center;font-size:12px;}}}.full-screen-box-body{flex:1;}&.full-screen-transition-enter{animation:", " ", "ms cubic-bezier(0.39,0.575,0.565,1);}&.full-screen-transition-leave{animation:", " ", "ms cubic-bezier(0.25,0.46,0.45,0.94) both;}}"], scaleUpCenter, enterTime, scaleDownCenter, leaveTime);
