// @ali/4ever-component/dist-web/Sidebar/Collapse.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { ArrowRightNormal } from '@ali/we-icon';
import { blockEvent } from '@ali/4ever-utils';
import CollapseContent from "./CollapseContent";
var CollapseWrapper = /*#__PURE__*/styled.div([""]);
var CommonContent = /*#__PURE__*/styled.div(["margin-bottom:12px;"]);
var CollapseTitleWrapper = /*#__PURE__*/styled.div(["position:relative;display:flex;align-items:center;cursor:", ";font-weight:500;font-size:14px;margin-bottom:12px;color:rgb(23,26,29);user-select:none;"], function (p) {
  return p.clickable ? 'pointer' : 'auto';
});
var CollapseTitleContent = /*#__PURE__*/styled.div(["flex:1;line-height:22px;"]);
var CollapseTitleFoldIconWrapper = /*#__PURE__*/styled.div(["display:flex;align-items:center;margin-right:6px;.collapse-arrow-icon{transform-origin:center center;transform:", ";transition:transform 0.3 ease;}"], function (p) {
  return !p.fold ? 'rotate(90deg)' : 'none';
});

var _ref = /*#__PURE__*/_createElement(ArrowRightNormal, {
  className: "collapse-arrow-icon"
});

var Collapse = function Collapse(props) {
  var renderTitle = props.renderTitle,
      _props$defaultFold = props.defaultFold,
      defaultFold = _props$defaultFold === void 0 ? true : _props$defaultFold,
      _props$useCollapse = props.useCollapse,
      useCollapse = _props$useCollapse === void 0 ? false : _props$useCollapse;

  var _React$useState = React.useState(useCollapse && defaultFold),
      fold = _React$useState[0],
      setFold = _React$useState[1];

  var handleClickTitle = React.useCallback(function (e) {
    if (!useCollapse) {
      return;
    }

    e.stopPropagation();
    e.preventDefault();
    setFold(!fold);
  }, [fold, useCollapse]);
  return /*#__PURE__*/_createElement(CollapseWrapper, null, /*#__PURE__*/_createElement(CollapseTitleWrapper, {
    onClick: handleClickTitle,
    onMouseDown: blockEvent,
    clickable: useCollapse
  }, useCollapse && /*#__PURE__*/_createElement(CollapseTitleFoldIconWrapper, {
    fold: fold
  }, _ref), /*#__PURE__*/_createElement(CollapseTitleContent, null, renderTitle())), useCollapse ? /*#__PURE__*/_createElement(CollapseContent, {
    fold: fold,
    animate: true,
    style: {
      marginBottom: 12
    }
  }, props.children) : /*#__PURE__*/_createElement(CommonContent, null, props.children));
};

export default Collapse;
