// @ali/4ever-component/dist-web/Sidebar/CollapseContent.js
// 复用的 HeadingGroupPlugin.Content
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import classnames from 'classnames';
var ContentBox = /*#__PURE__*/styled.div(["&.transition{max-height:", "px;transition:all 0.2s ease-in-out;overflow:hidden;&.fold{display:block;}}&.fold{max-height:0;overflow:hidden;display:none;}"], function (props) {
  return props.maxHeight;
});
var PocketBox = /*#__PURE__*/styled.div([":first-child > [data-cangjie-leaf-block] > *{margin-top:0 !important;}& >:first-child > [data-name=\"heading-group-title\"] > [data-name=\"heading-group-title-content\"] > [data-cangjie-leaf-block] > *{margin-top:0 !important;}"]);

var Content = function Content(props) {
  var fold = props.fold,
      _props$animate = props.animate,
      animate = _props$animate === void 0 ? true : _props$animate,
      style = props.style,
      className = props.className,
      dataName = props.dataName,
      children = props.children;
  var foldRef = React.useRef(fold);
  var contentRef = React.useRef(null);
  var pocketRef = React.useRef(null);
  var heightRef = React.useRef(0);

  var _React$useState = React.useState(fold),
      internalFold = _React$useState[0],
      setInternalFold = _React$useState[1];

  var _React$useState2 = React.useState(false),
      transition = _React$useState2[0],
      setTransition = _React$useState2[1];
  /**
   * 动画结束后移除动画类名
   */


  var handleTransitionEnd = React.useCallback(function () {
    setTransition(false);
  }, []);
  /**
   * 每次更新后刷新内容高度
   */

  React.useLayoutEffect(function () {
    if (pocketRef.current) {
      var pocket = pocketRef.current;
      heightRef.current = pocket.offsetHeight;
    }
  });
  /**
   * 折叠状态发生变化时，先开启 transition，再变更状态
   */

  React.useLayoutEffect(function () {
    if (fold !== foldRef.current) {
      if (animate) {
        setTransition(true);
      } else {
        setTransition(false);
        setInternalFold(fold);
      }

      foldRef.current = fold;
    }
  }, [animate, fold]);
  /**
   * 开启 transition 后变更 fold 状态
   */

  React.useLayoutEffect(function () {
    if (transition) {
      setInternalFold(fold);
    }
  }, [transition, fold]);
  var classNames = classnames({
    transition: transition,
    fold: internalFold
  });
  return /*#__PURE__*/_createElement(ContentBox, {
    ref: contentRef,
    className: classNames,
    style: style,
    maxHeight: heightRef.current,
    onTransitionEnd: handleTransitionEnd
  }, /*#__PURE__*/_createElement(PocketBox, {
    ref: pocketRef,
    "data-name": dataName,
    "data-heading-fold": fold,
    className: className
  }, children));
};

Content.displayName = 'Content';
export default Content;
