// @ali/4ever-component/dist-web/Sidebar/select.js
import React, { useState, useEffect } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import classnames from 'classnames';
import { blockEvent } from "./utils";
var SelectItems = /*#__PURE__*/styled.div(["display:flex;flex-direction:row;margin:0 4px;color:#171a1d;"]);
export var SelectItem = /*#__PURE__*/styled.div(["display:inline-block;min-width:80px;background-color:#f9fafc;text-align:center;border:1px solid rgba(126,134,142,0.16);border-radius:4px;margin:0 4px;cursor:default;overflow:hidden;user-select:none;&.selected{box-shadow:0px 0px 0px 2px rgba(0,106,254,0.12);background-color:#ffffff;border:1px solid #0089ff;}&:hover .title{background-color:rgba(126,134,142,0.12);}&.disabled{color:rgba(17,31,44,0.24) !important;cursor:not-allowed;.title{background-color:rgba(126,134,142,0.12);}}"]);
export var SelectItemTitle = /*#__PURE__*/styled.div(["font-size:12px;padding:0px 6px;white-space:nowrap;height:36px;display:flex;align-items:center;justify-content:left;"]);
export var SelectItemIcon = /*#__PURE__*/styled.span(["font-size:0;margin-right:4px;.weicon{font-size:20px;}"]);
export var SelectItemSubTitle = /*#__PURE__*/styled.div(["font-size:12px;color:rgba(17,31,44,0.56);background-color:#f6f6f7;height:16px;line-height:16px;border-radius:0px 0px 3px 3px;.subtitle{display:inline-block;transform:scale(0.9);}"]);
var TitleText = /*#__PURE__*/styled.span(["overflow:hidden;text-overflow:ellipsis;line-height:1;"]);

var StylesSelect = function StylesSelect(props) {
  var items = props.items,
      onChange = props.onChange,
      value = props.value,
      defaultValue = props.defaultValue,
      disabled = props.disabled;

  var _useState = useState(value || defaultValue || ''),
      val = _useState[0],
      setVal = _useState[1];

  var handleChange = function handleChange(v) {
    if (v !== val) {
      setVal(v);
      onChange(v);
    }
  };

  useEffect(function () {
    setVal(value || defaultValue || '');
  }, [value, defaultValue]);
  var selectItemStyle = React.useMemo(function () {
    return {
      width: 1 / items.length * 100 + "%"
    };
  }, [items.length]);
  var content = items.map(function (item) {
    return /*#__PURE__*/_createElement(SelectItem, {
      className: classnames({
        selected: item.key === val,
        disabled: disabled || false
      }),
      key: item.key,
      onClick: function onClick() {
        return handleChange(item.key);
      },
      onMouseDown: blockEvent,
      style: selectItemStyle
    }, /*#__PURE__*/_createElement(SelectItemTitle, {
      className: "title",
      style: item.titleStyle
    }, item.icon && /*#__PURE__*/_createElement(SelectItemIcon, null, item.icon), /*#__PURE__*/_createElement(TitleText, null, item.title)), item.subtitle && /*#__PURE__*/_createElement(SelectItemSubTitle, null, /*#__PURE__*/_createElement("span", {
      className: "subtitle"
    }, item.subtitle)));
  });
  return /*#__PURE__*/_createElement(SelectItems, null, content);
};

export default StylesSelect;
