// @ali/4ever-bi/dist-web/plugins/link/components/AnimatePortal.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled, { keyframes, css } from 'styled-components';
import { Transition } from 'react-transition-group';
import { Portal } from '@ali/4ever-component';
var fadeInTop = /*#__PURE__*/keyframes(["0%{transform:translateY(-12px);opacity:0;}100%{transform:translateY(0);opacity:1;}"]);
var fadeInTopAnimation = /*#__PURE__*/css(["animation:", " 0.25s cubic-bezier(0.455,0.030,0.515,0.955) both;"], fadeInTop);
var fadeOut = /*#__PURE__*/keyframes(["0%{opacity:1;}100%{opacity:0;}"]);
var fadeOutAnimation = /*#__PURE__*/css(["width:fit-content;height:fit-content;animation:", " 0.15s ease-out both;"], fadeOut);
var AnimateWrapper = /*#__PURE__*/styled.div(["position:absolute;", " ", ""], fadeInTopAnimation, function (props) {
  return (props.state === 'exiting' || props.state === 'exited') && fadeOutAnimation;
});

var AnimatePortal = function AnimatePortal(props) {
  var children = props.children,
      overlay = props.overlay,
      visible = props.visible,
      onOverlayHide = props.onOverlayHide,
      rest = _objectWithoutPropertiesLoose(props, ["children", "overlay", "visible", "onOverlayHide"]);

  var animatedOverlay = /*#__PURE__*/_createElement(Transition, {
    "in": visible,
    timeout: {
      enter: 250,
      exit: 150
    },
    onExited: onOverlayHide,
    unmountOnExit: true
  }, function (state) {
    return /*#__PURE__*/_createElement(AnimateWrapper, {
      state: state
    }, /*#__PURE__*/React.cloneElement(overlay));
  });

  return /*#__PURE__*/_createElement(Portal, _extends({}, rest, {
    overlay: animatedOverlay,
    visible: visible,
    destroyOverlayOnHide: false
  }), children);
};

export default AnimatePortal;
