// @ali/4ever-bi/dist-web/plugins/link/components/linkMenu.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { MenuItem, InlineMenuItem } from '@ali/we-design';
import { PluginRoles } from '@ali/4ever-bamboo';
import { wrapLink } from "../actions";
import { LINK_INSERT_HOT_KEY_TIP } from "../utils/hotKeyTips";

function withLinkMenu(Component) {
  return function Menu(props) {
    var menuItem = props.menuItem,
        controller = props.controller,
        locale = props.locale;

    var title = menuItem.title,
        rest = _objectWithoutPropertiesLoose(menuItem, ["title"]);

    var handleClick = function handleClick() {
      return controller.run('onAction', wrapLink());
    };

    return /*#__PURE__*/_createElement(Component, _extends({
      extra: LINK_INSERT_HOT_KEY_TIP
    }, rest, {
      onClick: handleClick,
      role: PluginRoles.link
    }), title || (locale == null ? void 0 : locale.buttonTitle));
  };
}

export var menu = withLinkMenu(MenuItem);
export var inlineMenu = withLinkMenu(InlineMenuItem);
