// @ali/4ever-plugin-table/dist-node/bi/mobile/components/toolbar/mobileColToolbar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MobileColToolbar = void 0;

var _react = _interopRequireDefault(require("react"));

var _tableColToolbarItem = require("../../../components/colToolbar/tableColToolbarItem");

var _constants = require("../../../constants");

var _hooks = require("../../../../utils/hooks");

var _actions = require("../../../actions");

var _colToolbarDragButton = require("./colToolbarDragButton");

const _createElement = /*#__PURE__*/_react.default.createElement;

const MobileColToolbar = props => {
  const {
    activeRange,
    selectRange,
    controller,
    visible,
    table,
    style: propsStyle
  } = props;
  const [colsWidth] = (0, _hooks.usePixelColsWidth)();

  const preventEvent = e => {
    e.stopPropagation();
    e.preventDefault();
  };

  const stopPropagation = e => {
    e.stopPropagation();
  };

  const handleSelectCol = (e, i) => {
    controller.run('onAction', (0, _actions.selectTableCols)(table, i, i));
    setTimeout(() => {
      controller.setIsFocused(false);
    });
  };

  const style = { ...propsStyle,
    visibility: visible ? 'visible' : 'hidden',
    width: colsWidth.reduce((acc, w) => acc + w, 0) + 1,
    height: _constants.TOOLBAR_ITEM_SIZE_MOBILE
  };
  return /*#__PURE__*/_createElement("div", {
    style: style
  }, colsWidth.map((w, i) => /*#__PURE__*/_createElement(_tableColToolbarItem.TableColToolbarItem, {
    style: {
      height: _constants.TOOLBAR_ITEM_SIZE_MOBILE,
      borderTopLeftRadius: i === 0 ? 4 : 0,
      borderTopRightRadius: i === colsWidth.length - 1 ? 4 : 0
    },
    key: `${i}_${w}`,
    selectRanges: selectRange ? [selectRange] : null,
    activeRange: activeRange,
    index: i,
    onClick: e => handleSelectCol(e, i),
    onTouchStart: preventEvent,
    onMouseDown: stopPropagation
  }, props.children, i === selectRange?.end && /*#__PURE__*/_createElement(_colToolbarDragButton.ColToolbarDragButton, {
    table: table,
    controller: controller,
    index: i
  }))));
};

exports.MobileColToolbar = MobileColToolbar;
