// @ali/4ever-bi/dist-node/plugins/formula/utils/importFont.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dynamicalFontImport = exports.FONTS = void 0;

var _KaTeX_AMSRegular = _interopRequireDefault(require("../fonts/KaTeX_AMS-Regular.woff2"));

var _KaTeX_AMSRegular2 = _interopRequireDefault(require("../fonts/KaTeX_AMS-Regular.woff"));

var _KaTeX_AMSRegular3 = _interopRequireDefault(require("../fonts/KaTeX_AMS-Regular.ttf"));

var _KaTeX_CaligraphicBold = _interopRequireDefault(require("../fonts/KaTeX_Caligraphic-Bold.woff2"));

var _KaTeX_CaligraphicBold2 = _interopRequireDefault(require("../fonts/KaTeX_Caligraphic-Bold.woff"));

var _KaTeX_CaligraphicBold3 = _interopRequireDefault(require("../fonts/KaTeX_Caligraphic-Bold.ttf"));

var _KaTeX_CaligraphicRegular = _interopRequireDefault(require("../fonts/KaTeX_Caligraphic-Regular.woff2"));

var _KaTeX_CaligraphicRegular2 = _interopRequireDefault(require("../fonts/KaTeX_Caligraphic-Regular.woff"));

var _KaTeX_CaligraphicRegular3 = _interopRequireDefault(require("../fonts/KaTeX_Caligraphic-Regular.ttf"));

var _KaTeX_FrakturBold = _interopRequireDefault(require("../fonts/KaTeX_Fraktur-Bold.woff2"));

var _KaTeX_FrakturBold2 = _interopRequireDefault(require("../fonts/KaTeX_Fraktur-Bold.woff"));

var _KaTeX_FrakturBold3 = _interopRequireDefault(require("../fonts/KaTeX_Fraktur-Bold.ttf"));

var _KaTeX_FrakturRegular = _interopRequireDefault(require("../fonts/KaTeX_Fraktur-Regular.woff2"));

var _KaTeX_FrakturRegular2 = _interopRequireDefault(require("../fonts/KaTeX_Fraktur-Regular.woff"));

var _KaTeX_FrakturRegular3 = _interopRequireDefault(require("../fonts/KaTeX_Fraktur-Regular.ttf"));

var _KaTeX_MainBold = _interopRequireDefault(require("../fonts/KaTeX_Main-Bold.woff2"));

var _KaTeX_MainBold2 = _interopRequireDefault(require("../fonts/KaTeX_Main-Bold.woff"));

var _KaTeX_MainBold3 = _interopRequireDefault(require("../fonts/KaTeX_Main-Bold.ttf"));

var _KaTeX_MainBoldItalic = _interopRequireDefault(require("../fonts/KaTeX_Main-BoldItalic.woff2"));

var _KaTeX_MainBoldItalic2 = _interopRequireDefault(require("../fonts/KaTeX_Main-BoldItalic.woff"));

var _KaTeX_MainBoldItalic3 = _interopRequireDefault(require("../fonts/KaTeX_Main-BoldItalic.ttf"));

var _KaTeX_MainItalic = _interopRequireDefault(require("../fonts/KaTeX_Main-Italic.woff2"));

var _KaTeX_MainItalic2 = _interopRequireDefault(require("../fonts/KaTeX_Main-Italic.woff"));

var _KaTeX_MainItalic3 = _interopRequireDefault(require("../fonts/KaTeX_Main-Italic.ttf"));

var _KaTeX_MainRegular = _interopRequireDefault(require("../fonts/KaTeX_Main-Regular.woff2"));

var _KaTeX_MainRegular2 = _interopRequireDefault(require("../fonts/KaTeX_Main-Regular.woff"));

var _KaTeX_MainRegular3 = _interopRequireDefault(require("../fonts/KaTeX_Main-Regular.ttf"));

var _KaTeX_MathBoldItalic = _interopRequireDefault(require("../fonts/KaTeX_Math-BoldItalic.woff2"));

var _KaTeX_MathBoldItalic2 = _interopRequireDefault(require("../fonts/KaTeX_Math-BoldItalic.woff"));

var _KaTeX_MathBoldItalic3 = _interopRequireDefault(require("../fonts/KaTeX_Math-BoldItalic.ttf"));

var _KaTeX_MathItalic = _interopRequireDefault(require("../fonts/KaTeX_Math-Italic.woff2"));

var _KaTeX_MathItalic2 = _interopRequireDefault(require("../fonts/KaTeX_Math-Italic.woff"));

var _KaTeX_MathItalic3 = _interopRequireDefault(require("../fonts/KaTeX_Math-Italic.ttf"));

var _KaTeX_SansSerifBold = _interopRequireDefault(require("../fonts/KaTeX_SansSerif-Bold.woff2"));

var _KaTeX_SansSerifBold2 = _interopRequireDefault(require("../fonts/KaTeX_SansSerif-Bold.woff"));

var _KaTeX_SansSerifBold3 = _interopRequireDefault(require("../fonts/KaTeX_SansSerif-Bold.ttf"));

var _KaTeX_SansSerifItalic = _interopRequireDefault(require("../fonts/KaTeX_SansSerif-Italic.woff2"));

var _KaTeX_SansSerifItalic2 = _interopRequireDefault(require("../fonts/KaTeX_SansSerif-Italic.woff"));

var _KaTeX_SansSerifItalic3 = _interopRequireDefault(require("../fonts/KaTeX_SansSerif-Italic.ttf"));

var _KaTeX_SansSerifRegular = _interopRequireDefault(require("../fonts/KaTeX_SansSerif-Regular.woff2"));

var _KaTeX_SansSerifRegular2 = _interopRequireDefault(require("../fonts/KaTeX_SansSerif-Regular.woff"));

var _KaTeX_SansSerifRegular3 = _interopRequireDefault(require("../fonts/KaTeX_SansSerif-Regular.ttf"));

var _KaTeX_ScriptRegular = _interopRequireDefault(require("../fonts/KaTeX_Script-Regular.woff2"));

var _KaTeX_ScriptRegular2 = _interopRequireDefault(require("../fonts/KaTeX_Script-Regular.woff"));

var _KaTeX_ScriptRegular3 = _interopRequireDefault(require("../fonts/KaTeX_Script-Regular.ttf"));

var _KaTeX_Size1Regular = _interopRequireDefault(require("../fonts/KaTeX_Size1-Regular.woff2"));

var _KaTeX_Size1Regular2 = _interopRequireDefault(require("../fonts/KaTeX_Size1-Regular.woff"));

var _KaTeX_Size1Regular3 = _interopRequireDefault(require("../fonts/KaTeX_Size1-Regular.ttf"));

var _KaTeX_Size2Regular = _interopRequireDefault(require("../fonts/KaTeX_Size2-Regular.woff2"));

var _KaTeX_Size2Regular2 = _interopRequireDefault(require("../fonts/KaTeX_Size2-Regular.woff"));

var _KaTeX_Size2Regular3 = _interopRequireDefault(require("../fonts/KaTeX_Size2-Regular.ttf"));

var _KaTeX_Size3Regular = _interopRequireDefault(require("../fonts/KaTeX_Size3-Regular.woff2"));

var _KaTeX_Size3Regular2 = _interopRequireDefault(require("../fonts/KaTeX_Size3-Regular.woff"));

var _KaTeX_Size3Regular3 = _interopRequireDefault(require("../fonts/KaTeX_Size3-Regular.ttf"));

var _KaTeX_Size4Regular = _interopRequireDefault(require("../fonts/KaTeX_Size4-Regular.woff2"));

var _KaTeX_Size4Regular2 = _interopRequireDefault(require("../fonts/KaTeX_Size4-Regular.woff"));

var _KaTeX_Size4Regular3 = _interopRequireDefault(require("../fonts/KaTeX_Size4-Regular.ttf"));

var _KaTeX_TypewriterRegular = _interopRequireDefault(require("../fonts/KaTeX_Typewriter-Regular.woff2"));

var _KaTeX_TypewriterRegular2 = _interopRequireDefault(require("../fonts/KaTeX_Typewriter-Regular.woff"));

var _KaTeX_TypewriterRegular3 = _interopRequireDefault(require("../fonts/KaTeX_Typewriter-Regular.ttf"));

/* eslint-disable max-len */

/* eslint-disable camelcase */
const FONTS = [{
  family: 'KaTeX_AMS',
  source: `url(${_KaTeX_AMSRegular.default}) format('woff2'), url(${_KaTeX_AMSRegular2.default}) format('woff'), url(${_KaTeX_AMSRegular3.default}) format('truetype')`,
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Caligraphic',
  source: `url(${_KaTeX_CaligraphicBold.default}) format('woff2'), url(${_KaTeX_CaligraphicBold2.default}) format('woff'), url(${_KaTeX_CaligraphicBold3.default}) format('truetype')`,
  descriptors: {
    weight: 'bold',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Caligraphic',
  source: `url(${_KaTeX_CaligraphicRegular.default}) format('woff2'), url(${_KaTeX_CaligraphicRegular2.default}) format('woff'), url(${_KaTeX_CaligraphicRegular3.default}) format('truetype') `,
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Fraktur',
  source: `url(${_KaTeX_FrakturBold.default}) format('woff2'), url(${_KaTeX_FrakturBold2.default}) format('woff'), url(${_KaTeX_FrakturBold3.default}) format('truetype')`,
  descriptors: {
    weight: 'bold',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Fraktur',
  source: `url(${_KaTeX_FrakturRegular.default}) format('woff2'), url(${_KaTeX_FrakturRegular2.default}) format('woff'), url(${_KaTeX_FrakturRegular3.default}) format('truetype')`,
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Main',
  source: `url(${_KaTeX_MainBold.default}) format('woff2'), url(${_KaTeX_MainBold2.default}) format('woff'), url(${_KaTeX_MainBold3.default}) format('truetype')`,
  descriptors: {
    weight: 'bold',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Main',
  source: `url(${_KaTeX_MainBoldItalic.default}) format('woff2'), url(${_KaTeX_MainBoldItalic2.default}) format('woff'), url(${_KaTeX_MainBoldItalic3.default}) format('truetype')`,
  descriptors: {
    weight: 'bold',
    style: 'italic'
  }
}, {
  family: 'KaTeX_Main',
  source: `url(${_KaTeX_MainItalic.default}) format('woff2'), url(${_KaTeX_MainItalic2.default}) format('woff'), url(${_KaTeX_MainItalic3.default}) format('truetype')`,
  descriptors: {
    weight: 'normal',
    style: 'italic'
  }
}, {
  family: 'KaTeX_Main',
  source: `url(${_KaTeX_MainRegular.default}) format('woff2'), url(${_KaTeX_MainRegular2.default}) format('woff'), url(${_KaTeX_MainRegular3.default}) format('truetype')`,
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Math',
  source: `url(${_KaTeX_MathBoldItalic.default}) format('woff2'), url(${_KaTeX_MathBoldItalic2.default}) format('woff'), url(${_KaTeX_MathBoldItalic3.default}) format('truetype')`,
  descriptors: {
    weight: 'bold',
    style: 'italic'
  }
}, {
  family: 'KaTeX_Math',
  source: `url(${_KaTeX_MathItalic.default}) format('woff2'), url(${_KaTeX_MathItalic2.default}) format('woff'), url(${_KaTeX_MathItalic3.default}) format('truetype')`,
  descriptors: {
    weight: 'normal',
    style: 'italic'
  }
}, {
  family: 'KaTeX_SansSerif',
  source: `url(${_KaTeX_SansSerifBold.default}) format('woff2'), url(${_KaTeX_SansSerifBold2.default}) format('woff'), url(${_KaTeX_SansSerifBold3.default}) format('truetype')`,
  descriptors: {
    weight: 'bold',
    style: 'normal'
  }
}, {
  family: 'KaTeX_SansSerif',
  source: `url(${_KaTeX_SansSerifItalic.default}) format('woff2'), url(${_KaTeX_SansSerifItalic2.default}) format('woff'), url(${_KaTeX_SansSerifItalic3.default}) format('truetype')`,
  descriptors: {
    weight: 'normal',
    style: 'italic'
  }
}, {
  family: 'KaTeX_SansSerif',
  source: `url(${_KaTeX_SansSerifRegular.default}) format('woff2'), url(${_KaTeX_SansSerifRegular2.default}) format('woff'), url(${_KaTeX_SansSerifRegular3.default}) format('truetype')`,
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Script',
  source: `url(${_KaTeX_ScriptRegular.default}) format('woff2'), url(${_KaTeX_ScriptRegular2.default}) format('woff'), url(${_KaTeX_ScriptRegular3.default}) format('truetype')`,
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Size1',
  source: `url(${_KaTeX_Size1Regular.default}) format('woff2'), url(${_KaTeX_Size1Regular2.default}) format('woff'), url(${_KaTeX_Size1Regular3.default}) format('truetype')`,
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Size2',
  source: `url(${_KaTeX_Size2Regular.default}) format('woff2'), url(${_KaTeX_Size2Regular2.default}) format('woff'), url(${_KaTeX_Size2Regular3.default}) format('truetype')`,
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Size3',
  source: `url(${_KaTeX_Size3Regular.default}) format('woff2'), url(${_KaTeX_Size3Regular2.default}) format('woff'), url(${_KaTeX_Size3Regular3.default}) format('truetype')`,
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Size4',
  source: `url(${_KaTeX_Size4Regular.default}) format('woff2'), url(${_KaTeX_Size4Regular2.default}) format('woff'), url(${_KaTeX_Size4Regular3.default}) format('truetype')`,
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}, {
  family: 'KaTeX_Typewriter',
  source: `url(${_KaTeX_TypewriterRegular.default}) format('woff2'), url(${_KaTeX_TypewriterRegular2.default}) format('woff'), url(${_KaTeX_TypewriterRegular3.default}) format('truetype')`,
  descriptors: {
    weight: 'normal',
    style: 'normal'
  }
}];
exports.FONTS = FONTS;

function importFont() {
  const styleElement = document.createElement('style');
  document.head.appendChild(styleElement);
  const styleSheet = styleElement.sheet;
  return function insertRule(style) {
    if (!styleSheet) {
      return;
    }

    if (typeof style === 'string') {
      styleSheet.insertRule(style, 0);
      return;
    }

    if (typeof style === 'object') {
      const {
        family,
        source,
        descriptors
      } = style;
      const rule = `
        @font-face {
          font-family: ${family};
          font-style: ${descriptors.style};
          font-weight: ${descriptors.weight};
          src: ${source};
        }
      `;
      styleSheet.insertRule(rule, 0);
    }
  };
}

const dynamicalFontImport = importFont();
exports.dynamicalFontImport = dynamicalFontImport;
