// @ali/4ever-zhi/dist-node/plugins/callout/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everBamboo = require("@ali/4ever-bamboo");

var _createRenderNode = _interopRequireDefault(require("./handlers/createRenderNode"));

var _createGroupNodes = _interopRequireDefault(require("./handlers/createGroupNodes"));

var _everPluginColorBlocks = require("@ali/4ever-plugin-color-blocks");

const {
  DEFAULT_STICKERS,
  Sticker
} = _everBamboo.StickerPlugin;

function createCalloutPlugin(config) {
  const {
    callout = {}
  } = config;
  const stickerConfig = {
    // @ts-ignore
    lang: 'zh_CN',
    // @ts-ignore
    customSticker: [DEFAULT_STICKERS],
    ...config.sticker
  };
  const sticker = new Sticker(stickerConfig);
  const params = {
    stickerConfig,
    sticker
  };
  return { ...(0, _everPluginColorBlocks.moFactory)({}),
    queries: {
      isNodeInColorBlocks: _everBamboo.CalloutPlugin.isNodeInColorBlocks
    },
    renderNode: (0, _createRenderNode.default)(params),
    getNodeStyle: _everBamboo.CalloutPlugin.getNodeStyle,
    groupNodes: (0, _createGroupNodes.default)(callout)
  };
}

var _default = createCalloutPlugin;
exports.default = _default;
