// @ali/4ever-bi/dist-web/plugins/clipboard/mobile.js
// 兼容
import ClipboardManager from '@ali/clipboard-manager';
import { mobileBiFactory, ClipboardPad // actions
, CLIPBOARD_INSERT_FRAGMENT, CLIPBOARD_SET_DECORATIONS, CLIPBOARD_DEL_SELECTION, CLIPBOARD_HIDE_PASTEPANEL, CLIPBOARD_HIDE_CUSTOM_PASTEPANEL, CLIPBOARD_MATCH_MODE, CLIPBOARD_TEXT_MODE, insertFragment, setDecorations, deleteAtRange, hidePastePanel, hideCustomPastePanel, insertMatchFragment, insertTextFragment // handler...
, copyNodes } from '@ali/4ever-plugin-clipboard';

var createClipboardPlugin = function createClipboardPlugin(configs) {
  return mobileBiFactory(configs);
};

var actions = {
  CLIPBOARD_INSERT_FRAGMENT: CLIPBOARD_INSERT_FRAGMENT,
  CLIPBOARD_SET_DECORATIONS: CLIPBOARD_SET_DECORATIONS,
  CLIPBOARD_DEL_SELECTION: CLIPBOARD_DEL_SELECTION,
  CLIPBOARD_HIDE_PASTEPANEL: CLIPBOARD_HIDE_PASTEPANEL,
  CLIPBOARD_HIDE_CUSTOM_PASTEPANEL: CLIPBOARD_HIDE_CUSTOM_PASTEPANEL,
  CLIPBOARD_MATCH_MODE: CLIPBOARD_MATCH_MODE,
  CLIPBOARD_TEXT_MODE: CLIPBOARD_TEXT_MODE,
  insertFragment: insertFragment,
  setDecorations: setDecorations,
  deleteAtRange: deleteAtRange,
  hidePastePanel: hidePastePanel,
  hideCustomPastePanel: hideCustomPastePanel,
  insertMatchFragment: insertMatchFragment,
  insertTextFragment: insertTextFragment
};
createClipboardPlugin.actions = actions; // hetu 使用；其他的 urls 引入较多非移动端用内容，暂时不导出

createClipboardPlugin.utils = {
  copyNodes: copyNodes
};
createClipboardPlugin.ClipboardPad = ClipboardPad;
createClipboardPlugin.ClipboardManager = ClipboardManager;
export default createClipboardPlugin;
