// @ali/4ever-plugin-mention/dist-web/bi/mobile/index.js
import _extends from "@babel/runtime/helpers/extends";
import { Commands } from '@ali/4ever-cangjie';
import toolbar from "./toolbar";
import renderNode from "./renderNode";
import onCangjieInput from "./onCangjieInput";
import onCangjieBlur from "./onCangjieBlur";
import insertMention from "./insertMention";
import { isSelectionInMention } from "../queries/isSelectionInMention";
import getMentionKeyword from "../queries/getMentionKeyword";
import insertMentionMark from "../commands/insertMention";
import { actions, isActions, MENTION_CANCEL_SELECT, MENTION_INSERT, MENTION_SELECT } from "../actions";
import removeMention from "./removeMention";
import { moFactory } from "../../mo";
import renderMark from "./renderMark";
export var mobileBiFactory = function mobileBiFactory(configs) {
  var mentionConfig = configs.mention || {};
  var mentionId = mentionConfig.getMentionId ? mentionConfig.getMentionId() : '';
  var extraItems = {};

  if (mentionConfig.renderMentionPanel) {
    // @ts-ignore
    extraItems.renderMark = renderMark(mentionConfig.renderMentionPanel);
  }

  return _extends({}, moFactory({}), {
    toolbar: toolbar(mentionConfig),
    commands: {
      insertMention: insertMention
    },
    actions: actions,
    isActions: isActions,
    renderNode: renderNode(mentionId, mentionConfig),
    onCangjieInput: onCangjieInput,
    onCangjieBlur: onCangjieBlur,
    onAction: function onAction(action, controller, next) {
      var type = action.type,
          payload = action.payload;

      if (type === MENTION_SELECT) {
        insertMentionMark(controller);
        return controller;
      }

      if (type === MENTION_INSERT) {
        var _ref = payload,
            suggestion = _ref.suggestion;
        controller.command(Commands.focus);
        return controller.command(insertMention, suggestion);
      }

      if (type === MENTION_CANCEL_SELECT) {
        return controller.command(removeMention);
      }

      return next();
    },
    queries: {
      isSelectionInMention: isSelectionInMention,
      getMentionKeyword: getMentionKeyword
    }
  }, extraItems);
};
