// @ali/4ever-plugin-mention/dist-web/bi/mobile/insertMention.js
import { Commands } from '@ali/4ever-cangjie';
import removeMention from "./removeMention";
import { nearsetAt } from "../utils";
export default (function (controller, results) {
  if (!results) return controller;
  var value = controller.value;

  if (controller.query('isCollapsedInInlineVoid')) {
    controller.command(Commands.moveToStartOfNextText);
  }

  var selection = value.selection,
      document = value.document;
  var start = selection.getStart(document);
  var at = nearsetAt(value.endText.text, start.offset);
  var mentionLength = start.offset - at;
  controller.command(Commands.deleteBackward, mentionLength);
  results.forEach(function (element) {
    controller.command(Commands.insertInline, {
      type: 'mention',
      data: {
        id: element.id,
        login: element.uid,
        name: element.name,
        avatarUrl: element.avatar
      }
    });
  });
  controller.command(Commands.moveToStartOfNextText).command(Commands.focus);
  return removeMention(controller);
});
