// @ali/4ever-plugin-embed/dist-web/utils/components/common/embedCard/styled.js
import styled from 'styled-components';
export var CardContentInnerWrapper = /*#__PURE__*/styled.div(["display:flex;margin-top:'';"]);
export var MobileCardContentInnerWrapper = /*#__PURE__*/styled.div(["display:block;"]);
export var CardContentWrapper = /*#__PURE__*/styled.div(["flex:1;min-width:0;"]);
export var ContentHeader = /*#__PURE__*/styled.div(["font-size:14px;color:", ";flex:1;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;"], function (props) {
  return props.mode === 'dark' ? '#F2F4F5' : '#111F2C';
});
export var ContentDesc = /*#__PURE__*/styled.div(["font-size:12px;color:", ";margin-right:16px;"], function (props) {
  return props.mode === 'dark' ? '#F2F4F5' : '#7A8189';
});
export var MobileEmbedCardWrapper = /*#__PURE__*/styled.div(["width:100%;height:54px;border:1px solid #D9DBDD;padding:8px 0 8px 16px;display:flex;border-radius:6px;background-color:#fff;box-sizing:border-box;"]);
export var CardIconWrapper = /*#__PURE__*/styled.div(["margin-right:16px;flex:0;margin-top:6px;"]);
export var PreviewIconWrapper = /*#__PURE__*/styled.div(["flex:0;padding:8px 16px;border-left:1px solid #E2E4E5;cursor:pointer;"]);
export var ProgressBar = /*#__PURE__*/styled.div(["width:100%;height:4px;background-color:#D9DBDD;border-radius:4px;"]);
export var ProgressBarInner = /*#__PURE__*/styled.div(["background-color:#0089FF;width:", ";height:100%;transition:width .6s;border-radius:4px;"], function (props) {
  return props.progress + "%";
});
export var ErrorDesc = /*#__PURE__*/styled.div(["font-size:12px;color:#D0021B;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;"]);
