// @ali/4ever-bi/dist-node/plugins/fonts/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _getFonts = _interopRequireDefault(require("./queries/getFonts"));

var actions = _interopRequireWildcard(require("./actions"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _createRenderToolbar = _interopRequireDefault(require("./createRenderToolbar"));

var _renderEditable = _interopRequireDefault(require("./renderEditable"));

const queries = {
  getFonts: _getFonts.default
};

function createFontsPlugin(config) {
  const fontsConfig = config.fonts || {};
  return {
    toolbar: (0, _toolbar.default)(fontsConfig),
    renderToolbar: (0, _createRenderToolbar.default)(fontsConfig.locale),
    renderEditable: _renderEditable.default,
    queries,
    onAction: _onAction.default,
    models: {
      klass: 'mark',
      type: 'fonts'
    }
  };
}

createFontsPlugin.actions = actions;
createFontsPlugin.queries = queries;
var _default = createFontsPlugin;
exports.default = _default;
