// @ali/4ever-bi/dist-node/plugins/subject/handlers/createRenderNode.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _subject = _interopRequireDefault(require("../components/subject"));

var _utils = require("../utils");

const _createElement = /*#__PURE__*/React.createElement;

function createRenderNode(config) {
  return function renderNode(props, controller, next) {
    const {
      node,
      attributes,
      isSelected,
      children
    } = props;

    if ((0, _utils.isSubject)(node)) {
      return /*#__PURE__*/_createElement(_subject.default, {
        controller: controller,
        node: node,
        attributes: attributes,
        isSelected: isSelected,
        config: config
      }, children());
    }

    return next();
  };
}

var _default = createRenderNode;
exports.default = _default;
