// @ali/4ever-plugin-heading/dist-web/mo/serializer/jsonMLToValue.js
import { Paragraph } from '@ali/4ever-plugin-paragraph';
import { getTagName, getAttributes } from '@ali/4ever-utils';
import Heading from "../models";
var tagToFactory = {
  h1: Heading.Heading1.create,
  h2: Heading.Heading2.create,
  h3: Heading.Heading3.create,
  h4: Heading.Heading4.create,
  h5: Heading.Heading5.create,
  h6: Heading.Heading6.create
};
export default {
  match: function match(node) {
    var tag = getTagName(node);
    return tagToFactory[tag];
  },
  convert: function convert(node) {
    var data = getAttributes(node) || {};

    if (data.ind && typeof data.ind.firstline === 'number') {
      data.ind.firstLine = data.ind.firstline;
      delete data.ind.firstline;
    } // 序列化时对不合法 level 进行矫正


    if (data.list) {
      var hasInvalidLevel = function hasInvalidLevel(list) {
        var level = list.level;
        return !/\d/g.test(String(level)) || level < 0;
      };

      if (hasInvalidLevel(data.list)) {
        data.list.level = 0;
      }
    }

    var factory = tagToFactory[getTagName(node)] || Paragraph.create;
    return factory({
      data: data
    });
  }
};
