// @ali/4ever-plugin-embed/dist-node/utils/bamboo.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  AudioPlayer: true,
  AudioPreview: true,
  VideoPlayer: true,
  EmbedFileCardWrapper: true,
  EmbedCard: true,
  MobileEmbedCard: true,
  EmbedDownloadButton: true,
  EmbedPreviewHeader: true,
  EmbedFailedCard: true,
  EmbedUploadProgressCard: true
};
Object.defineProperty(exports, "AudioPlayer", {
  enumerable: true,
  get: function () {
    return _audioPlayer.default;
  }
});
Object.defineProperty(exports, "AudioPreview", {
  enumerable: true,
  get: function () {
    return _audioPreview.default;
  }
});
Object.defineProperty(exports, "VideoPlayer", {
  enumerable: true,
  get: function () {
    return _videoPlayer.default;
  }
});
Object.defineProperty(exports, "EmbedFileCardWrapper", {
  enumerable: true,
  get: function () {
    return _embedFileCardWrapper.default;
  }
});
Object.defineProperty(exports, "EmbedCard", {
  enumerable: true,
  get: function () {
    return _embedCard.default;
  }
});
Object.defineProperty(exports, "MobileEmbedCard", {
  enumerable: true,
  get: function () {
    return _mobileEmbedCard.default;
  }
});
Object.defineProperty(exports, "EmbedDownloadButton", {
  enumerable: true,
  get: function () {
    return _embedDownloadButton.default;
  }
});
Object.defineProperty(exports, "EmbedPreviewHeader", {
  enumerable: true,
  get: function () {
    return _embedPreviewHeader.default;
  }
});
Object.defineProperty(exports, "EmbedFailedCard", {
  enumerable: true,
  get: function () {
    return _embedFailedCard.default;
  }
});
Object.defineProperty(exports, "EmbedUploadProgressCard", {
  enumerable: true,
  get: function () {
    return _embedUploadProgressCard.default;
  }
});

var _audioPlayer = _interopRequireDefault(require("./components/audio/audioPlayer"));

var _audioPreview = _interopRequireDefault(require("./components/audio/audioPreview"));

var _videoPlayer = _interopRequireDefault(require("./components/video/videoPlayer"));

var _embedFileCardWrapper = _interopRequireDefault(require("./components/common/embedFileCardWrapper"));

var _embedCard = _interopRequireDefault(require("./components/common/embedCard"));

var _mobileEmbedCard = _interopRequireDefault(require("./components/common/embedCard/mobileEmbedCard"));

var _embedDownloadButton = _interopRequireDefault(require("./components/common/embedDownloadButton"));

var _embedPreviewHeader = _interopRequireDefault(require("./components/common/embedPreviewHeader"));

var _embedFailedCard = _interopRequireDefault(require("./components/common/embedCard/embedFailedCard"));

var _embedUploadProgressCard = _interopRequireDefault(require("./components/common/embedCard/embedUploadProgressCard"));

var _fileTypeHelper = require("./fileTypeHelper");

Object.keys(_fileTypeHelper).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _fileTypeHelper[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _fileTypeHelper[key];
    }
  });
});

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});

var _role = require("./role");

Object.keys(_role).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _role[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _role[key];
    }
  });
});
