// @ali/4ever-utils/dist-node/heading/group.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateHeadingGroupType = generateHeadingGroupType;
exports.generateHeadingGroupKey = generateHeadingGroupKey;
exports.groupHeading = groupHeading;

var _everCangjie = require("@ali/4ever-cangjie");

/**
 * 生成 group type
 * @param type
 * @returns
 */
function generateHeadingGroupType(type) {
  return `${type}-group`;
}
/**
 * 生成 group key
 * @param key
 * @returns
 */


function generateHeadingGroupKey(key) {
  return `${key}-group`;
}

function defaultPass() {
  return false;
}
/**
 * 聚合一组节点
 * @param nodes 节点组
 * @param type group type
 * @param until 停止 group 条件
 * @param pass 跳过节点
 * @returns
 */


function groupHeading([start, end], parent, type, until, pass = defaultPass) {
  const first = parent.nodes[start];
  let i = start + 1;

  while (i < end) {
    const node = parent.nodes[i];

    if (until(node, first) && !pass(node)) {
      break;
    }

    i += 1;
  }

  const children = parent.nodes.slice(start, i);

  const result = _everCangjie.Block.create({
    type: generateHeadingGroupType(type),
    data: first.data,
    key: generateHeadingGroupKey(first.key),
    nodes: children
  });

  return {
    node: result,
    remainedRange: [i, end]
  };
}
