// @ali/4ever-utils/dist-node/heading/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "headTypeConvert", {
  enumerable: true,
  get: function () {
    return _heading.headTypeConvert;
  }
});
Object.defineProperty(exports, "headDefaultStyle", {
  enumerable: true,
  get: function () {
    return _heading.headDefaultStyle;
  }
});
Object.defineProperty(exports, "isHeadingElement", {
  enumerable: true,
  get: function () {
    return _heading.isHeadingElement;
  }
});
Object.defineProperty(exports, "generateHeadingGroupType", {
  enumerable: true,
  get: function () {
    return _group.generateHeadingGroupType;
  }
});
Object.defineProperty(exports, "generateHeadingGroupKey", {
  enumerable: true,
  get: function () {
    return _group.generateHeadingGroupKey;
  }
});
Object.defineProperty(exports, "groupHeading", {
  enumerable: true,
  get: function () {
    return _group.groupHeading;
  }
});

var _heading = require("./heading");

var _group = require("./group");
