// @ali/4ever-bi/dist-node/plugins/uuid/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onChange = _interopRequireDefault(require("./handlers/onChange"));

var actions = _interopRequireWildcard(require("./actions"));

var queries = _interopRequireWildcard(require("./queries"));

function createUUIDPlugin(config) {
  const {
    enable = false
  } = config.uuid || {}; // 若 UUID 未开启，则 UUID Plugin：
  // 1. 不再负责生成 UUID 的职责（交由仓颉底层实现）
  // 2. 通过 anchor query，兼容已生成的锚点

  if (!enable) {
    return {
      queries: {
        anchor: queries.anchor
      }
    };
  }

  return {
    onAction: _onAction.default,
    onChange: (0, _onChange.default)(config.uuid || {}),
    models: [],
    queries
  };
}

createUUIDPlugin.actions = actions;
var _default = createUUIDPlugin;
exports.default = _default;
