// @ali/we-design-next/dist-web/collaboration/collaborator-avatars/AvatarWithPopover.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
var _SplitLine;
var _excluded = ["renderMenu", "color", "description", "name", "avatar", "isPopoverVisible", "onPopoverVisibleChange"];
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { Popover } from "../../display/popover";
import { Dot } from "./styled";
import { Avatar as BaseAvatar } from "../../display/avatar";
import { Avatar } from "./Avatar";
import { AVATAR_SIZE } from "./constants";
import { COLORS } from "../../common";
var PopoverWrapper = /*#__PURE__*/styled.div(["min-width:100px;"]);
var ProfileInfoWrapper = /*#__PURE__*/styled.div(["display:flex;justify-content:space-between;padding:8px 12px;"]);
var ProfileInfoInnerLeft = /*#__PURE__*/styled.div(["display:flex;align-items:center;justify-content:flex-start;margin-right:8px;"]);
var ProfileInfoInnerCenter = /*#__PURE__*/styled.div(["display:flex;min-width:105px;align-items:center;justify-content:space-between;"]);
var ProfileInfoInnerRight = /*#__PURE__*/styled.div(["display:flex;align-items:center;justify-content:flex-end;"]);
var NickText = /*#__PURE__*/styled.span(["flex:1;width:50%;max-width:96px;font-weight:400;font-size:13px;color:", ";text-overflow:ellipsis;white-space:nowrap;overflow:hidden;margin-right:8px;"], COLORS.LIGHT.LEVEL1_BASE_COLOR);
var StatusText = /*#__PURE__*/styled.span(["font-weight:400;font-size:12px;margin-right:4px;color:", ";"], COLORS.LIGHT.LEVEL3_BASE_COLOR);
var SplitLine = /*#__PURE__*/styled.div(["height:1px;margin:0 4px;background:rgba(126,134,142,0.16);"]);
var ProfileInfo = function ProfileInfo(props) {
  var src = props.avatar,
    name = props.name,
    color = props.color,
    description = props.description;
  return /*#__PURE__*/_createElement(ProfileInfoWrapper, null, /*#__PURE__*/_createElement(ProfileInfoInnerLeft, null, /*#__PURE__*/_createElement(BaseAvatar, {
    name: name,
    src: src,
    size: AVATAR_SIZE
  })), /*#__PURE__*/_createElement(ProfileInfoInnerCenter, null, /*#__PURE__*/_createElement(NickText, null, name), description && /*#__PURE__*/_createElement(StatusText, null, description)), /*#__PURE__*/_createElement(ProfileInfoInnerRight, null, /*#__PURE__*/_createElement(Dot, {
    style: {
      background: color
    }
  })));
};
export var AvatarWithPopover = function AvatarWithPopover(props) {
  var renderMenu = props.renderMenu,
    color = props.color,
    description = props.description,
    name = props.name,
    src = props.avatar,
    isPopoverVisible = props.isPopoverVisible,
    onPopoverVisibleChange = props.onPopoverVisibleChange,
    rest = _objectWithoutPropertiesLoose(props, _excluded);
  var menu = isPopoverVisible && renderMenu ? renderMenu() : null;
  return /*#__PURE__*/_createElement(Popover, {
    visible: isPopoverVisible,
    onVisibleChange: onPopoverVisibleChange,
    animation: true,
    content: /*#__PURE__*/_createElement(PopoverWrapper, null, /*#__PURE__*/_createElement(ProfileInfo, {
      color: color,
      description: description,
      name: name,
      avatar: src
    }), menu ? /*#__PURE__*/_createElement(React.Fragment, null, _SplitLine || (_SplitLine = /*#__PURE__*/_createElement(SplitLine, null)), menu) : null),
    trigger: "mouse-enter",
    placement: "bottomLeft",
    placementOffset: [3, 0]
  }, /*#__PURE__*/_createElement(Avatar, _extends({
    name: name,
    src: src,
    color: color
  }, rest)));
};
