// @ali/we-design-next/dist-web/collaboration/collaborator-avatars/Avatar.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
var _excluded = ["color", "highlight"];
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { AvatarWrap, AvatarHighlightWrapper } from "./styled";
import { Avatar as BaseAvatar } from "../../display/avatar/Avatar";
import { AVATAR_SIZE } from "./constants";
export var Avatar = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var color = props.color,
    isHighlight = props.highlight,
    rest = _objectWithoutPropertiesLoose(props, _excluded);
  var wrapStyle = {
    '--hover-color': color,
    border: !color ? 'none' : undefined
  };
  var avata = /*#__PURE__*/_createElement(AvatarWrap, {
    style: wrapStyle,
    ref: ref,
    className: "wdn-avatar-group-item-wrapper"
  }, /*#__PURE__*/_createElement(BaseAvatar, _extends({
    shape: "circle",
    size: AVATAR_SIZE,
    style: {
      position: 'absolute',
      top: -1,
      right: -1
    }
  }, rest)));
  return isHighlight ? /*#__PURE__*/_createElement(AvatarHighlightWrapper, null, avata) : avata;
});
