// @ali/4ever-bi/dist-node/plugins/indent/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _increaseIndent = _interopRequireDefault(require("./commands/increaseIndent"));

var _decreaseIndent = _interopRequireDefault(require("./commands/decreaseIndent"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _actions = require("./actions");

var _createRenderToolbar = _interopRequireDefault(require("./createRenderToolbar"));

const commands = {
  increaseIndent: _increaseIndent.default,
  decreaseIndent: _decreaseIndent.default
};

function createIndentPlugin(config) {
  const indentConfig = config.indent || {};
  return {
    toolbar: (0, _toolbar.default)(indentConfig),
    renderToolbar: (0, _createRenderToolbar.default)(indentConfig.locale),
    menubar: (0, _menubar.default)(indentConfig),
    commands,
    actions: _actions.actions,
    onAction: _onAction.default,
    onKeyDown: _onKeyDown.default,
    models: []
  };
}

createIndentPlugin.actions = _actions.DEPRECATED_ACTIONS;
createIndentPlugin.commands = commands;
var _default = createIndentPlugin;
exports.default = _default;
