// @ali/we-design/dist-web/components/number/styled.js
import styled from 'styled-components';
import { withTheme } from "../theme";
import { Input } from "../input";
export var InputNumberInner = withTheme(styled.span(["position:relative;line-height:20px;display:inline-block;min-width:48px;border:1px solid ", ";border-radius:", ";"], function (p) {
  return p.theme.ui.commonLineLightColor;
}, function (p) {
  return p.theme.ui.radius;
}));
export var InputStyled = withTheme(styled(Input)(["box-sizing:border-box;border:0;padding:4px 16px 4px 4px;height:24px;&::-webkit-inner-spin-button{-webkit-appearance:none;margin:0;}"]));
export var Adjuster = /*#__PURE__*/styled.span(["position:absolute;right:0;top:0;bottom:0;width:12px;line-height:12px;z-index:0;"]);
export var IconWrapper = withTheme(styled.div(["box-sizing:border-box;position:relative;display:flex;flex-direction:column;justify-content:center;align-items:center;height:12px;cursor:pointer;&:hover{background-color:#eee;}&:active{background-color:#ddd;}.weicon{color:", ";&.disabled{color:", ";cursor:not-allowed;}}"], function (p) {
  return p.theme.ui.commonBaseColor3;
}, function (p) {
  return p.theme.ui.commonBaseColor4;
}));
export var UnitWrapper = /*#__PURE__*/styled.span(["margin-left:4px;user-select:none;"]);
export var NumberInputWrap = /*#__PURE__*/styled.span([""]);
