// @ali/4ever-plugin-highlight/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "biFactory", {
  enumerable: true,
  get: function () {
    return _index.biFactory;
  }
});
Object.defineProperty(exports, "mobileBiFactory", {
  enumerable: true,
  get: function () {
    return _mobile.mobileBiFactory;
  }
});
Object.defineProperty(exports, "HighlightRole", {
  enumerable: true,
  get: function () {
    return _role.HighlightRole;
  }
});
Object.defineProperty(exports, "HIGHLIGHT_SET", {
  enumerable: true,
  get: function () {
    return depreactedActions.HIGHLIGHT_SET;
  }
});
exports.depreactedQueries = exports.depreactedCommands = exports.depreactedActions = exports.OldBambooPlugin = void 0;

var _index = require("./bi/index");

var _mobile = require("./bi/mobile");

require("./utils/types");

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;

var _role = require("./utils/role");

var depreactedActions = _interopRequireWildcard(require("./bi/actions"));

exports.depreactedActions = depreactedActions;

var depreactedCommands = _interopRequireWildcard(require("./bi/commands"));

exports.depreactedCommands = depreactedCommands;

var depreactedQueries = _interopRequireWildcard(require("./bi/queries"));

exports.depreactedQueries = depreactedQueries;
