// @ali/4ever-bi/dist-node/plugins/italic/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _everComponent = require("@ali/4ever-component");

var _everPluginMarks = require("@ali/4ever-plugin-marks");

var _everPluginItalic = require("@ali/4ever-plugin-italic");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement("img", {
  width: "24",
  height: "24",
  src: "https://gw.alicdn.com/imgextra/i4/O1CN015Gu0pN1dNdhYp7ffg_!!6000000003724-55-tps-24-24.svg"
});

var _default = (0, _everPluginMarks.createMarkFactory)({
  moFactory: _everPluginItalic.moFactory,
  type: 'italic',
  hotKey: 'mod+i',
  delimiter: [{
    character: '*',
    size: 1
  }, {
    character: '_',
    size: 1,
    disableIntraword: true
  }],
  Button: _weToolbar.ItalicButton,
  Icon: () => _ref,
  toolbarIcon: /*#__PURE__*/_createElement(_everComponent.IconItalic, null),
  localeKeys: {
    titleKey: 'buttonTitle'
  }
});

exports.default = _default;
