// @ali/4ever-bi/dist-node/plugins/refer/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.actions = exports.default = void 0;

var _menubar = _interopRequireDefault(require("./menubar"));

var _insertRefer = _interopRequireDefault(require("./commands/insertRefer"));

var _createRenderNode = _interopRequireDefault(require("./createRenderNode"));

var actions = _interopRequireWildcard(require("./actions"));

exports.actions = actions;

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _everPluginRefer = require("@ali/4ever-plugin-refer");

const commands = {
  insertRefer: _insertRefer.default
};

function createReferPlugin(config) {
  const referConfig = config.refer || {};
  return { ...(0, _everPluginRefer.moFactory)({}),
    menubar: (0, _menubar.default)(),
    commands,
    renderNode: (0, _createRenderNode.default)(referConfig),
    onAction: _onAction.default,
    models: {
      klass: 'inline',
      type: 'refer'
    }
  };
}

createReferPlugin.actions = actions;
createReferPlugin.commands = commands;
var _default = createReferPlugin;
exports.default = _default;
