// @ali/4ever-bi/dist-node/plugins/translate/components/next/newTranslateButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everComponent = require("@ali/4ever-component");

const _createElement = /*#__PURE__*/_react.default.createElement;
var TranslationItemType;

(function (TranslationItemType) {
  TranslationItemType["wordTranslation"] = "wordTranslation";
  TranslationItemType["docTranslation"] = "docTranslation";
})(TranslationItemType || (TranslationItemType = {}));

const TranslationButton = props => {
  const {
    config
  } = props;
  const {
    locale = {}
  } = config;

  const handleChange = _react.default.useCallback(value => {
    if (config.onClickMenuItem) {
      config.onClickMenuItem(value);
    }
  }, [config]);

  const menuItems = _react.default.useMemo(() => {
    return [{
      key: TranslationItemType.wordTranslation,
      title: locale.wordTranslation || ''
    }, {
      key: TranslationItemType.docTranslation,
      title: locale.docTranslation || ''
    }];
  }, [locale.docTranslation, locale.wordTranslation]);

  return /*#__PURE__*/_createElement(_everComponent.TranslationSelect, {
    testid: "translate_button",
    onChange: handleChange,
    tooltip: locale?.buttonTitle,
    role: _everBamboo.PluginRoles.lineHeightSelect,
    title: locale?.title,
    menuItems: menuItems
  });
};

var _default = TranslationButton;
exports.default = _default;
