// @ali/4ever-bi/dist-node/plugins/translate/components/next/translateButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createStickerButton;

var _react = _interopRequireDefault(require("react"));

var _everComponent = require("@ali/4ever-component");

var _everBamboo = require("@ali/4ever-bamboo");

const _createElement = /*#__PURE__*/_react.default.createElement;
;

var _ref = /*#__PURE__*/_createElement(_everComponent.IconTranslate, null);

function createStickerButton(props) {
  const {
    controller,
    locale,
    icon,
    title,
    tooltip,
    badgeDot,
    disabled,
    config
  } = props;
  const {
    triggerTranslateSideMenu = () => {}
  } = config;

  const handleClick = () => {
    triggerTranslateSideMenu();
  };

  const isDisabled = disabled || controller.query('isSelectionInListSymbol') || false;
  const buttonConfig = {
    icon: icon || _ref,
    onClick: handleClick,
    title: title || locale?.title,
    tooltip: {
      title: title || locale?.title || '',
      ...tooltip
    },
    badgeDot,
    testid: "translate_button",
    role: _everBamboo.PluginRoles.translate,
    disabled: isDisabled
  };
  return buttonConfig;
}
