// @ali/4ever-bi/dist-node/plugins/bold/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _everPluginMarks = require("@ali/4ever-plugin-marks");

var _everComponent = require("@ali/4ever-component");

var _everPluginBold = require("@ali/4ever-plugin-bold");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement("img", {
  width: "24",
  height: "24",
  src: "https://gw.alicdn.com/imgextra/i1/O1CN01XqJvv81eCTf14uU8e_!!6000000003835-55-tps-24-24.svg"
});

var _default = (0, _everPluginMarks.createMarkFactory)({
  moFactory: _everPluginBold.moFactory,
  type: 'bold',
  hotKey: 'mod+b',
  delimiter: [{
    character: '*',
    size: 2
  }, {
    character: '_',
    size: 2,
    disableIntraword: true
  }],
  Button: _weToolbar.BoldButton,
  Icon: () => _ref,
  toolbarIcon: /*#__PURE__*/_createElement(_everComponent.IconBold, null),
  localeKeys: {
    titleKey: 'buttonTitle'
  }
});

exports.default = _default;
