// @ali/4ever-zhi/dist-node/plugins/link/linkCard/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everFactory = require("@ali/4ever-factory");

var _zh_CN = _interopRequireDefault(require("../../../locales/zh_CN"));

var _createRenderNode = _interopRequireDefault(require("./createRenderNode"));

var _everBi = require("@ali/4ever-bi");

var _everBamboo = require("@ali/4ever-bamboo");

const {
  copyLink
} = _everBi.LinkCardPlugin.utils;
const {
  LINK_CARD_COPY_LINK
} = _everBi.LinkCardPlugin.actions;
const createZhiCalendarCardPlugin = (0, _everFactory.zhiCardFactory)(config => {
  const linkConfig = {
    locale: { ..._zh_CN.default.link.locale
    },
    ...config.link
  };
  const serializer = config.clipboard?.serializer;
  return {
    cardType: 'link',
    renderNode: (0, _createRenderNode.default)(linkConfig, false),
    queries: {
      getMinTableColWidth: _everBamboo.LinkPlugin.getMinTableColWidthForBlockLinkCard
    },
    onAction: (action, controller, next) => {
      const {
        type,
        payload
      } = action;

      if (type === LINK_CARD_COPY_LINK) {
        const {
          node
        } = payload;
        return copyLink(controller, node, serializer || controller.serializer);
      }

      return next();
    }
  };
});
var _default = createZhiCalendarCardPlugin;
exports.default = _default;
