// @ali/we-design-next/dist-node/common/domAlign/utils/property-utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getTransformName = getTransformName;
exports.getTransformXY = getTransformXY;
exports.getTransitionProperty = getTransitionProperty;
exports.setTransformXY = setTransformXY;
exports.setTransitionProperty = setTransitionProperty;
let vendorPrefix;
const jsCssMap = {
  Webkit: '-webkit-',
  Moz: '-moz-',
  // IE did it wrong again ...
  ms: '-ms-',
  O: '-o-'
};
function getVendorPrefix() {
  if (vendorPrefix !== undefined) {
    return vendorPrefix;
  }
  vendorPrefix = '';
  const style = document.createElement('p').style;
  const testProp = 'Transform';
  for (const key in jsCssMap) {
    if (key + testProp in style) {
      vendorPrefix = key;
    }
  }
  return vendorPrefix;
}
function getTransitionName() {
  return getVendorPrefix() ? `${getVendorPrefix()}TransitionProperty` : 'transitionProperty';
}
function getTransformName() {
  return getVendorPrefix() ? `${getVendorPrefix()}Transform` : 'transform';
}
function setTransitionProperty(node, value) {
  const name = getTransitionName();
  if (name) {
    node.style[name] = value;
    if (name !== 'transitionProperty') {
      node.style.transitionProperty = value;
    }
  }
}
function setTransform(node, value) {
  const name = getTransformName();
  if (name) {
    node.style[name] = value;
    if (name !== 'transform') {
      node.style.transform = value;
    }
  }
}
function getTransitionProperty(node) {
  return node.style.transitionProperty || node.style[getTransitionName()];
}
function getTransformXY(node) {
  const style = window.getComputedStyle(node, null);
  const transform = style.getPropertyValue('transform') || style.getPropertyValue(getTransformName());
  if (transform && transform !== 'none') {
    const matrix = transform.replace(/[^0-9\-.,]/g, '').split(',');
    return {
      x: parseFloat(matrix[12] || matrix[4]),
      y: parseFloat(matrix[13] || matrix[5])
    };
  }
  return {
    x: 0,
    y: 0
  };
}
const matrix2d = /matrix\((.*)\)/;
const matrix3d = /matrix3d\((.*)\)/;
function setTransformXY(node, xy) {
  const style = window.getComputedStyle(node, null);
  const transform = style.getPropertyValue('transform') || style.getPropertyValue(getTransformName());
  if (transform && transform !== 'none') {
    let arr;
    const match2d = transform.match(matrix2d);
    if (match2d) {
      const match2dStr = match2d[1];
      arr = match2dStr.split(',').map(item => {
        return parseFloat(item);
      });
      arr[4] = xy.x;
      arr[5] = xy.y;
      setTransform(node, `matrix(${arr.join(',')})`);
    } else {
      const match3d = transform.match(matrix3d);
      if (!match3d) return;
      const match3dStr = match3d[1];
      arr = match3dStr.split(',').map(item => {
        return parseFloat(item);
      });
      arr[12] = xy.x;
      arr[13] = xy.y;
      setTransform(node, `matrix3d(${arr.join(',')})`);
    }
  } else {
    setTransform(node, `translateX(${xy.x}px) translateY(${xy.y}px) translateZ(0)`);
  }
}
