// @ali/we-design/dist-web/components/alert/styled.js
/* eslint-disable no-multiple-empty-lines */

/* eslint-disable prefer-template */

/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/explicit-function-return-type */
import styled from 'styled-components';
import { Ellipsis, resetComponent } from "../../utils/styled.mixin";
import { withTheme } from "../theme";
import { Button } from "../button";
export var AlertRoot = withTheme(styled.div(["", " width:100%;overflow:hidden;&.hidden{display:none;}.type-icon{font-size:", ";margin-right:", ";}.close-icon{margin-top:2px;color:", ";cursor:pointer;}&.close-mode{.close-icon{margin-top:2px;margin-left:", ";}}&.action-mode{.close-icon{display:none;}}&.info{border-color:", ";background-color:", ";.type-icon{color:", ";}}&.error{border-color:", ";background-color:", ";.type-icon{color:", ";}}&.warning{border-color:", ";background-color:", ";.type-icon{color:", ";}}"], resetComponent, function (_ref) {
  var theme = _ref.theme;
  return theme.ui.fontSizeLarge;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.marginMedium;
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.colorText3;
}, function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.margin;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.ui.primaryColor;
}, function (_ref6) {
  var theme = _ref6.theme;
  return theme.ui.primaryColor2;
}, function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.primaryColor;
}, function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.errorColor;
}, function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.errorColor2;
}, function (_ref10) {
  var theme = _ref10.theme;
  return theme.ui.errorColor;
}, function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.warningColor;
}, function (_ref12) {
  var theme = _ref12.theme;
  return theme.ui.warningColor2;
}, function (_ref13) {
  var theme = _ref13.theme;
  return theme.ui.warningColor;
}));
export var AlertWrap = withTheme(styled.div(["display:flex;align-items:flex-start;padding:", " ", ";line-height:1.6;font-size:", ";color:", ";"], function (_ref14) {
  var theme = _ref14.theme;
  return theme.ui.alertVerticalPadding;
}, function (_ref15) {
  var theme = _ref15.theme;
  return theme.ui.alertHorizontalPadding;
}, function (_ref16) {
  var theme = _ref16.theme;
  return theme.ui.fontSizeSmall;
}, function (_ref17) {
  var theme = _ref17.theme;
  return theme.ui.colorText5;
}));
export var MessageWrap = withTheme(styled.div(["flex:1 auto;width:0;display:flex;align-items:center;flex-direction:column;"]));
export var MessageContent = withTheme(styled.div(["display:flex;align-items:center;height:", ";width:100%;font-size:", ";"], function (_ref18) {
  var theme = _ref18.theme;
  return theme.ui.elementHeightSmall;
}, function (_ref19) {
  var size = _ref19.size,
      theme = _ref19.theme;
  return size === 'large' ? theme.ui.fontSize : theme.ui.fontSizeSmall;
}));
export var MessageText = withTheme(styled.div(["", " flex-shrink:1;"], Ellipsis));
export var DescriptionContent = withTheme(styled.div(["display:flex;align-items:center;width:100%;height:", ";font-size:", ";"], function (_ref20) {
  var theme = _ref20.theme;
  return theme.ui.elementHeightSmall;
}, function (_ref21) {
  var theme = _ref21.theme;
  return theme.ui.fontSizeSmall;
}));
export var DescriptionText = withTheme(styled.div(["", ""], Ellipsis));
export var TextButton = /*#__PURE__*/styled(Button)(["font-size:12px;cursor:pointer;padding:0 4px 0 4px;display:flex;align-items:center;justify-content:center;"]);
export var IconWrap = withTheme(styled.div(["height:", ";display:flex;align-items:center;"], function (_ref22) {
  var theme = _ref22.theme;
  return theme.ui.elementHeightSmall;
}));
export var ActionWrap = withTheme(styled(TextButton)(["height:", ";color:", ";margin-left:", ";margin-right:", ";cursor:pointer;&:hover{color:", ";}"], function (_ref23) {
  var theme = _ref23.theme;
  return theme.ui.elementHeightSmall;
}, function (_ref24) {
  var theme = _ref24.theme;
  return theme.ui.linkHoverColor;
}, function (_ref25) {
  var theme = _ref25.theme;
  return theme.ui.margin;
}, function (_ref26) {
  var theme = _ref26.theme;
  return theme.ui.margin;
}, function (_ref27) {
  var theme = _ref27.theme;
  return theme.ui.linkPressColor;
}));
export var CloseButton = withTheme(styled(TextButton)(["color:", ";width:24px;height:", ";margin-left:0;color:#0089ff;&:hover,&:focus{color:", ";}&.active{color:", ";}"], function (_ref28) {
  var theme = _ref28.theme;
  return theme.ui.colorText3;
}, function (_ref29) {
  var theme = _ref29.theme;
  return theme.ui.elementHeightSmall;
}, function (_ref30) {
  var theme = _ref30.theme;
  return theme.ui.colorText1;
}, function (_ref31) {
  var theme = _ref31.theme;
  return theme.ui.colorText1;
}));
