// @ali/we-design-next/dist-web/form/input-number/styled.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
var _excluded = ["disabled"],
  _excluded2 = ["disabled"];
import React, { memo } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { CSS_CURSOR_DISABLE } from "../../common";
import { COLORS } from "../../common/consts/colors";
var IconSvg = /*#__PURE__*/styled.svg(["cursor:auto;stroke-opacity:0.6;&:active{stroke-opacity:0.24;}flex-grow:1;", " ", " &:hover{background-color:rgba(126,134,142,0.12);}&:active{background-color:", ";}"], function (_ref) {
  var disabled = _ref.disabled;
  return disabled ? "color: " + COLORS.LIGHT.LEVEL4_BASE_COLOR + ";" : null;
}, CSS_CURSOR_DISABLE, COLORS.COMMON.OVERLAY_COLOR);
export var ArrowUpIcon = /*#__PURE__*/memo(function (props) {
  var disabled = props.disabled,
    rest = _objectWithoutPropertiesLoose(props, _excluded);
  return /*#__PURE__*/_createElement(IconSvg, _extends({
    width: "12",
    height: "12",
    viewBox: "0 0 12 12",
    fill: "none",
    disabled: disabled
  }, rest), /*#__PURE__*/_createElement("path", {
    d: "M9 7L6.04273 4L3 7",
    stroke: COLORS.LIGHT.LEVEL1_BASE_COLOR
  }));
});
ArrowUpIcon.displayName = 'InputNumberArrowUpIcon';
export var ArrowDownIcon = /*#__PURE__*/memo(function (props) {
  var disabled = props.disabled,
    rest = _objectWithoutPropertiesLoose(props, _excluded2);
  return /*#__PURE__*/_createElement(IconSvg, _extends({
    width: "12",
    height: "12",
    viewBox: "0 0 12 12",
    fill: "none",
    disabled: disabled
  }, rest), /*#__PURE__*/_createElement("path", {
    d: "M3 5L5.95727 8L9 5",
    stroke: COLORS.LIGHT.LEVEL1_BASE_COLOR
  }));
});
ArrowDownIcon.displayName = 'InputNumberArrowDownIcon';
export var NumberControlWrap = /*#__PURE__*/styled.div(["display:flex;flex-direction:column;background-color:#f2f2f6;border-left:1px solid rgb(205,205,205);height:100%;"]);
