// @ali/4ever-plugin-table/dist-web/utils/constants.js
export var SPACING = 16;
export var MOBILE_SPACING = 24;
export var SCROLL_BAR_HEIGHT = 8;
export var TOOLBAR_ITEM_SIZE = 8;
export var TOOLBAR_ITEM_SIZE_MOBILE = 16;
export var MIN_COL_WIDTH = 33;
export var DEFAULT_PCT_WIDTH = 100;
export var MIN_ROW_HEIGHT = 33; // 边距适中，A4 纸的宽度

export var DEFAULT_TOTAL_WIDTH_FOR_AUTOFIT = 650; // 降低以处理下边距过大的问题

export var QUICK_INSERT_BUTTON_WIDTH_HEIGHT = 18;
/**
 * realTable padding
 */

export var REALTABLE_PADDING = {
  top: 10,
  bottom: 0
}; // 表头行吸顶时，首行距离 scrollContainer 顶部空间

export var STICKY_ROW_TOP_HEIGHT = 30; // 表头行吸顶时，行工具栏左侧遮挡区域宽度

export var STICKY_ROW_LEFT_WIDTH = 10; // 表头行吸顶时，工具栏 z-index

export var STICKY_TOOLBAR_INDEX_MAP = {
  cornerToolbar: 22,
  colToolbar: 21,
  cornerButton: 20,
  rowToolbar: 19,
  stickyRow: 25,
  selection: 20
};
export var TABLE_HEADER_BG_COLOR = '#f5f6f7';
export var DEFAULT_BORDER_COLOR = '#D6D6D6';
export default {
  SPACING: SPACING,
  SCROLL_BAR_HEIGHT: SCROLL_BAR_HEIGHT,
  TOOLBAR_ITEM_SIZE: TOOLBAR_ITEM_SIZE,
  MIN_COL_WIDTH: MIN_COL_WIDTH,
  DEFAULT_PCT_WIDTH: DEFAULT_PCT_WIDTH,
  DEFAULT_TOTAL_WIDTH_FOR_AUTOFIT: DEFAULT_TOTAL_WIDTH_FOR_AUTOFIT,
  REALTABLE_PADDING: REALTABLE_PADDING,
  MIN_ROW_HEIGHT: MIN_ROW_HEIGHT,
  QUICK_INSERT_BUTTON_WIDTH_HEIGHT: QUICK_INSERT_BUTTON_WIDTH_HEIGHT,
  STICKY_ROW_TOP_HEIGHT: STICKY_ROW_TOP_HEIGHT,
  STICKY_ROW_LEFT_WIDTH: STICKY_ROW_LEFT_WIDTH,
  STICKY_TOOLBAR_INDEX_MAP: STICKY_TOOLBAR_INDEX_MAP,
  TABLE_HEADER_BG_COLOR: TABLE_HEADER_BG_COLOR,
  DEFAULT_BORDER_COLOR: DEFAULT_BORDER_COLOR,
  TOOLBAR_ITEM_SIZE_MOBILE: TOOLBAR_ITEM_SIZE_MOBILE
};
export var lineMap = {
  Thin: {
    lineType: 'single',
    size: 1
  },
  Medium: {
    lineType: 'single',
    size: 2
  },
  Thick: {
    lineType: 'single',
    size: 3
  },
  Dashed: {
    lineType: 'dashed',
    size: 1
  },
  MediumDashed: {
    lineType: 'dashed',
    size: 2
  }
};
export var DEFAULT_TABLE_BORDER_STYLE = 'Thin';
