// @ali/4ever-plugin-embed/dist-web/utils/fileTypeHelper.js
var _IconUrlMap, _FileIconMap;

import { DingdriveAiBetaNormal, DingdriveDingdocBetaNormal, DingdriveDingexcelBetaNormal, DingdriveExcelBetaNormal, DingdriveKeynoteBetaNormal, DingdriveNumbersBetaNormal, DingdriveOtherBetaNormal, DingdrivePageBetaNormal, DingdrivePdfBetaNormal, DingdrivePictureBetaNormal, DingdrivePptBetaNormal, DingdrivePsBetaNormal, DingdriveSketchBetaNormal, DingdriveTxtBetaNormal, DingdriveVideoBetaNormal, DingdriveVoiceBetaNormal, DingdriveWhitepageBetaNormal, DingdriveWordBetaNormal, DingdriveZipBetaNormal, DingnoteDocBetaNormal, DingnoteExcelBetaNormal } from '@ali/we-design';
export var FileTypeMap = {
  Word: 'word',
  Excel: 'excel',
  Ppt: 'ppt',
  Pdf: 'pdf',
  Zip: 'zip',
  Audio: 'audio',
  Video: 'video',
  Txt: 'txt',
  Sketch: 'sketch',
  Ps: 'ps',
  Picture: 'picture',
  Keynote: 'keynote',
  Page: 'page',
  Numbers: 'numbers',
  Ai: 'ai',
  App: 'app',
  DingSheet: 'dingsheet',
  DingDoc: 'dingdoc',
  WhiteBoard: 'whiteboard',
  DingNoteSheet: 'dingnotesheet',
  DingNoteDoc: 'dingnotedoc'
};
export var CanPreviewFileTypes = [FileTypeMap.Word, FileTypeMap.Excel, FileTypeMap.Ppt, FileTypeMap.Pdf, FileTypeMap.Audio, FileTypeMap.Video];
/* eslint-disable */

export function getFileTypeForRead(fileName) {
  if (!fileName) {
    return FileTypeMap.App;
  }

  var lowerCaseFileName = fileName.toLowerCase();

  if (lowerCaseFileName.endsWith('.doc')) {
    return FileTypeMap.Word;
  } else if (lowerCaseFileName.endsWith('.docx')) {
    return FileTypeMap.Word;
  } else if (lowerCaseFileName.endsWith('.xls')) {
    return FileTypeMap.Excel;
  } else if (lowerCaseFileName.endsWith('.xlsx')) {
    return FileTypeMap.Excel;
  } else if (lowerCaseFileName.endsWith('.ppt')) {
    return FileTypeMap.Ppt;
  } else if (lowerCaseFileName.endsWith('.pptx')) {
    return FileTypeMap.Ppt;
  } else if (lowerCaseFileName.endsWith('.zip')) {
    return FileTypeMap.Zip;
  } else if (lowerCaseFileName.endsWith('.mp4') || lowerCaseFileName.endsWith('.ogg') || lowerCaseFileName.endsWith('.mov')) {
    return FileTypeMap.Video;
  } else if (lowerCaseFileName.endsWith('.txt')) {
    return FileTypeMap.Txt;
  } else if (lowerCaseFileName.endsWith('.sketch')) {
    return FileTypeMap.Sketch;
  } else if (lowerCaseFileName.endsWith('.ps')) {
    return FileTypeMap.Ps;
  } else if (lowerCaseFileName.endsWith('.png') || lowerCaseFileName.endsWith('.jpg') || lowerCaseFileName.endsWith('.bmp')) {
    return FileTypeMap.Picture;
  } else if (lowerCaseFileName.endsWith('.pdf')) {
    return FileTypeMap.Pdf;
  } else if (lowerCaseFileName.endsWith('.key')) {
    return FileTypeMap.Keynote;
  } else if (lowerCaseFileName.endsWith('.pages')) {
    return FileTypeMap.Page;
  } else if (lowerCaseFileName.endsWith('.numbers')) {
    return FileTypeMap.Numbers;
  } else if (lowerCaseFileName.endsWith('.ai')) {
    return FileTypeMap.Ai;
  } else if (lowerCaseFileName.endsWith('.mp3')) {
    return FileTypeMap.Audio;
  } else if (lowerCaseFileName.endsWith('.adoc')) {
    return FileTypeMap.DingDoc;
  } else if (lowerCaseFileName.endsWith('.axls')) {
    return FileTypeMap.DingSheet;
  } else if (lowerCaseFileName.endsWith('.awbd')) {
    return FileTypeMap.WhiteBoard;
  } else {
    return FileTypeMap.App;
  }
} // 真实的换算应该按照 1024，但文件系统里显示是按照1000换算，所以跟系统保持同步

export var formatSize = function formatSize(size) {
  // 小于 1M，显示 KB
  if (size < 1 * 1000 * 1000) {
    return (size / 1000).toFixed(1) + "KB";
  }

  return (size / (1000 * 1000)).toFixed(1) + "MB";
};
export var IconUrlMap = (_IconUrlMap = {}, _IconUrlMap[FileTypeMap.Word] = 'https://img.alicdn.com/tfs/TB16zxhlHY1gK0jSZTEXXXDQVXa-24-24.svg', _IconUrlMap[FileTypeMap.Excel] = 'https://img.alicdn.com/tfs/TB1zi5Ik1bviK0jSZFNXXaApXXa-24-24.svg', _IconUrlMap[FileTypeMap.Zip] = 'https://img.alicdn.com/tfs/TB1HahflQY2gK0jSZFgXXc5OFXa-24-24.svg', _IconUrlMap[FileTypeMap.Video] = 'https://img.alicdn.com/tfs/TB1Ns0dlUH1gK0jSZSyXXXtlpXa-24-24.svg', _IconUrlMap[FileTypeMap.Txt] = 'https://img.alicdn.com/tfs/TB11_NclQL0gK0jSZFtXXXQCXXa-24-24.svg', _IconUrlMap[FileTypeMap.Sketch] = 'https://img.alicdn.com/tfs/TB1hllglNv1gK0jSZFFXXb0sXXa-24-24.svg', _IconUrlMap[FileTypeMap.Ps] = 'https://img.alicdn.com/tfs/TB1q9NblFY7gK0jSZKzXXaikpXa-24-24.svg', _IconUrlMap[FileTypeMap.Ppt] = 'https://img.alicdn.com/tfs/TB1KI0dlUH1gK0jSZSyXXXtlpXa-24-24.svg', _IconUrlMap[FileTypeMap.Picture] = 'https://img.alicdn.com/tfs/TB1HORdlNz1gK0jSZSgXXavwpXa-24-24.svg', _IconUrlMap[FileTypeMap.Pdf] = 'https://img.alicdn.com/tfs/TB1YsthlND1gK0jSZFKXXcJrVXa-24-24.svg', _IconUrlMap[FileTypeMap.Keynote] = 'https://gw.alicdn.com/tfs/TB1VLh0Bbj1gK0jSZFuXXcrHpXa-24-24.svg', _IconUrlMap[FileTypeMap.Page] = 'https://img.alicdn.com/tfs/TB1FyRdlNz1gK0jSZSgXXavwpXa-24-24.svg', _IconUrlMap[FileTypeMap.Numbers] = 'https://img.alicdn.com/tfs/TB16OBflO_1gK0jSZFqXXcpaXXa-24-24.svg', _IconUrlMap[FileTypeMap.Ai] = 'https://img.alicdn.com/tfs/TB1PIthlND1gK0jSZFKXXcJrVXa-24-24.svg', _IconUrlMap[FileTypeMap.Audio] = 'https://img.alicdn.com/tfs/TB11Vy.obj1gK0jSZFOXXc7GpXa-24-24.svg', _IconUrlMap[FileTypeMap.App] = 'https://img.alicdn.com/tfs/TB145BflO_1gK0jSZFqXXcpaXXa-24-24.svg', _IconUrlMap);
export var FileIconMap = (_FileIconMap = {}, _FileIconMap[FileTypeMap.Word] = DingdriveWordBetaNormal, _FileIconMap[FileTypeMap.Excel] = DingdriveExcelBetaNormal, _FileIconMap[FileTypeMap.Ppt] = DingdrivePptBetaNormal, _FileIconMap[FileTypeMap.Pdf] = DingdrivePdfBetaNormal, _FileIconMap[FileTypeMap.Zip] = DingdriveZipBetaNormal, _FileIconMap[FileTypeMap.Audio] = DingdriveVoiceBetaNormal, _FileIconMap[FileTypeMap.Video] = DingdriveVideoBetaNormal, _FileIconMap[FileTypeMap.Txt] = DingdriveTxtBetaNormal, _FileIconMap[FileTypeMap.Sketch] = DingdriveSketchBetaNormal, _FileIconMap[FileTypeMap.Ps] = DingdrivePsBetaNormal, _FileIconMap[FileTypeMap.Picture] = DingdrivePictureBetaNormal, _FileIconMap[FileTypeMap.Keynote] = DingdriveKeynoteBetaNormal, _FileIconMap[FileTypeMap.Page] = DingdrivePageBetaNormal, _FileIconMap[FileTypeMap.Numbers] = DingdriveNumbersBetaNormal, _FileIconMap[FileTypeMap.Ai] = DingdriveAiBetaNormal, _FileIconMap[FileTypeMap.App] = DingdriveOtherBetaNormal, _FileIconMap[FileTypeMap.DingDoc] = DingdriveDingdocBetaNormal, _FileIconMap[FileTypeMap.DingSheet] = DingdriveDingexcelBetaNormal, _FileIconMap[FileTypeMap.WhiteBoard] = DingdriveWhitepageBetaNormal, _FileIconMap[FileTypeMap.DingNoteDoc] = DingnoteDocBetaNormal, _FileIconMap[FileTypeMap.DingNoteSheet] = DingnoteExcelBetaNormal, _FileIconMap);
export var getIcon = function getIcon(type) {
  return IconUrlMap[type] || IconUrlMap['app'];
};
