// @ali/4ever-plugin-code/dist-node/bi/components/ToolbarButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _weIcon = require("@ali/we-icon");

var _actions = require("../actions");

var _role = require("../../utils/role");

const _createElement = /*#__PURE__*/React.createElement;
const iconStyle = {
  width: 20,
  height: 20
};

var _ref = /*#__PURE__*/_createElement(_weIcon.ToolbarCodeNormal, {
  style: iconStyle
});

const ToolbarButton = props => {
  const {
    controller,
    locale,
    ...rest
  } = props;
  const handleClick = React.useCallback(() => controller.run('onAction', (0, _actions.insertCode)()), [controller]);
  return /*#__PURE__*/_createElement(_weToolbar.IconButton, (0, _extends2.default)({}, rest, {
    title: locale.buttonTitle,
    tooltip: locale.buttonTitle,
    testid: "bi-toolbar-code-block",
    icon: _ref,
    onClick: handleClick,
    role: _role.CodeRole.code
  }));
};

var _default = /*#__PURE__*/React.memo(ToolbarButton);

exports.default = _default;
