// @ali/4ever-plugin-refblock/dist-node/components/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IconButtonTitle = exports.IconButtonWrapper = exports.Divider = exports.DropdownOverlay = exports.Title = exports.ListItem = exports.ListItemPlaceholder = exports.CopyFromPlaceholder = exports.CopyFrom = exports.ListItemContent = exports.ListItemIconWrapper = exports.ToolbarColumn = exports.Toolbar = exports.ToolbarWrapper = exports.RefBlockWrapper = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

function getBorderColor(isHost) {
  return isHost ? '#FF642B' : 'rgba(0, 137, 255, 0.48)';
}

const RefBlockWrapper = /*#__PURE__*/_styledComponents.default.div(["position:relative;padding:4px;border-width:1px;border-style:", ";border-color:", ";border-radius:8px;background-color:", ";", " @keyframes animateBorderOne{to{border-color:transparent;}}:hover{border-style:dotted;border-color:", ";}.refblock-content{> div:first-child{margin-top:0px !important;}> div:last-child{margin-bottom:0px !important;}}"], props => props.isFocused || props.$loading ? 'dotted' : 'solid', props => props.isFocused || props.$loading ? getBorderColor(props.isHost) : 'transparent', props => props.$loading ? '#f3f3f378' : 'inherit', props => props.$loading ? 'animation: 1s animateBorderOne ease infinite;' : '', props => getBorderColor(props.isHost));

exports.RefBlockWrapper = RefBlockWrapper;

const ToolbarWrapper = /*#__PURE__*/_styledComponents.default.div(["position:absolute;left:-100px;top:-100px;cursor:pointer;& .icon-button{cursor:pointer;}"]);

exports.ToolbarWrapper = ToolbarWrapper;

const Toolbar = /*#__PURE__*/_styledComponents.default.div(["position:relative;display:flex;justify-content:center;flex-direction:row;padding:5px 5px;border-radius:4px;transition:background-color 250ms;opacity:1;box-shadow:rgb(0 0 0 / 24%) 0px 0px 1px,rgb(0 0 0 / 16%) 0px 8px 24px;background-color:#fff;z-index:999;"]);

exports.Toolbar = Toolbar;

const ToolbarColumn = /*#__PURE__*/_styledComponents.default.div(["display:flex;font-size:13px;align-items:center;border-radius:4px;:hover{background-color:", ";}"], ({
  highlightOnHover
}) => highlightOnHover ? 'rgba(25,31,37,0.08)' : 'inherit');

exports.ToolbarColumn = ToolbarColumn;

const ListItemIconWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;align-items:center;justify-content:center;width:14px;height:16px;img{display:inline-block;max-width:100%;max-height:100%;}"]);

exports.ListItemIconWrapper = ListItemIconWrapper;

const ListItemContent = /*#__PURE__*/_styledComponents.default.div(["min-width:0;flex:1;width:208px;position:relative;display:flex;text-overflow:ellipsis;white-space:nowrap;overflow:hidden;flex-direction:column;margin-left:8px;a{color:#1f2329;display:inline-block;font-size:14px;text-overflow:ellipsis;white-space:pre;overflow:hidden;max-width:100%;text-decoration:none;}"]);

exports.ListItemContent = ListItemContent;

const CopyFrom = /*#__PURE__*/_styledComponents.default.div(["min-width:0;flex:1;min-width:100px;max-width:200px;position:relative;display:flex;text-overflow:ellipsis;white-space:nowrap;overflow:hidden;flex-direction:column;margin-left:8px;a{color:#1f2329;display:inline-block;font-size:14px;text-overflow:ellipsis;white-space:pre;overflow:hidden;max-width:100%;text-decoration:none;}"]);

exports.CopyFrom = CopyFrom;

const CopyFromPlaceholder = /*#__PURE__*/_styledComponents.default.div(["width:100px;height:20px;margin-left:8px;border-radius:4px;background-color:#F2F4F5;"]);

exports.CopyFromPlaceholder = CopyFromPlaceholder;
const ListItemPlaceholder = /*#__PURE__*/(0, _styledComponents.default)(CopyFromPlaceholder)(["width:208px;"]);
exports.ListItemPlaceholder = ListItemPlaceholder;

const ListItem = /*#__PURE__*/_styledComponents.default.div(["display:flex;align-items:center;flex-shrink:0;font-size:13px;font-weight:400;cursor:pointer;justify-content:center;padding:6px 10px;border-radius:4px;&:hover{background-color:rgba(126,134,142,0.12);}"]);

exports.ListItem = ListItem;

const Title = /*#__PURE__*/_styledComponents.default.div(["font-size:12px;color:rgba(17,31,44,0.56);line-height:17px;padding:6px;"]);

exports.Title = Title;

const DropdownOverlay = /*#__PURE__*/_styledComponents.default.div(["display:flex;flex-direction:column;max-height:375px;padding:6px 4px 6px 4px;overflow:overlay;"]);

exports.DropdownOverlay = DropdownOverlay;

const Divider = /*#__PURE__*/_styledComponents.default.div(["width:1px;margin:4px;background-color:rgba(17,31,44,0.08);"]);

exports.Divider = Divider;

const IconButtonWrapper = /*#__PURE__*/_styledComponents.default.span(["padding:10px 10px;"]);

exports.IconButtonWrapper = IconButtonWrapper;

const IconButtonTitle = /*#__PURE__*/_styledComponents.default.span(["margin-left:5px;font-size:13px;"]);

exports.IconButtonTitle = IconButtonTitle;
