// @ali/4ever-bi/dist-web/plugins/heading/components/headingGroup.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { FontSizeMaxSelect } from '@ali/we-toolbar';
import { PluginRoles } from '@ali/4ever-bamboo';
import { toggleBlockquote } from '@ali/4ever-utils';
import { setHeading } from "../actions";
;
var BLOCKQUOTE_KEY = 'blockquote';
var MemoFontSizeMaxSelect = /*#__PURE__*/React.memo(FontSizeMaxSelect);
export default function HeadingGroup(props) {
  var controller = props.controller;
  var type = controller.query('hasBlockQuote') ? BLOCKQUOTE_KEY : controller.query('getHeading');
  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var isDisabled = isCollapsedAtEdge || controller.query('isSelectionInHr') || controller.query('isSelectionInListSymbol');
  var handleChange = React.useCallback(function (headingStr) {
    if (headingStr === BLOCKQUOTE_KEY) {
      if (type !== 'paragraph') {
        // 变成引用同时去除标题
        controller.run('onAction', setHeading('paragraph'));
      }

      controller.run('onAction', toggleBlockquote());
    } else {
      if (type === BLOCKQUOTE_KEY) {
        // 变成标题同时去除引用
        controller.run('onAction', toggleBlockquote());
      }

      controller.run('onAction', setHeading(headingStr));
    }
  }, [controller, type]);
  return /*#__PURE__*/_createElement(MemoFontSizeMaxSelect, {
    testid: "bi-toolbar-heading-group",
    disabled: !!isDisabled,
    value: type,
    onChange: handleChange // 这个值是标题仅显示两项且工具栏未显示更多时的长度，若工具栏的内容变化，需要更新
    ,
    toolbarLength: 1290,
    role: PluginRoles.headingGroup
  });
}
