// @ali/4ever-bi/dist-web/plugins/heading/components/headingSimpleButton.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { HeadingSimpleSelect as OriginHeadingSimpleSelect } from '@ali/we-toolbar';
import { PluginRoles } from '@ali/4ever-bamboo';
import { setHeading } from "../actions";
import { enableBlockquote, disableBlockquote } from '@ali/4ever-utils';
;
var SUPPORT_HEADING = {
  paragraph: true,
  'heading-1': true,
  'heading-2': true,
  'heading-3': true,
  'heading-4': true,
  blockquote: true
};
export var isDisabled = function isDisabled(controller) {
  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInListSymbol') || controller.query('isSelectionInLinkCard') // fix: https://aone.alibaba-inc.com/v2/bug/35749970
  || controller.query('isSelectionInSubjectHeading');
  return Boolean(isDisabled);
};
export default function HeadingSimpleButton(props) {
  var controller = props.controller,
      locale = props.locale,
      size = props.size;
  var heading = controller.query('getBlockquoteState') ? 'blockquote' : controller.query('getHeading');
  var handleChange = React.useCallback(function (value) {
    if (value === 'blockquote') {
      controller.run('onAction', setHeading('paragraph'));
      controller.run('onAction', enableBlockquote());
    } else {
      controller.run('onAction', setHeading(value));
      controller.run('onAction', disableBlockquote());
    }
  }, [controller]);
  var HeadingSimpleSelect = React.useMemo(function () {
    return OriginHeadingSimpleSelect;
  }, []);
  return /*#__PURE__*/_createElement(HeadingSimpleSelect, {
    size: size,
    onChange: handleChange,
    value: SUPPORT_HEADING[heading] ? heading : 'paragraph',
    testid: "bi-toolbar-simple-heading",
    disabled: isDisabled(controller),
    tooltip: locale.buttonTitle,
    role: PluginRoles.headingSelect
  });
}
