// @ali/4ever-bi/dist-web/plugins/heading/components/mobileHeadingButton.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Button } from '@ali/we-design';
import { PluginRoles } from '@ali/4ever-bamboo';
import { setHeading } from "../actions";
export default function mobileHeadingButton(heading) {
  return function HeadingButton(props) {
    var controller = props.controller;
    var value = controller.value;
    var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
    var isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInLinkCard');

    var handleClick = function handleClick() {
      return controller.run('onAction', setHeading(heading));
    };

    var level = '';
    var levelMatch = heading == null ? void 0 : heading.match(/\d+/);

    if (levelMatch && levelMatch[0]) {
      level = levelMatch[0];
    }

    return /*#__PURE__*/_createElement(Button, {
      testid: "bi-toolbar-heading",
      icon: "H" + level,
      bordered: false,
      disabled: isDisabled,
      onClick: handleClick,
      role: PluginRoles.heading
    });
  };
}
