// @ali/4ever-plugin-subject/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "isSubjectHeading", {
  enumerable: true,
  get: function () {
    return _utils.isSubjectHeading;
  }
});
Object.defineProperty(exports, "isSubject", {
  enumerable: true,
  get: function () {
    return _utils.isSubject;
  }
});
Object.defineProperty(exports, "createSubject", {
  enumerable: true,
  get: function () {
    return _utils.createSubject;
  }
});
Object.defineProperty(exports, "isEmptyNode", {
  enumerable: true,
  get: function () {
    return _utils.isEmptyNode;
  }
});
Object.defineProperty(exports, "isNormalParagraph", {
  enumerable: true,
  get: function () {
    return _utils.isNormalParagraph;
  }
});
Object.defineProperty(exports, "uniqueId", {
  enumerable: true,
  get: function () {
    return _utils.uniqueId;
  }
});
exports.OldBambooPlugin = void 0;

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;

var _utils = require("./utils/utils");
