// @ali/4ever-zhi/dist-node/plugins/audio/components/audio.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

const _createElement = /*#__PURE__*/_react.default.createElement;
const {
  MobileEmbedCard,
  AudioPreview
} = _everBamboo.EmbedPlugin;

class Audio extends _react.PureComponent {
  constructor(props) {
    super(props);
    this.ref = /*#__PURE__*/_react.default.createRef();
    this.previewRef = /*#__PURE__*/_react.default.createRef();

    this.getNodeData = () => {
      const {
        node
      } = this.props;
      const {
        data
      } = node || {};
      return data;
    };

    this.onMediaLoadError = () => {
      const {
        src
      } = this.getNodeData();
      const {
        onMediaLoadError
      } = this.props;

      if (onMediaLoadError) {
        onMediaLoadError(src);
      }
    };

    this.onMediaLoadSuccess = () => {
      const {
        src
      } = this.getNodeData();
      const {
        onMediaLoadSuccess
      } = this.props;

      if (onMediaLoadSuccess) {
        onMediaLoadSuccess(src);
      }
    };

    this.onDownloadFile = () => {
      const {
        src,
        name,
        type,
        size
      } = this.getNodeData();
      const {
        onDownload
      } = this.props;

      if (onDownload) {
        onDownload(src, {
          name,
          type,
          size
        });
      }
    };

    const {
      allowDownload
    } = props;
    this.state = {
      allowDownload: typeof allowDownload === 'boolean' ? allowDownload : false
    };
  }

  componentDidMount() {
    const {
      allowDownload
    } = this.props;
    const {
      src
    } = this.getNodeData();

    if (typeof allowDownload === 'function') {
      allowDownload(src).then(res => {
        this.setState({
          allowDownload: res
        });
      });
    }
  }

  renderPreview() {
    const {
      allowDownload
    } = this.state;
    const {
      transformAudioURL,
      locale
    } = this.props;
    const {
      name,
      size,
      src
    } = this.getNodeData();
    return /*#__PURE__*/_createElement(AudioPreview, {
      allowDownload: allowDownload,
      onMediaLoadSuccess: this.onMediaLoadSuccess,
      onMediaLoadError: this.onMediaLoadError,
      transformAudioURL: transformAudioURL,
      locale: locale,
      name: name,
      size: size,
      src: src,
      onDownloadFile: this.onDownloadFile
    });
  }

  renderCard() {
    const {
      name,
      src,
      type,
      size
    } = this.getNodeData();
    const {
      onPreview
    } = this.props;
    return /*#__PURE__*/_createElement(MobileEmbedCard, {
      name: name,
      onPreview: onPreview,
      src: src,
      type: type,
      size: size
    });
  }

  render() {
    const {
      viewType
    } = this.getNodeData();
    const {
      attributes
    } = this.props;
    return /*#__PURE__*/_createElement(_react.default.Fragment, null, /*#__PURE__*/_createElement("div", (0, _extends2.default)({}, attributes, {
      ref: this.ref
    }), viewType === 'preview' ? this.renderPreview() : this.renderCard()));
  }

}

var _default = Audio;
exports.default = _default;
