// @ali/we-design/dist-web/components/date-picker/Control.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { SuperArrowLeftBetaNormal, SuperArrowRightBetaNormal, ArrowLeftThinBetaNormal, ArrowRightThinBetaNormal } from "../../components/icons";
import { ControlRoot, ControlPrevWrap, ControlNextWrap, ControlMain, ArrowWrap } from "./styled";
export var Control = function Control(props) {
  var children = props.children,
      hasPrev = props.hasPrev,
      hasNext = props.hasNext,
      hasSuperPrev = props.hasSuperPrev,
      hasSuperNext = props.hasSuperNext,
      onClickSuperNext = props.onClickSuperNext,
      onClickNext = props.onClickNext,
      onClickSuperPrev = props.onClickSuperPrev,
      onClickPrev = props.onClickPrev,
      testid = props.testid;
  return /*#__PURE__*/_createElement(ControlRoot, null, /*#__PURE__*/_createElement(ControlPrevWrap, null, hasSuperPrev && /*#__PURE__*/_createElement(ArrowWrap, {
    onClick: onClickSuperPrev
  }, /*#__PURE__*/_createElement(SuperArrowLeftBetaNormal, {
    className: "arrow-icon",
    testid: testid
  })), hasPrev && /*#__PURE__*/_createElement(ArrowWrap, {
    onClick: onClickPrev
  }, /*#__PURE__*/_createElement(ArrowLeftThinBetaNormal, {
    className: "arrow-icon",
    testid: testid
  }))), /*#__PURE__*/_createElement(ControlMain, {
    "data-testid": testid
  }, children), /*#__PURE__*/_createElement(ControlNextWrap, null, hasNext && /*#__PURE__*/_createElement(ArrowWrap, {
    onClick: onClickNext
  }, /*#__PURE__*/_createElement(ArrowRightThinBetaNormal, {
    className: "arrow-icon",
    testid: testid
  })), hasSuperNext && /*#__PURE__*/_createElement(ArrowWrap, {
    onClick: onClickSuperNext
  }, /*#__PURE__*/_createElement(SuperArrowRightBetaNormal, {
    className: "arrow-icon",
    testid: testid
  }))));
};
Control.defaultProps = {
  hasSuperNext: true,
  hasSuperPrev: true,
  hasNext: true,
  hasPrev: true,
  onClickSuperPrev: function onClickSuperPrev() {},
  onClickSuperNext: function onClickSuperNext() {},
  onClickPrev: function onClickPrev() {},
  onClickNext: function onClickNext() {}
};
