// @ali/4ever-bamboo/dist-node/plugins/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CalendarCardPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginCalendarCard.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "CardPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginCard.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "VoteCardPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginVoteCard.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "ColumnsPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginColumns.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "ColumnsRenderNodeEdit", {
  enumerable: true,
  get: function () {
    return _everPluginColumns.ColumnsRenderNodeEdit;
  }
});
Object.defineProperty(exports, "CommentPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginComment.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "DraggablePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginDraggable.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "EmbedPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginEmbed.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "HrPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginHr.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "ImagePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginImage.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "ImTagPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginImTag.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "FormulaPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginFormula.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "ListPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginList.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "Checkbox", {
  enumerable: true,
  get: function () {
    return _everPluginList.ListCheckBox;
  }
});
Object.defineProperty(exports, "IndentPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginIndent.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "LinkPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginLink.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "MarksPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.zhiFactory;
  }
});
Object.defineProperty(exports, "TextBackgroundPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginTextBackground.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "MentionPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginMention.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "ParagraphPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginParagraph.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "PcContextMenu", {
  enumerable: true,
  get: function () {
    return _everPluginPcContexMenu.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "ContextMenu", {
  enumerable: true,
  get: function () {
    return _everPluginPcContexMenu.PCContextMenu;
  }
});
Object.defineProperty(exports, "ReferPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginRefer.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "TablePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginTable.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "TagPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginTag.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "StickerPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "CalloutPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginCallout.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "HeadingGroupPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginHeadingGroup.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "SubjectPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginSubject.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "GroupFactory", {
  enumerable: true,
  get: function () {
    return _everPluginGroupFactory.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "ListGroupPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginListGroup.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "HeadingPlugin", {
  enumerable: true,
  get: function () {
    return _everPluginHeading.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "FindAndReplacePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginFindAndReplace.createFindAndReplacePlugin;
  }
});
Object.defineProperty(exports, "DocStylePlugin", {
  enumerable: true,
  get: function () {
    return _everPluginDocStyle.OldBambooPlugin;
  }
});
Object.defineProperty(exports, "HoverBlock", {
  enumerable: true,
  get: function () {
    return _everUtils.HoverBlock;
  }
});
exports.PluginRoles = exports.FileTypeMap = void 0;

var _everPluginCalendarCard = require("@ali/4ever-plugin-calendar-card");

var _everPluginCard = require("@ali/4ever-plugin-card");

var _everPluginVoteCard = require("@ali/4ever-plugin-vote-card");

var _everPluginColumns = require("@ali/4ever-plugin-columns");

var _everPluginComment = require("@ali/4ever-plugin-comment");

var _everPluginDraggable = require("@ali/4ever-plugin-draggable");

var _everPluginEmbed = require("@ali/4ever-plugin-embed");

var _everPluginHr = require("@ali/4ever-plugin-hr");

var _everPluginImage = require("@ali/4ever-plugin-image");

var _everPluginImTag = require("@ali/4ever-plugin-im-tag");

var _everPluginFormula = require("@ali/4ever-plugin-formula");

var _everPluginCode = require("@ali/4ever-plugin-code");

var _everPluginInlineCode = require("@ali/4ever-plugin-inline-code");

var _everPluginList = require("@ali/4ever-plugin-list");

var _everPluginIndent = require("@ali/4ever-plugin-indent");

var _everPluginLink = require("@ali/4ever-plugin-link");

var _everPluginMarks = require("@ali/4ever-plugin-marks");

var _everPluginTextBackground = require("@ali/4ever-plugin-text-background");

var _everPluginMention = require("@ali/4ever-plugin-mention");

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _everPluginPcContexMenu = require("@ali/4ever-plugin-pc-contex-menu");

var _everPluginRefer = require("@ali/4ever-plugin-refer");

var _everPluginTable = require("@ali/4ever-plugin-table");

var _everPluginTag = require("@ali/4ever-plugin-tag");

var _everPluginSticker = require("@ali/4ever-plugin-sticker");

var _everPluginCallout = require("@ali/4ever-plugin-callout");

var _everPluginHeadingGroup = require("@ali/4ever-plugin-heading-group");

var _everPluginSubject = require("@ali/4ever-plugin-subject");

var _everPluginGroupFactory = require("@ali/4ever-plugin-group-factory");

var _everPluginListGroup = require("@ali/4ever-plugin-list-group");

var _everPluginAlignment = require("@ali/4ever-plugin-alignment");

var _everPluginBlockquote = require("@ali/4ever-plugin-blockquote");

var _everPluginClearStyle = require("@ali/4ever-plugin-clear-style");

var _everPluginColor = require("@ali/4ever-plugin-color");

var _everPluginFonts = require("@ali/4ever-plugin-fonts");

var _everPluginFormatPainter = require("@ali/4ever-plugin-format-painter");

var _everPluginHeading = require("@ali/4ever-plugin-heading");

var _everPluginHighlight = require("@ali/4ever-plugin-highlight");

var _everPluginHistory = require("@ali/4ever-plugin-history");

var _everPluginLetterSpacing = require("@ali/4ever-plugin-letter-spacing");

var _everPluginLineHeight = require("@ali/4ever-plugin-line-height");

var _everPluginSz = require("@ali/4ever-plugin-sz");

var _everPluginTemplateButton = require("@ali/4ever-plugin-template-button");

var _everPluginTranslate = require("@ali/4ever-plugin-translate");

var _everPluginUnderline = require("@ali/4ever-plugin-underline");

var _everPluginVertAlign = require("@ali/4ever-plugin-vert-align");

var _everPluginRefblock = require("@ali/4ever-plugin-refblock");

var _everPluginFindAndReplace = require("@ali/4ever-plugin-find-and-replace");

var _everPluginHoverToolbar = require("@ali/4ever-plugin-hover-toolbar");

var _everPluginDocStyle = require("@ali/4ever-plugin-doc-style");

var _everUtils = require("@ali/4ever-utils");

const {
  FileTypeMap
} = _everPluginEmbed.OldBambooPlugin;
exports.FileTypeMap = FileTypeMap;
const PluginRoles = { ..._everPluginAlignment.OldBambooPlugin.AlignmentRole,
  ..._everPluginBlockquote.OldBambooPlugin.BlockquoteRole,
  ..._everPluginClearStyle.OldBambooPlugin.ClearStyleRole,
  ..._everPluginColor.OldBambooPlugin.ColorRole,
  ..._everPluginColumns.OldBambooPlugin.ColumnsRole,
  ..._everPluginComment.OldBambooPlugin.CommentRole,
  ..._everPluginDraggable.OldBambooPlugin.DraggableRole,
  ..._everPluginEmbed.OldBambooPlugin.EmbedRole,
  ..._everPluginFonts.OldBambooPlugin.FontRole,
  ..._everPluginFormatPainter.OldBambooPlugin.FormatPainterRole,
  ..._everPluginHeading.OldBambooPlugin.HeadingRole,
  ..._everPluginHighlight.OldBambooPlugin.HighlightRole,
  ..._everPluginHistory.HistoryRole,
  ..._everPluginHr.OldBambooPlugin.HrRole,
  ..._everPluginImage.OldBambooPlugin.ImageRole,
  ..._everPluginIndent.OldBambooPlugin.IndentRole,
  ..._everPluginLetterSpacing.OldBambooPlugin.LetterSpacingRole,
  ..._everPluginLineHeight.OldBambooPlugin.LineHeightRole,
  ..._everPluginLink.OldBambooPlugin.LinkRole,
  ..._everPluginList.OldBambooPlugin.ListRole,
  ..._everPluginParagraph.OldBambooPlugin.ParagraphRole,
  ..._everPluginRefer.OldBambooPlugin.ReferRole,
  ..._everPluginSticker.OldBambooPlugin.StickerRole,
  ..._everPluginSz.OldBambooPlugin.SzRole,
  ..._everPluginTable.OldBambooPlugin.TableRole,
  ..._everPluginTemplateButton.TemplateButtonRole,
  ..._everPluginTranslate.OldBambooPlugin.TranslateRole,
  ..._everPluginUnderline.OldBambooPlugin.UnderlineRole,
  ..._everPluginVertAlign.OldBambooPlugin.vertAlignRole,
  ..._everPluginCallout.OldBambooPlugin.CalloutRole,
  ..._everPluginMention.OldBambooPlugin.MentionRole,
  ..._everPluginRefblock.RefBlockRole,
  ..._everPluginCode.CodeRole,
  ..._everPluginInlineCode.InlineCodeRole,
  ..._everPluginHoverToolbar.HoverToolbarRole
};
exports.PluginRoles = PluginRoles;
