// @ali/4ever-bi/dist-node/plugins/onlineVideo/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.insertOnlineVideo = insertOnlineVideo;
exports.removeOnlineVideo = removeOnlineVideo;
exports.updateOnlineVideo = updateOnlineVideo;
exports.focusOnlineVideoEditor = focusOnlineVideoEditor;
exports.selectOnlineVideo = selectOnlineVideo;
exports.ONLINEVIDEO_SELECT = exports.ONLINEVIDEO_FOCUS_EDITOR = exports.ONLINEVIDEO_UPDATE = exports.ONLINEVIDEO_REMOVE = exports.ONLINEVIDEO_INSERT = void 0;
const ONLINEVIDEO_INSERT = Symbol('ONLINEVIDEO_INSERT');
exports.ONLINEVIDEO_INSERT = ONLINEVIDEO_INSERT;
const ONLINEVIDEO_REMOVE = Symbol('ONLINEVIDEO_REMOVE');
exports.ONLINEVIDEO_REMOVE = ONLINEVIDEO_REMOVE;
const ONLINEVIDEO_UPDATE = Symbol('ONLINEVIDEO_UPDATE');
exports.ONLINEVIDEO_UPDATE = ONLINEVIDEO_UPDATE;
const ONLINEVIDEO_FOCUS_EDITOR = Symbol('ONLINEVIDEO_FOCUS_EDITOR');
exports.ONLINEVIDEO_FOCUS_EDITOR = ONLINEVIDEO_FOCUS_EDITOR;
const ONLINEVIDEO_SELECT = Symbol('ONLINEVIDEO_SELECT');
exports.ONLINEVIDEO_SELECT = ONLINEVIDEO_SELECT;

function insertOnlineVideo() {
  return {
    type: ONLINEVIDEO_INSERT
  };
}

function removeOnlineVideo(node) {
  return {
    type: ONLINEVIDEO_REMOVE,
    payload: {
      node
    }
  };
}

function updateOnlineVideo(onlineVideo, data) {
  return {
    type: ONLINEVIDEO_UPDATE,
    payload: {
      node: onlineVideo,
      data
    }
  };
}

function focusOnlineVideoEditor(node) {
  return {
    type: ONLINEVIDEO_FOCUS_EDITOR,
    payload: {
      node
    }
  };
}

function selectOnlineVideo(node) {
  return {
    type: ONLINEVIDEO_SELECT,
    payload: {
      node
    }
  };
}
