// @ali/4ever-bi/dist-node/plugins/onlineVideo/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.actions = exports.default = void 0;

var _menubar = _interopRequireDefault(require("./menubar"));

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _insertOnlineVideo = _interopRequireDefault(require("./commands/insertOnlineVideo"));

var _createRenderNode = _interopRequireDefault(require("./createRenderNode"));

var actions = _interopRequireWildcard(require("./actions"));

exports.actions = actions;

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onlineVideoEditor = _interopRequireDefault(require("./components/onlineVideoEditor"));

var _OnlineVideoInlineToolbar = _interopRequireDefault(require("./components/OnlineVideoInlineToolbar"));

var _everPluginOnlineVideo = require("@ali/4ever-plugin-online-video");

const state = {};

function createOnlineVideoPlugin(config) {
  const onlineVideoConfig = config.onlineVideo || {};
  return { ...(0, _everPluginOnlineVideo.moFactory)({}),
    menubar: (0, _menubar.default)(),
    toolbar: (0, _toolbar.default)(config),
    commands: {
      insertOnlineVideo: (0, _insertOnlineVideo.default)(state)
    },
    renderNode: (0, _createRenderNode.default)(onlineVideoConfig, state, _onlineVideoEditor.default, _OnlineVideoInlineToolbar.default),
    onAction: (0, _onAction.default)(state),
    models: {
      klass: 'block',
      type: 'onlineVideo'
    }
  };
}

createOnlineVideoPlugin.actions = actions;
var _default = createOnlineVideoPlugin;
exports.default = _default;
