// @ali/we-design/dist-web/components/file-menu/styled.js
import styled from 'styled-components';
import { optionWrapStyle } from "../../utils/styled.mixin";
import { withTheme } from "../theme";
export var LeftWrap = /*#__PURE__*/styled.div(["padding:20px 0;"]);
export var RightWrap = /*#__PURE__*/styled.div(["padding:20px;border-left:1px solid rgba(17,31,44,0.08);"]);
export var TriggerWrapper = /*#__PURE__*/styled.div(["height:24px;width:57px;border-radius:4px;display:flex;justify-content:center;align-items:center;font-size:13px;cursor:pointer;&:hover,&.active{background-color:rgba(17,31,44,0.08);}.trigger-icon{font-size:14px;margin-right:4px;}"]);
export var SuffixWrapper = /*#__PURE__*/styled.div(["font-size:12px;color:rgba(17,31,44,0.40);"]);
export var Divider = withTheme(styled.div(["height:1px;margin:", " 8px;background-color:", ";"], function (_ref) {
  var theme = _ref.theme;
  return theme.ui.marginMedium;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.commonLineLightColor;
}));
export var MenuItemWrap = withTheme(styled.div(["", ";height:36px;padding:6px 8px;margin-bottom:2px;&.active{background-color:rgba(17,31,44,0.12);}.arrow-icon{position:relative;right:-2px;color:", ";&.disabled{color:", ";}}"], optionWrapStyle, function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.commonBaseColor3;
}, function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.disabledColor;
}));
export var MenuItemTitle = withTheme(styled.div(["min-width:120px;padding-right:8px;"]));
