// @ali/4ever-plugin-mention/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.moFactory;
  }
});
Object.defineProperty(exports, "zhiFactory", {
  enumerable: true,
  get: function () {
    return _index2.zhiFactory;
  }
});
Object.defineProperty(exports, "biFactory", {
  enumerable: true,
  get: function () {
    return _index3.biFactory;
  }
});
Object.defineProperty(exports, "mobileBiFactory", {
  enumerable: true,
  get: function () {
    return _mobile.mobileBiFactory;
  }
});
Object.defineProperty(exports, "Mention", {
  enumerable: true,
  get: function () {
    return _models.default;
  }
});
Object.defineProperty(exports, "ELEMENT_TYPE", {
  enumerable: true,
  get: function () {
    return _meta.ELEMENT_TYPE;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index4.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index4.createServerSerializationPlugin;
  }
});
Object.defineProperty(exports, "OldBiCommands", {
  enumerable: true,
  get: function () {
    return _commands.commands;
  }
});
Object.defineProperty(exports, "creatQuickInsertPlugin", {
  enumerable: true,
  get: function () {
    return _quickInsert.default;
  }
});
Object.defineProperty(exports, "isSelectionInMentionUtil", {
  enumerable: true,
  get: function () {
    return _utils.isSelectionInMention;
  }
});
Object.defineProperty(exports, "mobileInsertMentionCommand", {
  enumerable: true,
  get: function () {
    return _insertMention.default;
  }
});
exports.OldBambooPlugin = exports.OldBiActions = exports.mobileZhiFactory = void 0;

var _index = require("./mo/index");

var _index2 = require("./zhi/index");

var _index3 = require("./bi/index");

var _mobile = require("./bi/mobile");

require("./utils/types");

var _models = _interopRequireDefault(require("./mo/models"));

var _meta = require("./utils/meta");

var _index4 = _interopRequireWildcard(require("./mo/serializer/index"));

var _commands = require("./bi/commands");

var OldBiActions = _interopRequireWildcard(require("./bi/actions"));

exports.OldBiActions = OldBiActions;

var _quickInsert = _interopRequireDefault(require("./bi/quickInsert"));

var _utils = require("./bi/utils");

var _insertMention = _interopRequireDefault(require("./bi/mobile/insertMention"));

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;

const mobileZhiFactory = configs => (0, _index2.zhiFactory)(configs); // make types combination work


exports.mobileZhiFactory = mobileZhiFactory;
