// @ali/4ever-plugin-card/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.default;
  }
});
Object.defineProperty(exports, "Card", {
  enumerable: true,
  get: function () {
    return _models.default;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.createServerSerializationPlugin;
  }
});
Object.defineProperty(exports, "DEFAULT_MARGIN_TOP", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_MARGIN_TOP;
  }
});
Object.defineProperty(exports, "DEFAULT_MARGIN_BOTTOM", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_MARGIN_BOTTOM;
  }
});
Object.defineProperty(exports, "CardWrapper", {
  enumerable: true,
  get: function () {
    return _styled.CardWrapper;
  }
});
exports.OldBambooPlugin = void 0;

var _index = _interopRequireDefault(require("./mo/index"));

var _models = _interopRequireDefault(require("./mo/models"));

var _index2 = _interopRequireWildcard(require("./mo/serializer/index"));

var _constants = require("./utils/constants");

var _styled = require("./components/styled");

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;
