// @ali/we-design-next/dist-node/form/radio/Radio.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Radio = void 0;
var _react = _interopRequireDefault(require("react"));
var _useDefaultProp = require("../../common/react/useDefaultProp");
var _styled = require("./styled");
const _createElement = /*#__PURE__*/_react.default.createElement;
const Radio = /*#__PURE__*/_react.default.memo( /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    defaultChecked = false,
    checked: _checked,
    onChange,
    disabled = false,
    children
  } = props;
  const [checked, setChecked] = (0, _useDefaultProp.useDefaultProp)(defaultChecked, _checked, onChange, e => e.target.checked);
  return /*#__PURE__*/_createElement(_styled.RadioWrap, {
    disabled: !!disabled
  }, /*#__PURE__*/_createElement(_styled.RadioInput, {
    ref: ref,
    type: "radio",
    onChange: setChecked,
    disabled: disabled,
    checked: checked
  }), /*#__PURE__*/_createElement(_styled.RadioIcon, {
    disabled: !!disabled,
    checked: !!checked
  }), children && /*#__PURE__*/_createElement(_styled.RadioText, null, children));
}));
exports.Radio = Radio;
