// @ali/4ever-bi/dist-node/plugins/calendarCard/handlers/createMenu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _everFactory = require("@ali/4ever-factory");

const _createElement = /*#__PURE__*/_react.default.createElement;

const createMenubar = calendarConfig => {
  return function CalendarCardMenu(props) {
    const {
      menuItem,
      controller
    } = props;
    const {
      subjectPlaceholder,
      timePlaceholder,
      participantPlaceholder
    } = calendarConfig.locale || {};
    const {
      title,
      ...rest
    } = menuItem;
    const handleClick = (0, _react.useCallback)(() => {
      controller.run('onAction', (0, _everFactory.insertCard)({
        cardType: 'calendar',
        height: 102,
        metadata: {
          isPlaceholder: true,
          subject: subjectPlaceholder,
          time: timePlaceholder,
          topReceivers: [{
            isReceiverPlaceholder: true,
            name: participantPlaceholder,
            type: 2
          }]
        }
      }));
    }, [controller]);
    return /*#__PURE__*/_createElement(_weDesign.MenuItem, (0, _extends2.default)({}, rest, {
      onClick: handleClick
    }), title || '');
  };
};

var _default = createMenubar;
exports.default = _default;
