// @ali/4ever-plugin-heading/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.default;
  }
});
Object.defineProperty(exports, "Heading", {
  enumerable: true,
  get: function () {
    return _models.default;
  }
});
Object.defineProperty(exports, "H1", {
  enumerable: true,
  get: function () {
    return _styled.H1;
  }
});
Object.defineProperty(exports, "H2", {
  enumerable: true,
  get: function () {
    return _styled.H2;
  }
});
Object.defineProperty(exports, "H3", {
  enumerable: true,
  get: function () {
    return _styled.H3;
  }
});
Object.defineProperty(exports, "H4", {
  enumerable: true,
  get: function () {
    return _styled.H4;
  }
});
Object.defineProperty(exports, "H5", {
  enumerable: true,
  get: function () {
    return _styled.H5;
  }
});
Object.defineProperty(exports, "H6", {
  enumerable: true,
  get: function () {
    return _styled.H6;
  }
});
Object.defineProperty(exports, "createHeadingId", {
  enumerable: true,
  get: function () {
    return _createHeadingId.createHeadingId;
  }
});
Object.defineProperty(exports, "getHeadings", {
  enumerable: true,
  get: function () {
    return _getHeadings.getHeadings;
  }
});
Object.defineProperty(exports, "HeadingTypes", {
  enumerable: true,
  get: function () {
    return _constants.HeadingTypes;
  }
});
Object.defineProperty(exports, "HeadingRole", {
  enumerable: true,
  get: function () {
    return _constants.HeadingRole;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.createServerSerializationPlugin;
  }
});
Object.defineProperty(exports, "HEADING_SET", {
  enumerable: true,
  get: function () {
    return _actions.HEADING_SET;
  }
});
Object.defineProperty(exports, "HEADING_TRANSFORM", {
  enumerable: true,
  get: function () {
    return _actions.HEADING_TRANSFORM;
  }
});
Object.defineProperty(exports, "setHeading", {
  enumerable: true,
  get: function () {
    return _actions.setHeading;
  }
});
Object.defineProperty(exports, "transformHeading", {
  enumerable: true,
  get: function () {
    return _actions.transformHeading;
  }
});
exports.OldBambooPlugin = void 0;

var _index = _interopRequireDefault(require("./mo/index"));

var _models = _interopRequireDefault(require("./mo/models"));

var _styled = require("./components/styled");

var _createHeadingId = require("./utils/createHeadingId");

var _getHeadings = require("./utils/getHeadings");

var _constants = require("./utils/constants");

var _index2 = _interopRequireWildcard(require("./mo/serializer/index"));

var _actions = require("./bi/actions");

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;
