// @ali/4ever-plugin-image/dist-node/utils/components/image/MetaImage.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MetaImage = exports.HoverOutline = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _types = require("./types");

var _loading = _interopRequireDefault(require("../loading"));

var _getReloadSrc = _interopRequireDefault(require("../../utils/getReloadSrc"));

var _getBorder = _interopRequireDefault(require("../../utils/getBorder"));

var _aslService = require("../../utils/aslService");

var _imageBorder = _interopRequireDefault(require("../imageBorder"));

var _constants = require("../../constants");

const _createElement = /*#__PURE__*/_react.default.createElement;

const Img = /*#__PURE__*/_styledComponents.default.img(["display:inline-block;vertical-align:initial;user-select:none;image-orientation:from-image;"]);

const HoverOutline = /*#__PURE__*/_styledComponents.default.span(["position:absolute;width:100%;height:100%;left:0;top:0;border-color:transparent;transition:border-color 0.3s ease-in-out;pointer-events:none;"]);

exports.HoverOutline = HoverOutline;

const MetaImage = props => {
  const {
    style,
    data,
    imgSrc,
    onStatusChange,
    isMobile,
    onImgRef,
    onWrapperRef,
    disableLoading,
    forceLoading,
    ...restProps
  } = props;

  const [status, setStatus] = _react.default.useState(_types.Status.Loading);

  const {
    width,
    height,
    src: imageDataSrc,
    outline,
    radius,
    shadow
  } = data;

  const imgRef = _react.default.useRef();

  const handleLoad = _react.default.useCallback(() => {
    setStatus(_types.Status.Success);
    onStatusChange(_types.Status.Success);
  }, [imageDataSrc]);

  const handleError = _react.default.useCallback(() => {
    setStatus(_types.Status.Error);
    onStatusChange(_types.Status.Error);
  }, []);

  const borderStyle = {
    border: (0, _getBorder.default)(outline),
    borderRadius: radius,
    boxShadow: shadow
  };

  const onReload = () => {
    if (!imgRef.current) {
      return;
    }

    setStatus(_types.Status.Loading);
    const reloadSrc = (0, _getReloadSrc.default)(imgSrc);
    imgRef.current.src = reloadSrc;
  };

  const onRef = _react.default.useCallback(elem => {
    imgRef.current = elem;

    if (typeof onImgRef === 'function') {
      onImgRef(elem);
    } else if (onImgRef) {
      onImgRef.current = elem;
    }
  }, []);

  const isStatic = status === _types.Status.Success && !(0, _aslService.isValidSize)(width, height);
  const isUploading = !imageDataSrc && imgSrc.startsWith('blob:');
  return /*#__PURE__*/_createElement(_react.default.Fragment, null, /*#__PURE__*/_createElement(_imageBorder.default, {
    style: borderStyle,
    onRef: onWrapperRef,
    isStatic: isStatic
  }, imgSrc && /*#__PURE__*/_createElement(Img, (0, _extends2.default)({
    style: style,
    src: imgSrc,
    "data-src": imageDataSrc,
    "data-width": width,
    "data-height": height,
    "data-type": "image",
    "data-status": status,
    onLoad: handleLoad,
    onError: handleError,
    ref: onRef,
    "data-testid": "cangjie-image"
  }, restProps)), !isUploading && (!imgSrc || status !== _types.Status.Success && !disableLoading) && /*#__PURE__*/_createElement(_loading.default, {
    status: status,
    src: imageDataSrc,
    renderSrc: imgSrc,
    isMobile: isMobile,
    onReload: onReload,
    forceLoading: forceLoading,
    enableBlink: !!imgSrc
  })), /*#__PURE__*/_createElement(HoverOutline, {
    className: _constants.IMAGE_HOVER_BORDER,
    style: {
      borderRadius: radius
    }
  }));
};

exports.MetaImage = MetaImage;
