// @ali/4ever-bi/dist-web/plugins/image/components/next/createImageButton.js
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { IconPicture } from '@ali/4ever-component';
import { PluginRoles } from '@ali/4ever-bamboo';
import { selectImage } from "../../actions";
import { isButtonDisabled } from "../imageButton";

var handleClick = function handleClick(controller) {
  controller.run('onAction', selectImage());
};

var _ref = /*#__PURE__*/_createElement(IconPicture, null);

export default function createImageButton(props) {
  var controller = props.controller,
      locale = props.locale,
      icon = props.icon,
      title = props.title,
      tooltip = props.tooltip,
      badgeDot = props.badgeDot,
      disabled = props.disabled;
  var isDisabled = disabled || isButtonDisabled(controller);
  var buttonConfig = {
    icon: icon || _ref,
    onClick: handleClick,
    title: title || (locale == null ? void 0 : locale.quickInsertTitle),
    tooltip: _extends({
      title: title || (locale == null ? void 0 : locale.quickInsertTitle),
      description: 'Markdown: ![]() ' + ((locale == null ? void 0 : locale['space']) || 'Space')
    }, tooltip),
    badgeDot: badgeDot,
    disabled: isDisabled,
    role: PluginRoles.image,
    testid: "bi-toolbar-picture"
  };
  return buttonConfig;
}
