// @ali/4ever-bi/dist-web/plugins/link/linkCard/actions.js
/* #region  actionTypes */
export var LINK_CARD_PURE_LINK_TO_BLOCK_LINK_CARD = Symbol('LINK_CARD_PURE_LINK_TO_BLOCK_LINK_CARD');
export var LINK_CARD_TEXT_LINK_TO_BLOCK_LINK_CARD = Symbol('LINK_CARD_TEXT_LINK_TO_BLOCK_LINK_CARD');
export var LINK_CARD_BLOCK_LINK_CARD_TO_TEXT_LINK = Symbol('LINK_CARD_BLOCK_LINK_CARD_TO_TEXT_LINK');
export var LINK_CARD_COPY_LINK = Symbol('LINK_CARD_COPY_LINK');
export var LINK_CARD_CUT_LINK = Symbol('LINK_CARD_CUT_LINK');
export var LINK_CARD_DELETE_LINK = Symbol('LINK_CARD_DELETE_LINK');
export var LINK_CARD_UPDATE_DATA = Symbol('LINK_CARD_UPDATE_DATA');
export var LINK_CARD_UNLINK = Symbol('LINK_CARD_UNLINK');
export var LINK_CARD_INSERT_CARD = Symbol('LINK_CARD_INSERT_CARD');
/* #endregion */

/* #region  actions */

export function convertPureLinkToBlockLinkCard(link, offset, text, href, srcHref, cardInfo, needFetchLinkInfo) {
  return {
    type: LINK_CARD_PURE_LINK_TO_BLOCK_LINK_CARD,
    payload: {
      node: link,
      offset: offset,
      text: text,
      href: href,
      srcHref: srcHref,
      cardInfo: cardInfo,
      needFetchLinkInfo: needFetchLinkInfo
    }
  };
}
export function convertTextLinkToBlockLinkCard(node, text, href, cardInfo) {
  return {
    type: LINK_CARD_TEXT_LINK_TO_BLOCK_LINK_CARD,
    payload: {
      node: node,
      text: text,
      href: href,
      cardInfo: cardInfo
    }
  };
}
export function convertBlockLinkCardToTextLink(node, data) {
  return {
    type: LINK_CARD_BLOCK_LINK_CARD_TO_TEXT_LINK,
    payload: {
      node: node,
      data: data
    }
  };
}
export function copyBlockLinkCard(node) {
  return {
    type: LINK_CARD_COPY_LINK,
    payload: {
      node: node
    }
  };
}
export function cutBlockLinkCard(node) {
  return {
    type: LINK_CARD_CUT_LINK,
    payload: {
      node: node
    }
  };
}
export function deleteBlockLinkCard(node) {
  return {
    type: LINK_CARD_DELETE_LINK,
    payload: {
      node: node
    }
  };
}
export function updateBlockLinkCardData(node, text, href, cardInfo) {
  return {
    type: LINK_CARD_UPDATE_DATA,
    payload: {
      node: node,
      text: text,
      href: href,
      cardInfo: cardInfo
    }
  };
}
export function unlinkBlockCardData(node) {
  return {
    type: LINK_CARD_UNLINK,
    payload: {
      node: node
    }
  };
}
export function insertBlockLinkCard(text, href, cardInfo) {
  return {
    type: LINK_CARD_INSERT_CARD,
    payload: {
      text: text,
      href: href,
      cardInfo: cardInfo
    }
  };
}
/* #endregion */
