// @ali/4ever-bi/dist-web/plugins/link/linkCard/index.js
import _extends from "@babel/runtime/helpers/extends";
import locale from "../../../locales/zh_CN";
import { LinkPlugin } from '@ali/4ever-bamboo';
import { cardFactory } from '@ali/4ever-factory';
import createRenderNode from "./createRenderNode";
import careateOnAction from "./onAction";
import * as actions from "./actions";
import { copyLink } from "./commands";
export { actions };

function createBiLinkCardPlugin(bizConfig) {
  return cardFactory(function (bizConfig) {
    var _bizConfig$clipboard;

    var linkCardConfig = _extends({
      locale: _extends({}, locale.link.locale)
    }, bizConfig.link);

    var serializer = (_bizConfig$clipboard = bizConfig.clipboard) == null ? void 0 : _bizConfig$clipboard.serializer;
    var renderNode = createRenderNode(linkCardConfig, false);
    return {
      cardType: 'link',
      onAction: careateOnAction(linkCardConfig, serializer),
      renderNode: renderNode,
      queries: {
        getMinTableColWidth: LinkPlugin.getMinTableColWidthForBlockLinkCard
      }
    };
  })(bizConfig);
}

;
createBiLinkCardPlugin.actions = actions;
createBiLinkCardPlugin.utils = {
  copyLink: copyLink
};
export default createBiLinkCardPlugin;
