// @ali/we-ui/dist-web/utils/dom.js
import ReactDOM from 'react-dom';
export function getOffset(el) {
  var left = el.offsetLeft;
  var top = el.offsetTop;
  if (!el.offsetParent) return {
    left: left,
    top: top
  };
  var parent = el.offsetParent.tagName.toLocaleLowerCase();
  if (parent === 'body') return {
    left: left,
    top: top
  };
  var parentPos = getOffset(el.offsetParent);
  return {
    left: left + parentPos.left,
    top: top + parentPos.top
  };
}
export function elementContain(box, el) {
  if (!box) return false;
  var target = el;

  while (target) {
    if (target === box) return true;
    target = target.parentElement;
  }

  return false;
}
var mobileReg = /(android|iphone|symbianos|windows phone|ipad|ipod)/i;
export var isMobile = mobileReg.test(navigator.userAgent.toLocaleLowerCase());
var iosReg = /os ([\\.\\_\d]+) like mac os/i;
export var isIOS = iosReg.test(navigator.userAgent.toLocaleLowerCase());
/**
 * Return if a node is a DOM node. Else will return by `findDOMNode`
 */

export function findDOMNode(node) {
  if (node instanceof HTMLElement) {
    return node;
  } // eslint-disable-next-line react/no-find-dom-node


  return ReactDOM.findDOMNode(node);
}
