// @ali/4ever-plugin-color/dist-web/bi/components/next/colorButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
;
import { setColor } from "../../actions";
import { DEFAULT_COLOR } from "../../constants";
import { setColorPanelMode } from '@ali/4ever-plugin-pref';
import { FontColorPicker } from '@ali/4ever-component';
import { EColorPanelMode } from '@ali/we-toolbar';
export default function ColorButton(props) {
  var _controller$query;

  var controller = props.controller,
      locale = props.locale,
      rest = _objectWithoutPropertiesLoose(props, ["controller", "locale"]);

  var color = controller.query('getColor');
  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var handleChange = React.useCallback(function (colorValue) {
    return controller.run('onAction', setColor(colorValue));
  }, [controller]);
  var isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInLinkCard');
  var currentPreset = (_controller$query = controller.query('getColorPanelMode')) == null ? void 0 : _controller$query.mode;
  var handleSelectValueChange = React.useCallback(function (mode) {
    controller.run('onAction', setColorPanelMode({
      mode: mode === 'art' ? EColorPanelMode.art : EColorPanelMode.normal
    }));
  }, [controller]);
  return /*#__PURE__*/_createElement(FontColorPicker, _extends({
    testid: "bi-toolbar-color",
    disabled: !!isDisabled,
    color: color,
    onColorChange: handleChange,
    defaultColor: DEFAULT_COLOR,
    autoColor: DEFAULT_COLOR,
    title: locale.buttonTitle,
    tooltip: locale.buttonTitle,
    currentPreset: currentPreset,
    onCurrentPresetChange: handleSelectValueChange
  }, rest));
}
