// @ali/4ever-plugin-paragraph/dist-node/components/sidebar/common.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseInputNumberPara = parseInputNumberPara;
exports.checkInputValue = checkInputValue;
exports.UnitSelect = UnitSelect;
exports.commonSelectStyle = exports.inputNumberStyle = exports.commonInputNumberProps = exports.formatNumber100 = exports.formatNumber10 = void 0;

var React = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _everUtils = require("@ali/4ever-utils");

const _createElement = /*#__PURE__*/React.createElement;

const createNumberFormater = fix => {
  const f = fix || 100;
  return number => {
    if (number === '') return '';
    if (!number || typeof number === 'number') return `${number || 0}`;
    if (/\.$/.test(number)) return number;
    const res = Math.round(parseFloat(number) * f) / f;
    if ((0, _everUtils.isNaN)(res)) return '';
    return `${res}`;
  };
};

const unitSelectStyle = {
  width: 50,
  height: 28,
  fontSize: 13
};
const formatNumber10 = createNumberFormater(10);
exports.formatNumber10 = formatNumber10;
const formatNumber100 = createNumberFormater(100);
exports.formatNumber100 = formatNumber100;

function parseInputNumberPara(arg) {
  // 点击 上下 button 为 number
  if (typeof arg === 'number') return arg; // 按键盘上下键为 string

  if (typeof arg === 'string') return parseFloat(arg); // 按 enter 为 event

  if (arg?.target) return parseFloat(arg.target.value);
  return NaN;
}

const commonInputNumberProps = {
  min: 0,
  placeholder: '',
  step: 1
};
exports.commonInputNumberProps = commonInputNumberProps;

function checkInputValue(value, min) {
  if ((0, _everUtils.isNaN)(value)) return false;
  if (value < (min || 0)) return false;
  return true;
}

const inputNumberStyle = { ...unitSelectStyle,
  width: 60,
  margin: '0 5px'
};
exports.inputNumberStyle = inputNumberStyle;
const commonSelectStyle = {
  height: 28,
  width: 90,
  fontSize: 13,
  padding: '0 4px 0 0'
};
exports.commonSelectStyle = commonSelectStyle;

function UnitSelect(props) {
  const {
    value,
    units,
    onChange
  } = props;
  const options = units.map(unit => {
    return /*#__PURE__*/_createElement(_weDesign.Select.Option, {
      value: unit.key
    }, unit.value);
  });
  return /*#__PURE__*/_createElement(_weDesign.Select, {
    value: value,
    style: unitSelectStyle,
    onChange: onChange
  }, options);
}
