// @ali/4ever-plugin-paragraph/dist-node/mo/serializer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createSerializer;
exports.createServerSerializationPlugin = createServerSerializationPlugin;

var _jsonMLToValue = _interopRequireDefault(require("./jsonMLToValue"));

var _valueToJsonML = _interopRequireDefault(require("./valueToJsonML"));

var _htmlToJsonML = _interopRequireDefault(require("./htmlToJsonML"));

var _jsonMLToHTML = _interopRequireDefault(require("./jsonMLToHTML"));

var _valueToText = _interopRequireDefault(require("./valueToText"));

function createSerializer(configs) {
  return {
    jsonMLToValue: _jsonMLToValue.default,
    valueToJsonML: _valueToJsonML.default,
    htmlToJsonML: _htmlToJsonML.default,
    jsonMLToHTML: (0, _jsonMLToHTML.default)(configs),
    valueToText: _valueToText.default
  };
}

function createServerSerializationPlugin() {
  return {
    valueToJsonML: _valueToJsonML.default,
    jsonMLToValue: _jsonMLToValue.default
  };
}
