// @ali/4ever-bi/dist-web/plugins/autoClose/utils/coupleKey.js
import _extends from "@babel/runtime/helpers/extends";
import { createDict } from '@ali/4ever-utils';
var englishQuotes = createDict({
  '\'': true,
  '"': true,
  '`': true
});
var chineseQuotes = createDict({
  '“': true,
  '”': true,
  '‘': true,
  '’': true
});
var chineseLeftBrackets = createDict({
  '【': true,
  '（': true
});
var chineseRightBrackets = createDict({
  '】': true,
  '）': true
});
var englishLeftBrackets = createDict({
  '{': true,
  '[': true,
  '(': true
});
var englishRightBrackets = createDict({
  '}': true,
  ']': true,
  ')': true
});
var rightBrackets = createDict({
  '}': true,
  ']': true,
  ')': true,
  '】': true,
  '）': true
});
var coupleKey = createDict(_extends({}, englishLeftBrackets, englishRightBrackets, englishQuotes, chineseLeftBrackets, chineseRightBrackets, chineseQuotes));
var leftCoupleKey = createDict(_extends({}, englishLeftBrackets, chineseLeftBrackets, englishQuotes));
var coupleMap = createDict({
  '\'': '\'',
  '"': '"',
  '`': '`',
  '{': '}',
  '[': ']',
  '(': ')',
  '【': '】',
  '（': '）'
});
var couplePair = createDict({
  '\'\'': true,
  '""': true,
  '``': true,
  '{}': true,
  '[]': true,
  '()': true,
  '【】': true,
  '（）': true,
  '“”': true,
  '‘’': true
});
var chineseQuotePair = createDict({
  '‘': ['‘', '’'],
  '’': ['‘', '’'],
  '“': ['“', '”'],
  '”': ['“', '”']
});
export function isCoupleKey(key) {
  return coupleKey[key];
}
export function isLeftCoupleKey(key) {
  return leftCoupleKey[key];
}
export function isRightBracket(key) {
  return rightBrackets[key];
}
export function getRightCoupleKey(key) {
  return coupleMap[key];
}
export function isCouple(left, right) {
  return couplePair[left + right];
}
export function isEnglishQuoteKey(key) {
  return englishQuotes[key];
}
export function isChineseQuoteKey(key) {
  return chineseQuotes[key];
}
export function getChineseQuoteCP(key) {
  return chineseQuotePair[key];
}
