// @ali/we-design-next/dist-web/form/date-picker/panels/TimePanel/styled.js
import styled, { css } from 'styled-components';
import { PANEL_HEIGHT_WITH_NOW, PANEL_HEIGHT_WITHOUT_NOW, HEADER_HEIGHT } from "../DatePanel/styled";
import { COLORS, CSS_BORDER_BOX } from "../../../../common";
export var LI_HEIGHT = 28;
var UL_HEIGHT_WITH_NOW = PANEL_HEIGHT_WITH_NOW - HEADER_HEIGHT;
var UL_HEIGHT_WITHOUT_NOW = PANEL_HEIGHT_WITHOUT_NOW - HEADER_HEIGHT;
export var TimePanelWrapper = /*#__PURE__*/styled.div(["min-width:", ";height:", "px;.header-border{border-bottom:1px solid ", ";border-left:1px solid ", ";}"], function (_ref) {
  var showHeader = _ref.showHeader;
  return showHeader ? 'fit-content' : 'min-content';
}, function (_ref2) {
  var showNow = _ref2.showNow,
    showHeader = _ref2.showHeader;
  return showHeader ? showNow ? PANEL_HEIGHT_WITH_NOW : PANEL_HEIGHT_WITHOUT_NOW : UL_HEIGHT_WITH_NOW;
}, COLORS.OTHERS.LINE_LIGHT_COLOR, COLORS.OTHERS.LINE_LIGHT_COLOR);
export var ColumWrapper = /*#__PURE__*/styled.div(["display:flex;border-left:", ";"], function (_ref3) {
  var showHeader = _ref3.showHeader;
  return showHeader ? "1px solid " + COLORS.OTHERS.LINE_LIGHT_COLOR : 'unset';
});

// 定制版滚动条，滚动条高度是72而非60，因为纵向滚动条的滑块高度由内容决定，无法设置成60
// 以「时」的滑块高度72为准，保持一致
export var CSS_SCROLLBAR = /*#__PURE__*/css(["&::-webkit-scrollbar{width:8px;}&::-webkit-scrollbar-track{background:", ";}&::-webkit-scrollbar-thumb{background:rgba(199,200,201,1);border-radius:4px;height:72px;&:hover{background:rgba(162,163,165,1);}&:active{background:rgba(116,118,119,1);}}"], COLORS.TRANSPARENT);
export var UlWrapper = /*#__PURE__*/styled.ul(["", ";width:56px;height:", "px;overflow-y:overlay;", ";flex:1 0 auto;margin:0;padding:0;text-align:left;list-style:none;border-left:1px solid ", ";transition:background .3s;:first-child{border-left:unset;}"], CSS_BORDER_BOX, function (_ref4) {
  var showNow = _ref4.showNow;
  return showNow ? UL_HEIGHT_WITH_NOW : UL_HEIGHT_WITHOUT_NOW;
}, CSS_SCROLLBAR, COLORS.OTHERS.LINE_LIGHT_COLOR);
export var LiWrapper = /*#__PURE__*/styled.li(["", ";margin:0;padding:0;user-select:none;&:hover{background-color:#F2F4F5;cursor:default;}&.selected{background:", ";}"], CSS_BORDER_BOX, COLORS.COMMON.BLUR3_COLOR);
export var NumberWrapper = /*#__PURE__*/styled.div(["", ";display:block;width:100%;height:", "px;margin:0;text-align:center;font-weight:400;font-size:14px;line-height:", "px;border-radius:0;cursor:default;transition:background .3s;"], CSS_BORDER_BOX, LI_HEIGHT, LI_HEIGHT);
export var TitleLiWrapper = /*#__PURE__*/styled.li(["", ";margin:0;padding:0;position:sticky;top:0;background-color:", ";user-select:none;"], CSS_BORDER_BOX, COLORS.WHITE);
export var TitleWrapper = /*#__PURE__*/styled(NumberWrapper)(["font-size:12px;color:", ";"], COLORS.LIGHT.LEVEL3_BASE_COLOR);
