// @ali/we-design-next/dist-web/form/date-picker/panels/TimePanel/TimePanel.js
import React, { useCallback } from "react";
var _createElement = /*#__PURE__*/React.createElement;
import { useDayjsLocale, useTranslate } from "../../../../locale";
import { TimeColumn, generateUnits } from "./TimeColumn";
import { Header } from "../Header";
import { TimePanelWrapper, ColumWrapper } from "./styled";
import { DateController } from "../../date-util";
var hours = generateUnits(0, 23);
var minutes = generateUnits(0, 59);
var seconds = generateUnits(0, 59);
export var TimePanel = /*#__PURE__*/React.memo(function (props) {
  var value = props.value,
    _onSelect = props.onSelect,
    showNow = props.showNow,
    timeFormat = props.timeFormat,
    _props$showHeader = props.showHeader,
    showHeader = _props$showHeader === void 0 ? true : _props$showHeader;
  var dayjsLocale = useDayjsLocale();
  var t = useTranslate();
  var showSecond = timeFormat === 'hms';
  var format = showSecond ? 'HH:mm:ss' : 'HH:mm';
  var hour = value ? DateController.getHour(value) : -1;
  var minute = value ? DateController.getMinute(value) : -1;
  var second = value ? DateController.getSecond(value) : -1;
  var setTime = useCallback(function (newHour, newMinute, newSecond) {
    var curDate = value || DateController.getNow();
    var mergedHour = Math.max(0, newHour);
    var mergedMinute = Math.max(0, newMinute);
    var mergedSecond = Math.max(0, newSecond);
    var newDate = DateController.setHour(curDate.clone(), mergedHour);
    newDate = DateController.setMinute(newDate, mergedMinute);
    newDate = DateController.setSecond(newDate, mergedSecond);
    return newDate;
  }, [value]);
  return /*#__PURE__*/_createElement(TimePanelWrapper, {
    showNow: showNow,
    showHeader: showHeader
  }, showHeader ? /*#__PURE__*/_createElement(Header, {
    className: "header-border"
  }, value ? DateController.locale.format(dayjsLocale, value, format) : null) : null, /*#__PURE__*/_createElement(ColumWrapper, {
    showHeader: showHeader
  }, /*#__PURE__*/_createElement(TimeColumn, {
    value: hour,
    units: hours,
    onSelect: function onSelect(num) {
      return _onSelect(setTime(num, minute, second));
    },
    title: t('wdn_date_picker_hour'),
    showNow: showNow
  }), /*#__PURE__*/_createElement(TimeColumn, {
    value: minute,
    units: minutes,
    onSelect: function onSelect(num) {
      return _onSelect(setTime(hour, num, second));
    },
    title: t('wdn_date_picker_minute'),
    showNow: showNow
  }), showSecond ? /*#__PURE__*/_createElement(TimeColumn, {
    value: second,
    units: seconds,
    onSelect: function onSelect(num) {
      return _onSelect(setTime(hour, minute, num));
    },
    title: t('wdn_date_picker_second'),
    showNow: showNow
  }) : null));
});
