// @ali/4ever-plugin-marks/dist-node/mo/serializer/leaf.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createLeaf = void 0;

var _jsonMLAttrsToStyle = require("../../utils/jsonMLAttrsToStyle");

var _jsonMLAttrsToMarks = require("../../utils/jsonMLAttrsToMarks");

var _markToJsonMLAttrs = require("../../utils/markToJsonMLAttrs");

var _htmlToJsonMLAttrs = require("../../utils/htmlToJsonMLAttrs");

const createLeaf = configs => {
  /**
   *
   */
  const {
    marks: marksConfig = {}
  } = configs || {};
  const {
    allowedMarks
  } = marksConfig;
  const allowedMarksMap = allowedMarks?.reduce((acc, k) => {
    acc[k] = true;
    return acc;
  }, {});
  const leafRules = {
    jsonMLAttrsToStyle: (attrs, style) => {
      return (0, _jsonMLAttrsToStyle.jsonMLAttrsToStyle)(attrs, style);
    },
    htmlToJsonMLAttrs: (name, attrs, curMarks) => {
      let marks = (0, _htmlToJsonMLAttrs.htmlToJsonMLAttrs)(name, attrs, curMarks);

      if (allowedMarksMap) {
        marks = Object.keys(marks).reduce((acc, key) => {
          if (allowedMarksMap[key]) {
            acc[key] = marks[key];
          }

          return acc;
        }, {});
      }

      return marks;
    },
    jsonMLAttrsToMarks: attrs => {
      return (0, _jsonMLAttrsToMarks.jsonMLAttrsToMarks)(attrs);
    },
    markToJsonMLAttrs: mark => {
      return (0, _markToJsonMLAttrs.markToJsonMLAttrs)(mark);
    }
  };
  return leafRules;
};

exports.createLeaf = createLeaf;
