// @ali/4ever-plugin-table/dist-node/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  zhiTableCell: true,
  zhiTableRow: true,
  OldBambooPlugin: true,
  moFactory: true,
  zhiFactory: true,
  mobileZhiFactory: true,
  biFactory: true,
  mobileBiFactory: true,
  biActions: true,
  biActionTypes: true,
  Table: true,
  TableCell: true,
  TableRow: true,
  createSerializationPlugin: true,
  createServerSerializationPlugin: true,
  useContainerWidthResize: true,
  useRowIsSticky: true,
  RowIsStickyContext: true,
  isPureTableNode: true,
  isTableNode: true,
  getTableCellWidth: true,
  setSelectionByTable: true,
  getTable: true,
  getTableSelection: true,
  createEmptyParagraph: true,
  createEmptyTableCell: true,
  isSelectionInTableCell: true,
  setTableColsWidth: true,
  handleCangjiePaste: true,
  onCangjieInput: true,
  zhiTable: true,
  biTable: true,
  biTableRow: true,
  biTableCell: true,
  getAllTables: true,
  lineMap: true,
  DEFAULT_TABLE_BORDER_STYLE: true,
  DEFAULT_BORDER_COLOR: true,
  MIN_COL_WIDTH: true,
  TD: true
};
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.moFactory;
  }
});
Object.defineProperty(exports, "zhiFactory", {
  enumerable: true,
  get: function () {
    return _zhi.zhiFactory;
  }
});
Object.defineProperty(exports, "mobileZhiFactory", {
  enumerable: true,
  get: function () {
    return _zhi.mobileZhiFactory;
  }
});
Object.defineProperty(exports, "biFactory", {
  enumerable: true,
  get: function () {
    return _bi.biFactory;
  }
});
Object.defineProperty(exports, "mobileBiFactory", {
  enumerable: true,
  get: function () {
    return _bi.mobileBiFactory;
  }
});
Object.defineProperty(exports, "biActions", {
  enumerable: true,
  get: function () {
    return _actions.actions;
  }
});
Object.defineProperty(exports, "biActionTypes", {
  enumerable: true,
  get: function () {
    return _actions.actionTypes;
  }
});
Object.defineProperty(exports, "Table", {
  enumerable: true,
  get: function () {
    return _models.default;
  }
});
Object.defineProperty(exports, "TableCell", {
  enumerable: true,
  get: function () {
    return _tableCell2.default;
  }
});
Object.defineProperty(exports, "TableRow", {
  enumerable: true,
  get: function () {
    return _tableRow2.default;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.createServerSerializationPlugin;
  }
});
Object.defineProperty(exports, "useContainerWidthResize", {
  enumerable: true,
  get: function () {
    return _ContainerWidthResizeContext.useContainerWidthResize;
  }
});
Object.defineProperty(exports, "useRowIsSticky", {
  enumerable: true,
  get: function () {
    return _useRowIsSticky.useRowIsSticky;
  }
});
Object.defineProperty(exports, "RowIsStickyContext", {
  enumerable: true,
  get: function () {
    return _useRowIsSticky.RowIsStickyContext;
  }
});
Object.defineProperty(exports, "isPureTableNode", {
  enumerable: true,
  get: function () {
    return _isTableNode.isPureTableNode;
  }
});
Object.defineProperty(exports, "isTableNode", {
  enumerable: true,
  get: function () {
    return _isTableNode.isTableNode;
  }
});
Object.defineProperty(exports, "getTableCellWidth", {
  enumerable: true,
  get: function () {
    return _getTableCellWidth.default;
  }
});
Object.defineProperty(exports, "setSelectionByTable", {
  enumerable: true,
  get: function () {
    return _setSelectionByTable.default;
  }
});
Object.defineProperty(exports, "getTable", {
  enumerable: true,
  get: function () {
    return _getTable.default;
  }
});
Object.defineProperty(exports, "getTableSelection", {
  enumerable: true,
  get: function () {
    return _getTableSelection.default;
  }
});
Object.defineProperty(exports, "createEmptyParagraph", {
  enumerable: true,
  get: function () {
    return _createEmptyParagraph.default;
  }
});
Object.defineProperty(exports, "createEmptyTableCell", {
  enumerable: true,
  get: function () {
    return _createEmptyTableCell.default;
  }
});
Object.defineProperty(exports, "isSelectionInTableCell", {
  enumerable: true,
  get: function () {
    return _isSelectionInTableCell.default;
  }
});
Object.defineProperty(exports, "setTableColsWidth", {
  enumerable: true,
  get: function () {
    return _setTableColsWidth.default;
  }
});
Object.defineProperty(exports, "handleCangjiePaste", {
  enumerable: true,
  get: function () {
    return _handleCangjiePaste.handleCangjiePaste;
  }
});
Object.defineProperty(exports, "onCangjieInput", {
  enumerable: true,
  get: function () {
    return _onCangjieInput.default;
  }
});
Object.defineProperty(exports, "zhiTable", {
  enumerable: true,
  get: function () {
    return _table.default;
  }
});
Object.defineProperty(exports, "biTable", {
  enumerable: true,
  get: function () {
    return _table2.default;
  }
});
Object.defineProperty(exports, "biTableRow", {
  enumerable: true,
  get: function () {
    return _tableRow3.default;
  }
});
Object.defineProperty(exports, "biTableCell", {
  enumerable: true,
  get: function () {
    return _tableCell3.default;
  }
});
Object.defineProperty(exports, "getAllTables", {
  enumerable: true,
  get: function () {
    return _getSelectionContainers.getAllTables;
  }
});
Object.defineProperty(exports, "lineMap", {
  enumerable: true,
  get: function () {
    return _constants.lineMap;
  }
});
Object.defineProperty(exports, "DEFAULT_TABLE_BORDER_STYLE", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_TABLE_BORDER_STYLE;
  }
});
Object.defineProperty(exports, "DEFAULT_BORDER_COLOR", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_BORDER_COLOR;
  }
});
Object.defineProperty(exports, "MIN_COL_WIDTH", {
  enumerable: true,
  get: function () {
    return _constants.MIN_COL_WIDTH;
  }
});
Object.defineProperty(exports, "TD", {
  enumerable: true,
  get: function () {
    return _styled.TD;
  }
});
exports.OldBambooPlugin = exports.zhiTableRow = exports.zhiTableCell = void 0;

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;

var _index = require("./mo/index");

var _zhi = require("./zhi");

var _bi = require("./bi");

require("./bi/types");

require("./zhi/types");

require("./utils/types");

var _tableCell = _interopRequireDefault(require("./components/tableCell"));

var _tableRow = _interopRequireDefault(require("./components/tableRow"));

var _actions = require("./bi/actions");

var _models = _interopRequireDefault(require("./mo/models"));

var _tableCell2 = _interopRequireDefault(require("./mo/models/tableCell"));

var _tableRow2 = _interopRequireDefault(require("./mo/models/tableRow"));

var _index2 = _interopRequireWildcard(require("./mo/serializer/index"));

var _ContainerWidthResizeContext = require("./components/ContainerWidthResizeContext");

var _useRowIsSticky = require("./utils/hooks/useRowIsSticky");

var _isTableNode = require("./utils/utils/isTableNode");

var _getTableCellWidth = _interopRequireDefault(require("./utils/utils/getTableCellWidth"));

var _setSelectionByTable = _interopRequireDefault(require("./utils/utils/setSelectionByTable"));

var _getTable = _interopRequireDefault(require("./bi/utils/getTable"));

var _getTableSelection = _interopRequireDefault(require("./bi/utils/getTableSelection"));

var _createEmptyParagraph = _interopRequireDefault(require("./bi/utils/createEmptyParagraph"));

var _createEmptyTableCell = _interopRequireDefault(require("./bi/utils/createEmptyTableCell"));

var _isSelectionInTableCell = _interopRequireDefault(require("./bi/utils/isSelectionInTableCell"));

var _setTableColsWidth = _interopRequireDefault(require("./bi/utils/setTableColsWidth"));

var _handleCangjiePaste = require("./bi/utils/handleCangjiePaste");

var _onCangjieInput = _interopRequireDefault(require("./bi/handlers/onCangjieInput"));

var _table = _interopRequireDefault(require("./zhi/components/table"));

var _table2 = _interopRequireDefault(require("./bi/components/table"));

var _tableRow3 = _interopRequireDefault(require("./bi/components/tableRow"));

var _tableCell3 = _interopRequireDefault(require("./bi/components/tableCell"));

var _colsWidthConverter = require("./utils/utils/colsWidthConverter");

Object.keys(_colsWidthConverter).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _colsWidthConverter[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _colsWidthConverter[key];
    }
  });
});

var _getSelectionContainers = require("./common/queries/getSelectionContainers");

var _constants = require("./utils/constants");

var _styled = require("./components/styled");

// facotry
// Element Model
// serializer
// custom
const zhiTableCell = _tableCell.default;
exports.zhiTableCell = zhiTableCell;
const zhiTableRow = _tableRow.default;
exports.zhiTableRow = zhiTableRow;
