// @ali/4ever-plugin-marks/dist-node/utils/createMarkFactory/createRenderToolbar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = factory;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everComponent = require("@ali/4ever-component");

var _createMarkButton = _interopRequireDefault(require("./components/next/createMarkButton"));

const _createElement = /*#__PURE__*/React.createElement;

function factory(payload) {
  const {
    type,
    hotKey,
    delimiter,
    toolbarIcon: defaultIcon,
    localeKeys
  } = payload;
  const Component = (0, _createMarkButton.default)({
    type,
    hotKey,
    defaultIcon,
    localeKeys,
    delimiter
  });
  return function createRenderToolbar(locale) {
    return (props, controller) => {
      const {
        buttonName,
        extraProps
      } = props;

      if (buttonName === type) {
        return /*#__PURE__*/_createElement(_everComponent.ItemSlot, {
          display: _everComponent.SlotType.iconButton
        }, /*#__PURE__*/_createElement(Component, (0, _extends2.default)({
          controller: controller,
          locale: locale
        }, extraProps)));
      }

      return null;
    };
  };
}
