// @ali/4ever-plugin-marks/dist-node/utils/createMarkFactory/createContextMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createContextMenu;

var _lodash = require("lodash");

var _actions = require("./actions");

function createContextMenu(type, hotKey) {
  return function contextMenuPlugin(pluginConfig) {
    const locale = pluginConfig.locale || {};
    return function contextMenu(controller, next) {
      const menus = [];
      const {
        selection,
        document
      } = controller.value;
      let canToggle = false;

      try {
        // getFragmentAtRange throws unknow error
        // https://aone.alibaba-inc.com/v2/project/995011/bug/43531316
        const fragement = document.getFragmentAtRange(selection);
        canToggle = Boolean(fragement.text);
      } catch (e) {// ignore
      }

      if (canToggle && selection.isExpanded && controller.query(`canToggle${(0, _lodash.capitalize)(type)}`)) {
        menus.push({
          name: locale.buttonTitle,
          action: (0, _actions.toggleMark)(type),
          key: `mark-toggle-${type}`,
          options: {
            keepMenu: true,
            hotKey,
            group: 2,
            checked: controller.query(`is${(0, _lodash.capitalize)(type)}`)
          }
        });
      }

      const others = next() || [];
      return [...menus, ...others];
    };
  };
}
