// @ali/we-design-next/dist-node/form/date-picker/panels/DatePanel/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Th = exports.PanelWrapper = exports.PANEL_WIDTH = exports.PANEL_HEIGHT_WITH_NOW = exports.PANEL_HEIGHT_WITHOUT_NOW = exports.HEADER_HEIGHT = exports.BODY_HEIGHT = void 0;
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _common = require("../../../../common");
var _styled = require("../../styled");
const PANEL_WIDTH = 280;
exports.PANEL_WIDTH = PANEL_WIDTH;
const BODY_HEIGHT = 196;
exports.BODY_HEIGHT = BODY_HEIGHT;
const HEADER_HEIGHT = 52;
exports.HEADER_HEIGHT = HEADER_HEIGHT;
const PANEL_HEIGHT_WITH_NOW = HEADER_HEIGHT + BODY_HEIGHT + _styled.NOW_HEIGHT;
exports.PANEL_HEIGHT_WITH_NOW = PANEL_HEIGHT_WITH_NOW;
const PANEL_HEIGHT_WITHOUT_NOW = HEADER_HEIGHT + BODY_HEIGHT + _styled.SPACE_HEIGHT;
exports.PANEL_HEIGHT_WITHOUT_NOW = PANEL_HEIGHT_WITHOUT_NOW;
const PanelWrapper = /*#__PURE__*/_styledComponents.default.div(["min-width:", "px;height:", "px;.header-border{border-bottom:1px solid ", "}"], PANEL_WIDTH, ({
  showNow
}) => showNow ? PANEL_HEIGHT_WITH_NOW : PANEL_HEIGHT_WITHOUT_NOW, _common.COLORS.OTHERS.LINE_LIGHT_COLOR);
exports.PanelWrapper = PanelWrapper;
const Th = /*#__PURE__*/_styledComponents.default.th(["font-size:12px;height:30px;line-height:30px;font-weight:400;text-align:center;color:", ";"], _common.COLORS.LIGHT.LEVEL2_BASE_COLOR);
exports.Th = Th;
