// @ali/4ever-plugin-list/dist-web/bi/components/createListItem.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _extends from "@babel/runtime/helpers/extends";
import _createClass from "@babel/runtime/helpers/createClass";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { IndUtil, ShdUtil, theme } from '@ali/4ever-utils';
import { MoCommon } from '@ali/4ever-cangjie';
import { EXTRA_PADDING, utils as ListGroupPluginUtils } from '@ali/4ever-plugin-list-group';
import { setCheckStatus } from "../actions";
import { ListContext } from "../../component/context";
import { ROMAN_EXTRA_INDENT, SYMBOL_STYLE } from "../../utils/constant";
import getJustifyContent from "../../utils/getJustifyContent";
import { generateSymbol } from "../../utils/generateSymbol";
import { GroupListContainer, ListItemWrapper } from "../../utils/styled";
import { convertSymbolStyle } from "../../utils/getSymbolStyleByKey";
import { RenderStyledListItem } from "../../utils/renderStyledListItem";
var BorderStyleEnum = MoCommon.BorderStyleEnum;
var isListGroupAvailable = ListGroupPluginUtils.isListGroupAvailable,
    isNormalGroupNode = ListGroupPluginUtils.isNormalGroupNode,
    isInUnlistGroupContainer = ListGroupPluginUtils.isInUnlistGroupContainer;
var DEFAULT_BORDER_COLOR = '#000000';
var INDENT_SIZE = theme.indent.size;
var DEFAULT_SHD_STYLES = theme.DEFAULT_SHD_STYLES;
export default function createListItem(Paragraph, extraProps) {
  var _ref = extraProps || {},
      _ref$allowCustomCheck = _ref.allowCustomCheck,
      allowCustomCheck = _ref$allowCustomCheck === void 0 ? false : _ref$allowCustomCheck;

  var ListItem = /*#__PURE__*/function (_React$Component) {
    _inheritsLoose(ListItem, _React$Component);

    function ListItem() {
      var _this;

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;

      _this.handleChange = function (checked) {
        var _this$props = _this.props,
            controller = _this$props.controller,
            node = _this$props.node,
            onChange = _this$props.onChange;
        onChange && onChange(checked);
        controller.run('onAction', setCheckStatus(node, checked));
      };

      _this.getGroupListStyle = function () {
        var _this$props2 = _this.props,
            node = _this$props2.node,
            controller = _this$props2.controller;
        var shd = node.data.shd;

        if (!_this.isListGroupAvailable || !shd) {
          return {
            display: 'contents'
          };
        } //@ts-ignore


        var paddingLeft = DEFAULT_SHD_STYLES.paddingLeft + (_this.symbolAlign === 'right' ? ROMAN_EXTRA_INDENT : 0);
        var backgroundColor = isNormalGroupNode(node, controller) ? ShdUtil.transformShd(shd) : 'unset';
        var width = "calc(100% + " + ((DEFAULT_SHD_STYLES.paddingLeft || 0) + (DEFAULT_SHD_STYLES.paddingRight || 0)) + "px)";
        var justifyContent = getJustifyContent(_this.align);
        return _extends({}, DEFAULT_SHD_STYLES, {
          backgroundColor: backgroundColor,
          paddingLeft: paddingLeft,
          width: width,
          display: 'flex',
          justifyContent: justifyContent
        });
      };

      _this.onSelectSymbol = function () {
        var node = _this.props.node;
        var onSelectSymbol = _this.context.onSelectSymbol;
        onSelectSymbol && onSelectSymbol(_this.listConfig.listId, node.key);
      };

      return _this;
    }

    var _proto = ListItem.prototype;

    _proto.getStyle = function getStyle() {
      var _this$props3 = this.props,
          isSelected = _this$props3.isSelected,
          customStyle = _this$props3.customStyle,
          node = _this$props3.node,
          controller = _this$props3.controller,
          visible = _this$props3.visible;
      var _this$listConfig = this.listConfig,
          isTaskList = _this$listConfig.isTaskList,
          _this$listConfig$leve = _this$listConfig.level,
          level = _this$listConfig$leve === void 0 ? 0 : _this$listConfig$leve,
          listStyle = _this$listConfig.listStyle;
      var _ref2 = this.props.node.data,
          pi = _ref2.pi,
          _ref2$bdr = _ref2.bdr,
          bdr = _ref2$bdr === void 0 ? {} : _ref2$bdr;
      var _node$data = node.data,
          shd = _node$data.shd,
          ind = _node$data.ind;
      var paddingSpace = level * INDENT_SIZE;

      if (ind) {
        var _IndUtil$fromBlock = IndUtil.fromBlock(node),
            _IndUtil$fromBlock$fi = _IndUtil$fromBlock.firstLine,
            firstLine = _IndUtil$fromBlock$fi === void 0 ? 0 : _IndUtil$fromBlock$fi,
            _IndUtil$fromBlock$le = _IndUtil$fromBlock.left,
            left = _IndUtil$fromBlock$le === void 0 ? 0 : _IndUtil$fromBlock$le;

        paddingSpace += firstLine + left;
      } // 罗马数字统一右侧对齐


      if (listStyle && ['upperRoman', 'lowerRoman'].includes(listStyle.format)) {
        // 根据字体大小，动态计算罗马数字预留空间的宽度
        paddingSpace += this.isListGroupAvailable && shd ? 0 : ROMAN_EXTRA_INDENT;
      }

      var isTaskSelected = isTaskList && isSelected && !allowCustomCheck;

      var styleObj = _extends({
        paddingLeft: paddingSpace,
        backgroundColor: isTaskSelected && node.data.shd === undefined ? '#f4f4f4' : '',
        borderRadius: isTaskSelected ? '3px' : ''
      }, customStyle); // Pi-2.0：列表高度有时会被项目符号莫名撑高，且边距需要渲染在列表处才可以生效


      if (pi && pi.box) {
        var _pi$box = pi.box,
            height = _pi$box.height,
            _pi$box$margin = _pi$box.margin,
            margin = _pi$box$margin === void 0 ? {} : _pi$box$margin;
        var _margin$top = margin.top,
            top = _margin$top === void 0 ? 0 : _margin$top,
            _margin$bottom = margin.bottom,
            bottom = _margin$bottom === void 0 ? 0 : _margin$bottom;
        styleObj.height = height;
        styleObj.marginTop = top;
        styleObj.marginBottom = bottom;
      } // 列表的边框需要挂载外围的 DOM 上


      if (bdr.top) {
        var _bdr$top = bdr.top,
            _bdr$top$val = _bdr$top.val,
            val = _bdr$top$val === void 0 ? BorderStyleEnum.Single : _bdr$top$val,
            _bdr$top$sz = _bdr$top.sz,
            sz = _bdr$top$sz === void 0 ? 1 : _bdr$top$sz,
            _bdr$top$color = _bdr$top.color,
            bdrColor = _bdr$top$color === void 0 ? DEFAULT_BORDER_COLOR : _bdr$top$color,
            _bdr$top$space = _bdr$top.space,
            space = _bdr$top$space === void 0 ? 0 : _bdr$top$space;
        var borderStyle = val === BorderStyleEnum.Single ? 'solid' : val;
        var borderColor = bdrColor === 'auto' ? DEFAULT_BORDER_COLOR : bdrColor;
        styleObj.borderTop = sz + "px " + borderStyle + " " + borderColor;
        styleObj.paddingTop = space;
      }

      if (bdr.bottom) {
        var _bdr$bottom = bdr.bottom,
            _bdr$bottom$val = _bdr$bottom.val,
            _val = _bdr$bottom$val === void 0 ? BorderStyleEnum.Single : _bdr$bottom$val,
            _bdr$bottom$sz = _bdr$bottom.sz,
            _sz = _bdr$bottom$sz === void 0 ? 1 : _bdr$bottom$sz,
            _bdr$bottom$color = _bdr$bottom.color,
            _bdrColor = _bdr$bottom$color === void 0 ? DEFAULT_BORDER_COLOR : _bdr$bottom$color,
            _bdr$bottom$space = _bdr$bottom.space,
            _space = _bdr$bottom$space === void 0 ? 0 : _bdr$bottom$space;

        var _borderStyle = _val === BorderStyleEnum.Single ? 'solid' : _val;

        var _borderColor = _bdrColor === 'auto' ? DEFAULT_BORDER_COLOR : _bdrColor;

        styleObj.borderBottom = _sz + "px " + _borderStyle + " " + _borderColor;
        styleObj.paddingBottom = _space;
      }

      if (bdr.left) {
        var _bdr$left = bdr.left,
            _bdr$left$val = _bdr$left.val,
            _val2 = _bdr$left$val === void 0 ? BorderStyleEnum.Single : _bdr$left$val,
            _bdr$left$sz = _bdr$left.sz,
            _sz2 = _bdr$left$sz === void 0 ? 1 : _bdr$left$sz,
            _bdr$left$color = _bdr$left.color,
            _bdrColor2 = _bdr$left$color === void 0 ? DEFAULT_BORDER_COLOR : _bdr$left$color,
            _bdr$left$space = _bdr$left.space,
            _space2 = _bdr$left$space === void 0 ? 0 : _bdr$left$space;

        var _borderStyle2 = _val2 === BorderStyleEnum.Single ? 'solid' : _val2;

        var _borderColor2 = _bdrColor2 === 'auto' ? DEFAULT_BORDER_COLOR : _bdrColor2;

        styleObj.borderLeft = _sz2 + "px " + _borderStyle2 + " " + _borderColor2;
        styleObj.paddingLeft = _space2 + paddingSpace;
      }

      if (bdr.right) {
        var _bdr$right = bdr.right,
            _bdr$right$val = _bdr$right.val,
            _val3 = _bdr$right$val === void 0 ? BorderStyleEnum.Single : _bdr$right$val,
            _bdr$right$sz = _bdr$right.sz,
            _sz3 = _bdr$right$sz === void 0 ? 1 : _bdr$right$sz,
            _bdr$right$color = _bdr$right.color,
            _bdrColor3 = _bdr$right$color === void 0 ? DEFAULT_BORDER_COLOR : _bdr$right$color,
            _bdr$right$space = _bdr$right.space,
            _space3 = _bdr$right$space === void 0 ? 0 : _bdr$right$space;

        var _borderStyle3 = _val3 === BorderStyleEnum.Single ? 'solid' : _val3;

        var _borderColor3 = _bdrColor3 === 'auto' ? DEFAULT_BORDER_COLOR : _bdrColor3;

        styleObj.borderRight = _sz3 + "px " + _borderStyle3 + " " + _borderColor3;
        styleObj.paddingRight = _space3;
      }

      if (isInUnlistGroupContainer(node, controller)) {
        styleObj.paddingLeft = (Number(styleObj.paddingLeft) || 0) + EXTRA_PADDING;
      }

      if (shd && !this.isListGroupAvailable) {
        for (var key in DEFAULT_SHD_STYLES) {
          if (typeof styleObj[key] === 'number') {
            styleObj[key] += DEFAULT_SHD_STYLES[key];
          } else {
            styleObj[key] = DEFAULT_SHD_STYLES[key];
          }
        }

        styleObj.backgroundColor = ShdUtil.transformShd(shd);
      }

      if (visible === false) {
        styleObj.display = 'none';
      }

      return styleObj;
    };

    _proto.renderParagraph = function renderParagraph() {
      var _this$props4 = this.props,
          node = _this$props4.node,
          children = _this$props4.children,
          isSelected = _this$props4.isSelected,
          customStyle = _this$props4.customStyle,
          nodeRef = _this$props4.nodeRef,
          rest = _objectWithoutPropertiesLoose(_this$props4, ["node", "children", "isSelected", "customStyle", "nodeRef"]);

      return /*#__PURE__*/_createElement(Paragraph, _extends({}, rest, {
        ref: nodeRef,
        node: node,
        isList: this.isList,
        isSelected: isSelected
      }), children);
    };

    _proto.renderListItem = function renderListItem() {
      var _this$props5 = this.props,
          node = _this$props5.node,
          className = _this$props5.className,
          contentStyle = _this$props5.contentStyle,
          listItemRef = _this$props5.listItemRef;
      var _node$data2 = node.data,
          step = _node$data2.step,
          _node$data2$start = _node$data2.start,
          start = _node$data2$start === void 0 ? 1 : _node$data2$start;
      var _this$listConfig2 = this.listConfig,
          listId = _this$listConfig2.listId,
          isOrdered = _this$listConfig2.isOrdered,
          isTaskList = _this$listConfig2.isTaskList,
          isChecked = _this$listConfig2.isChecked,
          _this$listConfig2$lev = _this$listConfig2.level,
          level = _this$listConfig2$lev === void 0 ? 0 : _this$listConfig2$lev,
          listStyle = _this$listConfig2.listStyle,
          listStyleType = _this$listConfig2.listStyleType,
          hideSymbol = _this$listConfig2.hideSymbol,
          _this$listConfig2$sym = _this$listConfig2.symbolStyle,
          symbolStyle = _this$listConfig2$sym === void 0 ? SYMBOL_STYLE : _this$listConfig2$sym;
      var listSymbol = isTaskList ? '' : generateSymbol(level, Number(start), step, isOrdered, listStyleType, listStyle);
      var taskListConfig = {
        isTaskList: isTaskList,
        isChecked: isChecked,
        isOrdered: isOrdered,
        handleChange: this.handleChange
      };
      return /*#__PURE__*/_createElement(ListItemWrapper, {
        align: this.align,
        className: className,
        ref: listItemRef,
        style: this.getStyle(),
        contentStyle: contentStyle,
        "data-testid": "list",
        "data-start": start,
        "data-listid": listId,
        "data-level": level,
        "data-isordered": isOrdered,
        "data-istasklist": isTaskList,
        "data-ischecked": isChecked
      }, /*#__PURE__*/_createElement(GroupListContainer, {
        style: this.getGroupListStyle()
      }, /*#__PURE__*/_createElement(RenderStyledListItem, {
        hideSymbol: hideSymbol,
        listSymbol: listSymbol,
        symbolStyle: convertSymbolStyle(symbolStyle, isOrdered),
        isSelect: this.isSymbolSelected,
        symbolAlign: this.symbolAlign,
        onSelectSymbol: this.onSelectSymbol,
        taskListConfig: taskListConfig
      }, this.renderParagraph())));
    };

    _proto.render = function render() {
      return this.isList ? this.renderListItem() : this.renderParagraph();
    };

    _createClass(ListItem, [{
      key: "align",
      get: function get() {
        var node = this.props.node;
        return node.data.jc || 'left';
      }
    }, {
      key: "isList",
      get: function get() {
        var node = this.props.node;
        var listConfig = node.data.list;
        return !!listConfig;
      }
    }, {
      key: "symbolAlign",
      get: function get() {
        var listStyle = this.listConfig.listStyle;

        if (listStyle && ['upperRoman', 'lowerRoman'].includes(listStyle.format)) {
          return 'right';
        } else {
          return 'left';
        }
      }
    }, {
      key: "listConfig",
      get: function get() {
        return this.isList ? this.props.node.data.list : null;
      }
    }, {
      key: "isSymbolSelected",
      get: function get() {
        var node = this.props.node;
        return node.data.isSelected;
      }
    }, {
      key: "isListGroupAvailable",
      get: function get() {
        var _this$props6 = this.props,
            node = _this$props6.node,
            controller = _this$props6.controller;
        return isListGroupAvailable(controller) && !node.data.blockquote;
      }
    }]);

    return ListItem;
  }(React.Component);

  ListItem.contextType = ListContext;
  ;
  return /*#__PURE__*/React.forwardRef(function (props, ref) {
    return /*#__PURE__*/_createElement(ListItem, _extends({}, props, {
      nodeRef: ref
    }));
  });
}
