// @ali/4ever-bi/dist-web/plugins/underline/components/underlineMenu.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { MenuItem } from '@ali/we-design';
import { PluginRoles } from '@ali/4ever-bamboo';
;
import { toggleUnderline } from "../actions";
export default function UnderlineMenu(props) {
  var menuItem = props.menuItem,
      controller = props.controller;
  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;

  var title = menuItem.title,
      rest = _objectWithoutPropertiesLoose(menuItem, ["title"]);

  var isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInListSymbol');
  var isActive = controller.query('isUnderline');

  var handleClick = function handleClick() {
    return controller.run('onAction', toggleUnderline());
  };

  return /*#__PURE__*/_createElement(MenuItem, _extends({}, rest, {
    onClick: handleClick,
    disabled: isDisabled,
    selected: isActive,
    role: PluginRoles.underline
  }), title);
}
