// @ali/4ever-bi/dist-web/plugins/underline/components/underlineButtonMobile.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { toggleUnderline } from "../actions";

var isDisabled = function isDisabled(controller) {
  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var disabled = isCollapsedAtEdge;
  return Boolean(disabled);
};

export function MobileUnderlineButton(props) {
  var controller = props.controller,
      item = props.item,
      Wrapper = props.Wrapper;
  var isActive = controller.query('isUnderline');
  var handleClick = React.useCallback(function () {
    return controller.run('onAction', toggleUnderline());
  }, [controller]);
  return /*#__PURE__*/_createElement(Wrapper, {
    item: item,
    type: "horiz",
    icon: "underline",
    isActive: isActive,
    isDisabled: isDisabled(controller),
    onClick: handleClick
  });
}
export function MobileUnderlineBigButton(props) {
  var controller = props.controller,
      item = props.item,
      Wrapper = props.Wrapper;
  var isActive = controller.query('isUnderline');
  var handleClick = React.useCallback(function () {
    return controller.run('onAction', toggleUnderline());
  }, [controller]);
  return /*#__PURE__*/_createElement(Wrapper, {
    item: item,
    type: "big",
    icon: "underline",
    isFocused: isActive,
    isDisabled: isDisabled(controller),
    onClick: handleClick
  });
}
