// @ali/4ever-plugin-quickinsert/dist-node/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "QUICK_INSERT_INSERT", {
  enumerable: true,
  get: function () {
    return _actions.QUICK_INSERT_INSERT;
  }
});
Object.defineProperty(exports, "QUICK_INSERT_SHOW", {
  enumerable: true,
  get: function () {
    return _actions.QUICK_INSERT_SHOW;
  }
});
Object.defineProperty(exports, "QUICK_INSERT_SET", {
  enumerable: true,
  get: function () {
    return _actions.QUICK_INSERT_SET;
  }
});
Object.defineProperty(exports, "QUICK_INSERT_RESET", {
  enumerable: true,
  get: function () {
    return _actions.QUICK_INSERT_RESET;
  }
});
Object.defineProperty(exports, "QUICK_INSERT_REMOVE", {
  enumerable: true,
  get: function () {
    return _actions.QUICK_INSERT_REMOVE;
  }
});
Object.defineProperty(exports, "insertQuickInsert", {
  enumerable: true,
  get: function () {
    return _actions.insertQuickInsert;
  }
});
Object.defineProperty(exports, "showQuickInsert", {
  enumerable: true,
  get: function () {
    return _actions.showQuickInsert;
  }
});
Object.defineProperty(exports, "setQuickInsert", {
  enumerable: true,
  get: function () {
    return _actions.setQuickInsert;
  }
});
Object.defineProperty(exports, "resetQuickInsert", {
  enumerable: true,
  get: function () {
    return _actions.resetQuickInsert;
  }
});
Object.defineProperty(exports, "removeQuickInsert", {
  enumerable: true,
  get: function () {
    return _actions.removeQuickInsert;
  }
});

var _actions = require("./bi/actions");
