// @ali/we-ui/dist-web/components/menu/styled.js
import styled, { css } from 'styled-components';
import { Dropdown } from "../dropdown";
import { withTheme } from "../theme";
var hoverActive = /*#__PURE__*/css(["&:hover{background-color:", ";cursor:pointer;}"], function (_ref) {
  var theme = _ref.theme;
  return theme.ui.backgroundColor;
});
var handleItemCommon = /*#__PURE__*/css(["position:relative;min-width:150px;height:36px;font-size:13px;line-height:20px;padding:0 ", ";margin:0 ", ";color:", ";border-radius:", ";", " .icon{position:absolute;top:50%;right:23px;transform:translate(120%,-50%);}&.disabled{color:", ";&:hover{background-color:transparent;cursor:not-allowed;}}"], function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.paddingMedium;
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.marginMedium;
}, function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.commonBaseColor1;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.ui.radiusLarge;
}, function (_ref6) {
  var isHotkeyType = _ref6.isHotkeyType,
      hoverable = _ref6.hoverable,
      theme = _ref6.theme;
  return hoverable !== false && "\n    &:hover {\n      background-color: " + (isHotkeyType ? 'inherit' : theme.ui.commonBgHoverColor) + ";\n      cursor: pointer;\n    }\n  ";
}, function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.disabledColor;
});
export var MenuRoot = withTheme(styled.div(["box-sizing:border-box;font-size:13px;color:", ";&.vertical{min-width:150px;padding:8px 0;}&.horizontal{display:flex;}"], function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.commonActionIconColor;
}));
export var Divider = withTheme(styled.div(["height:1px;margin:", " 8px;background-color:", ";"], function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.marginMedium;
}, function (_ref10) {
  var theme = _ref10.theme;
  return theme.ui.commonLineLightColor;
}));
export var SubMenuWrap = withTheme(styled.div.attrs(function (_ref11) {
  var path = _ref11.path,
      menuId = _ref11.menuId;
  return {
    'data-path': path,
    'data-menuId': menuId
  };
})([""]));
export var SubMenuInnerWrap = withTheme(styled.div(["box-sizing:border-box;", " &.horizontal{height:28px;vertical-align:middle;display:inline-block;padding:", ";margin:0 ", ";border-radius:", ";}&.hotkeyHover{background-color:", " !important;}&.vertical{", "}&.opened{background-color:", ";}.arrow-icon{color:", ";&.disabled{color:", ";}}"], hoverActive, function (_ref12) {
  var theme = _ref12.theme;
  return theme.ui.paddingTiny;
}, function (_ref13) {
  var theme = _ref13.theme;
  return theme.ui.marginTiny;
}, function (_ref14) {
  var theme = _ref14.theme;
  return theme.ui.radiusMedium;
}, function (_ref15) {
  var theme = _ref15.theme;
  return theme.ui.commonBgHoverColor;
}, handleItemCommon, function (_ref16) {
  var theme = _ref16.theme;
  return theme.ui.backgroundColor;
}, function (_ref17) {
  var theme = _ref17.theme;
  return theme.ui.commonBaseColor3;
}, function (_ref18) {
  var theme = _ref18.theme;
  return theme.ui.disabledColor;
}));
export var TriggerWrap = withTheme(styled.div(["display:flex;align-items:center;height:100%;&.hasArrow{padding-right:24px;}"]));
export var TriggerIconWrap = withTheme(styled.div(["display:flex;align-items:center;height:100%;margin-right:8px;color:", ";"], function (_ref19) {
  var theme = _ref19.theme;
  return theme.ui.commonActionIconColor;
}));
export var MenuItemInner = withTheme(styled.div(["display:flex;align-items:center;&.selected{padding-right:35px;}"]));
export var MenuItemIconWrap = withTheme(styled.div(["display:flex;align-items:center;height:100%;margin-right:8px;color:", ";"], function (_ref20) {
  var theme = _ref20.theme;
  return theme.ui.commonActionIconColor;
}));
export var MenuItemContentWrap = withTheme(styled.div([""]));
export var TriggerTitleWrap = withTheme(styled.div([""]));
export var MenuItemWrap = withTheme(styled.div.attrs(function (_ref21) {
  var path = _ref21.path,
      menuId = _ref21.menuId;
  return {
    'data-path': path,
    'data-menuId': menuId
  };
})([".selected-icon{display:none;color:", ";font-size:12px;}&.hotkeyHover{background-color:", " !important;}&.vertical{", ";display:flex;align-items:center;&.selected{.selected-icon{display:inline-block;}}}&.horizontal{", " display:inline-block;padding:", ";margin:0 ", ";border-radius:", ";}"], function (_ref22) {
  var theme = _ref22.theme;
  return theme.ui.primaryColor;
}, function (_ref23) {
  var theme = _ref23.theme;
  return theme.ui.commonBgHoverColor;
}, handleItemCommon, hoverActive, function (_ref24) {
  var theme = _ref24.theme;
  return theme.ui.paddingTiny;
}, function (_ref25) {
  var theme = _ref25.theme;
  return theme.ui.marginTiny;
}, function (_ref26) {
  var theme = _ref26.theme;
  return theme.ui.radiusMedium;
}));
export var MenuItemGroupWrap = /*#__PURE__*/styled.div([""]);
export var MenuItemGroupTitle = withTheme(styled.div(["font-size:", ";color:", ";line-height:2;padding-left:", ";"], function (_ref27) {
  var theme = _ref27.theme;
  return theme.ui.fontSizeSmall;
}, function (_ref28) {
  var theme = _ref28.theme;
  return theme.ui.commonBaseColor3;
}, function (_ref29) {
  var theme = _ref29.theme;
  return theme.ui.paddingMedium;
}));
export var MenuItemGroupList = /*#__PURE__*/styled.div([""]);
export var SubMenuDropdown = /*#__PURE__*/styled(Dropdown)(["width:100%;"]);
