// @ali/4ever-bi/dist-node/plugins/callout/getSnippetConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everUtils = require("@ali/4ever-utils");

var _actions = require("./actions");

var _everPluginCallout = require("@ali/4ever-plugin-callout");

const getSnippetConfig = (_, next) => {
  const trigger = '[:：]{3}\\w*$';
  const list = Object.keys(_everPluginCallout.COLOR_MAP);
  const config = {
    trigger,
    triggerPrefix: '^$',
    triggerSuffix: '^$',
    fetchItems: async () => {
      return list.map(key => ({
        key,
        title: key
      }));
    },
    filterItems: async (text, items) => {
      const searchText = text.slice(3);

      if (!searchText) {
        return items;
      }

      return (0, _everUtils.fuzzySearch)(items, searchText, {
        key: 'key'
      });
    },
    insertSnippet: (controller, item) => {
      controller.run('onAction', (0, _actions.insertCallout)({
        text: '',
        type: item.key
      }));
      controller.dispatch('createTriggerMarkdownAction', {
        mdType: 'colorBlocks',
        info: item.key,
        from: 'snippet'
      });
    }
  };
  return {
    [trigger]: config,
    ...next()
  };
};

var _default = getSnippetConfig;
exports.default = _default;
