// @ali/4ever-bi/dist-node/plugins/callout/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everBamboo = require("@ali/4ever-bamboo");

var _createRenderNode = _interopRequireDefault(require("./handlers/createRenderNode"));

var commands = _interopRequireWildcard(require("./commands"));

var _createOnAction = _interopRequireDefault(require("./handlers/createOnAction"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _createGroupNodes = _interopRequireDefault(require("./handlers/createGroupNodes"));

var _createOnQuery = _interopRequireDefault(require("./handlers/createOnQuery"));

var _createOnCangjieSelect = _interopRequireDefault(require("./handlers/createOnCangjieSelect"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _onCangjieCopyOrCut = _interopRequireDefault(require("./handlers/onCangjieCopyOrCut"));

var _onCangjieInput = _interopRequireDefault(require("./handlers/onCangjieInput"));

var _queries = _interopRequireDefault(require("./queries"));

var _createRenderHoverToolbar = _interopRequireDefault(require("./createRenderHoverToolbar"));

var _getSnippetConfig = _interopRequireDefault(require("./getSnippetConfig"));

var actions = _interopRequireWildcard(require("./actions"));

var _everPluginColorBlocks = require("@ali/4ever-plugin-color-blocks");

const {
  DEFAULT_STICKERS,
  Sticker
} = _everBamboo.StickerPlugin;

function createCalloutPlugin(config) {
  const {
    callout = {}
  } = config;
  const {
    locale
  } = callout;
  const stickerConfig = {
    // @ts-ignore
    lang: 'zh_CN',
    // @ts-ignore
    customSticker: [DEFAULT_STICKERS],
    ...config.sticker
  };
  const sticker = new Sticker(stickerConfig);
  const params = {
    stickerConfig,
    sticker,
    locale: callout.locale
  };
  return { ...(0, _everPluginColorBlocks.moFactory)({}),
    renderNode: (0, _createRenderNode.default)(params),
    // Copy Cut 拦截作用于高亮块全选时
    onCangjieCopy: _onCangjieCopyOrCut.default,
    onCangjieCut: _onCangjieCopyOrCut.default,
    onCangjieInput: _onCangjieInput.default,
    onCangjieSelect: (0, _createOnCangjieSelect.default)(),
    onQuery: (0, _createOnQuery.default)(),
    groupNodes: (0, _createGroupNodes.default)(callout),
    onAction: (0, _createOnAction.default)(callout),
    getNodeStyle: _everBamboo.CalloutPlugin.getNodeStyle,
    // 作用于高亮块全选态变化时重渲
    // controller.value.data 变化可能不会引发重渲
    shouldNodeComponentUpdate: (prevProps, props, __, next) => {
      if (props.value.data.selectedCallout !== prevProps.value.data.selectedCallout && props.node.data?.calloutPr && (props.node.key === props.value.data.selectedCallout || props.node.key === prevProps.value.data.selectedCallout)) {
        return true;
      }

      return next();
    },
    // @ts-ignore 这里是 createCommands
    commands,
    menubar: _menubar.default,
    onKeyDown: _onKeyDown.default,
    queries: _queries.default,
    placeholder: {
      text: locale?.placeholder,
      when: (controller, node) => {
        return (0, _everBamboo.isEmptyParagraph)(node) && controller.query('getCalloutStartBlock', node) === node && controller.query('getCalloutEndBlock', node) === node;
      }
    },
    renderHoverToolbar: (0, _createRenderHoverToolbar.default)(callout),
    getSnippetConfig: _getSnippetConfig.default
  };
}

createCalloutPlugin.actions = actions;
var _default = createCalloutPlugin;
exports.default = _default;
