// @ali/we-design/dist-web/components/quick-insert/styled.js
import styled, { keyframes } from 'styled-components';
import { theme } from "../../components/theme";
export var HeaderWrapper = /*#__PURE__*/styled.div(["height:30px;margin-top:2px;padding:2px 0px 2px 10px;display:flex;align-items:center;justify-content:space-between;"]);
export var HeaderTitle = /*#__PURE__*/styled.span(["font-size:12px;width:fit-content;color:", ";display:flex;align-items:center;justify-content:center;opacity:", ";transition:opacity 600ms ease-in-out;"], theme.colorText4, function (p) {
  return p.isShowTitle ? 1 : 0;
});
export var HeaderButton = /*#__PURE__*/styled.span(["right:0;padding:0px 4px;font-size:12px;height:27px;margin-right:13px;width:fit-content;display:flex;border-radius:4px;align-items:center;justify-content:center;cursor:pointer;&:hover{background-color:", ";}&:active{background-color:", ";}"], theme.splitButtonHoverBgColor, theme.splitButtonActiveBgColor);
export var IconWrapper = /*#__PURE__*/styled.span(["color:rgba(17,31,44,0.56);display:flex;align-items:center;justify-content:center;"]);
export var ButtonText = /*#__PURE__*/styled.div(["padding-left:2px;"]);
export var QuickInsertWrapper = /*#__PURE__*/styled.div(["min-height:", ";height:", ";width:", ";display:flex;flex-direction:column;"], function (_ref) {
  var _ref$height = _ref.height,
      height = _ref$height === void 0 ? 347 : _ref$height;
  return height + "px";
}, function (_ref2) {
  var _ref2$height = _ref2.height,
      height = _ref2$height === void 0 ? 352 : _ref2$height;
  return height + "px";
}, function (_ref3) {
  var _ref3$width = _ref3.width,
      width = _ref3$width === void 0 ? 300 : _ref3$width;
  return width + "px";
});
export var ListWrapper = /*#__PURE__*/styled.div(["display:flex;flex-direction:column;flex:1;overflow-x:hidden;"]);
export var LoadingWrapper = /*#__PURE__*/styled.div(["display:flex;justify-content:center;align-items:center;height:30px;color:", ";"], theme.colorText2);
export var LoadingText = /*#__PURE__*/styled.div(["padding-left:4px;"]);
export var ItemWrapper = /*#__PURE__*/styled.div(["box-sizing:border-box;height:52px;min-height:52px;display:flex;flex-direction:row;align-items:center;padding-left:10px;background:", ";cursor:pointer;"], function (p) {
  return p.active ? theme.colorLine1 : 'transparent';
});
export var ItemContent = /*#__PURE__*/styled.div(["display:flex;flex-direction:column;padding:4px 6px;flex:1;"]);
export var ItemTitle = /*#__PURE__*/styled.div(["font-size:14px;color:", ";white-space:nowrap;overflow:hidden;text-overflow:ellipsis;max-width:250px;"], theme.colorText5);
export var ItemDescription = /*#__PURE__*/styled.span(["font-size:12px;color:", ";display:flex;flex-direction:row;"], theme.colorText3);
export var Description = /*#__PURE__*/styled.div(["font-size:12px;color:", ";display:flex;flex-direction:row;justify-content:space-between;align-items:center;width:264px;overflow:hidden;"], theme.colorText3);
export var DescriptionLeft = /*#__PURE__*/styled.span(["white-space:nowrap;max-width:180px;flex:1;overflow:hidden;text-overflow:ellipsis;"]);
export var DescriptionRight = /*#__PURE__*/styled.span(["white-space:nowrap;overflow:hidden;text-overflow:ellipsis;margin-right:10px;margin-left:2px;"]);
export var EllipsisText = /*#__PURE__*/styled.span(["flex:1;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;"]);
export var GroupTitle = /*#__PURE__*/styled.div(["box-sizing:border-box;height:30px;min-height:30px;padding-top:10px;padding-left:10px;font-size:12px;color:", ";"], theme.colorText4);
export var EmptyWrapper = /*#__PURE__*/styled.div(["display:flex;flex-direction:column;justify-content:center;align-items:center;flex:1;"]);
export var EmptyText = /*#__PURE__*/styled.div(["color:", ";width:230px;margin-top:20px;text-align:center;"], theme.colorText2);
export var ImageWrapper = /*#__PURE__*/styled.img(["width:100px;height:100px;"]);
export var MiddleEllipsisWrapper = /*#__PURE__*/styled.div(["display:flex;flex-direction:row;width:", "px;white-space:nowrap;"], function (p) {
  return p.width;
});
export var AvatarWrapper = /*#__PURE__*/styled.div(["height:32px;.avatar{width:32px;height:32px;}.avatar > div{border-radius:6px;border:0;}.avatar > div > div{border-radius:6px;}"]);
var loadingCircle = /*#__PURE__*/keyframes(["0%{background-position:100% 50%}to{background-position:0 50%}"]);
export var Skeleton = /*#__PURE__*/styled.div(["background:rgba(17,31,44,0.08);width:", ";height:", ";border-radius:", ";margin-top:", ";background:linear-gradient( 90deg,hsla(0,0%,74.5%,0.2) 25%,hsla(0,0%,50.6%,0.24) 37%,hsla(0,0%,74.5%,0.2) 63% );background-size:400% 100%;animation:", " 1.4s ease infinite;"], function (props) {
  return props.width;
}, function (props) {
  return props.height;
}, function (props) {
  return props.borderRadius || '9px';
}, function (props) {
  return props.marginTop || '0';
}, loadingCircle);
