// @ali/we-design/dist-web/components/quick-insert/type.js
// 插入类型
export var EInsertType;

(function (EInsertType) {
  EInsertType[EInsertType["Person"] = 0] = "Person";
  EInsertType[EInsertType["Group"] = 1] = "Group";
  EInsertType[EInsertType["Document"] = 2] = "Document";
  EInsertType[EInsertType["Folder"] = 3] = "Folder";
})(EInsertType || (EInsertType = {}));

export var FileTypeKey;

(function (FileTypeKey) {
  FileTypeKey[FileTypeKey["OFFICE_WORD"] = 0] = "OFFICE_WORD";
  FileTypeKey[FileTypeKey["OFFICE_EXCEL"] = 1] = "OFFICE_EXCEL";
  FileTypeKey[FileTypeKey["OFFICE_POWERPOINT"] = 2] = "OFFICE_POWERPOINT";
  FileTypeKey[FileTypeKey["DOC"] = 3] = "DOC";
  FileTypeKey[FileTypeKey["SHEET"] = 4] = "SHEET";
  FileTypeKey[FileTypeKey["WHITEBOARD"] = 5] = "WHITEBOARD";
  FileTypeKey[FileTypeKey["PRESENTATION"] = 6] = "PRESENTATION";
  FileTypeKey[FileTypeKey["MIND"] = 7] = "MIND";
  FileTypeKey[FileTypeKey["FOLDER"] = 8] = "FOLDER";
})(FileTypeKey || (FileTypeKey = {}));

export var fileTypeKey2LargeIcon = function fileTypeKey2LargeIcon(type) {
  switch (type) {
    case FileTypeKey.OFFICE_WORD:
      return 'https://img.alicdn.com/imgextra/i1/O1CN01NkJXrJ29ZblZQpgZo_!!6000000008082-55-tps-32-32.svg';

    case FileTypeKey.OFFICE_EXCEL:
      return 'https://img.alicdn.com/imgextra/i1/O1CN01RuCzQg1yjijsoL4LV_!!6000000006615-55-tps-32-32.svg';

    case FileTypeKey.OFFICE_POWERPOINT:
      return 'https://img.alicdn.com/imgextra/i1/O1CN01t0jjns1fiMry5bCiD_!!6000000004040-55-tps-32-32.svg';

    case FileTypeKey.DOC:
      return 'https://img.alicdn.com/imgextra/i1/O1CN01pG3xqM208HakjsQAZ_!!6000000006804-55-tps-32-32.svg';

    case FileTypeKey.SHEET:
      return 'https://img.alicdn.com/imgextra/i2/O1CN01DCPlBp1XxT00OCMdR_!!6000000002990-55-tps-32-32.svg';

    case FileTypeKey.WHITEBOARD:
      return 'https://img.alicdn.com/imgextra/i4/O1CN01V8EWMG1MRRWwdUoQ5_!!6000000001431-55-tps-32-32.svg';

    case FileTypeKey.PRESENTATION:
      return 'https://img.alicdn.com/imgextra/i4/O1CN01gClITa1khWDBEQTls_!!6000000004715-55-tps-32-32.svg';

    case FileTypeKey.MIND:
      return 'https://img.alicdn.com/imgextra/i1/O1CN01sH2OwL28G5GW0kYV0_!!6000000007904-55-tps-32-32.svg';

    case FileTypeKey.FOLDER:
      return 'https://img.alicdn.com/imgextra/i3/O1CN01hulOoL1psZkRw7RqX_!!6000000005416-55-tps-66-66.svg';
    // [FileTypeKey.SHEET]: 'https://img.alicdn.com/imgextra/i3/O1CN01SmA22l1stRLMYanKl_!!6000000005824-2-tps-128-128.png',
  }
}; // 基础列表项

export var EQuickInsertStatus;

(function (EQuickInsertStatus) {
  EQuickInsertStatus[EQuickInsertStatus["loading"] = 0] = "loading";
  EQuickInsertStatus[EQuickInsertStatus["ready"] = 1] = "ready";
  EQuickInsertStatus[EQuickInsertStatus["loadMore"] = 2] = "loadMore";
  EQuickInsertStatus[EQuickInsertStatus["skeleton"] = 3] = "skeleton";
})(EQuickInsertStatus || (EQuickInsertStatus = {}));
