// @ali/4ever-bi/dist-web/plugins/hr/index.js
import _extends from "@babel/runtime/helpers/extends";
import toolbar from "./toolbar";
import menubar from "./menubar";
import renderNode from "./renderNode";
import { DEPRECATED_ACTIONS, actions } from "./actions";
import onAction from "./handlers/onAction";
import onKeyDown from "./handlers/onKeyDown";
import insertHr from "./commands/insertHr";
import quickInsert from "./quickInsertConfig";
import { queries } from "./queries";
import * as utils from "./utils";
import createRenderHoverToolbar from "./createRenderHoverToolbar";
import { moFactory } from '@ali/4ever-plugin-hr';
var commands = {
  insertHr: insertHr
};

function createHrPlugin(config) {
  var hrConfig = config.hr || {};
  return _extends({}, moFactory({}), {
    toolbar: toolbar(hrConfig),
    menubar: menubar(),
    renderNode: renderNode,
    actions: actions,
    onAction: onAction,
    queries: queries,
    onKeyDown: onKeyDown,
    commands: commands,
    quickInsert: quickInsert,
    models: {
      klass: 'block',
      type: 'hr'
    },
    renderHoverToolbar: createRenderHoverToolbar(hrConfig)
  });
}

createHrPlugin.actions = DEPRECATED_ACTIONS;
createHrPlugin.commands = commands;
createHrPlugin.utils = utils;
export default createHrPlugin;
