// @ali/4ever-bi/dist-web/plugins/newlineGuide/components/styled.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
export var GuideWrapper = /*#__PURE__*/styled.div(["position:absolute;z-index:100;line-height:initial;height:21px;line-height:22px;font-size:11pt;transform:translateY(-50%);color:rgba(0,0,0,0.14);user-select:none;"]);
export var HolderText = /*#__PURE__*/styled.span(["padding-left:3px;"]);
export var HolderPlus = /*#__PURE__*/styled.span(["margin:0 2px;display:inline-block;"]);
export var HolderInsertText = /*#__PURE__*/styled.span(["text-decoration:underline;text-underline-position:under;cursor:pointer;margin-left:2px;pointer-events:auto;color:rgba(23,26,29,0.4);&:hover{color:#317ED0;}"]);
export var GuideMenuWrapper = /*#__PURE__*/styled.div(["width:206px;max-height:360px;overflow-y:auto;"]);
export var insertBtnStyle = {
  width: 19,
  height: 19,
  offset: 4
};
export var InsertBtnWrapper = /*#__PURE__*/styled.div(["display:flex;align-items:center;justify-content:center;width:19px;height:19px;border-radius:50%;background-color:#fff;cursor:pointer;border:1px solid rgba(126,134,142,0.24);&:hover{background-color:rgba(50,96,201,0.06);}&:active{background-color:rgba(50,96,201,0.09);}"]);
export var NewlineInsert = /*#__PURE__*/styled.div(["position:absolute;"]);
export var TooltipText = /*#__PURE__*/styled.div(["color:rgba(255,255,255,.56);display:flex;"]);
export var PrimaryText = /*#__PURE__*/styled.div(["margin-right:4px;color:#fff;"]);
export var AddIcon = /*#__PURE__*/_createElement("svg", {
  width: "9",
  height: "9",
  viewBox: "0 0 9 9",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/_createElement("path", {
  d: "M5 4V0H4V4H0V5H4V9H5V5H9V4H5Z",
  fill: "#171A1D",
  fillOpacity: "0.6"
}));
