// @ali/4ever-bi/dist-web/plugins/listGroup/index.js
import { ListGroupPlugin as BambooListGroup } from '@ali/4ever-bamboo';
import { groupFactory } from '@ali/4ever-factory';
var DRAG_BUTTON_OFFSET = -20;
var _BambooListGroup$hand = BambooListGroup.handlers,
    createRenderNode = _BambooListGroup$hand.createRenderNode,
    createGroupNodes = _BambooListGroup$hand.createGroupNodes;
var isFoldGroupNode = BambooListGroup.utils.isFoldGroupNode;
var ListGroupPlugin = groupFactory(function (config) {
  var _config$listGroup = config.listGroup,
      listGroup = _config$listGroup === void 0 ? {} : _config$listGroup;
  return {
    groupType: 'listGroup',
    renderNode: createRenderNode(listGroup),
    groupNodes: createGroupNodes(listGroup),
    queries: {
      isListGroupAvailable: function isListGroupAvailable() {
        return listGroup.ignore ? !listGroup.ignore() : true;
      },
      getLeftToolbarOffset: function getLeftToolbarOffset(controller, data) {
        var _node$data, _node$data$list;

        var node = data.node;

        if ((node == null ? void 0 : (_node$data = node.data) == null ? void 0 : (_node$data$list = _node$data.list) == null ? void 0 : _node$data$list.level) === 0 && isFoldGroupNode(node, controller)) {
          return {
            // 这里加上4是因为左侧工具栏的offset为6，但是在标题场景下是2
            x: DRAG_BUTTON_OFFSET + 4,
            y: 0
          };
        }

        return undefined;
      }
    }
  };
});
export default ListGroupPlugin;
