// @ali/4ever-plugin-template-button/dist-web/bi/actions/index.js
import { createActionHandlers } from "@ali/4ever-cangjie";
var TOGGLE_TEMPLATE_BUTTON_OPEN = Symbol('TOGGLE_TEMPLATE_BUTTON_OPEN');
var SET_TEMPLATE_BUTTON_DATA = Symbol('SET_TEMPLATE_BUTTON_DATA');
var INSERT_TEMPLATE_BUTTON = Symbol('INSERT_TEMPLATE_BUTTON');
var INSERT_FRAGMENT = Symbol('INSERT_FRAGMENT');
var DELETE_BUTTON = Symbol('DELETE_BUTTON');
var COPY_BUTTON = Symbol('COPY_BUTTON');
var CUT_BUTTON = Symbol('CUT_BUTTON');

var _createActionHandlers = createActionHandlers(INSERT_TEMPLATE_BUTTON),
    insertTemplateButton = _createActionHandlers[0],
    isInsertTemplateButton = _createActionHandlers[1];

var _createActionHandlers2 = createActionHandlers(TOGGLE_TEMPLATE_BUTTON_OPEN),
    toggleTemplateButton = _createActionHandlers2[0],
    isToggleTemplateButton = _createActionHandlers2[1];

var _createActionHandlers3 = createActionHandlers(SET_TEMPLATE_BUTTON_DATA),
    setTemplateButton = _createActionHandlers3[0],
    isSetTemplateButton = _createActionHandlers3[1];

var _createActionHandlers4 = createActionHandlers(INSERT_FRAGMENT),
    insertTemplateButtonFragment = _createActionHandlers4[0],
    isInsertTemplateButtonFragment = _createActionHandlers4[1];

var _createActionHandlers5 = createActionHandlers(DELETE_BUTTON),
    deleteTemplateButton = _createActionHandlers5[0],
    isDeleteTemplateButton = _createActionHandlers5[1];

var _createActionHandlers6 = createActionHandlers(COPY_BUTTON),
    copyTemplateButton = _createActionHandlers6[0],
    isCopyTemplateButton = _createActionHandlers6[1];

var _createActionHandlers7 = createActionHandlers(CUT_BUTTON),
    cutTemplateButton = _createActionHandlers7[0],
    isCutTemplateButton = _createActionHandlers7[1];

export var actions = {
  insertTemplateButton: insertTemplateButton,
  toggleTemplateButton: toggleTemplateButton,
  setTemplateButton: setTemplateButton,
  insertTemplateButtonFragment: insertTemplateButtonFragment,
  deleteTemplateButton: deleteTemplateButton,
  copyTemplateButton: copyTemplateButton,
  cutTemplateButton: cutTemplateButton
};
export var isActions = {
  isInsertTemplateButton: isInsertTemplateButton,
  isToggleTemplateButton: isToggleTemplateButton,
  isSetTemplateButton: isSetTemplateButton,
  isInsertTemplateButtonFragment: isInsertTemplateButtonFragment,
  isDeleteTemplateButton: isDeleteTemplateButton,
  isCopyTemplateButton: isCopyTemplateButton,
  isCutTemplateButton: isCutTemplateButton
};
