// @ali/dingtalk-jsapi/plugin/coolAppSdk/sendMessageToSingleChat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendMessageToSingleChat = void 0;
require("../../entry/union");
var mobile_1 = require("../../entry/mobile");
/** 提供以当前用户的身份将卡片消息发送到指定单聊的能力 */
function sendMessageToSingleChat(args) {
    var _a;
    var length = JSON.stringify(args).length;
    return mobile_1._invoke('biz.util.callComponent', {
        componentType: 'h5',
        params: {
            url: "/im/cool-app-component.html?corpId=" + encodeURIComponent((_a = args === null || args === void 0 ? void 0 : args.context) === null || _a === void 0 ? void 0 : _a.corpId) + "#/send-message-to-single-chat?params=" + encodeURIComponent(JSON.stringify({
                body: args,
                bodyLengthList: [length],
            })),
            target: 'float',
            title: '提示',
            wnId: 'sendMessageToSingleChat',
            panelHeight: 'percent83',
        },
    });
}
exports.sendMessageToSingleChat = sendMessageToSingleChat;
