// @ali/4ever-plugin-card/dist-web/mo/serializer/jsonMLToHTML/index.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { getTagName, getAttributes } from '@ali/4ever-utils';
import calendar from "./calendar";
import link from "./link";
var convertPlugins = {
  calendar: calendar,
  link: link
};
export default {
  name: 'card',
  match: function match(node) {
    var tag = getTagName(node);
    var attrs = getAttributes(node);
    return tag === 'card' && !!attrs && attrs.cardType;
  },
  convert: function convert(node, path) {
    if (path === void 0) {
      path = [];
    }

    // @ts-ignore
    var attrs = getAttributes(node) || {};
    var _ref = attrs,
        metadata = _ref.metadata,
        cardType = _ref.cardType;

    var _ref2 = metadata || {
      subject: ''
    },
        subject = _ref2.subject;

    var targetConvert = convertPlugins[cardType];
    if (targetConvert) return targetConvert(attrs, path); // 兜底方案：只显示calendar中的标题

    return /*#__PURE__*/_createElement("div", null, subject);
  }
};
