// @ali/4ever-plugin-draggable/dist-node/bi/utils/getDropHolderDataFromViewKeys.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getDropHolderDataFromViewKeys;

var _getClosestDom = _interopRequireDefault(require("./getClosestDom"));

var _types = require("../types");

function getDropHolderDataFromViewKeys(controller, event, viewKeys) {
  const {
    document
  } = controller.value;
  const startViewKey = viewKeys[0];
  const endViewKey = viewKeys[viewKeys.length - 1];
  if (startViewKey === endViewKey) return undefined;
  const startViewNode = document.getNode(startViewKey);
  const endViewNode = document.getNode(endViewKey);
  const startNode = (0, _getClosestDom.default)(startViewNode);
  const endNode = (0, _getClosestDom.default)(endViewNode);
  if (!startNode || !endNode) return undefined; // 处理跨页的情形

  const startNodeRect = startNode.getBoundingClientRect();
  const endNodeRect = endNode.getBoundingClientRect();
  const {
    clientY
  } = event;

  if (clientY > startNodeRect.top && clientY < startNodeRect.top + startNodeRect.height) {
    // 在前
    return {
      rect: {
        top: startNodeRect.top,
        left: startNodeRect.left,
        width: startNodeRect.width,
        height: startNodeRect.height
      },
      pagePosition: _types.DROP_POSITION.before
    };
  } else {
    // 在后
    return {
      rect: {
        top: endNodeRect.top,
        left: endNodeRect.left,
        width: endNodeRect.width,
        height: endNodeRect.height
      },
      pagePosition: _types.DROP_POSITION.after
    };
  }
}
