// @ali/4ever-plugin-basic/dist-web/bi/actions/index.js
import { createFocusToNodeAction, isFocusToNodeAction } from "./focusToNode";
import { createValidateLastNodeAction, isValidateLastNodeAction } from "./validateLastNode";
import { createInsertBlockWithoutExtraBlankAction, isInsertBlockWithoutExtraBlankAction } from "./insertBlockWithoutExtraBlank";
import { createMergeDataAction, isMergeDataAction } from "./mergeData";
import { createTriggerMarkdownAction, isTriggerMarkdownAction } from "./triggerMarkdown";
export { createFocusToNodeAction, createValidateLastNodeAction, createInsertBlockWithoutExtraBlankAction, createMergeDataAction, createTriggerMarkdownAction, isFocusToNodeAction, isInsertBlockWithoutExtraBlankAction, isValidateLastNodeAction, isMergeDataAction, isTriggerMarkdownAction };
export var actions = {
  createFocusToNodeAction: createFocusToNodeAction,
  createValidateLastNodeAction: createValidateLastNodeAction,
  createInsertBlockWithoutExtraBlankAction: createInsertBlockWithoutExtraBlankAction,
  createMergeDataAction: createMergeDataAction,
  createTriggerMarkdownAction: createTriggerMarkdownAction
};
export var isActions = {
  isFocusToNodeAction: isFocusToNodeAction,
  isInsertBlockWithoutExtraBlankAction: isInsertBlockWithoutExtraBlankAction,
  isValidateLastNodeAction: isValidateLastNodeAction,
  isMergeDataAction: isMergeDataAction,
  isTriggerMarkdownAction: isTriggerMarkdownAction
};
