// @ali/we-design-next/dist-node/form/checkbox/Checkbox.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Checkbox = void 0;
var _react = _interopRequireDefault(require("react"));
var _CheckboxIcon = require("./CheckboxIcon");
var _styled = require("./styled");
var _useDefaultProp = require("../../common/react/useDefaultProp");
const _createElement = /*#__PURE__*/_react.default.createElement;
const Checkbox = /*#__PURE__*/_react.default.memo( /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    defaultChecked = false,
    checked: _checked,
    onChange,
    children,
    indeterminate = false,
    disabled = false
  } = props;
  const [checked, setChecked] = (0, _useDefaultProp.useDefaultProp)(defaultChecked, _checked, onChange, e => {
    return e.target.checked;
  });
  return /*#__PURE__*/_createElement(_styled.CheckboxWrap, {
    disabled: disabled,
    checked: checked
  }, /*#__PURE__*/_createElement(_styled.CheckboxInput, {
    ref: ref,
    type: "checkbox",
    onChange: setChecked,
    disabled: disabled,
    checked: checked
  }), /*#__PURE__*/_createElement(_CheckboxIcon.CheckboxIcon, {
    disabled: disabled,
    checked: checked,
    indeterminate: indeterminate
  }), children && /*#__PURE__*/_createElement(_styled.CheckboxText, null, children));
}));
exports.Checkbox = Checkbox;
