// @ali/4ever-bi/dist-web/plugins/common/commonPlugin.js
// reexport for bi api, 插件改为从 4ever-utils 引入
import { isAtEndOfParagraph, isAtStartOfParagraph, isSelectionAtSameBlock, getPointAndBlockOnSelection, changeNullSelectionMark, generateKey, getClosestParagraph, getRandomId, hasMark, isCollapsedInEmptyText, isInsertInline, re, getMatches, getOffset, getMarkTextMatchInfo } from '@ali/4ever-utils';
var CommonPlugin = {
  utils: {
    toggleRangesMark: function toggleRangesMark(controller, mark) {
      return controller.command('toggleRangesMark', mark);
    },
    isAtEndOfParagraph: isAtEndOfParagraph,
    isAtStartOfParagraph: isAtStartOfParagraph,
    //isSection,
    isSelectionAtSameBlock: isSelectionAtSameBlock,
    //isSingleSection,
    getPointAndBlockOnSelection: getPointAndBlockOnSelection,
    //unwrapSingleSection,
    changeNullSelectionMark: changeNullSelectionMark,
    generateKey: generateKey,
    getClosestParagraph: getClosestParagraph,
    getRandomId: getRandomId,
    hasMark: hasMark,
    isCollapsedInEmptyText: isCollapsedInEmptyText,
    isInsertInline: isInsertInline,
    re: re,
    getMatches: getMatches,
    getOffset: getOffset,
    getMarkTextMatchInfo: getMarkTextMatchInfo
  }
};
export { CommonPlugin };
