// @ali/4ever-plugin-heading/dist-node/mo/serializer/jsonMLToValue.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _everUtils = require("@ali/4ever-utils");

var _models = _interopRequireDefault(require("../models"));

const tagToFactory = {
  h1: _models.default.Heading1.create,
  h2: _models.default.Heading2.create,
  h3: _models.default.Heading3.create,
  h4: _models.default.Heading4.create,
  h5: _models.default.Heading5.create,
  h6: _models.default.Heading6.create
};
var _default = {
  match: node => {
    const tag = (0, _everUtils.getTagName)(node);
    return tagToFactory[tag];
  },
  convert: node => {
    const data = (0, _everUtils.getAttributes)(node) || {};

    if (data.ind && typeof data.ind.firstline === 'number') {
      data.ind.firstLine = data.ind.firstline;
      delete data.ind.firstline;
    } // 序列化时对不合法 level 进行矫正


    if (data.list) {
      const hasInvalidLevel = list => {
        const {
          level
        } = list;
        return !/\d/g.test(String(level)) || level < 0;
      };

      if (hasInvalidLevel(data.list)) {
        data.list.level = 0;
      }
    }

    const factory = tagToFactory[(0, _everUtils.getTagName)(node)] || _everPluginParagraph.Paragraph.create;

    return factory({
      data
    });
  }
};
exports.default = _default;
