// @ali/4ever-bi/dist-web/plugins/vertAlign/index.js
import _extends from "@babel/runtime/helpers/extends";
import toolbar from "./toolbar";
import menubar from "./menubar";
import isSub from "./queries/isSub";
import isSup from "./queries/isSup";
import * as actions from "./actions";
import onAction from "./handlers/onAction";
import onKeyDown from "./handlers/onKeyDown";
import createOnCangjieInput from "./handlers/onCangjieInput";
import createRenderToolbar from "./createRenderToolbar";
import { moFactory } from '@ali/4ever-plugin-vert-align';
var queries = {
  isSub: isSub,
  isSup: isSup
};

function createVertAlignPlugin(config) {
  var vertAlignConfig = config.vertAlign || {};
  return _extends({}, moFactory({}), {
    toolbar: toolbar(vertAlignConfig),
    renderToolbar: createRenderToolbar(vertAlignConfig.locale),
    menubar: menubar(),
    queries: queries,
    onAction: onAction,
    onKeyDown: onKeyDown,
    onCangjieInput: createOnCangjieInput(),
    models: {
      klass: 'mark',
      type: 'vertAlign'
    }
  });
}

createVertAlignPlugin.actions = actions;
createVertAlignPlugin.queries = queries;
export default createVertAlignPlugin;
