// @ali/4ever-plugin-find-and-replace/dist-node/utils/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  getHighlightColor: true,
  isTextMatch: true,
  isReadonlyMatch: true,
  mergeData: true,
  isDel: true,
  setAncestorsUnfold: true,
  scrollIntoFocus: true,
  transferKey: true,
  isPi: true,
  ControllerManager: true
};
Object.defineProperty(exports, "getHighlightColor", {
  enumerable: true,
  get: function () {
    return _getHighlightColor.default;
  }
});
Object.defineProperty(exports, "isTextMatch", {
  enumerable: true,
  get: function () {
    return _isTextMatch.default;
  }
});
Object.defineProperty(exports, "isReadonlyMatch", {
  enumerable: true,
  get: function () {
    return _isReadonlyMatch.default;
  }
});
Object.defineProperty(exports, "mergeData", {
  enumerable: true,
  get: function () {
    return _mergeData.default;
  }
});
Object.defineProperty(exports, "isDel", {
  enumerable: true,
  get: function () {
    return _isDel.default;
  }
});
Object.defineProperty(exports, "setAncestorsUnfold", {
  enumerable: true,
  get: function () {
    return _setAncestorsUnfold.default;
  }
});
Object.defineProperty(exports, "scrollIntoFocus", {
  enumerable: true,
  get: function () {
    return _scrollIntoFocus.default;
  }
});
Object.defineProperty(exports, "transferKey", {
  enumerable: true,
  get: function () {
    return _transferKey.default;
  }
});
Object.defineProperty(exports, "isPi", {
  enumerable: true,
  get: function () {
    return _isPi.default;
  }
});
Object.defineProperty(exports, "ControllerManager", {
  enumerable: true,
  get: function () {
    return _ControllerManager.default;
  }
});

var _getHighlightColor = _interopRequireDefault(require("./getHighlightColor"));

var _isTextMatch = _interopRequireDefault(require("./isTextMatch"));

var _isReadonlyMatch = _interopRequireDefault(require("./isReadonlyMatch"));

var _mergeData = _interopRequireDefault(require("./mergeData"));

var _isDel = _interopRequireDefault(require("./isDel"));

var _setAncestorsUnfold = _interopRequireDefault(require("./setAncestorsUnfold"));

var _scrollIntoFocus = _interopRequireDefault(require("./scrollIntoFocus"));

var _transferKey = _interopRequireDefault(require("./transferKey"));

var _isPi = _interopRequireDefault(require("./isPi"));

var _ControllerManager = _interopRequireDefault(require("./ControllerManager"));

var _getDecorations = require("./getDecorations");

Object.keys(_getDecorations).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getDecorations[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getDecorations[key];
    }
  });
});
