// @ali/code-markdown/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.getDecorated = exports.default = exports.Plugins = exports.EditorMd = void 0;

var _editor = _interopRequireDefault(require("./editor"));

var _autoResize = _interopRequireDefault(require("./plugins/autoResize"));

var _codeBlock = _interopRequireDefault(require("./plugins/block/code-block"));

var _codeInline = _interopRequireDefault(require("./plugins/block/code-inline"));

var _quote = _interopRequireDefault(require("./plugins/block/quote"));

var _wrap = _interopRequireDefault(require("./plugins/block/wrap"));

var _clear = _interopRequireDefault(require("./plugins/clear"));

var _bold = _interopRequireDefault(require("./plugins/font/bold"));

var _italic = _interopRequireDefault(require("./plugins/font/italic"));

var _strikethrough = _interopRequireDefault(require("./plugins/font/strikethrough"));

var _underline = _interopRequireDefault(require("./plugins/font/underline"));

var _fullScreen = _interopRequireDefault(require("./plugins/fullScreen"));

var _header = _interopRequireDefault(require("./plugins/header"));

var _Image = _interopRequireDefault(require("./plugins/Image"));

var _link = _interopRequireDefault(require("./plugins/link"));

var _ordered = _interopRequireDefault(require("./plugins/list/ordered"));

var _unordered = _interopRequireDefault(require("./plugins/list/unordered"));

var _logger = _interopRequireDefault(require("./plugins/logger"));

var _modeToggle = _interopRequireDefault(require("./plugins/modeToggle"));

var _table = _interopRequireDefault(require("./plugins/table"));

var _emoji = _interopRequireDefault(require("./plugins/emoji"));

var _tabInsert = _interopRequireDefault(require("./plugins/tabInsert"));

var _Plugin = require("./plugins/Plugin");

exports.PluginComponent = _Plugin.PluginComponent;

var _index = _interopRequireDefault(require("./components/DropList/index"));

exports.DropList = _index.default;

var _toolbar = require("./toolbar");

exports.EditorToolbar = _toolbar.EditorToolbar;

var _previewOnly = require("./previewOnly");

exports.EditorPreview = _previewOnly.EditorPreview;

var _editorOnly = require("./editorOnly");

exports.EditorInput = _editorOnly.EditorInput;
exports.usePlugin = _editorOnly.use;

var _richText = _interopRequireDefault(require("./plugins/richText"));

var _decorate = _interopRequireDefault(require("./utils/decorate"));

exports.getDecorated = _decorate.default;

// 注册默认插件
_editor.default.use(_header.default);

_editor.default.use(_bold.default);

_editor.default.use(_italic.default); // Editor.use(FontUnderline);


_editor.default.use(_strikethrough.default);

_editor.default.use(_unordered.default);

_editor.default.use(_ordered.default);

_editor.default.use(_quote.default); // Editor.use(BlockWrap);
// Editor.use(BlockCodeInline);


_editor.default.use(_codeBlock.default);

_editor.default.use(_table.default);

_editor.default.use(_Image.default);

_editor.default.use(_link.default); // Editor.use(Clear);
// Editor.use(Logger);


_editor.default.use(_modeToggle.default); // Editor.use(FullScreen);


_editor.default.use(_tabInsert.default);

(0, _editorOnly.use)(_richText.default);
(0, _editorOnly.use)(_header.default);
(0, _editorOnly.use)(_bold.default);
(0, _editorOnly.use)(_italic.default); // use(FontUnderline);

(0, _editorOnly.use)(_strikethrough.default);
(0, _editorOnly.use)(_emoji.default);
(0, _editorOnly.use)(_table.default);
(0, _editorOnly.use)(_quote.default);
(0, _editorOnly.use)(_codeBlock.default);
(0, _editorOnly.use)(_link.default);
(0, _editorOnly.use)(_unordered.default);
(0, _editorOnly.use)(_ordered.default); // use(FullScreen);
// use(BlockWrap);
// use(BlockCodeInline);

(0, _editorOnly.use)(_Image.default);
(0, _editorOnly.use)(_modeToggle.default); // use(Clear);

(0, _editorOnly.use)(_logger.default);
(0, _editorOnly.use)(_tabInsert.default); // 导出声明
// 导出工具组件

var EditorMd = _previewOnly.EditorMdObj;
exports.EditorMd = EditorMd;
// 导出内置插件
var Plugins = {
  Header: _header.default,
  FontBold: _bold.default,
  FontItalic: _italic.default,
  FontUnderline: _underline.default,
  FontStrikethrough: _strikethrough.default,
  ListUnordered: _unordered.default,
  ListOrdered: _ordered.default,
  BlockQuote: _quote.default,
  BlockWrap: _wrap.default,
  BlockCodeInline: _codeInline.default,
  BlockCodeBlock: _codeBlock.default,
  Table: _table.default,
  Image: _Image.default,
  Link: _link.default,
  Clear: _clear.default,
  Logger: _logger.default,
  ModeToggle: _modeToggle.default,
  FullScreen: _fullScreen.default,
  AutoResize: _autoResize.default,
  TabInsert: _tabInsert.default
}; // 导出编辑器

exports.Plugins = Plugins;
var _default = _editor.default;
exports.default = _default;
