// @ali/4ever-plugin-card/dist-node/mo/serializer/jsonMLToHTML/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _everUtils = require("@ali/4ever-utils");

var _calendar = _interopRequireDefault(require("./calendar"));

var _link = _interopRequireDefault(require("./link"));

const _createElement = /*#__PURE__*/React.createElement;
const convertPlugins = {
  calendar: _calendar.default,
  link: _link.default
};
var _default = {
  name: 'card',
  match: node => {
    const tag = (0, _everUtils.getTagName)(node);
    const attrs = (0, _everUtils.getAttributes)(node);
    return tag === 'card' && !!attrs && attrs.cardType;
  },
  convert: (node, path = []) => {
    // @ts-ignore
    const attrs = (0, _everUtils.getAttributes)(node) || {};
    const {
      metadata,
      cardType
    } = attrs;
    const {
      subject
    } = metadata || {
      subject: ''
    };
    const targetConvert = convertPlugins[cardType];
    if (targetConvert) return targetConvert(attrs, path); // 兜底方案：只显示calendar中的标题

    return /*#__PURE__*/_createElement("div", null, subject);
  }
};
exports.default = _default;
