// @ali/4ever-plugin-table/dist-node/utils/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setTableMinColsWidthData = setTableMinColsWidthData;
exports.isTableSetTableMinColsWidthDataAction = isTableSetTableMinColsWidthDataAction;
exports.adjustColsWidth = adjustColsWidth;
exports.isTableAdjustColsWidthAction = isTableAdjustColsWidthAction;
exports.TABLE_ADJUST_COLS_WIDTH = exports.TABLE_SET_MIN_COLS_WIDTH_DATA = void 0;
const TABLE_SET_MIN_COLS_WIDTH_DATA = Symbol('TABLE_SET_MIN_COLS_WIDTH_DATA');
exports.TABLE_SET_MIN_COLS_WIDTH_DATA = TABLE_SET_MIN_COLS_WIDTH_DATA;

function setTableMinColsWidthData(data) {
  return {
    type: TABLE_SET_MIN_COLS_WIDTH_DATA,
    payload: {
      data
    }
  };
}

function isTableSetTableMinColsWidthDataAction(action) {
  return action.type === TABLE_SET_MIN_COLS_WIDTH_DATA;
}

const TABLE_ADJUST_COLS_WIDTH = Symbol('TABLE_ADJUST_COLS_WIDTH');
exports.TABLE_ADJUST_COLS_WIDTH = TABLE_ADJUST_COLS_WIDTH;

function adjustColsWidth(node, containerWidth) {
  return {
    type: TABLE_ADJUST_COLS_WIDTH,
    payload: {
      node,
      containerWidth
    }
  };
}

function isTableAdjustColsWidthAction(action) {
  return action.type === TABLE_ADJUST_COLS_WIDTH;
}
