// @ali/4ever-plugin-table/dist-node/utils/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_TABLE_BORDER_STYLE = exports.lineMap = exports.default = exports.DEFAULT_BORDER_COLOR = exports.TABLE_HEADER_BG_COLOR = exports.STICKY_TOOLBAR_INDEX_MAP = exports.STICKY_ROW_LEFT_WIDTH = exports.STICKY_ROW_TOP_HEIGHT = exports.REALTABLE_PADDING = exports.QUICK_INSERT_BUTTON_WIDTH_HEIGHT = exports.DEFAULT_TOTAL_WIDTH_FOR_AUTOFIT = exports.MIN_ROW_HEIGHT = exports.DEFAULT_PCT_WIDTH = exports.MIN_COL_WIDTH = exports.TOOLBAR_ITEM_SIZE_MOBILE = exports.TOOLBAR_ITEM_SIZE = exports.SCROLL_BAR_HEIGHT = exports.MOBILE_SPACING = exports.SPACING = void 0;
const SPACING = 16;
exports.SPACING = SPACING;
const MOBILE_SPACING = 24;
exports.MOBILE_SPACING = MOBILE_SPACING;
const SCROLL_BAR_HEIGHT = 8;
exports.SCROLL_BAR_HEIGHT = SCROLL_BAR_HEIGHT;
const TOOLBAR_ITEM_SIZE = 8;
exports.TOOLBAR_ITEM_SIZE = TOOLBAR_ITEM_SIZE;
const TOOLBAR_ITEM_SIZE_MOBILE = 16;
exports.TOOLBAR_ITEM_SIZE_MOBILE = TOOLBAR_ITEM_SIZE_MOBILE;
const MIN_COL_WIDTH = 33;
exports.MIN_COL_WIDTH = MIN_COL_WIDTH;
const DEFAULT_PCT_WIDTH = 100;
exports.DEFAULT_PCT_WIDTH = DEFAULT_PCT_WIDTH;
const MIN_ROW_HEIGHT = 33; // 边距适中，A4 纸的宽度

exports.MIN_ROW_HEIGHT = MIN_ROW_HEIGHT;
const DEFAULT_TOTAL_WIDTH_FOR_AUTOFIT = 650; // 降低以处理下边距过大的问题

exports.DEFAULT_TOTAL_WIDTH_FOR_AUTOFIT = DEFAULT_TOTAL_WIDTH_FOR_AUTOFIT;
const QUICK_INSERT_BUTTON_WIDTH_HEIGHT = 18;
/**
 * realTable padding
 */

exports.QUICK_INSERT_BUTTON_WIDTH_HEIGHT = QUICK_INSERT_BUTTON_WIDTH_HEIGHT;
const REALTABLE_PADDING = {
  top: 10,
  bottom: 0
}; // 表头行吸顶时，首行距离 scrollContainer 顶部空间

exports.REALTABLE_PADDING = REALTABLE_PADDING;
const STICKY_ROW_TOP_HEIGHT = 30; // 表头行吸顶时，行工具栏左侧遮挡区域宽度

exports.STICKY_ROW_TOP_HEIGHT = STICKY_ROW_TOP_HEIGHT;
const STICKY_ROW_LEFT_WIDTH = 10; // 表头行吸顶时，工具栏 z-index

exports.STICKY_ROW_LEFT_WIDTH = STICKY_ROW_LEFT_WIDTH;
const STICKY_TOOLBAR_INDEX_MAP = {
  cornerToolbar: 22,
  colToolbar: 21,
  cornerButton: 20,
  rowToolbar: 19,
  stickyRow: 25,
  selection: 20
};
exports.STICKY_TOOLBAR_INDEX_MAP = STICKY_TOOLBAR_INDEX_MAP;
const TABLE_HEADER_BG_COLOR = '#f5f6f7';
exports.TABLE_HEADER_BG_COLOR = TABLE_HEADER_BG_COLOR;
const DEFAULT_BORDER_COLOR = '#D6D6D6';
exports.DEFAULT_BORDER_COLOR = DEFAULT_BORDER_COLOR;
var _default = {
  SPACING,
  SCROLL_BAR_HEIGHT,
  TOOLBAR_ITEM_SIZE,
  MIN_COL_WIDTH,
  DEFAULT_PCT_WIDTH,
  DEFAULT_TOTAL_WIDTH_FOR_AUTOFIT,
  REALTABLE_PADDING,
  MIN_ROW_HEIGHT,
  QUICK_INSERT_BUTTON_WIDTH_HEIGHT,
  STICKY_ROW_TOP_HEIGHT,
  STICKY_ROW_LEFT_WIDTH,
  STICKY_TOOLBAR_INDEX_MAP,
  TABLE_HEADER_BG_COLOR,
  DEFAULT_BORDER_COLOR,
  TOOLBAR_ITEM_SIZE_MOBILE
};
exports.default = _default;
const lineMap = {
  Thin: {
    lineType: 'single',
    size: 1
  },
  Medium: {
    lineType: 'single',
    size: 2
  },
  Thick: {
    lineType: 'single',
    size: 3
  },
  Dashed: {
    lineType: 'dashed',
    size: 1
  },
  MediumDashed: {
    lineType: 'dashed',
    size: 2
  }
};
exports.lineMap = lineMap;
const DEFAULT_TABLE_BORDER_STYLE = 'Thin';
exports.DEFAULT_TABLE_BORDER_STYLE = DEFAULT_TABLE_BORDER_STYLE;
