// @ali/4ever-plugin-sticker/dist-node/bi/components/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PanelWrapper = exports.InlinePlaceholderContainer = exports.InlinePlaceholderWrapper = exports.StickerPanelItem = exports.StickerSplitLine = exports.StickerPannelWrapper = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

const StickerPannelWrapper = /*#__PURE__*/_styledComponents.default.div(["height:393px;overflow-y:scroll;display:grid;padding:8px;border-radius:8px;background:#ffffff;overflow:auto;[data-type='sticker'] div{cursor:pointer;}::-webkit-scrollbar{background-color:transparent;width:4px;}::-webkit-scrollbar-thumb{background:rgba(194,194,194,1);border-radius:10px;}::-webkit-scrollbar-thumb:hover{background:rgba(0,0,0,0.45);}::-webkit-scrollbar-track{background:transparent;}"]);

exports.StickerPannelWrapper = StickerPannelWrapper;

const StickerSplitLine = /*#__PURE__*/_styledComponents.default.hr(["width:100%;height:1px;border:none;background-color:rgba(126,134,142,0.16);grid-column-start:1;grid-column-end:", ";"], props => props.columns + 1);

exports.StickerSplitLine = StickerSplitLine;

const StickerPanelItem = /*#__PURE__*/_styledComponents.default.span(["display:inline-block;padding:8px;"]);

exports.StickerPanelItem = StickerPanelItem;

const InlinePlaceholderWrapper = /*#__PURE__*/_styledComponents.default.span(["& .inline-placeholder >:nth-child(2){margin:0;}& .inline-placeholder-icon-box{align-self:auto;}"]);

exports.InlinePlaceholderWrapper = InlinePlaceholderWrapper;

const InlinePlaceholderContainer = /*#__PURE__*/_styledComponents.default.div(["", " border-radius:3px;border:none;display:inline-block;overflow:hidden;font-size:inherit;color:rgba(23,26,29,0.4);background:none;padding:2px;position:relative;&.edit{cursor:pointer;&.active{background:rgba(126,134,142,0.16);}&:hover{background:rgba(126,134,142,0.12);}&.click{background:rgba(126,134,142,0.24);}&:active{background:rgba(126,134,142,0.16);}&.select{background:rgba(0,106,254,0.24);}}& >:nth-child(2){margin-left:4px;}"], props => ['width', 'height', 'line-height'].map(attr => `${attr}: ${props.sz}px;`));

exports.InlinePlaceholderContainer = InlinePlaceholderContainer;

const PanelWrapper = /*#__PURE__*/_styledComponents.default.div(["position:absolute;z-index:103;border:1px solid rgba(17,31,44,0.12);box-shadow:1px 3px 8px 0 rgb(0 0 0 / 10%);background-color:#ffffff;border-radius:4px;top:-999px;left:-999px;visibility:hidden;transform:scale(0.8);transform-origin:top left;transition:transform 0.1s ease-in-out;&.sticker-panel-visible{transform:scale(1);visibility:visible;}"]);

exports.PanelWrapper = PanelWrapper;
