// @ali/we-toolbar/dist-web/biz/splitButton/listSet/ListSet.js
import _extends from "@babel/runtime/helpers/extends";

var _ToolbarUnorderListNo, _ToolbarOrderListNorm;

import React, { useMemo, useState, useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { withTheme, useControlledValue, ToolbarOrderListNormal, ToolbarUnorderListNormal } from '@ali/we-design';
import styled from 'styled-components';
import SplitButton from "../../../common/splitButton";
import { classnames } from "../../../utils";
export var DEFAULT_BULLET_LIST_ARRAY = ['SCIR_ECIR_SREC', 'SFDI_SRIG_SREC', 'EREC_EREC_EREC', 'SARR_SDIO_SCIR', 'SSTA_ECIR_SREC', 'SRIG_ECIR_SREC'];
var DEFAULT_STICKER_LIST_ARRAY = ['VACATION_TEA_MILKTEA', 'RIGHT_NAIL_LOVE', 'FIRE_ROCKET_NAIL'];
export var DEFAULT_NUMBER_LIST_ARRAY = ['DEC_LEN_LROM_P', 'DEC_LEN_LROM_H', 'DEC_DEC_DEC_P', 'UEN_LEN_LROM_P', 'UROM_UEN_DEC_P', 'DDEC_LEN_LROM_P', 'ZH_DEC_LEN_D', 'ZH_DEC_LEN_A', 'ZH_ZH_DEC_D'];
export var LIST_STYLE_MAP = {
  SCIR_ECIR_SREC: 'https://img.alicdn.com/tfs/TB1mWdYbAcx_u4jSZFlXXXnUFXa-78-78.svg',
  SFDI_SRIG_SREC: 'https://img.alicdn.com/tfs/TB1hpepOpP7gK0jSZFjXXc5aXXa-78-78.svg',
  EREC_EREC_EREC: 'https://img.alicdn.com/tfs/TB1cJR4czMZ7e4jSZFOXXX7epXa-78-78.svg',
  SARR_SDIO_SCIR: 'https://img.alicdn.com/tfs/TB1aaWcOEY1gK0jSZFMXXaWcVXa-78-78.svg',
  SSTA_ECIR_SREC: 'https://img.alicdn.com/tfs/TB1VoX8OxD1gK0jSZFsXXbldVXa-78-78.svg',
  SRIG_ECIR_SREC: 'https://img.alicdn.com/tfs/TB1cqp4OAL0gK0jSZFtXXXQCXXa-78-78.svg',
  VACATION_TEA_MILKTEA: 'https://img.alicdn.com/imgextra/i4/O1CN01iGy3OC23bXk0A5V99_!!6000000007274-2-tps-240-240.png',
  RIGHT_NAIL_LOVE: 'https://img.alicdn.com/imgextra/i4/O1CN01rYMRza1OM46kvi2Yx_!!6000000001690-2-tps-240-240.png',
  FIRE_ROCKET_NAIL: 'https://img.alicdn.com/imgextra/i3/O1CN01QfDOFq1R49SyPKcpW_!!6000000002057-2-tps-240-240.png',
  DEC_LEN_LROM_P: 'https://img.alicdn.com/tfs/TB19lt9OEz1gK0jSZLeXXb9kVXa-78-78.svg',
  DEC_LEN_LROM_H: 'https://img.alicdn.com/tfs/TB18HydOAL0gK0jSZFAXXcA9pXa-78-78.svg',
  DEC_DEC_DEC_P: 'https://img.alicdn.com/tfs/TB1gHycOqL7gK0jSZFBXXXZZpXa-78-78.svg',
  UEN_LEN_LROM_P: 'https://img.alicdn.com/tfs/TB1VaOnOBr0gK0jSZFnXXbRRXXa-78-78.svg',
  UROM_UEN_DEC_P: 'https://img.alicdn.com/tfs/TB1xcqhOAY2gK0jSZFgXXc5OFXa-78-78.svg',
  DDEC_LEN_LROM_P: 'https://img.alicdn.com/tfs/TB1yjive8Bh1e4jSZFhXXcC9VXa-78-78.svg',
  ZH_DEC_LEN_D: 'https://img.alicdn.com/tfs/TB1bFQZ0aL7gK0jSZFBXXXZZpXa-82-82.svg',
  ZH_DEC_LEN_A: 'https://img.alicdn.com/tfs/TB11.iDj4vbeK8jSZPfXXariXXa-82-82.svg',
  ZH_ZH_DEC_D: 'https://img.alicdn.com/imgextra/i4/O1CN01XYORsY1OjQIxXuZxz_!!6000000001741-55-tps-82-82.svg'
};
var OverlayContent = /*#__PURE__*/styled.div(["width:272px;display:flex;flex-wrap:wrap;justify-content:flex-start;padding:4px;"]);
var ListStyleItemWrap = /*#__PURE__*/styled.div(["user-select:none;"]);
var ListStyleItem = withTheme(styled.img(["width:80px;height:80px;margin:4px;border:1px solid rgba(17,31,44,0.12);border-radius:2px;cursor:pointer;image-rendering:pixelated;&:hover{border:1px solid ", ";}&.actived{border:2px solid ", ";}"], function (_ref) {
  var theme = _ref.theme;
  return theme.ui.primaryColor;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.primaryColor;
}));

function getListRole(mode, key) {
  var prefix = mode === 'ul' ? 'listUnordered' : 'listOrdered';
  return prefix + "_" + key;
}

export var ListSet = function ListSet(props) {
  var _props$mode = props.mode,
      mode = _props$mode === void 0 ? 'ul' : _props$mode,
      defaultListStyle = props.defaultListStyle,
      listStyle = props.listStyle,
      onChange = props.onChange,
      onClick = props.onClick,
      disabled = props.disabled,
      title = props.title,
      active = props.active,
      size = props.size,
      enableSticker = props.enableSticker;
  var listStyleArray = useMemo(function () {
    if (mode !== 'ul') {
      return DEFAULT_NUMBER_LIST_ARRAY;
    } else {
      return enableSticker ? DEFAULT_BULLET_LIST_ARRAY.concat(DEFAULT_STICKER_LIST_ARRAY) : DEFAULT_BULLET_LIST_ARRAY;
    }
  }, [mode]);

  var _useControlledValue = useControlledValue(listStyle, defaultListStyle || listStyleArray[0]),
      value = _useControlledValue[0],
      setValue = _useControlledValue[1];

  var _useState = useState(false),
      overlayVisible = _useState[0],
      setOverlayVisible = _useState[1];

  var handleChange = useCallback(function (e) {
    if (onChange) {
      onChange(e);
    }

    setValue(e);
    setOverlayVisible(false);
  }, [onChange, setValue]);
  var handleVisibleChange = useCallback(function (visible) {
    setOverlayVisible(visible);
  }, []);
  var handleClickLeft = useCallback(function () {
    if (onClick) {
      onClick(value);
    }
  }, [onClick, value]);
  var getOverlay = useMemo(function () {
    return /*#__PURE__*/_createElement(OverlayContent, {
      "data-testid": "overlay-ListSet"
    }, listStyleArray.map(function (item) {
      return /*#__PURE__*/_createElement(ListStyleItemWrap, {
        key: item,
        onClick: function onClick() {
          return handleChange(item);
        },
        "data-role": getListRole(mode, item),
        role: "button"
      }, /*#__PURE__*/_createElement(ListStyleItem, {
        src: LIST_STYLE_MAP[item],
        alt: "list-style-item",
        className: classnames({
          actived: item === value
        }),
        "data-testid": "option-" + item
      }));
    }));
  }, [value]);
  var renderTriggerIcon = useMemo(function () {
    return mode === 'ul' ? _ToolbarUnorderListNo || (_ToolbarUnorderListNo = /*#__PURE__*/_createElement(ToolbarUnorderListNormal, null)) : _ToolbarOrderListNorm || (_ToolbarOrderListNorm = /*#__PURE__*/_createElement(ToolbarOrderListNormal, null));
  }, [mode]);
  return /*#__PURE__*/_createElement(SplitButton, _extends({
    overlay: getOverlay,
    onClickLeft: handleClickLeft,
    disabled: disabled,
    onActiveChange: handleVisibleChange,
    testid: "ListSet_" + mode,
    title: title || mode === 'ol' ? 'we_toolbar_order_list' : 'we_toolbar_unorder_list',
    size: size,
    icon: renderTriggerIcon,
    tooltip: mode === 'ol' ? 'we_toolbar_order_list' : 'we_toolbar_unorder_list'
  }, props, {
    active: overlayVisible,
    activeLeft: active
  }));
};
