// @ali/we-toolbar/dist-web/biz/splitButton/colorPicker/sketch/common/Hue.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _style;

import React, { Component, PureComponent } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import reactCSS from 'reactcss';
import * as hue from "../helpers/hue";
export var Hue = /*#__PURE__*/function (_ref) {
  _inheritsLoose(Hue, _ref);

  function Hue() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _ref.call.apply(_ref, [this].concat(args)) || this;

    _this.handleChange = function (e) {
      var change = hue.calculateChange(e, _this.props.direction, _this.props.hsl, _this.container);
      change && typeof _this.props.onChange === 'function' && _this.props.onChange(change, e);
    };

    _this.handleMouseDown = function (e) {
      _this.handleChange(e);

      window.addEventListener('mousemove', _this.handleChange);
      window.addEventListener('mouseup', _this.handleMouseUp);
    };

    _this.handleMouseUp = function () {
      _this.unbindEventListeners();
    };

    return _this;
  }

  var _proto = Hue.prototype;

  _proto.componentWillUnmount = function componentWillUnmount() {
    this.unbindEventListeners();
  };

  _proto.unbindEventListeners = function unbindEventListeners() {
    window.removeEventListener('mousemove', this.handleChange);
    window.removeEventListener('mouseup', this.handleMouseUp);
  };

  _proto.render = function render() {
    var _this2 = this;

    var _this$props$direction = this.props.direction,
        direction = _this$props$direction === void 0 ? 'horizontal' : _this$props$direction;
    var styles = reactCSS({
      "default": {
        hue: {
          absolute: '0px 0px 0px 0px',
          borderRadius: this.props.radius,
          boxShadow: this.props.shadow
        },
        container: {
          padding: '0 2px',
          position: 'relative',
          height: '100%',
          borderRadius: this.props.radius
        },
        pointer: {
          position: 'absolute',
          left: this.props.hsl.h * 100 / 360 + "%"
        },
        slider: {
          marginTop: '1px',
          width: '4px',
          borderRadius: '1px',
          height: '8px',
          boxShadow: '0 0 2px rgba(0, 0, 0, .6)',
          background: '#fff',
          transform: 'translateX(-2px)'
        }
      },
      vertical: {
        pointer: {
          left: '0px',
          top: -(this.props.hsl.h * 100 / 360) + 100 + "%"
        }
      }
    }, {
      vertical: direction === 'vertical'
    });
    return /*#__PURE__*/_createElement("div", {
      style: styles.hue
    }, /*#__PURE__*/_createElement("div", {
      className: "hue-" + direction,
      style: styles.container,
      ref: function ref(container) {
        _this2.container = container;
      },
      onMouseDown: this.handleMouseDown,
      onTouchMove: this.handleChange,
      onTouchStart: this.handleChange
    }, _style || (_style = /*#__PURE__*/_createElement("style", null, "\n            .hue-horizontal {\n              background: linear-gradient(to right, #f00 0%, #ff0 17%, #0f0\n                33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);\n              background: -webkit-linear-gradient(to right, #f00 0%, #ff0\n                17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);\n            }\n\n            .hue-vertical {\n              background: linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,\n                #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);\n              background: -webkit-linear-gradient(to top, #f00 0%, #ff0 17%,\n                #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);\n            }\n          ")), /*#__PURE__*/_createElement("div", {
      style: styles.pointer
    }, this.props.pointer ? /*#__PURE__*/_createElement(this.props.pointer, this.props) : /*#__PURE__*/_createElement("div", {
      style: styles.slider
    }))));
  };

  return Hue;
}(PureComponent || Component);
export default Hue;
