// @ali/4ever-plugin-calendar-card/dist-web/utils/components/index.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { ClockNormalNormal, DingdriveDingdocBetaNormalNormal, PositionNormalNormal } from '@ali/we-design';
import { IconInnerWrapper } from "./paragraphFactory";
import textParagraphFactory from "./textParagraphFactory";
import { CALENDAR_CARD_PADDING_LEN, CALENDAR_CARD_PARAGRAPH_MARGIN_TOP, DEFAULT_ICON_STYLE } from "../constants";
export var ContainerWrapper = /*#__PURE__*/styled.div(["cursor:default;color:#111F2C;font-size:14px;user-select:text;background-color:", ";border-radius:8px;.weicon{color:rgba(23,26,29,0.6);&.weicon-CalendarBetaNormal > svg{& > path{fill:rgba(23,26,29,0.6)!important;}}&.weicon-PositionNormal > svg{& > path{fill:rgba(23,26,29,0.6)!important;}}}&.placeholder{box-shadow:0px 0px 1px 0px rgba(0,0,0,0.12) inset;}&.placeholder > div{color:rgba(17,31,44,0.4);}&.card-calendar-selected{overflow:hidden;background-color:", ";box-shadow:0px 0px 0px 1px #006AFE,0px 0px 0px 3px rgba(0,106,254,0.12);}&:hover{overflow:hidden;}", "{padding-left:0;}"], function (props) {
  return props.backgroundColor || '#F6F6F7';
}, function (props) {
  return props.backgroundColor || '#F5F6F8';
}, IconInnerWrapper);
export var BodyWrapper = /*#__PURE__*/styled.div(["padding:", "px;position:relative;& > div{margin-top:", "px;}& > div:first-child{margin-top:0;}"], CALENDAR_CARD_PADDING_LEN, CALENDAR_CARD_PARAGRAPH_MARGIN_TOP);
export var CommentTextWrapper = /*#__PURE__*/styled.span(["overflow:hidden;text-overflow:ellipsis;display:-webkit-box;-webkit-line-clamp:6;-webkit-box-orient:vertical;"]);
export var SubjectPlaceholder = /*#__PURE__*/styled.span(["color:rgba(17,31,44,0.4);"]);
export var TimePlaceholder = /*#__PURE__*/styled.span(["color:rgba(17,31,44,0.4);cursor:pointer;"]);
export var TimeTextParagraph = textParagraphFactory( /*#__PURE__*/_createElement(ClockNormalNormal, {
  style: DEFAULT_ICON_STYLE
}));
export var PlaceTextParagraph = textParagraphFactory( /*#__PURE__*/_createElement(PositionNormalNormal, {
  style: DEFAULT_ICON_STYLE
}));
export var CommentTextParagraph = textParagraphFactory( /*#__PURE__*/_createElement(DingdriveDingdocBetaNormalNormal, {
  style: DEFAULT_ICON_STYLE
}));
export * from "./loadingWarpper";
export * from "./extraArea";
export * from "./receiversParagraph";
export * from "./subjectParagraph";
export { RefreshTipsTextWrapper } from "./extraArea/styled";
