// @ali/we-design-next/dist-web/collaboration/mention/MentionList.js
var _img;
import React, { useState, useMemo } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { FixedSizeList } from 'react-window';
import InfiniteLoader from 'react-window-infinite-loader';
import styled from 'styled-components';
import { useTranslate } from "../../locale";
import { MentionItem } from "./items/MemberItem";
import { MentionEmptyWrapper } from "./styled";
import { CSS_SCROLLBAR } from "../../common";
var VirtualList = /*#__PURE__*/styled(FixedSizeList)(["", " overflow:overlay;"], CSS_SCROLLBAR);
export var MentionList = /*#__PURE__*/React.memo(function (props) {
  var items = props.items,
    onMention = props.onMention,
    onLoadMore = props.onLoadMore;
  var _useState = useState(0),
    active = _useState[0],
    setActive = _useState[1];
  var itemData = useMemo(function () {
    return {
      items: items,
      active: active,
      onClick: function onClick(i) {
        return onMention(i);
      },
      onMouseEnter: function onMouseEnter(i) {
        return setActive(i);
      }
    };
  }, [items, active, setActive]);
  var t = useTranslate();
  return items.length > 0 ? /*#__PURE__*/_createElement(InfiniteLoader, {
    isItemLoaded: function isItemLoaded(i) {
      return items[i] !== undefined;
    },
    itemCount: 999999,
    loadMoreItems: function loadMoreItems(start, stop) {
      onLoadMore(start, stop - start + 1);
      return;
    }
  }, function (_ref) {
    var onItemsRendered = _ref.onItemsRendered,
      ref = _ref.ref;
    return /*#__PURE__*/_createElement(VirtualList, {
      ref: ref,
      width: 300,
      height: 300,
      itemCount: items.length,
      itemSize: 54,
      itemData: itemData,
      onItemsRendered: onItemsRendered,
      style: {
        // animation: 'fadeIn',
        // animationDuration: '.35s',
        overflowX: 'hidden',
        overflowY: 'overlay'
      },
      children: MentionItem
    });
  }) : /*#__PURE__*/_createElement(MentionEmptyWrapper, null, _img || (_img = /*#__PURE__*/_createElement("img", {
    src: "https://img.alicdn.com/imgextra/i2/O1CN019Tgrpl25H3CCoowv1_!!6000000007500-55-tps-90-90.svg",
    alt: ""
  })), t('wdn_mention_search_no_result'));
});
