// @ali/4ever-bi/dist-web/plugins/image/components/mobileImageButton.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { selectImage } from "../actions";

var isButtonDisabled = function isButtonDisabled(controller) {
  var _controller$value = controller.value,
      document = _controller$value.document,
      selection = _controller$value.selection;
  var isDisabled = document.getClosest(selection.anchor.key, function (n) {
    return controller.query('isVoid', n);
  });
  return Boolean(isDisabled);
};

export function MobileImage(props) {
  var controller = props.controller,
      Wrapper = props.Wrapper,
      item = props.item,
      locale = props.locale;

  var handleClick = function handleClick() {
    controller.run('onCangjieBlur');
    controller.run('onAction', selectImage());
  };

  return /*#__PURE__*/_createElement(Wrapper, {
    item: item,
    type: "line",
    icon: "image",
    lineText: locale == null ? void 0 : locale.quickInsertTitle,
    isDisabled: isButtonDisabled(controller),
    onClick: handleClick
  });
}
export function MobileBigImage(props) {
  var controller = props.controller,
      Wrapper = props.Wrapper,
      item = props.item;

  var handleClick = function handleClick() {
    controller.run('onCangjieBlur');
    controller.run('onAction', selectImage());
  };

  return /*#__PURE__*/_createElement(Wrapper, {
    item: item,
    type: "big",
    icon: "image",
    isDisabled: isButtonDisabled(controller),
    onClick: handleClick
  });
}
