// @ali/4ever-bi/dist-web/plugins/image/components/quickInsertPanel.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { QuickInsertCommonPanel as CommonPanel, QuickInsertButton as Button } from '@ali/4ever-component';
import { selectImage } from "../actions";

var placeholder = /*#__PURE__*/_createElement("svg", {
  width: "59px",
  height: "47px",
  viewBox: "0 0 59 47",
  version: "1.1"
}, /*#__PURE__*/_createElement("g", {
  stroke: "none",
  strokeWidth: "1",
  fill: "none",
  fillRule: "evenodd",
  opacity: "0.140276228"
}, /*#__PURE__*/_createElement("g", {
  transform: "translate(-1463.000000, -560.000000)",
  fill: "#000000",
  fillRule: "nonzero"
}, /*#__PURE__*/_createElement("g", {
  id: "Picture",
  transform: "translate(1457.000000, 548.000000)"
}, /*#__PURE__*/_createElement("rect", {
  opacity: "0",
  x: "0",
  y: "0",
  width: "71",
  height: "71"
}), /*#__PURE__*/_createElement("path", {
  d: "M11.9,12 L59.1,12 C63.0333333,12 65,13.9583333 65,17.875 L65,53.125 C65,57.0416667 63.0333333,59 59.1,59 L11.9,59 C7.96666673,59 6,57.0416667 6,53.125 L6,17.875 C6,13.9583333 7.96666668,12 11.9,12 Z",
  opacity: "0.3"
}), /*#__PURE__*/_createElement("polygon", {
  opacity: "0.3",
  points: "12 59 31 32 50 59"
}), /*#__PURE__*/_createElement("polygon", {
  points: "32 59 45.5 41 59 59"
}), /*#__PURE__*/_createElement("path", {
  d: "M50,25.5 C50,27.9852814 52.0147186,30 54.5,30 C56.9852814,30 59,27.9852814 59,25.5 C59,23.0147186 56.9852814,21 54.5,21 C52.0147186,21 50,23.0147186 50,25.5 Z",
  opacity: "0.3"
})))));

var QuickInsertPanel = function QuickInsertPanel(props) {
  var close = props.close,
      locale = props.locale;
  var handleInsert = React.useCallback(function () {
    return close().run('onAction', selectImage());
  }, [close]);
  return /*#__PURE__*/_createElement(CommonPanel, {
    placeholder: placeholder
  }, /*#__PURE__*/_createElement(Button, {
    onClick: handleInsert
  }, locale == null ? void 0 : locale.buttonTitle));
};

export default QuickInsertPanel;
