// @ali/4ever-bi/dist-web/plugins/image/components/recommendStyleItems.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import classnames from 'classnames';
import { ImagePlugin, PluginRoles } from '@ali/4ever-bamboo';
import logger from '@ali/4ever-logger';
import { isStyleActive } from "../utils/isStyleActive";
import { blockEvent } from '@ali/4ever-utils';
import { getRecommendStyleItems } from "../config";
export var RecommendStyleMenuType;

(function (RecommendStyleMenuType) {
  RecommendStyleMenuType["SMALL"] = "small";
  RecommendStyleMenuType["LARGE"] = "large";
})(RecommendStyleMenuType || (RecommendStyleMenuType = {}));

var IMAGE_EXAMPLE_URL = 'https://img.alicdn.com/imgextra/i3/O1CN016CBXo01OH1jCj0CB4_!!6000000001679-2-tps-224-160.png';
var EXAMPLE_WIDTH = 224;
var EXAMPLE_HEIGHT = 160;
var EXAMPLE_RATIO = EXAMPLE_HEIGHT / EXAMPLE_WIDTH;
var IMAGE_STYLE_MENU_SMALL = 'image-style-menu-small';
var IMAGE_STYLE_MENU_ITEM_SMALL = 'image-style-menu-item-small';
var IMAGE_STYLE_ITEM_SELECTED = 'image-style-item-selected';
var IMAGE_MARGIN_SMALL = 6;
var IMAGE_MARGIN_NORMAL = 8;
var ItemImage = /*#__PURE__*/styled.div(["background-image:url('", "'),linear-gradient(#FAFAFB,#F2F4F5);height:100%;box-sizing:border-box;background-size:cover;cursor:pointer;"], IMAGE_EXAMPLE_URL);
var ItemImageWrapper = /*#__PURE__*/styled.div(["position:absolute;top:", ";right:", ";width:", ";height:", ";"], function (p) {
  return p.margin + "px";
}, function (p) {
  return p.margin + "px";
}, function (p) {
  return "calc(100% - " + p.margin * 2 + "px)";
}, function (p) {
  return "calc(100% - " + p.margin * 2 + "px)";
});
var ItemImageBorderWrapper = /*#__PURE__*/styled.div(["width:100%;height:100%;overflow:hidden;"]);
var MenuWrapper = /*#__PURE__*/styled.div(["display:grid;grid-template-columns:repeat(3,80px);grid-row-gap:9px;grid-column-gap:3px;justify-content:space-around;&.", "{grid-template-columns:repeat(3,1fr);grid-row-gap:6px;justify-content:space-between;}"], IMAGE_STYLE_MENU_SMALL);
var ItemImageStatusWrapper = /*#__PURE__*/styled.div(["position:relative;display:flex;align-items:center;justify-content:center;padding-top:calc(calc(100% * ", ") + ", "px);border-radius:8px;background:transparent;transition:background 0.3s ease-in-out;&:hover{background:rgba(126,134,142,0.12);}&.", "{background:rgba(0,137,255,0.12);}&.", "{padding-top:calc(calc(100% * ", ") + ", "px);border-radius:5px;}"], EXAMPLE_RATIO, (1 - EXAMPLE_RATIO) * IMAGE_MARGIN_NORMAL * 2, IMAGE_STYLE_ITEM_SELECTED, IMAGE_STYLE_MENU_ITEM_SMALL, EXAMPLE_RATIO, (1 - EXAMPLE_RATIO) * IMAGE_MARGIN_SMALL * 2);

var RecommendItem = function RecommendItem(props) {
  var _classnames;

  var isActive = props.isActive,
      imageData = props.imageData,
      demoData = props.demoData,
      onClick = props.onClick,
      size = props.size,
      index = props.index;
  var _demoData$outline = demoData.outline,
      outline = _demoData$outline === void 0 ? {} : _demoData$outline,
      radius = demoData.radius,
      shadow = demoData.shadow;
  var outlineWidth = outline.width;
  var loggerKey = "" + (size === RecommendStyleMenuType.SMALL ? 'image-inline-toolbar-' : 'image-sidebar-') + index;
  var imageWrapperStyle = React.useMemo(function () {
    var border = ImagePlugin.getBorder(outline);
    return {
      border: border,
      boxShadow: shadow,
      borderRadius: radius || 0
    };
  }, [outlineWidth, shadow, radius]);
  var handleClick = React.useCallback(function (e) {
    e.preventDefault();
    onClick(imageData);
    logger.sum(loggerKey);
  }, [onClick, imageData, loggerKey]);
  return /*#__PURE__*/_createElement(ItemImageStatusWrapper, {
    className: classnames((_classnames = {}, _classnames[IMAGE_STYLE_ITEM_SELECTED] = isActive, _classnames[IMAGE_STYLE_MENU_ITEM_SMALL] = size === RecommendStyleMenuType.SMALL, _classnames)),
    "data-testid": "image-style-option-" + index
  }, /*#__PURE__*/_createElement(ItemImageWrapper, {
    margin: size === RecommendStyleMenuType.SMALL ? IMAGE_MARGIN_SMALL : IMAGE_MARGIN_NORMAL
  }, /*#__PURE__*/_createElement(ItemImageBorderWrapper, {
    style: imageWrapperStyle
  }, /*#__PURE__*/_createElement(ItemImage, {
    onClick: handleClick,
    onMouseDown: blockEvent,
    "data-testid": "image-recommend-style-" + index,
    "data-role": "" + PluginRoles.imageRecommendStyleItem + index
  }))));
};

export var RecommendStyleMenu = function RecommendStyleMenu(props) {
  var _classnames2;

  var _props$size = props.size,
      size = _props$size === void 0 ? RecommendStyleMenuType.LARGE : _props$size,
      data = props.data,
      onClick = props.onClick;
  var items = React.useMemo(getRecommendStyleItems, []);
  return /*#__PURE__*/_createElement(MenuWrapper, {
    className: classnames((_classnames2 = {}, _classnames2[IMAGE_STYLE_MENU_SMALL] = size === RecommendStyleMenuType.SMALL, _classnames2))
  }, items.map(function (item, index) {
    return /*#__PURE__*/_createElement(RecommendItem, {
      imageData: item.data,
      demoData: item.demo,
      isActive: isStyleActive(item.data, data || null),
      onClick: onClick,
      size: size,
      key: "image_recommend_style_item_" + index,
      index: index
    });
  }));
};
