// @ali/4ever-plugin-card/dist-node/mo/serializer/valueToText/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _models = _interopRequireDefault(require("../../models"));

var _projectCard = _interopRequireDefault(require("./projectCard"));

var _calendar = _interopRequireDefault(require("./calendar"));

var _url = _interopRequireDefault(require("./url"));

var _link = _interopRequireDefault(require("./link"));

const convertPlugins = {
  projectCard: _projectCard.default,
  calendar: _calendar.default,
  url: _url.default,
  link: _link.default
};
const valueToText = {
  name: 'card',

  match(node) {
    return _models.default.isCard(node);
  },

  convert(node) {
    const {
      data
    } = node;
    const {
      cardType,
      metadata
    } = data;
    const {
      subject
    } = metadata || {};
    const targetConvert = convertPlugins[cardType];
    if (targetConvert) return targetConvert(data);
    return subject || '';
  }

};
var _default = valueToText;
exports.default = _default;
