// @ali/we-design-next/dist-node/basic/badge/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.New = exports.DotWrapper = exports.Dot = exports.Beta = void 0;
var _react = require("react");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _colors = require("../../common/consts/colors");
const RED_DOT_CSS = `
  display: inline-block;
  box-sizing: border-box;
  width: 10px;
  height: 10px;
  border-radius: 5px;
  background-color: #FF5219;
  border: 2px solid white;
  vertical-align: top;
  top: 0px;
  right: 0px;
`;
const Dot = /*#__PURE__*/_styledComponents.default.span(["display:inline-block;", ""], RED_DOT_CSS);
exports.Dot = Dot;
const ClassWrapper = ({
  children,
  className
}) => {
  return className ? /*#__PURE__*/(0, _react.cloneElement)(children, {
    className: `${className} ${children.props.className || ''}`
  }) : children;
};
const DotWrapper = /*#__PURE__*/(0, _styledComponents.default)(ClassWrapper)(["&::after{content:'';", "}"], RED_DOT_CSS);
exports.DotWrapper = DotWrapper;
const BADGE_TEXT_CSS = `
  display: inline-block;
  font-size: 12px;
  line-height: 12px;
  text-align: center;
  height: 12px;
  border-radius: 2px;
  margin-left: 8px;
  padding: 0 2px;
  span {
    display: inline-block;
    transform: scale(0.83);
  }
`;
const New = /*#__PURE__*/_styledComponents.default.div(["color:rgb(255,146,0);background-color:rgba(255,146,0,0.12);", ""], BADGE_TEXT_CSS);
exports.New = New;
const Beta = /*#__PURE__*/_styledComponents.default.div(["color:", ";background-color:rgba(56,155,255,0.12);", ""], _colors.COLORS.COMMON.BLUE1_HOVER_COLOR, BADGE_TEXT_CSS);
exports.Beta = Beta;
