// @ali/4ever-plugin-link/dist-node/utils/bamboo.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  createDecorateNode: true,
  getLinkSz: true,
  getLinkMaxSz: true,
  LinkCss: true,
  normalizeLink: true,
  openLink: true,
  getMinTableColWidth: true,
  getMinTableColWidthForMobile: true,
  getMinTableColWidthForBlockLinkCard: true
};
Object.defineProperty(exports, "createDecorateNode", {
  enumerable: true,
  get: function () {
    return _createDecorateNode.default;
  }
});
Object.defineProperty(exports, "getLinkSz", {
  enumerable: true,
  get: function () {
    return _getLinkSz.getLinkSz;
  }
});
Object.defineProperty(exports, "getLinkMaxSz", {
  enumerable: true,
  get: function () {
    return _getLinkSz.getLinkMaxSz;
  }
});
Object.defineProperty(exports, "LinkCss", {
  enumerable: true,
  get: function () {
    return _styled.LinkCss;
  }
});
Object.defineProperty(exports, "normalizeLink", {
  enumerable: true,
  get: function () {
    return _openLink.normalizeLink;
  }
});
Object.defineProperty(exports, "openLink", {
  enumerable: true,
  get: function () {
    return _openLink.openLink;
  }
});
Object.defineProperty(exports, "getMinTableColWidth", {
  enumerable: true,
  get: function () {
    return _getMinTableColWidth.default;
  }
});
Object.defineProperty(exports, "getMinTableColWidthForMobile", {
  enumerable: true,
  get: function () {
    return _getMinTableColWidth.getMinTableColWidthForMobile;
  }
});
Object.defineProperty(exports, "getMinTableColWidthForBlockLinkCard", {
  enumerable: true,
  get: function () {
    return _getMinTableColWidth.getMinTableColWidthForBlockLinkCard;
  }
});

var _createDecorateNode = _interopRequireDefault(require("./createDecorateNode"));

var _getLinkSz = require("./getLinkSz");

var _styled = require("./styled");

var _openLink = require("./openLink");

var _getMinTableColWidth = _interopRequireWildcard(require("./getMinTableColWidth"));

var _role = require("./role");

Object.keys(_role).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _role[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _role[key];
    }
  });
});
