// @ali/4ever-plugin-node-select/dist-web/common/actions/index.js
import { createActionHandlers } from '@ali/4ever-cangjie';
export var CANCEL_SELECT_NODE = Symbol('cancel_select_node');
export var COPY_SELECTED_NODES = Symbol('copy_SELECTED_NODES');
export var CUT_SELECTED_NODES = Symbol('cut_SELECTED_NODES');
export var DELETE_SELECTED_NODES = Symbol('delete_SELECTED_NODES');

var _createActionHandlers = createActionHandlers(COPY_SELECTED_NODES),
    copySelectedNodes = _createActionHandlers[0],
    isCopySelectedNodes = _createActionHandlers[1];

export { copySelectedNodes, isCopySelectedNodes };

var _createActionHandlers2 = createActionHandlers(CUT_SELECTED_NODES),
    cutSelectedNodes = _createActionHandlers2[0],
    isCutSelectedNodes = _createActionHandlers2[1];

export { cutSelectedNodes, isCutSelectedNodes };

var _createActionHandlers3 = createActionHandlers(DELETE_SELECTED_NODES),
    deleteSelectedNodes = _createActionHandlers3[0],
    isDeleteSelectedNodes = _createActionHandlers3[1];

export { deleteSelectedNodes, isDeleteSelectedNodes };

var _createActionHandlers4 = createActionHandlers(CANCEL_SELECT_NODE),
    cancelSelectNode = _createActionHandlers4[0],
    isCancelSelectNode = _createActionHandlers4[1];

export { cancelSelectNode, isCancelSelectNode };
export var actions = {
  copySelectedNodes: copySelectedNodes,
  cutSelectedNodes: cutSelectedNodes,
  deleteSelectedNodes: deleteSelectedNodes,
  cancelSelectNode: cancelSelectNode
};
