// @ali/alife-logger-docs/lib/util.js
Date.now = Date.now || function() {
    return new Date().getTime();
};

var SEQUENCE = Date.now(), noop = function() {}, getCwarn = function() {
    var t = "object" == typeof console ? console.warn : noop;
    try {
        var e = {
            warn: t
        };
        e.warn.call(e);
    } catch (r) {
        return noop;
    }
    return t;
}, util = {
    noop: noop,
    warn: getCwarn(),
    key: "__bl",
    win: "object" == typeof window && window.document ? window : undefined,
    regionMap: {
        cn: "https://arms-retcode.aliyuncs.com/r.png?",
        sg: "https://arms-retcode-sg.aliyuncs.com/r.png?",
        sg_2: "https://retcode-sg-lazada.arms.aliyuncs.com/r.png?",
        daily: "http://arms-retcode-daily.alibaba.net/r.png?",
        daily_2: "https://arms-retcode-daily.alibaba.net/r.png?",
        us: "https://retcode-us-west-1.arms.aliyuncs.com/r.png?"
    },
    defaultImgUrl: "https://arms-retcode.aliyuncs.com/r.png?",
    createObject: function(t) {
        if (Object.create) return Object.create(t);
        var e = function() {};
        return e.prototype = t, new e();
    },
    each: function(t, e) {
        var r = 0, n = t.length;
        if (this.T(t, "Array")) for (;r < n && !1 !== e.call(t[r], t[r], r); r++) ; else for (r in t) if (!1 === e.call(t[r], t[r], r)) break;
        return t;
    },
    safetyCall: function(t, e, r) {
        if ("function" != typeof t) return r;
        try {
            return t.apply(this, e);
        } catch (n) {
            return r;
        }
    },
    T: function(t, e) {
        var r = Object.prototype.toString.call(t).substring(8).replace("]", "");
        return e ? r === e : r;
    },
    filterByRule: function(t, e) {
        if (!t) return "";
        if (!e) return t;
        var r = this, n = r.T(e);
        return "Function" === n ? r.safetyCall(e, [ t ], t) : "Array" === n ? (this.each(e, function(e) {
            t = r.filterByRule(t, e);
        }), t) : "Object" === n ? t.replace(e.rule, e.target || "") : t.replace(e, "");
    },
    ignoreByRule: function(t, e) {
        if (!t || !e) return !1;
        if ((this.isString(e) || e.source || "Function" === this.T(e)) && (e = [ e ]), !this.isArray(e)) return util.warn("[arms] invalid rules of ignore config, (list of) String/RegExp/Funcitons are available"), 
        !1;
        for (var r, n = [], o = 0, i = e.length; o < i; o++) if (r = e[o], this.isString(r)) n.push(r.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, "\\$1")); else if (r && r.source) n.push(r.source); else if (r && "Function" === this.T(r) && !0 === this.safetyCall(r, [ t ], !1)) return !0;
        var a = new RegExp(n.join("|"), "i");
        return !!(n.length && a.test && a.test(t));
    },
    J: function(t) {
        if (!t || "string" != typeof t) return t;
        var e = null;
        try {
            e = JSON.parse(t);
        } catch (r) {}
        return e;
    },
    pick: function(t) {
        return 1 === t || 1 === Math.ceil(Math.random() * t);
    },
    verifyConfig: function(t) {
        if ("sample" in t) {
            var e = t.sample, r = e;
            e && /^\d+(\.\d+)?%$/.test(e) && (r = parseInt(100 / parseFloat(e))), 0 < r && 1 > r && (r = parseInt(1 / r)), 
            r >= 1 && r <= 100 ? t.sample = r : delete t.sample;
        }
        return t;
    },
    on: function(t, e, r, n, o) {
        return t.addEventListener ? (o = o || !1, t.addEventListener(e, function i(a) {
            n && t.removeEventListener(e, i, o), r.call(this, a);
        }, o)) : t.attachEvent && t.attachEvent("on" + e, function a(o) {
            n && t.detachEvent("on" + e, a), r.call(this, o);
        }), this;
    },
    off: function(t, e, r) {
        return r ? (t.removeEventListener ? t.removeEventListener(e, r) : t.detachEvent && t.detachEvent(e, r), 
        this) : this;
    },
    delay: function(t, e) {
        return -1 === e ? (t(), null) : setTimeout(t, e || 0);
    },
    ext: function(t) {
        for (var e = 1, r = arguments.length; e < r; e++) {
            var n = arguments[e];
            for (var o in n) Object.prototype.hasOwnProperty.call(n, o) && (t[o] = n[o]);
        }
        return t;
    },
    sub: function(t, e) {
        var r = {};
        return this.each(t, function(t, n) {
            -1 !== e.indexOf(n) && (r[n] = t);
        }), r;
    },
    uu: function() {
        for (var t, e, r = 20, n = new Array(r), o = Date.now().toString(36).split(""); r-- > 0; ) e = (t = 36 * Math.random() | 0).toString(36), 
        n[r] = t % 3 ? e : e.toUpperCase();
        for (var i = 0; i < 8; i++) n.splice(3 * i + 2, 0, o[i]);
        return n.join("");
    },
    seq: function() {
        return (SEQUENCE++).toString(36);
    },
    decode: function(t) {
        try {
            t = decodeURIComponent(t);
        } catch (e) {}
        return t;
    },
    encode: function(t, e) {
        try {
            t = e ? encodeURIComponent(t).replace(/\(/g, "%28").replace(/\)/g, "%29") : encodeURIComponent(t);
        } catch (r) {}
        return t;
    },
    serialize: function(t) {
        t = t || {};
        var e = [];
        for (var r in t) Object.prototype.hasOwnProperty.call(t, r) && t[r] !== undefined && e.push(r + "=" + this.encode(t[r], "msg" === r));
        return e.join("&");
    },
    checkAPI: function(t, e) {
        if (!t || "string" != typeof t) return !1;
        var r = /arms-retcode[\w-]*\.aliyuncs/.test(t);
        return !r && e && (r = /(\.png)|(\.gif)|(alicdn\.com)/.test(t)), !r;
    },
    checkAutoError: function(t) {
        return !(!t || !t.message) && !/failed[\w\s]+fetch/i.test(t.message);
    },
    cutUrlSearch: function(t) {
        return t && "string" == typeof t ? t.replace(/^(https?:)?\/\//, "").replace(/\?.*$/, "") : "";
    },
    removeUrlSearch: function(t) {
        return t && "string" == typeof t ? t.replace(/\?.*$/, "") : "";
    },
    createFakeToString: function(t) {
        return function() {
            return t + "() { [native code] }";
        };
    },
    checkSameOrigin: function(t, e) {
        if (!e || !t) return !1;
        var r = "//" + e.split("/")[2];
        return t === e || t.slice(0, e.length + 1) === e + "/" || t === r || t.slice(0, r.length + 1) === r + "/" || !/^(\/\/|http:|https:).*/.test(t);
    },
    getRandIP: function() {
        for (var t = [], e = 0; e < 4; e++) {
            var r = Math.floor(256 * Math.random());
            t[e] = (r > 15 ? "" : "0") + r.toString(16);
        }
        return t.join("");
    },
    getSortNum: function(t) {
        return t ? (t += 1) >= 1e3 && t <= 9999 ? t : t < 1e3 ? t + 1e3 : t % 1e4 + 1e3 : 1e3;
    },
    getRandNum: function(t) {
        return t && "string" == typeof t ? t.length < 5 ? this.getNum(5) : t.substring(t.length - 5) : this.getNum(5);
    },
    getNum: function(t) {
        for (var e = [], r = 0; r < t; r++) {
            var n = Math.floor(16 * Math.random());
            e[r] = n.toString(16);
        }
        return e.join("");
    },
    getCurDomain: function() {
        return location && location.hostname || "";
    },
    parseFetchHeaders: function(t) {
        if (!t) return {};
        var e = {};
        try {
            if ("function" == typeof t.keys) for (var r = t.keys(), n = r.next(); !n.done; ) {
                var o = n.value;
                e[o] = t.get(o), n = r.next();
            } else e = t;
        } catch (i) {
            e = {};
        }
        return e;
    },
    parseXhrHeaders: function(t) {
        if (!t && "string" != typeof t) return {};
        var e = {};
        try {
            var r = t.split("\r\n");
            e = r.reduce(function(t, e) {
                var r = e.split(": ");
                return t[r[0]] = r[1], t;
            }, {});
        } catch (n) {
            e = {};
        }
        return e;
    },
    getQuerys: function(t) {
        if (!t) return "";
        var e = {}, r = [], n = "", o = "";
        try {
            var i = [];
            if (t.indexOf("?") >= 0 && (i = t.substring(t.indexOf("?") + 1, t.length).split("&")), 
            i.length > 0) for (var a in i) n = (r = i[a].split("="))[0], o = r[1], e[n] = o;
        } catch (c) {
            e = {};
        }
        return e;
    },
    getFetchSnapshot: function(t, e, r) {
        var n, o;
        try {
            var i = (t && "string" != typeof t[0] ? t[0].url : t[0]) || "", a = (t && "string" != typeof t[0] ? t[0] : t[1]) || {}, c = "POST" === a.method.toUpperCase() ? a.body : this.getQuerys(i);
            n = {
                originApi: i,
                method: a.method || "unknown",
                params: c,
                response: e || "",
                reqHeaders: this.parseFetchHeaders(a.headers || null),
                resHeaders: this.parseFetchHeaders(r)
            }, o = "function" == typeof encodeURIComponent && JSON && encodeURIComponent(JSON.stringify(n)) || "{}";
        } catch (s) {
            o = "{}";
        }
        return o;
    },
    getXhrSnapshot: function(t, e, r) {
        if (!t || !e || !r) return {};
        var n, o;
        try {
            n = {
                originApi: t,
                method: e,
                params: this.getQuerys(t),
                response: r.responseText || r.responseXml || "",
                reqHeaders: {},
                resHeaders: this.parseXhrHeaders("function" == typeof r.getAllResponseHeaders && r.getAllResponseHeaders() || "")
            }, o = "function" == typeof encodeURIComponent && JSON && encodeURIComponent(JSON.stringify(n)) || "{}";
        } catch (i) {
            o = "{}";
        }
        return o;
    },
    isRobot: function() {
        var t = [ "nuhk", "googlebot/", "googlebot-image", "yammybot", "openbot", "slurp", "msnbot", "ask jeeves/teoma", "ia_archiver", "baiduspider", "bingbot/" ];
        if (!navigator || "string" != typeof navigator.userAgent) return !1;
        try {
            for (var e = navigator.userAgent.toLowerCase(), r = 0; r < t.length; r++) {
                var n = t[r];
                if (e.lastIndexOf(n) >= 0) return !0;
            }
        } catch (o) {
            util.warn("[arms] useragent parse error");
        }
        return !1;
    },
    isFunction: function(t) {
        return "function" == typeof t;
    },
    isPlainObject: function(t) {
        return "[object Object]" === Object.prototype.toString.call(t);
    },
    isString: function(t) {
        return "[object String]" === Object.prototype.toString.call(t);
    },
    isArray: function(t) {
        return "[object Array]" === Object.prototype.toString.call(t);
    },
    joinRegExp: function(t) {
        for (var e, r = [], n = 0, o = t.length; n < o; n++) e = t[n], this.isString(e) ? r.push(e.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, "\\$1")) : e && e.source && r.push(e.source);
        return new RegExp(r.join("|"), "i");
    },
    reWriteMethod: function(t, e, r) {
        if (null !== t) {
            var n = t[e];
            t[e] = r(n);
        }
    }
};

module.exports = util;
