// @ali/dingtalk-jsapi/plugin/customAdd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.shortcutCustomAdd = void 0;
/**
 * 此插件用于生成桌面快捷方式：https://yuque.antfin-inc.com/docs/share/5a3723e0-060d-43cd-87d7-315132fc14ee?#
 * 调用示例 :
 * import { shortcutCustomAdd } from '@ali/dingtalk-jsapi/plugin/customAdd';
 * shortcutCustomAdd({
 *  appName: '杭州健康码',
 *  appIcon: 'https://gw.alicdn.com/tfs/TB1nUmbzUT1gK0jSZFhXXaAtVXa-1024-1024.png',
 *  appUrl: 'https://h5.dingtalk.com/healthAct/index.html'
 * })
 */
var env_1 = require("../lib/env");
var openLink_1 = require("../api/biz/util/openLink");
var otherApi_1 = require("../lib/otherApi");
var mobile_1 = require("../entry/mobile");
var dingtalkEnv = env_1.getENV();
var lowerUA = env_1.getUA().toLocaleLowerCase();
var iosShortcutServerUrl = 'https://h5.dingtalk.com/deskTopApp/index.html';
var shortcutCustomAdd = function (params) {
    if (params.appIcon && params.appName && (params.appUrl || params.scheme)) {
        if (dingtalkEnv.platform === env_1.ENV_ENUM.ios) {
            var regStrSaf = /os [\d._]*/gi;
            var verinfo = lowerUA.match(regStrSaf);
            var iosVersion = (verinfo + '').replace(/[^0-9|_.]/ig, '').replace(/_/ig, '.');
            if (otherApi_1.compareVersion('10.0.0', iosVersion, true)) {
                var url = '';
                if (params.appUrl) {
                    url = "dingtalk://dingtalkclient/action/jump?url=" + encodeURIComponent(iosShortcutServerUrl + "?appName=" + encodeURIComponent(params.appName) + "&appIcon=" + encodeURIComponent(params.appIcon) + "&appUrl=" + encodeURIComponent(params.appUrl));
                }
                else if (params.scheme) {
                    url = "dingtalk://dingtalkclient/action/jump?url=" + encodeURIComponent(iosShortcutServerUrl + "?appName=" + encodeURIComponent(params.appName) + "&appIcon=" + encodeURIComponent(params.appIcon) + "&scheme=" + encodeURIComponent(params.scheme));
                }
                openLink_1.default({
                    url: url,
                });
                return Promise.resolve({
                    errorCode: '0',
                });
            }
            else {
                return Promise.reject({
                    errorCode: '1',
                    errorMessage: 'ios 10以上的版本才支持该功能',
                });
            }
        }
        else if (dingtalkEnv.platform === env_1.ENV_ENUM.android) {
            var androidParams = {
                name: params.appName,
                iconUrl: params.appIcon,
            };
            if (params.appUrl) {
                androidParams.webUrl = params.appUrl;
            }
            else if (params.scheme) {
                androidParams.jumpUrl = params.scheme;
            }
            // @ts-ignore
            if (otherApi_1.compareVersion('5.0.15', dingtalkEnv.version, true)) {
                return mobile_1._invoke('internal.shortcut.customAdd', androidParams);
            }
            else {
                return Promise.reject({
                    errorCode: '2',
                    errorMessage: '安卓钉钉版本5.0.15以上支持该api',
                });
            }
        }
        else {
            return Promise.reject({
                errorCode: '3',
                errorMessage: '需要运行在移动端钉钉ios和安卓容器内',
            });
        }
    }
    else {
        return Promise.reject({
            errorCode: '4',
            errorMessage: 'appIcon、appName、和(appUrl|scheme)为必填项',
        });
    }
};
exports.shortcutCustomAdd = shortcutCustomAdd;
