// @ali/dingtalk-jsapi/plugin/environmentV2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDingTalkLite = exports.isMobile = exports.isXueXi = exports.isPCDingTalk = exports.isIOSDingTalk = exports.isIOS = exports.isAndroidDingTalk = exports.isAndroid = exports.isWeixin = exports.isInDingTalk = void 0;
/* tslint:disable:max-line-length */
/**
 * 相较于V1版本，为了性能以及与服务端命名保持一致，此版本更新为
 * 导出 变量 => 方法
 */
var env_1 = require("../lib/env");
var isDingTalkLite_1 = require("@ali/dingtalk-kit/lib/apis/miniProgram/platform/isDingTalkLite");
var isDingTalkLite_2 = require("@ali/dingtalk-kit/lib/apis/webview/platform/isDingTalkLite");
var isDingTalkLite_3 = require("@ali/dingtalk-kit/lib/apis/weex/platform/isDingTalkLite");
var dingtalkEnv = env_1.getENV();
var lowerUA = env_1.getUA().toLocaleLowerCase();
/** 判断是否是在DingTalk容器内 */
var isInDingTalk = function () {
    return dingtalkEnv.platform !== env_1.ENV_ENUM.notInDingTalk;
};
exports.isInDingTalk = isInDingTalk;
/** 判断是否在微信终端 */
var isWeixin = function () {
    return lowerUA.indexOf('micromessenger') !== -1;
};
exports.isWeixin = isWeixin;
/** 判断是否是android终端 */
var isAndroid = function () {
    return lowerUA.indexOf('android') !== -1 || lowerUA.indexOf('adr') !== -1;
};
exports.isAndroid = isAndroid;
/** 判断是否在android Dingtalk环境下 */
var isAndroidDingTalk = function () {
    return dingtalkEnv.platform === env_1.ENV_ENUM.android;
};
exports.isAndroidDingTalk = isAndroidDingTalk;
/** 判断是否是iOS终端 */
var isIOS = function () {
    return !!lowerUA.match(/\(i[^;]+;( U;)? CPU.+Mac OS X/i);
};
exports.isIOS = isIOS;
/** 判断是否在iOS Dingtalk环境下 */
var isIOSDingTalk = function () {
    return dingtalkEnv.platform === env_1.ENV_ENUM.ios;
};
exports.isIOSDingTalk = isIOSDingTalk;
/** 判断是否在PC Dingtalk环境下 */
var isPCDingTalk = function () {
    return dingtalkEnv.platform === env_1.ENV_ENUM.pc;
};
exports.isPCDingTalk = isPCDingTalk;
/** 判断是否是XueXi客户端下 */
var isXueXi = function () {
    return lowerUA.indexOf('device/xuexi') !== -1;
};
exports.isXueXi = isXueXi;
/** 判断是否是移动终端 */
var isMobile = function () {
    return exports.isIOSDingTalk() || exports.isAndroidDingTalk() ||
        /phone|pad|pod|iPhone|iPod|ios|iPad|Android|Mobile|BlackBerry|IEMobile|MQQBrowser|JUC|Fennec|wOSBrowser|BrowserNG|WebOS|Symbian|Windows Phone/i.test(lowerUA);
};
exports.isMobile = isMobile;
/** 判断是否是钉钉国际版 */
var isDingTalkLite = function () {
    return isDingTalkLite_1.isDingTalkLite() || isDingTalkLite_2.isDingTalkLite() || isDingTalkLite_3.isDingTalkLite();
};
exports.isDingTalkLite = isDingTalkLite;
