// @ali/dingtalk-jsapi/plugin/environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isMobile = exports.isXueXi = exports.isPCDingTalk = exports.isIOSDingTalk = exports.isIOS = exports.isAndroidDingTalk = exports.isAndroid = exports.isWeixin = exports.isInDingTalk = void 0;
/* tslint:disable:max-line-length */
/**
 * @deprecated 此插件不再维护，推荐使用 environmentV2 模块
 */
var env_1 = require("../lib/env");
var dingtalkEnv = env_1.getENV();
var lowerUA = env_1.getUA().toLocaleLowerCase();
/** 是否是在DingTalk容器内 */
exports.isInDingTalk = dingtalkEnv.platform !== env_1.ENV_ENUM.notInDingTalk;
/** 是否在微信终端 */
exports.isWeixin = lowerUA.indexOf('micromessenger') !== -1;
/** 是否是android终端 */
exports.isAndroid = lowerUA.indexOf('android') !== -1 || lowerUA.indexOf('adr') !== -1;
/** 是否在android Dingtalk环境下 */
exports.isAndroidDingTalk = dingtalkEnv.platform === env_1.ENV_ENUM.android;
/** 是否是iOS终端 */
exports.isIOS = !!lowerUA.match(/\(i[^;]+;( U;)? CPU.+Mac OS X/i);
/** 是否在iOS Dingtalk环境下 */
exports.isIOSDingTalk = dingtalkEnv.platform === env_1.ENV_ENUM.ios;
/** 是否在PC Dingtalk环境下 */
exports.isPCDingTalk = dingtalkEnv.platform === env_1.ENV_ENUM.pc;
/** 是否是XueXi客户端下 */
exports.isXueXi = lowerUA.indexOf('device/xuexi') !== -1;
/** 是否是移动终端 */
exports.isMobile = exports.isIOSDingTalk || exports.isAndroidDingTalk ||
    /phone|pad|pod|iPhone|iPod|ios|iPad|Android|Mobile|BlackBerry|IEMobile|MQQBrowser|JUC|Fennec|wOSBrowser|BrowserNG|WebOS|Symbian|Windows Phone/i.test(lowerUA);
