// @ali/4ever-plugin-comment/dist-web/utils/hooks/contextMenu.js
import { CommentRole } from "../role";
import { ACTION_COMMENT_CLICK } from "../actions";
import canComment from "../queries/canComment";
export default (function (configs) {
  return function contextMenu(controller, next) {
    var _configs$locale;

    if (!canComment(controller)) {
      return next();
    }

    var isDisabled = (configs == null ? void 0 : configs.isDisabled) && configs.isDisabled();
    var title = (configs == null ? void 0 : (_configs$locale = configs.locale) == null ? void 0 : _configs$locale.comment) || 'Comment';
    var item = {
      name: title,
      action: {
        type: ACTION_COMMENT_CLICK
      },
      key: 'comment',
      icon: 'https://gw.alicdn.com/imgextra/i2/O1CN01GtHgGq1jGfO1gqxmG_!!6000000004521-55-tps-24-24.svg',
      options: {
        group: 100,
        mobile: true,
        hotKey: controller.readOnly ? '' : 'mod+shift+m',
        disable: isDisabled
      },
      role: CommentRole.comment
    };
    return [item].concat(next());
  };
});
