// @ali/4ever-zhi/dist-node/plugins/imTag/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everFactory = require("@ali/4ever-factory");

const _createElement = /*#__PURE__*/_react.default.createElement;
const {
  ImTagContainer,
  IMTAG_MEASURE_STRING
} = _everBamboo.ImTagPlugin;
const createZhiImTagPlugin = (0, _everFactory.zhiTagFactory)(config => {
  const imTagConfig = config.imTag;
  const {
    locale = {
      imFrom: IMTAG_MEASURE_STRING
    },
    jumpToConversition
  } = imTagConfig || {};
  return {
    tagType: 'im',
    renderNode: props => {
      const node = props.node;
      const {
        height,
        width,
        metadata
      } = node.data;
      const {
        imTitle = '',
        cid
      } = metadata || {};

      const onClick = () => {
        if (jumpToConversition && cid) jumpToConversition(cid);
      };

      return /*#__PURE__*/_createElement(ImTagContainer, {
        width: width,
        height: height,
        text: `${locale.imFrom} ${imTitle}`,
        onClick: onClick
      });
    }
  };
});
var _default = createZhiImTagPlugin;
exports.default = _default;
