// @ali/4ever-plugin-comment/dist-node/utils/components/Bubbles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _everCangjie = require("@ali/4ever-cangjie");

var _constants = require("../utils/constants");

var _useDocSize = require("../hooks/useDocSize");

const _createElement = /*#__PURE__*/_react.default.createElement;
const {
  IS_MOBILE
} = _everCangjie.environment;

const Bubbles = /*#__PURE__*/_styledComponents.default.div(["float:right;position:relative;margin-right:-", "px;"], IS_MOBILE ? _constants.BUTTON_SIZE - 2 : _constants.BUTTON_SIZE + 4);

const PHWrapper = /*#__PURE__*/_styledComponents.default.div(["position:absolute;width:28px;bottom:0;right:0;"]);

function _default({
  setBottomMargin,
  updatePosition,
  block,
  children
}) {
  const bubblesRef = _react.default.useRef(null);

  const [phHeight, setHeight] = _react.default.useState(0);

  const [renderKey, updateRenderKey] = _react.default.useReducer(s => s + 1, 0);

  const onDocSizeChange = _react.default.useCallback(() => {
    updatePosition();
    updateRenderKey();
  }, [updateRenderKey, updatePosition]);

  (0, _useDocSize.useDocSize)(bubblesRef, onDocSizeChange);

  _react.default.useEffect(() => {
    const node = bubblesRef.current;
    const prev = bubblesRef.current?.previousElementSibling;

    if (!node || !prev) {
      return;
    }

    const {
      top: nodeTop
    } = node.getBoundingClientRect();
    const {
      bottom: prevBottom,
      height: rawHeight
    } = prev.getBoundingClientRect();
    const extra = Math.round(nodeTop - prevBottom);
    const height = Math.round(rawHeight);

    if (extra) {
      setBottomMargin(extra);
    }

    if (phHeight !== height) {
      setHeight(height);
    }
  }, [block, renderKey]);

  const style = _react.default.useMemo(() => ({
    height: `${phHeight}px`
  }), [phHeight]);

  return /*#__PURE__*/_createElement(Bubbles, {
    className: "data-comment-bubbles",
    "data-cache-ignore": true,
    ref: bubblesRef
  }, !IS_MOBILE && /*#__PURE__*/_createElement(PHWrapper, {
    style: style
  }), children);
}
