// @ali/4ever-plugin-list/dist-node/bi/createContextMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createContextMenu;

var _utils = require("./utils");

var _actions = require("./actions");

function createContextMenu() {
  return function contextMenuPlugin(pluginConfig) {
    const locale = pluginConfig.locale || {};
    return function contextMenu(controller, next) {
      const menus = [];
      const {
        selection
      } = controller.value;

      if (selection?.isCollapsed) {
        const currentItem = (0, _utils.getCurrentItem)(controller)[0];
        const listConfig = currentItem?.data?.list;

        if (listConfig && listConfig.isOrdered) {
          // 有序列表才有下列选项
          menus.push({
            name: locale.restartNumbering,
            action: (0, _actions.restartNumbering)(),
            key: 'list-restart-numbering',
            options: {
              mobile: false,
              group: 2
            }
          }, {
            name: locale.continueNumbering,
            action: (0, _actions.continueNumbering)(),
            key: 'list-continue-numbering',
            options: {
              mobile: false,
              group: 2
            }
          });
        }
      }

      const others = next() || [];
      return menus.concat(others);
    };
  };
}
