// @ali/4ever-mo/dist-node/plugins/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "marks", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "bold", {
  enumerable: true,
  get: function () {
    return _everPluginBold.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "blockquote", {
  enumerable: true,
  get: function () {
    return _everPluginBlockquote.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "br", {
  enumerable: true,
  get: function () {
    return _everPluginBr.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "code", {
  enumerable: true,
  get: function () {
    return _everPluginCode.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "heading", {
  enumerable: true,
  get: function () {
    return _everPluginHeading.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "hr", {
  enumerable: true,
  get: function () {
    return _everPluginHr.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "image", {
  enumerable: true,
  get: function () {
    return _everPluginImage.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "italic", {
  enumerable: true,
  get: function () {
    return _everPluginItalic.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "link", {
  enumerable: true,
  get: function () {
    return _everPluginLink.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "list", {
  enumerable: true,
  get: function () {
    return _everPluginList.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "mention", {
  enumerable: true,
  get: function () {
    return _everPluginMention.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "onlineVideo", {
  enumerable: true,
  get: function () {
    return _everPluginOnlineVideo.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "textBackground", {
  enumerable: true,
  get: function () {
    return _everPluginTextBackground.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "formula", {
  enumerable: true,
  get: function () {
    return _everPluginFormula.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "paragraph", {
  enumerable: true,
  get: function () {
    return _everPluginParagraph.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "placeholder", {
  enumerable: true,
  get: function () {
    return _everPluginPlaceholder.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "inlinePlaceholder", {
  enumerable: true,
  get: function () {
    return _everPluginInlinePlaceholder.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "refer", {
  enumerable: true,
  get: function () {
    return _everPluginRefer.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "sticker", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "strike", {
  enumerable: true,
  get: function () {
    return _everPluginStrike.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "vertAlign", {
  enumerable: true,
  get: function () {
    return _everPluginVertAlign.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "underline", {
  enumerable: true,
  get: function () {
    return _everPluginUnderline.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "table", {
  enumerable: true,
  get: function () {
    return _everPluginTable.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "card", {
  enumerable: true,
  get: function () {
    return _everPluginCard.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "tag", {
  enumerable: true,
  get: function () {
    return _everPluginTag.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "tagWithFormula", {
  enumerable: true,
  get: function () {
    return _everPluginTag.createSerializationWithFormulaPlugin;
  }
});
Object.defineProperty(exports, "embed", {
  enumerable: true,
  get: function () {
    return _everPluginEmbed.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "container", {
  enumerable: true,
  get: function () {
    return _everPluginContainer.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "colorBlock", {
  enumerable: true,
  get: function () {
    return _everPluginColorBlocks.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "templateButton", {
  enumerable: true,
  get: function () {
    return _everPluginTemplateButton.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "refblock", {
  enumerable: true,
  get: function () {
    return _everPluginRefblock.createSerializationPlugin;
  }
});
Object.defineProperty(exports, "inlineCode", {
  enumerable: true,
  get: function () {
    return _everPluginInlineCode.createSerializationPlugin;
  }
});
exports.default = void 0;

var _everPluginMarks = require("@ali/4ever-plugin-marks");

var _everPluginBold = require("@ali/4ever-plugin-bold");

var _everPluginBlockquote = require("@ali/4ever-plugin-blockquote");

var _everPluginBr = require("@ali/4ever-plugin-br");

var _everPluginCode = require("@ali/4ever-plugin-code");

var _everPluginHeading = require("@ali/4ever-plugin-heading");

var _everPluginHr = require("@ali/4ever-plugin-hr");

var _everPluginImage = require("@ali/4ever-plugin-image");

var _everPluginItalic = require("@ali/4ever-plugin-italic");

var _everPluginLink = require("@ali/4ever-plugin-link");

var _everPluginList = require("@ali/4ever-plugin-list");

var _everPluginMention = require("@ali/4ever-plugin-mention");

var _everPluginOnlineVideo = require("@ali/4ever-plugin-online-video");

var _everPluginTextBackground = require("@ali/4ever-plugin-text-background");

var _everPluginFormula = require("@ali/4ever-plugin-formula");

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _everPluginPlaceholder = require("@ali/4ever-plugin-placeholder");

var _everPluginInlinePlaceholder = require("@ali/4ever-plugin-inline-placeholder");

var _everPluginRefer = require("@ali/4ever-plugin-refer");

var _everPluginSticker = require("@ali/4ever-plugin-sticker");

var _everPluginStrike = require("@ali/4ever-plugin-strike");

var _everPluginVertAlign = require("@ali/4ever-plugin-vert-align");

var _everPluginUnderline = require("@ali/4ever-plugin-underline");

var _everPluginTable = require("@ali/4ever-plugin-table");

var _everPluginCard = require("@ali/4ever-plugin-card");

var _everPluginTag = require("@ali/4ever-plugin-tag");

var _everPluginEmbed = require("@ali/4ever-plugin-embed");

var _everPluginContainer = require("@ali/4ever-plugin-container");

var _everPluginColorBlocks = require("@ali/4ever-plugin-color-blocks");

var _everPluginTemplateButton = require("@ali/4ever-plugin-template-button");

var _everPluginRefblock = require("@ali/4ever-plugin-refblock");

var _everPluginInlineCode = require("@ali/4ever-plugin-inline-code");

var _default = {
  marks: _everPluginMarks.createSerializationPlugin,
  bold: _everPluginBold.createSerializationPlugin,
  formula: _everPluginFormula.createSerializationPlugin,
  blockquote: _everPluginBlockquote.createSerializationPlugin,
  code: _everPluginCode.createSerializationPlugin,
  br: _everPluginBr.createSerializationPlugin,
  tag: _everPluginTag.createSerializationPlugin,
  card: _everPluginCard.createSerializationPlugin,
  embed: _everPluginEmbed.createSerializationPlugin,
  heading: _everPluginHeading.createSerializationPlugin,
  hr: _everPluginHr.createSerializationPlugin,
  image: _everPluginImage.createSerializationPlugin,
  italic: _everPluginItalic.createSerializationPlugin,
  colorBlock: _everPluginColorBlocks.createSerializationPlugin,
  textBackground: _everPluginTextBackground.createSerializationPlugin,
  container: _everPluginContainer.createSerializationPlugin,
  link: _everPluginLink.createSerializationPlugin,
  list: _everPluginList.createSerializationPlugin,
  mention: _everPluginMention.createSerializationPlugin,
  onlineVideo: _everPluginOnlineVideo.createSerializationPlugin,
  refer: _everPluginRefer.createSerializationPlugin,
  table: _everPluginTable.createSerializationPlugin,
  strike: _everPluginStrike.createSerializationPlugin,
  inlineCode: _everPluginInlineCode.createSerializationPlugin,
  vertAlign: _everPluginVertAlign.createSerializationPlugin,
  underline: _everPluginUnderline.createSerializationPlugin,
  sticker: _everPluginSticker.createSerializationPlugin,
  placeholder: _everPluginPlaceholder.createSerializationPlugin,
  inlinePlaceholder: _everPluginInlinePlaceholder.createSerializationPlugin,
  refblock: _everPluginRefblock.createSerializationPlugin,
  // paragraph 放在次后
  paragraph: _everPluginParagraph.createSerializationPlugin
};
exports.default = _default;
