// @ali/4ever-mo/dist-node/plugins/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createEmptyParagraph", {
  enumerable: true,
  get: function () {
    return _everCangjie.createEmptyParagraph;
  }
});
Object.defineProperty(exports, "createTextWithString", {
  enumerable: true,
  get: function () {
    return _everCangjie.createTextWithString;
  }
});
Object.defineProperty(exports, "createEmptyText", {
  enumerable: true,
  get: function () {
    return _everCangjie.createEmptyText;
  }
});
Object.defineProperty(exports, "paragraphTags", {
  enumerable: true,
  get: function () {
    return _everUtils.paragraphTags;
  }
});
Object.defineProperty(exports, "markToAttr", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.markToAttr;
  }
});
Object.defineProperty(exports, "marksToAttrs", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.marksToAttrs;
  }
});
Object.defineProperty(exports, "attrsToMarks", {
  enumerable: true,
  get: function () {
    return _everPluginMarks.attrsToMarks;
  }
});
exports.normalizeChildren = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

var _everUtils = require("@ali/4ever-utils");

var _everPluginMarks = require("@ali/4ever-plugin-marks");

const normalizeChildren = _everUtils.normalizeJsonMLChildren;
exports.normalizeChildren = normalizeChildren;
