// @ali/4ever-mo/dist-node/plugins/valueJsonML.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everPluginCard = require("@ali/4ever-plugin-card");

var _everPluginTag = require("@ali/4ever-plugin-tag");

var _everPluginEmbed = require("@ali/4ever-plugin-embed");

var _everPluginCode = require("@ali/4ever-plugin-code");

var _everPluginTable = require("@ali/4ever-plugin-table");

var _everPluginHr = require("@ali/4ever-plugin-hr");

var _everPluginImage = require("@ali/4ever-plugin-image");

var _everPluginLink = require("@ali/4ever-plugin-link");

var _everPluginMention = require("@ali/4ever-plugin-mention");

var _everPluginOnlineVideo = require("@ali/4ever-plugin-online-video");

var _everPluginPlaceholder = require("@ali/4ever-plugin-placeholder");

var _everPluginInlinePlaceholder = require("@ali/4ever-plugin-inline-placeholder");

var _everPluginRefer = require("@ali/4ever-plugin-refer");

var _everPluginSticker = require("@ali/4ever-plugin-sticker");

var _everPluginParagraph = require("@ali/4ever-plugin-paragraph");

var _everPluginHeading = require("@ali/4ever-plugin-heading");

var _everPluginContainer = require("@ali/4ever-plugin-container");

var _everPluginRefblock = require("@ali/4ever-plugin-refblock");

var _everPluginInlineCode = require("@ali/4ever-plugin-inline-code");

var _everPluginMarks = require("@ali/4ever-plugin-marks");

var _default = {
  marks: _everPluginMarks.createServerSerializationPlugin,
  card: _everPluginCard.createServerSerializationPlugin,
  code: _everPluginCode.createServerSerializationPlugin,
  inlineCode: _everPluginInlineCode.createServerSerializationPlugin,
  embed: _everPluginEmbed.createServerSerializationPlugin,
  hr: _everPluginHr.createServerSerializationPlugin,
  image: _everPluginImage.createServerSerializationPlugin,
  link: _everPluginLink.createServerSerializationPlugin,
  mention: _everPluginMention.createServerSerializationPlugin,
  onlineVideo: _everPluginOnlineVideo.createServerSerializationPlugin,
  placeholder: _everPluginPlaceholder.createServerSerializationPlugin,
  inlinePlaceholder: _everPluginInlinePlaceholder.createServerSerializationPlugin,
  refer: _everPluginRefer.createServerSerializationPlugin,
  sticker: _everPluginSticker.createServerSerializationPlugin,
  table: _everPluginTable.createServerSerializationPlugin,
  tag: _everPluginTag.createServerSerializationPlugin,
  container: _everPluginContainer.createServerSerializationPlugin,
  refblock: _everPluginRefblock.createServerSerializationPlugin,
  heading: _everPluginHeading.createServerSerializationPlugin,
  paragraph: _everPluginParagraph.createServerSerializationPlugin
};
exports.default = _default;
