// @ali/we-design/dist-web/components/select/MultiSelectOption.js
import _extends from "@babel/runtime/helpers/extends";

var _SelectedNormalNormal;

/* eslint-disable no-nested-ternary */
import React, { useCallback, useContext, useEffect, useMemo, useState } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { SelectOptionWrap } from "./styled";
import { classnames } from "../../utils";
import { Tooltip } from "../../components/tooltip";
import { createMultiSelectContext } from "./SelectContext";
import { SelectedNormalNormal } from '@ali/we-icon';
var noopWrapperStyle = {};
export var getTitleContent = function getTitleContent(title) {
  if (typeof title === 'string' || !title) {
    return title;
  }

  return title.title;
};
export var getTitlePlacement = function getTitlePlacement(title) {
  if (typeof title === 'string' || !title) {
    return 'right';
  }

  return title.placement;
}; // eslint-disable-next-line max-lines-per-function

var MultiSelectOptionFC = function MultiSelectOptionFC(props) {
  var children = props.children,
      style = props.style,
      disabled = props.disabled,
      disabledTip = props.disabledTip,
      value = props.value,
      onSelect = props.onSelect,
      testid = props.testid,
      className = props.className,
      title = props.title,
      onClick = props.onClick;
  var SelectContext = createMultiSelectContext();
  var context = useContext(SelectContext);

  var _useState = useState(false),
      selected = _useState[0],
      updateSelected = _useState[1];

  if (!context) {
    throw new Error('[we-design] Select.Option 必须为 Select 的子组件');
  }

  useEffect(function () {
    if (value) {
      context.onUpdateOption(value, children);
    }

    return function () {
      if (value) {
        context.onRemoveOption(value);
      }
    }; // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value]);
  useEffect(function () {
    if (value) {
      context.onCacheOptionDisabled(value, !!disabled);
    }

    changeSelected();
    return function () {
      if (value) {
        context.onUncacheOptionDisabled(value);
      }
    }; // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value, disabled]); // 是否选中

  var changeSelected = useCallback(function () {
    if (typeof value === 'undefined') {
      return;
    }

    updateSelected(context.value.indexOf(value) !== -1);
  }, [context.value, value]);
  useEffect(function () {
    changeSelected();
  }, [changeSelected]); // 容器样式

  var optionWrapperStyle = context.optionWrapperStyle || noopWrapperStyle; // 输入过程中是否隐藏

  var hidden = context.isSearching && context.showOptionValues.indexOf(value) === -1; // 键盘事件中高亮

  var hovered = context && context.keyboardAt === value && value !== undefined;
  var showSelectOnRight = context.showSelectOnRight;
  /**
   * 点击选中
   *
   * @return {void}
   */

  var handleClick = useCallback(function (e) {
    if (value !== undefined) {
      context.onClickOption({
        value: value,
        disabled: !!disabled
      });
    } else if (onSelect) {
      onSelect();
    }

    changeSelected();
    onClick && onClick(e);
  }, [value, onSelect, changeSelected, onClick, context, disabled]); // 样式

  var styleCombined = useMemo(function () {
    return _extends({}, optionWrapperStyle, style);
  }, [optionWrapperStyle, style]);
  return hidden ? null : /*#__PURE__*/_createElement(Tooltip, {
    title: disabled && disabledTip ? disabledTip : getTitleContent(title),
    placement: disabled && disabledTip ? 'top' : getTitlePlacement(title)
  }, /*#__PURE__*/_createElement(SelectOptionWrap, {
    style: styleCombined,
    className: classnames(className, {
      selected: selected,
      disabled: disabled,
      hovered: hovered
    }),
    onClick: handleClick,
    "data-testid": testid && "option-" + testid,
    showSelectOnRight: showSelectOnRight
  }, !!context.selectable && !!selected && (_SelectedNormalNormal || (_SelectedNormalNormal = /*#__PURE__*/_createElement(SelectedNormalNormal, {
    className: "check-icon"
  }))), children));
};

MultiSelectOptionFC.defaultProps = {
  style: {}
}; // (MultiSelectOptionFC as React.FC).displayName = 'MultiSelectOption';

var MultiSelectOption = /*#__PURE__*/React.memo(MultiSelectOptionFC);
MultiSelectOption.displayName = 'MultiSelectOption';
export { MultiSelectOption };
