// @ali/we-design/dist-web/components/select/MultiSelectButton.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { MultiSelectInputStyled } from "./styled";

var stopPropagation = function stopPropagation(e) {
  e.stopPropagation();
};

export function MultiSelectButton(props) {
  var locale = props.locale,
      renderTrigger = props.renderTrigger,
      value = props.value,
      children = props.children,
      mode = props.mode,
      visible = props.visible,
      onFocus = props.onFocus,
      onBlur = props.onBlur,
      onEnter = props.onEnter,
      onChange = props.onChange,
      inputRef = props.inputRef,
      testid = props.testid; // 自定义显示区函数优先级最高

  if (typeof renderTrigger === 'function') {
    return renderTrigger(value);
  } // 支持输入的模式


  if (mode === 'search' || mode === 'input') {
    return /*#__PURE__*/_createElement(MultiSelectInputStyled, {
      ref: inputRef,
      className: "fix-box-shadow",
      placeholder: value,
      locale: locale,
      value: value,
      autoComplete: "off",
      readOnly: !visible,
      onFocus: onFocus,
      onChange: onChange,
      onBlur: onBlur,
      onEnter: onEnter,
      onMouseDown: visible ? stopPropagation : undefined,
      onClick: visible ? stopPropagation : undefined,
      "data-testid": testid && "input-" + testid
    });
  }

  return children;
}
