// @ali/we-design/dist-web/components/select/styled.js
import styled, { css } from 'styled-components';
import { Input } from "../../components/input";
import { resetComponent, Ellipsis } from "../../utils/styled.mixin";
import { withTheme } from "../theme";
var iconStyle = /*#__PURE__*/css(["text-transform:none;text-rendering:optimizeLegibility;line-height:0;text-align:center;"]);
export var SelectWrap = withTheme(styled.div(["", " display:inline-block;user-select:none;height:", ";padding:", ";border:1px solid transparent;border-radius:", ";transition:all 0.1s ease-in-out;cursor:pointer;outline:none;&.small{height:", ";}&.large{height:", ";padding:", ";}&:hover{background-color:", ";}&:active,&.actived{background-color:", ";outline:none;}&.bordered{border:1px solid rgba(17,31,44,0.12);}&.errored{border:1px solid #ff523c;}&.disabled{cursor:not-allowed;color:", ";background-color:transparent;&.weicon{color:", ";}&.select-arrow{color:", ";}}&.focused{border:1px solid #3296fa;background-color:#fff;&:hover{background-color:#fff;}}"], resetComponent, function (_ref) {
  var theme = _ref.theme;
  return theme.ui.elementHeight;
}, function (_ref2) {
  var showArrow = _ref2.showArrow;
  return showArrow && '0 4px 0 6px';
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.radius;
}, function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.elementHeightSmall;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.ui.elementHeightLarge;
}, function (_ref6) {
  var showArrow = _ref6.showArrow;
  return showArrow && '0 12px 0 12px';
}, function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.commonBgHoverColor;
}, function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.commonBgPressColor;
}, function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.disabledColor;
}, function (_ref10) {
  var theme = _ref10.theme;
  return theme.ui.disabledColor;
}, function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.disabledColor;
}));
export var StandardSelectWrap = withTheme(styled.div(["", " display:inline-block;user-select:none;height:", ";padding:", ";border:1px solid rgba(17,31,44,0.12);border-radius:", ";background-color:#fff;transition:all 0.1s ease-in-out;cursor:default;outline:none;&.small{height:", ";}&.large{height:", ";padding:", ";}&:active,&.actived{outline:none;}&:hover{border-color:#0089ff;}&.errored{border:1px solid #ff523c;}&.disabled{cursor:not-allowed;border-color:rgba(0,0,0,0.16);color:", ";background-color:#F7F6F6;:hover{border:1px solid rgba(0,0,0,0.16);}&.weicon{color:", ";}&.select-arrow{color:", ";}}&.focused{border:1px solid #0089ff;background-color:#fff;box-shadow:0px 0px 0px 3px rgba(0,106,254,0.12);&:hover{background-color:#fff;}}"], resetComponent, function (_ref12) {
  var theme = _ref12.theme;
  return theme.ui.elementHeight;
}, function (_ref13) {
  var showArrow = _ref13.showArrow;
  return showArrow && '0 4px 0 6px';
}, function (_ref14) {
  var theme = _ref14.theme;
  return theme.ui.radius;
}, function (_ref15) {
  var theme = _ref15.theme;
  return theme.ui.elementHeightSmall;
}, function (_ref16) {
  var theme = _ref16.theme;
  return theme.ui.elementHeightLarge;
}, function (_ref17) {
  var showArrow = _ref17.showArrow;
  return showArrow && '0 12px 0 12px';
}, function (_ref18) {
  var _theme$ui;

  var theme = _ref18.theme;
  return (theme == null ? void 0 : (_theme$ui = theme.ui) == null ? void 0 : _theme$ui.disabledColor) || 'rgba(17, 31, 44, 0.24)';
}, function (_ref19) {
  var _theme$ui2;

  var theme = _ref19.theme;
  return (theme == null ? void 0 : (_theme$ui2 = theme.ui) == null ? void 0 : _theme$ui2.disabledColor) || 'rgba(17, 31, 44, 0.24)';
}, function (_ref20) {
  var _theme$ui3;

  var theme = _ref20.theme;
  return (theme == null ? void 0 : (_theme$ui3 = theme.ui) == null ? void 0 : _theme$ui3.disabledColor) || 'rgba(17, 31, 44, 0.24)';
}));
export var MultiSelectWrapper = /*#__PURE__*/styled.div(["position:relative;"]);
export var DisableMask = /*#__PURE__*/styled.div(["position:absolute;top:0px;left:0px;width:100%;height:100%;cursor:no-drop;"]);
export var MultiSelectWrap = withTheme(styled.div(["", " display:inline-block;user-select:none;height:", ";padding:", ";border:1px solid transparent;border-radius:", ";background-color:#fff;transition:all 0.1s ease-in-out;cursor:default;outline:none;vertical-align:top;&.small{height:", ";}&.large{height:", ";padding:", ";}&:active,&.actived{outline:none;}&.bordered{border:1px solid rgba(17,31,44,0.12);}&:hover{border-color:#0089ff;}&.errored{border:1px solid #ff523c;}&.disabled{cursor:not-allowed;border-color:rgba(0,0,0,0.16);color:", ";background-color:#F7F6F6;:hover{border:1px solid rgba(0,0,0,0.16);}&.weicon{color:", ";}&.select-arrow{color:", ";}}&.focused{border:1px solid #0089ff;background-color:#fff;box-shadow:0px 0px 0px 3px rgba(0,106,254,0.12);&:hover{background-color:#fff;}}"], resetComponent, function (_ref21) {
  var _theme$ui4;

  var theme = _ref21.theme;
  return (theme == null ? void 0 : (_theme$ui4 = theme.ui) == null ? void 0 : _theme$ui4.elementHeight) || '28px';
}, function (_ref22) {
  var showArrow = _ref22.showArrow;
  return showArrow && '0 4px 0 6px';
}, function (_ref23) {
  var _theme$ui5;

  var theme = _ref23.theme;
  return (theme == null ? void 0 : (_theme$ui5 = theme.ui) == null ? void 0 : _theme$ui5.radius) || '4px';
}, function (_ref24) {
  var _theme$ui6;

  var theme = _ref24.theme;
  return (theme == null ? void 0 : (_theme$ui6 = theme.ui) == null ? void 0 : _theme$ui6.elementHeightSmall) || '24px';
}, function (_ref25) {
  var _theme$ui7;

  var theme = _ref25.theme;
  return (theme == null ? void 0 : (_theme$ui7 = theme.ui) == null ? void 0 : _theme$ui7.elementHeightLarge) || '32px';
}, function (_ref26) {
  var showArrow = _ref26.showArrow;
  return showArrow && '0 12px 0 12px';
}, function (_ref27) {
  var _theme$ui8;

  var theme = _ref27.theme;
  return (theme == null ? void 0 : (_theme$ui8 = theme.ui) == null ? void 0 : _theme$ui8.disabledColor) || 'rgba(17, 31, 44, 0.24)';
}, function (_ref28) {
  var _theme$ui9;

  var theme = _ref28.theme;
  return (theme == null ? void 0 : (_theme$ui9 = theme.ui) == null ? void 0 : _theme$ui9.disabledColor) || 'rgba(17, 31, 44, 0.24)';
}, function (_ref29) {
  var _theme$ui10;

  var theme = _ref29.theme;
  return (theme == null ? void 0 : (_theme$ui10 = theme.ui) == null ? void 0 : _theme$ui10.disabledColor) || 'rgba(17, 31, 44, 0.24)';
}));
export var TriggerWrap = withTheme(styled.div(["position:relative;height:100%;display:flex;align-items:center;justify-content:space-between;.weicon{display:block;color:", ";}.select-arrow{", " color:", ";flex-shrink:0;flex-grow:0;font-size:12px;transform:rotate(0deg);transform-origin:center center;margin-left:2px;&.opened{transform:rotate(180deg);}&.hidden{display:none;}}.multiple-select-overflow{position:relative;display:flex;flex:auto;flex-wrap:wrap;max-width:100%;&-item{flex:none;max-width:100%;}&-item-suffix{flex:1;position:unset;}}.multiple-select-item{font-size:13px;height:22px;background:#EBECED;border-radius:2px;margin:3px 3px 3px 0;padding:4px;display:flex;align-items:center;.multiple-select-item-content{margin-right:3px;}.multiple-select-item-remove{color:rgba(23,26,29,0.4);}}"], function (_ref30) {
  var theme = _ref30.theme;
  return theme.ui.commonActionIconColor;
}, iconStyle, function (_ref31) {
  var theme = _ref31.theme;
  return theme.ui.commonBaseColor2;
}));
export var TriggerContent = withTheme(styled.div(["", ";flex-grow:1;flex-shrink:1;"], Ellipsis));
var scroll = /*#__PURE__*/css(["&::-webkit-scrollbar{width:0px;}&:hover::-webkit-scrollbar{width:6px;}&::-webkit-scrollbar-track{}&::-webkit-scrollbar-thumb{background-color:", ";border-radius:6px;&:hover{background-color:", ";}&:active{background-color:", ";}}"], function (p) {
  var _p$theme, _p$theme$color;

  return ((_p$theme = p.theme) == null ? void 0 : (_p$theme$color = _p$theme.color) == null ? void 0 : _p$theme$color.scrollNormal) || 'rgba(199, 200, 201, 1)';
}, function (p) {
  var _p$theme2, _p$theme2$color;

  return ((_p$theme2 = p.theme) == null ? void 0 : (_p$theme2$color = _p$theme2.color) == null ? void 0 : _p$theme2$color.scrollHover) || 'rgba(162, 163, 165, 1)';
}, function (p) {
  var _p$theme3, _p$theme3$color;

  return ((_p$theme3 = p.theme) == null ? void 0 : (_p$theme3$color = _p$theme3.color) == null ? void 0 : _p$theme3$color.scrollClick) || 'rgba(116, 118, 119, 1)';
});
export var OverlayWrap = /*#__PURE__*/styled.div(["max-height:400px;overflow-y:overlay;", ";padding:4px 0;"], scroll);
export var SelectOptionWrap = withTheme(styled.div(["", " position:relative;display:flex;justify-content:space-between;align-items:center;padding:", " ", " ", " ", ";margin:0 4px;height:100%;min-width:100px;min-height:32px;font-size:13px;line-height:20px;white-space:nowrap;cursor:pointer;user-select:none;border-radius:4px;&:hover,&.hovered{background-color:", ";}&.selected{.check-icon{display:inline-block;}}&.disabled{color:", ";background-color:transparent;cursor:not-allowed;.check-icon{color:", ";background-color:transparent;cursor:not-allowed;}}.check-icon{", " color:", ";font-size:20px;display:none;color:rgb(23,26,29);position:absolute;left:", ";right:", ";top:50%;transform:translateY(-50%);}"], resetComponent, function (_ref32) {
  var theme = _ref32.theme;
  return theme.ui.paddingSmall;
}, function (_ref33) {
  var showSelectOnRight = _ref33.showSelectOnRight;
  return showSelectOnRight ? '28px' : '20px';
}, function (_ref34) {
  var theme = _ref34.theme;
  return theme.ui.paddingSmall;
}, function (_ref35) {
  var showSelectOnRight = _ref35.showSelectOnRight;
  return showSelectOnRight ? '8px' : '28px';
}, function (_ref36) {
  var theme = _ref36.theme;
  return theme.ui.backgroundColor;
}, function (_ref37) {
  var theme = _ref37.theme;
  return theme.ui.disabledColor;
}, function (_ref38) {
  var theme = _ref38.theme;
  return theme.ui.disabledColor;
}, iconStyle, function (_ref39) {
  var theme = _ref39.theme;
  return theme.ui.primaryColor;
}, function (_ref40) {
  var showSelectOnRight = _ref40.showSelectOnRight;
  return showSelectOnRight ? 'unset' : '4px';
}, function (_ref41) {
  var showSelectOnRight = _ref41.showSelectOnRight;
  return showSelectOnRight ? '4px' : 'unset';
}));
export var InputStyled = withTheme(styled(Input)(["border:0;background-color:transparent;color:inherit;font-size:inherit;line-height:inherit;padding-left:0;padding-right:0;&.fix-box-shadow{&.focused{box-shadow:none;}}& > input{font-size:inherit;color:inherit;line-height:inherit;}"]));
export var MultiSelectInputStyled = withTheme(styled(Input)(["border:0;background-color:transparent;color:inherit;font-size:inherit;line-height:inherit;padding-left:0;padding-right:0;&.fix-box-shadow{&.focused{box-shadow:none;}}& > input{font-size:inherit;color:inherit;line-height:inherit;padding:0;}"]));
export var EmptyWrap = withTheme(styled.div(["padding:60px 24px;"]));
export var OptionGroupWrap = withTheme(styled.div(["position:relative;&:after{display:block;content:' ';height:1px;background-color:", ";margin:4px 4px;}&:before{display:block;content:' ';height:1px;background-color:", ";width:80%;margin:6px auto;}&:nth-child(1){&:before{display:none;}}&:last-child{&:after{display:none;}}~ .select-group-wrapper{&:before{display:none;}}"], function (_ref42) {
  var theme = _ref42.theme;
  return theme.ui.commonLineLightColor;
}, function (_ref43) {
  var theme = _ref43.theme;
  return theme.ui.commonLineLightColor;
}));
export var ErrorTip = /*#__PURE__*/styled.div(["font-size:12px;line-height:18px;margin-top:4px;color:#ff5219;"]);
