// @ali/4ever-plugin-embed/dist-node/utils/components/audio/audioPreview.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = AudioPreview;

var _react = _interopRequireDefault(require("react"));

var _embedPreviewHeader = _interopRequireDefault(require("../common/embedPreviewHeader"));

var _embedDownloadButton = _interopRequireDefault(require("../common/embedDownloadButton"));

var _audioPlayer = _interopRequireDefault(require("./audioPlayer"));

var _fileTypeHelper = require("../../fileTypeHelper");

var _styled = require("./styled");

const _createElement = /*#__PURE__*/_react.default.createElement;

function AudioPreview(props) {
  const {
    onMediaLoadSuccess,
    onMediaLoadError,
    transformAudioURL,
    locale,
    allowDownload,
    name,
    size,
    src,
    onDownloadFile,
    isMobile
  } = props;
  return /*#__PURE__*/_createElement(_styled.AudioWrapper, {
    "data-type": "embed-audio"
  }, /*#__PURE__*/_createElement(_embedPreviewHeader.default, {
    type: _fileTypeHelper.FileTypeMap.Audio,
    name: name,
    size: size,
    renderExtra: () => /*#__PURE__*/_createElement(_embedDownloadButton.default, {
      title: locale?.title,
      onDownloadFile: onDownloadFile,
      allowDownload: allowDownload
    }),
    isMobile: isMobile
  }), /*#__PURE__*/_createElement(_audioPlayer.default, {
    transformAudioURL: transformAudioURL,
    onMediaLoadSuccess: onMediaLoadSuccess,
    onMediaLoadError: onMediaLoadError,
    src: src,
    isMobile: isMobile
  }));
}
