// @ali/4ever-plugin-embed/dist-node/utils/components/audio/audioPlayer.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

const _createElement = /*#__PURE__*/_react.default.createElement;

function AudioPlayer(props) {
  const {
    src,
    transformAudioURL,
    onMediaLoadError,
    onMediaLoadSuccess,
    isMobile
  } = props;
  const trueSrc = transformAudioURL ? transformAudioURL(src) : src;
  const handleAudioClick = (0, _react.useCallback)(e => {
    if (isMobile) {
      e.stopPropagation();
    }
  }, [isMobile]);
  return /*#__PURE__*/_createElement("audio", {
    controls: true,
    controlsList: "nodownload nofullscreen noremoteplayback",
    onCanPlay: onMediaLoadSuccess,
    onError: onMediaLoadError,
    onClick: handleAudioClick,
    onMouseDown: handleAudioClick
  }, /*#__PURE__*/_createElement("source", {
    src: trueSrc,
    type: "audio/mp3"
  }));
}

var _default = AudioPlayer;
exports.default = _default;
