// @ali/4ever-bi/dist-web/plugins/calendarCard/handlers/createMenu.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React, { useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { MenuItem } from '@ali/we-design';
import { insertCard } from '@ali/4ever-factory';

var createMenubar = function createMenubar(calendarConfig) {
  return function CalendarCardMenu(props) {
    var menuItem = props.menuItem,
        controller = props.controller;

    var _ref = calendarConfig.locale || {},
        subjectPlaceholder = _ref.subjectPlaceholder,
        timePlaceholder = _ref.timePlaceholder,
        participantPlaceholder = _ref.participantPlaceholder;

    var title = menuItem.title,
        rest = _objectWithoutPropertiesLoose(menuItem, ["title"]);

    var handleClick = useCallback(function () {
      controller.run('onAction', insertCard({
        cardType: 'calendar',
        height: 102,
        metadata: {
          isPlaceholder: true,
          subject: subjectPlaceholder,
          time: timePlaceholder,
          topReceivers: [{
            isReceiverPlaceholder: true,
            name: participantPlaceholder,
            type: 2
          }]
        }
      }));
    }, [controller]);
    return /*#__PURE__*/_createElement(MenuItem, _extends({}, rest, {
      onClick: handleClick
    }), title || '');
  };
};

export default createMenubar;
