// @ali/4ever-plugin-columns/dist-node/utils/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SIMPLE_HIGHLIGHTS = exports.DEFAULT_SPLIT = exports.STICK_WIDTH = exports.MIN_WIDTH = exports.DEAFULT_VALIGN = exports.COLUMN_SPACE = exports.HORIZ_PADDING = exports.CURSOR_SIZE = void 0;
// 用户的鼠标中间位置调整。用法请参考使用的场景，不建议修改此值。
const CURSOR_SIZE = 5; // 每一列的左右 padding。
// 最左一列左边、最右一列右边，如果没有背景色，就不要这个 padding，不然会显示文字对不齐。

exports.CURSOR_SIZE = CURSOR_SIZE;
const HORIZ_PADDING = 7; // 列间距。

exports.HORIZ_PADDING = HORIZ_PADDING;
const COLUMN_SPACE = 20; // 默认的对齐方式。

exports.COLUMN_SPACE = COLUMN_SPACE;
const DEAFULT_VALIGN = 'top'; // 每一个列的最小宽度。

exports.DEAFULT_VALIGN = DEAFULT_VALIGN;
const MIN_WIDTH = 50; // 拖拽粘性的容忍度。值越大，越容易触发分割杆的粘住。

exports.MIN_WIDTH = MIN_WIDTH;
const STICK_WIDTH = 6; // 分割杆的粘性比例点。当拖拽左侧的元素，达到对应的百分比时，会粘到对应的位置。

exports.STICK_WIDTH = STICK_WIDTH;
const DEFAULT_SPLIT = [0.1, 0.2, 0.25, 0.3, 1 / 3, 0.4, 0.5, 0.6, 2 / 3, 0.7, 0.75, 0.8, 0.9];
exports.DEFAULT_SPLIT = DEFAULT_SPLIT;
const SIMPLE_HIGHLIGHTS = ['#E6F3FF', '#FBF3DB', '#ECF9F0', '#FAEBDD', '#EAE4F2'];
exports.SIMPLE_HIGHLIGHTS = SIMPLE_HIGHLIGHTS;
