// @ali/4ever-zhi/dist-web/plugins/index.js
import { DocStylePlugin as docStyle } from '@ali/4ever-bamboo';
import blockquote from "./blockquote";
import { zhiFactory as textBackground } from '@ali/4ever-plugin-text-background';
import { zhiFactory as code } from '@ali/4ever-plugin-code';
import { zhiFactory as marks } from '@ali/4ever-plugin-marks';
import { zhiFactory as inlineCode } from '@ali/4ever-plugin-inline-code';
import { zhiFactory as table } from '@ali/4ever-plugin-table';
import { zhiFactory as contextMenu } from '@ali/4ever-plugin-pc-contex-menu';
import embed from "./embed";
import heading from "./heading";
import hr from "./hr";
import image from "./image";
import link from "./link";
import linkCard from "./link/linkCard/index";
import audio from "./audio";
import video from "./video";
import { zhiFactory as mention } from '@ali/4ever-plugin-mention';
import onlineVideo from "./onlineVideo";
import placeholder from "./placeholder";
import refer from "./refer";
import { zhiFactory as list } from '@ali/4ever-plugin-list';
import { zhiFactory as paragraph } from '@ali/4ever-plugin-paragraph';
import sticker from "./sticker";
import voteCard from "./voteCard";
import { zhiFactory as br } from '@ali/4ever-plugin-br';
import { zhiFactory as inlinePlaceholder } from '@ali/4ever-plugin-inline-placeholder';
import { zhiFactory as templateButton } from '@ali/4ever-plugin-template-button';
export default {
  br: br,
  blockquote: blockquote,
  audio: audio,
  video: video,
  embed: embed,
  code: code,
  hr: hr,
  image: image,
  link: link,
  linkCard: linkCard,
  textBackground: textBackground,
  marks: marks,
  inlineCode: inlineCode,
  templateButton: templateButton,
  mention: mention,
  placeholder: placeholder,
  heading: heading,
  table: table,
  onlineVideo: onlineVideo,
  refer: refer,
  contextMenu: contextMenu,
  list: list,
  inlinePlaceholder: inlinePlaceholder,
  paragraph: paragraph,
  sticker: sticker,
  voteCard: voteCard,
  // 放在其他插件之后兜底处理默认样式
  docStyle: docStyle
};
