// @ali/4ever-zhi/dist-web/plugins/mobile.js
import { DocStylePlugin as docStyle } from '@ali/4ever-bamboo';
import audio from "./audio/mobile";
import blockquote from "./blockquote";
import { zhiFactory as textBackground } from "@ali/4ever-plugin-text-background";
import { mobileZhiFactory as code } from '@ali/4ever-plugin-code';
import { mobileZhiFactory as inlineCode } from '@ali/4ever-plugin-inline-code';
import { zhiFactory as marks } from '@ali/4ever-plugin-marks';
import { mobileZhiFactory as table } from '@ali/4ever-plugin-table';
import embed from "./embed/mobile";
import heading from "./heading";
import hr from "./hr";
import image from "./image/mobile";
import link from "./link/mobile";
import linkCard from "./link/linkCard/mobile";
import { mobileZhiFactory as paragraph } from '@ali/4ever-plugin-paragraph';
import { mobileZhiFactory as mention } from '@ali/4ever-plugin-mention';
import onlineVideo from "./onlineVideo";
import placeholder from "./placeholder";
import refer from "./refer";
import sticker from "./sticker";
import { mobileZhiFactory as list } from '@ali/4ever-plugin-list';
import video from "./video/mobile";
import voteCard from "./voteCard/mobile";
import { zhiFactory as br } from '@ali/4ever-plugin-br';
import { zhiFactory as inlinePlaceholder } from '@ali/4ever-plugin-inline-placeholder';
import { zhiFactory as templateButton } from '@ali/4ever-plugin-template-button';
export default {
  br: br,
  blockquote: blockquote,
  video: video,
  embed: embed,
  code: code,
  hr: hr,
  image: image,
  link: link,
  inlineCode: inlineCode,
  linkCard: linkCard,
  textBackground: textBackground,
  marks: marks,
  templateButton: templateButton,
  mention: mention,
  heading: heading,
  placeholder: placeholder,
  table: table,
  onlineVideo: onlineVideo,
  refer: refer,
  list: list,
  sticker: sticker,
  audio: audio,
  inlinePlaceholder: inlinePlaceholder,
  paragraph: paragraph,
  // 放在其他插件之后兜底处理默认样式
  docStyle: docStyle,
  voteCard: voteCard
};
