// @ali/4ever-plugin-list/dist-web/bi/queries/index.js
import isEmpty from "./isEmpty";
import isTaskList from "./isTaskList";
import isOrderedList from "./isOrderedList";
import isUnorderedList from "./isUnorderedList";
import hasListSelection from "./hasListSelection";
import getListStyleText from "./getListStyleText";
import { isSelectionInList } from "./isSelectionInList";
import { isSelectionInListSymbol } from "./isSelectionInListSymbol";
import { isSelectionInTaskList } from "./isSelectionInTaskList";
import { getListCurrentItem } from "./getListCurrentItem";
import { getListChildItems } from "./getListChildItems";
import { isListSymbolSelected } from "./isListSymbolUnSelected";
import { getListDepthItems } from "./getListDepthItems";
import { getSymbolStyleByKey } from "./getSymbolStyleByKey";
export var queries = {
  isEmpty: isEmpty,
  isTaskList: isTaskList,
  isOrderedList: isOrderedList,
  isSelectionInList: isSelectionInList,
  isSelectionInListSymbol: isSelectionInListSymbol,
  isSelectionInTaskList: isSelectionInTaskList,
  isUnorderedList: isUnorderedList,
  hasListSelection: hasListSelection,
  getListStyleText: getListStyleText,
  getListCurrentItem: getListCurrentItem,
  getListChildItems: getListChildItems,
  isListSymbolSelected: isListSymbolSelected,
  getListDepthItems: getListDepthItems,
  getSymbolStyleByKey: getSymbolStyleByKey
};
