// @ali/we-design-next/dist-node/display/tabs/Tabs.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tabs = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _TabsBar = require("./TabsBar");
var _TabsContent = require("./TabsContent");
var _common = require("../../common");
const _createElement = /*#__PURE__*/_react.default.createElement;
const TabsWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;flex-direction:column;"]);
const getTabBarTabList = panes => {
  const tabList = [];
  _react.default.Children.forEach(panes, child => {
    if (child && /*#__PURE__*/_react.default.isValidElement(child) && child.type.displayName === 'TabPanel') {
      const {
        title,
        onClose,
        disabled,
        prefix
      } = child.props;
      tabList.push({
        key: child.key,
        title,
        onClose,
        disabled,
        prefix
      });
    }
  });
  return tabList;
};
const Tabs = /*#__PURE__*/_react.default.memo( /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  var _tabs$;
  const {
    children,
    onActiveTabChange,
    activeKey: activeKeyProp,
    defaultActiveKey,
    animated,
    keepMounted = true,
    ...rest
  } = props;
  const tabs = _react.default.useMemo(() => getTabBarTabList(children), [children]);
  const [activeKey, setActive] = (0, _common.useDefaultProp)(defaultActiveKey || ((_tabs$ = tabs[0]) === null || _tabs$ === void 0 ? void 0 : _tabs$.key), activeKeyProp, onActiveTabChange);
  return /*#__PURE__*/_createElement(TabsWrapper, (0, _extends2.default)({
    ref: ref
  }, rest), /*#__PURE__*/_createElement(_TabsBar.TabsBar, {
    tabs: tabs,
    onActiveTabChange: setActive,
    activeKey: activeKey,
    defaultActiveKey: defaultActiveKey,
    animated: animated
  }), /*#__PURE__*/_createElement(_TabsContent.TabsContent, {
    activeKey: activeKey,
    animated: animated,
    destroyInactiveTabPanel: !keepMounted
  }, children));
}));
exports.Tabs = Tabs;
