// @ali/we-design-next/dist-node/display/tabs/TabsContent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TabsContent = void 0;
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _common = require("../../common");
const _createElement = /*#__PURE__*/_react.default.createElement;
const TabsContentHolder = /*#__PURE__*/_styledComponents.default.div(["display:flex;position:relative;flex:auto;min-width:0;min-height:0;overflow:", ";"], p => p.animated ? 'hidden' : 'initial');
const TabsContentWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;width:100%;"]);
const TabPanel = /*#__PURE__*/_styledComponents.default.div(["flex:none;width:100%;&:not(.wdn-animated-tab-panel).wdn-inactive-tab-panel{display:none;}&.wdn-animated-tab-panel.wdn-inactive-tab-panel{height:0;overflow-y:hidden;visibility:hidden;}"]);
const TabsContent = props => {
  const {
    children,
    activeKey,
    animated,
    destroyInactiveTabPanel
  } = props;
  const newChildren = [];
  let index = 0;
  let activeIndex = 0;
  _react.default.Children.forEach(children, child => {
    if (!child || ! /*#__PURE__*/_react.default.isValidElement(child)) {
      return;
    }
    const isActive = activeKey === child.key;
    if (isActive) {
      activeIndex = index;
    }
    newChildren.push( /*#__PURE__*/_createElement(TabPanel, {
      className: (0, _common.mergeClassName)(!isActive ? 'wdn-inactive-tab-panel' : '', animated ? 'wdn-animated-tab-panel' : ''),
      key: child.key
    }, isActive || !destroyInactiveTabPanel ? child : null));
    index++;
  });
  const tabsContentStyle = animated ? {
    marginLeft: `-${activeIndex * 100}%`,
    transition: 'margin 0.3s ease'
  } : undefined;
  return /*#__PURE__*/_createElement(TabsContentHolder, {
    animated: animated
  }, /*#__PURE__*/_createElement(TabsContentWrapper, {
    style: tabsContentStyle
  }, newChildren));
};
exports.TabsContent = TabsContent;
