// @ali/4ever-bi/dist-node/plugins/link/queries/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.queries = void 0;

var _everBamboo = require("@ali/4ever-bamboo");

var _getText = _interopRequireDefault(require("./getText"));

var _isLinkCardSelected = _interopRequireDefault(require("./isLinkCardSelected"));

var _isSelectionInLink = _interopRequireDefault(require("./isSelectionInLink"));

var _isSelectionInLinkCard = _interopRequireDefault(require("./isSelectionInLinkCard"));

var _getSelectedLinkUrl = require("./getSelectedLinkUrl");

const queries = {
  getText: _getText.default,
  getLinkSz: _everBamboo.LinkPlugin.getLinkSz,
  getLinkMaxSz: _everBamboo.LinkPlugin.getLinkMaxSz,
  getSelectedLinkUrl: _getSelectedLinkUrl.getSelectedLinkUrl,
  isLinkCardSelected: _isLinkCardSelected.default,
  getMinTableColWidth: _everBamboo.LinkPlugin.getMinTableColWidth,
  isSelectionInLink: _isSelectionInLink.default,
  isSelectionInLinkCard: _isSelectionInLinkCard.default
};
exports.queries = queries;
