// @ali/we-toolbar/dist-web/biz/footer/biz/ratioSelect/RatioSelect.js
var _FooterReduceNormal, _FooterEnlargeNormal;

import React, { useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { FooterReduceNormal, FooterEnlargeNormal, Tooltip, Select, SelectOption, SelectOptionGroup, useControlledValue, KEY_CODES, classnames } from '@ali/we-design';
import { useTranslate } from '@ali/we-util';
import { RatioSelectWrap, RatioSelectInner, Button, SelectWrap, TriggerSpanWrap } from "./styled";
var RATIOS = [0.1, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2];
var triggerStyle = {
  width: 38,
  textAlign: 'center',
  borderRadius: 0,
  fontSize: 13,
  height: '100%'
}; // eslint-disable-next-line max-lines-per-function

export var RatioSelect = /*#__PURE__*/React.memo(function (props) {
  var locale = props.locale,
      onVisibleChange = props.onVisibleChange,
      onChange = props.onChange,
      step = props.step,
      zIndex = props.zIndex,
      placement = props.placement,
      disabled = props.disabled,
      customedList = props.customedList,
      customRatios = props.customRatios,
      animationPlacement = props.animationPlacement,
      zoomInShortcutText = props.zoomInShortcutText,
      zoomOutShortcutText = props.zoomOutShortcutText; // 国际化文案

  var t = useTranslate(locale);
  var placeholderTitle = t('we_toolbar_ratio_select_placeholder');
  var zoomInTitle = t('we_toolbar_ratio_select_zoom_in');
  var zoomOutTitle = t('we_toolbar_ratio_select_zoom_out');
  var addText = t('we_toolbar_ratio_select_add');
  var reduceText = t('we_toolbar_ratio_select_reduce'); // visible 受控

  var _useControlledValue = useControlledValue(props.visible, props.defaultVisible),
      visible = _useControlledValue[0],
      setVisible = _useControlledValue[1]; // value 受控


  var _useControlledValue2 = useControlledValue(props.value, props.defaultValue),
      value = _useControlledValue2[0],
      setValue = _useControlledValue2[1];

  var ratios = customRatios || RATIOS;
  /**
   * 可见性变化
   */

  var handleVisibleChange = useCallback(function (e) {
    if (onVisibleChange) {
      onVisibleChange(e);
    }

    setVisible(e);
  }, [setVisible, onVisibleChange]);
  /**
   * 值发生变化
   */

  var handleValueChange = useCallback(function (e) {
    if (onChange) {
      onChange(e);
    }

    setValue(e);
  }, [onChange, setValue]);
  /**
   * 减少比例
   */

  var handleReduce = useCallback(function () {
    if (disabled) return; // 浮点数精度问题消除

    var newValue = Math.round(value * 100) - Math.round(step * 100);

    if (newValue >= ratios[0] * 100) {
      handleValueChange(newValue / 100);
    } else {
      handleValueChange(ratios[0]);
    }
  }, [value, handleValueChange, step]);
  /**
   * 增加比例
   */

  var handleAdd = useCallback(function () {
    if (disabled) return; // 浮点数精度问题消除

    var newValue = Math.round(value * 100) + Math.round(step * 100);

    if (newValue <= ratios[ratios.length - 1] * 100) {
      handleValueChange(newValue / 100);
    } else {
      handleValueChange(ratios[ratios.length - 1]);
    }
  }, [value, handleValueChange, step]);
  /**
   * 如果是focus到减少按钮并回车
   */

  var handleMinusKeyDown = useCallback(function (e) {
    if (e.which === KEY_CODES.ENTER) {
      handleReduce();
    }
  }, [handleReduce]);
  /**
   * 如果是focus到增加按钮并回车
   */

  var handleAddKeyDown = useCallback(function (e) {
    if (e.which === KEY_CODES.ENTER) {
      handleAdd();
    }
  }, [handleAdd]);
  /**
   * 不是根据select来渲染
   */

  var renderRatio = useCallback(function (e) {
    return /*#__PURE__*/_createElement(TriggerSpanWrap, null, Math.round(e * 100), "%");
  }, []);
  return /*#__PURE__*/_createElement(RatioSelectWrap, null, /*#__PURE__*/_createElement(RatioSelectInner, null, /*#__PURE__*/_createElement(Tooltip, {
    title: zoomOutTitle + ("" + (zoomOutShortcutText ? " " + zoomOutShortcutText : ''))
  }, /*#__PURE__*/_createElement(Button, {
    tabIndex: 0,
    onKeyDown: handleMinusKeyDown,
    onClick: handleReduce,
    className: classnames({
      disabled: value === ratios[0] || disabled
    }),
    "data-testid": "RatioSelect_minus",
    "data-role": "footerZoomOut"
  }, _FooterReduceNormal || (_FooterReduceNormal = /*#__PURE__*/_createElement(FooterReduceNormal, null)))), /*#__PURE__*/_createElement(SelectWrap, null, /*#__PURE__*/_createElement(Select, {
    size: "small",
    value: value,
    visible: visible,
    onChange: handleValueChange,
    renderTrigger: renderRatio,
    onVisibleChange: handleVisibleChange,
    zIndex: zIndex,
    style: triggerStyle,
    showArrow: false,
    placement: placement,
    title: placeholderTitle,
    disabled: disabled,
    testid: "RatioSelect",
    animationPlacement: animationPlacement
  }, /*#__PURE__*/_createElement(SelectOptionGroup, null, ratios.map(function (ratio) {
    return /*#__PURE__*/_createElement(SelectOption, {
      key: "" + ratio,
      value: ratio,
      testid: "RatioSelect_" + ratio,
      role: "footerRatioSelectOption" + ratio
    }, ratio * 100, "%");
  })), customedList ? /*#__PURE__*/_createElement(SelectOptionGroup, null, customedList.map(function (item) {
    return /*#__PURE__*/_createElement(SelectOption, {
      key: "" + item.text,
      value: item.value,
      testid: "RatioSelect_" + item.text,
      role: "footerRatioSelectOption" + item.text
    }, item.text);
  })) : /*#__PURE__*/_createElement(React.Fragment, null), /*#__PURE__*/_createElement(SelectOption, {
    disabled: value >= ratios[ratios.length - 1],
    onSelect: handleAdd,
    testid: "RatioSelect_add"
  }, addText, step * 100, "%"), /*#__PURE__*/_createElement(SelectOption, {
    onSelect: handleReduce // @ts-ignore
    ,
    disabled: value <= ratios[0],
    testid: "RatioSelect_reduce"
  }, reduceText, step * 100, "%"))), /*#__PURE__*/_createElement(Tooltip, {
    title: zoomInTitle + ("" + (zoomInShortcutText ? " " + zoomInShortcutText : ''))
  }, /*#__PURE__*/_createElement(Button, {
    tabIndex: 0,
    onKeyDown: handleAddKeyDown,
    onClick: handleAdd,
    className: classnames({
      disabled: value === ratios[ratios.length - 1] || disabled
    }),
    "data-testid": "RatioSelect_plus",
    "data-role": "footerZoomIn"
  }, _FooterEnlargeNormal || (_FooterEnlargeNormal = /*#__PURE__*/_createElement(FooterEnlargeNormal, null))))));
});
RatioSelect.defaultProps = {
  defaultVisible: false,
  step: 0.05,
  defaultValue: 1,
  placement: 'top',
  disabled: false
};
