// @ali/we-design/dist-web/components/mobile-bottom-curtain/index.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import ReactDOM from 'react-dom';
import Curtain from "./Curtain";

var Ghost = /*#__PURE__*/function () {
  function Ghost() {
    this.ghost = null;
  }

  var _proto = Ghost.prototype;

  _proto.init = function init() {
    if (!this.ghost) {
      this.ghost = document.createElement('div');
      document.body.appendChild(this.ghost);
    }
  };

  _proto.getGhost = function getGhost() {
    return this.ghost;
  };

  _proto.removeGhost = function removeGhost() {
    this.ghost && document.body.removeChild(this.ghost);
    this.ghost = null;
  };

  return Ghost;
}();

var MobileBottomCurtain = function MobileBottomCurtain(props) {
  var container = props.container,
      children = props.children,
      extraProps = _objectWithoutPropertiesLoose(props, ["container", "children"]);

  var ghostRef = React.useRef(new Ghost());

  if (!container) {
    ghostRef.current.init();
  }

  React.useEffect(function () {
    return function () {
      ghostRef.current.removeGhost();
    };
  }, []);
  return /*#__PURE__*/ReactDOM.createPortal( /*#__PURE__*/_createElement(Curtain, extraProps, children), container || ghostRef.current.getGhost());
};

export { MobileBottomCurtain };
