// @ali/4ever-plugin-table/dist-node/bi/components/cornerToolbar/toolbarMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _weDesign = require("@ali/we-design");

const _createElement = /*#__PURE__*/React.createElement;

const Button = /*#__PURE__*/_styledComponents.default.span(["position:relative;display:flex;align-items:center;justify-content:center;color:rgba(17,31,44,0.72);margin:0;padding:0;width:28px;height:28px;font-size:16px;cursor:pointer;line-height:16px;border-radius:4px;&:hover{background-color:rgba(17,31,44,0.08);}"]);

var _ref = /*#__PURE__*/_createElement("span", null, /*#__PURE__*/_createElement(_weDesign.DeleteBetaNormal, null));

class ToolbarMenu extends React.Component {
  constructor(...args) {
    super(...args);

    this.handleMouseDown = e => {
      e.preventDefault(); // Prevent blur.

      e.stopPropagation();
      const {
        onMenuEvent
      } = this.props;
      onMenuEvent('delete-table');
    };

    this.highlightSelection = () => {
      const {
        setIsHighlightSelection
      } = this.props;
      setIsHighlightSelection(true);
    };

    this.cancelHighlightSelection = () => {
      const {
        setIsHighlightSelection
      } = this.props;
      setIsHighlightSelection(false);
    };
  }

  render() {
    const {
      style,
      locale = {}
    } = this.props;
    return /*#__PURE__*/_createElement(Button, {
      style: style,
      onMouseDown: this.handleMouseDown,
      onTouchEnd: this.handleMouseDown,
      onMouseOver: this.highlightSelection,
      onMouseLeave: this.cancelHighlightSelection
    }, /*#__PURE__*/_createElement(_weDesign.Tooltip, {
      placement: "top",
      title: locale.contextMenuDeleteTable
    }, _ref));
  }

}

exports.default = ToolbarMenu;
