// @ali/4ever-bamboo/dist-web/plugins/index.js
import _extends from "@babel/runtime/helpers/extends";
import { OldBambooPlugin as CalendarCardPlugin } from '@ali/4ever-plugin-calendar-card';
import { OldBambooPlugin as CardPlugin } from '@ali/4ever-plugin-card';
import { OldBambooPlugin as VoteCardPlugin } from '@ali/4ever-plugin-vote-card';
import { OldBambooPlugin as ColumnsPlugin } from '@ali/4ever-plugin-columns';
import { OldBambooPlugin as CommentPlugin } from '@ali/4ever-plugin-comment';
import { OldBambooPlugin as DraggablePlugin } from '@ali/4ever-plugin-draggable';
import { OldBambooPlugin as EmbedPlugin } from '@ali/4ever-plugin-embed';
import { OldBambooPlugin as HrPlugin } from '@ali/4ever-plugin-hr';
import { OldBambooPlugin as ImagePlugin } from '@ali/4ever-plugin-image';
import { OldBambooPlugin as ImTagPlugin } from '@ali/4ever-plugin-im-tag';
import { OldBambooPlugin as FormulaPlugin } from '@ali/4ever-plugin-formula';
import { CodeRole } from '@ali/4ever-plugin-code';
import { InlineCodeRole } from '@ali/4ever-plugin-inline-code';
import { OldBambooPlugin as ListPlugin } from '@ali/4ever-plugin-list';
import { OldBambooPlugin as IndentPlugin } from '@ali/4ever-plugin-indent';
import { OldBambooPlugin as LinkPlugin } from '@ali/4ever-plugin-link';
import { zhiFactory as MarksPlugin } from '@ali/4ever-plugin-marks';
import { OldBambooPlugin as TextBackgroundPlugin } from '@ali/4ever-plugin-text-background';
import { OldBambooPlugin as MentionPlugin } from '@ali/4ever-plugin-mention';
import { OldBambooPlugin as ParagraphPlugin } from '@ali/4ever-plugin-paragraph';
import { OldBambooPlugin as PcContextMenu, PCContextMenu as ContextMenu } from '@ali/4ever-plugin-pc-contex-menu';
import { OldBambooPlugin as ReferPlugin } from '@ali/4ever-plugin-refer';
import { OldBambooPlugin as TablePlugin } from '@ali/4ever-plugin-table';
import { OldBambooPlugin as TagPlugin } from '@ali/4ever-plugin-tag';
import { OldBambooPlugin as StickerPlugin } from '@ali/4ever-plugin-sticker';
import { OldBambooPlugin as CalloutPlugin } from '@ali/4ever-plugin-callout';
import { OldBambooPlugin as HeadingGroupPlugin } from '@ali/4ever-plugin-heading-group';
import { OldBambooPlugin as SubjectPlugin } from '@ali/4ever-plugin-subject';
import { OldBambooPlugin as GroupFactory } from '@ali/4ever-plugin-group-factory';
import { OldBambooPlugin as ListGroupPlugin } from '@ali/4ever-plugin-list-group';
import { OldBambooPlugin as AlignmentPlugin } from '@ali/4ever-plugin-alignment';
import { OldBambooPlugin as BlockquotePlugin } from '@ali/4ever-plugin-blockquote';
import { OldBambooPlugin as ClearStylePlugin } from '@ali/4ever-plugin-clear-style';
import { OldBambooPlugin as ColorPlugin } from '@ali/4ever-plugin-color';
import { OldBambooPlugin as FontsPlugin } from '@ali/4ever-plugin-fonts';
import { OldBambooPlugin as FormatPainterPlugin } from '@ali/4ever-plugin-format-painter';
import { OldBambooPlugin as HeadingPlugin } from '@ali/4ever-plugin-heading';
import { OldBambooPlugin as HighlightPlugin } from '@ali/4ever-plugin-highlight';
import { HistoryRole } from '@ali/4ever-plugin-history';
import { OldBambooPlugin as LetterSpacingPlugin } from '@ali/4ever-plugin-letter-spacing';
import { OldBambooPlugin as LineHeightPlugin } from '@ali/4ever-plugin-line-height';
import { OldBambooPlugin as SzPlugin } from '@ali/4ever-plugin-sz';
import { TemplateButtonRole } from '@ali/4ever-plugin-template-button';
import { OldBambooPlugin as TranslatePlugin } from '@ali/4ever-plugin-translate';
import { OldBambooPlugin as UnderlinePlugin } from '@ali/4ever-plugin-underline';
import { OldBambooPlugin as vertAlignPlugin } from '@ali/4ever-plugin-vert-align';
import { RefBlockRole } from '@ali/4ever-plugin-refblock';
import { createFindAndReplacePlugin as FindAndReplacePlugin } from '@ali/4ever-plugin-find-and-replace';
import { HoverToolbarRole } from '@ali/4ever-plugin-hover-toolbar';
export { ColumnsRenderNodeEdit } from '@ali/4ever-plugin-columns';
export { ContextMenu };
export { OldBambooPlugin as DocStylePlugin } from '@ali/4ever-plugin-doc-style';
export { ListCheckBox as Checkbox } from '@ali/4ever-plugin-list';
export { HoverBlock } from '@ali/4ever-utils';
var FileTypeMap = EmbedPlugin.FileTypeMap;
export { FileTypeMap };
export { CardPlugin, CalendarCardPlugin, ColumnsPlugin, CommentPlugin, DraggablePlugin, EmbedPlugin, HrPlugin, ImagePlugin, ImTagPlugin, IndentPlugin, LinkPlugin, ListPlugin, MarksPlugin, TextBackgroundPlugin, MentionPlugin, ParagraphPlugin, PcContextMenu, ReferPlugin, TablePlugin, TagPlugin, StickerPlugin, CalloutPlugin, HeadingPlugin, HeadingGroupPlugin, SubjectPlugin, GroupFactory, ListGroupPlugin, VoteCardPlugin, FindAndReplacePlugin, FormulaPlugin };
export var PluginRoles = _extends({}, AlignmentPlugin.AlignmentRole, BlockquotePlugin.BlockquoteRole, ClearStylePlugin.ClearStyleRole, ColorPlugin.ColorRole, ColumnsPlugin.ColumnsRole, CommentPlugin.CommentRole, DraggablePlugin.DraggableRole, EmbedPlugin.EmbedRole, FontsPlugin.FontRole, FormatPainterPlugin.FormatPainterRole, HeadingPlugin.HeadingRole, HighlightPlugin.HighlightRole, HistoryRole, HrPlugin.HrRole, ImagePlugin.ImageRole, IndentPlugin.IndentRole, LetterSpacingPlugin.LetterSpacingRole, LineHeightPlugin.LineHeightRole, LinkPlugin.LinkRole, ListPlugin.ListRole, ParagraphPlugin.ParagraphRole, ReferPlugin.ReferRole, StickerPlugin.StickerRole, SzPlugin.SzRole, TablePlugin.TableRole, TemplateButtonRole, TranslatePlugin.TranslateRole, UnderlinePlugin.UnderlineRole, vertAlignPlugin.vertAlignRole, CalloutPlugin.CalloutRole, MentionPlugin.MentionRole, RefBlockRole, CodeRole, InlineCodeRole, HoverToolbarRole);
