// @ali/4ever-plugin-history/dist-web/bi/components/next/redoButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { environment } from '@ali/4ever-cangjie';
import { HistoryRole } from "../../../utils/role";
import { RedoButton as Button } from '@ali/4ever-component';
import { redo } from "../../actions";
var SHORTCUTKEY = environment.IS_MAC ? '⌘+Shift+Z' : 'Ctrl+Shift+Z / Ctrl+Y';
export default function RedoButton(props) {
  var controller = props.controller,
      locale = props.locale,
      disabled = props.disabled,
      tooltipProp = props.tooltip,
      rest = _objectWithoutPropertiesLoose(props, ["controller", "locale", "disabled", "tooltip"]);

  var isDisabled = disabled || !controller.query('canRedo');
  var handleClick = React.useCallback(function () {
    return controller.run('onAction', redo());
  }, [controller]);
  var tooltip = React.useMemo(function () {
    return _extends({
      title: (locale == null ? void 0 : locale.redo) || 'Redo',
      shortCut: SHORTCUTKEY
    }, tooltipProp);
  }, [tooltipProp]);
  return /*#__PURE__*/_createElement(Button, _extends({
    testid: "bi-toolbar-redo",
    disabled: isDisabled,
    onClick: handleClick,
    role: HistoryRole.redo,
    title: (locale == null ? void 0 : locale.redo) || 'Redo',
    tooltip: tooltip
  }, rest));
}
