// @ali/we-design-next/dist-node/common/consts/colors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.COLORS = void 0;
const COLORS = {
  // 白色，黑色，透明，钉钉蓝
  WHITE: 'rgba(255, 255, 255, 1)',
  BLACK: 'rgba(0, 0, 0, 1)',
  TRANSPARENT: 'rgba(0, 0, 0, 0)',
  DING_BLUR: 'rgba(0, 137, 255, 1)',
  // Tokens

  LIGHT: {
    LEVEL1_BASE_COLOR: 'rgba(23, 26, 29, 1)',
    LEVEL2_BASE_COLOR: 'rgba(23, 26, 29, 0.6)',
    LEVEL3_BASE_COLOR: 'rgba(23, 26, 29, 0.4)',
    LEVEL4_BASE_COLOR: 'rgba(23, 26, 29, 0.24)',
    STAMP_COLOR: 'rgba(23, 26, 29, 0.08)',
    DOC_BG_COLOR: 'rgba(245, 246, 247, 1)',
    DOC_BG_LIGHT_COLOR: 'rgba(249, 250, 252, 1)',
    DOC_FG_COLOR: 'rgba(255, 255, 255, 1)',
    DAVINCI_BG_NEW: 'rgba(250, 252, 255, 1)',
    DAVINCI_BLUE_LIGHT: 'rgba(50, 96, 201, 0.09)',
    DAVINCI_BLUE_LIGHT_HOVER: 'rgba(50, 96, 201, 0.06)',
    DAVINCI_BLUE_LIGHT_ACTIVE: 'rgba(50, 96, 201, 0.09)',
    DAVINCI_BLUE_LIGHT_BUTTON_HOVER: 'rgba(50, 96, 201, 0.15)',
    DAVINCI_BLUE_LIGHT_BUTTON_ACTIVE: 'rgba(50, 96, 201, 0.12)'
  },
  DARK: {
    LEVEL1_BASE_COLOR: 'rgba(238, 242, 248, 1)',
    LEVEL2_BASE_COLOR: 'rgba(238, 242, 248, 0.6)',
    LEVEL3_BASE_COLOR: 'rgba(238, 242, 248, 0.4)',
    LEVEL4_BASE_COLOR: 'rgba(238, 242, 248, 0.24)',
    STAMP_COLOR: 'rgba(238, 242, 248, 0.12)',
    DOC_BG_COLOR: 'rgba(17, 18, 19, 1)',
    DOC_BG_LIGHT_COLOR: 'rgba(41, 43, 46, 1)',
    DOC_FG_COLOR: 'rgba(40, 40, 43, 1)',
    DAVINCI_BG_NEW: 'rgba(28, 29, 31, 1)',
    DAVINCI_BLUE_LIGHT: 'rgba(40, 44, 48, 1)',
    DAVINCI_BLUE_LIGHT_HOVER: 'rgba(126, 134, 142, 0.18)',
    DAVINCI_BLUE_LIGHT_ACTIVE: 'rgba(126, 134, 142, 0.15)',
    DAVINCI_BLUE_LIGHT_BUTTON_HOVER: 'rgba(126, 134, 142, 0.18)',
    DAVINCI_BLUE_LIGHT_BUTTON_ACTIVE: 'rgba(126, 134, 142, 0.15)'
  },
  COMMON: {
    BLUE1_COLOR: 'rgba(0, 127, 255, 1)',
    BLUE1_HOVER_COLOR: 'rgba(56, 155, 255, 1)',
    BLUE1_ACTIVE_COLOR: 'rgba(0, 93, 214, 1)',
    BLUE2_COLOR: 'rgba(0, 127, 255, 0.48)',
    BLUR3_COLOR: 'rgba(0, 127, 255, 0.12)',
    DANGER1_COLOR: 'rgba(255, 82, 25, 1)',
    DANGER1_HOVER_COLOR: 'rgba(232, 70, 17, 1)',
    DANGER1_ACTIVE_COLOR: 'rgba(211, 63, 14, 1)',
    LINK_COLOR: 'rgba(49, 126, 208, 1)',
    OVERLAY_COLOR: 'rgba(0, 0, 0, 0.16)',
    OVERLAY_HOVER_COLOR: 'rgba(0, 0, 0, 0.08)',
    OVERLAY_PRESS_COLOR: 'rgba(0, 0, 0, 0.16)',
    LINE_LIGHT_COLOR: 'rgba(126, 134, 142, 0.1)',
    LINE_HARD_COLOR: 'rgba(126, 134, 142, 0.18)',
    WATER1_COLOR: 'rgba(90, 200, 250, 1)'
  },
  OTHERS: {
    DAO_HOVER_BG_COLOR: 'rgba(126, 134, 142, 0.12)',
    LINE_LIGHT_COLOR: 'rgba(126, 134, 142, 0.16)',
    COMMON_DANGER2_COLOR: 'rgba(255, 82, 25, 0.48)',
    LINE_DIVIDER_COLOR: 'rgba(255, 255, 255, 0.1)',
    BASIC_WHITE2_COLOR: 'rgba(255, 255, 255, 0.4)',
    COMMON_DESCRIPTION_COLOR: 'rgba(255, 255, 255, 0.56)',
    MSG_BG_COLOR: 'rgba(69, 72, 74, 0.9)',
    BRAND1: 'rgba(45, 132, 250, 1)',
    // 通用背景色
    BG_COLOR: 'rgba(126, 134, 142, 0.16)',
    BG_ACTIVE: 'rgba(126, 134, 142, 0.24)',
    BG_FOCUS: 'rgba(126, 134, 142, 0.16)',
    // 边框色
    BORDER: 'rgba(126, 134, 142, 0.24)',
    DIVIDER: 'rgba(17, 31, 44, 0.08)',
    PRIMARY_HOVER: 'rgba(0, 126, 234, 1)'
  }
};
exports.COLORS = COLORS;
