// @ali/4ever-plugin-color-blocks/dist-node/mo/serializer/jsonMLToHTML.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _everUtils = require("@ali/4ever-utils");

var _meta = require("../../utils/meta");

const _createElement = /*#__PURE__*/React.createElement;

function getStyle(metadata) {
  const style = {
    padding: '10px 14px'
  };
  const {
    bgcolor = '#FEF8E8',
    border
  } = metadata || {};

  if (bgcolor) {
    style.backgroundColor = bgcolor;
  }

  if (border) {
    style.border = `1px solid ${border}`;
    style.borderRadius = '8px';
  }

  return style;
}

var _default = {
  name: 'container',
  match: node => {
    const tag = (0, _everUtils.getTagName)(node);
    const attrs = (0, _everUtils.getAttributes)(node);
    return tag === 'container' && attrs?.subType === _meta.ELEMENT_TYPE;
  },
  convert: (node, path, convert) => {
    const attrs = (0, _everUtils.getAttributes)(node) || {};
    const children = (0, _everUtils.getChildren)(node).map((child, index) => convert(child, [...path, index]));
    const style = getStyle(attrs?.metadata);
    return /*#__PURE__*/_createElement("div", {
      "data-type": "container",
      style: style
    }, children);
  }
};
exports.default = _default;
