// @ali/4ever-plugin-color-blocks/dist-node/mo/serializer/markdownToHTML.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _interopRequireWildcard2 = _interopRequireDefault(require("@babel/runtime/helpers/interopRequireWildcard"));

const validTypes = /^(warning|info|danger|success|tips)*$/;
const rule = {
  uses: [{
    plugin: () => Promise.resolve().then(() => (0, _interopRequireWildcard2.default)(require('markdown-it-container'))),
    data: ['colorBlocks', {
      validate: function (params) {
        return params.trim().match(validTypes);
      },
      render: function (tokens, idx, _options, env, slf) {
        const m = tokens[idx].info.trim().match(validTypes); // add a class to the opening tag

        if (tokens[idx].nesting === 1) {
          tokens[idx].attrJoin('class', `colorBlocks ${m[1] || ''}`);
        }

        return slf.renderToken(tokens, idx, _options, env, slf);
      }
    }]
  }]
};
var _default = rule;
exports.default = _default;
