// @ali/4ever-bi/dist-web/plugins/image/handlers/createContextMenu.js
import { Image } from '@ali/4ever-mo';
import { PcContextMenu } from '@ali/4ever-bamboo';
import { toggleImageSidebar } from "../actions";

function createContextMenu(config, sideBarTrigger) {
  var _config$locale = config.locale,
      locale = _config$locale === void 0 ? {} : _config$locale;
  return function contextMenu(controller, next) {
    var _controller$value = controller.value,
        document = _controller$value.document,
        selection = _controller$value.selection;
    var isCollapsed = selection.isCollapsed;

    if (!isCollapsed && sideBarTrigger) {
      var start = selection.getStart(document);
      var end = selection.getEnd(document);
      var node1 = document.getNextText(start.key);
      var node2 = document.getPreviousText(end.key);

      var closeContextMenu = function closeContextMenu(onActionController) {
        onActionController.run('onAction', PcContextMenu.closeContextMenu());
      };

      if (node1 && node2 && node1 === node2) {
        var node = document.getClosestVoid(node1.key, controller);

        if (Image.isImage(node)) {
          return [].concat(next(), [{
            name: (locale == null ? void 0 : locale.options) || 'Image options',
            key: 'image-options',
            action: toggleImageSidebar({
              node: node,
              toggle: true,
              callback: closeContextMenu
            }),
            options: {
              group: 2
            }
          }]);
        }
      }
    }

    return next();
  };
}

export default createContextMenu;
