// @ali/we-ui/dist-web/components/link/styled.js
import styled from 'styled-components';
import { withTheme } from "../theme";
var BaseWrapper = /*#__PURE__*/styled.div(["line-height:28px;font-size:14px;font-weight:normal;text-indent:0;border-radius:2px 2px;padding:8px 8px 8px 12px;"]);
export var EditorWrapper = withTheme(styled(BaseWrapper)(["display:flex;flex-direction:column;width:350px;background:", ";border:1px solid ", ";box-shadow:", ";color:", ";"], function (_ref) {
  var theme = _ref.theme;
  return theme.ui.colorGray1;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.ui.normalColor;
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.ui.boxShadow;
}, function (_ref4) {
  var theme = _ref4.theme;
  return theme.ui.colorText4;
}));
export var EditorContent = /*#__PURE__*/styled.div(["display:flex;flex-direction:column;"]);
export var EditorFooter = /*#__PURE__*/styled.div(["display:flex;justify-content:flex-end;align-items:center;margin-top:8px;"]);
export var ItemInput = withTheme(styled.div(["position:relative;line-height:28px;margin:0 3px;text-align:left;flex:0 0 auto;display:flex;padding:2px 0;input{color:inherit;line-height:24px;padding:0 3px;display:inline-block;border:1px solid ", ";outline:0;font-size:14px;letter-spacing:0.05em;border-radius:5px;flex:1;}"], function (_ref5) {
  var theme = _ref5.theme;
  return theme.ui.colorLine1;
}));
export var LinkText = withTheme(styled.span(["display:inline-block;min-width:200px;max-width:250px;line-height:28px;padding:0 5px;color:", ";overflow:hidden;text-overflow:ellipsis;white-space:nowrap;cursor:pointer;&:hover{color:", ";}"], function (_ref6) {
  var theme = _ref6.theme;
  return theme.ui.colorText4;
}, function (_ref7) {
  var theme = _ref7.theme;
  return theme.ui.colorBlue1;
}));
export var SaveButton = withTheme(styled.div(["border-radius:15px;padding:0px 12px;font-size:14px;background-color:", ";color:", ";cursor:pointer;&:hover,:focus{background-color:#3da5ff;}"], function (_ref8) {
  var theme = _ref8.theme;
  return theme.ui.colorBlue1;
}, function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.colorGray1;
}));
export var RemoveButton = /*#__PURE__*/styled.div(["display:flex;align-items:center;margin-right:10px;cursor:pointer;"]);
export var ToolbarWrapper = withTheme(styled(BaseWrapper)(["width:fit-content;display:flex;align-items:center;background:", ";border:1px solid ", ";box-shadow:", ";z-index:1;color:", ";.weicon{font-size:18px;vertical-align:middle;}"], function (_ref10) {
  var theme = _ref10.theme;
  return theme.ui.colorGray1;
}, function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.normalColor;
}, function (_ref12) {
  var theme = _ref12.theme;
  return theme.ui.boxShadow;
}, function (_ref13) {
  var theme = _ref13.theme;
  return theme.ui.colorText4;
}));
export var LinkHrefWrapper = withTheme(styled.span(["position:relative;flex:1;margin-left:12px;color:", ";height:24px;line-height:24px;:hover{cursor:pointer;}"], function (_ref14) {
  var theme = _ref14.theme;
  return theme.ui.linkColor;
}));
export var LinkHref = withTheme(styled.span(["display:inline-block;max-width:200px;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;color:", ";"], function (_ref15) {
  var ui = _ref15.theme.ui;
  return ui.primaryColor;
}));
export var ToolbarIconWrapper = withTheme(styled.span(["display:inline-block;margin-left:10px;cursor:pointer;border-radius:50%;width:24px;height:24px;line-height:24px;text-align:center;:hover{background-color:", ";}"], function (_ref16) {
  var theme = _ref16.theme;
  return theme.ui.colorFill4;
}));
