// @ali/4ever-plugin-heading/dist-node/components/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.H6 = exports.H5 = exports.H4 = exports.H3 = exports.H2 = exports.H1 = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _everUtils = require("@ali/4ever-utils");

const H1 = /*#__PURE__*/_styledComponents.default.h1(["", ""], (0, _everUtils.transformThemeObjToCssString)(_everUtils.theme['heading-1']));

exports.H1 = H1;

const H2 = /*#__PURE__*/_styledComponents.default.h2(["", ""], (0, _everUtils.transformThemeObjToCssString)(_everUtils.theme['heading-2']));

exports.H2 = H2;

const H3 = /*#__PURE__*/_styledComponents.default.h3(["", ""], (0, _everUtils.transformThemeObjToCssString)(_everUtils.theme['heading-3']));

exports.H3 = H3;

const H4 = /*#__PURE__*/_styledComponents.default.h4(["", ""], (0, _everUtils.transformThemeObjToCssString)(_everUtils.theme['heading-4']));

exports.H4 = H4;

const H5 = /*#__PURE__*/_styledComponents.default.h5(["", ""], (0, _everUtils.transformThemeObjToCssString)(_everUtils.theme['heading-5']));

exports.H5 = H5;

const H6 = /*#__PURE__*/_styledComponents.default.h6(["", ""], (0, _everUtils.transformThemeObjToCssString)(_everUtils.theme['heading-6']));

exports.H6 = H6;
