// @ali/4ever-zhi/dist-node/plugins/video/mobile/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createEmbedPlugin;

var React = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

const _createElement = /*#__PURE__*/React.createElement;
const {
  FileTypeMap,
  VideoPlayer,
  MobileEmbedCard,
  EmbedFailedCard
} = _everBamboo.EmbedPlugin;

function createEmbedPlugin(config) {
  const onDownloadFile = props => {
    const {
      node
    } = props;
    const {
      data
    } = node;
    const {
      src,
      name,
      type,
      size
    } = data;
    const videoConfig = config.video || {};
    const {
      downloadFile
    } = videoConfig;

    if (downloadFile) {
      downloadFile(src, {
        name,
        type,
        size
      });
    }
  };

  const handleMediaLoadError = src => {
    const videoConfig = config.video || {};
    const {
      onMediaLoadError
    } = videoConfig;

    if (onMediaLoadError) {
      onMediaLoadError(src);
    }
  };

  const handleMediaLoadSuccess = src => {
    const videoConfig = config.video || {};
    const {
      onMediaLoadSuccess
    } = videoConfig;

    if (onMediaLoadSuccess) {
      onMediaLoadSuccess(src);
    }
  };

  const renderPreview = props => {
    const {
      isSelected,
      node
    } = props;
    const {
      transformVideoURL,
      renderVideoMask,
      setTempPosterUrl,
      downloadFile,
      allowDownload,
      locale,
      enableVideoPreview = true,
      allowFullscreen
    } = config.video || {};

    if (!enableVideoPreview) {
      return renderCard(props);
    }

    return /*#__PURE__*/_createElement(VideoPlayer, {
      transformVideoURL: transformVideoURL,
      onMediaLoadSuccess: handleMediaLoadSuccess,
      onMediaLoadError: handleMediaLoadError,
      setTempPosterUrl: setTempPosterUrl,
      node: node,
      isSelected: isSelected,
      renderVideoMask: renderVideoMask,
      isMobile: true,
      allowDownload: downloadFile && allowDownload,
      onDownload: () => onDownloadFile(props),
      allowFullscreen: allowFullscreen,
      locale: locale
    });
  };

  const renderCard = props => {
    const {
      node
    } = props;
    const {
      data
    } = node;
    const {
      src,
      name,
      type,
      size
    } = data;
    const {
      onPreview
    } = config.video || {};
    return /*#__PURE__*/_createElement(MobileEmbedCard, {
      name: name,
      onPreview: onPreview,
      src: src,
      type: type,
      size: size
    });
  };

  return {
    renderNode: {
      embed: (props, _, next) => {
        const {
          node
        } = props;

        if (node.data.type === FileTypeMap.Video && node.data.src) {
          const {
            attributes
          } = props;
          const {
            data
          } = node;
          const {
            viewType
          } = data;
          return /*#__PURE__*/_createElement("div", attributes, viewType === 'preview' ? renderPreview(props) : renderCard(props));
        }

        if (node.data.type === FileTypeMap.Video && node.data.errorType) {
          return /*#__PURE__*/_createElement(EmbedFailedCard, {
            node: node,
            locale: config?.video?.locale || {}
          });
        }

        return next();
      }
    }
  };
}
