// @ali/4ever-plugin-inline-placeholder/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.moFactory;
  }
});
Object.defineProperty(exports, "InlinePlaceholder", {
  enumerable: true,
  get: function () {
    return _models.default;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.createServerSerializationPlugin;
  }
});
exports.biFactory = exports.zhiFactory = void 0;

var _index = require("./mo/index");

var _models = _interopRequireDefault(require("./mo/models"));

var _index2 = _interopRequireWildcard(require("./mo/serializer/index"));

// facotry
const zhiFactory = _index.moFactory;
exports.zhiFactory = zhiFactory;
const biFactory = _index.moFactory; // other export for api compatibility

exports.biFactory = biFactory;
