// @ali/4ever-plugin-columns/dist-web/utils/actions/index.js
export var ACTION_MERGE_COLUMNS = 'columns_merge_columns';
export var ACTION_INERT_COLUMNS = 'columns_insert_columns';
export var ACTION_INERT_COL = 'columns_insert_col';
export var ACTION_DELETE_COLUMN = 'columns_delete_column';
export var ACTION_CLEANUP_COLUMN = 'columns_cleanup_column';
export var ACTION_REMOVE_INVALID_CELLS = 'columns_remove_invalid_cells';
export var ACTION_SET_CELL_DATA = 'columns_set_cell_data';
export function mergeColumns(payload) {
  return {
    type: ACTION_MERGE_COLUMNS,
    payload: payload
  };
}
export function insertColumns(payload) {
  return {
    type: ACTION_INERT_COLUMNS,
    payload: payload
  };
}
export function insertCol(payload) {
  return {
    type: ACTION_INERT_COL,
    payload: payload
  };
}
export function deleteColumn(payload) {
  return {
    type: ACTION_DELETE_COLUMN,
    payload: payload
  };
}
export function cleanUpColumns() {
  return {
    type: ACTION_CLEANUP_COLUMN
  };
}
export function removeInvalidCells() {
  return {
    type: ACTION_REMOVE_INVALID_CELLS
  };
}
export function setCellData(payload) {
  return {
    type: ACTION_SET_CELL_DATA,
    payload: payload
  };
}
