// @ali/4ever-mo/dist-web/index.js
import Serializer, { moSerializer } from "./serializer";
import createMoPlugins from "./createMoPlugins";
import createCustomizedMoPlugins from "./createCustomizedMoPlugins";
import plugins from "./plugins";
import { createEmptyParagraph, createEmptyText, attrsToMarks, markToAttr, marksToAttrs } from "./plugins/utils";
import { Bold, Caps, Color, Fonts, Highlight, InlineCode, Italic, SmallCaps, Strike, Dstrike, Sz, Underline, VertAlign, Shd, Spacing, Kern, Em } from '@ali/4ever-plugin-marks';
export var pluginUtils = {
  createEmptyParagraph: createEmptyParagraph,
  createEmptyText: createEmptyText,
  attrsToMarks: attrsToMarks,
  markToAttr: markToAttr,
  marksToAttrs: marksToAttrs
};
export { MoCommon as Common, MoInterfaces as Interfaces } from '@ali/4ever-cangjie'; // serialization Plugins

export { paragraph as ParagraphPlugin, br as BrPlugin, code as CodePlugin, embed as EmbedPlugin, hr as HrPlugin, image as ImagePlugin, link as LinkPlugin, list as ListPlugin, mention as MentionPlugin, onlineVideo as OnlineVideoPlugin, placeholder as PlaceholderPlugin, inlineCode as InlineCodePlugin, refer as ReferPlugin, table as TablePlugin, tagWithFormula as TagPlugin, refblock as RefBlockPlugin } from "./plugins"; // 防止 breaking change 产生大规模影响

export var BasicPlugin = function BasicPlugin() {
  return {};
};
export * from "./type"; // reexport from cangjie mo base

export { MoState as State, DATA_CLIPBORAD_CANGJIE, SLASH_PLACEHOLDER, HtmlParser } from '@ali/4ever-cangjie'; // Models

export { Card } from '@ali/4ever-plugin-card';
export { Tag } from '@ali/4ever-plugin-tag';
export { Embed } from '@ali/4ever-plugin-embed';
export { Code } from '@ali/4ever-plugin-code';
export { Heading } from '@ali/4ever-plugin-heading';
export { Hr } from '@ali/4ever-plugin-hr';
export { Image } from '@ali/4ever-plugin-image';
export { Link } from '@ali/4ever-plugin-link';
export { Mention } from '@ali/4ever-plugin-mention';
export { InlineCode } from '@ali/4ever-plugin-inline-code';
export { Paragraph } from '@ali/4ever-plugin-paragraph';
export { Table, TableRow, TableCell } from '@ali/4ever-plugin-table';
export { OnlineVideo } from '@ali/4ever-plugin-online-video';
export { Placeholder } from '@ali/4ever-plugin-placeholder';
export { InlinePlaceholder } from '@ali/4ever-plugin-inline-placeholder';
export { Refer } from '@ali/4ever-plugin-refer';
export { Sticker } from '@ali/4ever-plugin-sticker';
export { Container } from '@ali/4ever-plugin-container';
export { TemplateButton } from '@ali/4ever-plugin-template-button';
export { ColorBlocks } from '@ali/4ever-plugin-color-blocks';
// Marks Models
export var Marks = {
  Bold: Bold,
  Caps: Caps,
  Color: Color,
  Fonts: Fonts,
  Highlight: Highlight,
  InlineCode: InlineCode,
  Italic: Italic,
  SmallCaps: SmallCaps,
  Strike: Strike,
  Dstrike: Dstrike,
  Sz: Sz,
  Underline: Underline,
  VertAlign: VertAlign,
  Shd: Shd,
  Spacing: Spacing,
  Kern: Kern,
  Em: Em
};
export { Bold as BoldMark, Caps as CapsMark, Color as ColorMark, Fonts as FontsMark, Highlight as HighlightMark, InlineCode as InlineCodeMark, Italic as ItalicMark, SmallCaps as SmallCapsMark, Strike as StrikeMark, Sz as SzMark, Underline as UnderlineMark, VertAlign as VertAlignMark, Shd as ShdMark } from '@ali/4ever-plugin-marks';
// Schema
export { default as schema } from "./schema"; // strictSchema

export { default as strictSchema, inlineTypes, blockTypes, createStrictSchemaOld as createStrictSchema } from "./strictSchema";
export { default as ServerSerializer, createServerMoPlugins, serverPlugins, moServerSerializer } from "./serializerValueJsonML";
export { moSerializer, plugins, Serializer, createMoPlugins, createCustomizedMoPlugins };
export { VERSION, KEY_OF_VERSION, DEFAULT_VERSION } from "./version";
export { default as migrate, getVersionFromCP, getVersionFromDocument, assertHasMigrations, normalizeError as normalizeMigrateError, MissingMigrationError } from "./migrate";
export { default as migrations } from "./migrations";
export { default as createMigration } from "./migrations/createMigration";
export { default as corrections } from "./corrections";
export default {
  Serializer: Serializer,
  createMoPlugins: createMoPlugins,
  createCustomizedMoPlugins: createCustomizedMoPlugins
};
