// @ali/4ever-mo/dist-web/schema.js
// [IMPORTANT] 本文件任何修改均必须通知文字团队全员，并通过架构师与 TL Review
import { createSchema } from '@ali/4ever-cangjie';
import { moFactory as Code } from '@ali/4ever-plugin-code';
import { moFactory as Hr } from '@ali/4ever-plugin-hr';
import { moFactory as Image } from '@ali/4ever-plugin-image';
import { moFactory as InlinePlaceholder } from '@ali/4ever-plugin-inline-placeholder';
import { moFactory as Link } from '@ali/4ever-plugin-link';
import { moFactory as Mention } from '@ali/4ever-plugin-mention';
import { moFactory as OnlineVideo } from '@ali/4ever-plugin-online-video';
import { moFactory as Placeholder } from '@ali/4ever-plugin-placeholder';
import { moFactory as Refer } from '@ali/4ever-plugin-refer';
import { moFactory as Sticker } from '@ali/4ever-plugin-sticker';
import { moFactory as Table } from '@ali/4ever-plugin-table';
import { moFactory as Card } from '@ali/4ever-plugin-card';
import { moFactory as Tag } from '@ali/4ever-plugin-tag';
import { moFactory as Embed } from '@ali/4ever-plugin-embed';
import { moFactory as Container } from '@ali/4ever-plugin-container';
import { moFactory as Colorblocks } from '@ali/4ever-plugin-color-blocks';
import { moFactory as RefBlock } from '@ali/4ever-plugin-refblock';
import { moFactory as InlineCode } from '@ali/4ever-plugin-inline-code';
import { moFactory as Br } from '@ali/4ever-plugin-br';
import { moFactory as Paragraph } from '@ali/4ever-plugin-paragraph';
import { moFactory as Heading } from '@ali/4ever-plugin-heading';
import { moFactory as TemplateButton } from '@ali/4ever-plugin-template-button'; // tab 插件不在 we-editor 中

var Tab = function Tab() {
  return {
    key: 'tab',
    schemas: {
      type: 'tab',
      isVoid: true,
      isBlock: false
    }
  };
}; // mo 里 schema 针对的规则是写死的, 标记为废弃态, 后续使用 controller.schema 和 controller.strictSchema


export var moPlugins = [// blocks
Paragraph, Heading, Code, Hr, OnlineVideo, Placeholder, Table, Card, Embed, TemplateButton, Colorblocks, Container, RefBlock, // inlines
Br, InlinePlaceholder, Image, Link, Mention, Refer, Sticker, InlineCode, Tag, Tab].map(function (f) {
  return f({});
});
/** @deprecated use controller.schema instead  */

var schema = createSchema(moPlugins);
export default schema;
