// @ali/4ever-zhi/dist-node/plugins/comment/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everBamboo = require("@ali/4ever-bamboo");

const {
  hooks,
  types,
  actions,
  commands,
  queries
} = _everBamboo.CommentPlugin;
const {
  lazyRenderStrategy
} = queries;

function createCommentPlugin(allConfigs) {
  const {
    createContextMenu,
    createOnKeyDown,
    createOnMouseMove,
    createOnAction,
    createOnChange,
    createRenderMark,
    createRenderNode,
    createRenderContent,
    createRenderEditable,
    createOnCangjieSelect,
    createOnCangjieCopyOrCut
  } = hooks;
  const configs = allConfigs.comment;
  return {
    queries: {
      lazyRenderStrategy
    },
    contextMenu: createContextMenu(configs),
    onKeyDown: createOnKeyDown(configs),
    onMouseMove: createOnMouseMove(configs),
    onAction: createOnAction(configs),
    onChange: createOnChange(configs),
    renderMark: createRenderMark(configs),
    renderNode: createRenderNode(configs),
    renderContent: createRenderContent(configs),
    renderEditable: createRenderEditable(configs),
    onCangjieSelect: createOnCangjieSelect(configs),
    onCangjieCopy: createOnCangjieCopyOrCut(),
    onCangjieCut: createOnCangjieCopyOrCut(),
    commands
  };
}

createCommentPlugin.types = types;
createCommentPlugin.actions = actions;
createCommentPlugin.commands = commands;
createCommentPlugin.queries = queries;
var _default = createCommentPlugin;
exports.default = _default;
