// @ali/4ever-plugin-draggable/dist-web/bi/index.js
import _extends from "@babel/runtime/helpers/extends";
import renderContent from "./renderContent";
import { actions, isActions } from "./actions";
import createRenderNode from "./createRenderNode";
import onAction from "./handlers/onAction";
import toolbar from "./toolbar";
import renderLeftToolbar from "./components/dragHander";
import { leftToolbarHandlers, handleDragChange } from "./leftToolbarHandlers";
import { queries } from "./queries";
import { key } from "../utils/meta";
import { commands } from "./commands";
export var biFactory = function biFactory(configs) {
  return {
    key: key,
    renderContent: renderContent(configs, handleDragChange),
    renderNode: createRenderNode(configs),
    queries: _extends({}, queries, {
      /**
       * TODO: 获取 configs 不应该是个 query，待确定历史原因
       */
      getDraggableConfig: function getDraggableConfig() {
        return configs.draggable;
      }
    }),
    onAction: onAction,
    actions: actions,
    isActions: isActions,
    commands: commands,
    models: [],
    toolbar: toolbar(configs),
    leftToolbar: renderLeftToolbar(configs),
    leftToolbarHandlers: leftToolbarHandlers
  };
};
