// @ali/4ever-bi/dist-node/plugins/vertAlign/mobile/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _isSub = _interopRequireDefault(require("../queries/isSub"));

var _isSup = _interopRequireDefault(require("../queries/isSup"));

var actions = _interopRequireWildcard(require("../actions"));

var _onAction = _interopRequireDefault(require("../handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("../handlers/onKeyDown"));

var _onCangjieInput = _interopRequireDefault(require("../handlers/onCangjieInput"));

var _everPluginVertAlign = require("@ali/4ever-plugin-vert-align");

const queries = {
  isSub: _isSub.default,
  isSup: _isSup.default
};

function createVertAlignPlugin() {
  return { ...(0, _everPluginVertAlign.moFactory)({}),
    queries,
    onAction: _onAction.default,
    onKeyDown: _onKeyDown.default,
    onCangjieInput: (0, _onCangjieInput.default)(),
    models: {
      klass: 'mark',
      type: 'vertAlign'
    }
  };
}

createVertAlignPlugin.actions = actions;
createVertAlignPlugin.queries = queries;
var _default = createVertAlignPlugin;
exports.default = _default;
