// @ali/4ever-bi/dist-node/plugins/sticker/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.insertSticker = insertSticker;
exports.setSticker = setSticker;
exports.setStickerPanelVisible = setStickerPanelVisible;
exports.moveCaret = moveCaret;
exports.setRange = setRange;
exports.insertMarkdownEmoji = insertMarkdownEmoji;
exports.INSERT_MARKDOWN_EMOJI = exports.STICKER_SET_RANGE = exports.STICKER_MOVE_CARET = exports.STICKER_PANEL_SET_VISIBLE = exports.STICKER_SET = exports.STICKER_INSERT = void 0;
const STICKER_INSERT = Symbol('STICKER_INSERT');
exports.STICKER_INSERT = STICKER_INSERT;
const STICKER_SET = Symbol('STICKER_SET');
exports.STICKER_SET = STICKER_SET;
const STICKER_PANEL_SET_VISIBLE = Symbol('STICKER_PANEL_SET_VISIBLE');
exports.STICKER_PANEL_SET_VISIBLE = STICKER_PANEL_SET_VISIBLE;
const STICKER_MOVE_CARET = Symbol('STICKER_MOVE_CARET');
exports.STICKER_MOVE_CARET = STICKER_MOVE_CARET;
const STICKER_SET_RANGE = Symbol('STICKER_SET_RANGE');
exports.STICKER_SET_RANGE = STICKER_SET_RANGE;
const INSERT_MARKDOWN_EMOJI = Symbol('INSERT_MARKDOWN_EMOJI');
exports.INSERT_MARKDOWN_EMOJI = INSERT_MARKDOWN_EMOJI;

function insertSticker(payload) {
  return {
    type: STICKER_INSERT,
    payload
  };
}

function setSticker(payload) {
  return {
    type: STICKER_SET,
    payload
  };
}

function setStickerPanelVisible(payload) {
  return {
    type: STICKER_PANEL_SET_VISIBLE,
    payload
  };
}

function moveCaret(payload) {
  return {
    type: STICKER_MOVE_CARET,
    payload
  };
}

function setRange(payload) {
  return {
    type: STICKER_SET_RANGE,
    payload
  };
}

function insertMarkdownEmoji(payload) {
  return {
    type: INSERT_MARKDOWN_EMOJI,
    payload
  };
}
