// @ali/4ever-bi/dist-node/plugins/sticker/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "MOVE_SET", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.MOVE_SET;
  }
});
Object.defineProperty(exports, "STICKER_NODE_TYPE", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.STICKER_NODE_TYPE;
  }
});
Object.defineProperty(exports, "STICKER_PANEL_COLUMMS", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.STICKER_PANEL_COLUMMS;
  }
});
Object.defineProperty(exports, "STICKER_PANEL_HEIGHT", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.STICKER_PANEL_HEIGHT;
  }
});
Object.defineProperty(exports, "STICKER_PANEL_STYLE", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.STICKER_PANEL_STYLE;
  }
});
Object.defineProperty(exports, "STICKER_PLACEHOLDER_DATA", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.STICKER_PLACEHOLDER_DATA;
  }
});
Object.defineProperty(exports, "STICKER_SZ_IN_PANEL", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.STICKER_SZ_IN_PANEL;
  }
});
Object.defineProperty(exports, "DEFAULT_LANG", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.DEFAULT_LANG;
  }
});
Object.defineProperty(exports, "DATA_STICKER", {
  enumerable: true,
  get: function () {
    return _everPluginSticker.DATA_STICKER;
  }
});

var _everPluginSticker = require("@ali/4ever-plugin-sticker");
