// @ali/4ever-bi/dist-node/plugins/sticker/getSnippetConfig.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _everCangjie = require("@ali/4ever-cangjie");

var _everPluginSticker = require("@ali/4ever-plugin-sticker");

var _everUtils = require("@ali/4ever-utils");

const _createElement = /*#__PURE__*/_react.default.createElement;

const EmojiSpan = /*#__PURE__*/_styledComponents.default.span(["display:inline-block;width:20px;height:20px;font-size:20px;margin-right:8px;"]);

const TitleSpan = /*#__PURE__*/_styledComponents.default.span(["display:inline-block;font-size:13px;color:#111F2C;.matched{color:#0089FF;}"]);

const getSnippetConfig = (_, next) => {
  const trigger = ':([^:\\W]|[\+\-])+$';
  const config = {
    trigger,
    triggerPrefix: '([^:]|^)$',
    fetchItems: async () => {
      const emoji = await _everPluginSticker.OldBambooPlugin.getEmojiData();
      return Object.keys(emoji).map(key => ({
        key,
        title: /*#__PURE__*/_createElement(TitleSpan, null, ":", key, ":"),
        prefix: /*#__PURE__*/_createElement(EmojiSpan, null, emoji[key]),
        value: emoji[key]
      }));
    },
    filterItems: async (text, items) => {
      const searchText = text.slice(1); // 性能考虑：至多筛选出 50 条

      return await (0, _everUtils.fuzzySearch)(items, searchText, {
        key: 'key',
        maxResults: 50
      });
    },
    insertSnippet: (controller, item) => {
      controller.command(_everCangjie.Commands.insertText, item.value);
      controller.dispatch('createTriggerMarkdownAction', {
        mdType: 'emoji',
        from: 'snippet'
      });
    }
  };
  return {
    [trigger]: config,
    ...next()
  };
};

var _default = getSnippetConfig;
exports.default = _default;
