// @ali/we-design/dist-web/components/spin/spinGrid.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { SpinGridCommonLine, SpinGridProgressLine, SpinGridWrapper, SpinGridCardWrapper, SpinGridLinesWrapper, SpinGridSimpleWrapper, SpinErrorIcon } from "./styled";
export var SpinGridBaseLine = function SpinGridBaseLine(props) {
  var showProgress = typeof props.percent === 'number';
  return /*#__PURE__*/_createElement(SpinGridCommonLine, {
    style: props.style,
    blink: props.blink && !showProgress
  }, showProgress && /*#__PURE__*/_createElement(SpinGridProgressLine, {
    percent: props.percent,
    blink: props.blink
  }));
};
SpinGridBaseLine.defaultProps = {
  blink: true
};
export var SpinGridInline = function SpinGridInline(props) {
  var style = props.style,
      testid = props.testid;
  return /*#__PURE__*/_createElement(SpinGridWrapper, {
    style: style,
    "data-testid": testid
  }, /*#__PURE__*/_createElement(SpinGridCommonLine, {
    blink: false,
    style: {
      width: 12,
      marginRight: 4
    }
  }), /*#__PURE__*/_createElement(SpinGridCommonLine, {
    style: {
      width: '60%'
    }
  }));
};
export var SpinGridCard = function SpinGridCard(props) {
  var style = props.style,
      icon = props.icon,
      blink = props.blink,
      testid = props.testid;
  return /*#__PURE__*/_createElement(SpinGridCardWrapper, {
    style: style,
    "data-testid": testid
  }, !icon ? /*#__PURE__*/_createElement(SpinGridCommonLine, {
    blink: false,
    style: {
      width: 40,
      height: 40,
      marginRight: 12
    }
  }) : /*#__PURE__*/_createElement(SpinErrorIcon, {
    className: "highlight-icon",
    style: {
      marginRight: 12
    }
  }, icon), /*#__PURE__*/_createElement(SpinGridLinesWrapper, null, /*#__PURE__*/_createElement(SpinGridCommonLine, {
    blink: blink,
    style: {
      width: '30%',
      marginBottom: 12
    }
  }), /*#__PURE__*/_createElement(SpinGridCommonLine, {
    blink: blink,
    style: {
      width: '60%'
    }
  })));
};
export var SpinGridPreviewCard = function SpinGridPreviewCard(props) {
  var style = props.style,
      testid = props.testid;
  return /*#__PURE__*/_createElement(SpinGridCardWrapper, {
    style: style,
    "data-testid": testid
  }, /*#__PURE__*/_createElement(SpinGridLinesWrapper, null, /*#__PURE__*/_createElement(SpinGridCommonLine, {
    blink: props.blink,
    style: {
      width: '30%',
      marginBottom: 28
    }
  }), /*#__PURE__*/_createElement(SpinGridCommonLine, {
    blink: props.blink,
    style: {
      width: '80%',
      marginBottom: 12
    }
  }), /*#__PURE__*/_createElement(SpinGridCommonLine, {
    blink: props.blink,
    style: {
      width: '40%'
    }
  })));
};
export var SpinGridSimpleCard = function SpinGridSimpleCard(props) {
  var style = props.style,
      testid = props.testid;
  return /*#__PURE__*/_createElement(SpinGridSimpleWrapper, {
    style: style,
    "data-testid": testid
  }, /*#__PURE__*/_createElement(SpinGridBaseLine, {
    blink: props.blink,
    style: {
      width: '30%',
      height: 8,
      minWidth: 8
    },
    percent: props.percent
  }));
};
