// @ali/4ever-plugin-formula/dist-node/mo/serializer/htmlToJsonML.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _uuid = require("uuid");

var _everCangjie = require("@ali/4ever-cangjie");

var _default = {
  name: 'formula',
  onOpenTag: (state, name, attrs) => {
    const {
      value,
      mode,
      type
    } = attrs;
    if (name !== 'span' || type !== 'formula') return false;
    const metadata = {
      id: (0, _uuid.v4)(),
      type: 'application/x-alidocs-plugin-formula',
      formula: value
    };
    const formulaTag = ['tag', {
      tagType: 'hetu',
      metadata
    }, (0, _everCangjie.createTextWithString)('')];
    const formula = mode === 'inline' ? formulaTag : ['p', {
      jc: 'center'
    }, (0, _everCangjie.createTextWithString)(''), formulaTag, (0, _everCangjie.createTextWithString)('')];
    state.push(formula);
    return true;
  },
  onCloseTag: (state, name, attrs) => {
    if (name !== 'span' || attrs.type !== 'formula') return false;
    const formula = state.pop();
    state.append(formula);
    return true;
  }
};
exports.default = _default;
