// @ali/4ever-plugin-pc-contex-menu/dist-web/utils/styled.js
import styled from 'styled-components';
import { constants } from '@ali/4ever-cangjie';
import { toPx } from "./util";
var MENU_SIZE = 174;
export var MenuWrapper = /*#__PURE__*/styled.div(["position:absolute;z-index:", ";opacity:0;background-color:#fff;box-shadow:0px 0px 1px rgba(0,0,0,0.24),0px 8px 24px rgba(0,0,0,0.16);border-radius:4px;will-change:top,right,bottom,left;"], constants.Z_INDEX_CONTEXTMENU);
export var MenuGroup = /*#__PURE__*/styled.div([""]);
export var MenuItem = /*#__PURE__*/styled.span(["display:", ";flex-direction:row;align-items:center;width:", ";border-radius:8px;"], function (_ref) {
  var inline = _ref.inline;
  return inline ? 'inline-flex' : 'flex';
}, function (_ref2) {
  var inline = _ref2.inline;
  return inline ? 'auto' : toPx(MENU_SIZE);
});
export var Icon = /*#__PURE__*/styled.span(["margin-right:4p;line-height:0;"]);
export var Name = /*#__PURE__*/styled.span([""]);
export var HotKey = /*#__PURE__*/styled.span(["color:", ";"], function (_ref3) {
  var disabled = _ref3.disabled;
  return disabled ? 'unset' : 'rgba(23, 26, 29, 0.4)';
});
export var Inflat = /*#__PURE__*/styled.span(["flex:1;"]);
