// @ali/4ever-bi/dist-node/plugins/vertAlign/components/next/subButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SubButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everComponent = require("@ali/4ever-component");

var _actions = require("../../actions");

var _utils = require("../../utils");

const _createElement = /*#__PURE__*/React.createElement;
;
;

function SubButton(props) {
  const {
    controller,
    locale,
    tooltip: tooltipProp,
    ...rest
  } = props;
  const isActive = controller.query('isSub');
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInMention') || controller.query('isSelectionInLink') || controller.query('isSelectionInListSymbol') || controller.query('isSelectionInLinkCard');
  const handleClick = React.useCallback(() => controller.run('onAction', (0, _actions.toggleSub)()), [controller]);
  const tooltip = React.useMemo(() => ({
    title: locale?.subButtonTitle || '',
    shortCut: _utils.SHORTCUT.sub,
    description: `Markdown: ~${locale?.text || 'Text'}~ ${locale?.space || 'Space'}`,
    ...tooltipProp
  }), [tooltipProp]);
  return /*#__PURE__*/_createElement(_everComponent.SubscriptButton, (0, _extends2.default)({
    testid: "bi-toolbar-sub",
    active: isActive,
    disabled: !!isDisabled,
    onClick: handleClick,
    title: locale?.subButtonTitle,
    tooltip: tooltip,
    role: _everBamboo.PluginRoles.vertAlignSub
  }, rest));
}
