// @ali/4ever-bi/dist-node/plugins/vertAlign/components/next/supButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SupButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everComponent = require("@ali/4ever-component");

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../../actions");

var _utils = require("../../utils");

const _createElement = /*#__PURE__*/React.createElement;
;
;
;

function SupButton(props) {
  const {
    controller,
    locale,
    tooltip: tooltipProp,
    ...rest
  } = props;
  const isActive = controller.query('isSup');
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInMention') || controller.query('isSelectionInLink') || controller.query('isSelectionInListSymbol') || controller.query('isSelectionInLinkCard');
  const handleClick = React.useCallback(() => controller.run('onAction', (0, _actions.toggleSup)()), [controller]);
  const tooltip = React.useMemo(() => ({
    title: locale?.supButtonTitle || '',
    shortCut: _utils.SHORTCUT.sup,
    description: `Markdown: ^${locale?.text || 'Text'}^ ${locale?.space || 'Space'}`,
    ...tooltipProp
  }), [tooltipProp]);
  return /*#__PURE__*/_createElement(_everComponent.SuperscriptButton, (0, _extends2.default)({
    testid: "bi-toolbar-sup",
    active: isActive,
    disabled: !!isDisabled,
    onClick: handleClick,
    role: _everBamboo.PluginRoles.vertAlignSup,
    title: locale?.supButtonTitle,
    tooltip: tooltip
  }, rest));
}
