// @ali/4ever-plugin-list/dist-node/bi/utils/getListData.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getListData;

var _regexp = require("../regexp");

var _listStyleMapping = require("../../utils/listStyleMapping");

/**
 * 根据指定的参数生成合法的 list 数据
 *
 * @export
 * @param {GetListDataParams} { type, level, listId, isChecked, listStyleType, symbolStyle }
 * @returns {{ list: ListProperties }}
 */
function getListData({
  type,
  level,
  listId,
  isChecked,
  listStyleType,
  symbolStyle
}) {
  const data = {
    listId,
    level,
    isOrdered: type === 'ol',
    isTaskList: type === 'tlist',
    isChecked: type === 'tlist' ? !!isChecked : undefined,
    listStyleType,
    symbolStyle
  };

  if (listStyleType) {
    if (_listStyleMapping.mapping[listStyleType]) {
      const [format, text, align] = _listStyleMapping.mapping[listStyleType][0]; // 默认取第一组

      data.listStyle = {
        format,
        text,
        align
      };
    } else if (_regexp.STICKER_CODE_REG.test(listStyleType)) {
      data.listStyle = {
        format: 'bullet',
        text: listStyleType,
        align: 'start'
      };
    }
  }

  return {
    list: data
  };
}
