// @ali/we-design-next/dist-web/display/tabs/TabsBarTab.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { IconClose } from "../../basic/icon";
import { TAB_MARGIN, TAB_BOTTOM_MARGIN } from "./constants";
import { mergeClassName, COLORS } from "../../common";
var TabsTabWrapper = /*#__PURE__*/styled.div(["position:relative;padding-right:", "px;margin-bottom:", "px;cursor:default;font-size:14px;line-height:20px;font-weight:400;user-select:none;color:", ";&:hover{color:", ";.wdn-tabs-tab-close-wrapper{display:block;cursor:pointer;}}&.wdn-tabs-tab-active{color:", ";text-shadow:0 0 .25px currentcolor;}&.wdn-tabs-tab-disabled{color:rgba(23,26,29,0.24) !important;cursor:not-allowed !important;}"], TAB_MARGIN, TAB_BOTTOM_MARGIN, COLORS.LIGHT.LEVEL2_BASE_COLOR, COLORS.LIGHT.LEVEL1_BASE_COLOR, COLORS.LIGHT.LEVEL1_BASE_COLOR);
var TitleWrapper = /*#__PURE__*/styled.span(["display:inline-block;"]);
var TabsTabContent = /*#__PURE__*/styled.div(["display:flex;align-items:center;padding:8px 0;"]);
var CloseWrapper = /*#__PURE__*/styled.div(["display:none;position:absolute;top:50%;right:", "px;transform:translate(100%,-50%);font-size:0;line-height:0;.wdn-icon{width:16px;height:16px;}"], TAB_MARGIN);
var TabsTabPrefix = /*#__PURE__*/styled.div(["display:flex;align-items:center;margin-right:6px;"]);
export var TabsBarTab = function TabsBarTab(props) {
  var activeKey = props.activeKey,
    onClose = props.onClose,
    title = props.title,
    tabKey = props.tabKey,
    prefix = props.prefix,
    onClick = props.onClick,
    onRef = props.onRef,
    disabled = props.disabled;
  var isActive = activeKey === tabKey;
  var handleClick = React.useCallback(function (e) {
    if (disabled) {
      return;
    }
    onClick(tabKey);
  }, [tabKey, disabled]);
  var handleRef = React.useCallback(function (elem) {
    onRef(tabKey, elem);
  }, [tabKey]);
  return /*#__PURE__*/_createElement(TabsTabWrapper, {
    className: mergeClassName(isActive ? 'wdn-tabs-tab-active' : '', disabled ? 'wdn-tabs-tab-disabled' : '', 'wdn-tabs-tab-wrapper')
  }, /*#__PURE__*/_createElement(TabsTabContent, {
    onClick: handleClick,
    ref: handleRef,
    "data-tab-key": tabKey
  }, prefix && /*#__PURE__*/_createElement(TabsTabPrefix, null, prefix), /*#__PURE__*/_createElement(TitleWrapper, null, title)), onClose && /*#__PURE__*/_createElement(CloseWrapper, {
    className: "wdn-tabs-tab-close-wrapper",
    onClick: onClose
  }, /*#__PURE__*/_createElement(IconClose, {
    style: {
      fontSize: 8
    }
  })));
};
