// @ali/we-design-next/dist-web/display/tabs/TabsContent.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { mergeClassName } from "../../common";
var TabsContentHolder = /*#__PURE__*/styled.div(["display:flex;position:relative;flex:auto;min-width:0;min-height:0;overflow:", ";"], function (p) {
  return p.animated ? 'hidden' : 'initial';
});
var TabsContentWrapper = /*#__PURE__*/styled.div(["display:flex;width:100%;"]);
var TabPanel = /*#__PURE__*/styled.div(["flex:none;width:100%;&:not(.wdn-animated-tab-panel).wdn-inactive-tab-panel{display:none;}&.wdn-animated-tab-panel.wdn-inactive-tab-panel{height:0;overflow-y:hidden;visibility:hidden;}"]);
export var TabsContent = function TabsContent(props) {
  var children = props.children,
    activeKey = props.activeKey,
    animated = props.animated,
    destroyInactiveTabPanel = props.destroyInactiveTabPanel;
  var newChildren = [];
  var index = 0;
  var activeIndex = 0;
  React.Children.forEach(children, function (child) {
    if (!child || ! /*#__PURE__*/React.isValidElement(child)) {
      return;
    }
    var isActive = activeKey === child.key;
    if (isActive) {
      activeIndex = index;
    }
    newChildren.push( /*#__PURE__*/_createElement(TabPanel, {
      className: mergeClassName(!isActive ? 'wdn-inactive-tab-panel' : '', animated ? 'wdn-animated-tab-panel' : ''),
      key: child.key
    }, isActive || !destroyInactiveTabPanel ? child : null));
    index++;
  });
  var tabsContentStyle = animated ? {
    marginLeft: "-" + activeIndex * 100 + "%",
    transition: 'margin 0.3s ease'
  } : undefined;
  return /*#__PURE__*/_createElement(TabsContentHolder, {
    animated: animated
  }, /*#__PURE__*/_createElement(TabsContentWrapper, {
    style: tabsContentStyle
  }, newChildren));
};
