// @ali/we-design-next/dist-web/display/tabs/Tabs.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
var _excluded = ["children", "onActiveTabChange", "activeKey", "defaultActiveKey", "animated", "keepMounted"];
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { TabsBar } from "./TabsBar";
import { TabsContent } from "./TabsContent";
import { useDefaultProp } from "../../common";
var TabsWrapper = /*#__PURE__*/styled.div(["display:flex;flex-direction:column;"]);
var getTabBarTabList = function getTabBarTabList(panes) {
  var tabList = [];
  React.Children.forEach(panes, function (child) {
    if (child && /*#__PURE__*/React.isValidElement(child) && child.type.displayName === 'TabPanel') {
      var _child$props = child.props,
        title = _child$props.title,
        onClose = _child$props.onClose,
        disabled = _child$props.disabled,
        prefix = _child$props.prefix;
      tabList.push({
        key: child.key,
        title: title,
        onClose: onClose,
        disabled: disabled,
        prefix: prefix
      });
    }
  });
  return tabList;
};
export var Tabs = /*#__PURE__*/React.memo( /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _tabs$;
  var children = props.children,
    onActiveTabChange = props.onActiveTabChange,
    activeKeyProp = props.activeKey,
    defaultActiveKey = props.defaultActiveKey,
    animated = props.animated,
    _props$keepMounted = props.keepMounted,
    keepMounted = _props$keepMounted === void 0 ? true : _props$keepMounted,
    rest = _objectWithoutPropertiesLoose(props, _excluded);
  var tabs = React.useMemo(function () {
    return getTabBarTabList(children);
  }, [children]);
  var _useDefaultProp = useDefaultProp(defaultActiveKey || ((_tabs$ = tabs[0]) == null ? void 0 : _tabs$.key), activeKeyProp, onActiveTabChange),
    activeKey = _useDefaultProp[0],
    setActive = _useDefaultProp[1];
  return /*#__PURE__*/_createElement(TabsWrapper, _extends({
    ref: ref
  }, rest), /*#__PURE__*/_createElement(TabsBar, {
    tabs: tabs,
    onActiveTabChange: setActive,
    activeKey: activeKey,
    defaultActiveKey: defaultActiveKey,
    animated: animated
  }), /*#__PURE__*/_createElement(TabsContent, {
    activeKey: activeKey,
    animated: animated,
    destroyInactiveTabPanel: !keepMounted
  }, children));
}));
