// @ali/4ever-plugin-subject/dist-web/utils/handlers/createOnCangjieSelect.js
import { Block, CangjieSelectEvent, Injection } from '@ali/4ever-cangjie';
import { toggleFold } from '@ali/4ever-utils';
import { getSubjectEndBlock, getSubjectStartBlock } from "../queries";
import { isSubjectHeading } from "../utils";
var SUBJECT_NORMALIZE = 'subjectNormalize';

function createOnCangjieSelect(options) {
  if (options === void 0) {
    options = {};
  }

  var _options = options,
      _options$inject = _options.inject,
      inject = _options$inject === void 0 ? false : _options$inject;
  return function onCangjieSelect(event, controller, next) {
    var _event$detail = event.detail,
        selection = _event$detail.selection,
        trigger = _event$detail.trigger;
    var _controller$value = controller.value,
        document = _controller$value.document,
        injections = _controller$value.injections;
    var focus = selection.focus;
    var block = document.getFurthestAncestor(focus.key);
    /**
     * Effects 1: 若光标 focus 进入到折叠议题区域内，需要自动展开议题
     * @description 此处需求是由于在内容折叠情况下，若用户光标选区进入，则无法看到鼠标聚焦区域，所以需要自动展开
     */
    // TODO: 这里的 query 会在每次 select 时消耗一次 O(N) 的查询性能，后续考虑在 DOM 上做一次 O(h) 的查询来优化

    if (trigger !== SUBJECT_NORMALIZE && Block.isBlock(block) && !isSubjectHeading(block)) {
      var subject = controller.query(getSubjectStartBlock, block); // 注入 Injection 数据

      if (inject && subject) {
        subject = Injection.injectIntoNode(injections, subject);
      } // 若议题节点折叠则展开


      if (subject && controller.userData.get(subject, 'fold')) {
        controller.run('onAction', toggleFold({
          node: subject,
          fold: false
        }));
      }
    }
    /**
     * Effects 2: 若选区跨越议题范围，则需要扩展选区范围至整个议题
     * @description 此处需求源于产品设计，议题标题不能被随意删除，由于标题的特例处理比较困难，因此我们通过扩展选区的方法使得用户
     * 会选择整个议题，这样使得用户在操作时能有明显的预期
     */


    var start = selection.getStart(document);
    var end = selection.getEnd(document);
    var startBlock = document.getFurthsestBlock(start.key);
    var endBlock = document.getFurthsestBlock(end.key); // 若光标的起始和结束不在同一个 Block 上

    if (trigger !== SUBJECT_NORMALIZE && startBlock && endBlock && startBlock !== endBlock) {
      var sel = selection;
      var startSubjectBlock = controller.query(getSubjectStartBlock, startBlock);
      var endSubjectBlock = controller.query(getSubjectStartBlock, endBlock); // 若光标的起始和结束不在同一个议题组内，则需要扩展选区

      if (startSubjectBlock !== endSubjectBlock) {
        // start 移动至起始议题的开头
        if (startSubjectBlock && !start.isAtStartOfNode(startSubjectBlock)) {
          sel = sel.moveStartToStartOfNode(startSubjectBlock, controller);
        } // end 移动至结束议题的末尾


        if (endSubjectBlock) {
          var subjectEndBlock = controller.query(getSubjectEndBlock, endSubjectBlock);

          if (subjectEndBlock && !end.isAtEndOfNode(subjectEndBlock)) {
            sel = sel.moveEndToEndOfNode(subjectEndBlock, controller);
          }
        }
      } // 若光标发生了变化，则更新


      if (sel !== selection) {
        return controller.run('onCangjieSelect', CangjieSelectEvent({
          selection: sel,
          trigger: SUBJECT_NORMALIZE
        }));
      }
    }

    return next();
  };
}

export default createOnCangjieSelect;
