// @ali/4ever-plugin-subject/dist-web/utils/handlers/createGroupNodes.js
import { Document } from '@ali/4ever-cangjie';
import { createSubject, isSubject, isSubjectHeading } from "../utils";

function createGroupNodes() {
  return function groupNodes(_ref, ctx, controller, next) {
    var start = _ref[0],
        end = _ref[1];
    var parent = ctx.parent,
        groups = ctx.groups;
    var first = parent.nodes[start];
    var parentGroup = groups[groups.length - 1]; // 如果不是一级节点，或者是文档末尾，则不分组

    if (!Document.isDocument(parent)) {
      return next();
    }

    if (isSubjectHeading(first) && !isSubject(parentGroup)) {
      var i = start + 1;

      while (i < end) {
        var node = parent.nodes[i];

        if (isSubjectHeading(node)) {
          break;
        }

        i += 1;
      }

      var children = parent.nodes.slice(start, i);
      var subject = createSubject({
        key: first.key + "-group",
        data: first.data,
        nodes: children
      });
      return {
        node: subject,
        remainedRange: [i, end]
      };
    }

    return next();
  };
}

export default createGroupNodes;
