// @ali/4ever-bi/dist-web/plugins/embed/mobile/createRenderNode.js
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import Embed from "./embed";
import { UploadOrFailedCard } from '@ali/4ever-factory';
export default function createRenderNode(config) {
  var embedConfig = config.embed || {};
  var embedLocaleConfig = embedConfig.locale,
      downloadFile = embedConfig.downloadFile,
      onPreview = embedConfig.onPreview,
      _embedConfig$enableFi = embedConfig.enableFilePreview,
      enableFilePreview = _embedConfig$enableFi === void 0 ? false : _embedConfig$enableFi;
  return {
    embed: function embed(props, controller, next) {
      var node = props.node;

      if (!node.data.src) {
        return /*#__PURE__*/_createElement(UploadOrFailedCard, _extends({}, props, {
          node: node,
          locale: embedLocaleConfig,
          controller: controller
        }));
      }

      if (node.data.type !== 'video' && node.data.type !== 'audio') {
        return /*#__PURE__*/_createElement(Embed, _extends({}, props, {
          controller: controller,
          onDownload: downloadFile,
          onPreview: onPreview,
          locale: embedLocaleConfig,
          enableFilePreview: enableFilePreview
        }));
      }

      return next();
    }
  };
}
