// @ali/4ever-bi/dist-web/plugins/indent/components/next/createIncreaseIndentButton.js
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { PluginRoles } from '@ali/4ever-bamboo';
;
import { IconIncreaseIndent } from '@ali/4ever-component';
;
import { INCREASE_HOT_KEY_TIP } from "../../utils/hotKeyTips";
import { increaseIndent } from "../../actions";
;

var _ref = /*#__PURE__*/_createElement(IconIncreaseIndent, null);

export default function createIncreaseIndentButton(props) {
  var controller = props.controller,
      locale = props.locale,
      icon = props.icon,
      title = props.title,
      tooltip = props.tooltip,
      badgeDot = props.badgeDot,
      disabled = props.disabled;

  var handleClick = function handleClick() {
    return controller.run('onAction', increaseIndent());
  };

  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var isDisabled = disabled || isCollapsedAtEdge || controller.query('isSelectionInHr') || controller.query('isSelectionInListSymbol');
  var buttonConfig = {
    icon: icon || _ref,
    onClick: handleClick,
    title: title || (locale == null ? void 0 : locale.rightButtonTitle),
    tooltip: _extends({
      title: title || (locale == null ? void 0 : locale.rightButtonTitle),
      shortCut: INCREASE_HOT_KEY_TIP
    }, tooltip),
    badgeDot: badgeDot,
    disabled: isDisabled,
    role: PluginRoles.indentIncrease,
    testid: "bi-toolbar-incindent"
  };
  return buttonConfig;
}
