// @ali/4ever-plugin-container/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.moFactory;
  }
});
Object.defineProperty(exports, "Container", {
  enumerable: true,
  get: function () {
    return _models.Container;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index2.createServerSerializationPlugin;
  }
});
exports.mobileBiFactory = exports.mobileZhiFactory = exports.biFactory = exports.zhiFactory = void 0;

var _index = require("./mo/index");

var _models = require("./mo/models");

var _index2 = _interopRequireWildcard(require("./mo/serializer/index"));

// facotry
const zhiFactory = _index.moFactory;
exports.zhiFactory = zhiFactory;
const biFactory = _index.moFactory;
exports.biFactory = biFactory;
const mobileZhiFactory = _index.moFactory;
exports.mobileZhiFactory = mobileZhiFactory;
const mobileBiFactory = _index.moFactory; // types
// other export for api compatibility

exports.mobileBiFactory = mobileBiFactory;
