// @ali/4ever-bi/dist-node/plugins/voteCard/handlers/createMenu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _everCangjie = require("@ali/4ever-cangjie");

var _dayjs = _interopRequireDefault(require("dayjs"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everMo = require("@ali/4ever-mo");

var _weDesign = require("@ali/we-design");

var _everFactory = require("@ali/4ever-factory");

const _createElement = /*#__PURE__*/_react.default.createElement;
const DEFAULT_DATA = {
  title: '',
  voteType: _everBamboo.VoteCardPlugin.EnumVoteType.RADIO,
  optionInfoList: [{
    content: '',
    pos: 1,
    optionTicket: 0,
    userOptionTicket: 0
  }, {
    content: '',
    pos: 2,
    optionTicket: 0,
    userOptionTicket: 0
  }],
  endTime: (0, _dayjs.default)().add(7, 'days').valueOf(),
  showTicketBeforeVoting: false
};

const createMenubar = () => {
  return function CalendarCardMenu(props) {
    const {
      menuItem,
      controller
    } = props;
    const {
      title,
      ...rest
    } = menuItem;
    const handleClick = (0, _react.useCallback)(() => {
      const node = _everCangjie.Block.create({
        type: _everMo.Card.TYPE,
        data: {
          cardType: 'vote',
          height: 400,
          metadata: DEFAULT_DATA
        }
      });

      controller.command(_everFactory.insertCardCommand, node).command(_everFactory.setCardInjectionCommand, node, {
        editing: true
      }).setIsFocused(false);
    }, [controller]);
    return /*#__PURE__*/_createElement(_weDesign.MenuItem, (0, _extends2.default)({}, rest, {
      onClick: handleClick
    }), title || '');
  };
};

var _default = createMenubar;
exports.default = _default;
