// @ali/4ever-bi/dist-node/plugins/listGroup/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everBamboo = require("@ali/4ever-bamboo");

var _everFactory = require("@ali/4ever-factory");

const DRAG_BUTTON_OFFSET = -20;
const {
  createRenderNode,
  createGroupNodes
} = _everBamboo.ListGroupPlugin.handlers;
const {
  isFoldGroupNode
} = _everBamboo.ListGroupPlugin.utils;
const ListGroupPlugin = (0, _everFactory.groupFactory)(config => {
  const {
    listGroup = {}
  } = config;
  return {
    groupType: 'listGroup',
    renderNode: createRenderNode(listGroup),
    groupNodes: createGroupNodes(listGroup),
    queries: {
      isListGroupAvailable: () => listGroup.ignore ? !listGroup.ignore() : true,
      getLeftToolbarOffset: (controller, data) => {
        const {
          node
        } = data;

        if (node?.data?.list?.level === 0 && isFoldGroupNode(node, controller)) {
          return {
            // 这里加上4是因为左侧工具栏的offset为6，但是在标题场景下是2
            x: DRAG_BUTTON_OFFSET + 4,
            y: 0
          };
        }

        return undefined;
      }
    }
  };
});
var _default = ListGroupPlugin;
exports.default = _default;
