// @ali/4ever-bi/dist-node/plugins/callout/components/Menu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inlineMenu = exports.menu = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _actions = require("../actions");

const _createElement = /*#__PURE__*/React.createElement;
const CALLOUT_MENU_STYLE = {
  whiteSpace: 'nowrap'
};

function WithCalloutMenu(Component) {
  return function CalloutMenu(props) {
    const {
      menuItem,
      controller
    } = props;
    const {
      action,
      title,
      ...rest
    } = menuItem;
    const {
      value
    } = controller;

    const handleClick = () => {
      if (action === 'format') {
        return controller.run('onAction', (0, _actions.transformCallout)());
      }

      controller.run('onAction', (0, _actions.insertCallout)());
    };

    const isDisabled = React.useMemo(() => controller.query('isSelectionInImage') || controller.query('isSelectionInLinkCard') || controller.query('isSelectionInSubjectHeading'), [controller, value]);
    return /*#__PURE__*/_createElement(Component, (0, _extends2.default)({
      disabled: isDisabled,
      style: CALLOUT_MENU_STYLE
    }, rest, {
      // @ts-ignore
      onClick: handleClick
    }), title || '');
  };
} // @ts-ignore


const menu = WithCalloutMenu(_weDesign.MenuItem); // @ts-ignore

exports.menu = menu;
const inlineMenu = WithCalloutMenu(_weDesign.InlineMenuItem);
exports.inlineMenu = inlineMenu;
