// @ali/4ever-bi/dist-node/plugins/callout/components/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmptyContainer = exports.Placeholder = exports.CalloutColorSettingItem = exports.CalloutClearSettingPanelWrapper = exports.CalloutColorSettingPanelLineWrapper = exports.CalloutColorSettingPanelWrapper = exports.CalloutSettingPanelWrapper = exports.CalloutInnerToolbarWrapper = exports.CalloutInnerToolbarOuterWrapper = exports.PaddingHolder = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

const PaddingHolder = /*#__PURE__*/_styledComponents.default.div(["display:flex;visibility:", ";opacity:", ";flex-direction:row;padding:0;position:absolute;z-index:200;top:3px;right:3px;"], props => props.showInnerToolbar ? 'visible' : 'hidden', props => props.visibile ? '1' : '0');

exports.PaddingHolder = PaddingHolder;

const CalloutInnerToolbarOuterWrapper = /*#__PURE__*/_styledComponents.default.div(["padding:", ";background-color:", ";border-radius:4px;"], ({
  bgcolor
}) => bgcolor === 'white' ? '0' : '1px', ({
  bgcolor
}) => bgcolor);

exports.CalloutInnerToolbarOuterWrapper = CalloutInnerToolbarOuterWrapper;

const CalloutInnerToolbarWrapper = /*#__PURE__*/_styledComponents.default.div(["cursor:pointer;border-radius:4px;transition:background-color 250ms;background-color:rgba(126,134,142,0.15);&:hover{background-color:rgba(126,134,142,0.12);}& div .icon-button{cursor:pointer;}"]);

exports.CalloutInnerToolbarWrapper = CalloutInnerToolbarWrapper;

const CalloutSettingPanelWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;flex-direction:column;width:125px;height:100px;border-radius:8px;background:#ffffff;box-shadow:0px 0px 1px rgba(0,0,0,0.24),0px 8px 24px rgba(0,0,0,0.16);"]);

exports.CalloutSettingPanelWrapper = CalloutSettingPanelWrapper;

const CalloutColorSettingPanelWrapper = /*#__PURE__*/_styledComponents.default.div(["display:inline-block;flex-direction:column;justify-content:center;align-items:center;flex-wrap:wrap;width:125px;height:71px;padding-top:7px;"]);

exports.CalloutColorSettingPanelWrapper = CalloutColorSettingPanelWrapper;

const CalloutColorSettingPanelLineWrapper = /*#__PURE__*/_styledComponents.default.div(["display:inline-block;width:125px;padding-left:7px;"]);

exports.CalloutColorSettingPanelLineWrapper = CalloutColorSettingPanelLineWrapper;

const CalloutClearSettingPanelWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;flex-direction:row;justify-content:center;align-items:center;width:121px;height:28px;color:#0089ff;border-top:solid 1px rgba(126,134,142,0.16);font-size:12px;margin:0 2px;cursor:pointer;"]);

exports.CalloutClearSettingPanelWrapper = CalloutClearSettingPanelWrapper;

const CalloutColorSettingItem = /*#__PURE__*/_styledComponents.default.div(["box-sizing:content-box;display:inline-block;text-align:center;line-height:22px;width:22px;height:22px;border-radius:4px;margin:2px 2px;color:rgba(23,26,29,0.4);font-size:17px;background-color:", ";border:", ";cursor:pointer;"], props => props.color, props => props.color === '#FFFFFF' ? '1px solid #EEEEEE' : '1px solid rgba(255,255,255,0)');

exports.CalloutColorSettingItem = CalloutColorSettingItem;

const Placeholder = /*#__PURE__*/_styledComponents.default.div(["position:absolute;color:rgba(23,26,29,0.4);padding-left:24px;top:0;left:2px;pointer-events:none;"]);

exports.Placeholder = Placeholder;

const EmptyContainer = /*#__PURE__*/_styledComponents.default.div([""]);

exports.EmptyContainer = EmptyContainer;
