// @ali/4ever-plugin-embed/dist-node/utils/components/common/embedPreviewHeader/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _fileTypeHelper = require("../../../fileTypeHelper");

var _styled = require("./styled");

const _createElement = /*#__PURE__*/_react.default.createElement;

function EmbedPreviewHeader(props) {
  const {
    name,
    size,
    type,
    mode,
    isMobile,
    renderExtra
  } = props;
  return /*#__PURE__*/_createElement(_styled.PreviewHeaderWrapper, {
    mode: mode
  }, /*#__PURE__*/_createElement(_styled.CardIconWrapper, null, /*#__PURE__*/_createElement("img", {
    src: (0, _fileTypeHelper.getIcon)(type),
    alt: "icon",
    style: isMobile ? {
      width: 23,
      height: 24
    } : {
      width: 16,
      height: 23
    }
  })), /*#__PURE__*/_createElement(_styled.CardContentWrapper, null, /*#__PURE__*/_createElement(_styled.CardContentInnerWrapper, null, /*#__PURE__*/_createElement(_styled.ContentHeader, {
    mode: mode
  }, name), /*#__PURE__*/_createElement(_styled.ContentDesc, {
    mode: mode
  }, (0, _fileTypeHelper.formatSize)(size)))), renderExtra && renderExtra());
}

var _default = EmbedPreviewHeader;
exports.default = _default;
