// @ali/4ever-plugin-embed/dist-node/utils/components/common/embedPreviewHeader/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ContentDesc = exports.ContentHeader = exports.CardContentInnerWrapper = exports.CardContentWrapper = exports.CardIconWrapper = exports.PreviewHeaderWrapper = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

const PreviewHeaderWrapper = /*#__PURE__*/_styledComponents.default.div(["width:100%;background-color:", ";padding:0 16px;height:32px;display:flex;border-top-left-radius:6px;border-top-right-radius:6px;line-height:23px;box-sizing:border-box;align-items:center;"], props => props.mode === 'dark' ? '#111F2C' : '#F6F8F9');

exports.PreviewHeaderWrapper = PreviewHeaderWrapper;

const CardIconWrapper = /*#__PURE__*/_styledComponents.default.div(["margin-right:16px;flex:0;line-height:100%;img{vertical-align:initial;}"]);

exports.CardIconWrapper = CardIconWrapper;

const CardContentWrapper = /*#__PURE__*/_styledComponents.default.div(["flex:1;min-width:0;"]);

exports.CardContentWrapper = CardContentWrapper;

const CardContentInnerWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;margin-top:'';"]);

exports.CardContentInnerWrapper = CardContentInnerWrapper;

const ContentHeader = /*#__PURE__*/_styledComponents.default.div(["font-size:14px;color:", ";flex:1;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;"], props => props.mode === 'dark' ? '#F2F4F5' : '#111F2C');

exports.ContentHeader = ContentHeader;

const ContentDesc = /*#__PURE__*/_styledComponents.default.div(["font-size:12px;color:", ";margin-right:16px;"], props => props.mode === 'dark' ? '#F2F4F5' : '#7A8189');

exports.ContentDesc = ContentDesc;
