// @ali/4ever-plugin-template-button/dist-node/bi/handlers/createRenderHoverToolbar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createRenderHoverToolbar = void 0;

var _react = _interopRequireDefault(require("react"));

var _everComponent = require("@ali/4ever-component");

var _role = require("../../utils/role");

var _mo = require("../../mo");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement(_everComponent.IconEdit, null);

const createRenderHoverToolbar = config => {
  const {
    locale
  } = config || {};
  return (props, controller, getNode) => {
    const {
      buttonName
    } = props;

    switch (buttonName) {
      case 'edit':
        return {
          title: locale?.editTooltip || 'Configuration Template',
          tooltip: locale?.editTooltip || 'Configuration Template',
          icon: _ref,
          onClick: () => {
            const node = getNode();
            if (!_mo.TemplateButton.isTemplateButton(node)) return;
            controller.dispatch('toggleTemplateButton', {
              key: node.key
            });
          },
          testid: 'templateButton-edit',
          role: _role.TemplateButtonRole.templateButton
        };

      default:
        return null;
    }
  };
};

exports.createRenderHoverToolbar = createRenderHoverToolbar;
