// @ali/4ever-plugin-comment/dist-node/utils/actions/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isCommentSelectedNodes = exports.commentSelectedNodes = exports.setSidebarVisible = exports.setComments = exports.setActiveId = exports.cancelComment = exports.addComment = exports.hideComment = exports.showComment = exports.ACTION_COMMENT_SELECTED_NODES = exports.ACTION_COMMENT_SET_SIDE_BAR = exports.ACTION_COMMENT_SET_COMMENTS = exports.ACTION_COMMENT_SET_ID = exports.ACTION_COMMENT_CANCEL = exports.ACTION_COMMENT_SHOW = exports.ACTION_COMMENT_ADD = exports.ACTION_COMMENT_CLICK = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

const ACTION_COMMENT_CLICK = Symbol('comment-click');
exports.ACTION_COMMENT_CLICK = ACTION_COMMENT_CLICK;
const ACTION_COMMENT_ADD = Symbol('comment-add');
exports.ACTION_COMMENT_ADD = ACTION_COMMENT_ADD;
const ACTION_COMMENT_SHOW = Symbol('comment-show');
exports.ACTION_COMMENT_SHOW = ACTION_COMMENT_SHOW;
const ACTION_COMMENT_CANCEL = Symbol('comment-cancel');
exports.ACTION_COMMENT_CANCEL = ACTION_COMMENT_CANCEL;
const ACTION_COMMENT_SET_ID = Symbol('comment-setid');
exports.ACTION_COMMENT_SET_ID = ACTION_COMMENT_SET_ID;
const ACTION_COMMENT_SET_COMMENTS = Symbol('comment-comments');
exports.ACTION_COMMENT_SET_COMMENTS = ACTION_COMMENT_SET_COMMENTS;
const ACTION_COMMENT_SET_SIDE_BAR = Symbol('comment-set-sidebar');
exports.ACTION_COMMENT_SET_SIDE_BAR = ACTION_COMMENT_SET_SIDE_BAR;
const ACTION_COMMENT_SELECTED_NODES = Symbol('comment-selected-nodes');
/**
 * 显示评论
 */

exports.ACTION_COMMENT_SELECTED_NODES = ACTION_COMMENT_SELECTED_NODES;

const showComment = (contentId, options) => ({
  type: ACTION_COMMENT_SHOW,
  payload: {
    contentId,
    options
  }
});
/**
 * 隐藏评论
 */


exports.showComment = showComment;

const hideComment = () => ({
  type: ACTION_COMMENT_SHOW,
  payload: undefined
});
/**
 * 添加评论
 */


exports.hideComment = hideComment;

const addComment = () => ({
  type: ACTION_COMMENT_ADD
});
/**
 * 取消正在添加的评论
 */


exports.addComment = addComment;

const cancelComment = () => ({
  type: ACTION_COMMENT_CANCEL
});
/**
 * 设置显示的 comment （showComment 会触发滚动，这里没有副作用）
 */


exports.cancelComment = cancelComment;

const setActiveId = (contentId, isAutoFocus = true) => ({
  type: ACTION_COMMENT_SET_ID,
  payload: {
    contentId,
    isAutoFocus
  }
});
/**
 * 设置 comments 数据
 */


exports.setActiveId = setActiveId;

const setComments = comments => ({
  type: ACTION_COMMENT_SET_COMMENTS,
  payload: {
    comments
  }
});
/**
 * 设置右侧栏是否可见
 */


exports.setComments = setComments;

const setSidebarVisible = visible => ({
  type: ACTION_COMMENT_SET_SIDE_BAR,
  payload: {
    visible
  }
});

exports.setSidebarVisible = setSidebarVisible;
const [commentSelectedNodes, isCommentSelectedNodes] = (0, _everCangjie.createActionHandlers)(ACTION_COMMENT_SELECTED_NODES);
exports.isCommentSelectedNodes = isCommentSelectedNodes;
exports.commentSelectedNodes = commentSelectedNodes;
