// @ali/we-biz-link-card/dist-web/index.js
import * as React from 'react';
import React__default, { useState, useCallback } from 'react';
import _inheritsLoose from '@babel/runtime/helpers/inheritsLoose';
import _wrapNativeSuper from '@babel/runtime/helpers/wrapNativeSuper';
import { AbstractCardLayout } from '@ali/we-toolbar';
import styled from 'styled-components';
import _extends from '@babel/runtime/helpers/extends';

var _rect, _path, _path2;
var URLLinkCardHeight = 80;
var ImagePlaceholder = function ImagePlaceholder() {
  var _useState = useState(false),
      isError = _useState[0],
      setError = _useState[1];

  var onImageError = useCallback(function () {
    setError(true);
  }, []);

  if (isError) {
    return /*#__PURE__*/React__default.createElement("svg", {
      width: "40",
      height: "40",
      viewBox: "0 0 40 40",
      style: {
        width: 64,
        height: 64
      },
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg"
    }, _rect || (_rect = /*#__PURE__*/React__default.createElement("rect", {
      width: "40",
      height: "40",
      rx: "6",
      fill: "#7E868E",
      "fill-opacity": "0.12"
    })), _path || (_path = /*#__PURE__*/React__default.createElement("path", {
      d: "M20.7071 23.5355L18.5858 25.6568C17.4142 26.8284 15.5147 26.8284 14.3432 25.6568C13.1716 24.4853 13.1716 22.5858 14.3432 21.4142L16.4645 19.2929C17.636 18.1213 19.5355 18.1213 20.7071 19.2929L21.4142 20L22.121 19.2925L21.4142 18.5858C19.8521 17.0237 17.3195 17.0237 15.7574 18.5858L13.636 20.7071C12.0739 22.2692 12.0739 24.8019 13.636 26.3639C15.1981 27.926 17.7308 27.926 19.2929 26.3639L21.4142 24.2426L20.7071 23.5355Z",
      fill: "#171A1D",
      "fill-opacity": "0.4"
    })), _path2 || (_path2 = /*#__PURE__*/React__default.createElement("path", {
      d: "M19.2929 16.4645L21.4142 14.3431C22.5782 13.1791 24.5602 13.2465 25.6569 14.3431C26.7535 15.4398 26.8209 17.4218 25.6569 18.5858L23.5355 20.7071C22.364 21.8787 20.4645 21.8787 19.2929 20.7071L18.5858 20L17.8789 20.7074L18.5858 21.4142C20.1479 22.9763 22.6805 22.9763 24.2426 21.4142L26.364 19.2929C27.9336 17.7232 27.8358 15.1079 26.364 13.636C24.8921 12.1642 22.2768 12.0664 20.7071 13.636L18.5858 15.7573L19.2929 16.4645Z",
      fill: "#171A1D",
      "fill-opacity": "0.4"
    })));
  }

  return /*#__PURE__*/React__default.createElement("img", {
    width: '64px',
    height: '64px',
    src: "https://img.alicdn.com/tfs/TB1vTtAoxvbeK8jSZPfXXariXXa-192-192.png",
    onError: onImageError
  });
};
var UnsupportedValueError = /*#__PURE__*/function (_Error) {
  _inheritsLoose(UnsupportedValueError, _Error);

  function UnsupportedValueError(value) {
    return _Error.call(this, 'Unsupported value: ' + value) || this;
  }

  return UnsupportedValueError;
}( /*#__PURE__*/_wrapNativeSuper(Error));

var _ImagePlaceholder$2;
var URLLinkCardLoading = function URLLinkCardLoading(_ref) {
  var isSelected = _ref.isSelected,
      cardStyle = _ref.cardStyle;
  // 如果是 indoc，因为有 border，所以需要减去 2px
  var height = cardStyle === 'popup' ? URLLinkCardHeight : URLLinkCardHeight - 2;
  return /*#__PURE__*/React.createElement(AbstractCardLayout, {
    isSelected: isSelected,
    height: height,
    type: "loading",
    line: "double",
    style: cardStyle === 'popup' ? {
      boxShadow: '0px 8px 24px 0px rgba(0, 0, 0, 0.16),0px 0px 1px 0px rgba(0, 0, 0, 0.24)',
      border: 'none'
    } : {},
    icon: _ImagePlaceholder$2 || (_ImagePlaceholder$2 = /*#__PURE__*/React.createElement(ImagePlaceholder, null))
  });
};

var _svg6, _svg7, _svg8;
var EditIcon = function EditIcon() {
  return _svg6 || (_svg6 = /*#__PURE__*/React__default.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    width: "20",
    height: "20",
    fill: "none",
    viewBox: "0 0 20 20"
  }, /*#__PURE__*/React__default.createElement("path", {
    fill: "#171A1D",
    fillRule: "evenodd",
    d: "M15.309 2.56a2.5 2.5 0 00-3.536 0L4.399 9.936a.5.5 0 00-.137.255l-.707 3.536a.5.5 0 00.588.588l3.536-.707a.5.5 0 00.255-.136l7.375-7.375a2.5 2.5 0 000-3.535zM5.213 10.536L11.42 4.33l2.12 2.12-6.206 6.207-2.652.53.53-2.65zm9.389-5.146l-.354.354-2.121-2.122.353-.353a1.5 1.5 0 112.121 2.121z",
    clipRule: "evenodd"
  }), /*#__PURE__*/React__default.createElement("path", {
    fill: "#171A1D",
    d: "M3 17h14v-1H3v1z"
  })));
};
var CopyIcon = function CopyIcon() {
  return _svg7 || (_svg7 = /*#__PURE__*/React__default.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    width: "20",
    height: "20",
    fill: "none",
    viewBox: "0 0 20 20"
  }, /*#__PURE__*/React__default.createElement("path", {
    fill: "#171A1D",
    d: "M8.5 2A1.5 1.5 0 007 3.5v2.182c0 .12.043.232.114.318H3.5A1.5 1.5 0 002 7.5v9A1.5 1.5 0 003.5 18h9a1.5 1.5 0 001.5-1.5V14h2.5a1.5 1.5 0 001.5-1.5v-9A1.5 1.5 0 0016.5 2h-8zM14 13V7.5A1.5 1.5 0 0012.5 6H7.886A.498.498 0 008 5.682V3.5a.5.5 0 01.5-.5h8a.5.5 0 01.5.5v9a.5.5 0 01-.5.5H14zM3 7.5a.5.5 0 01.5-.5h9a.5.5 0 01.5.5v9a.5.5 0 01-.5.5h-9a.5.5 0 01-.5-.5v-9z"
  })));
};
var UnlinkIcon = function UnlinkIcon() {
  return _svg8 || (_svg8 = /*#__PURE__*/React__default.createElement("svg", {
    width: "24",
    height: "24",
    viewBox: "0 0 24 24",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React__default.createElement("path", {
    d: "M19.0709 9.8824L15.889 13.0644L16.5961 13.7715L19.7781 10.5895C21.5354 8.83215 21.5354 5.98291 19.7781 4.22555C18.0207 2.46819 15.1715 2.46819 13.4141 4.22555L10.2321 7.40753L10.9392 8.11464L14.1212 4.93266C15.488 3.56582 17.7041 3.56582 19.0709 4.93266C20.4378 6.29949 20.4378 8.51557 19.0709 9.8824Z",
    fill: "#171A1D"
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M4.22171 13.4179L7.40369 10.236L8.11079 10.9431L4.92881 14.125C3.56198 15.4919 3.56198 17.708 4.92881 19.0748C6.29565 20.4416 8.51172 20.4416 9.87856 19.0748L13.0605 15.8928L13.7676 16.5999L10.5857 19.7819C8.82831 21.5393 5.97907 21.5393 4.22171 19.7819C2.46435 18.0245 2.46435 15.1753 4.22171 13.4179Z",
    fill: "#171A1D"
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M10.9392 13.7715L8.11073 16.5999L7.40363 15.8928L10.2321 13.0644L10.9392 13.7715Z",
    fill: "#171A1D"
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M16.596 8.11467L13.7676 10.9431L13.0605 10.236L15.8889 7.40756L16.596 8.11467Z",
    fill: "#171A1D"
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M4.64632 5.35355L18.6463 19.3536L19.3534 18.6464L5.35343 4.64645L4.64632 5.35355Z",
    fill: "#171A1D"
  })));
};

var _ImagePlaceholder$1, _CopyIcon$1, _EditIcon$1, _UnlinkIcon$1;
var URLLinkCardError = function URLLinkCardError(props) {
  var isSelected = props.isSelected,
      cardStyle = props.cardStyle,
      locale = props.locale,
      onRefresh = props.onRefresh,
      urlInfo = props.urlInfo,
      url = props.url,
      readonly = props.readonly,
      onOpenURL = props.onOpenURL,
      onSwitchStyle = props.onSwitchStyle,
      editBtnTestID = props.editBtnTestID,
      unlinkBtnTestID = props.unlinkBtnTestID,
      copyBtnTestID = props.copyBtnTestID,
      cutBtnTestID = props.cutBtnTestID,
      deleteBtnTestID = props.deleteBtnTestID,
      switchStyleBtnTestID = props.switchStyleBtnTestID,
      linkStyleBtnTestID = props.linkStyleBtnTestID,
      cardStyleBtnTestID = props.cardStyleBtnTestID,
      enablePreview = props.enablePreview,
      getContainer = props.getContainer;

  var handleRefresh = function handleRefresh(evt) {
    evt.stopPropagation();
    onRefresh();
  };

  var onStyleChange = function onStyleChange(num) {
    switch (num) {
      case 0:
        onSwitchStyle == null ? void 0 : onSwitchStyle('popup', 'text');
        break;

      case 1:
        onSwitchStyle == null ? void 0 : onSwitchStyle('indoc', 'abstract');
        break;

      case 2:
        onSwitchStyle == null ? void 0 : onSwitchStyle('indoc', 'preview');
        break;
    }
  }; // 如果是 indoc，因为有 border，所以需要减去 2px


  var height = cardStyle === 'popup' ? URLLinkCardHeight : URLLinkCardHeight - 2;
  return /*#__PURE__*/React__default.createElement(AbstractCardLayout, {
    isSelected: isSelected,
    height: height,
    type: "failure",
    title: (urlInfo == null ? void 0 : urlInfo.title) || locale.error,
    icon: _ImagePlaceholder$1 || (_ImagePlaceholder$1 = /*#__PURE__*/React__default.createElement(ImagePlaceholder, null)),
    toolbarVisible: cardStyle === 'popup' ? true : undefined,
    titleClick: function titleClick() {
      return onOpenURL == null ? void 0 : onOpenURL(url);
    },
    failureText: locale.errorDesc,
    getContainer: getContainer,
    refresh: true,
    style: cardStyle === 'popup' ? {
      boxShadow: '0px 8px 24px 0px rgba(0, 0, 0, 0.16),0px 0px 1px 0px rgba(0, 0, 0, 0.24)',
      border: 'none'
    } : {},
    refreshClick: handleRefresh,
    toolbarConfigs: readonly ? [{
      type: 'button',
      title: locale.copyTooltip,
      icon: _CopyIcon$1 || (_CopyIcon$1 = /*#__PURE__*/React__default.createElement(CopyIcon, null)),
      onClick: props.onCopy,
      testId: copyBtnTestID
    }] : [{
      type: 'button',
      title: locale.editTooltip,
      icon: _EditIcon$1 || (_EditIcon$1 = /*#__PURE__*/React__default.createElement(EditIcon, null)),
      onClick: props.onEdit,
      testId: editBtnTestID
    }, {
      type: 'button',
      title: locale.unlinkTooltip,
      icon: _UnlinkIcon$1 || (_UnlinkIcon$1 = /*#__PURE__*/React__default.createElement(UnlinkIcon, null)),
      onClick: props.onUnlink,
      testId: unlinkBtnTestID
    }, {
      type: 'style',
      options: enablePreview ? ['text', 'abstract', 'preview'] : ['text', 'abstract'],
      optionsTestId: {
        text: linkStyleBtnTestID || '',
        "abstract": cardStyleBtnTestID || '',
        preview: ''
      },
      onChange: onStyleChange,
      defaultValue: cardStyle === 'popup' ? 0 : 1,
      testId: switchStyleBtnTestID
    }, {
      type: 'more',
      menu: [[{
        name: 'delete',
        onClick: props.onDelete,
        testId: deleteBtnTestID
      }, {
        name: 'copy',
        onClick: props.onCopy,
        testId: copyBtnTestID
      }, {
        name: 'cut',
        onClick: props.onCut,
        testId: cutBtnTestID
      }]]
    }]
  });
};

function _createForOfIteratorHelperLoose(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (it) return (it = it.call(o)).next.bind(it); if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; return function () { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
// 按顺序匹配，常用的应该排在前面
var domainPatterns = [{
  name: '微博',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01W9wXqn1ErzTUwcWsH_!!6000000000406-2-tps-128-128.png',
  domains: ['weibo.com']
}, {
  name: '微信',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN01cBsJhj20L6awUwQwo_!!6000000006832-2-tps-128-128.png',
  domains: ['mp.weixin.qq.com']
}, {
  name: '知乎',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01cumaXw25IQLvlWE3J_!!6000000007503-2-tps-128-128.png',
  domains: ['zhihu.com']
}, {
  name: '百度',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01n0JkDW1cBtqqiRXkU_!!6000000003563-2-tps-128-128.png',
  domains: ['baidu.com']
}, {
  name: '天猫',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN01n3iF0u1RszLOtNmvJ_!!6000000002168-2-tps-128-128.png',
  domains: ['tmall.com']
}, {
  name: '腾讯视频',
  imgURL: 'https://img.alicdn.com/imgextra/i4/O1CN01k8st6n1ydJ6P13u59_!!6000000006601-2-tps-128-128.png',
  domains: ['v.qq.com']
}, // 其他腾讯网站需要放在前面
{
  name: '腾讯',
  imgURL: 'https://img.alicdn.com/imgextra/i4/O1CN01oHRnPd1cDjR9bLBcy_!!6000000003567-2-tps-128-128.png',
  domains: ['qq.com']
}, {
  name: '网易',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN01smqRuh1ZvIGLwYquB_!!6000000003256-2-tps-128-128.png',
  domains: ['163.com']
}, {
  name: '小米',
  imgURL: 'https://img.alicdn.com/imgextra/i4/O1CN0110r1AW22Dtan6JG57_!!6000000007087-2-tps-128-128.png',
  domains: ['mi.com']
}, {
  name: '淘宝',
  imgURL: 'https://img.alicdn.com/imgextra/i4/O1CN01d4MkJA1UuljIeoUSW_!!6000000002578-2-tps-128-128.png',
  domains: ['taobao.com']
}, {
  name: '抖音',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN01sjMj641QbIQezVJnk_!!6000000001994-2-tps-128-128.png',
  domains: ['tiktok.com', 'douyin.com']
}, {
  name: '58同城',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01XAYWZH1ZyxQlm8dR8_!!6000000003264-2-tps-128-128.png',
  domains: ['58.com']
}, {
  name: '爱奇艺',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01WAsGW71jD07Rh9GYQ_!!6000000004513-2-tps-128-128.png',
  domains: ['iqiyi.com']
}, {
  name: '贝壳',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN017VNJv61UhUFJAMC6w_!!6000000002549-2-tps-128-128.png',
  domains: ['ke.com']
}, {
  name: '语雀',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01wksTLs1CKRE716qJY_!!6000000000062-2-tps-128-128.png',
  domains: ['yuque.com', 'yuque.antfin-inc.com']
}, {
  name: 'WPS',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01diPy3g1NCcEqeb5oa_!!6000000001534-2-tps-128-128.png',
  domains: ['wps.cn', 'kdocs.cn']
}, {
  name: '豆瓣',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01ttYRz0201rxZhxtnx_!!6000000006790-2-tps-128-128.png',
  domains: ['douban.com']
}, {
  name: '京东',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01jvIDXA1N5kIP7CfLg_!!6000000001519-2-tps-128-128.png',
  domains: ['jd.com']
}, {
  name: '房天下',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN01sWwwyu1LlJ0SBDd6s_!!6000000001339-2-tps-128-128.png',
  domains: ['fang.com']
}, {
  name: '凤凰网',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01fXuGd51VXF062i9ky_!!6000000002662-2-tps-128-128.png',
  domains: ['ifeng.com']
}, {
  name: '酷狗',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01r1HCXR27pWCm6oDnh_!!6000000007846-2-tps-128-128.png',
  domains: ['kugou.com']
}, {
  name: '酷我',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01QWirqF1g3QzvA2rGc_!!6000000004086-2-tps-128-128.png',
  domains: ['kuwo.cn']
}, {
  name: '乐视',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN01S2fDsk1zey9n7jSbO_!!6000000006740-2-tps-128-128.png',
  domains: ['le.com']
}, {
  name: '拼多多',
  imgURL: 'https://img.alicdn.com/imgextra/i4/O1CN01l82Svz1jceIU2ocKL_!!6000000004569-2-tps-128-128.png',
  domains: ['pinduoduo.com']
}, {
  name: '汽车之家',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN01JHIQFw1R5WZDklXaa_!!6000000002060-2-tps-128-128.png',
  domains: ['autohome.com.cn']
}, {
  name: '搜狗',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN01qK2nRB1a4SBZpnhv6_!!6000000003276-2-tps-128-128.png',
  domains: ['sogou.com']
}, {
  name: '人民网',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN01Hh28do1uhC00rAG7i_!!6000000006068-2-tps-128-128.png',
  domains: ['people.com.cn', 'people.cn']
}, {
  name: '搜狐视频',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN01SRLj3r20i0OjZssLM_!!6000000006882-2-tps-128-128.png',
  domains: ['tv.sohu.com']
}, // 其他搜狐网站需要放在前面
{
  name: '搜狐',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN01koPtmw1XP7CkWQQwD_!!6000000002915-2-tps-128-128.png',
  domains: ['sohu.com']
}, {
  name: '钉钉文档',
  imgURL: 'https://img.alicdn.com/imgextra/i4/O1CN01fNUadl1USpdOdVcXZ_!!6000000002517-2-tps-128-128.png',
  domains: ['alidocs.dingtalk.com', 'pre-alidocs.dingtalk.com', 'notes.dingtalk.com', 'pre-notes.dingtalk.com']
}, // 其他钉钉域名需要放在前面
{
  name: '钉钉',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN01EZVcIF1KLpSTAhfuJ_!!6000000001148-2-tps-128-128.png',
  domains: ['dingtalk.com']
}, {
  name: 'Figma',
  imgURL: 'https://img.alicdn.com/imgextra/i4/O1CN01wyOys01QWiSvsCTQk_!!6000000001984-2-tps-120-120.png',
  domains: ['figma.com']
}, {
  name: 'Gist',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01G1D7pv23umHrOslt8_!!6000000007316-2-tps-120-120.png',
  domains: ['gist.github.com']
}, {
  name: 'Framer',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01ayhKvJ1oEtpiWLjBI_!!6000000005194-2-tps-120-120.png',
  domains: ['framer.com']
}, {
  name: 'GitHub',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01phOpN61Sbs9mTRDjG_!!6000000002266-2-tps-128-128.png',
  domains: ['github.com']
}, {
  name: 'Android',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN01Mcn2Wx29vagigUDHp_!!6000000008130-2-tps-128-128.png',
  domains: ['android.google.cn', 'android.com']
}, {
  name: 'Amazon',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN01aBUWBN1KnlW6Q8ieD_!!6000000001209-2-tps-128-128.png',
  domains: ['amazon.com', 'amazon.cn']
}, {
  name: 'Dribbble',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN015oMKux1x2NYNqJ5oz_!!6000000006385-2-tps-128-128.png',
  domains: ['dribbble.com']
}, {
  name: 'Apple',
  imgURL: 'https://img.alicdn.com/imgextra/i4/O1CN01C9Zstm23h2PODzFQ6_!!6000000007286-2-tps-128-128.png',
  domains: ['apple.com', 'apple.com.cn']
}, {
  name: 'Facebook',
  imgURL: 'https://img.alicdn.com/imgextra/i4/O1CN01CJWM7b1EDgcSK6xja_!!6000000000318-2-tps-128-130.png',
  domains: ['facebook.com']
}, {
  name: 'Evernote',
  imgURL: 'https://img.alicdn.com/imgextra/i4/O1CN014xDMp91iwVrIBuWFm_!!6000000004477-2-tps-128-128.png',
  domains: ['evernote.com', 'yinxiang.com']
}, {
  name: 'Gmail',
  imgURL: 'https://img.alicdn.com/imgextra/i4/O1CN01JVYr8L1ThCNZrURc0_!!6000000002413-2-tps-128-128.png',
  domains: ['gmail.com']
}, {
  name: 'Instagram',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN01JbPkIx1pMyR4UL4ep_!!6000000005347-2-tps-128-128.png',
  domains: ['instagram.com']
}, {
  name: 'Invision',
  imgURL: 'https://img.alicdn.com/imgextra/i4/O1CN01VejU3I1StHCxz6fM2_!!6000000002304-2-tps-128-128.png',
  domains: ['invisionapp.com']
}, {
  name: 'Line',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01FgoRKJ1ZVe5I4PLZq_!!6000000003200-2-tps-128-128.png',
  domains: ['line.me']
}, {
  name: 'MasterCard',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN01yBKJMg1dHE4vLJeu2_!!6000000003710-2-tps-128-128.png',
  domains: ['mastercard.us']
}, {
  name: 'PayPal',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN01ujIzlm1Li6ESOEaGl_!!6000000001332-2-tps-128-128.png',
  domains: ['paypal.com']
}, {
  name: 'Airbnb',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN01l4UAtG1UKaRSJfuN2_!!6000000002499-2-tps-128-128.png',
  domains: ['airbnb.com']
}, {
  name: 'LinkedIn',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01sKi6mb1garuRzbhIp_!!6000000004159-2-tps-128-128.png',
  domains: ['linkedin.com']
}, {
  name: 'GoogleDoc',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN01ZXIXix1JdOwCOa5Af_!!6000000001051-2-tps-128-128.png',
  domains: ['docs.google.com']
}, {
  name: 'Pinterest',
  imgURL: 'https://img.alicdn.com/imgextra/i4/O1CN01AeBCyr1QZ0Rf8Sxls_!!6000000001989-2-tps-128-128.png',
  domains: ['pinterest.com']
}, {
  name: 'StackOverflow',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN019aLzWa1CU3XuuALEM_!!6000000000083-2-tps-128-128.png',
  domains: ['stackoverflow.com']
}, {
  name: 'Tumblr',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN016iQilu1zv01DP9VLl_!!6000000006775-2-tps-128-128.png',
  domains: ['tumblr.com']
}, {
  name: 'GitLab',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN0160QQbE1IX9vIunvLn_!!6000000000902-2-tps-128-128.png',
  domains: ['gitlab.com']
}, {
  name: 'MicrosoftTeams',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN01sm0bb51dav6k8TBdQ_!!6000000003753-2-tps-128-128.png',
  domains: ['teams.microsoft.com']
}, // 其他 microsoft 的网站需要放在前面
{
  name: 'Microsoft',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN01LyQADC1va9Wp8eWpb_!!6000000006188-2-tps-128-128.png',
  domains: ['microsoft.com']
}, {
  name: 'ATA',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN01HTS1jt1vrYZpNPm2L_!!6000000006226-2-tps-128-128.png',
  domains: ['atatech.org']
}, {
  name: '阿里内外',
  imgURL: 'https://img.alicdn.com/imgextra/i3/O1CN01v9NcvL1NI6zVshbWT_!!6000000001546-2-tps-128-128.png',
  domains: ['work.alibaba-inc.com']
}, {
  name: '阿里味',
  imgURL: 'https://img.alicdn.com/imgextra/i2/O1CN01k1AzET21zEuw4YPem_!!6000000007055-2-tps-128-128.png',
  domains: ['ifocus.alibaba-inc.com']
}, // 注意内网子域名要放在前面
{
  name: '阿里内网',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN01wcCcSo1CRlYoHvLq9_!!6000000000078-2-tps-128-128.png',
  domains: ['alibaba-inc.com']
}, {
  name: 'Google',
  imgURL: 'https://img.alicdn.com/imgextra/i1/O1CN01xx1gKJ1uI0DLkDaGA_!!6000000006013-2-tps-128-128.png',
  domains: ['google.com']
}].map(function (p) {
  return _extends({}, p, {
    regexp: new RegExp("^(https?:\\/\\/)?(.+\\.)?(" + p.domains.map(function (d) {
      return d.replace(/\./g, '\\.');
    }).join('|') + ")($|\\/)")
  });
});
var patterns = [].concat(domainPatterns);

var getImageIfMatchPattern = function getImageIfMatchPattern(url) {
  for (var _iterator = _createForOfIteratorHelperLoose(patterns), _step; !(_step = _iterator()).done;) {
    var p = _step.value;

    if ('test' in p) {
      if (p.test(url)) {
        return p.imgURL;
      }
    } else {
      if (p.regexp.test(url)) {
        return p.imgURL;
      }
    }
  }

  return null;
};

var defaultImages = ['https://img.alicdn.com/imgextra/i1/O1CN01edSVl81Pl5HFhZ8qM_!!6000000001880-2-tps-128-128.png', 'https://img.alicdn.com/imgextra/i1/O1CN013BGmiH1Hr1U7QmB2x_!!6000000000810-2-tps-128-128.png', 'https://img.alicdn.com/imgextra/i3/O1CN01xOZYfH25Un31ouK0e_!!6000000007530-2-tps-128-128.png', 'https://img.alicdn.com/imgextra/i1/O1CN01Lp5IBv1O6wx7s7Zhk_!!6000000001657-2-tps-128-128.png', 'https://img.alicdn.com/imgextra/i4/O1CN01CEtLGO1YPrSuUpgod_!!6000000003052-2-tps-128-128.png'];

var getHostOfURL = function getHostOfURL(url) {
  try {
    return new URL(url).hostname;
  } catch (err) {
    return url;
  }
}; // 根据 input 计算一个 hash


var simpleHash = function simpleHash(input) {
  var result = 0;

  for (var i = 0; i < input.length; i++) {
    result = result * 7 + 13 * input.charCodeAt(i);
  }

  return result;
}; // 返回背景图片


var getImgURL = (function (url, info, type) {
  // 如果指定了 type 根据 type 返回
  if (type) {
    switch (type) {
      case 'figma':
        return 'https://img.alicdn.com/imgextra/i4/O1CN01wyOys01QWiSvsCTQk_!!6000000001984-2-tps-120-120.png';

      case 'gist':
        return 'https://img.alicdn.com/imgextra/i3/O1CN01G1D7pv23umHrOslt8_!!6000000007316-2-tps-120-120.png';

      case 'framer':
        return 'https://img.alicdn.com/imgextra/i3/O1CN01ayhKvJ1oEtpiWLjBI_!!6000000005194-2-tps-120-120.png';

      case 'yuque':
        return 'https://img.alicdn.com/imgextra/i3/O1CN01wksTLs1CKRE716qJY_!!6000000000062-2-tps-128-128.png';

      default:
        throw new UnsupportedValueError(type);
    }
  } // 如果 url 满足内置的模式，根据 url 返回


  var result = getImageIfMatchPattern(url);
  if (result != null) return result; // 返回接口抓取的图片

  if (info.imgURL !== '') return info.imgURL; // 返回默认图片

  var host = getHostOfURL(url);
  var hash = simpleHash(host);
  return defaultImages[hash % defaultImages.length];
});

var _ImagePlaceholder, _CopyIcon, _EditIcon, _UnlinkIcon;

var BgImg = /*#__PURE__*/styled.div(["background-color:rgba(23,26,29,0.04);background-image:url(", ");background-size:cover;background-position:50% 50%;border-radius:8px;width:", "px;height:", "px;"], function (p) {
  return p.url;
}, function (p) {
  return p.width;
}, function (p) {
  return p.height;
});
var URLLinkCardNormal = function URLLinkCardNormal(props) {
  var url = props.url,
      urlInfo = props.urlInfo,
      cardStyle = props.cardStyle,
      isMobile = props.isMobile,
      onSwitchStyle = props.onSwitchStyle,
      locale = props.locale,
      onOpenURL = props.onOpenURL,
      readonly = props.readonly,
      isSelected = props.isSelected,
      editBtnTestID = props.editBtnTestID,
      unlinkBtnTestID = props.unlinkBtnTestID,
      copyBtnTestID = props.copyBtnTestID,
      cutBtnTestID = props.cutBtnTestID,
      deleteBtnTestID = props.deleteBtnTestID,
      switchStyleBtnTestID = props.switchStyleBtnTestID,
      linkStyleBtnTestID = props.linkStyleBtnTestID,
      cardStyleBtnTestID = props.cardStyleBtnTestID,
      linkType = props.linkType,
      enablePreview = props.enablePreview,
      getContainer = props.getContainer;
  var title = urlInfo.title,
      desc = urlInfo.desc;
  var imgURL = getImgURL(url, urlInfo, linkType);

  var onStyleChange = function onStyleChange(num) {
    switch (num) {
      case 0:
        onSwitchStyle == null ? void 0 : onSwitchStyle('popup', 'text');
        break;

      case 1:
        onSwitchStyle == null ? void 0 : onSwitchStyle('indoc', 'abstract');
        break;

      case 2:
        onSwitchStyle == null ? void 0 : onSwitchStyle('indoc', 'preview');
        break;
    }
  };

  if (title === '') {
    title = desc;
    desc = url;

    if (title === '') {
      title = '链接';
    }
  } else if (desc === '') {
    desc = url;
  }

  var onClick = function onClick() {
    return onOpenURL == null ? void 0 : onOpenURL(url);
  };

  var cardDesc = []; // 只有一段文本

  if (cardStyle === 'popup') {
    cardDesc.push({
      text: desc,
      line: 2,
      onClick: desc == url ? onClick : undefined
    });
  } else {
    if (desc !== url) {
      cardDesc.push({
        text: desc
      });
      cardDesc.push({
        text: url,
        onClick: onClick
      });
    } else {
      cardDesc.push({
        text: url,
        line: 2,
        onClick: onClick
      });
    }
  }

  var icon = imgURL !== '' ? /*#__PURE__*/React.createElement(BgImg, {
    url: imgURL,
    width: 64,
    height: 64
  }) : _ImagePlaceholder || (_ImagePlaceholder = /*#__PURE__*/React.createElement(ImagePlaceholder, null));

  var toolbarVisible = function () {
    if (cardStyle === 'popup') {
      return true;
    }

    if (isMobile) return true;
    return undefined;
  }(); // 如果是 indoc，因为有 border，所以需要减去 2px


  var height = cardStyle === 'popup' ? URLLinkCardHeight : URLLinkCardHeight - 2;
  var toolbar = [];

  if (isMobile || readonly) {
    if (props.onCopy) {
      toolbar.push({
        type: 'button',
        title: locale.copyTooltip,
        icon: _CopyIcon || (_CopyIcon = /*#__PURE__*/React.createElement(CopyIcon, null)),
        onClick: props.onCopy,
        testId: copyBtnTestID
      });
    }
  } else {
    if (props.onEdit) {
      toolbar.push({
        type: 'button',
        title: locale.editTooltip,
        icon: _EditIcon || (_EditIcon = /*#__PURE__*/React.createElement(EditIcon, null)),
        onClick: props.onEdit,
        testId: editBtnTestID
      });
    }

    if (props.onUnlink) {
      toolbar.push({
        type: 'button',
        title: locale.unlinkTooltip,
        icon: _UnlinkIcon || (_UnlinkIcon = /*#__PURE__*/React.createElement(UnlinkIcon, null)),
        onClick: props.onUnlink,
        testId: unlinkBtnTestID
      });
    }

    if (props.onSwitchStyle) {
      toolbar.push({
        type: 'style',
        options: enablePreview ? ['text', 'abstract', 'preview'] : ['text', 'abstract'],
        optionsTestId: {
          text: linkStyleBtnTestID || '',
          "abstract": cardStyleBtnTestID || '',
          preview: ''
        },
        testId: switchStyleBtnTestID,
        onChange: onStyleChange,
        defaultValue: cardStyle === 'popup' ? 0 : 1
      });
    }

    if (props.onDelete || props.onCopy || props.onCut) {
      var menu = [];

      if (props.onDelete) {
        menu.push({
          name: 'delete',
          onClick: props.onDelete,
          testId: deleteBtnTestID
        });
      }

      if (props.onCopy) {
        menu.push({
          name: 'copy',
          onClick: props.onCopy,
          testId: copyBtnTestID
        });
      }

      if (props.onCut) {
        menu.push({
          name: 'cut',
          onClick: props.onCut,
          testId: cutBtnTestID
        });
      }

      toolbar.push({
        type: 'more',
        menu: [menu]
      });
    }
  }

  return /*#__PURE__*/React.createElement(AbstractCardLayout, {
    getContainer: getContainer,
    isSelected: isSelected,
    height: height,
    background: imgURL !== '' ? 'img' : 'color',
    color: "#f2f4fb",
    line: "multi",
    lineCount: 2,
    titleClick: function titleClick() {
      return onOpenURL == null ? void 0 : onOpenURL(url);
    },
    img: imgURL,
    icon: icon,
    title: title,
    style: cardStyle === 'popup' ? {
      boxShadow: '0px 8px 24px 0px rgba(0, 0, 0, 0.16),0px 0px 1px 0px rgba(0, 0, 0, 0.24)',
      border: 'none'
    } : {},
    desc: cardDesc,
    toolbarVisible: toolbarVisible,
    toolbarConfigs: toolbar
  });
};

var Wrapper = /*#__PURE__*/styled.div(["isolation:isolate;.card-layout-title{", "}.card-layout-top{background:linear-gradient( 270deg,rgba(243,246,255,1) 0%,rgba(241,244,255,1) 26.04%,rgba(244,246,249,1) 100% );}.card-layout-icon{margin:8px 16px 8px 8px !important;max-width:100% !important;max-height:100% !important;overflow:visible !important;}.card-layout-desc .item .text{white-space:'break-spaces';}"], function (p) {
  if (p.state === 'loading' && p.cardStyle === 'popup') {
    return 'margin-top: 5px !important;';
  }
});
var URLLinkCard = function URLLinkCard(props) {
  var stopPropagation = function stopPropagation(e) {
    e.stopPropagation();
  };

  var el = /*#__PURE__*/React.createElement(React.Fragment, null);

  switch (props.state) {
    case 'loading':
      el = /*#__PURE__*/React.createElement(URLLinkCardLoading, props);
      break;

    case 'error':
      el = /*#__PURE__*/React.createElement(URLLinkCardError, props);
      break;

    case 'normal':
      el = /*#__PURE__*/React.createElement(URLLinkCardNormal, props);
      break;

    default:
      throw new Error("[urlCard] unsupported state: " + props.state);
  }

  return /*#__PURE__*/React.createElement(Wrapper, {
    state: props.state,
    cardStyle: props.cardStyle,
    "data-testid": props.cardTestID,
    onMouseDown: stopPropagation,
    onTouchStart: stopPropagation
  }, el);
};

export { URLLinkCard, URLLinkCardHeight, domainPatterns };
