// @ali/4ever-utils/dist-node/theme/basic.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.blockMarginStyleInContainer = exports.lineHeight = exports.fontSize = void 0;
const fontSize = {
  DEFAULT: 11,
  MAX: 200,
  MIN: 5,
  PARTICLE: 0.5
};
exports.fontSize = fontSize;
const lineHeight = {
  ACTUAL_MAP: {
    1: 1.7,
    1.15: 2,
    1.5: 2.5,
    2: 3.4,
    2.5: 4.3,
    3: 5.1
  },
  DEFAULT: 1.7,
  MIN: 0.7
};
/**
 * 容器内（单元格、高亮块、分栏）：
 * 第一个块级元素不要 margin-top
 * 最后一个块级元素不要 margin-bottom
 * 以保证单元格上下左右间距一致
 * 使用性能最好的 css 选择器，放弃分页下的支持
 */

exports.lineHeight = lineHeight;
const blockMarginStyleInContainer = `
  & > [data-cangjie-edge-selectable='true']:first-child {
    margin-top: 0 !important;
  }

  & > [data-cangjie-edge-selectable='true']:last-child {
    margin-bottom: 0 !important;
  }

  & > [data-cangjie-leaf-block='true']:first-child {
    margin-top: 0 !important;
  }

  & > [data-cangjie-leaf-block='true']:last-child {
    margin-bottom: 0 !important;
  }
`;
exports.blockMarginStyleInContainer = blockMarginStyleInContainer;
