// @ali/4ever-utils/dist-node/theme/fonts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FALLBACK_STRATEGIES_CN = exports.FONTS_INFO = void 0;

/**
  * 通用字体族（去掉部分不用的）
  * https://developer.mozilla.org/en-US/docs/Web/CSS/font-family#%3Cgeneric-name%3E
*/

/**
 * 字重
 * https://developer.mozilla.org/en-US/docs/Web/CSS/font-weight#Common_weight_name_mapping
 * ```
 * 100 | 200 | 300 | 400 | 500 | 600 | 700 | 800 | 900 | 950
 * ```
 */

/**
 * 字符集（部分来源于统计）
 */

/**
 * 字体信息
 */
const FONTS_INFO = [
/** *************** 宋 **************** */
{
  family: 'SimSun',
  chinesefamily: '宋体',
  generic: 'serif'
}, {
  family: 'NSimSun',
  chinesefamily: '新宋体',
  generic: 'serif'
}, {
  family: 'STZhongsong',
  chinesefamily: '华文中宋',
  generic: 'serif',
  btbd: 1.137
}, {
  family: 'FZXiaoBiaoSong-B05S',
  chinesefamily: '方正小标宋简体',
  generic: 'serif',
  btbd: 1.1914
}, {
  family: 'Songti SC',
  generic: 'serif',
  postScriptName: 'STSongti-SC',
  chinesefamily: '宋体-简',
  weights: ['Light', 'Bold', 'Black']
}, {
  family: 'FZSY--SURROGATE-0',
  chinesefamily: '宋体-方正超大字符集',
  generic: 'serif',
  btbd: 1.127
}, {
  family: 'FZShuSong-Z01S',
  chinesefamily: '方正书宋简体',
  generic: 'serif',
  btbd: 1.148
}, {
  family: 'STSong',
  chinesefamily: '华文宋体',
  generic: 'serif'
}, {
  family: 'FZDaBiaoSong-B06S',
  chinesefamily: '方正大标宋简体',
  generic: 'serif',
  btbd: 1.113
}, {
  family: 'FZSong III-Z05S',
  chinesefamily: '方正宋三简体',
  generic: 'serif',
  btbd: 1.14
}, {
  family: 'PMingLiU',
  chinesefamily: '新細明體',
  generic: 'serif'
}, {
  family: 'Source Han Serif CN',
  chinesefamily: '思源宋体 CN',
  generic: 'serif',
  weights: ['Semibold', 'Light'],
  btbd: 1.437
}, {
  family: 'FZYaoti',
  chinesefamily: '方正姚体',
  generic: 'serif'
},
/** *************** 黑 **************** */
{
  family: 'Microsoft YaHei',
  chinesefamily: '微软雅黑',
  weights: ['Light', 'Bold'],
  generic: 'sans-serif',
  btbd: 1.32
}, {
  family: 'SimHei',
  chinesefamily: '黑体',
  generic: 'sans-serif'
}, {
  family: 'STXihei',
  chinesefamily: '华文细黑',
  generic: 'sans-serif',
  btbd: 1.079
}, {
  family: 'Dengxian',
  chinesefamily: '等线',
  generic: 'sans-serif',
  weights: ['Light', 'Bold'],
  btbd: 1.042
}, {
  family: 'PingFang SC',
  chinesefamily: '苹方-简',
  postScriptName: 'PingFangSC',
  generic: 'sans-serif',
  weights: ['Ultralight', 'Light', 'Thin', 'Medium', 'Semibold', 'Regular'],
  btbd: 1.4
}, {
  family: 'Source Han Sans CN',
  chinesefamily: '思源黑体 CN',
  generic: 'sans-serif',
  weights: ['Regular', 'Normal', 'Light', 'Medium', 'Bold', 'ExtraLight'],
  btbd: 1.375
}, {
  family: 'Microsoft JhengHei',
  chinesefamily: '微软正黑',
  generic: 'sans-serif',
  weights: ['Light'],
  btbd: 1.33
}, {
  family: 'Microsoft JhengHei',
  chinesefamily: '微軟正黑體',
  generic: 'sans-serif',
  btbd: 1.33
}, {
  family: 'Microsoft JhengHei UI',
  chinesefamily: '微软正黑 UI',
  generic: 'sans-serif',
  btbd: 1.33
}, {
  family: 'Adobe Heiti Std',
  chinesefamily: 'Adobe 黑体 Std R',
  generic: 'sans-serif',
  btbd: 1.25
}, {
  family: 'Microsoft YaHei UI',
  chinesefamily: '微软雅黑 UI',
  generic: 'sans-serif',
  btbd: 1.27
}, {
  family: 'FZHei-B01S',
  chinesefamily: '方正黑体简体',
  generic: 'sans-serif',
  btbd: 1.125
}, {
  family: 'Heiti SC',
  chinesefamily: '黑体-简',
  postScriptName: 'STHeitiSC',
  weights: ['Light', 'Medium'],
  generic: 'sans-serif'
}, {
  family: 'STHeiti',
  chinesefamily: '华文黑体',
  weights: ['Light'],
  generic: 'sans-serif'
}, {
  family: 'Hiragino Sans GB W3',
  chinesefamily: '冬青黑体简体中文 W3',
  generic: 'sans-serif',
  btbd: 1.162
}, {
  family: 'Hiragino Sans GB W6',
  chinesefamily: '冬青黑体简体中文 W6',
  generic: 'sans-serif',
  btbd: 1.24
},
/** *************** 仿宋 **************** */
{
  family: 'FangSong',
  chinesefamily: '仿宋',
  generic: 'fangsong',
  charSets: ['GBK', 'GB2312']
}, {
  family: 'FZFangSong-Z02S',
  chinesefamily: '方正仿宋简体',
  generic: 'fangsong',
  btbd: 1.125
}, {
  family: 'FZFangSong-Z02S',
  chinesefamily: '方正仿宋',
  generic: 'fangsong',
  charSets: ['GBK'],
  btbd: 1.125
}, {
  family: 'STFangsong',
  chinesefamily: '华文仿宋',
  generic: 'fangsong',
  btbd: 1.12
}, {
  family: 'Adobe Fangsong Std',
  chinesefamily: 'Adobe 仿宋 Std R',
  generic: 'fangsong',
  btbd: 1.253
}, {
  family: 'CHCFS',
  chinesefamily: '彩虹粗仿宋',
  generic: 'fangsong'
},
/** *************** 楷 **************** */
{
  family: 'KaiTi',
  chinesefamily: '楷体',
  charSets: ['GB2312'],
  generic: 'cursive'
}, {
  family: 'STKaiti',
  chinesefamily: '华文楷体',
  generic: 'cursive',
  btbd: 1.12
}, {
  family: 'FZKTK--GBK1-0',
  chinesefamily: '方正楷体简体',
  generic: 'cursive',
  btbd: 1.086
}, {
  family: 'FZKai-Z03S',
  chinesefamily: '方正楷体',
  generic: 'cursive',
  charSets: ['GBK'],
  btbd: 1.113
}, {
  family: 'FZShuTi-S05S',
  chinesefamily: '方正舒体',
  generic: 'cursive',
  btbd: 1.0976
}, {
  family: 'Kaiti SC',
  chinesefamily: '楷体-简',
  postScriptName: 'STKaitiSC',
  generic: 'cursive',
  weights: ['Regular', 'Bold', 'Black']
}, {
  family: 'Adobe Kaiti Std',
  chinesefamily: 'Adobe 楷体 Std R',
  generic: 'cursive'
}, {
  family: 'STXinwei',
  chinesefamily: '华文新魏',
  generic: 'cursive',
  btbd: 1.017
}, {
  family: 'Weibei SC',
  chinesefamily: '魏碑-简',
  generic: 'cursive',
  weights: ['Bold'],
  btbd: 1.32
}, {
  family: 'STXingkai',
  chinesefamily: '华文行楷',
  generic: 'cursive'
},
/** *************** 圆 **************** */
{
  family: 'YouYuan',
  chinesefamily: '幼圆',
  generic: 'serif',
  fallback: ['Yuanti SC']
}, {
  family: 'Yuanti SC',
  chinesefamily: '圆体-简',
  generic: 'serif',
  weights: ['Regular', 'Light', 'Bold'],
  fallback: ['YouYuan']
},
/** *************** 隶 **************** */
{
  family: 'LiShu',
  chinesefamily: '隶书',
  generic: 'serif',
  fallback: ['Libian SC']
}, {
  family: 'STLiti',
  chinesefamily: '华文隶书',
  generic: 'serif',
  fallback: ['Libian SC']
}, {
  family: 'Libian SC',
  chinesefamily: '隶变-简',
  generic: 'serif',
  fallback: ['LiShu']
},
/** *************** en **************** */
{
  /**
   * win: Calibri 等衬线体
   * mac: Times New Roman
   */
  family: 'Calibri',
  generic: 'serif',
  fallback: ['Times New Roman']
}, {
  family: 'Cambria Math',
  generic: 'serif',
  fallback: ['Times New Roman']
}, {
  family: 'MS Mincho',
  generic: 'serif',
  fallback: ['Times New Roman']
},
/**
 * win: MS Gothic 等非衬线体
 * mac: Arial
 */
{
  family: 'MS Gothic',
  generic: 'sans-serif',
  fallback: ['Arial']
}, {
  family: 'Segoe UI',
  generic: 'sans-serif',
  fallback: ['Arial']
},
/**
 * Fira Code 等宽字体
 * mac: Andale Mono
 */
{
  family: 'Fira Code',
  generic: 'monospace',
  fallback: ['Andale Mono']
}];
/**
 * 中文四大字体通用降级策略
 */

exports.FONTS_INFO = FONTS_INFO;
const FALLBACK_STRATEGIES_CN = {
  /**
   * win: SimSun NSimSun
   * mac: Songti SC STSong
   */
  serif: ['SimSun', 'Songti SC', 'NSimSun', 'STSong'],

  /**
   * win: Microsoft YaHei SimHei
   * mac: PingFang SC STHeiti
   */
  'sans-serif': ['Microsoft YaHei', 'PingFang SC', 'SimHei', 'STHeiti'],

  /**
   * win: FangSong FZFangSong-Z02S
   * mac: STFangsong
   */
  fangsong: ['FangSong', 'FZFangSong-Z02S', 'STFangsong'],

  /**
   * win: KaiTi STKaiti
   * mac: Kaiti SC
   */
  cursive: ['KaiTi', 'STKaiti', 'Kaiti SC']
};
exports.FALLBACK_STRATEGIES_CN = FALLBACK_STRATEGIES_CN;
