// @ali/4ever-utils/dist-node/theme/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "codeTheme", {
  enumerable: true,
  get: function () {
    return _code.default;
  }
});
Object.defineProperty(exports, "linkTheme", {
  enumerable: true,
  get: function () {
    return _link.default;
  }
});
Object.defineProperty(exports, "mentionTheme", {
  enumerable: true,
  get: function () {
    return _mention.default;
  }
});
Object.defineProperty(exports, "indent", {
  enumerable: true,
  get: function () {
    return _indent.default;
  }
});
Object.defineProperty(exports, "inlineCodeTheme", {
  enumerable: true,
  get: function () {
    return _inlineCode.default;
  }
});
Object.defineProperty(exports, "fontSize", {
  enumerable: true,
  get: function () {
    return _basic.fontSize;
  }
});
Object.defineProperty(exports, "lineHeight", {
  enumerable: true,
  get: function () {
    return _basic.lineHeight;
  }
});
Object.defineProperty(exports, "blockMarginStyleInContainer", {
  enumerable: true,
  get: function () {
    return _basic.blockMarginStyleInContainer;
  }
});
Object.defineProperty(exports, "spacing", {
  enumerable: true,
  get: function () {
    return _spacing.default;
  }
});
Object.defineProperty(exports, "defaultCompactSpacing", {
  enumerable: true,
  get: function () {
    return _spacing.defaultCompactSpacing;
  }
});
Object.defineProperty(exports, "defaultSpacing", {
  enumerable: true,
  get: function () {
    return _spacing.defaultSpacing;
  }
});
Object.defineProperty(exports, "DEFAULT_SHD_STYLES", {
  enumerable: true,
  get: function () {
    return _shd.default;
  }
});
exports.theme = void 0;

var _globalStyle = _interopRequireWildcard(require("./globalStyle"));

var _heading = _interopRequireDefault(require("./heading"));

var _paragraph = _interopRequireDefault(require("./paragraph"));

var _code = _interopRequireDefault(require("./code"));

var _embed = _interopRequireDefault(require("./embed"));

var _onlineVideo = _interopRequireDefault(require("./onlineVideo"));

var _hr = _interopRequireDefault(require("./hr"));

var _link = _interopRequireWildcard(require("./link"));

var _placeHolder = _interopRequireDefault(require("./placeHolder"));

var _table = _interopRequireDefault(require("./table"));

var _blockquote = _interopRequireDefault(require("./blockquote"));

var _mention = _interopRequireDefault(require("./mention"));

var _constant = _interopRequireDefault(require("./constant"));

var _refer = _interopRequireDefault(require("./refer"));

var _indent = _interopRequireDefault(require("./indent"));

var _inlineCode = _interopRequireDefault(require("./inlineCode"));

var _basic = require("./basic");

var _spacing = _interopRequireWildcard(require("./spacing"));

var _shd = _interopRequireDefault(require("./shd"));

const theme = { ..._heading.default,
  ..._table.default,
  ..._placeHolder.default,
  paragraph: _paragraph.default,
  code: _code.default,
  embed: _embed.default,
  inlineCode: _inlineCode.default,
  onlineVideo: _onlineVideo.default,
  hr: _hr.default,
  link: _link.default,
  PLACEHOLDER_LINK_PADDING_HORIZONTAL: _link.PLACEHOLDER_LINK_PADDING_HORIZONTAL,
  PLACEHOLDER_LINK_PADDING_VERTICAL: _link.PLACEHOLDER_LINK_PADDING_VERTICAL,
  blockquote: _blockquote.default,
  mention: _mention.default,
  globalStyle: _globalStyle.default,
  constant: _constant.default,
  refer: _refer.default,
  fontSize: _basic.fontSize,
  lineHeight: _basic.lineHeight,
  defaultKeyTypographyStyle: _globalStyle.defaultKeyTypographyStyle,
  spacing: _spacing.default,
  DEFAULT_SHD_STYLES: _shd.default,
  indent: _indent.default
};
exports.theme = theme;
