// @ali/4ever-utils/dist-node/theme/spacing.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.defaultCompactSpacing = exports.defaultSpacing = void 0;

/**
 * 定义 auto 的 spacing（用于普通段落和标题）
 * defaultBeforeAutospacing/defaultAfterAutospacing 用于还原 word 中的自动段间距
 * 相比 word（上 5pt，下 10pt）会稍微紧凑一点
 * defaultBeforeSpacin/defaultAfterSpacing 为钉钉文档定义的宽松间距
 */
const spacing = {
  defaultBeforeAutospacing: 12,
  // 9pt
  defaultAfterAutospacing: 12,
  // 9pt
  defaultBeforeSpacing: 8,
  // 6pt
  defaultAfterSpacing: 8,
  // 6pt
  defaultCompactSpacing: 2.667,
  // 2pt
  defaultLinePitch: 20.8,
  // 15.6pt
  defaultLine: 1
};
/**
 * 非普通段落的默认间距（例如 代码块、card 等）
 */

const defaultSpacing = {
  spacing: {
    before: spacing.defaultAfterAutospacing,
    after: spacing.defaultAfterAutospacing
  }
};
exports.defaultSpacing = defaultSpacing;
const defaultCompactSpacing = {
  spacing: {
    before: spacing.defaultCompactSpacing,
    after: spacing.defaultCompactSpacing
  }
};
exports.defaultCompactSpacing = defaultCompactSpacing;
var _default = spacing;
exports.default = _default;
