// @ali/4ever-plugin-heading/dist-web/mo/models/index.js
var _createDict;

import { Block } from '@ali/4ever-cangjie';
import { Paragraph } from '@ali/4ever-plugin-paragraph';
import { createDict } from '@ali/4ever-utils';
import Heading1 from "./heading1";
import Heading2 from "./heading2";
import Heading3 from "./heading3";
import Heading4 from "./heading4";
import Heading5 from "./heading5";
import Heading6 from "./heading6";
export var headingTypes = createDict((_createDict = {}, _createDict[Heading1.TYPE] = true, _createDict[Heading2.TYPE] = true, _createDict[Heading3.TYPE] = true, _createDict[Heading4.TYPE] = true, _createDict[Heading5.TYPE] = true, _createDict[Heading6.TYPE] = true, _createDict));
var Heading = {
  TYPES: Object.keys(headingTypes),
  Heading1: Heading1,
  Heading2: Heading2,
  Heading3: Heading3,
  Heading4: Heading4,
  Heading5: Heading5,
  Heading6: Heading6,
  isHeading: function isHeading(maybeHeading) {
    return Block.isBlock(maybeHeading) && headingTypes[maybeHeading.type];
  },
  isEmpty: function isEmpty(block) {
    var _block$data$ind;

    return headingTypes[block.type] && (block.data.jc || 'left') === 'left' && !((_block$data$ind = block.data.ind) != null && _block$data$ind.left) && block.text === '';
  },
  isCallout: function isCallout(maybeCallout) {
    return (Paragraph.isParagraph(maybeCallout) || Heading.isHeading(maybeCallout)) && !!maybeCallout.data.calloutPr;
  }
};
export default Heading;
