// @ali/4ever-component/dist-node/idleRender/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.idleRender = idleRender;

var React = _interopRequireWildcard(require("react"));

var _lodash = require("lodash");

var _everUtils = require("@ali/4ever-utils");

const _createElement = /*#__PURE__*/React.createElement;

function idleRender(Component, delay, options) {
  return class IdleComponent extends React.Component {
    constructor(...args) {
      super(...args);
      this.timeout = -1;
      this.debouncedForceUpdate = (0, _lodash.debounce)(() => {
        this.forceUpdate();
      }, delay, options);

      this.updateWhenIdle = () => {
        (0, _everUtils.cancelIdleCallback)(this.timeout);
        this.timeout = (0, _everUtils.requestIdleCallback)(() => {
          this.debouncedForceUpdate();
        });
      };
    }

    shouldComponentUpdate() {
      this.updateWhenIdle();
      return false;
    }

    componentWillUnmount() {
      this.debouncedForceUpdate.cancel();
      (0, _everUtils.cancelIdleCallback)(this.timeout);
    }

    render() {
      return /*#__PURE__*/_createElement(Component, this.props);
    }

  };
}
