// @ali/4ever-plugin-mention/dist-node/bi/commands/setMention.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

var _removeMention = _interopRequireDefault(require("./removeMention"));

var _utils = require("../utils");

var _default = (controller, suggestion, at) => {
  if (!suggestion) return controller;
  const {
    value
  } = controller;
  const {
    selection,
    document,
    injections
  } = value;
  const start = selection.getStart(document);
  const offset = (0, _utils.nearsetAt)(value.endText.text, start.offset);
  const mentionLength = start.offset - offset;

  const inline = _everCangjie.Inline.create({
    type: 'mention',
    data: {
      id: suggestion.id,
      login: suggestion.login,
      name: suggestion.name,
      avatarUrl: suggestion.avatarUrl
    }
  }); // 如果是 placeholder 类型的 mention 做 setData


  if (at) {
    const excludeInjections = injections.filter(i => i.key !== at);
    return controller.setInjections([...excludeInjections]).command(_everCangjie.Commands.setNodeByKey, at, {
      data: inline.data
    }).command(_everCangjie.Commands.focus).command(_everCangjie.Commands.moveToStartOfNextText);
  }

  controller.command(_everCangjie.Commands.deleteBackward, mentionLength).command(_everCangjie.Commands.focus).command(_everCangjie.Commands.insertInline, inline).command(_everCangjie.Commands.moveToStartOfNextText);
  return (0, _removeMention.default)(controller);
};

exports.default = _default;
