// @ali/4ever-bi/dist-node/plugins/sz/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setSz = setSz;
exports.adjustSz = adjustSz;
exports.DEPRECATED_ACTIONS = exports.isActions = exports.actions = exports.isAdjustSz = exports.isSetSz = exports.SZ_ADJUST = exports.SZ_SET = void 0;
const SZ_SET = Symbol('SZ_SET');
exports.SZ_SET = SZ_SET;
const SZ_ADJUST = Symbol('SZ_ADJUST');
exports.SZ_ADJUST = SZ_ADJUST;

function setSz(sz) {
  return {
    type: SZ_SET,
    payload: {
      sz
    }
  };
}

function adjustSz(isIncreasingSz) {
  return {
    type: SZ_ADJUST,
    payload: {
      isIncreasingSz
    }
  };
}

const isSetSz = action => {
  return action.type === SZ_SET;
};

exports.isSetSz = isSetSz;

const isAdjustSz = action => {
  return action.type === SZ_ADJUST;
};

exports.isAdjustSz = isAdjustSz;
const actions = {
  setSz,
  adjustSz
};
exports.actions = actions;
const isActions = {
  isSetSz,
  isAdjustSz
};
/** @deprecated */

exports.isActions = isActions;
const DEPRECATED_ACTIONS = {
  SZ_SET,
  SZ_ADJUST,
  setSz,
  adjustSz
};
exports.DEPRECATED_ACTIONS = DEPRECATED_ACTIONS;
