// @ali/we-design-next/dist-node/collaboration/network-status/NetworkStatus.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NetworkStatus = void 0;
var _react = _interopRequireDefault(require("react"));
var _styled = require("./styled");
var _dayjs = _interopRequireDefault(require("dayjs"));
var _locale = require("../../locale");
var _common = require("../../common");
var _localeData = _interopRequireDefault(require("dayjs/plugin/localeData"));
const _createElement = /*#__PURE__*/_react.default.createElement;
_dayjs.default.extend(_localeData.default);
const infoMap = {
  success: {
    color: _common.COLORS.LIGHT.LEVEL2_BASE_COLOR,
    icon: /*#__PURE__*/_createElement("svg", {
      width: "16",
      height: "16",
      fill: "currentColor",
      viewBox: "0 0 16 16",
      stroke: "none",
      xmlns: "http://www.w3.org/2000/svg"
    }, /*#__PURE__*/_createElement("path", {
      d: "M10 3C12.7614 3 15 5.23858 15 8C15 10.7614 12.7614 13 10 13H3.5C2.11929 13 1 11.8807 1 10.5C1 9.11929 2.11929 8 3.5 8C3.67163 8 3.83886 8.01722 4.0002 8.04995C4.00578 8.7324 4.12548 9.38866 4.34154 9.99996L5.28438 9.66671C5.10035 9.14605 5 8.58529 5 8C5 5.23858 7.23858 3 10 3ZM4.07525 7.04713C3.8879 7.01611 3.69571 7 3.5 7C1.567 7 0 8.567 0 10.5C0 12.433 1.567 14 3.5 14H10C13.3137 14 16 11.3137 16 8C16 4.68629 13.3137 2 10 2C7.01057 2 4.53176 4.18626 4.07525 7.04713ZM13.4536 6.85355L9.5 10.8071L7.14645 8.45355L7.85355 7.74645L9.5 9.39289L12.7464 6.14645L13.4536 6.85355Z"
    }))
  },
  info: {
    color: _common.COLORS.LIGHT.LEVEL3_BASE_COLOR,
    icon: /*#__PURE__*/_createElement("img", {
      width: "16",
      height: "16",
      src: "https://img.alicdn.com/imgextra/i3/O1CN01UGvMHI24U2rwXUkQt_!!6000000007393-1-tps-48-48.gif"
    })
  },
  error: {
    color: _common.COLORS.LIGHT.LEVEL3_BASE_COLOR,
    icon: /*#__PURE__*/_createElement("svg", {
      width: "16",
      height: "16",
      fill: "currentColor",
      viewBox: "0 0 16 16",
      stroke: "none",
      xmlns: "http://www.w3.org/2000/svg"
    }, /*#__PURE__*/_createElement("path", {
      d: "M3.63135 2.33783L14.6313 14.3378L15.3685 13.6621L4.3685 1.66211L3.63135 2.33783ZM16 7.99997C16 9.34387 15.5582 10.5846 14.8118 11.5847L14.1194 10.8346C14.6748 10.029 15 9.05247 15 7.99997C15 5.23855 12.7614 2.99997 10 2.99997C9.0906 2.99997 8.2379 3.24275 7.50319 3.66704L6.81072 2.91686C7.73462 2.33596 8.82807 1.99997 10 1.99997C13.3137 1.99997 16 4.68626 16 7.99997ZM5 7.99997C5 7.24413 5.16771 6.52746 5.46795 5.88515L4.74501 5.10197C4.41672 5.696 4.18623 6.35161 4.07525 7.0471C3.8879 7.01608 3.69571 6.99997 3.5 6.99997C1.567 6.99997 0 8.56697 0 10.5C0 12.433 1.567 14 3.5 14H10C10.8802 14 11.7162 13.8104 12.4693 13.4699L11.7462 12.6866C11.2026 12.8892 10.6142 13 10 13H3.5C2.11929 13 1 11.8807 1 10.5C1 9.11926 2.11929 7.99997 3.5 7.99997C3.67163 7.99997 3.83886 8.01719 4.0002 8.04992C4.00578 8.73237 4.12548 9.38863 4.34154 9.99993L5.28438 9.66668C5.10035 9.14602 5 8.58526 5 7.99997Z"
    }))
  },
  offline: {
    color: _common.COLORS.LIGHT.LEVEL3_BASE_COLOR,
    icon: /*#__PURE__*/_createElement("svg", {
      width: "16",
      height: "16",
      fill: "currentColor",
      viewBox: "0 0 16 16",
      stroke: "none",
      xmlns: "http://www.w3.org/2000/svg"
    }, /*#__PURE__*/_createElement("path", {
      d: "M2.64648 3.35356L12.6465 13.3536L13.3536 12.6465L3.35359 2.64645L2.64648 3.35356ZM12.4001 9.19952L11.7839 9.66243L9.26847 7.14703C10.5399 7.44723 11.6407 8.18861 12.4001 9.19952ZM8.00009 4C7.41536 4 6.84442 4.05902 6.29289 4.17145L5.46384 3.34239C6.27121 3.11925 7.12173 3 8.00009 3C11.1083 3 13.8679 4.49325 15.6003 6.79938L14.8007 7.39999C13.2491 5.33447 10.7805 4 8.00009 4ZM3.29717 5.41837L2.57752 4.69872C1.74441 5.27885 1.00872 5.98892 0.399902 6.79938L1.19944 7.39999C1.78063 6.62631 2.49047 5.9552 3.29717 5.41837ZM6.23714 8.35834L5.48591 7.60711C4.74264 7.9898 4.09797 8.53673 3.60009 9.19952L4.39963 9.80013C4.87211 9.17116 5.50488 8.67041 6.23714 8.35834ZM6.80028 11.5997C7.07309 11.2365 7.50905 11 8.00009 11C8.49114 11 8.92709 11.2365 9.19991 11.5997L9.50045 11.9997L8.00009 13.125L6.49974 11.9997L6.80028 11.5997Z"
    }))
  }
};
const NetworkStatus = /*#__PURE__*/_react.default.memo(props => {
  const {
    status = 'saved',
    title = '',
    collapse = false,
    onClick
  } = props;
  const t = (0, _locale.useTranslate)();
  const render = (type, txt) => {
    const {
      color,
      icon
    } = infoMap[type];
    return /*#__PURE__*/_createElement(_styled.Box, {
      style: {
        color
      },
      onClick: onClick
    }, icon, !collapse ? /*#__PURE__*/_createElement(_styled.TextWrapper, {
      color: color
    }, txt) : null);
  };
  if (title) {
    return render('success', title);
  }

  // 网络只关注3种关键状态
  switch (status) {
    case 'offline':
      // 6
      // 已离线
      return render('offline', t('wdn_network_status_offline'));
    case 'reconnecting':
      // 7
      // 正在重连...
      return render('info', t('wdn_network_status_reconnecting'));
    case 'token_expired':
      // 8
      // 页面已过期
      return render('error', t('wdn_network_status_expired'));
    case 'save-processing':
      return render('info', t('wdn_network_status_save_processing'));
    case 'saved':
    case 'saved-success':
      return render('success', t('wdn_network_status_saved'));
    case 'save-failed':
      return render('error', t('wdn_network_status_save_failed'));
    default:
      break;
  }
  return null;
});
exports.NetworkStatus = NetworkStatus;
