// @ali/4ever-plugin-table/dist-web/index.js
import * as OldBambooPlugin from "./utils/bamboo"; // facotry

import { moFactory } from "./mo/index";
import { zhiFactory, mobileZhiFactory } from "./zhi";
import { biFactory, mobileBiFactory } from "./bi";
import "./bi/types";
import "./zhi/types";
import "./utils/types";
import tableCell from "./components/tableCell";
import tableRow from "./components/tableRow";
export { actions as biActions, actionTypes as biActionTypes } from "./bi/actions";
export { moFactory, biFactory, zhiFactory, mobileBiFactory, mobileZhiFactory }; // Element Model

export { default as Table } from "./mo/models";
export { default as TableCell } from "./mo/models/tableCell";
export { default as TableRow } from "./mo/models/tableRow"; // serializer

export { default as createSerializationPlugin } from "./mo/serializer/index";
export { createServerSerializationPlugin } from "./mo/serializer/index"; // types

// custom
export { useContainerWidthResize } from "./components/ContainerWidthResizeContext";
export { useRowIsSticky, RowIsStickyContext } from "./utils/hooks/useRowIsSticky";
export { isPureTableNode, isTableNode } from "./utils/utils/isTableNode";
export { default as getTableCellWidth } from "./utils/utils/getTableCellWidth";
export { default as setSelectionByTable } from "./utils/utils/setSelectionByTable";
export { default as getTable } from "./bi/utils/getTable";
export { default as getTableSelection } from "./bi/utils/getTableSelection";
export { default as createEmptyParagraph } from "./bi/utils/createEmptyParagraph";
export { default as createEmptyTableCell } from "./bi/utils/createEmptyTableCell";
export { default as isSelectionInTableCell } from "./bi/utils/isSelectionInTableCell";
export { default as setTableColsWidth } from "./bi/utils/setTableColsWidth";
export { handleCangjiePaste } from "./bi/utils/handleCangjiePaste";
export { default as onCangjieInput } from "./bi/handlers/onCangjieInput";
export var zhiTableCell = tableCell;
export var zhiTableRow = tableRow;
export { default as zhiTable } from "./zhi/components/table";
export { default as biTable } from "./bi/components/table";
export { default as biTableRow } from "./bi/components/tableRow";
export { default as biTableCell } from "./bi/components/tableCell";
export * from "./utils/utils/colsWidthConverter";
export { getAllTables } from "./common/queries/getSelectionContainers";
export { lineMap, DEFAULT_TABLE_BORDER_STYLE, DEFAULT_BORDER_COLOR, MIN_COL_WIDTH } from "./utils/constants"; // other export for api compatibility

export {
/** @deprecated 只为 bamboo 兼容。标准是按需 export */
OldBambooPlugin };
export { TD } from "./components/styled";
