// @ali/4ever-zhi/dist-web/plugins/sticker/index.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Inline } from '@ali/4ever-cangjie';
import { StickerPlugin } from '@ali/4ever-bamboo';
import { Sticker as MoSticker } from '@ali/4ever-mo';
import StickerWrapper from "./stickerWrapper";
import { moFactory } from '@ali/4ever-plugin-sticker';
var DEFAULT_STICKERS = StickerPlugin.DEFAULT_STICKERS,
    Sticker = StickerPlugin.Sticker;

function createStickerPlugin(config) {
  var stickerConfig = _extends({
    lang: 'zh_CN',
    customSticker: [DEFAULT_STICKERS]
  }, config.sticker);

  var sticker = new Sticker(stickerConfig);
  return _extends({}, moFactory({}), {
    renderNode: function renderNode(props, controller, next) {
      var node = props.node;

      if (!Inline.isInline(node) || node.type !== MoSticker.TYPE) {
        return next();
      }

      return /*#__PURE__*/_createElement(StickerWrapper, {
        node: node,
        controller: controller,
        sticker: sticker
      });
    },
    models: {
      klass: 'inline',
      type: MoSticker.TYPE
    }
  });
}

createStickerPlugin.defaultSticker = DEFAULT_STICKERS;
export default createStickerPlugin;
