// @ali/4ever-zhi/dist-web/plugins/sticker/stickerWrapper.js
import _extends from "@babel/runtime/helpers/extends";
import React, { useMemo } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { StickerPlugin, unitsConverter } from '@ali/4ever-bamboo';
import { StickerPlugin as BiStickerPlugin } from '@ali/4ever-bi';
var StickerPlaceholder = BiStickerPlugin.components.StickerPlaceholder;
var DEFAULT_SZ = StickerPlugin.DEFAULT_SZ,
    StickerItem = StickerPlugin.StickerItem,
    getStickerStyle = StickerPlugin.getStickerStyle;

function StickerWrapper(props) {
  var node = props.node,
      sticker = props.sticker;

  var _useMemo = useMemo(function () {
    return getStickerStyle(node);
  }, [node]),
      _useMemo$sz = _useMemo.sz,
      sz = _useMemo$sz === void 0 ? DEFAULT_SZ : _useMemo$sz,
      highlight = _useMemo.highlight;

  return /*#__PURE__*/_createElement(StickerItem, {
    key: node.key,
    code: node.data.code,
    backgroundColor: highlight,
    sz: unitsConverter.toPX(sz * 1.5 + "pt"),
    placeholder: /*#__PURE__*/_createElement(StickerPlaceholder, _extends({
      readonly: true
    }, props)),
    sticker: sticker
  });
}

export default /*#__PURE__*/React.memo(StickerWrapper);
