// @ali/4ever-bi/dist-node/plugins/hoverBlock/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HIGHLIGHT_CLASS = exports.NO_BORDER_HOVER_HIGHLIGHT = exports.NO_BORDER_SELECT_HIGHLIGHT = exports.BORDER_BOX_ACTIVE_SELECT_HIGHLIGHT = exports.BORDER_BOX_ACTIVE_HIGHLIGHT = exports.BORDER_BOX_HOVER_HIGHLIGHT = void 0;
const BORDER_BOX_HOVER_HIGHLIGHT = 'bi-border-box-hover-highligh';
exports.BORDER_BOX_HOVER_HIGHLIGHT = BORDER_BOX_HOVER_HIGHLIGHT;
const BORDER_BOX_ACTIVE_HIGHLIGHT = 'bi-border-box-active-highligh';
exports.BORDER_BOX_ACTIVE_HIGHLIGHT = BORDER_BOX_ACTIVE_HIGHLIGHT;
const BORDER_BOX_ACTIVE_SELECT_HIGHLIGHT = 'bi-border-box-active-select-highligh';
exports.BORDER_BOX_ACTIVE_SELECT_HIGHLIGHT = BORDER_BOX_ACTIVE_SELECT_HIGHLIGHT;
const NO_BORDER_SELECT_HIGHLIGHT = 'bi-no-border-box-select-highlight';
exports.NO_BORDER_SELECT_HIGHLIGHT = NO_BORDER_SELECT_HIGHLIGHT;
const NO_BORDER_HOVER_HIGHLIGHT = 'bi-no-border-box-hover-highlight';
exports.NO_BORDER_HOVER_HIGHLIGHT = NO_BORDER_HOVER_HIGHLIGHT;
const HIGHLIGHT_CLASS = {
  borderBoxActive: BORDER_BOX_ACTIVE_HIGHLIGHT,
  borderBoxHover: BORDER_BOX_HOVER_HIGHLIGHT,
  borderBoxSelect: BORDER_BOX_ACTIVE_SELECT_HIGHLIGHT,
  noBorderBoxSelect: NO_BORDER_SELECT_HIGHLIGHT,
  noBorderBoxHover: NO_BORDER_HOVER_HIGHLIGHT
};
exports.HIGHLIGHT_CLASS = HIGHLIGHT_CLASS;
