// @ali/4ever-plugin-table/dist-web/zhi/index.js
import _extends from "@babel/runtime/helpers/extends";
import { moFactory } from "../mo/index";
import { createRenderNode } from "./handlers/createRenderNode";
import queries from "./queries";
import getNodeStyle from "../common/handlers/getNodeStyle";
import onAction from "../utils/handlers/onAction";
import shouldNodeComponentUpdate from "../common/handlers/shouldNodeComponentUpdate";
import createOnCangjieCopy from "./handlers/createOnCangjieCopy";
export var createZhiFactory = function createZhiFactory(isMobile) {
  return function (configs) {
    var tableConfig = configs.table || {};
    return _extends({}, moFactory({}), {
      queries: queries,
      renderNode: createRenderNode(tableConfig, isMobile),
      getNodeStyle: getNodeStyle,
      onAction: onAction,
      shouldNodeComponentUpdate: shouldNodeComponentUpdate,
      onCangjieCopy: createOnCangjieCopy(isMobile)
    });
  };
};
export var zhiFactory = createZhiFactory();
export var mobileZhiFactory = createZhiFactory(true);
