// @ali/4ever-bi/dist-node/plugins/indent/components/IncreaseIndentButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = IncreaseIndentButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../actions");

var _hotKeyTips = require("../utils/hotKeyTips");

const _createElement = /*#__PURE__*/React.createElement;
;

function IncreaseIndentButton(props) {
  const {
    controller,
    locale,
    ...rest
  } = props;
  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInHr') || controller.query('isSelectionInListSymbol');
  const handleClick = React.useCallback(() => controller.run('onAction', (0, _actions.increaseIndent)()), [controller]);
  return /*#__PURE__*/_createElement(_weToolbar.IncreaseIndent, (0, _extends2.default)({}, rest, {
    shortcutText: _hotKeyTips.INCREASE_HOT_KEY_TIP,
    testid: "bi-toolbar-incindent",
    onClick: handleClick,
    disabled: !!isDisabled,
    role: _everBamboo.PluginRoles.indentIncrease
  }));
}
