// @ali/4ever-plugin-embed/dist-node/utils/components/common/embedFileCardWrapper/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = EmbedCardWrapper;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _everComponent = require("@ali/4ever-component");

var _fileTypeHelper = require("../../../fileTypeHelper");

var _styled = require("./styled");

const _createElement = /*#__PURE__*/_react.default.createElement;

function EmbedCardWrapper(props) {
  const {
    node,
    isSelected,
    isUploadFailed,
    children,
    onRef,
    enableBorderHighlight
  } = props;
  const {
    type
  } = node.data;
  const wrapperRef = (0, _react.useRef)(null);
  const [isHover, hoverHandlers] = (0, _everComponent.useHover)(enableBorderHighlight);
  (0, _react.useEffect)(() => {
    if (onRef) {
      onRef(wrapperRef.current);
    }
  }, [onRef]);
  return /*#__PURE__*/_createElement(_styled.EmbedFileCardWrapper, (0, _extends2.default)({
    isSelected: isSelected,
    isUploadFailed: isUploadFailed,
    "data-testid": "embed-card",
    "data-type": `embed-${type}`,
    ref: wrapperRef,
    "data-hover-box": "borderBox",
    isHover: isHover
  }, hoverHandlers), /*#__PURE__*/_createElement(_styled.CardIconWrapper, null, /*#__PURE__*/_createElement("img", {
    src: (0, _fileTypeHelper.getIcon)(type),
    alt: "icon",
    style: {
      width: 23,
      height: 24
    }
  })), children);
}
