// @ali/4ever-plugin-node-select/dist-web/bi/createGetNodeSelectToolbarItems.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { MobileIconCopy, MobileIconCut, MobileIconDelete } from '@ali/4ever-component';

var _ref = /*#__PURE__*/_createElement(MobileIconCopy, {
  color: "rgb(92, 113, 221)"
});

var _ref2 = /*#__PURE__*/_createElement(MobileIconCut, {
  color: "rgb(75, 173, 112)"
});

var _ref3 = /*#__PURE__*/_createElement(MobileIconDelete, {
  color: "rgb(228, 117, 117)"
});

export var createGetNodeSelectToolbarItems = function createGetNodeSelectToolbarItems(locale) {
  return function (props, _, next) {
    var copyItem = {
      key: 'copy',
      icon: _ref,
      action: 'copySelectedNodes',
      priority: 3,
      title: locale.copy || 'Copy'
    };
    var cutItem = {
      key: 'cut',
      icon: _ref2,
      action: 'cutSelectedNodes',
      priority: 2,
      title: locale.cut || 'Cut'
    };
    var deleteItem = {
      key: 'delete',
      icon: _ref3,
      action: 'deleteSelectedNodes',
      priority: 4,
      title: locale["delete"] || 'Delete'
    };
    var res = [copyItem, cutItem, deleteItem];
    var nextItems = next();
    return [].concat(res, Array.isArray(nextItems) ? nextItems : []);
  };
};
