// @ali/4ever-plugin-history/dist-node/bi/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEPRECATED_ACTIONS = exports.isActions = exports.actions = exports.isRedo = exports.redo = exports.HISTORY_REDO = exports.isUndo = exports.undo = exports.HISTORY_UNDO = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

// undoAction
const HISTORY_UNDO = Symbol('undoAction');
exports.HISTORY_UNDO = HISTORY_UNDO;
const [undo, isUndo] = (0, _everCangjie.createActionHandlers)(HISTORY_UNDO); // redoAction

exports.isUndo = isUndo;
exports.undo = undo;
const HISTORY_REDO = Symbol('redoAction');
exports.HISTORY_REDO = HISTORY_REDO;
const [redo, isRedo] = (0, _everCangjie.createActionHandlers)(HISTORY_REDO);
exports.isRedo = isRedo;
exports.redo = redo;
const actions = {
  undo,
  redo
};
exports.actions = actions;
const isActions = {
  isUndo,
  isRedo
};
/** @deprecated */

exports.isActions = isActions;
const DEPRECATED_ACTIONS = {
  HISTORY_REDO,
  HISTORY_UNDO,
  undo,
  redo
};
exports.DEPRECATED_ACTIONS = DEPRECATED_ACTIONS;
