// @ali/4ever-plugin-history/dist-node/bi/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.biFactory = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _queries = require("./queries");

var _commands = require("./commands");

var _actions = require("./actions");

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _createRenderToolbar = _interopRequireDefault(require("./createRenderToolbar"));

var _meta = require("../utils/meta");

const biFactory = config => {
  const historyConfig = config.history || {};
  return {
    key: _meta.key,
    toolbar: (0, _toolbar.default)(historyConfig),
    renderToolbar: (0, _createRenderToolbar.default)(historyConfig),
    queries: {
      canUndo: historyConfig.canUndo ? historyConfig.canUndo : _queries.queries.canUndo,
      canRedo: historyConfig.canRedo ? historyConfig.canRedo : _queries.queries.canRedo
    },
    commands: _commands.commands,
    actions: _actions.actions,
    isActions: _actions.isActions,
    onAction: _onAction.default,
    onKeyDown: _onKeyDown.default,
    models: []
  };
};

exports.biFactory = biFactory;
