// @ali/4ever-plugin-paragraph/dist-node/bi/actions/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "setParagraphInd", {
  enumerable: true,
  get: function () {
    return _setParagraphInd.setParagraphInd;
  }
});
Object.defineProperty(exports, "isSetParagraphInd", {
  enumerable: true,
  get: function () {
    return _setParagraphInd.isSetParagraphInd;
  }
});
Object.defineProperty(exports, "PARAGRAPH_SET_IND", {
  enumerable: true,
  get: function () {
    return _setParagraphInd.PARAGRAPH_SET_IND;
  }
});
Object.defineProperty(exports, "setParagraphSpacing", {
  enumerable: true,
  get: function () {
    return _setParagraphSpacing.setParagraphSpacing;
  }
});
Object.defineProperty(exports, "isSetParagraphSpacing", {
  enumerable: true,
  get: function () {
    return _setParagraphSpacing.isSetParagraphSpacing;
  }
});
Object.defineProperty(exports, "PARAGRAPH_SET_SPACING", {
  enumerable: true,
  get: function () {
    return _setParagraphSpacing.PARAGRAPH_SET_SPACING;
  }
});
exports.isActions = exports.actions = void 0;

var _setParagraphInd = require("./setParagraphInd");

var _setParagraphSpacing = require("./setParagraphSpacing");

const actions = {
  setParagraphInd: _setParagraphInd.setParagraphInd,
  setParagraphSpacing: _setParagraphSpacing.setParagraphSpacing
};
exports.actions = actions;
const isActions = {
  isSetParagraphInd: _setParagraphInd.isSetParagraphInd,
  isSetParagraphSpacing: _setParagraphSpacing.isSetParagraphSpacing
};
exports.isActions = isActions;
