// @ali/4ever-bi/dist-web/plugins/fonts/components/next/fontsButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { useStyle } from '@ali/4ever-cangjie';
import { PluginRoles } from '@ali/4ever-bamboo';
import { FontFamilySelect } from '@ali/4ever-component';
;
import { setFonts } from "../../actions";
import { defaultFonts } from "../../utils/getActiveFontsMark";
import getFontsTip from "../../utils/getFontsTip";
export default function FontsButton(props) {
  var _docStyle$docDefaults, _docStyle$docDefaults2, _docStyle$docDefaults3;

  var controller = props.controller,
      locale = props.locale,
      _props$width = props.width,
      width = _props$width === void 0 ? 133 : _props$width,
      rest = _objectWithoutPropertiesLoose(props, ["controller", "locale", "width"]);

  var docStyle = useStyle();
  var fonts = controller.query('getFonts', docStyle == null ? void 0 : (_docStyle$docDefaults = docStyle.docDefaults) == null ? void 0 : (_docStyle$docDefaults2 = _docStyle$docDefaults.data) == null ? void 0 : (_docStyle$docDefaults3 = _docStyle$docDefaults2.rPr) == null ? void 0 : _docStyle$docDefaults3.fonts);
  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInLinkCard');
  var handleChange = React.useCallback(function (fontsValue) {
    return controller.run('onAction', setFonts(fontsValue));
  }, [controller]);
  var title = (locale == null ? void 0 : locale.buttonTitle) || '';

  if (fonts && fonts !== defaultFonts) {
    // 提示显示当前字体
    title += " " + getFontsTip(fonts, locale);
  }

  return /*#__PURE__*/_createElement(FontFamilySelect, _extends({
    testid: "bi-toolbar-fonts",
    disabled: !!isDisabled,
    value: fonts || (locale == null ? void 0 : locale.buttonTitle),
    onChange: handleChange,
    role: PluginRoles.fontFamilySelect,
    width: width,
    tooltip: title || 'Font Family'
  }, rest));
}
