// @ali/4ever-bi/dist-node/plugins/sz/components/next/SzButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SzButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everComponent = require("@ali/4ever-component");

var _actions = require("../../actions");

const _createElement = /*#__PURE__*/React.createElement;
;

function SzButton(props) {
  const {
    controller,
    locale,
    ...rest
  } = props;
  const defaultSz = (0, _everComponent.useDefaultSz)();
  const [_, setSzValue] = React.useState(undefined);
  let sz = controller.query('getSize', defaultSz);

  if ((0, _everBamboo.isNaN)(sz)) {
    sz = undefined;
  }

  const {
    isCollapsedAtEdge
  } = controller.value.selection;
  const isDisabled = isCollapsedAtEdge || controller.query('isSelectionInImage') || controller.query('isSelectionInHr') || controller.query('isSelectionInLinkCard');
  const handleChange = React.useCallback(size => {
    controller.run('onAction', (0, _actions.setSz)(size));
    setSzValue(size);
  }, [controller]);
  return /*#__PURE__*/_createElement(_everComponent.FontSizeSelect, (0, _extends2.default)({
    testid: "bi-toolbar-sz",
    disabled: isDisabled,
    value: sz,
    onChange: handleChange,
    tooltip: locale?.buttonTitle || '',
    role: _everBamboo.PluginRoles.sz
  }, rest));
}
