// @ali/4ever-bi/dist-node/plugins/sz/components/next/SzSingleButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createSzSmallerButton = createSzSmallerButton;
exports.createSzBiggerButton = createSzBiggerButton;

var React = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everComponent = require("@ali/4ever-component");

var _actions = require("../../actions");

var _getTipsByType = require("../../utils/getTipsByType");

const _createElement = /*#__PURE__*/React.createElement;

var _ref = /*#__PURE__*/_createElement(_everComponent.IconReduceFont, null);

function createSzSmallerButton(props) {
  const {
    controller,
    locale,
    icon,
    title,
    tooltip,
    badgeDot,
    disabled
  } = props;

  const handleClick = () => {
    controller.run('onAction', (0, _actions.adjustSz)(false));
  };

  const isDisabled = disabled || controller.query('isSelectionInLinkCard');
  const buttonConfig = {
    icon: icon || _ref,
    onClick: handleClick,
    title: title || locale?.buttonTitle,
    tooltip: {
      title: title || locale?.buttonTitle,
      shortCut: _getTipsByType.SHORTCUT.smaller,
      ...tooltip
    },
    badgeDot,
    disabled: !!isDisabled,
    testid: "ReduceFontSizeButton",
    role: _everBamboo.PluginRoles.szDecrease
  };
  return buttonConfig;
}

var _ref2 = /*#__PURE__*/_createElement(_everComponent.IconFont, null);

function createSzBiggerButton(props) {
  const {
    controller,
    locale,
    icon,
    title,
    tooltip,
    badgeDot,
    disabled
  } = props;

  const handleClick = () => {
    controller.run('onAction', (0, _actions.adjustSz)(true));
  };

  const isDisabled = disabled || controller.query('isSelectionInLinkCard');
  const buttonConfig = {
    icon: icon || _ref2,
    onClick: handleClick,
    title: title || locale?.buttonTitle,
    tooltip: {
      title: title || locale?.buttonTitle,
      shortCut: _getTipsByType.SHORTCUT.bigger,
      ...tooltip
    },
    badgeDot,
    disabled: !!isDisabled,
    testid: "add_fontsize_button",
    role: _everBamboo.PluginRoles.szIncrease
  };
  return buttonConfig;
}
