// @ali/we-design-next/dist-web/form/color-picker/ColorPickerGradientRow.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { COLORS } from "../../common";
import { Row, ColorSquareWrap } from "./styled";
import { isWhite } from "./isWhite";
var gradients = ['#FFFFFF', '#28B4BE__#3935DD', '#2449FE__#CA4BA7', '#FF7100__#F300AD', '#B16A4D__#D5A638', '#1890FF__#4FF3E0', '#F3AA00__#E43E29', '#EB47FF__#F95101', '#10CCF7__#AB44FB', '#F55B2D__#510101'];
export var PresetGradient = /*#__PURE__*/React.memo(function (props) {
  var value = props.value,
    onSelect = props.onSelect;
  return /*#__PURE__*/_createElement(Row, null, gradients.map(function (g, i) {
    var style = {};
    if (g.indexOf('__') === -1) {
      style.background = g;
      if (isWhite(g)) {
        style.border = "2px solid " + COLORS.OTHERS.BORDER;
      }
    } else {
      var _g$split = g.split('__'),
        c1 = _g$split[0],
        c2 = _g$split[1];
      style.backgroundImage = "linear-gradient(90deg, " + c1 + ", " + c2 + ")";
      style.backgroundOrigin = 'border-box';
    }
    return /*#__PURE__*/_createElement(ColorSquareWrap, {
      onClick: function onClick() {
        return onSelect(g);
      },
      style: style
    }, g === value ? /*#__PURE__*/_createElement("svg", {
      width: "14",
      height: "10",
      viewBox: "0 0 16 12",
      fill: "none"
    }, /*#__PURE__*/_createElement("path", {
      d: "M13.965 2.15533L5.63499 10.485L1.46999 6.32034",
      stroke: i === 0 ? '#000000' : '#ffffff',
      strokeOpacity: "0.6",
      strokeWidth: "2.38",
      strokeLinecap: "round",
      strokeLinejoin: "round"
    })) : null);
  }));
});
