// @ali/we-design-next/dist-web/form/color-picker/styled.js
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { COLORS, Divider, RowItem } from "../../common";
export var ColorPickerWrapper = /*#__PURE__*/styled.div(["user-select:none;margin:4px 8px;"]);
export var HeaderGroup = /*#__PURE__*/styled.div(["position:relative;display:flex;justify-content:space-between;align-items:center;padding:4px;"]);
export var AutoColorWrapper = /*#__PURE__*/styled.div(["padding:4px;display:flex;align-items:center;overflow:hidden;border-radius:4px;"]);
export var NoColorWrapper = /*#__PURE__*/styled(AutoColorWrapper)(["&:hover{background:", ";}transition:background 0.1s ease-in-out;"], COLORS.OTHERS.DAO_HOVER_BG_COLOR);
export var Group = /*#__PURE__*/styled.div(["position:relative;padding:4px 8px 8px 8px;"]);
export var RecentGroup = /*#__PURE__*/styled.div(["position:relative;padding:4px 8px;"]);
export var Row = /*#__PURE__*/styled.div(["position:relative;display:flex;width:100%;min-height:6px;margin-bottom:3px;&:last-child{margin-bottom:0;}"]);
export var AdvanceGroup = RowItem;
export var ColorSquareWrap = /*#__PURE__*/styled.span(["position:relative;display:inline-block;width:20px;height:20px;box-sizing:border-box;margin-right:3px;&:last-child{margin-right:0;}border:2px solid ", ";border-radius:4px 4px;flex:0 0 auto;background-color:", ";&:hover{padding:2px 2px;border:2px solid ", ";& svg{top:2px;left:2px;}}& svg{position:absolute;top:2px;left:2px;width:12px;height:12px;}& ._wdn-color-picker-text-a{display:none;}&:hover ._wdn-color-picker-text-a{display:flex;}"], function (_ref) {
  var alwaysShowBorder = _ref.alwaysShowBorder;
  return alwaysShowBorder ? COLORS.OTHERS.BORDER : COLORS.TRANSPARENT;
}, COLORS.WHITE, COLORS.OTHERS.PRIMARY_HOVER);
export var NullColorSquareWrap = /*#__PURE__*/styled(ColorSquareWrap)(["&:after{content:'';display:block;position:absolute;top:8px;left:-2px;width:22px;height:0;border-bottom:1px solid ", ";transform:rotate(-45deg);}&:hover{border:1px solid ", ";}border:1px solid ", ";overflow:hidden;margin:0;"], COLORS.COMMON.DANGER1_COLOR, COLORS.OTHERS.BORDER, COLORS.OTHERS.BORDER);
export { Divider };
export var ColorTextA = /*#__PURE__*/styled.span(["position:absolute;left:1px;right:0;bottom:0;top:1px;display:flex;justify-content:center;align-items:center;font-size:12px;"]);
export var SketchPickerWrap = /*#__PURE__*/styled.div(["input{box-sizing:content-box;}"]);
export var ArrowRight = /*#__PURE__*/React.memo(function () {
  return /*#__PURE__*/_createElement("svg", {
    width: "12",
    height: "12",
    viewBox: "0 0 12 12",
    fill: "none"
  }, /*#__PURE__*/_createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M7.54293 6.25004L3.39648 2.10359L4.10359 1.39648L8.95714 6.25004L4.10359 11.1036L3.39648 10.3965L7.54293 6.25004Z",
    fill: COLORS.LIGHT.LEVEL1_BASE_COLOR,
    fillOpacity: "0.6"
  }));
});
