// @ali/4ever-plugin-embed/dist-node/utils/components/common/embedCard/embedUploadProgressCard.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = EmbedUploadProgressCard;

var _react = _interopRequireWildcard(require("react"));

var _embedFileCardWrapper = _interopRequireDefault(require("../embedFileCardWrapper"));

var _styled = require("./styled");

const _createElement = /*#__PURE__*/_react.default.createElement;

function EmbedUploadProgressCard(props) {
  const {
    node,
    isSelected,
    enableBorderHighlight
  } = props;
  const {
    uploadPercent,
    name,
    errorType
  } = node.data;
  const getUploadProgress = (0, _react.useCallback)(() => {
    return uploadPercent ? uploadPercent * 0.95 : 0;
  }, [uploadPercent]);
  return /*#__PURE__*/_createElement(_embedFileCardWrapper.default, {
    node: node,
    isUploadFailed: !!errorType,
    isSelected: isSelected,
    enableBorderHighlight: enableBorderHighlight
  }, /*#__PURE__*/_createElement(_styled.CardContentWrapper, null, /*#__PURE__*/_createElement(_styled.CardContentInnerWrapper, null, /*#__PURE__*/_createElement(_styled.ContentHeader, null, name)), uploadPercent ? /*#__PURE__*/_createElement(_styled.ProgressBar, null, /*#__PURE__*/_createElement(_styled.ProgressBarInner, {
    progress: getUploadProgress()
  })) : null));
}
