// @ali/4ever-plugin-embed/dist-node/utils/components/common/embedCard/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styled = require("./styled");

var _embedDownloadButton = _interopRequireDefault(require("../embedDownloadButton"));

var _embedFileCardWrapper = _interopRequireDefault(require("../embedFileCardWrapper"));

var _fileTypeHelper = require("../../../fileTypeHelper");

const _createElement = /*#__PURE__*/_react.default.createElement;

function EmbedCard(props) {
  const {
    name,
    allowDownload,
    isSelected,
    node,
    locale,
    size,
    onDownloadFile
  } = props;
  const renderDesc = (0, _react.useCallback)(() => {
    return /*#__PURE__*/_createElement(_styled.ContentDesc, null, (0, _fileTypeHelper.formatSize)(size));
  }, [size]);
  return /*#__PURE__*/_createElement(_embedFileCardWrapper.default, {
    node: node,
    isSelected: isSelected
  }, /*#__PURE__*/_createElement(_react.default.Fragment, null, /*#__PURE__*/_createElement(_styled.CardContentWrapper, null, /*#__PURE__*/_createElement(_styled.CardContentInnerWrapper, null, /*#__PURE__*/_createElement(_styled.ContentHeader, null, name), renderDesc())), /*#__PURE__*/_createElement(_embedDownloadButton.default, {
    title: locale?.title,
    onDownloadFile: onDownloadFile,
    allowDownload: allowDownload
  })));
}

var _default = EmbedCard;
exports.default = _default;
