// @ali/4ever-plugin-embed/dist-node/utils/components/common/embedCard/mobileEmbedCard.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _fileTypeHelper = require("../../../fileTypeHelper");

var _styled = require("./styled");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement(_weDesign.PreviewBetaNormal, null);

function MobileEmbedCard(props) {
  const {
    name,
    onPreview,
    src,
    type,
    size
  } = props;
  const isClickedRef = (0, _react.useRef)(false);
  const fileType = (0, _fileTypeHelper.getFileTypeForRead)(name);

  const getIcon = () => {
    return _fileTypeHelper.IconUrlMap[type] || _fileTypeHelper.IconUrlMap['app'];
  };

  const renderDesc = () => {
    return /*#__PURE__*/_createElement(_styled.ContentDesc, null, (0, _fileTypeHelper.formatSize)(size));
  };

  const handleTouchStart = e => {
    if (_fileTypeHelper.CanPreviewFileTypes.includes(fileType)) {
      e.preventDefault();
      e.stopPropagation();
      isClickedRef.current = true;
    }
  };

  const handleTouchMove = e => {
    if (_fileTypeHelper.CanPreviewFileTypes.includes(fileType)) {
      e.preventDefault();
      e.stopPropagation();
      isClickedRef.current = false;
    }
  };

  const handleTouchEnd = e => {
    // TODO 此处限制了业务层的拓展，之后应当移除
    if (_fileTypeHelper.CanPreviewFileTypes.includes(fileType)) {
      e.preventDefault();
      e.stopPropagation();

      if (onPreview && isClickedRef.current) {
        onPreview(src);
        isClickedRef.current = false;
      }
    }
  };

  return /*#__PURE__*/_createElement(_styled.MobileEmbedCardWrapper, null, /*#__PURE__*/_createElement(_react.default.Fragment, null, /*#__PURE__*/_createElement(_styled.CardIconWrapper, null, /*#__PURE__*/_createElement("img", {
    src: getIcon(),
    alt: "icon",
    style: {
      width: 26,
      height: 28
    }
  })), /*#__PURE__*/_createElement(_styled.CardContentWrapper, null, /*#__PURE__*/_createElement(_styled.MobileCardContentInnerWrapper, null, /*#__PURE__*/_createElement(_styled.ContentHeader, null, name), renderDesc())), _fileTypeHelper.CanPreviewFileTypes.includes(fileType) && onPreview ? /*#__PURE__*/_createElement(_styled.PreviewIconWrapper, {
    onTouchStart: handleTouchStart,
    onTouchMove: handleTouchMove,
    onTouchEnd: handleTouchEnd
  }, _ref) : null));
}

var _default = MobileEmbedCard;
exports.default = _default;
