// @ali/4ever-plugin-embed/dist-node/utils/components/common/embedCard/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ErrorDesc = exports.ProgressBarInner = exports.ProgressBar = exports.PreviewIconWrapper = exports.CardIconWrapper = exports.MobileEmbedCardWrapper = exports.ContentDesc = exports.ContentHeader = exports.CardContentWrapper = exports.MobileCardContentInnerWrapper = exports.CardContentInnerWrapper = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

const CardContentInnerWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;margin-top:'';"]);

exports.CardContentInnerWrapper = CardContentInnerWrapper;

const MobileCardContentInnerWrapper = /*#__PURE__*/_styledComponents.default.div(["display:block;"]);

exports.MobileCardContentInnerWrapper = MobileCardContentInnerWrapper;

const CardContentWrapper = /*#__PURE__*/_styledComponents.default.div(["flex:1;min-width:0;"]);

exports.CardContentWrapper = CardContentWrapper;

const ContentHeader = /*#__PURE__*/_styledComponents.default.div(["font-size:14px;color:", ";flex:1;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;"], props => props.mode === 'dark' ? '#F2F4F5' : '#111F2C');

exports.ContentHeader = ContentHeader;

const ContentDesc = /*#__PURE__*/_styledComponents.default.div(["font-size:12px;color:", ";margin-right:16px;"], props => props.mode === 'dark' ? '#F2F4F5' : '#7A8189');

exports.ContentDesc = ContentDesc;

const MobileEmbedCardWrapper = /*#__PURE__*/_styledComponents.default.div(["width:100%;height:54px;border:1px solid #D9DBDD;padding:8px 0 8px 16px;display:flex;border-radius:6px;background-color:#fff;box-sizing:border-box;"]);

exports.MobileEmbedCardWrapper = MobileEmbedCardWrapper;

const CardIconWrapper = /*#__PURE__*/_styledComponents.default.div(["margin-right:16px;flex:0;margin-top:6px;"]);

exports.CardIconWrapper = CardIconWrapper;

const PreviewIconWrapper = /*#__PURE__*/_styledComponents.default.div(["flex:0;padding:8px 16px;border-left:1px solid #E2E4E5;cursor:pointer;"]);

exports.PreviewIconWrapper = PreviewIconWrapper;

const ProgressBar = /*#__PURE__*/_styledComponents.default.div(["width:100%;height:4px;background-color:#D9DBDD;border-radius:4px;"]);

exports.ProgressBar = ProgressBar;

const ProgressBarInner = /*#__PURE__*/_styledComponents.default.div(["background-color:#0089FF;width:", ";height:100%;transition:width .6s;border-radius:4px;"], props => `${props.progress}%`);

exports.ProgressBarInner = ProgressBarInner;

const ErrorDesc = /*#__PURE__*/_styledComponents.default.div(["font-size:12px;color:#D0021B;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;"]);

exports.ErrorDesc = ErrorDesc;
