// @ali/4ever-plugin-heading-group/dist-node/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "isFoldInGroup", {
  enumerable: true,
  get: function () {
    return _helpers.isFoldInGroup;
  }
});
Object.defineProperty(exports, "isUnfoldInGroup", {
  enumerable: true,
  get: function () {
    return _helpers.isUnfoldInGroup;
  }
});
Object.defineProperty(exports, "createAutoUnfold", {
  enumerable: true,
  get: function () {
    return _handlers.createAutoUnfold;
  }
});
Object.defineProperty(exports, "createRenderMark", {
  enumerable: true,
  get: function () {
    return _handlers.createRenderMark;
  }
});
Object.defineProperty(exports, "createCommentShow", {
  enumerable: true,
  get: function () {
    return _handlers.createCommentShow;
  }
});
Object.defineProperty(exports, "Title", {
  enumerable: true,
  get: function () {
    return _title.default;
  }
});
Object.defineProperty(exports, "Content", {
  enumerable: true,
  get: function () {
    return _content.default;
  }
});
Object.defineProperty(exports, "headingGroupList", {
  enumerable: true,
  get: function () {
    return _conditions.headingGroupList;
  }
});
Object.defineProperty(exports, "TOGGLE_FOLD", {
  enumerable: true,
  get: function () {
    return _actions.TOGGLE_FOLD;
  }
});
Object.defineProperty(exports, "toggleFold", {
  enumerable: true,
  get: function () {
    return _actions.toggleFold;
  }
});
Object.defineProperty(exports, "getNextViewGroupNode", {
  enumerable: true,
  get: function () {
    return _utils.getNextViewGroupNode;
  }
});
Object.defineProperty(exports, "getPreviousViewGroupNode", {
  enumerable: true,
  get: function () {
    return _utils.getPreviousViewGroupNode;
  }
});
Object.defineProperty(exports, "isHeadingGroupAvailable", {
  enumerable: true,
  get: function () {
    return _utils.isHeadingGroupAvailable;
  }
});
exports.OldBambooPlugin = void 0;

require("@ali/4ever-plugin-fold");

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;

var _helpers = require("./utils/helpers");

var _handlers = require("./utils/handlers");

var _title = _interopRequireDefault(require("./utils/components/title"));

var _content = _interopRequireDefault(require("./utils/components/content"));

var _conditions = require("./utils/conditions");

var _actions = require("./utils/actions");

var _utils = require("./utils/utils");
