// @ali/4ever-plugin-mention/dist-node/bi/mobile/insertMention.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

var _removeMention = _interopRequireDefault(require("./removeMention"));

var _utils = require("../utils");

var _default = (controller, results) => {
  if (!results) return controller;
  const {
    value
  } = controller;

  if (controller.query('isCollapsedInInlineVoid')) {
    controller.command(_everCangjie.Commands.moveToStartOfNextText);
  }

  const {
    selection,
    document
  } = value;
  const start = selection.getStart(document);
  const at = (0, _utils.nearsetAt)(value.endText.text, start.offset);
  const mentionLength = start.offset - at;
  controller.command(_everCangjie.Commands.deleteBackward, mentionLength);
  results.forEach(element => {
    controller.command(_everCangjie.Commands.insertInline, {
      type: 'mention',
      data: {
        id: element.id,
        login: element.uid,
        name: element.name,
        avatarUrl: element.avatar
      }
    });
  });
  controller.command(_everCangjie.Commands.moveToStartOfNextText).command(_everCangjie.Commands.focus);
  return (0, _removeMention.default)(controller);
};

exports.default = _default;
