// @ali/4ever-plugin-clipboard/dist-web/bi/actions.js
var CLIPBOARD_INSERT_FRAGMENT = Symbol('CLIPBOARD_INSERT_FRAGMENT');
var CLIPBOARD_SET_DECORATIONS = Symbol('CLIPBOARD_SET_DECORATIONS');
var CLIPBOARD_DEL_SELECTION = Symbol('CLIPBOARD_DEL_SELECTION');
var CLIPBOARD_HIDE_PASTEPANEL = Symbol('CLIPBOARD_HIDE_PASTEPANEL');
var CLIPBOARD_HIDE_CUSTOM_PASTEPANEL = Symbol('CLIPBOARD_HIDE_CUSTOM_PASTEPANEL');
var CLIPBOARD_MATCH_MODE = Symbol('CLIPBOARD_MATCH_MODE');
var CLIPBOARD_TEXT_MODE = Symbol('CLIPBOARD_TEXT_MODE');
export { CLIPBOARD_INSERT_FRAGMENT, CLIPBOARD_SET_DECORATIONS, CLIPBOARD_DEL_SELECTION, CLIPBOARD_HIDE_PASTEPANEL, CLIPBOARD_HIDE_CUSTOM_PASTEPANEL, CLIPBOARD_MATCH_MODE, CLIPBOARD_TEXT_MODE };
export function insertFragment(fragment) {
  return {
    type: CLIPBOARD_INSERT_FRAGMENT,
    payload: {
      fragment: fragment
    }
  };
}
export function setDecorations(decoration) {
  return {
    type: CLIPBOARD_SET_DECORATIONS,
    payload: {
      decoration: decoration
    }
  };
}
export function deleteAtRange() {
  return {
    type: CLIPBOARD_DEL_SELECTION
  };
}
export function hidePastePanel() {
  return {
    type: CLIPBOARD_HIDE_PASTEPANEL
  };
}
export function hideCustomPastePanel() {
  return {
    type: CLIPBOARD_HIDE_CUSTOM_PASTEPANEL
  };
}
export function insertMatchFragment(fragment) {
  return {
    type: CLIPBOARD_MATCH_MODE,
    payload: {
      fragment: fragment
    }
  };
}
export function insertTextFragment(fragment) {
  return {
    type: CLIPBOARD_TEXT_MODE,
    payload: {
      fragment: fragment
    }
  };
}
export var actions = {
  insertFragment: insertFragment,
  setDecorations: setDecorations,
  deleteAtRange: deleteAtRange,
  hidePastePanel: hidePastePanel,
  hideCustomPastePanel: hideCustomPastePanel,
  insertMatchFragment: insertMatchFragment,
  insertTextFragment: insertTextFragment
};
