// @ali/4ever-plugin-clipboard/dist-web/bi/index.js
import _extends from "@babel/runtime/helpers/extends";
import createOnCangjiePaste from "./handlers/createOnCangjiePaste";
import createOnCangjieCopyOrCut from "../common/handlers/createOnCangjieCopyOrCut";
import onCangjieInput from "./handlers/onCangjieInput";
import ClipboardPad from "../utils/ClipboardPad";
import createOnAction from "./handlers/onAction";
import onKeyDown from "./handlers/onKeyDown";
import createOnEditable from "./handlers/createOnEditable";
import { copyNodes } from "../common/queries/copyNodes";
import createGetClipboardData from "../common/queries/getClipboardData";
import { key } from "../utils/meta";
import { commands } from "./commands";
import { actions } from "./actions";
export var biFactory = function biFactory(configs) {
  var clipboard = configs.clipboard || {};
  var onCangjieCopyOrCut = createOnCangjieCopyOrCut(clipboard);
  return {
    onConstruct: function onConstruct(controller, next) {
      ClipboardPad.getClipboardPad(_extends({
        serializer: controller.serializer,
        schema: controller.strictSchema
      }, clipboard), clipboard.key);
      return next();
    },
    key: key,
    queries: {
      copyNodes: copyNodes,
      getClipboardData: createGetClipboardData(clipboard)
    },
    actions: actions,
    commands: commands,
    onAction: createOnAction(clipboard),
    onCangjieInput: onCangjieInput,
    onKeyDown: onKeyDown,
    onCangjieCopy: onCangjieCopyOrCut,
    onCangjieCut: onCangjieCopyOrCut,
    onCangjiePaste: createOnCangjiePaste(clipboard),
    renderEditable: createOnEditable(clipboard)
  };
};
