// @ali/4ever-zhi/dist-node/plugins/sticker/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _everCangjie = require("@ali/4ever-cangjie");

var _everBamboo = require("@ali/4ever-bamboo");

var _everMo = require("@ali/4ever-mo");

var _stickerWrapper = _interopRequireDefault(require("./stickerWrapper"));

var _everPluginSticker = require("@ali/4ever-plugin-sticker");

const _createElement = /*#__PURE__*/_react.default.createElement;
const {
  DEFAULT_STICKERS,
  Sticker
} = _everBamboo.StickerPlugin;

function createStickerPlugin(config) {
  const stickerConfig = {
    lang: 'zh_CN',
    customSticker: [DEFAULT_STICKERS],
    ...config.sticker
  };
  const sticker = new Sticker(stickerConfig);
  return { ...(0, _everPluginSticker.moFactory)({}),

    renderNode(props, controller, next) {
      const {
        node
      } = props;

      if (!_everCangjie.Inline.isInline(node) || node.type !== _everMo.Sticker.TYPE) {
        return next();
      }

      return /*#__PURE__*/_createElement(_stickerWrapper.default, {
        node: node,
        controller: controller,
        sticker: sticker
      });
    },

    models: {
      klass: 'inline',
      type: _everMo.Sticker.TYPE
    }
  };
}

createStickerPlugin.defaultSticker = DEFAULT_STICKERS;
var _default = createStickerPlugin;
exports.default = _default;
