// @ali/4ever-zhi/dist-node/plugins/sticker/stickerWrapper.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _everBamboo = require("@ali/4ever-bamboo");

var _everBi = require("@ali/4ever-bi");

const _createElement = /*#__PURE__*/_react.default.createElement;
const {
  StickerPlaceholder
} = _everBi.StickerPlugin.components;
const {
  DEFAULT_SZ,
  StickerItem,
  getStickerStyle
} = _everBamboo.StickerPlugin;

function StickerWrapper(props) {
  const {
    node,
    sticker
  } = props;
  const {
    sz = DEFAULT_SZ,
    highlight
  } = (0, _react.useMemo)(() => getStickerStyle(node), [node]);
  return /*#__PURE__*/_createElement(StickerItem, {
    key: node.key,
    code: node.data.code,
    backgroundColor: highlight,
    sz: _everBamboo.unitsConverter.toPX(`${sz * 1.5}pt`),
    placeholder: /*#__PURE__*/_createElement(StickerPlaceholder, (0, _extends2.default)({
      readonly: true
    }, props)),
    sticker: sticker
  });
}

var _default = /*#__PURE__*/_react.default.memo(StickerWrapper);

exports.default = _default;
