// @ali/4ever-plugin-paragraph/dist-node/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "moFactory", {
  enumerable: true,
  get: function () {
    return _index.default;
  }
});
Object.defineProperty(exports, "zhiFactory", {
  enumerable: true,
  get: function () {
    return _index2.default;
  }
});
Object.defineProperty(exports, "biFactory", {
  enumerable: true,
  get: function () {
    return _index3.default;
  }
});
Object.defineProperty(exports, "Paragraph", {
  enumerable: true,
  get: function () {
    return _models.default;
  }
});
Object.defineProperty(exports, "ELEMENT_TYPE", {
  enumerable: true,
  get: function () {
    return _meta.ELEMENT_TYPE;
  }
});
Object.defineProperty(exports, "createSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index4.default;
  }
});
Object.defineProperty(exports, "createServerSerializationPlugin", {
  enumerable: true,
  get: function () {
    return _index4.createServerSerializationPlugin;
  }
});
Object.defineProperty(exports, "setParagraphSpacing", {
  enumerable: true,
  get: function () {
    return _actions.setParagraphSpacing;
  }
});
Object.defineProperty(exports, "setParagraphInd", {
  enumerable: true,
  get: function () {
    return _actions.setParagraphInd;
  }
});
Object.defineProperty(exports, "PARAGRAPH_SET_SPACING", {
  enumerable: true,
  get: function () {
    return _actions.PARAGRAPH_SET_SPACING;
  }
});
Object.defineProperty(exports, "PARAGRAPH_SET_IND", {
  enumerable: true,
  get: function () {
    return _actions.PARAGRAPH_SET_IND;
  }
});
Object.defineProperty(exports, "isPureParagraphValue", {
  enumerable: true,
  get: function () {
    return _isPureParagraphValue.default;
  }
});
Object.defineProperty(exports, "ReadParagraph", {
  enumerable: true,
  get: function () {
    return _ReadParagraph.default;
  }
});
Object.defineProperty(exports, "EditParagraph", {
  enumerable: true,
  get: function () {
    return _EditParagraph.default;
  }
});
Object.defineProperty(exports, "P", {
  enumerable: true,
  get: function () {
    return _styled.Component;
  }
});
Object.defineProperty(exports, "ParagraphRole", {
  enumerable: true,
  get: function () {
    return _role.ParagraphRole;
  }
});
exports.OldBambooPlugin = exports.mobileZhiFactory = exports.mobileBiFactory = void 0;

var _index = _interopRequireDefault(require("./mo/index"));

var _index2 = _interopRequireDefault(require("./zhi/index"));

var _index3 = _interopRequireDefault(require("./bi/index"));

var _models = _interopRequireDefault(require("./mo/models"));

var _meta = require("./mo/meta");

var _index4 = _interopRequireWildcard(require("./mo/serializer/index"));

var _actions = require("./bi/actions");

var _isPureParagraphValue = _interopRequireDefault(require("./mo/serializer/utils/isPureParagraphValue"));

var _ReadParagraph = _interopRequireDefault(require("./components/ReadParagraph"));

var _EditParagraph = _interopRequireDefault(require("./components/EditParagraph"));

var _styled = require("./components/styled");

var OldBambooPlugin = _interopRequireWildcard(require("./utils/bamboo"));

exports.OldBambooPlugin = OldBambooPlugin;

var _role = require("./utils/role");

const mobileZhiFactory = _index2.default;
exports.mobileZhiFactory = mobileZhiFactory;
const mobileBiFactory = _index3.default; // facotry

exports.mobileBiFactory = mobileBiFactory;
