// @ali/4ever-bi/dist-node/plugins/image/components/next/createImageButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createImageButton;

var React = _interopRequireWildcard(require("react"));

var _everComponent = require("@ali/4ever-component");

var _everBamboo = require("@ali/4ever-bamboo");

var _actions = require("../../actions");

var _imageButton = require("../imageButton");

const _createElement = /*#__PURE__*/React.createElement;

const handleClick = controller => {
  controller.run('onAction', (0, _actions.selectImage)());
};

var _ref = /*#__PURE__*/_createElement(_everComponent.IconPicture, null);

function createImageButton(props) {
  const {
    controller,
    locale,
    icon,
    title,
    tooltip,
    badgeDot,
    disabled
  } = props;
  const isDisabled = disabled || (0, _imageButton.isButtonDisabled)(controller);
  const buttonConfig = {
    icon: icon || _ref,
    onClick: handleClick,
    title: title || locale?.quickInsertTitle,
    tooltip: {
      title: title || locale?.quickInsertTitle,
      description: 'Markdown: ![]() ' + (locale?.['space'] || 'Space'),
      ...tooltip
    },
    badgeDot,
    disabled: isDisabled,
    role: _everBamboo.PluginRoles.image,
    testid: "bi-toolbar-picture"
  };
  return buttonConfig;
}
