// @ali/we-design/dist-web/components/color-panel/ColorMatrix/styled.js
import styled, { css } from 'styled-components';
import { isMobile } from "../../../utils/dom";
import { NO_COLOR } from "../constants";
export var MatrixWrapper = /*#__PURE__*/styled.div(["display:grid;grid-template-columns:repeat(", ",", ");grid-template-rows:", ";gap:4px 4px;", ""], function (_ref) {
  var column = _ref.column;
  return column;
}, isMobile ? '1fr' : '22px', function (_ref2) {
  var isFirstRowDoubleGap = _ref2.isFirstRowDoubleGap;
  return isFirstRowDoubleGap ? '26px 22px' : '22px';
}, isMobile ? "\n    flex: 1;\n    grid-template-rows: 32px;\n    @media only screen and (max-device-width: 388px) {\n      grid-template-rows: 30px;\n    }\n    @media only screen and (max-device-width: 370px) {\n      grid-template-rows: 26px;\n    }\n  " : ''); // 色块基础样式

export var BasicColorBox = /*#__PURE__*/styled.div(["display:flex;justify-content:center;align-items:center;position:relative;box-sizing:border-box;width:22px;height:22px;background:#ffffff;border:2px solid rgba(126,134,142,0.16);border-radius:4px;cursor:pointer;&:hover{border:2px solid #0089ff;}", ""], isMobile ? "\n    width: 32px;\n    height: 32px;\n    @media only screen and (max-device-width: 388px) {\n      width: 30px;\n      height: 30px;\n    }\n    @media only screen and (max-device-width: 370px) {\n      width: 26px;\n      height: 26px;\n    }\n  " : ''); // 字体颜色选择色块

export var FontColorBox = /*#__PURE__*/styled(BasicColorBox)(["", ""], function (_ref3) {
  var isSelected = _ref3.isSelected;
  return isSelected ? css(["border:2px solid rgba(23,26,29,0.6);"]) : undefined;
}); // 边框颜色选择色块

export var BorderColorBox = /*#__PURE__*/styled(BasicColorBox)(["", ";"], function (_ref4) {
  var color = _ref4.color;
  return color !== NO_COLOR && "border: 2px solid " + color;
}); // 背景颜色选择色块

export var BackgroundColorBox = /*#__PURE__*/styled(BasicColorBox)(["", ";background:", ";"], function (_ref5) {
  var color = _ref5.color;
  return color !== '#FFFFFF' && color !== NO_COLOR && 'border: none';
}, function (_ref6) {
  var color = _ref6.color;
  return color;
}); // 带 A 的背景颜色选择色块

export var FontBackgroundColorBox = /*#__PURE__*/styled(BasicColorBox)(["", ";background:", ";"], function (_ref7) {
  var color = _ref7.color;
  return color !== '#FFFFFF' && color !== NO_COLOR && 'border: none';
}, function (_ref8) {
  var color = _ref8.color;
  return color;
});
export var Container = /*#__PURE__*/styled.div(["position:relative;display:flex;justify-content:space-between;align-items:center;"]);
export var MoreIconWrapper = /*#__PURE__*/styled.div(["position:absolute;right:-12px;cursor:pointer;svg{fill:#171a1d;}:hover svg{fill:#0089ff;}"]);
