// @ali/4ever-zhi/dist-node/plugins/embed/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EnlargeIconHide = exports.PreviewErrorRetry = exports.PreviewErrorText = exports.PreviewError = exports.PreviewContentWrapper = exports.PreviewHeaderWrapper = exports.EmbedFilePreviewWrapper = exports.ContentDesc = exports.ContentHeader = exports.CardContentInnerWrapper = exports.CardContentWrapper = exports.CardIconWrapper = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

const CardIconWrapper = /*#__PURE__*/_styledComponents.default.div(["margin-right:16px;flex:0;img{vertical-align:initial;}"]);

exports.CardIconWrapper = CardIconWrapper;

const CardContentWrapper = /*#__PURE__*/_styledComponents.default.div(["flex:1;min-width:0;"]);

exports.CardContentWrapper = CardContentWrapper;

const CardContentInnerWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;"]);

exports.CardContentInnerWrapper = CardContentInnerWrapper;

const ContentHeader = /*#__PURE__*/_styledComponents.default.div(["font-size:14px;color:", ";flex:1;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;"], props => props.mode === 'dark' ? '#F2F4F5' : '#111F2C');

exports.ContentHeader = ContentHeader;

const ContentDesc = /*#__PURE__*/_styledComponents.default.div(["font-size:12px;color:", ";margin-right:16px;"], props => props.mode === 'dark' ? '#F2F4F5' : '#7A8189');

exports.ContentDesc = ContentDesc;

const EmbedFilePreviewWrapper = /*#__PURE__*/_styledComponents.default.div(["position:relative;width:100%;display:flex;flex-direction:column;border:1px solid #ECEDEF;border-radius:6px;transition:height .6s;box-sizing:border-box;"]);

exports.EmbedFilePreviewWrapper = EmbedFilePreviewWrapper;

const PreviewHeaderWrapper = /*#__PURE__*/_styledComponents.default.div(["width:100%;background-color:", ";padding:4px 16px 4px 16px;height:32px;display:flex;border-top-left-radius:6px;border-top-right-radius:6px;line-height:23px;box-sizing:border-box;"], props => props.mode === 'dark' ? '#111F2C' : '#F6F8F9');

exports.PreviewHeaderWrapper = PreviewHeaderWrapper;

const PreviewContentWrapper = /*#__PURE__*/_styledComponents.default.div(["overflow:auto;flex:1;position:relative;justify-content:center;align-items:center;display:flex;background-color:#EDEDED;border-bottom-left-radius:6px;border-bottom-right-radius:6px;"]);

exports.PreviewContentWrapper = PreviewContentWrapper;

const PreviewError = /*#__PURE__*/_styledComponents.default.div(["text-align:center;display:none;"]);

exports.PreviewError = PreviewError;

const PreviewErrorText = /*#__PURE__*/_styledComponents.default.div(["font-size:13px;color:#414c56;display:inline-block;"]);

exports.PreviewErrorText = PreviewErrorText;

const PreviewErrorRetry = /*#__PURE__*/_styledComponents.default.span(["color:#3A454F;font-weight:bold;cursor:pointer;"]);

exports.PreviewErrorRetry = PreviewErrorRetry;

const EnlargeIconHide = /*#__PURE__*/_styledComponents.default.div(["position:absolute;bottom:0;right:0;border-bottom-right-radius:6px;width:100px;height:40px;background-color:#f3f3f3;z-index:3;display:none;transform:translateZ(0);"]);

exports.EnlargeIconHide = EnlargeIconHide;
