// @ali/4ever-bi/dist-web/plugins/alignment/index.js
import toolbar from "./toolbar";
import menubar from "./menubar";
import setAlignment from "./commands/setAlignment";
import getAlignment from "./queries/getAlignment";
import { DEPRECATED_ACTIONS, actions } from "./actions";
import onAction from "./handlers/onAction";
import onKeyDown from "./handlers/onKeyDown";
import sidebar from "./sidebar";
import createRenderHoverToolbar from "./createRenderHoverToolbar";
import createRenderToolbar from "./createRenderToolbar";
var commands = {
  setAlignment: setAlignment
};
var queries = {
  getAlignment: getAlignment
};

function createAlignmentPlugin(config) {
  var alignmentConfig = config.alignment || {};
  return {
    toolbar: toolbar(alignmentConfig),
    renderToolbar: createRenderToolbar(alignmentConfig.locale),
    sidebar: sidebar(config),
    menubar: menubar(alignmentConfig),
    commands: commands,
    queries: queries,
    actions: actions,
    onAction: onAction,
    onKeyDown: onKeyDown,
    models: [],
    renderHoverToolbar: createRenderHoverToolbar(alignmentConfig)
  };
}

createAlignmentPlugin.actions = DEPRECATED_ACTIONS;
createAlignmentPlugin.commands = commands;
createAlignmentPlugin.queries = queries;
export default createAlignmentPlugin;
