// @ali/4ever-plugin-subject/dist-node/utils/handlers/createGroupNodes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

var _utils = require("../utils");

function createGroupNodes() {
  return function groupNodes([start, end], ctx, controller, next) {
    const {
      parent,
      groups
    } = ctx;
    const first = parent.nodes[start];
    const parentGroup = groups[groups.length - 1]; // 如果不是一级节点，或者是文档末尾，则不分组

    if (!_everCangjie.Document.isDocument(parent)) {
      return next();
    }

    if ((0, _utils.isSubjectHeading)(first) && !(0, _utils.isSubject)(parentGroup)) {
      let i = start + 1;

      while (i < end) {
        const node = parent.nodes[i];

        if ((0, _utils.isSubjectHeading)(node)) {
          break;
        }

        i += 1;
      }

      const children = parent.nodes.slice(start, i);
      const subject = (0, _utils.createSubject)({
        key: `${first.key}-group`,
        data: first.data,
        nodes: children
      });
      return {
        node: subject,
        remainedRange: [i, end]
      };
    }

    return next();
  };
}

var _default = createGroupNodes;
exports.default = _default;
