// @ali/4ever-plugin-table/dist-node/bi/components/colToolbar/tableColToolbarItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TableColToolbarItem = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _hooks = require("../../../utils/hooks");

var _constants = require("../../constants");

var _rowColToolbarItemRect = require("../rowColToolbarItemRect");

const _createElement = /*#__PURE__*/React.createElement;

const TableColToolbarItem = props => {
  const {
    index,
    style: propsStyle,
    scale = 1,
    activeRange,
    selectRanges,
    deleteRange,
    children,
    onRectUpdate,
    ...rest
  } = props;
  const ref = React.useRef(null);
  const [colsWidth] = (0, _hooks.usePixelColsWidth)();
  React.useEffect(() => {
    const dom = ref.current;
    onRectUpdate?.(index, dom);
    return () => {
      onRectUpdate?.(index, dom, true);
    };
  });
  const width = colsWidth[index];
  const style = {
    width: index === 0 ? width + 1 : width,
    height: _constants.TOOLBAR_ITEM_SIZE / scale,
    borderTopRightRadius: index === colsWidth.length - 1 ? 3 / scale : '',
    float: 'left',
    ...propsStyle
  };
  return /*#__PURE__*/_createElement(_rowColToolbarItemRect.RowColToolbarItemRect, (0, _extends2.default)({
    ref: ref,
    index: index,
    type: "col",
    length: colsWidth.length,
    style: style,
    activeRange: activeRange,
    selectRanges: selectRanges,
    deleteRange: deleteRange
  }, rest), children);
};

exports.TableColToolbarItem = TableColToolbarItem;
