// @ali/4ever-plugin-comment/dist-web/utils/components/simplePopup/textInput.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Controller } from '@ali/4ever-cangjie';
var buttonDefaultStyle = {
  background: '#2364E8',
  border: 'none',
  color: 'white',
  borderRadius: '4px',
  padding: '4px 8px'
};
var inputDiv = {
  marginTop: '8px',
  textAlign: 'right'
};
export default function TextInput(props) {
  var contentId = props.contentId,
      sendComment = props.sendComment,
      onAdd = props.onAdd;
  var controller = Controller.useController();
  var textRef = React.useRef(null);

  var _React$useState = React.useState(false),
      isSending = _React$useState[0],
      setIsSending = _React$useState[1];

  var buttonStyle = _extends({}, buttonDefaultStyle, {
    opacity: isSending ? 0.5 : 1
  });

  var send = function send() {
    var _textRef$current;

    var content = (_textRef$current = textRef.current) == null ? void 0 : _textRef$current.value;

    if (content) {
      if (!sendComment) {
        throw Error('`Comment` plugin should specify sendComment callback');
      }

      setIsSending(true);
      sendComment({
        content: content,
        contentId: contentId
      }).then(function () {
        onAdd();

        if (textRef.current) {
          textRef.current.value = '';
        }

        setIsSending(false);
      })["catch"](function (error) {
        setIsSending(false);
      });
    }
  };

  var handleKeyDown = function handleKeyDown(event) {
    if (event.keyCode === 13 && event.metaKey) {
      event.preventDefault();
      send();
    }
  };

  return /*#__PURE__*/_createElement("div", null, /*#__PURE__*/_createElement("textarea", {
    "data-testid": "comment-input",
    ref: textRef,
    style: {
      width: '100%'
    },
    autoFocus: true,
    onKeyDown: handleKeyDown,
    placeholder: "\u6309 enter \u53D1\u9001\uFF0Cshift+enter \u6362\u884C"
  }), /*#__PURE__*/_createElement("div", {
    style: inputDiv
  }, /*#__PURE__*/_createElement("button", {
    "data-testid": "comment-send",
    onClick: send,
    style: buttonStyle
  }, isSending ? '发送中' : '评论')));
}
