// @ali/4ever-bi/dist-node/plugins/newlineGuide/components/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AddIcon = exports.PrimaryText = exports.TooltipText = exports.NewlineInsert = exports.InsertBtnWrapper = exports.insertBtnStyle = exports.GuideMenuWrapper = exports.HolderInsertText = exports.HolderPlus = exports.HolderText = exports.GuideWrapper = void 0;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

const _createElement = /*#__PURE__*/React.createElement;

const GuideWrapper = /*#__PURE__*/_styledComponents.default.div(["position:absolute;z-index:100;line-height:initial;height:21px;line-height:22px;font-size:11pt;transform:translateY(-50%);color:rgba(0,0,0,0.14);user-select:none;"]);

exports.GuideWrapper = GuideWrapper;

const HolderText = /*#__PURE__*/_styledComponents.default.span(["padding-left:3px;"]);

exports.HolderText = HolderText;

const HolderPlus = /*#__PURE__*/_styledComponents.default.span(["margin:0 2px;display:inline-block;"]);

exports.HolderPlus = HolderPlus;

const HolderInsertText = /*#__PURE__*/_styledComponents.default.span(["text-decoration:underline;text-underline-position:under;cursor:pointer;margin-left:2px;pointer-events:auto;color:rgba(23,26,29,0.4);&:hover{color:#317ED0;}"]);

exports.HolderInsertText = HolderInsertText;

const GuideMenuWrapper = /*#__PURE__*/_styledComponents.default.div(["width:206px;max-height:360px;overflow-y:auto;"]);

exports.GuideMenuWrapper = GuideMenuWrapper;
const insertBtnStyle = {
  width: 19,
  height: 19,
  offset: 4
};
exports.insertBtnStyle = insertBtnStyle;

const InsertBtnWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;align-items:center;justify-content:center;width:19px;height:19px;border-radius:50%;background-color:#fff;cursor:pointer;border:1px solid rgba(126,134,142,0.24);&:hover{background-color:rgba(50,96,201,0.06);}&:active{background-color:rgba(50,96,201,0.09);}"]);

exports.InsertBtnWrapper = InsertBtnWrapper;

const NewlineInsert = /*#__PURE__*/_styledComponents.default.div(["position:absolute;"]);

exports.NewlineInsert = NewlineInsert;

const TooltipText = /*#__PURE__*/_styledComponents.default.div(["color:rgba(255,255,255,.56);display:flex;"]);

exports.TooltipText = TooltipText;

const PrimaryText = /*#__PURE__*/_styledComponents.default.div(["margin-right:4px;color:#fff;"]);

exports.PrimaryText = PrimaryText;

const AddIcon = /*#__PURE__*/_createElement("svg", {
  width: "9",
  height: "9",
  viewBox: "0 0 9 9",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/_createElement("path", {
  d: "M5 4V0H4V4H0V5H4V9H5V5H9V4H5Z",
  fill: "#171A1D",
  fillOpacity: "0.6"
}));

exports.AddIcon = AddIcon;
