// @ali/4ever-utils/dist-web/utils/customEventUtils.js
/**
 * 由于文字中工具栏相关的组件比较多，比如行工具栏，左侧工具栏，部分插件还有自定义工具栏
 * 所以这里抽离出一个公用状态管理的工具，便于组件间共享一些状态
 */
export var LEGAL_EVENT = {
  leftToolbarEnable: 'leftToolbarEnable',
  // 浮动工具栏显示到时候需要隐藏掉左侧工具栏按钮, 表格插入行按钮展示的时候也需要将其隐藏掉
  hoverMenuVisible: 'hoverMenuVisible',
  contextMenuVisible: 'contextMenuVisible',
  // 右键菜单显示的时候，浮动工具栏不能显示
  // 加号或斜杠唤起的 快捷插入菜单显示 or 隐藏
  quickInsertMenuVisible: 'quickInsertMenuVisible'
};
export var getCustomEvent = function getCustomEvent(eventName, data, bubbles) {
  if (bubbles === void 0) {
    bubbles = true;
  }

  var customeEvent = new CustomEvent(eventName, {
    bubbles: bubbles,
    detail: data
  });
  return customeEvent;
};
export var dispatchCustomEvent = function dispatchCustomEvent(eventName, data, source, bubbles) {
  if (bubbles === void 0) {
    bubbles = true;
  }

  var customeEvent = getCustomEvent(eventName, data, bubbles);

  if (source && source != null && source.dispatchEvent) {
    source == null ? void 0 : source.dispatchEvent(customeEvent);
  }

  document.body.dispatchEvent(customeEvent);
};
