// @ali/4ever-plugin-refblock/dist-web/bi/commands/upgradeToRefBlock.js
import { Commands, EdgePoint, Selection } from '@ali/4ever-cangjie';
import { commands } from "../../common/commands";
import RefBlock, { RefBlockStatus } from "../../mo/models";
/**
 * 将选中的 blocks 升级为 refblock
 * @param controller
 * @param blocks
 * @param docKey
 * @returns
 */

export function upgradeToRefBlock(controller, blocks, data) {
  if (!blocks.length) {
    return controller;
  }

  return controller.withoutNormalizing(function () {
    // 先删除再插入
    blocks.slice(1).forEach(function (block) {
      controller.command(Commands.removeNodeByKey, block.key);
    }); // 替换第一个

    var refblock = RefBlock.create({
      data: data,
      nodes: blocks.map(function (block) {
        return block.regenerateKey();
      })
    });
    var point = EdgePoint.create({
      key: refblock.key,
      edge: EdgePoint.AFTER
    });
    var selection = Selection.create({
      anchor: point,
      focus: point
    });
    controller.command(Commands.replaceNodeByKey, blocks[0].key, refblock).command(Commands.select, selection).command(commands.setRefBlockInjectionByDocKey, data.docKey, {
      status: RefBlockStatus.loading
    }).command(Commands.focus);
  });
}
