// @ali/we-ui/dist-web/components/list-select/ListSelect.js
import React from 'react'; // react-window 包体积远小于 react-virtualized

var _createElement = /*#__PURE__*/React.createElement;
import { FixedSizeList } from 'react-window';
import memoize from 'memoize-one';
import { CheckboxGroup } from "../checkbox";
import { ListSelectWrap } from "./styled";
import { ListSelectItem } from "./ListSelectItem";
var createItemData = memoize(function (dataList, itemStyle) {
  return {
    dataList: dataList,
    itemStyle: itemStyle
  };
});
export var ListSelect = function ListSelect(props) {
  var dataList = props.dataList,
      onChange = props.onChange,
      width = props.width,
      height = props.height,
      _props$value = props.value,
      value = _props$value === void 0 ? [] : _props$value,
      _props$style = props.style,
      style = _props$style === void 0 ? {} : _props$style,
      itemHeight = props.itemHeight,
      testid = props.testid,
      itemStyle = props.itemStyle;
  var itemData = createItemData(dataList, itemStyle);
  return /*#__PURE__*/_createElement(ListSelectWrap, {
    style: style
  }, /*#__PURE__*/_createElement(CheckboxGroup, {
    onChange: onChange,
    value: value,
    testid: testid
  }, /*#__PURE__*/_createElement(FixedSizeList, {
    height: height,
    itemCount: dataList.length,
    itemData: itemData,
    itemSize: itemHeight,
    width: width,
    itemKey: function itemKey(index, data) {
      var item = data[index];
      return "FixedSizeList_" + ((item == null ? void 0 : item.value) || index);
    }
  }, ListSelectItem)));
};
ListSelect.defaultProps = {
  itemHeight: 20,
  value: []
};
