// @ali/4ever-component/dist-web/ToolbarNext/basicComponents/buttons/AlignButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { IconButton } from "./Basics";
import { IconAlignLeft, IconAlignCenter, IconAlignJustify, IconAlignRight } from "../icons";

var _ref = /*#__PURE__*/_createElement(IconAlignLeft, null);

export function AlignLeftButton(props) {
  var _props$icon = props.icon,
      icon = _props$icon === void 0 ? _ref : _props$icon,
      rest = _objectWithoutPropertiesLoose(props, ["icon"]);

  return /*#__PURE__*/_createElement(IconButton, _extends({
    testid: "align_left",
    icon: icon
  }, rest));
}

var _ref2 = /*#__PURE__*/_createElement(IconAlignRight, null);

export function AlignRightButton(props) {
  var _props$icon2 = props.icon,
      icon = _props$icon2 === void 0 ? _ref2 : _props$icon2,
      rest = _objectWithoutPropertiesLoose(props, ["icon"]);

  return /*#__PURE__*/_createElement(IconButton, _extends({
    testid: "align_right",
    icon: icon
  }, rest));
}

var _ref3 = /*#__PURE__*/_createElement(IconAlignCenter, null);

export function AlignCenterButton(props) {
  var _props$icon3 = props.icon,
      icon = _props$icon3 === void 0 ? _ref3 : _props$icon3,
      rest = _objectWithoutPropertiesLoose(props, ["icon"]);

  return /*#__PURE__*/_createElement(IconButton, _extends({
    testid: "align_center",
    icon: icon
  }, rest));
}

var _ref4 = /*#__PURE__*/_createElement(IconAlignJustify, null);

export function AlignJustifyButton(props) {
  var _props$icon4 = props.icon,
      icon = _props$icon4 === void 0 ? _ref4 : _props$icon4,
      rest = _objectWithoutPropertiesLoose(props, ["icon"]);

  return /*#__PURE__*/_createElement(IconButton, _extends({
    testid: "align_justify",
    icon: icon
  }, rest));
}
