// @ali/4ever-component/dist-web/ToolbarNext/basicComponents/buttons/HorizontalAlignSelect.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React, { useMemo, useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { IconAlignCenter, IconAlignLeft, IconAlignRight, IconAlignJustify } from "../icons";
import { IconSelect, IconSelectSingleLineDefaultWidth } from "./Basics";

var _ref = /*#__PURE__*/_createElement(IconAlignLeft, null);

var _ref2 = /*#__PURE__*/_createElement(IconAlignCenter, null);

var _ref3 = /*#__PURE__*/_createElement(IconAlignRight, null);

var _ref4 = /*#__PURE__*/_createElement(IconAlignJustify, null);

export var HorizontalAlignSelect = function HorizontalAlignSelect(props) {
  var onChange = props.onChange,
      options = props.options,
      role = props.role,
      value = props.value,
      tooltip = props.tooltip,
      rest = _objectWithoutPropertiesLoose(props, ["onChange", "options", "role", "value", "tooltip"]); // 垂直对齐的选项


  var normalizeOptions = useMemo(function () {
    return [{
      title: options.left.title,
      value: 'left',
      key: 'left',
      icon: options.left.icon || _ref,
      testid: 'option-left'
    }, {
      title: options.center.title,
      value: 'center',
      key: 'center',
      icon: options.center.icon || _ref2,
      testid: 'option-center'
    }, {
      title: options.right.title,
      value: 'right',
      key: 'right',
      icon: options.right.icon || _ref3,
      testid: 'option-right'
    }, {
      title: options.justify.title,
      value: 'justify',
      key: 'justify',
      icon: options.justify.icon || _ref4,
      testid: 'option-justify'
    }];
  }, [options]);
  var handleChange = useCallback(function (e) {
    if (onChange) {
      onChange(e);
    }
  }, [onChange]);
  return /*#__PURE__*/_createElement(IconSelect, _extends({
    value: value,
    onChange: handleChange,
    testid: "HorizontalAlignSelect",
    tooltip: tooltip,
    role: role,
    options: normalizeOptions,
    width: IconSelectSingleLineDefaultWidth
  }, rest));
};
