// @ali/4ever-component/dist-web/ToolbarNext/basicComponents/buttons/ListSet.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React, { useMemo, useState, useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled from 'styled-components';
import { SplitButton } from "./Basics";
import { IconOrderList, IconUnorderList } from "../icons";
import classnames from 'classnames';
var DEFAULT_BULLET_LIST_ARRAY = ['SCIR_ECIR_SREC', 'SFDI_SRIG_SREC', 'EREC_EREC_EREC', 'SARR_SDIO_SCIR', 'SSTA_ECIR_SREC', 'SRIG_ECIR_SREC'];
var DEFAULT_STICKER_LIST_ARRAY = ['VACATION_TEA_MILKTEA', 'RIGHT_NAIL_LOVE', 'FIRE_ROCKET_NAIL'];
var DEFAULT_NUMBER_LIST_ARRAY = ['DEC_LEN_LROM_P', 'DEC_LEN_LROM_H', 'DEC_DEC_DEC_P', 'UEN_LEN_LROM_P', 'UROM_UEN_DEC_P', 'DDEC_LEN_LROM_P', 'ZH_DEC_LEN_D', 'ZH_DEC_LEN_A', 'ZH_ZH_DEC_D'];
var LIST_STYLE_MAP = {
  SCIR_ECIR_SREC: 'https://img.alicdn.com/tfs/TB1mWdYbAcx_u4jSZFlXXXnUFXa-78-78.svg',
  SFDI_SRIG_SREC: 'https://img.alicdn.com/tfs/TB1hpepOpP7gK0jSZFjXXc5aXXa-78-78.svg',
  EREC_EREC_EREC: 'https://img.alicdn.com/tfs/TB1cJR4czMZ7e4jSZFOXXX7epXa-78-78.svg',
  SARR_SDIO_SCIR: 'https://img.alicdn.com/tfs/TB1aaWcOEY1gK0jSZFMXXaWcVXa-78-78.svg',
  SSTA_ECIR_SREC: 'https://img.alicdn.com/tfs/TB1VoX8OxD1gK0jSZFsXXbldVXa-78-78.svg',
  SRIG_ECIR_SREC: 'https://img.alicdn.com/tfs/TB1cqp4OAL0gK0jSZFtXXXQCXXa-78-78.svg',
  VACATION_TEA_MILKTEA: 'https://img.alicdn.com/imgextra/i4/O1CN01iGy3OC23bXk0A5V99_!!6000000007274-2-tps-240-240.png',
  RIGHT_NAIL_LOVE: 'https://img.alicdn.com/imgextra/i4/O1CN01rYMRza1OM46kvi2Yx_!!6000000001690-2-tps-240-240.png',
  FIRE_ROCKET_NAIL: 'https://img.alicdn.com/imgextra/i3/O1CN01QfDOFq1R49SyPKcpW_!!6000000002057-2-tps-240-240.png',
  DEC_LEN_LROM_P: 'https://img.alicdn.com/tfs/TB19lt9OEz1gK0jSZLeXXb9kVXa-78-78.svg',
  DEC_LEN_LROM_H: 'https://img.alicdn.com/tfs/TB18HydOAL0gK0jSZFAXXcA9pXa-78-78.svg',
  DEC_DEC_DEC_P: 'https://img.alicdn.com/tfs/TB1gHycOqL7gK0jSZFBXXXZZpXa-78-78.svg',
  UEN_LEN_LROM_P: 'https://img.alicdn.com/tfs/TB1VaOnOBr0gK0jSZFnXXbRRXXa-78-78.svg',
  UROM_UEN_DEC_P: 'https://img.alicdn.com/tfs/TB1xcqhOAY2gK0jSZFgXXc5OFXa-78-78.svg',
  DDEC_LEN_LROM_P: 'https://img.alicdn.com/tfs/TB1yjive8Bh1e4jSZFhXXcC9VXa-78-78.svg',
  ZH_DEC_LEN_D: 'https://img.alicdn.com/tfs/TB1bFQZ0aL7gK0jSZFBXXXZZpXa-82-82.svg',
  ZH_DEC_LEN_A: 'https://img.alicdn.com/tfs/TB11.iDj4vbeK8jSZPfXXariXXa-82-82.svg',
  ZH_ZH_DEC_D: 'https://img.alicdn.com/imgextra/i4/O1CN01XYORsY1OjQIxXuZxz_!!6000000001741-55-tps-82-82.svg'
};
var OverlayContent = /*#__PURE__*/styled.div(["width:272px;display:flex;flex-wrap:wrap;justify-content:flex-start;padding:4px;"]);
var ListStyleItemWrap = /*#__PURE__*/styled.div(["user-select:none;"]);
var ListStyleItem = /*#__PURE__*/styled.img(["width:80px;height:80px;margin:4px;border:1px solid rgba(17,31,44,0.12);border-radius:2px;cursor:pointer;image-rendering:pixelated;&:hover{border:1px solid #0089FF;}&.actived{border:2px solid #0089FF;}"]);

function getListRole(mode, key) {
  var prefix = mode === 'ul' ? 'listUnordered' : 'listOrdered';
  return prefix + "_" + key;
}

function StopEvent(event) {
  event.stopPropagation();
}

var _ref = /*#__PURE__*/_createElement(IconUnorderList, null);

var _ref2 = /*#__PURE__*/_createElement(IconOrderList, null);

export var ListSet = function ListSet(props) {
  var _props$mode = props.mode,
      mode = _props$mode === void 0 ? 'ul' : _props$mode,
      defaultListStyle = props.defaultListStyle,
      onChange = props.onChange,
      onClick = props.onClick,
      disabled = props.disabled,
      icon = props.icon,
      enableSticker = props.enableSticker,
      active = props.active,
      rest = _objectWithoutPropertiesLoose(props, ["mode", "defaultListStyle", "onChange", "onClick", "disabled", "icon", "enableSticker", "active"]);

  var listStyleArray = useMemo(function () {
    if (mode !== 'ul') {
      return DEFAULT_NUMBER_LIST_ARRAY;
    } else {
      return enableSticker ? DEFAULT_BULLET_LIST_ARRAY.concat(DEFAULT_STICKER_LIST_ARRAY) : DEFAULT_BULLET_LIST_ARRAY;
    }
  }, [mode]);

  var _useState = useState(defaultListStyle || listStyleArray[0]),
      value = _useState[0],
      setValue = _useState[1];

  var _useState2 = useState(false),
      overlayVisible = _useState2[0],
      setOverlayVisible = _useState2[1];

  var handleChange = useCallback(function (e) {
    if (onChange) {
      onChange(e);
    }

    setValue(e);
    setOverlayVisible(false);
  }, [onChange, setValue]);
  var handleVisibleChange = useCallback(function (visible) {
    setOverlayVisible(visible);
  }, []);
  var handleClickLeft = useCallback(function () {
    if (onClick) {
      onClick(value);
    }
  }, [onClick, value]);
  var getOverlay = useMemo(function () {
    return /*#__PURE__*/_createElement(OverlayContent, {
      "data-testid": "overlay-ListSet",
      onClick: StopEvent
    }, listStyleArray.map(function (item) {
      return /*#__PURE__*/_createElement(ListStyleItemWrap, {
        key: item,
        onClick: function onClick() {
          return handleChange(item);
        },
        "data-role": getListRole(mode, item),
        role: "button"
      }, /*#__PURE__*/_createElement(ListStyleItem, {
        src: LIST_STYLE_MAP[item],
        alt: "list-style-item",
        className: classnames({
          actived: item === value
        }),
        "data-testid": "option-" + item
      }));
    }));
  }, [value]);
  var renderTriggerIcon = useMemo(function () {
    if (icon) {
      return icon;
    }

    return mode === 'ul' ? _ref : _ref2;
  }, [mode, icon]);
  return /*#__PURE__*/_createElement(SplitButton, _extends({
    overlay: getOverlay,
    disabled: disabled,
    onClick: handleClickLeft,
    overlayVisible: overlayVisible,
    onVisibleChange: handleVisibleChange,
    icon: renderTriggerIcon,
    active: active || overlayVisible
  }, rest));
};
