// @ali/4ever-component/dist-web/ToolbarNext/basicComponents/buttons/ScriptSplitButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { SplitButton, List, Tooltip } from "./Basics";
import { IconSubscript, IconSuperscript, IconSelected } from "../icons";

var _ref = /*#__PURE__*/_createElement(IconSubscript, null);

var _ref2 = /*#__PURE__*/_createElement(IconSuperscript, null);

var _ref4 = /*#__PURE__*/_createElement(IconSuperscript, null);

var _ref5 = /*#__PURE__*/_createElement(IconSubscript, null);

var _ref6 = /*#__PURE__*/_createElement(IconSelected, null);

export var ScriptSplit = function ScriptSplit(props) {
  var onVisibleChange = props.onVisibleChange,
      curScript = props.curScript,
      onScriptChange = props.onScriptChange,
      disabled = props.disabled,
      testid = props.testid,
      _props$subItem = props.subItem,
      subItem = _props$subItem === void 0 ? {} : _props$subItem,
      _props$supItem = props.supItem,
      supItem = _props$supItem === void 0 ? {} : _props$supItem,
      icon = props.icon,
      rest = _objectWithoutPropertiesLoose(props, ["onVisibleChange", "curScript", "onScriptChange", "disabled", "testid", "subItem", "supItem", "icon"]);

  var _React$useState = React.useState(false),
      overlayVisible = _React$useState[0],
      setOverlayVisible = _React$useState[1];

  var handleOverlayVisible = React.useCallback(function (v) {
    setOverlayVisible(v);
    onVisibleChange && onVisibleChange(v);
  }, [onVisibleChange]);
  var splitButtonIcon = icon || (curScript === 'sub' ? _ref : _ref2);
  var handleClick = React.useCallback(function () {
    onScriptChange(curScript || 'super');
    handleOverlayVisible(false);
  }, [curScript, handleOverlayVisible]);
  var items = [_extends({
    key: 'super'
  }, supItem), _extends({
    key: 'sub'
  }, subItem)].map(function (_ref3) {
    var key = _ref3.key,
        title = _ref3.title,
        shortCut = _ref3.shortCut,
        description = _ref3.description;
    return {
      key: key,
      content: /*#__PURE__*/_createElement(Tooltip, {
        placement: "right",
        title: title + "(" + shortCut + ")",
        description: description
      }, /*#__PURE__*/_createElement(List.Item, {
        key: key,
        prefix: key === 'super' ? _ref4 : _ref5,
        suffix: curScript === key ? _ref6 : null,
        onClick: function onClick(e) {
          onScriptChange(key);
          handleOverlayVisible(false);
        }
      }, title))
    };
  });
  var tooltip = curScript === 'sub' ? subItem : supItem;
  return /*#__PURE__*/_createElement(SplitButton, _extends({
    onClick: handleClick,
    overlayVisible: overlayVisible,
    onVisibleChange: handleOverlayVisible,
    icon: splitButtonIcon,
    disabled: disabled,
    overlay: /*#__PURE__*/_createElement(List, {
      items: items
    }),
    autoClose: false,
    testid: testid,
    active: !!curScript,
    tooltip: tooltip
  }, rest));
};
