// @ali/4ever-component/dist-web/ToolbarNext/basicComponents/buttons/TableThemeButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { ColorPanel } from '@ali/we-design';
import { IconPalette } from "../icons";
import { IconMenu } from "./Basics";

var _ref = /*#__PURE__*/_createElement(IconPalette, null);

export var TableThemeButton = function TableThemeButton(props) {
  var tableBgColor = props.tableBgColor,
      _props$defaultTableBg = props.defaultTableBgColor,
      defaultTableBgColor = _props$defaultTableBg === void 0 ? '' : _props$defaultTableBg,
      onTableBgColorChange = props.onTableBgColorChange,
      tableBorderColor = props.tableBorderColor,
      _props$defaultTableBo = props.defaultTableBorderColor,
      defaultTableBorderColor = _props$defaultTableBo === void 0 ? '' : _props$defaultTableBo,
      onTableBorderColorChange = props.onTableBorderColorChange,
      restProps = _objectWithoutPropertiesLoose(props, ["tableBgColor", "defaultTableBgColor", "onTableBgColorChange", "tableBorderColor", "defaultTableBorderColor", "onTableBorderColorChange"]);

  var _React$useState = React.useState(false),
      overlayVisible = _React$useState[0],
      setOverlayVisible = _React$useState[1];

  var handleReset = function handleReset() {
    onTableBgColorChange(defaultTableBgColor);
    onTableBorderColorChange(defaultTableBorderColor);
    setOverlayVisible(false);
    return false;
  };

  return /*#__PURE__*/_createElement(IconMenu, _extends({
    icon: _ref,
    overlay: /*#__PURE__*/_createElement(ColorPanel, {
      backgroundColor: tableBgColor,
      onBackgroundColorClick: onTableBgColorChange,
      borderColor: tableBorderColor,
      onBorderColorClick: onTableBorderColorChange,
      enableResetButton: true,
      onResetButtonClick: handleReset
    }),
    overlayVisible: overlayVisible,
    onVisibleChange: setOverlayVisible,
    testid: "bi-toolbar-table-theme"
  }, restProps));
};
