// @ali/4ever-component/dist-web/ToolbarNext/basicComponents/buttons/VerticalAlignSelect.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React, { useMemo, useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { IconVerticalAlignBottom, IconVerticalAlignMiddle, IconVerticalAlignTop } from "../icons";
import { IconSelect } from "./Basics";

var _ref = /*#__PURE__*/_createElement(IconVerticalAlignTop, null);

var _ref2 = /*#__PURE__*/_createElement(IconVerticalAlignMiddle, null);

var _ref3 = /*#__PURE__*/_createElement(IconVerticalAlignBottom, null);

export var VerticalAlignSelect = function VerticalAlignSelect(props) {
  var onChange = props.onChange,
      options = props.options,
      value = props.value,
      rest = _objectWithoutPropertiesLoose(props, ["onChange", "options", "value"]); // 垂直对齐的选项


  var normalizeOptions = useMemo(function () {
    return [{
      title: options.top.title,
      value: 'top',
      key: 'top',
      icon: options.top.icon || _ref,
      testid: 'option-top'
    }, {
      title: options.middle.title,
      value: 'middle',
      key: 'middle',
      icon: options.middle.icon || _ref2,
      testid: 'option-middle'
    }, {
      title: options.bottom.title,
      value: 'bottom',
      key: 'bottom',
      icon: options.bottom.icon || _ref3,
      testid: 'option-bottom'
    }];
  }, [options]);
  var handleChange = useCallback(function (e) {
    if (onChange) {
      onChange(e);
    }
  }, [onChange]);
  return /*#__PURE__*/_createElement(IconSelect, _extends({
    value: value,
    onChange: handleChange,
    testid: "VerticalAlignSelect",
    options: normalizeOptions // width={IconSelectSingleLineDefaultWidth}

  }, rest));
};
