// @ali/4ever-factory/dist-node/tagFactory/createRenderElement.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderElement;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _everLogger = _interopRequireDefault(require("@ali/4ever-logger"));

var _everPluginTag = require("@ali/4ever-plugin-tag");

var _everComponent = require("@ali/4ever-component");

var _Container = _interopRequireDefault(require("./components/Container"));

var _tagStore = require("./utils/tagStore");

const _createElement = /*#__PURE__*/React.createElement;

const ChildrenComponent = ({
  childProps,
  controller,
  plugin,
  config
}) => {
  const {
    locale = {},
    renderNode,
    verticalAlign
  } = plugin;
  const {
    node
  } = childProps;
  const isOwner = (0, _tagStore.isOwnerAndFirstCreated)(node.key);
  React.useEffect(() => {
    if (isOwner) (0, _tagStore.setOwnerAndFirstCreated)(node.key, false);
  }, [node.key, isOwner]); // @ts-ignore

  const finalVerticalAlign = typeof verticalAlign === 'function' ? verticalAlign(node) : verticalAlign;
  return /*#__PURE__*/_createElement(_Container.default, (0, _extends2.default)({}, childProps, {
    locale: locale,
    verticalAlign: finalVerticalAlign,
    config: config
  }), renderNode({ ...childProps,
    isOwnerAndFirstCreated: isOwner
  }, controller));
};

function createRenderElement(plugin, config) {
  const {
    tagType: tagTypeConfig
  } = plugin;
  return {
    [_everPluginTag.Tag.TYPE]: (props, controller, next) => {
      const {
        node
      } = props;
      const {
        tagType
      } = node.data;

      if (tagTypeConfig !== tagType) {
        return next();
      }

      const errorTitle = `BI:TAG:${tagType}`;

      const fallback = /*#__PURE__*/_createElement(_everComponent.InlinePlaceholder, (0, _extends2.default)({}, props, {
        renderText: plugin.renderText
      }));

      try {
        return /*#__PURE__*/_createElement(_everComponent.ErrorBoundary, {
          errorTitle: errorTitle,
          fallback: fallback
        }, /*#__PURE__*/_createElement(ChildrenComponent, {
          childProps: props,
          controller: controller,
          plugin: plugin,
          config: config
        }));
      } catch (error) {
        // ErrorBoundary 只能捕获子组件错误，这里对组件本身加一层兜底处理
        _everLogger.default.error(errorTitle, error);

        return fallback;
      }
    }
  };
}
