// @ali/4ever-plugin-comment/dist-node/utils/hooks/contextMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _role = require("../role");

var _actions = require("../actions");

var _canComment = _interopRequireDefault(require("../queries/canComment"));

var _default = configs => {
  return function contextMenu(controller, next) {
    if (!(0, _canComment.default)(controller)) {
      return next();
    }

    const isDisabled = configs?.isDisabled && configs.isDisabled();
    const title = configs?.locale?.comment || 'Comment';
    const item = {
      name: title,
      action: {
        type: _actions.ACTION_COMMENT_CLICK
      },
      key: 'comment',
      icon: 'https://gw.alicdn.com/imgextra/i2/O1CN01GtHgGq1jGfO1gqxmG_!!6000000004521-55-tps-24-24.svg',
      options: {
        group: 100,
        mobile: true,
        hotKey: controller.readOnly ? '' : 'mod+shift+m',
        disable: isDisabled
      },
      role: _role.CommentRole.comment
    };
    return [item, ...next()];
  };
};

exports.default = _default;
