// @ali/4ever-plugin-comment/dist-node/utils/hooks/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createContextMenu", {
  enumerable: true,
  get: function () {
    return _contextMenu.default;
  }
});
Object.defineProperty(exports, "createOnKeyDown", {
  enumerable: true,
  get: function () {
    return _onKeyDown.default;
  }
});
Object.defineProperty(exports, "createOnMouseMove", {
  enumerable: true,
  get: function () {
    return _onMouseMove.default;
  }
});
Object.defineProperty(exports, "createOnCangjieInput", {
  enumerable: true,
  get: function () {
    return _onCangjieInput.default;
  }
});
Object.defineProperty(exports, "createOnAction", {
  enumerable: true,
  get: function () {
    return _onAction.default;
  }
});
Object.defineProperty(exports, "createOnChange", {
  enumerable: true,
  get: function () {
    return _onChange.default;
  }
});
Object.defineProperty(exports, "createRenderMark", {
  enumerable: true,
  get: function () {
    return _renderMark.default;
  }
});
Object.defineProperty(exports, "createRenderContent", {
  enumerable: true,
  get: function () {
    return _renderContent.default;
  }
});
Object.defineProperty(exports, "createRenderEditable", {
  enumerable: true,
  get: function () {
    return _renderEditable.default;
  }
});
Object.defineProperty(exports, "createRenderNode", {
  enumerable: true,
  get: function () {
    return _renderNode.default;
  }
});
Object.defineProperty(exports, "createOnCangjieSelect", {
  enumerable: true,
  get: function () {
    return _onCangjieSelect.default;
  }
});
Object.defineProperty(exports, "createOnCangjieCopyOrCut", {
  enumerable: true,
  get: function () {
    return _onCangjieCopyOrCut.default;
  }
});
Object.defineProperty(exports, "createGetNodeSelectToolbarItems", {
  enumerable: true,
  get: function () {
    return _getNodeSelectToolbarItems.default;
  }
});

var _contextMenu = _interopRequireDefault(require("./contextMenu"));

var _onKeyDown = _interopRequireDefault(require("./onKeyDown"));

var _onMouseMove = _interopRequireDefault(require("./onMouseMove"));

var _onCangjieInput = _interopRequireDefault(require("./onCangjieInput"));

var _onAction = _interopRequireDefault(require("./onAction"));

var _onChange = _interopRequireDefault(require("./onChange"));

var _renderMark = _interopRequireDefault(require("./renderMark"));

var _renderContent = _interopRequireDefault(require("./renderContent"));

var _renderEditable = _interopRequireDefault(require("./renderEditable"));

var _renderNode = _interopRequireDefault(require("./renderNode"));

var _onCangjieSelect = _interopRequireDefault(require("./onCangjieSelect"));

var _onCangjieCopyOrCut = _interopRequireDefault(require("./onCangjieCopyOrCut"));

var _getNodeSelectToolbarItems = _interopRequireDefault(require("./getNodeSelectToolbarItems"));
