// @ali/4ever-plugin-list/dist-node/utils/constant.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SYMBOL_MAGIC_SIZE_MAP = exports.SYMBOL_MAGIC_OFFSET_X_MAP = exports.SYMBOL_MAGIC_OFFSET_Y_MAP = exports.STICKER_REG = exports.TASK_BOX_WIDTH = exports.TASK_BORDER_LEFT_WIDTH = exports.TASK_PADDING = exports.TASK_EXTRA_BUTTONS_WIDTH = exports.SYMBOL_SZ = exports.SYMBOL_STYLE = exports.SYMBOL_FONT = exports.SYMBOL_CONTENT_SPACING = exports.ROMAN_FORMAT = exports.ROMAN_EXTRA_INDENT = void 0;
// 罗马类型
const ROMAN_FORMAT = ['upperRoman', 'lowerRoman']; // 罗马项目符号额外增加缩进值

exports.ROMAN_FORMAT = ROMAN_FORMAT;
const ROMAN_EXTRA_INDENT = 20; // 列表项目符号到内容之间的距离

exports.ROMAN_EXTRA_INDENT = ROMAN_EXTRA_INDENT;
const SYMBOL_CONTENT_SPACING = 8; // 项目符号默认字体

exports.SYMBOL_CONTENT_SPACING = SYMBOL_CONTENT_SPACING;
const SYMBOL_FONT = 'Arial'; // 列表右侧的按钮区域宽度

exports.SYMBOL_FONT = SYMBOL_FONT;
const TASK_EXTRA_BUTTONS_WIDTH = 120; // 列表看不见的左侧 border 宽度

exports.TASK_EXTRA_BUTTONS_WIDTH = TASK_EXTRA_BUTTONS_WIDTH;
const TASK_BORDER_LEFT_WIDTH = 10; // 代办 checbox 宽度

exports.TASK_BORDER_LEFT_WIDTH = TASK_BORDER_LEFT_WIDTH;
const TASK_BOX_WIDTH = 16; // 代办列表的 padding

exports.TASK_BOX_WIDTH = TASK_BOX_WIDTH;
const TASK_PADDING = 2; // 项目符号默认大小

exports.TASK_PADDING = TASK_PADDING;
const SYMBOL_SZ = 11;
exports.SYMBOL_SZ = SYMBOL_SZ;
const SYMBOL_STYLE = {// sz: SYMBOL_SZ,
};
exports.SYMBOL_STYLE = SYMBOL_STYLE;
const STICKER_REG = /^\[(.+)\]$/; // ● ■ 不是居中的 unicode 符号，需要使用 magic number 调整

exports.STICKER_REG = STICKER_REG;
const SYMBOL_MAGIC_OFFSET_Y_MAP = {
  '\u25cf': 4.5,
  '\u25cb': 4.5,
  '\u25a0': 12,
  '\u2605': 4.5
};
exports.SYMBOL_MAGIC_OFFSET_Y_MAP = SYMBOL_MAGIC_OFFSET_Y_MAP;
const SYMBOL_MAGIC_OFFSET_X_MAP = {
  '\u25c6': 1,
  '\u2605': 1
};
exports.SYMBOL_MAGIC_OFFSET_X_MAP = SYMBOL_MAGIC_OFFSET_X_MAP;
const SYMBOL_MAGIC_SIZE_MAP = {
  '\u25a0': 2,
  '\u25c6': 6,
  '\u2605': 2
};
exports.SYMBOL_MAGIC_SIZE_MAP = SYMBOL_MAGIC_SIZE_MAP;
