// @ali/4ever-plugin-list/dist-node/utils/getStickerSymbol.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getStickerSymbolWidth = getStickerSymbolWidth;
exports.default = getStickerSymbol;

var _react = _interopRequireDefault(require("react"));

var _everUtils = require("@ali/4ever-utils");

var _everPluginSticker = require("@ali/4ever-plugin-sticker");

const _createElement = /*#__PURE__*/_react.default.createElement;
let sticker;

function getStickerSymbolWidth(size) {
  const sz = parseInt(size, 10);
  return _everUtils.unitsConverter.toPX(`${(sz > 0 ? sz : 10) * 1.5}pt`);
}

function getStickerSymbol(code, size = '10pt') {
  if (!sticker) {
    sticker = new _everPluginSticker.StickerManager({
      customSticker: [_everPluginSticker.DEFAULT_STICKERS]
    });
  }

  return sticker.getInfoByCode(code) ? /*#__PURE__*/_createElement(_everPluginSticker.StickerItem, {
    code: code,
    sz: getStickerSymbolWidth(size),
    sticker: sticker
  }) : null;
}
