// @ali/4ever-plugin-list/dist-node/utils/listStyleMapping.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mapping = exports.DEFAULT_NUMBER_LIST_ARRAY = exports.DEFAULT_BULLET_LIST_ARRAY = exports.symbol_type_miscrosoft = exports.symbol_text_microsoft = void 0;

/**
 * 无序命名规则
 * 例：SCIR_ECIR_SREC, 形状 + 形状 + 形状
 *
 * 首字母：
 * S -> solid (■)
 * E -> empty (❏)
 *
 * 形状：
 * ● -> CIR -> circle
 * ❏ -> REC -> rect
 * ❖ -> FDI -> four diomand
 * ◆ -> DIO -> diomand
 * ➔ -> ARR -> arrow
 * ➢ -> RIG -> right
 * */
const symbol_text_microsoft = {
  Ø: '\u27a2',
  l: '\u25cf',
  u: '\u25c6',
  n: '\u25a0',
  '+': '+',
  '-': '-'
};
exports.symbol_text_microsoft = symbol_text_microsoft;
const symbol_type_miscrosoft = {
  Ø: 'SRIG_ECIR_SREC',
  l: 'SCIR_ECIR_SREC',
  u: 'SARR_SDIO_SCIR',
  n: 'SCIR_ECIR_SREC',
  '+': 'MINUS_MINUS_MINUS',
  '-': 'PLUS_PLUS_PLUS'
};
exports.symbol_type_miscrosoft = symbol_type_miscrosoft;
let DEFAULT_BULLET_LIST_ARRAY;
/**
 * 有序命名规则
 * 例： DEC_LEN_LROM_P ，符号类型 + 符号类型 + 符号类型 + 首个符号格式
 *
 * 符号类型：
 * DEC -> decimal
 * DDEC -> decimalZero
 * LEN -> lowerLetter
 * UEN -> upperLetter
 * LROM -> lowerRoman
 * UROM -> upperRoman
 * ZH -> chineseCountingThousand
 *
 * 符号格式：
 * P（point) -> xxx.
 * H（half)  -> xxx)
 * A（all)   -> (xxx)
 * D (don)   -> xxx、
 * */

exports.DEFAULT_BULLET_LIST_ARRAY = DEFAULT_BULLET_LIST_ARRAY;

(function (DEFAULT_BULLET_LIST_ARRAY) {
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["SCIR_ECIR_SREC"] = 0] = "SCIR_ECIR_SREC";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["SFDI_SRIG_SREC"] = 1] = "SFDI_SRIG_SREC";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["EREC_EREC_EREC"] = 2] = "EREC_EREC_EREC";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["SARR_SDIO_SCIR"] = 3] = "SARR_SDIO_SCIR";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["SSTA_ECIR_SREC"] = 4] = "SSTA_ECIR_SREC";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["SRIG_ECIR_SREC"] = 5] = "SRIG_ECIR_SREC";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["MINUS_MINUS_MINUS"] = 6] = "MINUS_MINUS_MINUS";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["PLUS_PLUS_PLUS"] = 7] = "PLUS_PLUS_PLUS";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["VACATION_TEA_MILKTEA"] = 8] = "VACATION_TEA_MILKTEA";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["RIGHT_NAIL_LOVE"] = 9] = "RIGHT_NAIL_LOVE";
  DEFAULT_BULLET_LIST_ARRAY[DEFAULT_BULLET_LIST_ARRAY["FIRE_ROCKET_NAIL"] = 10] = "FIRE_ROCKET_NAIL";
})(DEFAULT_BULLET_LIST_ARRAY || (exports.DEFAULT_BULLET_LIST_ARRAY = DEFAULT_BULLET_LIST_ARRAY = {}));

let DEFAULT_NUMBER_LIST_ARRAY;
/**
 * unicode映射表
 * ● -> \u25cf
 * ○ -> \u25cb
 * ■ -> \u25a0
 * ❖ -> \u2756
 * ❏ -> \u274f
 * ➔ -> \u2794
 * ◆  -> \u25c6
 * ★ -> \u2605
 * ➢ -> \u27a2
 * -  -> \u2012
 * +  ->  +
 * */

exports.DEFAULT_NUMBER_LIST_ARRAY = DEFAULT_NUMBER_LIST_ARRAY;

(function (DEFAULT_NUMBER_LIST_ARRAY) {
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["DEC_LEN_LROM_P"] = 0] = "DEC_LEN_LROM_P";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["DEC_LEN_LROM_H"] = 1] = "DEC_LEN_LROM_H";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["DEC_DEC_DEC_P"] = 2] = "DEC_DEC_DEC_P";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["UEN_LEN_LROM_P"] = 3] = "UEN_LEN_LROM_P";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["UROM_UEN_DEC_P"] = 4] = "UROM_UEN_DEC_P";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["DDEC_LEN_LROM_P"] = 5] = "DDEC_LEN_LROM_P";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["ZH_DEC_LEN_D"] = 6] = "ZH_DEC_LEN_D";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["ZH_DEC_LEN_A"] = 7] = "ZH_DEC_LEN_A";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["LEN_LROM_DEC_P"] = 8] = "LEN_LROM_DEC_P";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["DEC_LEN_LROM_A"] = 9] = "DEC_LEN_LROM_A";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["LEN_LROM_DEC_H"] = 10] = "LEN_LROM_DEC_H";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["DEC_LEN_LROM_D"] = 11] = "DEC_LEN_LROM_D";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["ZH_LEN_LROM_A"] = 12] = "ZH_LEN_LROM_A";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["ZH_LEN_LROM_D"] = 13] = "ZH_LEN_LROM_D";
  DEFAULT_NUMBER_LIST_ARRAY[DEFAULT_NUMBER_LIST_ARRAY["ZH_ZH_DEC_D"] = 14] = "ZH_ZH_DEC_D";
})(DEFAULT_NUMBER_LIST_ARRAY || (exports.DEFAULT_NUMBER_LIST_ARRAY = DEFAULT_NUMBER_LIST_ARRAY = {}));

const SCIR_ECIR_SREC = [['bullet', '\u25cf', 'left'], ['bullet', '\u25cb', 'left'], ['bullet', '\u25a0', 'left']];
const SFDI_SRIG_SREC = [['bullet', '\u2756', 'left'], ['bullet', '\u27a2', 'left'], ['bullet', '\u25a0', 'left']];
const EREC_EREC_EREC = [['bullet', '\u274f', 'left'], ['bullet', '\u274f', 'left'], ['bullet', '\u274f', 'left']];
const SARR_SDIO_SCIR = [['bullet', '\u2794', 'left'], ['bullet', '\u25c6', 'left'], ['bullet', '\u25cf', 'left']];
const SSTA_ECIR_SREC = [['bullet', '\u2605', 'left'], ['bullet', '\u25cb', 'left'], ['bullet', '\u25a0', 'left']];
const SRIG_ECIR_SREC = [['bullet', '\u27a2', 'left'], ['bullet', '\u25cb', 'left'], ['bullet', '\u25a0', 'left']];
const MINUS_MINUS_MINUS = [['bullet', '\u2012', 'left'], ['bullet', '\u2012', 'left'], ['bullet', '\u2012', 'left']];
const PLUS_PLUS_PLUS = [['bullet', '+', 'left'], ['bullet', '+', 'left'], ['bullet', '+', 'left']];
const VACATION_TEA_MILKTEA = [['bullet', '[休假]', 'left'], ['bullet', '[茶]', 'left'], ['bullet', '[奶茶]', 'left']];
const RIGHT_NAIL_LOVE = [['bullet', '[向右]', 'left'], ['bullet', '[钉子]', 'left'], ['bullet', '[爱心]', 'left']];
const FIRE_ROCKET_NAIL = [['bullet', '[火]', 'left'], ['bullet', '[火箭]', 'left'], ['bullet', '[钉子]', 'left']]; // ordered list

const DEC_LEN_LROM_P = [['decimal', '%1.', 'left'], ['lowerLetter', '%2.', 'left'], ['lowerRoman', '%3.', 'left']];
const DEC_LEN_LROM_H = [['decimal', '%1)', 'left'], ['lowerLetter', '%2)', 'left'], ['lowerRoman', '%3)', 'left']];
const DEC_DEC_DEC_P = [['decimal', '%1.', 'left'], ['decimal', '%1.%2', 'left'], ['decimal', '%1.%2.%3', 'left']];
const UEN_LEN_LROM_P = [['upperLetter', '%1.', 'left'], ['lowerLetter', '%2.', 'left'], ['lowerRoman', '%3.', 'left']];
const UROM_UEN_DEC_P = [['upperRoman', '%1.', 'left'], ['upperLetter', '%2.', 'left'], ['decimal', '%3.', 'left']];
const DDEC_LEN_LROM_P = [['decimalZero', '%1.', 'left'], ['lowerLetter', '%2.', 'left'], ['lowerRoman', '%3.', 'left']];
const ZH_LEN_LROM_D = [['chineseCountingThousand', '%1、', 'left'], ['lowerLetter', '%2)', 'left'], ['lowerRoman', '%3.', 'left']];
const ZH_LEN_LROM_A = [['chineseCountingThousand', '(%1)', 'left'], ['lowerLetter', '%2)', 'left'], ['lowerRoman', '%3.', 'left']];
const ZH_DEC_LEN_D = [['chineseCountingThousand', '%1、', 'left'], ['decimal', '%2.', 'left'], ['lowerLetter', '%3.', 'left']];
const ZH_DEC_LEN_A = [['chineseCountingThousand', '(%1)', 'left'], ['decimal', '%2.', 'left'], ['lowerLetter', '%3.', 'left']];
const ZH_ZH_DEC_D = [['chineseCountingThousand', '%1、', 'left'], ['chineseCountingThousand', '(%1)', 'left'], ['decimal', '%3.', 'left']];
const LEN_LROM_DEC_P = [['lowerLetter', '%2.', 'left'], ['lowerRoman', '%3.', 'left'], ['decimal', '%1.', 'left']];
const DEC_LEN_LROM_A = [['decimal', '(%1)', 'left'], ['lowerLetter', '%2)', 'left'], ['lowerRoman', '%3)', 'left']];
const LEN_LROM_DEC_H = [['lowerLetter', '%1)', 'left'], ['lowerRoman', '%2)', 'left'], ['decimal', '%3)', 'left']];
const DEC_LEN_LROM_D = [['decimal', '%1、', 'left'], ['lowerLetter', '%2.', 'left'], ['lowerRoman', '%3.', 'left']];
const mapping = {
  SCIR_ECIR_SREC,
  SFDI_SRIG_SREC,
  EREC_EREC_EREC,
  SARR_SDIO_SCIR,
  SSTA_ECIR_SREC,
  SRIG_ECIR_SREC,
  MINUS_MINUS_MINUS,
  PLUS_PLUS_PLUS,
  VACATION_TEA_MILKTEA,
  RIGHT_NAIL_LOVE,
  FIRE_ROCKET_NAIL,
  DEC_LEN_LROM_P,
  DEC_LEN_LROM_H,
  DEC_LEN_LROM_A,
  DEC_DEC_DEC_P,
  UEN_LEN_LROM_P,
  UROM_UEN_DEC_P,
  DDEC_LEN_LROM_P,
  ZH_LEN_LROM_D,
  ZH_LEN_LROM_A,
  ZH_DEC_LEN_D,
  ZH_DEC_LEN_A,
  ZH_ZH_DEC_D,
  LEN_LROM_DEC_P,
  LEN_LROM_DEC_H,
  DEC_LEN_LROM_D
};
exports.mapping = mapping;
