// @ali/4ever-bi/dist-web/plugins/columns/components/columnsMenuLegacy.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { ColumnsPlugin, PluginRoles } from '@ali/4ever-bamboo';
import { MenuItem } from '@ali/we-design';
var insertColumns = ColumnsPlugin.actions.insertColumns;
export default function getMenu(size, configs) {
  return function ColumnsMenu(props) {
    var menuItem = props.menuItem,
        controller = props.controller;

    var title = menuItem.title,
        rest = _objectWithoutPropertiesLoose(menuItem, ["title"]);

    var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;

    var handleClick = function handleClick() {
      controller.run('onAction', insertColumns({
        size: size,
        text: ''
      }));
    };

    return /*#__PURE__*/_createElement(MenuItem, _extends({}, rest, {
      disabled: isCollapsedAtEdge,
      onClick: handleClick,
      role: "" + PluginRoles.columns + size
    }), title || '');
  };
}
