// @ali/we-design-next/dist-node/office/guidance/MapAnchor.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MapAnchor = exports.Anchor = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactDom = _interopRequireDefault(require("react-dom"));
var _posIcon = require("./posIcon");
var _style = require("./style");
var _PosIcon, _div;
const _createElement = /*#__PURE__*/_react.default.createElement;
const Anchor = props => {
  const {
    left = 0,
    top = 0,
    duration = 500
  } = props;
  const dur = (duration / 1000).toFixed(3);
  return /*#__PURE__*/_createElement(_style.Pos, {
    style: {
      top: top - 20,
      left: left - 10,
      transition: `top ${dur}s ease-in-out, left ${dur}s ease-in-out`
    }
  }, _PosIcon || (_PosIcon = /*#__PURE__*/_createElement(_posIcon.PosIcon, null)), _div || (_div = /*#__PURE__*/_createElement("div", {
    className: "circle"
  })));
};
exports.Anchor = Anchor;
let current = null;
const MapAnchor = {
  place: (left, top) => {
    var _Anchor;
    MapAnchor.remove();
    requestAnimationFrame(() => {
      _reactDom.default.render(_Anchor || (_Anchor = /*#__PURE__*/_createElement(Anchor, {
        left: left,
        top: top,
        duration: 0
      })), ensureContainer());
      current = [left, top];
    });
  },
  move: (left, top) => {
    var _Anchor2;
    if (current === null) {
      return;
    }
    const [pLeft, pTop] = current;
    const dLeft = left - pLeft;
    const dTop = top - pTop;
    const dist = Math.sqrt(dLeft * dLeft + dTop * dTop);
    const duration = dist < 200 ? 200 : dist > 1000 ? 1000 : dist;
    requestAnimationFrame(() => {
      _reactDom.default.render(_Anchor2 || (_Anchor2 = /*#__PURE__*/_createElement(Anchor, {
        left: left,
        top: top,
        duration: duration
      })), ensureContainer());
      current = [left, top];
    });
  },
  remove: () => {
    _reactDom.default.unmountComponentAtNode(ensureContainer());
    current = null;
  }
};
exports.MapAnchor = MapAnchor;
const ensureContainer = () => {
  const id = 'wdn-map-anchor-root';
  const divEle = document.querySelector(`div#${id}`);
  if (!divEle) {
    const ele = document.createElement('div');
    ele.id = id;
    ele.style.position = 'absolute';
    ele.style.width = '100%';
    ele.style.top = '0px';
    ele.style.left = '0px';
    document.body.appendChild(ele);
  }
  ;
  return document.querySelector(`div#${id}`);
};
