// @ali/4ever-plugin-mention/dist-node/bi/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.biFactory = void 0;

var _index = require("../mo/index");

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _menubar = _interopRequireDefault(require("./menubar"));

var _renderNode = _interopRequireDefault(require("./renderNode"));

var _renderMark = _interopRequireDefault(require("./renderMark"));

var _onCangjieInput = _interopRequireDefault(require("./handlers/onCangjieInput"));

var _onKeyDown = _interopRequireDefault(require("./handlers/onKeyDown"));

var _onSelect = _interopRequireDefault(require("./handlers/onSelect"));

var _onCangjiePaste = _interopRequireDefault(require("./handlers/onCangjiePaste"));

var _createOnAction = _interopRequireDefault(require("./handlers/createOnAction"));

var _queries = require("./queries");

var _actions = require("./actions");

var _createRenderToolbar = _interopRequireDefault(require("./createRenderToolbar"));

const DEFAULT_THEME = {
  suggestionWidth: '200px',
  suggestionMaxHeight: '250px'
};

const biFactory = configs => {
  const mentionConfig = {
    theme: { ...DEFAULT_THEME,
      ...configs.mention?.popupOptions
    },
    ...configs.mention
  };
  const mentionId = mentionConfig.getMentionId ? mentionConfig.getMentionId() : '';
  return { ...(0, _index.moFactory)({}),
    toolbar: (0, _toolbar.default)(mentionConfig),
    renderToolbar: (0, _createRenderToolbar.default)(mentionConfig.locale),
    renderNode: (0, _renderNode.default)(mentionId, mentionConfig),
    renderMark: (0, _renderMark.default)(mentionConfig),
    onKeyDown: _onKeyDown.default,
    menubar: _menubar.default,
    onCangjieInput: _onCangjieInput.default,
    onCangjiePaste: (0, _onCangjiePaste.default)(configs),
    onSelect: _onSelect.default,
    onAction: (0, _createOnAction.default)(mentionConfig),
    locale: mentionConfig.locale || {},
    queries: _queries.queries,
    actions: _actions.actions,
    isActions: _actions.isActions
  };
};

exports.biFactory = biFactory;
