// @ali/we-ui/dist-web/components/tabs/TabBarTabsNode.js
/* eslint-disable @typescript-eslint/explicit-function-return-type */
import React, { useRef } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import warning from 'warning';
import { classnames } from "../../utils";
import { ClosedNormal } from "../icons";
import { TabBarTab, TabBarTabItemWrap, TabBarWrapper } from "./styled";

var TabBarTabsNode = function TabBarTabsNode(props) {
  var saveRef = props.saveRef,
      panes = props.panes,
      size = props.size,
      activeKey = props.activeKey,
      type = props.type,
      className = props.className,
      tabPosition = props.tabPosition,
      testid = props.testid;
  var rst = []; // 下面两个ref是为了缓存函数引用，根据key进行缓存
  // eslint-disable-next-line @typescript-eslint/no-explicit-any

  var onTabClicks = useRef({}); // eslint-disable-next-line @typescript-eslint/no-explicit-any

  var onTabCloses = useRef({}); // eslint-disable-next-line @typescript-eslint/no-explicit-any

  React.Children.forEach(panes, function (child) {
    if (!child) {
      return;
    }

    var key = child.key;
    var isActived = activeKey === key;

    if (!onTabClicks.current[key] || onTabClicks.current[key] !== props.onTabClick) {
      onTabClicks.current[key] = function (e) {
        if (props.onTabClick) {
          props.onTabClick(key, e);
        }
      };
    }

    if (!onTabCloses.current[key] || onTabCloses.current[key] !== props.onClose) {
      onTabCloses.current[key] = function (e) {
        e.stopPropagation();

        if (props.onClose) {
          props.onClose(key);
        }
      };
    }

    warning('tab' in child.props, 'There must be `tab` property on children of Tabs.');
    var id = props.id ? key + "-" + props.id : key;
    var closable = typeof child.props.closable !== 'undefined' ? !!child.props.closable : true;

    var node = /*#__PURE__*/_createElement(TabBarTab, {
      role: "tab",
      "aria-disabled": child.props.disabled ? 'true' : 'false',
      "aria-selected": isActived ? 'true' : 'false',
      "aria-controls": "tabpane-" + id,
      tabIndex: isActived ? 0 : -1,
      key: key,
      id: "tab-" + id,
      className: classnames(type, size, tabPosition, className, {
        disabled: !!child.props.disabled,
        actived: isActived
      }),
      ref: isActived ? saveRef('activeTab') : undefined,
      onClick: !isActived && !child.props.disabled ? onTabClicks.current[key] : undefined,
      "data-testid": testid + "_" + id
    }, /*#__PURE__*/_createElement(TabBarTabItemWrap, {
      className: classnames(type, size, tabPosition, className, {
        actived: isActived
      })
    }, child.props.tab, type === 'editable-card' && closable && /*#__PURE__*/_createElement(ClosedNormal, {
      onClick: onTabCloses.current[key],
      className: "card-tab-close-button",
      style: {
        fontSize: 14
      }
    }), type === 'editable-line' && closable && /*#__PURE__*/_createElement(ClosedNormal, {
      onClick: onTabCloses.current[key],
      className: "line-tab-close-button",
      style: {
        fontSize: 14
      }
    })));

    rst.push(node);
  });
  return /*#__PURE__*/_createElement(TabBarWrapper, {
    className: classnames(tabPosition && "" + tabPosition),
    ref: saveRef('navTabsContainer')
  }, rst);
};

export default TabBarTabsNode;
