// @ali/4ever-bi/dist-web/plugins/sz/components/next/SzSingleButton.js
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { PluginRoles } from '@ali/4ever-bamboo';
import { IconFont, IconReduceFont } from '@ali/4ever-component';
import { adjustSz } from "../../actions";
import { SHORTCUT } from "../../utils/getTipsByType";

var _ref = /*#__PURE__*/_createElement(IconReduceFont, null);

export function createSzSmallerButton(props) {
  var controller = props.controller,
      locale = props.locale,
      icon = props.icon,
      title = props.title,
      tooltip = props.tooltip,
      badgeDot = props.badgeDot,
      disabled = props.disabled;

  var handleClick = function handleClick() {
    controller.run('onAction', adjustSz(false));
  };

  var isDisabled = disabled || controller.query('isSelectionInLinkCard');
  var buttonConfig = {
    icon: icon || _ref,
    onClick: handleClick,
    title: title || (locale == null ? void 0 : locale.buttonTitle),
    tooltip: _extends({
      title: title || (locale == null ? void 0 : locale.buttonTitle),
      shortCut: SHORTCUT.smaller
    }, tooltip),
    badgeDot: badgeDot,
    disabled: !!isDisabled,
    testid: "ReduceFontSizeButton",
    role: PluginRoles.szDecrease
  };
  return buttonConfig;
}

var _ref2 = /*#__PURE__*/_createElement(IconFont, null);

export function createSzBiggerButton(props) {
  var controller = props.controller,
      locale = props.locale,
      icon = props.icon,
      title = props.title,
      tooltip = props.tooltip,
      badgeDot = props.badgeDot,
      disabled = props.disabled;

  var handleClick = function handleClick() {
    controller.run('onAction', adjustSz(true));
  };

  var isDisabled = disabled || controller.query('isSelectionInLinkCard');
  var buttonConfig = {
    icon: icon || _ref2,
    onClick: handleClick,
    title: title || (locale == null ? void 0 : locale.buttonTitle),
    tooltip: _extends({
      title: title || (locale == null ? void 0 : locale.buttonTitle),
      shortCut: SHORTCUT.bigger
    }, tooltip),
    badgeDot: badgeDot,
    disabled: !!isDisabled,
    testid: "add_fontsize_button",
    role: PluginRoles.szIncrease
  };
  return buttonConfig;
}
