// @ali/we-design-next/dist-node/collaboration/collaborator-avatars/AvatarCollaborators.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AvatarCollaborators = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _popover = require("../../display/popover");
var _styled = require("./styled");
var _common = require("../../common");
var _AvatarWithPopover = require("./AvatarWithPopover");
var _AvatarMoreList = require("./AvatarMoreList");
var _constants = require("./constants");
var _locale = require("../../locale");
var _i, _i2, _i3;
const _createElement = /*#__PURE__*/_react.default.createElement;
const PLACEMENT_OFFSET = [4, 0];
const AvatarCollaborators = /*#__PURE__*/_react.default.memo(props => {
  const {
    collaborators: items = [],
    max = _constants.DEFAULT_MAX_COLLABORATOR,
    onClick,
    highlight: isMoreButtonHighlight,
    renderMenu: customMorePanelTitle,
    renderCollaboratorMenu,
    ...rest
  } = props;
  const [visibleAvatarIndex, setVisibleAvatarIndex] = _react.default.useState(-1);

  // 数量
  const avatarCount = items.length;
  const isOverflowing = avatarCount > max;
  const t = (0, _locale.useTranslate)();
  const sortedItems = _react.default.useMemo(() => {
    return items.map((item, index) => ({
      ...item,
      index
    })).sort((a, b) => {
      const aValue = a.highlight ? 0 : 1;
      const bValue = b.highlight ? 0 : 1;
      return aValue - bValue;
    });
  }, [items]);
  return /*#__PURE__*/_createElement(_styled.AvatarCollaboratorsWrap, (0, _common.mergeRestProps)(rest, 'wdn-collaborator-avatars'), /*#__PURE__*/_createElement(_styled.GroupAvatarList, null, sortedItems.slice(0, isOverflowing ? max - 1 : max).map(item => {
    const {
      index: idx,
      key
    } = item;
    return /*#__PURE__*/_createElement(_styled.GroupAvatarItem, {
      className: "avatar-item",
      key: key
    }, /*#__PURE__*/_createElement(_AvatarWithPopover.AvatarWithPopover, (0, _extends2.default)({}, item, {
      renderMenu: renderCollaboratorMenu ? () => renderCollaboratorMenu(item) : undefined,
      color: item.color || _constants.COLOR_SCHEME[idx % _constants.COLOR_SCHEME_LEN],
      onClick: () => onClick === null || onClick === void 0 ? void 0 : onClick(item.key),
      isPopoverVisible: visibleAvatarIndex === idx,
      onPopoverVisibleChange: visible => {
        if (visible) {
          setVisibleAvatarIndex(idx);
        } else {
          setVisibleAvatarIndex(visibleAvatarIndex => {
            return visibleAvatarIndex === idx ? -1 : visibleAvatarIndex;
          });
        }
      }
    })));
  }), isOverflowing && /*#__PURE__*/_createElement(_styled.GroupAvatarItem, null, /*#__PURE__*/_createElement(_popover.Popover, {
    animation: true,
    onVisibleChange: v => {
      if (v) {
        setVisibleAvatarIndex('more');
      } else if (visibleAvatarIndex === 'more') {
        setVisibleAvatarIndex(-1);
      }
    },
    visible: visibleAvatarIndex === 'more',
    trigger: "click",
    placement: "bottomRight",
    placementOffset: PLACEMENT_OFFSET,
    content: /*#__PURE__*/_createElement(_styled.OverlayWrap, null, /*#__PURE__*/_createElement(_styled.CountWrap, null, !!customMorePanelTitle ? customMorePanelTitle() : t('wdn_avatar_group_total-collab-people', {
      value: avatarCount
    })), /*#__PURE__*/_createElement(_AvatarMoreList.AvatarMoreList, {
      items: sortedItems,
      onClick: onClick
    }))
  }, /*#__PURE__*/_createElement(_styled.GroupMoreItem, {
    className: (0, _common.mergeClassName)(visibleAvatarIndex === 'more' ? 'wdn-more-box-active' : '', isMoreButtonHighlight ? 'wdn-more-btn-highlight' : '')
  }, Number(avatarCount) < 999 ? /*#__PURE__*/_createElement(_styled.GroupMoreItemSpan, null, `${avatarCount}`) : /*#__PURE__*/_createElement(_react.default.Fragment, null, _i || (_i = /*#__PURE__*/_createElement("i", null)), _i2 || (_i2 = /*#__PURE__*/_createElement("i", null)), _i3 || (_i3 = /*#__PURE__*/_createElement("i", null))))))));
});
exports.AvatarCollaborators = AvatarCollaborators;
