// @ali/we-design-next/dist-node/collaboration/collaborator-avatars/AvatarWithPopover.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AvatarWithPopover = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _popover = require("../../display/popover");
var _styled = require("./styled");
var _avatar = require("../../display/avatar");
var _Avatar = require("./Avatar");
var _constants = require("./constants");
var _common = require("../../common");
var _SplitLine;
const _createElement = /*#__PURE__*/_react.default.createElement;
const PopoverWrapper = /*#__PURE__*/_styledComponents.default.div(["min-width:100px;"]);
const ProfileInfoWrapper = /*#__PURE__*/_styledComponents.default.div(["display:flex;justify-content:space-between;padding:8px 12px;"]);
const ProfileInfoInnerLeft = /*#__PURE__*/_styledComponents.default.div(["display:flex;align-items:center;justify-content:flex-start;margin-right:8px;"]);
const ProfileInfoInnerCenter = /*#__PURE__*/_styledComponents.default.div(["display:flex;min-width:105px;align-items:center;justify-content:space-between;"]);
const ProfileInfoInnerRight = /*#__PURE__*/_styledComponents.default.div(["display:flex;align-items:center;justify-content:flex-end;"]);
const NickText = /*#__PURE__*/_styledComponents.default.span(["flex:1;width:50%;max-width:96px;font-weight:400;font-size:13px;color:", ";text-overflow:ellipsis;white-space:nowrap;overflow:hidden;margin-right:8px;"], _common.COLORS.LIGHT.LEVEL1_BASE_COLOR);
const StatusText = /*#__PURE__*/_styledComponents.default.span(["font-weight:400;font-size:12px;margin-right:4px;color:", ";"], _common.COLORS.LIGHT.LEVEL3_BASE_COLOR);
const SplitLine = /*#__PURE__*/_styledComponents.default.div(["height:1px;margin:0 4px;background:rgba(126,134,142,0.16);"]);
const ProfileInfo = props => {
  const {
    avatar: src,
    name,
    color,
    description
  } = props;
  return /*#__PURE__*/_createElement(ProfileInfoWrapper, null, /*#__PURE__*/_createElement(ProfileInfoInnerLeft, null, /*#__PURE__*/_createElement(_avatar.Avatar, {
    name: name,
    src: src,
    size: _constants.AVATAR_SIZE
  })), /*#__PURE__*/_createElement(ProfileInfoInnerCenter, null, /*#__PURE__*/_createElement(NickText, null, name), description && /*#__PURE__*/_createElement(StatusText, null, description)), /*#__PURE__*/_createElement(ProfileInfoInnerRight, null, /*#__PURE__*/_createElement(_styled.Dot, {
    style: {
      background: color
    }
  })));
};
const AvatarWithPopover = props => {
  const {
    renderMenu,
    color,
    description,
    name,
    avatar: src,
    isPopoverVisible,
    onPopoverVisibleChange,
    ...rest
  } = props;
  const menu = isPopoverVisible && renderMenu ? renderMenu() : null;
  return /*#__PURE__*/_createElement(_popover.Popover, {
    visible: isPopoverVisible,
    onVisibleChange: onPopoverVisibleChange,
    animation: true,
    content: /*#__PURE__*/_createElement(PopoverWrapper, null, /*#__PURE__*/_createElement(ProfileInfo, {
      color: color,
      description: description,
      name: name,
      avatar: src
    }), menu ? /*#__PURE__*/_createElement(_react.default.Fragment, null, _SplitLine || (_SplitLine = /*#__PURE__*/_createElement(SplitLine, null)), menu) : null),
    trigger: "mouse-enter",
    placement: "bottomLeft",
    placementOffset: [3, 0]
  }, /*#__PURE__*/_createElement(_Avatar.Avatar, (0, _extends2.default)({
    name: name,
    src: src,
    color: color
  }, rest)));
};
exports.AvatarWithPopover = AvatarWithPopover;
