// @ali/we-design-next/dist-node/collaboration/collaborator-avatars/Avatar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Avatar = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _styled = require("./styled");
var _Avatar = require("../../display/avatar/Avatar");
var _constants = require("./constants");
const _createElement = /*#__PURE__*/_react.default.createElement;
const Avatar = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    color,
    highlight: isHighlight,
    ...rest
  } = props;
  const wrapStyle = {
    '--hover-color': color,
    border: !color ? 'none' : undefined
  };
  const avata = /*#__PURE__*/_createElement(_styled.AvatarWrap, {
    style: wrapStyle,
    ref: ref,
    className: "wdn-avatar-group-item-wrapper"
  }, /*#__PURE__*/_createElement(_Avatar.Avatar, (0, _extends2.default)({
    shape: "circle",
    size: _constants.AVATAR_SIZE,
    style: {
      position: 'absolute',
      top: -1,
      right: -1
    }
  }, rest)));
  return isHighlight ? /*#__PURE__*/_createElement(_styled.AvatarHighlightWrapper, null, avata) : avata;
});
exports.Avatar = Avatar;
