// @ali/we-design-next/dist-node/form/color-picker/sketch/common/interaction.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handleFocus = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
/* eslint-disable no-invalid-this */
const _createElement = /*#__PURE__*/_react.default.createElement;
const handleFocus = (Component, Span = 'span') => class Focus extends _react.default.Component {
  constructor(...args) {
    super(...args);
    this.state = {
      focus: false
    };
    this.handleFocus = () => this.setState({
      focus: true
    });
    this.handleBlur = () => this.setState({
      focus: false
    });
  }
  render() {
    return /*#__PURE__*/_createElement(Span, {
      onFocus: this.handleFocus,
      onBlur: this.handleBlur
    }, /*#__PURE__*/_createElement(Component, (0, _extends2.default)({}, this.props, this.state)));
  }
};
exports.handleFocus = handleFocus;
