// @ali/we-design-next/dist-node/form/color-picker/sketch/common/Swatch.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Swatch = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _reactcss = _interopRequireDefault(require("reactcss"));
var _interaction = require("./interaction");
var _Checkboard = _interopRequireDefault(require("./Checkboard"));
const _createElement = /*#__PURE__*/_react.default.createElement;
const ENTER = 13;
const Swatch = ({
  color,
  style,
  onClick = () => {},
  onHover,
  title = color,
  children,
  focus,
  focusStyle = {}
}) => {
  const transparent = color === 'transparent';
  const styles = (0, _reactcss.default)({
    default: {
      swatch: {
        background: color,
        height: '100%',
        width: '100%',
        cursor: 'pointer',
        position: 'relative',
        outline: 'none',
        ...style,
        ...(focus ? focusStyle : {})
      }
    }
  });
  const handleClick = e => onClick(color, e);
  const handleKeyDown = e => e.keyCode === ENTER && onClick(color, e);
  const handleHover = e => onHover(color, e);
  const optionalEvents = {};
  if (onHover) {
    optionalEvents.onMouseOver = handleHover;
  }
  return /*#__PURE__*/_createElement("div", (0, _extends2.default)({
    style: styles.swatch,
    onClick: handleClick,
    title: title,
    tabIndex: 0,
    onKeyDown: handleKeyDown
  }, optionalEvents), children, transparent && /*#__PURE__*/_createElement(_Checkboard.default, {
    borderRadius: styles.swatch.borderRadius,
    boxShadow: "inset 0 0 0 1px rgba(0,0,0,0.1)"
  }));
};
exports.Swatch = Swatch;
var _default = (0, _interaction.handleFocus)(Swatch);
exports.default = _default;
