// @ali/4ever-bi/dist-web/plugins/voteCard/handlers/createMenu.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React, { useCallback } from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { Block } from '@ali/4ever-cangjie';
import dayjs from 'dayjs';
import { VoteCardPlugin } from '@ali/4ever-bamboo';
import { Card } from '@ali/4ever-mo';
import { MenuItem } from '@ali/we-design';
import { insertCardCommand, setCardInjectionCommand } from '@ali/4ever-factory';
var DEFAULT_DATA = {
  title: '',
  voteType: VoteCardPlugin.EnumVoteType.RADIO,
  optionInfoList: [{
    content: '',
    pos: 1,
    optionTicket: 0,
    userOptionTicket: 0
  }, {
    content: '',
    pos: 2,
    optionTicket: 0,
    userOptionTicket: 0
  }],
  endTime: dayjs().add(7, 'days').valueOf(),
  showTicketBeforeVoting: false
};

var createMenubar = function createMenubar() {
  return function CalendarCardMenu(props) {
    var menuItem = props.menuItem,
        controller = props.controller;

    var title = menuItem.title,
        rest = _objectWithoutPropertiesLoose(menuItem, ["title"]);

    var handleClick = useCallback(function () {
      var node = Block.create({
        type: Card.TYPE,
        data: {
          cardType: 'vote',
          height: 400,
          metadata: DEFAULT_DATA
        }
      });
      controller.command(insertCardCommand, node).command(setCardInjectionCommand, node, {
        editing: true
      }).setIsFocused(false);
    }, [controller]);
    return /*#__PURE__*/_createElement(MenuItem, _extends({}, rest, {
      onClick: handleClick
    }), title || '');
  };
};

export default createMenubar;
