// @ali/4ever-plugin-hr/dist-web/mo/serializer/htmlToJsonML/parseFromWeb.js
import { createEmptyText } from '@ali/4ever-cangjie';
import { parseBorder } from "./utils";

function parseFromWeb() {
  return {
    name: 'hr',
    onOpenTag: function onOpenTag(state, name, attrs) {
      if (name !== 'hr') return false;
      var styleObj = attrs.styleObj;
      var lineStyle = styleObj.borderTop ? styleObj.borderTop : styleObj.borderBottom;

      if (lineStyle) {
        var lineStyleObj = parseBorder(lineStyle);
        var sz = lineStyleObj.sz,
            type = lineStyleObj.type,
            color = lineStyleObj.color;
        state.push(['hr', {
          type: type,
          sz: sz,
          color: color,
          width: '100%'
        }, createEmptyText()]);
      } else {
        state.push(['hr', createEmptyText()]);
      }

      return true;
    },
    onCloseTag: function onCloseTag(state, name) {
      if (name !== 'hr') return false;
      var hr = state.pop();
      state.append(hr);
      return true;
    }
  };
}

export default parseFromWeb;
