// @ali/we-design-next/dist-node/collaboration/mention/MentionList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MentionList = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactWindow = require("react-window");
var _reactWindowInfiniteLoader = _interopRequireDefault(require("react-window-infinite-loader"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _locale = require("../../locale");
var _MemberItem = require("./items/MemberItem");
var _styled = require("./styled");
var _common = require("../../common");
var _img;
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const _createElement = /*#__PURE__*/_react.default.createElement;
const VirtualList = /*#__PURE__*/(0, _styledComponents.default)(_reactWindow.FixedSizeList)(["", " overflow:overlay;"], _common.CSS_SCROLLBAR);
const MentionList = /*#__PURE__*/_react.default.memo(props => {
  const {
    items,
    onMention,
    onLoadMore
  } = props;
  const [active, setActive] = (0, _react.useState)(0);
  const itemData = (0, _react.useMemo)(() => ({
    items,
    active: active,
    onClick: i => onMention(i),
    onMouseEnter: i => setActive(i)
  }), [items, active, setActive]);
  const t = (0, _locale.useTranslate)();
  return items.length > 0 ? /*#__PURE__*/_createElement(_reactWindowInfiniteLoader.default, {
    isItemLoaded: i => items[i] !== undefined,
    itemCount: 999999,
    loadMoreItems: (start, stop) => {
      onLoadMore(start, stop - start + 1);
      return;
    }
  }, ({
    onItemsRendered,
    ref
  }) => /*#__PURE__*/_createElement(VirtualList, {
    ref: ref,
    width: 300,
    height: 300,
    itemCount: items.length,
    itemSize: 54,
    itemData: itemData,
    onItemsRendered: onItemsRendered,
    style: {
      // animation: 'fadeIn',
      // animationDuration: '.35s',
      overflowX: 'hidden',
      overflowY: 'overlay'
    },
    children: _MemberItem.MentionItem
  })) : /*#__PURE__*/_createElement(_styled.MentionEmptyWrapper, null, _img || (_img = /*#__PURE__*/_createElement("img", {
    src: "https://img.alicdn.com/imgextra/i2/O1CN019Tgrpl25H3CCoowv1_!!6000000007500-55-tps-90-90.svg",
    alt: ""
  })), t('wdn_mention_search_no_result'));
});
exports.MentionList = MentionList;
