// @ali/4ever-bi/dist-web/plugins/subject/components/content.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { environment, Commands } from '@ali/4ever-cangjie';
import { HeadingGroupPlugin, SubjectPlugin } from '@ali/4ever-bamboo';
import { isEmptyNode, isNormalParagraph } from "../utils";
var Content = HeadingGroupPlugin.Content;
var components = SubjectPlugin.components;
var Placeholder = components.Placeholder,
    InlinePlaceholder = components.InlinePlaceholder,
    InnerTextWrapper = components.InnerTextWrapper,
    InnerCardWrapper = components.InnerCardWrapper,
    SubjectPlaceholderWrapper = components.SubjectPlaceholderWrapper;

var SubjectContent = function SubjectContent(props) {
  var node = props.node,
      children = props.children,
      config = props.config,
      controller = props.controller,
      rest = _objectWithoutPropertiesLoose(props, ["node", "children", "config", "controller"]);

  var _node$data$subjectPr = node.data.subjectPr,
      subjectPr = _node$data$subjectPr === void 0 ? {} : _node$data$subjectPr;
  var name = subjectPr.name;
  var _config$locale = config.locale,
      locale = _config$locale === void 0 ? {} : _config$locale,
      renderInsertAttachmentComponent = config.renderInsertAttachmentComponent;
  var reviewConfig = config.review;

  var _ref = reviewConfig || {},
      ReviewPlaceholder = _ref.renderReviewPlaceholder;

  var reviewTextPlaceholder = locale.reviewTextPlaceholder,
      topicContentPlaceholder = locale.topicContentPlaceholder,
      topicContentPlaceholderMobile = locale.topicContentPlaceholderMobile,
      reviewTextPlaceholderMobile = locale.reviewTextPlaceholderMobile,
      conclusionContentPlaceholder = locale.conclusionContentPlaceholder;
  var node1 = node.nodes[1];

  var displayPlaceholder1 = function displayPlaceholder1(item) {
    if (name === 'review' || name === 'topic' || name === 'conclusion') {
      return isNormalParagraph(item) && isEmptyNode(item) && node.nodes.length === 2;
    }

    return false;
  };

  var _ref2 = /*#__PURE__*/_createElement(ReviewPlaceholder, {
    node: node
  });

  var renderFileChoosePlaceholder = function renderFileChoosePlaceholder() {
    var hasReviewTaskList = !!node.nodes.find(function (node) {
      var _data, _data$list;

      return node == null ? void 0 : (_data = node.data) == null ? void 0 : (_data$list = _data.list) == null ? void 0 : _data$list.isTaskList;
    }); // 没有taskList

    if (!hasReviewTaskList && ReviewPlaceholder) {
      return _ref2;
    }

    return null;
  };

  var focusToLine = React.useCallback(function () {
    controller.command(Commands.moveToStartOfNode, node1);
  }, [controller, node1]);

  var renderAttachmentPlaceholder = function renderAttachmentPlaceholder() {
    if (renderInsertAttachmentComponent) {
      return renderInsertAttachmentComponent({
        focusToLine: focusToLine
      });
    }

    return null;
  };

  var _ref3 = /*#__PURE__*/_createElement(InlinePlaceholder, {
    "data-placeholder": true
  }, reviewTextPlaceholderMobile);

  var _ref4 = /*#__PURE__*/_createElement(InnerTextWrapper, null, reviewTextPlaceholder);

  var _ref5 = /*#__PURE__*/_createElement(InlinePlaceholder, {
    "data-placeholder": true
  }, topicContentPlaceholderMobile);

  var _ref6 = /*#__PURE__*/_createElement(InnerTextWrapper, null, topicContentPlaceholder);

  var _ref7 = /*#__PURE__*/_createElement(InlinePlaceholder, {
    "data-placeholder": true
  }, conclusionContentPlaceholder);

  var placeholder = React.useMemo(function () {
    if (name === 'review') {
      if (environment.IS_MOBILE) {
        return _ref3;
      }

      return /*#__PURE__*/_createElement(SubjectPlaceholderWrapper, {
        "data-placeholder": true
      }, _ref4, /*#__PURE__*/_createElement(InnerCardWrapper, null, renderFileChoosePlaceholder()));
    } // 议题内容 placeholder


    if (name === 'topic') {
      if (environment.IS_MOBILE) {
        return _ref5;
      }

      return /*#__PURE__*/_createElement(SubjectPlaceholderWrapper, {
        "data-placeholder": true
      }, _ref6, /*#__PURE__*/_createElement(InnerCardWrapper, {
        className: "attachment"
      }, renderAttachmentPlaceholder()));
    } // 会议小结内容 placeholder


    if (name === 'conclusion') {
      return _ref7;
    }

    return null;
  }, [conclusionContentPlaceholder, name, reviewTextPlaceholder, topicContentPlaceholder, topicContentPlaceholderMobile, renderFileChoosePlaceholder]);

  if (children.length === 0) {
    return null;
  }

  return /*#__PURE__*/_createElement(Content, rest, node1 && /*#__PURE__*/_createElement(Placeholder, {
    controller: controller,
    node: node1,
    display: displayPlaceholder1,
    placeholder: placeholder
  }, children[0]), children.slice(1));
};

SubjectContent.displayName = 'SubjectContent';
export default SubjectContent;
