// @ali/4ever-bi/dist-web/plugins/subject/components/mentionButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import styled, { keyframes } from 'styled-components';
import classnames from 'classnames';
var clickIn = /*#__PURE__*/keyframes(["0%{transform:scale(1)}60%{transform:scale(1.2)}100%{transform:scale(1)}"]);
var zoomIn = /*#__PURE__*/keyframes(["0%{box-shadow:0 0 0 0px rgba(23,26,29,0.4);}100%{box-shadow:0 0 0 8px rgba(23,26,29,0);}"]);

var MentionIcon = /*#__PURE__*/_createElement("svg", {
  width: "16",
  height: "16",
  viewBox: "0 0 16 16",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/_createElement("path", {
  fill: "currentColor",
  fillRule: "evenodd",
  clipRule: "evenodd",
  d: "M11.1824 15.5556C10.3768 15.8241 9.23552 15.9775 7.98877 15.9775C3.15523 15.9775 0 12.8127 0 7.96C0 3.27031 3.24154 0 7.89287 0C12.4195 0 15.5652 2.92506 15.5652 7.13523C15.5652 10.2329 14.28 12.1414 12.1989 12.1414C11.0097 12.1414 10.137 11.4317 10.0315 10.3768H9.90684C9.47528 11.4701 8.58337 12.0743 7.42294 12.0743C5.43773 12.0743 4.0759 10.4343 4.0759 8.02713C4.0759 5.67749 5.45691 4.03754 7.42294 4.03754C8.49706 4.03754 9.43691 4.64173 9.80135 5.58159H9.92602V4.22935H11.0385V9.74381C11.0385 10.6645 11.518 11.1632 12.3907 11.1632C13.7046 11.1632 14.5102 9.61913 14.5102 7.11605C14.5102 3.46212 11.777 0.949446 7.80655 0.949446C3.89369 0.949446 1.04535 3.90328 1.04535 7.99836C1.04535 12.2277 3.81696 15.0281 7.99836 15.0281C9.31224 15.0281 10.5398 14.8651 11.1824 14.6061V15.5556ZM5.25552 8.0559C5.25552 9.86848 6.1666 11.0289 7.57638 11.0289C8.99576 11.0289 9.90684 9.83971 9.90684 8.00795C9.90684 6.20496 9.00535 5.06371 7.58597 5.06371C6.18578 5.06371 5.25552 6.25291 5.25552 8.0559Z"
}));

var ButtonBox = /*#__PURE__*/styled.button(["position:relative;display:inline-flex;pointer-events:auto;align-items:center;justify-content:center;height:24px;width:24px;border-radius:24px;cursor:pointer;background:none;outline:none;padding:0;border:0;transition:all 0.4s;transform-origin:50% 50%;color:rgba(23,26,29,0.4);transform:translate(0,1px);:hover{color:rgba(23,26,29);}&.click-in{animation-name:", ";animation-duration:0.3s;svg{border-radius:20px;animation-name:", ";animation-duration:0.3s;}}"], clickIn, zoomIn);

var MentionButton = function MentionButton(props) {
  var className = props.className,
      onClick = props.onClick,
      onAnimationEnd = props.onAnimationEnd,
      rest = _objectWithoutPropertiesLoose(props, ["className", "onClick", "onAnimationEnd"]);

  var _React$useState = React.useState(false),
      animate = _React$useState[0],
      setAnimate = _React$useState[1];

  var handleClick = React.useCallback(function (e) {
    if (typeof onClick === 'function') {
      onClick(e);
    }

    setAnimate(true);
  }, [onClick]);
  var handleAnimationEnd = React.useCallback(function (e) {
    if (typeof onAnimationEnd === 'function') {
      onAnimationEnd(e);
    }

    setAnimate(false);
  }, [onAnimationEnd]);
  var classNames = classnames(className, {
    'click-in': animate
  });
  return /*#__PURE__*/_createElement(ButtonBox, _extends({}, rest, {
    className: classNames,
    onClick: handleClick,
    onAnimationEnd: handleAnimationEnd
  }), MentionIcon);
};

export default MentionButton;
