// @ali/4ever-zhi/dist-web/plugins/subject/components/content.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { HeadingGroupPlugin, SubjectPlugin } from '@ali/4ever-bamboo';
import { isNormalParagraph } from "../utils";
var components = SubjectPlugin.components,
    utils = SubjectPlugin.utils;
var isEmptyNode = utils.isEmptyNode;
var InlinePlaceholder = components.InlinePlaceholder,
    Placeholder = components.Placeholder;
var Content = HeadingGroupPlugin.Content;

var SubjectContent = function SubjectContent(props) {
  var node = props.node,
      children = props.children,
      config = props.config,
      controller = props.controller,
      rest = _objectWithoutPropertiesLoose(props, ["node", "children", "config", "controller"]);

  var _node$data$subjectPr = node.data.subjectPr,
      subjectPr = _node$data$subjectPr === void 0 ? {} : _node$data$subjectPr;
  var name = subjectPr.name;
  var _config$locale = config.locale,
      locale = _config$locale === void 0 ? {} : _config$locale;
  var reviewTextPlaceholderMobile = locale.reviewTextPlaceholderMobile,
      topicContentPlaceholderMobile = locale.topicContentPlaceholderMobile,
      conclusionContentPlaceholder = locale.conclusionContentPlaceholder;
  var node1 = node.nodes[1];

  var displayPlaceholder1 = function displayPlaceholder1(item) {
    if (name === 'review' || name === 'topic' || name === 'conclusion') {
      return isNormalParagraph(item) && isEmptyNode(item) && node.nodes.length === 2;
    }

    return false;
  };

  var _ref = /*#__PURE__*/_createElement(InlinePlaceholder, {
    "data-placeholder": true
  }, reviewTextPlaceholderMobile);

  var _ref2 = /*#__PURE__*/_createElement(InlinePlaceholder, {
    "data-placeholder": true
  }, topicContentPlaceholderMobile);

  var _ref3 = /*#__PURE__*/_createElement(InlinePlaceholder, {
    "data-placeholder": true
  }, conclusionContentPlaceholder);

  var placeholder = React.useMemo(function () {
    if (name === 'review') {
      return _ref;
    } // 议题内容 placeholder


    if (name === 'topic') {
      return _ref2;
    } // 会议小结内容 placeholder


    if (name === 'conclusion') {
      return _ref3;
    }

    return null;
  }, [conclusionContentPlaceholder, name, topicContentPlaceholderMobile, reviewTextPlaceholderMobile]);

  if (children.length === 0) {
    return null;
  }

  return /*#__PURE__*/_createElement(Content, rest, node1 && /*#__PURE__*/_createElement(Placeholder, {
    controller: controller,
    node: node1,
    display: displayPlaceholder1,
    placeholder: placeholder
  }, children[0]), children.slice(1));
};

SubjectContent.displayName = 'SubjectContent';
export default SubjectContent;
