// @ali/4ever-bi/dist-node/plugins/embed/components/quickInsertPanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _everComponent = require("@ali/4ever-component");

var _everFactory = require("@ali/4ever-factory");

const _createElement = /*#__PURE__*/React.createElement;

const placeholder = /*#__PURE__*/_createElement("svg", {
  width: "61px",
  height: "60px",
  viewBox: "0 0 61 60",
  version: "1.1"
}, /*#__PURE__*/_createElement("g", {
  stroke: "none",
  strokeWidth: "1",
  fill: "none",
  fillRule: "evenodd"
}, /*#__PURE__*/_createElement("g", {
  transform: "translate(-201.000000, -206.000000)"
}, /*#__PURE__*/_createElement("g", {
  transform: "translate(193.000000, 200.000000)"
}, /*#__PURE__*/_createElement("g", {
  transform: "translate(7.291667, 6.083333)",
  id: "\u6587\u4EF6"
}, /*#__PURE__*/_createElement("g", {
  fillRule: "nonzero"
}, /*#__PURE__*/_createElement("path", {
  d: "M40.9543302,54.75 L6.86284853,54.75 C3.55152086,54.75 0.866216571,52.0674231 0.862851623,48.7560972 L0.819432862,6.02922666 C0.815388028,2.71552056 3.49894835,0.026500659 6.81265514,0.0231332923 C6.81288097,0.0231330628 6.81310681,0.023132846 6.81333326,0.0238101101 L33.1730769,0 L33.1730769,0 L46.9543302,12.5546 L46.9543302,48.75 C46.9543302,52.0637085 44.2680387,54.75 40.9543302,54.75 Z",
  fill: "#F2F4F5"
}), /*#__PURE__*/_createElement("path", {
  d: "M33.1782995,6.55460188 L33.1730769,0 L33.1730769,0 L46.9951923,12.5546 L39.1830803,12.5546 C35.8712386,12.5498173 33.185721,9.86643868 33.1782995,6.55460188 Z",
  fill: "#D9DBDD"
}), /*#__PURE__*/_createElement("path", {
  d: "M35.9374999,20.1305787 C35.9374999,20.8387282 35.3130723,21.4127967 34.5428018,21.4127967 L11.5309161,21.4127967 C10.7606456,21.4127967 10.1362179,20.8387282 10.1362179,20.1305787 C10.1362179,19.4224292 10.7606456,18.8483607 11.5309161,18.8483607 L34.5431185,18.8483607 C35.3132653,18.8485214 35.9374999,19.4225429 35.9374999,20.1305787 L35.9374999,20.1305787 Z M35.9374999,29.747214 C35.9374999,30.4553635 35.3130723,31.0294321 34.5428018,31.0294321 L11.5309161,31.0294321 C10.7606456,31.0294321 10.136218,30.4553635 10.136218,29.747214 C10.136218,29.0390646 10.7606456,28.464996 11.5309161,28.464996 L34.5431185,28.464996 C35.3132653,28.4651567 35.9374999,29.0391782 35.9374999,29.747214 L35.9374999,29.747214 Z M35.9374999,40.0046672 C35.9374999,40.7128167 35.3130723,41.2868852 34.5428018,41.2868852 L11.5309161,41.2868852 C10.7606456,41.2868852 10.1362179,40.7128167 10.1362179,40.0046672 C10.1362179,39.2965177 10.7606456,38.7224492 11.5309161,38.7224492 L34.5431185,38.7224492 C35.3132653,38.7226099 35.9374999,39.2966314 35.9374999,40.0046672 Z",
  fillOpacity: "0.08",
  fill: "#111F2C"
}))), /*#__PURE__*/_createElement("g", {
  transform: "translate(31.250000, 33.458333)"
}, /*#__PURE__*/_createElement("path", {
  d: "M3.75000003,0 L33.75,0 C36.25,0 37.5,1.33536585 37.5,4.00609755 L37.5,28.0426829 C37.5,30.7134146 36.25,32.0487805 33.75,32.0487805 L3.75000003,32.0487805 C1.25000004,32.0487805 0,30.7134146 0,28.0426829 L0,4.00609755 C0,1.33536585 1.25000001,0 3.75000003,0 Z",
  fill: "#E7E9EB",
  fillRule: "nonzero"
}), /*#__PURE__*/_createElement("g", {
  id: "Picture",
  fill: "#D9DBDD",
  fillRule: "nonzero"
}, /*#__PURE__*/_createElement("path", {
  d: "M7.31707317,0.101388889 L7.317,5.88738889 L12.804,5.88738889 L12.804878,0.101388889 L14.6341463,0.101388889 L14.634,5.88738889 L20.121,5.88738889 L20.1219512,0.101388889 L21.9512195,0.101388889 L21.951,5.88738889 L28.353,5.88738889 L28.3536585,0.101388889 L30.1829268,0.101388889 L30.182,5.88738889 L37.5,5.88797425 L37.5,7.66846206 L0,7.66846206 L0,5.88797425 L5.487,5.88738889 L5.48780488,0.101388889 L7.31707317,0.101388889 Z"
}), /*#__PURE__*/_createElement("path", {
  d: "M0.00446020146,24.3333333 L37.5,24.4273365 L37.4955398,26.3611111 L31.097,26.3443333 L31.097561,32.0487805 L29.2682927,32.0487805 L29.268,26.3393333 L22.865,26.3233333 L22.8658537,32.0487805 L21.0365854,32.0487805 L21.036,26.3193333 L15.548,26.3053333 L15.5487805,32.0487805 L13.7195122,32.0487805 L13.719,26.3003333 L8.231,26.2873333 L8.23170732,32.0487805 L6.40243902,32.0487805 L6.402,26.2823333 L0,26.2671079 L0.00446020146,24.3333333 Z"
}), /*#__PURE__*/_createElement("path", {
  d: "M14.9136179,19.4936538 L14.9136179,13.4453706 C14.9136179,12.4755047 15.6998495,11.689273 16.6697154,11.689273 C16.9809369,11.689273 17.2865678,11.7719816 17.5553163,11.9289307 L22.733641,14.9530724 C23.5711476,15.4421762 23.8535838,16.5176065 23.36448,17.3551131 C23.2120105,17.6161909 22.9947189,17.8334825 22.733641,17.985952 L17.5553163,21.0100937 C16.7178097,21.4991975 15.6423795,21.2167613 15.1532756,20.3792547 C14.9963265,20.1105062 14.9136179,19.8048753 14.9136179,19.4936538 Z"
})))))));

const Tip = /*#__PURE__*/_styledComponents.default.div(["width:200px;padding:15px 0px;"]);

const QuickInsertPanel = props => {
  const {
    editor,
    locale = {},
    close
  } = props;
  const handleInsert = React.useCallback(() => close(editor).run('onAction', (0, _everFactory.selectFile)()), [editor, close]);
  return /*#__PURE__*/_createElement(_everComponent.QuickInsertCommonPanel, {
    placeholder: placeholder
  }, /*#__PURE__*/_createElement(_everComponent.QuickInsertButton, {
    onClick: handleInsert
  }, locale.buttonTitle), /*#__PURE__*/_createElement(Tip, null, locale.insertTip));
};

var _default = QuickInsertPanel;
exports.default = _default;
