// @ali/4ever-bi/dist-node/plugins/embed/components/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreviewErrorRetry = exports.PreviewErrorText = exports.PreviewError = exports.ErrorDesc = exports.ProgressBarInner = exports.ProgressBar = exports.EmbedFilePreviewMask = exports.Rotation = exports.Copy = exports.Delete = exports.PreviewView = exports.CardView = exports.PreviewContentWrapper = exports.EmbedFilePreviewWrapper = exports.ContentDesc = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

const ContentDesc = /*#__PURE__*/_styledComponents.default.div(["font-size:12px;color:", ";margin-right:16px;"], props => props.mode === 'dark' ? '#F2F4F5' : '#7A8189');

exports.ContentDesc = ContentDesc;

const EmbedFilePreviewWrapper = /*#__PURE__*/_styledComponents.default.div(["position:relative;width:100%;display:flex;flex-direction:column;border:1px solid ", ";border-radius:6px;transition:height .6s;"], props => props.isSelected ? '#D9DBDD' : '#ECEDEF');

exports.EmbedFilePreviewWrapper = EmbedFilePreviewWrapper;

const PreviewContentWrapper = /*#__PURE__*/_styledComponents.default.div(["overflow:auto;flex:1;position:relative;justify-content:center;align-items:center;display:flex;background-color:#EDEDED;border-bottom-left-radius:6px;border-bottom-right-radius:6px;"]);

exports.PreviewContentWrapper = PreviewContentWrapper;

const CardView = /*#__PURE__*/_styledComponents.default.span(["flex:1;cursor:pointer;:hover{color:#3296fa;}"]);

exports.CardView = CardView;

const PreviewView = /*#__PURE__*/_styledComponents.default.span(["flex:1;cursor:pointer;:hover{color:#3296fa;}"]);

exports.PreviewView = PreviewView;

const Delete = /*#__PURE__*/_styledComponents.default.span(["flex:1;cursor:pointer;:hover{color:#3296fa;}"]);

exports.Delete = Delete;

const Copy = /*#__PURE__*/_styledComponents.default.span(["flex:1;cursor:pointer;:hover{color:#3296fa;}"]);

exports.Copy = Copy;

const Rotation = /*#__PURE__*/_styledComponents.default.span(["flex:1;cursor:pointer;:hover{color:#3296fa;}"]);

exports.Rotation = Rotation;

const EmbedFilePreviewMask = /*#__PURE__*/_styledComponents.default.div(["position:absolute;left:0;right:0;top:0;bottom:0;z-index:2;"]);

exports.EmbedFilePreviewMask = EmbedFilePreviewMask;

const ProgressBar = /*#__PURE__*/_styledComponents.default.div(["width:100%;height:4px;background-color:#D9DBDD;border-radius:4px;"]);

exports.ProgressBar = ProgressBar;

const ProgressBarInner = /*#__PURE__*/_styledComponents.default.div(["background-color:#0089FF;width:", ";height:100%;transition:width .6s;border-radius:4px;"], props => `${props.progress}%`);

exports.ProgressBarInner = ProgressBarInner;

const ErrorDesc = /*#__PURE__*/_styledComponents.default.div(["font-size:12px;color:#D0021B;"]);

exports.ErrorDesc = ErrorDesc;

const PreviewError = /*#__PURE__*/_styledComponents.default.div(["text-align:center;"]);

exports.PreviewError = PreviewError;

const PreviewErrorText = /*#__PURE__*/_styledComponents.default.div(["font-size:13px;color:#414c56;display:inline-block;"]);

exports.PreviewErrorText = PreviewErrorText;

const PreviewErrorRetry = /*#__PURE__*/_styledComponents.default.span(["color:#3A454F;font-weight:bold;cursor:pointer;"]);

exports.PreviewErrorRetry = PreviewErrorRetry;
