// @ali/4ever-bi/dist-node/plugins/strike/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _everPluginMarks = require("@ali/4ever-plugin-marks");

var _everComponent = require("@ali/4ever-component");

var _everPluginStrike = require("@ali/4ever-plugin-strike");

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement("img", {
  width: "24",
  height: "24",
  src: "https://gw.alicdn.com/imgextra/i4/O1CN01DTJU0l1wAKtfRmFD2_!!6000000006267-55-tps-24-24.svg"
});

var _default = (0, _everPluginMarks.createMarkFactory)({
  moFactory: _everPluginStrike.moFactory,
  type: 'strike',
  hotKey: 'mod+shift+x',
  delimiter: [{
    character: '~',
    size: 2
  }],
  Button: _weToolbar.StrikethroughButton,
  Icon: () => _ref,
  toolbarIcon: /*#__PURE__*/_createElement(_everComponent.IconStrikethrough, null),
  localeKeys: {
    titleKey: 'buttonTitle'
  }
});

exports.default = _default;
