// @ali/dingtalk-jsapi/lib/bridge/weexNbBridge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.weexNbBridge = void 0;
var weex_1 = require("./weex");
function AndroidBridgeCall(apiName, options, callback) {
    var bridge = weex_1.requireModule('wxNbBridge');
    bridge.exec({
        action: apiName,
        data: options,
    }, function (res) {
        if (callback) {
            callback(res);
        }
    });
}
var useAlipayJSBridge = {
    call: AndroidBridgeCall,
};
if (typeof AlipayJSBridge !== 'undefined') {
    useAlipayJSBridge = AlipayJSBridge;
}
var weexNbBridge = function (method, params) {
    return new Promise(function (resolve, reject) {
        var m = method;
        var onSuccess = params.onSuccess;
        var onFail = params.onFail;
        delete params.onSuccess;
        delete params.onFail;
        var mArr = m.split('.');
        var actionName = mArr.pop() || '';
        var serviceName = mArr.join('.');
        useAlipayJSBridge.call('ddExec', {
            actionName: actionName,
            serviceName: serviceName,
            args: params,
        }, function (res) {
            if (!res || !res.success) {
                if (onFail) {
                    onFail();
                }
                return reject();
            }
            try {
                // 兼容ios返回content是空的情况，不能当做fail处理
                var content = res.content;
                if (!content) {
                    if (onSuccess) {
                        onSuccess();
                    }
                }
                else {
                    content = JSON.parse(res.content);
                    if (onSuccess) {
                        onSuccess(content);
                    }
                }
                return resolve(content);
            }
            catch (e) {
                if (onFail) {
                    onFail(e);
                }
                return reject(e);
            }
        });
    });
};
exports.weexNbBridge = weexNbBridge;
