// @ali/we-toolbar/dist-web/biz/layout/toolbar/styled.js
import styled, { keyframes, css } from 'styled-components';
import { withTheme } from '@ali/we-design';
import { MODE_TYPE, LAYOUT_TYPE } from "../interface";
var fadeIn = /*#__PURE__*/keyframes(["from{opacity:0;}to{opacity:1;}"]);
var fadeOut = /*#__PURE__*/keyframes(["from{opacity:1;}to{opacity:0;}"]);
var fadeInUp = /*#__PURE__*/keyframes(["   0%{opacity:0;transform:translate3d(0,20px,0);}100%{opacity:1;transform:translate3d(0,0,0);}"]);
var fadeInDown = /*#__PURE__*/keyframes(["0%{opacity:0;transform:translate3d(0,-20px,0);}100%{opacity:1;transform:translate3d(0,0,0);}"]);
var scaleUpTr = /*#__PURE__*/keyframes(["0%{transform:scale(0.5);transform-origin:100% 0%;opacity:0;}100%{transform:scale(1);transform-origin:100% 0%;opacity:1;}"]);
export var ColumnWrapper = withTheme(styled.div(["display:flex;flex-direction:column;justify-content:space-between;margin-right:6px;:last-child{margin-right:0;}"]));
export var GroupWrapper = withTheme(styled.div(["display:flex;", " padding:0 ", ";justify-content:space-between;flex-direction:", ";", "{}"], function (_ref) {
  var hideBorder = _ref.hideBorder;
  return !hideBorder && css(["border-right:1px solid ", ";"], function (_ref2) {
    var theme = _ref2.theme;
    return theme.ui.commonBgPressColor;
  });
}, function (_ref3) {
  var theme = _ref3.theme,
      clearPadding = _ref3.clearPadding;
  return clearPadding ? 0 : theme.ui.paddingSmall;
}, function (_ref4) {
  var layoutType = _ref4.layoutType;
  return layoutType !== LAYOUT_TYPE.row ? 'row' : 'column';
}, ColumnWrapper));
export var RowWrapper = withTheme(styled.div(["display:flex;margin-bottom:4px;:last-child{margin-bottom:0px;}", ";"], function (_ref5) {
  var mode = _ref5.mode;

  if (mode === MODE_TYPE.doubleLine) {
    var cs = ['.split-button-layout', '.icon-menu-layout', '.icon-button-layout', '.icon-select-layout', '.text-select-layout'];
    return css(["", " .text-select-layout ~ .text-select-layout{margin-left:4px;}"], cs.map(function (o) {
      return cs.map(function (oo) {
        return o + " ~ " + oo;
      });
    }).join(',') + "{ margin-left: 6px;}");
  } else {
    return css(["box-sizing:content-box;.split-button-layout,.icon-select-layout{border-left:2px solid transparent;border-right:2px solid transparent;}> *{margin-right:4px;}> *:last-child{margin-right:0;}"]);
  }
}));
export var ToolbarWrapper = withTheme(styled.div(["padding-left:", ";display:flex;justify-content:space-between;background-color:", ";position:relative;height:", ";overflow:hidden;.show-left-wrapper{animation-duration:200ms;animation-fill-mode:both;animation-name:", ";}.toolbar-hide{display:none;}.group-wrapper-enter{animation-duration:200ms;animation-fill-mode:forwards;animation-name:", ";}.group-wrapper-leave{animation-duration:200ms;animation-fill-mode:forwards;animation-name:", ";}.single,.double{display:flex;}.toolbar-left-enter.single{animation-duration:300ms;animation-fill-mode:forwards;animation-name:", ";}.toolbar-left-enter.double{animation-duration:300ms;animation-fill-mode:forwards;animation-name:", ";}.toolbar-left-leave.single,.toolbar-left-leave.double{animation-duration:50ms;animation-fill-mode:forwards;animation-name:", ";top:0;position:absolute;z-index:1;}"], function (_ref6) {
  var theme = _ref6.theme;
  return theme.ui.paddingHug;
}, function (_ref7) {
  var backgroundColor = _ref7.backgroundColor;
  return backgroundColor || '#f9fafe';
}, function (_ref8) {
  var mode = _ref8.mode;
  return mode === MODE_TYPE.doubleLine ? '60px' : '36px';
}, fadeIn, fadeIn, fadeOut, fadeInUp, fadeInDown, fadeOut));
export var ToolbarLeftWrapper = withTheme(styled.div(["padding:", " 0;display:flex;position:relative;.group-transition:last-child .group-wrapper{border-right:none;}.group-transition:first-child .group-wrapper{padding:0 6px 0 0;}.more-btn{margin-left:4px;}"], function (_ref9) {
  var theme = _ref9.theme;
  return theme.ui.paddingTiny;
}));
export var ToolbarRightWrapper = withTheme(styled.div.attrs(function () {
  return {
    'data-testid': 'toolbar-change-mode'
  };
})(["display:flex;align-items:center;padding-right:24px;background-color:", ";position:absolute;right:0;height:100%;.icon-wrap{width:20px;height:20px;border-radius:4px;&:hover{background-color:", ";}&:active{background-color:", ";}.weicon{font-size:20px;color:rgba(58,60,63,1);transition:transform 500ms;transform:rotate( ", " );}}"], function (_ref10) {
  var backgroundColor = _ref10.backgroundColor;
  return backgroundColor || '#f9fafe';
}, function (_ref11) {
  var theme = _ref11.theme;
  return theme.ui.toolbarButtonHover;
}, function (_ref12) {
  var theme = _ref12.theme;
  return theme.ui.toolbarButtonClick;
}, function (_ref13) {
  var mode = _ref13.mode;
  return mode === MODE_TYPE.singleLine ? '0deg' : '180deg';
}));
export var OffScreenWrapper = withTheme(styled.div(["position:absolute;left:-10000px;display:flex;flex-direction:column;"]));
export var MoreOverlay = withTheme(styled.div(["display:flex;background-color:#f7f8f9;border:1px solid rgba(17,31,44,0.08);box-shadow:1px 3px 8px 0 rgba(0,0,0,0.1);border-radius:6px;&.scale-up{animation:", " 150ms cubic-bezier(0.39,0.575,0.565,1) forwards;}&.scale-out{animation-duration:50ms;animation-fill-mode:forwards;animation-name:", ";}", "{opacity:1;.group-transition:first-child .group-wrapper{padding:0 6px;}}"], scaleUpTr, fadeOut, ToolbarLeftWrapper));
