// @ali/4ever-plugin-mention/dist-web/bi/components/next/mentionButton.js
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { IconMention } from '@ali/4ever-component';
import { activeMention } from "../../actions";
import isSelectionInMention from "../../utils/isSelectionInMention";

var isMentionDisabled = function isMentionDisabled(controller) {
  var value = controller.value;
  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var isDisabled = isCollapsedAtEdge || isSelectionInMention(value) || controller.query('isSelectionInLink') || controller.query('isSelectionInImage') || controller.query('isSelectionInLinkCard') || controller.query('isSelectionInHr');
  return Boolean(isDisabled);
};

var _ref = /*#__PURE__*/_createElement(IconMention, null);

export default function createMentionButton(props) {
  var controller = props.controller,
      locale = props.locale,
      icon = props.icon,
      title = props.title,
      tooltip = props.tooltip,
      badgeDot = props.badgeDot,
      disabled = props.disabled;

  var handleClick = function handleClick() {
    return controller.run('onAction', activeMention());
  };

  var isDisabled = disabled || isMentionDisabled(controller);
  var buttonConfig = {
    icon: icon || _ref,
    onClick: handleClick,
    title: title || (locale == null ? void 0 : locale.buttonTitle),
    tooltip: tooltip,
    badgeDot: badgeDot,
    disabled: isDisabled,
    testid: "bi-toolbar-mention"
  };
  return buttonConfig;
}
