// @ali/4ever-plugin-table/dist-node/bi/components/toolbarButtons/tableButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TableButton;

var React = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _role = require("../../../utils/role");

var _actions = require("../../actions");

var _everComponent = require("@ali/4ever-component");

const _createElement = /*#__PURE__*/React.createElement;

var _ref = /*#__PURE__*/_createElement(_weDesign.TableNormal, null);

var _ref2 = /*#__PURE__*/_createElement(_weDesign.TableLightNormal, null);

function TableButton(props) {
  const {
    controller,
    locale,
    toolbarMode
  } = props;
  const isDisabled = !!controller.query('canInsertTable');
  const handleClick = React.useCallback(() => controller.run('onAction', (0, _actions.insertTable)(3, 3)), [controller]);
  return /*#__PURE__*/_createElement(_weDesign.Button, {
    title: locale.buttonTitle,
    bordered: false,
    className: "bi-toolbar-table",
    icon: toolbarMode === _everComponent.ToolbarMode.single ? _ref : _ref2,
    disabled: isDisabled,
    onClick: handleClick,
    role: _role.TableRole.table
  });
}
