// @ali/4ever-plugin-table/dist-node/bi/components/toolbarButtons/tableSimpleMergeButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TableSimpleMergeButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _models = _interopRequireDefault(require("../../../mo/models"));

var _weToolbar = require("@ali/we-toolbar");

var _weDesign = require("@ali/we-design");

var _role = require("../../../utils/role");

var _actions = require("../../actions");

var _getTable = _interopRequireDefault(require("../../utils/getTable"));

var _getTableSelection = _interopRequireDefault(require("../../utils/getTableSelection"));

const _createElement = /*#__PURE__*/_react.default.createElement;

var _ref = /*#__PURE__*/_createElement(_weDesign.ToolbarMergeCellNormal, null);

function TableSimpleMergeButton(props) {
  const {
    controller,
    locale,
    ...rest
  } = props;
  const table = (0, _getTable.default)(controller.value);
  const selection = (0, _getTableSelection.default)(controller);

  const isSelectionSupportSplit = _react.default.useMemo(() => {
    return !!table && controller.query('isSelectionSupportSplit', {
      visible: true,
      node: table
    }); // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [table, selection, controller]);

  const isSelectionSupportMerge = _react.default.useMemo(() => {
    return !!table && controller.query('isSelectionSupportMerge', {
      visible: true,
      node: table
    }); // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [table, selection, controller]);

  const title = locale.contextMenuMerge;

  const handleClick = () => {
    controller.run('onAction', (0, _actions.mergeTableCells)(table));
  }; // MergeButton SplitButton 二选一逻辑，非常业务化
  // 但是条件又属于与表格组件逻辑强关联
  // 目前除了在此处直接蒸发，没有其他方式实现


  if (!isSelectionSupportMerge && isSelectionSupportSplit) {
    return null;
  }

  return /*#__PURE__*/_createElement(_weToolbar.IconButton, (0, _extends2.default)({}, rest, {
    icon: _ref,
    title: title,
    tooltip: title,
    testid: "bi-toolbar-tablesimplemerge",
    onClick: handleClick,
    disabled: !isSelectionSupportMerge,
    role: _role.TableRole.tableMerge
  }));
}

TableSimpleMergeButton.needRender = function (controller) {
  const selection = (0, _getTableSelection.default)(controller);
  if (!selection) return false;
  const node = controller.value.document.getNode(selection.key);
  if (!(node && _models.default.isTable(node))) return false;
  const tableRowCount = node.nodes.length - 1;
  const tableColCount = (node.data.colsWidth || []).length - 1;
  const {
    startRowIndex,
    startColIndex,
    endRowIndex,
    endColIndex
  } = selection;
  const isSelectWholeRow = startColIndex === 0 && endColIndex >= tableColCount;
  const isSelectWholeCol = startRowIndex === 0 && endRowIndex >= tableRowCount;
  const isSelectWholeTable = isSelectWholeRow && isSelectWholeCol;
  const isSelectSingleCell = startColIndex === endColIndex && startRowIndex === endRowIndex;

  if (isSelectWholeTable || isSelectSingleCell) {
    return false;
  }

  return true;
};
