// @ali/4ever-plugin-table/dist-node/bi/components/toolbarButtons/tablePickerButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TablePickerButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weDesign = require("@ali/we-design");

var _weToolbar = require("@ali/we-toolbar");

var _role = require("../../../utils/role");

var _isSelectionInTableCell = _interopRequireDefault(require("../../utils/isSelectionInTableCell"));

var _actions = require("../../actions");

const _createElement = /*#__PURE__*/React.createElement;

function TablePickerButton(props) {
  const {
    controller,
    locale,
    toolbarMode,
    ...rest
  } = props;
  const {
    value
  } = controller;
  const tableSelection = controller.query('hasTableSelection'); // @ts-ignore

  const isDisabled = value.selection.isBlurred || value.data.edgeSelection || tableSelection || (0, _isSelectionInTableCell.default)(value) || controller.query('isSelectionInLink') || controller.query('isSelectionInMention') || controller.query('isSelectionInListSymbol');
  const handleChange = React.useCallback((row, col) => {
    controller.run('onAction', (0, _actions.insertTable)(row, col));
  }, [controller]);

  var _ref = /*#__PURE__*/_createElement(_weDesign.TablePanel, {
    row: 10,
    col: 10,
    onPick: handleChange
  });

  const tablePanel = React.useMemo(() => _ref, [handleChange]);
  return /*#__PURE__*/_createElement(_weToolbar.TableButton, (0, _extends2.default)({}, rest, {
    testid: "bi-toolbar-table-picker",
    overlay: tablePanel,
    disabled: isDisabled,
    autoClose: true,
    role: _role.TableRole.tablePicker
  }));
}
