// @ali/4ever-plugin-table/dist-node/bi/components/toolbarButtons/tableVAlignButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TableVAlignButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _weToolbar = require("@ali/we-toolbar");

var _role = require("../../../utils/role");

var _getTable = _interopRequireDefault(require("../../utils/getTable"));

var _getTableSelection = _interopRequireDefault(require("../../utils/getTableSelection"));

var _actions = require("../../actions");

var _getVAlign = _interopRequireDefault(require("../../utils/getVAlign"));

var _constants = require("../../constants");

const _createElement = /*#__PURE__*/React.createElement;
const SUPPORT_ALIGNMENT = {
  middle: true,
  top: true,
  bottom: true
};

function TableVAlignButton(props) {
  const {
    controller,
    locale,
    ...rest
  } = props;
  let alignment = (0, _getVAlign.default)(controller);
  alignment = SUPPORT_ALIGNMENT[alignment] ? alignment : _constants.DEFAULT_TABLE_ALIGNMENT;
  const selection = (0, _getTableSelection.default)(controller);
  let tip = locale.vAlignMiddle;

  if (alignment === 'top') {
    tip = locale.vAlignTop;
  } else if (alignment === 'bottom') {
    tip = locale.vAlignBottom;
  }

  const handleChange = vAlign => {
    const table = (0, _getTable.default)(controller.value);
    controller.run('onAction', (0, _actions.setTableCellVAlign)(table, selection, vAlign));
  };

  return /*#__PURE__*/_createElement(_weToolbar.VerticalAlignSelect, (0, _extends2.default)({}, rest, {
    title: tip,
    testid: "bi-toolbar-valign",
    defaultValue: _constants.DEFAULT_TABLE_ALIGNMENT,
    value: alignment,
    disabled: !selection,
    onChange: handleChange,
    role: _role.TableRole.tableVAlignSelect
  }));
}
