// @ali/4ever-bi/dist-web/plugins/indent/components/IncreaseIndentButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import * as React from 'react';
var _createElement = /*#__PURE__*/React.createElement;
import { IncreaseIndent as Button } from '@ali/we-toolbar';
import { PluginRoles } from '@ali/4ever-bamboo';
import { increaseIndent } from "../actions";
import { INCREASE_HOT_KEY_TIP } from "../utils/hotKeyTips";
;
export default function IncreaseIndentButton(props) {
  var controller = props.controller,
      locale = props.locale,
      rest = _objectWithoutPropertiesLoose(props, ["controller", "locale"]);

  var isCollapsedAtEdge = controller.value.selection.isCollapsedAtEdge;
  var isDisabled = isCollapsedAtEdge || controller.query('isSelectionInHr') || controller.query('isSelectionInListSymbol');
  var handleClick = React.useCallback(function () {
    return controller.run('onAction', increaseIndent());
  }, [controller]);
  return /*#__PURE__*/_createElement(Button, _extends({}, rest, {
    shortcutText: INCREASE_HOT_KEY_TIP,
    testid: "bi-toolbar-incindent",
    onClick: handleClick,
    disabled: !!isDisabled,
    role: PluginRoles.indentIncrease
  }));
}
