// @ali/4ever-bi/dist-web/plugins/link/linkCard/commands/unwrapLinkCard.js
import _extends from "@babel/runtime/helpers/extends";
import { Inline, Text, Commands, Block } from '@ali/4ever-cangjie';
import removeLink from "./removeLink";
export default function unwrapLinkCard(controller, node) {
  var _ref = node.data.metadata,
      text = _ref.text,
      href = _ref.href,
      cardInfo = _ref.cardInfo; // 将 text link 转换为 block link card

  var nodeText = text.length ? text : href;
  controller.command(Commands.moveToEndOfNode, node);
  var inline = Inline.create({
    type: 'link',
    nodes: [Text.create(nodeText)],
    data: {
      href: href,
      cardInfo: _extends({}, cardInfo, {
        displayType: 'link'
      })
    }
  }); // 将 BlockLinkCard 转换为 textLink 后，调用原有的 unwrapLink 方法
  // 这里需要先插入，再删除，否则会报找不到 path 的异常

  controller.command(Commands.moveToEndOfNode, node).command(Commands.insertBlock, {
    type: Block.DEFAULT_TYPE,
    nodes: [inline]
  }).command(Commands.moveToRangeOfNode, inline).command('unwrapLink').command(removeLink, node);
  return controller;
}
