// @ali/4ever-bi/dist-node/plugins/letterSpacing/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _setLetterSpacing = _interopRequireDefault(require("./commands/setLetterSpacing"));

var _letterSpacingMenu = _interopRequireDefault(require("./components/letterSpacingMenu"));

var _getLetterSpacing = _interopRequireDefault(require("./queries/getLetterSpacing"));

var actions = _interopRequireWildcard(require("./actions"));

var _onAction = _interopRequireDefault(require("./handlers/onAction"));

const commands = {
  setLetterSpacing: _setLetterSpacing.default
};
const queries = {
  getLetterSpacing: _getLetterSpacing.default
};

function createLetterSpacingPlugin() {
  return {
    menubar: {
      menu: _letterSpacingMenu.default
    },
    commands,
    queries,
    actions,
    onAction: _onAction.default,
    models: {
      klass: 'mark',
      type: 'spacing'
    }
  };
}

createLetterSpacingPlugin.actions = actions;
createLetterSpacingPlugin.commands = commands;
createLetterSpacingPlugin.queries = queries;
var _default = createLetterSpacingPlugin;
exports.default = _default;
