// @ali/4ever-bi/dist-node/plugins/image/mobile/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _renderNode = _interopRequireDefault(require("./renderNode"));

var _insertImage = _interopRequireDefault(require("./insertImage"));

var _actions = require("../actions");

var _insertImageFile = _interopRequireDefault(require("../commands/insertImageFile"));

var _onCommand = _interopRequireDefault(require("../handlers/onCommand"));

var _createRenderEditable = _interopRequireDefault(require("../createRenderEditable"));

var _queries = require("../queries");

var _onCangjieInput = _interopRequireDefault(require("../handlers/onCangjieInput"));

var _everPluginImage = require("@ali/4ever-plugin-image");

const state = {};
let commands = {};

function createImagePlugin(config) {
  const imageConfig = config.image || {};
  commands = {
    insertImage: (0, _insertImage.default)(state),
    insertImageFile: (0, _insertImageFile.default)(imageConfig, state)
  };
  return { ...(0, _everPluginImage.moFactory)({}),
    toolbar: (0, _toolbar.default)(imageConfig),
    renderNode: (0, _renderNode.default)(imageConfig, state),
    renderEditable: (0, _createRenderEditable.default)(imageConfig),
    onCommand: (0, _onCommand.default)(config, state),
    onCangjieInput: (0, _onCangjieInput.default)(imageConfig, state),
    commands,
    queries: _queries.queries,
    actions: _actions.actions,
    models: {
      klass: 'inline',
      type: 'image'
    }
  };
}

createImagePlugin.actions = _actions.DEPRECATED_ACTIONS;
createImagePlugin.commands = commands;
var _default = createImagePlugin;
exports.default = _default;
