// @ali/4ever-utils/dist-node/utils/formats/common.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getDefaultHeadStyle = getDefaultHeadStyle;
exports.getDefaultMarks = getDefaultMarks;
exports.CommonFormatUtil = exports.replaceObject = void 0;

var _lodash = require("lodash");

var _everCangjie = require("@ali/4ever-cangjie");

var _heading = require("../../heading/heading");

var _sz = require("./sz");

function getDefaultHeadStyle(type) {
  return _heading.headDefaultStyle[(0, _heading.headTypeConvert)(type)];
}

function getDefaultMarks(node, defaultParagraphSz) {
  if (!_everCangjie.Block.isBlock(node)) return [];
  const defaultMarks = [_everCangjie.Mark.create({
    type: 'sz',
    data: {
      value: _sz.SzUtil.fromHeadingType(node.type, defaultParagraphSz),
      szUnit: 'pt'
    }
  })];

  if (getDefaultHeadStyle(node.type)) {
    defaultMarks.push(_everCangjie.Mark.create({
      type: 'bold'
    }));
  }

  return defaultMarks;
}

const replaceObject = (obj, val, pathes) => {
  const stack = pathes.split('.'); // PERF: 避免引用的对象更改，性能不好

  const copyed = (0, _lodash.cloneDeep)(obj);
  if (stack.length < 1) return copyed;
  let target = copyed;

  while (stack.length > 1) {
    const key = stack.shift() || '';
    target[key] = target[key] || {};
    target = target[key];
  }

  target[stack.shift() || ''] = val;
  return copyed;
};

exports.replaceObject = replaceObject;
const CommonFormatUtil = {
  getDefaultHeadStyle,
  getDefaultMarks,
  replaceObject
};
exports.CommonFormatUtil = CommonFormatUtil;
