// @ali/4ever-utils/dist-node/utils/formats/gradientColor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.shouldDegradeGradient = exports.setGradientStyle = exports.notSupportClipText = exports.transferGradientColor = exports.isGradientColor = exports.GRADIENT_STYLE_FLAG = exports.GRADIENT_COLOR_STYLE = exports.GRADIENT_FLAG = void 0;

var _everCangjie = require("@ali/4ever-cangjie");

const GRADIENT_FLAG = '__';
exports.GRADIENT_FLAG = GRADIENT_FLAG;
const GRADIENT_COLOR_STYLE = {
  // chrome 会自动去掉 webkit 前缀，需在 snapshot 和 全局 手动处理
  // https://github.com/facebook/react/issues/14200
  WebkitBackgroundClip: 'text',
  backgroundClip: 'text',
  WebkitTextFillColor: 'transparent',
  backgroundImage: 'linear-gradient(90deg, #404040, #404040)'
};
exports.GRADIENT_COLOR_STYLE = GRADIENT_COLOR_STYLE;
const GRADIENT_STYLE_FLAG = 'WebkitBackgroundClip';
exports.GRADIENT_STYLE_FLAG = GRADIENT_STYLE_FLAG;
const CHROME_VERSION_DIVIDE = 72;

const isGradientColor = color => {
  // 判断是否为渐变色，渐变色格式为 ‘#color1__#color2’
  return color && color.includes(GRADIENT_FLAG);
}; // 将 #color1_#color2 转换为 [color1, color2]


exports.isGradientColor = isGradientColor;

const transferGradientColor = color => {
  if (!isGradientColor(color)) {
    return [];
  }

  return color.split(GRADIENT_FLAG);
}; // safari 和 低版本(72-) chrome 对 -webkit-text-clip 支持有问题, 导致渐变色换行不显示，需特殊处理


exports.transferGradientColor = transferGradientColor;

const notSupportClipText = () => {
  const {
    getChromeVersion,
    IS_SAFARI
  } = _everCangjie.environment;
  const chromeVersion = getChromeVersion();
  const isOldChrome = chromeVersion !== -1 && chromeVersion < CHROME_VERSION_DIVIDE;
  return isOldChrome || IS_SAFARI;
};

exports.notSupportClipText = notSupportClipText;

const setGradientStyle = (color, style) => {
  // 渐变色
  const [startColor, endColor] = transferGradientColor(color);
  Object.keys(GRADIENT_COLOR_STYLE).forEach(key => {
    style[key] = GRADIENT_COLOR_STYLE[key];
  });
  style.backgroundImage = `linear-gradient(90deg, ${startColor}, ${endColor})`; // 使用第一个颜色兜底

  style.color = startColor; // safari 和 低版本(72-) chrome 渐变色换行不显示，需添加一个属性

  if (notSupportClipText()) {
    style.WebkitBoxDecorationBreak = 'clone';
  }
}; // 存在下划线/删除线/非透明背景时，渐变色降级为第一个颜色


exports.setGradientStyle = setGradientStyle;

const shouldDegradeGradient = style => {
  if (!(GRADIENT_STYLE_FLAG in style)) {
    return false;
  }

  const hasLineDecoration = ('textDecorationLine' in style);
  const hasBackground = style.backgroundColor && style.backgroundColor !== 'transparent';
  return hasLineDecoration || hasBackground;
};

exports.shouldDegradeGradient = shouldDegradeGradient;
var _default = {
  GRADIENT_COLOR_STYLE,
  GRADIENT_STYLE_FLAG,
  isGradientColor,
  transferGradientColor,
  setGradientStyle,
  notSupportClipText,
  shouldDegradeGradient
};
exports.default = _default;
